--- Mesa-4.0.3/configure.remove-rpath	2002-07-15 16:50:23.000000000 +0200
+++ Mesa-4.0.3/configure	2002-07-15 17:14:46.000000000 +0200
@@ -5302,8 +5302,10 @@ EOF
   esac
 
   if test "$ld_shlibs" = yes; then
-    runpath_var=LD_RUN_PATH
-    hardcode_libdir_flag_spec='${wl}--rpath ${wl}$libdir'
+    # avoid hardcoding RPATH.
+    #runpath_var=LD_RUN_PATH
+    #hardcode_libdir_flag_spec='${wl}--rpath ${wl}$libdir'
+    hardcode_minus_L=no
     export_dynamic_flag_spec='${wl}--export-dynamic'
     case $host_os in
     cygwin* | mingw* | pw32*)
