diff -Naur php-4.3.4/acinclude.m4 php-4.3.4.oden/acinclude.m4
--- php-4.3.4/acinclude.m4	2003-10-03 06:29:24.000000000 +0200
+++ php-4.3.4.oden/acinclude.m4	2004-01-09 20:14:48.000000000 +0100
@@ -180,7 +180,7 @@
   unset ac_new_flags
   for i in [$]$1; do
     case [$]i in
-    -L/usr/lib|-L/usr/lib/[)] ;;
+    -L/usr/lib|-L/usr/lib/|-L/usr/lib64|-L/usr/lib64/) ;;
     *[)] ac_new_flags="[$]ac_new_flags [$]i" ;;
     esac
   done
@@ -836,7 +836,7 @@
 dnl add a library to linkpath/runpath
 dnl
 AC_DEFUN([PHP_ADD_LIBPATH],[
-  if test "$1" != "/usr/lib"; then
+  if test "$1" != "/usr/lib" && test "$1" != "/usr/lib64"; then
     PHP_EXPAND_PATH($1, ai_p)
     ifelse([$2],,[
       _PHP_ADD_LIBPATH_GLOBAL([$ai_p])
@@ -1538,6 +1538,40 @@
   	fi
 ])
 
+dnl PHP_CHECK_LIBNAME
+dnl
+dnl Determine libdir name.
+dnl
+AC_DEFUN([PHP_CHECK_LIBNAME], [
+  AC_REQUIRE([AC_CANONICAL_HOST])dnl
+  AC_REQUIRE([AC_CANONICAL_BUILD])dnl
+  AC_MSG_CHECKING([libdir name])
+  case $host in
+  *-*-linux*)
+ # Test if the compiler is 64bit
+ echo 'int i;' > conftest.$ac_ext
+ php_cv_cc_64bit_output=no
+ if AC_TRY_EVAL(ac_compile); then
+   case `/usr/bin/file conftest.$ac_objext` in
+   *"ELF 64"*)
+     php_cv_cc_64bit_output=yes
+     ;;
+   esac
+ fi
+ rm -rf conftest*
+ ;;
+  esac
+  
+  case $host_cpu:$php_cv_cc_64bit_output in
+  powerpc64:yes | s390x:yes | sparc64:yes | x86_64:yes)
+ PHP_LIBNAME="lib64"
+ ;;
+  *:*)
+ PHP_LIBNAME="lib"
+ ;;
+  esac
+  AC_MSG_RESULT([$PHP_LIBNAME])
+])
 
 dnl
 dnl PHP_CHECK_LIBRARY(library, function [, action-found [, action-not-found [, extra-libs]]])
@@ -1625,8 +1659,8 @@
       if test -r $i/include/openssl/evp.h; then
         OPENSSL_INCDIR=$i/include
       fi
-      if test -r $i/lib/libssl.a -o -r $i/lib/libssl.$SHLIB_SUFFIX_NAME; then
-        OPENSSL_LIBDIR=$i/lib
+      if test -r $i/$PHP_LIBNAME/libssl.a -o -r $i/$PHP_LIBNAME/libssl.$SHLIB_SUFFIX_NAME; then
+        OPENSSL_LIBDIR=$i/$PHP_LIBNAME
       fi
       test -n "$OPENSSL_INCDIR" && test -n "$OPENSSL_LIBDIR" && break
     done
@@ -1731,8 +1765,8 @@
       AC_MSG_ERROR([Please specify the install prefix of iconv with --with-iconv=<DIR>])
     fi
   
-    if test -f $ICONV_DIR/lib/lib$iconv_lib_name.a ||
-       test -f $ICONV_DIR/lib/lib$iconv_lib_name.$SHLIB_SUFFIX_NAME
+    if test -f $ICONV_DIR/$PHP_LIBNAME/lib$iconv_lib_name.a ||
+       test -f $ICONV_DIR/$PHP_LIBNAME/lib$iconv_lib_name.$SHLIB_SUFFIX_NAME
     then
       PHP_CHECK_LIBRARY($iconv_lib_name, libiconv, [
         found_iconv=yes
@@ -1742,10 +1776,10 @@
           found_iconv=yes
           PHP_DEFINE(HAVE_ICONV)
         ], [], [
-          -L$ICONV_DIR/lib
+          -L$ICONV_DIR/$PHP_LIBNAME
         ])
       ], [
-        -L$ICONV_DIR/lib
+        -L$ICONV_DIR/$PHP_LIBNAME
       ])
     fi
   fi
diff -Naur php-4.3.4/aclocal.m4 php-4.3.4.oden/aclocal.m4
--- php-4.3.4/aclocal.m4	2003-11-03 00:27:49.000000000 +0100
+++ php-4.3.4.oden/aclocal.m4	2004-01-09 20:14:48.000000000 +0100
@@ -180,7 +180,7 @@
   unset ac_new_flags
   for i in [$]$1; do
     case [$]i in
-    -L/usr/lib|-L/usr/lib/[)] ;;
+    -L/usr/lib|-L/usr/lib/|-L/usr/lib64|-L/usr/lib64/) ;;
     *[)] ac_new_flags="[$]ac_new_flags [$]i" ;;
     esac
   done
@@ -836,7 +836,7 @@
 dnl add a library to linkpath/runpath
 dnl
 AC_DEFUN([PHP_ADD_LIBPATH],[
-  if test "$1" != "/usr/lib"; then
+  if test "$1" != "/usr/lib" && test "$1" != "/usr/lib64"; then
     PHP_EXPAND_PATH($1, ai_p)
     ifelse([$2],,[
       _PHP_ADD_LIBPATH_GLOBAL([$ai_p])
@@ -1538,6 +1538,41 @@
   	fi
 ])
 
+dnl
+dnl PHP_CHECK_LIBNAME
+dnl
+dnl Determine libdir name.
+dnl
+AC_DEFUN([PHP_CHECK_LIBNAME], [
+  AC_REQUIRE([AC_CANONICAL_HOST])dnl
+  AC_REQUIRE([AC_CANONICAL_BUILD])dnl
+  AC_MSG_CHECKING([libdir name])
+  case $host in
+  *-*-linux*)
+ # Test if the compiler is 64bit
+ echo 'int i;' > conftest.$ac_ext
+ php_cv_cc_64bit_output=no
+ if AC_TRY_EVAL(ac_compile); then
+   case `/usr/bin/file conftest.$ac_objext` in
+   *"ELF 64"*)
+     php_cv_cc_64bit_output=yes
+     ;;
+   esac
+ fi
+ rm -rf conftest*
+ ;;
+  esac
+  
+  case $host_cpu:$php_cv_cc_64bit_output in
+  powerpc64:yes | s390x:yes | sparc64:yes | x86_64:yes)
+ PHP_LIBNAME="lib64"
+ ;;
+  *:*)
+ PHP_LIBNAME="lib"
+ ;;
+  esac
+  AC_MSG_RESULT([$PHP_LIBNAME])
+])
 
 dnl
 dnl PHP_CHECK_LIBRARY(library, function [, action-found [, action-not-found [, extra-libs]]])
@@ -1625,8 +1660,8 @@
       if test -r $i/include/openssl/evp.h; then
         OPENSSL_INCDIR=$i/include
       fi
-      if test -r $i/lib/libssl.a -o -r $i/lib/libssl.$SHLIB_SUFFIX_NAME; then
-        OPENSSL_LIBDIR=$i/lib
+      if test -r $i/$PHP_LIBNAME/libssl.a -o -r $i/$PHP_LIBNAME/libssl.$SHLIB_SUFFIX_NAME; then
+        OPENSSL_LIBDIR=$i/$PHP_LIBNAME
       fi
       test -n "$OPENSSL_INCDIR" && test -n "$OPENSSL_LIBDIR" && break
     done
@@ -1731,8 +1766,8 @@
       AC_MSG_ERROR([Please specify the install prefix of iconv with --with-iconv=<DIR>])
     fi
   
-    if test -f $ICONV_DIR/lib/lib$iconv_lib_name.a ||
-       test -f $ICONV_DIR/lib/lib$iconv_lib_name.$SHLIB_SUFFIX_NAME
+    if test -f $ICONV_DIR/$PHP_LIBNAME/lib$iconv_lib_name.a ||
+       test -f $ICONV_DIR/$PHP_LIBNAME/lib$iconv_lib_name.$SHLIB_SUFFIX_NAME
     then
       PHP_CHECK_LIBRARY($iconv_lib_name, libiconv, [
         found_iconv=yes
@@ -1742,10 +1777,10 @@
           found_iconv=yes
           PHP_DEFINE(HAVE_ICONV)
         ], [], [
-          -L$ICONV_DIR/lib
+          -L$ICONV_DIR/$PHP_LIBNAME
         ])
       ], [
-        -L$ICONV_DIR/lib
+        -L$ICONV_DIR/$PHP_LIBNAME
       ])
     fi
   fi
diff -Naur php-4.3.4/configure.in php-4.3.4.oden/configure.in
--- php-4.3.4/configure.in	2003-11-02 23:15:05.000000000 +0100
+++ php-4.3.4.oden/configure.in	2004-01-09 20:14:48.000000000 +0100
@@ -128,6 +128,10 @@
 PHP_RUNPATH_SWITCH
 
 PHP_PROG_RE2C
+
+dnl check for libdir name
+PHP_CHECK_LIBNAME
+
 AC_PROG_RANLIB
 AC_PROG_LN_S
 AC_PROG_AWK
diff -Naur php-4.3.4/ext/dba/config.m4 php-4.3.4.oden/ext/dba/config.m4
--- php-4.3.4/ext/dba/config.m4	2003-10-03 07:25:33.000000000 +0200
+++ php-4.3.4.oden/ext/dba/config.m4	2004-01-09 20:18:32.000000000 +0100
@@ -133,8 +133,8 @@
 dnl parameters(version, library list, function)
 AC_DEFUN(PHP_DBA_DB_CHECK,[
   for LIB in $2; do
-    if test -f $THIS_PREFIX/lib/lib$LIB.a -o -f $THIS_PREFIX/lib/lib$LIB.$SHLIB_SUFFIX_NAME; then
-      PHP_TEMP_LDFLAGS(-L$THIS_PREFIX/lib -l$LIB,[
+    if test -f $THIS_PREFIX/$PHP_LIBNAME/lib$LIB.a -o -f $THIS_PREFIX/$PHP_LIBNAME/lib$LIB.$SHLIB_SUFFIX_NAME; then
+      PHP_TEMP_LDFLAGS(-L$THIS_PREFIX/$PHP_LIBNAME -l$LIB,[
         AC_TRY_LINK([
 #include "$THIS_INCLUDE"
         ],[
@@ -326,7 +326,7 @@
           fi
           AC_DEFINE(DBA_DBM, 1, [ ]) 
           THIS_LIBS=$LIB
-        ], [], [-L$THIS_PREFIX/lib])
+        ], [], [-L$THIS_PREFIX/$PHP_LIBNAME])
         if test -n "$THIS_LIBS"; then
           break
         fi
@@ -368,7 +368,7 @@
           AC_DEFINE_UNQUOTED(CDB_INCLUDE_FILE, "$THIS_INCLUDE", [ ])
           AC_DEFINE(DBA_CDB, 1, [ ]) 
           THIS_LIBS=$LIB
-        ], [], [-L$THIS_PREFIX/lib])
+        ], [], [-L$THIS_PREFIX/$PHP_LIBNAME])
         if test -n "$THIS_LIBS"; then
           break
         fi
diff -Naur php-4.3.4/ext/gd/config.m4 php-4.3.4.oden/ext/gd/config.m4
--- php-4.3.4/ext/gd/config.m4	2003-10-01 04:53:57.000000000 +0200
+++ php-4.3.4.oden/ext/gd/config.m4	2004-01-09 20:14:48.000000000 +0100
@@ -51,7 +51,7 @@
   if test "$PHP_JPEG_DIR" != "no"; then
 
     for i in $PHP_JPEG_DIR /usr/local /usr; do
-      test -f $i/lib/libjpeg.$SHLIB_SUFFIX_NAME -o -f $i/lib/libjpeg.a && GD_JPEG_DIR=$i && break
+      test -f $i/$PHP_LIBNAME/libjpeg.$SHLIB_SUFFIX_NAME -o -f $i/$PHP_LIBNAME/libjpeg.a && GD_JPEG_DIR=$i && break
     done
 
     if test -z "$GD_JPEG_DIR"; then
@@ -61,11 +61,11 @@
     PHP_CHECK_LIBRARY(jpeg,jpeg_read_header,
     [
       PHP_ADD_INCLUDE($GD_JPEG_DIR/include)
-      PHP_ADD_LIBRARY_WITH_PATH(jpeg, $GD_JPEG_DIR/lib, GD_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(jpeg, $GD_JPEG_DIR/$PHP_LIBNAME, GD_SHARED_LIBADD)
     ],[
       AC_MSG_ERROR([Problem with libjpeg.(a|so). Please check config.log for more information.]) 
     ],[
-      -L$GD_JPEG_DIR/lib
+      -L$GD_JPEG_DIR/$PHP_LIBNAME
     ])
   else 
     AC_MSG_RESULT([If configure fails try --with-jpeg-dir=<DIR>])
@@ -76,7 +76,7 @@
   if test "$PHP_PNG_DIR" != "no"; then
 
     for i in $PHP_PNG_DIR /usr/local /usr; do
-      test -f $i/lib/libpng.$SHLIB_SUFFIX_NAME -o -f $i/lib/libpng.a && GD_PNG_DIR=$i && break
+      test -f $i/$PHP_LIBNAME/libpng.$SHLIB_SUFFIX_NAME -o -f $i/$PHP_LIBNAME/libpng.a && GD_PNG_DIR=$i && break
     done
 
     if test -z "$GD_PNG_DIR"; then
@@ -94,12 +94,12 @@
     PHP_CHECK_LIBRARY(png,png_write_image,
     [
       PHP_ADD_INCLUDE($GD_PNG_DIR/include)
-      PHP_ADD_LIBRARY_WITH_PATH(z, $PHP_ZLIB_DIR/lib, GD_SHARED_LIBADD)
-      PHP_ADD_LIBRARY_WITH_PATH(png, $GD_PNG_DIR/lib, GD_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(z, $PHP_ZLIB_DIR/$PHP_LIBNAME, GD_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(png, $GD_PNG_DIR/$PHP_LIBNAME, GD_SHARED_LIBADD)
     ],[
       AC_MSG_ERROR([Problem with libpng.(a|so) or libz.(a|so). Please check config.log for more information.]) 
     ],[
-      -L$PHP_ZLIB_DIR/lib -lz -L$GD_PNG_DIR/lib
+      -L$PHP_ZLIB_DIR/$PHP_LIBNAME -lz -L$GD_PNG_DIR/$PHP_LIBNAME
     ])
 
   else 
@@ -111,7 +111,7 @@
   if test "$PHP_XPM_DIR" != "no"; then
 
     for i in $PHP_XPM_DIR /usr/local /usr/X11R6 /usr; do
-      test -f $i/lib/libXpm.$SHLIB_SUFFIX_NAME -o -f $i/lib/libXpm.a && GD_XPM_DIR=$i && break
+      test -f $i/$PHP_LIBNAME/libXpm.$SHLIB_SUFFIX_NAME -o -f $i/$PHP_LIBNAME/libXpm.a && GD_XPM_DIR=$i && break
     done
 
     if test -z "$GD_XPM_DIR"; then
@@ -129,12 +129,12 @@
     PHP_CHECK_LIBRARY(Xpm,XpmFreeXpmImage, 
     [
       PHP_ADD_INCLUDE($GD_XPM_INC)
-      PHP_ADD_LIBRARY_WITH_PATH(Xpm, $GD_XPM_DIR/lib, GD_SHARED_LIBADD)
-      PHP_ADD_LIBRARY_WITH_PATH(X11, $GD_XPM_DIR/lib, GD_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(Xpm, $GD_XPM_DIR/$PHP_LIBNAME, GD_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(X11, $GD_XPM_DIR/$PHP_LIBNAME, GD_SHARED_LIBADD)
     ],[
       AC_MSG_ERROR([Problem with libXpm.(a|so) or libX11.(a|so). Please check config.log for more information.]) 
     ],[
-      -L$GD_XPM_DIR/lib -lX11
+      -L$GD_XPM_DIR/$PHP_LIBNAME -lX11
     ])
   else 
     AC_MSG_RESULT(If configure fails try --with-xpm-dir=<DIR>)
@@ -163,7 +163,7 @@
       fi
       if test -n "$TTF_DIR" ; then
         AC_DEFINE(HAVE_LIBTTF,1,[ ])
-        PHP_ADD_LIBRARY_WITH_PATH(ttf, $TTF_DIR/lib, GD_SHARED_LIBADD)
+        PHP_ADD_LIBRARY_WITH_PATH(ttf, $TTF_DIR/$PHP_LIBNAME, GD_SHARED_LIBADD)
       fi
       if test -z "$TTF_INC_DIR"; then
         TTF_INC_DIR=$TTF_DIR/include
@@ -187,7 +187,7 @@
     done
     
     if test -n "$FREETYPE2_DIR" ; then
-      PHP_ADD_LIBRARY_WITH_PATH(freetype, $FREETYPE2_DIR/lib, GD_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(freetype, $FREETYPE2_DIR/$PHP_LIBNAME, GD_SHARED_LIBADD)
       PHP_ADD_INCLUDE($FREETYPE2_DIR/include)
       PHP_ADD_INCLUDE($FREETYPE2_INC_DIR)
       AC_DEFINE(USE_GD_IMGSTRTTF, 1, [ ])
@@ -215,11 +215,11 @@
     [
       AC_DEFINE(HAVE_LIBT1,1,[ ])
       PHP_ADD_INCLUDE($GD_T1_DIR/include)
-      PHP_ADD_LIBRARY_WITH_PATH(t1, $GD_T1_DIR/lib, GD_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(t1, $GD_T1_DIR/$PHP_LIBNAME, GD_SHARED_LIBADD)
     ],[
       AC_MSG_ERROR([Problem with libt1.(a|so). Please check config.log for more information.]) 
     ],[
-      -L$GD_T1_DIR/lib
+      -L$GD_T1_DIR/$PHP_LIBNAME
     ])
   fi
 ])
@@ -358,7 +358,7 @@
   done
 
 dnl Library path
-  for i in lib/gd1.3 lib/gd lib gd1.3 gd ""; do
+  for i in lib/gd1.3 lib/gd $PHP_LIBNAME gd1.3 gd ""; do
     test -f "$PHP_GD/$i/libgd.$SHLIB_SUFFIX_NAME" -o -f "$PHP_GD/$i/libgd.a" && GD_LIB="$PHP_GD/$i"
   done
 
diff -Naur php-4.3.4/ext/gettext/config.m4 php-4.3.4.oden/ext/gettext/config.m4
--- php-4.3.4/ext/gettext/config.m4	2003-10-01 04:53:58.000000000 +0200
+++ php-4.3.4.oden/ext/gettext/config.m4	2004-01-09 20:14:48.000000000 +0100
@@ -14,7 +14,7 @@
     AC_MSG_ERROR(Cannot locate header file libintl.h)
   fi
 
-  GETTEXT_LIBDIR=$GETTEXT_DIR/lib
+  GETTEXT_LIBDIR=$GETTEXT_DIR/$PHP_LIBNAME
   GETTEXT_INCDIR=$GETTEXT_DIR/include
   
   O_LDFLAGS=$LDFLAGS
diff -Naur php-4.3.4/ext/imap/config.m4 php-4.3.4.oden/ext/imap/config.m4
--- php-4.3.4/ext/imap/config.m4	2003-10-03 07:25:35.000000000 +0200
+++ php-4.3.4.oden/ext/imap/config.m4	2004-01-09 20:14:48.000000000 +0100
@@ -71,7 +71,7 @@
     fi
 
     for i in $SEARCH_PATHS; do
-      if test -f $i/lib/libkrb5.a || test -f $i/lib/libkrb5.$SHLIB_SUFFIX_NAME; then
+      if test -f $i/$PHP_LIBNAME/libkrb5.a || test -f $i/$PHP_LIBNAME/libkrb5.$SHLIB_SUFFIX_NAME; then
         PHP_KERBEROS_DIR=$i
         break
       fi
@@ -84,7 +84,7 @@
       ])
     fi
     AC_DEFINE(HAVE_IMAP_KRB,1,[ ])
-    PHP_ADD_LIBPATH($PHP_KERBEROS_DIR/lib, IMAP_SHARED_LIBADD)
+    PHP_ADD_LIBPATH($PHP_KERBEROS_DIR/$PHP_LIBNAME, IMAP_SHARED_LIBADD)
     PHP_ADD_LIBRARY(gssapi_krb5, 1, IMAP_SHARED_LIBADD)
     PHP_ADD_LIBRARY(krb5, 1, IMAP_SHARED_LIBADD)
     PHP_ADD_LIBRARY(k5crypto, 1, IMAP_SHARED_LIBADD)
@@ -116,11 +116,11 @@
   AC_MSG_CHECKING([whether SSL libraries are needed for c-client])
 
   if test "$PHP_IMAP_SSL" != "no"; then
-    AC_MSG_RESULT([$PHP_IMAP_SSL/lib])
+    AC_MSG_RESULT([$PHP_IMAP_SSL/$PHP_LIBNAME])
     AC_DEFINE(HAVE_IMAP_SSL,1,[ ])
     PHP_ADD_LIBRARY_DEFER(ssl,    1, IMAP_SHARED_LIBADD)
     PHP_ADD_LIBRARY_DEFER(crypto, 1, IMAP_SHARED_LIBADD)
-    PHP_ADD_LIBPATH($PHP_IMAP_SSL/lib, IMAP_SHARED_LIBADD)
+    PHP_ADD_LIBPATH($PHP_IMAP_SSL/$PHP_LIBNAME, IMAP_SHARED_LIBADD)
   else
     AC_MSG_RESULT(no)
   fi
@@ -175,7 +175,7 @@
 
     for lib in c-client4 c-client imap; do
       IMAP_LIB=$lib
-      IMAP_LIB_CHK(lib)
+      IMAP_LIB_CHK($PHP_LIBNAME)
       IMAP_LIB_CHK(c-client)
     done
 
diff -Naur php-4.3.4/ext/ldap/config.m4 php-4.3.4.oden/ext/ldap/config.m4
--- php-4.3.4/ext/ldap/config.m4	2003-07-01 02:02:29.000000000 +0200
+++ php-4.3.4.oden/ext/ldap/config.m4	2004-01-09 20:14:48.000000000 +0100
@@ -6,7 +6,7 @@
   if test -f $1/include/ldap.h; then
     LDAP_DIR=$1
     LDAP_INCDIR=$1/include
-    LDAP_LIBDIR=$1/lib
+    LDAP_LIBDIR=$1/$PHP_LIBNAME
   elif test -f $1/include/umich-ldap/ldap.h; then
     LDAP_DIR=$1
     LDAP_INCDIR=$1/include/umich-ldap
diff -Naur php-4.3.4/ext/mysql/config.m4 php-4.3.4.oden/ext/mysql/config.m4
--- php-4.3.4/ext/mysql/config.m4	2003-07-14 12:43:51.000000000 +0200
+++ php-4.3.4.oden/ext/mysql/config.m4	2004-01-09 20:14:48.000000000 +0100
@@ -113,7 +113,7 @@
 
   MYSQL_MODULE_TYPE=external
 
-  for i in lib lib/mysql; do
+  for i in $PHP_LIBNAME $PHP_LIBNAME/mysql; do
     MYSQL_LIB_CHK($i)
   done
 
diff -Naur php-4.3.4/ext/pgsql/config.m4 php-4.3.4.oden/ext/pgsql/config.m4
--- php-4.3.4/ext/pgsql/config.m4	2002-06-26 15:07:40.000000000 +0200
+++ php-4.3.4.oden/ext/pgsql/config.m4	2004-01-09 20:14:48.000000000 +0100
@@ -29,7 +29,7 @@
       fi
     done
 
-    for j in lib lib/pgsql lib/postgres lib/postgresql ""; do
+    for j in $PHP_LIBNAME lib/pgsql lib/postgres lib/postgresql ""; do
       if test -f "$i/$j/libpq.so" || test -f "$i/$j/libpq.a"; then 
         PGSQL_LIBDIR=$i/$j
       fi
diff -Naur php-4.3.4/ext/snmp/config.m4 php-4.3.4.oden/ext/snmp/config.m4
--- php-4.3.4/ext/snmp/config.m4	2003-09-23 10:16:47.000000000 +0200
+++ php-4.3.4.oden/ext/snmp/config.m4	2004-01-09 20:14:48.000000000 +0100
@@ -46,12 +46,12 @@
         test -f $i/snmp/include/ucd-snmp/snmp.h && SNMP_INCDIR=$i/snmp/include/ucd-snmp
       done
       for i in /usr /usr/snmp /usr/local /usr/local/snmp; do
-        test -f $i/lib/libsnmp.a -o -f $i/lib/libsnmp.$SHLIB_SUFFIX_NAME && SNMP_LIBDIR=$i/lib
+        test -f $i/$PHP_LIBNAME/libsnmp.a -o -f $i/$PHP_LIBNAME/libsnmp.$SHLIB_SUFFIX_NAME && SNMP_LIBDIR=$i/$PHP_LIBNAME
       done
     else
       SNMP_INCDIR=$PHP_SNMP/include
       test -d $PHP_SNMP/include/ucd-snmp && SNMP_INCDIR=$PHP_SNMP/include/ucd-snmp
-      SNMP_LIBDIR=$PHP_SNMP/lib
+      SNMP_LIBDIR=$PHP_SNMP/$PHP_LIBNAME
     fi
 
     if test -z "$SNMP_INCDIR"; then
diff -Naur php-4.3.4/ext/xml/config.m4 php-4.3.4.oden/ext/xml/config.m4
--- php-4.3.4/ext/xml/config.m4	2003-10-03 07:25:42.000000000 +0200
+++ php-4.3.4.oden/ext/xml/config.m4	2004-01-09 20:14:48.000000000 +0100
@@ -31,7 +31,7 @@
     PHP_NEW_EXTENSION(xml, xml.c, $ext_shared)
 
     for i in $PHP_XML $PHP_EXPAT_DIR; do
-      if test -f $i/lib/libexpat.a -o -f $i/lib/libexpat.$SHLIB_SUFFIX_NAME ; then
+      if test -f $i/$PHP_LIBNAME/libexpat.a -o -f $i/$PHP_LIBNAME/libexpat.$SHLIB_SUFFIX_NAME ; then
         EXPAT_DIR=$i
       fi
     done
@@ -41,7 +41,7 @@
     fi
 
     PHP_ADD_INCLUDE($EXPAT_DIR/include)
-    PHP_ADD_LIBRARY_WITH_PATH(expat, $EXPAT_DIR/lib, XML_SHARED_LIBADD)
+    PHP_ADD_LIBRARY_WITH_PATH(expat, $EXPAT_DIR/$PHP_LIBNAME, XML_SHARED_LIBADD)
     PHP_SUBST(XML_SHARED_LIBADD)
   fi
 fi
diff -Naur php-4.3.4/scripts/Makefile.frag php-4.3.4.oden/scripts/Makefile.frag
--- php-4.3.4/scripts/Makefile.frag	2003-09-19 07:35:21.000000000 +0200
+++ php-4.3.4.oden/scripts/Makefile.frag	2004-01-09 20:14:48.000000000 +0100
@@ -4,7 +4,7 @@
 #
 
 phpincludedir = $(includedir)/php
-phpbuilddir = $(prefix)/lib/php/build
+phpbuilddir = $(libdir)/php/build
 
 BUILD_FILES = \
 	scripts/phpize.m4 \
diff -Naur php-4.3.4/scripts/phpize.in php-4.3.4.oden/scripts/phpize.in
--- php-4.3.4/scripts/phpize.in	2003-08-27 19:57:32.000000000 +0200
+++ php-4.3.4.oden/scripts/phpize.in	2004-01-09 20:14:48.000000000 +0100
@@ -1,7 +1,7 @@
 #! /bin/sh
 
 prefix='@prefix@'
-phpdir="$prefix/lib/php/build"
+phpdir="@libdir@/php/build"
 includedir="$prefix/include/php"
 builddir="`pwd`"
 
diff -Naur php-4.3.4/scripts/phpize.m4 php-4.3.4.oden/scripts/phpize.m4
--- php-4.3.4/scripts/phpize.m4	2003-09-03 10:07:51.000000000 +0200
+++ php-4.3.4.oden/scripts/phpize.m4	2004-01-09 20:14:48.000000000 +0100
@@ -47,6 +47,7 @@
 AC_PROG_CC_C_O
 
 PHP_RUNPATH_SWITCH
+PHP_CHECK_LIBNAME
 PHP_SHLIB_SUFFIX_NAME
 PHP_WITH_PHP_CONFIG
 
