--- tetex-src-2.0.2/texk/tetex/texconfig.orig	2003-03-01 10:13:18.000000000 +0100
+++ tetex-src-2.0.2/texk/tetex/texconfig	2003-03-03 20:49:00.000000000 +0100
@@ -443,7 +443,6 @@
 {
   echo "Running mf to create plain base ..."
   fmtutil --byfmt mf
-  fmtutil --byfmt mfw
   echo "Done."; echo
 }
 
--- tetex-src-2.0.2/texk/tetex/texconfig.orig	2003-09-06 09:44:03.000000000 +0200
+++ tetex-src-2.0.2/texk/tetex/texconfig	2003-09-05 23:44:25.000000000 +0200
@@ -1040,12 +1040,12 @@
 fontmaking := 0;
 myexit;
 eof
-  mf '\mode='"$1"';  \input ./mftmp' </dev/null | awk '$1 == "XDPI" || $1 == "YDPI" { print $3 }'
+  mf-nowin '\mode='"$1"';  \input ./mftmp' </dev/null | awk '$1 == "XDPI" || $1 == "YDPI" { print $3 }'
 }
 
 mode_name()
 {
-  mf '\mode='"$1"';  \input modename' | sed -n '/^MODENAME/!d; s/.*= //; s/_.*//; p'
+  mf-nowin '\mode='"$1"';  \input modename' | sed -n '/^MODENAME/!d; s/.*= //; s/_.*//; p'
 }
 
 list_formats()
@@ -1451,7 +1451,7 @@
 do_print_setup()
 {
   echo '==================== binaries found by searching $PATH ==================='
-  echo_locate_binary tex etex dvipdfm pdftex omega mf mpost tcdialog dvips xdvi kpsewhich mktexpk mktextfm
+  echo_locate_binary tex etex dvipdfm pdftex omega mf-nowin mpost tcdialog dvips xdvi kpsewhich mktexpk mktextfm
   echo
   echo '=========================== Kpathsea variables ==========================='
   echo_show_variable TEXMF TEXMFMAIN VARTEXMF
