--- gettext-0.11.5/src/msgfmt.c.i18n	2003-06-03 10:52:59.000000000 -0400
+++ gettext-0.11.5/src/msgfmt.c	2003-06-03 10:54:10.000000000 -0400
@@ -1446,9 +1446,13 @@
 	 translations are equal or different.  This is for consistency
 	 with msgmerge, msgcat and others.  The user can use the
 	 msguniq program to get rid of duplicates.  */
-      po_gram_error_at_line (msgid_pos, _("duplicate message definition"));
-      po_gram_error_at_line (&mp->pos, _("\
+
+      /* LSB/LI18NUX expects duplicates to be ignored if -o specified */
+      if (output_file_name == NULL) {
+	po_gram_error_at_line (msgid_pos, _("duplicate message definition"));
+	po_gram_error_at_line (&mp->pos, _("\
 ...this is the location of the first definition"));
+      }
 
       /* We don't need the just constructed entries' parameter string
 	 (allocated in po-gram.y).  */
