--- ypserv-2.10/lib/access.c.nomap	2003-06-13 14:09:14.000000000 +0200
+++ ypserv-2.10/lib/access.c	2003-11-06 11:13:00.000000000 +0100
@@ -190,6 +190,8 @@
 		  status = -1;
 	      ypdb_close (dbp);
 	    }
+	  else
+	    status = -4;
 	}
     }
 
--- ypserv-2.10/ypserv/server.c.nomap	2003-08-15 10:55:05.000000000 +0200
+++ ypserv-2.10/ypserv/server.c	2003-11-06 11:10:05.000000000 +0100
@@ -177,6 +177,11 @@
 	    log_msg ("\t-> Ignored (not a valid domain)");
 	  result->stat = YP_NODOM;
 	  break;
+	case -4:
+	  if (debug_flag)
+	    log_msg ("\t-> Ignored (map name not found)");
+	  result->stat = YP_NOMAP;
+	  break;
 	}
       return TRUE;
     }
