--- man-1.5k/src/man.conf.in.mandirs	2002-07-17 13:08:40.000000000 -0600
+++ man-1.5k/src/man.conf.in	2004-06-15 23:34:18.000000000 -0600
@@ -36,10 +36,11 @@
 #
 # Every automatically generated MANPATH includes these fields
 #
-MANPATH	/usr/man
 MANPATH /usr/share/man
-MANPATH	/usr/local/man
 MANPATH	/usr/X11R6/man
+MANPATH	/usr/local/share/man
+MANPATH /usr/kerberos/man
+MANPATH	/usr/man
 #
 # Uncomment if you want to include one of these by default
 #
@@ -65,11 +66,13 @@
 MANPATH_MAP	/sbin			/usr/share/man
 MANPATH_MAP	/usr/bin		/usr/share/man
 MANPATH_MAP	/usr/sbin		/usr/share/man
-MANPATH_MAP	/usr/local/bin		/usr/local/man
-MANPATH_MAP	/usr/local/sbin		/usr/local/man
+MANPATH_MAP	/usr/local/bin		/usr/local/share/man
+MANPATH_MAP	/usr/local/sbin		/usr/local/share/man
 MANPATH_MAP	/usr/X11R6/bin		/usr/X11R6/man
 MANPATH_MAP	/usr/bin/X11		/usr/X11R6/man
 MANPATH_MAP	/usr/bin/mh		/usr/share/man
+MANPATH_MAP	/usr/kerberos/bin	/usr/kerberos/man
+MANPATH_MAP	/usr/kerberos/sbin	/usr/kerberos/man
 #
 # NOAUTOPATH keeps man from automatically adding directories that look like
 # manual page directories to the path.
