--- man-1.5i/src/man.c.security	Mon Mar 19 22:59:15 2001
+++ man-1.5i/src/man.c	Tue Apr 24 16:13:47 2001
@@ -1110,6 +1110,20 @@
      extern int optind;
 #endif
 
+     char *s;
+#define CHECK(p, l) s=getenv(p); if(s && (strlen(s)>l)) { fprintf(stderr, "ERROR: Environment variable %s too long!\n", p); exit(1); }
+
+     CHECK("LANG", 32);
+     CHECK("MANPAGER", 128);
+     CHECK("PAGER", 128);
+     CHECK("SYSTEM", 64);
+     CHECK("MANROFFSEQ", 128);
+     CHECK("MANSECT", 128);
+     CHECK("MANPL", 128);
+     CHECK("MAN_HP_DIREXT", 128);
+     CHECK("LANGUAGE", 128);
+     CHECK("LC_MESSAGES", 128);
+
 #ifndef __FreeBSD__ 
      /* Slaven Rezif: FreeBSD-2.2-SNAP does not recognize LC_MESSAGES. */
      setlocale(LC_MESSAGES, "");
--- man-1.5i/src/makewhatis.sh.security	Mon Mar 19 17:15:24 2001
+++ man-1.5i/src/makewhatis.sh	Tue Apr 24 16:48:26 2001
@@ -36,8 +36,8 @@
 # In case both /usr/man and /usr/share/manexist, the former is local
 # and should be first.
 # It is a bug to add /var/cache/man to DEFCATPATH.
-DEFMANPATH=/usr/man:/usr/share/man
-DEFCATPATH=/usr/man/preformat:/usr/man:/usr/share/man/preformat:/usr/share/man
+DEFMANPATH=/usr/share/man:/usr/man
+DEFCATPATH=/var/cache/man:/usr/man/preformat:/usr/share/man:/usr/man
 
 # In case /usr is read-only, make /usr/man/whatis (etc) a symlink to
 # something like /var/cache/man/whatis.
@@ -52,12 +52,12 @@
 # We try here to be careful (and avoid preconstructed symlinks)
 # in case makewhatis is run as root, by creating a subdirectory of /tmp.
 
-TMPFILEDIR=/tmp/whatis.tmp.dir.$$
-rm -rf TMPFILEDIR
-if ! mkdir -m 0700 $TMPFILEDIR; then
-    echo Could not create $TMPFILEDIR
-    exit 1;
+TMPFILEDIR=`mktemp -d /tmp/makewhatisXXXXXX`
+if [ $? -ne 0 ]; then
+	echo "$0: Can't create temp file, exiting..."
+	exit 1
 fi
+chmod 0700 $TMPFILEDIR
 TMPFILE=$TMPFILEDIR/w
 
 # make sure TMPFILEDIR is deleted if program is killed or terminates
