--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/input/mouse/mouse.c.mouse-twice	2003-02-04 16:21:18.000000000 +0100
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/input/mouse/mouse.c	2003-02-17 08:13:30.000000000 +0100
@@ -991,7 +991,17 @@
     /* OS/2 has a mouse handled by the OS - it cannot fail here */
 
     /* Check if the device can be opened. */
-    pInfo->fd = xf86OpenSerial(pInfo->options);
+    if (pInfo->fd == -1) {
+      pInfo->fd = xf86OpenSerial(pInfo->options);
+    }
+    
+    /* FL [Wed Nov 27 04:07:13 EST 2002]
+     * Try a second time if it fails to workaround a kernel
+     * bug that make the mouse device disappear */
+    if (pInfo->fd == -1) {
+      pInfo->fd = xf86OpenSerial(pInfo->options);
+    }
+    
     if (pInfo->fd == -1) {
 	if (xf86GetAllowMouseOpenFail())
 	    xf86Msg(X_WARNING, "%s: cannot open input device\n", pInfo->name);
@@ -1004,9 +1014,9 @@
 	    return pInfo;
 	}
     }
-    xf86CloseSerial(pInfo->fd);
+    /* xf86CloseSerial(pInfo->fd); */
 #endif
-    pInfo->fd = -1;
+    /* pInfo->fd = -1; */
 
     if (!(mPriv = (pointer) xcalloc(sizeof(mousePrivRec), 1)))
 	return pInfo;
@@ -1570,7 +1580,10 @@
 	break;
 
     case DEVICE_ON:
-	pInfo->fd = xf86OpenSerial(pInfo->options);
+	if (pInfo->fd == -1)
+	    pInfo->fd = xf86OpenSerial(pInfo->options);
+	if (pInfo->fd == -1)
+	    pInfo->fd = xf86OpenSerial(pInfo->options);
 	if (pInfo->fd == -1)
 	    xf86Msg(X_WARNING, "%s: cannot open input device\n", pInfo->name);
 	else {
