--- itcl3.2/itcl/Makefile.in.makecfg	Mon Aug 28 23:58:00 2000
+++ itcl3.2/itcl/Makefile.in	Tue Aug 28 00:12:27 2001
@@ -23,8 +23,9 @@
 # library that we are building.
 #========================================================================
 
-lib_BINARIES=$(itcl_LIB_FILE) $(itclstub_LIB_FILE)
-BINARIES=$(lib_BINARIES)
+lib_BINARIES=libitcl.so.0 libitcl.so $(itclstub_LIB_FILE)
+bin_BINARIES=itclsh
+BINARIES=$(lib_BINARIES) $(bin_BINARIES)
 
 #========================================================================
 # Enumerate the names of the source files included in this package.
@@ -147,7 +148,7 @@
 includedir = @includedir@
 oldincludedir = /usr/include
 
-DESTDIR =
+DESTDIR = $$INSTALL_ROOT
 
 pkgdatadir = $(datadir)/@PACKAGE@@VERSION@
 pkglibdir = $(libdir)/@PACKAGE@@VERSION@
@@ -155,8 +156,8 @@
 
 top_builddir = .
 
-#INSTALL = $(top_srcdir)/../config/install-sh -c
-INSTALL = $(TCLSH_PROG) `@CYGPATH@ $(srcdir)/../config/installFile.tcl` -c
+INSTALL = $(top_srcdir)/../config/install-sh -c
+#INSTALL = $(TCLSH_PROG) `@CYGPATH@ $(srcdir)/../config/installFile.tcl` -c
 INSTALL_PROGRAM = @INSTALL_PROGRAM@
 INSTALL_DATA = @INSTALL_DATA@
 INSTALL_SCRIPT = @INSTALL_SCRIPT@
@@ -174,22 +175,22 @@
 VERSION = @VERSION@
 CC = @CC@
 CFLAGS_DEBUG = @CFLAGS_DEBUG@
-CFLAGS_DEFAULT = @CFLAGS_DEFAULT@
-CFLAGS_OPTIMIZE = @CFLAGS_OPTIMIZE@
+CFLAGS_DEFAULT = -O2
+CFLAGS_OPTIMIZE = -O2
 CLEANFILES = @CLEANFILES@
 EXEEXT = @EXEEXT@
 LDFLAGS_DEBUG = @LDFLAGS_DEBUG@
 LDFLAGS_DEFAULT = @LDFLAGS_DEFAULT@
 LDFLAGS_OPTIMIZE = @LDFLAGS_OPTIMIZE@
-MAKE_LIB = @MAKE_LIB@
-MAKE_SHARED_LIB = @MAKE_SHARED_LIB@
-MAKE_STATIC_LIB = @MAKE_STATIC_LIB@
+MAKE_LIB = ${SHLIB_LD} -o $@ $(OBJECTS) ${SHLIB_LDFLAGS} ${SHLIB_LD_LIBS}
+MAKE_SHARED_LIB = ${SHLIB_LD} -o $@ $(OBJECTS) ${SHLIB_LDFLAGS} ${SHLIB_LD_LIBS}
+MAKE_STATIC_LIB = ${STLIB_LD} $@ $(OBJECTS)
 OBJEXT = @OBJEXT@
 RANLIB = @RANLIB@
 SHLIB_CFLAGS = @SHLIB_CFLAGS@
-SHLIB_LD = @SHLIB_LD@
+SHLIB_LD = gcc -pipe -shared -Wl,-soname,libitcl.so.0
 SHLIB_LDFLAGS = @SHLIB_LDFLAGS@
-SHLIB_LD_LIBS = @SHLIB_LD_LIBS@
+SHLIB_LD_LIBS = @TCL_BUILD_STUB_LIB_SPEC@
 STLIB_LD = @STLIB_LD@
 TCL_BIN_DIR = @TCL_BIN_DIR@
 TCL_DEFS = @TCL_DEFS@
@@ -269,16 +270,16 @@
 #========================================================================
 
 install-libraries: libraries
-	$(mkinstalldirs) $(includedir)
+	$(mkinstalldirs) $(DESTDIR)$(includedir)
 	@echo "Installing header files in $(includedir)"
 	@for i in $(GENERIC_HDRS) ; do \
 	    echo "Installing $$i" ; \
-	    $(INSTALL_DATA) $$i $(includedir) ; \
+	    $(INSTALL_DATA) $$i $(DESTDIR)$(includedir) ; \
 	done;
 	@echo "Installing library files in $(ITCL_LIBRARY)"
 	@for i in $(srcdir)/library/*.tcl ; do \
 	    echo "Installing $$i" ; \
-	    $(INSTALL_DATA) $$i $(ITCL_LIBRARY) ; \
+	    $(INSTALL_DATA) $$i $(DESTDIR)$(ITCL_LIBRARY) ; \
 	done;
 
 #========================================================================
@@ -287,13 +288,13 @@
 #========================================================================
 
 install-doc: doc
-	$(mkinstalldirs) $(mandir)/mann
+	$(mkinstalldirs) $(DESTDIR)$(mandir)/mann
 	@echo "Installing man pages in $(mandir)"
 	@for i in $(srcdir)/doc/*.n; \
 	    do \
 	    echo "Installing $$i"; \
-	    rm -f $(mandir)/mann/$$i; \
-	    $(INSTALL_DATA) $$i $(mandir)/mann ; \
+	    rm -f $(DESTDIR)$(mandir)/mann/$$i; \
+	    $(INSTALL_DATA) $$i $(DESTDIR)$(mandir)/mann ; \
 	    done
 
 test: $(TCLSH_PROG)
@@ -316,16 +317,22 @@
 # name from "exampleA" to your package's name.
 #========================================================================
 
-$(itcl_LIB_FILE): $(itcl_OBJECTS)
-	-rm -f $(itcl_LIB_FILE)
-	@MAKE_LIB@
-	$(RANLIB) $(itcl_LIB_FILE)
+libitcl.so.0: $(itcl_OBJECTS)
+	-rm -f $@
+	${SHLIB_LD} -o $@ $? ${SHLIB_LDFLAGS} ${SHLIB_LD_LIBS}
+	$(RANLIB) $@
+
+libitcl.so: libitcl.so.0
+	ln -sf $< $@
 
 $(itclstub_LIB_FILE): $(itclstub_OBJECTS)
 	-rm -f $(itclstub_LIB_FILE)
-	@MAKE_STATIC_LIB@
+	${STLIB_LD} $@ $?
 	$(RANLIB) $(itclstub_LIB_FILE)
 
+itclsh: tclAppInit.o
+	$(CC) -o $@ $? -L. -litcl @TCL_BUILD_STUB_LIB_SPEC@ -ltcl
+
 #========================================================================
 # We need to enumerate the list of .c to .o lines here.
 # Unfortunately, there does not seem to be any other way to do this
@@ -443,7 +450,7 @@
 	    $(RANLIB) $(DESTDIR)$(libdir)/$$p; \
 	  else :; fi; \
 	done
-	$(INSTALL_DATA) pkgIndex.tcl $(pkglibdir)
+	$(INSTALL_DATA) pkgIndex.tcl $(DESTDIR)$(pkglibdir)
 
 #========================================================================
 # Install binary executables (e.g. .exe files)
@@ -454,8 +461,8 @@
 install-bin-binaries: installdirs
 	@list='$(bin_BINARIES)'; for p in $$list; do \
 	  if test -f $$p; then \
-	    echo " $(INSTALL_DATA) $$p $(DESTDIR)$(bindir)/$$p"; \
-	    $(INSTALL_DATA) $$p $(DESTDIR)$(bindir)/$$p; \
+	    echo " $(INSTALL_PROGRAM) $$p $(DESTDIR)$(bindir)/$$p"; \
+	    $(INSTALL_PROGRAM) $$p $(DESTDIR)$(bindir)/$$p; \
 	  else :; fi; \
 	done
 
@@ -481,7 +488,7 @@
 	$(mkinstalldirs)  $(DESTDIR)$(libdir)
 	$(mkinstalldirs)  $(DESTDIR)$(bindir)
 	$(mkinstalldirs)  $(DESTDIR)$(pkglibdir)
-	$(mkinstalldirs)  $(ITCL_LIBRARY)
+	$(mkinstalldirs)  $(DESTDIR)$(ITCL_LIBRARY)
 
 .PHONY: all binaries clean depend distclean doc install installdirs \
 libraries test
--- itcl3.2/itk/Makefile.in.makecfg	Mon Aug 28 23:58:00 2000
+++ itcl3.2/itk/Makefile.in	Tue Aug 28 00:13:33 2001
@@ -23,8 +23,9 @@
 # library that we are building.
 #========================================================================
 
-lib_BINARIES=$(itk_LIB_FILE) $(itkstub_LIB_FILE)
-BINARIES=$(lib_BINARIES)
+lib_BINARIES=libitk.so.0 libitk.so $(itkstub_LIB_FILE)
+bin_BINARIES=itkwish
+BINARIES=$(lib_BINARIES) $(bin_BINARIES)
 
 #========================================================================
 # Enumerate the names of the source files included in this package.
@@ -36,9 +37,9 @@
 GENERIC_DIR=$(srcdir)/generic
 PLATFORM_DIR=@PLATFORM_DIR@
 
-WIN_SOURCES=$(PLATFORM_DIR)/tclAppInit.c \
+WIN_SOURCES=$(PLATFORM_DIR)/tkAppInit.c \
 	$(PLATFORM_DIR)/dllEntryPoint.c
-UNIX_SOURCES=$(PLATFORM_DIR)/tclAppInit.c
+UNIX_SOURCES=$(PLATFORM_DIR)/tkAppInit.c
 GENERIC_SOURCES = $(GENERIC_DIR)/itk_cmds.c \
 	$(GENERIC_DIR)/itk_option.c \
 	$(GENERIC_DIR)/itk_archetype.c \
@@ -84,7 +85,7 @@
 itkstub_LIB_FILE = @itkstub_LIB_FILE@
 $(itkstub_LIB_FILE)_OBJECTS = $(itkstub_OBJECTS)
 
-itclstub_LIB_FILE = @itclstub_LIB_FILE@
+itclstub_LIB_FILE = ../itcl/libitclstub.a
 
 tkstub_LIB_SPEC = @tkstub_LIB_SPEC@
 
@@ -135,7 +136,7 @@
 includedir = @includedir@
 oldincludedir = /usr/include
 
-DESTDIR =
+DESTDIR = $$INSTALL_ROOT
 
 pkgdatadir = $(datadir)/@PACKAGE@@VERSION@
 pkglibdir = $(libdir)/@PACKAGE@@VERSION@
@@ -143,8 +144,8 @@
 
 top_builddir = .
 
-#INSTALL = $(top_srcdir)/../config/install-sh -c
-INSTALL = $(TCLSH_PROG) `@CYGPATH@ $(top_srcdir)/../config/installFile.tcl` -c
+INSTALL = $(top_srcdir)/../config/install-sh -c
+#INSTALL = $(TCLSH_PROG) `@CYGPATH@ $(top_srcdir)/../config/installFile.tcl` -c
 INSTALL_PROGRAM = @INSTALL_PROGRAM@
 INSTALL_DATA = @INSTALL_DATA@
 INSTALL_SCRIPT = @INSTALL_SCRIPT@
@@ -162,22 +163,22 @@
 VERSION = @VERSION@
 CC = @CC@
 CFLAGS_DEBUG = @CFLAGS_DEBUG@
-CFLAGS_DEFAULT = @CFLAGS_DEFAULT@
-CFLAGS_OPTIMIZE = @CFLAGS_OPTIMIZE@
+CFLAGS_DEFAULT = -O2
+CFLAGS_OPTIMIZE = -O2
 CLEANFILES = @CLEANFILES@
 EXEEXT = @EXEEXT@
 LDFLAGS_DEBUG = @LDFLAGS_DEBUG@
 LDFLAGS_DEFAULT = @LDFLAGS_DEFAULT@
 LDFLAGS_OPTIMIZE = @LDFLAGS_OPTIMIZE@
-MAKE_LIB = @MAKE_LIB@
-MAKE_SHARED_LIB = @MAKE_SHARED_LIB@
-MAKE_STATIC_LIB = @MAKE_STATIC_LIB@
-OBJEXT = @OBJEXT@
+MAKE_LIB = ${SHLIB_LD} -o $@ $(itk_OBJECTS) ${SHLIB_LDFLAGS} ${SHLIB_LD_LIBS}
+MAKE_SHARED_LIB = ${SHLIB_LD} -o $@ $(itk_OBJECTS) ${SHLIB_LDFLAGS} ${SHLIB_LD_LIBS}
+MAKE_STATIC_LIB = ${STLIB_LD} $@ $(itk_OBJECTS)
+OBJEXT = o
 RANLIB = @RANLIB@
 SHLIB_CFLAGS = @SHLIB_CFLAGS@
-SHLIB_LD = @SHLIB_LD@
+SHLIB_LD = gcc -pipe -shared -Wl,-soname,libitk.so.0
 SHLIB_LDFLAGS = @SHLIB_LDFLAGS@
-SHLIB_LD_LIBS = @SHLIB_LD_LIBS@
+SHLIB_LD_LIBS = @TCL_BUILD_STUB_LIB_SPEC@ -L@TK_SRC_DIR@/unix -ltk
 STLIB_LD = @STLIB_LD@
 TCL_BIN_DIR = @TCL_BIN_DIR@
 TCL_DEFS = @TCL_DEFS@
@@ -311,16 +312,22 @@
 # name from "exampleA" to your package's name.
 #========================================================================
 
-$(itk_LIB_FILE): $(itk_OBJECTS)
-	-rm -f $(itk_LIB_FILE)
-	@MAKE_LIB@
-	$(RANLIB) $(itk_LIB_FILE)
+libitk.so.0: $(itk_OBJECTS)
+	-rm -f $@
+	${SHLIB_LD} -o $@ $(itk_OBJECTS) ${SHLIB_LDFLAGS} $(itclstub_LIB_FILE) ${SHLIB_LD_LIBS}
+	$(RANLIB) $@
+
+libitk.so: libitk.so.0
+	ln -sf $< $@
 
 $(itkstub_LIB_FILE): $(itkstub_OBJECTS)
 	-rm -f $(itkstub_LIB_FILE)
-	@MAKE_STATIC_LIB@
+	${STLIB_LD} $@ $(itk_OBJECTS)
 	$(RANLIB) $(itkstub_LIB_FILE)
 
+itkwish: tkAppInit.$(OBJEXT)
+	$(CC) -o $@ $? -L. -litk -L../itcl -litcl @TCL_BUILD_STUB_LIB_SPEC@ -ltcl @TK_BUILD_STUB_LIB_SPEC@ -ltk
+
 #========================================================================
 # We need to enumerate the list of .c to .o lines here.
 # Unfortunately, there does not seem to be any other way to do this
@@ -338,8 +345,8 @@
 dllEntryPoint.$(OBJEXT): $(WIN_DIR)/dllEntryPoint.c
 	$(COMPILE) -c `@CYGPATH@ $(WIN_DIR)/dllEntryPoint.c` -o $@
 
-tclAppInit.$(OBJEXT): $(PLATFORM_DIR)/tclAppInit.c
-	$(COMPILE) -c `@CYGPATH@ $(PLATFORM_DIR)/tclAppInit.c` -o $@
+tkAppInit.$(OBJEXT): $(PLATFORM_DIR)/tkAppInit.c
+	$(COMPILE) -c `@CYGPATH@ $(PLATFORM_DIR)/tkAppInit.c` -o $@
 
 itkStubInit.$(OBJEXT): $(GENERIC_DIR)/itkStubInit.c
 	$(COMPILE) -c `@CYGPATH@ $(GENERIC_DIR)/itkStubInit.c` -o $@
@@ -428,8 +435,8 @@
 install-bin-binaries: installdirs
 	@list='$(bin_BINARIES)'; for p in $$list; do \
 	  if test -f $$p; then \
-	    echo " $(INSTALL_DATA) $$p $(DESTDIR)$(bindir)/$$p"; \
-	    $(INSTALL_DATA) $$p $(DESTDIR)$(bindir)/$$p; \
+	    echo " $(INSTALL_PROGRAM) $$p $(DESTDIR)$(bindir)/$$p"; \
+	    $(INSTALL_PROGRAM) $$p $(DESTDIR)$(bindir)/$$p; \
 	  else :; fi; \
 	done
 
--- itcl3.2/itk/configure.makecfg	Tue Aug 28 00:35:35 2001
+++ itcl3.2/itk/configure	Tue Aug 28 00:36:53 2001
@@ -2224,6 +2224,7 @@
 s%@TK_SRC_DIR@%$TK_SRC_DIR%g
 s%@TK_LIB_FILE@%$TK_LIB_FILE%g
 s%@TK_XINCLUDES@%$TK_XINCLUDES%g
+s%@TK_BUILD_STUB_LIB_SPEC@%$TK_BUILD_STUB_LIB_SPEC%g
 s%@ITCL_INCLUDES@%$ITCL_INCLUDES%g
 s%@itclstub_LIB_FILE@%$itclstub_LIB_FILE%g
 s%@tkstub_LIB_SPEC@%$tkstub_LIB_SPEC%g
