--- mpage-2.5.4/Makefile.config	2004-05-30 15:38:31.000000000 -0400
+++ mpage-2.5.4/Makefile	2005-08-21 16:21:15.000000000 -0400
@@ -39,8 +39,8 @@ CC=gcc
 #          A4 for European A4
 #  See glob.c for all possibilities
 #
-# PAGESIZE=Letter
-PAGESIZE=A4
+PAGESIZE=Letter
+# PAGESIZE=A4
 
 #
 # Define your spooler type
@@ -50,11 +50,11 @@ PAGESIZE=A4
 SPOOL_TYPE=BSD_SPOOLER
 
 # PREFIX=e:/usr		# OS/2 type
-PREFIX=/usr/local
-#PREFIX=/usr
+# PREFIX=/usr/local
+PREFIX=/usr
 BINDIR=$(PREFIX)/bin
 LIBDIR=$(PREFIX)/share
-MANDIR=$(PREFIX)/man/man1
+MANDIR=$(PREFIX)/share/man/man1
 
 #
 # A default encoding is given in encoding.h. Setting ENCODING=1 will
@@ -107,7 +107,7 @@ O = .o
 # If you are using gcc, you probably don't need to change anything here.
 
 # Linux:
-CFLAGS = -O2 -s $(DEFS) -Wall
+CFLAGS = -s $(DEFS) $(RPM_OPT_FLAGS) -Wall
 
 # AIX (xlC on aix 4):
 #CFLAGS = -O2 -s $(DEFS)
@@ -178,10 +178,10 @@ mpage.ps:	mpage.1
 # add your proper install stuff
 #
 install:
-	if [ ! -d $(LIBDIR)/mpage ] ; then mkdir -p $(LIBDIR)/mpage ; fi
-	if [ ! -d $(BINDIR) ] ; then mkdir -p $(BINDIR) ; fi
-	if [ ! -d $(MANDIR) ] ; then mkdir -p $(MANDIR) ; fi
-	cp mpage$(E) $(BINDIR)
-	cp mpage.1 $(MANDIR)
-	-cp Encodings/* $(LIBDIR)/mpage
-	-chmod 644 $(LIBDIR)/mpage/*
+	if [ ! -d $(DESTDIR)$(LIBDIR)/mpage ] ; then mkdir -p $(DESTDIR)$(LIBDIR)/mpage ; fi
+	if [ ! -d $(DESTDIR)$(BINDIR) ] ; then mkdir -p $(DESTDIR)$(BINDIR) ; fi
+	if [ ! -d $(DESTDIR)$(MANDIR) ] ; then mkdir -p $(DESTDIR)$(MANDIR) ; fi
+	cp mpage$(E) $(DESTDIR)$(BINDIR)/
+	cp mpage.1 $(DESTDIR)$(MANDIR)/
+	-cp Encodings/* $(DESTDIR)$(LIBDIR)/mpage/
+	-chmod 644 $(DESTDIR)$(LIBDIR)/mpage/*
