--- nfs-utils-1.0.7/support/nfs/xlog.c.save	2003-07-24 21:53:59.000000000 -0400
+++ nfs-utils-1.0.7/support/nfs/xlog.c	2005-03-19 11:51:40.236851728 -0500
@@ -167,7 +167,7 @@ xlog(int kind, const char *fmt, ...)
 	}
 	if (!logged || foreground) {
 		if (!logged && log_fp == NULL) {
-			syslog(LOG_DEBUG, "%s", buff);
+			syslog(LOG_INFO, "%s", buff);
 		} else if (log_fp != NULL) {
 #ifdef VERBOSE_PRINTF
 			time(&now);
