Index: xmlsave.c
===================================================================
RCS file: /cvs/gnome/gnome-xml/xmlsave.c,v
retrieving revision 1.27
retrieving revision 1.28
diff -u -p -r1.27 -r1.28
--- libxml2/xmlsave.c	3 Sep 2005 14:56:36 -0000	1.27
+++ libxml2/xmlsave.c	12 Sep 2005 00:03:43 -0000	1.28
@@ -1086,6 +1086,11 @@ xhtmlNodeDumpOutput(xmlSaveCtxtPtr ctxt,
     xmlOutputBufferPtr buf;
 
     if (cur == NULL) return;
+    if ((cur->type == XML_DOCUMENT_NODE) ||
+        (cur->type == XML_HTML_DOCUMENT_NODE)) {
+        xmlDocContentDumpOutput(ctxt, (xmlDocPtr) cur);
+	return;
+    }
     if (cur->type == XML_XINCLUDE_START)
 	return;
     if (cur->type == XML_XINCLUDE_END)
Index: parser.c
===================================================================
RCS file: /cvs/gnome/gnome-xml/parser.c,v
retrieving revision 1.438
retrieving revision 1.439
diff -u -p -r1.438 -r1.439
--- libxml2/parser.c	4 Sep 2005 21:38:59 -0000	1.438
+++ libxml2/parser.c	9 Sep 2005 10:56:28 -0000	1.439
@@ -601,181 +601,181 @@ int
 xmlHasFeature(xmlFeature feature)
 {
     switch (feature) {
-	case XML_FEATURE_THREAD:
+	case XML_WITH_THREAD:
 #ifdef LIBXML_THREAD_ENABLED
 	    return(1);
 #else
 	    return(0);
 #endif
-        case XML_FEATURE_TREE:
+        case XML_WITH_TREE:
 #ifdef LIBXML_TREE_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_OUTPUT:
+        case XML_WITH_OUTPUT:
 #ifdef LIBXML_OUTPUT_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_PUSH:
+        case XML_WITH_PUSH:
 #ifdef LIBXML_PUSH_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_READER:
+        case XML_WITH_READER:
 #ifdef LIBXML_READER_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_PATTERN:
+        case XML_WITH_PATTERN:
 #ifdef LIBXML_PATTERN_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_WRITER:
+        case XML_WITH_WRITER:
 #ifdef LIBXML_WRITER_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_SAX1:
+        case XML_WITH_SAX1:
 #ifdef LIBXML_SAX1_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_FTP:
+        case XML_WITH_FTP:
 #ifdef LIBXML_FTP_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_HTTP:
+        case XML_WITH_HTTP:
 #ifdef LIBXML_HTTP_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_VALID:
+        case XML_WITH_VALID:
 #ifdef LIBXML_VALID_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_HTML:
+        case XML_WITH_HTML:
 #ifdef LIBXML_HTML_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_LEGACY:
+        case XML_WITH_LEGACY:
 #ifdef LIBXML_LEGACY_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_C14N:
+        case XML_WITH_C14N:
 #ifdef LIBXML_C14N_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_CATALOG:
+        case XML_WITH_CATALOG:
 #ifdef LIBXML_CATALOG_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_XPATH:
+        case XML_WITH_XPATH:
 #ifdef LIBXML_XPATH_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_XPTR:
+        case XML_WITH_XPTR:
 #ifdef LIBXML_XPTR_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_XINCLUDE:
+        case XML_WITH_XINCLUDE:
 #ifdef LIBXML_XINCLUDE_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_ICONV:
+        case XML_WITH_ICONV:
 #ifdef LIBXML_ICONV_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_ISO8859X:
+        case XML_WITH_ISO8859X:
 #ifdef LIBXML_ISO8859X_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_UNICODE:
+        case XML_WITH_UNICODE:
 #ifdef LIBXML_UNICODE_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_REGEXP:
+        case XML_WITH_REGEXP:
 #ifdef LIBXML_REGEXP_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_AUTOMATA:
+        case XML_WITH_AUTOMATA:
 #ifdef LIBXML_AUTOMATA_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_EXPR:
+        case XML_WITH_EXPR:
 #ifdef LIBXML_EXPR_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_SCHEMAS:
+        case XML_WITH_SCHEMAS:
 #ifdef LIBXML_SCHEMAS_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_SCHEMATRON:
+        case XML_WITH_SCHEMATRON:
 #ifdef LIBXML_SCHEMATRON_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_MODULES:
+        case XML_WITH_MODULES:
 #ifdef LIBXML_MODULES_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_DEBUG:
+        case XML_WITH_DEBUG:
 #ifdef LIBXML_DEBUG_ENABLED
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_DEBUG_MEM:
+        case XML_WITH_DEBUG_MEM:
 #ifdef DEBUG_MEMORY_LOCATION
             return(1);
 #else
             return(0);
 #endif
-        case XML_FEATURE_DEBUG_RUN:
+        case XML_WITH_DEBUG_RUN:
 #ifdef LIBXML_DEBUG_RUNTIME
             return(1);
 #else
@@ -9432,8 +9432,8 @@ xmlCheckCdataPush(const xmlChar *utf, in
 	    if (ix + 2 > len) return(ix);
 	    if ((utf[ix+1] & 0xc0 ) != 0x80)
 	        return(-ix);
-	    codepoint = (utf[0] & 0x1f) << 6;
-	    codepoint |= utf[1] & 0x3f;
+	    codepoint = (utf[ix] & 0x1f) << 6;
+	    codepoint |= utf[ix+1] & 0x3f;
 	    if (!xmlIsCharQ(codepoint))
 	        return(-ix);
 	    ix += 2;
@@ -9442,9 +9442,9 @@ xmlCheckCdataPush(const xmlChar *utf, in
 	    if (((utf[ix+1] & 0xc0) != 0x80) ||
 	        ((utf[ix+2] & 0xc0) != 0x80))
 		    return(-ix);
-	    codepoint = (utf[0] & 0xf) << 12;
-	    codepoint |= (utf[1] & 0x3f) << 6;
-	    codepoint |= utf[2] & 0x3f;
+	    codepoint = (utf[ix] & 0xf) << 12;
+	    codepoint |= (utf[ix+1] & 0x3f) << 6;
+	    codepoint |= utf[ix+2] & 0x3f;
 	    if (!xmlIsCharQ(codepoint))
 	        return(-ix);
 	    ix += 3;
@@ -9454,10 +9454,10 @@ xmlCheckCdataPush(const xmlChar *utf, in
 	        ((utf[ix+2] & 0xc0) != 0x80) ||
 		((utf[ix+3] & 0xc0) != 0x80))
 		    return(-ix);
-	    codepoint = (utf[0] & 0x7) << 18;
-	    codepoint |= (utf[1] & 0x3f) << 12;
-	    codepoint |= (utf[2] & 0x3f) << 6;
-	    codepoint |= utf[3] & 0x3f;
+	    codepoint = (utf[ix] & 0x7) << 18;
+	    codepoint |= (utf[ix+1] & 0x3f) << 12;
+	    codepoint |= (utf[ix+2] & 0x3f) << 6;
+	    codepoint |= utf[ix+3] & 0x3f;
 	    if (!xmlIsCharQ(codepoint))
 	        return(-ix);
 	    ix += 4;
Index: include/libxml/parser.h
===================================================================
RCS file: /cvs/gnome/gnome-xml/include/libxml/parser.h,v
retrieving revision 1.121
retrieving revision 1.122
diff -u -p -r1.121 -r1.122
--- libxml2/include/libxml/parser.h	4 Sep 2005 21:39:02 -0000	1.121
+++ libxml2/include/libxml/parser.h	12 Sep 2005 09:20:31 -0000	1.122
@@ -1171,39 +1171,40 @@ XMLPUBFUN xmlDocPtr XMLCALL
  *
  * Used to examine the existance of features that can be enabled
  * or disabled at compile-time.
+ * They used to be called XML_FEATURE_xxx but this clashed with Expat
  */
 typedef enum {
-    XML_FEATURE_THREAD = 1,
-    XML_FEATURE_TREE = 2,
-    XML_FEATURE_OUTPUT = 3,
-    XML_FEATURE_PUSH = 4,
-    XML_FEATURE_READER = 5,
-    XML_FEATURE_PATTERN = 6,
-    XML_FEATURE_WRITER = 7,
-    XML_FEATURE_SAX1 = 8,
-    XML_FEATURE_FTP = 9,
-    XML_FEATURE_HTTP = 10,
-    XML_FEATURE_VALID = 11,
-    XML_FEATURE_HTML = 12,
-    XML_FEATURE_LEGACY = 13,
-    XML_FEATURE_C14N = 14,
-    XML_FEATURE_CATALOG = 15,
-    XML_FEATURE_XPATH = 16,
-    XML_FEATURE_XPTR = 17,
-    XML_FEATURE_XINCLUDE = 18,
-    XML_FEATURE_ICONV = 19,
-    XML_FEATURE_ISO8859X = 20,
-    XML_FEATURE_UNICODE = 21,
-    XML_FEATURE_REGEXP = 22,
-    XML_FEATURE_AUTOMATA = 23,
-    XML_FEATURE_EXPR = 24,
-    XML_FEATURE_SCHEMAS = 25,
-    XML_FEATURE_SCHEMATRON = 26,
-    XML_FEATURE_MODULES = 27,
-    XML_FEATURE_DEBUG = 28,
-    XML_FEATURE_DEBUG_MEM = 29,
-    XML_FEATURE_DEBUG_RUN = 30,
-    XML_FEATURE_NONE = 99999 /* just to be sure of allocation size */
+    XML_WITH_THREAD = 1,
+    XML_WITH_TREE = 2,
+    XML_WITH_OUTPUT = 3,
+    XML_WITH_PUSH = 4,
+    XML_WITH_READER = 5,
+    XML_WITH_PATTERN = 6,
+    XML_WITH_WRITER = 7,
+    XML_WITH_SAX1 = 8,
+    XML_WITH_FTP = 9,
+    XML_WITH_HTTP = 10,
+    XML_WITH_VALID = 11,
+    XML_WITH_HTML = 12,
+    XML_WITH_LEGACY = 13,
+    XML_WITH_C14N = 14,
+    XML_WITH_CATALOG = 15,
+    XML_WITH_XPATH = 16,
+    XML_WITH_XPTR = 17,
+    XML_WITH_XINCLUDE = 18,
+    XML_WITH_ICONV = 19,
+    XML_WITH_ISO8859X = 20,
+    XML_WITH_UNICODE = 21,
+    XML_WITH_REGEXP = 22,
+    XML_WITH_AUTOMATA = 23,
+    XML_WITH_EXPR = 24,
+    XML_WITH_SCHEMAS = 25,
+    XML_WITH_SCHEMATRON = 26,
+    XML_WITH_MODULES = 27,
+    XML_WITH_DEBUG = 28,
+    XML_WITH_DEBUG_MEM = 29,
+    XML_WITH_DEBUG_RUN = 30,
+    XML_WITH_NONE = 99999 /* just to be sure of allocation size */
 } xmlFeature;
 
 XMLPUBFUN int XMLCALL
Index: xmllint.c
===================================================================
RCS file: /cvs/gnome/gnome-xml/xmllint.c,v
retrieving revision 1.159
retrieving revision 1.160
diff -u -p -r1.159 -r1.160
--- libxml2/xmllint.c	8 Sep 2005 21:39:47 -0000	1.159
+++ libxml2/xmllint.c	12 Sep 2005 09:20:28 -0000	1.160
@@ -2713,36 +2713,36 @@ static void parseAndPrintFile(char *file
 static void showVersion(const char *name) {
     fprintf(stderr, "%s: using libxml version %s\n", name, xmlParserVersion);
     fprintf(stderr, "   compiled with: ");
-    if (xmlHasFeature(XML_FEATURE_THREAD)) fprintf(stderr, "Threads ");
-    if (xmlHasFeature(XML_FEATURE_TREE)) fprintf(stderr, "Tree ");
-    if (xmlHasFeature(XML_FEATURE_OUTPUT)) fprintf(stderr, "Output ");
-    if (xmlHasFeature(XML_FEATURE_PUSH)) fprintf(stderr, "Push ");
-    if (xmlHasFeature(XML_FEATURE_READER)) fprintf(stderr, "Reader ");
-    if (xmlHasFeature(XML_FEATURE_PATTERN)) fprintf(stderr, "Patterns ");
-    if (xmlHasFeature(XML_FEATURE_WRITER)) fprintf(stderr, "Writer ");
-    if (xmlHasFeature(XML_FEATURE_SAX1)) fprintf(stderr, "SAXv1 ");
-    if (xmlHasFeature(XML_FEATURE_FTP)) fprintf(stderr, "FTP "); 
-    if (xmlHasFeature(XML_FEATURE_HTTP)) fprintf(stderr, "HTTP "); 
-    if (xmlHasFeature(XML_FEATURE_VALID)) fprintf(stderr, "DTDValid ");
-    if (xmlHasFeature(XML_FEATURE_HTML)) fprintf(stderr, "HTML "); 
-    if (xmlHasFeature(XML_FEATURE_LEGACY)) fprintf(stderr, "Legacy "); 
-    if (xmlHasFeature(XML_FEATURE_C14N)) fprintf(stderr, "C14N "); 
-    if (xmlHasFeature(XML_FEATURE_CATALOG)) fprintf(stderr, "Catalog "); 
-    if (xmlHasFeature(XML_FEATURE_XPATH)) fprintf(stderr, "XPath "); 
-    if (xmlHasFeature(XML_FEATURE_XPTR)) fprintf(stderr, "XPointer "); 
-    if (xmlHasFeature(XML_FEATURE_XINCLUDE)) fprintf(stderr, "XInclude "); 
-    if (xmlHasFeature(XML_FEATURE_ICONV)) fprintf(stderr, "Iconv "); 
-    if (xmlHasFeature(XML_FEATURE_ISO8859X)) fprintf(stderr, "ISO8859X "); 
-    if (xmlHasFeature(XML_FEATURE_UNICODE)) fprintf(stderr, "Unicode "); 
-    if (xmlHasFeature(XML_FEATURE_REGEXP)) fprintf(stderr, "Regexps "); 
-    if (xmlHasFeature(XML_FEATURE_AUTOMATA)) fprintf(stderr, "Automata "); 
-    if (xmlHasFeature(XML_FEATURE_EXPR)) fprintf(stderr, "Expr "); 
-    if (xmlHasFeature(XML_FEATURE_SCHEMAS)) fprintf(stderr, "Schemas "); 
-    if (xmlHasFeature(XML_FEATURE_SCHEMATRON)) fprintf(stderr, "Schematron "); 
-    if (xmlHasFeature(XML_FEATURE_MODULES)) fprintf(stderr, "Modules "); 
-    if (xmlHasFeature(XML_FEATURE_DEBUG)) fprintf(stderr, "Debug "); 
-    if (xmlHasFeature(XML_FEATURE_DEBUG_MEM)) fprintf(stderr, "MemDebug "); 
-    if (xmlHasFeature(XML_FEATURE_DEBUG_RUN)) fprintf(stderr, "RunDebug "); 
+    if (xmlHasFeature(XML_WITH_THREAD)) fprintf(stderr, "Threads ");
+    if (xmlHasFeature(XML_WITH_TREE)) fprintf(stderr, "Tree ");
+    if (xmlHasFeature(XML_WITH_OUTPUT)) fprintf(stderr, "Output ");
+    if (xmlHasFeature(XML_WITH_PUSH)) fprintf(stderr, "Push ");
+    if (xmlHasFeature(XML_WITH_READER)) fprintf(stderr, "Reader ");
+    if (xmlHasFeature(XML_WITH_PATTERN)) fprintf(stderr, "Patterns ");
+    if (xmlHasFeature(XML_WITH_WRITER)) fprintf(stderr, "Writer ");
+    if (xmlHasFeature(XML_WITH_SAX1)) fprintf(stderr, "SAXv1 ");
+    if (xmlHasFeature(XML_WITH_FTP)) fprintf(stderr, "FTP "); 
+    if (xmlHasFeature(XML_WITH_HTTP)) fprintf(stderr, "HTTP "); 
+    if (xmlHasFeature(XML_WITH_VALID)) fprintf(stderr, "DTDValid ");
+    if (xmlHasFeature(XML_WITH_HTML)) fprintf(stderr, "HTML "); 
+    if (xmlHasFeature(XML_WITH_LEGACY)) fprintf(stderr, "Legacy "); 
+    if (xmlHasFeature(XML_WITH_C14N)) fprintf(stderr, "C14N "); 
+    if (xmlHasFeature(XML_WITH_CATALOG)) fprintf(stderr, "Catalog "); 
+    if (xmlHasFeature(XML_WITH_XPATH)) fprintf(stderr, "XPath "); 
+    if (xmlHasFeature(XML_WITH_XPTR)) fprintf(stderr, "XPointer "); 
+    if (xmlHasFeature(XML_WITH_XINCLUDE)) fprintf(stderr, "XInclude "); 
+    if (xmlHasFeature(XML_WITH_ICONV)) fprintf(stderr, "Iconv "); 
+    if (xmlHasFeature(XML_WITH_ISO8859X)) fprintf(stderr, "ISO8859X "); 
+    if (xmlHasFeature(XML_WITH_UNICODE)) fprintf(stderr, "Unicode "); 
+    if (xmlHasFeature(XML_WITH_REGEXP)) fprintf(stderr, "Regexps "); 
+    if (xmlHasFeature(XML_WITH_AUTOMATA)) fprintf(stderr, "Automata "); 
+    if (xmlHasFeature(XML_WITH_EXPR)) fprintf(stderr, "Expr "); 
+    if (xmlHasFeature(XML_WITH_SCHEMAS)) fprintf(stderr, "Schemas "); 
+    if (xmlHasFeature(XML_WITH_SCHEMATRON)) fprintf(stderr, "Schematron "); 
+    if (xmlHasFeature(XML_WITH_MODULES)) fprintf(stderr, "Modules "); 
+    if (xmlHasFeature(XML_WITH_DEBUG)) fprintf(stderr, "Debug "); 
+    if (xmlHasFeature(XML_WITH_DEBUG_MEM)) fprintf(stderr, "MemDebug "); 
+    if (xmlHasFeature(XML_WITH_DEBUG_RUN)) fprintf(stderr, "RunDebug "); 
     fprintf(stderr, "\n");
 }
 
Index: testapi.c
===================================================================
RCS file: /cvs/gnome/gnome-xml/testapi.c,v
retrieving revision 1.63
retrieving revision 1.64
diff -u -p -r1.63 -r1.64
--- libxml2/testapi.c	4 Sep 2005 21:39:00 -0000	1.63
+++ libxml2/testapi.c	12 Sep 2005 09:20:28 -0000	1.64
@@ -1054,10 +1054,10 @@ static void des_xmlElementTypeVal(int no
 
 #define gen_nb_xmlFeature 4
 static xmlFeature gen_xmlFeature(int no, int nr ATTRIBUTE_UNUSED) {
-    if (no == 1) return(XML_FEATURE_AUTOMATA);
-    if (no == 2) return(XML_FEATURE_C14N);
-    if (no == 3) return(XML_FEATURE_CATALOG);
-    if (no == 4) return(XML_FEATURE_DEBUG);
+    if (no == 1) return(XML_WITH_AUTOMATA);
+    if (no == 2) return(XML_WITH_C14N);
+    if (no == 3) return(XML_WITH_CATALOG);
+    if (no == 4) return(XML_WITH_DEBUG);
     return(0);
 }
 
