Index: agent/agent_registry.c
===================================================================
RCS file: /cvsroot/net-snmp/net-snmp/agent/agent_registry.c,v
retrieving revision 5.32
diff -u -r5.32 agent_registry.c
--- agent/agent_registry.c	17 Nov 2005 22:03:02 -0000	5.32
+++ agent/agent_registry.c	20 Dec 2005 21:21:38 -0000
@@ -1015,7 +1015,6 @@
         if (child == NULL)      /* Didn't find the given name */
             break;
     }
-    netsnmp_subtree_free(myptr);
 
     memset(&reg_parms, 0x0, sizeof(reg_parms));
     reg_parms.name = name;
@@ -1028,6 +1027,7 @@
     snmp_call_callbacks(SNMP_CALLBACK_APPLICATION,
                         SNMPD_CALLBACK_UNREGISTER_OID, &reg_parms);
 
+    netsnmp_subtree_free(myptr);
     netsnmp_set_lookup_cache_size(old_lookup_cache_val);
     invalidate_lookup_cache(context);
     return MIB_UNREGISTERED_OK;
