--- src/output-plugins/spo_log_tcpdump.c	2004-09-13 19:44:49.000000000 +0200
+++ src/output-plugins/spo_log_tcpdump.c.oden	2005-03-11 19:27:52.052226865 +0100
@@ -296,8 +296,7 @@
     //strftime(timebuf,9,"%m%d@%H%M",loc_time);
 
     if(data->filename[0] == '/')
-        value = snprintf(logdir, STD_BUF-1, "%s.%lu", data->filename, 
-                (unsigned long)curr_time);
+        value = snprintf(logdir, STD_BUF-1, "%s", data->filename);
     else
         value = snprintf(logdir, STD_BUF-1, "%s/%s.%lu", pv.log_dir, 
                 data->filename, (unsigned long)curr_time);
