--- ./src/auto/configure.gcc31	2003-06-01 15:42:39.000000000 +0200
+++ ./src/auto/configure	2003-06-24 13:04:16.000000000 +0200
@@ -1189,7 +1191,14 @@
 if test -z "$have_local_include" -a -d /usr/local/include; then
   tt=`echo "$CPPFLAGS" | sed -e 's+-I/usr/local/include ++g' -e 's+-I/usr/local/include$++g'`
   if test "$tt" = "$CPPFLAGS"; then
-    CPPFLAGS="$CPPFLAGS -I/usr/local/include"
+    # Don't do this if gcc looks in /usr/local/include by default! gcc 3.1+ is rather
+    # unforgiving about these.
+    gccinc=`$CC -v -xc /dev/null -E 2>&1 >/dev/null \
+      | sed -n '/include.*search starts here/,/End of search/p' \
+      | grep -v '\(include.*search starts here\|End of search list\)'`
+    if ! echo $gccinc |grep -q /usr/local/include; then
+      CPPFLAGS="$CPPFLAGS -I/usr/local/include"
+    fi
   fi
 fi
 
--- ./src/configure.in.gcc31	2003-05-31 22:30:02.000000000 +0200
+++ ./src/configure.in	2003-06-24 13:03:25.000000000 +0200
@@ -91,7 +91,8 @@
   rm -f conftest.c conftest.o
 fi
 if test -z "$have_local_lib" -a -d /usr/local/lib; then
-  tt=`echo "$LDFLAGS" | sed -e 's+-L/usr/local/lib ++g' -e 's+-L/usr/local/lib$++g'`
+  tt=`echo "$CPPFLAGS" | sed -e 's+-I/usr/local/include ++g' -e \
+ 's+-I/usr/local/include$++g'`
   if test "$tt" = "$LDFLAGS"; then
     LDFLAGS="$LDFLAGS -L/usr/local/lib"
   fi
@@ -99,7 +100,15 @@
 if test -z "$have_local_include" -a -d /usr/local/include; then
   tt=`echo "$CPPFLAGS" | sed -e 's+-I/usr/local/include ++g' -e 's+-I/usr/local/include$++g'`
   if test "$tt" = "$CPPFLAGS"; then
-    CPPFLAGS="$CPPFLAGS -I/usr/local/include"
+    # Don't do this if gcc looks in /usr/local/include by default! gcc 3.1+ is
+    # rather
+    # unforgiving about these.
+    gccinc=`$CC -v -xc /dev/null -E 2>&1 >/dev/null \
+      | sed -n '/include.*search starts here/,/End of search/p' \
+      | grep -v '\(include.*search starts here\|End of search list\)'`
+    if ! echo $gccinc |grep -q /usr/local/include; then
+      CPPFLAGS="$CPPFLAGS -I/usr/local/include"
+    fi
   fi
 fi
 
