--- ./vimrc_example.vim.tv2	2002-10-31 10:06:39.000000000 -0500
+++ ./runtime/vimrc_example.vim	2002-11-01 16:49:49.000000000 -0500
@@ -18,6 +18,12 @@
 " This must be first, because it changes other options as a side effect.
 set nocompatible
 
+
+" Make shift-insert work like in Xterm
+map <S-Insert> <MiddleMouse>
+map! <S-Insert> <MiddleMouse>
+
+
 " allow backspacing over everything in insert mode
 set backspace=indent,eol,start
 
