
Fix for #118020.

Upstream-HEAD: committed
Upstream-2.0: needed
Upstream-PR: 27731
Upstream-Status: Medium risk since nobody understands this code upstream

--- httpd-2.0.48/server/vhost.c.vhost
+++ httpd-2.0.48/server/vhost.c
@@ -543,12 +543,6 @@
                          ic->sar->virthost, ic->sar->host_port);
             *pic = ic->next;
         }
-        else if (ic->names == NULL) {
-            /* if server != NULL and names == NULL then we're done
-             * looking at NameVirtualHosts
-             */
-            break;
-        }
         else {
             pic = &ic->next;
         }
