diff -uNr -U 5 pam_ldap-180/pam_ldap.c pam_ldap-181/pam_ldap.c
--- pam_ldap-180/pam_ldap.c	2005-08-17 19:35:13.000000000 -0300
+++ pam_ldap-181/pam_ldap.c	2006-04-30 13:41:10.000000000 -0300
@@ -3817,38 +3821,38 @@
   switch (session->info->policy_error)
     {
     case POLICY_ERROR_SUCCESS:
       break;
     case POLICY_ERROR_PASSWORD_EXPIRED:
+    case POLICY_ERROR_CHANGE_AFTER_RESET:
       _conv_sendmsg (appconv,
 		     "You are required to change your LDAP password immediately.",
 		     PAM_ERROR_MSG, no_warn);
 #ifdef LINUX
       rc = success = PAM_AUTHTOKEN_REQD;
 #else
       rc = success = PAM_NEW_AUTHTOK_REQD;
 #endif /* LINUX */
       break;
     case POLICY_ERROR_ACCOUNT_LOCKED:
-    case POLICY_ERROR_CHANGE_AFTER_RESET:
     case POLICY_ERROR_PASSWORD_MOD_NOT_ALLOWED:
     case POLICY_ERROR_MUST_SUPPLY_OLD_PASSWORD:
     case POLICY_ERROR_INSUFFICIENT_PASSWORD_QUALITY:
     case POLICY_ERROR_PASSWORD_TOO_SHORT:
     case POLICY_ERROR_PASSWORD_TOO_YOUNG:
     case POLICY_ERROR_PASSWORD_INSUFFICIENT:
       _conv_sendmsg (appconv,
 		     policy_error_table[session->info->policy_error],
 		     PAM_ERROR_MSG, no_warn);
-      rc = success = PAM_PERM_DENIED;
+      return PAM_PERM_DENIED;
       break;
     default:
       snprintf (buf, sizeof buf,
 		"Unknown password policy error %d received.",
 		session->info->policy_error);
       _conv_sendmsg (appconv, buf, PAM_ERROR_MSG, no_warn);
-      rc = success = PAM_PERM_DENIED;
+      return PAM_PERM_DENIED;
       break;
     }
 
   /*
    * Warnings.  First, check if we've got a non-zero warning time
