diff -urN icewm-1.2.10pre11/src/default.h icewm-1.2.10pre11-modified/src/default.h
--- icewm-1.2.10pre11/src/default.h	2003-08-10 11:39:40.000000000 +0200
+++ icewm-1.2.10pre11-modified/src/default.h	2003-08-20 18:08:54.000000000 +0200
@@ -186,24 +186,24 @@
 
 XSV(const char *, kdeDataDir,                   KDEDIR)
 XSV(const char *, mailBoxPath,                  0)
-XSV(const char *, mailCommand,                  "xterm -name pine -title PINE -e pine")
-XSV(const char *, mailClassHint,                "pine.XTerm")
+XSV(const char *, mailCommand,                  "xvt -name mutt -title PINE -e mutt")
+XSV(const char *, mailClassHint,                "mutt.XTerm")
 XSV(const char *, newMailCommand,               0)
 XSV(const char *, lockCommand,                  "xlock")
 XSV(const char *, clockCommand,                 "xclock -name icewm -title Clock")
 XSV(const char *, clockClassHint,               "icewm.XClock")
 XSV(const char *, runDlgCommand,                0)
 XSV(const char *, openCommand,                  0)
-XSV(const char *, terminalCommand,              "xterm")
+XSV(const char *, terminalCommand,              "xvt")
 XSV(const char *, logoutCommand,                0)
 XSV(const char *, logoutCancelCommand,          0)
 XSV(const char *, shutdownCommand,              "shutdown -h now")
 XSV(const char *, rebootCommand,                "shutdown -r now")
-XSV(const char *, cpuCommand,                   "xterm -name top -title Process\\ Status -e top")
+XSV(const char *, cpuCommand,                   "xvt -name top -title Process\\ Status -e top")
 XSV(const char *, cpuClassHint,                 "top.XTerm")
-XSV(const char *, netCommand,                   "xterm -name netstat -title 'Network Status' -e netstat -c")
+XSV(const char *, netCommand,                   "xvt -name netstat -title 'Network Status' -e netstat -c")
 XSV(const char *, netClassHint,                 "netstat.XTerm")
-XSV(const char *, netDevice,                    "ppp0 eth0")
+XSV(const char *, netDevice,                    "eth1 eth0 ppp0")
 XSV(const char *, addressBarCommand,            0)
 #ifdef CONFIG_I18N
 XSV(const char *, fmtTime,                      "%X")
diff -urN icewm-1.2.10pre11/src/yprefs.h icewm-1.2.10pre11-modified/src/yprefs.h
--- icewm-1.2.10pre11/src/yprefs.h	2003-08-10 11:39:40.000000000 +0200
+++ icewm-1.2.10pre11-modified/src/yprefs.h	2003-08-20 18:00:45.000000000 +0200
@@ -30,7 +30,7 @@
 #ifdef CONFIG_XFREETYPE
 XIV(bool, haveXft,                              true)
 #endif
-XSV(const char *, iconPath,                     0)
+XSV(const char *, iconPath,                     "/usr/share/icons:/usr/share/icons/mini:/usr/share/icons/locolr/16x16/apps/:/usr/share/icons/hicolor/16x16/apps/:/usr/X11R6/lib/X11/icewm/icons")
 XSV(const char *, libDir,                       LIBDIR)
 XSV(const char *, configDir,                    CFGDIR)
 #define CONFIG_DEFAULT_THEME "icedesert/default.theme"
