--- mod_auth_mysql.c.orig	2003-02-22 22:58:57.000000000 -0500
+++ mod_auth_mysql.c	2003-02-22 23:00:05.000000000 -0500
@@ -268,7 +268,8 @@
     db_host = m->mysqlhost;
   }
 
-  mysql_handle=mysql_connect(&mysql_conn,db_host,m->mysqluser,m->mysqlpasswd);
+  mysql_init(&mysql_conn);
+  mysql_handle=mysql_real_connect(&mysql_conn,db_host,m->mysqluser,m->mysqlpasswd,NULL, 0, NULL, 0);
 
   if (mysql_handle) {
 
