--- make-3.80/configure.no-hires-timestamp	2002-10-03 20:18:48.000000000 +0200
+++ make-3.80/configure	2003-07-25 13:35:36.000000000 +0200
@@ -6724,7 +6724,7 @@ if { (eval echo "$as_me:$LINENO: \"$ac_c
   ac_status=$?
   echo "$as_me:$LINENO: \$? = $ac_status" >&5
   (exit $ac_status); }; }; then
-  make_cv_file_timestamp_hi_res=yes
+  make_cv_file_timestamp_hi_res=no
 else
   echo "$as_me: failed program was:" >&5
 cat conftest.$ac_ext >&5
