--- autofs-4.0.0/modules/lookup_multi.c.multiarg	Wed Mar 28 07:08:23 2001
+++ autofs-4.0.0/modules/lookup_multi.c	Fri Oct  5 15:41:29 2001
@@ -53,6 +53,7 @@
 
   if ( !(*context = ctxt = malloc(sizeof(struct lookup_context))) )
     goto nomem;
+	memset(ctxt, 0, sizeof(struct lookup_context));
 
   if ( argc < 1 ) {
     syslog(LOG_CRIT, MODPREFIX "No map list");
@@ -65,22 +66,22 @@
   }
 
   if ( !(ctxt->m = malloc(ctxt->n*sizeof(struct module_info))) ||
-       !(ctxt->argl = malloc((argc+1)*sizeof(const char **))) )
+       !(ctxt->argl = malloc((argc+1)*sizeof(const char *))) )
     goto nomem;
+  memset(ctxt->m, 0, ctxt->n*sizeof(struct module_info));
 
-  memcpy(ctxt->argl, argv, (argc+1)*sizeof(const char **));
+  memcpy(ctxt->argl, argv, (argc+1)*sizeof(const char *));
 
-  for ( i = j = an = 0 ; argv[an] ; an++ ) {
-    if ( j == 0 ) {
+  for ( i = j = an = 0 ; ctxt->argl[an] ; an++ ) {
+    if ( ctxt->m[i].argc == 0 ) {
       ctxt->m[i].argv = &ctxt->argl[an];
       ctxt->m[i].argc = 0;
     }
     if ( !strcmp(ctxt->argl[an], "--") ) {
       ctxt->argl[an] = NULL;
-      i++; j = 0;
+      i++;
     } else {
       ctxt->m[i].argc++;
-      j++;
     }
   }
 
@@ -98,7 +99,8 @@
     
     if ( !(ctxt->m[i].mod = open_lookup(map, MODPREFIX,
 					mapfmt ? mapfmt : my_mapfmt,
-					ctxt->m[i].argc-1, ctxt->m[i].argv+1)) )
+					ctxt->m[i].argc - 1,
+					ctxt->m[i].argv + 1)) )
       return 1;
   }
 
