--- mpage.c.orig	2003-03-10 15:41:02.000000000 +0100
+++ mpage.c	2003-03-10 15:41:58.000000000 +0100
@@ -44,7 +44,8 @@
     struct sheet *thesheet;
     char outcommand[LINESIZE]; /* the command which is the output filter */
 
-    current_locale = setlocale(LC_ALL,"");
+    setlocale(LC_ALL,"");
+    current_locale = setlocale(LC_ALL,NULL);
 
    
 #ifdef __EMX__
--- text.c.orig	2003-03-10 15:43:10.000000000 +0100
+++ text.c	2003-03-10 15:44:10.000000000 +0100
@@ -400,7 +400,7 @@
      */
     while (gathering && textp - text < LINESIZE - 5) {
 #ifdef KANJI
-        if (current_locale && !strncmp(current_locale,"ja_JP",5)) {
+        if (current_locale && !strncmp(current_locale,"ja",2)) {
             if (prevkanj) {
                 *kp = prevkanj;
                 if (prevkanj >= 0x0100)
