--- ./src/touch.c.FBopts	2003-08-09 19:49:23.000000000 +0200
+++ ./src/touch.c	2003-12-10 15:49:11.000000000 +0100
@@ -75,6 +75,9 @@
 /* New time to use when setting time. */
 static struct timespec newtime;
 
+/* Time modifier, allowing to "make file x 10 seconds older than file y" */
+static int time_modifier = 0;
+
 /* File to use for -r. */
 static char *ref_file;
 
@@ -94,6 +97,8 @@
   {"no-create", no_argument, 0, 'c'},
   {"date", required_argument, 0, 'd'},
   {"file", required_argument, 0, 'r'}, /* FIXME: phase out --file */
+  {"forward", required_argument, 0, 'F'},
+  {"backward", required_argument, 0, 'B'},
   {"reference", required_argument, 0, 'r'},
   {GETOPT_HELP_OPTION_DECL},
   {GETOPT_VERSION_OPTION_DECL},
@@ -168,9 +173,19 @@
 
   if (amtime_now)
     {
-      /* Pass NULL to utime so it will not fail if we just have
-	 write access to the file, but don't own it.  */
-      status = utime (file, NULL);
+      if (time_modifier == 0)
+	{
+          /* Pass NULL to utime so it will not fail if we just have
+	  write access to the file, but don't own it.  */
+          status = utime (file, NULL);
+	}
+      else
+	{
+	  struct utimbuf t;
+	  t.actime = time(NULL) + time_modifier;
+	  t.modtime = time(NULL) + time_modifier;
+	  status = utime (file, &t);
+	}
     }
   else
     {
@@ -182,9 +197,9 @@
 
       if (use_ref)
 	{
-	  timespec[0].tv_sec = ref_stats.st_atime;
+	  timespec[0].tv_sec = ref_stats.st_atime + time_modifier;
 	  timespec[0].tv_nsec = TIMESPEC_NS (ref_stats.st_atim);
-	  timespec[1].tv_sec = ref_stats.st_mtime;
+	  timespec[1].tv_sec = ref_stats.st_mtime + time_modifier;
 	  timespec[1].tv_nsec = TIMESPEC_NS (ref_stats.st_mtim);
 	}
       else
@@ -192,13 +207,13 @@
 
       if (!(change_times & CH_ATIME))
 	{
-	  timespec[0].tv_sec = sbuf.st_atime;
+	  timespec[0].tv_sec = sbuf.st_atime + time_modifier;
 	  timespec[0].tv_nsec = TIMESPEC_NS (sbuf.st_atim);
 	}
 
       if (!(change_times & CH_MTIME))
 	{
-	  timespec[1].tv_sec = sbuf.st_mtime;
+	  timespec[1].tv_sec = sbuf.st_mtime + time_modifier;
 	  timespec[1].tv_nsec = TIMESPEC_NS (sbuf.st_mtim);
 	}
 
@@ -245,8 +260,10 @@
 "), stdout);
       fputs (_("\
   -a                     change only the access time\n\
+  -B SEC, --backward=SEC date back SEC seconds\n\
   -c, --no-create        do not create any files\n\
   -d, --date=STRING      parse STRING and use it instead of current time\n\
+  -F SEC, --forward=SEC  date forward SEC seconds\n\
   -f                     (ignored)\n\
   -m                     change only the modification time\n\
 "), stdout);
@@ -284,13 +301,21 @@
 
   change_times = no_create = use_ref = posix_date = flexible_date = 0;
 
-  while ((c = getopt_long (argc, argv, "acd:fmr:t:", longopts, NULL)) != -1)
+  while ((c = getopt_long (argc, argv, "B:F:acd:fmr:t:", longopts, NULL)) != -1)
     {
       switch (c)
 	{
 	case 0:
 	  break;
 
+	case 'B':
+	  time_modifier = -atoi(optarg);
+	  break;
+
+	case 'F':
+	  time_modifier = atoi(optarg);
+	  break;
+
 	case 'a':
 	  change_times |= CH_ATIME;
 	  break;
--- ./doc/coreutils.texi.FBopts	2003-12-10 15:47:03.000000000 +0100
+++ ./doc/coreutils.texi	2003-12-10 15:47:03.000000000 +0100
@@ -7891,6 +7891,14 @@
 @opindex use @r{time, changing}
 Change the access time only.
 
+@item -B
+@itemx --backward=seconds
+@opindex -B
+@opindex --backward
+Modify the time by going back @var{seconds} seconds.
+For example, @code{touch -r foo -B 5 bar} will make the file bar 5 seconds
+older than file foo.
+
 @item -c
 @itemx --no-create
 @opindex -c
@@ -7905,6 +7913,14 @@
 Use @var{time} instead of the current time.  It can contain month names,
 time zones, @samp{am} and @samp{pm}, etc.  @xref{Date input formats}.
 
+@item -F
+@itemx --forward=seconds
+@opindex -F
+@opindex --forward
+Modify the time by going forward @var{seconds} seconds.
+For example, @code{touch -r foo -F 5 bar} will make the file bar 5 seconds
+newer than file foo.
+
 @item -f
 @opindex -f
 @cindex BSD @command{touch} compatibility
--- ./doc/coreutils.info.FBopts	2003-12-10 15:47:03.000000000 +0100
+++ ./doc/coreutils.info	2003-12-10 15:47:03.000000000 +0100
@@ -6338,6 +6338,12 @@
 `--time=use'
      Change the access time only.
 
+`-B'
+`--backward=seconds'
+     Modify the time by going back `seconds' seconds.
+     For example, `touch -r foo -B 5 bar' will make the file bar 5 seconds
+     older than file foo.
+
 `-c'
 `--no-create'
      Do not create files that do not exist.
@@ -6347,6 +6353,12 @@
      Use TIME instead of the current time.  It can contain month names,
      time zones, `am' and `pm', etc.  *Note Date input formats::.
 
+`-F'
+'--forward=seconds'
+     Modify the time by going forward `seconds' seconds.
+     For example, `touch -r foo -F 5 bar' will make the file bar 5 seconds
+     newer than file foo.
+
 `-f'
      Ignored; for compatibility with BSD versions of `touch'.
 
--- ./man/touch.1.FBopts	2003-09-02 22:02:36.000000000 +0200
+++ ./man/touch.1	2003-12-10 15:47:03.000000000 +0100
@@ -15,12 +15,22 @@
 \fB\-a\fR
 change only the access time
 .TP
+\fB\-B\fR, \fB\-\-backward\fR=\fISECONDS\fR
+Modify the time by going back SECONDS seconds.
+For example, \fBtouch -r foo -B 5 bar\fR will make the file bar 5 seconds
+older than file foo.
+.TP
 \fB\-c\fR, \fB\-\-no\-create\fR
 do not create any files
 .TP
 \fB\-d\fR, \fB\-\-date\fR=\fISTRING\fR
 parse STRING and use it instead of current time
 .TP
+\fB\-F\fR, \fB\-\-forward\fR=\fISECONDS\fR
+Modify the time by going forward SECONDS seconds.
+For example, \fBtouch -r foo -F 5 bar\fR will make the file bar 5 seconds
+newer than file foo.
+.TP
 \fB\-f\fR
 (ignored)
 .TP
