--- util-linux-2.11b/fdisk/sfdisk.c.390	Wed May  9 20:50:30 2001
+++ util-linux-2.11b/fdisk/sfdisk.c	Wed May  9 20:50:59 2001
@@ -131,7 +131,7 @@
  *
  * Note: we use 512-byte sectors here, irrespective of the hardware ss.
  */
-#if !defined (__alpha__) && !defined (__ia64__)
+#if !defined (__alpha__) && !defined (__ia64__) && !defined(__s390x__)
 static
 _syscall5(int,  _llseek,  uint,  fd, ulong, hi, ulong, lo,
        loff_t *, res, uint, wh);
@@ -143,7 +143,7 @@
     in = ((loff_t) s << 9);
     out = 1;
 
-#if !defined (__alpha__) && !defined (__ia64__)
+#if !defined (__alpha__) && !defined (__ia64__) && !defined(__s390x__)
     if (_llseek (fd, in>>32, in & 0xffffffff, &out, SEEK_SET) != 0) {
 #else
     if ((out = lseek(fd, in, SEEK_SET)) != in) {
