--- util-linux-ng-2.13-rc3/mount/mount.c.set-as-encrypted	2007-08-25 14:10:04.000000000 +0200
+++ util-linux-ng-2.13-rc3/mount/mount.c	2007-08-25 14:11:21.000000000 +0200
@@ -107,11 +107,12 @@ struct opt_map {
 #define MS_OWNER	0x10000000
 #define MS_GROUP	0x08000000
 #define MS_PAMCONSOLE	0x04000000
+#define MS_ENCRYPT	0x00030000
 #define MS_COMMENT	0x02000000
 #define MS_LOOP		0x00010000
 
 /* Options that we keep the mount system call from seeing.  */
-#define MS_NOSYS	(MS_NOAUTO|MS_USERS|MS_USER|MS_COMMENT|MS_LOOP|MS_PAMCONSOLE)
+#define MS_NOSYS	(MS_NOAUTO|MS_USERS|MS_USER|MS_COMMENT|MS_ENCRYPT|MS_LOOP|MS_PAMCONSOLE)
 
 /* Options that we keep from appearing in the options field in the mtab.  */
 #define MS_NOMTAB	(MS_REMOUNT|MS_NOAUTO|MS_USERS|MS_USER|MS_PAMCONSOLE)
@@ -150,6 +151,7 @@ static const struct opt_map opt_map[] = 
   { "noowner",	0, 1, MS_OWNER  },	/* Device owner has no special privs */
   { "group",	0, 0, MS_GROUP  },	/* Let the group of the device mount */
   { "nogroup",	0, 1, MS_GROUP  },	/* Device group has no special privs */
+  { "encrypted", 0, 0, MS_ENCRYPT },	/* Device encrypted */
   { "_netdev",	0, 0, MS_COMMENT},	/* Device requires network */
   { "comment",	0, 0, MS_COMMENT},	/* fstab comment only (kudzu,_netdev)*/
 
