--- e2fsprogs-1.36/lib/ss/mk_cmds.sh.in.strip-me	2004-09-17 19:54:21.000000000 -0400
+++ e2fsprogs-1.36/lib/ss/mk_cmds.sh.in	2005-02-28 08:05:30.627684418 -0500
@@ -3,7 +3,9 @@
 #
 
 DIR="${DIR-@datadir@/ss}"
+# BUILD start
 SS_DIR="@SS_DIR@"
+# BUILD end
 AWK=@AWK@
 SED=@SED@
 
@@ -17,12 +19,16 @@ if test -n "$_SS_DIR_OVERRIDE" ; then
 fi
 
 if test ! -f $DIR/ct_c.sed || test ! -f $DIR/ct_c.awk ; then
+# BUILD start
     DIR="$SS_DIR"
 #    echo "Falling back to $DIR..."
     if test ! -f "$DIR/ct_c.sed" || test ! -f "$DIR/ct_c.awk" ; then
+# BUILD end
 	echo "mk_cmds: Couldn't find mk_cmds's template files."
 	exit 1
+# BUILD start
     fi
+# BUILD end
 fi
 
 FILE="$1"
--- e2fsprogs-1.36/lib/ss/Makefile.in.strip-me	2005-02-05 15:48:48.000000000 -0500
+++ e2fsprogs-1.36/lib/ss/Makefile.in	2005-02-28 08:12:43.461996408 -0500
@@ -120,6 +120,10 @@ mk_cmds: $(DEP_SUBSTITUTE) $(srcdir)/mk_
 	@$(SUBSTITUTE) $(srcdir)/mk_cmds.sh.in mk_cmds
 	@$(CHMOD) +x mk_cmds
 
+mk_cmds.install: mk_cmds
+	@perl -ne '/^# BUILD start/ ... /^# BUILD end/ or print' $< > $@
+	@$(CHMOD) +x $@
+
 ss.pc: $(srcdir)/ss.pc.in $(top_builddir)/config.status
 	@echo "	CONFIG.STATUS $@"
 	@cd $(top_builddir); CONFIG_FILES=lib/ss/ss.pc ./config.status
@@ -130,7 +134,7 @@ installdirs::
 		$(DESTDIR)$(includedir)/ss $(DESTDIR)$(datadir)/ss \
 		$(DESTDIR)$(bindir) $(DESTDIR)$(libdir)/pkgconfig
 
-install:: libss.a $(INSTALL_HFILES) installdirs ss_err.h mk_cmds ss.pc
+install:: libss.a $(INSTALL_HFILES) installdirs ss_err.h mk_cmds ss.pc mk_cmds.install
 	@echo "	INSTALL_DATA $(DESTDIR)$(libdir)/libss.a"
 	@$(INSTALL_DATA) libss.a $(DESTDIR)$(libdir)/libss.a
 	@-$(RANLIB) $(DESTDIR)$(libdir)/libss.a
@@ -147,7 +151,7 @@ install:: libss.a $(INSTALL_HFILES) inst
 		$(INSTALL_DATA) $(srcdir)/$$i $(DESTDIR)$(datadir)/ss/$$i; \
 	done
 	@echo "	INSTALL $(bindir)/mk_cmds"
-	@$(INSTALL) mk_cmds $(DESTDIR)$(bindir)/mk_cmds
+	@$(INSTALL) mk_cmds.install $(DESTDIR)$(bindir)/mk_cmds
 	@echo "	INSTALL_DATA $(man1dir)/mk_cmds.1"
 	@$(INSTALL_DATA) $(srcdir)/mk_cmds.1 $(DESTDIR)$(man1dir)/mk_cmds.1
 	@echo "	INSTALL_DATA $(libdir)/pkgconfig/ss.pc"
--- e2fsprogs-1.36/lib/et/compile_et.sh.in.strip-me	2004-09-17 19:54:21.000000000 -0400
+++ e2fsprogs-1.36/lib/et/compile_et.sh.in	2005-02-28 08:04:15.009430590 -0500
@@ -3,12 +3,14 @@
 #
 AWK=@AWK@
 DIR="${DIR-@datadir@/et}"
+# BUILD start
 ET_DIR="@ET_DIR@"
 
 if test "$1" = "--build-tree" ; then
     shift;
     DIR="$ET_DIR"
 fi
+# BUILD end
 
 if test "x$1" = x ; then
     echo "Usage: compile_et file"
@@ -16,12 +18,16 @@ if test "x$1" = x ; then
 fi
 
 if test ! -f "$DIR/et_h.awk" || test ! -f "$DIR/et_c.awk" ; then
+# BUILD start
     DIR="$ET_DIR"
 #    echo "Falling back to $DIR..."
     if test ! -f "$DIR/et_h.awk" || test ! -f "$DIR/et_c.awk" ; then
+# BUILD end
 	echo "compile_et: Couldn't find compile_et's template files."
 	exit 1
+# BUILD start
     fi
+# BUILD end
 fi
 
 ROOT=`echo $1 | sed -e s/.et$//`
--- e2fsprogs-1.36/lib/et/Makefile.in.strip-me	2005-02-05 15:48:48.000000000 -0500
+++ e2fsprogs-1.36/lib/et/Makefile.in	2005-02-28 08:12:59.244180014 -0500
@@ -59,6 +59,10 @@ compile_et: $(DEP_SUBSTITUTE) $(srcdir)/
 	@$(SUBSTITUTE) $(srcdir)/compile_et.sh.in compile_et
 	@$(CHMOD) +x compile_et
 
+compile_et.install: compile_et
+	@perl -ne '/^# BUILD start/ ... /^# BUILD end/ or print' $< > $@
+	@$(CHMOD) +x $@
+
 com_err.ps : com_err.dvi
 com_err.dvi: com_err.texinfo
 
@@ -80,7 +84,7 @@ installdirs::
 		$(DESTDIR)$(bindir) $(DESTDIR)$(man1dir) \
 		$(DESTDIR)$(man3dir) $(DESTDIR)$(libdir)/pkgconfig
 
-install:: compile_et libcom_err.a $(HFILES) installdirs com_err.pc
+install:: compile_et libcom_err.a $(HFILES) installdirs com_err.pc compile_et.install
 	@echo "	INSTALL_DATA $(libdir)/libcom_err.a"
 	@$(INSTALL_DATA) libcom_err.a $(DESTDIR)$(libdir)/libcom_err.a
 	@-$(RANLIB) $(DESTDIR)$(libdir)/libcom_err.a
@@ -94,7 +98,7 @@ install:: compile_et libcom_err.a $(HFIL
 		$(INSTALL_DATA) $(srcdir)/$$i $(DESTDIR)$(datadir)/et/$$i; \
 	done
 	@echo "	INSTALL_SCRIPT $(bindir)/compile_et"
-	@$(INSTALL_SCRIPT) compile_et $(DESTDIR)$(bindir)/compile_et
+	@$(INSTALL_SCRIPT) compile_et.install $(DESTDIR)$(bindir)/compile_et
 	@echo "	INSTALL_DATA $(man3dir)/com_err.3"
 	@$(INSTALL_DATA) $(srcdir)/com_err.3 $(DESTDIR)$(man3dir)/com_err.3
 	@echo "	INSTALL_DATA $(man1dir)/compile_et.1"
