--- libtool-1.5.6/ltmain.in.old-libtool	2004-06-21 17:36:50.232711104 +0800
+++ libtool-1.5.6/ltmain.in	2004-06-21 18:52:42.100721984 +0800
@@ -242,8 +242,18 @@
 #####################################
 
 # Darwin sucks
+#eval std_shrext=\"$shrext_cmds\"
+
+# And fixing for Darwin sucks for everybody else
+if test -z "$shrext_cmds" && test -n "$shrext"; then
+	eval shrext_cmds=\"$shrext\"
+fi
 eval std_shrext=\"$shrext_cmds\"
 
+# This value is evaluated to 32768, so place it here as a compatilibity hack
+# because older libtool.m4 didn't define this variable
+test -z "$max_cmd_len" && max_cmd_len=32768
+
 # Parse our command line options once, thoroughly.
 while test "$#" -gt 0
 do
