--- mkbootdisk-1.4.5/mkbootdisk.orig	Fri Jul 26 14:47:15 2002
+++ mkbootdisk-1.4.5/mkbootdisk	Fri Jul 26 14:48:07 2002
@@ -30,6 +30,21 @@
     exit $1
 }
 
+clean() {
+	umount $MOUNTDIR
+	rmdir $MOUNTDIR
+	rm $TMPINITRD
+}
+
+detect_error() {
+    # Detect "no space left on device" errors
+    if [ "$?" != 0 ]; then
+	echo "Error !"
+	clean
+	exit 1
+    fi
+}
+
 while [ $# -gt 0 ]; do
     case $1 in
 	--device)
@@ -187,10 +202,16 @@
 
 [ -n "$verbose" ] && echo -n "Copying /boot/vmlinuz-$kernel... "
 cp -p /boot/vmlinuz-$kernel $MOUNTDIR/vmlinuz
+detect_error
 [ -n "$verbose" ] && echo "done."
 
 [ -n "$verbose" ] && echo -n "Creating initrd image... "
-/sbin/mkinitrd $mkinitrdargs $witheth --ifneeded $MOUNTDIR/initrd.img $kernel
+TMPINITRD=`mktemp /tmp/mkbootdisk-initrd.XXXXXX`
+/sbin/mkinitrd -f $mkinitrdargs $witheth --ifneeded $TMPINITRD $kernel
+if [ -f $TMPINITRD ]; then
+    cp $TMPINITRD $MOUNTDIR/initrd.img
+    detect_error
+fi
 [ -n "$verbose" ] && echo "done."
 
 [ -n "$verbose" ] && echo -n "Setting up syslinux... "
@@ -210,6 +231,7 @@
 	kernel vmlinuz
 	append $INITRDARG root=$rootdev $appendargs
 EOF
+detect_error
 
 chmod 644 $MOUNTDIR/syslinux.cfg
 
@@ -220,8 +242,8 @@
 "linux <params>", followed by <return> if you like.
 
 EOF
+detect_error
 
 [ -n "$verbose" ] && echo "done."
 
-umount $MOUNTDIR
-rmdir $MOUNTDIR
+clean
