diff -Naur bind-9.4.1-P1/lib/isc/unix/socket.c bind-9.4.1-P1.new/lib/isc/unix/socket.c
--- bind-9.4.1-P1/lib/isc/unix/socket.c	2006-06-06 08:56:09.000000000 +0800
+++ bind-9.4.1-P1.new/lib/isc/unix/socket.c	2007-09-04 22:31:16.339152000 +0800
@@ -1492,7 +1492,7 @@
 		return (ISC_R_UNEXPECTED);
 	}
 
-#ifdef SO_BSDCOMPAT
+#if 0
 	if (type != isc_sockettype_unix &&
 	    setsockopt(sock->fd, SOL_SOCKET, SO_BSDCOMPAT,
 		       (void *)&on, sizeof(on)) < 0) {
