--- bind-9.3.1rc1/lib/bind/irs/irs_data.c.fix_h_errno	2004-11-29 20:15:43.000000000 -0500
+++ bind-9.3.1rc1/lib/bind/irs/irs_data.c	2005-03-09 21:05:52.000000000 -0500
@@ -222,12 +222,7 @@
 void
 __h_errno_set(struct __res_state *res, int err) {
 
-
-#if (__GLIBC__ > 2 || __GLIBC__ == 2 &&  __GLIBC_MINOR__ >= 3)
-	res->res_h_errno = err;
-#else
 	h_errno = res->res_h_errno = err;
-#endif
 }
 
 #endif /*__BIND_NOSTATIC*/
--- bind-9.3.1rc1/lib/bind/resolv/res_query.c.fix_h_errno	2004-03-16 07:34:19.000000000 -0500
+++ bind-9.3.1rc1/lib/bind/resolv/res_query.c	2005-03-09 21:53:34.000000000 -0500
@@ -192,6 +192,8 @@
 		}
 		return (-1);
 	}
+	if( n > 0 )
+		RES_SET_H_ERRNO(statp,0);
 	return (n);
 }
 
