--- mtools-3.9.8/mformat.c.fpons	2002-11-22 17:32:59.000000000 +0100
+++ mtools-3.9.8/mformat.c	2002-11-22 17:35:35.000000000 +0100
@@ -28,6 +28,7 @@
 
 #define _LINUX_STRING_H_
 #define kdev_t int
+#include "asm/atomic.h"
 #include "linux/fs.h"
 #undef _LINUX_STRING_H_
 
--- mtools-3.9.8/mpartition.c.fpons	2002-11-22 17:36:34.000000000 +0100
+++ mtools-3.9.8/mpartition.c	2002-11-22 17:36:50.000000000 +0100
@@ -20,6 +20,7 @@
 
 #define _LINUX_STRING_H_
 #define kdev_t int
+#include "asm/atomic.h"
 #include "linux/fs.h"
 #undef _LINUX_STRING_H_
 
