--- Mesa-3.5/configure.in.orig	Wed Oct 17 15:59:16 2001
+++ Mesa-3.5/configure.in	Wed Oct 17 16:00:34 2001
@@ -144,10 +144,7 @@
   dnl Optimization flags
   if test "x$enable_debug" = xno && test "x$enable_prof" = xno; then
     if test "x$GCC" = xyes; then
-      CFLAGS="$CFLAGS -fomit-frame-pointer -ffast-math -fexpensive-optimizations -fstrict-aliasing"
-      case "$host" in
-      i*86-*-*) CFLAGS="$CFLAGS -malign-loops=2 -malign-jumps=2 -malign-functions=2";;
-      esac     
+      CFLAGS="$CFLAGS $RPM_OPT_FLAGS -fstrict-aliasing"
     fi
   fi
 fi
