Index: util-linux-2.12q/hwclock/rtc.c
===================================================================
--- util-linux-2.12q.orig/hwclock/rtc.c	2004-12-15 21:21:48.000000000 +0100
+++ util-linux-2.12q/hwclock/rtc.c	2005-04-29 14:58:18.000000000 +0200
@@ -41,8 +41,8 @@ struct sparc_rtc_time
         int     year;   /* Year (0-99) */
 };
 
-#define RTCGET _IOR('p', 20, struct sparc_rtc_time)
-#define RTCSET _IOW('p', 21, struct sparc_rtc_time)
+#define RTCGET _IOR('p', 20, sparc_rtc_time)
+#define RTCSET _IOW('p', 21, sparc_rtc_time)
 
 
 /* non-sparc stuff */
Index: util-linux-2.12q/partx/gpt.c
===================================================================
--- util-linux-2.12q.orig/partx/gpt.c	2002-08-04 02:20:16.000000000 +0200
+++ util-linux-2.12q/partx/gpt.c	2005-04-29 14:58:18.000000000 +0200
@@ -41,7 +41,7 @@
 #define BLKGETLASTSECT  _IO(0x12,108)   /* get last sector of block device */
 #define BLKGETSIZE _IO(0x12,96)	        /* return device size */
 #define BLKSSZGET  _IO(0x12,104)	/* get block device sector size */
-#define BLKGETSIZE64 _IOR(0x12,114,sizeof(uint64_t))	/* return device size in bytes (u64 *arg) */
+#define BLKGETSIZE64 _IOR(0x12,114,uint64_t)	/* return device size in bytes (u64 *arg) */
 
 struct blkdev_ioctl_param {
         unsigned int block;
