Index: util-linux-2.12q/mount/mount.c
===================================================================
--- util-linux-2.12q.orig/mount/mount.c	2005-04-29 14:51:09.000000000 +0200
+++ util-linux-2.12q/mount/mount.c	2005-04-29 15:03:11.000000000 +0200
@@ -915,6 +915,7 @@ try_mount_one (const char *spec0, const 
   /*
    * Call mount.TYPE for types that require a separate mount program.
    * For the moment these types are ncpfs and smbfs. Maybe also vxfs.
+   * For the moment these types are ncpfs and smbfs and davfs. Maybe also vxfs.
    * All such special things must occur isolated in the types string.
    */
   if (check_special_mountprog(spec, node, types, flags, extra_opts, &status)) {
@@ -1269,7 +1270,12 @@ mount_one (const char *spec, const char 
 		spec = nspec;
 
 	if (types == NULL && !mounttype && !is_existing_file(spec)) {
-		if (strchr (spec, ':') != NULL) {
+		if((!strncmp(spec, "http://", 7))||(!strncmp(spec, "https://", 8))) {
+			types = "davfs";
+			if (verbose)
+				printf(_("mount: no type was given - "
+					"I'll assume dav because of the http[s]:// prefix\n"));
+		} else if (strchr (spec, ':') != NULL) {
 			types = "nfs";
 			if (verbose)
 				printf(_("mount: no type was given - "
