--- util-linux-2.12/mount/swapon.c.swaponsymlink	2004-01-30 12:16:46.062769289 +0100
+++ util-linux-2.12/mount/swapon.c	2004-01-30 12:18:38.475564719 +0100
@@ -624,8 +624,16 @@
 					}
 					continue;
 				}
-				for (i = 0; i < numSwaps; i++)
-				     if (!strcmp(swapList[i], fstab->mnt_fsname)) break;
+				for (i = 0; i < numSwaps; i++) {
+				     struct stat swapped, swapping;
+
+				     if (stat (swapList[i], &swapped) == 0
+					 && stat (fstab->mnt_fsname, &swapping) == 0) {
+				         if (swapped.st_dev == swapping.st_dev
+					     && swapped.st_ino == swapping.st_ino)
+					     break;
+				     }
+				}
 
 				if ((i == numSwaps) && !is_in_proc_swaps(special) && 
 				    (!ifexists || !access(special, R_OK))) {
