--- autoconf-2.60/man/Makefile.in~	2006-12-29 17:08:35.000000000 -0700
+++ autoconf-2.60/man/Makefile.in	2006-12-29 17:08:35.000000000 -0700
@@ -368,19 +368,6 @@
 	    (cd $$dir && $(MAKE) $(AM_MAKEFLAGS) $$program) || exit; \
 	  done;; \
 	esac
-	echo "Updating man page $@"
-	PATH="$(top_builddir)/tests@PATH_SEPARATOR@$(top_srcdir)/config@PATH_SEPARATOR@$$PATH"; \
-	export PATH; \
-	$(HELP2MAN) \
-	    --include=$*.x \
-	    --include=$(srcdir)/common.x \
-	    --output=$@.t `echo '$*' | sed 's,.*/,,'`
-	if sed $(remove_time_stamp) $@ >$@a.t 2>/dev/null && \
-	   sed $(remove_time_stamp) $@.t | cmp $@a.t - >/dev/null 2>&1; then \
-		touch $@; \
-	else \
-		mv $@.t $@; \
-	fi
 	rm -f $@*.t
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
 # Otherwise a system limit (for SysV at least) may be exceeded.
--- autoconf-2.60/man/Makefile.am~	2006-12-29 17:08:56.000000000 -0700
+++ autoconf-2.60/man/Makefile.am	2006-12-29 17:08:56.000000000 -0700
@@ -63,17 +63,4 @@
 	    (cd $$dir && $(MAKE) $(AM_MAKEFLAGS) $$program) || exit; \
 	  done;; \
 	esac
-	echo "Updating man page $@"
-	PATH="$(top_builddir)/tests@PATH_SEPARATOR@$(top_srcdir)/config@PATH_SEPARATOR@$$PATH"; \
-	export PATH; \
-	$(HELP2MAN) \
-	    --include=$*.x \
-	    --include=$(srcdir)/common.x \
-	    --output=$@.t `echo '$*' | sed 's,.*/,,'`
-	if sed $(remove_time_stamp) $@ >$@a.t 2>/dev/null && \
-	   sed $(remove_time_stamp) $@.t | cmp $@a.t - >/dev/null 2>&1; then \
-		touch $@; \
-	else \
-		mv $@.t $@; \
-	fi
 	rm -f $@*.t
--- autoconf-2.60/Makefile.in~	2006-12-29 17:09:27.000000000 -0700
+++ autoconf-2.60/Makefile.in	2006-12-29 17:09:27.000000000 -0700
@@ -94,7 +94,7 @@
 EMACSLOADPATH = @EMACSLOADPATH@
 EXPR = @EXPR@
 GREP = @GREP@
-HELP2MAN = @HELP2MAN@
+#HELP2MAN = @HELP2MAN@
 INSTALL_DATA = @INSTALL_DATA@
 INSTALL_PROGRAM = @INSTALL_PROGRAM@
 INSTALL_SCRIPT = @INSTALL_SCRIPT@
