--- gmp-4.1.2/mpz/gcd_ui.c.mpz_gcd_ui-retval	2002-05-22 00:15:44.000000000 +0200
+++ gmp-4.1.2/mpz/gcd_ui.c	2003-07-10 15:59:43.000000000 +0200
@@ -58,8 +58,9 @@ mpz_gcd_ui (mpz_ptr w, mpz_srcptr u, uns
 	    }
 	  SIZ(w) = un;
 	}
-      /* We can't return any useful result for gcd(big,0).  */
-      return un > 1 ? 0 : PTR(u)[0];
+      /* Return u if it fits a ulong, otherwise 0. */
+      res = PTR(u)[0];
+      return (un == 1 && res <= ULONG_MAX ? res : 0);
     }
   else
     res = mpn_gcd_1 (PTR(u), un, (mp_limb_t) v);
