Index: squid/src/repl/heap/store_heap_replacement.c
diff -c squid/src/repl/heap/store_heap_replacement.c:1.7 squid/src/repl/heap/store_heap_replacement.c:1.7.2.1
*** squid/src/repl/heap/store_heap_replacement.c:1.7	Tue Oct 23 23:26:22 2001
--- squid/src/repl/heap/store_heap_replacement.c	Sun Feb 13 14:23:19 2005
***************
*** 79,85 ****
  	tie = 1.0 - exp((double) (e->lastref - squid_curtime) / 86400.0);
      key = age + (double) e->refcount - tie;
      debug(81, 3) ("HeapKeyGen_StoreEntry_LFUDA: %s refcnt=%d lastref=%ld age=%f tie=%f -> %f\n",
! 	storeKeyText(e->hash.key), (int)e->refcount, e->lastref, age, tie, key);
      if (e->mem_obj && e->mem_obj->url)
  	debug(81, 3) ("HeapKeyGen_StoreEntry_LFUDA: url=%s\n",
  	    e->mem_obj->url);
--- 79,85 ----
  	tie = 1.0 - exp((double) (e->lastref - squid_curtime) / 86400.0);
      key = age + (double) e->refcount - tie;
      debug(81, 3) ("HeapKeyGen_StoreEntry_LFUDA: %s refcnt=%d lastref=%ld age=%f tie=%f -> %f\n",
! 	storeKeyText(e->hash.key), (int) e->refcount, (long int) e->lastref, age, tie, key);
      if (e->mem_obj && e->mem_obj->url)
  	debug(81, 3) ("HeapKeyGen_StoreEntry_LFUDA: url=%s\n",
  	    e->mem_obj->url);
***************
*** 115,121 ****
      double tie = (e->lastref > 1) ? (1.0 / e->lastref) : 1.0;
      key = age + ((double) e->refcount / size) - tie;
      debug(81, 3) ("HeapKeyGen_StoreEntry_GDSF: %s size=%f refcnt=%d lastref=%ld age=%f tie=%f -> %f\n",
! 	storeKeyText(e->hash.key), size, (int)e->refcount, e->lastref, age, tie, key);
      if (e->mem_obj && e->mem_obj->url)
  	debug(81, 3) ("HeapKeyGen_StoreEntry_GDSF: url=%s\n",
  	    e->mem_obj->url);
--- 115,121 ----
      double tie = (e->lastref > 1) ? (1.0 / e->lastref) : 1.0;
      key = age + ((double) e->refcount / size) - tie;
      debug(81, 3) ("HeapKeyGen_StoreEntry_GDSF: %s size=%f refcnt=%d lastref=%ld age=%f tie=%f -> %f\n",
! 	storeKeyText(e->hash.key), size, (int) e->refcount, (long int) e->lastref, age, tie, key);
      if (e->mem_obj && e->mem_obj->url)
  	debug(81, 3) ("HeapKeyGen_StoreEntry_GDSF: url=%s\n",
  	    e->mem_obj->url);
Index: squid/src/carp.c
diff -c squid/src/carp.c:1.15.2.2 squid/src/carp.c:1.15.2.3
*** squid/src/carp.c:1.15.2.2	Thu Sep 30 03:23:30 2004
--- squid/src/carp.c	Sun Feb 13 13:54:09 2005
***************
*** 126,132 ****
  	combined_hash += combined_hash * 0x62531965;
  	combined_hash = ROTATE_LEFT(combined_hash, 21);
  	combined_hash = combined_hash * tp->carp.load_multiplier;
! 	debug(39, 3) ("carpSelectParent: %s combined_hash %d\n",
  	    tp->host, combined_hash);
  	if ((combined_hash > high_score) && neighborUp(tp)) {
  	    p = tp;
--- 126,132 ----
  	combined_hash += combined_hash * 0x62531965;
  	combined_hash = ROTATE_LEFT(combined_hash, 21);
  	combined_hash = combined_hash * tp->carp.load_multiplier;
! 	debug(39, 3) ("carpSelectParent: %s combined_hash %ld\n",
  	    tp->host, combined_hash);
  	if ((combined_hash > high_score) && neighborUp(tp)) {
  	    p = tp;
Index: squid/src/external_acl.c
diff -c squid/src/external_acl.c:1.1.2.31 squid/src/external_acl.c:1.1.2.32
*** squid/src/external_acl.c:1.1.2.31	Sat Aug 14 15:01:33 2004
--- squid/src/external_acl.c	Sun Feb 13 14:23:19 2005
***************
*** 231,237 ****
  	    if (member) {
  		/* Split in header and member */
  		*member++ = '\0';
! 		if (!isalnum(*member))
  		    format->separator = *member++;
  		else
  		    format->separator = ',';
--- 231,237 ----
  	    if (member) {
  		/* Split in header and member */
  		*member++ = '\0';
! 		if (!isalnum((int) *member))
  		    format->separator = *member++;
  		else
  		    format->separator = ',';
Index: squid/src/main.c
diff -c squid/src/main.c:1.345.2.15 squid/src/main.c:1.345.2.17
*** squid/src/main.c:1.345.2.15	Tue Dec 28 05:54:35 2004
--- squid/src/main.c	Sun Feb 13 14:19:44 2005
***************
*** 836,842 ****
  	return 0;
      if (kill(pid, 0) < 0)
  	return 0;
!     debug(0, 0) ("Squid is already running!  Process ID %d\n", pid);
      return 1;
  }
  
--- 836,842 ----
  	return 0;
      if (kill(pid, 0) < 0)
  	return 0;
!     debug(0, 0) ("Squid is already running!  Process ID %ld\n", (long int) pid);
      return 1;
  }
  
***************
*** 853,859 ****
--- 853,861 ----
      int status;
  #endif
      pid_t pid;
+ #ifdef TIOCNOTTY
      int i;
+ #endif
      int nullfd;
      if (*(argv[0]) == '(')
  	return;
Index: squid/src/net_db.c
diff -c squid/src/net_db.c:1.158.2.4 squid/src/net_db.c:1.158.2.5
*** squid/src/net_db.c:1.158.2.4	Thu Jan  2 16:21:57 2003
--- squid/src/net_db.c	Sun Feb 13 14:23:19 2005
***************
*** 621,627 ****
      }
      debug(38, 3) ("netdbExchangeHandleReply: used %d entries, (x %d bytes) == %d bytes total\n",
  	nused, rec_sz, nused * rec_sz);
!     debug(38, 3) ("netdbExchangeHandleReply: seen %d, used %d\n", ex->seen, ex->used);
      if (EBIT_TEST(ex->e->flags, ENTRY_ABORTED)) {
  	debug(38, 3) ("netdbExchangeHandleReply: ENTRY_ABORTED\n");
  	netdbExchangeDone(ex);
--- 621,627 ----
      }
      debug(38, 3) ("netdbExchangeHandleReply: used %d entries, (x %d bytes) == %d bytes total\n",
  	nused, rec_sz, nused * rec_sz);
!     debug(38, 3) ("netdbExchangeHandleReply: seen %ld, used %ld\n", (long int) ex->seen, (long int) ex->used);
      if (EBIT_TEST(ex->e->flags, ENTRY_ABORTED)) {
  	debug(38, 3) ("netdbExchangeHandleReply: ENTRY_ABORTED\n");
  	netdbExchangeDone(ex);
Index: squid/src/stat.c
diff -c squid/src/stat.c:1.351.2.11 squid/src/stat.c:1.351.2.12
*** squid/src/stat.c:1.351.2.11	Tue Dec  7 18:00:56 2004
--- squid/src/stat.c	Sun Feb 13 14:19:44 2005
***************
*** 567,584 ****
      mp = mallinfo();
      storeAppendPrintf(sentry, "Memory usage for %s via mallinfo():\n",
  	appname);
!     storeAppendPrintf(sentry, "\tTotal space in arena:  %6d KB\n",
! 	mp.arena >> 10);
!     storeAppendPrintf(sentry, "\tOrdinary blocks:       %6d KB %6d blks\n",
! 	mp.uordblks >> 10, mp.ordblks);
!     storeAppendPrintf(sentry, "\tSmall blocks:          %6d KB %6d blks\n",
! 	mp.usmblks >> 10, mp.smblks);
!     storeAppendPrintf(sentry, "\tHolding blocks:        %6d KB %6d blks\n",
! 	mp.hblkhd >> 10, mp.hblks);
!     storeAppendPrintf(sentry, "\tFree Small blocks:     %6d KB\n",
! 	mp.fsmblks >> 10);
!     storeAppendPrintf(sentry, "\tFree Ordinary blocks:  %6d KB\n",
! 	mp.fordblks >> 10);
      t = mp.uordblks + mp.usmblks + mp.hblkhd;
      storeAppendPrintf(sentry, "\tTotal in use:          %6d KB %d%%\n",
  	t >> 10, percent(t, mp.arena + mp.hblkhd));
--- 567,584 ----
      mp = mallinfo();
      storeAppendPrintf(sentry, "Memory usage for %s via mallinfo():\n",
  	appname);
!     storeAppendPrintf(sentry, "\tTotal space in arena:  %6ld KB\n",
! 	(long int) mp.arena >> 10);
!     storeAppendPrintf(sentry, "\tOrdinary blocks:       %6ld KB %6ld blks\n",
! 	(long int) mp.uordblks >> 10, (long int) mp.ordblks);
!     storeAppendPrintf(sentry, "\tSmall blocks:          %6ld KB %6ld blks\n",
! 	(long int) mp.usmblks >> 10, (long int) mp.smblks);
!     storeAppendPrintf(sentry, "\tHolding blocks:        %6ld KB %6ld blks\n",
! 	(long int) mp.hblkhd >> 10, (long int) mp.hblks);
!     storeAppendPrintf(sentry, "\tFree Small blocks:     %6ld KB\n",
! 	(long int) mp.fsmblks >> 10);
!     storeAppendPrintf(sentry, "\tFree Ordinary blocks:  %6ld KB\n",
! 	(long int) mp.fordblks >> 10);
      t = mp.uordblks + mp.usmblks + mp.hblkhd;
      storeAppendPrintf(sentry, "\tTotal in use:          %6d KB %d%%\n",
  	t >> 10, percent(t, mp.arena + mp.hblkhd));
Index: squid/src/tools.c
diff -c squid/src/tools.c:1.213.2.10 squid/src/tools.c:1.213.2.11
*** squid/src/tools.c:1.213.2.10	Sat Aug 14 15:01:33 2004
--- squid/src/tools.c	Sun Feb 13 14:19:44 2005
***************
*** 132,149 ****
  	return;
      mp = mallinfo();
      fprintf(debug_log, "Memory usage for %s via mallinfo():\n", appname);
!     fprintf(debug_log, "\ttotal space in arena:  %6d KB\n",
! 	mp.arena >> 10);
!     fprintf(debug_log, "\tOrdinary blocks:       %6d KB %6d blks\n",
! 	mp.uordblks >> 10, mp.ordblks);
!     fprintf(debug_log, "\tSmall blocks:          %6d KB %6d blks\n",
! 	mp.usmblks >> 10, mp.smblks);
!     fprintf(debug_log, "\tHolding blocks:        %6d KB %6d blks\n",
! 	mp.hblkhd >> 10, mp.hblks);
!     fprintf(debug_log, "\tFree Small blocks:     %6d KB\n",
! 	mp.fsmblks >> 10);
!     fprintf(debug_log, "\tFree Ordinary blocks:  %6d KB\n",
! 	mp.fordblks >> 10);
      t = mp.uordblks + mp.usmblks + mp.hblkhd;
      fprintf(debug_log, "\tTotal in use:          %6d KB %d%%\n",
  	t >> 10, percent(t, mp.arena));
--- 132,149 ----
  	return;
      mp = mallinfo();
      fprintf(debug_log, "Memory usage for %s via mallinfo():\n", appname);
!     fprintf(debug_log, "\ttotal space in arena:  %6ld KB\n",
! 	(long int) mp.arena >> 10);
!     fprintf(debug_log, "\tOrdinary blocks:       %6ld KB %6ld blks\n",
! 	(long int) mp.uordblks >> 10, (long int) mp.ordblks);
!     fprintf(debug_log, "\tSmall blocks:          %6ld KB %6ld blks\n",
! 	(long int) mp.usmblks >> 10, (long int) mp.smblks);
!     fprintf(debug_log, "\tHolding blocks:        %6ld KB %6ld blks\n",
! 	(long int) mp.hblkhd >> 10, (long int) mp.hblks);
!     fprintf(debug_log, "\tFree Small blocks:     %6ld KB\n",
! 	(long int) mp.fsmblks >> 10);
!     fprintf(debug_log, "\tFree Ordinary blocks:  %6ld KB\n",
! 	(long int) mp.fordblks >> 10);
      t = mp.uordblks + mp.usmblks + mp.hblkhd;
      fprintf(debug_log, "\tTotal in use:          %6d KB %d%%\n",
  	t >> 10, percent(t, mp.arena));
