diff -urNp procmail-3.22.orig/config.h procmail-3.22/config.h
--- procmail-3.22.orig/config.h	Tue Sep 11 06:53:50 2001
+++ procmail-3.22/config.h	Tue Sep 25 04:12:28 2001
@@ -78,8 +78,8 @@
 			 "list","slist","lists","news",0}
 
 /*#define NO_fcntl_LOCK		/* uncomment any of these three if you	     */
-/*#define NO_lockf_LOCK		/* definitely do not want procmail to make   */
-/*#define NO_flock_LOCK		/* use of those kernel-locking methods	     */
+#define NO_lockf_LOCK		/* definitely do not want procmail to make   */
+#define NO_flock_LOCK		/* use of those kernel-locking methods	     */
 				/* If you set LOCKINGTEST to a binary number
 	than there's no need to set these.  These #defines are only useful
 	if you want to disable particular locking styles but are unsure which
