--- shadow-20000826/lib/defines.h.chmou	Sat Aug 26 14:27:17 2000
+++ shadow-20000826/lib/defines.h	Fri Dec 22 09:58:03 2000
@@ -140,7 +140,7 @@
 
 /* cleaner than lots of #ifdefs everywhere - use this as follows:
    SYSLOG((LOG_CRIT, "user %s cracked root", user)); */
-#if ENABLE_NLS
+#if 0
 /* Temporarily set LC_TIME to "C" to avoid strange dates in syslog.
    This is a workaround for a more general syslog(d) design problem -
    syslogd should log the current system time for each event, and not
