--- util-linux-2.11h/fdisk/fdisk.c.chmou	Mon Aug  6 17:51:39 2001
+++ util-linux-2.11h/fdisk/fdisk.c	Mon Aug  6 17:57:30 2001
@@ -45,6 +45,8 @@
 
 #include <linux/hdreg.h>       /* for HDIO_GETGEO */
 
+#define O_LARGEFILE	0100000
+
 #include "nls.h"
 #include "common.h"
 #include "fdisk.h"
--- util-linux-2.12/mount/fstab.c.orig  2004-01-30 13:48:56.273874331 +0100
+++ util-linux-2.12/mount/fstab.c       2004-01-30 13:58:12.494513734 +0100
@@ -20,6 +20,8 @@
 
 #define PROC_MOUNTS            "/proc/mounts"
 
+#define O_LARGEFILE    0100000
+
 
 /* Information about mtab. ------------------------------------*/
 static int have_mtab_info = 0;
