--- console-tools-0.2.3/console-data-1999.08.29/keymaps/i386/azerty/fr-latin1.kmap.chmou	Sun Aug 29 21:15:30 1999
+++ console-tools-0.2.3/console-data-1999.08.29/keymaps/i386/azerty/fr-latin1.kmap	Tue Jan 16 10:30:43 2001
@@ -56,6 +56,7 @@
 keycode  41 = twosuperior     
 	control keycode  41 = nul             
 	alt     keycode  41 = Meta_grave      
+	Shift	keycode	 41 = tilde
 keycode  42 = Shift           
 keycode  43 = asterisk         mu              
 	control keycode  43 = Control_backslash
--- console-tools-0.2.3/console-data-1999.08.29/keymaps/i386/azerty/fr.kmap.chmou	Sun Aug 29 21:15:30 1999
+++ console-tools-0.2.3/console-data-1999.08.29/keymaps/i386/azerty/fr.kmap	Tue Jan 16 10:30:51 2001
@@ -56,6 +56,7 @@
 keycode  41 = asterisk         asciitilde      
 	control keycode  41 = nul             
 	alt     keycode  41 = Meta_grave      
+	Shift	keycode	 41 = tilde
 keycode  42 = Shift           
 keycode  43 = asterisk         numbersign      
 	control keycode  43 = Control_backslash
--- console-tools-0.2.3/console-data-1999.08.29/keymaps/i386/azerty/fr-latin0.kmap.chmou	Sun Aug 29 21:15:30 1999
+++ console-tools-0.2.3/console-data-1999.08.29/keymaps/i386/azerty/fr-latin0.kmap	Tue Jan 16 10:31:08 2001
@@ -235,7 +235,7 @@
 keycode  41 =  +onehalf       +onequarter	guillemotleft	guillemotright
 	control keycode 41 = nul
 	alt     keycode 41 = Meta_grave
-
+	Shift	keycode	 41 = tilde
 keycode  42 = CapsShift
 
 keycode  43 = +asterisk         +mu		threesuperior	yen
