--- src/makewhatis.sh.orig	2002-05-22 17:17:08.000000000 +0200
+++ src/makewhatis.sh	2002-05-22 18:58:23.000000000 +0200
@@ -130,10 +130,15 @@
 fi
 catpath=`echo ${catpath} | tr : ' '`
 
+#WHATIS_DIR=$DESTDIR/var/cache/man/`echo $here|sed -e 's!.*/man/!!g'`
+WHATIS_DIR=$DESTDIR/var/cache/man/$LANG
+[[ -d $WHATIS_DIR ]] || mkdir -p $WHATIS_DIR/
+
 # first truncate all the whatis files that will be created new,
 # then only update - we might visit the same directory twice
 if [ x$update = x ]; then
-   cp /dev/null /var/cache/man/whatis
+   mkdir -p $WHATIS_DIR/
+   /bin/echo -n > $WHATIS_DIR/whatis
 fi
 
 for pages in man cat
@@ -356,13 +360,13 @@
 
        cd $here
 
-       if [ -f /var/cache/man/whatis ]
+       if [ -f $WHATIS_DIR/whatis ]
        then
-	 cat /var/cache/man/whatis >> $TMPFILE
+	 cat $WHATIS_DIR/whatis >> $TMPFILE
        fi
-       sed '/^$/d' < $TMPFILE | sort | uniq > /var/cache/man/whatis
+       sed '/^$/d' < $TMPFILE | sort | uniq > $WHATIS_DIR/whatis
 
-       chmod 644 /var/cache/man/whatis
+       chmod 644 $WHATIS_DIR/whatis
        rm $TMPFILE
    done
 done
