--- man-1.5m2/src/man.conf.in.avx	2007-04-27 20:53:36.574280750 -0600
+++ man-1.5m2/src/man.conf.in	2007-04-27 20:55:14.460398250 -0600
@@ -36,9 +36,8 @@
 #
 # Every automatically generated MANPATH includes these fields
 #
-MANPATH	/usr/man
 MANPATH /usr/share/man
-MANPATH	/usr/local/man
+MANPATH	/usr/local/share/man
 MANPATH	/usr/X11R6/man
 #
 # Uncomment if you want to include one of these by default
@@ -65,8 +64,8 @@
 MANPATH_MAP	/sbin			/usr/share/man
 MANPATH_MAP	/usr/bin		/usr/share/man
 MANPATH_MAP	/usr/sbin		/usr/share/man
-MANPATH_MAP	/usr/local/bin		/usr/local/man
-MANPATH_MAP	/usr/local/sbin		/usr/local/man
+MANPATH_MAP	/usr/local/bin		/usr/local/share/man
+MANPATH_MAP	/usr/local/sbin		/usr/local/share/man
 MANPATH_MAP	/usr/X11R6/bin		/usr/X11R6/man
 MANPATH_MAP	/usr/bin/X11		/usr/X11R6/man
 MANPATH_MAP	/usr/bin/mh		/usr/share/man
