--- dietlibc-0.27/test/cycles.c.ppc-rdtsc	2004-12-14 17:00:58.000000000 -0500
+++ dietlibc-0.27/test/cycles.c	2004-12-14 17:18:25.007616012 -0500
@@ -12,6 +12,16 @@
   asm volatile ("rdtsc" : "=a" (l), "=d" (h));	\
   dst = (((uint64_t)h) << 32) | l;				\
 } while (0)
+#elif defined (__powerpc64__)
+#define RDTSC(dst) asm volatile ("mftb %0" : "=r" (dst))
+#elif defined (__powerpc__)
+#define RDTSC(dst) do { \
+  uint32_t chk, tbl, tbu; \
+  /* The code below is as suggested in Motorola reference manual for 32 bits PPCs. */ \
+  __asm__ __volatile__ ("1: mftbu %0; mftb %1; mftbu %2; cmpw %2,%0; bne 1b" \
+    : "=r" (tbu), "=r" (tbl), "=r" (chk) ); \
+  dst = ((uint64_t)tbu << 32) | tbl; \
+} while (0)
 #else
 #error "Unimplemented rdtsc"
 #endif
