--- pam-0.74/modules/pam_console/chmod.c.fred	Fri Mar  9 02:00:21 2001
+++ pam-0.74/modules/pam_console/chmod.c	Thu Mar 15 16:54:27 2001
@@ -161,7 +161,8 @@
     {
       if(mntent->mnt_dir &&
          mntent->mnt_fsname &&
-	 (fnmatch(dir, mntent->mnt_dir, 0) == 0))
+	 (fnmatch(dir, mntent->mnt_dir, 0) == 0) &&
+	 strcmp(dir, mntent->mnt_fsname))
         {
           errors |= change_file(mntent->mnt_fsname, changes, TRUE, user, group);
         }
