--- Linux-PAM-0.77/modules/pam_xauth/pam_xauth.c.xauth-groups	2003-07-08 05:16:28.000000000 +0200
+++ Linux-PAM-0.77/modules/pam_xauth/pam_xauth.c	2003-08-25 17:08:40.000000000 +0200
@@ -67,7 +67,7 @@
  * given input on stdin, and storing any output it generates. */
 static int
 run_coprocess(const char *input, char **output,
-	      uid_t uid, gid_t gid, const char *command, ...)
+	      uid_t uid, gid_t gid, const char *name, const char *command, ...)
 {
 	int ipipe[2], opipe[2], i;
 	char buf[LINE_MAX];
@@ -102,10 +102,18 @@
 		/* We're the child. */
 		char *args[10];
 		const char *tmp;
+		
+		/* Mandrake Linux specific:
+		 * we need to set the extra groups because in high security levels
+		 * access to /usr/X11R6/bin dir is controlled by a group */
+		setgroups(0, NULL);
+		initgroups(name, gid);
+
 		/* Drop privileges. */
 		setgid(gid);
-		setgroups(0, NULL);
 		setuid(uid);
+
+		
 		/* Initialize the argument list. */
 		memset(&args, 0, sizeof(args));
 		/* Set the pipe descriptors up as stdin and stdout, and close
@@ -130,6 +138,7 @@
 		}
 		/* Run the command. */
 		execvp(command, args);
+		syslog(LOG_ERR, "pam_xauth: execvp of %s failed: %m", command);
 		/* Never reached. */
 		exit(1);
 	}
@@ -404,7 +413,7 @@
 		       (unsigned long) getgid());
 	}
 	if (run_coprocess(NULL, &cookie,
-			  getuid(), getgid(),
+			  getuid(), getgid(), rpwd->pw_name,
 			  xauth, "-f", cookiefile, "nlist", display,
 			  NULL) == 0) {
 		/* Check that we got a cookie.  If not, we get creative. */
@@ -457,7 +466,7 @@
 						       (unsigned long) getgid());
 					}
 					run_coprocess(NULL, &cookie,
-						      getuid(), getgid(),
+						      getuid(), getgid(), rpwd->pw_name,
 						      xauth, "-f", cookiefile,
 						      "nlist", t, NULL);
 				}
@@ -556,7 +565,7 @@
 			       (unsigned long) tpwd->pw_gid);
 		}
 		run_coprocess(cookie, &tmp,
-			      tpwd->pw_uid, tpwd->pw_gid,
+			      tpwd->pw_uid, tpwd->pw_gid, tpwd->pw_name,
 			      xauth, "-f", cookiefile, "nmerge", "-", NULL);
 
 		/* We don't need to keep a copy of these around any more. */
