diff -urN mtools-3.9.6-orig/fat.c mtools-3.9.6/fat.c
--- mtools-3.9.6-orig/fat.c	Fri Apr 16 21:27:19 1999
+++ mtools-3.9.6/fat.c	Sat Aug  7 12:29:56 1999
@@ -608,6 +608,10 @@
 		 * and 0xff.  Do not consider this as an error */
 		return 0;
 	
+
+	/* Some Atari ST disks are actually using f7/f9; ignore this
+	   "sanity check" to read them...
+	
 	if((address[0] != boot->descr && boot->descr >= 0xf0 &&
 	    (address[0] != 0xf9 || boot->descr != 0xf0)) ||
 	   address[0] < 0xf0) {
@@ -616,7 +620,7 @@
 				address[0],
 				boot->descr);
 		return -1;
-	}
+	} */
 
 	if(address[1] != 0xff || address[2] != 0xff){
 		fprintf(stderr,"Initial byte of fat is not 0xff\n");
