--- programs/pluto/server.c~	2006-06-27 17:42:32.000000000 -0600
+++ programs/pluto/server.c	2006-06-27 17:42:32.000000000 -0600
@@ -777,7 +777,7 @@
 		    q = alloc_thing(struct iface_port, "struct iface_port");
 		    id = alloc_thing(struct iface_dev, "struct iface_dev");
 
-		    LIST_INSERT_HEAD(&interface_dev, id, id_entry)
+		    LIST_INSERT_HEAD(&interface_dev, id, id_entry);
 
 		    q->ip_dev = id;
 		    id->id_rname = clone_str(ifp->name, "real device name");
