--- modutils-2.4.22/util/alias.h.chmou	2003-03-05 16:21:55.000000000 +0100
+++ modutils-2.4.22/util/alias.h	2003-03-05 16:21:59.000000000 +0100
@@ -246,7 +246,9 @@
 char *above[] =
 {
 	"hid keybdev mousedev",
+#if FIXME
 	"usbmouse hid",
+#endif
 	"wacom evdev",
 	NULL			/* marks the end of the list! */
 };
@@ -274,6 +276,18 @@
 };
 
 /*
+ * This is the list of pre-defined "pre-install"s,
+ * used to execute commands before loading modules.
+ * /etc/modules.conf can add entries but not remove them.
+ */
+char *pre_install[] =
+{
+        "usbmouse    /sbin/modprobe hid",
+	NULL 
+};
+
+
+/*
  * This is the list of pre-defined "pre-remove"s,
  * used to execute commands before unloading modules.
  * /etc/modules.conf can add entries but not remove them.
--- modutils-2.4.22/util/config.c.chmou	2003-03-05 16:21:55.000000000 +0100
+++ modutils-2.4.22/util/config.c	2003-03-05 16:21:59.000000000 +0100
@@ -702,6 +702,13 @@
 				decode_exec(xstrdup(post_install[i]), EXEC_POST_INSTALL);
 		
 		/*
+		 * Build predef preinstalls
+		 */
+		if (all && pre_install[0])
+			for (i=0 ; pre_install[i] ; i++ )
+				decode_exec(xstrdup(pre_install[i]), EXEC_PRE_INSTALL);
+		
+		/*
 		 * Build predef preremoves
 		 */
 		if (all && pre_remove[0])
