--- old/drivers/usb/storage/usb.c	2003-09-02 07:56:15.000000000 +0200
+++ old/drivers/usb/storage/usb.c.new	2003-09-02 08:04:16.000000000 +0200
@@ -844,7 +844,8 @@ static void * storage_probe(struct usb_d
 			ss->transport_name = "Bulk";
 			ss->transport = usb_stor_Bulk_transport;
 			ss->transport_reset = usb_stor_Bulk_reset;
-			ss->max_lun = usb_stor_Bulk_max_lun(ss);
+			if (!(ss->flags & US_FL_SINGLE_LUN))
+			        ss->max_lun = usb_stor_Bulk_max_lun(ss);
 			break;
 
 #ifdef CONFIG_USB_STORAGE_HP8200e
@@ -1034,6 +1035,10 @@ static void * storage_probe(struct usb_d
 		ss->htmplt.module = THIS_MODULE;
 		scsi_register_module(MODULE_SCSI_HA, &(ss->htmplt));
 
+		/* some device need reset process */
+		if (ss->flags & US_FL_INIT_RESET)
+			ss->transport_reset(ss);
+
 		/* lock access to the data structures */
 		down(&us_list_semaphore);
 
--- old/drivers/usb/storage/usb.h	2003-09-02 07:56:15.000000000 +0200
+++ old/drivers/usb/storage/usb.h.new	2003-09-02 08:06:22.000000000 +0200
@@ -104,6 +104,7 @@ struct us_unusual_dev {
 #define US_FL_FIX_INQUIRY     0x00000040 /* INQUIRY response needs fixing */
 #define US_FL_FIX_CAPACITY    0x00000080 /* READ_CAPACITY response too big */
 #define US_FL_SL_IDE_BUG      0x00000100 /* ScanLogic usb-ide workaround */
+#define US_FL_INIT_RESET      0x00000120 /* reset process when initialize */
 
 #define USB_STOR_STRING_LEN 32
 
