diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/arch/i386/kernel/head.S w1/arch/i386/kernel/head.S
--- w2/arch/i386/kernel/head.S	2003-08-20 19:14:10.000000000 +0200
+++ w1/arch/i386/kernel/head.S	2003-08-18 23:28:17.000000000 +0200
@@ -124,7 +124,7 @@ startup_32:
 	popfl
 	jmp checkCPUtype
 1:
-#endif CONFIG_SMP
+#endif /* CONFIG_SMP */
 
 /*
  * Clear BSS first so that there are no surprises...
