diff -up q1/drivers/mtd/devices/Config.in.orig q1/drivers/mtd/devices/Config.in
--- q1/drivers/mtd/devices/Config.in.orig	2002-07-30 14:57:26.000000000 +0200
+++ q1/drivers/mtd/devices/Config.in	2002-07-30 14:57:03.000000000 +0200
@@ -25,7 +25,9 @@ fi
 dep_tristate '  MTD emulation using block device' CONFIG_MTD_BLKMTD $CONFIG_MTD
 
 comment 'Disk-On-Chip Device Drivers'
-   dep_tristate '  M-Systems Disk-On-Chip 1000' CONFIG_MTD_DOC1000 $CONFIG_MTD
+   if [ "$CONFIG_SMP" != "y" ]; then
+      dep_tristate '  M-Systems Disk-On-Chip 1000' CONFIG_MTD_DOC1000 $CONFIG_MTD
+   fi
    dep_tristate '  M-Systems Disk-On-Chip 2000 and Millennium' CONFIG_MTD_DOC2000 $CONFIG_MTD
    dep_tristate '  M-Systems Disk-On-Chip Millennium-only alternative driver (see help)' CONFIG_MTD_DOC2001 $CONFIG_MTD
    if [ "$CONFIG_MTD_DOC2001" = "y" -o "$CONFIG_MTD_DOC2000" = "y" ]; then
