diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/drivers/md/raid5.c w1/drivers/md/raid5.c
--- w2/drivers/md/raid5.c	2003-08-20 19:13:56.000000000 +0200
+++ w1/drivers/md/raid5.c	2003-08-19 01:41:27.000000000 +0200
@@ -265,7 +265,9 @@ static struct stripe_head *get_active_st
 			 * If someone sets the buffer size to something else, we will need to
 			 * assert that we want to change it again
 			 */
+#if RAID5_DEBUG
 			int oldsize = conf->buffer_size;
+#endif
 			PRINTK("get_stripe %ld/%d buffer_size is %d, %d active\n", sector, size, conf->buffer_size, atomic_read(&conf->active_stripes));
 			if (size==0)
 				wait_event_lock_irq(conf->wait_for_stripe,
