diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff 2.4.19-rc3/drivers/media/video/cpia.h test/drivers/media/video/cpia.h
--- 2.4.19-rc3/drivers/media/video/cpia.h	2002-07-17 11:21:08.000000000 +0200
+++ test/drivers/media/video/cpia.h	2002-07-22 19:06:06.000000000 +0200
@@ -378,8 +378,7 @@ void cpia_unregister_camera(struct cam_d
 /* ErrorCode */
 #define ERROR_FLICKER_BELOW_MIN_EXP     0x01 /*flicker exposure got below minimum exposure */
 
-#define ALOG(lineno,fmt,args...) printk(fmt,lineno,##args)
-#define LOG(fmt,args...) ALOG((__LINE__),KERN_INFO __FILE__":"__FUNCTION__"(%d):"fmt,##args)
+#define LOG(fmt,args...) printk(KERN_INFO __FILE__":%s(%d):"fmt, __FUNCTION__, __LINE__, ##args)
 
 #ifdef _CPIA_DEBUG_
 #define ADBG(lineno,fmt,args...) printk(fmt, jiffies, lineno, ##args)
