diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/drivers/ide/pci/amd74xx.c w1/drivers/ide/pci/amd74xx.c
--- w2/drivers/ide/pci/amd74xx.c	2003-08-20 19:13:42.000000000 +0200
+++ w1/drivers/ide/pci/amd74xx.c	2003-08-19 00:47:57.000000000 +0200
@@ -389,11 +389,6 @@ static unsigned int __init init_chipset_
 	return 0;
 }
 
-static unsigned int __init ata66_amd74xx(ide_hwif_t *hwif)
-{
-	return ((amd_enabled & amd_80w) >> hwif->channel) & 1;
-}
-
 static void __init init_hwif_amd74xx(ide_hwif_t *hwif)
 {
 	int i;
