diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/drivers/net/ns83820.c w1/drivers/net/ns83820.c
--- w2/drivers/net/ns83820.c	2003-08-12 10:25:03.000000000 +0200
+++ w1/drivers/net/ns83820.c	2003-08-20 10:33:03.000000000 +0200
@@ -1585,6 +1585,7 @@ static void ns83820_run_bist(struct ns83
 	dprintk("%s: done %s in %d loops\n", dev->net_dev.name, name, loops);
 }
 
+#ifdef PHY_CODE_IS_FINISHED
 static void ns83820_mii_write_bit(struct ns83820 *dev, int bit)
 {
 	/* drive MDC low */
@@ -1757,6 +1758,7 @@ static void ns83820_probe_phy(struct ns8
 		dprintk("version: 0x%04x 0x%04x\n", a, b);
 	}
 }
+#endif
 
 static int __devinit ns83820_init_one(struct pci_dev *pci_dev, const struct pci_device_id *id)
 {
