diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/drivers/net/tokenring/3c359.c w1/drivers/net/tokenring/3c359.c
--- w2/drivers/net/tokenring/3c359.c	2003-06-13 16:51:35.000000000 +0200
+++ w1/drivers/net/tokenring/3c359.c	2003-08-20 10:48:05.000000000 +0200
@@ -389,7 +389,7 @@ static int xl_hw_reset(struct net_device
     	struct xl_private *xl_priv = (struct xl_private *)dev->priv ;
 	u8 *xl_mmio = xl_priv->xl_mmio ; 
 	unsigned long t ; 
-	u16 i ; 
+	u32 i ; /* we compare agains 0xFFFF */
     	u16 result_16 ; 
 	u8 result_8 ;
 	u16 start ; 
