--- linux-newsplash/drivers/video/fbcon.c.bootlogo-ver	Tue Jan 29 16:21:21 2002
+++ linux-newsplash/drivers/video/fbcon.c	Tue Jan 29 16:21:59 2002
@@ -50,6 +50,8 @@
  *
  *    - Implement 16 plane mode (iplan2p16)
  *
+ *  Alterations marked G.S. contributed by Gerard Sharp (gsharp@ihug.co.nz)
+ *  Essentially a line of text above the Boot Logo stating kernel version. :)
  *
  *  This file is subject to the terms and conditions of the GNU General Public
  *  License.  See the file COPYING in the main directory of this archive for
@@ -58,6 +60,7 @@
 
 #undef FBCONDEBUG
 
+#include <linux/version.h>
 #include <linux/config.h>
 #include <linux/module.h>
 #include <linux/types.h>
@@ -664,7 +667,8 @@
 	int cnt;
 	int step;
     
-    	logo_lines = (LOGO_H + fontheight(p) - 1) / fontheight(p);
+	/* G.S. Make logo 1 (text) line taller - for version text */
+	logo_lines = (LOGO_H + (fontheight(p)*2) - 1) / fontheight(p);
     	q = (unsigned short *)(conp->vc_origin + conp->vc_size_row * old_rows);
     	step = logo_lines * old_cols;
     	for (r = q - logo_lines * old_cols; r < q; r++)
@@ -1553,6 +1557,15 @@
     if (p->dispsw->clear_margins && vt_cons[unit]->vc_mode == KD_TEXT)
 	p->dispsw->clear_margins(conp, p, 0);
     if (logo_shown == -2) {
+	/* G.S.: Display a line above the Boot Logo to state what
+	* version of the kernel we are booting.
+	*/
+	char welcometext[] = "Linux " UTS_RELEASE;
+	unsigned short welcomestring[ sizeof(welcometext) ];
+	int i;
+	for(i = 0; i < sizeof(welcometext); i++)
+	    scr_writew(0x0700 | welcometext[i], &welcomestring[i]);
+	fbcon_putcs( conp, welcomestring, sizeof(welcometext), 0, 0 );
 	logo_shown = fg_console;
 	fbcon_show_logo(); /* This is protected above by initmem_freed */
 	update_region(fg_console,
@@ -2150,6 +2163,8 @@
     if (!fb)
 	return 0;
 	
+    /* G.S. : Move Penguin(s) down one line; for version line */
+    fb += fontheight(p) * line;
     /*
      * Set colors if visual is PSEUDOCOLOR and we have enough colors, or for
      * DIRECTCOLOR
