diff -ruN linux-2.4.19/drivers/md.orig/md.c linux-2.4.19/drivers/md/md.c
--- linux-2.4.19/drivers/md.orig/md.c	2002-09-07 23:22:22.000000000 +0200
+++ linux-2.4.19/drivers/md/md.c	2002-09-07 23:24:18.000000000 +0200
@@ -3515,7 +3515,10 @@
 	mdp_disk_t *spare;
 	struct md_list_head *tmp;
 
-	printk(KERN_INFO "md: recovery thread got woken up ...\n");
+#ifndef MODULE
+	if(console_loglevel > 8)
+#endif
+	  printk(KERN_INFO "md: recovery thread got woken up ...\n");
 restart:
 	ITERATE_MDDEV(mddev,tmp) {
 		sb = mddev->sb;
@@ -3592,7 +3595,10 @@
 		md_update_sb(mddev);
 		goto restart;
 	}
-	printk(KERN_INFO "md: recovery thread finished ...\n");
+#ifndef MODULE
+	if(console_loglevel > 8)
+#endif
+	  printk(KERN_INFO "md: recovery thread finished ...\n");
 
 }
 
@@ -3605,7 +3611,10 @@
 	if ((code == MD_SYS_DOWN) || (code == MD_SYS_HALT)
 					|| (code == MD_SYS_POWER_OFF)) {
 
-		printk(KERN_INFO "md: stopping all md devices.\n");
+#ifndef MODULE
+	        if(console_loglevel > 8)
+#endif
+		  printk(KERN_INFO "md: stopping all md devices.\n");
 
 		ITERATE_MDDEV(mddev,tmp)
 			do_md_stop (mddev, 1);
