diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/drivers/net/wireless/orinoco.c w1/drivers/net/wireless/orinoco.c
--- w2/drivers/net/wireless/orinoco.c	2003-08-07 12:17:47.000000000 +0200
+++ w1/drivers/net/wireless/orinoco.c	2003-08-07 00:10:34.000000000 +0200
@@ -3602,6 +3602,7 @@ static int orinoco_ioctl_getspy(struct n
 	return 0;
 }
 
+static char version[];
 static int netdev_ethtool_ioctl(struct net_device *dev, void *useraddr)
 {
 	u32 ethcmd;
