diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/drivers/char/amd76x_pm.c w1/drivers/char/amd76x_pm.c
--- w2/drivers/char/amd76x_pm.c	2003-08-12 10:25:02.000000000 +0200
+++ w1/drivers/char/amd76x_pm.c	2003-08-19 01:36:05.000000000 +0200
@@ -474,7 +474,7 @@ activate_amd76x_NTH(int enable, int rati
 }
 #endif
 
-
+#ifdef AMD76X_POS
 /*
  * Activate sleep state via its ACPI register (PM1_CNT).
  */
@@ -490,7 +490,6 @@ activate_amd76x_SLP(int type)
 }
 
 
-#ifdef AMD76X_POS
 /*
  * Wrapper function to activate POS sleep state.
  */
