diff -urNp --exclude-from=/home/quintela/config/misc/dontdiff q1/drivers/ide/pci/piix.c q2/drivers/ide/pci/piix.c
--- q1/drivers/ide/pci/piix.c	2003-10-17 21:52:07.000000000 +0200
+++ q2/drivers/ide/pci/piix.c	2003-10-17 23:46:17.000000000 +0200
@@ -878,9 +878,7 @@ static struct pci_device_id piix_pci_tbl
 	{ PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_82801EB_11,PCI_ANY_ID, PCI_ANY_ID, 0, 0, 15},
 	{ PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_82801E_11, PCI_ANY_ID, PCI_ANY_ID, 0, 0, 16},
 	{ PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_82801DB_10,PCI_ANY_ID, PCI_ANY_ID, 0, 0, 17},
-#ifndef CONFIG_SCSI_SATA
  	{ PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_82801EB_1, PCI_ANY_ID, PCI_ANY_ID, 0, 0, 18},
-#endif /* !CONFIG_SCSI_SATA */
 	{ 0, },
 };
 
