diff -uNp q1/drivers/pci/quirks.c.orig q1/drivers/pci/quirks.c
--- q1/drivers/pci/quirks.c.orig	2003-10-18 01:11:41.000000000 +0200
+++ q1/drivers/pci/quirks.c	2003-10-18 01:12:12.000000000 +0200
@@ -713,7 +713,10 @@ static void __init quirk_intel_ide_combi
 	 */
 	switch (pdev->device) {
 	/* PCI ids taken from drivers/scsi/ata_piix.c */
+#if 0
 	case 0x24d1:
+/* we need this to be handled by ide */
+#endif
 	case 0x24df:
 	case 0x25a3:
 	case 0x25b0:
