diff -ru linux-2.4.21-orig/drivers/ieee1394/pcilynx.c linux-2.4.21-i2c/drivers/ieee1394/pcilynx.c
--- linux-2.4.21-orig/drivers/ieee1394/pcilynx.c	2003-06-16 09:30:27.000000000 +0200
+++ linux-2.4.21-i2c/drivers/ieee1394/pcilynx.c	2003-07-10 12:02:22.000000000 +0200
@@ -138,6 +138,7 @@
 }; 
 
 static struct i2c_adapter bit_ops = {
+	.owner                  = THIS_MODULE,
 	.id 			= 0xAA, //FIXME: probably we should get an id in i2c-id.h
 	.client_register	= bit_reg,
 	.client_unregister	= bit_unreg,
