diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w3/drivers/usb/ov511/ov511_core.c w2/drivers/usb/ov511/ov511_core.c
--- w3/drivers/usb/ov511/ov511_core.c	2003-05-11 09:50:50.000000000 +0200
+++ w2/drivers/usb/ov511/ov511_core.c	2003-08-15 04:01:24.000000000 +0200
@@ -52,7 +52,7 @@
 #  include "tuner.h"
 #else
 #  if LINUX_VERSION_CODE < KERNEL_VERSION(2, 5, 0)
-#    include "../media/video/tuner.h"
+#    include "../../media/video/tuner.h"
 #  else
 #    include <media/tuner.h>
 #  endif
