diff -X ../../KDIFX -burpN -X ../../KDIFX linux-vanilla/fs/proc/inode.c linux/fs/proc/inode.c
--- linux-vanilla/fs/proc/inode.c	2002-09-25 01:49:40.000000000 +0200
+++ linux/fs/proc/inode.c	2002-11-25 19:00:21.000000000 +0100
@@ -186,6 +186,7 @@ struct super_block *proc_read_super(stru
 	s->s_blocksize_bits = 10;
 	s->s_magic = PROC_SUPER_MAGIC;
 	s->s_op = &proc_sops;
+	s->s_maxbytes = ~0ULL;
 	
 	root_inode = proc_get_inode(s, PROC_ROOT_INO, &proc_root);
 	if (!root_inode)
