diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff orig/3rdparty/p4b_smbus/p4b_smbus.c q5/3rdparty/p4b_smbus/p4b_smbus.c
--- orig/3rdparty/p4b_smbus/p4b_smbus.c	2002-08-23 14:56:43.000000000 +0200
+++ q5/3rdparty/p4b_smbus/p4b_smbus.c	2002-08-23 13:45:21.000000000 +0200
@@ -60,11 +60,6 @@
 #error ERROR - You must have 'Support for hot-pluggable devices' enabled in your kernel (under 'general setup')!!
 #endif
 
-/* Deal with CONFIG_MODVERSIONS */
-#ifdef CONFIG_MODVERSIONS
-#define MODVERSIONS
-#include <linux/modversions.h>
-#endif
 
 #include <linux/module.h>
 #include <linux/kernel.h>
