diff -up t2/3rdparty/p4b_smbus/p4b_smbus.c.orig t2/3rdparty/p4b_smbus/p4b_smbus.c
--- t2/3rdparty/p4b_smbus/p4b_smbus.c.orig	2002-09-06 11:29:12.000000000 +0200
+++ t2/3rdparty/p4b_smbus/p4b_smbus.c	2002-09-06 11:29:26.000000000 +0200
@@ -88,7 +88,7 @@ rmm    from lm_sensors-2.3.3: 
  */
 
 #define ICH2 PCI_DEVICE_ID_INTEL_82801BA_0
-#define ICH2_SMBUS PCI_DEVICE_ID_INTEL_82801BA_3
+#define ICH2_SMBUS PCI_DEVICE_ID_INTEL_82801BA_2
 
 
 #ifndef PCI_DEVICE_ID_INTEL_82801DB_0
