diff -urNp --exclude-from=/home/quintela/config/misc/dontdiff c2/3rdparty/prism25/include/wlan/wlan_compat.h c1/3rdparty/prism25/include/wlan/wlan_compat.h
--- c2/3rdparty/prism25/include/wlan/wlan_compat.h	2002-11-08 06:20:33.000000000 -0500
+++ c1/3rdparty/prism25/include/wlan/wlan_compat.h	2002-11-22 14:47:02.000000000 -0500
@@ -124,7 +124,7 @@
 #endif
 
 #if defined(__KERNEL__)
-#if defined(__i386__) || defined(__i486__) || defined(__i586__) || defined(__i686__)
+#if defined(__i386__) || defined(__i486__) || defined(__i586__) || defined(__i686__) || defined(__x86_64__)
 	#define WLAN_CPU_FAMILY		WLAN_Ix86
 	#define WLAN_CPU_CORE		WLAN_I386CORE
 	#define WLAN_CPU_PART		WLAN_I386PART
