diff -uNp t1/3rdparty/atmelwlan/src/Pcmcia_Pci/fastvnet_cs.c.orig t1/3rdparty/atmelwlan/src/Pcmcia_Pci/fastvnet_cs.c
--- t1/3rdparty/atmelwlan/src/Pcmcia_Pci/fastvnet_cs.c.orig	2003-02-06 01:50:33.000000000 +0100
+++ t1/3rdparty/atmelwlan/src/Pcmcia_Pci/fastvnet_cs.c	2003-02-06 01:50:41.000000000 +0100
@@ -30,6 +30,7 @@
 #include "vnet.h"
 #include <linux/delay.h>
 #include <linux/wireless.h>
+#include <asm/uaccess.h>
 #include "vnetioctl.h"
 #include "interrupt.h"
 
diff -uNp t1/3rdparty/atmelwlan/src/usb/vnetusba.c.orig t1/3rdparty/atmelwlan/src/usb/vnetusba.c
--- t1/3rdparty/atmelwlan/src/usb/vnetusba.c.orig	2003-02-06 01:51:03.000000000 +0100
+++ t1/3rdparty/atmelwlan/src/usb/vnetusba.c	2003-02-06 01:51:12.000000000 +0100
@@ -29,6 +29,7 @@
 #include <linux/kmod.h>		/* for call_usermodehelper only */
 #include "vnetusba.h"
 #include <linux/string.h>
+#include <asm/uaccess.h>
 #include "vnetioctl.h"
 #include "config.h"
 
