diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/3rdparty/dxr3/adv717x.c w1/3rdparty/dxr3/adv717x.c
--- w2/3rdparty/dxr3/adv717x.c	2003-04-15 07:19:18.000000000 +0200
+++ w1/3rdparty/dxr3/adv717x.c	2003-08-07 11:49:29.000000000 +0200
@@ -41,6 +41,7 @@
 #include <linux/version.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 #include <linux/video_encoder.h>
 
@@ -99,8 +100,6 @@ MODULE_PARM_DESC(color_bars, "If you set
 static int adv717x_attach_adapter(struct i2c_adapter *adapter);
 int adv717x_detach_client(struct i2c_client *client);
 int adv717x_command(struct i2c_client *client, unsigned int cmd, void *arg);
-void adv717x_inc_use (struct i2c_client *client);
-void adv717x_dec_use (struct i2c_client *client);
 
 #define CHIP_ADV7175A 1
 #define CHIP_ADV7170  2
@@ -118,14 +117,13 @@ struct adv717x_data_s {
 
 /* This is the driver that will be inserted */
 static struct i2c_driver adv717x_driver = {
-  /* name */		"ADV717X video encoder driver",
-  /* id */		I2C_DRIVERID_ADV717X,
-  /* flags */		I2C_DF_NOTIFY,
-  /* attach_adapter */  &adv717x_attach_adapter,
-  /* detach_client */   &adv717x_detach_client,
-  /* command */		&adv717x_command,
-  /* inc_use */		&adv717x_inc_use,
-  /* dec_use */		&adv717x_dec_use
+	.owner	= THIS_MODULE,
+	.name	= "ADV717X video encoder driver",
+	.id	= I2C_DRIVERID_ADV717X,
+	.flags	= I2C_DF_NOTIFY,
+	.attach_adapter	= adv717x_attach_adapter,
+	.detach_client	= adv717x_detach_client,
+	.command	= &adv717x_command,
 };
 
 int adv717x_id = 0;
@@ -497,22 +495,6 @@ int adv717x_command(struct i2c_client *c
 	return 0;
 }
 
-/* Nothing here yet */
-void adv717x_inc_use (struct i2c_client *client)
-{
-#ifdef MODULE
-	MOD_INC_USE_COUNT;
-#endif
-}
-
-/* Nothing here yet */
-void adv717x_dec_use (struct i2c_client *client)
-{
-#ifdef MODULE
-	MOD_DEC_USE_COUNT;
-#endif
-}
-
 /* ----------------------------------------------------------------------- */
 
 
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/3rdparty/dxr3/bt865.c w1/3rdparty/dxr3/bt865.c
--- w2/3rdparty/dxr3/bt865.c	2003-04-15 07:19:18.000000000 +0200
+++ w1/3rdparty/dxr3/bt865.c	2003-08-07 11:50:53.000000000 +0200
@@ -42,6 +42,7 @@
 #include <linux/version.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 #include <linux/video_encoder.h>
 
@@ -69,8 +70,6 @@ EXPORT_NO_SYMBOLS;
 static int bt865_attach_adapter(struct i2c_adapter *adapter);
 int bt865_detach_client(struct i2c_client *client);
 int bt865_command(struct i2c_client *client, unsigned int cmd, void *arg);
-void bt865_inc_use (struct i2c_client *client);
-void bt865_dec_use (struct i2c_client *client);
 static int bt865_setup(struct i2c_client *client);
 
 struct bt865_data_s {
@@ -86,14 +85,13 @@ struct bt865_data_s {
 
 /* This is the driver that will be inserted */
 static struct i2c_driver bt865_driver = {
-  /* name */		"BT865 video encoder driver",
-  /* id */		I2C_DRIVERID_BT865,
-  /* flags */		I2C_DF_NOTIFY,
-  /* attach_adapter */  &bt865_attach_adapter,
-  /* detach_client */   &bt865_detach_client,
-  /* command */		&bt865_command,
-  /* inc_use */		&bt865_inc_use,
-  /* dec_use */		&bt865_dec_use
+	.owner	= THIS_MODULE,
+	.name	= "BT865 video encoder driver",
+	.id	= I2C_DRIVERID_BT865,
+	.flags	= I2C_DF_NOTIFY,
+	.attach_adapter	= bt865_attach_adapter,
+	.detach_client	= bt865_detach_client,
+	.command	= bt865_command,
 };
 
 int bt865_id = 0;
@@ -961,22 +959,6 @@ int bt865_command(struct i2c_client *cli
 	return 0;
 }
 
-/* Nothing here yet */
-void bt865_inc_use (struct i2c_client *client)
-{
-#ifdef MODULE
-	MOD_INC_USE_COUNT;
-#endif
-}
-
-/* Nothing here yet */
-void bt865_dec_use (struct i2c_client *client)
-{
-#ifdef MODULE
-	MOD_DEC_USE_COUNT;
-#endif
-}
-
 /* ----------------------------------------------------------------------- */
 
 
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/3rdparty/dxr3/em8300_audio.c w1/3rdparty/dxr3/em8300_audio.c
--- w2/3rdparty/dxr3/em8300_audio.c	2003-04-15 07:19:18.000000000 +0200
+++ w1/3rdparty/dxr3/em8300_audio.c	2003-08-07 11:42:09.000000000 +0200
@@ -15,6 +15,7 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_reg.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/3rdparty/dxr3/em8300_dicom.c w1/3rdparty/dxr3/em8300_dicom.c
--- w2/3rdparty/dxr3/em8300_dicom.c	2003-04-15 07:19:18.000000000 +0200
+++ w1/3rdparty/dxr3/em8300_dicom.c	2003-08-07 11:42:12.000000000 +0200
@@ -14,6 +14,7 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_reg.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/3rdparty/dxr3/em8300_fifo.c w1/3rdparty/dxr3/em8300_fifo.c
--- w2/3rdparty/dxr3/em8300_fifo.c	2003-04-15 07:19:18.000000000 +0200
+++ w1/3rdparty/dxr3/em8300_fifo.c	2003-08-07 11:42:15.000000000 +0200
@@ -41,6 +41,7 @@
 #include <linux/version.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_reg.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/3rdparty/dxr3/em8300_i2c.c w1/3rdparty/dxr3/em8300_i2c.c
--- w2/3rdparty/dxr3/em8300_i2c.c	2003-04-15 07:19:18.000000000 +0200
+++ w1/3rdparty/dxr3/em8300_i2c.c	2003-08-07 11:47:24.000000000 +0200
@@ -14,6 +14,7 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_reg.h"
@@ -116,14 +116,6 @@ static int em8300_i2c_unreg(struct i2c_c
 	return 0;
 }
 
-static void em8300_i2c_inc(struct i2c_adapter *adapter)
-{	
-}
-
-static void em8300_i2c_dec(struct i2c_adapter *adapter)
-{	
-}
-
 /* ----------------------------------------------------------------------- */
 /* I2C functions							   */
 /* ----------------------------------------------------------------------- */
@@ -172,12 +173,11 @@ int em8300_i2c_init(struct em8300_s *em)
 	
 	em->i2c_data_1.data = pdata;
 
+	em->i2c_ops_1.owner = THIS_MODULE;
 	strcpy(em->i2c_ops_1.name, "EM8300 I2C bus 1");
 	em->i2c_ops_1.id = I2C_HW_B_EM8300;
 	em->i2c_ops_1.algo = NULL;
 	em->i2c_ops_1.algo_data = &em->i2c_data_1;
-	em->i2c_ops_1.inc_use = em8300_i2c_inc;
-	em->i2c_ops_1.dec_use = em8300_i2c_dec;
 	em->i2c_ops_1.client_register = em8300_i2c_reg;
 	em->i2c_ops_1.client_unregister = em8300_i2c_unreg;
 	em->i2c_ops_1.data = em;
@@ -207,12 +207,11 @@ int em8300_i2c_init(struct em8300_s *em)
 	
 	em->i2c_data_2.data = pdata;
 
+	em->i2c_ops_2.owner = THIS_MODULE;
 	strcpy(em->i2c_ops_2.name, "EM8300 I2C bus 2");
 	em->i2c_ops_2.id = I2C_HW_B_EM8300;
 	em->i2c_ops_2.algo = NULL;
 	em->i2c_ops_2.algo_data = &em->i2c_data_2;
-	em->i2c_ops_2.inc_use = em8300_i2c_inc;
-	em->i2c_ops_2.dec_use = em8300_i2c_dec;
 	em->i2c_ops_2.client_register = em8300_i2c_reg;
 	em->i2c_ops_2.client_unregister = em8300_i2c_unreg;
 	em->i2c_ops_2.data = em;
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/3rdparty/dxr3/em8300_ioctl.c w1/3rdparty/dxr3/em8300_ioctl.c
--- w2/3rdparty/dxr3/em8300_ioctl.c	2003-04-15 07:19:18.000000000 +0200
+++ w1/3rdparty/dxr3/em8300_ioctl.c	2003-08-07 11:43:05.000000000 +0200
@@ -14,6 +14,7 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_reg.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/3rdparty/dxr3/em8300_main.c w1/3rdparty/dxr3/em8300_main.c
--- w2/3rdparty/dxr3/em8300_main.c	2003-04-16 03:26:13.000000000 +0200
+++ w1/3rdparty/dxr3/em8300_main.c	2003-08-07 11:43:10.000000000 +0200
@@ -54,6 +54,7 @@
 #endif
 
 #include <asm/uaccess.h>
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_version.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/3rdparty/dxr3/em8300_misc.c w1/3rdparty/dxr3/em8300_misc.c
--- w2/3rdparty/dxr3/em8300_misc.c	2003-04-15 07:19:18.000000000 +0200
+++ w1/3rdparty/dxr3/em8300_misc.c	2003-08-07 11:43:14.000000000 +0200
@@ -14,6 +14,7 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_reg.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/3rdparty/dxr3/em8300_spu.c w1/3rdparty/dxr3/em8300_spu.c
--- w2/3rdparty/dxr3/em8300_spu.c	2003-04-15 07:19:18.000000000 +0200
+++ w1/3rdparty/dxr3/em8300_spu.c	2003-08-07 11:43:18.000000000 +0200
@@ -14,6 +14,7 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_reg.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/3rdparty/dxr3/em8300_ucode.c w1/3rdparty/dxr3/em8300_ucode.c
--- w2/3rdparty/dxr3/em8300_ucode.c	2003-04-15 07:19:18.000000000 +0200
+++ w1/3rdparty/dxr3/em8300_ucode.c	2003-08-07 11:43:22.000000000 +0200
@@ -14,6 +14,7 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_reg.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/3rdparty/dxr3/em8300_video.c w1/3rdparty/dxr3/em8300_video.c
--- w2/3rdparty/dxr3/em8300_video.c	2003-04-15 07:19:18.000000000 +0200
+++ w1/3rdparty/dxr3/em8300_video.c	2003-08-07 11:43:26.000000000 +0200
@@ -14,6 +14,7 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_reg.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/3rdparty/dxr3/em9010.c w1/3rdparty/dxr3/em9010.c
--- w2/3rdparty/dxr3/em9010.c	2003-04-15 07:19:18.000000000 +0200
+++ w1/3rdparty/dxr3/em9010.c	2003-08-07 11:43:30.000000000 +0200
@@ -14,6 +14,7 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 #include <linux/time.h>
 
