diff -urNp --exclude-from=/home/quintela/config/misc/dontdiff t2/arch/i386/config.in t1/arch/i386/config.in
--- t2/arch/i386/config.in	2003-03-14 14:26:30.000000000 +0100
+++ t1/arch/i386/config.in	2003-03-14 14:24:31.000000000 +0100
@@ -456,6 +456,9 @@ if [ "$CONFIG_VT" = "y" ]; then
    endmenu
 fi
 
+# Alsa needs usb
+source drivers/usb/Config.in
+
 mainmenu_option next_comment
 comment 'Sound'
 
@@ -465,8 +468,6 @@ if [ "$CONFIG_SOUND" != "n" ]; then
 fi
 endmenu
 
-source drivers/usb/Config.in
-
 source net/bluetooth/Config.in
 
 mainmenu_option next_comment
