--- linux/fs/super.c.org	Thu Sep 13 10:05:24 2001
+++ linux/fs/super.c	Thu Sep 13 10:11:09 2001
@@ -268,6 +268,21 @@
 			fs = NULL;
 		read_unlock(&file_systems_lock);
 	}
+	if (!fs) {	
+		char *foo;
+		foo = kmalloc(strlen(name)+6,GFP_NOFS);
+		if (!foo) 
+			return NULL;
+		sprintf(foo,"free%s",name);
+		if (request_module(foo) == 0) {
+			read_lock(&file_systems_lock);
+			fs = *(find_filesystem(name));
+			if (fs && !try_inc_mod_count(fs->owner))   
+				fs = NULL;
+			read_unlock(&file_systems_lock);
+		}
+		kfree(foo);
+	}
 	return fs;
 }
 
