--- postfix-2.3.4/conf/postfix-files.config	2006-07-19 19:01:07.000000000 -0600
+++ postfix-2.3.4/conf/postfix-files	2006-12-09 19:49:22.000000000 -0700
@@ -101,8 +101,6 @@
 $sendmail_path:f:root:-:755
 $newaliases_path:l:$sendmail_path
 $mailq_path:l:$sendmail_path
-$config_directory/LICENSE:f:root:-:644
-$config_directory/TLS_LICENSE:f:root:-:644
 $config_directory/access:f:root:-:644:p
 $config_directory/aliases:f:root:-:644:p
 $config_directory/bounce.cf.default:f:root:-:644
@@ -113,6 +111,7 @@
 $config_directory/header_checks:f:root:-:644:p
 $config_directory/install.cf:f:root:-:644:o
 $config_directory/main.cf.default:f:root:-:644
+$config_directory/main.cf.dist:f:root:-:644
 $config_directory/main.cf:f:root:-:644:p
 $config_directory/makedefs.out:f:root:-:644
 $config_directory/master.cf:f:root:-:644:p
@@ -205,6 +204,7 @@
 $sample_directory/sample-pcre-access.cf:f:root:-:644:o:o
 $sample_directory/sample-pcre-body.cf:f:root:-:644:o:o
 $sample_directory/sample-pcre-header.cf:f:root:-:644:o:o
+$sample_directory/sample-pgsql.cf:f:root:-:644:o
 $sample_directory/sample-pgsql-aliases.cf:f:root:-:644:o
 $sample_directory/sample-qmqpd.cf:f:root:-:644:o
 $sample_directory/sample-rate.cf:f:root:-:644:o
@@ -255,7 +255,6 @@
 $readme_directory/PGSQL_README:f:root:-:644
 $readme_directory/QMQP_README:f:root:-:644
 $readme_directory/QSHAPE_README:f:root:-:644
-$readme_directory/RELEASE_NOTES:f:root:-:644
 $readme_directory/RESTRICTION_CLASS_README:f:root:-:644
 $readme_directory/SASL_README:f:root:-:644
 $readme_directory/SCHEDULER_README:f:root:-:644
--- postfix-2.3.4/conf/main.cf.config	2006-03-24 19:38:54.000000000 -0700
+++ postfix-2.3.4/conf/main.cf	2006-12-09 19:47:43.000000000 -0700
@@ -39,7 +39,7 @@
 # daemon programs (i.e. programs listed in the master.cf file). This
 # directory must be owned by root.
 #
-daemon_directory = /usr/libexec/postfix
+daemon_directory = /usr/lib/postfix
 
 # QUEUE AND PROCESS OWNERSHIP
 #
@@ -104,6 +104,7 @@
 #
 # Note: you need to stop/start Postfix when this parameter changes.
 #
+inet_interfaces = localhost
 #inet_interfaces = all
 #inet_interfaces = $myhostname
 #inet_interfaces = $myhostname, localhost
@@ -151,7 +152,7 @@
 #
 # See also below, section "REJECTING MAIL FOR UNKNOWN LOCAL USERS".
 #
-#mydestination = $myhostname, localhost.$mydomain, localhost
+mydestination = $myhostname, localhost.$mydomain, localhost
 #mydestination = $myhostname, localhost.$mydomain, localhost, $mydomain
 #mydestination = $myhostname, localhost.$mydomain, localhost, $mydomain,
 #	mail.$mydomain, www.$mydomain, ftp.$mydomain
@@ -238,7 +239,7 @@
 # 
 #mynetworks_style = class
 #mynetworks_style = subnet
-#mynetworks_style = host
+mynetworks_style = host
 
 # Alternatively, you can specify the mynetworks list by hand, in
 # which case Postfix ignores the mynetworks_style setting.
@@ -373,7 +374,7 @@
 # "postfix reload" to eliminate the delay.
 #
 #alias_maps = dbm:/etc/aliases
-#alias_maps = hash:/etc/aliases
+alias_maps = hash:/etc/aliases
 #alias_maps = hash:/etc/aliases, nis:mail.aliases
 #alias_maps = netinfo:/aliases
 
@@ -413,7 +414,7 @@
 # system type.
 #
 #mail_spool_directory = /var/mail
-#mail_spool_directory = /var/spool/mail
+mail_spool_directory = /var/spool/mail
 
 # The mailbox_command parameter specifies the optional external
 # command to use instead of mailbox delivery. The command is run as
--- postfix-2.3.4/conf/master.cf.config	2006-03-18 07:40:01.000000000 -0700
+++ postfix-2.3.4/conf/master.cf	2006-12-09 19:47:43.000000000 -0700
@@ -51,29 +51,6 @@
 # and other message envelope options.
 # ====================================================================
 #
-# maildrop. See the Postfix MAILDROP_README file for details.
-# Also specify in main.cf: maildrop_destination_recipient_limit=1
-#
-maildrop  unix  -       n       n       -       -       pipe
-  flags=DRhu user=vmail argv=/usr/local/bin/maildrop -d ${recipient}
-#
-# The Cyrus deliver program has changed incompatibly, multiple times.
-#
-old-cyrus unix  -       n       n       -       -       pipe
-  flags=R user=cyrus argv=/cyrus/bin/deliver -e -m ${extension} ${user}
-# Cyrus 2.1.5 (Amos Gouaux)
-# Also specify in main.cf: cyrus_destination_recipient_limit=1
-cyrus     unix  -       n       n       -       -       pipe
-  user=cyrus argv=/cyrus/bin/deliver -e -r ${sender} -m ${extension} ${user}
-#
-# See the Postfix UUCP_README file for configuration details.
-#
-uucp      unix  -       n       n       -       -       pipe
-  flags=Fqhu user=uucp argv=uux -r -n -z -a$sender - $nexthop!rmail ($recipient)
 #
 # Other external delivery methods.
 #
-ifmail    unix  -       n       n       -       -       pipe
-  flags=F user=ftn argv=/usr/lib/ifmail/ifmail -r $nexthop ($recipient)
-bsmtp     unix  -       n       n       -       -       pipe
-  flags=Fq. user=foo argv=/usr/local/sbin/bsmtp -f $sender $nexthop $recipient
--- postfix-2.3.4/src/global/mail_params.h.config	2006-09-03 13:21:00.000000000 -0600
+++ postfix-2.3.4/src/global/mail_params.h	2006-12-09 19:47:43.000000000 -0700
@@ -427,11 +427,7 @@
   * Local delivery: alias databases.
   */
 #define VAR_ALIAS_MAPS		"alias_maps"
-#ifdef HAS_NIS
-#define DEF_ALIAS_MAPS		ALIAS_DB_MAP ", nis:mail.aliases"
-#else
 #define DEF_ALIAS_MAPS		ALIAS_DB_MAP
-#endif
 extern char *var_alias_maps;
 
  /*
--- postfix-2.3.4/src/util/sys_defs.h.config	2006-09-28 16:55:47.000000000 -0600
+++ postfix-2.3.4/src/util/sys_defs.h	2006-12-09 19:47:43.000000000 -0700
@@ -659,7 +659,7 @@
 #define HAS_FSYNC
 #define HAS_DB
 #define DEF_DB_TYPE	"hash"
-#define ALIAS_DB_MAP	"hash:/etc/aliases"
+#define ALIAS_DB_MAP	"hash:/etc/postfix/aliases"
 #define HAS_NIS
 #define GETTIMEOFDAY(t)	gettimeofday(t,(struct timezone *) 0)
 #define ROOT_PATH	"/bin:/usr/bin:/sbin:/usr/sbin"
@@ -673,7 +673,7 @@
 #define NATIVE_MAILQ_PATH "/usr/bin/mailq"
 #define NATIVE_NEWALIAS_PATH "/usr/bin/newaliases"
 #define NATIVE_COMMAND_DIR "/usr/sbin"
-#define NATIVE_DAEMON_DIR "/usr/libexec/postfix"
+#define NATIVE_DAEMON_DIR "/usr/lib/postfix"
 #if __GLIBC__ >= 2 && __GLIBC_MINOR__ >= 1
 #define SOCKADDR_SIZE	socklen_t
 #define SOCKOPT_SIZE	socklen_t
@@ -707,7 +707,7 @@
 #define HAS_FSYNC
 #define HAS_DB
 #define DEF_DB_TYPE	"hash"
-#define ALIAS_DB_MAP	"hash:/etc/aliases"
+#define ALIAS_DB_MAP	"hash:/etc/postfix/aliases"
 #define HAS_NIS
 #define GETTIMEOFDAY(t)	gettimeofday(t,(struct timezone *) 0)
 #define ROOT_PATH	"/bin:/usr/bin:/sbin:/usr/sbin"
@@ -763,7 +763,7 @@
 #define NATIVE_MAILQ_PATH "/usr/bin/mailq"
 #define NATIVE_NEWALIAS_PATH "/usr/bin/newaliases"
 #define NATIVE_COMMAND_DIR "/usr/sbin"
-#define NATIVE_DAEMON_DIR "/usr/libexec/postfix"
+#define NATIVE_DAEMON_DIR "/usr/lib/postfix"
 #endif
 
 #ifdef HPUX10
