--- Mesa-3.3/glx/servGL/serverglx/glx_log.c.chmou	Fri Feb 25 09:08:21 2000
+++ Mesa-3.3/glx/servGL/serverglx/glx_log.c	Mon Oct 16 01:35:30 2000
@@ -137,7 +137,7 @@
          switch (*sptr) {
             case 'f':		/* float */
                {
-                  float f = va_arg(ap, float);
+                  double f = va_arg(ap, double);
                   fprintf(fPtr, "%1.2f", f);
                   break;
                }
@@ -179,7 +179,7 @@
                }
             case 'c':		/* char */
                {
-                  char c = va_arg(ap, char);
+                  int c = va_arg(ap, int);
                   fprintf(fPtr, "%c", c);
                   break;
                }
