diff -u linux/arch/i386/config.in.chmou linux/arch/i386/config.in
--- linux/arch/i386/config.in.chmou	Thu Jan 24 13:42:10 2002
+++ linux/arch/i386/config.in	Thu Jan 24 13:42:12 2002
@@ -407,7 +407,7 @@
 
 mainmenu_option next_comment
 comment 'Kernel hacking'
-
+bool 'Make the kernel for install system (only useful for distro makers)' CONFIG_BOOT_KERNEL
 bool 'Kernel debugging' CONFIG_DEBUG_KERNEL
 if [ "$CONFIG_DEBUG_KERNEL" != "n" ]; then
    bool '  Debug high memory support' CONFIG_DEBUG_HIGHMEM
diff -u linux/arch/i386/Makefile.chmou linux/arch/i386/Makefile
--- linux/arch/i386/Makefile.chmou	Thu Jan 24 13:31:21 2002
+++ linux/arch/i386/Makefile	Thu Jan 24 13:43:06 2002
@@ -86,6 +86,10 @@
 CFLAGS += -march=i586
 endif
 
+ifdef CONFIG_BOOT_KERNEL
+CFLAGS += -Os -falign-functions=4
+endif
+
 HEAD := arch/i386/kernel/head.o arch/i386/kernel/init_task.o
 
 SUBDIRS += arch/i386/kernel arch/i386/mm arch/i386/lib
