diff -urNp --exclude-from=/home/quintela/config/misc/dontdiff u2/drivers/usb/w9968cf.c u1/drivers/usb/w9968cf.c
--- u2/drivers/usb/w9968cf.c	2003-11-11 09:52:51.000000000 +0100
+++ u1/drivers/usb/w9968cf.c	2003-11-13 14:46:45.000000000 +0100
@@ -441,9 +441,6 @@ static int w9968cf_i2c_attach_inform(str
 static int w9968cf_i2c_detach_inform(struct i2c_client*);
 static int w9968cf_i2c_control(struct i2c_adapter*, unsigned int cmd,
                                unsigned long arg);
-static void w9968cf_i2c_inc_use(struct i2c_adapter*);
-static void w9968cf_i2c_dec_use(struct i2c_adapter*);
-
 /* Memory management */
 static inline unsigned long kvirt_to_pa(unsigned long adr);
 static void* rvmalloc(unsigned long size);
@@ -1832,17 +1829,6 @@ w9968cf_i2c_control(struct i2c_adapter* 
 }
 
 
-static void w9968cf_i2c_inc_use(struct i2c_adapter* adap)
-{
-	MOD_INC_USE_COUNT;
-}
-
-
-static void w9968cf_i2c_dec_use(struct i2c_adapter* adap)
-{
-	MOD_DEC_USE_COUNT;
-}
-
 
 static int w9968cf_i2c_init(struct w9968cf_device* cam)
 {
@@ -1857,9 +1846,8 @@ static int w9968cf_i2c_init(struct w9968
 	};
 
 	static struct i2c_adapter adap = {
+		.owner = 	     THIS_MODULE,
 		.id =                I2C_ALGO_SMBUS | I2C_HW_SMBUS_W9968CF,
-		.inc_use =           w9968cf_i2c_inc_use,
-		.dec_use =           w9968cf_i2c_dec_use,
 		.client_register =   w9968cf_i2c_attach_inform,
 		.client_unregister = w9968cf_i2c_detach_inform,
 		.algo =              &algo,
