diff -uNp c1/include/asm-x86_64/bitops.h.orig c1/include/asm-x86_64/bitops.h
--- c1/include/asm-x86_64/bitops.h.orig	2002-11-22 13:00:59.000000000 +0100
+++ c1/include/asm-x86_64/bitops.h	2002-11-22 13:01:18.000000000 +0100
@@ -355,6 +355,13 @@ static __inline__ unsigned long ffz(unsi
 	return word;
 }
 
+/*
+ * fls: find last bit set.
+ */
+
+#define fls(x) generic_fls(x)
+
+
 #ifdef __KERNEL__
 
 /**
