--- chkconfig-1.3.4/chkconfig.c.chmou	Sat Sep 23 14:35:50 2000
+++ chkconfig-1.3.4/chkconfig.c	Sat Sep 23 14:37:20 2000
@@ -63,7 +63,11 @@
 	readServiceError(rc, name);
 	return 1;
     }
-    if (s.type == TYPE_XINETD) return 0;
+    if (s.type == TYPE_XINETD) {
+            setXinetdService(s, 0);
+            system("/sbin/service xinetd reload >/dev/null 2>&1");
+            return 0;
+    }
 
     for (level = 0; level < 7; level++) {
 	    if (!findServiceEntries(name, level, &globres)) {
@@ -181,7 +185,11 @@
 	return 1;
     }
 	
-    if (s.type == TYPE_XINETD) return 0;
+    if (s.type == TYPE_XINETD) {
+            setXinetdService(s, 1);
+            system("/sbin/service xinetd reload >/dev/null 2>&1");
+            return 0;
+    }
     if (s.isLSB && (s.sPriority == s.kPriority == -1))
 		frobDependencies(&s);
     
