--- xpm-3.4k/lib/Imakefile.s_popen	2004-11-29 09:46:52.000000000 -0700
+++ xpm-3.4k/lib/Imakefile	2004-11-29 09:46:52.000000000 -0700
@@ -110,7 +110,8 @@
 	 CrBufFrI.o CrDatFrP.o CrPFrBuf.o RdFToI.o WrFFrI.o \
 	 CrBufFrP.o CrIFrBuf.o CrPFrDat.o RdFToP.o WrFFrP.o \
 	 CrDatFrI.o CrIFrDat.o RdFToDat.o WrFFrDat.o \
-	 Attrib.o CrIFrP.o CrPFrI.o Image.o Info.o RdFToBuf.o WrFFrBuf.o
+	 Attrib.o CrIFrP.o CrPFrI.o Image.o Info.o RdFToBuf.o WrFFrBuf.o \
+	 s_popen.o
 
        INCLUDES = -I.
        LINTLIBS = $(LINTXTOLL) $(LINTXLIB) 
--- xpm-3.4k/lib/WrFFrI.c.s_popen	2004-11-29 09:58:51.000000000 -0700
+++ xpm-3.4k/lib/WrFFrI.c	2004-11-29 10:01:19.000000000 -0700
@@ -326,10 +326,7 @@
 #ifndef NO_ZPIPE
 	size_t len = strlen(filename);
 
-	if(len == 0                        ||
-	   filename[0] == '/'              ||
-	   strstr(filename, "../") != NULL ||
-	   filename[len-1] == '/')
+	if(len == 0)
 		return(XpmOpenFailed);
 
 	if (len > 2 && !strcmp(".Z", filename + (len - 2))) {
