Index: squid/src/main.c
diff -c squid/src/main.c:1.345.2.13 squid/src/main.c:1.345.2.15
*** squid/src/main.c:1.345.2.13	Wed Dec 17 14:10:30 2003
--- squid/src/main.c	Tue Dec 28 05:54:35 2004
***************
*** 586,592 ****
  main(int argc, char **argv)
  {
      int errcount = 0;
-     int n;			/* # of GC'd objects */
      int loop_delay;
      mode_t oldmask;
  #if defined(_SQUID_MSWIN_) || defined(_SQUID_CYGWIN_)
--- 586,591 ----
***************
*** 701,710 ****
  	watch_child(argv);
      setMaxFD();
  
-     if (opt_catch_signals)
- 	for (n = Squid_MaxFD; n > 2; n--)
- 	    close(n);
- 
      /* init comm module */
      comm_init();
      comm_select_init();
--- 700,705 ----
***************
*** 892,900 ****
  	dup2(nullfd, 1);
  	dup2(nullfd, 2);
      }
-     /* Close all else */
-     for (i = 3; i < Squid_MaxFD; i++)
- 	close(i);
      for (;;) {
  	mainStartScript(argv[0]);
  	if ((pid = fork()) == 0) {
--- 887,892 ----
