--- gmp-4.1.3/mpn/x86_64/gmp-mparam.h.x86_64	2004-06-17 10:32:46.729621649 +0200
+++ gmp-4.1.3/mpn/x86_64/gmp-mparam.h	2004-06-17 10:32:46.729621649 +0200
@@ -0,0 +1,27 @@
+/* gmp-mparam.h -- Compiler/machine parameter header file.
+
+Copyright (C) 1991, 1993, 1994 Free Software Foundation, Inc.
+
+This file is part of the GNU MP Library.
+
+The GNU MP Library is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or (at your
+option) any later version.
+
+The GNU MP Library is distributed in the hope that it will be useful, but
+WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
+License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with the GNU MP Library; see the file COPYING.LIB.  If not, write to
+the Free Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
+MA 02111-1307, USA. */
+
+#define BITS_PER_MP_LIMB 64
+#define BYTES_PER_MP_LIMB 8
+#define BITS_PER_LONGINT 64
+#define BITS_PER_INT 32
+#define BITS_PER_SHORTINT 16
+#define BITS_PER_CHAR 8
--- gmp-4.1.3/longlong.h.x86_64	2004-04-22 19:03:00.000000000 +0200
+++ gmp-4.1.3/longlong.h	2004-06-17 10:32:46.731621684 +0200
@@ -708,7 +708,7 @@ extern UWtype __MPN(udiv_qrnnd) _PROTO (
 #endif
 #endif /* 80x86 */
 
-#if defined (__x86_64__) && W_TYPE_SIZE == 64
+#if defined (__x86_64__) && !defined (NO_ASM) && W_TYPE_SIZE == 64
 #define add_ssaaaa(sh, sl, ah, al, bh, bl) \
   __asm__ ("addq %5,%1\n\tadcq %3,%0"					\
 	   : "=r" ((UDItype)(sh)), "=&r" ((UDItype)(sl))		\
@@ -723,10 +723,12 @@ extern UWtype __MPN(udiv_qrnnd) _PROTO (
   __asm__ ("mulq %3"							\
 	   : "=a" (w0), "=d" (w1)					\
 	   : "%0" ((UDItype)(u)), "rm" ((UDItype)(v)))
+#if 0
 #define udiv_qrnnd(q, r, n1, n0, dx) /* d renamed to dx avoiding "=d" */\
   __asm__ ("divq %4"		     /* stringification in K&R C */	\
 	   : "=a" (q), "=d" (r)						\
 	   : "0" ((UDItype)(n0)), "1" ((UDItype)(n1)), "rm" ((UDItype)(dx)))
+#endif
 #define count_leading_zeros(count, x)					\
   do {									\
     UDItype __cbtmp;							\
--- gmp-4.1.3/configure.x86_64	2004-04-28 02:29:27.000000000 +0200
+++ gmp-4.1.3/configure	2004-06-17 10:32:46.747621969 +0200
@@ -2186,6 +2186,10 @@ echo "include_mpn(\`arm/arm-defs.m4')" >
     path="i960"
     ;;
 
+  x86_64*-*-*)
+    path="x86_64"
+    ;;
+
   ia64*-*-*)
     path="ia64"
 
