--- expect5.32/configure.libdir	2000-08-29 05:58:00.000000000 +0200
+++ expect5.32/configure	2002-06-27 19:58:34.000000000 +0200
@@ -1132,7 +1132,7 @@ if test $ac_cv_sys_long_file_names = no;
 fi
 
 EXP_BUILD_LIB_SPEC="-L`pwd` -lexpect${EXP_LIB_VERSION}${DBGX}"
-EXP_LIB_SPEC="-L\${INSTALL_ROOT}\${exec_prefix}/lib -lexpect${EXP_LIB_VERSION}${DBGX}"
+EXP_LIB_SPEC="-L\${INSTALL_ROOT}\${libdir} -lexpect${EXP_LIB_VERSION}${DBGX}"
 EXP_UNSHARED_LIB_FILE=libexpect${EXP_LIB_VERSION}${DBGX}.a
 
 # The TCL_SHARED_LIB_SUFFIX macro below relies on the DBGX macro,
@@ -1186,6 +1186,7 @@ LIB_RUNTIME_DIR='${LIB_RUNTIME_DIR}'
 # If Tcl and Expect are installed in different places, adjust the library
 # search path to reflect this.
 
+# (gb) don't care about those ones?
 if test x"$no_tk" = x"true" ; then
   if test "$TCL_EXEC_PREFIX" != "$exec_prefix"; then
     LIB_RUNTIME_DIR="${LIB_RUNTIME_DIR}:${TCL_EXEC_PREFIX}/lib"
