diff -BrauN chpax-0.5/Changelog chpax-0.5-omer/Changelog
--- chpax-0.5/Changelog	2003-10-04 18:42:23.000000000 -0500
+++ chpax-0.5-omer/Changelog	1969-12-31 18:00:00.000000000 -0600
@@ -1,15 +0,0 @@
- *
- * Nov 2002 : Added multi{options,files} cmdline, zeroflag, nicer output 
- * (+ double output if flags are changed and -v is specified), more error 
- * handling, more explicit error messages and return values
- * 
- * Jan 2003 : Packaging, Port to Sparc/ELF64 : flags now stand in e_ident[EI_PAX] and 
- * e_ident[EI_PAX + 1], old chpax version is needed for clearing old flags .
- *
- * Feb 2003: Added alpha support.
- *
- * Mar 2003: Added parisc support.
- *
- * Apr 2003: Added ppc support.
- *
- * Oct 2003: Added amd64, ia64 and make install support, short format flags printout.
diff -BrauN chpax-0.5/ChangeLog chpax-0.5-omer/ChangeLog
--- chpax-0.5/ChangeLog	1969-12-31 18:00:00.000000000 -0600
+++ chpax-0.5-omer/ChangeLog	2003-10-04 18:42:23.000000000 -0500
@@ -0,0 +1,15 @@
+ *
+ * Nov 2002 : Added multi{options,files} cmdline, zeroflag, nicer output 
+ * (+ double output if flags are changed and -v is specified), more error 
+ * handling, more explicit error messages and return values
+ * 
+ * Jan 2003 : Packaging, Port to Sparc/ELF64 : flags now stand in e_ident[EI_PAX] and 
+ * e_ident[EI_PAX + 1], old chpax version is needed for clearing old flags .
+ *
+ * Feb 2003: Added alpha support.
+ *
+ * Mar 2003: Added parisc support.
+ *
+ * Apr 2003: Added ppc support.
+ *
+ * Oct 2003: Added amd64, ia64 and make install support, short format flags printout.
diff -BrauN chpax-0.5/chpax.h chpax-0.5-omer/chpax.h
--- chpax-0.5/chpax.h	2003-10-04 18:42:23.000000000 -0500
+++ chpax-0.5-omer/chpax.h	2003-12-24 16:00:16.000000000 -0600
@@ -4,6 +4,24 @@
  * The PaX project : http://pageexec.virtualave.net/
  * 
  */
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#else
+#  error "you need to run ./configure before compiling"
+#endif
+#if (! defined HAVE_FCNTL_H) || \
+    (! defined HAVE_STDLIB_H) || \
+    (! defined HAVE_STRING_H) || \
+    (! defined HAVE_SYS_TYPES_H) || \
+    (! defined HAVE_UNISTD_H) || \
+    (! defined STDC_HEADERS)
+#  warning "you are missing at least one standard header; the build will fail"
+#endif
+#if (! defined HAVE_LINUX_A_OUT_H) || \
+    (! defined HAVE_LINUX_ELF_H)
+#  warning "you are missing at least one Linux header; the build will fail"
+#endif
+
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
@@ -13,7 +32,7 @@
 #include <linux/elf.h>
 #include <linux/a.out.h>
 
-#define	CHPAX_VERSION		"0.5"
+#define	CHPAX_VERSION		PACKAGE_VERSION
 
 #define HF_PAX_PAGEEXEC         1    /* 0: Paging based non-exec pages */
 #define HF_PAX_EMUTRAMP         2    /* 0: Emulate trampolines */
diff -BrauN chpax-0.5/configure.in chpax-0.5-omer/configure.in
--- chpax-0.5/configure.in	1969-12-31 18:00:00.000000000 -0600
+++ chpax-0.5-omer/configure.in	2003-12-24 16:41:20.000000000 -0600
@@ -0,0 +1,27 @@
+#                                               -*- Autoconf -*-
+# Process this file with autoconf to produce a configure script.
+
+AC_PREREQ(2.57)
+AC_INIT(chpax, 0.5, chpax-autotools@omershenker.net)
+AC_CONFIG_SRCDIR([aout.c])
+AM_CONFIG_HEADER([config.h])
+AC_CANONICAL_SYSTEM
+AM_INIT_AUTOMAKE([1.7 foreign dist-bzip2])
+
+# Checks for programs.
+AC_PROG_CC
+AC_PROG_INSTALL
+
+# Checks for libraries.
+
+# Checks for header files.
+AC_HEADER_STDC
+AC_CHECK_HEADERS([fcntl.h stdlib.h string.h unistd.h linux/elf.h linux/a.out.h])
+
+# Checks for typedefs, structures, and compiler characteristics.
+
+# Checks for library functions.
+AC_FUNC_MEMCMP
+
+AC_CONFIG_FILES([Makefile])
+AC_OUTPUT
diff -BrauN chpax-0.5/Makefile chpax-0.5-omer/Makefile
--- chpax-0.5/Makefile	2003-10-04 18:41:51.000000000 -0500
+++ chpax-0.5-omer/Makefile	1969-12-31 18:00:00.000000000 -0600
@@ -1,33 +0,0 @@
-##
-## Makefile for chpax.c
-##
-## The PaX project : http://pageexec.virtualave.net/
-##
-
-SRC	= chpax.c io.c elf32.c elf64.c aout.c flags.c
-OBJ	= $(SRC:.c=.o)
-RM	= rm -f 
-TARGET	= chpax
-CC	= gcc
-CFLAGS	= -Wall -W -g3
-STRIP	= strip
-#STRIP	= touch
-CP	= cp
-DESTDIR =
-MANDIR	= /usr/share/man/man1/
-MKDIR	= mkdir -p
-
-all	: $(OBJ)
-	$(CC) $(CFLAGS) $(OBJ) -o $(TARGET)
-
-install	: all
-	$(STRIP) $(TARGET)
-	$(MKDIR) $(DESTDIR)/sbin/ $(DESTDIR)$(MANDIR)
-	$(CP) $(TARGET) $(DESTDIR)/sbin/
-	$(CP) $(TARGET).1 $(DESTDIR)$(MANDIR)
-
-clean	:
-	$(RM) *.o *~ \#* *\#
-
-fclean	: clean
-	$(RM) $(TARGET)
diff -BrauN chpax-0.5/Makefile.am chpax-0.5-omer/Makefile.am
--- chpax-0.5/Makefile.am	1969-12-31 18:00:00.000000000 -0600
+++ chpax-0.5-omer/Makefile.am	2003-12-24 16:09:22.000000000 -0600
@@ -0,0 +1,24 @@
+## Process this file with automake to create Makefile.in.
+
+sbin_PROGRAMS = chpax
+
+chpax_SOURCES = \
+	aout.c \
+	chpax.c \
+	chpax.h \
+	elf32.c \
+	elf64.c \
+	flags.c \
+	io.c
+
+chpax_DEPENDENCIES = \
+	$(chpax_SOURCES) \
+        config.h
+
+dist_man_MANS = chpax.1
+
+EXTRA_DIST = \
+	README \
+	ChangeLog
+
+AM_CFLAGS = -Wall -W
