--- ./src/url.c.url_password	2003-12-11 10:30:58.000000000 +0100
+++ ./src/url.c	2003-12-11 10:32:38.000000000 +0100
@@ -466,12 +466,19 @@
 static int
 url_skip_credentials (const char *url)
 {
   /* Look for '@' that comes before terminators, such as '/', '?',
      '#', or ';'.  */
-  const char *p = (const char *)strpbrk (url, "@/?#;");
+  const char *p, *pp;
+  
-  if (!p || *p != '@')
-    return 0;
-  return p + 1 - url;
+  pp = url;
+  while ((p = (const char *)strpbrk (pp, "/?@")) != NULL) {
+    if (*p != '@')
+      break;
+    /* Found '@' character so go on with possible next '@'. */
+    pp = p + 1;
+  }
+  
+  return pp != url ? pp - url: 0;
 }
 
 /* Parse credentials contained in [BEG, END).  The region is expected
