--- glibc-2.3.2/sysdeps/generic/dl-sysdep.c.propolice	2003-10-24 17:17:36.000000000 -0600
+++ glibc-2.3.2/sysdeps/generic/dl-sysdep.c	2003-10-24 17:17:36.000000000 -0600
@@ -51,7 +51,8 @@
 int __libc_enable_secure = 0;
 INTVARDEF(__libc_enable_secure)
 int __libc_multiple_libcs = 0;	/* Defining this here avoids the inclusion
-				   of init-first.  */
+int __libc_argc = 0;				   of init-first.  */
+char **_libc_argv = 0;
 /* This variable contains the lowest stack address ever used.  */
 void *__libc_stack_end;
 static ElfW(auxv_t) *_dl_auxv;
