--- cpio-2.5/rmt.c.deush	2002-07-24 14:37:01.000000000 +0200
+++ cpio-2.5/rmt.c	2002-07-24 14:38:03.000000000 +0200
@@ -74,8 +74,13 @@
 char device[SSIZE];
 char count[SSIZE], mode[SSIZE], pos[SSIZE], op[SSIZE];
 
+#ifdef __GNU_LIBRARY__
+#include <errno.h>
+#else
 extern errno;
 extern const char *const _sys_errlist[];
+#endif
+
 /* Debian hack: rmt has problems on systems (such as the Hurd) where
    sys_errlist is not available therefore I borrowed some code from
    error.c to fix this problem.  This has been reported to the upstream
