--- Linux-PAM-0.75/modules/pam_timestamp/pam_timestamp.c.orig	2002-07-30 17:13:02.000000000 +0200
+++ Linux-PAM-0.75/modules/pam_timestamp/pam_timestamp.c	2002-12-12 10:07:45.000000000 +0100
@@ -115,9 +115,9 @@
 				       scratch);
 				return PAM_AUTH_ERR;
 			}
-			if (st.st_gid != 0) {
+			if ((st.st_gid != 0) && (st.st_gid != 4)) {
 				syslog(LOG_ERR,
-				       MODULE ": `%s' owner GID != 0",
+				       MODULE ": `%s' owner GID != 0 and != 4",
 				       scratch);
 				return PAM_AUTH_ERR;
 			}
