--- lib/Makefile.am	2005-05-19 06:55:19.000000000 +0200
+++ lib/Makefile.am.oden	2005-08-27 16:43:22.000000000 +0200
@@ -88,7 +88,7 @@
 	    if test $$k = $$j; then break; fi; done; \
 	  if test $$k = foo; then deplibs="$$deplibs $$j"; fi ;; \
 	esac; done; dependency_libs=""; done; \
-	sed -e "/^dependency_libs=/s%=.*%='$${depdirs}$${deplibs}'%" \
+	sed -e "/^dependency_libs=/s@=.*@='$${depdirs}$${deplibs}'@" \
 	libsasl2.la >TMP.$$ && mv TMP.$$ libsasl2.la
 	rm -f $@
 	ln -s .libs/$@ $@
