--- xpm-3.4k/lib/Imakefile.shlib	1998-03-19 20:50:59.000000000 +0100
+++ xpm-3.4k/lib/Imakefile	2002-06-26 19:05:09.000000000 +0200
@@ -31,7 +31,7 @@ XCOMM
 
 XCOMM default locations
 #ifndef XpmLibDir
-#define XpmLibDir $(USRLIBDIR)
+#define XpmLibDir $(SHLIBDIR)
 #endif
 #ifndef XpmIncDir
 #define XpmIncDir $(INCDIR)
--- xpm-3.4k/sxpm/Imakefile.shlib	1998-03-19 20:51:01.000000000 +0100
+++ xpm-3.4k/sxpm/Imakefile	2002-06-26 19:03:01.000000000 +0200
@@ -32,7 +32,7 @@ XCOMM
 
 XCOMM default locations
 #ifndef XpmLibDir
-#define XpmLibDir $(USRLIBDIR)
+#define XpmLibDir $(SHLIBDIR)
 #endif
 #ifndef XpmBinDir
 #define XpmBinDir $(BINDIR)
