--- lilo-22.5.9/device.c.pix	2004-09-08 16:49:13.000000000 +0200
+++ lilo-22.5.9/device.c	2004-09-09 12:16:14.406158442 +0200
@@ -829,7 +829,7 @@
 
     if (ret && fatal) {
 	fprintf(stderr, "Aborting ...\n");
-	exit(0);
+	exit(1);
     }
     
     return ret;
--- lilo-22.5.9/bsect.c.pix	2004-09-08 16:49:13.000000000 +0200
+++ lilo-22.5.9/bsect.c	2004-09-09 12:16:36.324032126 +0200
@@ -516,7 +516,7 @@
 	fprintf(errstd, "WARNING: '-F' override used. Filesystem on  %s  may be destroyed.\n", boot_devnam);
 #else
 	fprintf(stderr, "WARNING: '-F' override used. Filesystem on  %s  may be destroyed.\n", boot_devnam);
-	if (!yesno("\nProceed? ",0)) exit(0);
+	if (!yesno("\nProceed? ",0)) exit(1);
 #endif
 	ireloc=PTW_OKAY;
     }
--- lilo-22.5.9/edit.c.pix	2003-08-29 17:54:41.000000000 +0200
+++ lilo-22.5.9/edit.c	2004-09-09 12:16:51.520944565 +0200
@@ -687,7 +687,7 @@
     }
 
     printf("Transfer parameters from '%s' to '%s'", config_file, bitmap_file);
-    if (yesno("?", 0)==0) exit(0);
+    if (yesno("?", 0)==0) exit(1);
 
     if (verbose > 0) printf("%s bitmap file:  %s\n", opt, bitmap_file);
     
--- lilo-22.5.9/partition.c.pix	2004-09-08 16:49:12.000000000 +0200
+++ lilo-22.5.9/partition.c	2004-09-09 12:17:42.284652212 +0200
@@ -225,7 +225,7 @@
 	if (!dos && !cfg_get_flag(cf_options,"ignore-table"))
 	    die("You may proceed by using either '-P ignore' or 'ignore-table'");
 #else
-	if (!yesno("\nProceed? ", 0)) exit(0);
+	if (!yesno("\nProceed? ", 0)) exit(1);
 #endif
     }
     cyl = part_table[part].cyl+((part_table[part].sector >> 6) << 8);
