--- cvs-1.11.6/src/add.c.errno	2003-04-30 18:21:21.000000000 +0200
+++ cvs-1.11.6/src/add.c	2003-06-04 20:28:09.000000000 +0200
@@ -357,8 +357,7 @@
 	    if (dirp == NULL)
 		error (1, errno, "cannot read directory %s", finfo.repository);
 	    found_name = NULL;
-	    errno = 0;
-	    while ((dp = CVS_READDIR (dirp)) != NULL)
+	    while (errno = 0, (dp = CVS_READDIR (dirp)) != NULL)
 	    {
 		if (cvs_casecmp (dp->d_name, finfo.file) == 0)
 		{
--- cvs-1.11.6/src/filesubr.c.errno	2003-03-19 21:44:49.000000000 +0100
+++ cvs-1.11.6/src/filesubr.c	2003-06-04 20:30:15.000000000 +0200
@@ -494,8 +494,7 @@
 		 */
 		return -1;
 
-	    errno = 0;
-	    while ((dp = CVS_READDIR (dirp)) != NULL)
+	    while (errno = 0, (dp = CVS_READDIR (dirp)) != NULL)
 	    {
 		char *buf;
 
@@ -1074,8 +1073,7 @@
 		error ( 1, errno, "cannot read directory %s", dir );
 	    }
 	}
-	errno = 0;
-	while ( ( dp = CVS_READDIR ( dirp ) ) != NULL )
+	while (	errno = 0, ( dp = CVS_READDIR ( dirp ) ) != NULL )
 	{
 	    if ( cvs_casecmp ( dp->d_name, file ) == 0 )
 	    {
--- cvs-1.11.6/src/import.c.errno	2003-01-31 21:15:11.000000000 +0100
+++ cvs-1.11.6/src/import.c	2003-06-04 20:28:09.000000000 +0200
@@ -427,8 +427,7 @@
     }
     else
     {
-	errno = 0;
-	while ((dp = CVS_READDIR (dirp)) != NULL)
+	while (errno = 0, (dp = CVS_READDIR (dirp)) != NULL)
 	{
 	    if (strcmp (dp->d_name, ".") == 0 || strcmp (dp->d_name, "..") == 0)
 		goto one_more_time_boys;
--- cvs-1.11.6/src/find_names.c.errno	2001-09-24 17:16:05.000000000 +0200
+++ cvs-1.11.6/src/find_names.c	2003-06-04 20:28:09.000000000 +0200
@@ -267,8 +267,7 @@
 	return (1);
 
     /* read the dir, grabbing the ,v files */
-    errno = 0;
-    while ((dp = CVS_READDIR (dirp)) != NULL)
+    while (errno = 0, (dp = CVS_READDIR (dirp)) != NULL)
     {
 	if (CVS_FNMATCH (RCSPAT, dp->d_name, 0) == 0) 
 	{
@@ -330,8 +329,7 @@
 	return (1);
 
     /* read the dir, grabbing sub-dirs */
-    errno = 0;
-    while ((dp = CVS_READDIR (dirp)) != NULL)
+    while (errno = 0, (dp = CVS_READDIR (dirp)) != NULL)
     {
 	if (strcmp (dp->d_name, ".") == 0 ||
 	    strcmp (dp->d_name, "..") == 0 ||
--- cvs-1.11.6/src/lock.c.errno	2003-02-28 22:46:48.000000000 +0100
+++ cvs-1.11.6/src/lock.c	2003-06-04 20:28:09.000000000 +0200
@@ -657,7 +657,7 @@
 
 	ret = 0;
 	errno = 0;
-	while ((dp = CVS_READDIR (dirp)) != NULL)
+	while (errno = 0, (dp = CVS_READDIR (dirp)) != NULL)
 	{
 	    if (CVS_FNMATCH (CVSRFLPAT, dp->d_name, 0) == 0)
 	    {
--- cvs-1.11.6/src/update.c.errno	2003-04-26 21:51:11.000000000 +0200
+++ cvs-1.11.6/src/update.c	2003-06-04 20:28:09.000000000 +0200
@@ -1101,8 +1101,7 @@
 	error (0, errno, "cannot open directory %s for empty check", dir);
 	return (0);
     }
-    errno = 0;
-    while ((dp = CVS_READDIR (dirp)) != NULL)
+    while (errno = 0, (dp = CVS_READDIR (dirp)) != NULL)
     {
 	if (strcmp (dp->d_name, ".") != 0
 	    && strcmp (dp->d_name, "..") != 0)
