--- tiff-v3.6.1/libtiff/tif_strip.c.can-2005-2452	2005-08-15 14:22:21.385786958 -0600
+++ tiff-v3.6.1/libtiff/tif_strip.c	2005-08-15 14:28:43.456791430 -0600
@@ -126,20 +126,26 @@
                               ycbcrsubsampling + 0, 
                               ycbcrsubsampling + 1 );
 
-		w = TIFFroundup(td->td_imagewidth, ycbcrsubsampling[0]);
-		scanline =
-		    TIFFhowmany8(multiply(tif, w, td->td_bitspersample, "TIFFVStripSize"));
 		samplingarea = ycbcrsubsampling[0]*ycbcrsubsampling[1];
+		if (samplingarea == 0) {
+			TIFFError(tif->tif_name, "Invalid YCbCr subsampling");
+			return 0;
+		}
+		
+		w = TIFFroundup(td->td_imagewidth, ycbcrsubsampling[0]);
+		scanline = TIFFhowmany8(multiply(tif, w, td->td_bitspersample, 
+						 "TIFFVStripSize"));
 		nrows = TIFFroundup(nrows, ycbcrsubsampling[1]);
 		/* NB: don't need TIFFhowmany here 'cuz everything is rounded */
 		scanline = multiply(tif, nrows, scanline, "TIFFVStripSize");
 		return ((tsize_t)
 			summarize(tif, scanline,
-				multiply(tif, 2, scanline / samplingarea,
-				"TIFFVStripSize"), "TIFFVStripSize"));
+				  multiply(tif, 2, scanline / samplingarea,
+					   "TIFFVStripSize"), "TIFFVStripSize"));
 	} else
 #endif
-		return ((tsize_t) multiply(tif, nrows, TIFFScanlineSize(tif), "TIFFVStripSize"));
+		return ((tsize_t) multiply(tif, nrows, TIFFScanlineSize(tif), 
+					   "TIFFVStripSize"));
 }
 
 
--- tiff-v3.6.1/libtiff/tif_tile.c.can-2005-2452	2005-08-15 14:29:04.556979081 -0600
+++ tiff-v3.6.1/libtiff/tif_tile.c	2005-08-15 14:34:13.564155430 -0600
@@ -205,6 +205,10 @@
 					  "TIFFVTileSize"));
 		tsize_t samplingarea =
 		    td->td_ycbcrsubsampling[0]*td->td_ycbcrsubsampling[1];
+		if (samplingarea == 0) {
+			TIFFError(tif->tif_name, "Invalid YCbCr subsampling");
+			return 0;
+		}
 		nrows = TIFFroundup(nrows, td->td_ycbcrsubsampling[1]);
 		/* NB: don't need TIFFhowmany here 'cuz everything is rounded */
 		tilesize = multiply(tif, nrows, rowsize, "TIFFVTileSize");
