--- cracklib,2.7/cracklib/Makefile.makevars	2002-06-25 13:14:18.000000000 +0200
+++ cracklib,2.7/cracklib/Makefile	2002-06-25 13:14:18.000000000 +0200
@@ -8,8 +8,10 @@
 
 LIB	= libcrack.so
 OBJ	= fascist.o packlib.o rules.o stringlib.o
-CFLAGS	= $(RPM_OPT_FLAGS) -g -I../cracklib -DIN_CRACKLIB -fPIC
+CFLAGS	= $(RPM_OPT_FLAGS) -I../cracklib -DIN_CRACKLIB -fPIC
 LD	= ld
+includedir	= /usr/include
+libdir		= /usr/lib
 
 $(LIB):	$(OBJ) Makefile
 	$(LD) -shared -soname $(LIB).$(MAJOR) -o $(LIB).$(VERSION) $(OBJ) -lc
@@ -21,6 +23,6 @@ clean:
 	-rm -f $(OBJ) $(LIB) $(LIB).$(VERSION) *~
 
 install: $(LIB) crack.h
-	install -m 755 $(LIB).$(VERSION) $(ROOT)/usr/lib
-	ln -sf $(LIB).$(VERSION) $(ROOT)/usr/lib/$(LIB)
-	install -m 644 crack.h $(ROOT)/usr/include
+	install -m 755 $(LIB).$(VERSION) $(ROOT)$(libdir)
+	ln -sf $(LIB).$(VERSION) $(ROOT)$(libdir)/$(LIB)
+	install -m 644 crack.h $(ROOT)$(includedir)
--- cracklib,2.7/util/Makefile.makevars	2002-06-25 13:14:18.000000000 +0200
+++ cracklib,2.7/util/Makefile	2002-06-25 13:14:18.000000000 +0200
@@ -17,6 +17,7 @@
 CFLAGS	= $(RPM_OPT_FLAGS) -I../cracklib '-DCRACKLIB_DICTPATH="$(DICTPATH)"'
 LDFLAGS	= -L../cracklib -lcrack
 LIBS	= ../cracklib/libcrack.so
+sbindir = /usr/sbin
 
 all:	packer unpacker testnum teststr testlib
 	touch all
@@ -41,4 +42,4 @@ clean:
 	-rm teststr testnum testlib packer unpacker
 
 install: all create-cracklib-dict
-	install -m 755 mkdict packer create-cracklib-dict $(ROOT)/usr/sbin
+	install -m 755 mkdict packer create-cracklib-dict $(ROOT)/$(sbindir)
--- cracklib,2.7/Makefile.makevars	2002-06-25 13:14:18.000000000 +0200
+++ cracklib,2.7/Makefile	2002-06-25 13:20:05.000000000 +0200
@@ -6,6 +6,9 @@
 # and upwards.
 ###
 
+libdir = /usr/lib
+datadir = /usr/share
+
 ###
 # cracklib version
 MAJOR=2
@@ -16,12 +19,12 @@ export MAJOR MINOR VERSION
 ###
 # set this to the absolute path (less extn) of compressed dict.
 
-DICTPATH="/usr/lib/cracklib_dict"
+DICTPATH="$(libdir)/cracklib_dict"
 
 ###
 # Set this to the path of one or more files continaing wordlists.
 
-SRCDICTS=/usr/share/dict/words /usr/share/dict/extra.words
+SRCDICTS=$(datadir)/dict/words $(datadir)/dict/extra.words
 
 ###
 # If you have installed the cracklib-dicts directory, use this
