--- Linux-PAM-0.75/modules/pam_time/pam_time.c.fred	Sun Sep 29 12:11:16 2002
+++ Linux-PAM-0.75/modules/pam_time/pam_time.c	Sun Sep 29 14:48:45 2002
@@ -577,8 +577,7 @@
 	D(("PAM_TTY not set, probing stdin"));
 	tty = ttyname(STDIN_FILENO);
 	if (tty == NULL) {
-	    _log_err("couldn't get the tty name");
-	    return PAM_ABORT;
+	    tty = "none";
 	}
 	if (pam_set_item(pamh, PAM_TTY, tty) != PAM_SUCCESS) {
 	    _log_err("couldn't set tty name");
