Index: util-linux-2.12q/mount/swapon.c
===================================================================
--- util-linux-2.12q.orig/mount/swapon.c	2005-04-29 14:29:32.000000000 +0200
+++ util-linux-2.12q/mount/swapon.c	2005-04-29 14:41:44.000000000 +0200
@@ -860,6 +860,8 @@ main_swapoff(int argc, char *argv[]) {
 					loop = opt + 5;
 				if (strncmp(opt, "encryption=", 11) == 0)
 					encryption = opt + 11;
+				if (strncmp(opt, "encrypted", 9) == 0)
+					skip=1;
 			}
 			if (loop && encryption) {
 				if (!is_in_proc_swaps(loop)) {
