--- openldap-2.3.7/servers/slapd/slap.h~	2005-08-26 02:18:03.000000000 +0200
+++ openldap-2.3.7/servers/slapd/slap.h	2005-09-09 17:35:18.000000000 +0200
@@ -74,10 +74,10 @@
 #define SLAP_ORDERED_PRETTYNORM
 #define SLAP_AUTHZ_SYNTAX
 
+#endif
 #ifdef ENABLE_REWRITE
 #define SLAP_AUTH_REWRITE	1 /* use librewrite for sasl-regexp */
 #endif
-#endif
 
 #if defined(LDAP_SLAPI) && !defined(SLAP_OVERLAY_ACCESS)
 #define SLAP_OVERLAY_ACCESS
