--- js/src/config/Linux_All.mk.va_copy	2003-05-15 20:21:08.000000000 -0400
+++ js/src/config/Linux_All.mk	2004-09-29 16:19:38.478184537 -0400
@@ -86,8 +86,12 @@ endif
 JS_EDITLINE = 1
 
 ifeq ($(CPU_ARCH),x86_64)
-# Use VA_COPY() standard macro on x86-64
-# FIXME: better use it everywhere
+# Use va_copy() standard macro on x86-64
+OS_CFLAGS += -DHAVE_VA_COPY
+endif
+
+ifeq ($(CPU_ARCH),ppc)
+# Use va_copy() standard macro on ppc too
 OS_CFLAGS += -DHAVE_VA_COPY
 endif
 
--- js/src/jsprf.c.va_copy	2003-02-23 01:59:13.000000000 -0500
+++ js/src/jsprf.c	2004-09-29 16:21:06.093858503 -0400
@@ -51,7 +51,7 @@
 ** and requires array notation.
 */
 #ifdef HAVE_VA_COPY
-#define VARARGS_ASSIGN(foo, bar)        VA_COPY(foo,bar)
+#define VARARGS_ASSIGN(foo, bar)        va_copy(foo,bar)
 #elif defined(HAVE_VA_LIST_AS_ARRAY)
 #define VARARGS_ASSIGN(foo, bar)        foo[0] = bar[0]
 #else
