--- ncurses-5.4/progs/modules.deps	2000-01-01 20:30:45.000000000 -0500
+++ ncurses-5.4/progs/modules	2004-09-17 09:53:51.651481467 -0400
@@ -33,9 +33,9 @@
 
 @ base
 clear		progs		$(srcdir)	../include/term.h
-tic		progs		$(srcdir)	../include/term.h $(INCDIR)/tic.h $(srcdir)/dump_entry.h
+tic		progs		$(srcdir)	transform.h ../include/term.h $(INCDIR)/tic.h $(srcdir)/dump_entry.h
 toe		progs		$(srcdir)	../include/term.h $(INCDIR)/tic.h $(srcdir)/dump_entry.h
 dump_entry	progs		$(srcdir)	../include/term.h $(INCDIR)/tic.h $(srcdir)/dump_entry.h ../include/parametrized.h $(INCDIR)/capdefaults.c termsort.c
 infocmp		progs		$(srcdir)	../include/term.h $(INCDIR)/tic.h $(srcdir)/dump_entry.h
-tput		progs		$(srcdir)	../include/term.h
-tset		progs		$(srcdir)	../include/term.h
+tput		progs		$(srcdir)	transform.h ../include/term.h
+tset		progs		$(srcdir)	transform.h ../include/term.h
