Index: servers/slapd/overlays/syncprov.c
===================================================================
RCS file: /repo/OpenLDAP/pkg/ldap/servers/slapd/overlays/syncprov.c,v
retrieving revision 1.102
retrieving revision 1.103
diff -u -r1.102 -r1.103
--- servers/slapd/overlays/syncprov.c	26 Aug 2005 19:42:48 -0000	1.102
+++ servers/slapd/overlays/syncprov.c	2 Sep 2005 08:08:44 -0000	1.103
@@ -427,7 +427,9 @@
 	fop.ors_filter = fc->fss->s_op->ors_filter;
 	fop.ors_filterstr = fc->fss->s_op->ors_filterstr;
 
+	fop.o_req_dn = fc->fss->s_op->o_req_dn;
 	fop.o_req_ndn = fc->fss->s_op->o_req_ndn;
+	fop.o_authz = fc->fss->s_op->o_authz;
 
 	fop.o_bd->bd_info = on->on_info->oi_orig;
 	rc = fop.o_bd->be_search( &fop, &frs );
