--- util-linux-2.12q/mount/mount.c.devpts-mode	2005-08-26 11:33:32.000000000 -0400
+++ util-linux-2.12q/mount/mount.c	2005-08-26 11:44:44.000000000 -0400
@@ -476,7 +476,7 @@ static void remove_non_handled_options(c
    static char *fs_using_iocharset[] = { "iso9660", "udf", "vfat", "msdos", "ntfs", "jfs", "smbfs", "befs", "cifs", NULL };
    static char *fs_using_codepage[] = { "iso9660", "vfat", "msdos", "smbfs", NULL };
    static char *fs_using_umask[] = { "iso9660", "udf", "vfat", "msdos", "ntfs", "hfs", "hfsplus", "hpfs", NULL };
-   static char *fs_using_mode[] = { "iso9660", "ncpfs", "affs", NULL };
+   static char *fs_using_mode[] = { "iso9660", "ncpfs", "affs", "devpts", NULL };
    static char *fs_using_nls[] = { "ntfs", NULL };
    static char *fs_using_utf8[] = { "iso9660", "ntfs", "vfat", NULL };
 
