--- Net_SSLeay.pm-1.25/SSLeay.pm.fpons	2003-08-22 14:20:48.000000000 +0200
+++ Net_SSLeay.pm-1.25/SSLeay.pm	2003-08-22 14:26:50.000000000 +0200
@@ -1605,15 +1605,15 @@
 
 sub tcp_read_all {
     my ($how_much) = @_;
-    $how_much = 2000000000 unless $how_much;
+    $how_much = undef unless $how_much;
     my ($n, $got, $errs);
     my $reply = '';
 
-    while ($how_much > 0) {
-	$n = sysread(SSLCAT_S,$got,$how_much);
+    while (! defined $how_much || $how_much > 0) {
+	$n = sysread(SSLCAT_S,$got,$how_much > 0 ? $how_much : 512);
 	warn "Read error: $! ($n,$how_much)" unless defined $n;
 	last if !$n;  # EOF
-	$how_much -= $n;
+	defined $how_much and $how_much -= $n;
 	debug_read(\$reply, \$got) if $trace>1;
 	$reply .= $got;
     }
