diff -up q1/drivers/message/i2o/i2o_pci.c.orig q1/drivers/message/i2o/i2o_pci.c
--- q1/drivers/message/i2o/i2o_pci.c.orig	2002-07-30 15:33:09.000000000 +0200
+++ q1/drivers/message/i2o/i2o_pci.c	2002-07-30 15:35:10.000000000 +0200
@@ -345,6 +345,8 @@ int __init i2o_pci_scan(void)
 }
 
 
+void __init i2o_sys_init(void);
+
 /**
  *	i2o_pci_core_attach	-	PCI initialisation for I2O
  *
@@ -386,4 +386,3 @@ MODULE_PARM(dpt, "i");
 MODULE_PARM_DESC(dpt, "Set this if you want to drive DPT cards normally handled by dpt_i2o");
 module_init(i2o_pci_core_attach);
 module_exit(i2o_pci_core_detach);
- 
\ No newline at end of file
