diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/drivers/ide/pci/generic.h w1/drivers/ide/pci/generic.h
--- w2/drivers/ide/pci/generic.h	2003-08-20 10:56:54.000000000 +0200
+++ w1/drivers/ide/pci/generic.h	2003-08-20 19:20:17.000000000 +0200
@@ -148,6 +148,7 @@ static ide_pci_device_t generic_chipsets
 	}
 };
 
+#if 0
 static ide_pci_device_t unknown_chipset[] __devinitdata = {
 	{	/* 0 */
 		.vendor		= 0,
@@ -170,5 +171,5 @@ static ide_pci_device_t unknown_chipset[
 	}
 
 };
-
+#endif
 #endif /* IDE_GENERIC_H */
