diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/drivers/char/keyboard.c w1/drivers/char/keyboard.c
--- w2/drivers/char/keyboard.c	2003-08-20 19:14:10.000000000 +0200
+++ w1/drivers/char/keyboard.c	2003-08-20 20:29:50.000000000 +0200
@@ -732,10 +732,7 @@ static void do_fn(unsigned char value, c
 {
 	if (up_flag)
 		return;
-	if (value < SIZE(func_table)) {
-		if (func_table[value])
-			puts_queue(func_table[value]);
-	} else
-		printk(KERN_ERR "do_fn called with value=%d\n", value);
+	if (func_table[value])
+		puts_queue(func_table[value]);
 }
 
