--- linux-2.4.19/drivers/char/console.c.console-fix	2003-05-22 21:25:53.000000000 +0200
+++ linux-2.4.19/drivers/char/console.c	2003-05-26 15:24:25.000000000 +0200
@@ -259,7 +259,7 @@ static void scrup(int currcons, unsigned
 		return;
 	d = (unsigned short *) (origin+video_size_row*t);
 	s = (unsigned short *) (origin+video_size_row*(t+nr));
-	scr_memcpyw(d, s, (b-t-nr) * video_size_row);
+	scr_memmovew(d, s, (b-t-nr) * video_size_row);
 	scr_memsetw(d + (b-t-nr) * video_num_columns, video_erase_char, video_size_row*nr);
 }
 
