diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' t2/arch/i386/config.in t1/arch/i386/config.in
--- t2/arch/i386/config.in	2003-06-30 12:30:08.000000000 +0200
+++ t1/arch/i386/config.in	2003-06-30 16:38:37.000000000 +0200
@@ -211,7 +211,7 @@ if [ "$CONFIG_CPU_FREQ" = "y" ]; then
    fi
    dep_tristate ' VIA Cyrix III Longhaul' CONFIG_X86_LONGHAUL $CONFIG_CPU_FREQ_TABLE
    dep_tristate ' Intel Speedstep (ICH)' CONFIG_X86_SPEEDSTEP_ICH  $CONFIG_CPU_FREQ_TABLE
-   dep_tristate ' Intel Pentium-M Enhanced SpeedStep' CONFIG_X86_SPEEDSTEP_CENTRINO
+   dep_tristate ' Intel Pentium-M Enhanced SpeedStep' CONFIG_X86_SPEEDSTEP_CENTRINO $CONFIG_CPU_FREQ_TABLE
    dep_tristate ' Intel Pentium 4 clock modulation' CONFIG_X86_P4_CLOCKMOD $CONFIG_CPU_FREQ_TABLE
    tristate ' Transmeta LongRun' CONFIG_X86_LONGRUN
    tristate ' Cyrix MediaGX/NatSemi Geode Suspend Modulation' CONFIG_X86_GX_SUSPMOD
