--- t2/drivers/misc/aes-c.c	2003-02-07 02:18:19.000000000 +0100
+++ t1/drivers/misc/aes-c.c	2003-02-07 04:21:06.000000000 +0100
@@ -1397,6 +1397,6 @@ void aes_decrypt(const aes_context *cx, 
     state_out(out_blk, b0);
 }
 
-EXPORT_SYMBOL(aes_set_key);
-EXPORT_SYMBOL(aes_encrypt);
-EXPORT_SYMBOL(aes_decrypt);
+EXPORT_SYMBOL_NOVERS(aes_set_key);
+EXPORT_SYMBOL_NOVERS(aes_encrypt);
+EXPORT_SYMBOL_NOVERS(aes_decrypt);
--- t2/drivers/misc/aes-i586-ksym.c	2003-02-07 04:35:20.000000000 +0100
+++ t1/drivers/misc/aes-i586-ksym.c	2003-02-07 04:37:47.000000000 +0100
@@ -1,5 +1,14 @@
 #include <linux/module.h>
 #include <linux/aes.h>
-EXPORT_SYMBOL(aes_set_key);
-EXPORT_SYMBOL(aes_encrypt);
-EXPORT_SYMBOL(aes_decrypt);
+
+MODULE_AUTHOR("Dr Brian Gladman <brg@gladman.uk.net>");
+MODULE_DESCRIPTION("AES encryption functions");
+/* This is to make happy modprobe, mailed author about it.
+   This is temporary until he answers. Juan Quintela.
+ */
+MODULE_LICENSE("Dual BSD/GPL");
+
+
+EXPORT_SYMBOL_NOVERS(aes_set_key);
+EXPORT_SYMBOL_NOVERS(aes_encrypt);
+EXPORT_SYMBOL_NOVERS(aes_decrypt);
