--- linux-2.4.21/drivers/ieee1394/sbp2.c.orig	Thu Jul 24 18:40:52 2003
+++ linux-2.4.21/drivers/ieee1394/sbp2.c	Thu Jul 24 18:41:56 2003
@@ -231,7 +231,7 @@
  * enable this define to make use of it. This provides better hotplug
  * support. The mentioned patch is not part of the kernel proper though,
  * because it is considered somewhat of a hack. */
-//#define SBP2_USE_SCSI_ADDREM_HACK
+#define SBP2_USE_SCSI_ADDREM_HACK
 
 
 /*
