diff -uNp w2/net/Makefile.de02-2.orig w2/net/Makefile
--- w2/net/Makefile.de02-2.orig	2002-08-03 02:39:46.000000000 +0200
+++ w2/net/Makefile	2003-08-07 12:16:33.000000000 +0200
@@ -33,6 +33,7 @@ subdir-$(CONFIG_BRIDGE)		+= bridge
 subdir-$(CONFIG_IPX)		+= ipx
 subdir-$(CONFIG_ATALK)		+= appletalk
 subdir-$(CONFIG_WAN_ROUTER)	+= wanrouter
+subdir-$(CONFIG_CRYPTOAPI)	+= cryptoapi
 subdir-$(CONFIG_X25)		+= x25
 subdir-$(CONFIG_LAPB)		+= lapb
 subdir-$(CONFIG_NETROM)		+= netrom
diff -uNp w2/net/Config.in.de02-2.orig w2/net/Config.in
--- w2/net/Config.in.de02-2.orig	2003-08-05 00:43:15.000000000 +0200
+++ w2/net/Config.in	2003-08-07 12:17:03.000000000 +0200
@@ -51,6 +51,8 @@ if [ "$CONFIG_EXPERIMENTAL" = "y" ]; the
 fi
 tristate '802.1Q VLAN Support' CONFIG_VLAN_8021Q
 
+source net/cryptoapi/Config.in
+
 comment ' '
 tristate 'The IPX protocol' CONFIG_IPX
 if [ "$CONFIG_IPX" != "n" ]; then
