diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/net/ipsec/ipsec_rcv.c w1/net/ipsec/ipsec_rcv.c
--- w2/net/ipsec/ipsec_rcv.c	2003-06-29 05:59:39.000000000 +0200
+++ w1/net/ipsec/ipsec_rcv.c	2003-08-18 19:29:23.000000000 +0200
@@ -1320,7 +1320,7 @@ ipsec_rcv(struct sk_buff *skb, struct de
 						     (des_cblock *)iv, 0);
 				break;
 #endif /* CONFIG_IPSEC_ENC_3DES */
-			default:
+			default:;
 			}
 			next_header = idat[ilen - 1];
 			padlen = idat[ilen - 2];
