--- linux-2.4.21-orig/drivers/i2c/scx200_acb.c	2003-06-16 09:30:27.000000000 +0200
+++ linux-2.4.21-i2c/drivers/i2c/scx200_acb.c	2003-07-11 10:55:56.000000000 +0200
@@ -408,16 +408,6 @@
 	return 0;
 }
 
-static void scx200_acb_inc_use(struct i2c_adapter *adapter)
-{
-	MOD_INC_USE_COUNT;
-}
-
-static void scx200_acb_dec_use(struct i2c_adapter *adapter)
-{
-	MOD_DEC_USE_COUNT;
-}
-
 /* For now, we only handle combined mode (smbus) */
 static struct i2c_algorithm scx200_acb_algorithm = {
 	name:		"NatSemi SCx200 ACCESS.bus",
@@ -479,11 +469,10 @@
 	memset(iface, 0, sizeof(*iface));
 	adapter = &iface->adapter;
 	adapter->data = iface;
+	adapter->owner = THIS_MODULE;
 	sprintf(adapter->name, "SCx200 ACB%d", index);
 	adapter->id = I2C_ALGO_SMBUS;
 	adapter->algo = &scx200_acb_algorithm;
-	adapter->inc_use = scx200_acb_inc_use;
-	adapter->dec_use = scx200_acb_dec_use;
 	adapter->client_register = scx200_acb_reg;
 	adapter->client_unregister = scx200_acb_unreg;
 
