diff -ru linux-2.4.21-orig/drivers/acorn/char/i2c.c linux-2.4.21-i2c/drivers/acorn/char/i2c.c
--- linux-2.4.21-orig/drivers/acorn/char/i2c.c	2001-03-03 03:38:37.000000000 +0100
+++ linux-2.4.21-i2c/drivers/acorn/char/i2c.c	2003-07-10 09:46:14.000000000 +0200
@@ -203,11 +203,12 @@
 }
 
 static struct i2c_adapter ioc_ops = {
-	name:			"IOC/IOMD",
-	id:			I2C_HW_B_IOC,
-	algo_data:		&ioc_data,
-	client_register:	ioc_client_reg,
-	client_unregister:	ioc_client_unreg
+	.owner             = THIS_MODULE,
+	.name              = "IOC/IOMD",
+	.id                = I2C_HW_B_IOC,
+	.algo_data         = &ioc_data,
+	.client_register   = ioc_client_reg,
+	.client_unregister = ioc_client_unreg
 };
 
 static int __init i2c_ioc_init(void)
diff -ru linux-2.4.21-orig/drivers/acorn/char/pcf8583.c linux-2.4.21-i2c/drivers/acorn/char/pcf8583.c
--- linux-2.4.21-orig/drivers/acorn/char/pcf8583.c	2003-06-16 09:30:12.000000000 +0200
+++ linux-2.4.21-i2c/drivers/acorn/char/pcf8583.c	2003-07-10 09:56:05.000000000 +0200
@@ -225,12 +225,13 @@
 }
 
 static struct i2c_driver pcf8583_driver = {
-	name:		"PCF8583",
-	id:		I2C_DRIVERID_PCF8583,
-	flags:		I2C_DF_NOTIFY,
-	attach_adapter:	pcf8583_probe,
-	detach_client:	pcf8583_detach,
-	command:	pcf8583_command
+	.owner          = THIS_MODULE,
+	.name           = "PCF8583",
+	.id 	        = I2C_DRIVERID_PCF8583,
+	.flags          = I2C_DF_NOTIFY,
+	.attach_adapter = pcf8583_probe,
+	.detach_client  = pcf8583_detach,
+	.command        = pcf8583_command
 };
 
 static __init int pcf8583_init(void)
