diff -urNp --exclude-from=/home/quintela/config/misc/dontdiff q2/arch/i386/kernel/dmi_scan.c q1/arch/i386/kernel/dmi_scan.c
--- q2/arch/i386/kernel/dmi_scan.c	2003-10-09 23:54:51.000000000 +0200
+++ q1/arch/i386/kernel/dmi_scan.c	2003-10-09 23:58:30.000000000 +0200
@@ -1053,6 +1053,12 @@ static __initdata struct dmi_blacklist d
 			/* newer BIOS, Revision 1011, does work */
 			MATCH(DMI_BIOS_VERSION, "ASUS A7V ACPI BIOS Revision 1007"),
 			NO_MATCH }},
+	{ acpi_disable, "Hewlett Packard d325 ", {
+			MATCH(DMI_BOARD_VENDOR, "Hewlett-Packard"),
+			MATCH(DMI_BOARD_NAME, "0830h"),
+			NO_MATCH, NO_MATCH }},
+
+
 #endif	// CONFIG_ACPI_BOOT
 
 	{ NULL, }
