diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/net/ipv4/netfilter/ip_conntrack_pptp.c w1/net/ipv4/netfilter/ip_conntrack_pptp.c
--- w2/net/ipv4/netfilter/ip_conntrack_pptp.c	2003-08-20 19:13:46.000000000 +0200
+++ w1/net/ipv4/netfilter/ip_conntrack_pptp.c	2003-08-18 19:37:48.000000000 +0200
@@ -59,7 +59,6 @@ DECLARE_LOCK(ip_pptp_lock);
 
 static int pptp_expectfn(struct ip_conntrack *ct)
 {
-	struct ip_conntrack_expect *exp, *other_exp;
 	struct ip_conntrack *master;
 
 	DEBUGP("increasing timeouts\n");
