From: Chmouel Boudjnah <chmouel@mandrakesoft.com>
Subject: [kernel] undef DEBUG for usb kaweth
To: Mandrake Kernel <kernel@linux-mandrake.com>
Date: 18 Jun 2001 16:03:50 +0200


Juan please apply.

- Disable debug for kaweth usb device.

--- linux/drivers/usb/kaweth.c.chmou    Mon Jun 18 14:40:50 2001
+++ linux/drivers/usb/kaweth.c  Mon Jun 18 16:03:16 2001
@@ -56,7 +56,7 @@
 #include <linux/types.h>
 #include <asm/semaphore.h>
 
-#define DEBUG
+#undef DEBUG
 
 #ifdef DEBUG
 #define kaweth_dbg(format, arg...) printk(KERN_DEBUG __FILE__ ": " format "\n" ,##arg)

