--- /usr/src/linux-2.4.21-0.pre3.1mdk/drivers/scsi/sd.c	2003-01-16 19:18:11.000000000 +0300
+++ linux-2.4.21-0.pre3.1mdk/drivers/scsi/sd.c	2003-01-21 21:29:55.000000000 +0300
@@ -867,8 +867,9 @@
 		if( the_result != 0
 		    && ((driver_byte(the_result) & DRIVER_SENSE) != 0)
 		    && SRpnt->sr_sense_buffer[2] == UNIT_ATTENTION
+		    && SRpnt->sr_sense_buffer[7] >= 6
 		    && SRpnt->sr_sense_buffer[12] == 0x3A ) {
-			rscsi_disks[i].capacity = 0x1fffff;
+			rscsi_disks[i].capacity = 0;
 			sector_size = 512;
 			rscsi_disks[i].device->changed = 1;
 			rscsi_disks[i].ready = 0;
@@ -950,29 +951,40 @@
 	 */
 
 	if (the_result) {
-		printk("%s : READ CAPACITY failed.\n"
-		       "%s : status = %x, message = %02x, host = %d, driver = %02x \n",
-		       nbuff, nbuff,
-		       status_byte(the_result),
-		       msg_byte(the_result),
-		       host_byte(the_result),
-		       driver_byte(the_result)
-		    );
-		if (driver_byte(the_result) & DRIVER_SENSE)
-			print_req_sense("sd", SRpnt);
-		else
-			printk("%s : sense not available. \n", nbuff);
-
-		printk("%s : block size assumed to be 512 bytes, disk size 1GB.  \n",
-		       nbuff);
-		rscsi_disks[i].capacity = 0x1fffff;
-		sector_size = 512;
-
-		/* Set dirty bit for removable devices if not ready -
-		 * sometimes drives will not report this properly. */
-		if (rscsi_disks[i].device->removable &&
-		    SRpnt->sr_sense_buffer[2] == NOT_READY)
+		if (rscsi_disks[i].device->removable
+		    && ((driver_byte(the_result) & DRIVER_SENSE) != 0)
+		    && SRpnt->sr_sense_buffer[2] == NOT_READY
+		    && SRpnt->sr_sense_buffer[7] >= 6
+		    && SRpnt->sr_sense_buffer[12] == 0x3A ) {
+			rscsi_disks[i].capacity = 0;
+			sector_size = 512;
 			rscsi_disks[i].device->changed = 1;
+			rscsi_disks[i].ready = 0;
+		} else {
+			printk("%s : READ CAPACITY failed.\n"
+			       "%s : status = %x, message = %02x, host = %d, driver = %02x \n",
+			       nbuff, nbuff,
+			       status_byte(the_result),
+			       msg_byte(the_result),
+			       host_byte(the_result),
+			       driver_byte(the_result)
+			    );
+			if (driver_byte(the_result) & DRIVER_SENSE)
+				print_req_sense("sd", SRpnt);
+			else
+				printk("%s : sense not available. \n", nbuff);
+
+			printk("%s : block size assumed to be 512 bytes, disk size 1GB.  \n",
+			       nbuff);
+			rscsi_disks[i].capacity = 0x1fffff;
+			sector_size = 512;
+
+			/* Set dirty bit for removable devices if not ready -
+			 * sometimes drives will not report this properly. */
+			if (rscsi_disks[i].device->removable &&
+			    SRpnt->sr_sense_buffer[2] == NOT_READY)
+				rscsi_disks[i].device->changed = 1;
+		}
 
 	} else {
 		/*
