--- linux-2.4.22/include/linux/brlock.h.hb06-6.orig	2002-11-29 00:53:15.000000000 +0100
+++ linux-2.4.22/include/linux/brlock.h	2003-08-28 03:16:06.000000000 +0200
@@ -45,7 +45,7 @@ enum brlock_indices {
 #include <linux/cache.h>
 #include <linux/spinlock.h>
 
-#if defined(__i386__) || defined(__ia64__) || defined(__x86_64__)
+#if defined(__i386__) || defined(__ia64__)
 #define __BRLOCK_USE_ATOMICS
 #else
 #undef __BRLOCK_USE_ATOMICS
--- linux-2.4.22/drivers/net/wan/Config.in.hb06-6.orig	2003-06-13 16:51:35.000000000 +0200
+++ linux-2.4.22/drivers/net/wan/Config.in	2003-08-28 03:16:06.000000000 +0200
@@ -42,8 +42,9 @@ fi
 # The Etinc driver has not been tested as non-modular yet.
 #
 
-   dep_tristate '  Etinc PCISYNC serial board support (EXPERIMENTAL)' CONFIG_DSCC4 m
-
+   if [ "$CONFIG_X86_64" != "y" ]; then	
+      dep_tristate '  Etinc PCISYNC serial board support (EXPERIMENTAL)' CONFIG_DSCC4 m
+   fi
 
 #
 # Lan Media's board. Currently 1000, 1200, 5200, 5245
@@ -55,7 +56,9 @@ fi
 # ATI's sync/async boards. Currently 2520, 4020, 4520, 8520 -- 0 in second digit means async only
 #
 
+   if [ "$CONFIG_X86_64" != "y" ]; then	
    tristate '  Aurora Technology, Inc. synchronous asynchronous PCI cards V2' CONFIG_ATI_XX20
+   fi
 
 # There is no way to detect a Sealevel board. Force it modular
 
@@ -108,7 +111,7 @@ fi
 	 if [ "$CONFIG_EXPERIMENTAL" = "y" ]; then
 	    dep_tristate '    Cyclom 2X(tm) cards (EXPERIMENTAL)' CONFIG_CYCLADES_SYNC $CONFIG_WAN_ROUTER_DRIVERS $CONFIG_ISA
 	    if [ "$CONFIG_CYCLADES_SYNC" != "n" ]; then
-	       bool '      Cyclom 2X X.25 support (EXPERIMENTAL)' CONFIG_CYCLOMX_X25
+	       bool '      Cyclom 2X X.25 support' CONFIG_CYCLOMX_X25
 	    fi
 	 fi
       fi
--- linux-2.4.22/drivers/net/tokenring/Config.in.hb06-6.orig	2002-08-03 02:39:44.000000000 +0200
+++ linux-2.4.22/drivers/net/tokenring/Config.in	2003-08-28 03:16:06.000000000 +0200
@@ -5,6 +5,11 @@
 mainmenu_option next_comment
 comment 'Token Ring devices'
 
+#x86-64 disabled by now because the drivers look largely not 64bit clean
+if [ "$CONFIG_X86_64" = "y" ]; then
+   define_bool CONFIG_TR n
+else
+
 # So far, we only have PCI, ISA, and MCA token ring devices
 if [ "$CONFIG_PCI" = "y" -o "$CONFIG_ISA" = "y" -o "$CONFIG_MCA" = "y" ]; then
    bool 'Token Ring driver support' CONFIG_TR
@@ -12,6 +17,8 @@ else
    define_bool CONFIG_TR n
 fi
 
+fi
+
 if [ "$CONFIG_TR" != "n" ]; then
    if [ "$CONFIG_ISA" = "y" -o "$CONFIG_MCA" = "y" ]; then
       tristate '  IBM Tropic chipset based adapter support' CONFIG_IBMTR
--- linux-2.4.22/drivers/net/de620.c.hb06-6.orig	2003-06-13 16:51:34.000000000 +0200
+++ linux-2.4.22/drivers/net/de620.c	2003-08-28 03:16:06.000000000 +0200
@@ -170,6 +170,9 @@ typedef unsigned char byte;
 #define PRINTK(x) /**/
 #endif
 
+#ifdef __x86_64__
+#define inline      /* work around compiler problem */
+#endif
 
 /*
  * Force media with insmod:
--- linux-2.4.22/drivers/scsi/Config.in.hb06-6.orig	2003-08-28 03:10:07.000000000 +0200
+++ linux-2.4.22/drivers/scsi/Config.in	2003-08-28 03:16:06.000000000 +0200
@@ -41,11 +41,11 @@ fi
 if [ "$CONFIG_PCI" = "y" ]; then
    dep_tristate '3ware Hardware ATA-RAID support' CONFIG_BLK_DEV_3W_XXXX_RAID $CONFIG_SCSI
 fi
-dep_tristate '7000FASST SCSI support' CONFIG_SCSI_7000FASST $CONFIG_SCSI
+dep_tristate '7000FASST SCSI support' CONFIG_SCSI_7000FASST $CONFIG_SCSI $CONFIG_ISA
 dep_tristate 'ACARD SCSI support' CONFIG_SCSI_ACARD $CONFIG_SCSI
-dep_tristate 'Adaptec AHA152X/2825 support' CONFIG_SCSI_AHA152X $CONFIG_SCSI
-dep_tristate 'Adaptec AHA1542 support' CONFIG_SCSI_AHA1542 $CONFIG_SCSI
-dep_tristate 'Adaptec AHA1740 support' CONFIG_SCSI_AHA1740 $CONFIG_SCSI
+dep_tristate 'Adaptec AHA152X/2825 support' CONFIG_SCSI_AHA152X $CONFIG_SCSI $CONFIG_ISA
+dep_tristate 'Adaptec AHA1542 support' CONFIG_SCSI_AHA1542 $CONFIG_SCSI $CONFIG_ISA
+dep_tristate 'Adaptec AHA1740 support' CONFIG_SCSI_AHA1740 $CONFIG_SCSI $CONFIG_EISA
 if [ "$CONFIG_EXPERIMENTAL" = "y" ]; then
    dep_tristate 'Adaptec AACRAID support (EXPERIMENTAL)' CONFIG_SCSI_AACRAID $CONFIG_SCSI $CONFIG_PCI
 fi
@@ -62,8 +62,12 @@ if [ "$CONFIG_X86_64" != "y" ]; then
   dep_tristate 'Adaptec I2O RAID support ' CONFIG_SCSI_DPT_I2O $CONFIG_SCSI
 fi
 dep_tristate 'AdvanSys SCSI support' CONFIG_SCSI_ADVANSYS $CONFIG_SCSI
-dep_tristate 'Always IN2000 SCSI support' CONFIG_SCSI_IN2000 $CONFIG_SCSI
-dep_tristate 'AM53/79C974 PCI SCSI support' CONFIG_SCSI_AM53C974 $CONFIG_SCSI $CONFIG_PCI
+dep_tristate 'Always IN2000 SCSI support' CONFIG_SCSI_IN2000 $CONFIG_SCSI $CONFIG_ISA
+# does not use pci dma and seems to be isa/onboard only for old machines
+if [ "$CONFIG_X86_64" != "y" ]; then
+   dep_tristate 'AM53/79C974 PCI SCSI support' CONFIG_SCSI_AM53C974 $CONFIG_SCSI $CONFIG_PCI
+fi
+
 dep_tristate 'LSI MegaRAID support (driver 1.18 series)' CONFIG_SCSI_MEGARAID $CONFIG_SCSI
 dep_tristate 'LSI MegaRAID support (driver 2.00 series)' CONFIG_SCSI_MEGARAID2 $CONFIG_SCSI
 
@@ -75,7 +79,7 @@ if [ "$CONFIG_PCI" = "y" ]; then
    dep_tristate 'Compaq Fibre Channel 64-bit/66Mhz HBA support' CONFIG_SCSI_CPQFCTS $CONFIG_SCSI
 fi
 dep_tristate 'DMX3191D SCSI support' CONFIG_SCSI_DMX3191D $CONFIG_SCSI $CONFIG_PCI
-dep_tristate 'DTC3180/3280 SCSI support' CONFIG_SCSI_DTC3280 $CONFIG_SCSI
+dep_tristate 'DTC3180/3280 SCSI support' CONFIG_SCSI_DTC3280 $CONFIG_SCSI $CONFIG_ISA
 dep_tristate 'EATA ISA/EISA/PCI (DPT and generic EATA/DMA-compliant boards) support' CONFIG_SCSI_EATA $CONFIG_SCSI
 if [ "$CONFIG_SCSI_EATA" != "n" ]; then
    bool '  enable tagged command queueing' CONFIG_SCSI_EATA_TAGGED_QUEUE
@@ -131,7 +135,7 @@ if [ "$CONFIG_PPSCSI" != "n" ]; then
   dep_tristate '  Shuttle EPST adapter' CONFIG_PPSCSI_EPST $CONFIG_PPSCSI
 fi
 
-dep_tristate 'NCR53c406a SCSI support' CONFIG_SCSI_NCR53C406A $CONFIG_SCSI
+dep_tristate 'NCR53c406a SCSI support' CONFIG_SCSI_NCR53C406A $CONFIG_SCSI $CONFIG_ISA
 if [ "$CONFIG_MCA" = "y" ]; then
    dep_tristate 'NCR Dual 700 MCA SCSI support' CONFIG_SCSI_NCR_D700 $CONFIG_SCSI
    if [ "$CONFIG_SCSI_NCR_D700" != "n" ]; then
@@ -187,11 +191,11 @@ fi
 if [ "$CONFIG_MCA" = "y" ]; then
    dep_tristate 'NCR MCA 53C9x SCSI support' CONFIG_SCSI_MCA_53C9X $CONFIG_SCSI
 fi
-dep_tristate 'PAS16 SCSI support' CONFIG_SCSI_PAS16 $CONFIG_SCSI
+dep_tristate 'PAS16 SCSI support' CONFIG_SCSI_PAS16 $CONFIG_SCSI $CONFIG_ISA
 dep_tristate 'PCI2000 support' CONFIG_SCSI_PCI2000 $CONFIG_SCSI
 dep_tristate 'PCI2220i support' CONFIG_SCSI_PCI2220I $CONFIG_SCSI
-dep_tristate 'PSI240i support' CONFIG_SCSI_PSI240I $CONFIG_SCSI
-dep_tristate 'Qlogic FAS SCSI support' CONFIG_SCSI_QLOGIC_FAS $CONFIG_SCSI
+dep_tristate 'PSI240i support' CONFIG_SCSI_PSI240I $CONFIG_SCSI $CONFIG_ISA
+dep_tristate 'Qlogic FAS SCSI support' CONFIG_SCSI_QLOGIC_FAS $CONFIG_SCSI $CONFIG_ISA
 if [ "$CONFIG_PCI" = "y" ]; then
    dep_tristate 'Qlogic ISP SCSI support' CONFIG_SCSI_QLOGIC_ISP $CONFIG_SCSI
    dep_tristate 'Qlogic ISP FC SCSI support' CONFIG_SCSI_QLOGIC_FC $CONFIG_SCSI
@@ -201,24 +205,26 @@ if [ "$CONFIG_PCI" = "y" ]; then
    dep_tristate 'Qlogic QLA 1280 SCSI support' CONFIG_SCSI_QLOGIC_1280 $CONFIG_SCSI
    dep_tristate 'Qlogic QLA 2100 driver support' CONFIG_SCSI_QLOGIC_QLA2100 $CONFIG_SCSI
 fi
-if [ "$CONFIG_X86" = "y" ]; then
+if [ "$CONFIG_X86" = "y" -a "$CONFIG_ISA" = "y" ]; then
    dep_tristate 'Seagate ST-02 and Future Domain TMC-8xx SCSI support' CONFIG_SCSI_SEAGATE $CONFIG_SCSI
 fi
-dep_tristate 'Simple 53c710 SCSI support (Compaq, NCR machines)' CONFIG_SCSI_SIM710 $CONFIG_SCSI
-dep_tristate 'Symbios 53c416 SCSI support' CONFIG_SCSI_SYM53C416 $CONFIG_SCSI
+if [ "$CONFIG_ISA" = "y" -o "$CONFIG_MCA" = "y" -a "$CONFIG_X86_64" != "y" ]; then
+  dep_tristate 'Simple 53c710 SCSI support (Compaq, NCR machines)' CONFIG_SCSI_SIM710 $CONFIG_SCSI
+fi
+dep_tristate 'Symbios 53c416 SCSI support' CONFIG_SCSI_SYM53C416 $CONFIG_SCSI $CONFIG_ISA
 if [ "$CONFIG_PCI" = "y" ]; then
    dep_tristate 'Tekram DC390(T) and Am53/79C974 SCSI support' CONFIG_SCSI_DC390T $CONFIG_SCSI
    if [ "$CONFIG_SCSI_DC390T" != "n" ]; then
       bool '  _omit_ support for non-DC390 adapters' CONFIG_SCSI_DC390T_NOGENSUPP
    fi
 fi
-dep_tristate 'Trantor T128/T128F/T228 SCSI support' CONFIG_SCSI_T128 $CONFIG_SCSI
+dep_tristate 'Trantor T128/T128F/T228 SCSI support' CONFIG_SCSI_T128 $CONFIG_SCSI $CONFIG_ISA
 dep_tristate 'UltraStor 14F/34F support' CONFIG_SCSI_U14_34F $CONFIG_SCSI
    if [ "$CONFIG_SCSI_U14_34F" != "n" ]; then
       bool '  enable elevator sorting' CONFIG_SCSI_U14_34F_LINKED_COMMANDS
       int  '  maximum number of queued commands' CONFIG_SCSI_U14_34F_MAX_TAGS 8
    fi
-if [ "$CONFIG_X86" = "y" ]; then
+if [ "$CONFIG_X86" = "y" -a "$CONFIG_ISA" = "y" ]; then
    dep_tristate 'UltraStor SCSI support' CONFIG_SCSI_ULTRASTOR $CONFIG_SCSI
 fi
 dep_tristate 'Workbit NinjaSCSI-32Bi/UDE support' CONFIG_SCSI_NSP32 $CONFIG_SCSI
--- linux-2.4.22/drivers/scsi/ips.c.hb06-6.orig	2003-08-28 02:43:35.000000000 +0200
+++ linux-2.4.22/drivers/scsi/ips.c	2003-08-28 03:16:25.000000000 +0200
@@ -199,7 +199,7 @@ MODULE_PARM(ips, "s");
 #define IPS_VERSION_HIGH        "6.10"
 #define IPS_VERSION_LOW         ".24 "
 
-#if !defined(__i386__) && !defined(__ia64__)
+#if !defined(__i386__) && !defined(__ia64__) && !defined(__x86_64__)
 #error "This driver has only been tested on the x86/ia64 platforms"
 #endif
 
