--- linux-2.4.22/arch/x86_64/kernel/e820.c.amd64-acpi-on	2003-10-29 06:47:04.000000000 +0100
+++ linux-2.4.22/arch/x86_64/kernel/e820.c	2003-10-30 11:48:45.000000000 +0100
@@ -538,6 +538,8 @@ void __init parse_mem_cmdline (char ** c
 			iommu_setup(from+6); 
 		} 	
 #endif
+ 		else if (!memcmp(from, "acpi=on", 7))
+  			acpi_disabled = 0;
  		else if (!memcmp(from, "acpi=off", 8))
   			acpi_disabled = 1;
 		else if (!memcmp(from,"maxcpus=0",9)) {
