diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' w2/drivers/char/agp/agpgart_be.c w1/drivers/char/agp/agpgart_be.c
--- w2/drivers/char/agp/agpgart_be.c	2003-08-21 03:11:00.000000000 +0200
+++ w1/drivers/char/agp/agpgart_be.c	2003-08-21 02:58:58.000000000 +0200
@@ -5268,7 +5268,7 @@ static int ati_create_gatt_pages(int nr_
 #define GET_PAGE_DIR_IDX(addr) (GET_PAGE_DIR_OFF(addr) - \
 	GET_PAGE_DIR_OFF(agp_bridge.gart_bus_addr))
 #define GET_GATT_OFF(addr) ((addr & 0x003ff000) >> 12)
-#undef  GET_GATT(addr)
+#undef  GET_GATT
 #define GET_GATT(addr) (ati_generic_private.gatt_pages[\
 	GET_PAGE_DIR_IDX(addr)]->remapped)
 
