diff -Nurp 2425/3rdparty/iteraid.old/iteraid.c 2425/3rdparty/iteraid/iteraid.c
--- 2425/3rdparty/iteraid.old/iteraid.c	2004-04-02 17:59:45.000000000 +0300
+++ 2425/3rdparty/iteraid/iteraid.c	2004-04-02 18:06:46.000000000 +0300
@@ -234,9 +234,9 @@ MODULE_DESCRIPTION ("ITE IT8212 RAID Con
 #include <asm/irq.h>
 #include <asm/uaccess.h>
 
-#include "sd.h"
-#include "scsi.h"
-#include "hosts.h"
+#include "../../drivers/scsi/sd.h"
+#include "../../drivers/scsi/scsi.h"
+#include "../../drivers/scsi/hosts.h"
 
 #include "iteraid.h"
 
@@ -5941,4 +5941,4 @@ itedev_close(struct inode *inode,  struc
  * each time a new driver is written.
  */
 static Scsi_Host_Template driver_template = ITERAID;
-#include "scsi_module.c"
+#include "../../drivers/scsi/scsi_module.c"
