--- 2.4.26/3rdparty/iteraid/iteraid.c.old	2004-04-29 21:31:42.000000000 +0300
+++ 2.4.26/3rdparty/iteraid/iteraid.c	2004-04-29 21:33:17.324572495 +0300
@@ -245,9 +245,9 @@
 #include <asm/irq.h>
 #include <asm/uaccess.h>
 
-#include "sd.h"
-#include "scsi.h"
-#include "hosts.h"
+#include "../../drivers/scsi/sd.h"
+#include "../../drivers/scsi/scsi.h"
+#include "../../drivers/scsi/hosts.h"
 
 #include "iteraid.h"
 
@@ -5852,4 +5852,4 @@
  * each time a new driver is written.
  */
 static Scsi_Host_Template driver_template = ITERAID;
-#include "scsi_module.c"
+#include "../../drivers/scsi/scsi_module.c"
