diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff orig/3rdparty/sam9407/audio.c w1/3rdparty/sam9407/audio.c
--- orig/3rdparty/sam9407/audio.c	2002-08-27 19:31:27.000000000 +0200
+++ w1/3rdparty/sam9407/audio.c	2002-08-27 18:43:53.000000000 +0200
@@ -670,7 +670,7 @@ static int audioFeedThread(void *arg)
 		}
 	      }
 	      break;
-	    default: /* make compiler happy */
+	    default:; /* make compiler happy */
 	    }
 
 	    up(&chp->accessLock);
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff orig/3rdparty/sam9407/cards.c w1/3rdparty/sam9407/cards.c
--- orig/3rdparty/sam9407/cards.c	2002-08-27 19:31:27.000000000 +0200
+++ w1/3rdparty/sam9407/cards.c	2002-08-27 18:44:15.000000000 +0200
@@ -247,7 +247,7 @@ static void findPNPCards(void)
 	      v=isapnp_read_byte(ISAPNP_CFG_IRQ+2*resource->idx);
 	      break;
 #endif
-	    default:
+	    default:;
 	      /* make compiler happy */
 	    }
 	    isapnp_cfg_end();
@@ -310,7 +310,7 @@ static int getPCIResources(SamPCIInfo *p
 	  v=b;
 	break;
 #endif
-      default:
+      default:;
 	/* make compiler happy */
       }
     }
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff orig/3rdparty/sam9407/hw_maxi64.c w1/3rdparty/sam9407/hw_maxi64.c
--- orig/3rdparty/sam9407/hw_maxi64.c	2002-08-27 19:31:27.000000000 +0200
+++ w1/3rdparty/sam9407/hw_maxi64.c	2002-08-27 18:46:13.000000000 +0200
@@ -186,7 +186,7 @@ static int maxi64AudioInputHandler(SamCa
       break;
     case SAM_MIXER_AUDIO_INPUT_DIGITAL:
       audioInput=3;
-    default: /* make compiler happy */
+    default:; /* make compiler happy */
     }
   }
 
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff orig/3rdparty/sam9407/midi.c w1/3rdparty/sam9407/midi.c
--- orig/3rdparty/sam9407/midi.c	2002-08-27 19:31:27.000000000 +0200
+++ w1/3rdparty/sam9407/midi.c	2002-08-27 18:44:31.000000000 +0200
@@ -211,7 +211,7 @@ void samMidiInput(SamCard *card, unsigne
     }
     SAM_END_CRITICAL(&card->midi.recSpinLock, flags);
     break;
-  default:
+  default:;
     /* SamMidiPortOwnerNone */
   }
 }
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff orig/3rdparty/sam9407/mixer.c w1/3rdparty/sam9407/mixer.c
--- orig/3rdparty/sam9407/mixer.c	2002-08-27 19:31:27.000000000 +0200
+++ w1/3rdparty/sam9407/mixer.c	2002-08-27 18:44:59.000000000 +0200
@@ -146,7 +146,7 @@ static __inline__ unsigned ctrlConvertVa
   case SamCtrlTypeValueHalfPlus:
     result=(result+1)>>1;
     break;
-  default:
+  default:;
     /* make compiler happy */
   }
 
@@ -870,7 +870,7 @@ int samCtrlsSet(SamCard *card, int chann
       case SamChannelStateClosed:
       case SamChannelStateClosing:
 	sendCmds=0;
-      default:
+      default:;
 	/* make compiler happy */
       }
     }
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff orig/3rdparty/sam9407/mod.c w1/3rdparty/sam9407/mod.c
--- orig/3rdparty/sam9407/mod.c	2002-08-27 19:31:27.000000000 +0200
+++ w1/3rdparty/sam9407/mod.c	2002-08-27 18:45:24.000000000 +0200
@@ -472,7 +472,7 @@ void samCheckPlaybackMod(SamCard *card)
 	  case SamModLoopReverseSignInvert:
 	    format|=0x02;
 	    break;
-	  default: /* SamModLoopForward */
+	  default:; /* SamModLoopForward */
 	  }
 
 	  error=samTalk(card,
@@ -538,7 +538,7 @@ void samCheckPlaybackMod(SamCard *card)
 		    SamSendCtrl, 4, SAM_CMD_VOI_FILT, event.filter.voice, event.filter.q, event.filter.fc,
 		    SamEndOfScript);
       break;
-    default:
+    default:;
       /* make compiler happy */
     }
     if(error<0)
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff orig/3rdparty/sam9407/sbank.c w1/3rdparty/sam9407/sbank.c
--- orig/3rdparty/sam9407/sbank.c	2002-08-27 19:31:27.000000000 +0200
+++ w1/3rdparty/sam9407/sbank.c	2002-08-27 18:45:48.000000000 +0200
@@ -576,7 +576,7 @@ static void fillupMidiTable(SamCard *car
 	    memcpy(drumDef, prioItem->u.drum.ptr, sizeof(DrumDef));
 	    *(char **)&dest+=sizeof(DrumDef);
 	    break;
-	  default:
+	  default:;
 	    /* PriorityItemIgnore:
 	     * will never be called that way,
 	     * but we want to keep the compiler happy
