diff -up w1/3rdparty/Portman_PCS/Portman_PCS.c.orig w1/3rdparty/Portman_PCS/Portman_PCS.c
--- w1/3rdparty/Portman_PCS/Portman_PCS.c.orig	2002-08-29 14:35:42.000000000 +0200
+++ w1/3rdparty/Portman_PCS/Portman_PCS.c	2002-08-29 14:38:21.000000000 +0200
@@ -20,13 +20,13 @@
 #include <linux/init.h>
 #include <linux/timer.h>
 #include <asm/uaccess.h>
+#include <asm/io.h>
 #include <linux/string.h>
 
 #include <linux/serial.h>
 #include <linux/serial_reg.h>
 #include <linux/serialP.h>
 
-#include <sys/io.h>
 #include <linux/delay.h>
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2,4,0)
@@ -70,6 +70,7 @@ static wait_queue_head_t PORTMAN_PCS_buf
 
 static unsigned char evt_prec = 0x00;	//past event read from tty
 
+static int play_event (unsigned char *q);
 
 /**********************************************************************************
 *   Serial Port write/read
