diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/3rdparty/mod_dvb/em8300/em8300_dvb.c w1/3rdparty/mod_dvb/em8300/em8300_dvb.c
--- w2/3rdparty/mod_dvb/em8300/em8300_dvb.c	2003-04-17 00:22:16.000000000 +0200
+++ w1/3rdparty/mod_dvb/em8300/em8300_dvb.c	2003-08-07 11:59:47.000000000 +0200
@@ -23,6 +23,7 @@
 #include <linux/version.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "modules/encoder.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w5/3rdparty/mod_dvb/em8300/modules/adv717x.c w4/3rdparty/mod_dvb/em8300/modules/adv717x.c
--- w5/3rdparty/mod_dvb/em8300/modules/adv717x.c	2003-04-17 00:22:16.000000000 +0200
+++ w4/3rdparty/mod_dvb/em8300/modules/adv717x.c	2003-08-07 18:35:21.000000000 +0200
@@ -40,6 +40,7 @@
 #include <linux/version.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 #include <linux/video_encoder.h>
 
@@ -82,8 +83,6 @@ MODULE_PARM(color_bars, "i");
 static int adv717x_attach_adapter(struct i2c_adapter *adapter);
 int adv717x_detach_client(struct i2c_client *client);
 int adv717x_command(struct i2c_client *client, unsigned int cmd, void *arg);
-void adv717x_inc_use (struct i2c_client *client);
-void adv717x_dec_use (struct i2c_client *client);
 
 #define CHIP_ADV7175A 1
 #define CHIP_ADV7170  2
@@ -101,14 +100,13 @@ struct adv717x_data_s {
 
 /* This is the driver that will be inserted */
 static struct i2c_driver adv717x_driver = {
-  /* name */		"ADV717X video encoder driver",
-  /* id */		I2C_DRIVERID_ADV717X,
-  /* flags */		I2C_DF_NOTIFY,
-  /* attach_adapter */  &adv717x_attach_adapter,
-  /* detach_client */   &adv717x_detach_client,
-  /* command */		&adv717x_command,
-  /* inc_use */		&adv717x_inc_use,
-  /* dec_use */		&adv717x_dec_use
+	.owner	= THIS_MODULE,
+	.name	= "ADV717X video encoder driver",
+	.id	= I2C_DRIVERID_ADV717X,
+	.flags	= I2C_DF_NOTIFY,
+	.attach_adapter	= adv717x_attach_adapter,
+	.detach_client	= adv717x_detach_client,
+	.command	= adv717x_command,
 };
 
 int adv717x_id = 0;
@@ -478,22 +476,6 @@ int adv717x_command(struct i2c_client *c
 	return 0;
 }
 
-/* Nothing here yet */
-void adv717x_inc_use (struct i2c_client *client)
-{
-#ifdef MODULE
-	MOD_INC_USE_COUNT;
-#endif
-}
-
-/* Nothing here yet */
-void adv717x_dec_use (struct i2c_client *client)
-{
-#ifdef MODULE
-	MOD_DEC_USE_COUNT;
-#endif
-}
-
 /* ----------------------------------------------------------------------- */
 
 
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w5/3rdparty/mod_dvb/em8300/modules/adv717x_eeprom.c w4/3rdparty/mod_dvb/em8300/modules/adv717x_eeprom.c
--- w5/3rdparty/mod_dvb/em8300/modules/adv717x_eeprom.c	2003-04-17 00:22:16.000000000 +0200
+++ w4/3rdparty/mod_dvb/em8300/modules/adv717x_eeprom.c	2003-08-07 18:36:32.000000000 +0200
@@ -40,6 +40,7 @@
 #include <linux/version.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 #include <linux/video_encoder.h>
 
@@ -61,19 +62,16 @@ MODULE_LICENSE("GPL");
 static int eeprom_attach_adapter(struct i2c_adapter *adapter);
 int eeprom_detach_client(struct i2c_client *client);
 int eeprom_command(struct i2c_client *client, unsigned int cmd, void *arg);
-void eeprom_inc_use (struct i2c_client *client);
-void eeprom_dec_use (struct i2c_client *client);
 
 /* This is the driver that will be inserted */
 static struct i2c_driver eeprom_driver = {
-  /* name */		"EEPROM video encoder driver",
-  /* id */		I2C_DRIVERID_AT24Cxx,
-  /* flags */		I2C_DF_NOTIFY,
-  /* attach_adapter */	&eeprom_attach_adapter,
-  /* detach_client */	&eeprom_detach_client,
-  /* command */		&eeprom_command,
-  /* inc_use */		&eeprom_inc_use,
-  /* dec_use */		&eeprom_dec_use
+	.owner	= THIS_MODULE,
+	.name	= "EEPROM video encoder driver",
+	.id	= I2C_DRIVERID_AT24Cxx,
+	.flags	= I2C_DF_NOTIFY,
+	.attach_adapter	= eeprom_attach_adapter,
+	.detach_client	= eeprom_detach_client,
+	.command	= eeprom_command,
 };
 
 int eeprom_id = 0;
@@ -155,22 +153,6 @@ int eeprom_command(struct i2c_client *cl
 	return 0;
 }
 
-/* Nothing here yet */
-void eeprom_inc_use (struct i2c_client *client)
-{
-#ifdef MODULE
-	MOD_INC_USE_COUNT;
-#endif
-}
-
-/* Nothing here yet */
-void eeprom_dec_use (struct i2c_client *client)
-{
-#ifdef MODULE
-	MOD_DEC_USE_COUNT;
-#endif
-}
-
 /* ----------------------------------------------------------------------- */
 
 
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w5/3rdparty/mod_dvb/em8300/modules/bt865.c w4/3rdparty/mod_dvb/em8300/modules/bt865.c
--- w5/3rdparty/mod_dvb/em8300/modules/bt865.c	2003-04-17 00:22:16.000000000 +0200
+++ w4/3rdparty/mod_dvb/em8300/modules/bt865.c	2003-08-07 18:37:41.000000000 +0200
@@ -41,6 +41,7 @@
 #include <linux/version.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 #include <linux/video_encoder.h>
 
@@ -67,8 +68,6 @@ MODULE_LICENSE("GPL");
 static int bt865_attach_adapter(struct i2c_adapter *adapter);
 int bt865_detach_client(struct i2c_client *client);
 int bt865_command(struct i2c_client *client, unsigned int cmd, void *arg);
-void bt865_inc_use (struct i2c_client *client);
-void bt865_dec_use (struct i2c_client *client);
 static int bt865_setup(struct i2c_client *client);
 
 struct bt865_data_s {
@@ -84,14 +83,13 @@ struct bt865_data_s {
 
 /* This is the driver that will be inserted */
 static struct i2c_driver bt865_driver = {
-  /* name */		"BT865 video encoder driver",
-  /* id */		I2C_DRIVERID_BT865,
-  /* flags */		I2C_DF_NOTIFY,
-  /* attach_adapter */  &bt865_attach_adapter,
-  /* detach_client */   &bt865_detach_client,
-  /* command */		&bt865_command,
-  /* inc_use */		&bt865_inc_use,
-  /* dec_use */		&bt865_dec_use
+	.owner	= THIS_MODULE,
+	.name	= "BT865 video encoder driver",
+	.id	= I2C_DRIVERID_BT865,
+	.flags	= I2C_DF_NOTIFY,
+	.attach_adapter	= bt865_attach_adapter,
+	.detach_client	= bt865_detach_client,
+	.command	= bt865_command,
 };
 
 int bt865_id = 0;
@@ -959,22 +957,6 @@ int bt865_command(struct i2c_client *cli
 	return 0;
 }
 
-/* Nothing here yet */
-void bt865_inc_use (struct i2c_client *client)
-{
-#ifdef MODULE
-	MOD_INC_USE_COUNT;
-#endif
-}
-
-/* Nothing here yet */
-void bt865_dec_use (struct i2c_client *client)
-{
-#ifdef MODULE
-	MOD_DEC_USE_COUNT;
-#endif
-}
-
 /* ----------------------------------------------------------------------- */
 
 
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w5/3rdparty/mod_dvb/em8300/modules/em8300_audio.c w4/3rdparty/mod_dvb/em8300/modules/em8300_audio.c
--- w5/3rdparty/mod_dvb/em8300/modules/em8300_audio.c	2003-04-17 00:22:16.000000000 +0200
+++ w4/3rdparty/mod_dvb/em8300/modules/em8300_audio.c	2003-08-07 18:33:11.000000000 +0200
@@ -15,6 +15,7 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_reg.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w5/3rdparty/mod_dvb/em8300/modules/em8300_dicom.c w4/3rdparty/mod_dvb/em8300/modules/em8300_dicom.c
--- w5/3rdparty/mod_dvb/em8300/modules/em8300_dicom.c	2003-04-17 00:22:16.000000000 +0200
+++ w4/3rdparty/mod_dvb/em8300/modules/em8300_dicom.c	2003-08-07 18:33:13.000000000 +0200
@@ -14,6 +14,7 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_reg.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w5/3rdparty/mod_dvb/em8300/modules/em8300_fifo.c w4/3rdparty/mod_dvb/em8300/modules/em8300_fifo.c
--- w5/3rdparty/mod_dvb/em8300/modules/em8300_fifo.c	2003-04-17 00:22:16.000000000 +0200
+++ w4/3rdparty/mod_dvb/em8300/modules/em8300_fifo.c	2003-08-07 18:33:16.000000000 +0200
@@ -41,6 +41,7 @@
 #include <linux/version.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_reg.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w5/3rdparty/mod_dvb/em8300/modules/em8300_i2c.c w4/3rdparty/mod_dvb/em8300/modules/em8300_i2c.c
--- w5/3rdparty/mod_dvb/em8300/modules/em8300_i2c.c	2003-04-17 00:22:16.000000000 +0200
+++ w4/3rdparty/mod_dvb/em8300/modules/em8300_i2c.c	2003-08-07 18:38:38.000000000 +0200
@@ -14,6 +14,7 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_reg.h"
@@ -123,13 +124,6 @@ static int em8300_i2c_unreg(struct i2c_c
 	return 0;
 }
 
-static void em8300_i2c_inc(struct i2c_adapter *adapter)
-{	
-}
-
-static void em8300_i2c_dec(struct i2c_adapter *adapter)
-{	
-}
 
 /* ----------------------------------------------------------------------- */
 /* I2C functions							   */
@@ -178,12 +172,12 @@ int em8300_i2c_init(struct em8300_s *em)
 	
 	em->i2c_data_1.data = pdata;
 
+	em->i2c_ops_1.owner = THIS_MODULE;
 	strcpy(em->i2c_ops_1.name, "EM8300 I2C bus 1");
 	em->i2c_ops_1.id = I2C_HW_B_EM8300;
 	em->i2c_ops_1.algo = NULL;
 	em->i2c_ops_1.algo_data = &em->i2c_data_1;
-	em->i2c_ops_1.inc_use = em8300_i2c_inc;
-	em->i2c_ops_1.dec_use = em8300_i2c_dec;
+
 	em->i2c_ops_1.client_register = em8300_i2c_reg;
 	em->i2c_ops_1.client_unregister = em8300_i2c_unreg;
 	em->i2c_ops_1.data = em;
@@ -213,12 +207,11 @@ int em8300_i2c_init(struct em8300_s *em)
 	
 	em->i2c_data_2.data = pdata;
 
+	em->i2c_ops_2.owner = THIS_MODULE;
 	strcpy(em->i2c_ops_2.name, "EM8300 I2C bus 2");
 	em->i2c_ops_2.id = I2C_HW_B_EM8300;
 	em->i2c_ops_2.algo = NULL;
 	em->i2c_ops_2.algo_data = &em->i2c_data_2;
-	em->i2c_ops_2.inc_use = em8300_i2c_inc;
-	em->i2c_ops_2.dec_use = em8300_i2c_dec;
 	em->i2c_ops_2.client_register = em8300_i2c_reg;
 	em->i2c_ops_2.client_unregister = em8300_i2c_unreg;
 	em->i2c_ops_2.data = em;
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w5/3rdparty/mod_dvb/em8300/modules/em8300_ioctl.c w4/3rdparty/mod_dvb/em8300/modules/em8300_ioctl.c
--- w5/3rdparty/mod_dvb/em8300/modules/em8300_ioctl.c	2003-04-17 00:22:16.000000000 +0200
+++ w4/3rdparty/mod_dvb/em8300/modules/em8300_ioctl.c	2003-08-07 18:33:21.000000000 +0200
@@ -14,6 +14,7 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_reg.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w5/3rdparty/mod_dvb/em8300/modules/em8300_main.c w4/3rdparty/mod_dvb/em8300/modules/em8300_main.c
--- w5/3rdparty/mod_dvb/em8300/modules/em8300_main.c	2003-04-17 00:22:16.000000000 +0200
+++ w4/3rdparty/mod_dvb/em8300/modules/em8300_main.c	2003-08-07 18:33:25.000000000 +0200
@@ -42,6 +42,7 @@
 #include <linux/version.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "encoder.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w5/3rdparty/mod_dvb/em8300/modules/em8300_misc.c w4/3rdparty/mod_dvb/em8300/modules/em8300_misc.c
--- w5/3rdparty/mod_dvb/em8300/modules/em8300_misc.c	2003-04-17 00:22:16.000000000 +0200
+++ w4/3rdparty/mod_dvb/em8300/modules/em8300_misc.c	2003-08-07 18:33:29.000000000 +0200
@@ -14,6 +14,7 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_reg.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w5/3rdparty/mod_dvb/em8300/modules/em8300_spu.c w4/3rdparty/mod_dvb/em8300/modules/em8300_spu.c
--- w5/3rdparty/mod_dvb/em8300/modules/em8300_spu.c	2003-04-17 00:22:16.000000000 +0200
+++ w4/3rdparty/mod_dvb/em8300/modules/em8300_spu.c	2003-08-07 18:33:31.000000000 +0200
@@ -14,6 +14,7 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_reg.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w5/3rdparty/mod_dvb/em8300/modules/em8300_ucode.c w4/3rdparty/mod_dvb/em8300/modules/em8300_ucode.c
--- w5/3rdparty/mod_dvb/em8300/modules/em8300_ucode.c	2003-04-17 00:22:16.000000000 +0200
+++ w4/3rdparty/mod_dvb/em8300/modules/em8300_ucode.c	2003-08-07 18:33:34.000000000 +0200
@@ -14,6 +14,7 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_reg.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w5/3rdparty/mod_dvb/em8300/modules/em8300_video.c w4/3rdparty/mod_dvb/em8300/modules/em8300_video.c
--- w5/3rdparty/mod_dvb/em8300/modules/em8300_video.c	2003-04-17 00:22:16.000000000 +0200
+++ w4/3rdparty/mod_dvb/em8300/modules/em8300_video.c	2003-08-07 18:33:36.000000000 +0200
@@ -14,6 +14,7 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 
 #include "em8300_reg.h"
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w5/3rdparty/mod_dvb/em8300/modules/em9010.c w4/3rdparty/mod_dvb/em8300/modules/em9010.c
--- w5/3rdparty/mod_dvb/em8300/modules/em9010.c	2003-04-17 00:22:16.000000000 +0200
+++ w4/3rdparty/mod_dvb/em8300/modules/em9010.c	2003-08-07 18:33:39.000000000 +0200
@@ -14,6 +14,7 @@
 #include <asm/io.h>
 #include <asm/uaccess.h>
 
+#include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 #include <linux/time.h>
 
