diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/dc395x_trm/dc395x_trm.h w1/3rdparty/dc395x_trm/dc395x_trm.h
--- w2/3rdparty/dc395x_trm/dc395x_trm.h	2002-08-30 00:34:18.000000000 +0200
+++ w1/3rdparty/dc395x_trm/dc395x_trm.h	2003-08-20 20:06:33.000000000 +0200
@@ -64,7 +64,7 @@
 #endif
 
 #if (LINUX_VERSION_CODE < KERNEL_VERSION (2,1,70)) || defined(OLD_EH)
-# define EH_FUNC	abort:		DC395x_abort,		\	
+# define EH_FUNC	abort:		DC395x_abort,		\
 			reset:		DC395x_reset,
 # define NORM_REC_LVL 1
 #else
