diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/thinkpad/smapi_call.S w1/3rdparty/thinkpad/smapi_call.S
--- w2/3rdparty/thinkpad/smapi_call.S	2003-04-08 02:59:07.000000000 +0200
+++ w1/3rdparty/thinkpad/smapi_call.S	2003-08-18 20:33:24.000000000 +0200
@@ -47,7 +47,7 @@ SMAPI_BIOS_do:
 
 	movl	$0,%eax
 	movw	%cs,wSegSmapiBios
-	lcall	farpxSmapiBios
+	lcall	*farpxSmapiBios
 
 	add	$0x10,%esp
 	popl	%ebp
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/thinkpad/thinkpadpm.c w1/3rdparty/thinkpad/thinkpadpm.c
--- w2/3rdparty/thinkpad/thinkpadpm.c	2003-04-08 02:59:07.000000000 +0200
+++ w1/3rdparty/thinkpad/thinkpadpm.c	2003-08-18 20:34:57.000000000 +0200
@@ -204,7 +204,7 @@ static u8 apm_bios_call(u32 func, u32 eb
 	__asm__ __volatile__(APM_DO_ZERO_SEGS
 		"pushl %%edi\n\t"
 		"pushl %%ebp\n\t"
-		"lcall %%cs:" SYMBOL_NAME_STR(apm_bios_entry) "\n\t"
+		"lcall *%%cs:" SYMBOL_NAME_STR(apm_bios_entry) "\n\t"
 		"setc %%al\n\t"
 		"popl %%ebp\n\t"
 		"popl %%edi\n\t"
@@ -251,7 +251,7 @@ static u8 apm_bios_call_simple(u32 func,
 		__asm__ __volatile__(APM_DO_ZERO_SEGS
 			"pushl %%edi\n\t"
 			"pushl %%ebp\n\t"
-			"lcall %%cs:" SYMBOL_NAME_STR(apm_bios_entry) "\n\t"
+			"lcall *%%cs:" SYMBOL_NAME_STR(apm_bios_entry) "\n\t"
 			"setc %%bl\n\t"
 			"popl %%ebp\n\t"
 			"popl %%edi\n\t"
