diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/arm/Config.help w2/sound/arm/Config.help
--- w1/sound/arm/Config.help	1970-01-01 01:00:00.000000000 +0100
+++ w2/sound/arm/Config.help	2003-08-13 23:48:50.000000000 +0200
@@ -0,0 +1,3 @@
+CONFIG_SND_SA11XX_UDA1341
+  Say Y or M if you have a Compaq iPaq H3x00 handheld computer and want
+  to use its Philips UDA 1341 audio chip.
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/arm/Config.in w2/sound/arm/Config.in
--- w1/sound/arm/Config.in	1970-01-01 01:00:00.000000000 +0100
+++ w2/sound/arm/Config.in	2003-08-13 23:48:50.000000000 +0200
@@ -0,0 +1,10 @@
+# ALSA ARM drivers
+
+mainmenu_option next_comment
+comment 'ALSA ARM devices'
+
+if [ "$CONFIG_ARCH_SA1100" = "y" ]; then
+  dep_tristate 'SA11xx UDA1341TS driver (H3600)' CONFIG_SND_SA11XX_UDA1341 $CONFIG_SND $CONFIG_L3
+fi
+
+endmenu
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/arm/Kconfig w2/sound/arm/Kconfig
--- w1/sound/arm/Kconfig	2002-10-31 14:44:20.000000000 +0100
+++ w2/sound/arm/Kconfig	1970-01-01 01:00:00.000000000 +0100
@@ -1,14 +0,0 @@
-# ALSA ARM drivers
-
-menu "ALSA ARM devices"
-	depends on SND!=n && ARM
-
-config SND_SA11XX_UDA1341
-	tristate "SA11xx UDA1341TS driver (H3600)"
-	depends on ARCH_SA1100 && SND && L3
-	help
-	  Say Y or M if you have a Compaq iPaq H3x00 handheld computer and want
-	  to use its Philips UDA 1341 audio chip.
-
-endmenu
-
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/Config.help w2/sound/Config.help
--- w1/sound/Config.help	1970-01-01 01:00:00.000000000 +0100
+++ w2/sound/Config.help	2003-08-13 23:48:50.000000000 +0200
@@ -0,0 +1,7 @@
+CONFIG_SOUND_PRIME
+  Say 'Y' or 'M' to enable Open Sound System drivers.
+
+CONFIG_SOUND_SND
+  Say 'Y' or 'M' to enable Advanced Linux Sound Architecture (ALSA) drivers.
+  You need to install also alsa-lib and alsa-utils packages available on
+  the ALSA website at <http://www.alsa-project.org>.
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/Config.in w2/sound/Config.in
--- w1/sound/Config.in	1970-01-01 01:00:00.000000000 +0100
+++ w2/sound/Config.in	2003-08-13 23:51:56.000000000 +0200
@@ -0,0 +1,38 @@
+# sound/Config.in
+#
+
+mainmenu_option next_comment
+comment 'Advanced Linux Sound Architecture'
+
+dep_tristate 'Advanced Linux Sound Architecture' CONFIG_SND $CONFIG_SOUND
+if [ "$CONFIG_SND" != "n" ]; then
+  source sound/core/Config.in
+  source sound/drivers/Config.in
+fi
+if [ "$CONFIG_SND" != "n" -a "$CONFIG_ISA" = "y" ]; then
+  source sound/isa/Config.in
+fi
+if [ "$CONFIG_SND" != "n" -a "$CONFIG_PCI" = "y" ]; then
+  source sound/pci/Config.in
+fi
+if [ "$CONFIG_SND" != "n" -a "$CONFIG_PPC" = "y" ]; then
+  source sound/ppc/Config.in
+fi
+if [ "$CONFIG_SND" != "n" -a "$CONFIG_ARM" = "y" ]; then
+  source sound/arm/Config.in
+fi
+if [ "$CONFIG_SND" != "n" ]; then
+  source sound/pcmcia/Config.in
+fi
+# the following will depenend on the order of config.
+# here assuming USB is defined before ALSA
+if [ "$CONFIG_SND" != "n" -a "$CONFIG_USB" != "n" ]; then
+  source sound/usb/Config.in
+fi
+if [ "$CONFIG_SND" != "n" ]; then
+  if [ "$CONFIG_SPARC32" = "y" -o "$CONFIG_SPARC64" = "y" ]; then
+    source sound/sparc/Config.in
+  fi
+fi
+
+endmenu
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/core/Config.help w2/sound/core/Config.help
--- w1/sound/core/Config.help	1970-01-01 01:00:00.000000000 +0100
+++ w2/sound/core/Config.help	2003-08-13 23:48:50.000000000 +0200
@@ -0,0 +1,40 @@
+CONFIG_SND_SEQUENCER
+  Say 'Y' or 'M' to enable MIDI sequencer and router support. This feature
+  allows routing and enqueing MIDI events. Events can be processed at given
+  time.
+
+CONFIG_SND_SEQ_DUMMY
+  Say 'Y' or 'M' to enable dummy sequencer client. This client is a simple
+  midi-through client. All normal input events are redirected to output port
+  immediately.
+
+CONFIG_SND_OSSEMUL
+  Say 'Y' to enable OSS (Open Sound System) API emulation code.
+
+CONFIG_SND_MIXER_OSS
+  Say 'Y' or 'M' to enable mixer OSS API emulation (/dev/mixer*).
+
+CONFIG_SND_PCM_OSS
+  Say 'Y' or 'M' to enable digital audio (PCM) OSS API emulation (/dev/dsp*).
+
+CONFIG_SND_SEQUENCER_OSS
+  Say 'Y' or 'M' to enable OSS sequencer emulation (both /dev/sequencer and
+  /dev/music interfaces).
+
+CONFIG_SND_RTCTIMER
+  Say 'Y' or 'M' to enable RTC timer support for ALSA. ALSA code uses RTC
+  timer as precise timing source and maps the RTC timer to the ALSA's timer
+  interface. ALSA sequencer code can also use this timing source.
+
+CONFIG_SND_VERBOSE_PRINTK
+  Say 'Y' to enable verbose log messages. These messages will help to
+  identify source file and position containing printed messages.
+
+CONFIG_SND_DEBUG
+  Say 'Y' to enable ALSA debug code.
+
+CONFIG_SND_DEBUG_MEMORY
+  Say 'Y' to enable debugging of memory allocation.
+
+CONFIG_SND_DEBUG_DETECTION
+  Say 'Y' to enable debugging of hardware detection.
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/core/Config.in w2/sound/core/Config.in
--- w1/sound/core/Config.in	1970-01-01 01:00:00.000000000 +0100
+++ w2/sound/core/Config.in	2003-08-13 23:48:50.000000000 +0200
@@ -0,0 +1,26 @@
+# ALSA soundcard-configuration
+
+if [ "$CONFIG_X86_64" = "y" -a "$CONFIG_IA32_EMULATION" = "y" ]; then
+  dep_tristate '  Emulation for 32-bit applications' CONFIG_SND_BIT32_EMUL $CONFIG_SND
+fi
+if [ "$CONFIG_PPC64" = "y" ]; then
+  dep_tristate '  Emulation for 32-bit applications' CONFIG_SND_BIT32_EMUL $CONFIG_SND
+fi
+if [ "$CONFIG_SPARC64" = "y" ]; then
+  dep_tristate '  Emulation for 32-bit applications' CONFIG_SND_BIT32_EMUL $CONFIG_SND
+fi
+dep_tristate '  Sequencer support' CONFIG_SND_SEQUENCER $CONFIG_SND
+if [ "$CONFIG_SND_SEQUENCER" != "n" ]; then
+  dep_tristate '  Sequencer dummy client' CONFIG_SND_SEQ_DUMMY $CONFIG_SND_SEQUENCER
+fi
+dep_mbool '  OSS API emulation' CONFIG_SND_OSSEMUL $CONFIG_SND
+if [ "$CONFIG_SND_OSSEMUL" = "y" ]; then
+  dep_tristate '    OSS Mixer API' CONFIG_SND_MIXER_OSS $CONFIG_SND
+  dep_tristate '    OSS PCM (digital audio) API' CONFIG_SND_PCM_OSS $CONFIG_SND
+  dep_mbool '    OSS Sequencer API' CONFIG_SND_SEQUENCER_OSS $CONFIG_SND_SEQUENCER
+fi
+dep_tristate '  RTC Timer support' CONFIG_SND_RTCTIMER $CONFIG_SND $CONFIG_RTC
+bool '  Verbose printk' CONFIG_SND_VERBOSE_PRINTK
+bool '  Debug' CONFIG_SND_DEBUG
+dep_bool '    Debug memory' CONFIG_SND_DEBUG_MEMORY $CONFIG_SND_DEBUG
+dep_bool '    Debug detection' CONFIG_SND_DEBUG_DETECT $CONFIG_SND_DEBUG
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/core/Kconfig w2/sound/core/Kconfig
--- w1/sound/core/Kconfig	2003-06-10 16:07:20.000000000 +0200
+++ w2/sound/core/Kconfig	1970-01-01 01:00:00.000000000 +0100
@@ -1,77 +0,0 @@
-# ALSA soundcard-configuration
-config SND_BIT32_EMUL
-	tristate "Emulation for 32-bit applications"
-	depends on SND && (SPARC64 || PPC64 || X86_64 && IA32_EMULATION)
-
-config SND_SEQUENCER
-	tristate "Sequencer support"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to enable MIDI sequencer and router support. This feature
-	  allows routing and enqueing MIDI events. Events can be processed at given
-	  time.
-
-config SND_SEQ_DUMMY
-	tristate "Sequencer dummy client"
-	depends on SND_SEQUENCER
-	help
-	  Say 'Y' or 'M' to enable dummy sequencer client. This client is a simple
-	  midi-through client. All normal input events are redirected to output port
-	  immediately.
-
-config SND_OSSEMUL
-	bool "OSS API emulation"
-	depends on SND
-	help
-	  Say 'Y' to enable OSS (Open Sound System) API emulation code.
-
-config SND_MIXER_OSS
-	tristate "OSS Mixer API"
-	depends on SND_OSSEMUL && SND
-	help
-	  Say 'Y' or 'M' to enable mixer OSS API emulation (/dev/mixer*).
-
-config SND_PCM_OSS
-	tristate "OSS PCM (digital audio) API"
-	depends on SND_OSSEMUL && SND
-	help
-	  Say 'Y' or 'M' to enable digital audio (PCM) OSS API emulation (/dev/dsp*).
-
-config SND_SEQUENCER_OSS
-	bool "OSS Sequencer API"
-	depends on SND_OSSEMUL && SND_SEQUENCER
-	help
-	  Say 'Y' to enable OSS sequencer emulation (both /dev/sequencer and
-	  /dev/music interfaces).
-
-config SND_RTCTIMER
-	tristate "RTC Timer support"
-	depends on SND && RTC
-	help
-	  Say 'Y' or 'M' to enable RTC timer support for ALSA. ALSA code uses RTC
-	  timer as precise timing source and maps the RTC timer to the ALSA's timer
-	  interface. ALSA sequencer code can also use this timing source.
-
-config SND_VERBOSE_PRINTK
-	bool "Verbose printk"
-	depends on SND
-	help
-	  Say 'Y' to enable verbose log messages. These messages will help to
-	  identify source file and position containing printed messages.
-
-config SND_DEBUG
-	bool "Debug"
-	depends on SND
-	help
-	  Say 'Y' to enable ALSA debug code.
-
-config SND_DEBUG_MEMORY
-	bool "Debug memory"
-	depends on SND_DEBUG
-	help
-	  Say 'Y' to enable debugging of memory allocation.
-
-config SND_DEBUG_DETECT
-	bool "Debug detection"
-	depends on SND_DEBUG
-
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/drivers/Config.help w2/sound/drivers/Config.help
--- w1/sound/drivers/Config.help	1970-01-01 01:00:00.000000000 +0100
+++ w2/sound/drivers/Config.help	2003-08-13 23:48:50.000000000 +0200
@@ -0,0 +1,18 @@
+CONFIG_SND_DUMMY
+  Say 'Y' or 'M' to include dummy driver. This driver does nothing, but
+  emulates various mixer controls and PCM devices.
+  
+CONFIG_SND_VIRMIDI
+  Say 'Y' or 'M' to include virtual MIDI driver. This driver allows to
+  connect applications using raw MIDI devices to sequencer.
+
+CONFIG_SND_MTPAV
+  Say 'Y' or 'M' to include support for MOTU MidiTimePiece AV multiport
+  MIDI adapter.
+
+CONFIG_SND_SERIAL_U16550
+  Say 'Y' or 'M' to include support for MIDI serial port driver. It works
+  with serial UARTs 16550 and better.
+
+CONFIG_SND_MPU401
+  Say 'Y' or 'M' to include support for MPU401 hardware using UART access.
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/drivers/Config.in w2/sound/drivers/Config.in
--- w1/sound/drivers/Config.in	1970-01-01 01:00:00.000000000 +0100
+++ w2/sound/drivers/Config.in	2003-08-13 23:48:50.000000000 +0200
@@ -0,0 +1,12 @@
+# ALSA generic drivers
+
+mainmenu_option next_comment
+comment 'Generic devices'
+
+dep_tristate 'Dummy (/dev/null) soundcard' CONFIG_SND_DUMMY $CONFIG_SND
+dep_tristate 'Virtual MIDI soundcard' CONFIG_SND_VIRMIDI $CONFIG_SND $CONFIG_SND_SEQUENCER
+dep_tristate 'MOTU MidiTimePiece AV multiport MIDI' CONFIG_SND_MTPAV $CONFIG_SND
+dep_tristate 'UART16550 - MIDI only driver' CONFIG_SND_SERIAL_U16550 $CONFIG_SND
+dep_tristate 'Generic MPU-401 UART driver' CONFIG_SND_MPU401 $CONFIG_SND
+
+endmenu
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/drivers/Kconfig w2/sound/drivers/Kconfig
--- w1/sound/drivers/Kconfig	2002-10-31 14:44:21.000000000 +0100
+++ w2/sound/drivers/Kconfig	1970-01-01 01:00:00.000000000 +0100
@@ -1,42 +0,0 @@
-# ALSA generic drivers
-
-menu "Generic devices"
-	depends on SND!=n
-
-config SND_DUMMY
-	tristate "Dummy (/dev/null) soundcard"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include dummy driver. This driver does nothing, but
-	  emulates various mixer controls and PCM devices.
-	  
-
-config SND_VIRMIDI
-	tristate "Virtual MIDI soundcard"
-	depends on SND_SEQUENCER
-	help
-	  Say 'Y' or 'M' to include virtual MIDI driver. This driver allows to
-	  connect applications using raw MIDI devices to sequencer.
-
-config SND_MTPAV
-	tristate "MOTU MidiTimePiece AV multiport MIDI"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for MOTU MidiTimePiece AV multiport
-	  MIDI adapter.
-
-config SND_SERIAL_U16550
-	tristate "UART16550 - MIDI only driver"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for MIDI serial port driver. It works
-	  with serial UARTs 16550 and better.
-
-config SND_MPU401
-	tristate "Generic MPU-401 UART driver"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for MPU401 hardware using UART access.
-
-endmenu
-
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/isa/Config.help w2/sound/isa/Config.help
--- w1/sound/isa/Config.help	1970-01-01 01:00:00.000000000 +0100
+++ w2/sound/isa/Config.help	2003-08-13 23:48:50.000000000 +0200
@@ -0,0 +1,99 @@
+CONFIG_SND_AD1816A
+  Say 'Y' or 'M' to include support for Analog Devices SoundPort AD1816A or
+  compatible sound chips.
+
+CONFIG_SND_AD1848
+  Say 'Y' or 'M' to include support for AD1848 (Analog Devices) or CS4248 
+  (Cirrus Logic - Crystal Semiconductors) chips. Please, for newer chips
+  from Cirrus Logic, use CS4231, CS4232 or CS4236+ driver.
+
+CONFIG_SND_CS4231
+  Say 'Y' or 'M' to include support for CS4231 chips from Cirrus Logic -
+  Crystal Semiconductors.
+
+CONFIG_SND_CS4232
+  Say 'Y' or 'M' to include support for CS4232 chips from Cirrus Logic -
+  Crystal Semiconductors.
+
+CONFIG_SND_CS4236
+  Say 'Y' or 'M' to include support for CS4235,CS4236,CS4237B,CS4238B,CS4239
+  chips from Cirrus Logic - Crystal Semiconductors.
+
+CONFIG_SND_ES968
+  Say 'Y' or 'M' to include support for ESS AudioDrive ES968 chip.
+
+CONFIG_SND_ES1688
+  Say 'Y' or 'M' to include support for ESS AudioDrive ES688 or ES1688 chips.
+
+CONFIG_SND_ES18XX
+  Say 'Y' or 'M' to include support for ESS AudioDrive ES18xx chips.
+
+CONFIG_SND_GUSCLASSIC
+  Say 'Y' or 'M' to include support for Gravis UltraSound Classic soundcard.
+
+CONFIG_SND_GUSEXTREME
+  Say 'Y' or 'M' to include support for Gravis UltraSound Extreme soundcard.
+
+CONFIG_SND_GUSMAX
+  Say 'Y' or 'M' to include support for Gravis UltraSound MAX soundcard.
+
+CONFIG_SND_INTERWAVE
+  Say 'Y' or 'M' to include support for AMD InterWave based soundcards
+  (Gravis UltraSound Plug & Play, STB SoundRage32, MED3210, Dynasonic Pro,
+  Panasonic PCA761AW).
+
+CONFIG_SND_INTERWAVE_STB
+  Say 'Y' or 'M' to include support for AMD InterWave based soundcards
+  with TEA6330T bass and treble regulator (UltraSound 32-Pro).
+
+CONFIG_SND_OPTI92X_AD1848
+  Say 'Y' or 'M' to include support for Opti92x soundcards equiped with
+  AD1848 codec.
+
+CONFIG_SND_OPTI92X_CS4231
+  Say 'Y' or 'M' to include support for Opti92x soundcards equiped with
+  CS4231 codec.
+
+CONFIG_SND_OPTI93X
+  Say 'Y' or 'M' to include support for Opti93x soundcards.
+
+CONFIG_SND_SB8
+  Say 'Y' or 'M' to include support for Sound Blaster 1.0/2.0/Pro (8-bit)
+  soundcards or 100% compatible from Creative.
+
+CONFIG_SND_SB16
+  Say 'Y' or 'M' to include support for Sound Blaster 16 (including
+  Plug and Play version).
+
+CONFIG_SND_SBAWE
+  Say 'Y' or 'M' to include support for Sound Blaster AWE (including
+  Plug and Play version).
+
+CONFIG_SND_SB16_CSP
+  Say 'Y' to include support for CSP core. This special coprocessor
+  can do variable tasks like various compression and decompression
+  algorithms.
+
+CONFIG_SND_WAVEFRONT
+  Say 'Y' or 'M' to include support for Turtle Beach Maui, Tropez
+  and Tropez+ soundcards based on Wavefront chip.
+
+CONFIG_SND_ALS100
+  Say 'Y' or 'M' to include support for Avance Logic ALS100, ALS110,
+  ALS120 and ALS200 soundcards.
+
+CONFIG_SND_AZT2320
+  Say 'Y' or 'M' to include support for Aztech Systems AZT2320 soundcard.
+
+CONFIG_SND_CMI8330
+  Say 'Y' or 'M' to include support for C-Media CMI8330 based soundcards.
+
+CONFIG_SND_DT019X
+  Say 'Y' or 'M' to include support for Diamond Technologies DT-019X and
+  Avance Logic ALS-007 soundcards.
+
+CONFIG_SND_OPL3SA2
+  Say 'Y' or 'M' to include support for Yamaha OPL3SA2 or OPL3SA3 chips.
+
+CONFIG_SND_SGALAXY
+  Say 'Y' or 'M' to include support for Aztech Sound Galaxy.
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/isa/Config.in w2/sound/isa/Config.in
--- w1/sound/isa/Config.in	1970-01-01 01:00:00.000000000 +0100
+++ w2/sound/isa/Config.in	2003-08-13 23:49:11.000000000 +0200
@@ -0,0 +1,37 @@
+# ALSA ISA drivers
+
+mainmenu_option next_comment
+comment 'ISA devices'
+
+dep_tristate 'Analog Devices SoundPort AD1816A' CONFIG_SND_AD1816A $CONFIG_SND $CONFIG_ISAPNP
+dep_tristate 'Generic AD1848/CS4248 driver' CONFIG_SND_AD1848 $CONFIG_SND
+dep_tristate 'Generic Cirrus Logic CS4231 driver' CONFIG_SND_CS4231 $CONFIG_SND
+dep_tristate 'Generic Cirrus Logic CS4232 driver' CONFIG_SND_CS4232 $CONFIG_SND
+dep_tristate 'Generic Cirrus Logic CS4236+ driver' CONFIG_SND_CS4236 $CONFIG_SND
+dep_tristate 'Generic ESS ES968 driver' CONFIG_SND_ES968 $CONFIG_SND $CONFIG_ISAPNP
+dep_tristate 'Generic ESS ES688/ES1688 driver' CONFIG_SND_ES1688 $CONFIG_SND
+dep_tristate 'Generic ESS ES18xx driver' CONFIG_SND_ES18XX $CONFIG_SND
+dep_tristate 'Gravis UltraSound Classic' CONFIG_SND_GUSCLASSIC $CONFIG_SND
+dep_tristate 'Gravis UltraSound Extreme' CONFIG_SND_GUSEXTREME $CONFIG_SND
+dep_tristate 'Gravis UltraSound MAX' CONFIG_SND_GUSMAX $CONFIG_SND
+dep_tristate 'AMD InterWave, Gravis UltraSound PnP' CONFIG_SND_INTERWAVE $CONFIG_SND
+dep_tristate 'AMD InterWave + TEA6330T (UltraSound 32-Pro)' CONFIG_SND_INTERWAVE_STB $CONFIG_SND
+dep_tristate 'OPTi 82C92x - AD1848' CONFIG_SND_OPTI92X_AD1848 $CONFIG_SND
+dep_tristate 'OPTi 82C92x - CS4231' CONFIG_SND_OPTI92X_CS4231 $CONFIG_SND
+dep_tristate 'OPTi 82C93x' CONFIG_SND_OPTI93X $CONFIG_SND
+dep_tristate 'Sound Blaster 1.0/2.0/Pro (8-bit)' CONFIG_SND_SB8 $CONFIG_SND
+dep_tristate 'Sound Blaster 16 (PnP)' CONFIG_SND_SB16 $CONFIG_SND
+dep_tristate 'Sound Blaster AWE (32,64) (PnP)' CONFIG_SND_SBAWE $CONFIG_SND
+if [ "$CONFIG_SND_SB16" != "n" -o "$CONFIG_SND_SBAWE" != "n" ]; then
+  bool '  Sound Blaster 16/AWE CSP support' CONFIG_SND_SB16_CSP
+fi
+dep_tristate 'Turtle Beach Maui,Tropez,Tropez+ (Wavefront)' CONFIG_SND_WAVEFRONT $CONFIG_SND
+dep_tristate 'Avance Logic ALS100/ALS120' CONFIG_SND_ALS100 $CONFIG_SND $CONFIG_ISAPNP
+dep_tristate 'Aztech Systems AZT2320' CONFIG_SND_AZT2320 $CONFIG_SND $CONFIG_ISAPNP
+dep_tristate 'C-Media CMI8330' CONFIG_SND_CMI8330 $CONFIG_SND
+dep_tristate 'Diamond Technologies DT-019X, Avance Logic ALS-007' CONFIG_SND_DT019X $CONFIG_SND $CONFIG_ISAPNP
+dep_tristate 'Yamaha OPL3-SA2/SA3' CONFIG_SND_OPL3SA2 $CONFIG_SND
+dep_tristate 'Aztech Sound Galaxy' CONFIG_SND_SGALAXY $CONFIG_SND
+dep_tristate 'Ensoniq SoundScape PnP' CONFIG_SND_SSCAPE $CONFIG_SND
+
+endmenu
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/isa/Kconfig w2/sound/isa/Kconfig
--- w1/sound/isa/Kconfig	2003-05-20 14:42:44.000000000 +0200
+++ w2/sound/isa/Kconfig	1970-01-01 01:00:00.000000000 +0100
@@ -1,201 +0,0 @@
-# ALSA ISA drivers
-
-menu "ISA devices"
-	depends on SND!=n && ISA
-
-config SND_AD1816A
-	tristate "Analog Devices SoundPort AD1816A"
-	depends on SND && ISAPNP
-	help
-	  Say 'Y' or 'M' to include support for Analog Devices SoundPort AD1816A or
-	  compatible sound chips.
-
-config SND_AD1848
-	tristate "Generic AD1848/CS4248 driver"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for AD1848 (Analog Devices) or CS4248 
-	  (Cirrus Logic - Crystal Semiconductors) chips. Please, for newer chips
-	  from Cirrus Logic, use CS4231, CS4232 or CS4236+ driver.
-
-config SND_CS4231
-	tristate "Generic Cirrus Logic CS4231 driver"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for CS4231 chips from Cirrus Logic -
-	  Crystal Semiconductors.
-
-config SND_CS4232
-	tristate "Generic Cirrus Logic CS4232 driver"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for CS4232 chips from Cirrus Logic -
-	  Crystal Semiconductors.
-
-config SND_CS4236
-	tristate "Generic Cirrus Logic CS4236+ driver"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for CS4235,CS4236,CS4237B,CS4238B,CS4239
-	  chips from Cirrus Logic - Crystal Semiconductors.
-
-config SND_PC98_CS4232
-	tristate "NEC PC9800 CS4232 driver"
-	depends on SND && X86_PC9800
-	help
-	  Say 'Y' or 'M' to include support for NEC PC-9801/PC-9821 on-board
-	  soundchip based on CS4232.
-
-config SND_ES968
-	tristate "Generic ESS ES968 driver"
-	depends on SND && ISAPNP
-	help
-	  Say 'Y' or 'M' to include support for ESS AudioDrive ES968 chip.
-
-config SND_ES1688
-	tristate "Generic ESS ES688/ES1688 driver"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for ESS AudioDrive ES688 or ES1688 chips.
-
-config SND_ES18XX
-	tristate "Generic ESS ES18xx driver"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for ESS AudioDrive ES18xx chips.
-
-config SND_GUSCLASSIC
-	tristate "Gravis UltraSound Classic"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Gravis UltraSound Classic soundcard.
-
-config SND_GUSEXTREME
-	tristate "Gravis UltraSound Extreme"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Gravis UltraSound Extreme soundcard.
-
-config SND_GUSMAX
-	tristate "Gravis UltraSound MAX"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Gravis UltraSound MAX soundcard.
-
-config SND_INTERWAVE
-	tristate "AMD InterWave, Gravis UltraSound PnP"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for AMD InterWave based soundcards
-	  (Gravis UltraSound Plug & Play, STB SoundRage32, MED3210, Dynasonic Pro,
-	  Panasonic PCA761AW).
-
-config SND_INTERWAVE_STB
-	tristate "AMD InterWave + TEA6330T (UltraSound 32-Pro)"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for AMD InterWave based soundcards
-	  with TEA6330T bass and treble regulator (UltraSound 32-Pro).
-
-config SND_OPTI92X_AD1848
-	tristate "OPTi 82C92x - AD1848"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Opti92x soundcards equiped with
-	  AD1848 codec.
-
-config SND_OPTI92X_CS4231
-	tristate "OPTi 82C92x - CS4231"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Opti92x soundcards equiped with
-	  CS4231 codec.
-
-config SND_OPTI93X
-	tristate "OPTi 82C93x"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Opti93x soundcards.
-
-config SND_SB8
-	tristate "Sound Blaster 1.0/2.0/Pro (8-bit)"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Sound Blaster 1.0/2.0/Pro (8-bit)
-	  soundcards or 100% compatible from Creative.
-
-config SND_SB16
-	tristate "Sound Blaster 16 (PnP)"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Sound Blaster 16 (including
-	  Plug and Play version).
-
-config SND_SBAWE
-	tristate "Sound Blaster AWE (32,64) (PnP)"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Sound Blaster AWE (including
-	  Plug and Play version).
-
-config SND_SB16_CSP
-	bool "Sound Blaster 16/AWE CSP support"
-	depends on SND_SB16 || SND_SBAWE
-	help
-	  Say 'Y' to include support for CSP core. This special coprocessor
-	  can do variable tasks like various compression and decompression
-	  algorithms.
-
-config SND_WAVEFRONT
-	tristate "Turtle Beach Maui,Tropez,Tropez+ (Wavefront)"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Turtle Beach Maui, Tropez
-	  and Tropez+ soundcards based on Wavefront chip.
-
-config SND_ALS100
-	tristate "Avance Logic ALS100/ALS120"
-	depends on SND && ISAPNP
-	help
-	  Say 'Y' or 'M' to include support for Avance Logic ALS100, ALS110,
-	  ALS120 and ALS200 soundcards.
-
-config SND_AZT2320
-	tristate "Aztech Systems AZT2320"
-	depends on SND && ISAPNP
-	help
-	  Say 'Y' or 'M' to include support for Aztech Systems AZT2320 soundcard.
-
-config SND_CMI8330
-	tristate "C-Media CMI8330"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for C-Media CMI8330 based soundcards.
-
-config SND_DT019X
-	tristate "Diamond Technologies DT-019X, Avance Logic ALS-007"
-	depends on SND && ISAPNP
-	help
-	  Say 'Y' or 'M' to include support for Diamond Technologies DT-019X and
-	  Avance Logic ALS-007 soundcards.
-
-config SND_OPL3SA2
-	tristate "Yamaha OPL3-SA2/SA3"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Yamaha OPL3SA2 or OPL3SA3 chips.
-
-config SND_SGALAXY
-	tristate "Aztech Sound Galaxy"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Aztech Sound Galaxy.
-
-config SND_SSCAPE
-	tristate "Ensoniq SoundScape PnP driver"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Ensoniq SoundScape PnP
-	  soundcard.
-
-endmenu
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/Kconfig w2/sound/Kconfig
--- w1/sound/Kconfig	2003-05-23 18:00:37.000000000 +0200
+++ w2/sound/Kconfig	1970-01-01 01:00:00.000000000 +0100
@@ -1,48 +0,0 @@
-# sound/Config.in
-#
-
-menu "Advanced Linux Sound Architecture"
-	depends on SOUND!=n
-
-config SND
-	tristate "Advanced Linux Sound Architecture"
-	depends on SOUND
-
-source "sound/core/Kconfig"
-
-source "sound/drivers/Kconfig"
-
-source "sound/isa/Kconfig"
-
-source "sound/pci/Kconfig"
-
-source "sound/ppc/Kconfig"
-
-source "sound/arm/Kconfig"
-
-# the following will depenend on the order of config.
-# here assuming USB is defined before ALSA
-source "sound/usb/Kconfig"
-
-# the following will depenend on the order of config.
-# here assuming PCMCIA is defined before ALSA
-source "sound/pcmcia/Kconfig"
-
-source "sound/sparc/Kconfig"
-
-source "sound/parisc/Kconfig"
-
-endmenu
-
-menu "Open Sound System"
-	depends on SOUND!=n
-
-config SOUND_PRIME
-	tristate "Open Sound System (DEPRECATED)"
-	depends on SOUND
-	help
-	  Say 'Y' or 'M' to enable Open Sound System drivers.
-
-source "sound/oss/Kconfig"
-
-endmenu
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/parisc/Kconfig w2/sound/parisc/Kconfig
--- w1/sound/parisc/Kconfig	2003-05-23 12:07:28.000000000 +0200
+++ w2/sound/parisc/Kconfig	1970-01-01 01:00:00.000000000 +0100
@@ -1,13 +0,0 @@
-# ALSA PA-RISC drivers
-
-menu "ALSA PA-RISC devices"
-	depends on SND!=n && PARISC
-
-config SND_HARMONY
-	tristate "Harmony/Vivace sound chip"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Harmony/Vivace soundchip
-	  on HP712s, 715/new and many other GSC based machines.
-
-endmenu
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/pci/Config.help w2/sound/pci/Config.help
--- w1/sound/pci/Config.help	1970-01-01 01:00:00.000000000 +0100
+++ w2/sound/pci/Config.help	2003-08-13 23:48:50.000000000 +0200
@@ -0,0 +1,89 @@
+CONFIG_SND_ALI5451
+  Say 'Y' or 'M' to include support for ALI PCI Audio M5451 sound core.
+
+CONFIG_SND_CS46XX
+  Say 'Y' or 'M' to include support for Cirrus Logic CS4610 / CS4612 /
+  CS4614 / CS4615 / CS4622 / CS4624 / CS4630 / CS4280 chips.
+
+CONFIG_SND_CS46XX_NEW_DSP
+  Say 'Y' to use a new DSP image for SPDIF and dual codecs.
+
+CONFIG_SND_CS4281
+  Say 'Y' or 'M' to include support for Cirrus Logic CS4281.
+
+CONFIG_SND_EMU10K1
+  Say 'Y' or 'M' to include support for Sound Blaster PCI 512, Live!,
+  Audigy and E-mu APS (partially supported).
+
+CONFIG_SND_KORG1212
+  Say 'Y' or 'M' to include support for Korg 1212IO.
+
+CONFIG_SND_NM256
+  Say 'Y' or 'M' to include support for NeoMagic NM256AV/ZX chips.
+
+CONFIG_SND_RME96
+  Say 'Y' or 'M' to include support for RME Digi96, Digi96/8 and
+  Digi96/8 PRO/PAD/PST.
+
+CONFIG_SND_RME32
+  Say 'Y' or 'M' to include support for RME Digi32, Digi32/8 and
+  Digi32 PRO audio devices.
+
+CONFIG_SND_RME9652
+  Say 'Y' or 'M' to include support for RME Hammerfall (RME Digi9652 /
+  Digi9636) soundcards.
+
+CONFIG_SND_HDSP
+  Say 'Y' or 'M' to include support for RME Hammerfall DSP Audio
+  soundcards.
+
+CONFIG_SND_TRIDENT
+  Say 'Y' or 'M' to include support for Trident 4D-Wave DX/NX and
+  SiS 7018 soundcards.
+
+CONFIG_SND_YMFPCI
+  Say 'Y' or 'M' to include support for Yamaha PCI audio chips - 
+  YMF724, YMF724F, YMF740, YMF740C, YMF744, YMF754.
+
+CONFIG_SND_ALS4000
+  Say 'Y' or 'M' to include support for Avance Logic ALS4000.
+
+CONFIG_SND_CMIPCI
+  Say 'Y' or 'M' to include support for C-Media CMI8338 and 8738 PCI
+  soundcards.
+
+CONFIG_SND_ENS1370
+  Say 'Y' or 'M' to include support for Ensoniq AudioPCI ES1370.
+
+CONFIG_SND_ENS1371
+  Say 'Y' or 'M' to include support for Ensoniq AudioPCI ES1371 and
+  Sound Blaster PCI 64 or 128 soundcards.
+
+CONFIG_SND_ES1938
+  Say 'Y' or 'M' to include support for ESS Solo-1 (ES1938, ES1946)
+  soundcard.
+
+CONFIG_SND_ES1968
+  Say 'Y' or 'M' to include support for ESS Maestro 1/2/2E.
+
+CONFIG_SND_MAESTRO3
+  Say 'Y' or 'M' to include support for ESS Maestro 3 (Allegro) soundcard.
+
+CONFIG_SND_FM801
+  Say 'Y' or 'M' to include support for ForteMedia FM801 based soundcards.
+
+CONFIG_SND_ICE1712
+  Say 'Y' or 'M' to include support for ICE1712 (Envy24) based soundcards.
+  Currently supported hardware is: MidiMan M Audio - Delta 1010(LT), Dio 2496,
+  Delta 66/44, Audiophile 24/96; Hoontech SoundTrack DSP 24 (Value);
+  TerraTec - EWX 24/96, EWS 88MT, EWS 88D, DMX 6Fire.
+
+CONFIG_SND_INTEL8X0
+  Say 'Y' or 'M' to include support for Intel8x0 based soundcards,
+  SiS 7012, AMD768/8111, NVidia NForce and ALi 5455 chips.
+
+CONFIG_SND_SONICVIBES
+  Say 'Y' or 'M' to include support for S3 SonicVibes based soundcards.
+
+CONFIG_SND_VIA82XX
+  Say 'Y' or 'M' to include support for VIA VT82C686A/B, VT8233 South Bridge.
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/pci/Config.in w2/sound/pci/Config.in
--- w1/sound/pci/Config.in	1970-01-01 01:00:00.000000000 +0100
+++ w2/sound/pci/Config.in	2003-08-13 23:49:23.000000000 +0200
@@ -0,0 +1,35 @@
+# ALSA PCI drivers
+
+mainmenu_option next_comment
+comment 'PCI devices'
+
+dep_tristate 'ALi PCI Audio M5451' CONFIG_SND_ALI5451 $CONFIG_SND
+dep_tristate 'Cirrus Logic (Sound Fusion) CS4280/CS461x/CS462x/CS463x' CONFIG_SND_CS46XX $CONFIG_SND $CONFIG_SOUND_GAMEPORT
+dep_mbool '  Cirrus Logic (Sound Fusion) New DSP support (EXPERIMENTAL)' CONFIG_SND_CS46XX_NEW_DSP $CONFIG_SND_CS46XX $CONFIG_EXPERIMENTAL
+dep_tristate 'Cirrus Logic (Sound Fusion) CS4281' CONFIG_SND_CS4281 $CONFIG_SND $CONFIG_SOUND_GAMEPORT
+dep_tristate 'EMU10K1 (SB Live! & Audigy, E-mu APS)' CONFIG_SND_EMU10K1 $CONFIG_SND
+dep_tristate 'Korg 1212 IO' CONFIG_SND_KORG1212 $CONFIG_SND
+dep_tristate 'NeoMagic NM256AV/ZX' CONFIG_SND_NM256 $CONFIG_SND
+dep_tristate 'RME Digi32, 32/8, 32 PRO' CONFIG_SND_RME32 $CONFIG_SND
+dep_tristate 'RME Digi96, 96/8, 96/8 PRO' CONFIG_SND_RME96 $CONFIG_SND
+dep_tristate 'RME Digi9652 (Hammerfall)' CONFIG_SND_RME9652 $CONFIG_SND
+dep_tristate 'RME Hammerfall DSP Audio' CONFIG_SND_HDSP $CONFIG_SND
+dep_tristate 'Trident 4D-Wave DX/NX; SiS 7018' CONFIG_SND_TRIDENT $CONFIG_SND $CONFIG_SOUND_GAMEPORT
+dep_tristate 'Yamaha YMF724/740/744/754' CONFIG_SND_YMFPCI $CONFIG_SND
+dep_tristate 'Avance Logic ALS4000' CONFIG_SND_ALS4000 $CONFIG_SND
+dep_tristate 'C-Media 8738, 8338' CONFIG_SND_CMIPCI $CONFIG_SND
+dep_tristate '(Creative) Ensoniq AudioPCI 1370' CONFIG_SND_ENS1370 $CONFIG_SND
+dep_tristate '(Creative) Ensoniq AudioPCI 1371/1373' CONFIG_SND_ENS1371 $CONFIG_SND
+dep_tristate 'ESS ES1938/1946 (Solo-1)' CONFIG_SND_ES1938 $CONFIG_SND $CONFIG_SOUND_GAMEPORT
+dep_tristate 'ESS ES1968/1978 (Maestro-1/2/2E)' CONFIG_SND_ES1968 $CONFIG_SND
+dep_tristate 'ESS Allegro/Maestro3' CONFIG_SND_MAESTRO3 $CONFIG_SND
+dep_tristate 'ForteMedia FM801' CONFIG_SND_FM801 $CONFIG_SND
+dep_tristate 'ICEnsemble ICE1712 (Envy24)' CONFIG_SND_ICE1712 $CONFIG_SND
+dep_tristate 'ICE/VT1724 (Envy24HT)' CONFIG_SND_ICE1724 $CONFIG_SND
+dep_tristate 'Intel i8x0/MX440, SiS 7012; Ali 5455; NForce Audio; AMD768/8111' CONFIG_SND_INTEL8X0 $CONFIG_SND
+dep_tristate 'S3 SonicVibes' CONFIG_SND_SONICVIBES $CONFIG_SND $CONFIG_SOUND_GAMEPORT
+dep_tristate 'VIA 82C686A/B, 8233 South Bridge' CONFIG_SND_VIA82XX $CONFIG_SND
+dep_tristate 'Digigram VX222/Mic' CONFIG_SND_VX222 $CONFIG_SND
+dep_tristate 'Aztech AZF3328 / PCI168 (EXPERIMENTAL)' CONFIG_SND_AZT3328 $CONFIG_SND $CONFIG_EXPERIMENTAL
+
+endmenu
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/pci/Kconfig w2/sound/pci/Kconfig
--- w1/sound/pci/Kconfig	2003-06-02 14:59:40.000000000 +0200
+++ w2/sound/pci/Kconfig	1970-01-01 01:00:00.000000000 +0100
@@ -1,193 +0,0 @@
-# ALSA PCI drivers
-
-menu "PCI devices"
-	depends on SND!=n && PCI
-
-config SND_ALI5451
-	tristate "ALi PCI Audio M5451"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for ALI PCI Audio M5451 sound core.
-
-config SND_AZT3328
-	tristate "Aztech AZF3328 / PCI168 (EXPERIMENTAL)"
-	depends on SND && EXPERIMENTAL
-	help
-	  Say 'Y' or 'M' to include support for Aztech AZF3328 (PCI168) soundcards.
-
-config SND_CS46XX
-	tristate "Cirrus Logic (Sound Fusion) CS4280/CS461x/CS462x/CS463x"
-	depends on SND && SOUND_GAMEPORT
-	help
-	  Say 'Y' or 'M' to include support for Cirrus Logic CS4610 / CS4612 /
-	  CS4614 / CS4615 / CS4622 / CS4624 / CS4630 / CS4280 chips.
-
-config SND_CS46XX_NEW_DSP
-	bool "Cirrus Logic (Sound Fusion) New DSP support (EXPERIMENTAL)"
-	depends on SND_CS46XX && EXPERIMENTAL
-	help
-	  Say 'Y' to use a new DSP image for SPDIF and dual codecs.
-
-config SND_CS4281
-	tristate "Cirrus Logic (Sound Fusion) CS4281"
-	depends on SND && SOUND_GAMEPORT
-	help
-	  Say 'Y' or 'M' to include support for Cirrus Logic CS4281.
-
-config SND_EMU10K1
-	tristate "EMU10K1 (SB Live! & Audigy, E-mu APS)"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Sound Blaster PCI 512, Live!,
-	  Audigy and E-mu APS (partially supported).
-
-config SND_KORG1212
-	tristate "Korg 1212 IO"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Korg 1212IO.
-
-config SND_NM256
-	tristate "NeoMagic NM256AV/ZX"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for NeoMagic NM256AV/ZX chips.
-
-config SND_RME32
-	tristate "RME Digi32, 32/8, 32 PRO"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for RME Digi32, Digi32 PRO and
-	  Digi32/8 (Sek'd Prodif32, Prodif96 and Prodif Gold) audio devices.
-
-config SND_RME96
-	tristate "RME Digi96, 96/8, 96/8 PRO"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for RME Digi96, Digi96/8 and
-	  Digi96/8 PRO/PAD/PST.
-
-config SND_RME9652
-	tristate "RME Digi9652 (Hammerfall)"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for RME Hammerfall (RME Digi9652 /
-	  Digi9636) soundcards.
-
-config SND_HDSP
-	tristate "RME Hammerfall DSP Audio"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for RME Hammerfall DSP Audio
-	  soundcards.
-
-config SND_TRIDENT
-	tristate "Trident 4D-Wave DX/NX; SiS 7018"
-	depends on SND && SOUND_GAMEPORT
-	help
-	  Say 'Y' or 'M' to include support for Trident 4D-Wave DX/NX and
-	  SiS 7018 soundcards.
-
-config SND_YMFPCI
-	tristate "Yamaha YMF724/740/744/754"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Yamaha PCI audio chips - 
-	  YMF724, YMF724F, YMF740, YMF740C, YMF744, YMF754.
-
-config SND_ALS4000
-	tristate "Avance Logic ALS4000"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Avance Logic ALS4000.
-
-config SND_CMIPCI
-	tristate "C-Media 8738, 8338"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for C-Media CMI8338 and 8738 PCI
-	  soundcards.
-
-config SND_ENS1370
-	tristate "(Creative) Ensoniq AudioPCI 1370"
-	depends on SND && SOUND_GAMEPORT
-	help
-	  Say 'Y' or 'M' to include support for Ensoniq AudioPCI ES1370.
-
-config SND_ENS1371
-	tristate "(Creative) Ensoniq AudioPCI 1371/1373"
-	depends on SND && SOUND_GAMEPORT
-	help
-	  Say 'Y' or 'M' to include support for Ensoniq AudioPCI ES1371 and
-	  Sound Blaster PCI 64 or 128 soundcards.
-
-config SND_ES1938
-	tristate "ESS ES1938/1946/1969 (Solo-1)"
-	depends on SND && SOUND_GAMEPORT
-	help
-	  Say 'Y' or 'M' to include support for ESS Solo-1 (ES1938, ES1946, ES1969)
-	  soundcard.
-
-config SND_ES1968
-	tristate "ESS ES1968/1978 (Maestro-1/2/2E)"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for ESS Maestro 1/2/2E.
-
-config SND_MAESTRO3
-	tristate "ESS Allegro/Maestro3"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for ESS Maestro 3 (Allegro) soundcard.
-
-config SND_FM801
-	tristate "ForteMedia FM801"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for ForteMedia FM801 based soundcards.
-
-config SND_ICE1712
-	tristate "ICEnsemble ICE1712 (Envy24)"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for ICE1712 (Envy24) based soundcards.
-	  Currently supported hardware is: MidiMan M Audio - Delta 1010(LT), Dio 2496,
-	  Delta 66/44, Audiophile 24/96; Hoontech SoundTrack DSP 24 (Value);
-	  TerraTec - EWX 24/96, EWS 88MT, EWS 88D, DMX 6Fire.
-
-config SND_ICE1724
-	tristate "ICE/VT1724 (Envy24HT)"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for ICE/VT1724 (Envy24HT) based
-	  soundcards.
-	  Currently supported hardware is: MidiMan M Audio - Revolution 7.1,
-	  AMP Ltd AUDIO2000.
-
-config SND_INTEL8X0
-	tristate "Intel i8x0/MX440, SiS 7012; Ali 5455; NForce Audio; AMD768/8111"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Intel8x0 based soundcards,
-	  SiS 7012, AMD768/8111, NVidia NForce and ALi 5455 chips.
-
-config SND_SONICVIBES
-	tristate "S3 SonicVibes"
-	depends on SND && SOUND_GAMEPORT
-	help
-	  Say 'Y' or 'M' to include support for S3 SonicVibes based soundcards.
-
-config SND_VIA82XX
-	tristate "VIA 82C686A/B, 8233 South Bridge"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for VIA VT82C686A/B, VT8233 South Bridge.
-
-config SND_VX222
-	tristate "Digigram VX222"
-	depends on SND
-	help
-	  Say 'Y' or 'M' to include support for Digigram VX222 soundcards.
-
-endmenu
-
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/pcmcia/Config.in w2/sound/pcmcia/Config.in
--- w1/sound/pcmcia/Config.in	1970-01-01 01:00:00.000000000 +0100
+++ w2/sound/pcmcia/Config.in	2003-08-13 23:51:56.000000000 +0200
@@ -0,0 +1,9 @@
+# ALSA PCMCIA drivers
+
+mainmenu_option next_comment
+comment 'ALSA PCMCIA devices'
+
+dep_tristate 'Digigram VX-pocket driver' CONFIG_SND_VXPOCKET $CONFIG_SND $CONFIG_PCMCIA
+dep_tristate 'Digigram VX-pocket 440 driver' CONFIG_SND_VXP440 $CONFIG_SND $CONFIG_PCMCIA
+
+endmenu
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/pcmcia/Kconfig w2/sound/pcmcia/Kconfig
--- w1/sound/pcmcia/Kconfig	2003-05-23 18:00:45.000000000 +0200
+++ w2/sound/pcmcia/Kconfig	1970-01-01 01:00:00.000000000 +0100
@@ -1,18 +0,0 @@
-# ALSA PCMCIA drivers
-
-menu "PCMCIA devices"
-	depends on SND!=n && PCMCIA
-
-config SND_VXPOCKET
-	tristate "Digigram VXpocket"
-	depends on SND && PCMCIA
-	help
-	  Say 'Y' or 'M' to include support for Digigram VXpocket soundcard.
-
-config SND_VXP440
-	tristate "Digigram VXpocket 440"
-	depends on SND && PCMCIA
-	help
-	  Say 'Y' or 'M' to include support for Digigram VXpocket 440 soundcard.
-
-endmenu
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/ppc/Config.in w2/sound/ppc/Config.in
--- w1/sound/ppc/Config.in	1970-01-01 01:00:00.000000000 +0100
+++ w2/sound/ppc/Config.in	2003-08-13 23:48:50.000000000 +0200
@@ -0,0 +1,8 @@
+# ALSA PowerMac drivers
+
+mainmenu_option next_comment
+comment 'ALSA PowerMac devices'
+
+dep_tristate 'PowerMac (AWACS, DACA, Burgundy, Tumbler, Keywest)' CONFIG_SND_POWERMAC $CONFIG_SND
+
+endmenu
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/ppc/Kconfig w2/sound/ppc/Kconfig
--- w1/sound/ppc/Kconfig	2002-10-31 14:44:21.000000000 +0100
+++ w2/sound/ppc/Kconfig	1970-01-01 01:00:00.000000000 +0100
@@ -1,11 +0,0 @@
-# ALSA PowerMac drivers
-
-menu "ALSA PowerMac devices"
-	depends on SND!=n && PPC
-
-config SND_POWERMAC
-	tristate "PowerMac (AWACS, DACA, Burgundy, Tumbler, Keywest)"
-	depends on SND
-
-endmenu
-
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/sparc/Config.in w2/sound/sparc/Config.in
--- w1/sound/sparc/Config.in	1970-01-01 01:00:00.000000000 +0100
+++ w2/sound/sparc/Config.in	2003-08-13 23:48:50.000000000 +0200
@@ -0,0 +1,13 @@
+# ALSA Sparc drivers
+
+mainmenu_option next_comment
+comment 'ALSA Sparc devices'
+
+if [ "$CONFIG_SBUS" != "n" ]; then
+  dep_tristate 'Sun AMD7930' CONFIG_SND_SUN_AMD7930 $CONFIG_SND
+#  dep_tristate 'Sun DBRI' CONFIG_SND_SUN_DBRI $CONFIG_SND
+fi
+
+dep_tristate 'Sun CS4231' CONFIG_SND_SUN_CS4231 $CONFIG_SND
+
+endmenu
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/sparc/Kconfig w2/sound/sparc/Kconfig
--- w1/sound/sparc/Kconfig	2002-10-31 14:44:21.000000000 +0100
+++ w2/sound/sparc/Kconfig	1970-01-01 01:00:00.000000000 +0100
@@ -1,16 +0,0 @@
-# ALSA Sparc drivers
-
-menu "ALSA Sparc devices"
-	depends on SND!=n && (SPARC32 || SPARC64)
-
-config SND_SUN_AMD7930
-	tristate "Sun AMD7930"
-	depends on SBUS && SND
-
-#  dep_tristate 'Sun DBRI' CONFIG_SND_SUN_DBRI $CONFIG_SND
-config SND_SUN_CS4231
-	tristate "Sun CS4231"
-	depends on SND
-
-endmenu
-
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/usb/Config.help w2/sound/usb/Config.help
--- w1/sound/usb/Config.help	1970-01-01 01:00:00.000000000 +0100
+++ w2/sound/usb/Config.help	2003-08-13 23:48:50.000000000 +0200
@@ -0,0 +1,2 @@
+CONFIG_SND_USB_AUDIO
+  Say 'Y' or 'M' to include support for USB audio and USB MIDI devices.
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/usb/Config.in w2/sound/usb/Config.in
--- w1/sound/usb/Config.in	1970-01-01 01:00:00.000000000 +0100
+++ w2/sound/usb/Config.in	2003-08-13 23:48:50.000000000 +0200
@@ -0,0 +1,8 @@
+# ALSA USB drivers
+
+mainmenu_option next_comment
+comment 'ALSA USB devices'
+
+dep_tristate 'USB Audio/MIDI driver' CONFIG_SND_USB_AUDIO $CONFIG_SND
+
+endmenu
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w1/sound/usb/Kconfig w2/sound/usb/Kconfig
--- w1/sound/usb/Kconfig	2002-12-10 16:22:07.000000000 +0100
+++ w2/sound/usb/Kconfig	1970-01-01 01:00:00.000000000 +0100
@@ -1,13 +0,0 @@
-# ALSA USB drivers
-
-menu "ALSA USB devices"
-	depends on SND!=n && USB!=n
-
-config SND_USB_AUDIO
-	tristate "USB Audio/MIDI driver"
-	depends on SND && USB
-	help
-	  Say 'Y' or 'M' to include support for USB audio and USB MIDI devices.
-
-endmenu
-
