--- linux-2.4.19/include/sound/adriver.h.orig	2002-11-15 11:53:15.000000000 +0100
+++ linux-2.4.19/include/sound/adriver.h	2002-11-15 11:54:32.000000000 +0100
@@ -223,18 +223,18 @@
 #include <linux/usb.h>
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 5, 0)
-inline static struct urb *usb_alloc_urb_wrapper(int iso_packets, int flags)
+inline static struct urb *usb_alloc_urb_wrapper(int iso_packets)
 {
 	return usb_alloc_urb(iso_packets);
 }
-inline static int usb_submit_urb_wrapper(struct urb *urb, int flags)
+inline static int usb_submit_urb_wrapper(struct urb *urb)
 {
 	return usb_submit_urb(urb);
 }
 #undef usb_alloc_urb
 #undef usb_submit_urb
-#define usb_alloc_urb(n,flags) usb_alloc_urb_wrapper(n,flags)
-#define usb_submit_urb(p,flags) usb_submit_urb_wrapper(p,flags)
+#define usb_alloc_urb(n) usb_alloc_urb_wrapper(n)
+#define usb_submit_urb(p) usb_submit_urb_wrapper(p)
 #define OLD_USB
 #endif /* LINUX_VERSION_CODE < 2.5.0 */
 
