--- linux-2.4.21/sound/core/pnp/pnp.c.pnp-buildfixes	2003-08-10 13:09:08.000000000 +0200
+++ linux-2.4.21/sound/core/pnp/pnp.c	2003-08-10 15:23:59.000000000 +0200
@@ -137,6 +137,7 @@ static int parse_id(const char * id, uns
 	return 0;
 }
 
+#ifdef CONFIG_PNP
 struct pnp_dev * pnp_request_card_device(struct pnp_card_link *clink, const char * id, struct pnp_dev * from)
 {
 	unsigned short vendor, function;
@@ -368,6 +369,7 @@ int pnp_activate_dev(struct pnp_dev *dev
 
 	return dev->p.activate((struct isapnp_dev *)dev);
 }
+#endif
 
 static int __init pnp_init(void)
 {
