diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/sound/isa/msnd/msnd.c w1/sound/isa/msnd/msnd.c
--- w2/sound/isa/msnd/msnd.c	2003-08-20 19:14:09.000000000 +0200
+++ w1/sound/isa/msnd/msnd.c	2003-08-18 20:30:28.000000000 +0200
@@ -304,4 +304,4 @@ void cleanup_module(void)
 {
 }
 #endif
-*/
\ No newline at end of file
+*/
