diff -Nurp -X /Data/dontdiff -X /Data/dontdiff2 patched/sound/isa/opl3sa2.c patched/sound.cvs/isa/opl3sa2.c
--- patched/sound/isa/opl3sa2.c	2003-07-08 13:41:34.000000000 +0300
+++ patched/sound.cvs/isa/opl3sa2.c	2003-08-13 16:01:22.000000000 +0300
@@ -752,6 +752,7 @@ static int __devinit snd_opl3sa2_probe(i
 		err = -ENOMEM;
 		goto __error;
 	}
+	spin_lock_init(&chip->reg_lock);
 	chip->irq = -1;
 	if ((err = snd_device_new(card, SNDRV_DEV_LOWLEVEL, chip, &ops)) < 0)
 		goto __error;
