diff -uNp l1/net/sctp/socket.c.orig l1/net/sctp/socket.c
--- l1/net/sctp/socket.c.orig	2004-05-17 13:15:37.398074272 +0200
+++ l1/net/sctp/socket.c	2004-05-17 13:15:55.569311824 +0200
@@ -1516,18 +1516,7 @@ SCTP_STATIC int sctp_setsockopt(struct s
 
 	switch (optname) {
 	case SCTP_SOCKOPT_DEBUG_NAME:
-		/* BUG! we don't ever seem to free this memory. --jgrimm */
-		if (NULL == (tmp = kmalloc(optlen + 1, GFP_KERNEL))) {
-			retval = -ENOMEM;
-			goto out_unlock;
-		}
-
-		if (copy_from_user(tmp, optval, optlen)) {
-			retval = -EFAULT;
-			goto out_unlock;
-		}
-		tmp[optlen] = '\000';
-		sctp_sk(sk)->ep->debug_name = tmp;
+		/* do nothing */
 		break;
 
 	case SCTP_SOCKOPT_BINDX_ADD:
