--- RPC/Server.php.orig	2005-08-19 18:09:41.000000000 -0400
+++ RPC/Server.php	2005-08-19 18:19:19.000000000 -0400
@@ -227,11 +227,12 @@
         $parser = xml_parser_create($XML_RPC_defencoding);
 
         $XML_RPC_xh[$parser] = array();
-        $XML_RPC_xh[$parser]['st'] = "";
-        $XML_RPC_xh[$parser]['cm'] = 0;
         $XML_RPC_xh[$parser]['isf'] = 0;
+        $XML_RPC_xh[$parser]['isf_reason'] = 0;
         $XML_RPC_xh[$parser]['params'] = array();
         $XML_RPC_xh[$parser]['method'] = "";
+        $XML_RPC_xh[$parser]['stack'] = array();
+        $XML_RPC_xh[$parser]['valuestack'] = array();
 
         $plist = '';
 
@@ -249,14 +250,19 @@
                                               xml_error_string(xml_get_error_code($parser)),
                                               xml_get_current_line_number($parser)));
             xml_parser_free($parser);
+        } else if ($XML_RPC_xh[$parser]['isf']>1) {
+            $r = new XML_RPC_Response(0, $XML_RPC_err['invalid_request'],
+                                         $XML_RPC_str['invalid_request']
+                                            . ': ' . $XML_RPC_xh[$parser]['isf_reason']);
+            xml_parser_free($parser);
         } else {
             xml_parser_free($parser);
             $m = new XML_RPC_Message($XML_RPC_xh[$parser]['method']);
             // now add parameters in
             for ($i = 0; $i < sizeof($XML_RPC_xh[$parser]['params']); $i++) {
                 // print "<!-- " . $XML_RPC_xh[$parser]['params'][$i]. "-->\n";
-                $plist .= "$i - " . $XML_RPC_xh[$parser]['params'][$i] . " \n";
-                eval('$m->addParam(' . $XML_RPC_xh[$parser]['params'][$i] . ");");
+                $plist .= "$i - " . var_export($XML_RPC_xh[$parser]['params'][$i], true) . " \n";
+                @$m->addParam($XML_RPC_xh[$parser]['params'][$i]);
             }
             XML_RPC_Server_debugmsg($plist);
             // now to deal with the method
