--- libpng-1.2.6/scripts/libpng.pc.in.lib64	2004-08-15 08:15:22.000000000 -0400
+++ libpng-1.2.6/scripts/libpng.pc.in	2004-09-30 08:01:39.050599596 -0400
@@ -1,7 +1,7 @@
 
 prefix=@PREFIX@
 exec_prefix=${prefix}
-libdir=${exec_prefix}/lib
+libdir=@LIBDIR@
 includedir=${exec_prefix}/include
 
 Name: libpng12
--- libpng-1.2.6/scripts/makefile.linux.lib64	2004-09-30 07:59:16.232790000 -0400
+++ libpng-1.2.6/scripts/makefile.linux	2004-09-30 08:06:16.362539378 -0400
@@ -71,7 +71,7 @@ libpng.a: $(OBJS)
 	$(RANLIB) $@
 
 libpng.pc:
-	cat scripts/libpng.pc.in | sed -e s\!@PREFIX@!$(prefix)! > libpng.pc
+	cat scripts/libpng.pc.in | sed -e "s,@PREFIX@,$(prefix),;s,@LIBDIR@,$(LIBPATH)," > libpng.pc
 
 libpng-config:
 	( cat scripts/libpng-config-head.in; \
