--- samba-3.0.5/source/aclocal.m4.lib64	2004-07-20 18:28:16.000000000 +0200
+++ samba-3.0.5/source/aclocal.m4	2004-08-13 00:16:37.633039232 +0200
@@ -546,7 +546,7 @@
   ac_new_flags=""
   for i in [$]$1; do
     case [$]i in
-    -L/usr/lib|-L/usr/lib/) ;;
+    -L/usr/lib|-L/usr/lib/|-L/usr/lib64|-L/usr/lib64/) ;;
     *) ac_new_flags="[$]ac_new_flags [$]i" ;;
     esac
   done
@@ -679,7 +679,7 @@
 dnl LIB_ADD_DIR(LDFLAGS, $LIBDIR)
 dnl This function doesn't add -L/usr/lib into LDFLAGS
 AC_DEFUN(LIB_ADD_DIR,[
-if test "$2" != "/usr/lib" ; then
+if test "$2" != "/usr/lib" -a "$2" != "/usr/lib64"; then
     $1="$$1 -L$2"
 fi
 ])
--- samba-3.0.5/source/configure.in.lib64	2004-07-20 18:28:13.000000000 +0200
+++ samba-3.0.5/source/configure.in	2004-08-13 00:25:56.634372001 +0200
@@ -345,6 +345,33 @@
 
 AC_CANONICAL_SYSTEM
 
+dnl Determine LIB directory name
+case $host in
+*-*-linux*)
+  # Test if the compiler is 64bit
+  echo 'int i;' > conftest.$ac_ext
+  samba_cv_cc_64bit_output=no
+  if AC_TRY_EVAL(ac_compile); then
+    case `/usr/bin/file conftest.$ac_objext` in
+    *"ELF 64"*)
+     samba_cv_cc_64bit_output=yes
+     ;;
+    esac
+  fi
+  rm -rf conftest*
+  ;;
+esac
+
+case $host_cpu:$samba_cv_cc_64bit_output in
+powerpc64:yes | s390x:yes | sparc64:yes | x86_64:yes)
+  LIB="lib64"
+  ;;
+*:*)
+  LIB="lib"
+  ;;
+esac
+AC_SUBST(LIB)
+
 dnl Add #include for broken IRIX header files
   case "$host_os" in
 	*irix6*) AC_ADD_INCLUDE(<standards.h>)
@@ -868,7 +895,7 @@
     _cppflags=${CPPFLAGS}
 
     # Add additional search path
-    LDFLAGS="-L$with_readline/lib $LDFLAGS"
+    LDFLAGS="-L$with_readline/$LIB $LDFLAGS"
     CPPFLAGS="-I$with_readline/include $CPPFLAGS"
 
     AC_CHECK_HEADERS(readline.h history.h readline/readline.h)
@@ -879,7 +906,7 @@
        AC_CHECK_LIB(${termlib}, tgetent, [TERMLIBS="-l${termlib}"; break])
       done
       AC_CHECK_LIB(readline, rl_callback_handler_install,
-       [TERMLDFLAGS="-L$with_readline/lib"
+       [TERMLDFLAGS="-L$with_readline/$LIB"
        TERMCPPFLAGS="-I$with_readline/include"
        CPPFLAGS="-I$with_readline/include $CPPFLAGS"
        TERMLIBS="-lreadline $TERMLIBS"
@@ -1682,7 +1709,7 @@
 
 dnl Check lib and lib32 library variants to cater for IRIX ABI-specific
 dnl installation paths.
-    for l in "lib" "lib32" ; do
+    for l in "$LIB" "lib32" ; do
         LDFLAGS="$LDFLAGS -L$i/$l"
         LIBS=
         export LDFLAGS LIBS CPPFLAGS
@@ -1708,7 +1735,7 @@
         # check for iconv in libc
         ic_save_LIBS="$LIBS"
         if test x"$ICONV_PATH_SPEC" = "xyes" ; then
-           LIBS="$LIBS -L$ICONV_LOCATION/lib"
+           LIBS="$LIBS -L$ICONV_LOCATION/$LIB"
         fi
         if test x"$jm_cv_lib_iconv" != x; then
            LIBS="$LIBS -l$jm_cv_lib_iconv"
@@ -2568,7 +2595,7 @@
         AC_MSG_RESULT($withval)
         KRB5_CFLAGS="-I$withval/include"
         KRB5_CPPFLAGS="-I$withval/include"
-        KRB5_LDFLAGS="-L$withval/lib"
+        KRB5_LDFLAGS="-L$withval/$LIB"
         FOUND_KRB5=yes
         ;;
       esac ],
--- samba-3.0.5/source/Makefile.in.lib64	2004-07-20 18:27:57.000000000 +0200
+++ samba-3.0.5/source/Makefile.in	2004-08-13 00:16:37.639039375 +0200
@@ -58,6 +58,7 @@
 BINDIR = @bindir@
 # sbindir is mapped to bindir when compiling SAMBA in 2.0.x compatibility mode.
 SBINDIR = @sbindir@
+LIB = @LIB@
 LIBDIR = @libdir@
 VFSLIBDIR = $(LIBDIR)/vfs
 PDBLIBDIR = $(LIBDIR)/pdb
@@ -1278,9 +1279,9 @@
 	@$(SHELL) $(srcdir)/script/installswat.sh $(DESTDIR)$(SWATDIR) $(srcdir)
 
 installclientlib: installdirs libsmbclient
-	@$(SHELL) $(srcdir)/script/installdirs.sh $(DESTDIR)${prefix}/lib
-	-$(INSTALLCLIENTCMD_SH) bin/libsmbclient.@SHLIBEXT@ $(DESTDIR)${prefix}/lib
-	-$(INSTALLCLIENTCMD_A) bin/libsmbclient.a $(DESTDIR)${prefix}/lib
+	@$(SHELL) $(srcdir)/script/installdirs.sh $(DESTDIR)${prefix}/$(LIB)
+	-$(INSTALLCLIENTCMD_SH) bin/libsmbclient.@SHLIBEXT@ $(DESTDIR)${prefix}/$(LIB)
+	-$(INSTALLCLIENTCMD_A) bin/libsmbclient.a $(DESTDIR)${prefix}/$(LIB)
 	@$(SHELL) $(srcdir)/script/installdirs.sh $(DESTDIR)${prefix}/include
 	-$(INSTALLCMD) $(srcdir)/include/libsmbclient.h $(DESTDIR)${prefix}/include
 
