--- Linux-PAM-0.77/modules/pam_console/config.y.console-reset	2003-07-10 22:50:26.000000000 +0200
+++ Linux-PAM-0.77/modules/pam_console/config.y	2003-07-10 22:54:22.000000000 +0200
@@ -394,7 +394,7 @@
     }
     g_free(dirname);
     return ret;
-
+}
 
 
 STATIC int
@@ -415,13 +415,13 @@
 	    if (pwd == NULL) {
 		_pam_log(LOG_ERR, FALSE, "getpwnam failed for %s",
 			 c->revert_owner ? c->revert_owner : "root");
-		return -1;
+		pwd->pw_uid = 0;
 	    }
 	    grp = _pammodutil_getgrnam(pamh, c->revert_group ? c->revert_group : "root");
 	    if (grp == NULL) {
                 _pam_log(LOG_ERR, FALSE, "getgrnam failed for %s",
                          c->revert_group ? c->revert_group : "root");
-                return -1;
+		grp->gr_gid = 0;
             }
 	    if (c->device_class->list)
 		chmod_files(c->revert_mode ? c->revert_mode : "0600",
--- Linux-PAM-0.77/modules/pam_console/pam_console_apply.c.console-reset	2003-07-08 05:16:28.000000000 +0200
+++ Linux-PAM-0.77/modules/pam_console/pam_console_apply.c	2003-07-10 22:50:26.000000000 +0200
@@ -41,8 +41,9 @@
 	va_list args;
 	if (debug_p && !debug) return;
         va_start(args, format);
-	fprintf(stderr, format, args);
+	vfprintf(stderr, format, args);
 	va_end(args);
+	fprintf(stderr, "\n");
 }
 
 int
