--- XFree86-4.1.0/xc/programs/glxinfo/Imakefile.dl	Mon Apr  2 16:58:45 2001
+++ XFree86-4.1.0/xc/programs/glxinfo/Imakefile	Wed Sep 19 15:00:26 2001
@@ -2,7 +2,7 @@
 
 #include <Threads.tmpl>
         DEPLIBS = $(DEPGLULIB) $(DEPGLXLIB) $(DEPXLIB)
-LOCAL_LIBRARIES = $(GLULIB) $(GLXLIB) $(XLIB) 
+LOCAL_LIBRARIES = $(GLULIB) $(GLXLIB) $(XLIB) DlLibrary
   SYS_LIBRARIES = MathLibrary
 
 SimpleProgramTarget(glxinfo)
