--- ppp-2.4.1/pppd/Makefile.linux	Sun Apr  8 08:35:04 2001
+++ ppp-2.4.1/pppd/Makefile.geoff	Sun Apr  8 08:35:29 2001
@@ -4,8 +4,8 @@
 #
 
 # Default installation locations
-BINDIR = /usr/sbin
-MANDIR = /usr/man
+BINDIR = $(DESTDIR)/usr/sbin
+MANDIR = $(DESTDIR)/usr/man
 
 PPPDSRCS = main.c magic.c fsm.c lcp.c ipcp.c upap.c chap.c md5.c ccp.c \
 	   ipxcp.c auth.c options.c sys-linux.c md4.c chap_ms.c cbcp.c \
@@ -25,10 +25,10 @@
 include .depend
 endif
 
-# CC = gcc
+CC = gcc
 #
 COPTS = -O2 -pipe -Wall -g
-LIBS =
+LIBS = -lutil
 
 ifneq ($(wildcard /usr/lib/libcrypt.*),)
 LIBS += -lcrypt
@@ -49,7 +49,7 @@
 #FILTER=y
 
 HAS_SHADOW=y
-#USE_PAM=y
+USE_PAM=y
 #HAVE_INET6=y
 
 PLUGIN=y
@@ -112,13 +112,11 @@
 endif
 
 
-INSTALL= install -o root
+INSTALL= install
 
 install: pppd
 	mkdir -p $(BINDIR) $(MANDIR)
-	$(INSTALL) -s -c -m 555 pppd $(BINDIR)/pppd
-	if chgrp pppusers $(BINDIR)/pppd 2>/dev/null; then \
-	  chmod o-rx,u+s $(BINDIR)/pppd; fi
+	$(INSTALL)  -m 555 pppd $(BINDIR)/pppd
 	$(INSTALL) -c -m 444 pppd.8 $(MANDIR)/man8
 
 pppd: $(PPPDOBJS)
