--- expect5.32/exp_command.c.alpha	Mon Aug 28 23:58:00 2000
+++ expect5.32/exp_command.c	Thu Jul 12 13:52:00 2001
@@ -794,14 +794,18 @@
 	    return TCL_ERROR;
 	}
 	if (mode & TCL_READABLE) {
-	    if (TCL_ERROR == Tcl_GetChannelHandle(channel, TCL_READABLE, (ClientData) &rfd)) {
+	    ClientData rfd_data;
+	    if (TCL_ERROR == Tcl_GetChannelHandle(channel, TCL_READABLE, &rfd_data)) {
 		return TCL_ERROR;
 	    }
+	    rfd = (int)rfd_data;
 	}
 	if (mode & TCL_WRITABLE) {
-	    if (TCL_ERROR == Tcl_GetChannelHandle(channel, TCL_WRITABLE, (ClientData) &wfd)) {
+	    ClientData wfd_data;
+	    if (TCL_ERROR == Tcl_GetChannelHandle(channel, TCL_WRITABLE, &wfd_data)) {
 		return TCL_ERROR;
 	    }    
+	    wfd = (int)wfd_data;
 	}
 	master = ((mode & TCL_READABLE)?rfd:wfd);
 
@@ -1215,7 +1219,7 @@
     Tcl_DStringFree(&dstring);
     if (esPtr) {
         exp_close(interp,esPtr);
-	waitpid(esPtr->pid,&esPtr->wait,0);
+	waitpid(esPtr->pid,(int *) &esPtr->wait,0);
 	if (esPtr->registered) {
 	    Tcl_UnregisterChannel(interp,esPtr->channel);
 	}
@@ -2555,7 +2559,7 @@
 			if (rc != TCL_OK) return(rc);
 		    }
 
-		    result = waitpid(esPtr->pid,&esPtr->wait,0);
+		    result = waitpid(esPtr->pid,(int *) &esPtr->wait,0);
 		    if (result == esPtr->pid) break;
 		    if (result == -1) {
 			if (errno == EINTR) continue;
@@ -2589,7 +2593,7 @@
 	    for (fp=forked_proc_base;fp;fp=fp->next) {
 		if (fp->link_status == not_in_use) continue;
 	restart:
-		result = waitpid(fp->pid,&fp->wait_status,WNOHANG);
+		result = waitpid(fp->pid,(int *) &fp->wait_status,WNOHANG);
 		if (result == fp->pid) {
 		    waited_on_forked_process = 1;
 		    break;
