--- cups-1.1.16pre1/scheduler/cups.pam.chmou Mon Mar  6 21:33:43 2000
+++ cups-1.1.16pre1/scheduler/cups.pam       Tue Sep 26 11:49:54 2000
@@ -1,2 +1,2 @@
-auth	required	/lib/security/pam_pwdb.so nullok shadow
-account	required	/lib/security/pam_pwdb.so
+auth	required	/lib/security/pam_stack.so service=system-auth
+account	required	/lib/security/pam_stack.so service=system-auth
--- cups-1.1.16pre1/conf/pam.conf.in.orig	2002-09-10 01:43:08.000000000 +0200
+++ cups-1.1.16pre1/conf/pam.conf.in	2002-09-10 01:44:15.000000000 +0200
@@ -1,2 +1,2 @@
-auth	required	@PAMMOD@ nullok shadow
-account	required	@PAMMOD@
+auth	required	/lib/security/pam_stack.so service=system-auth
+account	required	/lib/security/pam_stack.so service=system-auth
