diff -Naur php-4.3.4RC3/ext/standard/url_scanner.c php-4.3.4RC3.oden/ext/standard/url_scanner.c
--- php-4.3.4RC3/ext/standard/url_scanner.c	2003-03-11 00:09:38.000000000 +0100
+++ php-4.3.4RC3.oden/ext/standard/url_scanner.c	2003-10-29 20:45:29.000000000 +0100
@@ -88,7 +88,7 @@
 
 /* {{{ url_adapt_ext
  */
-char *url_adapt_ext(const char *src, uint srclen, const char *name, const char *val, size_t *newlen)
+char *url_adapt_ext(const char *src, uint srclen, const char *name, const char *val, uint *newlen)
 {
 	char buf[1024];
 
@@ -100,7 +100,7 @@
 
 /* {{{ url_adapt
  */
-char *url_adapt(const char *src, size_t srclen, const char *data, size_t *newlen)
+char *url_adapt(const char *src, uint srclen, const char *data, uint *newlen)
 {
 	char *out,*outp;
 	int maxl,n;
diff -Naur php-4.3.4RC3/ext/standard/url_scanner_ex.c php-4.3.4RC3.oden/ext/standard/url_scanner_ex.c
--- php-4.3.4RC3/ext/standard/url_scanner_ex.c	2003-10-29 14:28:07.000000000 +0100
+++ php-4.3.4RC3.oden/ext/standard/url_scanner_ex.c	2003-10-29 20:45:29.000000000 +0100
@@ -834,7 +834,7 @@
 }
 
 
-static char *url_adapt_ext(const char *src, size_t srclen, size_t *newlen, zend_bool do_flush TSRMLS_DC)
+static char *url_adapt_ext(const char *src, uint srclen, uint *newlen, zend_bool do_flush TSRMLS_DC)
 {
 	url_adapt_state_ex_t *ctx;
 	char *retval;
diff -Naur php-4.3.4RC3/ext/standard/url_scanner.h php-4.3.4RC3.oden/ext/standard/url_scanner.h
--- php-4.3.4RC3/ext/standard/url_scanner.h	2002-12-31 17:35:35.000000000 +0100
+++ php-4.3.4RC3.oden/ext/standard/url_scanner.h	2003-10-29 20:45:29.000000000 +0100
@@ -23,7 +23,7 @@
 int php_url_scanner_activate(TSRMLS_D);
 int php_url_scanner_deactivate(TSRMLS_D);
 
-char *url_adapt(const char *src, size_t srclen, const char *data, size_t *newlen);
+char *url_adapt(const char *src, uint srclen, const char *data, uint *newlen);
 
 enum url_state { 
 	STATE_NORMAL, 
