--- tcl8.3.3/unix/tclUnixFCmd.c.cruft	Tue Apr  4 04:05:57 2000
+++ tcl8.3.3/unix/tclUnixFCmd.c	Fri Jul  6 10:59:52 2001
@@ -1090,7 +1090,7 @@
 
     groupPtr = getgrgid(statBuf.st_gid);		/* INTL: Native. */
     if (groupPtr == NULL) {
-	*attributePtrPtr = Tcl_NewIntObj(statBuf.st_gid);
+	*attributePtrPtr = Tcl_NewIntObj((int)statBuf.st_gid);
     } else {
 	Tcl_DString ds;
 	CONST char *utf;
@@ -1141,7 +1141,7 @@
 
     pwPtr = getpwuid(statBuf.st_uid);			/* INTL: Native. */
     if (pwPtr == NULL) {
-	*attributePtrPtr = Tcl_NewIntObj(statBuf.st_uid);
+	*attributePtrPtr = Tcl_NewIntObj((int)statBuf.st_uid);
     } else {
 	Tcl_DString ds;
 	CONST char *utf;
@@ -1190,7 +1190,7 @@
 	return TCL_ERROR;
     }
 
-    sprintf(returnString, "%0#5lo", (statBuf.st_mode & 0x00007FFF));
+    sprintf(returnString, "%0#5lo", ((unsigned long)(statBuf.st_mode) & 0x7FFFUL));
 
     *attributePtrPtr = Tcl_NewStringObj(returnString, -1);
     
