--- tcl8.3.3/doc/registry.n.heiierarchy	Sat Mar 18 13:55:00 2000
+++ tcl8.3.3/doc/registry.n	Sat Mar 18 13:55:20 2000
@@ -55,7 +55,7 @@
 If the optional \fIvalueName\fR argument is present, the specified
 value under \fIkeyName\fR will be deleted from the registry.  If the
 optional \fIvalueName\fR is omitted, the specified key and any subkeys
-or values beneath it in the registry heirarchy will be deleted.  If
+or values beneath it in the registry hierarchy will be deleted.  If
 the key could not be deleted then an error is generated.  If the key
 did not exist, the command has no effect.
 .TP
