--- pam-0.75/modules/pam_console/pam_console.c.fred	Fri Jun 29 22:11:46 2001
+++ pam-0.75/modules/pam_console/pam_console.c	Mon Sep 24 21:13:36 2001
@@ -407,7 +407,7 @@
     if (!configfileparsed) { parse_file(consoleperms); configfileparsed = 1; }
 
     /* return success quietly if not a terminal login */
-    if (!check_console_name(tty, allow_nonroot_tty)) return PAM_SUCCESS;
+    if (!check_console_name(tty, allow_nonroot_tty) && tty[0] != ':') return PAM_SUCCESS;
 
     lockfile = _do_malloc(strlen(consolerefs) + strlen(username) + 2);
     sprintf(lockfile, "%s%s", consolerefs, username); /* trusted data */
