--- rpm-4.2/platform.in.opensls	2004-02-03 20:11:57.000000000 -0700
+++ rpm-4.2/platform.in	2004-02-03 20:12:48.000000000 -0700
@@ -217,7 +217,7 @@
 @mandrake@%{nil}
 @mandrake@
 @mandrake@%_add_user_helper /usr/share/rpm-helper/add-user
-@mandrake@%_pre_useradd() %_add_user_helper %{name} $1 %{1} %{2} %{3} \
+@mandrake@%_pre_useradd() %_add_user_helper %{name} $1 %{1} %{2} %{3} %{4} \
 @mandrake@%{nil}
 @mandrake@
 @mandrake@%_del_user_helper /usr/share/rpm-helper/del-user
@@ -225,7 +225,7 @@
 @mandrake@%{nil}
 @mandrake@
 @mandrake@%_add_group_helper /usr/share/rpm-helper/add-group
-@mandrake@%_pre_groupadd() %_add_group_helper %{name} $1 %{1} %{?2:%2} \
+@mandrake@%_pre_groupadd() %_add_group_helper %{name} $1 %{1} %{2} %{?3:%3} \
 @mandrake@%{nil}
 @mandrake@
 @mandrake@%_del_group_helper /usr/share/rpm-helper/del-group
@@ -239,3 +239,32 @@
 @mandrake@%_del_shell_helper /usr/share/rpm-helper/del-shell
 @mandrake@%_preun_shelldel() %_del_shell_helper %{name} $1 \
 @mandrake@%{nil}
+
+@mandrake@#---------------------------------------------------------------------
+@mandrake@#	OpenSLS vendor specific macros
+@mandrake@# 	Vincent Danen <vdanen@opensls.org>
+@mandrake@#
+@mandrake@#
+@mandrake@%_srvdir		%{_var}/service
+@mandrake@%_srvlogdir		%{_var}/log/supervise
+@mandrake@%_vendor		OpenSLS
+@mandrake@%_real_vendor		opensls
+@mandrake@%_target_platform	%{_target_cpu}-%{_real_vendor}-%{_target_os}%{?_gnu}
+@mandrake@
+@mandrake@# RPM extension
+@mandrake@%_ext			sls
+@mandrake@
+@mandrake@# Default to building with propolice
+@mandrake@%build_propolice %([ -n "$STACK_PROTECTOR" ] && echo 1 || echo 0)
+@mandrake@
+@mandrake@%_add_srv_helper /usr/share/rpm-helper/add-srv
+@mandrake@%_post_srv() %_add_srv_helper %{name} $1 %{1} \
+@mandrake@%{nil}
+@mandrake@
+@mandrake@%_del_srv_helper /usr/share/rpm-helper/del-srv
+@mandrake@%_preun_srv() %_del_srv_helper %{name} $1 %{1} \
+@mandrake@%{nil}
+@mandrake@
+@mandrake@%_mkafterboot	if [[ -f %{_datadir}/afterboot/mkafterboot ]]; then %{_datadir}/afterboot/mkafterboot; fi \
+@mandrake@%{nil}
+@mandrake@
