--- rpm-4.2.2/lib/depends.c.provides	2003-12-01 22:58:10.000000000 +0100
+++ rpm-4.2.2/lib/depends.c	2004-01-19 09:16:38.000000000 +0100
@@ -254,7 +254,7 @@
 
     /* On upgrade, erase older packages of same color (if any). */
 
-    mi = rpmtsInitIterator(ts, RPMTAG_PROVIDENAME, rpmteN(p), 0);
+    mi = rpmtsInitIterator(ts, RPMTAG_NAME, rpmteN(p), 0);
     while((oh = rpmdbNextIterator(mi)) != NULL) {
 
 	/* Ignore colored packages not in our rainbow. */
@@ -288,7 +288,8 @@
 	if (!strcmp(rpmteN(p), Name))
 	    continue;
 
-	mi = rpmtsInitIterator(ts, RPMTAG_PROVIDENAME, Name, 0);
+	/* avoid obsoleting virtual provides (Mandrake packaging policies is not compliant) */
+	mi = rpmtsInitIterator(ts, RPMTAG_NAME, Name, 0);
 
 	xx = rpmdbPruneIterator(mi,
 	    ts->removedPackages, ts->numRemovedPackages, 1);
@@ -304,8 +305,7 @@
 	     * If no obsoletes version info is available, match all names.
 	     */
 	    if (rpmdsEVR(obsoletes) == NULL
-	     || rpmdsAnyMatchesDep(oh, obsoletes, _rpmds_nopromote))
-		if (rpmVersionCompare(h, oh))
+	     || rpmdsNVRMatchesDep(oh, obsoletes, _rpmds_nopromote))
 		    xx = removePackage(ts, oh, rpmdbGetIteratorOffset(mi), pkgKey);
 	}
 	mi = rpmdbFreeIterator(mi);
--- rpm-4.2.2/lib/rpmts.c.provides	2004-01-19 09:14:42.000000000 +0100
+++ rpm-4.2.2/lib/rpmts.c	2004-01-19 09:14:43.000000000 +0100
@@ -879,6 +879,8 @@
 	}
 	rootLen = strlen(rootDir);
 
+	/* FIXME: if multiple / are present, other code will be unhappy */
+
 /*@-branchstate@*/
 	/* Make sure that rootDir has trailing / */
 	if (!(rootLen && rootDir[rootLen - 1] == '/')) {
--- rpm-4.2.2/rpmdb/rpmdb.c.provides	2004-01-19 09:14:42.000000000 +0100
+++ rpm-4.2.2/rpmdb/rpmdb.c	2004-01-19 09:14:43.000000000 +0100
@@ -3667,7 +3667,7 @@
     }
     dbpath = rootdbpath = rpmGetPath(prefix, tfn, NULL);
     if (!(prefix[0] == '/' && prefix[1] == '\0'))
-	dbpath += strlen(prefix);
+        dbpath += strlen(prefix) - 1; /* FIXME: a trailing / is always added */
     tfn = _free(tfn);
 
     /*@-nullpass@*/
