Index: util-linux-2.12q/mount/Makefile
===================================================================
--- util-linux-2.12q.orig/mount/Makefile	2004-12-22 10:32:08.000000000 +0100
+++ util-linux-2.12q/mount/Makefile	2005-04-29 13:59:34.000000000 +0200
@@ -1,7 +1,7 @@
 include ../make_include
 include ../MCONFIG
 
-DEFINES = -DHAVE_NFS $(BLKID_DEFINE)
+DEFINES = -DHAVE_NFS $(BLKID_DEFINE) -D_FILE_OFFSET_BITS=64
 
 RPCSVCDIR = rpcsvc
 RPC_CFLAGS = -Wno-unused 
