--- clamav-0.87/etc/clamd.conf.avx	2005-06-23 16:03:04.000000000 -0400
+++ clamav-0.87/etc/clamd.conf	2005-09-27 23:40:59.000000000 -0400
@@ -4,14 +4,11 @@
 ##
 
 
-# Comment or remove the line below.
-Example
-
 # Uncomment this option to enable logging.
 # LogFile must be writable for the user running daemon.
 # A full path is required.
 # Default: disabled
-#LogFile /tmp/clamd.log
+LogFile /dev/stderr
 
 # By default the log file is locked for writing - the lock protects against
 # running clamd multiple times (if want to run another clamd, please
@@ -31,7 +28,7 @@
 
 # Log time with each message.
 # Default: disabled
-#LogTime
+LogTime
 
 # Also log clean files. Useful in debugging but drastically increases the
 # log size.
@@ -49,27 +46,27 @@
 
 # Enable verbose logging.
 # Default: disabled
-#LogVerbose
+LogVerbose
 
 # This option allows you to save a process identifier of the listening
 # daemon (main thread).
 # Default: disabled
-#PidFile /var/run/clamd.pid
+PidFile /var/run/clamav/clamd.pid
 
 # Optional path to the global temporary directory.
 # Default: system specific (usually /tmp or /var/tmp).
-#TemporaryDirectory /var/tmp
+TemporaryDirectory /var/lib/clamav/tmp
 
 # Path to the database directory.
 # Default: hardcoded (depends on installation options)
-#DatabaseDirectory /var/lib/clamav
+DatabaseDirectory /var/lib/clamav
 
 # The daemon works in a local OR a network mode. Due to security reasons we
 # recommend the local mode.
 
 # Path to a local socket file the daemon will listen on.
 # Default: disabled
-LocalSocket /tmp/clamd
+LocalSocket /var/lib/clamav/clamd.socket
 
 # Remove stale socket after unclean shutdown.
 # Default: disabled
@@ -107,12 +104,12 @@
 
 # Maximal number of threads running at the same time.
 # Default: 10
-#MaxThreads 20
+MaxThreads 64
 
 # Waiting for data from a client socket will timeout after this time (seconds).
 # Value of 0 disables the timeout.
 # Default: 120
-#ReadTimeout 300
+ReadTimeout 300
 
 # Waiting for a new job will timeout after this time (seconds).
 # Default: 30
@@ -124,11 +121,11 @@
 
 # Follow directory symlinks.
 # Default: disabled
-#FollowDirectorySymlinks
+FollowDirectorySymlinks
 
 # Follow regular file symlinks.
 # Default: disabled
-#FollowFileSymlinks
+FollowFileSymlinks
 
 # Perform internal sanity check (database integrity and freshness).
 # Default: 1800 (30 min)
@@ -141,18 +138,18 @@
 
 # Run as a selected user (clamd must be started by root).
 # Default: disabled
-#User clamav
+User clamav
 
 # Initialize supplementary group access (clamd must be started by root).
 # Default: disabled
-#AllowSupplementaryGroups
+AllowSupplementaryGroups
 
 # Stop daemon when libclamav reports out of memory condition.
 #ExitOnOOM
 
 # Don't fork into background.
 # Default: disabled
-#Foreground
+Foreground
 
 # Enable debug messages in libclamav.
 # Default: disabled
@@ -179,7 +176,7 @@
 # required for decompression of popular executable packers such as UPX, FSG,
 # and Petite.
 # Default: enabled
-#ScanPE
+ScanPE
 
 # With this option clamav will try to detect broken executables and mark
 # them as Broken.Executable
@@ -193,7 +190,7 @@
 
 # This option enables scanning of Microsoft Office document macros.
 # Default: enabled
-#ScanOLE2
+ScanOLE2
 
 ##
 ## Mail files
@@ -201,7 +198,7 @@
 
 # Enable internal e-mail scanner.
 # Default: enabled
-#ScanMail
+ScanMail
 
 # If an email contains URLs ClamAV can download and scan them.
 # WARNING: This option may open your system to a DoS attack.
@@ -216,7 +213,7 @@
 
 # Perform HTML normalisation and decryption of MS Script Encoder code.
 # Default: enabled
-#ScanHTML
+ScanHTML
 
 
 ##
@@ -225,7 +222,7 @@
 
 # ClamAV can scan within archives and compressed files.
 # Default: enabled
-#ScanArchive
+ScanArchive
 
 # Due to license issues libclamav does not support RAR 3.0 archives (only the
 # old 2.0 format is supported). Because some users report stability problems
@@ -307,3 +304,4 @@
 # Value of 0 disables the limit.
 # Default: 5M
 #ClamukoMaxFileSize 10M
+
--- clamav-0.87/etc/freshclam.conf.avx	2005-09-15 18:27:13.000000000 -0400
+++ clamav-0.87/etc/freshclam.conf	2005-09-27 23:41:16.000000000 -0400
@@ -4,22 +4,18 @@
 ## This file may be optionally merged with clamd.conf.
 ##
 
-
-# Comment or remove the line below.
-Example
-
 # Path to the database directory.
 # WARNING: It must match clamd.conf's directive!
 # Default: hardcoded (depends on installation options)
-#DatabaseDirectory /var/lib/clamav
+DatabaseDirectory /var/lib/clamav
 
 # Path to the log file (make sure it has proper permissions)
 # Default: disabled
-#UpdateLogFile /var/log/freshclam.log
+UpdateLogFile /dev/stderr
 
 # Enable verbose logging.
 # Default: disabled
-#LogVerbose
+LogVerbose
 
 # Use system logger (can work together with UpdateLogFile).
 # Default: disabled
@@ -32,22 +28,22 @@
 
 # This option allows you to save the process identifier of the daemon
 # Default: disabled
-#PidFile /var/run/freshclam.pid
+PidFile /var/run/clamav/freshclam.pid
 
 # By default when started freshclam drops privileges and switches to the
 # "clamav" user. This directive allows you to change the database owner.
 # Default: clamav (may depend on installation options)
-#DatabaseOwner clamav
+DatabaseOwner clamav
 
 # Initialize supplementary group access (freshclam must be started by root).
 # Default: disabled
-#AllowSupplementaryGroups
+AllowSupplementaryGroups
 
 # Use DNS to verify virus database version. Freshclam uses DNS TXT records
 # to verify database and software versions. With this directive you can change
 # the database verification domain.
 # Default: enabled, pointing to current.cvd.clamav.net
-#DNSDatabaseInfo current.cvd.clamav.net
+DNSDatabaseInfo current.cvd.clamav.net
 
 # Uncomment the following line and replace XY with your country
 # code. See http://www.iana.org/cctld/cctld-whois.htm for the full list.
@@ -62,11 +58,11 @@
 
 # How many attempts to make before giving up.
 # Default: 3 (per mirror)
-#MaxAttempts 5
+MaxAttempts 5
 
 # Number of database checks per day.
 # Default: 12 (every two hours)
-#Checks 24
+Checks 12
 
 # Proxy settings
 # Default: disabled
@@ -85,7 +81,7 @@
 #NotifyClamd
 # By default it uses the hardcoded configuration file but you can force an
 # another one.
-#NotifyClamd /config/file/path
+NotifyClamd /etc/clamd.conf
 
 # Run command after successful database update.
 # Default: disabled
@@ -102,7 +98,7 @@
 
 # Don't fork into background.
 # Default: disabled
-#Foreground
+Foreground
 
 # Enable debug messages in libclamav.
 # Default: disabled
