--- wget-1.8.2/src/log.c.logstdout	2002-05-18 05:05:19.000000000 +0200
+++ wget-1.8.2/src/log.c	2003-01-22 22:36:55.000000000 +0100
@@ -560,11 +560,18 @@
 {
   if (file)
     {
-      logfp = fopen (file, appendp ? "a" : "w");
-      if (!logfp)
+      if (strcmp(file, "-"))
 	{
-	  perror (opt.lfilename);
-	  exit (1);
+          logfp = fopen (file, appendp ? "a" : "w");
+          if (!logfp)
+	    {
+	      perror (opt.lfilename);
+	      exit (1);
+	    }
+        }
+      else
+        {
+	  logfp = stdout;
 	}
     }
   else
