--- openldap-2.3.4/configure.in.bak	2005-06-10 17:11:24.000000000 +0200
+++ openldap-2.3.4/configure.in	2005-08-05 17:21:01.338855823 +0200
@@ -2231,7 +2231,7 @@
 			[have_setproctitle=yes
 			LUTIL_LIBS="$LUTIL_LIBS -lutil"],
 			[have_setproctitle=no
-			LIBOBJS="$LIBOBJS setproctitle.o"
+			AC_LIBOBJ([setproctitle])
 			LIBSRCS="$LIBSRCS setproctitle.c"])])
 
 	if test $have_setproctitle = yes ; then
