--- conf/cupsd.conf.in.orig	2002-06-19 16:38:09.000000000 +0200
+++ conf/cupsd.conf.in	2002-06-19 22:04:03.000000000 +0200
@@ -308,7 +308,7 @@
 # the value of the TMPDIR environment variable.
 #
 
-#TempDir @CUPS_REQUESTS@/tmp
+TempDir @CUPS_REQUESTS@/tmp
 
 #
 # FilterLimit: sets the maximum cost of all job filters that can be run
@@ -446,7 +446,7 @@
 #BrowseAddress x.y.255.255
 #BrowseAddress x.255.255.255
 #BrowseAddress 255.255.255.255
-#BrowseAddress @LOCAL
+BrowseAddress @LOCAL
 #BrowseAddress @IF(name)
 
 #
@@ -487,6 +487,9 @@
 
 #BrowseAllow address
 #BrowseDeny address
+BrowseDeny All
+BrowseAllow 127.0.0.1
+BrowseAllow @LOCAL
 
 #
 # BrowseInterval: the time between browsing updates in seconds.  Default
@@ -506,7 +509,7 @@
 #
 
 #BrowseOrder allow,deny
-#BrowseOrder deny,allow
+BrowseOrder deny,allow
 
 #
 # BrowsePoll: poll the named server(s) for printers
@@ -674,6 +677,7 @@
 Order Deny,Allow
 Deny From All
 Allow From 127.0.0.1
+Allow From @LOCAL
 </Location>
 
 #<Location /classes>
