--- ifplugd-0.15/libdaemon-0.2/src/dfork.c.fred	2003-07-10 18:21:55.000000000 +0200
+++ ifplugd-0.15/libdaemon-0.2/src/dfork.c	2003-08-20 07:48:20.000000000 +0200
@@ -72,22 +72,7 @@
     } else if (pid == 0) {
         // First child
 
-        fclose(pipe_out);
-        
-        if (_null_open(O_RDONLY, 0) < 0) {
-            daemon_log(LOG_ERR, "Failed to open /dev/null for STDIN: %s", strerror(errno));
-            goto fail;
-        }
-        
-        if (_null_open(O_WRONLY, 1) < 0) {
-            daemon_log(LOG_ERR, "Failed to open /dev/null for STDOUT: %s", strerror(errno));
-            goto fail;
-        }
-        
-        if (_null_open(O_WRONLY, 2) < 0) {
-            daemon_log(LOG_ERR, "Failed to open /dev/null for STDERR: %s", strerror(errno));
-            goto fail;
-        }
+        fclose(pipe_out);        
         
         setsid();
         umask(0777);
@@ -131,6 +116,18 @@
 }
 
 int daemon_retval_init(void) {
+    if (_null_open(O_RDONLY, 0) < 0) {
+        daemon_log(LOG_ERR, "Failed to open /dev/null for STDIN: %s", strerror(errno));
+    }
+        
+    if (_null_open(O_WRONLY, 1) < 0) {
+        daemon_log(LOG_ERR, "Failed to open /dev/null for STDOUT: %s", strerror(errno));
+    }
+        
+    if (_null_open(O_WRONLY, 2) < 0) {
+        daemon_log(LOG_ERR, "Failed to open /dev/null for STDERR: %s", strerror(errno));
+    }
+
     if (pipe(_daemon_retval_pipe) < 0)
         return -1;
 
