--- mod_auth_radius.c.orig	2003-02-28 15:46:31.000000000 -0400
+++ mod_auth_radius.c	2003-02-28 15:48:46.000000000 -0400
@@ -2112,7 +2112,12 @@
 	if (gethostname(hostname, MAXHOSTNAMELEN) == 0) {
 
 	    local_address = get_ip_addr(p, hostname);
-	    local_ip.s_addr = local_address->s_addr;
+            if(local_address != NULL) {
+	      local_ip.s_addr = local_address->s_addr;
+            } else {
+	      local_ip.s_addr = 16777343;
+	      fprintf(stderr, "mod_auth_radius: WARNING! INVALID HOSTNAME! Assuming %s\n", inet_ntoa(local_ip));
+            }            
 	    apr_snprintf(cookie_name, 256, "%s-%s", RADIUS_COOKIE_NAME,
 		inet_ntoa(local_ip));
 
