--- nasm-0.98.39/output/outieee.c.can-2005-1194	2005-01-15 15:16:08.000000000 -0700
+++ nasm-0.98.39/output/outieee.c	2005-05-18 14:17:34.697320901 -0600
@@ -1120,7 +1120,7 @@
     va_list ap;
 
     va_start(ap, format);
-    vsprintf(buffer, format, ap);
+    vsnprintf(buffer, sizeof(buffer), format, ap);
     l = strlen(buffer);
     for (i = 0; i < l; i++)
         if ((buffer[i] & 0xff) > 31)
