--- openldap-2.0.27/servers/slapd/Makefile.in.md5	2003-04-29 12:13:03.000000000 -0600
+++ openldap-2.0.27/servers/slapd/Makefile.in	2003-04-29 12:13:03.000000000 -0600
@@ -45,8 +45,9 @@
 # $(LTHREAD_LIBS) must be last
 XLIBS = libbackends.a -lavl -lldbm -lldif -llutil -lldap_r -llber
 XXLIBS = $(LDBM_LIBS) $(SLAPD_LIBS) \
+	 $(LUTIL_LIBS) \
 	 $(SECURITY_LIBS) \
-	 $(LDIF_LIBS) $(LUTIL_LIBS)
+	 $(LDIF_LIBS)
 XXXLIBS = $(LTHREAD_LIBS) $(MODULES_LIBS)
 
 BUILD_OPT = "--enable-slapd"
