--- installkernel.sls	2004-06-15 22:18:21.000000000 -0600
+++ installkernel	2004-06-15 22:18:21.000000000 -0600
@@ -211,7 +211,7 @@
 [[ -z "$NOLINK" ]] && [[ -n "$REMOVE" ]] && do_load_loop_module
 [[ -z "$NOINITRD" ]] && [[ -z "$REMOVE" ]] && $sharfile/make-initrd $NOLINK $version
 if [[ $HOSTTYPE == *586 ]];then
-    [[  ! -e /boot/grub/menu.lst && ! -e /etc/lilo.conf ]] && exit 0 #this is for install time.
+    [[  ! -e /boot/grub/grub.conf && ! -e /etc/lilo.conf ]] && exit 0 #this is for install time.
 elif [[ $HOSTTYPE == *powerpc ]];then
      [[  ! -e /etc/yaboot.conf ]] && exit 0
 fi
--- rebootin.sls	2004-06-15 22:19:29.000000000 -0600
+++ rebootin	2004-06-15 22:19:29.000000000 -0600
@@ -15,7 +15,7 @@
 
 my $lilo_conf = "/etc/lilo.conf";
 
-my $grub_menu = "/boot/grub/menu.lst";
+my $grub_menu = "/boot/grub/grub.conf";
 my $grub_once = "/boot/grub/menu.once";
 my @entries;
 my $bootloader  = `/usr/sbin/detectloader -q` or die "Can't detect your bootloader\n";chomp $bootloader;
@@ -64,7 +64,7 @@
 
     grep { m|^altconfigfile\s.*grub/menu\.once| } @l or die "rebootin is of no use without altconfigfile\n";
     list_entries() if $list;
-    @entries > 0 or die "bad menu.lst (no entry found)\n";
+    @entries > 0 or die "bad grub.conf (no entry found)\n";
     @ARGV == 1 or usage();
 
     for (my $i = 0; $i < @entries; $i++) {
--- rebootin.8.sls	2004-06-15 22:20:03.000000000 -0600
+++ rebootin.8	2004-06-15 22:20:03.000000000 -0600
@@ -26,7 +26,7 @@
 Don't reboot immediately just set the flags for next reboot.
 .SH FILES
 /etc/lilo.conf
-/boot/grub/menu.lst
+/boot/grub/grub.conf
 .br
 The \fBrebootin\fP executable (perl script)
 .PP
--- grub.sls	2004-06-15 22:21:18.000000000 -0600
+++ grub	2004-06-15 22:21:18.000000000 -0600
@@ -50,7 +50,7 @@
 
 # we keep entry as hash for future use.
 my (%main, %entry);
-my $grub_conf = $ENV{GRUB_CONF} || "/boot/grub/menu.lst";
+my $grub_conf = $ENV{GRUB_CONF} || "/boot/grub/grub.conf";
 
 `cp -f $grub_conf ${grub_conf}.old` if -f $grub_conf && !$debug;
 
