--- dovecot-example.conf.avx_config	2007-06-28 09:19:49.604003500 +0300
+++ dovecot-example.conf	2007-06-28 09:19:14.365801250 +0300
@@ -16,7 +16,7 @@
 # --with-ssldir=/etc/ssl
 
 # Base directory where to store runtime data.
-#base_dir = /var/run/dovecot/
+base_dir = /var/lib/dovecot/
 
 # Protocols we want to be serving: imap imaps pop3 pop3s
 # If you only want to use dovecot-auth, you can set this to "none".
@@ -61,7 +61,7 @@
 
 # Log file to use for error messages, instead of sending them to syslog.
 # /dev/stderr can be used to log into stderr.
-#log_path = 
+log_path = /dev/stderr 
 
 # Log file to use for informational and debug messages.
 # Default is the same as log_path.
@@ -91,8 +91,8 @@
 # dropping root privileges, so keep the key file unreadable by anyone but
 # root. Included doc/mkcert.sh can be used to easily generate self-signed
 # certificate, just make sure to update the domains in dovecot-openssl.cnf
-#ssl_cert_file = /etc/ssl/certs/dovecot.pem
-#ssl_key_file = /etc/ssl/private/dovecot.pem
+#ssl_cert_file = /etc/pki/tls/certs/dovecot.pem
+#ssl_key_file = /etc/pki/tls/private/dovecot.pem
 
 # If key file is password protected, give the password here. Alternatively
 # give it when starting dovecot with -p parameter.
@@ -128,7 +128,7 @@
 # which login needs to be able to connect to. The sockets are created when
 # running as root, so you don't have to worry about permissions. Note that
 # everything in this directory is deleted when Dovecot is started.
-#login_dir = /var/run/dovecot/login
+login_dir = /var/lib/dovecot/login
 
 # chroot login process to the login_dir. Only reason not to do this is if you
 # wish to run the whole Dovecot without roots. <doc/wiki/Rootless.txt>
@@ -138,7 +138,7 @@
 # and don't use it anywhere else. The user must also belong to a group where
 # only it has access, it's used to control access for authentication process.
 # Note that this user is NOT used to access mails. <doc/wiki/UserIds.txt>
-#login_user = dovecot
+login_user = dovecot
 
 # Set max. process size in megabytes. If you don't use
 # login_process_per_connection you might need to grow this.
