--- openldap-2.0.3/libraries/liblutil/passwd.c.krb5-1.1	Wed Sep  6 20:11:25 2000
+++ openldap-2.0.3/libraries/liblutil/passwd.c	Fri Sep 15 14:29:25 2000
@@ -753,7 +753,7 @@
 
 		krb5_free_principal( context, client );
 		krb5_free_principal( context, server );
-		krb5_free_cred_contents( context, &creds );
+		krb5_free_creds( context, &creds );
 		krb5_free_context( context );
 
 		rtn = !!ret;
--- openldap-2.0.3/configure.in.krb5-1.1	Tue Sep 12 20:42:32 2000
+++ openldap-2.0.3/configure.in	Fri Sep 15 14:29:25 2000
@@ -763,6 +763,7 @@
 	AC_CHECK_HEADERS(krb5.h)
 
 	if test $ac_cv_header_krb5_h = yes ; then
+		AC_DEFINE(HAVE_KERBEROS_V)
 		dnl lazy check for Heimdal Kerberos
 		AC_CHECK_HEADERS(heim_err.h)
 		if test $ac_cv_header_heim_err_h = yes ; then
@@ -774,9 +775,9 @@
 		if test $krb5_impl = mit; then
 			AC_CHECK_LIB(krb5, main,
 				[have_krb5=yes
-				KRB5_LIBS="-lkrb5 -lcrypto -lcom_err"],
+				KRB5_LIBS="-lkrb5 -lk5crypto -lcom_err"],
 				[have_krb5=no],
-				[-lcrypto -lcom_err])
+				[-lk5crypto -lcom_err])
 
 		elif test $krb5_impl = heimdal; then
 			AC_CHECK_LIB(krb5, main,
@@ -822,7 +823,7 @@
 		if test $krb5_impl = mit; then
 			AC_CHECK_LIB(krb4, main, [have_k425=yes
 				KRB4_LIBS="-lkrb4 -ldes425"], [have_k425=no],
-				[-ldes425 -lkrb5 -lcrypto -lcom_err])
+				[-ldes425 -lkrb5 -lk5crypto -lcom_err])
 
 		elif test $krb5_impl = heimdal; then
 			AC_CHECK_LIB(krb4, main, [have_k425=yes
--- openldap-2.0.3/acconfig.h.krb5-1.1	Fri Sep 15 14:37:46 2000
+++ openldap-2.0.3/acconfig.h	Fri Sep 15 14:38:17 2000
@@ -44,6 +44,9 @@
 /* define this if TIOCGWINSZ is defined in sys/ioctl.h */
 #undef GWINSZ_IN_SYS_IOCTL
 
+/* define this if you have some flavor of Kerberos 5 */
+#undef HAVE_KERBEROS_V
+
 /* These are defined in ldap_features.h */
 /*
 	LDAP_API_FEATURE_X_OPENLDAP_REENTRANT
