Index: squid/src/enums.h
diff -c squid/src/enums.h:1.203.2.15 squid/src/enums.h:1.203.2.16
*** squid/src/enums.h:1.203.2.15	Fri Mar 25 19:50:52 2005
--- squid/src/enums.h	Sat Sep 10 19:05:52 2005
***************
*** 180,186 ****
  
  /* recognized or "known" header fields; @?@ add more! */
  typedef enum {
!     HDR_ACCEPT,
      HDR_ACCEPT_CHARSET,
      HDR_ACCEPT_ENCODING,
      HDR_ACCEPT_LANGUAGE,
--- 180,187 ----
  
  /* recognized or "known" header fields; @?@ add more! */
  typedef enum {
!     HDR_UNKNOWN = -1,
!     HDR_ACCEPT = 0,
      HDR_ACCEPT_CHARSET,
      HDR_ACCEPT_ENCODING,
      HDR_ACCEPT_LANGUAGE,
