--- XFree86-4.2.0/xc/lib/GL/mesa/src/drv/tdfx/tdfx_context.c.fred	Wed Dec 12 19:34:21 2001
+++ XFree86-4.2.0/xc/lib/GL/mesa/src/drv/tdfx/tdfx_context.c	Wed Feb  6 15:59:23 2002
@@ -494,7 +494,7 @@
  */
 GLboolean tdfxInitGlide(tdfxContextPtr tmesa)
 {
-   static const char *defaultGlide = "libglide3.so";
+   static const char *defaultGlide = "libglide3.so.3";
    const char *libName;
    void *libHandle;
 
