--- tetex-src-2.0.2/texk/tetex/allcm.badscript	2003-03-10 12:21:40.000000000 +0000
+++ tetex-src-2.0.2/texk/tetex/allcm	2003-03-10 12:24:02.000000000 +0000
@@ -74,10 +74,9 @@
   echo '\usepackage['$encoding']{fontenc}'
 }
 
-cd /tmp
-(umask 077; mkdir tmp$$) || exit 1
-cd tmp$$ || exit 1
-trap "rc=$?; cd /; rm -f /tmp/tmp$$/*; rmdir /tmp/tmp$$; trap '' 0; exit $rc" 0 1 2 3 7 13 15
+TMPD=`mktemp -d /tmp/allcm.XXXXXX` || exit
+cd $TMPD
+trap "rc=$?; cd /; rm -f $TMPD/*; rmdir /$TMPD; trap '' 0; exit $rc" 0 1 2 3 7 13 15
 
 echo >&2
 echo "---------------------------------------------------------------------" >&2
@@ -107,5 +106,5 @@
 $DVIPS ${1+"$@"} -f allcm12 > /dev/null
 
 cd /
-rm -f /tmp/tmp$$/*
-rmdir /tmp/tmp$$
+rm -f $TMPD/*
+rmdir $TMPD
--- tetex-src-2.0.2/texk/tetex/fmtutil.badscript	2003-03-10 12:24:08.000000000 +0000
+++ tetex-src-2.0.2/texk/tetex/fmtutil	2003-03-10 12:39:33.000000000 +0000
@@ -76,8 +76,6 @@
 
   trap 'cleanup' 1 2 3 7 13 15
   needsCleanup=true
-  (umask 077; mkdir "$tmpdir") \
-    || abort "could not create directory \`$tmpdir'"
 }
 
 ###############################################################################
@@ -376,7 +374,8 @@
   : ${KPSE_DOT=$thisdir}
   export KPSE_DOT
 
-  tmpdir=${TMP-/tmp}/$progname.$$
+  tmpdir=`mktemp -d ${TMP-/tmp}/$progname.XXXXXX` \
+    || abort "could not create secure temporary directory"
   setupTmpDir
   cd "$tmpdir" || {
     false  # some systems need this to set nonzero $?
--- tetex-src-2.0.2/texk/tetex/fontimport.badscript	2003-03-10 12:40:44.000000000 +0000
+++ tetex-src-2.0.2/texk/tetex/fontimport	2003-03-10 12:42:30.000000000 +0000
@@ -146,9 +146,10 @@
   $MT_MKTEXDIR "$DESTDIR"
   test -d $DESTDIR || continue
 
-  $OP $i $DESTDIR/pktmp.$$ || continue
+  TMPFILE=`mktemp $DESTDIR/pktmp.XXXXXX` || continue
+  $OP $i "$TMPFILE" || continue
 
-  mv -f $DESTDIR/pktmp.$$ $DESTDIR/$file || continue
+  mv -f "$TMPFILE" $DESTDIR/$file || continue
   chmod $FILE_PERMS $DESTDIR/$file || continue
   echo "$DESTDIR/$file"
   $MT_MKTEXUPD $DESTDIR $file
@@ -188,9 +189,10 @@
   $MT_MKTEXDIR "$DESTDIR"
   test -d $DESTDIR || continue                               
 
-  $OP $i $DESTDIR/tfmtmp.$$ || continue
+  TMPFILE=`mktemp $DESTDIR/tfmtmp.XXXXXX` || continue
+  $OP $i "$TMPFILE" || continue
 
-  mv -f $DESTDIR/tfmtmp.$$ $DESTDIR/$file || continue
+  mv -f "$TMPFILE" $DESTDIR/$file || continue
   chmod $FILE_PERMS $DESTDIR/$file || continue
   echo "$DESTDIR/$file"
   $MT_MKTEXUPD $DESTDIR $file
--- tetex-src-2.0.2/texk/tetex/texconfig.badscript	2003-03-10 12:42:39.000000000 +0000
+++ tetex-src-2.0.2/texk/tetex/texconfig	2003-03-10 12:44:36.000000000 +0000
@@ -63,7 +63,7 @@
 export KPSE_DOT
 
 progname=`basename $0`
-tmpdir=${TMP-/tmp}/texconf$$
+tmpdir=`mktemp -d "${TMP-/tmp}/texconf.XXXXXX"` || exit 1
 tmpmenu="$tmpdir"/menu.out
 tmpans="$tmpdir"/answer.out
 std_pager=more
@@ -177,8 +177,6 @@
 
   trap 'cleanup' 1 2 3 7 13 15
   needsCleanup=true
-  (umask 077; mkdir "$tmpdir") \
-    || abort "could not create directory \`$tmpdir'"
 }
 
 ###############################################################################
@@ -1384,7 +1382,7 @@
         return 1;; 
   esac
 
-  tmpfile=hy$$
+  tmpfile=`mktemp hy.XXXXXX` || exit 1
   sed 's@[%].*@@; s@^[ 	]*@@; /^$/d' $language | sort > $tmpfile-1
   $EDIT $language
   sed 's@[%].*@@; s@^[ 	]*@@; /^$/d' $language | sort > $tmpfile-2
@@ -1713,7 +1711,7 @@
 {
   exist_test -f "$fmtutilcnf" "fmtutil.cnf file"
 
-  tmpfile=fmt$$
+  tmpfile=`mktemp fmt.XXXXXX` || exit 1
   sed 's@[%#].*@@; s@^[ 	]*@@; /^$/d' $fmtutilcnf | sort > $tmpfile-1
   
   test -z "$NO_CLEAR" && tty >/dev/null 2>&1 && clear;
--- tetex-src-2.0.2/texk/tetex/texlinks.badscript	2003-03-10 12:45:26.000000000 +0000
+++ tetex-src-2.0.2/texk/tetex/texlinks	2003-03-10 12:50:08.000000000 +0000
@@ -83,8 +83,6 @@
 
   trap 'cleanup' 1 2 3 7 13 15
   needsCleanup=true
-  (umask 077; mkdir "$tmpdir") \
-    || abort "could not create directory \`$tmpdir'"
 }
 
 
@@ -182,12 +180,12 @@
   test -z "$kpsewhich" && return 1
   symlinkdir=`echo $kpsewhich | sed 's@/*kpsewhich$@@'`
   access -w "$symlinkdir" || return 1
-  touch "$symlinkdir/tl$$"
-  if test -f "$selfautoloc/tl$$"; then
-    rm -f "$symlinkdir/tl$$"
+  tmpfile=`cd "$symlinkdir"; mktemp tl.XXXXXX`
+  if test -f "$selfautoloc/$tmpfile"; then
+    rm -f "$symlinkdir/$tmpfile"
     return 1
   else
-    rm -f "$symlinkdir/tl$$"
+    rm -f "$symlinkdir/$tmpfile"
     return 0
   fi
 }
@@ -200,7 +198,7 @@
 upd_symlinkdir()
 {
   search_symlinkdir || return 0
-  for i in `sed 's@ .*@@' cnf_file_ln.$$`; do
+  for i in `sed 's@ .*@@' $tmpdir/cnf_file_ln`; do
     install_link "$selfautoloc/$i" "$symlinkdir/$i"
   done
 }
@@ -242,7 +240,7 @@
   test -f "$cnf_file" || abort "config file \`$cnf' not found"
 
 
-  tmpdir=${TMP-/tmp}/texlinks.$$
+  tmpdir=`mktemp -d ${TMP-/tmp}/texlinks.XXXXXX` || exit 1
   setupTmpDir
   cd "$tmpdir" || {
     false  # some systems need this to set nonzero $?
@@ -250,7 +248,7 @@
   }
 
   sed '/^[ 	]*#/d; /^[ 	]*$/d' $cnf_file \
-    | awk '{print $1, $2}' > cnf_file_ln.$$
+    | awk '{print $1, $2}' > $tmpdir/cnf_file_ln
 
   if test -z "$dirs"; then
     if test $multiplatform = true; then
@@ -268,8 +266,8 @@
   for d in $dirs; do
     access -w $d \
       || { errmsg "$d: no write permissions. Skipping..."; continue; }
-    # cnf_file_ln.$$ has lines with "format engine" pairs
-    set x `cat cnf_file_ln.$$`; shift
+    # $tmpdir/cnf_file_ln has lines with "format engine" pairs
+    set x `cat $tmpdir/cnf_file_ln`; shift
     while test $# != 0; do
       fmt=$1; engine=$2; shift; shift
 
--- tetex-src-2.0.2/texk/xdvik/xdvizilla.badscript	2003-03-10 12:15:20.000000000 +0000
+++ tetex-src-2.0.2/texk/xdvik/xdvizilla	2003-03-10 12:16:34.000000000 +0000
@@ -33,7 +33,7 @@
 case "$FILETYPE" in
 
   *"gzip compressed data"*)
-    FILE=/tmp/xdvizilla$$
+    FILE=$(mktemp /tmp/xdvizilla.XXXXXX)
     gunzip -c "$1" > $FILE
     [ -n "$NO_RM" ] || rm -f -- "$1"
     NO_RM=
@@ -41,7 +41,7 @@
     ;;
 
   *"compressed data"* | *"compress'd data"*)
-    FILE=/tmp/xdvizilla$$
+    FILE=$(mktemp /tmp/xdvizilla.XXXXXX)
     uncompress -c "$1" > $FILE
     [ -n "$NO_RM" ] || rm -f -- "$1"
     NO_RM=
@@ -60,7 +60,7 @@
 case "$FILETYPE" in
 
   *" tar archive")
-    TARDIR=/tmp/xdvitar$$
+    TARDIR=$(mktemp /tmp/xdvitar.XXXXXX)
     mkdir $TARDIR
     cat "$FILE" | (cd $TARDIR; tar xf -)
     DVINAME=`tar tf "$FILE" | grep '\.dvi$' | head -1`
--- tetex-src-2.0.2/texk/kpathsea/mktexlsr.badscript	2003-03-10 12:17:09.000000000 +0000
+++ tetex-src-2.0.2/texk/kpathsea/mktexlsr	2003-03-10 12:18:24.000000000 +0000
@@ -95,9 +95,8 @@
   # Skip if we cannot write the file:
   access -w "$db_file" || { echo "$progname: $db_file: no write permission. Skipping..." >&2; continue; }
 
+  db_dir_tmp=`mktemp -d "$db_dir/lsR.XXXXXX"` \
+    || { echo "$progname: could not create secure temporary directory. Skipping..." >&2; continue; }
-  db_dir_tmp="$db_dir/lsR$$.tmp"
-  (umask 077 && mkdir "$db_dir_tmp" ) \
-    || { echo "$progname: could not create directory '$db_dir_tmp'. Skipping..." >&2; continue; }
   db_file_tmp="$db_dir_tmp/lsR$$.tmp"
   rm -f "$db_file_tmp"
 
--- tetex-src-2.0.2/texk/kpathsea/mktexmf.badscript	2003-03-10 12:18:30.000000000 +0000
+++ tetex-src-2.0.2/texk/kpathsea/mktexmf	2003-03-10 12:19:57.000000000 +0000
@@ -87,50 +87,52 @@
   exit 0
 fi
 
+TMPFILE=`mktemp mf.XXXXXX` || exit 1
+
 case "$name" in
   ec*|tc*)
-    cat > "mf$$.tmp" <<END
+    cat > "$TMPFILE" <<END
 if unknown exbase: input exbase fi;
 gensize:=$realsize;
 generate $rootname;
 END
     ;;
   dc*)
-    cat > "mf$$.tmp" <<END
+    cat > "$TMPFILE" <<END
 if unknown dxbase: input dxbase fi;
 gensize:=$realsize;
 generate $rootname;
 END
     ;;
   cs*|lcsss*|icscsc*|icstt*|ilcsss*)
-    cat > "mf$$.tmp" <<END
+    cat > "$TMPFILE" <<END
 input cscode
 use_driver;
 END
     ;;
   wn[bcdfirstuv]*|rx[bcdfiorstuvx][bcfhilmostx]*|l[abcdhl][bcdfiorstuvx]*)
-    cat > "mf$$.tmp" <<END
+    cat > "$TMPFILE" <<END
 input fikparm;
 END
     ;;
   g[lorst][ijmtwx][cinou]*)
     # A small superset of the names of the cbgreek fonts.
-    cat > "mf$$.tmp" <<END
+    cat > "$TMPFILE" <<END
 gensize:=$realsize;
 input $rootname;
 END
     ;;
   *)
-    cat > "mf$$.tmp" <<END
+    cat > "$TMPFILE" <<END
 design_size := $realsize;
 input $rootname;
 END
     ;;
 esac
 
-chmod `kpsestat -xst,go-w .` "mf$$.tmp"
+chmod `kpsestat -xst,go-w .` "$TMPFILE"
 rm -f "$mfname"
-mv "mf$$.tmp" "$mfname"
+mv "$TMPFILE" "$mfname"
 
 echo "$destdir/$mfname" >$STDOUT
 echo "$progname: $destdir/$mfname: successfully generated." >&2
--- tetex-src-2.0.2/texk/dviljk/dvihp.badscript	2003-03-10 12:20:30.000000000 +0000
+++ tetex-src-2.0.2/texk/dviljk/dvihp	2003-03-10 12:21:22.000000000 +0000
@@ -7,9 +7,7 @@
 : ${SPOOL=lpr}          # used to print an LJ file
 : ${TMP=/tmp}        # for the dvicopy output
 
-TMPDIR="${TMP}"/dvihp$$
-(umask 077; mkdir "$TMPDIR") \
-  || { echo "cannot create directory \`$TMPDIR'."; exit 1; }
+TMPDIR=`mktemp -d ${TMPDIR}/dvihp.XXXXXX` || { echo "cannot create secure temporary directory."; exit 1; }
 
 trap '
   cd "${TMP}"
--- tetex-src-2.0.2/texk/dvipdfm/dvipdft.badscript	2003-03-12 12:03:23.000000000 +0000
+++ tetex-src-2.0.2/texk/dvipdfm/dvipdft	2003-03-12 12:05:44.000000000 +0000
@@ -41,12 +41,11 @@
 # Run dvipdfm with the fastest options for the first pass
 if dvipdfm $ARGS -e -z0 $FILENAME; then
 
-  TMP=/tmp/dvipdft.$$
+  TMP=`mktemp -d /tmp/dvipdft.XXXXXX` \
+    || { echo "$0: could not create directory $TMP" >&2; exit 1; }
   export TMP
   trap 'rm -f "$TMP"/*rmdir "$TMP"
         echo "Interrupted!" >&2; exit 1' 1 2 3 7 13 15
-  (umask 077; mkdir "$TMP") \
-    || { echo "$0: could not create directory $TMP" >&2; exit 1; }
 
   if gs -r10 -dNOPAUSE -dBATCH -sDEVICE=png256 \
         -sOutputFile=$TMP/$PDFFILENAME.%d $PDFFILENAME.pdf
--- tetex-src-2.0.2/texk/tetex/updmap.badscript	2003-03-13 11:04:47.000000000 +0000
+++ tetex-src-2.0.2/texk/tetex/updmap	2003-03-13 11:07:15.000000000 +0000
@@ -552,7 +552,6 @@
 
   trap 'cleanup' 1 2 3 7 13 15
   needsCleanup=true
-  mkdir "$tmpdir" || abort "could not create directory \`$tmpdir'"
 }
 
 ###############################################################################
@@ -873,7 +872,8 @@
   outputdir=
   TEXMFMAIN=`kpsewhich --expand-var='$TEXMFMAIN'`
 
-  tmpdir=${TMP-/tmp}/$progname.$$
+  tmpdir=`mktemp -d ${TMP-/tmp}/$progname.XXXXXX` || \
+	abort "could not create directory"
   tmp1=$tmpdir/a
   tmp2=$tmpdir/b
   tmp3=$tmpdir/c
