--- am-utils-6.0.4/conf/nfs_prot/nfs_prot_linux.h.nfs3	Tue Jan 23 15:03:05 2001
+++ am-utils-6.0.4/conf/nfs_prot/nfs_prot_linux.h	Tue Jan 23 15:04:43 2001
@@ -45,6 +45,10 @@
 #ifndef _AMU_NFS_PROT_H
 #define _AMU_NFS_PROT_H
 
+#ifndef _LINUX_NFS3_H
+# define _LINUX_NFS3_H
+#endif /* don't include <linux/nfs3.h> */
+
 #ifdef HAVE_RPCSVC_NFS_PROT_H
 # include <rpcsvc/nfs_prot.h>
 #endif /* HAVE_RPCSVC_NFS_PROT_H */
