--- dietlibc-0.29/test/strncmp.c.fix-strncmp	2007-05-04 15:30:25.000000000 +0200
+++ dietlibc-0.29/test/strncmp.c	2007-05-04 15:30:25.000000000 +0200
@@ -0,0 +1,27 @@
+#include <stdlib.h>
+#include <string.h>
+#include <limits.h>
+
+int main(void)
+{
+  const char foo[] = "foo";
+
+  if (strncmp(foo, "bar", 3) == 0)
+    abort();
+  if (strncmp(foo, "foo", 3) != 0)
+    abort();
+  if (strncmp(foo, "", (size_t)-1) == 0)
+    abort();
+  if (strncmp(foo, "fo", 3) == 0)
+    abort();
+  if (strncmp(foo, "fo", 2) != 0)
+    abort();
+  if (strncmp(foo, "fo", 10) <= 0)
+    abort();
+  if (strncmp(foo, "fo", ULONG_MAX) <= 0)
+    abort();
+  if (strncmp(foo, "fo", (size_t)-1) <= 0)
+    abort();
+
+  return 0;
+}
--- dietlibc-0.29/test/Makefile.fix-strncmp	2007-05-04 15:28:55.000000000 +0200
+++ dietlibc-0.29/test/Makefile	2007-05-04 15:33:52.000000000 +0200
@@ -14,7 +14,7 @@ glob grent hasmntopt hello iconv if_name
 memccpy memchr memcmp memcpy memrchr memusage mktime mmap_test pipe printf printftest \
 protoent prototypes putenv pwent rand48 read1 readdir regex select sendfile servent siglist \
 speed spent sprintf sscanf stdarg strcasecmp strcmp strlen strncat strncpy strptime strrchr \
-strstr strtol sysenter ungetc waitpid
+strstr strtol strncmp sysenter ungetc waitpid
 
 test: $(TESTPROGRAMS)
 
@@ -49,6 +49,10 @@ timetests:
 	$(MAKE) -C time CC="$(CC)" CFLAGS="$(CFLAGS)" DIET=$(DIET) LCOMPAT=$(LCOMPAT)
 	
 	
+str%: str%.c
+	$(DIET) $(CC) $(CFLAGS) -fno-builtin $^ -o $@
+mem%: mem%.c
+	$(DIET) $(CC) $(CFLAGS) -fno-builtin $^ -o $@
 %: %.c
 	$(DIET) $(CC) $(CFLAGS) $^ -o $@
 
--- dietlibc-0.29/lib/strncmp.c.fix-strncmp	2005-04-10 15:59:32.000000000 +0200
+++ dietlibc-0.29/lib/strncmp.c	2007-05-04 15:32:02.000000000 +0200
@@ -6,12 +6,12 @@
 int strncmp(const char *s1, const char *s2, size_t n) {
   register const unsigned char* a=(const unsigned char*)s1;
   register const unsigned char* b=(const unsigned char*)s2;
-  register const unsigned char* fini=a+n;
-  while (a<fini) {
+  while (n>0) {
     register int res=*a-*b;
     if (res) return res;
     if (!*a) return 0;
     ++a; ++b;
+    --n;
   }
   return 0;
 }
