--- chkauth.avx	2003-10-17 12:40:22.000000000 -0400
+++ chkauth	2005-03-28 21:12:27.000000000 -0500
@@ -85,7 +85,7 @@
  
 auth       required       pam_nologin.so
 auth       required       pam_env.so
-auth       sufficient     pam_pwdb.so likeauth nullok
+auth       sufficient     pam_unix.so likeauth nullok
 auth       sufficient     pam_ldap.so use_first_pass
 auth       required       pam_deny.so
  
@@ -94,7 +94,7 @@
 account    required       pam_deny.so
  
 password   required       pam_cracklib.so retry=3
-password   sufficient     pam_pwdb.so nullok use_authtok md5 shadow use_first_pass
+password   sufficient     pam_unix.so nullok use_authtok md5 shadow use_first_pass
 password   sufficient     pam_ldap.so use_authtok
 password   required       pam_deny.so
 
@@ -192,7 +192,7 @@
 }
 
 sub exit_with_usage() {
-    print "Usage drakeauth [file|yp|ldap] ...\n";
+    print "Usage chkauth [file|yp|ldap] ...\n";
     print "file : [noshadow|shadow] [-m md5|crypt] (default shadow, md5)\n";
     print "yp : domainname [-s server]\n";
     print "ldap : [tls|nossl|ssl] [-s server]* [-D base dn]\n";
@@ -236,6 +236,7 @@
     my @serveurs=();
     my $basedn="";
     my $sslmode="tls";
+    my $need_pkg="0";
 
 # parse arg
     while (@ARGV) {
@@ -267,16 +268,19 @@
     if ($PACKAGE_CHECK) {	
 	for $i (@packages) {
 	    if (check_package($i)) {
-		print STDERR ("You have to install $i package for do this\n");
-		exit (2);
+		print STDERR ("You must install the '$i' package to do this\n");
+		$need_pkg="1";
 	    }
 	}
+	if ($need_pkg eq "1") {
+	    exit (2);
+	}
     }
     
     if ($BINARIES_CHECK) {
 	for $i (@binaries) {
 	    if (! -x $i ) {				
-		print STDERR ("You have to have install $i for do this\n"); 	
+		print STDERR ("You must have '$i' installed for this\n"); 	
 		exit (2);
 	    }	    
 	}
@@ -296,7 +300,7 @@
 
     exit_with_usage() if (($basedn eq "") || (scalar(@serveurs) < 1));
 
-    print "LDAP: serveur=@serveurs: basedn=$basedn serveur number ". scalar(@serveurs) . "\n";
+    print "LDAP: server=@serveurs: basedn=$basedn serveur number ". scalar(@serveurs) . "\n";
     set_nsswitch_priority("ldap");
     pam_change ("ldap");
     pam_conf_setup (($sslmode, $basedn, @serveurs));
