--- openldap-2.3.5/build/top.mk.libtool1.5	2005-08-22 13:25:45.000000000 +0200
+++ openldap-2.3.5/build/top.mk	2005-08-22 13:34:09.000000000 +0200
@@ -70,7 +70,7 @@
 
 MKVERSION = $(top_srcdir)/build/mkversion -v "$(VERSION)"
 
-LIBTOOL = @LIBTOOL@
+LIBTOOL = @LIBTOOL@ --tag=CC
 LIBRELEASE = @OPENLDAP_LIBRELEASE@
 LIBVERSION = @OPENLDAP_LIBVERSION@
 LTVERSION = -release $(LIBRELEASE) -version-info $(LIBVERSION)
--- openldap-2.3.5/contrib/slapd-modules/smbk5pwd/Makefile.libtool1.5	2005-08-22 13:25:45.000000000 +0200
+++ openldap-2.3.5/contrib/slapd-modules/smbk5pwd/Makefile	2005-08-22 13:25:45.000000000 +0200
@@ -9,7 +9,7 @@
 # top-level directory of the distribution or, alternatively, at
 # <http://www.OpenLDAP.org/license.html>.
 
-LIBTOOL=../../../libtool
+LIBTOOL=../../../libtool --tag=CC
 OPT=-g -O2
 CC=gcc
 
@@ -23,7 +23,7 @@
 
 HEIMDAL_LIB=-L/usr/heimdal/lib -lkrb5 -lkadm5srv
 SSL_LIB=-lcrypto
-LDAP_LIB=-L../../../libraries/libldap_r/.libs -L../../../libraries/liblber/.libs -lldap_r -llber
+LDAP_LIB=../../../libraries/libldap_r/libldap_r.la ../../../libraries/liblber/liblber.la
 LIBS=$(LDAP_LIB) $(SSL_LIB)
 
 all:	smbk5pwd.la
