--- rsbac-admin-1.2.5/main/libs/Makefile.soname	2006-02-07 16:34:13.000000000 -0700
+++ rsbac-admin-1.2.5/main/libs/Makefile	2006-02-07 16:34:13.000000000 -0700
@@ -71,7 +71,7 @@
         
 $(RSBACSO): $(FILES_LIBS)
 	$(E) "LIB\t\t$<"
-	$(CC) -shared $(CFLAGS) $(LDFLAGS) $(DEFINES) -o $(RSBACSO) $(FILES_LIBS)
+	$(CC) -shared $(CFLAGS) $(LDFLAGS) $(DEFINES) -Wl,-soname,librsbac.so.1 -o $(RSBACSO) $(FILES_LIBS)
 	$(E) "SYM\t\t$(RSBACSO)"
 	ln -s -f $(RSBACSO) librsbac.so
 
