--- chkauth.annvix	2004-06-30 12:33:27.000000000 -0600
+++ chkauth	2004-06-30 12:33:58.000000000 -0600
@@ -85,7 +85,7 @@
  
 auth       required       pam_nologin.so
 auth       required       pam_env.so
-auth       sufficient     pam_pwdb.so likeauth nullok
+auth       sufficient     pam_unix.so likeauth nullok
 auth       sufficient     pam_ldap.so use_first_pass
 auth       required       pam_deny.so
  
@@ -94,7 +94,7 @@
 account    required       pam_deny.so
  
 password   required       pam_cracklib.so retry=3
-password   sufficient     pam_pwdb.so nullok use_authtok md5 shadow use_first_pass
+password   sufficient     pam_unix.so nullok use_authtok md5 shadow use_first_pass
 password   sufficient     pam_ldap.so use_authtok
 password   required       pam_deny.so
 
