--- src/bin/pgaccess/lib/tables.tcl.old	Thu Aug  2 13:36:27 2001
+++ src/bin/pgaccess/lib/tables.tcl	Thu Aug  2 13:36:47 2001
@@ -544,7 +544,8 @@
 		set PgAcVar(mw,$wn,newrec_values) [lreplace $PgAcVar(mw,$wn,newrec_values) $sfp $sfp]
 	}			
 	lappend PgAcVar(mw,$wn,newrec_fields) "\"$fld\""
-	lappend PgAcVar(mw,$wn,newrec_values) '$fldval'
+	regsub -all {'} $fldval '' fldvalfixed
+	lappend PgAcVar(mw,$wn,newrec_values) '$fldvalfixed'
 	# Remove the untouched tag from the object
 	$wn.c dtag $PgAcVar(mw,$wn,id_edited) unt
 		$wn.c itemconfigure $PgAcVar(mw,$wn,id_edited) -fill red
