--- postgresql-7.2.1/src/interfaces/perl5/GNUmakefile.pix	2001-08-27 00:28:04.000000000 +0200
+++ postgresql-7.2.1/src/interfaces/perl5/GNUmakefile	2002-07-10 12:45:15.000000000 +0200
@@ -36,9 +36,7 @@
 	$(PERL) $< \
 	  INC='-I$(srcdir) -I$(libpq_srcdir) $(filter -I%, $(CPPFLAGS))' \
 	  LIBS="-L$$abs_libpq_builddir -lpq" \
-	  INSTALLSITEARCH='$$(DESTDIR)$(perl_installsitearch)' \
-	  INSTALLSITELIB='$$(DESTDIR)$(perl_installsitelib)' \
-	  INSTALLMAN3DIR='$$(DESTDIR)$(perl_installman3dir)' \
+	  INSTALLDIRS=vendor \
 	  MAKEFILE="$$abs_builddir/Makefile"
 
 .PHONY: libpq-all
@@ -61,11 +59,8 @@
 # We need to massage the packlist after installation to get the
 # DESTDIR out of there.
 install: all installdirs
-	@if test -w $(DESTDIR)$(perl_installsitearch); then \
-	  $(MAKE) -f Makefile pure_install DESTDIR=$(DESTDIR) && \
-	  mv $(DESTDIR)$(perl_installsitearch)/auto/Pg/.packlist fake-packlist && \
-	  sed 's,^$(DESTDIR),,' fake-packlist >$(DESTDIR)$(perl_installsitearch)/auto/Pg/.packlist && \
-	  rm fake-packlist; \
+	if test -w $(DESTDIR)$(perl_installsitearch); then \
+	  $(MAKE) -f Makefile pure_install PREFIX=$(DESTDIR)/usr; \
 	else \
 	  $(install-warning-msg); \
 	fi
