--- openldap-2.0.9/servers/slapd/slapd.conf.old	Sat May 19 16:19:12 2001
+++ openldap-2.0.9/servers/slapd/slapd.conf	Sat May 19 16:19:38 2001
@@ -3,9 +3,35 @@
 # See slapd.conf(5) for details on configuration options.
 # This file should NOT be world readable.
 #
-include		%SYSCONFDIR%/schema/core.schema
+# Modified by Christian Zoffoli <czoffoli@linux-mandrake.com>
+# Version 0.2
+#
+
+include	%DATADIR%/schema/core.schema
+include	%DATADIR%/schema/cosine.schema
+include	%DATADIR%/schema/corba.schema 
+include	%DATADIR%/schema/inetorgperson.schema
+include	%DATADIR%/schema/java.schema 
+include	%DATADIR%/schema/krb5-kdc.schema
+include %DATADIR%/schema/kerberosobject.schema
+include	%DATADIR%/schema/misc.schema
+include	%DATADIR%/schema/nadf.schema 
+include	%DATADIR%/schema/nis.schema
+include	%DATADIR%/schema/openldap.schema 
+
+#include %DATADIR%/schema/rfc822-MailMember.schema
+#include %DATADIR%/schema/pilot.schema
+#include %DATADIR%/schema/autofs.schema
+#include %DATADIR%/schema/samba.schema
+#include %DATADIR%/schema/qmail.schema
+#include %DATADIR%/schema/mull.schema
+
+include	%SYSCONFDIR%/schema/local.schema
+
 
 # Define global ACLs to disable default read access.
+include 	%SYSCONFDIR%/slapd.access.conf
+
 
 # Do not enable referrals until AFTER you have a working directory
 # service AND an understanding of referrals.
@@ -14,28 +40,53 @@
 pidfile		%LOCALSTATEDIR%/slapd.pid
 argsfile	%LOCALSTATEDIR%/slapd.args
 
-# Load dynamic backend modules:
-# modulepath	%MODULEDIR%
-# moduleload	back_ldap.la
-# moduleload	back_ldbm.la
-# moduleload	back_passwd.la
-# moduleload	back_shell.la
+modulepath	%MODULEDIR%
+#moduleload      back_dnssrv.la
+#moduleload      back_ldap.la
+#moduleload      back_passwd.la
+#moduleload      back_sql.la
+
+# To allow TLS-enabled connections, create /usr/share/ssl/certs/slapd.pem
+# and uncomment the following lines.
+#TLSCipherSuite         HIGH:MEDIUM:+SSLv2
+TLSCertificateFile      %SYSCONFDIR%/ldap.pem
+TLSCertificateKeyFile   %SYSCONFDIR%/ldap.pem
+TLSCACertificateFile    %SYSCONFDIR%/ldap.pem
 
 #######################################################################
 # ldbm database definitions
 #######################################################################
 
 database	ldbm
-suffix		"dc=my-domain,dc=com"
+suffix		"dc=example,dc=com"
 #suffix		"o=My Organization Name,c=US"
-rootdn		"cn=Manager,dc=my-domain,dc=com"
+rootdn		"cn=Manager,dc=example,dc=com"
 #rootdn		"cn=Manager,o=My Organization Name,c=US"
+
 # Cleartext passwords, especially for the rootdn, should
-# be avoid.  See slappasswd(8) and slapd.conf(5) for details.
+# be avoided.  See slappasswd(8) and slapd.conf(5) for details.
 # Use of strong authentication encouraged.
-rootpw		secret
-# The database directory MUST exist prior to running slapd AND 
-# should only be accessible by the slapd/tools. Mode 700 recommended.
-directory	%LOCALSTATEDIR%/openldap-ldbm
-# Indices to maintain
-index	objectClass	eq
+# rootpw		secret
+# rootpw		{crypt}ijFYNcSNctBYg
+
+# The database directory MUST exist prior to running slapd AND 
+# should only be accessable by the slapd/tools. Mode 700 recommended.
+directory	/var/lib/ldap
+
+# Indices to maintain
+#index	objectClass				eq
+index	objectClass,uid,uidNumber,gidNumber	eq
+index	cn,mail,surname,givenname		eq,subinitial
+
+
+# Basic ACL
+access to attr=userPassword
+        by self write
+        by anonymous auth
+        by dn="uid=root,ou=People,dc=example,dc=com" write
+        by * none
+ 
+access to *
+        by dn="uid=root,ou=People,dc=example,dc=com" write
+        by * read
+
