--- mkinitrd-3.5.18/mkinitrd,scsimult	2004-05-25 17:14:13.000000000 -0600
+++ mkinitrd-3.5.18/mkinitrd	2004-05-25 17:14:13.000000000 -0600
@@ -447,7 +447,7 @@
 	# support standard alias form
 	# support probeall form as found with Mandrake 8.1 and superior
 	# support canonical probeall translation for module-init-tools
-	scsimodules=$(grep -E '^[[:space:]]*(alias|probeall|install)[[:space:]]+scsi_hostadapter' $modulefile | sed 's/^.*scsi_hostadapter//;s/\/sbin\/modprobe//g;s/;//g;s/\/bin\/true//;s/||//')
+	scsimodules=$(grep -E '^[[:space:]]*(alias|probeall|install)[[:space:]]+scsi_hostadapter' $modulefile | sed 's/^.*scsi_hostadapter[0-9]*//;s/\/sbin\/modprobe//g;s/;//g;s/\/bin\/true//;s/||//')
 
 	if [ -n "$scsimodules" ]; then
 	    SAVEMODULES=$MODULES
