--- mkinitrd-3.5.18/mkinitrd.fred	2004-03-11 10:18:01.000000000 +0100
+++ mkinitrd-3.5.18/mkinitrd	2004-03-11 10:20:26.653600614 +0100
@@ -54,7 +54,7 @@
 NB_INODES=100
 POSTSCSIMODS="sd_mod"
 fstab="/etc/fstab"
-IGNOREMODS="$IGNOREMODS ppa imm ide-scsi $initrdfs usb-storage"
+IGNOREMODS="$IGNOREMODS ppa imm ide-scsi $initrdfs"
 
 [[ -f /etc/sysconfig/bootsplash ]] && source /etc/sysconfig/bootsplash
 [[ -n $SPLASH ]] && splash=$SPLASH
