--- util-linux-2.12r/mount/swapon.c.label	2005-11-22 12:06:43.000000000 +0100
+++ util-linux-2.12r/mount/swapon.c	2005-12-26 17:11:38.000000000 +0100
@@ -649,7 +649,7 @@
 			struct stat swapped, swapping;
 
 			if (stat (swapList[i], &swapped) == 0
-					&& stat (fstab->mnt_fsname, &swapping) == 0) {
+					&& stat (special, &swapping) == 0) {
 				if (swapped.st_dev == swapping.st_dev
 						&& swapped.st_ino == swapping.st_ino)
 					break;
@@ -657,10 +657,10 @@
 		}
 
  		for (i = 0; i < numSwaps; i++)
-		     if (!strcmp(swapList[i], fstab->mnt_fsname)) break;
+		     if (!strcmp(swapList[i], special)) break;
         
 		if (!((!skip) && (i == numSwaps) && 
-		         (!ifexists || !access(fstab->mnt_fsname, R_OK))))
+		         (!ifexists || !access(special, R_OK))))
 			continue;
 
  		if (loop && encryption) {
