--- shadow-20000902/src/useradd.c.org	Thu Sep 13 15:22:04 2001
+++ shadow-20000902/src/useradd.c	Thu Sep 13 15:22:27 2001
@@ -1950,13 +1950,14 @@
 	 * to that group, use useradd -g username username.
 	 * --bero
 	 */
+/*
 	if (! (nflg || gflg)) {
 	    if (getgrnam(user_name)) {
                 fprintf(stderr, _("%s: group %s exists - if you want to add this user to that group, use -g.\n"), Prog, user_name);
 		exit(E_NAME_IN_USE);
 	    }
 	}
-
+*/
 	/*
 	 * Do the hard stuff - open the files, create the user entries,
 	 * create the home directory, then close and update the files.
