diff -Naur mysql-4.1.3-beta/acinclude.m4 mysql-4.1.3-beta.oden/acinclude.m4
--- mysql-4.1.3-beta/acinclude.m4	2004-06-27 23:56:49.000000000 +0200
+++ mysql-4.1.3-beta.oden/acinclude.m4	2004-07-18 23:53:40.640597907 +0200
@@ -783,9 +783,9 @@
        fi
       done
 
-      for d in /usr/ssl/lib /usr/local/ssl/lib /usr/lib/openssl \
-/usr/lib /usr/lib64 /opt/ssl/lib /opt/openssl/lib \
-/usr/freeware/lib32 /usr/local/lib/ ; do
+      for d in /usr/ssl/$MYSQL_LIB /usr/local/ssl/$MYSQL_LIB /usr/$MYSQL_LIB/openssl \
+/usr/$MYSQL_LIB /usr/lib64 /opt/ssl/$MYSQL_LIB /opt/openssl/$MYSQL_LIB \
+/usr/freeware/lib32 /usr/local/$MYSQL_LIB/ ; do
       if test -f $d/libssl.a || test -f $d/libssl.so || test -f $d/libssl.dylib ; then
         OPENSSL_LIB=$d
       fi
@@ -1119,7 +1119,7 @@
 dnl echo ["MYSQL_CHECK_BDB_DIR ($1)"]
   dir="$1"
 
-  MYSQL_CHECK_INSTALLED_BDB([$dir/include], [$dir/lib])
+  MYSQL_CHECK_INSTALLED_BDB([$dir/include], [$dir/$MYSQL_LIB])
 
   if test X"$bdb_dir_ok" != Xinstalled; then
     # test to see if it's a source dir
@@ -1653,4 +1653,3 @@
 ])# _AC_PROG_CXX_EXIT_DECLARATION
 
 dnl ---------------------------------------------------------------------------
-
diff -Naur mysql-4.1.3-beta/configure.in mysql-4.1.3-beta.oden/configure.in
--- mysql-4.1.3-beta/configure.in	2004-06-27 23:56:57.000000000 +0200
+++ mysql-4.1.3-beta.oden/configure.in	2004-07-18 23:54:45.874573992 +0200
@@ -2211,6 +2211,32 @@
 fi
 AC_SUBST(tools_dirs)
 
+# Check for native library dir name
+case $host in
+*-*-linux*)
+  # Test if the compiler is 64bit
+  echo 'int i;' > conftest.$ac_ext
+  mysql_cv_cc_64bit_output=no
+  if AC_TRY_EVAL(ac_compile); then
+    case `/usr/bin/file conftest.$ac_objext` in
+    *"ELF 64"*)
+      mysql_cv_cc_64bit_output=yes
+      ;;
+    esac
+  fi
+  rm -rf conftest*
+  ;;
+esac
+
+case $host_cpu:$mysql_cv_cc_64bit_output in
+powerpc64:yes | s390x:yes | sparc64:yes | x86_64:yes)
+  MYSQL_LIB="lib64"
+  ;;
+*:*)
+  MYSQL_LIB="lib"
+  ;;
+esac
+
 #MYSQL_CHECK_CPU
 MYSQL_CHECK_MYSQLFS
 MYSQL_CHECK_VIO
