--- mkinitrd.nodietinsmod	2006-04-17 18:25:59.000000000 -0600
+++ mkinitrd	2006-04-17 18:27:02.000000000 -0600
@@ -549,19 +549,23 @@
 rm -rf $MNTPOINT/lost+found
 
 inst /sbin/nash "$MNTIMAGE/bin/nash"
-if is_dietlibc_arch; then
+#if is_dietlibc_arch; then
+#    if [[ -n "$kernel25" ]]; then
+#	inst /sbin/insmod-25-DIET "$MNTIMAGE/bin/insmod"
+#    else
+#	inst /sbin/insmod-DIET "$MNTIMAGE/bin/insmod"
+#    fi
+#else
     if [[ -n "$kernel25" ]]; then
-	inst /sbin/insmod-25-DIET "$MNTIMAGE/bin/insmod"
-    else
-	inst /sbin/insmod-DIET "$MNTIMAGE/bin/insmod"
-    fi
-else
-    if [[ -n "$kernel25" ]]; then
-	inst /sbin/insmod.static-25 "$MNTIMAGE/bin/insmod"
+        if [ -f /sbin/insmod.static-25 ]; then
+            inst /sbin/insmod.static-25 "$MNTIMAGE/bin/insmod"
+        else
+            inst /sbin/insmod.static "$MNTIMAGE/bin/insmod"
+        fi
     else
 	inst /sbin/insmod.static "$MNTIMAGE/bin/insmod"
     fi
-fi
+#fi
 ln -s ../bin/nash $MNTIMAGE/sbin/modprobe
 
 for MODULE in $MODULES; do
