--- mailx-8.1.1/Makefile.help-files	2002-11-11 06:44:59.000000000 -0500
+++ mailx-8.1.1/Makefile	2002-11-11 06:49:27.000000000 -0500
@@ -42,4 +42,5 @@ default: all
 	install -c -m 2755 -s $(PROG) $(DESTDIR)/usr/bin/
 	install -c -m 644 $(MFILES) $(DESTDIR)/usr/man/man1/
 	cd misc && install -c -m 644 $(EFILES) $(DESTDIR)/etc/
-	cd misc && install -c -m 644 $(SFILES) $(DESTDIR)/usr/lib/
+	mkdir -p $(DESTDIR)/usr/share/mailx/
+	cd misc && install -c -m 644 $(SFILES) $(DESTDIR)/usr/share/mailx/
--- mailx-8.1.1/pathnames.h.help-files	2002-11-11 06:44:59.000000000 -0500
+++ mailx-8.1.1/pathnames.h	2002-11-11 06:47:16.000000000 -0500
@@ -39,8 +39,8 @@
 
 #include <paths.h>
 /* mail installed files */
-#define _PATH_HELP	"/usr/lib/mail.help"
-#define _PATH_TILDE	"/usr/lib/mail.tildehelp"
+#define _PATH_HELP	"/usr/share/mailx/mail.help"
+#define _PATH_TILDE	"/usr/share/mailx/mail.tildehelp"
 #define _PATH_MASTER_RC	"/etc/mail.rc"
 
 /* mail runtime files */
