--- ppp-2.4.3/linux/Makefile.top.etcppp	2005-04-22 19:01:22.832038987 +0200
+++ ppp-2.4.3/linux/Makefile.top	2005-04-22 19:01:24.943061198 +0200
@@ -19,7 +19,7 @@
 	cd pppdump; $(MAKE) $(MFLAGS) all
 	cd contrib/pppgetpass; $(MAKE) $(MFLAGS) all
 
-install: $(BINDIR) $(RUNDIR) $(LOGDIR) $(MANDIR)/man8 install-progs install-devel
+install: $(BINDIR) $(RUNDIR) $(LOGDIR) $(MANDIR)/man8 install-progs install-devel install-etcppp
 
 install-progs:
 	cd chat; $(MAKE) $(MFLAGS) install
