--- courier-imap-3.0.8/authlib/authmksock.c.overflow	2004-06-08 19:44:17.000000000 -0300
+++ courier-imap-3.0.8/authlib/authmksock.c	2005-09-08 17:21:49.000000000 -0300
@@ -29,7 +29,8 @@
 	if (argc < 2)	exit(1);
 	if (fd < 0)	exit(1);
 	skun.sun_family=AF_UNIX;
-	strcpy(skun.sun_path, argv[1]);
+	strncpy(skun.sun_path, argv[1], sizeof(skun.sun_path) - 1);
+	skun.sun_path[sizeof(skun.sun_path) - 1] = '\0';
 	unlink(skun.sun_path);
 	if (bind(fd, (const struct sockaddr *)&skun, sizeof(skun)) ||
 		listen(fd, SOMAXCONN))
