--- rpm-4.2.2/python/rpmdb/Makefile.am.fred	2003-12-16 06:00:41.000000000 +0100
+++ rpm-4.2.2/python/rpmdb/Makefile.am	2004-01-16 06:49:12.000000000 +0100
@@ -10,8 +10,8 @@
 	__init__.py dbobj.py db.py dbrecio.py dbshelve.py dbtables.py dbutils.py
 
 # XXX grrr, sys.path[1] wrong value in python-2.3.2
-pylibdir = $(shell python -c 'import sys; print sys.path[2]')
+pythondir = $(shell python -c "from distutils import sysconfig; print sysconfig.get_python_lib(0,0,prefix='$(prefix)')")
 
-rpmdbdir = $(pylibdir)/site-packages/rpmdb
+rpmdbdir = $(pythondir)/rpmdb
 rpmdb_SCRIPTS = \
 	__init__.py dbobj.py db.py dbrecio.py dbshelve.py dbtables.py dbutils.py
--- rpm-4.2.2/python/Makefile.am.fred	2004-01-01 18:51:27.000000000 +0100
+++ rpm-4.2.2/python/Makefile.am	2004-01-16 06:50:49.000000000 +0100
@@ -4,7 +4,6 @@
 
 LINT = splint
 
-pylibdir = ${prefix}/lib@MARK64@/python@WITH_PYTHON_VERSION@
 pyincdir = $(prefix)/include/python@WITH_PYTHON_VERSION@
 
 SUBDIRS = rpmdb test
@@ -34,10 +33,10 @@
 
 LDADD =
 
-pythondir = $(pylibdir)/site-packages
+pythondir = $(shell python -c "from distutils import sysconfig; print sysconfig.get_python_lib(0,0,prefix='$(prefix)')")
 python_PROGRAMS = rpmmodule.so
 
-rpmdbdir = $(pylibdir)/site-packages/rpmdb
+rpmdbdir = $(pythondir)/rpmdb
 rpmdb_PROGRAMS = _rpmdb.so
 
 noinst_PROGRAMS = poptmodule.so
