--- rpm-4.2/scripts/brp-redhat.in.mdk	2000-02-22 04:09:55.000000000 +0100
+++ rpm-4.2/scripts/brp-redhat.in	2004-01-15 21:10:39.000000000 +0100
@@ -1,13 +1,17 @@
 #!/bin/sh
 
-# These are the build root policies that Red Hat invokes at the end
+# These are the build root policies that Mandrake Linux invokes at the end
 # of the %install scriptlet.
 
-# Compress man pages (Red Hat uses GNU gzip)
-@RPMCONFIGDIR@/brp-compress
+if [ -x /usr/share/spec-helper/spec-helper ];then
+    /usr/share/spec-helper/spec-helper
+else
+    # Compress man pages (Mandrake Linux uses GNU bzip2)
+    @RPMCONFIGDIR@/brp-compress
 
-# Strip ELF binaries (Red Hat uses GNU binutils).
-@RPMCONFIGDIR@/brp-strip
+    # Strip ELF binaries (Mandrake Linux uses GNU binutils).
+    @RPMCONFIGDIR@/brp-strip
 
-# Strip even more sections (Red Hat uses GNU binutils).
-@RPMCONFIGDIR@/brp-strip-comment-note
+    # Strip even more sections (Mandrake Linux uses GNU binutils).
+    @RPMCONFIGDIR@/brp-strip-comment-note
+fi
--- rpm-4.2/scripts/brp-compress.mdk	2004-01-15 21:10:37.000000000 +0100
+++ rpm-4.2/scripts/brp-compress	2004-01-15 21:13:20.000000000 +0100
@@ -8,8 +8,8 @@
 cd $RPM_BUILD_ROOT
 
 # Compress man pages
-COMPRESS="gzip -9 -n"
-COMPRESS_EXT=.gz
+COMPRESS="bzip2 -9 -n"
+COMPRESS_EXT=.bz2
 
 for d in ./usr/man/man* ./usr/man/*/man* ./usr/info \
 	./usr/share/man/man* ./usr/share/man/*/man* ./usr/share/info \
--- rpm-4.2/scripts/cpanflute.mdk	2001-04-22 10:58:08.000000000 +0200
+++ rpm-4.2/scripts/cpanflute	2004-01-15 21:10:39.000000000 +0100
@@ -10,7 +10,7 @@
 
 # set default options, then get options
 my %options=();
-$options{'email'}=(getpwuid($<))[0] . "\@redhat.com" ;
+$options{'email'}=(getpwuid($<))[0] . "\@linux-mandrake.com" ;
 GetOptions(\%options, "email=s", "n=s", "create") || exit 1;
 
 my $InputFile = $ARGV[0];
--- rpm-4.2/lib/rpmrc.c.mdk	2004-01-15 21:10:38.000000000 +0100
+++ rpm-4.2/lib/rpmrc.c	2004-01-15 21:10:39.000000000 +0100
@@ -524,7 +524,7 @@
     addMacro(NULL, "_preScriptEnvironment",NULL, prescriptenviron,RMIL_DEFAULT);
 
     setVarDefault(-1,			"_topdir",
-		"/usr/src/redhat",	"%{_usr}/src/redhat");
+		"/usr/src/RPM",		"%{_usr}/src/RPM");
     setVarDefault(-1,			"_tmppath",
 		"/var/tmp",		"%{_var}/tmp");
     setVarDefault(-1,			"_dbpath",
--- rpm-4.2/GROUPS.mdk	1999-03-16 21:04:26.000000000 +0100
+++ rpm-4.2/GROUPS	2004-01-15 21:10:39.000000000 +0100
@@ -1,29 +1,84 @@
-Amusements/Games
-Amusements/Graphics
-Applications/Archiving
-Applications/Communications
-Applications/Databases
-Applications/Editors
-Applications/Emulators
-Applications/Engineering
-Applications/File
-Applications/Internet
-Applications/Multimedia
-Applications/Productivity
-Applications/Publishing
-Applications/System
-Applications/Text
-Development/Debuggers
-Development/Languages
-Development/Libraries
-Development/System
-Development/Tools
-Documentation
-System Environment/Base
-System Environment/Daemons
-System Environment/Kernel
-System Environment/Libraries
-System Environment/Shells
-User Interface/Desktops
-User Interface/X
-User Interface/X Hardware Support
+Accessibility
+Archiving/Backup
+Archiving/Cd burning
+Archiving/Compression
+Archiving/Other
+Books/Computer books
+Books/Faqs
+Books/Howtos
+Books/Litterature
+Books/Other
+Communications
+Databases
+Development/C
+Development/C++
+Development/Databases
+Development/GNOME and GTK+
+Development/Java
+Development/KDE and QT
+Development/Kernel
+Development/Other
+Development/Perl
+Development/Python
+Editors
+Emulators
+File tools
+Games/Adventure
+Games/Arcade
+Games/Boards
+Games/Cards
+Games/Other
+Games/Puzzles
+Games/Sports
+Games/Strategy
+Graphical desktop/Enlightenment
+Graphical desktop/FVWM based
+Graphical desktop/GNOME
+Graphical desktop/Icewm
+Graphical desktop/KDE
+Graphical desktop/Other
+Graphical desktop/Sawfish
+Graphical desktop/Window Maker
+Graphics
+Monitoring
+Networking/Chat
+Networking/File transfer
+Networking/IRC
+Networking/Instant messaging
+Networking/Mail
+Networking/News
+Networking/Other
+Networking/Remote access
+Networking/WWW
+Office
+Publishing
+Sciences/Astronomy
+Sciences/Biology
+Sciences/Chemistry
+Sciences/Computer science
+Sciences/Geosciences
+Sciences/Mathematics
+Sciences/Other
+Sciences/Physics
+Shells
+Sound
+System/Base
+System/Configuration/Boot and Init
+System/Configuration/Hardware
+System/Configuration/Networking
+System/Configuration/Other
+System/Configuration/Packaging
+System/Configuration/Printing
+System/Fonts/Console
+System/Fonts/True type
+System/Fonts/Type1
+System/Fonts/X11 bitmap
+System/Internationalization
+System/Kernel and hardware
+System/Libraries
+System/Servers
+System/XFree86
+Terminals
+Text tools
+Toys
+Video
--- rpm-4.2/rpmrc.in.mdk	2004-01-15 21:10:39.000000000 +0100
+++ rpm-4.2/rpmrc.in	2004-01-15 21:15:48.000000000 +0100
@@ -11,56 +11,57 @@
 #############################################################
 # Values for RPM_OPT_FLAGS for various platforms
 
-optflags: i386 -O2 -g -march=i386 -mcpu=i686
-optflags: i486 -O2 -g -march=i486
-optflags: i586 -O2 -g -march=i586
-optflags: i686 -O2 -g -march=i686
-optflags: athlon -O2 -g -march=athlon
-optflags: ia64 -O2 -g
-optflags: x86_64 -O2 -g
-optflags: amd64 -O2 -g
-
-optflags: alpha -O2 -g -mieee
-optflags: alphaev5 -O2 -g -mieee -mcpu=ev5
-optflags: alphaev56 -O2 -g -mieee -mcpu=ev56
-optflags: alphapca56 -O2 -g -mieee -mcpu=pca56
-optflags: alphaev6 -O2 -g -mieee -mcpu=ev6
-optflags: alphaev67 -O2 -g -mieee -mcpu=ev67
-
-optflags: sparc -O2 -g -m32 -mtune=ultrasparc
-optflags: sparcv9 -O2 -g -m32 -mcpu=ultrasparc
-optflags: sparc64 -O2 -g -m64 -mcpu=ultrasparc
-
-optflags: m68k -O2 -g -fomit-frame-pointer
-
-optflags: ppc -O2 -g -fsigned-char
-optflags: ppciseries -O2 -g -fsigned-char
-optflags: ppcpseries -O2 -g -fsigned-char
-optflags: ppc64 -O2 -g -fsigned-char
-
-optflags: parisc -O2 -g -mpa-risc-1-0
-optflags: hppa1.0 -O2 -g -mpa-risc-1-0
-optflags: hppa1.1 -O2 -g -mpa-risc-1-0
-optflags: hppa1.2 -O2 -g -mpa-risc-1-0
-optflags: hppa2.0 -O2 -g -mpa-risc-1-0
-
-optflags: mips -O2 -g
-optflags: mipsel -O2 -g
-
-optflags: armv3l -O2 -g -fsigned-char -fomit-frame-pointer -march=armv3
-optflags: armv4b -O2 -g -fsigned-char -fomit-frame-pointer -march=armv4
-optflags: armv4l -O2 -g -fsigned-char -fomit-frame-pointer -march=armv4
-
-optflags: atarist -O2 -g -fomit-frame-pointer
-optflags: atariste -O2 -g -fomit-frame-pointer
-optflags: ataritt -O2 -g -fomit-frame-pointer
-optflags: falcon -O2 -g -fomit-frame-pointer
-optflags: atariclone -O2 -g -fomit-frame-pointer
-optflags: milan -O2 -g -fomit-frame-pointer
-optflags: hades -O2 -g -fomit-frame-pointer
+optflags: i386 -O2 -fomit-frame-pointer -pipe -march=i386 %{debugcflags}
+optflags: i486 -O2 -fomit-frame-pointer -pipe -march=i486 %{debugcflags}
+optflags: k6   -O2 -fomit-frame-pointer -pipe -march=k6 %{debugcflags}
+optflags: i586 -O2 -fomit-frame-pointer -pipe -march=i586 -mcpu=pentiumpro %{debugcflags}
+optflags: i686 -O2 -fomit-frame-pointer -pipe -march=i686 %{debugcflags}
+optflags: athlon -O2 -fomit-frame-pointer -pipe -march=athlon %{debugcflags}
+optflags: ia64 -O2 -pipe %{debugcflags}
+optflags: x86_64 -O2 -pipe %{debugcflags}
+  
+  # XXX Please note that -mieee has been added in rpm-3.0.5.
+optflags: alpha -O2 -mieee -mcpu=ev5 -pipe %{debugcflags}
+optflags: alphaev5 -O2 -mieee -mcpu=ev5 -pipe %{debugcflags}
+optflags: alphaev56 -O2 -mieee -mcpu=ev56 -pipe %{debugcflags}
+optflags: alphapca56 -O2 -mieee -mcpu=pca56 -pipe %{debugcflags}
+optflags: alphaev6 -O2 -mieee -mcpu=ev6 -pipe %{debugcflags}
+optflags: alphaev67 -O2 -mieee -mcpu=ev67 -pipe %{debugcflags}
+
+optflags: sparc -O2 -m32 -mtune=ultrasparc %{debugcflags}
+optflags: sparcv9 -O2 -m32 -mcpu=ultrasparc %{debugcflags}
+optflags: sparc64 -O2 -m64 -mcpu=ultrasparc %{debugcflags}
+
+optflags: m68k -O2 %{debugcflags} -fomit-frame-pointer
+
+optflags: ppc -O2 -fsigned-char -frename-registers -mcpu=750 -mtune=7450 -pipe %{debugcflags}
+optflags: ppciseries -O2 %{debugcflags} -fsigned-char
+optflags: ppcpseries -O2 %{debugcflags} -fsigned-char
+optflags: ppc64 -O2 %{debugcflags} -fsigned-char
+
+optflags: parisc -O2 %{debugcflags} -mpa-risc-1-0
+optflags: hppa1.0 -O2 %{debugcflags} -mpa-risc-1-0
+optflags: hppa1.1 -O2 %{debugcflags} -mpa-risc-1-0
+optflags: hppa1.2 -O2 %{debugcflags} -mpa-risc-1-0
+optflags: hppa2.0 -O2 %{debugcflags} -mpa-risc-1-0
+
+optflags: mips -O2 %{debugcflags}
+optflags: mipsel -O2 %{debugcflags}
+
+optflags: armv3l -O2 %{debugcflags} -fsigned-char -fomit-frame-pointer -march=armv3
+optflags: armv4b -O2 %{debugcflags} -fsigned-char -fomit-frame-pointer -march=armv4
+optflags: armv4l -O2 %{debugcflags} -fsigned-char -fomit-frame-pointer -march=armv4
+
+optflags: atarist -O2 %{debugcflags} -fomit-frame-pointer
+optflags: atariste -O2 %{debugcflags} -fomit-frame-pointer
+optflags: ataritt -O2 %{debugcflags} -fomit-frame-pointer
+optflags: falcon -O2 %{debugcflags} -fomit-frame-pointer
+optflags: atariclone -O2 %{debugcflags} -fomit-frame-pointer
+optflags: milan -O2 %{debugcflags} -fomit-frame-pointer
+optflags: hades -O2 %{debugcflags} -fomit-frame-pointer
 
-optflags: s390 -O2 -g
-optflags: s390x -O2 -g
+optflags: s390 -O2 %{debugcflags}
+optflags: s390x -O2 %{debugcflags}
 
 #############################################################
 # Canonical arch names and numbers
@@ -126,6 +127,8 @@
 arch_canon:	sh: sh		17
 arch_canon:	xtensa: xtensa	18
 
+arch_canon:	x86_64:	x86_64	19
+
 #############################################################
 # Canonical OS names and numbers
 
@@ -164,10 +167,11 @@
 buildarchtranslate: osfmach3_i486: i386
 buildarchtranslate: osfmach3_i386: i386
 
-buildarchtranslate: athlon: i386
-buildarchtranslate: i686: i386
-buildarchtranslate: i586: i386
-buildarchtranslate: i486: i386
+buildarchtranslate: athlon: i586
+buildarchtranslate: i686: i586
+buildarchtranslate: k6: i586
+buildarchtranslate: i586: i586
+buildarchtranslate: i486: i486
 buildarchtranslate: i386: i386
 
 buildarchtranslate: alphaev5: alpha
@@ -216,6 +220,7 @@
 
 arch_compat: athlon: i686
 arch_compat: i686: i586
+arch_compat: k6: i586
 arch_compat: i586: i486
 arch_compat: i486: i386
 arch_compat: i386: noarch
@@ -305,6 +310,7 @@
 
 buildarch_compat: athlon: i686
 buildarch_compat: i686: i586
+buildarch_compat: k6: i486
 buildarch_compat: i586: i486
 buildarch_compat: i486: i386
 buildarch_compat: i386: noarch
--- rpm-4.2/platform.in.mdk	2003-01-08 21:30:50.000000000 +0100
+++ rpm-4.2/platform.in	2004-01-15 21:10:39.000000000 +0100
@@ -90,11 +90,18 @@
 @mandrake@# XXX These values, particularly vendor, need to be regularized, please.
 @mandrake@%_vendor		MandrakeSoft
 @mandrake@%_real_vendor		mandrake
-@mandrake@%_target_platform	%{_target_cpu}-%{_real_vendor}-%{_target_os}
+@mandrake@%_target_platform	%{_target_cpu}-%{_real_vendor}-%{_target_os}%{?_gnu}
+@mandrake@
+@mandrake@# Mandrake Linux version
+@mandrake@# - "9.1"    =>  910
+@mandrake@# - "10.2.2" => 1022
+@mandrake@# (user may copy the following line in specfile)
+@mandrake@%mdkversion		%(perl -pe '/(\\d+)\\.(\\d)\\.?(\\d)?/; $_="$1$2".($3||0)' /etc/mandrake-release)
 @mandrake@
 @mandrake@# XXX I'll make these the default linux values soon as I can.
 @mandrake@%_libexecdir		%{_libdir}
-@mandrake@%_localstatedir		%{_var}/lib
+@mandrake@%_localstatedir	%{_var}/lib
+@mandrake@%_sysconfdir		/etc
 @mandrake@
 @mandrake@# XXX A copy of brp-mandrake, please. I'd suggest break down by
 @mandrake@#     functionality rather than vendor <shrug>.
@@ -102,43 +109,133 @@
 @mandrake@    @RPMCONFIGDIR@/brp-mandrake \
 @mandrake@%{nil}
 @mandrake@
+@mandrake@%configure2_5x \
+@mandrake@  CONFIGURE_TOP="${CONFIGURE_TOP:-.}"; \
+@mandrake@  CFLAGS="${CFLAGS:-%optflags}" ; export CFLAGS ; \
+@mandrake@  CXXFLAGS="${CXXFLAGS:-%optflags}" ; export CXXFLAGS ; \
+@mandrake@  FFLAGS="${FFLAGS:-%optflags}" ; export FFLAGS ; \
+@mandrake@  %{?__cputoolize:%{__cputoolize} -c $CONFIGURE_TOP} ; \
+@mandrake@  %{?__libtoolize:(cd $CONFIGURE_TOP; [ -f configure.in -o -f configure.ac ] && %{__libtoolize} --copy --force)} ; \
+@mandrake@  $CONFIGURE_TOP/configure --build=%{_target_platform} \\\
+@mandrake@ 	--prefix=%{_prefix} \\\
+@mandrake@	--exec-prefix=%{_exec_prefix} \\\
+@mandrake@	--bindir=%{_bindir} \\\
+@mandrake@	--sbindir=%{_sbindir} \\\
+@mandrake@	--sysconfdir=%{_sysconfdir} \\\
+@mandrake@	--datadir=%{_datadir} \\\
+@mandrake@	--includedir=%{_includedir} \\\
+@mandrake@	--libdir=%{_libdir} \\\
+@mandrake@	--libexecdir=%{_libexecdir} \\\
+@mandrake@	--localstatedir=%{_localstatedir} \\\
+@mandrake@	--sharedstatedir=%{_sharedstatedir} \\\
+@mandrake@	--mandir=%{_mandir} \\\
+@mandrake@	--infodir=%{_infodir}
+@mandrake@
+@mandrake@%makeinstall_std make DESTDIR=%{?buildroot:%{buildroot}} install
+@mandrake@
+@mandrake@%old_makeinstall \
+@mandrake@    make \\\
+@mandrake@	prefix=%{?buildroot:%{buildroot}}%{_prefix} \\\
+@mandrake@	exec_prefix=%{?buildroot:%{buildroot}}%{_exec_prefix} \\\
+@mandrake@	bindir=%{?buildroot:%{buildroot}}%{_bindir} \\\
+@mandrake@	sbindir=%{?buildroot:%{buildroot}}%{_sbindir} \\\
+@mandrake@	sysconfdir=%{?buildroot:%{buildroot}}%{_sysconfdir} \\\
+@mandrake@	datadir=%{?buildroot:%{buildroot}}%{_datadir} \\\
+@mandrake@	includedir=%{?buildroot:%{buildroot}}%{_includedir} \\\
+@mandrake@	libdir=%{?buildroot:%{buildroot}}%{_libdir} \\\
+@mandrake@	libexecdir=%{?buildroot:%{buildroot}}%{_libexecdir} \\\
+@mandrake@	localstatedir=%{?buildroot:%{buildroot}}%{_localstatedir} \\\
+@mandrake@	sharedstatedir=%{?buildroot:%{buildroot}}%{_sharedstatedir} \\\
+@mandrake@	mandir=%{?buildroot:%{buildroot}}%{_mandir} \\\
+@mandrake@	infodir=%{?buildroot:%{buildroot}}%{_infodir} \\\
+@mandrake@    install
+@mandrake@
+@mandrake@%_create_ghostfile_helper /usr/share/rpm-helper/create-file
+@mandrake@%create_ghostfile() %_create_ghostfile_helper %{name} $1 %{1} %{2} %{3} %{4} \
+@mandrake@%{nil}
+@mandrake@
 @mandrake@# make
 @mandrake@%_make_bin make
-@mandrake@%make if [ -z "$NPROCS" -a -f /proc/stat ]; then NPROCS=`egrep -c ^cpu[0-9]+ /proc/stat || :`; fi \
-@mandrake@if [ -z "$NPROCS" -o "$NPROCS" -le "0" ]; then \
-@mandrake@  NPROCS=1 \
-@mandrake@fi \
-@mandrake@%{_make_bin} -j$NPROCS
+@mandrake@%make %{_make_bin} %_smp_mflags
+@mandrake@
+@mandrake@#	Default extension to use.
+@mandrake@%_extension		.bz2
 @mandrake@
 @mandrake@# Menu directories
-@mandrake@%_menudir %{_libdir}/menu
+@mandrake@%_menudir %{_prefix}/lib/menu
 @mandrake@%_iconsdir %{_datadir}/icons
 @mandrake@%_miconsdir %{_datadir}/icons/mini
 @mandrake@%_liconsdir %{_datadir}/icons/large
 @mandrake@
 @mandrake@# Update Menu
 @mandrake@%_update_menus_bin %{_bindir}/update-menus
-@mandrake@%update_menus if [ -x %{_update_menus_bin} ]; then %{_update_menus_bin} || true ; fi
+@mandrake@%update_menus if [ -x %{_update_menus_bin} ]; then %{_update_menus_bin} || true ; fi \
+@mandrake@%{nil}
 @mandrake@
 @mandrake@# Clean Menu
-@mandrake@%clean_menus if [ "$1" = "0" -a -x %{_update_menus_bin} ]; then %{_update_menus_bin} || true ; fi
+@mandrake@%clean_menus if [ "$1" = "0" -a -x %{_update_menus_bin} ]; then %{_update_menus_bin} || true ; fi \
+@mandrake@%{nil}
 @mandrake@
 @mandrake@# Update Window Managers session.
 @mandrake@%_fndsession_bin %{_sbindir}/fndSession
-@mandrake@%make_session if [ -x %{_fndsession_bin} ]; then %{_fndsession_bin} || true ; fi
+@mandrake@%make_session if [ -x %{_fndsession_bin} ]; then %{_fndsession_bin} || true ; fi \
+@mandrake@%{nil}
 @mandrake@
 @mandrake@#	Where install-info is located
 @mandrake@%__install_info		/sbin/install-info
 @mandrake@
 @mandrake@#	Macros to install info-files.
-@mandrake@%_install_info()	%{__install_info} %{_infodir}/%{1}%{_extension} --dir=%{_infodir}/dir\;
+@mandrake@%_install_info()	if [[ -f %{_infodir}/%{1}%{_extension} ]];then %{__install_info} %{_infodir}/%{1}%{_extension} --dir=%{_infodir}/dir;fi \
+@mandrake@%{nil}
 @mandrake@
 @mandrake@#	Macros to remove info-files. (keep the "alone on its line ';'")
-@mandrake@%_remove_install_info() if [ "$1" = "0" ]; then %{__install_info} \\\
-@mandrake@ %{_infodir}/%{1}%{_extension} --dir=%{_infodir}/dir --remove ; fi\
+@mandrake@%_remove_install_info() if [ "$1" = "0" ]; then if [[ -f %{_infodir}/%{1}%{_extension} ]];then %{__install_info} %{_infodir}/%{1}%{_extension} --dir=%{_infodir}/dir --remove ;fi; fi \
 @mandrake@%{nil}
 @mandrake@
 @mandrake@# Games macros
 @mandrake@%_gamesdir	games
 @mandrake@%_gamesbindir   %{_prefix}/%{_gamesdir}
 @mandrake@%_gamesdatadir  %{_datadir}/%{_gamesdir}
+@mandrake@
+@mandrake@# Servers opt flags
+@mandrake@%serverbuild export CFLAGS="%optflags -fno-omit-frame-pointer"; export CXXFLAGS="%optflags -fno-omit-frame-pointer"; export RPM_OPT_FLAGS="%optflags -fno-omit-frame-pointer" \
+@mandrake@%{nil}
+@mandrake@
+@mandrake@# Macro: %{mklibname <name> [<major> [<minor>]] [-s] [-d]}
+@mandrake@# - %{mklibname test}				=> lib64test, on a lib64 platform
+@mandrake@# - %{mklibname test 1 -d}		=> libtest1-devel
+@mandrake@# - %{mklibname test 1 -d 0 -s}	=> libtest1_0-static-devel
+@mandrake@%mklibname(ds)	%{_lib}%{1}%{?2:%{2}}%{?3:_%{3}}%{-s:-static}%{-d:-devel}
+@mandrake@
+@mandrake@# initscripts macros
+@mandrake@%_add_service_helper /usr/share/rpm-helper/add-service
+@mandrake@%_post_service() %_add_service_helper %{name} $1 %{1} \
+@mandrake@%{nil}
+@mandrake@
+@mandrake@%_del_service_helper /usr/share/rpm-helper/del-service
+@mandrake@%_preun_service() %_del_service_helper %{name} $1 %{1} \
+@mandrake@%{nil}
+@mandrake@
+@mandrake@%_add_user_helper /usr/share/rpm-helper/add-user
+@mandrake@%_pre_useradd() %_add_user_helper %{name} $1 %{1} %{2} %{3} \
+@mandrake@%{nil}
+@mandrake@
+@mandrake@%_del_user_helper /usr/share/rpm-helper/del-user
+@mandrake@%_postun_userdel() %_del_user_helper %{name} $1 %{1} \
+@mandrake@%{nil}
+@mandrake@
+@mandrake@%_add_group_helper /usr/share/rpm-helper/add-group
+@mandrake@%_pre_groupadd() %_add_group_helper %{name} $1 %{1} %{?2:%2} \
+@mandrake@%{nil}
+@mandrake@
+@mandrake@%_del_group_helper /usr/share/rpm-helper/del-group
+@mandrake@%_postun_groupdel() %_del_group_helper %{name} $1 %{1} \
+@mandrake@%{nil}
+@mandrake@
+@mandrake@%_add_shell_helper /usr/share/rpm-helper/add-shell
+@mandrake@%_post_shelladd() %_add_shell_helper %{name} $1 %{1} \
+@mandrake@%{nil}
+@mandrake@
+@mandrake@%_del_shell_helper /usr/share/rpm-helper/del-shell
+@mandrake@%_preun_shelldel() %_del_shell_helper %{name} $1 \
+@mandrake@%{nil}
--- rpm-4.2/macros.in.mdk	2004-01-15 21:10:39.000000000 +0100
+++ rpm-4.2/macros.in	2004-01-15 21:10:39.000000000 +0100
@@ -81,6 +81,7 @@
 
 # XXX avoid failures if tools are not installed when rpm is built.
 %__libtoolize		libtoolize
+%__cputoolize		cputoolize
 %__aclocal		aclocal
 %__autoheader		autoheader
 %__automake		automake
@@ -131,6 +132,14 @@
 %endif\
 %{nil}
 
+#%_enable_debug_packages 1
+
+%debugcflags %{?_enable_debug_packages:-g}
+
+%install %{?_enable_debug_packages:%{?buildsubdir:%{debug_package}}}\
+%%install\
+%{nil}
+
 %_defaultdocdir		%{_usr}/doc
 
 #	The path to the gzip executable (legacy, use %{__gzip} instead).
@@ -168,7 +177,7 @@
 %_tmppath		%{_var}/tmp
 
 #	Path to top of build area.
-%_topdir		%{_usrsrc}/redhat
+%_topdir		%{_usrsrc}/RPM
 
 #	The path to the unzip executable (legacy, use %{__unzip} instead).
 %_unzipbin		%{__unzip}
@@ -285,6 +294,25 @@
 #
 #%vendor
 
+#
+# Script gets packaged file list on input and buildroot as first parameter.
+# Returns list of unpackaged files, i.e. files in $RPM_BUILD_ROOT not packaged.
+#
+# Note: Disable (by commenting out) for legacy compatibility.
+%__check_files         @prefix@/lib/rpm/check-files %{buildroot}
+
+#
+# Should unpackaged files in a build root terminate a build?
+#
+# Note: The default value should be 0 for legacy compatibility.
+%_unpackaged_files_terminate_build	1
+
+#
+# Should missing %doc files in the build directory terminate a build?
+#
+# Note: The default value should be 0 for legacy compatibility.
+%_missing_doc_files_terminate_build	1
+
 #==============================================================================
 # ---- Build configuration macros.
 #
@@ -321,7 +349,7 @@
 
 #
 # Use internal dependency generator rather than external helpers?
-%_use_internal_dependency_generator	1
+%_use_internal_dependency_generator	0
 
 #
 # Filter GLIBC_PRIVATE Provides:/Requires:
@@ -333,8 +361,8 @@
 # Note: Used iff _use_internal_dependency_generator is zero.
 #%__find_provides	@RPMCONFIGDIR@/rpmdeps --provides
 #%__find_requires	@RPMCONFIGDIR@/rpmdeps --requires
-%__find_provides	@RPMCONFIGDIR@/find-provides
-%__find_requires	@RPMCONFIGDIR@/find-requires
+%__find_provides	@RPMCONFIGDIR@/filter.sh '%{?_provides_exceptions:%{_provides_exceptions}}%{!?_provides_exceptions: }' @RPMCONFIGDIR@/find-provides
+%__find_requires	@RPMCONFIGDIR@/filter.sh '%{?_requires_exceptions:%{_requires_exceptions}}%{!?_requires_exceptions: }' @RPMCONFIGDIR@/find-requires %{?buildroot:%{buildroot}} %{?_target_cpu:%{_target_cpu}}
 #%__find_conflicts	???
 #%__find_obsoletes	???
 
@@ -965,11 +993,13 @@
 #------------------------------------------------------------------------------
 # This is an improved version of %configure (from PLD team).
 %configure \
+  CONFIGURE_TOP="${CONFIGURE_TOP:-.}"; \
   CFLAGS="${CFLAGS:-%optflags}" ; export CFLAGS ; \
   CXXFLAGS="${CXXFLAGS:-%optflags}" ; export CXXFLAGS ; \
   FFLAGS="${FFLAGS:-%optflags}" ; export FFLAGS ; \
-  ./configure --host=%{_host} --build=%{_build} \\\
-	--target=%{_target_platform} \\\
+  %{?__cputoolize:%{__cputoolize} -c $CONFIGURE_TOP} ; \
+  %{?__libtoolize:(cd $CONFIGURE_TOP; [ -f configure.in -o -f configure.ac ] && %{__libtoolize} --copy --force)} ; \
+  $CONFIGURE_TOP/configure %{_target_platform} \\\
 	--program-prefix=%{?_program_prefix} \\\
  	--prefix=%{_prefix} \\\
 	--exec-prefix=%{_exec_prefix} \\\
@@ -1061,28 +1091,34 @@
   %{-C:cd ${_mydir}; unset _mydir}
 
 #------------------------------------------------------------------------------
-# Useful perl macros (from Artur Frysiak <wiget@t17.ds.pwr.wroc.pl>)
+# Useful perl macros 
+# (initially from Artur Frysiak <wiget@t17.ds.pwr.wroc.pl>)
+# (modified by Pixel <pixel@mandrakesoft.com>)
 #
-# For example, these can be used as (from ImageMagick.spec from PLD site)
 #	[...]
-#	BuildPrereq: perl
+#	Group: Development/Perl
 #	[...]
-#	%package perl
-#	Summary: libraries and modules for access to ImageMagick from perl
-#	Group: Development/Languages/Perl
-#	Requires: %{name} = %{version}
-#	%requires_eq    perl
+#	%build
+#	perl Makefile.PL INSTALLDIRS=vendor
+#	%make
 #	[...]
 #	%install
-#	rm -fr $RPM_BUILD_ROOT
-#	install -d $RPM_BUILD_ROOT/%{perl_sitearch}
+#	rm -rf $RPM_BUILD_ROOT
+#	make install PREFIX="$RPM_BUILD_ROOT/%{prefix}"
 #	[...]
-#	%files perl
+#	%files
 #	%defattr(644,root,root,755)
-#	%{perl_sitearch}/Image
-#	%dir %{perl_sitearch}/auto/Image
+#	%{perl_vendorlib}/CGI
+# or
+#	%{perl_vendorarch}/DBI
+#	%{perl_vendorarch}/auto/DBI
+#	%{_mandir}/*/*
 #
 %requires_eq()	%(LC_ALL="C" echo '%*' | xargs -r rpm -q --qf 'Requires: %%{name} = %%{epoch}:%%{version}\\n' | sed -e 's/ (none):/ /' -e 's/ 0:/ /' | grep -v "is not")
+%perl_vendorarch	%(eval "`perl -V:installvendorarch`"; echo $installvendorarch)
+%perl_vendorlib	%(eval "`perl -V:installvendorlib`"; echo $installvendorlib)
+#
+# kept for compatibility, but should not be used, especially the ugly perl_archlib!
 %perl_sitearch	%(eval "`perl -V:installsitearch`"; echo $installsitearch)
 %perl_archlib	%(eval "`perl -V:installarchlib`"; echo $installarchlib)
 
--- rpm-4.2/installplatform.mdk	2004-01-15 21:10:39.000000000 +0100
+++ rpm-4.2/installplatform	2004-01-15 21:18:35.000000000 +0100
@@ -28,18 +28,21 @@
 RPMRC_GNU="`$RPM --eval '%{_gnu}'`"
 target_platform="`$RPM --eval '%{?_gnu:%undefine _gnu}%{_target_platform}'|sed -e "$canonarch_sed"`"
 target="`$RPM --eval '%{_target}'|sed -e "$canonarch_sed"`"
+targetcpu="`$RPM --eval '%{_target_cpu}'`"
 
 case "$arch" in
-  i[3456]86|athlon) SUBSTS='s_i386_i386_ s_i386_i486_ s_i386_i586_ s_i386_i686_ s_i386_athlon_' ;;
+  i[3456]86|athlon) SUBSTS="s_${arch}_i386_ s_${arch}_i486_ s_${arch}_i586_ s_${arch}_k6_ s_${arch}_i686_ s_${arch}_athlon_" ;;
   alpha*) SUBSTS='s_alpha_alpha_ s_alpha_alphaev5_ s_alpha_alphaev56_ s_alpha_alphapca56_ s_alpha_alphaev6_ s_alpha_alphaev67_' ;;
   sparc*) SUBSTS='s_sparc\(64\|v9\)_sparc_ s_sparc64_sparcv9_;s_sparc\([^v]\|$\)_sparcv9\1_ s_sparcv9_sparc64_;s_sparc\([^6]\|$\)_sparc64\1_' ;;
-  powerpc*|ppc*) SUBSTS='s_ppc64_ppc_ s_ppc\([^6ip]\|$\)_ppc64\1_ s_ppc\([^6ip]\|$\)_ppciseries_ s_ppc\([^6ip]\|$\)_ppcpseries_ s_ppc\([^6ip]\|$\)_ppc64iseries_ s_ppc\([^6ip]\|$\)_ppc64pseries_' ;;
+  powerpc*|ppc*) SUBSTS='s_ppc64_ppc_ s_ppc\([^6]\|$\)_ppc64\1_ s_${arch}_ppciseries_ s_${arch}_ppcpseries_' ;;
+  x86_64*) SUBSTS='s/x86_64/x86_64/' ;;
   s390*) SUBSTS='s_s390x_s390_ s_s390\([^x]\|$\)_s390x\1_' ;;
   *) SUBSTS=y___ ;;
 esac
 
 for SUBST in $SUBSTS ; do
   ARCH=`echo $arch | sed -e $SUBST`
+  SUBST=`echo $SUBST | sed -e "s@${arch}@${targetcpu}@"`
   TARGET_PLATFORM=`echo $target_platform | sed -e $SUBST`
   TARGET=`echo $target | sed -e $SUBST`
   LIB=lib
@@ -52,7 +55,7 @@
   RPMRC_OPTFLAGS="`echo $RPMRC_OPTFLAGS | sed -e 's, ,\ ,g'`"
   case $RPMRC_OPTFLAGS in
   *-g*) ;;
-  *) RPMRC_OPTFLAGS="$RPMRC_OPTFLAGS -g" ;;
+  *) RPMRC_OPTFLAGS="$RPMRC_OPTFLAGS" ;;
   esac
 
   ARCH_INSTALL_POST='%{nil}'
@@ -85,7 +88,6 @@
     ;;
   mandrake)
     VENDORSED='-e s,^@mandrake@,,'
-    RPMRC_OPTFLAGS="`echo $RPMRC_OPTFLAGS | sed -e s/i386/i686/`"
     ;;
   conectiva)
     VENDORSED='-e s,^@conectiva@,,'
@@ -103,6 +105,7 @@
 	-e "s,@ARCH_INSTALL_POST@,$ARCH_INSTALL_POST," \
 	-e "s,@DEFAULTDOCDIR@,$DEFAULTDOCDIR," \
 	-e '/\${\w*:-/!s,\${,%{_,' \
+	-e '/\(FLAGS\|CONFIGURE_TOP\):-/!s,\${,%{_,' \
 	$MULTILIBSED \
 	$VENDORSED \
   | grep -v '^@' \
--- rpm-4.2/rpmpopt.in.mdk	2004-01-15 21:10:39.000000000 +0100
+++ rpm-4.2/rpmpopt.in	2004-01-15 21:10:39.000000000 +0100
@@ -147,28 +147,27 @@
 # Popt glue to preserve legacy CLI behavior.
 #
 # XXX popt exec parsing doesn't honor POPT_ARGFLAG_ONEDASH
-# XXX Build mode aliases, removed in rpm-4.1.
-#rpm	exec --bp		rpmb -bp
-#rpm	exec --bc		rpmb -bc
-#rpm	exec --bi		rpmb -bi
-#rpm	exec --bl		rpmb -bl
-#rpm	exec --ba		rpmb -ba
-#rpm	exec --bb		rpmb -bb
-#rpm	exec --bs		rpmb -bs
-#rpm	exec --tp		rpmb -tp
-#rpm	exec --tc		rpmb -tc
-#rpm	exec --ti		rpmb -ti
-#rpm	exec --tl		rpmb -tl
-#rpm	exec --ta		rpmb -ta
-#rpm	exec --tb		rpmb -tb
-#rpm	exec --ts		rpmb -ts
-#rpm	exec --rebuild		rpmb --rebuild
-#rpm	exec --recompile	rpmb --recompile
-#rpm	exec --clean		rpmb --clean
-#rpm	exec --rmsource		rpmb --rmsource
-#rpm	exec --rmspec		rpmb --rmspec
-#rpm	exec --target		rpmb --target
-#rpm	exec --short-circuit	rpmb --short-circuit
+rpm	exec --bp		rpmb -bp
+rpm	exec --bc		rpmb -bc
+rpm	exec --bi		rpmb -bi
+rpm	exec --bl		rpmb -bl
+rpm	exec --ba		rpmb -ba
+rpm	exec --bb		rpmb -bb
+rpm	exec --bs		rpmb -bs
+rpm	exec --tp		rpmb -tp
+rpm	exec --tc		rpmb -tc
+rpm	exec --ti		rpmb -ti
+rpm	exec --tl		rpmb -tl
+rpm	exec --ta		rpmb -ta
+rpm	exec --tb		rpmb -tb
+rpm	exec --ts		rpmb -ts
+rpm	exec --rebuild		rpmb --rebuild
+rpm	exec --recompile	rpmb --recompile
+rpm	exec --clean		rpmb --clean
+rpm	exec --rmsource		rpmb --rmsource
+rpm	exec --rmspec		rpmb --rmspec
+rpm	exec --target		rpmb --target
+rpm	exec --short-circuit	rpmb --short-circuit
 
 rpm	exec --initdb		rpmd --initdb
 rpm	exec --rebuilddb	rpmd --rebuilddb
