diff -uNr orig/arch/i386/Makefile linux-2.4.20/arch/i386/Makefile
--- orig/arch/i386/Makefile	2003-05-01 16:34:16.000000000 -0500
+++ linux-2.4.20/arch/i386/Makefile	2003-05-01 16:48:56.000000000 -0500
@@ -53,24 +53,60 @@
 CFLAGS += -march=i586
 endif
 
+ifdef CONFIG_M586MMX31
+CFLAGS += -march=pentium -mmmx
+endif
+
 ifdef CONFIG_M686
 CFLAGS += -march=i686
 endif
 
+ifdef CONFIG_MPPRO31
+CFLAGS += -march=pentiumpro
+endif
+
+ifdef CONFIG_M68631
+CFLAGS += -march=pentium2
+endif
+
 ifdef CONFIG_MPENTIUMIII
 CFLAGS += -march=i686
 endif
 
+ifdef CONFIG_MPENTIUMIII31
+CFLAGS += -march=pentium3
+endif
+
 ifdef CONFIG_MPENTIUM4
 CFLAGS += -march=i686
 endif
 
+ifdef CONFIG_MPENTIUM431
+CFLAGS += -march=pentium4
+endif
+
 ifdef CONFIG_MK6
 CFLAGS += $(call check_gcc,-march=k6,-march=i586)
 endif
 
+ifdef CONFIG_MK6231
+CFLAGS += -march=k6-2
+endif
+
 ifdef CONFIG_MK7
-CFLAGS += $(call check_gcc,-march=athlon,-march=i686 -malign-functions=4)
+CFLAGS += -march=i686
+endif
+
+ifdef CONFIG_MK731
+CFLAGS += -march=athlon
+endif
+
+ifdef CONFIG_MXP31
+CFLAGS += -march=athlon-xp
+endif
+
+ifdef CONFIG_MMP31
+CFLAGS += -march=athlon-mp
 endif
 
 ifdef CONFIG_MCRUSOE
diff -uNr orig/arch/i386/config.in linux-2.4.20/arch/i386/config.in
--- orig/arch/i386/config.in	2003-05-01 16:34:16.000000000 -0500
+++ linux-2.4.20/arch/i386/config.in	2003-05-01 16:53:02.000000000 -0500
@@ -34,11 +34,20 @@
 	 586/K5/5x86/6x86/6x86MX		CONFIG_M586 \
 	 Pentium-Classic			CONFIG_M586TSC \
 	 Pentium-MMX				CONFIG_M586MMX \
+	 Pentium-MMX(gcc>31)                    CONFIG_M586MMX31 \
 	 Pentium-Pro/Celeron/Pentium-II		CONFIG_M686 \
+	 Pentium-Pro(gcc>31)                    CONFIG_MPPRO31 \
+	 Pentium-II/Celeron(gcc>31)             CONFIG_M68631 \
 	 Pentium-III/Celeron(Coppermine)	CONFIG_MPENTIUMIII \
+	 Pentium-III/Celeron(Coppermine)(gcc>31)        CONFIG_MPENTIUMIII31 \
 	 Pentium-4				CONFIG_MPENTIUM4 \
+ 	 Pentium-4(gcc>31)                      CONFIG_MPENTIUM431 \
 	 K6/K6-II/K6-III			CONFIG_MK6 \
+	 K6-II/K6-III(gcc>31)                   CONFIG_MK6231 \
 	 Athlon/Duron/K7			CONFIG_MK7 \
+	 Athlon/Duron/K7(>gcc31)                        CONFIG_MK731 \
+         AthlonXP(gcc>31)                       CONFIG_MXP31 \
+         AthlonMP(gcc>31)                       CONFIG_MMP31 \
 	 Opteron/Athlon64/Hammer/K8		CONFIG_MK8 \
 	 Elan					CONFIG_MELAN \
 	 Crusoe					CONFIG_MCRUSOE \
@@ -100,6 +109,14 @@
    define_bool CONFIG_X86_PPRO_FENCE y
    define_bool CONFIG_X86_F00F_WORKS_OK n
 fi
+if [ "$CONFIG_M586MMX31" = "y" ]; then
+   define_int  CONFIG_X86_L1_CACHE_SHIFT 5
+   define_bool CONFIG_X86_USE_STRING_486 y
+   define_bool CONFIG_X86_ALIGNMENT_16 y
+   define_bool CONFIG_X86_HAS_TSC y
+   define_bool CONFIG_X86_GOOD_APIC y
+   define_bool CONFIG_X86_PPRO_FENCE y
+fi
 if [ "$CONFIG_M686" = "y" ]; then
    define_int  CONFIG_X86_L1_CACHE_SHIFT 5
    define_bool CONFIG_X86_HAS_TSC y
@@ -109,6 +126,39 @@
    define_bool CONFIG_X86_PPRO_FENCE y
    define_bool CONFIG_X86_F00F_WORKS_OK y
 fi
+if [ "$CONFIG_MPENTIUM431" = "y" ]; then
+    define_int  CONFIG_X86_L1_CACHE_SHIFT 7
+    define_bool CONFIG_X86_HAS_TSC y
+    define_bool CONFIG_X86_GOOD_APIC y
+    define_bool CONFIG_X86_PGE y
+    define_bool CONFIG_X86_USE_PPRO_CHECKSUM y
+ fi
+
+if [ "$CONFIG_MK6231" = "y" ]; then
+    define_int  CONFIG_X86_L1_CACHE_SHIFT 5
+    define_bool CONFIG_X86_ALIGNMENT_16 y
+    define_bool CONFIG_X86_HAS_TSC y
+    define_bool CONFIG_X86_USE_PPRO_CHECKSUM y
+ fi
+
+if [ "$CONFIG_MPPRO31" = "y" ]; then
+   define_int  CONFIG_X86_L1_CACHE_SHIFT 5
+   define_bool CONFIG_X86_HAS_TSC y
+   define_bool CONFIG_X86_GOOD_APIC y
+   define_bool CONFIG_X86_PGE y
+   define_bool CONFIG_X86_USE_PPRO_CHECKSUM y
+   define_bool CONFIG_X86_PPRO_FENCE y
+   define_bool CONFIG_X86_F00F_WORKS_OK y
+fi
+if [ "$CONFIG_M68631" = "y" ]; then
+   define_int  CONFIG_X86_L1_CACHE_SHIFT 5
+   define_bool CONFIG_X86_HAS_TSC y
+   define_bool CONFIG_X86_GOOD_APIC y
+   define_bool CONFIG_X86_PGE y
+   define_bool CONFIG_X86_USE_PPRO_CHECKSUM y
+   define_bool CONFIG_X86_PPRO_FENCE y
+   define_bool CONFIG_X86_F00F_WORKS_OK y
+fi
 if [ "$CONFIG_MPENTIUMIII" = "y" ]; then
    define_int  CONFIG_X86_L1_CACHE_SHIFT 5
    define_bool CONFIG_X86_HAS_TSC y
@@ -117,6 +167,13 @@
    define_bool CONFIG_X86_USE_PPRO_CHECKSUM y
    define_bool CONFIG_X86_F00F_WORKS_OK y
 fi
+if [ "$CONFIG_MPENTIUMIII31" = "y" ]; then
+   define_int  CONFIG_X86_L1_CACHE_SHIFT 5
+   define_bool CONFIG_X86_HAS_TSC y
+   define_bool CONFIG_X86_GOOD_APIC y
+   define_bool CONFIG_X86_PGE y
+   define_bool CONFIG_X86_USE_PPRO_CHECKSUM y
+fi
 if [ "$CONFIG_MPENTIUM4" = "y" ]; then
    define_int  CONFIG_X86_L1_CACHE_SHIFT 7
    define_bool CONFIG_X86_HAS_TSC y
@@ -145,6 +202,30 @@
    define_bool CONFIG_X86_USE_PPRO_CHECKSUM y
    define_bool CONFIG_X86_F00F_WORKS_OK y
 fi
+if [ "$CONFIG_MK731" = "y" ]; then
+   define_int  CONFIG_X86_L1_CACHE_SHIFT 6
+   define_bool CONFIG_X86_HAS_TSC y
+   define_bool CONFIG_X86_GOOD_APIC y
+   define_bool CONFIG_X86_USE_3DNOW y
+   define_bool CONFIG_X86_PGE y
+   define_bool CONFIG_X86_USE_PPRO_CHECKSUM y
+fi
+if [ "$CONFIG_MXP31" = "y" ]; then
+   define_int  CONFIG_X86_L1_CACHE_SHIFT 6
+   define_bool CONFIG_X86_HAS_TSC y
+   define_bool CONFIG_X86_GOOD_APIC y
+   define_bool CONFIG_X86_USE_3DNOW y
+   define_bool CONFIG_X86_PGE y
+   define_bool CONFIG_X86_USE_PPRO_CHECKSUM y
+fi
+if [ "$CONFIG_MMP31" = "y" ]; then
+   define_int  CONFIG_X86_L1_CACHE_SHIFT 6
+   define_bool CONFIG_X86_HAS_TSC y
+   define_bool CONFIG_X86_GOOD_APIC y
+   define_bool CONFIG_X86_USE_3DNOW y
+   define_bool CONFIG_X86_PGE y
+   define_bool CONFIG_X86_USE_PPRO_CHECKSUM y
+fi
 if [ "$CONFIG_MELAN" = "y" ]; then
    define_int  CONFIG_X86_L1_CACHE_SHIFT 4
    define_bool CONFIG_X86_USE_STRING_486 y
diff -uNr orig/arch/i386/lib/mmx.c linux-2.4.20/arch/i386/lib/mmx.c
--- orig/arch/i386/lib/mmx.c	2003-05-01 16:34:16.000000000 -0500
+++ linux-2.4.20/arch/i386/lib/mmx.c	2003-05-01 16:43:57.000000000 -0500
@@ -121,7 +121,8 @@
 	return p;
 }
 
-#ifdef CONFIG_MK7
+#if defined(CONFIG_MK7) || defined(CONFIG_MK731) \
+			|| defined(CONFIG_MXP31) || defined(CONFIG_MMP31)
 
 /*
  *	The K7 has streaming cache bypass load/store. The Cyrix III, K6 and
diff -uNr orig/include/asm-i386/processor.h linux-2.4.20/include/asm-i386/processor.h
--- orig/include/asm-i386/processor.h	2003-05-01 16:35:25.000000000 -0500
+++ linux-2.4.20/include/asm-i386/processor.h	2003-05-01 16:54:38.000000000 -0500
@@ -488,7 +488,7 @@
 #define cpu_relax()	rep_nop()
 
 /* Prefetch instructions for Pentium III and AMD Athlon */
-#if defined(CONFIG_MPENTIUMIII) || defined (CONFIG_MPENTIUM4)
+#if defined(CONFIG_MPENTIUMIII) || defined(CONFIG_MPENTIUMIII31)
 
 #define ARCH_HAS_PREFETCH
 extern inline void prefetch(const void *x)
