diff -Naur php-5.1.0RC1/scripts/Makefile.frag php-5.1.0RC1.oden/scripts/Makefile.frag
--- php-5.1.0RC1/scripts/Makefile.frag	2005-08-08 18:06:48.000000000 +0200
+++ php-5.1.0RC1.oden/scripts/Makefile.frag	2005-09-20 16:12:16.000000000 +0200
@@ -4,7 +4,7 @@
 #
 
 phpincludedir = $(includedir)/php
-phpbuilddir = $(libdir)/build
+phpbuilddir = $(libdir)/php/build
 
 BUILD_FILES = \
 	scripts/phpize.m4 \
diff -Naur php-5.1.0RC1/scripts/phpize.in php-5.1.0RC1.oden/scripts/phpize.in
--- php-5.1.0RC1/scripts/phpize.in	2005-08-08 18:58:12.000000000 +0200
+++ php-5.1.0RC1.oden/scripts/phpize.in	2005-09-20 16:13:00.000000000 +0200
@@ -3,7 +3,7 @@
 # Variable declaration
 prefix='@prefix@'
 exec_prefix="`eval echo @exec_prefix@`"
-phpdir="`eval echo @libdir@`/build"
+phpdir="`eval echo @libdir@`/php/build"
 includedir="`eval echo @includedir@`/php"
 builddir="`pwd`"
 
