--- acinclude.m4	2007-05-24 23:40:41.000000000 +0200
+++ acinclude.m4.oden	2007-06-20 13:13:58.000000000 +0200
@@ -429,6 +429,7 @@
 dnl internal, don't use
 AC_DEFUN([_PHP_ADD_LIBPATH_GLOBAL],[
   PHP_RUN_ONCE(LIBPATH, $1, [
+    test "x$PHP_RPATH" != "xno" &&
     test -n "$ld_runpath_switch" && LDFLAGS="$LDFLAGS $ld_runpath_switch$1"
     LDFLAGS="$LDFLAGS -L$1"
     PHP_RPATHS="$PHP_RPATHS $1"
@@ -447,7 +448,11 @@
       _PHP_ADD_LIBPATH_GLOBAL([$ai_p])
     ],[
       if test "$ext_shared" = "yes"; then
+        if test "x$PHP_RPATH" = "xno"; then
+          $2="-L$ai_p [$]$2"
+        else
         $2="$ld_runpath_switch$ai_p -L$ai_p [$]$2"
+        fi
       else
         _PHP_ADD_LIBPATH_GLOBAL([$ai_p])
       fi
