diff -Naur php-5.1.0RC4/Makefile.global php-5.1.0RC4.oden/Makefile.global
--- php-5.1.0RC4/Makefile.global	2005-11-03 04:19:32.000000000 +0100
+++ php-5.1.0RC4.oden/Makefile.global	2005-11-03 04:17:55.000000000 +0100
@@ -51,7 +51,7 @@
 install-su: install-pear
 
 test: all 
-	-@if test ! -z "$(PHP_EXECUTABLE)" && test -x "$(PHP_EXECUTABLE)"; then \
+	@if test ! -z $(PHP_EXECUTABLE) && test -x $(PHP_EXECUTABLE); then \
 		TEST_PHP_EXECUTABLE=$(PHP_EXECUTABLE) \
 		TEST_PHP_SRCDIR=$(top_srcdir) \
 		CC="$(CC)" \
@@ -60,7 +60,7 @@
 		TEST_PHP_EXECUTABLE=$(top_builddir)/$(SAPI_CLI_PATH) \
 		TEST_PHP_SRCDIR=$(top_srcdir) \
 		CC="$(CC)" \
-			$(top_builddir)/$(SAPI_CLI_PATH) -d 'open_basedir=' -d 'safe_mode=0' -d 'output_buffering=0' -d 'memory_limit=-1' $(top_srcdir)/run-tests.php $(TESTS); \
+			$(top_builddir)/$(SAPI_CLI_PATH) -d 'open_basedir=' -d 'safe_mode=0' -n -d 'output_buffering=0' -d 'memory_limit=-1' $(top_srcdir)/run-tests.php $(TESTS); \
 	else \
 		echo "ERROR: Cannot run tests without CLI sapi."; \
 	fi
diff -Naur php-5.1.0RC4/ext/standard/tests/file/bug26615.phpt php-5.1.0RC4.oden/ext/standard/tests/file/bug26615.phpt
--- php-5.1.0RC4/ext/standard/tests/file/bug26615.phpt	2005-06-18 04:24:37.000000000 +0200
+++ php-5.1.0RC4.oden/ext/standard/tests/file/bug26615.phpt	2005-11-03 04:16:14.000000000 +0100
@@ -6,7 +6,7 @@
 <?php
 $out = array();
 $status = -1;
-exec($_ENV['TEST_PHP_EXECUTABLE'].' -r \'for($i=1;$i<=5000;$i++) print "$i\n";\' | tr \'\n\' \' \'', $out, $status);
+exec($_ENV['TEST_PHP_EXECUTABLE'].' -n -r \'for($i=1;$i<=5000;$i++) print "$i\n";\' | tr \'\n\' \' \'', $out, $status);
 print_r($out);
 ?>
 --EXPECT--
diff -Naur php-5.1.0RC4/ext/standard/tests/file/bug26938.phpt php-5.1.0RC4.oden/ext/standard/tests/file/bug26938.phpt
--- php-5.1.0RC4/ext/standard/tests/file/bug26938.phpt	2004-01-21 23:48:28.000000000 +0100
+++ php-5.1.0RC4.oden/ext/standard/tests/file/bug26938.phpt	2005-11-03 04:16:14.000000000 +0100
@@ -5,7 +5,7 @@
 $out = array();
 $status = -1;
 $php = getenv('TEST_PHP_EXECUTABLE');
-exec($php . ' -r \'' 
+exec($php . ' -n -r \'' 
      . '$lengths = array(10,20000,10000,5,10000,3);'
      . 'foreach($lengths as $length) {'
      . '  for($i=0;$i<$length;$i++) print chr(65+$i % 27);'
diff -Naur php-5.1.0RC4/ext/standard/tests/file/proc_open01.phpt php-5.1.0RC4.oden/ext/standard/tests/file/proc_open01.phpt
--- php-5.1.0RC4/ext/standard/tests/file/proc_open01.phpt	2003-07-15 19:28:41.000000000 +0200
+++ php-5.1.0RC4.oden/ext/standard/tests/file/proc_open01.phpt	2005-11-03 04:16:14.000000000 +0100
@@ -9,7 +9,7 @@
 	die("no php executable defined");
 } 
 $proc = proc_open(
-	$php,
+	$php . " -n",
 	array(0 => array('pipe', 'r'), 1 => array('pipe', 'w')),
 	$pipes
 );
diff -Naur php-5.1.0RC4/ext/xml/tests/bug32001.phpt php-5.1.0RC4.oden/ext/xml/tests/bug32001.phpt
--- php-5.1.0RC4/ext/xml/tests/bug32001.phpt	2005-06-29 13:52:18.000000000 +0200
+++ php-5.1.0RC4.oden/ext/xml/tests/bug32001.phpt	2005-11-03 04:16:14.000000000 +0100
@@ -157,7 +157,7 @@
 
 if (XML_SAX_IMPL == 'libxml') {
   $php = getenv('TEST_PHP_EXECUTABLE');
-  preg_match("/^libxml2 Version.*\$/im", `$php -i`, $match);
+  preg_match("/^libxml2 Version.*\$/im", `$php -i -n`, $match);
   echo $match[0], "\n";
 } else {
   echo "libxml2 Version => NONE\n";  
