diff -Naur nfs-utils-1.0.7/utils/statd/monitor.c nfs-utils-1.0.7.oden/utils/statd/monitor.c
--- nfs-utils-1.0.7/utils/statd/monitor.c	2004-09-15 05:08:01.000000000 +0200
+++ nfs-utils-1.0.7.oden/utils/statd/monitor.c	2005-01-09 10:16:22.775988738 +0100
@@ -15,6 +15,7 @@
 #include <string.h>
 #include <unistd.h>
 #include <sys/stat.h>
+#include <errno.h>
 #include <arpa/inet.h>
 #include "misc.h"
 #include "statd.h"
@@ -172,7 +173,7 @@
 	sprintf(path, "%s/%s", SM_DIR, mon_name);
 	if ((fd = open(path, O_WRONLY|O_SYNC|O_CREAT, S_IRUSR|S_IWUSR)) < 0) {
 		/* Didn't fly.  We won't monitor. */
-		note(N_ERROR, "creat(%s) failed: %m", path);
+		note(N_ERROR, "creat(%s) failed: %s", path, strerror (errno));
 		nlist_free(NULL, clnt);
 		free(path);
 		goto failure;
diff -Naur nfs-utils-1.0.7/utils/statd/statd.c nfs-utils-1.0.7.oden/utils/statd/statd.c
--- nfs-utils-1.0.7/utils/statd/statd.c	2004-12-06 02:25:27.000000000 +0100
+++ nfs-utils-1.0.7.oden/utils/statd/statd.c	2005-01-09 10:28:56.146983044 +0100
@@ -194,8 +194,10 @@
 	struct stat st;
 
 	if (stat(SM_DIR, &st) == -1 &&
-	    stat(DIR_BASE, &st) == -1)
+	    stat(DIR_BASE, &st) == -1) {
 		st.st_uid = 0;
+		st.st_gid = 0;
+	}
 
 	if (st.st_uid == 0) {
 		note(N_WARNING, "statd running as root. chown %s to choose different user\n",
diff -Naur nfs-utils-1.0.7/utils/statd/svc_run.c nfs-utils-1.0.7.oden/utils/statd/svc_run.c
--- nfs-utils-1.0.7/utils/statd/svc_run.c	2004-12-06 02:25:28.000000000 +0100
+++ nfs-utils-1.0.7.oden/utils/statd/svc_run.c	2005-01-09 10:17:06.706982708 +0100
@@ -123,7 +123,8 @@
 			if (errno == EINTR || errno == ECONNREFUSED
 			 || errno == ENETUNREACH || errno == EHOSTUNREACH)
 				continue;
-			note(N_ERROR, "my_svc_run() - select: %m");
+			note(N_ERROR, "my_svc_run() - select: %s",
+				strerror (errno));
 			return;
 
 		case 0:
