--- util-linux-2.12a/MCONFIG.rhconfig	2004-03-04 21:14:17.000000000 +0100
+++ util-linux-2.12a/MCONFIG	2004-07-19 18:18:29.321062317 +0200
@@ -16,21 +16,21 @@
 # If HAVE_PAM is set to "yes", then login, chfn, chsh, and newgrp
 # will use PAM for authentication. Additionally, passwd will not be
 # installed as it is not PAM aware.
-HAVE_PAM=no
+HAVE_PAM=yes
 
 # If HAVE_SELINUX is set to "yes", the login will make sure the user is
 # logged into an appropriate security context
-HAVE_SELINUX=no
+HAVE_SELINUX=yes
 
 # If HAVE_SHADOW is set to "yes", then login, chfn, chsh, newgrp, passwd,
 # and vipw will not be built or installed from the login-utils
 # subdirectory.  
-HAVE_SHADOW=yes
+HAVE_SHADOW=no
 
 # If HAVE_PASSWD is set to "yes", then passwd will not be built or
 # installed from the login-utils subdirectory (but login, chfn, chsh,
 # newgrp, and vipw *will* be installed).
-HAVE_PASSWD=no
+HAVE_PASSWD=yes
 
 # If you use chfn and chsh from this package, REQUIRE_PASSWORD will require
 # non-root users to enter the account password before updating /etc/passwd.
@@ -83,7 +83,7 @@
 # to the current user, allowing her to make a screendump and do other
 # nifty things on the console, but also allowing him to keep an open
 # file descriptor after logging out to trick the next user.
-ALLOW_VCS_USE=yes
+ALLOW_VCS_USE=no
 
 # If HAVE_RESET is set to "yes", then reset won't be installed.  The version
 # of reset that comes with the ncurses package is less aggressive.
@@ -106,6 +106,9 @@
 # DISABLE_NLS to "yes".
 DISABLE_NLS=no
 
+# If ADD_RAW is set to "yes", then raw gets built
+ADD_RAW=yes
+
 # Different optimizations for different cpus.
 # gcc 3.0 likes options -mcpu=i486 instead of -m486
 ifeq "$(ARCH)" "intel"
@@ -137,9 +140,9 @@
 
 # Set HAVE_SLANG to yes if you have slang (and prefer to use that for cfdisk)
 # (If neither HAVE_NCURSES nor HAVE_SLANG is defined, cfdisk is not made.)
-# HAVE_SLANG=yes
+HAVE_SLANG=yes
 # There is a subdirectory /usr/include/slang containing slcurses.h
-# SLANGFLAGS=-I/usr/include/slang
+SLANGFLAGS=-I/usr/include/slang
 # No such subdirectory - slcurses.h lives in /usr/include
 #   (no extra definition required).
 LIBSLANG=-lslang
@@ -202,7 +205,7 @@
 INSTALLBIN=	$(INSTALL) -m $(BINMODE)
 INSTALLMAN=	$(INSTALL) -m $(MANMODE)
 INSTALLDAT=	$(INSTALL) -m $(DATMODE)
-INSTALLSUID=    $(INSTALL) -m $(SUIDMODE) -o root
+INSTALLSUID=    $(INSTALL) -m $(SUIDMODE)
 
 ifeq "$(DISABLE_NLS)" "yes"
 NLSFLAGS = -DDISABLE_NLS
