diff -X ../../KDIFX -burpN -X ../../KDIFX linux-vanilla/fs/open.c linux/fs/open.c
--- linux-vanilla/fs/open.c	2002-11-30 00:37:12.000000000 +0100
+++ linux/fs/open.c	2002-11-30 09:59:56.000000000 +0100
@@ -785,6 +785,9 @@ asmlinkage long sys_open(const char * fi
 	char * tmp;
 	int fd, error;
 
+#ifdef __x86_64__ /* should really get an generic macro for that */ 
+	if (!(current->thread.flags & THREAD_IA32))
+#endif
 #if BITS_PER_LONG != 32
 	flags |= O_LARGEFILE;
 #endif
