--- linux-2.4.20/kernel/timer.c.no-generic-do_gettimeoffset_safe	2003-02-07 12:33:49.000000000 -0500
+++ linux-2.4.20/kernel/timer.c	2003-02-07 12:47:33.000000000 -0500
@@ -475,7 +475,7 @@ static void second_overflow(void)
 }
 
 #ifdef __x86_64__
-extern unsigned int (*do_gettimeoffset_safe)(void);
+extern unsigned int (*do_gettimeoffset)(void);
 #endif
 
 /* in the NTP reference this is called "hardclock()" */
@@ -484,7 +484,7 @@ static void update_wall_time_one_tick(vo
 #ifdef __x86_64__
 	long orig_usec;
 
-	orig_usec = xtime.tv_usec + do_gettimeoffset_safe();
+	orig_usec = xtime.tv_usec + do_gettimeoffset();
 #endif
 
 	if ( (time_adjust_step = time_adjust) != 0 ) {
@@ -522,7 +522,7 @@ static void update_wall_time_one_tick(vo
 		xtime.tv_usec += ltemp;
 	}
 #ifdef __x86_64__
-	while (xtime.tv_usec + do_gettimeoffset_safe() < orig_usec);
+	while (xtime.tv_usec + do_gettimeoffset() < orig_usec);
 #endif
 }
 
