diff -uNr words-2/usr/dict/linux.words words-2/usr/dict/linux.words
--- words-2/usr/dict/linux.words	Thu Apr 13 17:57:01 2000
+++ words-2/usr/dict/linux.words	Thu Apr 13 17:56:48 2000
@@ -1436,6 +1436,8 @@
 allot
 allotment
 allotments
+allotrope
+allotropic
 allots
 allotted
 allotter
@@ -2187,6 +2189,7 @@
 apartments
 apathetic
 apathy
+Apatosaurus
 ape
 aped
 aperiodic
@@ -2491,6 +2494,7 @@
 arcane
 arced
 arch
+Archaeopteryx
 archaic
 archaically
 archaicness
@@ -6628,6 +6632,7 @@
 burrows
 burrs
 bursa
+bursar
 bursitis
 burst
 burstiness
@@ -6818,6 +6823,9 @@
 cafe
 cafes
 cafeteria
+caffein
+caffeinated
+caffeine
 cage
 caged
 cager
@@ -7177,6 +7185,7 @@
 Carboloy
 carbon
 carbonate
+carbonated
 carbonates
 carbonation
 Carbondale
@@ -12650,6 +12659,7 @@
 dimensioning
 dimensions
 dimes
+dimetrodon
 diminish
 diminished
 diminishes
@@ -12698,6 +12708,7 @@
 dip
 diphtheria
 diphthong
+Diplodocus
 diploma
 diplomacy
 diplomas
@@ -17685,6 +17696,7 @@
 froze
 frozen
 frozenly
+fructose
 Fruehauf
 frugal
 frugally
@@ -17876,6 +17888,7 @@
 gaiters
 Gaithersburg
 galactic
+galactose
 Galahad
 Galapagos
 Galatea
@@ -18477,6 +18490,7 @@
 glowing
 glowingly
 glows
+glucose
 glue
 glued
 glues
@@ -23608,6 +23622,7 @@
 lacquered
 lacquers
 lacrosse
+lactose
 lacy
 lad
 ladder
@@ -26035,6 +26050,7 @@
 mesh
 meson
 Mesopotamia
+mesosphere
 Mesozoic
 mesquite
 mess
@@ -27084,6 +27100,7 @@
 muck
 mucker
 mucking
+mucilage
 mucus
 mud
 Mudd
@@ -32345,6 +32362,8 @@
 psychotherapist
 psychotherapy
 psychotic
+Pteranodon
+Pterodactyl
 Ptolemaic
 Ptolemaists
 Ptolemy
@@ -40043,6 +40062,7 @@
 suckle
 suckling
 sucks
+sucrose
 suction
 Sudan
 Sudanese
@@ -41369,6 +41389,7 @@
 Thermofax
 thermometer
 thermometers
+thermosphere
 thermostat
 thermostats
 these
@@ -43668,6 +43689,7 @@
 Vela
 Velasquez
 Vella
+Velociraptor
 velocities
 velocity
 velvet
