--- contrib/redhat/sshd.pam.annvix	2004-08-19 10:22:53.000000000 -0600
+++ contrib/redhat/sshd.pam	2004-08-19 10:22:53.000000000 -0600
@@ -1,4 +1,5 @@
 #%PAM-1.0
+auth       required     pam_listfile.so item=user sense=deny file=/etc/ssh/denyusers.pam
 auth       required     pam_stack.so service=system-auth
 auth       required     pam_nologin.so
 account    required     pam_stack.so service=system-auth
--- sshd_config.avx	2005-07-26 07:54:57.000000000 -0400
+++ sshd_config	2005-09-26 04:12:08.000000000 -0400
@@ -11,7 +11,7 @@
 # default value.
 
 #Port 22
-#Protocol 2,1
+Protocol 2
 #AddressFamily any
 #ListenAddress 0.0.0.0
 #ListenAddress ::
@@ -19,8 +19,8 @@
 # HostKey for protocol version 1
 #HostKey /etc/ssh/ssh_host_key
 # HostKeys for protocol version 2
-#HostKey /etc/ssh/ssh_host_rsa_key
-#HostKey /etc/ssh/ssh_host_dsa_key
+HostKey /etc/ssh/ssh_host_rsa_key
+HostKey /etc/ssh/ssh_host_dsa_key
 
 # Lifetime and size of ephemeral version 1 server key
 #KeyRegenerationInterval 1h
@@ -34,9 +34,9 @@
 # Authentication:
 
 #LoginGraceTime 2m
-#PermitRootLogin yes
+PermitRootLogin no
 #StrictModes yes
-#MaxAuthTries 6
+MaxAuthTries 4
 
 #RSAAuthentication yes
 #PubkeyAuthentication yes
@@ -79,7 +79,7 @@
 # ChallengeResponseAuthentication=no
 #UsePAM no
 
-#AllowTcpForwarding yes
+AllowTcpForwarding no
 #GatewayPorts no
 #X11Forwarding no
 #X11DisplayOffset 10
@@ -88,7 +88,7 @@
 #PrintLastLog yes
 #TCPKeepAlive yes
 #UseLogin no
-#UsePrivilegeSeparation yes
+UsePrivilegeSeparation yes
 #PermitUserEnvironment no
 #Compression delayed
 #ClientAliveInterval 0
--- ssh_config.avx	2005-02-08 17:46:48.000000000 -0500
+++ ssh_config	2005-09-26 04:13:13.000000000 -0400
@@ -37,3 +37,8 @@
 #   Cipher 3des
 #   Ciphers aes128-cbc,3des-cbc,blowfish-cbc,cast128-cbc,arcfour,aes192-cbc,aes256-cbc
 #   EscapeChar ~
+
+Host *
+  Protocol 2,1
+  StrictHostKeyChecking ask
+  HashKnownHosts yes
