--- itcl3.2/itcl/configure.in.libdir	2002-06-28 12:24:05.000000000 +0200
+++ itcl3.2/itcl/configure.in	2002-06-28 12:24:05.000000000 +0200
@@ -284,7 +284,7 @@ case "`uname -s`" in
 	;;
     *)
 	if test "${SHARED_BUILD}" = "1" ; then
-	    SHLIB_LD_LIBS="${TCL_STUB_LIB_SPEC}"
+	    SHLIB_LD_LIBS="${TCL_BUILD_STUB_LIB_SPEC}"
 	    eval "${PACKAGE}_LIB_FILE=lib${PACKAGE}${SHARED_LIB_SUFFIX}"
 	    RANLIB=:
 	else
--- itcl3.2/itcl/Makefile.in.libdir	2000-08-29 05:58:00.000000000 +0200
+++ itcl3.2/itcl/Makefile.in	2002-06-28 12:50:03.000000000 +0200
@@ -150,7 +150,7 @@ oldincludedir = /usr/include
 DESTDIR =
 
 pkgdatadir = $(datadir)/@PACKAGE@@VERSION@
-pkglibdir = $(libdir)/@PACKAGE@@VERSION@
+pkglibdir = $(prefix)/lib/@PACKAGE@@VERSION@
 pkgincludedir = $(includedir)/@PACKAGE@@VERSION@
 
 top_builddir = .
--- itcl3.2/itk/configure.in.libdir	2002-06-28 12:24:05.000000000 +0200
+++ itcl3.2/itk/configure.in	2002-06-28 12:24:05.000000000 +0200
@@ -323,7 +323,7 @@ case "`uname -s`" in
 	;;
     *)
 	if test "${SHARED_BUILD}" = "1" ; then
-	    SHLIB_LD_LIBS="${tkstub_LIB_SPEC} ${TCL_STUB_LIB_SPEC} ../itcl/${itclstub_LIB_FILE}"
+	    SHLIB_LD_LIBS="${tkstub_LIB_SPEC} ${TCL_BUILD_STUB_LIB_SPEC} ../itcl/${itclstub_LIB_FILE}"
 	    eval "${PACKAGE}_LIB_FILE=lib${PACKAGE}${SHARED_LIB_SUFFIX}"
 	    RANLIB=:
 	else
--- itcl3.2/itk/Makefile.in.libdir	2000-08-29 05:58:00.000000000 +0200
+++ itcl3.2/itk/Makefile.in	2002-06-28 12:50:40.000000000 +0200
@@ -138,7 +138,7 @@ oldincludedir = /usr/include
 DESTDIR =
 
 pkgdatadir = $(datadir)/@PACKAGE@@VERSION@
-pkglibdir = $(libdir)/@PACKAGE@@VERSION@
+pkglibdir = $(prefix)/lib/@PACKAGE@@VERSION@
 pkgincludedir = $(includedir)/@PACKAGE@@VERSION@
 
 top_builddir = .
--- itcl3.2/iwidgets2.2.0/configure.in.libdir	2000-08-29 05:58:00.000000000 +0200
+++ itcl3.2/iwidgets2.2.0/configure.in	2002-06-28 12:24:05.000000000 +0200
@@ -67,7 +67,7 @@ AC_ARG_WITH(itcl, [  --with-itcl=DIR    
         itcl_search=$withval, itcl_search=`cd ${srcdir}/../itcl; pwd`)
 
 ITCL_LIB_DIR=""
-for dir in $exec_prefix/lib $TCL_LIB_DIR $itcl_search ; do
+for dir in ${libdir} $TCL_LIB_DIR $itcl_search ; do
     if test -r $dir/library/itcl.tcl; then
         ITCL_LIB_DIR=$dir
 	ITCL_SRC_DIR=$dir
@@ -88,7 +88,7 @@ AC_ARG_WITH(itk, [  --with-itk=DIR      
         itk_search=$withval, itk_search=`cd ${srcdir}/../itk; pwd`)
 
 ITK_LIB_DIR=""
-for dir in $exec_prefix/lib $TCL_LIB_DIR $itk_search ; do
+for dir in ${libdir} $TCL_LIB_DIR $itk_search ; do
     if test -r $dir/library/itk.tcl; then
         ITK_LIB_DIR=$dir
         ITK_SRC_DIR=$dir
--- itcl3.2/iwidgets3.0.0/configure.in.libdir	2000-08-29 05:58:00.000000000 +0200
+++ itcl3.2/iwidgets3.0.0/configure.in	2002-06-28 12:24:05.000000000 +0200
@@ -67,7 +67,7 @@ AC_ARG_WITH(itcl, [  --with-itcl=DIR    
         itcl_search=$withval, itcl_search=`cd ${srcdir}/../itcl; pwd`)
 
 ITCL_LIB_DIR=""
-for dir in $exec_prefix/lib $TCL_LIB_DIR $itcl_search ; do
+for dir in ${libdir} $TCL_LIB_DIR $itcl_search ; do
     if test -r $dir/library/itcl.tcl; then
         ITCL_LIB_DIR=$dir
         ITCL_SRC_DIR=$dir
@@ -88,7 +88,7 @@ AC_ARG_WITH(itk, [  --with-itk=DIR      
         itk_search=$withval, itk_search=`cd ${srcdir}/../itk; pwd`)
 
 ITK_LIB_DIR=""
-for dir in $exec_prefix/lib $TCL_LIB_DIR $itk_search ; do
+for dir in ${libdir} $TCL_LIB_DIR $itk_search ; do
     if test -r $dir/library/itk.tcl; then
         ITK_LIB_DIR=$dir
         ITK_SRC_DIR=$dir
