--- tclx8.3.5/tcl/unix/Makefile.in.buildhelp	2000-08-10 04:21:48.000000000 +0200
+++ tclx8.3.5/tcl/unix/Makefile.in	2004-12-03 13:04:17.875340456 +0100
@@ -121,7 +121,7 @@
     stringfile.o     tcllib.o         fmath.o          buildhelp.o \
     buildidx.o       autoload.o       tclx.o
 
-HELP_DIR = @HELP_DIR@/tcl
+HELP_DIR = $(srcbasedir)/help/tcl
 
 TEST_OBJS = tclTest.o tclTestObj.o tclUnixTest.o tclXtest.o \
 	tclThreadTest.o tclXunixTest.o
@@ -311,14 +311,14 @@
 BUILDHELP=@BUILDHELP@
 buildhelp: tcl tcl.tndx
 	rm -rf ${HELP_DIR} help help.tmp
-	if test ! -d @HELP_DIR@ ; then \
-	    mkdir -p @HELP_DIR@ ; \
-	fi
+	#if test ! -d @HELP_DIR@ ; then \
+	#    mkdir -p @HELP_DIR@ ; \
+	#fi
 	mkdir -p ${HELP_DIR}
-	${BLDMANHELP} ${TCL_SRC_DIR}/doc ${TCLX_TOOLS_SRC_DIR}/tclmanpages \
+	#${BLDMANHELP} ${TCL_SRC_DIR}/doc ${TCLX_TOOLS_SRC_DIR}/tclmanpages \
 	    ${HELP_DIR} Tcl.brf
-	${RUNTCL} -c "buildhelp ${HELP_DIR} TclX.brf ${TCLX_DOC_DIR}/TclX.n"
-	${INSTCOPY} ${HELP_DIR} ${INSTALL_ROOT}${TCLX_INST_RUNTIME}/help
+	#${RUNTCL} -c "buildhelp ${HELP_DIR} TclX.brf ${TCLX_DOC_DIR}/TclX.n"
+	#${INSTCOPY} ${HELP_DIR} ${INSTALL_ROOT}${TCLX_INST_RUNTIME}/help
 
 #------------------------------------------------------------------------------
 
--- tclx8.3.5/tk/unix/Makefile.in.buildhelp	2000-08-10 04:21:48.000000000 +0200
+++ tclx8.3.5/tk/unix/Makefile.in	2004-12-03 13:04:37.270391960 +0100
@@ -72,7 +72,7 @@
 LIBTCLX = ../../tcl/unix/${TCLX_STLIB_NAME}
 TCL = ../../tcl/unix/tcl
 
-HELP_DIR = @HELP_DIR@/tk
+HELP_DIR = $(srcbasedir)/tk/help
 
 TEST_OBJS = tkTest.o tkSquare.o tclThreadTest.o tkXunixTest.o
 
@@ -203,13 +203,13 @@
 BUILDHELP=@BUILDHELP@
 buildhelp:
 	rm -rf ${HELP_DIR} help help.tmp
-	if test ! -d @HELP_DIR@ ; then \
-	    mkdir -p @HELP_DIR@ ;\
-	fi
+	#if test ! -d @HELP_DIR@ ; then \
+	#    mkdir -p @HELP_DIR@ ;\
+	#fi
 	mkdir -p ${HELP_DIR}
-	${BLDMANHELP} ${TK_SRC_DIR}/doc ${TCLX_TOOLS_SRC_DIR}/tkmanpages \
+	#${BLDMANHELP} ${TK_SRC_DIR}/doc ${TCLX_TOOLS_SRC_DIR}/tkmanpages \
 	    ${HELP_DIR} Tk.brf
-	${INSTCOPY} ${HELP_DIR} ${INSTALL_ROOT}${TKX_INST_RUNTIME}/help
+	#${INSTCOPY} ${HELP_DIR} ${INSTALL_ROOT}${TKX_INST_RUNTIME}/help
 
 #------------------------------------------------------------------------------
 
