--- mysql-4.0.11a-gamma/scripts/mysql_install_db.sh.max	2003-03-12 10:57:56.000000000 +0100
+++ mysql-4.0.11a-gamma/scripts/mysql_install_db.sh	2003-03-12 11:10:41.000000000 +0100
@@ -81,10 +81,10 @@
   execdir=@libexecdir@ 
 else
   bindir="$basedir/bin"
-if test -x "$basedir/libexec/mysqld"
+if test -x "$basedir/libexec/mysqld" -o -x "$basedir/libexec/mysqld-max"
 then
   execdir="$basedir/libexec"
-elif test -x "@libexecdir@/mysqld"
+elif test -x "@libexecdir@/mysqld" -o -x "@libexecdir@/mysqld-max" 
 then
   execdir="@libexecdir@"
 else
@@ -94,16 +94,24 @@
 
 mdata=$ldata/mysql
 
-if test ! -x $execdir/mysqld
+# Use the mysqld-max binary by default if the user doesn't specify a binary
+if test -x $execdir/mysqld-max; 
 then
-  if test "$IN_RPM" -eq 1
-  then
-    echo "FATAL ERROR $execdir/mysqld not found!"
-    exit 1
+  MYSQLD=mysqld-max
+else
+  if test -x $execdir/mysqld
+    then
+    MYSQLD=mysqld
   else
-    echo "Didn't find $execdir/mysqld"
-    echo "You should do a 'make install' before executing this script"
-    exit 1
+    if test "$IN_RPM" -eq 1
+    then
+      echo "FATAL ERROR $execdir/mysqld not found!"
+      exit 1
+    else
+      echo "Didn't find $execdir/mysqld"
+      echo "You should do a 'make install' before executing this script"
+      exit 1
+    fi
   fi
 fi
 
@@ -307,7 +315,7 @@
 fi
 
 echo "Installing all prepared tables"
-if eval "$execdir/mysqld $defaults --bootstrap --skip-grant-tables \
+if eval "$execdir/$MYSQLD $defaults --bootstrap --skip-grant-tables \
          --basedir=$basedir --datadir=$ldata --skip-innodb --skip-bdb $args" << END_OF_DATA
 use mysql;
 $c_d
@@ -369,7 +377,7 @@
   echo
   echo "Examine the logs in $ldata for more information."
   echo "You can also try to start the mysqld daemon with:"
-  echo "$execdir/mysqld --skip-grant &"
+  echo "$execdir/$MYSQLD --skip-grant &"
   echo "You can use the command line tool"
   echo "$bindir/mysql to connect to the mysql"
   echo "database and look at the grant tables:"
