--- scripts/mysql_install_db.sh.max	2004-09-06 16:29:38.000000000 -0600
+++ scripts/mysql_install_db.sh	2004-10-21 20:19:01.000000000 -0600
@@ -95,10 +95,10 @@
   pkgdatadir=@pkgdatadir@
 else
   bindir="$basedir/bin"
-if test -x "$basedir/libexec/mysqld"
+if test -x "$basedir/libexec/mysqld" -o -x "$basedir/libexec/mysqld-max"
 then
   execdir="$basedir/libexec"
-elif test -x "@libexecdir@/mysqld"
+elif test -x "@libexecdir@/mysqld" -o -x "@libexecdir@/mysqld-max" 
 then
   execdir="@libexecdir@"
 else
@@ -118,16 +118,28 @@
   scriptdir="./scripts"
 fi
 
-if test ! -x $mysqld
+# Use the mysqld-max binary by default if the user doesn't specify a binary
+if test -x $mysqld-max
 then
-  if test "$in_rpm" = 1
+  MYSQLD=mysqld-max
+else
+  if test -x $mysqld
   then
-    echo "FATAL ERROR $mysqld not found!"
-    exit 1
+    MYSQLD=mysqld
   else
     echo "Didn't find $mysqld"
     echo "You should do a 'make install' before executing this script"
     exit 1
+      
+    if test "$in_rpm" = 1
+    then
+      echo "FATAL ERROR $mysqld not found!"
+      exit 1
+    else
+      echo "Didn't find $mysqld"
+      echo "You should do a 'make install' before executing this script"
+      exit 1
+    fi
   fi
 fi
 
@@ -340,7 +352,7 @@
 fi
 
 echo "Installing all prepared tables"
-if eval "$mysqld $defaults $mysqld_opt --bootstrap --skip-grant-tables \
+if eval "$execdir/$MYSQLD $defaults $mysqld_opt --bootstrap --skip-grant-tables \
          --basedir=$basedir --datadir=$ldata --skip-innodb --skip-bdb \
          $args" << END_OF_DATA
 use mysql;
@@ -403,7 +415,7 @@
   echo
   echo "Examine the logs in $ldata for more information."
   echo "You can also try to start the mysqld daemon with:"
-  echo "$execdir/mysqld --skip-grant &"
+  echo "$execdir/$MYSQLD --skip-grant &"
   echo "You can use the command line tool"
   echo "$bindir/mysql to connect to the mysql"
   echo "database and look at the grant tables:"
