--- zlib-1.1.4/configure.multibuild	1998-07-08 14:19:35.000000000 -0400
+++ zlib-1.1.4/configure	2003-01-30 01:34:53.000000000 -0500
@@ -18,9 +18,10 @@
 # If you have problems, try without defining CC and CFLAGS before reporting
 # an error.
 
+top_srcdir=`dirname $0`
 LIBS=libz.a
 SHAREDLIB=libz.so
-VER=`sed -n -e '/VERSION "/s/.*"\(.*\)".*/\1/p' < zlib.h`
+VER=`sed -n -e '/VERSION "/s/.*"\(.*\)".*/\1/p' < $top_srcdir/zlib.h`
 AR=${AR-"ar rc"}
 RANLIB=${RANLIB-"ranlib"}
 prefix=${prefix-/usr/local}
@@ -72,8 +73,8 @@ if test "$gcc" -eq 1 && ($cc -c $cflags 
   SFLAGS=${CFLAGS-"-fPIC -O3"}
   CFLAGS="$cflags"
   case `(uname -s || echo unknown) 2>/dev/null` in
-  Linux | linux) LDSHARED=${LDSHARED-"gcc -shared -Wl,-soname,libz.so.1"};;
-  *)             LDSHARED=${LDSHARED-"gcc -shared"};;
+  Linux | linux) LDSHARED=${LDSHARED-"$CC -shared -Wl,-soname,libz.so.1"};;
+  *)             LDSHARED=${LDSHARED-"$CC -shared"};;
   esac
 else
   # find system name and corresponding cc options
@@ -195,7 +196,7 @@ esac
 rm -f $test.[co] $test$shared_ext
 
 # udpate Makefile
-sed < Makefile.in "
+sed < $top_srcdir/Makefile.in "
 /^CC *=/s%=.*%=$CC%
 /^CFLAGS *=/s%=.*%=$CFLAGS%
 /^CPP *=/s%=.*%=$CPP%
@@ -210,3 +211,10 @@ sed < Makefile.in "
 /^libdir *=/s%=.*%=$libdir%
 /^includedir *=/s%=.*%=$includedir%
 " > Makefile
+
+# symlink sources
+if test "$top_srcdir" != "."; then
+  for f in $top_srcdir/*.[ch]; do
+    ln -s $f `basename $f`
+  done
+fi
