--- mkbootdisk-1.5.1/mkbootdisk-mdk	2004-05-13 20:53:22.729145617 +0200
+++ mkbootdisk-1.5.1/mkbootdisk	2004-05-13 20:55:03.941933806 +0200
@@ -6,7 +6,8 @@
 
 pause=yes
 unset kernel
-device=/dev/fd0
+device=/dev/floppy/0 ; [ -e $device ] || device=/dev/fd0
+loop=/dev/loop/7 ; [ -e $device ] || device=/dev/loop7
 unset verbose
 unset kernelargs
 unset mountopts
@@ -95,7 +96,7 @@
 
 
 # yves from chmou fix loopback devices
-rootdev=`awk '$1 ~ /^[^#]/ && $2 == "/" { print $4 ~ /loop/ ? "/dev/loop7" : $1 }' /etc/fstab`
+rootdev=`awk '$1 ~ /^[^#]/ && $2 == "/" { print $4 ~ /loop/ ? ENVIRON["loop"] : $1 }' /etc/fstab`
 
 if [ $(echo $rootdev | cut -c1-6) = "LABEL=" ]; then
     rootlabel=$(echo $rootdev | cut -c7-)
