--- perl-5.8.1-RC4/Configure.pix	2003-08-01 14:24:02.000000000 +0200
+++ perl-5.8.1-RC4/Configure	2003-08-03 22:00:34.000000000 +0200
@@ -1218,6 +1218,13 @@
 : no include file wanted by default
 inclwanted=''
 
+: determine libdir name
+: Should be */lib64 on x86-64, sparc64, ppc64
+case `arch` in
+  x86_64) lib='lib64';;
+  *) lib='lib';;
+esac
+
 siteman1dir=''
 siteman3dir=''
 sitescript=''
@@ -1229,12 +1236,12 @@
 : change the next line if compiling for Xenix/286 on Xenix/386
 xlibpth='/usr/lib/386 /lib/386'
 : Possible local library directories to search.
-loclibpth="/usr/local/lib /opt/local/lib /usr/gnu/lib"
-loclibpth="$loclibpth /opt/gnu/lib /usr/GNU/lib /opt/GNU/lib"
+loclibpth="/usr/local/$lib /opt/local/$lib /usr/gnu/$lib"
+loclibpth="$loclibpth /opt/gnu/$lib /usr/GNU/$lib /opt/GNU/$lib"
 
 : general looking path for locating libraries
-glibpth="/lib /usr/lib $xlibpth"
-glibpth="$glibpth /usr/ccs/lib /usr/ucblib /usr/local/lib"
+glibpth="/$lib /usr/$lib $xlibpth"
+glibpth="$glibpth /usr/ccs/$lib /usr/ucblib /usr/local/$lib"
 test -f /usr/shlib/libc.so && glibpth="/usr/shlib $glibpth"
 test -f /shlib/libc.so     && glibpth="/shlib $glibpth"
 
@@ -10282,9 +10289,9 @@
 		else
 			for net in net socket
 			do
-				if test -f /usr/lib/lib$net$_a; then
-					( ($nm $nm_opt /usr/lib/lib$net$_a | eval $nm_extract) ||  \
-					$ar t /usr/lib/lib$net$_a) 2>/dev/null >> libc.list
+				if test -f /usr/$lib/lib$net$_a; then
+					( ($nm $nm_opt /usr/$lib/lib$net$_a | eval $nm_extract) ||  \
+					$ar t /usr/$lib/lib$net$_a) 2>/dev/null >> libc.list
 					if $contains socket libc.list >/dev/null 2>&1; then
 						d_socket="$define"
 						socketlib="-l$net"
@@ -21104,6 +21111,7 @@
 ldflags_uselargefiles='$ldflags_uselargefiles'
 ldlibpthname='$ldlibpthname'
 less='$less'
+lib='$lib'
 lib_ext='$lib_ext'
 libc='$libc'
 libperl='$libperl'
