--- man-1.5j/src/man.conf.in.mandirs	Tue Nov 20 00:37:50 2001
+++ man-1.5j/src/man.conf.in	Thu Nov 22 14:46:51 2001
@@ -36,10 +36,11 @@
 #
 # Every automatically generated MANPATH includes these fields
 #
-MANPATH	/usr/man
 MANPATH /usr/share/man
-MANPATH	/usr/local/man
 MANPATH	/usr/X11R6/man
+MANPATH	/usr/local/man
+MANPATH /usr/kerberos/man
+MANPATH	/usr/man
 #
 # Uncomment if you want to include one of these by default
 #
@@ -70,6 +71,8 @@
 MANPATH_MAP	/usr/X11R6/bin		/usr/X11R6/man
 MANPATH_MAP	/usr/bin/X11		/usr/X11R6/man
 MANPATH_MAP	/usr/bin/mh		/usr/share/man
+MANPATH_MAP	/usr/kerberos/bin	/usr/kerberos/man
+MANPATH_MAP	/usr/kerberos/sbin	/usr/kerberos/man
 #
 # NOAUTOPATH keeps man from automatically adding directories that look like
 # manual page directories to the path.
