--- ./src/man.c.security	2003-09-05 11:49:13.000000000 +0200
+++ ./src/man.c	2003-12-11 18:56:49.000000000 +0100
@@ -1217,6 +1217,20 @@
 #endif
 
 
+     char *s;
+#define CHECK(p, l) s=getenv(p); if(s && (strlen(s)>l)) { fprintf(stderr, "ERROR: Environment variable %s too long!\n", p); exit(1); }
+
+     CHECK("LANG", 32);
+     CHECK("MANPAGER", 128);
+     CHECK("PAGER", 128);
+     CHECK("SYSTEM", 64);
+     CHECK("MANROFFSEQ", 128);
+     CHECK("MANSECT", 128);
+     CHECK("MANPL", 128);
+     CHECK("MAN_HP_DIREXT", 128);
+     CHECK("LANGUAGE", 128);
+     CHECK("LC_MESSAGES", 128);
+
 #ifndef __FreeBSD__ 
      /* Slaven Rezif: FreeBSD-2.2-SNAP does not recognize LC_MESSAGES. */
      setlocale(LC_CTYPE, "");	/* used anywhere? maybe only isdigit()? */
--- ./src/makewhatis.sh.security	2003-09-05 11:48:13.000000000 +0200
+++ ./src/makewhatis.sh	2003-12-11 19:15:42.000000000 +0100
@@ -41,7 +41,7 @@
 # and should be first.
 # It is a bug to add /var/cache/man to DEFCATPATH.
 dm=
-for d in /usr/man /usr/share/man
+for d in /usr/share/man /usr/man
 do
     if [ -d $d ]; then
 	if [ x$dm = x ]; then dm=$d; else dm=$dm:$d; fi
@@ -49,7 +49,7 @@
 done
 DEFMANPATH=$dm
 dc=
-for d in /usr/man/preformat /usr/man /usr/share/man/preformat /usr/share/man
+for d in /var/cache/man /usr/share/man/preformat /usr/man/preformat /usr/share/man /usr/man
 do
     if [ -d $d ]; then
 	if [ x$dc = x ]; then dc=$d; else dc=$dc:$d; fi
@@ -72,12 +72,12 @@
 # We try here to be careful (and avoid preconstructed symlinks)
 # in case makewhatis is run as root, by creating a subdirectory of /tmp.
 
-TMPFILEDIR=/tmp/whatis.tmp.dir.$$
-rm -rf TMPFILEDIR
-if ! mkdir -m 0700 $TMPFILEDIR; then
-    echo Could not create $TMPFILEDIR
-    exit 1;
+TMPFILEDIR=`mktemp -d /tmp/makewhatisXXXXXX`
+if [ $? -ne 0 ]; then
+	echo "$0: Can't create temp file, exiting..."
+	exit 1
 fi
+chmod 0700 $TMPFILEDIR
 TMPFILE=$TMPFILEDIR/w
 
 # make sure TMPFILEDIR is deleted if program is killed or terminates
