--- autofs-4.0.0pre10/include/linux/auto_fs.h.64bit-fixes	2001-03-28 07:08:23.000000000 +0200
+++ autofs-4.0.0pre10/include/linux/auto_fs.h	2003-11-13 17:35:14.000000000 +0100
@@ -45,7 +45,7 @@
  * If so, 32-bit user-space code should be backwards compatible.
  */
 
-#if defined(__sparc__) || defined(__mips__)
+#if !defined(__alpha__) && !defined(__ia64__)   
 typedef unsigned int autofs_wqt_t;
 #else
 typedef unsigned long autofs_wqt_t;
--- autofs-4.0.0pre10/modules/mount_nfs.c.64bit-fixes	2003-11-13 17:29:49.000000000 +0100
+++ autofs-4.0.0pre10/modules/mount_nfs.c	2003-11-13 17:29:49.000000000 +0100
@@ -73,7 +73,7 @@ int mount_mount(const char *root, const 
   struct sockaddr_in saddr, laddr;
   int sock, local, err;
   int nosymlink = 0;
-  size_t len;
+  socklen_t len;
 
   syslog(LOG_DEBUG, MODPREFIX " root=%s name=%s what=%s, fstype=%s, options=%s",
 	 root, name, what, fstype, options);
