--- shadow-4.0.3/libmisc/copydir.c.preserve	2001-06-28 20:27:29.000000000 +0200
+++ shadow-4.0.3/libmisc/copydir.c	2002-11-19 00:02:26.000000000 +0100
@@ -125,7 +125,8 @@
  */
 
 int
-copy_tree(const char *src_root, const char *dst_root, uid_t uid, gid_t gid)
+copy_tree(const char *src_root, const char *dst_root, uid_t uid, gid_t gid,
+	  int preserve)
 {
 	char	src_name[1024];
 	char	dst_name[1024];
@@ -135,6 +136,7 @@
 	int	err = 0;
 	int	cnt;
 	int	set_orig = 0;
+	int	open_flags;
 	struct	DIRECT	*ent;
 	struct	stat	sb;
 	struct	link_name *lp;
@@ -206,7 +208,7 @@
 			chown (dst_name, uid == (uid_t) -1 ? sb.st_uid:uid,
 				gid == (gid_t) -1 ? sb.st_gid:gid);
 
-			if (copy_tree (src_name, dst_name, uid, gid)) {
+			if (copy_tree (src_name, dst_name, uid, gid, preserve)) {
 				err++;
 				break;
 			}
@@ -293,7 +295,11 @@
 			err++;
 			break;
 		}
-		if ((ofd = open (dst_name, O_WRONLY|O_CREAT|O_TRUNC, 0)) < 0 ||
+		open_flags = O_WRONLY | O_CREAT | O_TRUNC;
+ 		if (preserve) {
+ 			open_flags |= O_EXCL;
+ 		}
+ 		if ((ofd = open (dst_name, open_flags, 0)) < 0 ||
 			chown (dst_name, uid == (uid_t) -1 ? sb.st_uid:uid,
 					gid == (gid_t) -1 ? sb.st_gid:gid) ||
 				chmod (dst_name, sb.st_mode & 07777)) {
--- shadow-4.0.3/lib/prototypes.h.preserve	2000-08-26 20:27:17.000000000 +0200
+++ shadow-4.0.3/lib/prototypes.h	2002-11-19 00:02:26.000000000 +0100
@@ -55,7 +55,7 @@
 extern int is_listed(const char *, const char *, int);
 
 /* copydir.c */
-extern int copy_tree(const char *, const char *, uid_t, gid_t);
+extern int copy_tree(const char *, const char *, uid_t, gid_t, int);
 extern int remove_tree(const char *);
 
 /* encrypt.c */
--- shadow-4.0.3/lib/defines.h.preserve	2001-11-19 10:23:52.000000000 +0100
+++ shadow-4.0.3/lib/defines.h	2002-11-19 00:02:26.000000000 +0100
@@ -47,6 +47,10 @@
 # include <errno.h>
 #endif
 
+#if HAVE_ERRNO_H
+# include <errno.h>
+#endif
+
 /* Solaris 2.4 defines __SVR4, but not SVR4 -j. */
 
 #ifdef __SVR4
--- shadow-4.0.3/src/usermod.c.preserve	2002-11-19 00:02:26.000000000 +0100
+++ shadow-4.0.3/src/usermod.c	2002-11-19 00:02:26.000000000 +0100
@@ -1543,7 +1543,7 @@
 				}
 				if (copy_tree (user_home, user_newhome,
 					       uflg ? user_newid : -1,
-					       gflg ? user_newgid : -1) ==
+					       gflg ? user_newgid : -1, 0) ==
 				    0 && remove_tree (user_home) == 0
 				    && rmdir (user_home) == 0)
 					return;
--- shadow-4.0.3/src/useradd.c.preserve	2002-11-19 00:02:26.000000000 +0100
+++ shadow-4.0.3/src/useradd.c	2002-11-19 00:03:27.000000000 +0100
@@ -2078,7 +2078,7 @@
 
 	if (mflg) {
 		create_home ();
-		copy_tree (def_template, user_home, user_id, user_gid);
+		copy_tree (def_template, user_home, user_id, user_gid,1);
 	} /* Stupid warning removed for the innocent's protection */
  	    /*
  	     * The whole idea about breaking some stupid scripts by creating a new
