--- shadow-4.0.3/libmisc/nscd.c.nscd	2001-11-06 17:13:20.000000000 +0100
+++ shadow-4.0.3/libmisc/nscd.c	2002-11-19 00:00:12.000000000 +0100
@@ -1,3 +1,30 @@
+/* club nscd so it invalidates its cache */
+/* notting@redhat.com 19990408           */
+
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/types.h>
+
+
+int hup_nscd() {
+    FILE *pidfile;
+    char *pidbuf=calloc(256,sizeof(char));
+    int pid;
+    
+    if ((pidfile=fopen("/var/run/nscd.pid","r"))!=NULL) {
+	pidbuf=fgets(pidbuf,256,pidfile);
+	fclose(pidfile);
+	if (pidbuf!=NULL) {
+	    pid=atoi(pidbuf);
+	    free(pidbuf);
+	    if (pid>0) {
+		return kill(pid,SIGHUP);
+	    }
+	}
+    }
+    return -1;
+}
 /* Copyright (c) 1999 SuSE GmbH Nuerenberg, Germany
    Author: Thorsten Kukuk <kukuk@suse.de> */
 
--- shadow-4.0.3/src/groupmod.c.nscd	2002-01-05 16:41:43.000000000 +0100
+++ shadow-4.0.3/src/groupmod.c	2002-11-19 00:00:12.000000000 +0100
@@ -592,6 +592,8 @@
 
 	close_files ();
 
+        hup_nscd ();
+
 #ifdef USE_PAM
 	if (retval == PAM_SUCCESS) {
 		retval = pam_chauthtok (pamh, 0);
--- shadow-4.0.3/src/groupdel.c.nscd	2000-10-09 21:02:20.000000000 +0200
+++ shadow-4.0.3/src/groupdel.c	2002-11-19 00:00:12.000000000 +0100
@@ -396,6 +396,8 @@
 	grp_update ();
 
 	close_files ();
+        
+	hup_nscd ();
 
 #ifdef USE_PAM
 	if (retval == PAM_SUCCESS) {
--- shadow-4.0.3/src/userdel.c.nscd	2002-11-18 23:58:02.000000000 +0100
+++ shadow-4.0.3/src/userdel.c	2002-11-19 00:00:12.000000000 +0100
@@ -943,6 +943,8 @@
 			 */
 
 			user_cancel (user_name); close_files ();
+
+			hup_nscd ();
 #ifdef USE_PAM
 			if (retval == PAM_SUCCESS) {
 			retval = pam_chauthtok (pamh, 0);
--- shadow-4.0.3/src/gpasswd.c.nscd	2002-01-05 16:41:43.000000000 +0100
+++ shadow-4.0.3/src/gpasswd.c	2002-11-19 00:00:12.000000000 +0100
@@ -648,5 +648,6 @@
 	endsgent ();
 #endif
 #endif
+        hup_nscd ();
 	exit (0);
  /*NOTREACHED*/}
--- shadow-4.0.3/src/newusers.c.nscd	2002-01-05 16:41:43.000000000 +0100
+++ shadow-4.0.3/src/newusers.c	2002-11-19 00:00:12.000000000 +0100
@@ -615,6 +615,7 @@
 		spw_unlock ();
 #endif
 	(void) pw_unlock ();
+        hup_nscd ();
 
 #ifdef USE_PAM
 	if (retval == PAM_SUCCESS) {
--- shadow-4.0.3/src/useradd.c.nscd	2002-11-18 23:58:37.000000000 +0100
+++ shadow-4.0.3/src/useradd.c	2002-11-19 00:01:58.000000000 +0100
@@ -2072,6 +2072,7 @@
  	     */
   
 	close_files ();
+        hup_nscd ();
 
 #ifdef USE_PAM
 	if (retval == PAM_SUCCESS) {
--- shadow-4.0.3/src/usermod.c.nscd	2002-01-05 16:41:44.000000000 +0100
+++ shadow-4.0.3/src/usermod.c	2002-11-19 00:00:12.000000000 +0100
@@ -1708,6 +1708,7 @@
 	if (pampw == NULL) {
 		retval = PAM_USER_UNKNOWN;
 	}
+        hup_nscd();
 
 	if (retval == PAM_SUCCESS) {
 		retval =
--- shadow-4.0.3/src/groupadd.c.nscd	2002-11-18 23:58:02.000000000 +0100
+++ shadow-4.0.3/src/groupadd.c	2002-11-19 00:00:12.000000000 +0100
@@ -593,6 +593,8 @@
 	grp_update ();
 
 	close_files ();
+        
+	hup_nscd ();
 
 #ifdef USE_PAM
 	if (retval == PAM_SUCCESS) {
