diff -ruN linux-2.4.19/drivers/scsi.orig/scsi.c linux-2.4.19/drivers/scsi/scsi.c
--- linux-2.4.19/drivers/scsi.orig/scsi.c	2002-09-07 22:17:55.000000000 +0200
+++ linux-2.4.19/drivers/scsi/scsi.c	2002-09-07 22:26:47.000000000 +0200
@@ -2424,10 +2424,16 @@
 
 		/* Load upper level device handler of some kind */
 	case MODULE_SCSI_DEV:
+                /* Don't request scsi_hostadapter module when scsi_mod is  */
+                /* a module because that claimed an error at initrd        */
+                /* That is obselete when it is compiled as module because  */
+                /* scsi_mod is loaded before hostadapter via dependence    */
+#ifndef MODULE
 #ifdef CONFIG_KMOD
 		if (scsi_hosts == NULL)
 			request_module("scsi_hostadapter");
 #endif
+#endif
 		return scsi_register_device_module((struct Scsi_Device_Template *) ptr);
 		/* The rest of these are not yet implemented */
 
