--- linux-2.4.28/drivers/usb/serial/usbserial.c.du01.orig	2004-11-17 04:54:21.000000000 -0700
+++ linux-2.4.28/drivers/usb/serial/usbserial.c	2004-11-25 12:01:32.000000000 -0700
@@ -664,7 +664,7 @@
 	dbg("%s - port %d", __FUNCTION__, port->number);
 
 	/* if disconnect beat us to the punch here, there's nothing to do */
-	if (tty->driver_data) {
+	if (tty && tty->driver_data) {
 		/*
 		 * XXX The right thing would be to wait for the output to drain.
 		 * But we are not sufficiently daring to experiment in 2.4.
@@ -679,9 +679,10 @@
 		down (&port->sem);
 		if (!tty->driver_data) /* woopsie, disconnect, now what */ ;
 #endif
+		tty->driver_data = NULL;
 		__serial_close(port, filp);
 	}
-
+	port->tty = NULL;
 	up (&port->sem);
 }
 
