--- linux-2.4.21/drivers/video/matrox/i2c-matroxfb.c.dv20.orig	2003-06-13 16:51:37.000000000 +0200
+++ linux-2.4.21/drivers/video/matrox/i2c-matroxfb.c	2003-08-24 01:21:54.000000000 +0200
@@ -87,29 +87,21 @@ static int matroxfb_gpio_getscl(void* da
 	return (matroxfb_read_gpio(b->minfo) & b->mask.clock) ? 1 : 0;
 }
 
-static void matroxfb_dh_inc_use(struct i2c_adapter* dummy) {
-	MOD_INC_USE_COUNT;
-}
-
-static void matroxfb_dh_dec_use(struct i2c_adapter* dummy) {
-	MOD_DEC_USE_COUNT;
-}
-
 static struct i2c_adapter matrox_i2c_adapter_template =
 {
+	.owner	=	THIS_MODULE,
 	.id =		I2C_HW_B_G400,
-	.inc_use =	matroxfb_dh_inc_use,
-	.dec_use =	matroxfb_dh_dec_use,
 };
 
 static struct i2c_algo_bit_data matrox_i2c_algo_template =
 {
-	NULL,
-	matroxfb_gpio_setsda,
-	matroxfb_gpio_setscl,
-	matroxfb_gpio_getsda,
-	matroxfb_gpio_getscl,
-	10, 10, 100,
+	.setsda	= matroxfb_gpio_setsda,
+	.setscl	= matroxfb_gpio_setscl,
+	.getsda	= matroxfb_gpio_getsda,
+	.getscl	= matroxfb_gpio_getscl,
+	.udelay	= 10,
+	.mdelay	= 10,
+	.timeout	= 100,
 };
 
 static int i2c_bus_reg(struct i2c_bit_adapter* b, struct matrox_fb_info* minfo, 
--- linux-2.4.21/drivers/video/matrox/matroxfb_maven.c.dv20.orig	2003-06-13 16:51:37.000000000 +0200
+++ linux-2.4.21/drivers/video/matrox/matroxfb_maven.c	2003-08-24 01:21:54.000000000 +0200
@@ -1246,14 +1246,6 @@ static unsigned short normal_i2c[] = { M
 static unsigned short normal_i2c_range[] = { MAVEN_I2CID, MAVEN_I2CID, I2C_CLIENT_END };
 I2C_CLIENT_INSMOD;
 
-static void maven_inc_use(struct i2c_client* clnt) {
-	MOD_INC_USE_COUNT;
-}
-
-static void maven_dec_use(struct i2c_client* clnt) {
-	MOD_DEC_USE_COUNT;
-}
-
 static struct i2c_driver maven_driver;
 
 static int maven_detect_client(struct i2c_adapter* adapter, int address, unsigned short flags,
@@ -1320,14 +1312,13 @@ static int maven_command(struct i2c_clie
 static int maven_driver_registered = 0;
 
 static struct i2c_driver maven_driver={
-	"maven",
-	I2C_DRIVERID_MGATVO,
-	I2C_DF_NOTIFY,
-	maven_attach_adapter,
-	maven_detach_client,
-	maven_command,
-	maven_inc_use,
-	maven_dec_use
+	.owner	= THIS_MODULE,
+	.name	= "maven",
+	.id	= I2C_DRIVERID_MGATVO,
+	.flags	= I2C_DF_NOTIFY,
+	.attach_adapter	= maven_attach_adapter,
+	.detach_client	= maven_detach_client,
+	.command	= maven_command,
 };
 
 /* ************************** */
--- linux-2.4.21/drivers/media/video/adv7175.c.dv20.orig	2001-09-30 21:26:06.000000000 +0200
+++ linux-2.4.21/drivers/media/video/adv7175.c	2003-08-24 01:21:54.000000000 +0200
@@ -439,23 +439,25 @@ static int adv7175_command(struct i2c_de
 /* ----------------------------------------------------------------------- */
 
 static struct i2c_driver i2c_driver_adv7175 = {
-	"adv7175",		/* name */
-	I2C_DRIVERID_VIDEOENCODER,	/* ID */
-	I2C_ADV7175, I2C_ADV7175 + 3,
-
-	adv7175_attach,
-	adv7175_detach,
-	adv7175_command
+	.name	= "adv7175",
+	.id	= I2C_DRIVERID_VIDEOENCODER,
+	.addr_l	= I2C_ADV7175,
+	.addr_h = I2C_ADV7175 + 3,
+
+	.attach	= adv7175_attach,
+	.detach	= adv7175_detach,
+	.command	= adv7175_command
 };
 
 static struct i2c_driver i2c_driver_adv7176 = {
-	"adv7175",		/* name */
-	I2C_DRIVERID_VIDEOENCODER,	/* ID */
-	I2C_ADV7176, I2C_ADV7176 + 3,
-
-	adv7175_attach,
-	adv7175_detach,
-	adv7175_command
+	.name	= "adv7175",
+	.id	= I2C_DRIVERID_VIDEOENCODER,
+	.addr_l	= I2C_ADV7176,
+	.addr_h	= I2C_ADV7176 + 3,
+
+	.attach	= adv7175_attach,
+	.detach	= adv7175_detach,
+	.command	= adv7175_command
 };
 
 EXPORT_NO_SYMBOLS;
--- linux-2.4.21/drivers/media/video/bt819.c.dv20.orig	2001-09-30 21:26:06.000000000 +0200
+++ linux-2.4.21/drivers/media/video/bt819.c	2003-08-24 01:21:54.000000000 +0200
@@ -471,13 +471,14 @@ static int bt819_command(struct i2c_devi
 /* ----------------------------------------------------------------------- */
 
 static struct i2c_driver i2c_driver_bt819 = {
-	"bt819",		/* name */
-	I2C_DRIVERID_VIDEODECODER,	/* ID */
-	I2C_BT819, I2C_BT819 + 1,
-
-	bt819_attach,
-	bt819_detach,
-	bt819_command
+	.name	= "bt819",		/* name */
+	.id	= I2C_DRIVERID_VIDEODECODER,	/* ID */
+	.addr_l	= I2C_BT819,
+	.addr_h	= I2C_BT819 + 1,
+
+	.attach	= bt819_attach,
+	.detach	= bt819_detach,
+	.command	= bt819_command
 };
 
 EXPORT_NO_SYMBOLS;
--- linux-2.4.21/drivers/media/video/bt856.c.dv20.orig	2001-09-30 21:26:06.000000000 +0200
+++ linux-2.4.21/drivers/media/video/bt856.c	2003-08-24 01:21:54.000000000 +0200
@@ -279,12 +279,13 @@ static int bt856_command(struct i2c_devi
 /* ----------------------------------------------------------------------- */
 
 static struct i2c_driver i2c_driver_bt856 = {
-	"bt856",		/* name */
-	I2C_DRIVERID_VIDEOENCODER,	/* ID */
-	I2C_BT856, I2C_BT856 + 1,
-	bt856_attach,
-	bt856_detach,
-	bt856_command
+	.name	= "bt856",		/* name */
+	.id	= I2C_DRIVERID_VIDEOENCODER,	/* ID */
+	.addr_l	= I2C_BT856,
+	.addr_h	= I2C_BT856 + 1,
+	.attach = bt856_attach,
+	.detach	= bt856_detach,
+	.command	= bt856_command
 };
 
 EXPORT_NO_SYMBOLS;
--- linux-2.4.21/drivers/media/video/saa5249.c.dv20.orig	2001-09-30 21:26:06.000000000 +0200
+++ linux-2.4.21/drivers/media/video/saa5249.c	2003-08-24 01:21:54.000000000 +0200
@@ -258,21 +258,19 @@ static int saa5249_command(struct i2c_cl
 
 static struct i2c_driver i2c_driver_videotext = 
 {
-	IF_NAME,		/* name */
-	I2C_DRIVERID_SAA5249, /* in i2c.h */
-	I2C_DF_NOTIFY,
-	saa5249_probe,
-	saa5249_detach,
-	saa5249_command
+	.owner	= THIS_MODULE,
+	.name   = IF_NAME,		/* name */
+	.id	= I2C_DRIVERID_SAA5249, /* in i2c.h */
+	.flags	= I2C_DF_NOTIFY,
+	.attach_adapter = saa5249_probe,
+	.detach_client	= saa5249_detach,
+	.command	= saa5249_command
 };
 
 static struct i2c_client client_template = {
-	"(unset)",
-	-1,
-	0,
-	0,
-	NULL,
-	&i2c_driver_videotext
+	.name	= "(unset)",
+	.id	= -1,
+	.driver	= &i2c_driver_videotext,
 };
 
 /*
diff -uNp q1/drivers/media/video/saa7110.c.dv22.orig q1/drivers/media/video/saa7110.c
--- q1/drivers/media/video/saa7110.c.dv22.orig	2003-12-26 10:52:11.000000000 +0100
+++ q1/drivers/media/video/saa7110.c	2003-12-29 14:50:31.000000000 +0100
@@ -402,14 +402,15 @@ int saa7110_command(struct i2c_device *d
 
 static struct i2c_driver i2c_driver_saa7110 =
 {
-	"saa7110",			/* name */
+	.name	= "saa7110",
 
-	I2C_DRIVERID_VIDEODECODER,	/* in i2c-old.h */
-	I2C_SAA7110, I2C_SAA7110+1,	/* Addr range */
-
-	saa7110_attach,
-	saa7110_detach,
-	saa7110_command
+	.id	= I2C_DRIVERID_VIDEODECODER,
+	.addr_l	= I2C_SAA7110,
+	.addr_h	= I2C_SAA7110+1,
+
+	.attach	= saa7110_attach,
+	.detach	= saa7110_detach,
+	.command	= saa7110_command
 };
 
 EXPORT_NO_SYMBOLS;
--- linux-2.4.21/drivers/media/video/saa7185.c.dv20.orig	2001-09-30 21:26:06.000000000 +0200
+++ linux-2.4.21/drivers/media/video/saa7185.c	2003-08-24 01:21:54.000000000 +0200
@@ -383,13 +383,14 @@ static int saa7185_command(struct i2c_de
 /* ----------------------------------------------------------------------- */
 
 static struct i2c_driver i2c_driver_saa7185 = {
-	"saa7185",		/* name */
-	I2C_DRIVERID_VIDEOENCODER,	/* ID */
-	I2C_SAA7185, I2C_SAA7185 + 1,
-
-	saa7185_attach,
-	saa7185_detach,
-	saa7185_command
+	.name	= "saa7185",
+	.id	= I2C_DRIVERID_VIDEOENCODER,
+	.addr_l	= I2C_SAA7185,
+	.addr_h	= I2C_SAA7185 + 1,
+
+	.attach	= saa7185_attach,
+	.detach	= saa7185_detach,
+	.command	= saa7185_command
 };
 
 EXPORT_NO_SYMBOLS;
--- linux-2.4.21/drivers/media/video/tuner-3036.c.dv20.orig	2001-09-30 21:26:06.000000000 +0200
+++ linux-2.4.21/drivers/media/video/tuner-3036.c	2003-08-24 01:21:54.000000000 +0200
@@ -185,22 +185,20 @@ tuner_probe(struct i2c_adapter *adap)
 static struct i2c_driver 
 i2c_driver_tuner = 
 {
-	"sab3036",		/* name       */
-	I2C_DRIVERID_SAB3036,	/* ID         */
-        I2C_DF_NOTIFY,
-	tuner_probe,
-	tuner_detach,
-	tuner_command
+	.owner	= THIS_MODULE,
+	.name	= "sab3036",		/* name       */
+	.id	= I2C_DRIVERID_SAB3036,	/* ID         */
+        .flags	= I2C_DF_NOTIFY,
+	.attach_adapter	= tuner_probe,
+	.detach_client	= tuner_detach,
+	.command	= tuner_command
 };
 
 static struct i2c_client client_template =
 {
-        "SAB3036",		/* name       */
-        -1,
-        0,
-        0,
-        NULL,
-        &i2c_driver_tuner
+	.name	= "SAB3036",
+        .id	= -1,
+        .driver	= &i2c_driver_tuner
 };
 
 EXPORT_NO_SYMBOLS;
