--- linux-2.4.21/3rdparty/pdc-ultra/pdc618_mod.c.pdc20319	2003-08-11 17:07:53.000000000 +0200
+++ linux-2.4.21/3rdparty/pdc-ultra/pdc618_mod.c	2003-08-11 18:17:40.000000000 +0200
@@ -469,8 +469,8 @@ int pdc618_findcards(void) 
 			}
 #endif
 
-			/* class code is 0x0180 or PDC378*/
-			if(pdc618_pci_dev->class>>8 == PCI_CLASS_STORAGE_OTHER || PDC618_DEVID[class] == DID378 || PDC618_DEVID[class] == DID376) {
+			/* class code is 0x0180 or 0x104 or PDC378*/
+			if((pdc618_pci_dev->class>>8) == PCI_CLASS_STORAGE_OTHER || (pdc618_pci_dev->class>>8) == PCI_CLASS_STORAGE_RAID || PDC618_DEVID[class] == DID378) {
 				pdc618_adapter_t  *pada;
 
 				/* pdc376 class:0x0104 */
@@ -1206,6 +1206,7 @@ int pdc618_proc_info(char *buffer, char 
 			pdc618_copy_info(&info, "Adapter%d - ", i+1);
 			switch (id) {
 				case DID318:
+				case DID319:
 					pdc618_copy_info(&info, "SATA150 TX4  ");
 					break;
 				case DID375:
@@ -2189,6 +2190,9 @@ static int pdc618dev_ioctl(struct inode 
 				case DID318:
 					buf->adaptertype = SATA318;
 					break;
+				case DID319:
+					buf->adaptertype = SATA319;
+					break;
 				case DID378:
 					buf->adaptertype = SATA378;
 					break;
--- linux-2.4.21/3rdparty/pdc-ultra/pdc618_mod.h.pdc20319	2003-08-11 17:07:53.000000000 +0200
+++ linux-2.4.21/3rdparty/pdc-ultra/pdc618_mod.h	2003-08-11 18:12:23.000000000 +0200
@@ -24,6 +24,7 @@
 #define MAXSGTABLE  32
 
 #define DID318      0x3318
+#define DID319      0x3319
 #define DID375      0x3375
 #define DID376      0x3376
 #define DID378      0x3373
@@ -49,7 +50,7 @@
 #define STIO_READ_FLASHMEM	0x00004006
 #define STIO_WRITE_FLASHMEM	0x00004007
 
-unsigned short PDC618_DEVID[6] = { DID618, DID318, DID375, DID376, DID378, 0};
+unsigned short PDC618_DEVID[] = { DID618, DID318, DID319, DID375, DID376, DID378, 0};
 
 typedef struct _pdc618_adapter pdc618_adapter_t;
 typedef struct _pdc618_channel pdc618_channel_t;
@@ -124,6 +125,7 @@ typedef enum _hardwaretype {
 	SATA375,			/* SATA150 TX2plus */
 	SATA376,			/* SATA150 ??? */
 	SATA318,			/* SATA150 TX4 */
+	SATA319,			/* SATA150 TX4 */
 	SATA378				/* SATA 378 */
 } hardwaretype;
 
