--- contrib/codepage.c.old	2003-07-08 02:41:31.000000000 +0200
+++ contrib/codepage.c	2003-07-08 02:41:58.000000000 +0200
@@ -227,9 +227,7 @@
 	if (optc)
 	  return 0;
 
-        fprintf(stderr, "\
-Warning: CP format is a hack!\n
-         The files produced may or may not be usable!\n");
+        fprintf(stderr, "Warning: CP format is a hack!\nThe files produced may or may not be usable!\n");
 
 	sprintf(outfile, "%d.cp", CPEntryHeader.codepage);
 	if ((out = fopen(outfile, "w")) == NULL) {
