--- binutils-2.15.92.0.2/ld/testsuite/ld-selective/selective.exp.x86_64-testsuite	2004-07-28 00:36:11.000000000 -0400
+++ binutils-2.15.92.0.2/ld/testsuite/ld-selective/selective.exp	2004-12-16 15:45:00.624000398 -0500
@@ -42,8 +42,8 @@ set seltests {
   {selective1 C   1.c  {}       {}    {dropme1 dropme2} {}}
   {selective2 C   2.c  {}       {}    {foo} {mips*-*}}
   {selective3 C   2.c  {-u foo} {foo} {{foo 0}} {mips*-*}}
-  {selective4 C++ 3.cc {}       {start a A::foo() B::foo()} {A::bar()} {mips*-*}}
-  {selective5 C++ 4.cc {}       {start a A::bar()} {A::foo() B::foo()} {mips*-*}}
+  {selective4 C++ 3.cc {}       {start a A::foo() B::foo()} {A::bar()} {mips*-* x86_64*-*}}
+  {selective5 C++ 4.cc {}       {start a A::bar()} {A::foo() B::foo()} {mips*-* x86_64*-*}}
   {selective6 C++ 5.cc {}       {start a A::bar()}
     {A::foo() B::foo() dropme1() dropme2()} {*-*-*}}
 }
