Index: ssl_engine_kernel.c
===================================================================
RCS file: /home/cvs/httpd-2.0/modules/ssl/ssl_engine_kernel.c,v
retrieving revision 1.105
diff -u -r1.105 ssl_engine_kernel.c
--- httpd-2.0/modules/ssl/ssl_engine_kernel.c	5 Mar 2004 02:44:40 -0000	1.105
+++ httpd-2.0/modules/ssl/ssl_engine_kernel.c	25 May 2004 10:00:01 -0000
@@ -807,8 +807,7 @@
     SSLConnRec *sslconn = myConnConfig(r->connection);
     SSLSrvConfigRec *sc = mySrvConfig(r->server);
     SSLDirConfigRec *dc = myDirConfig(r);
-    char buf1[MAX_STRING_LEN], buf2[MAX_STRING_LEN];
-    char *clientdn;
+    char *clientdn, *buf;
     const char *auth_line, *username, *password;
 
     /*
@@ -886,14 +885,14 @@
      * adding the string "xxj31ZMTZzkVA" as the password in the user file.
      * This is just the crypted variant of the word "password" ;-)
      */
-    apr_snprintf(buf1, sizeof(buf1), "%s:password", clientdn);
-    ssl_util_uuencode(buf2, buf1, FALSE);
-
-    apr_snprintf(buf1, sizeof(buf1), "Basic %s", buf2);
-    apr_table_set(r->headers_in, "Authorization", buf1);
+    buf = apr_pstrcat(r->pool, "Basic ", 
+                      ap_pbase64encode(r->pool, 
+                                       apr_pstrcat(r->pool, clientdn, 
+                                                   ":password", NULL)), NULL);
+    apr_table_set(r->headers_in, "Authorization", buf);
 
     ap_log_error(APLOG_MARK, APLOG_INFO, 0, r->server,
-                 "Faking HTTP Basic Auth header: \"Authorization: %s\"", buf1);
+                 "Faking HTTP Basic Auth header: \"Authorization: %s\"", buf);
 
     return DECLINED;
 }
