--- ifplugd-0.14/src/ifplugd.c.fred	2003-07-06 22:48:13.000000000 +0200
+++ ifplugd-0.14/src/ifplugd.c	2003-07-07 21:31:54.000000000 +0200
@@ -379,7 +379,7 @@
             goto finish;
 
     if (daemonize && wait_on_fork) {
-        char c = status == IFSTATUS_UP ? 2 : (status == IFSTATUS_DOWN ? 3 : 1);
+        char c = status == IFSTATUS_UP ? 1 : (status == IFSTATUS_DOWN ? 2 : 3);
         daemon_retval_send(c);
         send_retval = 0;
     }
@@ -774,7 +774,7 @@
                 }
             
             
-            return c;
+            return c - 1;
         }
     }
 
