--- nfs-utils-1.0.6/support/export/Makefile.pie	2003-09-15 01:29:08.000000000 +0200
+++ nfs-utils-1.0.6/support/export/Makefile	2004-02-12 16:18:05.178180168 +0100
@@ -14,6 +14,8 @@
 
 include $(TOP)rules.mk
 
+CFLAGS += -fpie
+
 $(RPCHDRS) $(RPCSRCS): mount.x
 	$(RM) $(RPCHDRS) $(RPCSRCS)
 	$(RPCGEN) -h -o mount.h $<
--- nfs-utils-1.0.6/support/misc/Makefile.pie	2003-05-30 06:43:39.000000000 +0200
+++ nfs-utils-1.0.6/support/misc/Makefile	2004-02-12 16:18:05.179180016 +0100
@@ -9,3 +9,5 @@
 
 install::
 	@:
+
+CFLAGS += -fpie
--- nfs-utils-1.0.6/support/nfs/Makefile.pie	2003-03-26 01:29:08.000000000 +0100
+++ nfs-utils-1.0.6/support/nfs/Makefile	2004-02-12 16:18:05.180179864 +0100
@@ -12,3 +12,5 @@
 
 install::
 	@:
+
+CFLAGS += -fpie
--- nfs-utils-1.0.6/utils/lockd/Makefile.pie	2000-02-26 07:55:18.000000000 +0100
+++ nfs-utils-1.0.6/utils/lockd/Makefile	2004-02-12 16:18:05.181179712 +0100
@@ -10,3 +10,6 @@
 MAN8	= lockd
 
 include $(TOP)rules.mk
+
+CFLAGS += -fpie
+LDFLAGS += -pie
--- nfs-utils-1.0.6/utils/mountd/Makefile.pie	2003-05-21 08:22:41.000000000 +0200
+++ nfs-utils-1.0.6/utils/mountd/Makefile	2004-02-12 16:18:05.182179560 +0100
@@ -10,3 +10,6 @@
 MAN8	= mountd
 
 include $(TOP)rules.mk
+
+CFLAGS += -fpie
+LDFLAGS += -pie
--- nfs-utils-1.0.6/utils/nfsd/Makefile.pie	2004-02-12 17:07:55.875525064 +0100
+++ nfs-utils-1.0.6/utils/nfsd/Makefile	2004-02-12 17:08:22.726443104 +0100
@@ -10,6 +10,9 @@
 MAN8	= nfsd
 include $(TOP)rules.mk
 
+CFLAGS += -fpie
+LDFLAGS += -pie
+
 # 
 # all::	nfsd
 # 	@echo "Done."
--- nfs-utils-1.0.6/utils/statd/Makefile.pie	2003-09-15 01:29:39.000000000 +0200
+++ nfs-utils-1.0.6/utils/statd/Makefile	2004-02-12 16:18:05.183179408 +0100
@@ -37,6 +37,8 @@
 include $(TOP)rules.mk
 
 AFLAGS	+= -Wno-unused
+CFLAGS += -fpie
+LDFLAGS += -pie
 
 $(RPCHDRS) $(RPCSRCS):	sm_inter.x
 	$(RM) $(RPCHDRS) $(RPCSRCS)
