--- tetex-src-2.0/texk/web2c/mp.ch~	2003-02-03 16:25:15.000000000 +0100
+++ tetex-src-2.0/texk/web2c/mp.ch	2003-02-03 16:25:15.000000000 +0100
@@ -147,7 +147,7 @@
 @!max_font_dimen=50; {maximum number of \&{fontdimen} parameters}
 @#
 @!inf_main_memory = 2999;
-@!sup_main_memory = 8000000;
+@!sup_main_memory = 32000000;
 
 @!inf_max_strings = 2500;
 @!sup_max_strings = ssup_max_strings;
@@ -159,7 +159,7 @@
 @!sup_path_size = 300000;
 
 @!inf_pool_size = 32000;
-@!sup_pool_size = 10000000;
+@!sup_pool_size = 40000000;
 @!inf_pool_free = 1000;
 @!sup_pool_free = sup_pool_size;
 @!inf_string_vacancies = 8000;
