--- tetex-src-2.0.2/texk/web2c/pdftexdir/mapfile.c.badc	2003-02-17 13:19:51.000000000 +0100
+++ tetex-src-2.0.2/texk/web2c/pdftexdir/mapfile.c	2004-02-08 21:34:18.000000000 +0100
@@ -161,12 +161,13 @@
     int hash_code_lower = 0;
     int i;
     for (i=0; *str; i++, str++) {
-        hash_code_upper ^= (*str << 8) + *++str;
+        hash_code_upper ^= (*str << 8) + str[1];
+        ++str;
         if (!*str || !*++str) {
             break;
         }
-        hash_code_lower ^= (*str << 8) + *++str;
-        if (!*str) {
+        hash_code_lower ^= (*str << 8) + str[1];
+        if (!*++str) {
             break;
         }
     }
--- tetex-src-2.0.2/texk/dvipsk/emspecial.c.badc	2002-09-24 20:24:21.000000000 +0200
+++ tetex-src-2.0.2/texk/dvipsk/emspecial.c	2004-02-08 21:34:18.000000000 +0100
@@ -1049,7 +1049,8 @@
 		if (!hi) {   /* we are about to place the low 4 bits */
 		    ch = ((ch>>4)&0x0f) | ((ch<<4)&0xf0); /* swap order */
 		    i++;
-		    *s++ = (*s & 0xf0) | (ch & 0x0f);
+		    *s = (*s & 0xf0) | (ch & 0x0f);
+		    ++s;
 		    hi = TRUE;
 		    cnt--;
 		}
--- tetex-src-2.0.2/texk/ps2pkm/token.c.badc	1999-02-06 23:00:18.000000000 +0100
+++ tetex-src-2.0.2/texk/ps2pkm/token.c	2004-02-08 21:34:18.000000000 +0100
@@ -36,14 +36,14 @@
 #include "tokst.h"
 #include "hdigit.h"
  
-extern int T1Getc(struct F_FILE *f);
-extern int T1Ungetc(int c,struct F_FILE *f);
 /*
  * -------------------------------------------------------------------
  * Globals
  * -------------------------------------------------------------------
  */
  
+extern int  T1Getc(F_FILE * ), T1Ungetc( int,F_FILE *);
+
 /* These variables are set by the caller */
 char           *tokenStartP;   /* Pointer to token buffer in VM */
 char           *tokenMaxP;     /* Pointer to last byte in buffer + 1 */
@@ -108,7 +108,7 @@
   if (exponent < 0) {
     power = 0.1;
     value = (exponent & 1 ? power : 1.0);
-    exponent = -(++exponent >> 1); /* portable C for -(exponent/2) */
+    exponent = -((exponent+1) >> 1); /* portable C for -(exponent/2) */
   }
   else {
     power = 10.0;
