--- pciutils-2.1.10/lib/configure.x86_64	2000-04-21 13:32:26.000000000 +0200
+++ pciutils-2.1.10/lib/configure	2002-06-26 17:20:22.000000000 +0200
@@ -45,6 +45,11 @@ case $sys in
 						echo >>$c '#define HAVE_PM_INTEL_CONF'
 						ok=1
 						;;
+			x86_64)
+#				echo_n " i386-ports"
+				echo >> $c '#define HAVE_64BIT_ADDRESS'
+#				ok=1
+				;;
 				alpha|ia64)	echo >>$c '#define HAVE_64BIT_ADDRESS'
 #						echo_n " syscalls"
 #						echo >>$c '#define HAVE_PM_SYSCALLS'
