--- pure-ftpd-1.0.16b/configuration-file/pure-ftpd.conf.in.slsconf	2003-08-03 07:30:34.000000000 -0600
+++ pure-ftpd-1.0.16b/configuration-file/pure-ftpd.conf.in	2004-03-03 15:29:26.000000000 -0700
@@ -9,7 +9,7 @@
 # instead of command-line options, please run the
 # following command :
 #
-# @prefix@/sbin/pure-config.pl @prefix@/etc/pure-ftpd.conf
+# /usr/sbin/pure-config.pl /etc/pure-ftpd/pure-ftpd.conf
 #
 # Please don't forget to have a look at documentation at
 # http://www.pureftpd.org/documentation.html for a complete list of
@@ -43,7 +43,7 @@
 
 # Fork in background
 
-Daemonize                   yes
+Daemonize                   no
 
 
 
@@ -107,23 +107,23 @@
 
 # LDAP configuration file (see README.LDAP)
 
-# LDAPConfigFile                /etc/pureftpd-ldap.conf
+# LDAPConfigFile                /etc/pure-ftpd/pureftpd-ldap.conf
 
 
 
 # MySQL configuration file (see README.MySQL)
 
-# MySQLConfigFile               /etc/pureftpd-mysql.conf
+# MySQLConfigFile               /etc/pure-ftpd/pureftpd-mysql.conf
 
 
 # Postgres configuration file (see README.PGSQL)
 
-# PGSQLConfigFile               /etc/pureftpd-pgsql.conf
+# PGSQLConfigFile               /etc/pure-ftpd/pureftpd-pgsql.conf
 
 
 # PureDB user database (see README.Virtual-Users)
 
-# PureDB                        /etc/pureftpd.pdb
+# PureDB                        /etc/pure-ftpd/pureftpd.pdb
 
 
 # Path to pure-authd socket (see README.Authentication-Modules)
@@ -134,7 +134,7 @@
 
 # If you want to enable PAM authentication, uncomment the following line
 
-# PAMAuthentication             yes
+PAMAuthentication             yes
 
 
 
@@ -301,7 +301,7 @@
 # fw.c9x.org - jedi [13/Dec/1975:19:36:39] "GET /ftp/linux.tar.bz2" 200 21809338
 # This log file can then be processed by www traffic analyzers.
 
-# AltLog                     clf:/var/log/pureftpd.log
+AltLog                     clf:/var/log/pureftpd.log
 
 
 
