--- lilo-22.5.6.orig/second.S	2003-05-25 13:48:11.000000000 +0000
+++ lilo-22.5.6/second.S	2003-06-09 15:31:02.000000000 +0000
@@ -2635,7 +2635,8 @@
 ; try the E820 memory map first
 	xor	edx,edx			; flag nothing found
 	xor	ebx,ebx
-	jmp	e8go
+;       jmp     e8go                    ; Circumvented due to BIOS problems
+        br      no_e801			; Skip to backup plan
 e8go2:	or	ebx,ebx			; test for end
 	jz	e8go5
 e8go:	push	edx			; save best prospect
