--- console-tools-0.2.3/fontfiletools/fonts2virfont.c.old	Fri Aug 10 21:42:37 2001
+++ console-tools-0.2.3/fontfiletools/fonts2virfont.c	Fri Aug 10 21:43:07 2001
@@ -18,7 +18,7 @@
 
 void usage()
 {
-  version();
+  version(progname);
   
   fprintf (stderr, 
 	   "Usage: "
--- console-tools-0.2.3/vttools/vt-is-UTF8.c.old	Fri Aug 10 21:44:30 2001
+++ console-tools-0.2.3/vttools/vt-is-UTF8.c	Fri Aug 10 21:44:48 2001
@@ -51,7 +51,7 @@
       switch (result)
 	{
 	case 'V':
-	  version();
+	  version(progname);
 	  exit (0);
 	case 'h':
 	  usage();
--- console-tools-0.2.3/kbdtools/dumpkeys.c.old	Fri Aug 10 21:42:06 2001
+++ console-tools-0.2.3/kbdtools/dumpkeys.c	Fri Aug 10 21:43:28 2001
@@ -613,7 +613,7 @@
 	  verbose = 1;
 	  break;
 	case 'V':
-	  version();
+	  version(progname);
 	  exit(0);
 	case 'c':
 	  if ((set_charset(optarg)) != 0)
--- console-tools-0.2.3/kbdtools/showkey.c.old	Fri Aug 10 21:43:40 2001
+++ console-tools-0.2.3/kbdtools/showkey.c	Fri Aug 10 21:43:55 2001
@@ -158,7 +158,7 @@
 	    badusage(_("timeout value must be numeric"));
 	  break;
 	case 'V':
-	  version();
+	  version(progname);
 	  exit(0);
 	case 'h':
 	case '?':
--- console-tools-0.2.3/screenfonttools/consolechars.c.old	Fri Aug 10 21:34:54 2001
+++ console-tools-0.2.3/screenfonttools/consolechars.c	Fri Aug 10 21:44:22 2001
@@ -44,7 +44,7 @@
 
 static void usage()
 {
-  version();
+  version(progname);
   printf(_("Usage:  %s [options] [commands]\n"), progname);
   OPTIONS_ARE();
   OPT(" -v --verbose        ", _("List operations as they are done"));
@@ -171,7 +171,7 @@
       switch (result)
 	{
 	case 'V':
-	  version();
+	  version(progname);
 	  exit (EX_OK);
 	case 'h':
 	  usage();
