diff -urNp dump-0.4b24.orig/MCONFIG.in dump-0.4b24/MCONFIG.in
--- dump-0.4b24.orig/MCONFIG.in	Wed Sep 12 14:40:07 2001
+++ dump-0.4b24/MCONFIG.in	Sun Sep 30 17:17:43 2001
@@ -44,7 +44,7 @@ GINC=		-I$(top_builddir) -I$(top_srcdir)
 # indicate where the ext2fs library can be found (this is not needed if you
 # have run `make install-libs' in the e2fsprogs source directory).
 #GLIBDIR=	-L/usr/src/e2fsprogs-0.5c/lib
-GLIBS=		$(GLIBDIR) -L../compat/lib -lcompat -lext2fs -lcom_err
+GLIBS=		-L../lib $(GLIBDIR) -L../compat/lib -lcompat -lext2fs -lcom_err
 
 #
 # Definitions (don't change them unless you know what you are doing)
