--- lilo-22.5.9/loader.h.pix	2003-05-31 23:28:37.000000000 +0200
+++ lilo-22.5.9/loader.h	2004-07-05 14:07:28.603197354 +0200
@@ -11,6 +11,7 @@
 
 extern BUILTIN_FILE First;
 extern BUILTIN_FILE Second;
+extern BUILTIN_FILE SecondGraphic;
 extern BUILTIN_FILE Third;
 extern BUILTIN_FILE Bitmap;
 extern BUILTIN_FILE Mbr;
--- lilo-22.5.9/loader.S.pix	2003-08-15 05:05:42.000000000 +0200
+++ lilo-22.5.9/loader.S	2004-07-05 14:07:28.603197354 +0200
@@ -6,6 +6,7 @@
 
 	global	First
 	global	Second
+	global	SecondGraphic
 	global	Third
 	global	Bitmap
 	global	Mbr
@@ -25,6 +26,12 @@
 #endif
 SecondEnd	equ	$
 
+SecondGraphic		dd	SecondGraphicEnd-$-4
+#ifdef LCF_BUILTIN
+incbin "second-graphic.b"
+#endif
+SecondGraphicEnd	equ	$
+
 Third		dd	ThirdEnd-$-4
 #ifdef LCF_BUILTIN
 incbin	"third.b"
--- lilo-22.5.9/Makefile.pix	2004-04-09 01:36:15.000000000 +0200
+++ lilo-22.5.9/Makefile	2004-07-05 14:08:05.523302004 +0200
@@ -84,7 +84,7 @@
 LD86=ld86 -0
 NASM=nasm
 
-CFLAGS=$(OPT) -Wall -g $(PCONFIG)
+CFLAGS=$(OPT) -Wall $(RPM_OPT_FLAGS) $(PCONFIG) $(AFLAGS)
 LDFLAGS=#-Xlinker -qmagic
 LIBS=
 
@@ -96,7 +96,7 @@
 LIS=chain.lis disk.lis dparam.lis os2_d.lis first.lis second.lis \
   bootsect.lis third.lis mbr.lis bitmap.lis loader.lis pseudo.lis \
   mbr2.lis pseudo1.lis
-BFILES=first.b second.b third.b bitmap.b chain.b mbr.b mbr2.b \
+BFILES=first.b second.b third.b second-graphic.b bitmap.b chain.b mbr.b mbr2.b \
   bootsect.b pseudo.b pseudo1.b
 CFILES=boot.c bsect.c cfg.c common.c device.c edit.c geometry.c \
   identify.c lilo.c map.c partition.c probe.c raid.c shs2.c temp.c
@@ -180,7 +180,7 @@
 edit:		$(EDIT)
 		$(CC) -Wall -s -O -DSTANDALONE -o edit $(EDIT) $(LDFLAGS)
 
-loader.o:	loader.s first.b second.b third.b bitmap.b mbr.b \
+loader.o:	loader.s first.b second.b third.b second-graphic.b bitmap.b mbr.b \
 			chain.b os2_d.b mbr2.b
 		$(NASM) -f elf -o loader.o -l loader.lis  loader.s
 
@@ -263,6 +263,9 @@
 boot-text.b:	first.b second.b
 		(dd if=first.b bs=512 conv=sync; dd if=second.b) >boot-text.b
 
+boot-graphic.b:	first.b second-graphic.b
+		(dd if=first.b bs=512 conv=sync; dd if=second-graphic.b) >boot-graphic.b
+
 boot-menu.b:	first.b third.b
 		(dd if=first.b bs=512 conv=sync; dd if=third.b) >boot-menu.b
 
@@ -273,6 +276,7 @@
 		$(CPP) -C -traditional -DLILO_ASM -o common.s common.h
 
 bitmap.o: bitmap.s common.s
+second-graphic.o: second-graphic.s common.s
 third.o: third.s common.s
 second.o: second.s common.s
 first.o: first.s common.s
@@ -285,6 +289,10 @@
 			graph.S menu.S strlen.S bitmap.S display4.S
 		$(CPP) $(PCONFIG) -DTEXT=0xce60f1b7 second.S -o second.s
 
+second-graphic.s:	second.S mdk_graphic.S read.S volume.S mapper.S biosdata.S shs3.S bdata.h lilo.h version.h \
+			graph.S menu.S strlen.S bitmap.S display4.S
+		$(CPP) $(PCONFIG) -DTEXT=0x68db3d5c -DMDK_GRAPHIC second.S -o second-graphic.s
+
 third.s:	second.S read.S volume.S mapper.S biosdata.S bdata.h lilo.h version.h \
 			graph.S menu.S strlen.S crt.S
 		$(CPP) $(PCONFIG) -DMENU=0x662efe71 second.S -o third.s
--- lilo-22.5.9/bsect.c.pix	2004-03-11 01:59:45.000000000 +0100
+++ lilo-22.5.9/bsect.c	2004-07-05 14:07:28.605197360 +0200
@@ -84,14 +84,29 @@
 
 static PASSWORD *pwsave = NULL;
 
+int message_is_graphic()
+{
+  char is_graphic = 0;
+  char *message = cfg_get_strg(cf_options,"message");
+  int fd = open(message, O_RDONLY);
+  if (fd > 0) {
+    char c;
+    is_graphic = read(fd, &c, 1) == 1 && c == 0x0E;
+    close(fd);
+  }
+  return is_graphic;
+}
+
+
 BUILTIN_FILE *select_loader(void)
 {
     BUILTIN_FILE *loader = &Third;	/* MENU interface is the default */
     char *install = cfg_get_strg(cf_options,"install");
     char *bitmap  = cfg_get_strg(cf_options,"bitmap");
 
-    if (!install) {
-	if (bitmap) loader = &Bitmap;
+    if (message_is_graphic()) loader = &SecondGraphic;
+    else if (!install) {
+	if (bitmap) loader = &Bitmap;	
     } 
     else if (strstr(install,"text")) loader = &Second;  /* text interface */
     else if (strstr(install,"menu")) loader = &Third;
@@ -637,6 +652,7 @@
 	loader = select_loader();
 	if (verbose > 0) {
 	    printf("Using %s secondary loader\n",
+		loader==&SecondGraphic ? "GRAPHIC" :
 		loader==&Bitmap ? "BITMAP" :
 		loader==&Third  ? "MENU" :
 		"TEXT" );
