--- ocaml-3.05/asmrun/Makefile.pix	2002-07-29 18:39:23.000000000 +0200
+++ ocaml-3.05/asmrun/Makefile	2002-07-29 18:40:23.000000000 +0200
@@ -18,8 +18,8 @@
 CC=$(NATIVECC)
 FLAGS=-I../byterun -DNATIVE_CODE -DTARGET_$(ARCH) -DSYS_$(SYSTEM) 
 CFLAGS=$(FLAGS) -O $(NATIVECCCOMPOPTS)
-DFLAGS=$(FLAGS) -g -DDEBUG $(NATIVECCCOMPOPTS)
-PFLAGS=$(FLAGS) -pg -O -DPROFILING $(NATIVECCPROFOPTS)
+DFLAGS=$(FLAGS) -g -DDEBUG
+PFLAGS=$(FLAGS) -pg -O -DPROFILING
 
 COBJS=startup.o main.o fail.o roots.o globroots.o signals.o \
   misc.o freelist.o major_gc.o minor_gc.o memory.o alloc.o compare.o ints.o \
