--- gcc-3.3.1/gcc/p/rts/Makefile.in.gpc-serialize-build	2003-04-22 18:31:24.000000000 +0200
+++ gcc-3.3.1/gcc/p/rts/Makefile.in	2003-07-18 12:07:33.000000000 +0200
@@ -113,6 +113,9 @@ DUMMY=# Don't let configure substitute `
         sharedlib. sharedlib.no sharedlib.@with_shared$(DUMMY)@ sharedlib.yes \
         mostlyclean clean distclean extraclean maintainer-clean
 
+# FIXME: workaround parallel build problems for now
+.NOTPARALLEL: $(PAS_SRCS:%.pas=%.o) $(PAS_SRCS:%.pas=%.lo)
+
 install:
 	@echo "There is currently no separate installation for the GNU Pascal Compiler's" >&2
 	@echo "Run Time System, because it is usually built and installed together" >&2
