--- gcc-3.3.1/gcc/gcse.c.no-store-motion	2003-07-21 18:26:00.000000000 +0200
+++ gcc-3.3.1/gcc/gcse.c	2003-07-24 19:36:48.000000000 +0200
@@ -915,7 +915,8 @@ gcse_main (f, file)
   end_alias_analysis ();
   allocate_reg_info (max_reg_num (), FALSE, FALSE);
 
-  if (!optimize_size && flag_gcse_sm)
+  /* Store motion disabled until it is fixed.  */
+  if (0 && !optimize_size && flag_gcse_sm)
     store_motion ();
 
   /* Record where pseudo-registers are set.  */
