--- squid-2.6.STABLE6/src/auth/negotiate/auth_negotiate.c.orig	2006-07-31 01:27:04.000000000 +0200
+++ squid-2.6.STABLE6/src/auth/negotiate/auth_negotiate.c	2007-01-17 20:44:49.000000000 +0100
@@ -651,7 +651,7 @@
 authNegotiateAuthenticated(auth_user_request_t * auth_user_request)
 {
     negotiate_request_t *negotiate_request = auth_user_request->scheme_data;
-    if (negotiate_request->auth_state == AUTHENTICATE_STATE_FINISHED)
+    if (negotiate_request->auth_state == AUTHENTICATE_STATE_DONE)
 	return 1;
     debug(29, 9) ("User not fully authenticated.\n");
     return 0;
--- squid-2.6.STABLE6/src/auth/ntlm/auth_ntlm.c.orig	2006-07-31 01:27:04.000000000 +0200
+++ squid-2.6.STABLE6/src/auth/ntlm/auth_ntlm.c	2007-01-17 20:44:49.000000000 +0100
@@ -427,7 +427,7 @@
 
     auth_user = auth_user_request->auth_user;
     assert(auth_user != NULL);
-    assert(auth_user->auth_type == AUTH_NEGOTIATE);
+    assert(auth_user->auth_type == AUTH_NTLM);
     ntlm_user = auth_user_request->auth_user->scheme_data;
 
     if (ntlm_request->authserver == NULL)
@@ -508,7 +508,7 @@
     assert(ntlm_request);
     assert(handler);
     assert(data);
-    assert(auth_user->auth_type == AUTH_NEGOTIATE);
+    assert(auth_user->auth_type == AUTH_NTLM);
     debug(29, 9) ("authenticateNTLMStart: auth state '%d'\n", ntlm_request->auth_state);
     sent_string = ntlm_request->client_blob;
 
@@ -581,7 +581,7 @@
     dlink_node *node;
     assert(auth_user_request->auth_user == NULL);
     auth_user_request->auth_user = authenticateAuthUserNew("ntlm");
-    auth_user_request->auth_user->auth_type = AUTH_NEGOTIATE;
+    auth_user_request->auth_user->auth_type = AUTH_NTLM;
     auth_user_request->auth_user->scheme_data = memPoolAlloc(ntlm_user_pool);
     auth_user_request->scheme_data = memPoolAlloc(ntlm_request_pool);
     memset(auth_user_request->scheme_data, '\0', sizeof(ntlm_request_t));
@@ -607,7 +607,7 @@
 authNTLMAuthenticated(auth_user_request_t * auth_user_request)
 {
     ntlm_request_t *ntlm_request = auth_user_request->scheme_data;
-    if (ntlm_request->auth_state == AUTHENTICATE_STATE_FINISHED)
+    if (ntlm_request->auth_state == AUTHENTICATE_STATE_DONE)
 	return 1;
     debug(29, 9) ("User not fully authenticated.\n");
     return 0;
@@ -624,7 +624,7 @@
 
     auth_user = auth_user_request->auth_user;
     assert(auth_user);
-    assert(auth_user->auth_type == AUTH_NEGOTIATE);
+    assert(auth_user->auth_type == AUTH_NTLM);
     assert(auth_user->scheme_data != NULL);
     assert(auth_user_request->scheme_data != NULL);
     ntlm_user = auth_user->scheme_data;
@@ -661,7 +661,7 @@
 	ntlm_request->auth_state = AUTHENTICATE_STATE_INITIAL;
 	safe_free(ntlm_request->client_blob);
 	ntlm_request->client_blob = xstrdup(blob);
-	conn->auth_type = AUTH_NEGOTIATE;
+	conn->auth_type = AUTH_NTLM;
 	conn->auth_user_request = auth_user_request;
 	ntlm_request->conn = conn;
 	/* and lock for the connection duration */
