--- linux-2.4.26/drivers/ide/ide.c.di97.orig	2004-07-17 14:35:09.631134451 +0300
+++ linux-2.4.26/drivers/ide/ide.c	2004-07-17 14:37:50.865288766 +0300
@@ -2980,7 +2980,8 @@ static int ide_notify_reboot (struct not
 			return NOTIFY_DONE;
 	}
 
-	printk(KERN_INFO "flushing ide devices: ");
+	if(console_loglevel > 8)
+	  printk(KERN_INFO "flushing ide devices: ");
 
 	for (i = 0; i < MAX_HWIFS; i++) {
 		hwif = &ide_hwifs[i];
@@ -2992,7 +2993,8 @@ static int ide_notify_reboot (struct not
 				continue;
 
 			/* set the drive to standby */
-			printk("%s ", drive->name);
+			if(console_loglevel > 8)
+			    printk("%s ", drive->name);
 #ifdef	CONFIG_ALPHA
 			/* On Alpha, halt(8) doesn't actually turn the machine
 			   off, it puts you into the sort of firmware monitor.
@@ -3018,7 +3020,8 @@ static int ide_notify_reboot (struct not
 			continue;
 		}
 	}
-	printk("\n");
+	if(console_loglevel > 8)
+	  printk("\n");
 	return NOTIFY_DONE;
 }
 
