diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' t2/drivers/block/Config.in t1/drivers/block/Config.in
--- t2/drivers/block/Config.in	2003-02-01 03:09:12.000000000 +0100
+++ t1/drivers/block/Config.in	2003-02-01 01:55:57.000000000 +0100
@@ -41,7 +41,7 @@ dep_tristate 'Micro Memory MM5415 Batter
 
 tristate 'Loopback device support' CONFIG_BLK_DEV_LOOP
 if [ "$CONFIG_BLK_DEV_LOOP" != "n" ]; then
-   bool '  AES encrypted loop device support' CONFIG_BLK_DEV_LOOP_AES
+   tristate '  AES encrypted loop device support' CONFIG_BLK_DEV_LOOP_AES
 fi
 dep_tristate 'Network block device support' CONFIG_BLK_DEV_NBD $CONFIG_NET
 
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' t2/drivers/block/loop.c t1/drivers/block/loop.c
--- t2/drivers/block/loop.c	2003-02-01 03:09:12.000000000 +0100
+++ t1/drivers/block/loop.c	2003-02-01 01:55:57.000000000 +0100
@@ -160,7 +160,7 @@ struct loop_func_table xor_funcs = { 
 	init: xor_status
 }; 	
 
-#if CONFIG_BLK_DEV_LOOP_AES
+#if defined(CONFIG_BLK_DEV_LOOP_AES) || defined(CONFIG_BLK_DEV_LOOP_AES_MODULE)
 int transfer_aes(struct loop_device *lo, int cmd, char *raw_buf,
 		char *loop_buf, int size, int devSect)
 {
@@ -258,13 +258,13 @@ static struct loop_func_table funcs_aes 
 	init:		keySetup_aes,
 	release:	keyClean_aes
 };
-#endif /* CONFIG_BLK_DEV_LOOP_AES */
+#endif /* CONFIG_BLK_DEV_LOOP_AES || CONFIG_BLK_DEV_LOOP_AES_MODULE */
 
 /* xfer_funcs[0] is special - its release function is never called */ 
 struct loop_func_table *xfer_funcs[MAX_LO_CRYPT] = {
 	&none_funcs,
 	&xor_funcs,
-#if CONFIG_BLK_DEV_LOOP_AES
+#if defined(CONFIG_BLK_DEV_LOOP_AES) || defined(CONFIG_BLK_DEV_LOOP_AES_MODULE)
 	[LO_CRYPT_AES] = &funcs_aes,
 #endif
 };
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' t2/drivers/misc/Makefile t1/drivers/misc/Makefile
--- t2/drivers/misc/Makefile	2003-02-01 03:09:12.000000000 +0100
+++ t1/drivers/misc/Makefile	2003-02-01 02:09:17.000000000 +0100
@@ -14,7 +14,9 @@
 
 O_TARGET := misc.o
 
-ifeq ($(CONFIG_BLK_DEV_LOOP_AES),y)
+list-multi := aes.o
+
+ifneq ($(CONFIG_BLK_DEV_LOOP_AES),n)
 AES_PENTIUM_ASM=n
 ifeq ($(CONFIG_X86),y)
 ifneq ($(CONFIG_X86_64),y)
@@ -25,15 +27,23 @@ endif
 endif
 endif
 endif
+
+export-objs	+= aes-c.o aes-i586-ksym.o
+
 ifeq ($(AES_PENTIUM_ASM),y)
-  export-objs	+= aes-i586-ksym.o
-  obj-y 	+= aes-i586.o aes-i586-ksym.o
+  aes-objs := aes-i586.o aes-i586-ksym.o
 else
-  export-objs	+= aes.o
-  obj-y 	+= aes.o
+  aes-objs := aes-c.o
 endif
 endif
 
+obj-$(CONFIG_BLK_DEV_LOOP_AES) = aes.o
+
 include $(TOPDIR)/Rules.make
 
 fastdep:
+
+# Link rules for multi-part drivers.
+
+aes.o: $(aes-objs)
+	$(LD) -r -o $@ $(aes-objs)
