--- linux-2.4.20/drivers/char/amd768_rng.c.hb02.orig	2003-04-14 22:55:48.000000000 -0400
+++ linux-2.4.20/drivers/char/amd768_rng.c	2003-04-14 23:14:53.000000000 -0400
@@ -15,6 +15,7 @@
 	This software may be used and distributed according to the terms
         of the GNU General Public License, incorporated herein by reference.
 
+    2002 Andi Kleen. Added AMD 8111 PCI-ID. 
  */
 
 
@@ -218,7 +219,7 @@ static int __init rng_init_one (struct p
 	rnen|=(1<<7);	/* PMIO enable */
 	pci_write_config_byte(dev, 0x41, rnen);
 
-	printk(KERN_INFO PFX "AMD768 system management I/O registers at 0x%X.\n", pmbase);
+	printk(KERN_INFO PFX "AMD system management I/O registers at 0x%X.\n", pmbase);
 	DPRINTK ("EXIT, returning 0\n");
 	return 0;
 
@@ -239,6 +240,7 @@ err_out:
  */
 static struct pci_device_id rng_pci_tbl[] __initdata = {
 	{ 0x1022, 0x7443, PCI_ANY_ID, PCI_ANY_ID, },
+	{ 0x1022, 0x7455, PCI_ANY_ID, PCI_ANY_ID, },
 	{ 0x1022, 0x746b, PCI_ANY_ID, PCI_ANY_ID, },
 	{ 0, },
 };
