diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff ../linux-2.4.20-rc3/arch/x86_64/config.in c2/arch/x86_64/config.in
--- ../linux-2.4.20-rc3/arch/x86_64/config.in	2002-11-22 23:36:55.000000000 +0100
+++ c2/arch/x86_64/config.in	2002-11-25 11:37:23.000000000 +0100
@@ -230,6 +230,8 @@ endmenu
 
 source drivers/usb/Config.in
 
+source lib/Config.in
+
 mainmenu_option next_comment
 comment 'Kernel hacking'
 
