--- linux-2.4.22/arch/x86_64/config.in.amd64-boot-kernel	2003-10-24 11:01:26.000000000 +0200
+++ linux-2.4.22/arch/x86_64/config.in	2003-10-24 12:40:22.000000000 +0200
@@ -234,7 +234,7 @@ source lib/Config.in
 
 mainmenu_option next_comment
 comment 'Kernel hacking'
-
+bool 'Make the kernel for install system (only usefull for distro makers)' CONFIG_BOOT_KERNEL
 bool 'Kernel debugging' CONFIG_DEBUG_KERNEL
 if [ "$CONFIG_DEBUG_KERNEL" != "n" ]; then
    bool '  Debug memory allocations' CONFIG_DEBUG_SLAB
--- linux-2.4.22/arch/x86_64/Makefile.amd64-boot-kernel	2003-10-24 11:01:07.000000000 +0200
+++ linux-2.4.22/arch/x86_64/Makefile	2003-10-24 12:41:27.000000000 +0200
@@ -55,6 +55,10 @@ CFLAGS += -fno-asynchronous-unwind-table
 endif
 CFLAGS += $(call check_gcc,-fno-unit-at-a-time,)
 
+ifdef CONFIG_BOOT_KERNEL
+CFLAGS += -Os
+endif
+
 HEAD := arch/x86_64/kernel/head.o arch/x86_64/kernel/head64.o arch/x86_64/kernel/init_task.o
 
 SUBDIRS := arch/x86_64/tools $(SUBDIRS) arch/x86_64/kernel arch/x86_64/mm arch/x86_64/lib
