--- linux-2.4.22/arch/x86_64/kernel/setup.c.amd64-pcmcia	2003-10-29 06:47:04.000000000 +0100
+++ linux-2.4.22/arch/x86_64/kernel/setup.c	2003-10-30 11:30:09.000000000 +0100
@@ -197,6 +197,7 @@ void __init setup_arch(char **cmdline_p)
 {
 	int i;
 	unsigned long kernel_end; 
+	unsigned long low_mem_size;
 
  	ROOT_DEV = to_kdev_t(ORIG_ROOT_DEV);
  	drive_info = DRIVE_INFO;
@@ -332,11 +333,13 @@ void __init setup_arch(char **cmdline_p)
 	for (i = 0; i < STANDARD_IO_RESOURCES; i++)
 		request_resource(&ioport_resource, standard_io_resources+i);
 
-	/* We put PCI memory up to make sure VALID_PAGE with DISCONTIGMEM
-	   never returns true for it */ 
+	/* FIXME: We CAN NOT put PCI memory up to make sure VALID_PAGE with
+	 * DISCONTIGMEM never returns true for it. */
 
 	/* Tell the PCI layer not to allocate too close to the RAM area.. */
-	pci_mem_start = IOMAP_START;
+	low_mem_size = ((end_pfn << PAGE_SHIFT) + 0xfffff) & ~0xfffff;
+	if (low_mem_size > pci_mem_start)
+		pci_mem_start = low_mem_size;
 
 #ifdef CONFIG_GART_IOMMU
 	iommu_hole_init();
