--- linux-2.4.21/3rdparty/pdc-ultra/pdc618_mod.c.pdc-ultra-no-mmio	2003-06-08 18:10:00.000000000 +0200
+++ linux-2.4.21/3rdparty/pdc-ultra/pdc618_mod.c	2003-08-11 07:43:37.000000000 +0200
@@ -2198,6 +2198,8 @@ static int pdc618dev_ioctl(struct inode 
 
 		break;
 		}
+
+#ifdef _MMIO_
 		/* find flash mem type */
 		case STIO_FIND_FLASHMEM: {
 		tag_flash_t *buf = (tag_flash_t *)args;
@@ -2259,6 +2261,8 @@ static int pdc618dev_ioctl(struct inode 
 
 		break;
 		}
+#endif
+
 		default:
 			break;
 	}
