--- man-1.5j/src/makewhatis.sh.quoting	Tue Feb 26 14:01:59 2002
+++ man-1.5j/src/makewhatis.sh	Tue Feb 26 14:02:42 2002
@@ -198,7 +198,7 @@
 	      }
 	      sub(/\..*/, "", progname);
 	      if (use_zcat) {
-		pipe_cmd = "zcat " filename;
+		pipe_cmd = "zcat \"" filename "\"";
 		# try to avoid suspicious stuff
 		if (filename ~ /[;&|`$(]/) {
 		  print "ignored strange file name " filename " in " curdir > "/dev/stderr";
