--- beecrypt-3.1.0/python/Makefile.am.rh	2003-08-17 02:46:57.000000000 -0400
+++ beecrypt-3.1.0/python/Makefile.am	2003-12-22 21:56:47.764588621 -0500
@@ -1,14 +1,12 @@
-# Makefile for rpm library.
-
 AUTOMAKE_OPTIONS = 1.4 foreign
 
 LINT = splint
 
-PYVER= 2.2
+PYVER= 2.3
 
 SUBDIRS = test
 
-pylibdir = $(shell python -c 'import sys; print sys.path[1]')
+pylibdir = $(shell python -c 'import sys; print sys.path[2]')
 pyincdir = $(prefix)/include/python${PYVER}
 
 EXTRA_DIST = debug-py.c
