--- ppp-2.4.1/pppd/plugins/Makefile.linux.libdir	2001-03-06 23:21:18.000000000 -0500
+++ ppp-2.4.1/pppd/plugins/Makefile.linux	2002-08-09 06:43:04.000000000 -0400
@@ -3,6 +3,9 @@ CFLAGS	= -g -O2 -I.. -I../../include -fP
 LDFLAGS	= -shared
 INSTALL	= install
 
+LIBDIR = /usr/lib
+PLUGINDIR = $(LIBDIR)/pppd
+
 all:	minconn.so passprompt.so
 
 minconn.so: minconn.c
@@ -11,9 +14,7 @@ minconn.so: minconn.c
 passprompt.so: passprompt.c
 	$(CC) -o $@ $(LDFLAGS) $(CFLAGS) passprompt.c
 
-LIBDIR	= /usr/lib/pppd
-
 install: minconn.so passprompt.so
 	version=`awk -F '"' '/VERSION/ { print $$2; }' ../patchlevel.h`; \
-	$(INSTALL) -d $(LIBDIR)/$$version; \
-	$(INSTALL) $? $(LIBDIR)/$$version
\ No newline at end of file
+	$(INSTALL) -d $(PLUGINDIR)/$$version; \
+	$(INSTALL) $? $(PLUGINDIR)/$$version
--- ppp-2.4.1/pppd/Makefile.linux.libdir	2002-08-09 06:29:40.000000000 -0400
+++ ppp-2.4.1/pppd/Makefile.linux	2002-08-09 06:40:46.000000000 -0400
@@ -6,6 +6,7 @@
 # Default installation locations
 BINDIR = $(DESTDIR)/usr/sbin
 MANDIR = $(DESTDIR)/usr/man
+LIBDIR = $(DESTDIR)/usr/lib
 
 PPPDSRCS = main.c magic.c fsm.c lcp.c ipcp.c upap.c chap.c md5.c ccp.c \
 	   ipxcp.c auth.c options.c sys-linux.c md4.c chap_ms.c cbcp.c \
@@ -30,7 +31,7 @@ CC = gcc
 COPTS = -Wall $(RPM_OPT_FLAGS)
 LIBS = -lutil
 
-ifneq ($(wildcard /usr/lib/libcrypt.*),)
+ifneq ($(wildcard $(LIBDIR)/libcrypt.*),)
 LIBS += -lcrypt
 endif
 
@@ -38,7 +39,7 @@ endif
 # MS-CHAP authentication protocol.
 CHAPMS=y
 USE_CRYPT=y
-ifneq ($(wildcard /usr/lib/libcrypt.*),)
+ifneq ($(wildcard $(LIBDIR)/libcrypt.*),)
 HAVE_CRYPT_H=y
 endif
 
@@ -78,7 +79,7 @@ PLUGIN=y
 
 INCLUDE_DIRS= -I../include
 
-COMPILE_FLAGS= -D_linux_=1 -DHAVE_PATHS_H -DIPX_CHANGE -DHAVE_MULTILINK -DHAVE_MMAP
+COMPILE_FLAGS= -D_linux_=1 -DHAVE_PATHS_H -DIPX_CHANGE -DHAVE_MULTILINK -DHAVE_MMAP -DLIBDIR=\"$(LIBDIR)\"
 
 CFLAGS= $(COPTS) $(COMPILE_FLAGS) $(INCLUDE_DIRS)
 
--- ppp-2.4.1/pppd/pathnames.h.libdir	2001-03-08 00:15:37.000000000 -0500
+++ ppp-2.4.1/pppd/pathnames.h	2002-08-09 06:41:37.000000000 -0400
@@ -53,5 +53,8 @@
 #endif /* __STDC__ */
 
 #ifdef PLUGIN
-#define _PATH_PLUGIN	"/usr/lib/pppd/" VERSION
+#ifndef LIBDIR
+#define LIBDIR "/usr/lib"
+#endif
+#define _PATH_PLUGIN	LIBDIR "/pppd/" VERSION
 #endif /* PLUGIN */
