--- ./devfsd.conf~	2004-01-07 12:19:28.000000000 +0300
+++ ./devfsd.conf	2004-01-13 00:00:09.862234392 +0300
@@ -21,6 +21,9 @@
 REGISTER	.*		MKNEWCOMPAT
 UNREGISTER	.*		RMNEWCOMPAT
 
+# This stops attemtping to modload on every access to /dev/ptsN on 2.6
+LOOKUP		^pts([0-9]+)$	CFUNCTION GLOBAL mksymlink pts/\1 $devname
+
 # Enable module autoloading. You may comment this out if you don't use
 # autoloading
 LOOKUP		.*		MODLOAD
