--- rpm-4.2.3/lib/rpminstall.c.dont-install-delta-rpms	2005-01-13 11:56:02.000000000 +0100
+++ rpm-4.2.3/lib/rpminstall.c	2005-01-27 23:12:13.711991171 +0100
@@ -452,6 +452,7 @@ if (fileURL[0] == '=') {
 	 eiu->fnp++, eiu->prevx++)
     {
 	const char * fileName;
+	const char * payloadFormat;
 
 	rpmMessage(RPMMESS_DEBUG, "============== %s\n", *eiu->fnp);
 	(void) urlPath(*eiu->fnp, &fileName);
@@ -492,6 +493,17 @@ if (fileURL[0] == '=') {
 	    /*@switchbreak@*/ break;
 	}
 
+	payloadFormat = 0;
+	if (!headerGetEntry(eiu->h, RPMTAG_PAYLOADFORMAT, NULL, (void **)&payloadFormat, NULL))
+	    payloadFormat = 0;
+	if (payloadFormat && strcmp(payloadFormat, "cpio") != 0) {
+	    if (!strcmp(payloadFormat, "drpm"))
+		rpmMessage(RPMMESS_ERROR, _("%s is a deltarpm, create a real rpm from it first!\n"), *eiu->fnp);
+	    else
+		rpmMessage(RPMMESS_ERROR, _("%s contains no cpio payload\n"), *eiu->fnp);
+	    eiu->numFailed++; *eiu->fnp = NULL;
+	    continue;
+	}
 	eiu->isSource = headerIsEntry(eiu->h, RPMTAG_SOURCEPACKAGE);
 
 	if (eiu->isSource) {
