--- chkauth-0.1/chkauth.vdanen	2003-05-05 09:49:43.000000000 -0600
+++ chkauth-0.1/chkauth	2003-05-05 09:49:43.000000000 -0600
@@ -73,9 +73,9 @@
 password   sufficient     /lib/security/pam_unix.so nullok use_authtok md5 shadow use_first_pass
 password   required       /lib/security/pam_deny.so
  
-session     required      /lib/security/pam_limits.so
-session     sufficient      /lib/security/pam_unix.so
-session   required       /lib/security/pam_deny.so
+session    required       /lib/security/pam_limits.so
+session    sufficient     /lib/security/pam_unix.so
+session    required       /lib/security/pam_deny.so
 EOF
     ;
     }
@@ -85,23 +85,23 @@
  
 auth       required       /lib/security/pam_nologin.so
 auth       required       /lib/security/pam_env.so
-auth       sufficient     /lib/security/pam_ldap.so
 auth       sufficient     /lib/security/pam_unix.so likeauth nullok try_first_pass
+auth       sufficient     /lib/security/pam_ldap.so use_first_pass
 auth       required       /lib/security/pam_deny.so
  
-account    sufficient     /lib/security/pam_ldap.so
 account    sufficient     /lib/security/pam_unix.so
+account    sufficient     /lib/security/pam_ldap.so
 account    required       /lib/security/pam_deny.so
  
 password   required       /lib/security/pam_cracklib.so retry=3
-password  sufficient      /lib/security/pam_ldap.so
 password   sufficient     /lib/security/pam_unix.so nullok use_authtok md5 shadow use_first_pass
+password   sufficient     /lib/security/pam_ldap.so use_authtok
 password   required       /lib/security/pam_deny.so
- 
-session     required      /lib/security/pam_limits.so
-session     sufficient      /lib/security/pam_ldap.so
-session     sufficient      /lib/security/pam_unix.so
-session   required       /lib/security/pam_deny.so
+
+session    required       /lib/security/pam_limits.so
+session    sufficient     /lib/security/pam_unix.so
+session    sufficient     /lib/security/pam_ldap.so
+session    required       /lib/security/pam_deny.so
 EOF
     ;
     }
