diff -Naur ntp-4.2.0/arlib/configure.in ntp-4.2.0.oden/arlib/configure.in
--- ntp-4.2.0/arlib/configure.in	2003-07-17 12:27:11.000000000 +0200
+++ ntp-4.2.0.oden/arlib/configure.in	2004-06-27 18:17:06.062059750 +0200
@@ -6,7 +6,7 @@
 
 # Checks for programs.
 AC_PROG_CC
-AC_PROG_RANLIB
+AC_PROG_LIBTOOL
 
 # Checks for libraries.
 AC_CHECK_FUNC(gethostent, , AC_CHECK_LIB(nsl, gethostent, , , -lsocket))
iff -Naur ntp-4.2.0/configure.in ntp-4.2.0.oden/configure.in
--- ntp-4.2.0/configure.in	2004-06-27 18:17:17.598903520 +0200
+++ ntp-4.2.0.oden/configure.in	2004-06-27 18:16:57.307178310 +0200
@@ -177,7 +177,7 @@
 AC_PROG_GCC_TRADITIONAL
 AC_C_VOLATILE
 AC_ISC_POSIX
-AC_PROG_RANLIB
+AC_PROG_LIBTOOL
 AC_PATH_PROG(PATH_SH, sh)
 AC_PATH_PROG(PATH_PERL, perl)
 
