--- linux-2.6.1/include/asm-i386/string.h.force-inline-memcmp	2004-01-19 04:28:26.282777000 +0100
+++ linux-2.6.1/include/asm-i386/string.h	2004-01-19 12:48:40.429933216 +0100
@@ -313,7 +313,23 @@ __asm__ __volatile__(
 return dest;
 }
 
+#if defined(__OPTIMIZE_SIZE__)
+/* workaround newer gcc for now */
+#define __HAVE_ARCH_MEMCMP
+static inline __attribute__((always_inline))
+int memcmp(const void * cs,const void * ct,size_t count)
+{
+	const unsigned char *su1, *su2;
+	int res = 0;
+
+	for( su1 = cs, su2 = ct; 0 < count; ++su1, ++su2, count--)
+		if ((res = *su1 - *su2) != 0)
+			break;
+	return res;
+}
+#else
 #define memcmp __builtin_memcmp
+#endif
 
 #define __HAVE_ARCH_MEMCHR
 static inline void * memchr(const void * cs,int c,size_t count)
