diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/drivers/ide/pci/siimage.c w1/drivers/ide/pci/siimage.c
--- w2/drivers/ide/pci/siimage.c	2003-08-12 10:25:02.000000000 +0200
+++ w1/drivers/ide/pci/siimage.c	2003-08-19 00:48:23.000000000 +0200
@@ -62,6 +62,7 @@ static int pdev_is_sata(struct pci_dev *
 			return 0;
 	}
 	BUG();
+	return 0;
 }
  
 /**
