diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/fs/befs/btree.c w1/fs/befs/btree.c
--- w2/fs/befs/btree.c	2003-06-13 16:51:37.000000000 +0200
+++ w1/fs/befs/btree.c	2003-08-18 17:47:02.000000000 +0200
@@ -85,7 +85,7 @@ typedef struct {
 } befs_btree_node;
 
 /* local constants */
-const static befs_off_t befs_bt_inval = 0xffffffffffffffff;
+const static befs_off_t befs_bt_inval = 0xffffffffffffffffULL;
 
 /* local functions */
 static int befs_btree_seekleaf(struct super_block *sb, befs_data_stream * ds,
