diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/drivers/video/matrox/matroxfb_g450.c w1/drivers/video/matrox/matroxfb_g450.c
--- w2/drivers/video/matrox/matroxfb_g450.c	2003-06-13 16:51:37.000000000 +0200
+++ w1/drivers/video/matrox/matroxfb_g450.c	2003-08-18 17:44:30.000000000 +0200
@@ -558,7 +558,6 @@ static int matroxfb_g450_program(void* m
 }
 
 static int matroxfb_g450_verify_mode(void* md, u_int32_t arg) {
-	MINFO_FROM(md);
 	
 	switch (arg) {
 		case MATROXFB_OUTPUT_MODE_PAL:
