diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/drivers/char/ip2main.c w1/drivers/char/ip2main.c
--- w2/drivers/char/ip2main.c	2002-11-29 00:53:12.000000000 +0100
+++ w1/drivers/char/ip2main.c	2003-08-19 01:34:54.000000000 +0200
@@ -988,7 +988,7 @@ retry:
 static void __init
 ip2_init_board( int boardnum )
 {
-	int i,rc;
+	int i;
 	int nports = 0, nboxes = 0;
 	i2ChanStrPtr pCh;
 	i2eBordStrPtr pB = i2BoardPtrTable[boardnum];
