diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/drivers/usb/storage/sddr09.c w1/drivers/usb/storage/sddr09.c
--- w2/drivers/usb/storage/sddr09.c	2003-08-12 10:25:08.000000000 +0200
+++ w1/drivers/usb/storage/sddr09.c	2003-08-19 02:13:39.000000000 +0200
@@ -444,6 +444,7 @@ sddr09_test_unit_ready(struct us_data *u
  * byte 0: opcode: 03
  * byte 4: data length
  */
+#if 0
 static int
 sddr09_request_sense(struct us_data *us, unsigned char *sensebuf, int buflen) {
 	unsigned char command[12] = {
@@ -465,7 +466,7 @@ sddr09_request_sense(struct us_data *us,
 
 	return result;
 }
-
+#endif
 /*
  * Read Command: 12 bytes.
  * byte 0: opcode: E8
