diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/drivers/mtd/maps/amd76xrom.c w1/drivers/mtd/maps/amd76xrom.c
--- w2/drivers/mtd/maps/amd76xrom.c	2003-06-13 16:51:34.000000000 +0200
+++ w1/drivers/mtd/maps/amd76xrom.c	2003-08-19 01:43:35.000000000 +0200
@@ -178,7 +178,9 @@ err_out_iounmap:
 	iounmap((void *)(info->window_addr));
 err_out_free_mmio_region:
 	release_mem_region(window->start, window->size);
+#if 0
 err_out_none:
+#endif
 	return -ENODEV;
 }
 
