diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/fs/intermezzo/file.c w1/fs/intermezzo/file.c
--- w2/fs/intermezzo/file.c	2002-11-29 00:53:15.000000000 +0100
+++ w1/fs/intermezzo/file.c	2003-08-20 15:51:28.000000000 +0200
@@ -78,16 +78,19 @@ static int presto_open_upcall(int minor,
         pathlen = MYPATHLEN(buffer, path);
         
         CDEBUG(D_FILE, "de %p, dd %p\n", de, dd);
+
         if (dd->remote_ino == 0) {
                 rc = presto_get_fileid(minor, fset, de);
+                if (dd->remote_ino == 0)
+                        CERROR("get_fileid failed %d, ino: %Lx, fetching by name\n", rc,
+                               dd->remote_ino);
+
         }
         memset (&info, 0, sizeof(info));
         if (dd->remote_ino > 0) {
                 info.remote_ino = dd->remote_ino;
                 info.remote_generation = dd->remote_generation;
-        } else
-                CERROR("get_fileid failed %d, ino: %Lx, fetching by name\n", rc,
-                       dd->remote_ino);
+        }
 
         rc = izo_upc_open(minor, pathlen, path, fset->fset_name, &info);
         PRESTO_FREE(buffer, PAGE_SIZE);
