diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/fs/coda/dir.c w1/fs/coda/dir.c
--- w2/fs/coda/dir.c	2003-08-12 10:25:08.000000000 +0200
+++ w1/fs/coda/dir.c	2003-08-20 15:42:31.000000000 +0200
@@ -589,8 +589,7 @@ static int coda_venus_readdir(struct fil
 			break;
 		}
 		/* validate whether the directory file actually makes sense */
-		if (vdir->d_reclen < vdir_size + vdir->d_namlen ||
-		    vdir->d_namlen > CODA_MAXNAMLEN) {
+		if (vdir->d_reclen < vdir_size + vdir->d_namlen) {
 			printk("coda_venus_readdir: Invalid dir: %ld\n",
 			       filp->f_dentry->d_inode->i_ino);
 			ret = -EBADF;
