diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/drivers/net/fc/iph5526.c w1/drivers/net/fc/iph5526.c
--- w2/drivers/net/fc/iph5526.c	2003-06-13 16:51:34.000000000 +0200
+++ w1/drivers/net/fc/iph5526.c	2003-08-20 10:43:02.000000000 +0200
@@ -3595,12 +3595,8 @@ u_int service_params;
 			/* This helps to maintain the target_id no matter what your
 			 *  Fibre Channel ID is.
 			 */
-			if (temp->target_id == 0xFF) {
-				if (fi->g.no_of_targets <= MAX_SCSI_TARGETS)
-					temp->target_id = fi->g.no_of_targets++;
-				else
-					T_MSG("MAX TARGETS reached!");
-			}
+			if (temp->target_id == 0xFF)
+				temp->target_id = fi->g.no_of_targets++;
 			else
 				DPRINTK1("Target_id %d already present", temp->target_id);
 #if DEBUG_5526_2
