diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/drivers/scsi/AM53C974.c w1/drivers/scsi/AM53C974.c
--- w2/drivers/scsi/AM53C974.c	2001-09-30 21:26:07.000000000 +0200
+++ w1/drivers/scsi/AM53C974.c	2003-08-20 17:25:28.000000000 +0200
@@ -1612,7 +1612,7 @@ static void AM53C974_information_transfe
 		for (i = 0; i < sizeof(hostdata->last_message); i++)
 			hostdata->last_message[i] = hostdata->msgout[i];
 		if ((hostdata->msgout[0] == 0) || INSIDE(hostdata->msgout[0], 0x02, 0x1F) ||
-		    INSIDE(hostdata->msgout[0], 0x80, 0xFF))
+		    (hostdata->msgout[0] >= 0x80))
 			len = 1;
 		else {
 			if (hostdata->msgout[0] == EXTENDED_MESSAGE) {
