diff -urNp --exclude-from=/home/quintela/config/misc/dontdiff q2/arch/i386/boot/video.S q3/arch/i386/boot/video.S
--- q2/arch/i386/boot/video.S	2001-07-05 20:28:16.000000000 +0200
+++ q3/arch/i386/boot/video.S	2004-02-16 12:53:53.000000000 +0100
@@ -128,6 +128,7 @@ video:	pushw	%ds		# We use different seg
 
 	leaw	badmdt, %si			# Invalid mode ID
 	call	prtstr
+	jmp	vid1
 vid2:	call	mode_menu
 vid1:
 #ifdef CONFIG_VIDEO_RETAIN
@@ -536,7 +537,7 @@ check_vesa:
 	int	$0x10
 	addb	$VIDEO_FIRST_VESA>>8, %bh
 	cmpw	$0x004f, %ax
-	jnz	setbad
+	jnz	check_vesa_retry
 
 	movb	(%di), %al			# Check capabilities.
 	andb	$0x19, %al
@@ -546,20 +547,32 @@ check_vesa:
 	movb	(%di), %al			# Check capabilities.
 	andb	$0x99, %al
 	cmpb	$0x99, %al
-	jnz	_setbad				# Doh! No linear frame buffer.
-
+	jnz	check_vesa_retry		# Doh! No linear frame buffer.
+	pushw	%bx
+	
 	subb	$VIDEO_FIRST_VESA>>8, %bh
 	orw	$0x4000, %bx			# Use linear frame buffer
 	movw	$0x4f02, %ax			# VESA BIOS mode set call
 	int	$0x10
 	cmpw	$0x004f, %ax			# AL=4f if implemented
-	jnz	_setbad				# AH=0 if OK
+	popw	%bx
+	jnz	check_vesa_retry		# AH=0 if OK
 
 	movb	$1, graphic_mode		# flag graphic mode
 	movb	$0, do_restore			# no screen restore
 	stc
 	ret
 
+check_vesa_retry:
+	movw	%bx,%ax
+	cmpw	$0x120+VIDEO_FIRST_VESA, %ax
+	jae	_setbad
+	and	$0xfff0, %ax
+	cmpw	$0x110+VIDEO_FIRST_VESA, %bx
+	adc	$-3, %bx
+	cmpw	%ax, %bx
+	jae	check_vesa
+	
 _setbad:	jmp	setbad          	# Ugly...
 
 # Recalculate vertical display end registers -- this fixes various
@@ -1908,7 +1921,7 @@ prompt:		.byte	0x0d, 0x0a
 
 unknt:		.asciz	"Unknown mode ID. Try again."
 
-badmdt:		.ascii	"You passed an undefined mode number."
+badmdt:		.ascii	"Using VGA16 mode."
 		.byte	0x0d, 0x0a, 0
 
 vesaer:		.ascii	"Error: Scanning of VESA modes failed. Please "
