diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/fs/binfmt_elf.c w1/fs/binfmt_elf.c
--- w2/fs/binfmt_elf.c	2003-08-20 19:14:10.000000000 +0200
+++ w1/fs/binfmt_elf.c	2003-08-19 01:04:52.000000000 +0200
@@ -131,7 +131,7 @@ create_elf_tables(char *p, int argc, int
 	elf_caddr_t *envp;
 	elf_addr_t *sp, *csp;
 	char *stack_top;
-	char *k_platform, *u_platform;
+	char *k_platform, *u_platform = NULL;
 	long hwcap;
 	size_t platform_len = 0;
 	size_t len;
