diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/arch/i386/kernel/mpparse.c w1/arch/i386/kernel/mpparse.c
--- w2/arch/i386/kernel/mpparse.c	2003-08-29 13:54:53.000000000 +0200
+++ w1/arch/i386/kernel/mpparse.c	2003-08-29 16:34:44.000000000 +0200
@@ -1021,7 +1021,7 @@ static int __init mp_find_ioapic (
 			return i;
 	}
 
-	printk(KERN_ERR "ERROR: Unable to locate IOAPIC for IRQ %d/n", irq);
+	printk(KERN_ERR "ERROR: Unable to locate IOAPIC for IRQ %d\n", irq);
 
 	return -1;
 }
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/arch/x86_64/kernel/mpparse.c w1/arch/x86_64/kernel/mpparse.c
--- w2/arch/x86_64/kernel/mpparse.c	2003-08-29 13:55:17.000000000 +0200
+++ w1/arch/x86_64/kernel/mpparse.c	2003-08-29 18:42:33.000000000 +0200
@@ -713,7 +713,7 @@ static int __init mp_find_ioapic (
 			return i;
 	}
 
-	printk(KERN_ERR "ERROR: Unable to locate IOAPIC for IRQ %d/n", irq);
+	printk(KERN_ERR "ERROR: Unable to locate IOAPIC for IRQ %d\n", irq);
 
 	return -1;
 }
