diff -urN linux-2.4.18/drivers/char/pc_keyb.c lia64-2.4/drivers/char/pc_keyb.c
--- linux-2.4.18/drivers/char/pc_keyb.c	Mon Nov 26 11:18:32 2001
+++ lia64-2.4/drivers/char/pc_keyb.c	Fri Mar 29 17:16:46 2002
@@ -802,6 +802,15 @@
 {
 	int status;
 
+	/*
+	 * This is not really IA-64 specific.  Probably ought to be done on all platforms
+	 * that are (potentially) legacy-free.
+	 */
+	if (kbd_read_status() == 0xff && kbd_read_input() == 0xff) {
+		kbd_exists = 0;
+		return "No keyboard controller preset";
+	}
+
 	/*
 	 * Test the keyboard interface.
 	 * This seems to be the only way to get it going.
@@ -904,6 +915,8 @@
 		char *msg = initialize_kbd();
 		if (msg)
 			printk(KERN_WARNING "initialize_kbd: %s\n", msg);
+		if (!kbd_exists)
+			return;
 	}
 
 #if defined CONFIG_PSMOUSE
