diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/ciphers/cipher-3des.c w1/net/cryptoapi/ciphers/cipher-3des.c
--- w2/net/cryptoapi/ciphers/cipher-3des.c	2002-11-21 16:44:51.000000000 +0100
+++ w1/net/cryptoapi/ciphers/cipher-3des.c	2003-08-07 11:23:08.000000000 +0200
@@ -47,7 +47,7 @@
 #include <linux/errno.h>
 #include <linux/init.h>
 #include <asm/byteorder.h>
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 
 #ifdef MODULE_LICENSE
 MODULE_LICENSE ("GPL");
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/ciphers/cipher-aes.c w1/net/cryptoapi/ciphers/cipher-aes.c
--- w2/net/cryptoapi/ciphers/cipher-aes.c	2002-11-21 16:44:51.000000000 +0100
+++ w1/net/cryptoapi/ciphers/cipher-aes.c	2003-08-07 11:23:12.000000000 +0200
@@ -27,7 +27,7 @@
 #include <linux/init.h>
 #include <linux/types.h>
 #include <linux/wordops.h>
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 #include <asm/byteorder.h>
 
 #ifdef MODULE_LICENSE
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/ciphers/cipher-blowfish.c w1/net/cryptoapi/ciphers/cipher-blowfish.c
--- w2/net/cryptoapi/ciphers/cipher-blowfish.c	2002-11-21 16:44:51.000000000 +0100
+++ w1/net/cryptoapi/ciphers/cipher-blowfish.c	2003-08-07 11:23:16.000000000 +0200
@@ -6,7 +6,7 @@
 #include <linux/errno.h>
 #include <linux/init.h>
 #include <asm/byteorder.h>
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 
 #ifdef MODULE_LICENSE
 MODULE_LICENSE("GPL");
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/ciphers/cipher-blowfish_old.c w1/net/cryptoapi/ciphers/cipher-blowfish_old.c
--- w2/net/cryptoapi/ciphers/cipher-blowfish_old.c	2002-11-21 16:44:51.000000000 +0100
+++ w1/net/cryptoapi/ciphers/cipher-blowfish_old.c	2003-08-07 11:23:21.000000000 +0200
@@ -2,7 +2,7 @@
 #include <linux/errno.h>
 #include <linux/init.h>
 #include <asm/byteorder.h>
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 
 #ifdef MODULE_LICENSE
 MODULE_LICENSE("unknown");
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/ciphers/cipher-cast5.c w1/net/cryptoapi/ciphers/cipher-cast5.c
--- w2/net/cryptoapi/ciphers/cipher-cast5.c	2002-11-21 16:44:51.000000000 +0100
+++ w1/net/cryptoapi/ciphers/cipher-cast5.c	2003-08-07 11:23:27.000000000 +0200
@@ -20,7 +20,7 @@
 #include <linux/init.h>
 #include <linux/types.h>
 
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 
 #ifdef MODULE_LICENSE
 MODULE_LICENSE("GPL");
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/ciphers/cipher-des.c w1/net/cryptoapi/ciphers/cipher-des.c
--- w2/net/cryptoapi/ciphers/cipher-des.c	2002-11-21 16:44:51.000000000 +0100
+++ w1/net/cryptoapi/ciphers/cipher-des.c	2003-08-07 11:23:30.000000000 +0200
@@ -15,7 +15,7 @@
 #include <linux/errno.h>
 #include <linux/init.h>
 #include <asm/byteorder.h>
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 #include <linux/wordops.h>
 
 #ifdef MODULE_LICENSE
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/ciphers/cipher-dfc.c w1/net/cryptoapi/ciphers/cipher-dfc.c
--- w2/net/cryptoapi/ciphers/cipher-dfc.c	2002-11-21 16:44:51.000000000 +0100
+++ w1/net/cryptoapi/ciphers/cipher-dfc.c	2003-08-07 11:23:35.000000000 +0200
@@ -104,7 +104,7 @@ Mean:         1749 cycles =    14.6 mbit
 #include <linux/module.h>
 #include <linux/init.h>
 #include <linux/types.h>
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 #include <asm/byteorder.h>
 
 #ifdef MODULE_LICENSE
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/ciphers/cipher-gost.c w1/net/cryptoapi/ciphers/cipher-gost.c
--- w2/net/cryptoapi/ciphers/cipher-gost.c	2002-11-21 16:44:51.000000000 +0100
+++ w1/net/cryptoapi/ciphers/cipher-gost.c	2003-08-07 11:23:38.000000000 +0200
@@ -29,7 +29,7 @@
 #include <linux/errno.h>
 #include <linux/init.h>
 
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 #include <linux/wordops.h>
 
 #include <asm/byteorder.h>
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/ciphers/cipher-idea.c w1/net/cryptoapi/ciphers/cipher-idea.c
--- w2/net/cryptoapi/ciphers/cipher-idea.c	2002-11-21 16:44:51.000000000 +0100
+++ w1/net/cryptoapi/ciphers/cipher-idea.c	2003-08-07 11:23:43.000000000 +0200
@@ -41,7 +41,7 @@
 #include <linux/errno.h>
 #include <linux/init.h>
 #include <asm/byteorder.h>
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 #include <linux/string.h>
 
 #ifdef MODULE_LICENSE
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/ciphers/cipher-mars.c w1/net/cryptoapi/ciphers/cipher-mars.c
--- w2/net/cryptoapi/ciphers/cipher-mars.c	2002-11-21 16:44:51.000000000 +0100
+++ w1/net/cryptoapi/ciphers/cipher-mars.c	2003-08-07 11:23:49.000000000 +0200
@@ -31,7 +31,7 @@
 #include <linux/init.h>
 #include <linux/types.h>
 #include <linux/wordops.h>
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 
 #ifdef MODULE_LICENSE
 MODULE_LICENSE("Dual BSD/GPL");
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/ciphers/cipher-null.c w1/net/cryptoapi/ciphers/cipher-null.c
--- w2/net/cryptoapi/ciphers/cipher-null.c	2002-11-21 16:44:51.000000000 +0100
+++ w1/net/cryptoapi/ciphers/cipher-null.c	2003-08-07 11:23:53.000000000 +0200
@@ -22,7 +22,7 @@
  */
 
 #include <asm/byteorder.h>
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 #include <linux/errno.h>
 #include <linux/init.h>
 #include <linux/module.h>
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/ciphers/cipher-rc5.c w1/net/cryptoapi/ciphers/cipher-rc5.c
--- w2/net/cryptoapi/ciphers/cipher-rc5.c	2002-11-21 16:44:51.000000000 +0100
+++ w1/net/cryptoapi/ciphers/cipher-rc5.c	2003-08-07 11:23:59.000000000 +0200
@@ -38,7 +38,7 @@
 #include <linux/init.h>  
 #include <linux/string.h> 
 
-#include <linux/crypto.h> 
+#include <linux/cryptoapi.h> 
 #include <linux/wordops.h> 
 
 #ifdef MODULE_LICENSE
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/ciphers/cipher-rc6.c w1/net/cryptoapi/ciphers/cipher-rc6.c
--- w2/net/cryptoapi/ciphers/cipher-rc6.c	2002-11-21 16:44:51.000000000 +0100
+++ w1/net/cryptoapi/ciphers/cipher-rc6.c	2003-08-07 11:24:04.000000000 +0200
@@ -20,7 +20,7 @@
 #include <linux/init.h>
 #include <linux/types.h>
 #include <linux/wordops.h>
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 
 #ifdef MODULE_LICENSE
 MODULE_LICENSE("Dual BSD/GPL");
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/ciphers/cipher-serpent.c w1/net/cryptoapi/ciphers/cipher-serpent.c
--- w2/net/cryptoapi/ciphers/cipher-serpent.c	2002-11-21 16:44:51.000000000 +0100
+++ w1/net/cryptoapi/ciphers/cipher-serpent.c	2003-08-07 11:24:08.000000000 +0200
@@ -19,7 +19,7 @@
 #include <linux/init.h>
 #include <linux/types.h>
 
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 #include <linux/wordops.h>
 
 #include <asm/byteorder.h>
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/ciphers/cipher-twofish.c w1/net/cryptoapi/ciphers/cipher-twofish.c
--- w2/net/cryptoapi/ciphers/cipher-twofish.c	2002-11-21 16:44:51.000000000 +0100
+++ w1/net/cryptoapi/ciphers/cipher-twofish.c	2003-08-07 11:24:13.000000000 +0200
@@ -29,7 +29,7 @@
 #include <linux/module.h>
 #include <linux/init.h>
 #include <linux/types.h>
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 
 #ifdef MODULE_LICENSE
 MODULE_LICENSE("GPL");
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/Config.in w1/net/cryptoapi/Config.in
--- w2/net/cryptoapi/Config.in	2003-08-07 10:29:32.000000000 +0200
+++ w1/net/cryptoapi/Config.in	2003-08-07 10:41:07.000000000 +0200
@@ -6,13 +6,13 @@ comment 'Cryptography support (CryptoAPI
 tristate 'CryptoAPI support' CONFIG_CRYPTOAPI
 
 dep_mbool 'Cipher Algorithms' CONFIG_CIPHERS $CONFIG_CRYPTOAPI
-source net/crypto/ciphers/Config.in
+source net/cryptoapi/ciphers/Config.in
 
 dep_mbool 'Digest Algorithms' CONFIG_DIGESTS $CONFIG_CRYPTOAPI
-source net/crypto/digests/Config.in
+source net/cryptoapi/digests/Config.in
 
 dep_mbool 'Crypto Devices' CONFIG_CRYPTODEV $CONFIG_CRYPTOAPI
-source net/crypto/drivers/Config.in
+source net/cryptoapi/drivers/Config.in
 
 endmenu
 
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/cryptoapi.c w1/net/cryptoapi/cryptoapi.c
--- w2/net/cryptoapi/cryptoapi.c	2002-11-21 16:44:51.000000000 +0100
+++ w1/net/cryptoapi/cryptoapi.c	2003-08-07 11:22:05.000000000 +0200
@@ -23,7 +23,7 @@
 #include <linux/init.h>
 #include <linux/config.h>
 #include <linux/string.h>
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 #include <linux/proc_fs.h>
 
 #ifdef CONFIG_KMOD
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/digests/digest-md5.c w1/net/cryptoapi/digests/digest-md5.c
--- w2/net/cryptoapi/digests/digest-md5.c	2002-11-21 16:44:52.000000000 +0100
+++ w1/net/cryptoapi/digests/digest-md5.c	2003-08-07 11:24:22.000000000 +0200
@@ -28,7 +28,7 @@
 #include <linux/init.h>
 #include <linux/string.h>
 
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 
 #include <asm/uaccess.h>
 #include <asm/byteorder.h>
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/digests/digest-ripemd160.c w1/net/cryptoapi/digests/digest-ripemd160.c
--- w2/net/cryptoapi/digests/digest-ripemd160.c	2002-11-21 16:44:52.000000000 +0100
+++ w1/net/cryptoapi/digests/digest-ripemd160.c	2003-08-07 11:24:25.000000000 +0200
@@ -19,7 +19,7 @@
 
 #include <asm/uaccess.h>
 
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 #include <linux/wordops.h>
 
 #ifdef MODULE_LICENSE
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/digests/digest-sha1.c w1/net/cryptoapi/digests/digest-sha1.c
--- w2/net/cryptoapi/digests/digest-sha1.c	2002-11-21 16:44:52.000000000 +0100
+++ w1/net/cryptoapi/digests/digest-sha1.c	2003-08-07 11:24:30.000000000 +0200
@@ -15,7 +15,7 @@
 #include <linux/module.h>
 #include <linux/init.h>
 #include <linux/types.h>
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 
 #include <asm/uaccess.h>
 #include <asm/byteorder.h>
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/digests/digest-sha256.c w1/net/cryptoapi/digests/digest-sha256.c
--- w2/net/cryptoapi/digests/digest-sha256.c	2002-11-21 16:44:52.000000000 +0100
+++ w1/net/cryptoapi/digests/digest-sha256.c	2003-08-07 11:24:33.000000000 +0200
@@ -19,7 +19,7 @@
 
 #include <asm/uaccess.h>
 
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 #include <linux/wordops.h>
 
 #ifdef MODULE_LICENSE
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/digests/digest-sha384.c w1/net/cryptoapi/digests/digest-sha384.c
--- w2/net/cryptoapi/digests/digest-sha384.c	2002-11-21 16:44:52.000000000 +0100
+++ w1/net/cryptoapi/digests/digest-sha384.c	2003-08-07 11:24:37.000000000 +0200
@@ -19,7 +19,7 @@
 
 #include <asm/uaccess.h>
 
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 #include <linux/wordops.h>
 
 #ifdef MODULE_LICENSE
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff w2/net/cryptoapi/digests/digest-sha512.c w1/net/cryptoapi/digests/digest-sha512.c
--- w2/net/cryptoapi/digests/digest-sha512.c	2002-11-21 16:44:52.000000000 +0100
+++ w1/net/cryptoapi/digests/digest-sha512.c	2003-08-07 11:24:40.000000000 +0200
@@ -19,7 +19,7 @@
 
 #include <asm/uaccess.h>
 
-#include <linux/crypto.h>
+#include <linux/cryptoapi.h>
 #include <linux/wordops.h>
 
 #ifdef MODULE_LICENSE

