diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff orig/net/ipsec/ipsec_tunnel.c q5/net/ipsec/ipsec_tunnel.c
--- orig/net/ipsec/ipsec_tunnel.c	2002-08-23 15:51:20.000000000 +0200
+++ q5/net/ipsec/ipsec_tunnel.c	2002-08-22 23:16:33.000000000 +0200
@@ -990,7 +990,7 @@ ipsec_tunnel_start_xmit(struct sk_buff *
 				}
 				stats->tx_dropped++;
 			}
-			default:
+			default:;
 				/* XXX what do we do with an unknown shunt spi? */
 			} /* switch (ntohl(outgoing_said.spi)) */
 			goto cleanup;
@@ -1724,7 +1724,7 @@ ipsec_tunnel_start_xmit(struct sk_buff *
 				case 1:
 					iph->tos = 0;
 					break;
-				default:
+				default:;
 				}
 #ifdef NET_21
 #ifdef NETDEV_23
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff orig/net/ipsec/libfreeswan/pfkey_v2_parse.c q5/net/ipsec/libfreeswan/pfkey_v2_parse.c
--- orig/net/ipsec/libfreeswan/pfkey_v2_parse.c	2002-08-23 15:52:25.000000000 +0200
+++ q5/net/ipsec/libfreeswan/pfkey_v2_parse.c	2002-08-22 23:13:07.000000000 +0200
@@ -1198,7 +1198,7 @@ pfkey_msg_parse(struct sadb_msg *pfkey_m
 				  pfkey_v2_sadb_type_string(pfkey_msg->sadb_msg_type));
 			SENDERR(EINVAL);
 		}
-	default:
+	default:;
 	}
 	
 	/* errno must not be set in downward messages */
@@ -1445,7 +1445,7 @@ pfkey_msg_parse(struct sadb_msg *pfkey_m
 				SENDERR(EINVAL);
 			}
 			break;
-		default:
+		default:;
 		}
 		if(ntohl(((struct sadb_sa*)(extensions[SADB_EXT_SA]))->sadb_sa_spi) <= 255) {
 			DEBUGGING(PF_KEY_DEBUG_PARSE_PROBLEM,
@@ -1454,7 +1454,7 @@ pfkey_msg_parse(struct sadb_msg *pfkey_m
 				ntohl(((struct sadb_sa*)(extensions[SADB_EXT_SA]))->sadb_sa_spi));
 			SENDERR(EINVAL);
 		}
-	default:	
+	default:;
 	}
 errlab:
 
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff orig/net/ipsec/pfkey_v2_parser.c q5/net/ipsec/pfkey_v2_parser.c
--- orig/net/ipsec/pfkey_v2_parser.c	2002-08-23 15:51:20.000000000 +0200
+++ q5/net/ipsec/pfkey_v2_parser.c	2002-08-22 23:16:04.000000000 +0200
@@ -474,7 +474,7 @@ pfkey_address_process(struct sadb_ext *p
 				    "uh, tdb_said.dst doesn't do address family=%d yet, said will be invalid.\n",
 				    s->sa_family);
 		}
-	default:
+	default:;
 	}
 	
 	/* XXX check if port!=0 */
@@ -3659,7 +3659,7 @@ pfkey_msg_interp(struct sock *sk, struct
 				    extr.tdb->tdb_said.proto);
 		}
 		break;
-	default:
+	default:;
 	}
 	
 	/* The NULL below causes the default extension parsers to be used */
