--- linux-2.4.24/drivers/media/video/bt832.c.dl07.orig	2003-06-13 16:51:34.000000000 +0200
+++ linux-2.4.24/drivers/media/video/bt832.c	2004-01-19 19:37:48.104119504 +0100
@@ -257,6 +257,7 @@ bt832_command(struct i2c_client *client,
 /* ----------------------------------------------------------------------- */
 
 static struct i2c_driver driver = {
+        .owner          = THIS_MODULE,
         .name           = "i2c bt832 driver",
         .id             = -1, /* FIXME */
         .flags          = I2C_DF_NOTIFY,
--- linux-2.4.24/drivers/media/video/msp3400.c.dl07.orig	2004-01-19 19:24:01.245821072 +0100
+++ linux-2.4.24/drivers/media/video/msp3400.c	2004-01-19 19:37:48.106119200 +0100
@@ -1257,6 +1257,7 @@ static int msp_probe(struct i2c_adapter 
 static int msp_command(struct i2c_client *client, unsigned int cmd, void *arg);
 
 static struct i2c_driver driver = {
+        .owner          = THIS_MODULE,
         .name           = "i2c msp3400 driver",
         .id             = I2C_DRIVERID_MSP3400,
         .flags          = I2C_DF_NOTIFY,
--- linux-2.4.24/drivers/media/video/tda7432.c.dl07.orig	2003-11-28 19:26:20.000000000 +0100
+++ linux-2.4.24/drivers/media/video/tda7432.c	2004-01-19 19:37:54.402162056 +0100
@@ -520,6 +520,7 @@ static int tda7432_command(struct i2c_cl
 }
 
 static struct i2c_driver driver = {
+        .owner           = THIS_MODULE,
         .name            = "i2c tda7432 driver",
 	.id              = I2C_DRIVERID_TDA7432,
         .flags           = I2C_DF_NOTIFY,
--- linux-2.4.24/drivers/media/video/tda9875.c.dl07.orig	2003-11-28 19:26:20.000000000 +0100
+++ linux-2.4.24/drivers/media/video/tda9875.c	2004-01-19 19:37:54.402162056 +0100
@@ -391,6 +391,7 @@ static int tda9875_command(struct i2c_cl
 
 
 static struct i2c_driver driver = {
+        .owner          = THIS_MODULE,
         .name           = "i2c tda9875 driver",
         .id             = I2C_DRIVERID_TDA9875,
         .flags          = I2C_DF_NOTIFY,
