diff -ru linux-2.4.21-orig/drivers/pcmcia/sa1100_stork.c linux-2.4.21-i2c/drivers/pcmcia/sa1100_stork.c
--- linux-2.4.21-orig/drivers/pcmcia/sa1100_stork.c	2003-06-16 09:30:01.000000000 +0200
+++ linux-2.4.21-i2c/drivers/pcmcia/sa1100_stork.c	2003-07-10 12:15:01.000000000 +0200
@@ -24,7 +24,6 @@
 #include <linux/init.h>
 #include <linux/kernel.h>
 #include <linux/sched.h>
-#include <linux/i2c.h>
 
 #include <asm/hardware.h>
 #include <asm/irq.h>
