--- linux-2.4.21/drivers/video/matrox/i2c-matroxfb.c.dv20.orig	2003-06-13 16:51:37.000000000 +0200
+++ linux-2.4.21/drivers/video/matrox/i2c-matroxfb.c	2003-08-24 01:21:54.000000000 +0200
@@ -87,29 +87,21 @@ static int matroxfb_gpio_getscl(void* da
 	return (matroxfb_read_gpio(b->minfo) & b->mask.clock) ? 1 : 0;
 }
 
-static void matroxfb_dh_inc_use(struct i2c_adapter* dummy) {
-	MOD_INC_USE_COUNT;
-}
-
-static void matroxfb_dh_dec_use(struct i2c_adapter* dummy) {
-	MOD_DEC_USE_COUNT;
-}
-
 static struct i2c_adapter matrox_i2c_adapter_template =
 {
+	.owner	=	THIS_MODULE,
 	.id =		I2C_HW_B_G400,
-	.inc_use =	matroxfb_dh_inc_use,
-	.dec_use =	matroxfb_dh_dec_use,
 };
 
 static struct i2c_algo_bit_data matrox_i2c_algo_template =
 {
-	NULL,
-	matroxfb_gpio_setsda,
-	matroxfb_gpio_setscl,
-	matroxfb_gpio_getsda,
-	matroxfb_gpio_getscl,
-	10, 10, 100,
+	.setsda	= matroxfb_gpio_setsda,
+	.setscl	= matroxfb_gpio_setscl,
+	.getsda	= matroxfb_gpio_getsda,
+	.getscl	= matroxfb_gpio_getscl,
+	.udelay	= 10,
+	.mdelay	= 10,
+	.timeout	= 100,
 };
 
 static int i2c_bus_reg(struct i2c_bit_adapter* b, struct matrox_fb_info* minfo, 
--- linux-2.4.21/drivers/video/matrox/matroxfb_maven.c.dv20.orig	2003-06-13 16:51:37.000000000 +0200
+++ linux-2.4.21/drivers/video/matrox/matroxfb_maven.c	2003-08-24 01:21:54.000000000 +0200
@@ -1246,14 +1246,6 @@ static unsigned short normal_i2c[] = { M
 static unsigned short normal_i2c_range[] = { MAVEN_I2CID, MAVEN_I2CID, I2C_CLIENT_END };
 I2C_CLIENT_INSMOD;
 
-static void maven_inc_use(struct i2c_client* clnt) {
-	MOD_INC_USE_COUNT;
-}
-
-static void maven_dec_use(struct i2c_client* clnt) {
-	MOD_DEC_USE_COUNT;
-}
-
 static struct i2c_driver maven_driver;
 
 static int maven_detect_client(struct i2c_adapter* adapter, int address, unsigned short flags,
@@ -1320,14 +1312,13 @@ static int maven_command(struct i2c_clie
 static int maven_driver_registered = 0;
 
 static struct i2c_driver maven_driver={
-	"maven",
-	I2C_DRIVERID_MGATVO,
-	I2C_DF_NOTIFY,
-	maven_attach_adapter,
-	maven_detach_client,
-	maven_command,
-	maven_inc_use,
-	maven_dec_use
+	.owner	= THIS_MODULE,
+	.name	= "maven",
+	.id	= I2C_DRIVERID_MGATVO,
+	.flags	= I2C_DF_NOTIFY,
+	.attach_adapter	= maven_attach_adapter,
+	.detach_client	= maven_detach_client,
+	.command	= maven_command,
 };
 
 /* ************************** */
--- linux-2.4.21/drivers/usb/bt819-new.c.dv20.orig	2003-08-24 01:21:53.000000000 +0200
+++ linux-2.4.21/drivers/usb/bt819-new.c	2003-08-24 01:21:54.000000000 +0200
@@ -558,21 +558,19 @@ static int bt819_command(struct i2c_clie
 /* ----------------------------------------------------------------------- */
 
 static struct i2c_driver driver = {
-	"i2c bt819 driver",		/* name */
-	I2C_DRIVERID_BT819,	/* ID */
-	I2C_DF_NOTIFY, 
-	bt819_probe,
-	bt819_detach,
-	bt819_command
+	.owner	= THIS_MODULE,
+	.name	="i2c bt819 driver",
+	.id	= I2C_DRIVERID_BT819,
+	.flags	= I2C_DF_NOTIFY, 
+	.attach_adapter	= bt819_probe,
+	.detach_client	= bt819_detach,
+	.command	= bt819_command
 };
 
 static struct i2c_client client_template = {
-	"(unset)",		/* name */
-	-1,
-	0,
-	0,
-	NULL,
-	&driver
+	.name	= "(unset)",
+	.id	= -1,
+	.driver	= &driver
 };
 
 EXPORT_NO_SYMBOLS;
--- linux-2.4.21/drivers/usb/i2c-algo-usb.c.dv20.orig	2003-08-24 01:21:53.000000000 +0200
+++ linux-2.4.21/drivers/usb/i2c-algo-usb.c	2003-08-24 01:21:54.000000000 +0200
@@ -163,14 +163,12 @@ static u32 usb_func(struct i2c_adapter *
 /* -----exported algorithm data: -------------------------------------	*/
 
 static struct i2c_algorithm i2c_usb_algo = {
-	"USB algorithm",
-	I2C_ALGO_BIT,		/* FIXME */
-	usb_xfer,
-	NULL,
-	NULL,			/* slave_xmit           */
-	NULL,			/* slave_recv           */
-	algo_control,		/* ioctl                */
-	usb_func,		/* functionality        */
+	.owner	= THIS_MODULE,
+	.name	= "USB algorithm",
+	.id	= I2C_ALGO_BIT,
+	.master_xfer	= usb_xfer,
+	.algo_control	= algo_control,		/* ioctl                */
+	.functionality	= usb_func,		/* functionality        */
 };
 
 /* 
--- linux-2.4.21/drivers/usb/saa7111-new.c.dv20.orig	2003-08-24 01:21:53.000000000 +0200
+++ linux-2.4.21/drivers/usb/saa7111-new.c	2003-08-24 01:21:54.000000000 +0200
@@ -459,21 +459,19 @@ static int saa7111_command(struct i2c_cl
 /* ----------------------------------------------------------------------- */
 
 static struct i2c_driver saa7111_driver = {
-	"SAA7111",
-	I2C_DRIVERID_SAA7111A,	/* ????? */
-	I2C_DF_NOTIFY,
-	saa7111_probe,
-	saa7111_detach,
-	saa7111_command,
+	.owner	= THIS_MODULE,
+	.name	= "SAA7111",
+	.id	= I2C_DRIVERID_SAA7111A,
+	.flags	= I2C_DF_NOTIFY,
+	.attach_adapter	= saa7111_probe,
+	.detach_client	= saa7111_detach,
+	.command	= saa7111_command,
 };
 
 static struct i2c_client client_template = {
-	"(unset)",		/* name */
-	I2C_DRIVERID_SAA7111A,
-	0,
-	0,
-	NULL,
-	&saa7111_driver
+	.name	= "(unset)",
+	.id	= I2C_DRIVERID_SAA7111A,
+	.driver	= &saa7111_driver
 };
 
 
--- linux-2.4.21/drivers/usb/saa7113-new.c.dv20.orig	2003-08-24 01:21:53.000000000 +0200
+++ linux-2.4.21/drivers/usb/saa7113-new.c	2003-08-24 01:21:54.000000000 +0200
@@ -456,36 +456,20 @@ static int saa7113_command(struct i2c_cl
 } 
 
 
-void inc_use (struct i2c_client *client)
-{
-#ifdef MODULE
-        MOD_INC_USE_COUNT;
-#endif
-}
-
-void dec_use (struct i2c_client *client)
-{
-#ifdef MODULE
-        MOD_DEC_USE_COUNT;
-#endif
-}
-
 static struct i2c_driver saa7113_driver = {
-        "SAA7113",
-        I2C_DRIVERID_SAA7113,
-        I2C_DF_NOTIFY,
-        saa7113_probe,
-        saa7113_detach,
-        saa7113_command,
+        .owner	= THIS_MODULE,
+        .name	= "SAA7113",
+        .id	= I2C_DRIVERID_SAA7113,
+        .flags	= I2C_DF_NOTIFY,
+        .attach_adapter	= saa7113_probe,
+	.detach_client	= saa7113_detach,
+        .command	= saa7113_command,
 };
 
 static struct i2c_client client_template = {
-        "(unset)",
-        I2C_DRIVERID_SAA7113,
-        0,
-        0,
-        NULL,
-        &saa7113_driver,
+        .name	= "(unset)",
+        .id	= I2C_DRIVERID_SAA7113,
+        .driver	= &saa7113_driver,
 };
 
 
--- linux-2.4.21/drivers/usb/usbvision.c.dv20.orig	2003-08-24 01:21:53.000000000 +0200
+++ linux-2.4.21/drivers/usb/usbvision.c	2003-08-24 01:21:54.000000000 +0200
@@ -1814,16 +1814,6 @@ static int usbvision_write_reg(struct us
 /* I2C functions                                                           */
 /* ----------------------------------------------------------------------- */
 
-static void usbvision_inc_use(struct i2c_adapter *adap)
-{
-	MOD_INC_USE_COUNT;
-}
-
-static void usbvision_dec_use(struct i2c_adapter *adap)
-{
-	MOD_DEC_USE_COUNT;
-}
-
 static void call_i2c_clients(struct usb_usbvision *usbvision, unsigned int cmd,
 			     void *arg)
 {
@@ -1840,54 +1830,6 @@ static void call_i2c_clients(struct usb_
 	}
 }
 
-static int attach_inform(struct i2c_client *client)
-{
-	struct usb_usbvision *usbvision = (struct usb_usbvision *) client->adapter->data;
-	int i;
-
-	for (i = 0; i < I2C_CLIENTS_MAX; i++) {
-		if (usbvision->i2c_clients[i] == NULL ||
-		    usbvision->i2c_clients[i]->driver->id ==
-		    client->driver->id) {
-			usbvision->i2c_clients[i] = client;
-			break;
-		}
-	}
-
-	if ((usbvision->have_tuner) && (usbvision->tuner_type != -1)) {
-			
-		call_i2c_clients(usbvision, TUNER_SET_TYPE, &usbvision->tuner_type);
-		call_i2c_clients(usbvision, VIDIOCSCHAN, &usbvision->vchan);
-	}
-	call_i2c_clients(usbvision, DECODER_SET_PICTURE, &usbvision->vpic);
-/*	usbvision_muxsel(usbvision, usbvision->channel, usbvision->vchan.norm);
-*/
-	if (verbose)
-		printk("usbvision: device %d: i2c attach [%s]\n", usbvision->nr,
-		       client->name);
-
-	return 0;
-}
-
-static int detach_inform(struct i2c_client *client)
-{
-	struct usb_usbvision *usbvision =
-	    (struct usb_usbvision *) client->adapter->data;
-	int i;
-
-	if (verbose)
-		printk("usbvision: device %d: i2c detach [%s]\n", usbvision->nr,
-		       client->name);
-	for (i = 0; i < I2C_CLIENTS_MAX; i++) {
-		if (NULL != usbvision->i2c_clients[i] &&
-		    usbvision->i2c_clients[i]->driver->id ==
-		    client->driver->id) {
-			usbvision->i2c_clients[i] = NULL;
-			break;
-		}
-	}
-	return 0;
-}
 
 static int
 usbvision_i2c_read_max4(struct usb_usbvision *usbvision, unsigned char addr,
@@ -2057,31 +1999,22 @@ static int usbvision_i2c_read(void *data
 }
 
 static struct i2c_algo_usb_data i2c_algo_template = {
-	NULL,
-	usbvision_i2c_read,
-	usbvision_i2c_write,
-	10, 10, 100,
+	.inb	= usbvision_i2c_read,
+	.outb	= usbvision_i2c_write,
+	.udelay	= 10,
+	.mdelay	= 10,
+	.timeout	= 100,
 };
 
 static struct i2c_adapter i2c_adap_template = {
-	"usbvision",
-	I2C_HW_B_BT848,		/* FIXME */
-	NULL,
-	NULL,
-	usbvision_inc_use,
-	usbvision_dec_use,
-	attach_inform,
-	detach_inform,
-	NULL,
+	.owner	= THIS_MODULE,
+	.name	= "usbvision",
+	.id	= I2C_HW_B_BT848,
 };
 
 static struct i2c_client i2c_client_template = {
-	"usbvision internal",
-	-1,
-	0,
-	0,
-	NULL,
-	NULL
+	.name	= "usbvision internal",
+	.id	= -1,
 };
 
 static int usbvision_init_i2c(struct usb_usbvision *usbvision)
--- linux-2.4.21/drivers/media/video/bttv9/bttv9-if.c.dv20.orig	2003-04-16 07:59:22.000000000 +0200
+++ linux-2.4.21/drivers/media/video/bttv9/bttv9-if.c	2003-08-24 01:21:54.000000000 +0200
@@ -191,18 +191,6 @@ static int bttv_bit_getsda(void *data)
 	return state;
 }
 
-#if LINUX_VERSION_CODE < KERNEL_VERSION(2,5,54)
-static void bttv_inc_use(struct i2c_adapter *adap)
-{
-	MOD_INC_USE_COUNT;
-}
-
-static void bttv_dec_use(struct i2c_adapter *adap)
-{
-	MOD_DEC_USE_COUNT;
-}
-#endif
-
 static int attach_inform(struct i2c_client *client)
 {
         struct bttv *btv = i2c_get_adapdata(client->adapter);
@@ -242,12 +230,7 @@ static struct i2c_algo_bit_data bttv_i2c
 };
 
 static struct i2c_adapter bttv_i2c_adap_template = {
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,5,54)
 	.owner             = THIS_MODULE,
-#else
-	.inc_use           = bttv_inc_use,
-	.dec_use           = bttv_dec_use,
-#endif
 	I2C_DEVNAME("bt848"),
 	.id                = I2C_HW_B_BT848,
 	.client_register   = attach_inform,
--- linux-2.4.21/drivers/media/video/adv7175.c.dv20.orig	2001-09-30 21:26:06.000000000 +0200
+++ linux-2.4.21/drivers/media/video/adv7175.c	2003-08-24 01:21:54.000000000 +0200
@@ -439,23 +439,25 @@ static int adv7175_command(struct i2c_de
 /* ----------------------------------------------------------------------- */
 
 static struct i2c_driver i2c_driver_adv7175 = {
-	"adv7175",		/* name */
-	I2C_DRIVERID_VIDEOENCODER,	/* ID */
-	I2C_ADV7175, I2C_ADV7175 + 3,
-
-	adv7175_attach,
-	adv7175_detach,
-	adv7175_command
+	.name	= "adv7175",
+	.id	= I2C_DRIVERID_VIDEOENCODER,
+	.addr_l	= I2C_ADV7175,
+	.addr_h = I2C_ADV7175 + 3,
+
+	.attach	= adv7175_attach,
+	.detach	= adv7175_detach,
+	.command	= adv7175_command
 };
 
 static struct i2c_driver i2c_driver_adv7176 = {
-	"adv7175",		/* name */
-	I2C_DRIVERID_VIDEOENCODER,	/* ID */
-	I2C_ADV7176, I2C_ADV7176 + 3,
-
-	adv7175_attach,
-	adv7175_detach,
-	adv7175_command
+	.name	= "adv7175",
+	.id	= I2C_DRIVERID_VIDEOENCODER,
+	.addr_l	= I2C_ADV7176,
+	.addr_h	= I2C_ADV7176 + 3,
+
+	.attach	= adv7175_attach,
+	.detach	= adv7175_detach,
+	.command	= adv7175_command
 };
 
 EXPORT_NO_SYMBOLS;
--- linux-2.4.21/drivers/media/video/bt819.c.dv20.orig	2001-09-30 21:26:06.000000000 +0200
+++ linux-2.4.21/drivers/media/video/bt819.c	2003-08-24 01:21:54.000000000 +0200
@@ -471,13 +471,14 @@ static int bt819_command(struct i2c_devi
 /* ----------------------------------------------------------------------- */
 
 static struct i2c_driver i2c_driver_bt819 = {
-	"bt819",		/* name */
-	I2C_DRIVERID_VIDEODECODER,	/* ID */
-	I2C_BT819, I2C_BT819 + 1,
-
-	bt819_attach,
-	bt819_detach,
-	bt819_command
+	.name	= "bt819",		/* name */
+	.id	= I2C_DRIVERID_VIDEODECODER,	/* ID */
+	.addr_l	= I2C_BT819,
+	.addr_h	= I2C_BT819 + 1,
+
+	.attach	= bt819_attach,
+	.detach	= bt819_detach,
+	.command	= bt819_command
 };
 
 EXPORT_NO_SYMBOLS;
--- linux-2.4.21/drivers/media/video/bt856.c.dv20.orig	2001-09-30 21:26:06.000000000 +0200
+++ linux-2.4.21/drivers/media/video/bt856.c	2003-08-24 01:21:54.000000000 +0200
@@ -279,12 +279,13 @@ static int bt856_command(struct i2c_devi
 /* ----------------------------------------------------------------------- */
 
 static struct i2c_driver i2c_driver_bt856 = {
-	"bt856",		/* name */
-	I2C_DRIVERID_VIDEOENCODER,	/* ID */
-	I2C_BT856, I2C_BT856 + 1,
-	bt856_attach,
-	bt856_detach,
-	bt856_command
+	.name	= "bt856",		/* name */
+	.id	= I2C_DRIVERID_VIDEOENCODER,	/* ID */
+	.addr_l	= I2C_BT856,
+	.addr_h	= I2C_BT856 + 1,
+	.attach = bt856_attach,
+	.detach	= bt856_detach,
+	.command	= bt856_command
 };
 
 EXPORT_NO_SYMBOLS;
--- linux-2.4.21/drivers/media/video/bttv-if.c.dv20.orig	2003-08-24 00:54:49.000000000 +0200
+++ linux-2.4.21/drivers/media/video/bttv-if.c	2003-08-24 01:21:54.000000000 +0200
@@ -190,16 +190,6 @@ static int bttv_bit_getsda(void *data)
 	return state;
 }
 
-static void bttv_inc_use(struct i2c_adapter *adap)
-{
-	MOD_INC_USE_COUNT;
-}
-
-static void bttv_dec_use(struct i2c_adapter *adap)
-{
-	MOD_DEC_USE_COUNT;
-}
-
 static int attach_inform(struct i2c_client *client)
 {
         struct bttv *btv = i2c_get_adapdata(client->adapter);
@@ -241,8 +231,7 @@ static struct i2c_algo_bit_data bttv_i2c
 };
 
 static struct i2c_adapter bttv_i2c_adap_template = {
-	.inc_use           = bttv_inc_use,
-	.dec_use           = bttv_dec_use,
+	.owner             = THIS_MODULE,
 	I2C_DEVNAME("bt848"),
 	.id                = I2C_HW_B_BT848,
 	.client_register   = attach_inform,
--- linux-2.4.21/drivers/media/video/cx88-i2c.c.dv20.orig	2003-08-24 01:21:54.000000000 +0200
+++ linux-2.4.21/drivers/media/video/cx88-i2c.c	2003-08-24 01:28:51.000000000 +0200
@@ -77,18 +77,6 @@ static int cx8800_bit_getsda(void *data)
 
 /* ----------------------------------------------------------------------- */
 
-#ifndef I2C_PEC
-static void cx8800_inc_use(struct i2c_adapter *adap)
-{
-	MOD_INC_USE_COUNT;
-}
-
-static void cx8800_dec_use(struct i2c_adapter *adap)
-{
-	MOD_DEC_USE_COUNT;
-}
-#endif
-
 static int attach_inform(struct i2c_client *client)
 {
         struct cx8800_dev *dev = i2c_get_adapdata(client->adapter);
@@ -122,15 +110,7 @@ static struct i2c_algo_bit_data cx8800_i
 /* ----------------------------------------------------------------------- */
 
 static struct i2c_adapter cx8800_i2c_adap_template = {
-#ifdef I2C_PEC
 	.owner             = THIS_MODULE,
-#else
-	.inc_use           = cx8800_inc_use,
-	.dec_use           = cx8800_dec_use,
-#endif
-#ifdef I2C_ADAP_CLASS_TV_ANALOG
-	.class             = I2C_ADAP_CLASS_TV_ANALOG,
-#endif
 	I2C_DEVNAME("cx2388x"),
 	.id                = I2C_HW_B_BT848,
 	.client_register   = attach_inform,
--- linux-2.4.21/drivers/media/video/i2c-compat.h.dv20.orig	2003-08-24 00:54:49.000000000 +0200
+++ linux-2.4.21/drivers/media/video/i2c-compat.h	2003-08-24 01:21:54.000000000 +0200
@@ -3,6 +3,7 @@
  * i2c modules with tons of #ifdefs
  */
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,5,66)
 
 static inline void *i2c_get_adapdata(struct i2c_adapter *dev)
 {
@@ -45,6 +46,7 @@ static inline void i2c_clients_command(s
 	}
 }
 
+#endif
 
 /*
  * Local variables:
--- linux-2.4.21/drivers/media/video/saa5249.c.dv20.orig	2001-09-30 21:26:06.000000000 +0200
+++ linux-2.4.21/drivers/media/video/saa5249.c	2003-08-24 01:21:54.000000000 +0200
@@ -258,21 +258,19 @@ static int saa5249_command(struct i2c_cl
 
 static struct i2c_driver i2c_driver_videotext = 
 {
-	IF_NAME,		/* name */
-	I2C_DRIVERID_SAA5249, /* in i2c.h */
-	I2C_DF_NOTIFY,
-	saa5249_probe,
-	saa5249_detach,
-	saa5249_command
+	.owner	= THIS_MODULE,
+	.name   = IF_NAME,		/* name */
+	.id	= I2C_DRIVERID_SAA5249, /* in i2c.h */
+	.flags	= I2C_DF_NOTIFY,
+	.attach_adapter = saa5249_probe,
+	.detach_client	= saa5249_detach,
+	.command	= saa5249_command
 };
 
 static struct i2c_client client_template = {
-	"(unset)",
-	-1,
-	0,
-	0,
-	NULL,
-	&i2c_driver_videotext
+	.name	= "(unset)",
+	.id	= -1,
+	.driver	= &i2c_driver_videotext,
 };
 
 /*
--- linux-2.4.21/drivers/media/video/saa7110.c.dv20.orig	2001-11-09 23:01:22.000000000 +0100
+++ linux-2.4.21/drivers/media/video/saa7110.c	2003-08-24 01:21:54.000000000 +0200
@@ -402,14 +402,15 @@ int saa7110_command(struct i2c_device *d
 
 static struct i2c_driver i2c_driver_saa7110 =
 {
-	"saa7110",			/* name */
+	.name	= "saa7110",			/* name */
 
-	I2C_DRIVERID_VIDEODECODER,	/* in i2c.h */
-	I2C_SAA7110, I2C_SAA7110+1,	/* Addr range */
-
-	saa7110_attach,
-	saa7110_detach,
-	saa7110_command
+	.id	= I2C_DRIVERID_VIDEODECODER,	/* in i2c.h */
+	.addr_l	= I2C_SAA7110,
+	.addr_h	= I2C_SAA7110 + 1,
+
+	.attach	= saa7110_attach,
+	.detach	= saa7110_detach,
+	.command	= saa7110_command
 };
 
 EXPORT_NO_SYMBOLS;
--- linux-2.4.21/drivers/media/video/saa7134-i2c.c.dv20.orig	2003-08-24 01:21:54.000000000 +0200
+++ linux-2.4.21/drivers/media/video/saa7134-i2c.c	2003-08-24 01:21:54.000000000 +0200
@@ -313,18 +313,6 @@ static u32 functionality(struct i2c_adap
 	return I2C_FUNC_SMBUS_EMUL;
 }
 
-#if LINUX_VERSION_CODE < KERNEL_VERSION(2,5,54)
-static void inc_use(struct i2c_adapter *adap)
-{
-	MOD_INC_USE_COUNT;
-}
-
-static void dec_use(struct i2c_adapter *adap)
-{
-	MOD_DEC_USE_COUNT;
-}
-#endif
-
 static int attach_inform(struct i2c_client *client)
 {
         struct saa7134_dev *dev = client->adapter->algo_data;
@@ -343,12 +331,7 @@ static struct i2c_algorithm saa7134_algo
 };
 
 static struct i2c_adapter saa7134_adap_template = {
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,5,54)
 	.owner         = THIS_MODULE,
-#else
-	.inc_use       = inc_use,
-	.dec_use       = dec_use,
-#endif
 	.name	       = "saa7134",
 	.id            = I2C_ALGO_SAA7134,
 	.algo          = &saa7134_algo,
--- linux-2.4.21/drivers/media/video/saa7185.c.dv20.orig	2001-09-30 21:26:06.000000000 +0200
+++ linux-2.4.21/drivers/media/video/saa7185.c	2003-08-24 01:21:54.000000000 +0200
@@ -383,13 +383,14 @@ static int saa7185_command(struct i2c_de
 /* ----------------------------------------------------------------------- */
 
 static struct i2c_driver i2c_driver_saa7185 = {
-	"saa7185",		/* name */
-	I2C_DRIVERID_VIDEOENCODER,	/* ID */
-	I2C_SAA7185, I2C_SAA7185 + 1,
-
-	saa7185_attach,
-	saa7185_detach,
-	saa7185_command
+	.name	= "saa7185",
+	.id	= I2C_DRIVERID_VIDEOENCODER,
+	.addr_l	= I2C_SAA7185,
+	.addr_h	= I2C_SAA7185 + 1,
+
+	.attach	= saa7185_attach,
+	.detach	= saa7185_detach,
+	.command	= saa7185_command
 };
 
 EXPORT_NO_SYMBOLS;
--- linux-2.4.21/drivers/media/video/tda9887.c.dv20.orig	2003-08-24 01:21:54.000000000 +0200
+++ linux-2.4.21/drivers/media/video/tda9887.c	2003-08-24 01:21:54.000000000 +0200
@@ -438,9 +438,7 @@ tda9887_command(struct i2c_client *clien
 /* ----------------------------------------------------------------------- */
 
 static struct i2c_driver driver = {
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,5,54)
 	.owner          = THIS_MODULE,
-#endif
         .name           = "i2c tda9887 driver",
         .id             = -1, /* FIXME */
         .flags          = I2C_DF_NOTIFY,
--- linux-2.4.21/drivers/media/video/tuner-3036.c.dv20.orig	2001-09-30 21:26:06.000000000 +0200
+++ linux-2.4.21/drivers/media/video/tuner-3036.c	2003-08-24 01:21:54.000000000 +0200
@@ -185,22 +185,20 @@ tuner_probe(struct i2c_adapter *adap)
 static struct i2c_driver 
 i2c_driver_tuner = 
 {
-	"sab3036",		/* name       */
-	I2C_DRIVERID_SAB3036,	/* ID         */
-        I2C_DF_NOTIFY,
-	tuner_probe,
-	tuner_detach,
-	tuner_command
+	.owner	= THIS_MODULE,
+	.name	= "sab3036",		/* name       */
+	.id	= I2C_DRIVERID_SAB3036,	/* ID         */
+        .flags	= I2C_DF_NOTIFY,
+	.attach_adapter	= tuner_probe,
+	.detach_client	= tuner_detach,
+	.command	= tuner_command
 };
 
 static struct i2c_client client_template =
 {
-        "SAB3036",		/* name       */
-        -1,
-        0,
-        0,
-        NULL,
-        &i2c_driver_tuner
+	.name	= "SAB3036",
+        .id	= -1,
+        .driver	= &i2c_driver_tuner
 };
 
 EXPORT_NO_SYMBOLS;
--- linux-2.4.21/drivers/media/video/tvmixer.c.dv20.orig	2003-08-24 00:54:49.000000000 +0200
+++ linux-2.4.21/drivers/media/video/tvmixer.c	2003-08-24 01:21:54.000000000 +0200
@@ -192,8 +192,8 @@ static int tvmixer_open(struct inode *in
 
 	/* lock bttv in memory while the mixer is in use  */
 	file->private_data = mix;
-	if (client->adapter->inc_use)
-		client->adapter->inc_use(client->adapter);
+	if (client->adapter->owner)
+		__MOD_INC_USE_COUNT(client->adapter->owner);
         return 0;
 }
 
@@ -207,8 +207,8 @@ static int tvmixer_release(struct inode 
 		return -ENODEV;
 	}
 
-	if (client->adapter->dec_use)
-		client->adapter->dec_use(client->adapter);
+	if (client->adapter->owner)
+		__MOD_DEC_USE_COUNT(client->adapter->owner);
 	return 0;
 }
 
--- linux-2.4.21/drivers/media/video/tw98.c.dv20.orig	2003-08-24 01:21:54.000000000 +0200
+++ linux-2.4.21/drivers/media/video/tw98.c	2003-08-24 01:21:54.000000000 +0200
@@ -2,7 +2,7 @@
 #define LM_VERSION "1.0.0"
 
 #include <linux/module.h>
-#include <linux/malloc.h>
+#include <linux/slab.h>
 #include <linux/init.h>
 #include <linux/i2c.h>
 
@@ -373,29 +373,14 @@ static int tw98_detach(struct i2c_client
 	return 0;
 }
 
-void tw98_inc_use(struct i2c_client *client)
-{
-#ifdef MODULE
-	MOD_INC_USE_COUNT;
-#endif
-}
-
-void tw98_dec_use(struct i2c_client *client)
-{
-#ifdef MODULE
-	MOD_DEC_USE_COUNT;
-#endif
-}
-
 static struct i2c_driver driver = {
-	"tw98 driver",
-	I2C_DRIVERID_TW98,
-	I2C_DF_NOTIFY,
-	tw98_attach,
-	tw98_detach,
-	tw98_command,
-	tw98_inc_use,
-	tw98_dec_use
+	.owner          = THIS_MODULE,
+	.name           = "tw98 driver",
+	.id             = I2C_DRIVERID_TW98,
+	.flags          = I2C_DF_NOTIFY,
+	.attach_adapter = tw98_attach,
+	.detach_client  = tw98_detach,
+	.command        = tw98_command,
 };
 
 EXPORT_NO_SYMBOLS;
