--- linux-2.4.21/drivers/media/video/cx88-video.c~	2003-08-24 05:00:59.000000000 +0200
+++ linux-2.4.21/drivers/media/video/cx88-video.c	2003-08-24 15:57:31.000000000 +0200
@@ -663,7 +663,7 @@
 
 	if (STATE_NEEDS_INIT == buf->vb.state) {
 		init_buffer = 1;
-		if (0 != (rc = videobuf_iolock(dev->pci,&buf->vb,NULL)))
+		if (0 != (rc = videobuf_iolock(dev->pci,&buf->vb)))
 			goto fail;
 	}
 
@@ -1634,6 +1634,18 @@
 	return 0;
 }
 
+struct video_device *video_device_alloc(void)
+{
+        struct video_device *vfd;
+
+        vfd = kmalloc(sizeof(*vfd),GFP_KERNEL);
+        if (NULL == vfd)
+                return NULL;
+        memset(vfd,0,sizeof(*vfd));
+        return vfd;
+}
+#define video_device_release(d) kfree(d)
+
 static struct video_device *vdev_init(struct cx8800_dev *dev,
 				      struct video_device *template,
 				      char *type)
