diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff orig/3rdparty/m7101/m7101.c q5/3rdparty/m7101/m7101.c
--- orig/3rdparty/m7101/m7101.c	2002-08-23 14:56:33.000000000 +0200
+++ q5/3rdparty/m7101/m7101.c	2002-08-23 13:43:23.000000000 +0200
@@ -111,11 +105,8 @@ MODULE_PARM(debug, "i");
 /* status, used to indicate that io space needs to be freed */
 static struct pci_dev *m7101 = NULL;
 static int m7101_inserted = FALSE;
-extern void cleanup_module();
  
 static rwlock_t m7101_lock = RW_LOCK_UNLOCKED;
-static unsigned long m7101_lock_flags = 0;
-
 
 #ifdef M7101_DEBUG
 static void
@@ -362,14 +353,17 @@ int init_module(void)
 }
 
 
-void cleanup_module()
+void cleanup_module(void)
 {
-	write_lock_irqsave(m7101_lock, m7101_lock_flags);
+
+	unsigned long flags;
+
+	write_lock_irqsave(&m7101_lock, flags);
 	if (m7101_inserted) {
 		pci_remove_device(m7101);
 		m7101_inserted = FALSE;
 	}
-	write_unlock_irqrestore(m7101_lock, m7101_lock_flags);
+	write_unlock_irqrestore(&m7101_lock, flags);
 
 	if (NULL != m7101)
 		kfree(m7101);
