--- linux-2.4.21/3rdparty/atmelwlan/src/Pcmcia_Pci/fastvnet_cs.c.orig	2003-08-24 15:26:08.000000000 +0200
+++ linux-2.4.21/3rdparty/atmelwlan/src/Pcmcia_Pci/fastvnet_cs.c	2003-08-24 15:32:16.000000000 +0200
@@ -1380,7 +1380,7 @@
 	      C80211_MGMT_AAN_OPENSYSTEM)
 	    wrq->u.data.flags |= IW_ENCODE_OPEN;
 	  wrq->u.data.length =
-	    (Adapter->WepInfo.EncryptionLevel == WEP_64BIT) ? 5 : 13;
+	    ((Adapter->WepInfo.EncryptionLevel == WEP_64BIT) ? 5 : 13);
 	}
 
       copy_to_user (wrq->u.data.pointer,
@@ -2061,14 +2061,13 @@
   if (ret != 0)
     return 0;
 
-  DbgPrint (__FILE__ " " __FUNCTION__ " Io1 %04x, Irq %d\n", ioaddr,
+  DbgPrint (__FILE__ " %s Io1 %04x, Irq %d\n", __FUNCTION__, ioaddr,
 	    dev->irq);
 
   local->Adapter.IoBase = ioaddr;
   local->Adapter.dev = dev;
-  IF_LOUD (DbgPrint
-	   (__FILE__ " " __FUNCTION__ "Net Device : 0x%08x\n ",
-	    local->Adapter.dev);) local->Adapter.IsUp = TRUE;
+  IF_LOUD (DbgPrint(__FILE__ " %s Net Device : 0x%08x\n ", __FUNCTION__, local->Adapter.dev);)
+    local->Adapter.IsUp = TRUE;
   spin_lock_init (&local->slock);
   InitAdapter (&local->Adapter);
 
@@ -2090,7 +2089,7 @@
 
   register_netdev (local->Adapter.dev);
 
-  DbgPrint (__FUNCTION__ " Mac : %02x:%02x:%02x:%02x:%02x:%02x\n",
+  DbgPrint ("%s Mac : %02x:%02x:%02x:%02x:%02x:%02x\n", __FUNCTION__,
 	    local->Adapter.dev->dev_addr[0], local->Adapter.dev->dev_addr[1],
 	    local->Adapter.dev->dev_addr[2], local->Adapter.dev->dev_addr[3],
 	    local->Adapter.dev->dev_addr[4], local->Adapter.dev->dev_addr[5]);
--- linux-2.4.21/3rdparty/atmelwlan/src/includes/pcmcia/vnet.h.orig	2003-08-24 00:33:18.000000000 +0200
+++ linux-2.4.21/3rdparty/atmelwlan/src/includes/pcmcia/vnet.h	2003-08-24 15:30:41.000000000 +0200
@@ -491,6 +491,10 @@
 	//
 } VNet_ADAPTER, * PVNet_ADAPTER;
 
+VOID VnetSleep(ULONG stime);
+VOID RxIndicatePacket(PVNet_ADAPTER Adapter, PUCHAR pRxBuf, ULONG PacketLength, ULONG IndicateLen);
+VOID VNet_netif_wake_queue(PVNet_ADAPTER Adapter);
+
 #define VNET_ASSERT(A)	{}
 #include "stdinclude.h"
 #endif
