diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/exaudio/mixer.c w1/3rdparty/exaudio/mixer.c
--- w2/3rdparty/exaudio/mixer.c	2003-06-24 04:24:57.000000000 +0200
+++ w1/3rdparty/exaudio/mixer.c	2003-08-18 20:59:24.000000000 +0200
@@ -198,7 +198,7 @@ static int get_rec_src(struct mixerdev *
       case 1: mask = 1<<SOUND_MIXER_LINE; break;
       case 2: mask = 1<<SOUND_MIXER_MIC;  break;
       case 3: mask = 1<<SOUND_MIXER_IMIX; break;
-      default:
+      default:;
       }
       return(mask);
     }
