diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/plustek/scale.c w1/3rdparty/plustek/scale.c
--- w2/3rdparty/plustek/scale.c	2002-03-25 22:04:38.000000000 +0100
+++ w1/3rdparty/plustek/scale.c	2003-08-18 20:37:22.000000000 +0200
@@ -131,4 +131,4 @@ void ScaleX( pScanData ps, pUChar inBuf,
 	}
 }
 
-/* END SCALE.C...............................................................*/
\ No newline at end of file
+/* END SCALE.C...............................................................*/
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/plustek/tpa.c w1/3rdparty/plustek/tpa.c
--- w2/3rdparty/plustek/tpa.c	2002-03-25 22:03:19.000000000 +0100
+++ w1/3rdparty/plustek/tpa.c	2003-08-18 20:37:15.000000000 +0200
@@ -1148,4 +1148,4 @@ void TPAP98003Reshading( pScanData ps )
     ps->AsicReg.RD_Pixels = 5400;
 }
 
-/* END TPA.C ................................................................*/
\ No newline at end of file
+/* END TPA.C ................................................................*/
