--- linux-2.4.22/include/sound/adriver.h~	2003-08-28 04:06:30.000000000 +0200
+++ linux-2.4.22/include/sound/adriver.h	2003-08-28 10:35:07.000000000 +0200
@@ -185,12 +185,14 @@
 #undef synchronize_irq
 #define synchronize_irq(irq)	synchronize_irq_wrapper(irq)
 #endif /* LINUX_VERSION_CODE < 2.5.28 */
+#if 0
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2, 5, 68)
 #define IRQ_NONE	/*void*/
 #define IRQ_HANDLED	/*void*/
 #define IRQ_RETVAL(x)	/*void*/
 typedef void irqreturn_t;
 #endif /* LINUX_VERSION_CODE < 2.5.68 */
+#endif
 
 #ifndef min
 /*
