--- linux-2.4.20/drivers/net/hamradio/soundmodem/sm.c.za01-5.orig	2001-04-18 23:40:05.000000000 +0200
+++ linux-2.4.20/drivers/net/hamradio/soundmodem/sm.c	2003-04-16 04:12:05.000000000 +0200
@@ -634,6 +634,7 @@ MODULE_PARM_DESC(midiio, "soundmodem PTT
 
 MODULE_AUTHOR("Thomas M. Sailer, sailer@ife.ee.ethz.ch, hb9jnx@hb9w.che.eu");
 MODULE_DESCRIPTION("Soundcard amateur radio modem driver");
+MODULE_LICENSE("GPL");
 
 /* --------------------------------------------------------------------- */
 
--- linux-2.4.20/drivers/char/ecc.c.za01-5.orig	2003-04-16 04:04:15.000000000 +0200
+++ linux-2.4.20/drivers/char/ecc.c	2003-04-16 04:12:05.000000000 +0200
@@ -982,3 +982,5 @@ int init_module(void) {
 
 	return 0; 
 }
+MODULE_LICENSE("Unknown")
+;
