diff -uNp t2/grsecurity/Makefile.orig t2/grsecurity/Makefile
--- t2/grsecurity/Makefile.orig	2003-08-05 00:27:10.000000000 +0200
+++ t2/grsecurity/Makefile	2003-08-05 00:28:30.000000000 +0200
@@ -8,17 +8,13 @@
 
 O_TARGET := grsec.o
 
-obj-y = grsec_chdir.o grsec_chroot.o grsec_exec.o grsec_fifo.o grsec_fork.o \
+obj-$(CONFIG_GRKERNSEC) = grsec_chdir.o grsec_chroot.o grsec_exec.o grsec_fifo.o grsec_fork.o \
 	grsec_mount.o grsec_rand.o grsec_sig.o grsec_sock.o grsec_sysctl.o \
 	grsec_time.o grsec_tpe.o grsec_ipc.o grsec_link.o
 
-ifeq ($(CONFIG_GRKERNSEC),y)
-obj-y += grsec_init.o grsum.o gracl.o gracl_ip.o gracl_segv.o obsd_rand.o \
+obj-$(CONFIG_GRKERNSEC) += grsec_init.o grsum.o gracl.o gracl_ip.o gracl_segv.o obsd_rand.o \
 	gracl_cap.o gracl_alloc.o gracl_shm.o grsec_mem.o gracl_fs.o \
 	gracl_learn.o
 obj-$(CONFIG_GRKERNSEC_RESLOG) += gracl_res.o
-else
-obj-y += grsec_disabled.o
-endif
 
 include $(TOPDIR)/Rules.make
