--- XFree86-4.3/xc/programs/xdm/file.c.fred	2001-12-14 21:01:21.000000000 +0100
+++ XFree86-4.3/xc/programs/xdm/file.c	2004-02-24 14:16:31.075342400 +0100
@@ -168,6 +168,12 @@
     type = args[1];
     argv = args + 2;
     /*
+     * Support kdm extended syntax to reserve a server for future use but do nothing
+     */
+    if (strcmp(argv[0], "reserve") == 0) {
+      return;
+    }
+    /*
      * extended syntax; if the second argument doesn't
      * exactly match a legal display type and the third
      * argument does, use the second argument as the
