--- xorg-x11/xc/lib/Xft/xftdpy.c	10 Mar 2004 22:07:29 -0000	1.23
+++ xorg-x11/xc/lib/Xft/xftdpy.c	2 Mar 2005 05:32:14 -0000	1.26
@@ -101,6 +97,7 @@ _XftDisplayInfoGet (Display *dpy, FcBool
 
     info->display = dpy;
     info->defaults = 0;
+    info->solidFormat = 0;
     info->hasRender = (XRenderQueryExtension (dpy, &event_base, &error_base) &&
 		       (XRenderFindVisualFormat (dpy, DefaultVisual (dpy, DefaultScreen (dpy))) != 0));
     info->use_free_glyphs = FcTrue;
@@ -375,6 +373,10 @@ _XftDefaultInit (Display *dpy)
 	goto bail1;
     if (!_XftDefaultInitBool (dpy, pat, FC_ANTIALIAS))
 	goto bail1;
+#ifdef FC_EMBOLDEN
+    if (!_XftDefaultInitBool (dpy, pat, FC_EMBOLDEN))
+	goto bail1;
+#endif
     if (!_XftDefaultInitBool (dpy, pat, FC_AUTOHINT))
 	goto bail1;
 #ifdef FC_HINT_STYLE
@@ -471,6 +473,14 @@ XftDefaultSubstitute (Display *dpy, int 
 			   XftDefaultGetBool (dpy, FC_ANTIALIAS, screen,
 					      True));
     }
+#ifdef FC_EMBOLDEN
+    if (FcPatternGet (pattern, FC_EMBOLDEN, 0, &v) == FcResultNoMatch)
+    {
+	FcPatternAddBool (pattern, FC_EMBOLDEN,
+			   XftDefaultGetBool (dpy, FC_EMBOLDEN, screen,
+					      False));
+    }
+#endif
     if (FcPatternGet (pattern, FC_HINTING, 0, &v) == FcResultNoMatch)
     {
 	FcPatternAddBool (pattern, FC_HINTING,
--- xorg-x11/xc/lib/Xft/xftfreetype.c	19 Mar 2004 20:26:51 -0000	1.42
+++ xorg-x11/xc/lib/Xft/xftfreetype.c	28 Feb 2005 19:38:20 -0000	1.46
@@ -527,6 +524,20 @@ XftFontInfoFill (Display *dpy, _Xconst F
 	goto bail1;
     }
 
+#ifdef FC_EMBOLDEN
+    switch (FcPatternGetBool (pattern, FC_EMBOLDEN, 0, &fi->embolden)) {
+    case FcResultNoMatch:
+	fi->embolden = FcFalse;
+	break;
+    case FcResultMatch:
+	break;
+    default:
+	goto bail1;
+    }
+#else
+    fi->embolden = FcFalse;
+#endif
+    
 #ifdef FC_HINT_STYLE
     switch (FcPatternGetInteger (pattern, FC_HINT_STYLE, 0, &hint_style)) {
     case FcResultNoMatch:
--- xorg-x11/xc/lib/Xft/xftglyphs.c	1 May 2003 00:13:17 -0000	1.26
+++ xorg-x11/xc/lib/Xft/xftglyphs.c	2 Mar 2005 05:32:14 -0000	1.29
@@ -173,6 +169,13 @@ XftFontLoadGlyphs (Display	    *dpy,
 		
 	glyphslot = face->glyph;
 
+#if HAVE_FT_GLYPHSLOT_EMBOLDEN
+	/*
+	 * Embolden if required
+	 */
+	if (font->info.embolden) FT_GlyphSlot_Embolden(glyphslot);
+#endif
+
 	/*
 	 * Compute glyph metrics from FreeType information
 	 */
@@ -220,10 +223,10 @@ XftFontLoadGlyphs (Display	    *dpy,
 	height = TRUNC( top - bottom );
 
 	/*
-	 * Try to keep monospace fonts ink-inside
+	 * Clip charcell glyphs to the bounding box
 	 * XXX transformed?
 	 */
-	if (font->info.spacing != FC_PROPORTIONAL && !font->info.transform)
+	if (font->info.spacing >= FC_CHARCELL && !font->info.transform)
 	{
 	    if (font->info.load_flags & FT_LOAD_VERTICAL_LAYOUT)
 	    {
@@ -267,7 +270,7 @@ XftFontLoadGlyphs (Display	    *dpy,
 	xftg->metrics.x = -TRUNC(left);
 	xftg->metrics.y = TRUNC(top);
 
-	if (font->info.spacing != FC_PROPORTIONAL)
+	if (font->info.spacing >= FC_MONO)
 	{
 	    if (font->info.transform)
 	    {
--- xorg-x11/xc/lib/Xft/xftint.h	10 Mar 2004 22:07:29 -0000	1.39
+++ xorg-x11/xc/lib/Xft/xftint.h	28 Feb 2005 19:38:20 -0000	1.41
@@ -109,6 +124,7 @@ struct _XftFontInfo {
      */
     FT_F26Dot6		xsize, ysize;	/* pixel size */
     FcBool		antialias;	/* doing antialiasing */
+    FcBool		embolden;	/* force emboldening */
     int			rgba;		/* subpixel order */
     FT_Matrix		matrix;		/* glyph transformation matrix */
     FcBool		transform;	/* non-identify matrix? */
