--- utempter-0.5.2/utempter.c.biarch-utmp	1999-04-08 12:19:36.000000000 -0400
+++ utempter-0.5.2/utempter.c	2003-04-09 09:38:43.000000000 -0400
@@ -151,7 +151,15 @@ int main(int argc, const char ** argv) {
 
     strncpy(utx.ut_id, id, sizeof(utx.ut_id));
 
-    gettimeofday(&utx.ut_tv, NULL);
+    if (sizeof (utx.ut_tv) == sizeof (struct timeval))
+      gettimeofday((struct timeval *) &utx.ut_tv, NULL);
+    else
+      {
+        struct timeval tv;
+        gettimeofday(&tv, NULL);
+        utx.ut_tv.tv_sec = tv.tv_sec;
+        utx.ut_tv.tv_usec = tv.tv_usec;
+      }
 
     pututxline(&utx);
     updwtmpx(_PATH_WTMP, &utx);
