--- krb5-1.2.7/src/config/pre.in.deps	2000-05-03 23:22:31.000000000 +0200
+++ krb5-1.2.7/src/config/pre.in	2003-04-07 16:12:37.000000000 +0200
@@ -4,7 +4,7 @@
 WHAT = unix
 SHELL=/bin/sh
 
-all:: all-$(WHAT)
+all:: all-$(WHAT)-base
 
 clean:: clean-$(WHAT)
 
@@ -12,9 +12,15 @@ install:: install-$(WHAT)
 
 check:: check-$(WHAT)
 
-all-unix:: all-prerecurse
-all-unix:: all-recurse
-all-unix:: all-postrecurse
+all-unix-base:
+	$(MAKE) all-prerecurse
+	$(MAKE) all-recurse
+	$(MAKE) all-postrecurse
+	$(MAKE) includes
+	$(MAKE) all-unix
+
+includes::
+all-unix::
 
 all-prerecurse::
 all-postrecurse::
--- krb5-1.2.7/src/lib/gssapi/generic/Makefile.in.deps	1999-10-26 03:48:43.000000000 +0200
+++ krb5-1.2.7/src/lib/gssapi/generic/Makefile.in	2003-04-07 16:15:39.000000000 +0200
@@ -20,9 +20,9 @@ EHDRDIR= $(BUILDTOP)$(S)include$(S)gssap
 HDRS=	$(EHDRDIR)$(S)gssapi.h \
 	$(EHDRDIR)$(S)gssapi_generic.h
 
-$(EHDRDIR)$(S)gssapi.h: gssapi.h
+$(EHDRDIR)$(S)gssapi.h: $(EHDRDIR) gssapi.h
 	$(CP) gssapi.h $@
-$(EHDRDIR)$(S)gssapi_generic.h: $(srcdir)$(S)gssapi_generic.h
+$(EHDRDIR)$(S)gssapi_generic.h: $(EHDRDIR) $(srcdir)$(S)gssapi_generic.h
 	$(CP) $(srcdir)$(S)gssapi_generic.h $@
 
 $(OUTPRE)gssapi_err_generic.$(OBJEXT): gssapi_err_generic.c
@@ -103,8 +103,7 @@ EXPORTED_BUILT_HEADERS= gssapi.h
 
 $(OBJS): $(EXPORTED_HEADERS) $(ETHDRS)
 
-all-unix:: includes-unix
-all-unix:: $(EXPORTED_HEADERS) $(ETHDRS) $(HDRS)
+includes:: $(EXPORTED_HEADERS) $(ETHDRS) $(HDRS)
 all-unix:: all-libobjs
 
 
@@ -124,8 +123,8 @@ clean-windows::
 	$(RM) $(HDRS)
 	-if exist $(EHDRDIR)\nul rmdir $(EHDRDIR)
 
-includes-unix::
-	@if test -d $(EHDRDIR); then :; else (set -x; mkdir $(EHDRDIR)) fi
+$(EHDRDIR):
+	@(set -x; mkdir $@)
 
 # Krb5InstallHeaders($(EXPORTED_HEADERS), $(KRB5_INCDIR)/krb5)
 install::
--- krb5-1.2.7/src/lib/kadm5/Makefile.in.deps	2000-02-22 22:15:45.000000000 +0100
+++ krb5-1.2.7/src/lib/kadm5/Makefile.in	2003-04-07 16:34:49.000000000 +0200
@@ -91,7 +91,7 @@ clean-unix::
 	$(RM) $(HDRS)
 
 all-prerecurse:: includes
-all-prerecurse:: all-libobjs
+	$(MAKE) all-libobjs
 
 all-windows:: $(OBJS)
 
--- krb5-1.2.7/src/lib/rpc/Makefile.in.deps	2000-02-22 22:15:46.000000000 +0100
+++ krb5-1.2.7/src/lib/rpc/Makefile.in	2003-04-07 16:32:58.000000000 +0200
@@ -173,7 +173,7 @@ HDRS=	$(HDRDIR)/auth.h \
 	$(HDRDIR)/xdr.h
 
 all-prerecurse:: includes
-all-prerecurse:: all-liblinks
+	$(MAKE) all-liblinks
 
 all-windows:: $(OBJS)
 
--- krb5-1.2.7/src/util/profile/Makefile.in.deps	2003-04-07 16:08:40.000000000 +0200
+++ krb5-1.2.7/src/util/profile/Makefile.in	2003-04-07 16:13:08.000000000 +0200
@@ -51,8 +51,9 @@ STOBJLISTS=OBJS.ST
 
 PROFILE_HDR=$(BUILDTOP)$(S)include$(S)profile.h
 
-all-unix:: includes test_parse test_profile
+all-unix:: test_parse test_profile
 all-unix:: all-libs
+
 all-windows:: $(PROFILE_HDR)
 
 $(PROFILE_HDR):: profile.h
--- krb5-1.2.7/src/util/ss/Makefile.in.deps	1999-10-26 03:49:53.000000000 +0200
+++ krb5-1.2.7/src/util/ss/Makefile.in	2003-04-07 16:08:57.000000000 +0200
@@ -117,7 +117,7 @@ includes:: $(SRC_HDRS_DEP) $(BUILD_HDRS)
 clean-unix::
 	$(RM) $(HDRS)
 
-std_rqs.c: std_rqs.ct
+std_rqs.c: mk_cmds std_rqs.ct
 
 ss_err.h: ss_err.et
 
