--- pciutils-2.2.1/lib/types.h.pix	2005-09-21 13:43:02.000000000 +0200
+++ pciutils-2.2.1/lib/types.h	2006-01-09 11:06:49.000000000 +0100
@@ -6,7 +6,7 @@
  *	Can be freely distributed and used under the terms of the GNU GPL.
  */
 
-#include <sys/types.h>
+#include <stdint.h>
 
 #ifndef PCI_HAVE_Uxx_TYPES
 
@@ -15,9 +15,9 @@
 typedef unsigned __int16 u16;
 typedef unsigned __int32 u32;
 #else
-typedef u_int8_t u8;
-typedef u_int16_t u16;
-typedef u_int32_t u32;
+typedef uint8_t u8;
+typedef uint16_t u16;
+typedef uint32_t u32;
 #endif
 
 #ifdef PCI_HAVE_64BIT_ADDRESS
