--- console-tools-0.2.3/doc/Makefile.am._db	1999-04-09 00:11:11.000000000 +0200
+++ console-tools-0.2.3/doc/Makefile.am	2002-05-07 11:17:34.000000000 +0200
@@ -13,7 +13,9 @@
 
 MAINTAINERCLEANFILES = lct*.html lct.txt kbd.FAQ.txt
 
-all: lct.txt lct.html kbd.FAQ.txt kbd.FAQ.html
+all: lct.txt kbd.FAQ.txt 
+
+# lct.html kbd.FAQ.html
 
 %.html: $(srcdir)/%.sgml
 	sgml2html $<
--- console-tools-0.2.3/doc/Makefile.in._db	1999-10-25 22:58:12.000000000 +0200
+++ console-tools-0.2.3/doc/Makefile.in	2002-05-07 11:17:44.000000000 +0200
@@ -329,7 +329,9 @@
 maintainer-clean-generic clean mostlyclean distclean maintainer-clean
 
 
-all: lct.txt lct.html kbd.FAQ.txt kbd.FAQ.html
+all: lct.txt kbd.FAQ.txt 
+
+# lct.html kbd.FAQ.html
 
 %.html: $(srcdir)/%.sgml
 	sgml2html $<
