--- Linux-PAM-0.77/modules/pam_console/config.y.console-reset	2004-09-07 17:48:06.085931320 +0200
+++ Linux-PAM-0.77/modules/pam_console/config.y	2004-09-07 17:49:14.031602008 +0200
@@ -394,13 +394,13 @@
     }
     g_free(dirname);
     return ret;
-
+}
 
 
 STATIC int
 reset_permissions(pam_handle_t *pamh, const char *consolename, int nonroot_ok) {
-    struct passwd *pwd;
-    struct group *grp;
+    struct passwd *pwd, p;
+    struct group *grp, g;
     config *c;
     GSList *cl;
 
@@ -415,13 +415,15 @@
 	    if (pwd == NULL) {
 		_pam_log(LOG_ERR, FALSE, "getpwnam failed for %s",
 			 c->revert_owner ? c->revert_owner : "root");
-		return -1;
+		p.pw_uid = 0;
+		pwd = &p;
 	    }
 	    grp = _pammodutil_getgrnam(pamh, c->revert_group ? c->revert_group : "root");
 	    if (grp == NULL) {
                 _pam_log(LOG_ERR, FALSE, "getgrnam failed for %s",
                          c->revert_group ? c->revert_group : "root");
-                return -1;
+		g.gr_gid = 0;
+		grp = &g;
             }
 	    if (c->device_class->list)
 		chmod_files(c->revert_mode ? c->revert_mode : "0600",
--- Linux-PAM-0.77/modules/pam_console/pam_console_apply.c.console-reset	2004-09-07 17:48:06.064934512 +0200
+++ Linux-PAM-0.77/modules/pam_console/pam_console_apply.c	2004-09-07 17:48:06.091930408 +0200
@@ -41,8 +41,9 @@
 	va_list args;
 	if (debug_p && !debug) return;
         va_start(args, format);
-	fprintf(stderr, format, args);
+	vfprintf(stderr, format, args);
 	va_end(args);
+	fprintf(stderr, "\n");
 }
 
 int
