--- libpng-1.2.10/scripts/makefile.linux.mdkconf	2006-04-23 20:45:40.000000000 +0200
+++ libpng-1.2.10/scripts/makefile.linux	2006-06-15 19:09:42.000000000 +0200
@@ -31,12 +31,6 @@ RM_F=/bin/rm -f
 prefix=/usr/local
 exec_prefix=$(prefix)
 
-# Where the zlib library and include files are located.
-#ZLIBLIB=/usr/local/lib
-#ZLIBINC=/usr/local/include
-ZLIBLIB=../zlib
-ZLIBINC=../zlib
-
 ALIGN=
 # for i386:
 #ALIGN=-malign-loops=2 -malign-functions=2
@@ -46,12 +40,11 @@ WARNMORE=-Wwrite-strings -Wpointer-arith
 	-Wstrict-prototypes -Wmissing-prototypes #-Wconversion
 
 # for pgcc version 2.95.1, -O3 is buggy; don't use it.
+CFLAGS=$(RPM_OPT_FLAGS) -Wall
 
-CFLAGS=-I$(ZLIBINC) -Wall -O3 -funroll-loops \
-	$(ALIGN) # $(WARNMORE) -g -DPNG_DEBUG=5
-
-LDFLAGS=-L. -Wl,-rpath,. -L$(ZLIBLIB) -Wl,-rpath,$(ZLIBLIB) -lpng12 -lz -lm
-LDFLAGS_A=-L$(ZLIBLIB) -Wl,-rpath,$(ZLIBLIB) libpng.a -lz -lm
+LIBPNG_LDFLAGS=-lz -lm
+LDFLAGS=-L. -Wl,-rpath,. -lpng12
+LDFLAGS_A=libpng.a $(LIBPNG_LDFLAGS)
 
 INCPATH=$(prefix)/include
 LIBPATH=$(exec_prefix)/lib
@@ -97,9 +90,9 @@ libpng-config:
 	( cat scripts/libpng-config-head.in; \
 	echo prefix=\"$(prefix)\"; \
 	echo I_opts=\"-I$(INCPATH)/$(LIBNAME)\"; \
-	echo L_opts=\"-L$(LIBPATH)\"; \
-	echo R_opts=\"-Wl,-rpath,$(LIBPATH)\"; \
-	echo libs=\"-lpng12 -lz -lm\"; \
+	echo L_opts=\"\"; \
+	echo R_opts=\"\"; \
+	echo libs=\"-lpng12\"; \
 	cat scripts/libpng-config-body.in ) > libpng-config
 	chmod +x libpng-config
 
@@ -110,11 +103,12 @@ $(LIBSOMAJ): $(LIBSOVER)
 	$(LN_SF) $(LIBSOVER) $(LIBSOMAJ)
 
 $(LIBSOVER): $(OBJSDLL)
-	$(CC) -shared -Wl,-soname,$(LIBSOMAJ) -o $(LIBSOVER) $(OBJSDLL)
+	$(CC) -shared -Wl,-soname,$(LIBSOMAJ) -o $(LIBSOVER) $(LIBPNG_LDFLAGS) $(OBJSDLL)
 
 $(OLDSOVER): $(OBJSDLL)
 	$(CC) -shared -Wl,-soname,$(OLDSOMAJ) \
 	-o $(OLDSOVER) \
+	$(LIBPNG_LDFLAGS) \
 	$(OBJSDLL)
 
 pngtest: pngtest.o $(LIBSO)
@@ -208,9 +202,8 @@ test-dd:
 	./pngtestd pngtest.png
 
 test-installed:
-	$(CC) -I$(ZLIBINC) \
+	$(CC) \
 	   `$(BINPATH)/$(LIBNAME)-config --cflags` pngtest.c \
-	   -L$(ZLIBLIB) -Wl,-rpath,$(ZLIBLIB) \
 	   -o pngtesti `$(BINPATH)/$(LIBNAME)-config --ldflags`
 	./pngtesti pngtest.png
 
