--- samba-3.0.22/packaging/Mandrake/samba.pamd.avx	2006-06-24 13:54:59.000000000 -0600
+++ samba-3.0.22/packaging/Mandrake/samba.pamd	2006-06-24 13:54:59.000000000 -0600
@@ -1,5 +1,5 @@
 #%PAM-1.0
-auth       required	/lib/security/pam_nologin.so
-auth       required	/lib/security/pam_stack.so service=system-auth
-account    required	/lib/security/pam_stack.so service=system-auth
-session    required	/lib/security/pam_stack.so service=system-auth
+auth		required	pam_nologin.so
+auth		include		system-auth
+account		include		system-auth
+session		include		system-auth
