--- samba-3.0.11/examples/LDAP/smbldap-tools-0.8.6/smbldap.conf.orig	2005-01-17 10:22:35.000000000 +0200
+++ samba-3.0.11/examples/LDAP/smbldap-tools-0.8.6/smbldap.conf	2005-02-10 14:10:19.568741675 +0200
@@ -82,19 +82,19 @@
 
 # LDAP Suffix
 # Ex: suffix=dc=IDEALX,dc=ORG
-suffix="dc=idealx,dc=org"
+suffix="dc=example,dc=com"
 
 # Where are stored Users
 # Ex: usersdn="ou=Users,dc=IDEALX,dc=ORG"
-usersdn="ou=Users,${suffix}"
+usersdn="ou=People,${suffix}"
 
 # Where are stored Computers
 # Ex: computersdn="ou=Computers,dc=IDEALX,dc=ORG"
-computersdn="ou=Computers,${suffix}"
+computersdn="ou=Hosts,${suffix}"
 
 # Where are stored Groups
 # Ex groupsdn="ou=Groups,dc=IDEALX,dc=ORG"
-groupsdn="ou=Groups,${suffix}"
+groupsdn="ou=Group,${suffix}"
 
 # Where are stored Idmap entries (used if samba is a domain member server)
 # Ex groupsdn="ou=Idmap,dc=IDEALX,dc=ORG"
@@ -156,29 +156,31 @@
 # Ex: \\My-PDC-netbios-name\homes\%U
 # Just set it to a null string if you want to use the smb.conf 'logon home'
 # directive and/or disable roaming profiles
-userSmbHome="\\PDC-SMB3\homes\%U"
+#userSmbHome="\\PDC-SMB3\homes\%U"
+userSmbHome=""
 
 # The UNC path to profiles locations (%U username substitution)
 # Ex: \\My-PDC-netbios-name\profiles\%U
 # Just set it to a null string if you want to use the smb.conf 'logon path'
 # directive and/or disable roaming profiles
-userProfile="\\PDC-SMB3\profiles\%U"
+#userProfile="\\PDC-SMB3\profiles\%U"
+userProfile=""
 
 # The default Home Drive Letter mapping
 # (will be automatically mapped at logon time if home directory exist)
 # Ex: H: for H:
-userHomeDrive="H:"
+#userHomeDrive="H:"
 
 # The default user netlogon script name (%U username substitution)
 # if not used, will be automatically username.cmd
 # make sure script file is edited under dos
 # Ex: %U.cmd
 # userScript="startup.cmd" # make sure script file is edited under dos
-userScript="%U.cmd"
+#userScript="%U.cmd"
 
 # Domain appended to the users "mail"-attribute
 # when smbldap-useradd -M is used
-mailDomain="idealx.com"
+mailDomain="example.com"
 
 ##############################################################################
 #
--- samba-3.0.11/examples/LDAP/smbldap-tools-0.8.6/smbldap_bind.conf.orig	2005-01-17 10:22:35.000000000 +0200
+++ samba-3.0.11/examples/LDAP/smbldap-tools-0.8.6/smbldap_bind.conf	2005-02-10 14:11:48.825348427 +0200
@@ -5,8 +5,7 @@
 # master ldap for writing access and a slave ldap server for reading access
 # By default, we will use the same DN (so it will work for standard Samba
 # release)
-slaveDN="cn=Manager,dc=idealx,dc=org"
+slaveDN="cn=root,dc=example,dc=com"
 slavePw="secret"
-masterDN="cn=Manager,dc=idealx,dc=org"
+masterDN="cn=root,dc=example,dc=com"
 masterPw="secret"
-
