--- smbldap-tools-0.9.1/smbldap.conf.orig	2005-05-31 10:49:26.000000000 +0200
+++ smbldap-tools-0.9.1/smbldap.conf	2005-09-18 21:59:31.000000000 +0200
@@ -39,7 +39,7 @@
 # Domain name the Samba server is in charged.
 # If not defined, parameter is taking from smb.conf configuration file
 # Ex: sambaDomain="IDEALX-NT"
-sambaDomain="IDEALX-NT"
+sambaDomain="example"
 
 ##############################################################################
 #
@@ -85,34 +85,34 @@
 
 # CA certificate
 # see "man Net::LDAP" in start_tls section for more details
-cafile="/etc/opt/IDEALX/smbldap-tools/ca.pem"
+cafile="/etc/ssl/cacert.pem"
 
 # certificate to use to connect to the ldap server
 # see "man Net::LDAP" in start_tls section for more details
-clientcert="/etc/opt/IDEALX/smbldap-tools/smbldap-tools.pem"
+clientcert=""
 
 # key certificate to use to connect to the ldap server
 # see "man Net::LDAP" in start_tls section for more details
-clientkey="/etc/opt/IDEALX/smbldap-tools/smbldap-tools.key"
+clientkey=""
 
 # LDAP Suffix
 # Ex: suffix=dc=IDEALX,dc=ORG
-suffix="dc=idealx,dc=org"
+suffix="dc=example,dc=com"
 
 # Where are stored Users
 # Ex: usersdn="ou=Users,dc=IDEALX,dc=ORG"
 # Warning: if 'suffix' is not set here, you must set the full dn for usersdn
-usersdn="ou=Users,${suffix}"
+usersdn="ou=People,${suffix}"
 
 # Where are stored Computers
 # Ex: computersdn="ou=Computers,dc=IDEALX,dc=ORG"
 # Warning: if 'suffix' is not set here, you must set the full dn for computersdn
-computersdn="ou=Computers,${suffix}"
+computersdn="ou=Hosts,${suffix}"
 
 # Where are stored Groups
 # Ex: groupsdn="ou=Groups,dc=IDEALX,dc=ORG"
 # Warning: if 'suffix' is not set here, you must set the full dn for groupsdn
-groupsdn="ou=Groups,${suffix}"
+groupsdn="ou=Group,${suffix}"
 
 # Where are stored Idmap entries (used if samba is a domain member server)
 # Ex: groupsdn="ou=Idmap,dc=IDEALX,dc=ORG"
@@ -123,7 +123,7 @@
 # If not defined, entries are stored in sambaDomainName object.
 # Ex: sambaUnixIdPooldn="sambaDomainName=${sambaDomain},${suffix}"
 # Ex: sambaUnixIdPooldn="cn=NextFreeUnixId,${suffix}"
-sambaUnixIdPooldn="sambaDomainName=IDEALX-NT,${suffix}"
+sambaUnixIdPooldn="sambaDomainName=example,${suffix}"
 
 # Default scope Used
 scope="sub"
@@ -181,29 +181,29 @@
 # Just set it to a null string if you want to use the smb.conf 'logon home'
 # directive and/or disable roaming profiles
 # Ex: userSmbHome="\\PDC-SMB3\%U"
-userSmbHome="\\PDC-SRV\%U"
+userSmbHome=""
 
 # The UNC path to profiles locations (%U username substitution)
 # Just set it to a null string if you want to use the smb.conf 'logon path'
 # directive and/or disable roaming profiles
 # Ex: userProfile="\\PDC-SMB3\profiles\%U"
-userProfile="\\PDC-SRV\profiles\%U"
+userProfile=""
 
 # The default Home Drive Letter mapping
 # (will be automatically mapped at logon time if home directory exist)
 # Ex: userHomeDrive="H:"
-userHomeDrive="H:"
+userHomeDrive=""
 
 # The default user netlogon script name (%U username substitution)
 # if not used, will be automatically username.cmd
 # make sure script file is edited under dos
 # Ex: userScript="startup.cmd" # make sure script file is edited under dos
-userScript="logon.bat"
+userScript=""
 
 # Domain appended to the users "mail"-attribute
 # when smbldap-useradd -M is used
 # Ex: mailDomain="idealx.com"
-mailDomain="idealx.com"
+mailDomain="example.com"
 
 ##############################################################################
 #
