--- gzip-1.2.4a/zgrep.in.can-2005-0758	1993-06-23 01:51:32.000000000 -0600
+++ gzip-1.2.4a/zgrep.in	2005-05-18 14:22:29.935181653 -0600
@@ -6,7 +6,7 @@
 
 PATH="BINDIR:$PATH"; export PATH
 
-prog=`echo $0 | sed 's|.*/||'`
+prog=`echo "$0" | sed 's|.*/||'`
 case "$prog" in
 	*egrep)	grep=${EGREP-egrep}	;;
 	*fgrep)	grep=${FGREP-fgrep}	;;
@@ -53,7 +53,7 @@
 res=0
 for i do
   if test $list -eq 1; then
-    gzip -cdfq "$i" | $grep $opt "$pat" > /dev/null && echo $i
+    gzip -cdfq "$i" | $grep $opt "$pat" > /dev/null && printf "%s\n" "$i"
     r=$?
   elif test $# -eq 1 -o $silent -eq 1; then
     gzip -cdfq "$i" | $grep $opt "$pat"
