--- modutils-2.4.26/util/alias.h~	2004-01-20 15:19:04.259991580 +0100
+++ modutils-2.4.26/util/alias.h	2004-01-20 15:24:55.151072766 +0100
@@ -153,6 +153,21 @@
 	"char-major-206 osst",  /* OnStream SCSI tape */
 	"char-major-216 rfcomm",
 
+	/* agpgart like 2.6 */
+	"ali-agp agpgart"
+	"ati-agp agpgart"
+	"amd-k7-agp agpgart"
+	"amd64-agp agpgart"
+	"alpha-agp agpgart"
+	"hp-agp agpgart"
+	"i460-agp agpgart"
+	"intel-agp agpgart"
+	"nvidia-agp agpgart"
+	"sis-agp agpgart"
+	"sworks-agp agpgart"
+	"uninorth-agp agpgart"
+	"via-agp agpgart"
+
 	"dos msdos",
 	"dummy0 dummy",
 	"dummy1 dummy",
