--- php-4.3.9/scripts/phpize.in.phpize	2004-09-29 22:09:34.000000000 -0600
+++ php-4.3.9/scripts/phpize.in	2004-09-29 22:14:30.000000000 -0600
@@ -4,10 +4,11 @@
 phpdir="@libdir@/php/build"
 includedir="$prefix/include/php"
 builddir="`pwd`"
+export WANT_AUTOCONF_2_5="1"
 
 FILES_BUILD="mkdep.awk scan_makefile_in.awk shtool"
 FILES="acinclude.m4 Makefile.global"
-CLEAN_FILES="$FILES *.lo *.la *.o .deps .libs/ build/ include/ modules/ install-sh \
+CLEAN_FILES="$FILES *.so *.lo *.la *.o .deps .libs/ build/ include/ modules/  autom*.cache install-sh \
     mkinstalldirs missing config.nice config.sub config.guess configure configure.in \
 	aclocal.m4 config.h config.h.in conftest* ltmain.sh libtool config.cache \
 	config.log config.status Makefile Makefile.fragments Makefile.objects confdefs.h"
@@ -48,24 +49,14 @@
 (cd "$phpdir" && cp $FILES_BUILD "$builddir"/build)
 (cd "$phpdir" && cp $FILES "$builddir")
 
-sed \
--e "s#@prefix@#$prefix#" \
-< "$phpdir/phpize.m4" > configure.in
+cp $phpdir/phpize.m4 configure.in
 
 touch install-sh mkinstalldirs missing
 
-aclocal || exit 1
+aclocal-1.7 || exit 1
 autoconf || exit 1
 autoheader || exit 1
-test -x "$builddir/build/shtool" || chmod +x "$builddir/build/shtool"
-if test ! -x "$builddir/build/shtool"; then
-  echo "shtool at '$builddir/build/shtool' not executable. "
-  echo "Make sure that the file exists and is executable and then rerun this script. "
-  echo 
-  exit 1
-fi
-libtoolize=`$builddir/build/shtool path glibtoolize libtoolize`
-$libtoolize -f -c || exit 1
+libtoolize --copy --force || exit 1
 
 # dumping API NOs:
 PHP_API_VERSION=`egrep '#define PHP_API_VERSION' $includedir/main/php.h|sed 's/#define PHP_API_VERSION//'`

diff -Naur php-4.3.6/scripts/phpize.m4 php-4.3.6.oden/scripts/phpize.m4
--- php-4.3.6/scripts/phpize.m4	2004-05-26 13:55:20.910139576 +0200
+++ php-4.3.6.oden/scripts/phpize.m4	2004-05-26 13:55:52.210381216 +0200
@@ -71,7 +71,7 @@
 CPPFLAGS="$CPPFLAGS -DHAVE_CONFIG_H"
 CFLAGS_CLEAN='$(CFLAGS)'
 
-test "$prefix" = "NONE" && prefix="/usr/local"
+test "$prefix" = "NONE" && prefix="/usr"
 test "$exec_prefix" = "NONE" && exec_prefix='$(prefix)'
 
 PHP_SUBST(PHP_MODULES)
