--- linux-2.4.22/kernel/kallsyms.c.amd64-kallsyms	2003-10-30 14:38:07.000000000 +0100
+++ linux-2.4.22/kernel/kallsyms.c	2003-10-31 01:39:35.000000000 +0100
@@ -38,7 +38,6 @@
 #include <linux/module.h>
 #include <linux/string.h>
 #include <linux/kallsyms.h>
-#include <linux/kdb.h>
 
 /* These external symbols are only set on kernels compiled with
  * CONFIG_KALLSYMS.
--- linux-2.4.22/arch/x86_64/config.in.amd64-kallsyms	2003-10-30 14:38:41.000000000 +0100
+++ linux-2.4.22/arch/x86_64/config.in	2003-10-31 01:38:27.000000000 +0100
@@ -246,6 +246,7 @@ if [ "$CONFIG_DEBUG_KERNEL" != "n" ]; th
    bool '  IOMMU debugging' CONFIG_IOMMU_DEBUG
    bool '  IOMMU leak tracing' CONFIG_IOMMU_LEAK
    bool '  Probalistic stack overflow check' CONFIG_DEBUG_STACKOVERFLOW
+   bool '  Load all symbols for debugging' CONFIG_KALLSYMS
 fi
 endmenu
 
