--- lsof-4.57.orig/scripts/idrlogin.perl5
+++ lsof-4.57/scripts/idrlogin.perl5
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl
+#!/usr/bin/perl
 #
 # $Id: idrlogin.perl5,v 1.4 2000/07/14 17:04:31 abe Exp $
 #
--- lsof-4.57.orig/scripts/count_pf.perl5
+++ lsof-4.57/scripts/count_pf.perl5
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl
+#!/usr/bin/perl
 #
 # count_pf.perl5 -- run lsof in repeat mode and count processes and
 #		    files
--- lsof-4.57.orig/scripts/big_brother.perl5
+++ lsof-4.57/scripts/big_brother.perl5
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl -w
+#!/usr/bin/perl -w
 #+##############################################################################
 #                                                                              #
 # File: big_brother.perl                                                       #
--- lsof-4.57.orig/scripts/list_NULf.perl5
+++ lsof-4.57/scripts/list_NULf.perl5
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl
+#!/usr/bin/perl
 #
 # $Id: list_NULf.perl5,v 1.5 2000/07/14 17:03:37 abe Exp $
 #
--- lsof-4.57.orig/scripts/list_fields.perl
+++ lsof-4.57/scripts/list_fields.perl
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl4
+#!/usr/bin/perl4
 #
 # $Id: list_fields.perl,v 1.5 2000/07/14 17:03:37 abe Exp $
 #
--- lsof-4.57.orig/scripts/identd.perl5
+++ lsof-4.57/scripts/identd.perl5
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl
+#!/usr/bin/perl
 ###################################################################
 # identd.perl5 : An implementation of RFC 1413 Ident Server
 #                using Vic Abell's lsof.
--- lsof-4.57.orig/scripts/shared.perl5
+++ lsof-4.57/scripts/shared.perl5
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl
+#!/usr/bin/perl
 #
 # $Id: shared.perl5,v 1.3 2000/07/14 17:03:37 abe Exp $
 #
--- lsof-4.57.orig/scripts/idrlogin.perl.orig	2001-11-18 13:21:00.000000000 +0100
+++ lsof-4.57.orig/scripts/idrlogin.perl	2002-07-10 12:26:39.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl
+#!/usr/bin/perl
 #
 # $Id: idrlogin.perl,v 1.5 2001/11/18 12:20:46 abe Exp $
 #
--- lsof-4.57.orig/scripts/count_pf.perl.orig	2001-11-12 18:40:55.000000000 +0100
+++ lsof-4.57.orig/scripts/count_pf.perl	2002-07-10 12:26:53.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl
+#!/usr/bin/perl
 #
 # count_pf.perl-- run lsof in repeat mode and count processes and
 #		  files
--- lsof-4.57.orig/scripts/watch_a_file.perl.orig	2001-11-12 18:41:44.000000000 +0100
+++ lsof-4.57.orig/scripts/watch_a_file.perl	2002-07-10 12:26:57.000000000 +0200
@@ -1,4 +1,4 @@
-#!/usr/local/bin/perl
+#!/usr/bin/perl
 #
 # watch_a_file.perl -- use lsof -F output to watch a specific file
 #		       (or file system)
