diff -ur krb5-1.2.7/src/include/krb5.hin krb5-1.2.7/src/include/krb5.hin
--- krb5-1.2.7/src/include/krb5.hin	2002-04-16 19:47:53.000000000 -0400
+++ krb5-1.2.7/src/include/krb5.hin	2003-02-03 17:09:14.000000000 -0500
@@ -326,7 +326,7 @@
 #define	krb5_princ_size(context, princ) (princ)->length
 #define	krb5_princ_type(context, princ) (princ)->type
 #define	krb5_princ_name(context, princ) (princ)->data
-#define	krb5_princ_component(context, princ,i) ((princ)->data + i)
+#define	krb5_princ_component(context, princ,i) (i < krb5_princ_size(context, princ) ? ((princ)->data + i) : NULL)
 
 /*
  * end "base-defs.h"
