--- ./terminal.c.fixendkey	2002-03-04 18:23:09.000000000 +0100
+++ ./terminal.c	2002-07-23 21:45:01.000000000 +0200
@@ -313,7 +313,7 @@
   { "ei", &_rl_term_ei },
   { "ic", &_rl_term_ic },
   { "im", &_rl_term_im },
-  { "kH", &_rl_term_kH },	/* home down ?? */
+  { "@", &_rl_term_kH },	/* home down ?? */
   { "kI", &_rl_term_kI },	/* insert */
   { "kd", &_rl_term_kd },
   { "ke", &_rl_term_ke },	/* end keypad mode */
