--- gcc-3.3.1/gcc/toplev.c.default-O2-options	2003-07-21 13:58:53.000000000 +0200
+++ gcc-3.3.1/gcc/toplev.c	2003-07-21 14:05:04.000000000 +0200
@@ -5101,7 +5101,6 @@ parse_options_and_default_flags (argc, a
       flag_reorder_functions = 1;
       flag_value_histograms = 1;
       flag_value_profile_transformations = 1;
-      flag_unit_at_time = 1;
     }
 
   if (optimize >= 3)
@@ -5111,6 +5110,7 @@ parse_options_and_default_flags (argc, a
       flag_tracer = 1;
       flag_web = 1;
       flag_unswitch_loops = 1;
+      flag_unit_at_time = 1;
     }
 
   if (optimize < 2 || optimize_size)
