--- gcc-3.2.2/gcc/config/ia64/linux.h.cross-inhibit_libc	2002-05-23 17:24:38.000000000 -0400
+++ gcc-3.2.2/gcc/config/ia64/linux.h	2003-03-16 14:48:45.000000000 -0500
@@ -53,6 +53,7 @@
 /* Do code reading to identify a signal frame, and set the frame
    state data appropriately.  See unwind-dw2.c for the structs.  */
 
+#ifndef inhibit_libc
 #ifdef IN_LIBGCC2
 #include <signal.h>
 #include <sys/ucontext.h>
@@ -116,3 +117,4 @@
       goto SUCCESS;							\
     }
 #endif /* IN_LIBGCC2 */
+#endif /* inhibit_libc */
--- gcc-3.2.2/gcc/config/rs6000/linux.h.cross-inhibit_libc	2003-01-29 14:56:15.000000000 -0500
+++ gcc-3.2.2/gcc/config/rs6000/linux.h	2003-03-16 14:47:22.000000000 -0500
@@ -80,6 +80,7 @@ Boston, MA 02111-1307, USA.  */
 /* Do code reading to identify a signal frame, and set the frame
    state data appropriately.  See unwind-dw2.c for the structs.  */
 
+#ifndef inhibit_libc
 #ifdef IN_LIBGCC2
 #include <signal.h>
 #include <sys/ucontext.h>
@@ -154,4 +155,4 @@ enum { SIGNAL_FRAMESIZE = 64 };
     (FS)->retaddr_column = CR0_REGNO;					\
     goto SUCCESS;							\
   } while (0)
-
+#endif
