--- shadow-4.0.3/src/useradd.c.mailspool	2003-02-25 00:19:52.000000000 +0100
+++ shadow-4.0.3/src/useradd.c	2003-02-25 00:23:01.000000000 +0100
@@ -235,6 +235,7 @@
 static void lastlog_reset (uid_t);
 static void usr_update (void);
 static void create_home (void);
+static void create_mail (void);
 
 /*
  * fail_exit - undo as much as possible
@@ -1757,6 +1758,47 @@
 		grp_update ();
 }
 
+/*
+ * create_mail - create the user's mail spool
+ *
+ *	create_mail() creates the user's mail spool if it does not
+ *	already exist. It will be created mode 660 owned by the user
+ *	and group 'mail' (if it exists, otherwise we bail)
+ */
+static void create_mail (void)
+{
+	char *ms;
+	int fd;
+	struct group *mail;
+	gid_t mail_gid;
+	mode_t mode;
+
+	mail = getgrnam("mail");
+	if (mail == NULL) {
+	    fprintf(stderr, "No group named \"mail\" exists, creating mail spool with mode 0600.\n");
+	    mode = 0600;
+	    mail_gid = user_gid;
+	} else {
+	    mode = 0660;
+	    mail_gid = mail->gr_gid;
+	}
+
+	ms = malloc(strlen(user_name) + 11);
+	if (ms != NULL) {
+		sprintf(ms, "/var/mail/%s", user_name);
+		if (access(ms, R_OK) != 0) {
+			fd = open(ms, O_CREAT|O_EXCL|O_WRONLY|O_TRUNC);
+			if (fd != -1) {
+				fchown(fd, user_id, mail_gid);
+				fchmod(fd, mode);
+				close(fd);
+			}
+		}
+	}
+
+	free(ms);
+}
+
 /* a fake something */
 static char *empty_list = NULL;
 
@@ -2079,6 +2121,7 @@
 	if (mflg) {
 		create_home ();
 		copy_tree (def_template, user_home, user_id, user_gid,1);
+		create_mail ();
 	} /* Stupid warning removed for the innocent's protection */
  	    /*
  	     * The whole idea about breaking some stupid scripts by creating a new
