--- shadow-4.0.3/src/userdel.c.vipw	2002-11-14 22:44:14.000000000 +0100
+++ shadow-4.0.3/src/userdel.c	2002-11-14 22:46:27.000000000 +0100
@@ -621,6 +621,11 @@
 			}
 			}
 
+	if(!stat("/etc/ptmp", &s)) {
+		fputs("vipw lockfile (/etc/ptmp) is present.", stderr);
+		fail_exit(1);
+	}
+
 #ifdef SHADOWPWD
 			if (is_shadow_pwd && sp_dbm_present ()
 			    && !sp_dbm_remove (user_name))
@@ -776,6 +781,7 @@
 			int main (int argc, char **argv) {
 			struct passwd *pwd; int arg; int errors = 0;
 			struct group *gr;
+           	        struct  stat    s;
 
 #ifdef USE_PAM
 			pam_handle_t * pamh = NULL;
--- shadow-4.0.3/src/useradd.c.vipw	2002-11-14 22:44:14.000000000 +0100
+++ shadow-4.0.3/src/useradd.c	2002-11-14 22:47:12.000000000 +0100
@@ -1856,6 +1856,11 @@
 	}
 #ifdef	NDBM
 
+	if(!stat("/etc/ptmp", &s)) {
+		puts("vipw lockfile (/etc/ptmp) is present!");
+		fail_exit(1);
+	}
+	
 	/*
 	 * Update the DBM group file with the new entry as well.
 	 */
@@ -1919,6 +1924,7 @@
 	struct passwd *pampw;
 	int retval;
 #endif
+ 	struct stat s;
 	/*
 	 * Get my name so that I can use it to report errors.
 	 */
