--- tetex-src-3.0/libs/xpdf/xpdf/Decrypt.cc.CAN-2005-0064	2004-01-22 02:26:45.000000000 +0100
+++ tetex-src-3.0/libs/xpdf/xpdf/Decrypt.cc	2005-02-08 10:09:10.899949464 +0100
@@ -135,6 +135,9 @@ GBool Decrypt::makeFileKey2(int encVersi
   int len, i, j;
   GBool ok;
 
+  // truncate keyLength when it won't fit tmpKey
+  keyLength = keyLength > sizeof(tmpKey) ? sizeof(tmpKey) : keyLength;
+
   // generate file key
   buf = (Guchar *)gmalloc(68 + fileID->getLength());
   if (userPassword) {
