Fix two security bugs
OpenPKG-2004.014-mysql, CAN-2004-0381, CAN-2004-0388

Index: scripts/mysqlbug.sh
--- scripts/mysqlbug.sh.orig	2004-02-10 19:15:54.000000000 +0100
+++ scripts/mysqlbug.sh	2004-04-14 21:12:44.000000000 +0200
@@ -254,7 +254,7 @@
 if cmp -s $TEMP $TEMP.x
 then
   echo "File not changed, no bug report submitted."
-  cp $TEMP /tmp/failed-mysql-bugreport
+  mv -f $TEMP /tmp/failed-mysql-bugreport
   echo "The raw bug report exists in /tmp/failed-mysql-bugreport"
   echo "If you use this remember that the first lines of the report now is a lie.."
   exit 1
Index: scripts/mysqld_multi.sh
--- scripts/mysqld_multi.sh.orig	2004-02-10 19:16:06.000000000 +0100
+++ scripts/mysqld_multi.sh	2004-04-14 22:08:34.000000000 +0200
@@ -9,7 +9,7 @@
 $opt_config_file   = undef();
 $opt_example       = 0;
 $opt_help          = 0;
-$opt_log           = "/tmp/mysqld_multi.log";
+$opt_log           = "$ENV{HOME}/mysqld_multi.log";
 $opt_mysqladmin    = "@bindir@/mysqladmin";
 $opt_mysqld        = "@libexecdir@/mysqld";
 $opt_no_log        = 0;
