Index: common.c
===================================================================
RCS file: /cvs/src/sasl/lib/common.c,v
retrieving revision 1.103
retrieving revision 1.104
diff -u -r1.103 -r1.104
--- lib/common.c	6 Jul 2004 16:03:05 -0000	1.103
+++ lib/common.c	22 Sep 2004 21:35:34 -0000	1.104
@@ -1838,7 +1838,10 @@
   if (! path)
     return SASL_BADPARAM;
 
-  *path = getenv(SASL_PATH_ENV_VAR);
+  /* Honor external variable only in a safe environment */
+  if (getuid() == geteuid() && getgid() == getegid())
+    *path = getenv(SASL_PATH_ENV_VAR);
+
   if (! *path)
     *path = PLUGINDIR;
 
