We still want a dynamic gssapi plugin.

--- cyrus-sasl-2.1.17/configure.in	2003-12-17 18:23:20.000000000 -0500
+++ cyrus-sasl-2.1.17/configure.in	2003-12-17 18:23:23.000000000 -0500
@@ -479,7 +479,11 @@
 SASL_PLAIN_CHK
 
 if test "$gssapi" != "no"; then
-  AC_DEFINE(STATIC_GSSAPIV2,[],[Link GSSAPI Staticly])
+  SASL_MECHS="$SASL_MECHS libgssapiv2.la"
+  if test "$enable_static" = yes; then
+    SASL_STATIC_OBJS="$SASL_STATIC_OBJS ../plugins/gssapi.o"
+    AC_DEFINE(STATIC_GSSAPIV2,[],[Link GSSAPI Staticly])
+  fi
 fi
 
 dnl ANONYMOUS
