diff -Naur net/ipsvd-0.9.6/src/Makefile net.oden/ipsvd-0.9.6/src/Makefile
--- net/ipsvd-0.9.6/src/Makefile	2004-08-02 21:46:18.000000000 +0200
+++ net.oden/ipsvd-0.9.6/src/Makefile	2004-08-04 20:54:43.791267112 +0200
@@ -24,24 +24,9 @@
 	./load ipsvd-cdb ipsvd_fmt.o ipsvd_check.o cdb.a dns.a unix.a byte.a \
 	  time.a `cat socket.lib`
 
-sslio: load sslio.o uidgid.o unix.a byte.a time.a matrixssl.a
-	./load sslio uidgid.o unix.a byte.a time.a matrixssl.a
-
-matrixssl.a: makelib matrixssl
-	install -m0644 Makefile.matrixssl matrixssl/src/Makefile.matrixssl
-	( cd matrixssl/src && $(MAKE) -fMakefile.matrixssl matrixssl.a )
-	install -m0644 matrixssl/src/matrixssl.a matrixssl.a
-
-matrixSsl.h: matrixssl
-	rm -f matrixSsl.h
-	ln -s matrixssl/matrixSsl.h matrixSsl.h
-
-matrixssl: matrixssl.tar.gz compile load
-	gzip -dc matrixssl.tar.gz |tar xf -
-	sed -e 's/#define USE_MULTITHREADING/#undef USE_MULTITHREADING/' \
-	  <matrixssl/src/matrixConfig.h >matrixConfig.h && \
-	    mv -f matrixConfig.h matrixssl/src/matrixConfig.h
-	touch matrixssl
+sslio: load sslio.o uidgid.o unix.a byte.a time.a /usr/lib/dietlibc/lib/libmatrixssl.a
+	cp /usr/lib/dietlibc/include/matrixSsl.h .
+	./load sslio uidgid.o unix.a byte.a time.a /usr/lib/dietlibc/lib/libmatrixssl.a
 
 check-tcpsvd: load check-tcpsvd.o unix.a byte.a socket.lib
 	./load check-tcpsvd unix.a byte.a `cat socket.lib`
@@ -61,7 +46,7 @@
 ipsvd-cdb.o: compile sysdeps ipsvd-cdb.c
 	./compile ipsvd-cdb.c
 
-sslio.o: compile sysdeps matrixSsl.h sslio.c
+sslio.o: compile sysdeps sslio.c
 	./compile sslio.c
 
 check-tcpsvd.o: compile check-tcpsvd.c
@@ -105,7 +90,6 @@
 
 command.linux: sysdeps
 	( ( test "`cut -c1-5 <systype`" != 'linux' && echo nolinux ) || \
-	  ( test ! -r matrixssl.tar.gz && echo nomatrixssl ) || \
 	  echo sslio ) >command.linux
 
 nolinux:
diff -Naur net/ipsvd-0.9.6/src/Makefile.matrixssl net.oden/ipsvd-0.9.6/src/Makefile.matrixssl
--- net/ipsvd-0.9.6/src/Makefile.matrixssl	2004-08-02 21:46:18.000000000 +0200
+++ net.oden/ipsvd-0.9.6/src/Makefile.matrixssl	1970-01-01 01:00:00.000000000 +0100
@@ -1,12 +0,0 @@
-include Makefile
-
-CC=../../compile
-LD=../../load
-LDFLAGS=
-DFLAGS=
-
-matrixssl.a: $(OBJECTS)
-	rm -f matrixssl.a
-	../../makelib matrixssl.a $(OBJECTS)
-#	ar cr libmatrixssl.a $(OBJECTS)
-#	ranlib libmatrixssl.a
