diff -ru tar-1.13.25.orig/Makefile.am tar-1.13.25/Makefile.am
--- tar-1.13.25.orig/Makefile.am	2001-08-29 07:23:19.000000000 +0200
+++ tar-1.13.25/Makefile.am	2002-08-09 05:32:35.000000000 +0200
@@ -18,7 +18,7 @@
 ## Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 ## 02111-1307, USA.
 
-AUTOMAKE_OPTIONS = gnits dist-bzip2 dist-shar
+AUTOMAKE_OPTIONS = gnits dist-shar
 ACLOCAL_AMFLAGS = -I m4
 EXTRA_DIST = ChangeLog.1 PORTS
 SUBDIRS = doc intl lib m4 src scripts po tests
