--- rpm-4.2.2/file/autogen.sh.fred	2003-12-16 06:00:41.000000000 +0100
+++ rpm-4.2.2/file/autogen.sh	2004-01-16 09:16:41.000000000 +0100
@@ -16,9 +16,9 @@
 recent versions of libtool/autoconf/automake.
 "
 
-[ "`libtoolize --version | head -1`" != "$LTV" ] && echo "$USAGE" && exit 1
-[ "`autoconf --version | head -1`" != "$ACV" ] && echo "$USAGE" && exit 1
-[ "`automake --version | head -1 | sed -e 's/1\.4[a-z]/1.4/'`" != "$AMV" ] && echo "$USAGE" && exit 1
+#[ "`libtoolize --version | head -1`" != "$LTV" ] && echo "$USAGE" && exit 1
+#[ "`autoconf --version | head -1`" != "$ACV" ] && echo "$USAGE" && exit 1
+#[ "`automake --version | head -1 | sed -e 's/1\.4[a-z]/1.4/'`" != "$AMV" ] && echo "$USAGE" && exit 1
 
 libtoolize --copy --force
 aclocal
