--- rpm-4.2/python/rpmdb/Makefile.in.fred	2003-08-07 16:54:19.000000000 +0200
+++ rpm-4.2/python/rpmdb/Makefile.in	2003-08-07 17:26:50.000000000 +0200
@@ -207,7 +207,7 @@
 
 PYVER = @WITH_PYTHON_VERSION@
 
-pylibdir = $(shell python -c 'import sys; print sys.path[1]')
+pylibdir = $(shell python -c 'import sys; print sys.path[2]')
 
 EXTRA_DIST = \
 	__init__.py dbobj.py db.py dbrecio.py dbshelve.py dbtables.py dbutils.py
--- rpm-4.2/python/Makefile.in.fred	2003-08-07 16:54:19.000000000 +0200
+++ rpm-4.2/python/Makefile.in	2003-08-07 17:27:07.000000000 +0200
@@ -209,7 +209,7 @@
 
 PYVER = @WITH_PYTHON_VERSION@
 
-pylibdir = $(shell python -c 'import sys; print sys.path[1]')
+pylibdir = $(shell python -c 'import sys; print sys.path[2]')
 pyincdir = $(prefix)/include/python${PYVER}
 
 SUBDIRS = rpmdb test
