--- rpm-4.0.4/rpmdb/db3.c.fred	2002-08-29 10:54:55.000000000 +0200
+++ rpm-4.0.4/rpmdb/db3.c	2002-08-30 14:15:22.000000000 +0200
@@ -1199,8 +1203,7 @@
 	     * acquire a lock, but permit failures, as some other
 	     * DBENV player may already have acquired the lock.
 	     */
-	    if (rc == 0 && dbi->dbi_lockdbfd &&
-		(!dbi->dbi_use_dbenv || _lockdbfd++ == 0))
+	    if (rc == 0 && dbi->dbi_lockdbfd)
 	    {
 		int fdno = -1;
 
