--- rpm-4.2/platform.in.opensls	2004-01-12 16:07:00.000000000 -0700
+++ rpm-4.2/platform.in	2004-01-12 16:48:48.000000000 -0700
@@ -239,3 +239,20 @@
 @mandrake@%_del_shell_helper /usr/share/rpm-helper/del-shell
 @mandrake@%_preun_shelldel() %_del_shell_helper %{name} $1 \
 @mandrake@%{nil}
+
+@mandrake@#---------------------------------------------------------------------
+@mandrake@#	OpenSLS vendor specific macros
+@mandrake@# 	Vincent Danen <vdanen@opensls.org>
+@mandrake@#
+@mandrake@#
+@mandrake@%_srvdir		%{_var}/service
+@mandrake@%_srvlogdir		%{_var}/log/supervise
+@mandrake@%_vendor		OpenSLS
+@mandrake@%_real_vendor		opensls
+@mandrake@%_target_platform	%{_target_cpu}-%{_real_vendor}-%{_target_os}%{?_gnu}
+@mandrake@
+@mandrake@# RPM extension
+@mandrake@%_ext			sls
+@mandrake@
+@mandrake@# Default to building with propolice
+@mandrake@%build_propolice %([ -n "$STACK_PROTECTOR" ] && echo 1 || echo 0)
