--- netpbm-10.26/lib/Makefile.libm	2004-12-07 23:32:37.000000000 -0500
+++ netpbm-10.26/lib/Makefile	2005-02-16 07:58:13.385624569 -0500
@@ -112,7 +112,7 @@ $(SONAME): libnetpbm.$(NETPBMLIBSUFFIX).
 	$(SYMLINK) $< $@
 libnetpbm.$(NETPBMLIBSUFFIX).$(MAJ).$(MIN): $(LIBOBJECTS) $(LIBOBJECTS_X)
 	$(LD) $(LDSHLIB) -o $@ $(LIBOBJECTS) $(LIBOBJECTS_X) \
-          $(SHLIB_CLIB) $(LADD)
+          $(SHLIB_CLIB) $(LADD) $(MATHLIB)
 endif
 
 ifeq ($(NETPBMLIBTYPE),dll)
