diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff t1/include/linux/fs.h w1/include/linux/fs.h
--- t1/include/linux/fs.h	2002-09-04 18:44:57.000000000 +0200
+++ w1/include/linux/fs.h	2002-09-04 19:13:46.000000000 +0200
@@ -1234,6 +1234,7 @@ extern void end_buffer_io_sync(struct bu
 
 /* reiserfs_writepage needs this */
 extern void set_buffer_async_io(struct buffer_head *bh) ;
+extern void end_buffer_io_sync(struct buffer_head *bh, int uptodate);
 
 #define BUF_CLEAN	0
 #define BUF_LOCKED	1	/* Buffers scheduled for write */
