--- perl-5.8.1-RC3/lib/ExtUtils/MM_Unix.pm.pix	2003-08-01 21:04:41.000000000 +0200
+++ perl-5.8.1-RC3/lib/ExtUtils/MM_Unix.pm	2003-08-01 21:05:05.000000000 +0200
@@ -620,6 +620,26 @@
     my $self = shift;
 
     # Initialize DESTDIR
+
+    # [Pixel] special rpmbuild adaptation to automagically set DESTDIR,
+    # [Pixel] so that perl modules build correctly with no modification
+    # [Pixel] this will ultimately be removed when perl module rpm packages
+    # [Pixel] use %makeinstall_std.
+    #
+    # [Pixel] A warning is added in ExtUtils::Install::install() to inform the
+    # [Pixel] packager of the correct way
+    if ($ENV{RPM_BUILD_ROOT}) {
+	if ($self->{PREFIX}) {
+	    # [Pixel] don't set DESTDIR when PREFIX since in that case,
+	    # [Pixel] MakeMaker handles PREFIX correctly in "make install
+	    # [Pixel] PREFIX=$RPM_BUILD_ROOT/usr"
+	    my $warn = 'do not set PREFIX in "perl Makefile.PL PREFIX=...", use %makeinstall_std instead';
+	    warn "*" x 80, "\nWARNING: $warn\n", "*" x 80, "\n";
+	} else {
+	    $self->{DESTDIR} ||= $ENV{RPM_BUILD_ROOT};
+	}
+    }
+
     $self->{DESTDIR} ||= '';
 
     # Make DEST variables.
--- perl-5.8.1-RC3/lib/ExtUtils/Install.pm.pix	2003-07-10 08:11:41.000000000 +0200
+++ perl-5.8.1-RC3/lib/ExtUtils/Install.pm	2003-08-01 21:04:41.000000000 +0200
@@ -84,6 +84,20 @@
     $verbose ||= 0;
     $nonono  ||= 0;
 
+    # [Pixel] help migration from "make PREFIX=$RPM_BUILD_ROOT/usr install" to "%makeinstall_std"
+    # [Pixel] see also the changes in ExtUtils::MM_Unix::init_DEST()
+    if ($ENV{RPM_BUILD_ROOT}) {
+	my $warn;
+	if ($ENV{MAKEFLAGS} =~ /PREFIX=/ && $ENV{MAKEFLAGS} !~ /DESTDIR=/) {
+	    $warn = 'use "%makeinstall_std" instead of "make PREFIX=$RPM_BUILD_ROOT/usr install" (which is obsolete)';
+	} elsif ($ENV{MAKEFLAGS} =~ /PREFIX=/) {
+	    $warn = 'using both PREFIX and DESTDIR is weird (in "make install")';
+	} elsif ($ENV{MAKEFLAGS} !~ /DESTDIR=/) {
+	    $warn = 'you should use DESTDIR (in "make install") (neither PREFIX nor DESTDIR is set, this is weird)';
+	}
+	warn "*" x 80, "\nWARNING: $warn\n", "*" x 80, "\n" if $warn;
+    }
+
     use Cwd qw(cwd);
     use ExtUtils::Packlist;
     use File::Basename qw(dirname);
