--- lib/Pod/Perldoc/ToMan.pm.orig	2004-06-23 17:18:23.000000000 +0200
+++ lib/Pod/Perldoc/ToMan.pm	2004-07-07 10:08:47.130148000 +0200
@@ -60,7 +60,7 @@
       unless -e $pod2man;
   }
 
-  my $command = "$pod2man $switches --lax $file | $render -man";
+  my $command = "$pod2man $switches --lax $file | $render -c -man";
          # no temp file, just a pipe!
 
   # Thanks to Brendan O'Dea for contributing the following block
@@ -77,6 +77,8 @@
   # don't have a -c switch, so that adding it here would presumably
   # be a Bad Thing   -- sburke@cpan.org
 
+  # [Pixel] our man do have the -c switch, using it!!
+
   $command .= " | col -x" if Pod::Perldoc::IS_HPUX;
   
   defined(&Pod::Perldoc::DEBUG)
