diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff q2/drivers/sound/awe_wave.c q1/drivers/sound/awe_wave.c
--- q2/drivers/sound/awe_wave.c	2001-09-30 21:26:08.000000000 +0200
+++ q1/drivers/sound/awe_wave.c	2002-07-31 00:34:51.000000000 +0200
@@ -4791,7 +4791,18 @@ static struct {
 	{0}
 };
 
-MODULE_DEVICE_TABLE(isapnp, isapnp_awe_list);
+static struct isapnp_device_id id_table[] __initdata = {
+	{	ISAPNP_ANY_ID, ISAPNP_ANY_ID,
+		ISAPNP_VENDOR('C','T','L'), ISAPNP_FUNCTION(0x0021), 0 },
+	{	ISAPNP_ANY_ID, ISAPNP_ANY_ID,
+		ISAPNP_VENDOR('C','T','L'), ISAPNP_FUNCTION(0x0022), 0 },
+	{	ISAPNP_ANY_ID, ISAPNP_ANY_ID,
+		ISAPNP_VENDOR('C','T','L'), ISAPNP_FUNCTION(0x0023), 0 },
+	{0}
+};
+
+
+MODULE_DEVICE_TABLE(isapnp, id_table);
 
 static struct pci_dev *idev = NULL;
 
