--- linux-2.4.22/drivers/sound/ali5455.c.bh05.orig	2003-09-06 23:06:08.000000000 +0200
+++ linux-2.4.22/drivers/sound/ali5455.c	2003-09-06 23:06:09.000000000 +0200
@@ -2936,7 +2936,9 @@ static u16 ali_ac97_get(struct ac97_code
 	struct ali_card *card = dev->private_data;
 	int count1 = 100;
 	char val;
-	unsigned short int data, count, addr1, addr2;
+	u16 count, addr1;
+	u16 addr2 = -1;
+	u16 data = -1;
 
 	spin_lock(&card->ac97_lock);
 	while (count1-- && (inl(card->iobase + ALI_CAS) & 0x80000000))
