diff -urN --exclude-from=/home/mitica/quintela/config/misc/dontdiff work/include/linux/kernel.h 2.4.19-rc1-x86_64/include/linux/kernel.h
--- work/include/linux/kernel.h	2002-07-04 14:21:15.000000000 +0200
+++ 2.4.19-rc1-x86_64/include/linux/kernel.h	2002-07-04 16:34:03.000000000 +0200
@@ -185,6 +185,8 @@
 	char _f[20-2*sizeof(long)-sizeof(int)];	/* Padding: libc5 uses this.. */
 };
 
+#define warn_if_not_ulong(x) do { unsigned long foo; (void) (&(x) == &foo); } while (0)
+
 #define BUG_ON(condition) do { if (unlikely((condition)!=0)) BUG(); } while(0)
 
 #endif /* _LINUX_KERNEL_H */
