--- linux-2.4.21/arch/i386/kernel/setup.c.orig	2003-08-24 04:36:45.000000000 +0200
+++ linux-2.4.21/arch/i386/kernel/setup.c	2003-08-24 04:38:26.000000000 +0200
@@ -856,6 +856,10 @@
 			if (!acpi_force) acpi_disabled = 1; 
 		} 
 
+		/* "acpi=on" re-enables both ACPI table parsing and interpreter init */
+		else if (!memcmp(from, "acpi=on", 7))
+			acpi_disabled = 0;
+
                 /* disable IO-APIC */
                 else if (!memcmp(from, "noapic", 6))
                         disable_ioapic_setup();
