--- linux-2.4.21-orig/drivers/i2c/scx200_i2c.c	2003-06-16 09:30:27.000000000 +0200
+++ linux-2.4.21-i2c/drivers/i2c/scx200_i2c.c	2003-07-11 10:54:39.000000000 +0200
@@ -76,16 +76,6 @@
 	return 0;
 }
 
-static void scx200_i2c_inc_use(struct i2c_adapter *adap)
-{
-	MOD_INC_USE_COUNT;
-}
-
-static void scx200_i2c_dec_use(struct i2c_adapter *adap)
-{
-	MOD_DEC_USE_COUNT;
-}
-
 /* ------------------------------------------------------------------------
  * Encapsulate the above functions in the correct operations structure.
  * This is only done when more than one hardware adapter is supported.
@@ -101,11 +91,10 @@
 };
 
 static struct i2c_adapter scx200_i2c_ops = {
+	.owner             = THIS_MODULE,
 	.name              = "NatSemi SCx200 I2C",
 	.id		   = I2C_HW_B_VELLE,
 	.algo_data	   = &scx200_i2c_data,
-	.inc_use	   = scx200_i2c_inc_use,
-	.dec_use	   = scx200_i2c_dec_use,
 	.client_register   = scx200_i2c_reg,
 	.client_unregister = scx200_i2c_unreg,
 };
