diff -ruN linux/arch/i386/kernel.orig/dmi_scan.c linux/arch/i386/kernel/dmi_scan.c
--- linux/arch/i386/kernel.orig/dmi_scan.c	Wed May 15 16:40:19 2002
+++ linux/arch/i386/kernel/dmi_scan.c	Wed May 15 16:46:56 2002
@@ -491,6 +491,20 @@
 }
 
 /*
+ *      keyboard reset for legacy free like option kbd-reset
+ */
+
+static int __init kbd_reset_legacy_free(struct dmi_blacklist *d)
+{
+	extern int kbd_startup_reset;
+	if (!kbd_startup_reset) {
+		kbd_startup_reset=1;
+		printk("Keyboard reset : Legacy Free detected\n");
+	}
+	return 0;
+}
+
+/*
  *	Simple "print if true" callback
  */
  
@@ -748,6 +762,18 @@
 			MATCH(DMI_PRODUCT_VERSION, "HP OmniBook XE3 GF"),
 			NO_MATCH, NO_MATCH
 			} },
+
+        { kbd_reset_legacy_free, "e-pc 43", { /* Keyboard reset */
+                        MATCH(DMI_SYS_VENDOR, "Hewlett-Packard"),
+                        MATCH(DMI_PRODUCT_NAME, "e-pc 43"),
+                        NO_MATCH, NO_MATCH
+                        } },
+
+        { kbd_reset_legacy_free, "Evo D-510 e-pc", { /* Keyboard reset */
+                        MATCH(DMI_SYS_VENDOR, "COMPAQ"),
+                        MATCH(DMI_PRODUCT_NAME, "Compaq Evo D510 e-pc"),
+                        NO_MATCH, NO_MATCH
+                        } },
 
 	/*
 	 *	Generic per vendor APM settings
