--- linux-2.4.20/include/asm-arm/arch-sa1100/SA-1101.h.za02-2.orig	2001-02-13 17:13:44.000000000 -0500
+++ linux-2.4.20/include/asm-arm/arch-sa1100/SA-1101.h	2003-04-14 23:22:08.000000000 -0400
@@ -143,7 +143,7 @@
 #define VMCCR_SDTCTest	    Fld(7,24)	  /* stale data timeout counter */
 #define VMCCR_ForceSelfRef  (1<<31)	  /* Force self refresh */
 
-#endif LANGUAGE == C
+#endif /* LANGUAGE == C */
 
 
 /* Update FIFO
--- linux-2.4.20/include/asm-parisc/som.h.za02-2.orig	2000-12-05 15:29:39.000000000 -0500
+++ linux-2.4.20/include/asm-parisc/som.h	2003-04-14 23:22:08.000000000 -0400
@@ -5,4 +5,4 @@
 #include <linux/som.h>
 
 
-#endif /* _ASM_PARISC_SOM_H
+#endif /* _ASM_PARISC_SOM_H */
--- linux-2.4.20/include/asm-cris/eshlibld.h.za02-2.orig	2001-04-06 13:51:19.000000000 -0400
+++ linux-2.4.20/include/asm-cris/eshlibld.h	2003-04-14 23:22:08.000000000 -0400
@@ -109,6 +109,6 @@ shlibmod_fork(struct shlibdep **deps);
 # define shlibmod_fork(x) 1
 #endif /* ! SHARE_LIB_CORE */
 
-#endif _cris_relocate_h
+#endif /* _cris_relocate_h */
 /********************** END OF FILE eshlibld.h *****************************/
 
--- linux-2.4.20/drivers/char/pcxx.c.za02-2.orig	2002-11-28 18:53:12.000000000 -0500
+++ linux-2.4.20/drivers/char/pcxx.c	2003-04-14 23:22:08.000000000 -0400
@@ -123,7 +123,7 @@ MODULE_PARM(altpin,      "1-4i");
 MODULE_PARM(numports,    "1-4i");
 # endif
 
-#endif MODULE
+#endif /* MODULE */
 
 static int numcards = 1;
 static int nbdevs = 0;
--- linux-2.4.20/drivers/sbus/char/aurora.c.za02-2.orig	2002-11-28 18:53:14.000000000 -0500
+++ linux-2.4.20/drivers/sbus/char/aurora.c	2003-04-14 23:22:08.000000000 -0400
@@ -674,7 +674,9 @@ static void aurora_check_modem(struct Au
 /* We don't have such things yet. My aurora board has DTR and RTS swapped, but that doesn't count in this driver. Let's hope
  * Aurora didn't made any boards with CTS or DSR broken...
  */
-/* #ifdef AURORA_BRAIN_DAMAGED_CTS
+
+#if 0
+#ifdef AURORA_BRAIN_DAMAGED_CTS
 	if (mcr & MCR_CTSCHG)  {
 		if (aurora_in(bp, CD180_MSVR) & MSVR_CTS)  {
 			tty->hw_stopped = 0;
@@ -699,7 +701,8 @@ static void aurora_check_modem(struct Au
 		}
 		sbus_writeb(port->SRER, &bp->r[chip]->r[CD180_SRER]);
 	}
-#endif AURORA_BRAIN_DAMAGED_CTS */
+#endif /* AURORA_BRAIN_DAMAGED_CTS */
+#endif /* 0 */
 	
 	/* Clear change bits */
 	sbus_writeb(0, &bp->r[chip]->r[CD180_MCR]);
--- linux-2.4.20/drivers/s390/char/tape.c.za02-2.orig	2001-07-25 17:12:02.000000000 -0400
+++ linux-2.4.20/drivers/s390/char/tape.c	2003-04-14 23:22:08.000000000 -0400
@@ -845,7 +845,7 @@ tape_init (void)
 #endif
 #ifdef CONFIG_DEVFS_FS
         tape_devfs_root_entry=devfs_mk_dir (NULL, "tape", NULL);
-#endif CONFIG_DEVFS_FS
+#endif /* CONFIG_DEVFS_FS */
 
 #ifdef TAPE_DEBUG
         debug_text_event (tape_debug_area,3,"dev detect");
@@ -994,7 +994,7 @@ cleanup_module (void)
 	}
 #ifdef CONFIG_DEVFS_FS
         devfs_unregister (tape_devfs_root_entry);
-#endif CONFIG_DEVFS_FS
+#endif /* CONFIG_DEVFS_FS */
 #ifdef CONFIG_PROC_FS
 #if (LINUX_VERSION_CODE > KERNEL_VERSION(2,3,98))
 	remove_proc_entry ("tapedevices", &proc_root);
--- linux-2.4.20/drivers/s390/net/netiucv.c.za02-2.orig	2002-11-28 18:53:14.000000000 -0500
+++ linux-2.4.20/drivers/s390/net/netiucv.c	2003-04-14 23:22:08.000000000 -0400
@@ -1746,7 +1746,7 @@ netiucv_proc_destroy_main(void)
 	proc_net_unregister(netiucv_dir.low_ino);
 #endif
 }
-#endif MODULE
+#endif /* MODULE */
 
 /**
  * Create a device specific subdirectory in /proc/net/iucv/ with the
