diff -Naur autofs-5.0.1-separate-config-files/samples/auto.master autofs-5.0.1-comment-default-master-map.patch/samples/auto.master
--- autofs-5.0.1-separate-config-files/samples/auto.master	2007-01-12 15:00:54.000000000 +0100
+++ autofs-5.0.1-comment-default-master-map.patch/samples/auto.master	2007-01-12 15:05:22.000000000 +0100
@@ -6,8 +6,8 @@
 # key [ -mount-options-separated-by-comma ] location
 # For details of the format look at autofs(5).
 #
-/misc	/etc/auto.misc
-/net	-hosts
+#/misc	/etc/autofs/auto.misc
+#/net	-hosts
 #
 # Include central master map if it can be found using
 # nsswitch sources.
