--- /dev/null	2001-01-02 16:30:38.000000000 -0700
+++ add-srv	2004-01-27 12:33:51.000000000 -0700
@@ -0,0 +1,33 @@
+#!/bin/sh
+
+# Annvix script to restart supervised service via rpm scriptlet
+
+if [ $# != 3 ]; then
+    echo "usage: $0 <pkg name> <number installed> <service name>" 1>&2
+    exit 1
+fi
+
+pkg=$1				# name of the package
+num=$2				# number of packages installed
+srv=$3				# name of the service
+
+if [ $num != 1 ]; then
+    # Upgrade mode: restart the service if already running
+    SRVSTAT=`/usr/bin/svstat /service/$srv|grep -q up; echo $?`
+    if [ "$SRVSTAT" = "0" ]; then
+	/usr/sbin/srv restart $srv || :
+	# restart services that depend of portmap
+	if [ $srv = portmap ]; then
+	    for s in amd autofs bootparamd clusternfs mcserv nfs nfslock ypserv ypbind yppasswdd ypxfrd; do
+		if [ -d /service/$s ]; then
+		    PSRVSTAT=`/usr/bin/svstat /service/$s|grep -q up; echo $?`
+		    if [ "$PSRVSTAT" = "0" ]; then
+			/usr/sbin/srv restart $srv || :
+		    fi
+		fi
+	    done
+	fi
+    fi
+fi
+
+# add-srv ends here
--- /dev/null	2001-01-02 16:30:38.000000000 -0700
+++ del-srv	2004-01-27 12:33:52.000000000 -0700
@@ -0,0 +1,21 @@
+#!/bin/sh
+
+# Annvix script to remove supervised service via rpm scriptlet
+
+if [ $# != 3 ]; then
+    echo "usage: $0 <pkg name> <number installed> <service name>" 1>&2
+    exit 1
+fi
+
+pkg=$1				# name of the package
+num=$2				# number of packages installed
+srv=$3				# name of the service
+
+if [ $num = 0 ]; then
+    if [ -d /service/$srv ]; then
+	/usr/sbin/srv stop $srv >/dev/null 2>&1 || :
+	/usr/sbin/srv del $srv
+    fi
+fi
+
+# del-srv ends here
--- Makefile~	2004-01-27 10:48:32.000000000 -0700
+++ Makefile	2004-01-27 10:48:32.000000000 -0700
@@ -15,7 +15,8 @@
 FILES = AUTHORS README README.CVS COPYING ChangeLog Makefile \
        $(PACKAGE).spec $(SCRIPTS)
 SCRIPTS = add-user del-user add-service del-service create-file \
-	add-group del-group add-shell del-shell verify-shell
+	add-group del-group add-shell del-shell verify-shell \
+	add-srv del-srv
 
 LIBDIR=/usr/share/rpm-helper
 
--- add-group.annvix	2002-12-22 16:08:41.000000000 -0700
+++ add-group	2004-02-03 20:17:46.000000000 -0700
@@ -11,16 +11,17 @@
 #---------------------------------------------------------------
 
 if [ $# -lt 3 ]; then
-    echo "usage: $0 <pkg name> <num installed> <group name> [<user1>,<user2>...]" 1>&2
+    echo "usage: $0 <pkg name> <num installed> <group name> <gid> [<user1>,<user2>...]" 1>&2
     exit 1
 fi
 
 pkg=$1				# name of the package
 num=$2				# number of packages installed
 name=$3				# name of the group
-users=$4			# users to add to this group
+gid=$4				# gid of group
+users=$5			# users to add to this group
 
-/usr/sbin/groupadd -r $name > /dev/null 2>&1
+/usr/sbin/groupadd -g $gid -r $name > /dev/null 2>&1
 
 if [ -n "$users" ]; then
     SAVED_IFS="$IFS"
--- /add-user.annvix	2002-12-22 14:46:07.000000000 -0700
+++ add-user	2004-02-03 20:17:50.000000000 -0700
@@ -11,7 +11,7 @@
 #---------------------------------------------------------------
 
 if [ $# -lt 5 ]; then
-    echo "usage: $0 <pkg name> <num installed> <user name> <home dir> <shell>" 1>&2
+    echo "usage: $0 <pkg name> <num installed> <user name> <home dir> <shell> <uid>" 1>&2
     exit 1
 fi
 
@@ -20,8 +20,9 @@
 name=$3				# name of the user
 dir=$4				# home directory
 shell=$5			# shell
+uid=$6				# uid to use
 
-/usr/sbin/useradd -r -M -s $shell -d $dir -c "system user for $pkg" $name > /dev/null 2>&1
+/usr/sbin/useradd -r -M -u $uid -s $shell -d $dir -c "system user for $pkg" $name > /dev/null 2>&1
 
 exit 0
 
