diff -urNp tclx8.3.orig/tcl/unix/Makefile.in tclx8.3/tcl/unix/Makefile.in
--- tclx8.3.orig/tcl/unix/Makefile.in	Tue Aug 29 05:58:00 2000
+++ tclx8.3/tcl/unix/Makefile.in	Wed Feb 27 06:51:33 2002
@@ -121,7 +121,7 @@ STANDALONE_OBJS = \
     stringfile.o     tcllib.o         fmath.o          buildhelp.o \
     buildidx.o       autoload.o       tclx.o
 
-HELP_DIR = @HELP_DIR@/tcl
+HELP_DIR = $(srcbasedir)/help/tcl
 
 TEST_OBJS = tclTest.o tclTestObj.o tclUnixTest.o tclXtest.o \
 	tclThreadTest.o tclXunixTest.o
@@ -311,14 +311,14 @@ testx.verbose: all
 BUILDHELP=@BUILDHELP@
 buildhelp: tcl tcl.tndx
 	rm -rf ${HELP_DIR} help help.tmp
-	if test ! -d @HELP_DIR@ ; then \
-	    mkdir @HELP_DIR@ ; \
-	fi
-	mkdir ${HELP_DIR}
-	${BLDMANHELP} ${TCL_SRC_DIR}/doc ${TCLX_TOOLS_SRC_DIR}/tclmanpages \
+	#if test ! -d @HELP_DIR@ ; then \
+	#    mkdir @HELP_DIR@ ; \
+	#fi
+	mkdir -p ${HELP_DIR}
+	#${BLDMANHELP} ${TCL_SRC_DIR}/doc ${TCLX_TOOLS_SRC_DIR}/tclmanpages \
 	    ${HELP_DIR} Tcl.brf
-	${RUNTCL} -c "buildhelp ${HELP_DIR} TclX.brf ${TCLX_DOC_DIR}/TclX.n"
-	${INSTCOPY} ${HELP_DIR} ${INSTALL_ROOT}${TCLX_INST_RUNTIME}/help
+	#${RUNTCL} -c "buildhelp ${HELP_DIR} TclX.brf ${TCLX_DOC_DIR}/TclX.n"
+	#${INSTCOPY} ${HELP_DIR} ${INSTALL_ROOT}${TCLX_INST_RUNTIME}/help
 
 #------------------------------------------------------------------------------
 
diff -urNp tclx8.3.orig/tk/unix/Makefile.in tclx8.3/tk/unix/Makefile.in
--- tclx8.3.orig/tk/unix/Makefile.in	Tue Aug 29 05:58:00 2000
+++ tclx8.3/tk/unix/Makefile.in	Wed Feb 27 06:51:50 2002
@@ -72,7 +72,7 @@ STANDALONE_OBJS = tkx.o
 LIBTCLX = ../../tcl/unix/${TCLX_STLIB_NAME}
 TCL = ../../tcl/unix/tcl
 
-HELP_DIR = @HELP_DIR@/tk
+HELP_DIR = $(srcbasedir)/tk/help
 
 TEST_OBJS = tkTest.o tkSquare.o tclThreadTest.o tkXunixTest.o
 
@@ -203,13 +203,13 @@ test.static: all
 BUILDHELP=@BUILDHELP@
 buildhelp:
 	rm -rf ${HELP_DIR} help help.tmp
-	if test ! -d @HELP_DIR@ ; then \
-	    mkdir @HELP_DIR@ ;\
+	#if test ! -d @HELP_DIR@ ; then \
+	#    mkdir @HELP_DIR@ ;\
 	fi
-	mkdir ${HELP_DIR}
-	${BLDMANHELP} ${TK_SRC_DIR}/doc ${TCLX_TOOLS_SRC_DIR}/tkmanpages \
+	mkdir -p ${HELP_DIR}
+	#${BLDMANHELP} ${TK_SRC_DIR}/doc ${TCLX_TOOLS_SRC_DIR}/tkmanpages \
 	    ${HELP_DIR} Tk.brf
-	${INSTCOPY} ${HELP_DIR} ${INSTALL_ROOT}${TKX_INST_RUNTIME}/help
+	#${INSTCOPY} ${HELP_DIR} ${INSTALL_ROOT}${TKX_INST_RUNTIME}/help
 
 #------------------------------------------------------------------------------
 
