--- Manip.pm.orig	2005-06-02 17:47:42.000000000 +0200
+++ Manip.pm	2005-06-03 09:58:42.986099352 +0200
@@ -460,7 +460,7 @@ sub Date_Init {
                     [$Lang{$L}{"DoMH"},1]);
 
     @tmp=();
-    foreach $tmp (keys %{ $Lang{$L}{"DoMH"} }) {
+    foreach $tmp (sort keys %{ $Lang{$L}{"DoMH"} }) {
       $tmp2=$Lang{$L}{"DoMH"}{$tmp};
       if ($tmp2<6) {
         $Lang{$L}{"WoMH"}{$tmp} = $tmp2;
@@ -777,7 +777,7 @@ sub Date_Init {
       $val=$Cnf{$key};
       push(@tmp,"$key=$val");
     }
-    foreach $key (keys %Cnf) {
+    foreach $key (sort keys %Cnf) {
       next  if (exists $tmp{$key});
       $val=$Cnf{$key};
       push(@tmp,"$key=$val");
@@ -1087,7 +1087,7 @@ sub ParseDateString {
     }
 
     # Check for some special types of dates (next, prev)
-    foreach $from (keys %{ $Lang{$L}{"Repl"} }) {
+    foreach $from (sort keys %{ $Lang{$L}{"Repl"} }) {
       $to=$Lang{$L}{"Repl"}{$from};
       s/(^|[^a-z])$from($|[^a-z])/$1$to$2/i;
     }
@@ -1663,7 +1663,7 @@ sub ParseDateDelta {
     }
     $workweek=7  if ($Curr{"Mode"} != 2);
 
-    foreach $from (keys %{ $Lang{$Cnf{"Language"}}{"Repl"} }) {
+    foreach $from (sort keys %{ $Lang{$Cnf{"Language"}}{"Repl"} }) {
       $to=$Lang{$Cnf{"Language"}}{"Repl"}{$from};
       s/(^|[^a-z])$from($|[^a-z])/$1$to$2/i;
     }
@@ -3865,7 +3865,7 @@ sub Date_InitHash {
 
   # Create the regular expression
   if ($regexp) {
-    @list=keys(%$hash);
+    @list=sort keys(%$hash);
     @list=sort sortByLength(@list)  if ($sort);
     if ($escape) {
       foreach $val (@list) {
@@ -3958,7 +3958,7 @@ sub Date_InitLists {
 
   # Create the regular expression
   if ($regexp) {
-    @list=keys(%hash);
+    @list=sort keys(%hash);
     @list=sort sortByLength(@list)  if ($sort);
     if ($escape) {
       foreach $ele (@list) {
@@ -4144,7 +4144,7 @@ sub Date_Regexp {
   @list=split(/&&&/,$list);
   if ($keys) {
     %hash=@list;
-    @list=keys %hash;
+    @list=sort keys %hash;
   } elsif ($key0 or $key1 or $val0 or $val1) {
     $i=0;
     $i=1  if ($key1 or $val1);
@@ -5028,7 +5028,7 @@ sub Date_UpdateHolidays {
   my($date,$delta,$err)=();
   my($key,@tmp,$tmp);
 
-  foreach $key (keys %{ $Holiday{"desc"} }) {
+  foreach $key (sort keys %{ $Holiday{"desc"} }) {
     @tmp=&Recur_Split($key);
     if (@tmp) {
       $tmp=&ParseDateString("${year}010100:00:00");
@@ -5211,7 +5211,7 @@ sub Events_Calc {
       $tmp =~ s/^!//;
       delete $tmp{$tmp};
     }
-    push(@ret,$date,[ keys %tmp ]);
+    push(@ret,$date,[ sort keys %tmp ]);
   }
 
   %tmp = @ret;
