--- device-mapper.1.02.07/lib/datastruct/hash.c.bluca	2006-01-09 21:35:24.000000000 +0100
+++ device-mapper.1.02.07/lib/datastruct/hash.c	2006-05-21 15:39:49.000000000 +0200
@@ -68,14 +68,14 @@
 	return n;
 }
 
-static unsigned long _hash(const unsigned char *str, unsigned len)
+static unsigned long _hash(const char *str, unsigned len)
 {
 	unsigned long h = 0, g;
 	unsigned i;
 
 	for (i = 0; i < len; i++) {
 		h <<= 4;
-		h += _nums[*str++];
+		h += _nums[(unsigned char) *str++];
 		g = h & ((unsigned long) 0xf << 16u);
 		if (g) {
 			h ^= g >> 16u;
--- device-mapper.1.02.07/lib/ioctl/libdm-iface.c.bluca	2006-04-03 17:56:02.000000000 +0200
+++ device-mapper.1.02.07/lib/ioctl/libdm-iface.c	2006-05-21 15:23:07.000000000 +0200
@@ -1542,7 +1542,7 @@
 		dmi->flags |= DM_SKIP_BDGET_FLAG;
 
 	log_debug("dm %s %s %s%s%s %s%.0d%s%.0d%s"
-		  "%s%c %.0llu %s [%u]",
+		  "%s%c %.0" PRIu64 " %s [%u]",
 		  _cmd_data_v4[dmt->type].name,
 		  dmi->name, dmi->uuid, dmt->newname ? " " : "",
 		  dmt->newname ? dmt->newname : "",
