--- device-mapper.1.02.07/dmeventd/Makefile.in.pk	2006-05-21 14:55:35.000000000 +0200
+++ device-mapper.1.02.07/dmeventd/Makefile.in	2006-05-21 15:11:22.000000000 +0200
@@ -55,7 +55,7 @@
 
 install_pkgconfig:
 	$(INSTALL) -D $(OWNER) $(GROUP) -m 444 libdevmapper-event.pc \
-		$(usrlibdir)/pkgconfig/devmapper-event.pc
+		$(libdir)/pkgconfig/devmapper-event.pc
 
 install_static: libdevmapper-event.a
 	$(INSTALL) -D $(OWNER) $(GROUP) -m 555 $(STRIP) $< \
--- device-mapper.1.02.07/lib/Makefile.in.pk	2006-05-21 14:55:35.000000000 +0200
+++ device-mapper.1.02.07/lib/Makefile.in	2006-05-21 15:11:28.000000000 +0200
@@ -88,7 +88,7 @@
 
 install_pkgconfig:
 	$(INSTALL) -D $(OWNER) $(GROUP) -m 444 libdevmapper.pc \
-		$(usrlibdir)/pkgconfig/devmapper.pc
+		$(libdir)/pkgconfig/devmapper.pc
 
 install_ioctl_static: ioctl/libdevmapper.a
 	$(INSTALL) -D $(OWNER) $(GROUP) -m 555 $(STRIP) $< \
--- device-mapper.1.02.07/lib/libdevmapper.pc.in.pk	2006-04-19 17:23:10.000000000 +0200
+++ device-mapper.1.02.07/lib/libdevmapper.pc.in	2006-05-21 15:07:01.000000000 +0200
@@ -8,4 +8,4 @@
 Version: @DM_LIB_VERSION@
 Cflags: -I${includedir} 
 Libs: -L${libdir} -ldevmapper
-Libs.private: -lselinux -lsepol
+Libs.private: @SELIBS@
--- device-mapper.1.02.07/configure.in.pk	2006-05-21 14:55:35.000000000 +0200
+++ device-mapper.1.02.07/configure.in	2006-05-21 15:06:10.000000000 +0200
@@ -229,6 +229,7 @@
 	if test x$HAVE_SEPOL = xyes; then
 		AC_DEFINE([HAVE_SEPOL], 1, [Define to 1 if sepol_check_context is available.])
 		LIBS="-lsepol $LIBS"
+		SELIBS="-lsepol $SELIBS"
 	fi
 
 	AC_CHECK_LIB(selinux, is_selinux_enabled, HAVE_SELINUX=yes, HAVE_SELINUX=no)
@@ -236,6 +237,7 @@
 	if test x$HAVE_SELINUX = xyes; then
 		AC_DEFINE([HAVE_SELINUX], 1, [Define to 1 if selinux is available.])
 		LIBS="-lselinux $LIBS"
+		SELIBS="-lselinux $SELIBS"
 	else
 		AC_MSG_WARN(Disabling selinux)
 	fi
@@ -366,6 +368,7 @@
 AC_SUBST(OWNER)
 AC_SUBST(GROUP)
 AC_SUBST(LIBS)
+AC_SUBST(SELIBS)
 AC_SUBST(interface)
 AC_SUBST(kerneldir)
 AC_SUBST(missingkernel)
