--- util-linux-2.11o/misc-utils/Makefile.nochkdupexe	Sun Jul  1 20:59:08 2001
+++ util-linux-2.11o/misc-utils/Makefile	Mon Aug  6 17:48:27 2001
@@ -11,7 +11,7 @@
 
 # Where to put man pages?
 
-MAN1= 		cal.1 chkdupexe.1 ddate.1 logger.1 look.1 mcookie.1 \
+MAN1= 		cal.1 ddate.1 logger.1 look.1 mcookie.1 \
 		namei.1 rename.1 script.1 whereis.1
 
 # Where to put binaries?
@@ -19,7 +19,7 @@
 
 BIN=
 
-USRBIN=		cal chkdupexe ddate logger look mcookie \
+USRBIN=		cal ddate logger look mcookie \
 		namei rename script whereis
 
 MAYBE=		reset setterm
@@ -79,7 +79,6 @@
 
 cal.o: $(LIB)/errs.h
 cal: cal.o $(ERR_O)
-chkdupexe: chkdupexe.pl
 ddate: ddate.o
 kill: kill.o procs.o
 logger: logger.o
