diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/crypto.c w1/3rdparty/bcm5820/crypto.c
--- w2/3rdparty/bcm5820/crypto.c	2002-08-08 15:33:43.000000000 +0200
+++ w1/3rdparty/bcm5820/crypto.c	2003-08-20 16:08:40.000000000 +0200
@@ -175,7 +175,9 @@ do_encrypt(ubsec_DeviceContext_t pContex
   unsigned int in_packet_size;
   unsigned int out_packet_size;
   int MacSize=0;
+#ifndef GOTOSLEEP
   unsigned long delay_total_us;
+#endif
   ubsec_CipherCommandInfo_t *ubsec_commands = NULL;
 
 #ifndef STATIC_ALLOC_OF_CRYPTO_BUFFERS
@@ -604,7 +606,9 @@ if ( UBSEC_USING_MAC(at->flags) ){
     udelay(1);
   }
 #endif
+#ifdef DEBUG_TIMEOUT
 skip_error_ret:
+#endif
 
   if (UBSEC_USING_MAC( at->flags ) && (at->flags & UBSEC_DECODE)) {
 
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/dispatch.c w1/3rdparty/bcm5820/dispatch.c
--- w2/3rdparty/bcm5820/dispatch.c	2003-08-20 19:14:04.000000000 +0200
+++ w1/3rdparty/bcm5820/dispatch.c	2003-08-20 16:07:06.000000000 +0200
@@ -38,8 +38,6 @@
  * Dec   2001 SRM stats ioctl
  */
 
-#define MODULE
-
 #define NUMBER_OF_KEY_MCRS 128
 #define NUMBER_OF_CRYPTO_MCRS 512 
 
@@ -139,15 +137,11 @@ ubsec_ioctl(struct inode *inode,struct f
   int                   status = 0;
   int                   deadlockctr = 0;
   
-  unsigned short        value;
-
   
   
   /* Simple round robin scheduling of device. We need to increment
      first since the keysetup command may block. */
   
- TheBeginning:
-
   deadlockctr = 0;
   do {
 
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/hash.c w1/3rdparty/bcm5820/hash.c
--- w2/3rdparty/bcm5820/hash.c	2002-08-08 15:33:43.000000000 +0200
+++ w1/3rdparty/bcm5820/hash.c	2003-08-20 16:08:09.000000000 +0200
@@ -178,7 +178,9 @@ ubsec_hash(
   volatile CommandContext_t  CommandContext;
   CommandContext_t          *pCommandContext   = (CommandContext_t *)&CommandContext;
   int                            NumCommands       = 1;
+#ifndef GOTOSLEEP
   unsigned long		         delayTotalUs      = 0;
+#endif
   ubsec_FragmentInfo_pt          pSourceFragments  = NULL;
   unsigned char *                pUserSourceBuffer = NULL;
   unsigned int                   SourceBufferBytes = 0;
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/pkey.c w1/3rdparty/bcm5820/pkey.c
--- w2/3rdparty/bcm5820/pkey.c	2002-08-08 15:33:43.000000000 +0200
+++ w1/3rdparty/bcm5820/pkey.c	2003-08-20 16:09:45.000000000 +0200
@@ -69,7 +69,6 @@ int ubsec_keysetup(ubsec_DeviceContext_t
   unsigned char			*pkey_buf = NULL;
   ubsec_key_io_t                KeyIOInfoForDSA;
   int                           add_dsa_buf_bytes = 0;
-  int i;
   int timeout;
   int message_alignment;  
 
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/rng.c w1/3rdparty/bcm5820/rng.c
--- w2/3rdparty/bcm5820/rng.c	2002-08-08 15:33:43.000000000 +0200
+++ w1/3rdparty/bcm5820/rng.c	2003-08-20 16:12:32.000000000 +0200
@@ -102,7 +102,9 @@ ubsec_rng(ubsec_DeviceContext_t *ubsecCo
   ubsec_RNGCommandInfo_pt	kcmd;
   ubsec_RNGCommandParams_pt	pRngparams=NULL, 
 				pIOparams = NULL;
+#ifndef GOTOSLEEP
   unsigned long			delay_total_us;
+#endif
   unsigned char			*RngLoc;
   unsigned int			num_commands=1;
   ubsec_rng_io_t		RngCommand;
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/sslarc4.c w1/3rdparty/bcm5820/sslarc4.c
--- w2/3rdparty/bcm5820/sslarc4.c	2002-08-08 15:33:43.000000000 +0200
+++ w1/3rdparty/bcm5820/sslarc4.c	2003-08-20 16:13:04.000000000 +0200
@@ -184,7 +184,9 @@ ubsec_sslarc4(
   volatile  CommandContext_t  CommandContext;
   CommandContext_t          *pCommandContext   = (CommandContext_t *)&CommandContext;
   int                           NumCommands       = 1;
+#ifndef GOTOSLEEP
   unsigned long		        delayTotalUs      = 0;
+#endif
   ubsec_FragmentInfo_pt         pSourceFragments  = NULL;
   unsigned char *               pUserSourceBuffer;
   unsigned int                  SourceBufferBytes;
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/sslcipher.c w1/3rdparty/bcm5820/sslcipher.c
--- w2/3rdparty/bcm5820/sslcipher.c	2002-08-08 15:33:43.000000000 +0200
+++ w1/3rdparty/bcm5820/sslcipher.c	2003-08-20 16:31:50.000000000 +0200
@@ -171,12 +171,13 @@ ubsec_sslcipher(
   volatile  CommandContext_t  CommandContext;
   CommandContext_t          *pCommandContext   = (CommandContext_t *)&CommandContext;
   int                            NumCommands       = 1;
+#ifndef GOTOSLEEP
   unsigned long		         delayTotalUs      = 0;
+#endif
   ubsec_FragmentInfo_pt          pSourceFragments  = NULL;
   unsigned char *                pUserSourceBuffer = NULL;
   unsigned int                   SourceBufferBytes = 0;
   int                            error             = 0;
-  int                            i                 = 0;
 
 #ifndef STATIC_ALLOC_OF_CRYPTO_BUFFERS
   
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/sslmac.c w1/3rdparty/bcm5820/sslmac.c
--- w2/3rdparty/bcm5820/sslmac.c	2002-08-08 15:33:43.000000000 +0200
+++ w1/3rdparty/bcm5820/sslmac.c	2003-08-20 16:32:37.000000000 +0200
@@ -180,7 +180,9 @@ ubsec_sslmac(
   volatile  CommandContext_t  CommandContext;
   CommandContext_t          *pCommandContext   = (CommandContext_t *)&CommandContext;
   int                        NumCommands       = 1;
+#ifndef GOTOSLEEP
   unsigned long		     delayTotalUs      = 0;
+#endif
   ubsec_FragmentInfo_pt      pSourceFragments  = NULL;
   unsigned char *            pUserSourceBuffer = NULL;
   unsigned int               SourceBufferBytes = 0;
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/tlsmac.c w1/3rdparty/bcm5820/tlsmac.c
--- w2/3rdparty/bcm5820/tlsmac.c	2002-08-08 15:33:43.000000000 +0200
+++ w1/3rdparty/bcm5820/tlsmac.c	2003-08-20 16:33:33.000000000 +0200
@@ -166,7 +166,9 @@ ubsec_tlsmac(
   volatile  CommandContext_t  CommandContext;
   CommandContext_t          *pCommandContext   = (CommandContext_t *)&CommandContext;
   int                           NumCommands       = 1;
+#ifndef GOTOSLEEP
   unsigned long		        delayTotalUs      = 0;
+#endif
   ubsec_FragmentInfo_pt         pSourceFragments  = NULL;
   unsigned char *               pUserSourceBuffer = NULL;
   unsigned int                  SourceBufferBytes = 0;
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/ubsdh.c w1/3rdparty/bcm5820/ubsdh.c
--- w2/3rdparty/bcm5820/ubsdh.c	2002-08-08 15:33:43.000000000 +0200
+++ w1/3rdparty/bcm5820/ubsdh.c	2003-08-20 16:37:16.000000000 +0200
@@ -75,9 +75,11 @@ DH_SetupPublicParams(VOLATILE MasterComm
   ubsec_Status_t Status=UBSEC_STATUS_SUCCESS;
   int NgLen;
   int NormalizeLen,NrmBits;
+#if defined(UBS_ENABLE_KEY_SWAP)
   int element;
-  unsigned char *pNg;
   UBS_UINT32 *longkey;
+#endif
+  unsigned char *pNg;
   ubsec_MemAddress_t PhysAddr;
 
   PacketIndex = pMCR->NumberOfPackets; 
@@ -347,9 +349,11 @@ DH_SetupSharedParams(VOLATILE MasterComm
   DH_REC_CtxCmdBuf_pt pDHRecCtx;
   VOLATILE int             	PacketIndex;
   int NgLen;
+#if defined(UBS_ENABLE_KEY_SWAP)
   int element;
-  int NormalizeLen,NrmBits=0;
   UBS_UINT32 *longkey;
+#endif
+  int NormalizeLen,NrmBits=0;
   ubsec_MemAddress_t PhysAddr;
 
   PacketIndex = pMCR->NumberOfPackets; 
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/ubsdsa.c w1/3rdparty/bcm5820/ubsdsa.c
--- w2/3rdparty/bcm5820/ubsdsa.c	2002-08-08 15:33:43.000000000 +0200
+++ w1/3rdparty/bcm5820/ubsdsa.c	2003-08-20 16:37:50.000000000 +0200
@@ -74,9 +74,11 @@ DSA_SetupSignParams(MasterCommand_pt pMC
   int                  PacketIndex;
   int fragnum; 
   int Offset;
+#if defined(UBS_ENABLE_KEY_SWAP)
   int element;
-  int NormBits,NormalizeLen;
   UBS_UINT32 *longkey;
+#endif
+  int NormBits,NormalizeLen;
   ubsec_MemAddress_t PhysAddr;
 
   PacketIndex = pMCR->NumberOfPackets;
@@ -388,9 +390,11 @@ DSA_SetupVerifyParams(MasterCommand_pt p
   int                  PacketIndex;
   int fragnum; 
   int Offset;
+#if defined(UBS_ENABLE_KEY_SWAP)
   int element;
-  int NormBits,NormalizeLen;
   UBS_UINT32 *longkey;
+#endif
+  int NormBits,NormalizeLen;
   ubsec_MemAddress_t PhysAddr;
 
   /* First do a sanity check for an excessively long input fragment list */
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/ubskey.c w1/3rdparty/bcm5820/ubskey.c
--- w2/3rdparty/bcm5820/ubskey.c	2002-08-08 15:33:43.000000000 +0200
+++ w1/3rdparty/bcm5820/ubskey.c	2003-08-20 16:41:19.000000000 +0200
@@ -278,9 +278,14 @@ ubsec_KeyCommand(ubsec_DeviceContext_t C
 void KeyFinishResult(unsigned long Context,ubsec_Status_t Result)
 {
   KeyContext_pt pContext=(KeyContext_pt)Context;
+#ifndef UBSEC_HW_NORMALIZE
   long NormBits_RNG;
-  int key_num, element_num, native_endianess;
+#endif
+#if defined(UBS_ENABLE_KEY_SWAP) 
+  int element_num;
   UBS_UINT32 *key_array;
+#endif
+  int key_num, native_endianess;
 
   /* We are assuming that the KeyLength field value passed in refers to a key array element that
      is not less significant (lower array index) than the actual most significant array element,
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/ubskeytest.c w1/3rdparty/bcm5820/ubskeytest.c
--- w2/3rdparty/bcm5820/ubskeytest.c	2002-08-08 15:33:43.000000000 +0200
+++ w1/3rdparty/bcm5820/ubskeytest.c	2003-08-20 16:42:57.000000000 +0200
@@ -37,10 +37,6 @@
 
 #include "ubsincl.h"
 
-#ifndef ASYNC
-static void callback(unsigned long PacketContext,ubsec_Status_t Result);
-#endif
-
 /* computing Y=G**X/N */
 
 /*  1. base DHP g - G */
@@ -98,7 +94,7 @@ ubsec_TestKeyDevice(ubsec_DeviceContext_
   unsigned char* pUserX;
   unsigned char* pY;
   int num_packets=1;
-  int i, Status;
+  int Status;
   UBS_UINT32 SaveConfig;
   DeviceInfo_pt pDevice=(DeviceInfo_pt)Ubsec_Context;
 
@@ -276,7 +272,9 @@ static ubsec_Status_t
 CompareKeyResults(void)
 {
   unsigned char *pX, *pY;
+#if defined(UBS_OVERRIDE_LONG_KEY_MODE)
   int i;
+#endif
 
   pY = (unsigned char*)OS_GetVirtualAddress(Y_Handle);
   pX = (unsigned char*)OS_GetVirtualAddress(X_Handle);
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/ubslinux.h w1/3rdparty/bcm5820/ubslinux.h
--- w2/3rdparty/bcm5820/ubslinux.h	2002-08-08 15:33:43.000000000 +0200
+++ w1/3rdparty/bcm5820/ubslinux.h	2003-08-20 20:06:13.000000000 +0200
@@ -106,12 +106,6 @@ extern unsigned long LinuxGetVirtualAddr
 /*************************** 
  * Critical section functions
  ***************************/ 
-#if 0
-extern void 		LinuxInitCriticalSection(OS_DeviceInfo_t);
-extern unsigned long 	LinuxEnterCriticalSection(OS_DeviceInfo_t);
-extern unsigned long 	LinuxTestCriticalSection(OS_DeviceInfo_t);
-extern void  		LinuxLeaveCriticalSection(OS_DeviceInfo_t);
-#endif
 
 #define OS_InitCriticalSection(pDevice)     	       LinuxInitCriticalSection(pDevice->OsDeviceInfo)
 #define OS_EnterCriticalSection(pDevice,SaveConfig)    LinuxEnterCriticalSection(pDevice->OsDeviceInfo)
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/ubsmath.c w1/3rdparty/bcm5820/ubsmath.c
--- w2/3rdparty/bcm5820/ubsmath.c	2002-08-08 15:33:43.000000000 +0200
+++ w1/3rdparty/bcm5820/ubsmath.c	2003-08-20 16:53:39.000000000 +0200
@@ -63,8 +63,10 @@ ubsec_MathCommand(ubsec_DeviceContext_t 
   VOLATILE DataBufChainList_t   *FragPtr, *NextFragPtr;
   int DataLength;
   int NormalizeLen,NrmBits=0;
+#if defined(UBS_ENABLE_KEY_SWAP) 
   int element;
   UBS_UINT32 *longkey;
+#endif
   ubsec_MemAddress_t PhysAddr;
   CallBackInfo_pt pCompletionContext;
 
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/ubsrng.c w1/3rdparty/bcm5820/ubsrng.c
--- w2/3rdparty/bcm5820/ubsrng.c	2002-08-08 15:33:43.000000000 +0200
+++ w1/3rdparty/bcm5820/ubsrng.c	2003-08-20 16:54:34.000000000 +0200
@@ -241,7 +241,7 @@ ubsec_RNGCommand(ubsec_DeviceContext_t C
 void RNGFinishResult(unsigned long Context,ubsec_Status_t Result)
 {
   KeyContext_pt pContext=(KeyContext_pt)Context;
-  UBS_UINT32 rng_mask, *key_array;
+  UBS_UINT32 rng_mask;
 
   if ((Result==UBSEC_STATUS_SUCCESS) && (pContext->ResultRNG != NULL)) {
     
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/ubsrsa.c w1/3rdparty/bcm5820/ubsrsa.c
--- w2/3rdparty/bcm5820/ubsrsa.c	2002-08-08 15:33:43.000000000 +0200
+++ w1/3rdparty/bcm5820/ubsrsa.c	2003-08-20 16:55:29.000000000 +0200
@@ -77,10 +77,11 @@ RSA_SetupPublicParams(MasterCommand_pt p
   VOLATILE Packet_t 		*pPacket;
   Pub_RSA_CtxCmdBuf_pt		pRSACtx;
   VOLATILE int                  PacketIndex;
-  ubsec_Status_t Status=UBSEC_STATUS_SUCCESS;
   int ModNLen,NormBits;
+#if defined(UBS_ENABLE_KEY_SWAP)
   int element;
   UBS_UINT32 *longkey;
+#endif
   ubsec_MemAddress_t PhysAddr;
 
   PacketIndex = pMCR->NumberOfPackets;
@@ -245,9 +246,11 @@ RSA_SetupPrivateParams(MasterCommand_pt 
   VOLATILE Pri_RSA_CtxCmdBuf_t	*pRSACtx;
   int                  PacketIndex;
   int Offset;
+#if defined(UBS_ENABLE_KEY_SWAP)
   int element;
-  int ParamLen,NormBits;
   UBS_UINT32 *longkey;
+#endif
+  int ParamLen,NormBits;
   ubsec_MemAddress_t PhysAddr;
 
   PacketIndex = pMCR->NumberOfPackets;
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/ubsssl.c w1/3rdparty/bcm5820/ubsssl.c
--- w2/3rdparty/bcm5820/ubsssl.c	2002-08-08 15:33:43.000000000 +0200
+++ w1/3rdparty/bcm5820/ubsssl.c	2003-08-20 16:56:08.000000000 +0200
@@ -58,8 +58,6 @@ ubsec_SSLCommand(ubsec_DeviceContext_t C
   VOLATILE SSL_CryptoContext_t *pSSLCryptoContext;
   VOLATILE ARC4_CryptoContext_t *pARC4Context;
   VOLATILE Hash_Context_t *pHashContext;
-  int i;
-  long *plong;
   VOLATILE int             PacketIndex;
 
   int CommandIndex=0;
diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/3rdparty/bcm5820/ubstest.c w1/3rdparty/bcm5820/ubstest.c
--- w2/3rdparty/bcm5820/ubstest.c	2002-08-08 15:33:43.000000000 +0200
+++ w1/3rdparty/bcm5820/ubstest.c	2003-08-18 21:02:19.000000000 +0200
@@ -37,9 +37,6 @@
 
 #include "ubsincl.h"
 
-#ifndef ASYNC
-static void callback(unsigned long PacketContext,ubsec_Status_t Result);
-#endif
 #define UBS_SELFTEST_MD5   /* Undefine for SHA1 authentication */
 
 static unsigned char InitialVector[]={1, 2, 3, 4, 5, 6, 7, 8 };
