--- setarch-1.3/setarch.c.linux64	2003-09-02 14:14:37.000000000 +0200
+++ setarch-1.3/setarch.c	2004-02-11 12:21:12.085952030 +0100
@@ -25,6 +25,9 @@ set_arch(const char *pers, unsigned long
     char *target_arch, *result_arch;
   } transitions[] = {
     {PER_LINUX32, "linux32", NULL},
+#if defined(__LP64__)
+    {PER_LINUX, "linux64", NULL},
+#endif
 #if defined(__powerpc__) || defined(__powerpc64__)
     {PER_LINUX32, "ppc32", "ppc"},
     {PER_LINUX32, "ppc", "ppc"},
