--- sh-utils-2.0.12/src/Makefile.am.hname	2002-07-31 16:10:05.000000000 +0200
+++ sh-utils-2.0.12/src/Makefile.am	2002-07-31 16:10:27.000000000 +0200
@@ -7,7 +7,7 @@
 bin_SCRIPTS = groups @OPTIONAL_BIN_ZCRIPTS@
 
 bin_PROGRAMS = basename date dirname echo env expr factor false \
-  hostname id kill logname pathchk printenv printf pwd seq sleep tee \
+  id kill logname pathchk printenv printf pwd seq sleep tee \
   test true tty whoami yes @OPTIONAL_BIN_PROGS@
 
 localedir = $(datadir)/locale
--- sh-utils-2.0.12/man/Makefile.am.hname	2002-07-31 16:10:41.000000000 +0200
+++ sh-utils-2.0.12/man/Makefile.am	2002-07-31 16:10:54.000000000 +0200
@@ -1,7 +1,7 @@
 ## Process this file with automake to produce Makefile.in -*-Makefile-*-
 dist_man_MANS = \
   basename.1 chroot.1 date.1 dirname.1 echo.1 env.1 expr.1 factor.1 \
-  false.1 groups.1 hostid.1 hostname.1 id.1 logname.1 nice.1 nohup.1 \
+  false.1 groups.1 hostid.1 id.1 logname.1 nice.1 nohup.1 \
   pathchk.1 pinky.1 printenv.1 printf.1 pwd.1 seq.1 sleep.1 stty.1 su.1 \
   tee.1 test.1 true.1 tty.1 uname.1 uptime.1 users.1 who.1 whoami.1 yes.1
 man_aux = $(dist_man_MANS:.1=.x)
@@ -23,7 +23,6 @@
 false.1:	$(common_dep)	$(srcdir)/false.x	../src/false.c
 groups.1:	$(common_dep)	$(srcdir)/groups.x	../src/groups.sh
 hostid.1:	$(common_dep)	$(srcdir)/hostid.x	../src/hostid.c
-hostname.1:	$(common_dep)	$(srcdir)/hostname.x	../src/hostname.c
 id.1:		$(common_dep)	$(srcdir)/id.x		../src/id.c
 logname.1:	$(common_dep)	$(srcdir)/logname.x	../src/logname.c
 nice.1:		$(common_dep)	$(srcdir)/nice.x	../src/nice.c
