diff -X ../../KDIFX -burpN -X ../../KDIFX linux-vanilla/kernel/timer.c linux/kernel/timer.c
--- linux-vanilla/kernel/timer.c	2002-11-30 00:37:19.000000000 +0100
+++ linux/kernel/timer.c	2002-11-30 10:00:19.000000000 +0100
@@ -474,9 +474,19 @@ static void second_overflow(void)
 #endif
 }

+#ifdef __x86_64__
+extern unsigned int (*do_gettimeoffset_safe)(void);
+#endif
+
 /* in the NTP reference this is called "hardclock()" */
 static void update_wall_time_one_tick(void)
 {
+#ifdef __x86_64__
+	long orig_usec;
+
+	orig_usec = xtime.tv_usec + do_gettimeoffset_safe();
+#endif
+
 	if ( (time_adjust_step = time_adjust) != 0 ) {
 	    /* We are doing an adjtime thing. 
 	     *
@@ -511,6 +517,9 @@ static void update_wall_time_one_tick(vo
 		time_phase -= ltemp << SHIFT_SCALE;
 		xtime.tv_usec += ltemp;
 	}
+#ifdef __x86_64__
+	while (xtime.tv_usec + do_gettimeoffset_safe() < orig_usec);
+#endif
 }
 
 /*
