--- automake-1.9.4/tests/Makefile.in.avx	2005-02-28 21:16:21.000000000 -0700
+++ automake-1.9.4/tests/Makefile.in	2005-02-28 21:16:21.000000000 -0700
@@ -173,7 +173,6 @@
 autohdr3.test \
 autohdr4.test \
 auxdir.test \
-auxdir2.test \
 auxdir3.test \
 auxdir4.test \
 backsl.test \
@@ -231,7 +230,6 @@
 cond14.test \
 cond15.test \
 cond16.test \
-cond17.test \
 cond18.test \
 cond19.test \
 cond20.test \
@@ -627,7 +625,6 @@
 targetclash.test \
 txinfo.test \
 txinfo2.test \
-txinfo3.test \
 txinfo4.test \
 txinfo5.test \
 txinfo6.test \
@@ -635,20 +632,12 @@
 txinfo8.test \
 txinfo9.test \
 txinfo10.test \
-txinfo13.test \
-txinfo16.test \
 txinfo17.test \
-txinfo18.test \
 txinfo19.test \
 txinfo20.test \
 txinfo21.test \
-txinfo22.test \
-txinfo23.test \
-txinfo24.test \
-txinfo25.test \
 txinfo26.test \
 txinfo27.test \
-txinfo28.test \
 txinfo29.test \
 transform.test \
 unused.test \
@@ -660,7 +649,6 @@
 version3.test \
 version4.test \
 version6.test \
-version7.test \
 vpath.test \
 vtexi.test \
 vtexi2.test \
