--- shadow-20000902/libmisc/Makefile.am.nscd	Sat Sep  2 20:08:15 2000
+++ shadow-20000902/libmisc/Makefile.am	Tue May 22 18:11:26 2001
@@ -14,7 +14,7 @@
  chowndir.c chowntty.c console.c copydir.c entry.c env.c failure.c \
  fields.c getdate.y hushed.c isexpired.c limits.c list.c log.c \
  login_access.c login_desrpc.c login_krb.c loginprompt.c mail.c motd.c \
- myname.c obscure.c pam_pass.c pwd2spwd.c pwdcheck.c pwd_init.c rlogin.c \
+ myname.c nscd.c obscure.c pam_pass.c pwd2spwd.c pwdcheck.c pwd_init.c rlogin.c \
  salt.c setugid.c setup.c setupenv.c shell.c strtoday.c suauth.c sub.c \
  sulog.c ttytype.c tz.c ulimit.c utmp.c valid.c xmalloc.c
 
--- shadow-20000902/libmisc/nscd.c.nscd	Tue May 22 18:11:26 2001
+++ shadow-20000902/libmisc/nscd.c	Tue May 22 18:11:26 2001
@@ -0,0 +1,27 @@
+/* club nscd so it invalidates its cache */
+/* notting@redhat.com 19990408           */
+
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/types.h>
+
+
+int hup_nscd() {
+    FILE *pidfile;
+    char *pidbuf=calloc(256,sizeof(char));
+    int pid;
+    
+    if ((pidfile=fopen("/var/run/nscd.pid","r"))!=NULL) {
+	pidbuf=fgets(pidbuf,256,pidfile);
+	fclose(pidfile);
+	if (pidbuf!=NULL) {
+	    pid=atoi(pidbuf);
+	    free(pidbuf);
+	    if (pid>0) {
+		return kill(pid,SIGHUP);
+	    }
+	}
+    }
+    return -1;
+}
--- shadow-20000902/libmisc/Makefile.in.nscd	Sat Sep  2 21:26:01 2000
+++ shadow-20000902/libmisc/Makefile.in	Tue May 22 18:11:26 2001
@@ -112,7 +112,7 @@
 INCLUDES = -I$(top_srcdir)/libmisc -I$(top_srcdir)/lib
 DEFS = -DLOCALEDIR=\"$(localedir)\" -I. -I$(srcdir) -I.. @DEFS@
 
-libmisc_a_SOURCES = addgrps.c age.c basename.c chkname.c chkshell.c  chowndir.c chowntty.c console.c copydir.c entry.c env.c failure.c  fields.c getdate.y hushed.c isexpired.c limits.c list.c log.c  login_access.c login_desrpc.c login_krb.c loginprompt.c mail.c motd.c  myname.c obscure.c pam_pass.c pwd2spwd.c pwdcheck.c pwd_init.c rlogin.c  salt.c setugid.c setup.c setupenv.c shell.c strtoday.c suauth.c sub.c  sulog.c ttytype.c tz.c ulimit.c utmp.c valid.c xmalloc.c
+libmisc_a_SOURCES = addgrps.c age.c basename.c chkname.c chkshell.c  chowndir.c chowntty.c console.c copydir.c entry.c env.c failure.c  fields.c getdate.y hushed.c isexpired.c limits.c list.c log.c  login_access.c login_desrpc.c login_krb.c loginprompt.c mail.c motd.c  myname.c nscd.c obscure.c pam_pass.c pwd2spwd.c pwdcheck.c pwd_init.c rlogin.c  salt.c setugid.c setup.c setupenv.c shell.c strtoday.c suauth.c sub.c  sulog.c ttytype.c tz.c ulimit.c utmp.c valid.c xmalloc.c
 
 mkinstalldirs = $(SHELL) $(top_srcdir)/mkinstalldirs
 CONFIG_HEADER = ../config.h
--- shadow-20000902/src/groupmod.c.nscd	Sat Sep  2 20:40:44 2000
+++ shadow-20000902/src/groupmod.c	Tue May 22 18:11:26 2001
@@ -543,6 +543,7 @@
 	grp_update ();
 
 	close_files ();
+        hup_nscd ();
 	exit(E_SUCCESS);
 	/*NOTREACHED*/
 }
--- shadow-20000902/src/groupdel.c.nscd	Sat Sep  2 20:40:43 2000
+++ shadow-20000902/src/groupdel.c	Tue May 22 18:11:26 2001
@@ -346,6 +346,7 @@
 	grp_update ();
 
 	close_files ();
+        hup_nscd ();
 	exit(errors == 0 ? E_SUCCESS : E_GRP_UPDATE);
 	/*NOTREACHED*/
 }
--- shadow-20000902/src/userdel.c.nscd	Tue May 22 18:11:26 2001
+++ shadow-20000902/src/userdel.c	Tue May 22 18:11:26 2001
@@ -944,6 +944,7 @@
 	user_cancel(user_name);
 
 	close_files ();
+        hup_nscd ();
 
 	exit(errors ? E_HOMEDIR : E_SUCCESS);
 	/*NOTREACHED*/
--- shadow-20000902/src/gpasswd.c.nscd	Sat Sep  2 20:40:43 2000
+++ shadow-20000902/src/gpasswd.c	Tue May 22 18:11:26 2001
@@ -656,6 +656,7 @@
 	endsgent ();
 #endif
 #endif
+        hup_nscd ();
 	exit (0);
 	/*NOTREACHED*/
 }
--- shadow-20000902/src/newusers.c.nscd	Sat Sep  2 20:02:07 2000
+++ shadow-20000902/src/newusers.c	Tue May 22 18:11:26 2001
@@ -562,6 +562,7 @@
 		spw_unlock();
 #endif
 	(void) pw_unlock ();
+        hup_nscd ();
 
 	exit (0);
 	/*NOTREACHED*/
--- shadow-20000902/src/useradd.c.nscd	Tue May 22 18:11:26 2001
+++ shadow-20000902/src/useradd.c	Tue May 22 18:12:21 2001
@@ -1964,6 +1964,8 @@
 
 	close_files ();
 
+	hup_nscd ();
+
 	exit(E_SUCCESS);
 	/*NOTREACHED*/
 }
--- shadow-20000902/src/usermod.c.nscd	Sat Sep  2 20:40:44 2000
+++ shadow-20000902/src/usermod.c	Tue May 22 18:11:26 2001
@@ -1696,6 +1696,7 @@
 			user_id, user_newid,
 			user_gid, gflg ? user_newgid:user_gid);
 	}
+        hup_nscd();
 
 	if (grp_err)
 		exit(E_GRP_UPDATE);
--- shadow-20000902/src/groupadd.c.nscd	Tue May 22 18:11:26 2001
+++ shadow-20000902/src/groupadd.c	Tue May 22 18:11:26 2001
@@ -547,6 +547,7 @@
 	grp_update();
 
 	close_files();
+        hup_nscd();
 	exit(E_SUCCESS);
 	/*NOTREACHED*/
 }
