--- audit-1.6.1/configure.ac.avx	2007-12-11 20:04:45.891369000 -0700
+++ audit-1.6.1/configure.ac	2007-12-11 20:04:51.767736250 -0700
@@ -108,7 +108,6 @@
 if test x$use_apparmor != xno ; then
       AC_DEFINE(WITH_APPARMOR,1,[Define if you want to enable AppArmor events.])fi
 
-AC_CONFIG_SUBDIRS([system-config-audit])
 AC_OUTPUT(Makefile lib/Makefile auparse/Makefile auparse/test/Makefile src/Makefile src/mt/Makefile swig/Makefile docs/Makefile init.d/Makefile audisp/Makefile audisp/configs/Makefile bindings/Makefile bindings/python/Makefile)
 
 echo .
--- audit-1.6.1/Makefile.am.avx	2007-12-11 20:05:07.372711500 -0700
+++ audit-1.6.1/Makefile.am	2007-12-11 20:05:07.372711500 -0700
@@ -21,8 +21,7 @@
 #   Rickard E. (Rik) Faith <faith@redhat.com>
 #
 
-SUBDIRS = lib auparse src/mt src audisp swig bindings init.d docs \
-	system-config-audit
+SUBDIRS = lib auparse src/mt src audisp swig bindings init.d docs
 EXTRA_DIST = ChangeLog AUTHORS NEWS README README-install audit.spec \
 	sample.rules contrib/capp.rules contrib/nispom.rules \
 	contrib/lspp.rules contrib/skeleton.c contrib/avc_snap \
