--- sfio-1999/src/lib/sfio/sfsetbuf.c.implicit	Wed Jun  2 15:54:49 1999
+++ sfio-1999/src/lib/sfio/sfsetbuf.c	Mon Oct 15 15:01:07 2001
@@ -18,6 +18,12 @@
 _END_EXTERNS_
 #endif
 
+#if _lib_memalign
+_BEGIN_EXTERNS_
+extern Void_t* memalign _ARG_((size_t, size_t));
+_END_EXTERNS_
+#endif
+
 /*	Set a (new) buffer for a stream.
 **	If size < 0, it is assigned a suitable value depending on the
 **	kind of stream. The actual buffer size allocated is dependent
