diff -urN icewm-1.2.7/lib/keys.in icewm-1.2.7-modified/lib/keys.in
--- icewm-1.2.7/lib/keys.in	2003-03-08 20:37:51.000000000 +0100
+++ icewm-1.2.7-modified/lib/keys.in	2003-04-03 12:51:02.000000000 +0200
@@ -9,13 +9,12 @@
 # You'll have to omit XK_ prefixs and to replace XF86XK_ prefixes by
 # XF86. Valid modifiers are Alt, Ctrl, Shift, Meta, Super and Hyper.
 #
-key "Alt+Ctrl+t"		xterm
 key "Alt+Ctrl+f"		fte
 key "Alt+Ctrl+e"		nedit
 key "Alt+Ctrl+g"		gimp
-key "Alt+Ctrl+n"		netscape -noraise -remote openBrowser
-key "Alt+Ctrl+b"		netscape -noraise -remote openBookmarks
-key "Alt+Ctrl+m"		netscape -noraise -remote openURL(mailto:,new-window) 
+key "Alt+Ctrl+n"		mozilla -noraise -remote openBrowser
+key "Alt+Ctrl+b"		mozilla -noraise -remote openBookmarks
+key "Alt+Ctrl+m"		mozilla -noraise -remote openURL(mailto:,new-window) 
 
 key "Alt+Ctrl+KP_Divide"	aumix -v -5		# lower volume
 key "Alt+Ctrl+KP_Multiply"	aumix -v +5		# raise volume
@@ -27,10 +26,10 @@
 key "XF86Standby"		killall -QUIT icewm
 key "XF86AudioLowerVolume"	aumix -v -5
 key "XF86AudioRaiseVolume"	aumix -v +5
-key "XF86AudioMute"		aumix -v 0
+key "XF86AudioMute"		aumix -v sh -c "if aumix -v q | grep -q 'vol 0,'; then aumix -L > /dev/null; else aumix -S ; aumix -v 0; fi"
 key "XF86AudioPlay"		cdplay play 1
 key "XF86AudioStop"		cdplay stop
 key "XF86HomePage"		netscape -noraise -remote openHomepage
-key "XF86Mail"			netscape -noraise -remote openURL(mailto:,new-window) 
-key "XF86Search"		netscape -noraise -remote openURL(http://www.google.com/)
+key "XF86Mail"			mozilla -noraise -remote openURL(mailto:,new-window) 
+key "XF86Search"		mozilla -noraise -remote openURL(http://www.google.com/)
 key "XF86Eject"			eject
diff -urN icewm-1.2.7/lib/toolbar.in icewm-1.2.7-modified/lib/toolbar.in
--- icewm-1.2.7/lib/toolbar.in	2003-03-08 20:37:51.000000000 +0100
+++ icewm-1.2.7-modified/lib/toolbar.in	2003-04-03 12:51:39.000000000 +0200
@@ -4,6 +4,6 @@
 # since modifications to this file will be discarded when you
 # (re)install icewm.
 #
-prog XTerm xterm xterm
-prog FTE fte fte
-prog Netscape netscape netscape
+prog Terminal terminals_section.png xvt
+prog Emacs emacs.png emacs
+prog Mozilla mozilla.png mozilla
