--- pkg-config-0.19/main.c.biarch	2005-05-21 11:14:47.000000000 +0200
+++ pkg-config-0.19/main.c	2005-08-11 11:32:28.000000000 +0200
@@ -29,6 +29,11 @@
 #include <string.h>
 #include <ctype.h>
 #include <stdio.h>
+#if defined(__linux__)
+#include <sys/stat.h>
+#include <sys/syscall.h>
+#include <sys/personality.h>
+#endif
 
 #ifdef G_OS_WIN32
 #define STRICT
@@ -169,6 +174,17 @@
   return FALSE;
 }
 
+#if defined(__linux__)
+static int
+is_dir(const char *path)
+{
+  struct stat st;
+  if (stat(path, &st) < 0)
+    return 0;
+  return S_ISDIR(st.st_mode);
+}
+#endif
+
 int
 main (int argc, char **argv)
 {
@@ -200,6 +216,9 @@
   char **search_dirs;
   char **iter;
   gboolean need_newline;
+#if defined(__linux__)
+  long personality;
+#endif
 
   const char *pkgname;
   Package *pkg;
@@ -289,7 +308,15 @@
 	{
 	  add_search_dirs(search_path, G_SEARCHPATH_SEPARATOR_S);
 	}
-      
+     
+#if defined(__linux__)
+	personality = syscall(SYS_personality, 0xffffffff);
+	if ((personality & PER_MASK) == PER_LINUX32 && is_dir(LIBDIR_PATH32))
+		add_search_dirs(PKG_CONFIG_PC_PATH32, G_SEARCHPATH_SEPARATOR_S);
+	else 
+#endif
+	add_search_dirs(PKG_CONFIG_PC_PATH, G_SEARCHPATH_SEPARATOR_S);
+ 
       add_search_dirs(PKG_CONFIG_PC_PATH, G_SEARCHPATH_SEPARATOR_S);
     }
 
--- pkg-config-0.19/configure.in.biarch	2005-07-16 12:22:31.000000000 +0200
+++ pkg-config-0.19/configure.in	2005-08-11 11:31:16.000000000 +0200
@@ -18,6 +18,10 @@
    AC_MSG_RESULT([$pc_path])],
  [pc_path="$libdir/pkgconfig:$datadir/pkgconfig"])
  AC_SUBST([pc_path])
+ libdir_path32="$libdir/pkgconfig/32"
+ AC_DEFINE_UNQUOTED(LIBDIR_PATH32, ["$libdir_path32"], [libdir pkgconfig path for 32 personality])
+ pc_path32="$libdir/pkgconfig/32:$datadir/pkgconfig"
+ AC_DEFINE_UNQUOTED(PKG_CONFIG_PC_PATH32, ["$pc_path32"], [path for 32 personality])
 dnl AC_DEFINE_UNQUOTED(PKG_CONFIG_PC_PATH,["$pc_path"],[Default search path for .pc files])
 
 ])
