--- pciutils-2.2.1/lspci.c.pix	2005-12-18 17:16:24.000000000 +0100
+++ pciutils-2.2.1/lspci.c	2005-12-18 17:18:03.000000000 +0100
@@ -117,6 +117,9 @@
     die("Unable to read the standard configuration space header");
   if ((d->config[PCI_HEADER_TYPE] & 0x7f) == PCI_HEADER_TYPE_CARDBUS)
     {
+      /* Chmou: Only when we are root */
+      if (getuid() == 0) 
+
       /* For cardbus bridges, we need to fetch 64 bytes more to get the
        * full standard header... */
       if (config_fetch(d, 64, 64))
