--- XFree86-4.2.1/xc/programs/Xserver/hw/xfree86/os-support/shared/std_kbdEv.c.fred	1999-05-07 04:56:23.000000000 +0200
+++ XFree86-4.2.1/xc/programs/Xserver/hw/xfree86/os-support/shared/std_kbdEv.c	2002-09-11 09:57:25.000000000 +0200
@@ -30,17 +30,28 @@
 #include "xf86Priv.h"
 #include "xf86_OSlib.h"
 
+extern int errno;
+
 void
 xf86KbdEvents()
 {
 	unsigned char rBuf[64];
 	int nBytes, i;
 
+	errno = 0;
+	
 	if ((nBytes = read( xf86Info.consoleFd, (char *)rBuf, sizeof(rBuf)))
 	    > 0)
 	{
 		for (i = 0; i < nBytes; i++)
 			xf86PostKbdEvent(rBuf[i]);
+	} else {
+		static int saved_errno = 0;
+		
+		if (saved_errno != errno) {
+			saved_errno = errno;
+			ErrorF("xf86KbdEvents %s\n", strerror(errno));
+		}
 	}
 }
 
