--- XFree86-4.2.1/xc/programs/Xserver/hw/xfree86/drivers/chips/ct_driver.c.chips-CT69000-noaccel	2002-01-04 16:22:27.000000000 -0500
+++ XFree86-4.2.1/xc/programs/Xserver/hw/xfree86/drivers/chips/ct_driver.c	2003-03-26 19:17:20.000000000 -0500
@@ -1486,6 +1486,20 @@ chipsPreInitHiQV(ScrnInfoPtr pScrn, int 
 			   "rgb bits %d\n", val);
 	}
     }
+    /* FIXME: Disable 2D acceleration on C&T 69000 by default, since it is
+     * reported to be broken, but nobody who has this hardware has narrowed
+     * it down to individual acceleration primitives yet.  This is a Red Hat
+     * workaround for a bug reported in bugzilla at:
+     * https://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=74841
+     * Mike A. Harris <mharris@redhat.com>
+     */
+    if (cPtr->Chipset == CHIPS_CT69000 && (cPtr->Flags & ChipsAccelSupport) ) {
+	cPtr->Flags &= ~ChipsAccelSupport;
+	xf86DrvMsg(pScrn->scrnIndex, X_WARNING,
+	    "Acceleration is disabled by default on C&T 69000 as it has been reported\n"
+	    "to be broken: https://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=74841\n");
+    }
+
     if ((cPtr->Flags & ChipsAccelSupport) &&
 	(xf86ReturnOptValBool(cPtr->Options, OPTION_NOACCEL, FALSE))) {
 	cPtr->Flags &= ~ChipsAccelSupport;
