--- ./src/Makefile.am.gege	Sun Mar  4 06:39:40 2001
+++ ./src/Makefile.am	Thu Jan 24 22:36:35 2002
@@ -5,13 +5,11 @@
 
 # We let configure figure out whether we need
 # regex.c, strtoul.c, strtoull.c, and strtoumax.c.
-bin_PROGRAMS = grep egrep fgrep
+bin_PROGRAMS = grep
 base_sources = grep.c grep.h \
                dfa.c dfa.h \
                kwset.c  kwset.h \
                search.c getpagesize.h system.h
-egrep_SOURCES = $(base_sources) egrepmat.c
-fgrep_SOURCES = $(base_sources) fgrepmat.c
 grep_SOURCES  = $(base_sources) grepmat.c
 
 LDADD = @INTLLIBS@ ../lib/libgreputils.a
--- ./src/grep.c.gege	Mon Nov 19 14:26:52 2001
+++ ./src/grep.c	Thu Jan 24 22:37:23 2002
@@ -1280,6 +1280,11 @@
   if (program_name && strrchr (program_name, '/'))
     program_name = strrchr (program_name, '/') + 1;
 
+  if (!strcmp(program_name, "egrep"))
+	  setmatcher ("egrep");
+  if (!strcmp(program_name, "fgrep"))
+	  setmatcher ("fgrep");
+
 #if defined(__MSDOS__) || defined(_WIN32)
   /* DOS and MS-Windows use backslashes as directory separators, and usually
      have an .exe suffix.  They also have case-insensitive filesystems.  */
