diff -ruN iproute2.0/ip/iproute.c iproute2.1/ip/iproute.c
--- iproute2.0/ip/iproute.c	2001-08-06 02:31:52.000000000 +0200
+++ iproute2.1/ip/iproute.c	2003-07-11 21:46:56.000000000 +0200
@@ -620,6 +620,8 @@
 		}
 		rtnh->rtnh_len = sizeof(*rtnh);
 		rtnh->rtnh_ifindex = 0;
+		rtnh->rtnh_flags = 0;
+		rtnh->rtnh_hops = 0;
 		rta->rta_len += rtnh->rtnh_len;
 		parse_one_nh(rta, rtnh, &argc, &argv);
 		rtnh = RTNH_NEXT(rtnh);
