--- e2fsprogs-1.39/configure.in.dm	2006-04-08 21:36:21.000000000 -0400
+++ e2fsprogs-1.39/configure.in	2006-07-06 14:11:32.000000000 -0400
@@ -494,11 +494,15 @@
 	AC_DEFINE(HAVE_DEVMAPPER)
 	echo "Enabling device-mapper support"
 
-	DEVMAPPER_REQ='libselinux libsepol'
-	DEVMAPPER_PC_LIBS='-ldevmapper'
-	DEVMAPPER_LIBS='-ldevmapper -lselinux -lsepol -lpthread'
-	STATIC_DEVMAPPER_LIBS='/usr/lib/libdevmapper.a /usr/lib/libselinux.a /usr/lib/libsepol.a -lpthread'
-
+	PKG_PROG_PKG_CONFIG()
+ 
+	AC_CHECK_LIB(devmapper, dm_tree_create,
+		[DEVMAPPER_LIBS=`$PKG_CONFIG --libs devmapper`; 
+		 STATIC_DEVMAPPER_LIBS=`$PKG_CONFIG --static --libs devmapper`;
+		 DEVMAPPER_REQ="devmapper";
+		 DEVMAPPER_PC_LIBS="-ldevmapper"],
+		[AC_MSG_ERROR([device-mapper library not found])],
+		[$DEVMAPPER_LIBS])
 fi]
 ,
 echo "Disabling device-mapper support by default"
