--- dev86-0.16.3/bcc/bcc.c.bccpaths	Wed Mar 20 23:33:47 2002
+++ dev86-0.16.3/bcc/bcc.c	Mon May 27 08:01:49 2002
@@ -158,13 +158,12 @@
 #define LOCALPREFIX     	/usr
 #endif
 char * localprefix = QUOT(LOCALPREFIX);
+char * default_include = "-I~/lib/bcc/include";
 #ifndef L_TREE
-char * default_include = "-I~/include";
 char * default_libdir0 = "-L~/lib/bcc/i86/";
 char * default_libdir3 = "-L~/lib/bcc/i386/";
 char * optim_rules     = "-d~/lib/bcc/i86";
 #else
-char * default_include = "-I~/include";
 char * default_libdir0 = "-L~/lib/";
 char * default_libdir3 = "-L~/lib/i386/";
 char * optim_rules     = "-d~/lib";
