--- Python2.4/setup.py	2004-12-03 01:23:56.206613864 +0100
+++ Python2.4/setup.py	2004-12-03 01:23:37.935248377 +0100
@@ -238,9 +238,6 @@
         return sys.platform
 
     def detect_modules(self):
-        # Ensure that /usr/local is always used
-        add_dir_to_list(self.compiler.library_dirs, '/usr/local/lib')
-        add_dir_to_list(self.compiler.include_dirs, '/usr/local/include')
 
         # Add paths to popular package managers on OS X/darwin
         if sys.platform == "darwin":
