diff -Naur php-4.3.6/ext/standard/credits.c php-4.3.6.oden/ext/standard/credits.c
--- php-4.3.6/ext/standard/credits.c	2003-08-07 17:41:43.000000000 +0200
+++ php-4.3.6.oden/ext/standard/credits.c	2004-05-26 13:31:39.758187512 +0200
@@ -108,6 +108,16 @@
 		php_info_print_table_end();
 	}
 
+	if (flag & PHP_CREDITS_RPM_PACKAGING) {
+		/* RPM Packaging Team */
+		php_info_print_table_start();
+		php_info_print_table_colspan_header(2, "Mandrakelinux PHP RPM Packaging Team");
+		php_info_print_table_header(2, "Name", "Organization");
+		CREDIT_LINE("Oden Eriksson", "MandrakeSoft, Imega System AB");
+		CREDIT_LINE("Jean-Michel Dault", "MandrakeSoft");
+		php_info_print_table_end();
+	}
+
 	if (flag & PHP_CREDITS_FULLPAGE) {
 		PUTS("</div></body></html>\n");
 	}
diff -Naur php-4.3.6/ext/standard/credits.h php-4.3.6.oden/ext/standard/credits.h
--- php-4.3.6/ext/standard/credits.h	2002-12-31 17:35:25.000000000 +0100
+++ php-4.3.6.oden/ext/standard/credits.h	2004-05-26 13:23:59.656133656 +0200
@@ -33,6 +33,7 @@
 #define PHP_CREDITS_FULLPAGE		(1<<5)
 #define PHP_CREDITS_QA				(1<<6)
 #define PHP_CREDITS_WEB				(1<<7)
+#define PHP_CREDITS_RPM_PACKAGING		(1<<8)
 #define PHP_CREDITS_ALL				0xFFFFFFFF
 
 #endif /* HAVE_CREDITS_DEFS */
diff -Naur php-4.3.6/ext/standard/info.c php-4.3.6.oden/ext/standard/info.c
--- php-4.3.6/ext/standard/info.c	2004-03-15 17:39:53.000000000 +0100
+++ php-4.3.6.oden/ext/standard/info.c	2004-05-26 13:23:59.658133352 +0200
@@ -832,6 +832,7 @@
 	REGISTER_LONG_CONSTANT("CREDITS_DOCS",	PHP_CREDITS_DOCS, CONST_PERSISTENT|CONST_CS);
 	REGISTER_LONG_CONSTANT("CREDITS_FULLPAGE",	PHP_CREDITS_FULLPAGE, CONST_PERSISTENT|CONST_CS);
 	REGISTER_LONG_CONSTANT("CREDITS_QA",	PHP_CREDITS_QA, CONST_PERSISTENT|CONST_CS);
+	REGISTER_LONG_CONSTANT("CREDITS_RPM_PACKAGING",	PHP_CREDITS_RPM_PACKAGING, CONST_PERSISTENT|CONST_CS);
 	REGISTER_LONG_CONSTANT("CREDITS_ALL",	PHP_CREDITS_ALL, CONST_PERSISTENT|CONST_CS);
 }
 /* }}} */
diff -Naur php-4.3.6/ext/standard/info.h php-4.3.6.oden/ext/standard/info.h
--- php-4.3.6/ext/standard/info.h	2003-04-22 05:27:01.000000000 +0200
+++ php-4.3.6.oden/ext/standard/info.h	2004-05-26 13:23:59.659133200 +0200
@@ -46,6 +46,7 @@
 #define PHP_CREDITS_FULLPAGE		(1<<5)
 #define PHP_CREDITS_QA				(1<<6)
 #define PHP_CREDITS_WEB             (1<<7)
+#define PHP_CREDITS_RPM_PACKAGING               (1<<8)
 #define PHP_CREDITS_ALL				0xFFFFFFFF
 
 #endif /* HAVE_CREDITS_DEFS */
