--- aa-eventd.avx	2006-08-30 18:06:45.000000000 -0600
+++ aa-eventd	2006-08-30 18:06:45.000000000 -0600
@@ -46,7 +46,7 @@
 my $cfgfile     = "$cfgdir/notify.cfg";
 my $errlog      = "$dbdir/event-dispatch.log";
 
-my $logfile     = "/var/log/audit/audit.log";
+my $logfile     = "/var/log/system/audit/current";
 my $syslogfile  = "/var/log/messages";
 
 ##########################################################################
--- genprof.avx	2006-08-30 18:07:19.000000000 -0600
+++ genprof	2006-08-30 18:07:19.000000000 -0600
@@ -124,7 +124,7 @@
 my $logmark = "";
 my $done_profiling = 0;
 
-$syslog = 0 if ( -e "/var/log/audit/audit.log" );
+$syslog = 0 if ( -e "/var/log/system/audit/current" );
 
 while(not $done_profiling) {
   if ( $syslog ) { 
@@ -175,7 +175,7 @@
 
 sub last_audit_entry_time {
   
-  local $_ = `tail -1 /var/log/audit/audit.log`;
+  local $_ = `tail -1 /var/log/system/audit/current`;
   my $logmark;
   if ( /^*msg\=audit\((\d+\.\d+\:\d+).*\).*$/ ) {
     $logmark = $1;
--- SubDomain.pm.avx	2006-04-09 15:49:30.000000000 -0600
+++ SubDomain.pm	2006-08-30 18:08:41.000000000 -0600
@@ -69,8 +69,8 @@
 
 # where do we get our log messages from?
 our $filename;
-if(-f "/var/log/audit/audit.log") {
-  $filename = "/var/log/audit/audit.log";
+if(-f "/var/log/system/audit/current") {
+  $filename = "/var/log/system/audit/current";
 } elsif(-f "/etc/slackware-version") {
   $filename = "/var/log/syslog";
 } else {
