--- dosfstools-2.10/dosfsck/dosfsck.h~	2003-06-14 22:03:20.000000000 +0200
+++ dosfstools-2.10/dosfsck/dosfsck.h	2004-04-17 13:33:21.032047228 +0200
@@ -13,7 +13,7 @@
 #define _LINUX_STAT_H		/* hack to avoid inclusion of <linux/stat.h> */
 #define _LINUX_STRING_H_	/* hack to avoid inclusion of <linux/string.h>*/
 #define _LINUX_FS_H             /* hack to avoid inclusion of <linux/fs.h> */
-#include <linux/msdos_fs.h>
+#include "msdos_fs.h"
 
 /* 2.1 kernels use le16_to_cpu() type functions for CF_LE_W & Co., but don't
  * export this macros, only __le16_to_cpu(). */
--- dosfstools-2.10/dosfsck/file.c~	2003-06-14 22:03:20.000000000 +0200
+++ dosfstools-2.10/dosfsck/file.c	2004-04-17 13:35:44.781208547 +0200
@@ -15,7 +15,7 @@
 #define _LINUX_STAT_H		/* hack to avoid inclusion of <linux/stat.h> */
 #define _LINUX_STRING_H_	/* hack to avoid inclusion of <linux/string.h>*/
 #define _LINUX_FS_H             /* hack to avoid inclusion of <linux/fs.h> */
-#include <linux/msdos_fs.h>
+#include "msdos_fs.h"
 
 #include "common.h"
 #include "file.h"
