--- ncurses-5.3/ncurses/curses.priv.h.utf8	2003-07-10 14:03:33.000000000 +0800
+++ ncurses-5.3/ncurses/curses.priv.h	2003-07-10 14:36:21.000000000 +0800
@@ -155,6 +155,12 @@
 #define USE_SIGWINCH 0
 #endif
 
+#if USE_WIDEC_SUPPORT
+#if NEED_WCHAR_H
+#include <wchar.h>
+#endif
+#endif
+
 /*
  * If desired, one can configure this, disabling environment variables that
  * point to custom terminfo/termcap locations.
--- ncurses-5.3/test/test.priv.h.utf8	2003-07-10 14:03:32.000000000 +0800
+++ ncurses-5.3/test/test.priv.h	2003-07-10 14:42:17.000000000 +0800
@@ -151,6 +151,12 @@
 #endif
 #endif
 
+#if USE_WIDEC_SUPPORT
+#if NEED_WCHAR_H
+#include <wchar.h>
+#endif
+#endif
+
 #ifndef HAVE_TYPE_ATTR_T
 #if !USE_WIDEC_SUPPORT
 #define attr_t long
