--- termcap.xtermchanges	Mon Mar 20 02:21:17 2000
+++ termcap	Mon Mar 20 02:28:12 2000
@@ -1723,178 +1723,123 @@
 	:kr=\EOC:ks=\E[?1h\E=:ku=\EOA:le=^H:md=\E[1m:me=\E[m:\
 	:mr=\E[7m:nd=\E[C:se=\E[m:sf=^J:so=\E[7m:sr=\EM:ta=^I:\
 	:ue=\E[m:up=\E[A:us=\E[4m:
-# Compatible with the R5 xterm 
-# (from the XFree86 3.2 distribution, <blink=@> removed)
-# added khome/kend, rmir/smir, rmul/smul based on the R5 xterm code - TD
-# corrected typos in rs2 string - TD
-xterm-r5|xterm R5 version:\
-	:am:bs:km:ms:xn:\
-	:co#80:it#8:li#24:\
-	:@7=\E[4~:AL=\E[%dL:DC=\E[%dP:DL=\E[%dM:DO=\E[%dB:\
-	:F1=\E[23~:F2=\E[24~:IC=\E[%d@:Km=\E[M:LE=\E[%dD:\
-	:RI=\E[%dC:UP=\E[%dA:al=\E[L:bl=^G:cd=\E[J:ce=\E[K:\
-	:cl=\E[H\E[2J:cm=\E[%i%d;%dH:cr=^M:cs=\E[%i%d;%dr:\
-	:ct=\E[3g:dc=\E[P:dl=\E[M:do=^J:ei=\E[4l:ho=\E[H:ic=\E[@:\
-	:im=\E[4h:k0=\EOq:k1=\E[11~:k2=\E[12~:k3=\E[13~:k4=\E[14~:\
+xf|xterm-xfree86|XFree86 xterm:\
+	:is=\E[!p\E[?3;4l\E[4l\E>:\
+	:rs=\E[!p\E[?3;4l\E[4l\E>:\
+	:AL=\E[%dL:DL=\E[%dM:DC=\E[%dP:DO=\E[%dB:UP=\E[%dA:\
+	:LE=\E[%dD:RI=\E[%dC:\
+	:al=\E[L:am:bl=^G:\
+	:cd=\E[J:ce=\E[K:cl=\E[H\E[2J:cm=\E[%i%d;%dH:co#80:\
+	:cs=\E[%i%d;%dr:ct=\E[3g:\
+	:dc=\E[P:dl=\E[M:ho=\E[H:\
+	:im=\E[4h:ei=\E[4l:mi:\
+	:ks=\E[?1h\E=:ke=\E[?1l\E>:\
+	:k1=\EOP:k2=\EOQ:k3=\EOR:k4=\EOS:\
 	:k5=\E[15~:k6=\E[17~:k7=\E[18~:k8=\E[19~:k9=\E[20~:\
-	:k;=\E[21~:kA=\E[30~:kD=\E[3~:kE=\E[8~:kI=\E[2~:kL=\E[31~:\
-	:kN=\E[6~:kP=\E[5~:kb=^H:kd=\EOB:ke=\E[?1l\E>:kh=\E[1~:\
-	:kl=\EOD:kr=\EOC:ks=\E[?1h\E=:ku=\EOA:le=^H:md=\E[1m:\
-	:me=\E[m:mr=\E[7m:nd=\E[C:\
-	:r2=\E>\E[?1;3;4;5;6l\E[4l\E[?7h\E[m\E[r\E[2J\E[H:\
-	:rc=\E8:\
-	:..sa=\E[%?%p1%t;7%;%?%p2%t;4%;%?%p3%t;7%;%?%p4%t;5%;%?%p6%t;1%;m:\
-	:sc=\E7:se=\E[m:sf=^J:so=\E[7m:sr=\EM:st=\EH:ta=^I:ue=\E[m:\
-	:up=\E[A:us=\E[4m:
+	:k;=\E[21~:F1=\E[23~:F2=\E[24~:\
+	:kn#12:\
+	:kH=\E[4~::@7=\E[4~:kh=\E[1~:\
+	:@0=\E[1~:kI=\E[2~:kD=\177:\
+	:*6=\E[4~:kP=\E[5~:kN=\E[6~:\
+	:km:\
+	:kb=^H:ku=\EOA:kd=\EOB:kr=\EOC:kl=\EOD:\
+	:li#24:md=\E[1m:me=\E[m^O:mr=\E[7m:ms:nd=\E[C:\
+	:eA=\E)0:as=^N:ae=^O:ml=\El:mu=\Em:\
+	:sc=\E7:rc=\E8:sf=\n:so=\E[7m:se=\E[27m:sr=\EM:st=\EH:\
+	:ti=\E7\E[?47h:te=\E[2J\E[?47l\E8:\
+	:vi=\E[?25l:ve=\E[?25h:\
+	:up=\E[A:us=\E[4m:ue=\E[24m:xn:\
+	:ut:Co#8:op=\E[39;49m:AB=\E[4%dm:AF=\E[3%dm:\
+	:pa#64:Sf=\E[3%dm:Sb=\E[4%dm:
+v1|xterm-24|xterms|vs100|24x80 xterm:\
+	:li#24:\
+	:tc=xterm:
+v2|xterm-65|65x80 xterm:\
+	:li#65:tc=xterm:
+vb|xterm-bold|xterm with bold for underline:\
+	:us=\E[1m:tc=xterm:
+vb|xterm-boldso|xterm with bold for standout:\
+	:se=\E[m:so=\E[1m:tc=xterm:
+vm|xtermm|monochrome xterm:\
+	:F1=\E[23~:F2=\E[24~:F3=\E[25~:F4=\E[26~:F5=\E[28~:\
+	:F6=\E[29~:F7=\E[31~:F8=\E[32~:F9=\E[33~:FA=\E[34~:\
+	:kn#20:\
+	:st@:ut@:Co@:NC@:op@:AB@:AF@:pa@:Sf@:Sb@:tc=xterm:
+#
+# This should work for the commonly used "color xterm" variations (XFree86
+# xterm, color_xterm, nxterm, rxvt).  You may have trouble with this using
+# conventional termcap because ncurses reports it is longer than 1023
+# characters.  That is because resolving it adds the alternate character-set
+# (you can suppress it with "ac@").
+vc|xterm-color|generic "ANSI" color xterm:\
+	:Co#8:NC@:pa#64:op=\E[m:AB=\E[4%dm:AF=\E[3%dm:tc=xterm-r6:
+xterm-redhat|xterm with modifications to follow Debian keyboard policy:\
+	:kb=\177:kD=\E[3~:F3=\E[25~:F4=\E[26~:F5=\E[28~:\
+	:F6=\E[29~:F7=\E[31~:F8=\E[32~:F9=\E[33~:FA=\E[34~:\
+	:tc=xterm-xfree86:
+vt|xterm-vt220|xterm emulating vt220:\
+	:tc=xterm-xfree86:
+#
+# vi may work better with this entry, because vi doesn't use insert mode much.+ # |xterm-ic|xterm-vi|xterm with insert character instead of insert mode:\
+vi|xterm-ic|xterm-vi|xterm with insert char:\
+	:im@:ei@:mi@:ic=\E[@:IC=\E[%d@:tc=xterm:
+#
 # Compatible with the R6 xterm
-# (from XFree86 3.2 distribution, <acsc> and :it: added, <blink@> removed)
-# added khome/kend - TD
-# (untranslatable capabilities removed to fit entry within 1023 bytes)
-# (sgr removed to fit entry within 1023 bytes)
-# (terminfo-only capabilities suppressed to fit entry within 1023 bytes)
-xterm-r6|xterm-old|xterm X11R6 version:\
-	:am:bs:km:mi:ms:xn:\
-	:co#80:it#8:li#24:\
-	:AL=\E[%dL:DC=\E[%dP:DL=\E[%dM:DO=\E[%dB:LE=\E[%dD:\
-	:RI=\E[%dC:UP=\E[%dA:ae=^O:al=\E[L:as=^N:bl=^G:cd=\E[J:\
-	:ce=\E[K:cl=\E[H\E[2J:cm=\E[%i%d;%dH:cr=^M:\
-	:cs=\E[%i%d;%dr:ct=\E[3g:dc=\E[P:dl=\E[M:do=^J:ei=\E[4l:\
-	:ho=\E[H:im=\E[4h:\
-	:is=\E7\E[r\E[m\E[?7h\E[?1;3;4;6l\E[4l\E8\E>:k1=\EOP:\
-	:k2=\EOQ:k3=\EOR:k4=\EOS:k5=\E[15~:k6=\E[17~:k7=\E[18~:\
-	:k8=\E[19~:k9=\E[20~:kD=\E[3~:kI=\E[2~:kN=\E[6~:kP=\E[5~:\
-	:kb=^H:kd=\EOB:ke=\E[?1l\E>:kl=\EOD:kr=\EOC:ks=\E[?1h\E=:\
-	:ku=\EOA:le=^H:md=\E[1m:me=\E[m:mr=\E[7m:nd=\E[C:rc=\E8:\
-	:sc=\E7:se=\E[m:sf=^J:so=\E[7m:sr=\EM:ta=^I:\
-	:te=\E[2J\E[?47l\E8:ti=\E7\E[?47h:ue=\E[m:up=\E[A:\
-	:us=\E[4m:
-# This is the base xterm entry for the xterm supplied with XFree86 3.2 & up.
-# The name has been changed and some aliases have been removed.
-# (untranslatable capabilities removed to fit entry within 1023 bytes)
-# (sgr removed to fit entry within 1023 bytes)
-# (terminfo-only capabilities suppressed to fit entry within 1023 bytes)
-xterm-xf86-v32|xterm terminal emulator (XFree86 3.2 Window System):\
-	:am:bs:km:mi:ms:xn:\
-	:co#80:it#8:li#24:\
-	:AL=\E[%dL:DC=\E[%dP:DL=\E[%dM:DO=\E[%dB:IC=\E[%d@:\
-	:K1=\EOw:K2=\EOy:K3=\EOu:K4=\EOq:K5=\EOs:LE=\E[%dD:\
-	:RI=\E[%dC:UP=\E[%dA:ae=^O:al=\E[L:as=^N:bl=^G:bt=\E[Z:\
-	:cd=\E[J:ce=\E[K:cl=\E[H\E[2J:cm=\E[%i%d;%dH:cr=^M:\
-	:cs=\E[%i%d;%dr:ct=\E[3g:dc=\E[P:dl=\E[M:do=^J:ec=\E[%dX:\
-	:ei=\E[4l:ho=\E[H:ic=\E[@:im=\E[4h:\
+r6|xterm-r6|xterm-old|X11R6 xterm:\
 	:is=\E7\E[r\E[m\E[?7h\E[?1;3;4;6l\E[4l\E8\E>:\
-	:k1=\E[11~:k2=\E[12~:k3=\E[13~:k4=\E[14~:k5=\E[15~:\
-	:k6=\E[17~:k7=\E[18~:k8=\E[19~:k9=\E[20~:kD=\177:kI=\E[2~:\
-	:kN=\E[6~:kP=\E[5~:kb=^H:kd=\EOB:ke=\E[?1l\E>:kh=\EOH:\
-	:kl=\EOD:kr=\EOC:ks=\E[?1h\E=:ku=\EOA:le=^H:md=\E[1m:\
-	:me=\E[m\017:mr=\E[7m:nd=\E[C:rc=\E8:sc=\E7:se=\E[27m:\
-	:sf=^J:so=\E[7m:sr=\EM:st=\EH:ta=^I:te=\E[2J\E[?47l\E8:\
-	:ti=\E7\E[?47h:ue=\E[24m:up=\E[A:us=\E[4m:\
-	:vb=\E[?5h\E[?5l:ve=\E[?25h:vi=\E[?25l:vs=\E[?25h:
-
-# This is the stock xterm entry supplied with XFree86 3.3, which uses VT100
-# codes for F1-F4 except while in VT220 mode.
-xterm-xf86-v33|xterm terminal emulator (XFree86 3.3 Window System):\
+	:rs=\E7\E[r\E[m\E[?7h\E[?1;3;4;6l\E[4l\E8\E>:\
+	:AL=\E[%dL:DL=\E[%dM:DC=\E[%dP:DO=\E[%dB:UP=\E[%dA:\
+	:LE=\E[%dD:RI=\E[%dC:\
+	:al=\E[L:am:bl=^G:\
+	:bs:cd=\E[J:ce=\E[K:cl=\E[H\E[2J:cm=\E[%i%d;%dH:co#80:\
+	:cs=\E[%i%d;%dr:ct=\E[3g:dc=\E[P:dl=\E[M:ho=\E[H:\
+	:im=\E[4h:ei=\E[4l:mi:\
+	:ks=\E[?1h\E=:ke=\E[?1l\E>:\
+	:@7=\E[4~:kh=\E[1~:\
 	:k1=\EOP:k2=\EOQ:k3=\EOR:k4=\EOS:\
-	:tc=xterm-xf86-v32:
-
-# This version was released in XFree86 3.3.3 (November 1998).
-# Besides providing printer support, it exploits a new feature that allows
-# xterm to use terminfo-based descriptions with the titeInhibit resource.
-xterm-xf86-v333|xterm terminal emulator (XFree86 3.3.3 Window System):\
-	:5i:\
-	:*6@:@0@:@7=\E[4~:ei=:ic@:im=:is=\E[\041p\E[?3;4l\E[4l\E>:\
-	:kD=\E[3~:kh=\E[1~:mb=\E[5m:mk=\E[8m:pf=\E[4i:po=\E[5i:\
-	:ps=\E[i:r1=\Ec:r2=\E[\041p\E[?3;4l\E[4l\E>:\
-	:..sa=\E[0%?%p1%p6%|%t;1%;%?%p2%t;4%;%?%p1%p3%|%t;7%;%?%p4%t;5%;%?%p7%t;8%;m%?%p9%t\016%e\017%;:\
-	:te=\E[?1047l\E[?1048l:ti=\E[?1048h\E[?1047h:\
-	:tc=xterm-xf86-v33:
-
-# This beta version will probably be released in XFree86 4.0.
-xterm-xf86-v40|xterm terminal emulator (XFree86 4.0 Window System):\
-	:@7=\EOF:K1=\EOH:K2=\EOE:K3=\E[5~:K4=\EOF:K5=\E[6~:kD=\177:\
-	:kh=\EOH:te=\E[?1049l:ti=\E[?1049h:\
-	:tc=xterm-xf86-v333:
-
-xterm-xfree86|xterm-new|xterm terminal emulator (XFree86 4.0 Window System):\
-	:tc=xterm-xf86-v40:
-
-# From: David J. MacKenzie <djm@va.pubnix.com>, 14 Nov 1997
-xterm-xi|xterm on XI Graphics Accelerated X under BSD/OS 3.1:\
-	:se=\E[m:ue=\E[m:\
-	:tc=xterm-xf86-v33:
-
-# This is one of the variants of XFree86 3.3 xterm, updated for 4.0 (T.Dickey)
-xterm-16color|xterm with 16 colors like aixterm:\
-	:Co#16:NC#32:pa#256:\
-	:AB=\E[%?%p1%{8}%<%t%p1%{40}%+%e%p1%{92}%+%;%dm:\
-	:AF=\E[%?%p1%{8}%<%t%p1%{30}%+%e%p1%{82}%+%;%dm:\
-	:..Sb=%p1%{8}%/%{6}%*%{4}%+\E[%d%p1%{8}%m%Pa%?%ga%{1}%=%t4%e%ga%{3}%=%t6%e%ga%{4}%=%t1%e%ga%{6}%=%t3%e%ga%d%;m:\
-	:..Sf=%p1%{8}%/%{6}%*%{3}%+\E[%d%p1%{8}%m%Pa%?%ga%{1}%=%t4%e%ga%{3}%=%t6%e%ga%{4}%=%t1%e%ga%{6}%=%t3%e%ga%d%;m:\
-	:tc=xterm-xf86-v40:
-
-# This is another variant, for XFree86 4.0 xterm (T.Dickey)
-# This is an 8-bit version of xterm, which emulates DEC vt220 with ANSI color.
-# To use it, your decTerminalID resource must be set to 200 or above.
-#
-#	HTS	\E H	\210
-#	RI	\E M	\215
-#	SS3	\E O	\217
-#	CSI	\E [	\233
-#
-# (untranslatable capabilities removed to fit entry within 1023 bytes)
-# (sgr removed to fit entry within 1023 bytes)
-# (terminfo-only capabilities suppressed to fit entry within 1023 bytes)
-xterm-8bit|xterm terminal emulator 8-bit controls (X Window System):\
-	:am:bs:km:mi:ms:xn:\
-	:co#80:it#8:li#24:\
-	:AL=\233%dL:DC=\233%dP:DL=\233%dM:DO=\233%dB:IC=\233%d@:\
-	:K1=\217w:K2=\217y:K3=\217u:K4=\217q:K5=\217s:LE=\233%dD:\
-	:RI=\233%dC:UP=\233%dA:ae=^O:al=\233L:as=^N:bl=^G:bt=\233Z:\
-	:cd=\233J:ce=\233K:cl=\233H\2332J:cm=\233%i%d;%dH:cr=^M:\
-	:cs=\233%i%d;%dr:ct=\2333g:dc=\233P:dl=\233M:do=^J:\
-	:ec=\233%dX:ei=\2334l:ho=\233H:im=\2334h:\
-	:is=\E7\E G\233r\233m\233?7h\233?1;3;4;6l\2334l\E8\E>:\
-	:k1=\23311~:k2=\23312~:k3=\23313~:k4=\23314~:k5=\23315~:\
-	:k6=\23317~:k7=\23318~:k8=\23319~:k9=\23320~:kD=\2333~:\
-	:kI=\2332~:kN=\2336~:kP=\2335~:kb=^H:kd=\217B:\
-	:ke=\233?1l\E>:kh=\2331~:kl=\217D:kr=\217C:ks=\233?1h\E=:\
-	:ku=\217A:le=^H:mb=\2335m:md=\2331m:me=\233m^O:mr=\2337m:\
-	:nd=\233C:rc=\E8:sc=\E7:se=\23327m:sf=^J:so=\2337m:sr=\215:\
-	:st=\210:ta=^I:te=\233?1049l:ti=\233?1049h:ue=\23324m:\
-	:up=\233A:us=\2334m:vb=\233?5h\233?5l:ve=\233?25h:\
-	:vi=\233?25l:vs=\233?25h:
-
-xterm-24|vs100|xterms|xterm terminal emulator (X Window System):\
-	:li#24:tc=xterm:
-
-# This is xterm for ncurses.
-xterm|xterm terminal emulator (X Window System):\
-	:Km=\E[M:\
-	:ac=``aaffggiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz{{||}}~~:\
-	:u6=\E[%i%d;%dR:u7=\E[6n:u8=\E[?1;2c:u9=\E[c:\
-	:tc=xterm-r6:
-
-# These entries allow access to the X titlebar and icon name as a status line. 
-# Note that twm (and possibly window managers descended from it such as tvtwm, 
-# ctwm, and vtwm) track windows by icon-name; thus, you don't want to mess
-# with it.
-xterm+sl|access X title line and icon name:\
-	:hs:\
-	:ws#40:\
-	:ds=\E]0;\007:fs=^G:ts=\E]0;:tc=xterm:
-xterm+sl-twm|access X title line (pacify twm-descended window managers):\
-	:hs:\
-	:ws#40:\
-	:ds=\E]2;\007:fs=^G:ts=\E]2;:tc=xterm:
+	:k5=\E[15~:k6=\E[17~:k7=\E[18~:k8=\E[19~:k9=\E[20~:\
+	:k;=\E[21~:\
+	:F1=\E[23~:F2=\E[24~:F3=\E[25~:F4=\E[26~:F5=\E[28~:\
+	:F6=\E[29~:F7=\E[31~:F8=\E[32~:F9=\E[33~:FA=\E[34~:\
+	:kn#20:\
+	:@0=\E[1~:kI=\E[2~:kD=\E[3~:\
+	:*6=\E[4~:kP=\E[5~:kN=\E[6~:\
+	:km:\
+	:kb=^H:ku=\EOA:kd=\EOB:kr=\EOC:kl=\EOD:\
+	:li#24:md=\E[1m:me=\E[m:mr=\E[7m:ms:nd=\E[C:pt:\
+	:eA=\E)0:as=^N:ae=^O:\
+	:ml=\El:mu=\Em:\
+	:sc=\E7:rc=\E8:sf=\n:so=\E[7m:se=\E[m:sr=\EM:\
+	:ti=\E7\E[?47h:te=\E[2J\E[?47l\E8:\
+	:up=\E[A:us=\E[4m:ue=\E[m:xn:
+#
+# Compatible with the R5 xterm
+r5|xterm-r5|X11R5 xterm X11R5:\
+	:AL=\E[%dL:DC=\E[%dP:DL=\E[%dM:DO=\E[%dB:IC=\E[%d@:UP=\E[%dA:\
+	:al=\E[L:am:\
+	:bs:cd=\E[J:ce=\E[K:cl=\E[H\E[2J:cm=\E[%i%d;%dH:co#80:\
+	:cs=\E[%i%d;%dr:ct=\E[3g:\
+	:dc=\E[P:dl=\E[M:\
+	:im=\E[4h:ei=\E[4l:mi:\
+	:ho=\E[H:\
+	:is=\E[r\E[m\E[2J\E[H\E[?7h\E[?1;3;4;6l\E[4l:\
+	:rs=\E>\E[?1;3;4;5;6l\E[4l\E[?7h\E[m\E[r\E[2J\E[H:\
+	:k1=\E[11~:k2=\E[12~:k3=\E[13~:k4=\E[14~:kb=^H:kd=\EOB:ke=\E[?1l\E>:\
+	:kl=\EOD:km:kn#4:kr=\EOC:ks=\E[?1h\E=:ku=\EOA:\
+	:@7=\E[4~:kh=\E[1~:\
+	:li#24:md=\E[1m:me=\E[m:mr=\E[7m:ms:nd=\E[C:pt:\
+	:sc=\E7:rc=\E8:sf=\n:so=\E[7m:se=\E[m:sr=\EM:\
+	:te=\E[2J\E[?47l\E8:ti=\E7\E[?47h:\
+	:up=\E[A:us=\E[4m:ue=\E[m:xn:
+#
+# This is the only entry which you should have to customize, since "xterm"
+# is widely used for a variety of incompatible terminal emulations including
+# color_xterm and rxvt.
+v0|xterm|X11 terminal emulator:\
+	:tc=xterm-redhat:
+#       :tc=xterm-xfree86:
+#       :tc=xterm-r6:
 
-#
-# The following xterm variants don't depend on your base version
-#
-# xterm with bold instead of underline
-xterm-bold|xterm terminal emulator (X11R6 Window System) standout w/bold:\
-	:so=\E[7m:us=\E[1m:\
-	:tc=xterm:
 # (kterm: this had extension capabilities ":KJ:TY=ascii:" -- esr)
 # (kterm should not invoke DEC Graphics as the alternate character set
 #  -- Kenji Rikitake)
@@ -1912,40 +1857,6 @@
 	:te@:ti@:\
 	:tc=xterm:
 
-# This describes the capabilities of color_xterm, an xterm variant from
-# before ECMA-64 color support was folded into the main-line xterm release.
-# This entry is straight from color_xterm's maintainer.
-# From: Jacob Mandelson <jlm@ugcs.caltech.edu>, 09 Nov 1996
-# The README's with the distribution also say that it supports SGR 21, 24, 25
-# and 27, but they are not present in the terminfo or termcap.
-# (untranslatable capabilities removed to fit entry within 1023 bytes)
-# (sgr removed to fit entry within 1023 bytes)
-# (terminfo-only capabilities suppressed to fit entry within 1023 bytes)
-color_xterm|cx|cx100|color_xterm color terminal emulator for X:\
-	:am:bs:km:mi:ms:xn:\
-	:co#80:it#8:li#65:\
-	:AL=\E[%dL:DC=\E[%dP:DL=\E[%dM:DO=\E[%dB:IC=\E[%d@:\
-	:K1=\EOw:K2=\EOu:K3=\EOy:K4=\EOq:K5=\EOs:LE=\E[%dD:\
-	:RI=\E[%dC:UP=\E[%dA:ae=^O:al=\E[L:as=^N:bl=^G:cd=\E[J:\
-	:ce=\E[K:cl=\E[H\E[2J:cm=\E[%i%d;%dH:cr=^M:\
-	:cs=\E[%i%d;%dr:dc=\E[P:dl=\E[M:do=^J:ei=\E[4l:ho=\E[H:\
-	:i1=\E[r\E[m\E[?7h\E[?4;6l\E[4l:im=\E[4h:k1=\E[11~:\
-	:k2=\E[12~:k3=\E[13~:k4=\E[14~:k5=\E[15~:k6=\E[17~:\
-	:k7=\E[18~:k8=\E[19~:k9=\E[20~:kI=\E[2~:kN=\E[6~:kP=\E[5~:\
-	:kb=^H:kd=\EOB:kh=\E[7~:kl=\EOD:kr=\EOC:ku=\EOA:le=^H:\
-	:md=\E[1m:me=\E[m:mr=\E[7m:nd=\E[C:rc=\E8:sc=\E7:se=\E[27m:\
-	:sf=^J:so=\E[7m:sr=\EM:ta=^I:te=\E>\E[?41;1r:\
-	:ti=\E[?1;41s\E[?1;41h\E=:ue=\E[24m:up=\E[A:us=\E[4m:
-
-# The 'nxterm' distributed with Redhat Linux is a slight rehack of
-# xterm-sb_right-ansi-3d, which implements ANSI colors, but does not support
-# SGR 39 or 49.  SGR 0 does reset colors (along with everything else).  This
-# description is "compatible" with color_xterm, rxvt and XFree86 xterm, except
-# that each of those implements the home, end, delete keys differently.
-nxterm|xterm-color|generic color xterm:\
-	:NC@:\
-	:op=\E[m:tc=xterm-r6:tc=klone+color:
-
 # From: Thomas Dickey <dickey@clark.net> 04 Oct 1997
 # Updated: Oezguer Kesim <kesim@math.fu-berlin.de> 02 Nov 1997
 # Notes:
@@ -2023,16 +1934,6 @@
 	:..Sf=\E[3%?%p1%{1}%=%t4%e%p1%{3}%=%t6%e%p1%{4}%=%t1%e%p1%{6}%=%t3%e%p1%d%;m:\
 	:op=\E[100m:\
 	:tc=xtermm:
-
-# From: David J. MacKenzie <djm@va.pubnix.com> 20 Apr 1995
-# Here's a termcap entry I've been using for xterm_color, which comes
-# with BSD/OS 2.0, and the X11R6 contrib tape too I think.  Besides the
-# color stuff, I also have a status line defined as the window manager
-# title bar. [I have translated it to terminfo -- ESR]
-xterm-pcolor|xterm with color used for highlights and status line:\
-	:md=\E[1m\E[43m:mr=\E[7m\E[34m:so=\E[7m\E[31m:\
-	:us=\E[4m\E[42m:\
-	:tc=xterm+sl:tc=xterm-r6:
 
 # HP ships this, except for the pb#9600 which was merged in from BSD termcap.
 # (hpterm: added empty <acsc>, we have no idea what ACS chars look like --esr)
