--- tmdns-0.1/server/info.c.fred	2003-01-15 07:18:41.000000000 +0100
+++ tmdns-0.1/server/info.c	2003-01-17 23:53:26.000000000 +0100
@@ -352,6 +352,7 @@
 			    (struct sockaddr_in *)ifnow->ifa_addr;
 
 		    memcpy(&inaddr, &(addr->sin_addr) , sizeof(struct in_addr));
+		    if (ib[0] == 169 && ib[1] == 254) {
 		    debug("  %s\n", inet_ntoa(inaddr) );
     		    ll_add(records , newARec(namebuf,&inaddr));
 
@@ -360,6 +361,7 @@
 		    debug("Reverse addrees is %s\n" , revname );
 		    ll_add(records , newPtrRec(namebuf,revname));
 		    free(revname);
+		    }
 		}
 		break;
 
