--- shadow-4.0.12/configure.in.passwd	2006-06-30 21:22:54.000000000 -0600
+++ shadow-4.0.12/configure.in	2006-06-30 21:25:19.000000000 -0600
@@ -15,6 +15,7 @@
 AM_MAINTAINER_MODE
 
 dnl Checks for programs.
+AC_GNU_SOURCE
 AC_PROG_CC
 AC_ISC_POSIX
 AC_PROG_LN_S
@@ -34,7 +35,7 @@
 	utime.h ulimit.h sys/resource.h gshadow.h shadow.h lastlog.h \
 	locale.h rpc/key_prot.h)
 
-AC_CHECK_FUNCS(a64l fchmod fchown fsync getgroups gethostname getspnam \
+AC_CHECK_FUNCS(a64l fchmod fchown fsync getgroups gethostname getspnam setlocale \
 	gettimeofday getusershell getutent initgroups lchown lckpwdf lstat \
 	memcpy memset setgroups sigaction strchr updwtmp updwtmpx)
 AC_SYS_LARGEFILE
@@ -181,14 +182,7 @@
 AC_DEFINE_UNQUOTED(FAILLOG_FILE, "$shadow_cv_logdir/faillog",
 	[Path for faillog file.])
 
-AC_CACHE_CHECK([location of the passwd program], shadow_cv_passwd_dir,
-[if test -f /usr/bin/passwd; then
-	shadow_cv_passwd_dir=/usr/bin
-else
-	shadow_cv_passwd_dir=/bin
-fi])
-AC_DEFINE_UNQUOTED(PASSWD_PROGRAM, "$shadow_cv_passwd_dir/passwd",
-	[Path to passwd program.])
+AC_DEFINE_UNQUOTED(PASSWD_PROGRAM, "/usr/bin/passwd")
 
 dnl XXX - quick hack, should disappear before anyone notices :).
 AC_DEFINE(USE_SYSLOG, 1, [Define to use syslog().])
@@ -255,7 +249,7 @@
 	AC_CHECK_LIB(pam, pam_start,
 		[AC_DEFINE(USE_PAM, 1, [Define to support Pluggable Authentication Modules])
 		LIBPAM="-lpam"
-		AC_CHECK_LIB(pam_misc, main,
+		AC_CHECK_LIB(pam_misc, misc_conv,
 			[LIBPAM="$LIBPAM -lpam_misc"],
 			AC_MSG_ERROR(libpam_misc is missing)
 		)],
