--- modules/pam_console/console.perms.mdk	2004-02-06 14:14:19.000000000 -0700
+++ modules/pam_console/console.perms	2004-02-06 14:14:19.000000000 -0700
@@ -55,9 +55,9 @@
 # permission definitions
 <console>  0660 <serial>     0660 root.tty
 <console>  0660 <floppy>     0660 root.floppy
-<console>  0600 <sound>      0600 root.audio
+<console>  0600 <sound>      0600 root.root
 <console>  0600 <cdrom>      0660 root.cdrom
-<console>  0600 <pilot>      0660 root.uucp
+<console>  0600 <pilot>      0660 root.root
 <console>  0600 <jaz>        0660 root.disk
 <console>  0600 <zip>        0660 root.disk
 <console>  0600 <ls120>      0660 root.disk
@@ -70,7 +70,7 @@
 <console>  0600 <fb>         0600 root
 <console>  0600 <kbd>        0600 root
 <console>  0600 <joystick>   0600 root
-<console>  0600 <v4l>        0600 root.video
+<console>  0600 <v4l>        0600 root.root
 <console>  0700 <gpm>	     0700 root
 <console>  0600 <mainboard>  0600 root
 <console>  0660 <burner>     0660 root.cdwriter
