--- Docs/Makefile.am	2004-10-23 09:28:43.000000000 +0200
+++ Docs/Makefile.am.oden	2004-10-30 21:05:11.699102754 +0200
@@ -18,9 +18,9 @@
 			Support/generate-mirror-listing.pl
 info_TEXINFOS =		manual.texi
 
-targets =		manual.txt mysql.info manual.html
+targets =		manual.txt mysql.info manual.html manual-split.html
 
-BUILT_SOURCES =		$(targets) manual_toc.html include.texi
+BUILT_SOURCES =		$(targets) manual_toc.html manual-split_toc.html include.texi
 EXTRA_DIST =		$(noinst_SCRIPTS) $(BUILT_SOURCES) mysqld_error.txt \
 			INSTALL-BINARY reservedwords.texi internals.texi
 
@@ -67,6 +67,13 @@
 	cd $(srcdir) && @PERL@ $(srcdir)/Support/texi2html $(TEXI2HTML_FLAGS) $<
 manual_toc.html: manual.html
 
+# HTML, several files
+manual-split.html:	manual.texi include.texi $(srcdir)/Support/texi2html
+	mkdir -p $(srcdir)/manual-split
+	cd $(srcdir)/manual-split && @PERL@ ../Support/texi2html \
+	$(TEXI2HTML_FLAGS) -glossary -menu -split_chapter ../$<
+manual-split_toc.html: manual-split.html
+
 # PDF, Portable Document Format
 manual.pdf: manual.texi
 	sed -e 's|@image{[^}]*} *||g' <$< >manual-tmp.texi
