diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff 2.4.19-rc3/drivers/message/i2o/i2o_block.c test/drivers/message/i2o/i2o_block.c
--- 2.4.19-rc3/drivers/message/i2o/i2o_block.c	2002-07-22 18:43:30.000000000 +0200
+++ test/drivers/message/i2o/i2o_block.c	2002-07-22 18:48:09.000000000 +0200
@@ -506,6 +506,7 @@ static int i2ob_merge_requests(request_q
 	return 1;
 }
 
+#if 0
 static int i2ob_flush(struct i2o_controller *c, struct i2ob_device *d, int unit)
 {
 	unsigned long msg;
@@ -531,7 +532,7 @@ static int i2ob_flush(struct i2o_control
 	i2o_post_message(c,m);
 	return 0;
 }
-			
+#endif			
 /*
  *	OSM reply handler. This gets all the message replies
  */
