diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/drivers/isdn/eicon/fourbri.c w1/drivers/isdn/eicon/fourbri.c
--- w2/drivers/isdn/eicon/fourbri.c	2001-09-30 21:26:05.000000000 +0200
+++ w1/drivers/isdn/eicon/fourbri.c	2003-08-20 10:13:59.000000000 +0200
@@ -337,7 +337,8 @@ static int diva_server_4bri_load(card_t 
 static int diva_server_4bri_start(card_t *card, byte *channels)
 {
 	byte *ctl;
-	byte *shared, i;
+	byte *shared;
+	int i;
 	int adapter_num;
 
 	DPRINTF(("divas: start Diva Server 4BRI"));
