diff -urp linux-2.4.25/fs/proc.old/generic.c linux-2.4.25/fs/proc/generic.c
--- linux-2.4.25/fs/proc.old/generic.c	2004-05-20 16:13:47.000000000 +0300
+++ linux-2.4.25/fs/proc/generic.c	2004-05-21 02:09:06.000000000 +0300
@@ -443,6 +443,11 @@ static struct proc_dir_entry *proc_creat
 	ent->namelen = len;
 	ent->mode = mode;
 	ent->nlink = nlink;
+
+ /* make /proc/kmsg belong to klogd group (id=84) so klogd can read it chrooted */
+	if (name == "kmsg")
+		ent->gid = 84;
+
  out:
 	return ent;
 }
diff -urp linux-2.4.25/fs/proc.old/proc_misc.c linux-2.4.25/fs/proc/proc_misc.c
--- linux-2.4.25/fs/proc.old/proc_misc.c	2004-05-21 12:19:25.000000000 +0300
+++ linux-2.4.25/fs/proc/proc_misc.c	2004-05-21 12:28:26.000000000 +0300
@@ -655,7 +655,8 @@ void __init proc_misc_init(void)
 	proc_symlink("mounts", NULL, "self/mounts");
 
 	/* And now for trickier ones */
-	entry = create_proc_entry("kmsg", S_IRUSR, &proc_root);
+	/* make sure /proc/kmsg is group readable for chrooted klogd */
+	entry = create_proc_entry("kmsg", S_IRUSR|S_IRGRP, &proc_root);
 	if (entry)
 		entry->proc_fops = &proc_kmsg_operations;
 	create_seq_entry("cpuinfo", 0, &proc_cpuinfo_operations);
