--- tiff-v3.6.1/libtiff/tif_dirread.c.idefense	2005-03-22 21:09:40.422721794 +0100
+++ tiff-v3.6.1/libtiff/tif_dirread.c	2005-03-22 21:10:54.746331203 +0100
@@ -67,7 +67,11 @@
 	char	*cp = NULL;
 	tsize_t	bytes = nmemb * elem_size;
 
-	if (elem_size && bytes / elem_size == nmemb)
+	/*
+	 * XXX: Check for integer overflow.
+	 */
+	
+	if (nmemb && elem_size && bytes / elem_size == nmemb)
 		cp = (char*)_TIFFmalloc(bytes);
 
 	if (cp == NULL)
@@ -1112,8 +1116,13 @@
 	static const char mesg[] = "to fetch tag value";
 	int ok = 0;
 	const TIFFFieldInfo* fip = _TIFFFieldWithTag(tif, dp->tdir_tag);
-
-	if (dp->tdir_count > 1) {		/* array of values */
+	/*
+	 * another potential crash case
+	 */
+	if (dp->tdir_count == -1) {
+		TIFFError(tif->tif_name, "Invalid count for tag %u", dp->tdir_tag);
+		return (ok);
+	} else if (dp->tdir_count > 1) {		/* array of values */
 		char* cp = NULL;
 
 		switch (dp->tdir_type) {
@@ -1340,7 +1349,7 @@
 	      nstrips, sizeof (uint32), "for strip array")) == NULL)
 		return (0);
 	lp = *lpp;
-        memset( lp, 0, sizeof(uint32) * nstrips );
+        _TIFFmemset( lp, 0, sizeof(uint32) * nstrips );
 
 	if (dir->tdir_type == (int)TIFF_SHORT) {
 		/*
@@ -1352,7 +1361,6 @@
 			return (0);
 		if( (status = TIFFFetchShortArray(tif, dir, dp)) != 0 ) {
                     int i;
-                    
                     for( i = 0; i < nstrips && i < (int) dir->tdir_count; i++ )
                     {
                         lp[i] = dp[i];
--- tiff-v3.6.1/libtiff/tif_fax3.c.idefense	2005-03-22 21:09:40.408721488 +0100
+++ tiff-v3.6.1/libtiff/tif_fax3.c	2005-03-22 21:09:40.438722142 +0100
@@ -442,8 +442,12 @@
 {
 	char	*cp = NULL;
 	tsize_t	bytes = nmemb * elem_size;
+	
+	/*
+	 * XXX: Check for integer overflow.
+	 */
 
-	if (elem_size && bytes / elem_size == nmemb)
+	if (nmemb && elem_size && bytes / elem_size == nmemb)
 		cp = (char*) _TIFFmalloc(bytes);
 
 	if (cp == NULL)
--- tiff-v3.6.1/tools/tiffdump.c.idefense	2003-09-03 07:23:38.000000000 +0200
+++ tiff-v3.6.1/tools/tiffdump.c	2005-03-22 21:09:40.439722164 +0100
@@ -316,6 +316,10 @@
 			continue;
 		}
 		space = dp->tdir_count * datawidth[dp->tdir_type];
+		if (space <= 0) {
+			Error("Invalid count for tag %u", dp->tdir_tag);
+			goto done;
+		}            
 		if (space <= 4) {
 			switch (dp->tdir_type) {
 			case TIFF_FLOAT:
