--- ./jobs.c.ia64	2002-05-09 17:56:20.000000000 +0200
+++ ./jobs.c	2002-07-23 22:08:10.000000000 +0200
@@ -28,6 +28,7 @@
 #include <stdio.h>
 #include <signal.h>
 #include <errno.h>
+#include <string.h>
 
 #if defined (HAVE_UNISTD_H)
 #  include <unistd.h>
--- ./Makefile.in.ia64	2002-05-31 19:44:23.000000000 +0200
+++ ./Makefile.in	2002-07-23 22:08:35.000000000 +0200
@@ -107,7 +107,7 @@
 
 CFLAGS = @CFLAGS@
 CFLAGS_FOR_BUILD = @CFLAGS_FOR_BUILD@
-CPPFLAGS = @CPPFLAGS@
+CPPFLAGS = -D_GNU_SOURCE @CPPFLAGS@
 CPPFLAGS_FOR_BUILD = @CPPFLAGS_FOR_BUILD@
 LOCAL_CFLAGS = @LOCAL_CFLAGS@ ${DEBUG}
 DEFS = @DEFS@
