--- console-tools-0.2.3/console-data-1999.08.29/keymaps/i386/include/compose.inc.ppc	Tue Feb 26 11:21:49 2002
+++ console-tools-0.2.3/console-data-1999.08.29/keymaps/i386/include/compose.inc	Tue Feb 26 11:22:27 2002
@@ -3,7 +3,6 @@
 #
 # (c) 1999 Bernhard Rosenkränzer <bero@mandrakesoft.com>
 
-keycode  99 = Compose
-keycode 100 = AltGr
-        alt     keycode 100 = Compose
+keycode 125 = AltGr
+        shift     keycode 125 = Compose
 
