--- glibc-2.2.5/locale/iso-4217.def.compat-EUR-currencies	2002-02-28 10:29:58.000000000 +0100
+++ glibc-2.2.5/locale/iso-4217.def	2002-04-15 17:13:00.000000000 +0200
@@ -6,6 +6,7 @@
  *
  * !!! The list has to be sorted !!!
  */
+DEFINE_INT_CURR("ADP ")		/* Andorran Peseta -> EUR  */
 DEFINE_INT_CURR("AED ")		/* United Arab Emirates Dirham  */
 DEFINE_INT_CURR("AFA ")		/* Afghanistan Afgani  */
 DEFINE_INT_CURR("ALL ")		/* Albanian Lek  */
@@ -13,12 +14,14 @@
 DEFINE_INT_CURR("ANG ")		/* Netherlands Antilles  */
 DEFINE_INT_CURR("AOA ")		/* Angolan Kwanza  */
 DEFINE_INT_CURR("ARS ")		/* Argentine Peso  */
-DEFINE_INT_CURR("AUD ")		/* Australian Dollar  */
+DEFINE_INT_CURR("ATS ")		/* Austrian Schilling  */
+DEFINE_INT_CURR("AUD ")		/* Australian Dollar -> EUR  */
 DEFINE_INT_CURR("AWG ")		/* Aruba Guilder  */
 DEFINE_INT_CURR("AZM ")		/* Azerbaijan Manat  */
 DEFINE_INT_CURR("BAK ")		/* Bosnian and Herzegovina Convertible Mark  */
 DEFINE_INT_CURR("BBD ")		/* Barbados Dollar  */
 DEFINE_INT_CURR("BDT ")		/* Bangladesh Taka  */
+DEFINE_INT_CURR("BEF ")		/* Belgian Franc -> EUR  */
 DEFINE_INT_CURR("BGL ")		/* Bulgarian Lev  */
 DEFINE_INT_CURR("BHD ")		/* Bahraini Dinar  */
 DEFINE_INT_CURR("BIF ")		/* Burundi Franc  */
@@ -42,6 +45,7 @@
 DEFINE_INT_CURR("CVE ")		/* Cape Verde Escudo  */
 DEFINE_INT_CURR("CYP ")		/* Cypriot Pound  */
 DEFINE_INT_CURR("CZK ")		/* Czech Koruna  */
+DEFINE_INT_CURR("DEM ")		/* German Mark -> EUR  */
 DEFINE_INT_CURR("DJF ")		/* Djibouti Franc  */
 DEFINE_INT_CURR("DKK ")		/* Danish Krone (Faroe Islands, Greenland)  */
 DEFINE_INT_CURR("DOP ")		/* Dominican Republic  */
@@ -49,16 +53,20 @@
 DEFINE_INT_CURR("EEK ")		/* Estonian Kroon  */
 DEFINE_INT_CURR("EGP ")		/* Egyptian Pound  */
 DEFINE_INT_CURR("ERN ")		/* Eritrean Nakfa  */
+DEFINE_INT_CURR("ESP ")		/* Spanish Peseta -> EUR  */
 DEFINE_INT_CURR("ETB ")		/* Ethiopian Birr  */
 DEFINE_INT_CURR("EUR ")		/* European Union Euro  */
+DEFINE_INT_CURR("FIM ")		/* Finnish Markka -> EUR  */
 DEFINE_INT_CURR("FJD ")		/* Fiji Dollar  */
 DEFINE_INT_CURR("FKP ")		/* Falkland Islands Pound (Malvinas)  */
+DEFINE_INT_CURR("FRF ")		/* French Franc -> EUR  */
 DEFINE_INT_CURR("GBP ")		/* British Pound  */
 DEFINE_INT_CURR("GEL ")		/* Georgia Lari  */
 DEFINE_INT_CURR("GHC ")		/* Ghana Cedi  */
 DEFINE_INT_CURR("GIP ")		/* Gibraltar Pound  */
 DEFINE_INT_CURR("GMD ")		/* Gambian Dalasi  */
 DEFINE_INT_CURR("GNF ")		/* Guinea Franc  */
+DEFINE_INT_CURR("GRD ")		/* Greek Drachma -> EUR  */
 DEFINE_INT_CURR("GTQ ")		/* Guatemala Quetzal  */
 DEFINE_INT_CURR("GYD ")		/* Guyana Dollar  */
 DEFINE_INT_CURR("HKD ")		/* Hong Kong Dollar  */
@@ -67,12 +75,14 @@
 DEFINE_INT_CURR("HTG ")		/* Haiti Gourde  */
 DEFINE_INT_CURR("HUF ")		/* Hungarian Forint  */
 DEFINE_INT_CURR("IDR ")		/* Indonesia Rupiah  */
+DEFINE_INT_CURR("IEP ")		/* Irish Pound -> EUR  */
 DEFINE_INT_CURR("ILS ")		/* Israeli Shekel  */
 DEFINE_INT_CURR("IMP ")		/* Isle of Man Pounds  */
 DEFINE_INT_CURR("INR ")		/* Indian Rupee (Bhutan)  */
 DEFINE_INT_CURR("IQD ")		/* Iraqi Dinar  */
 DEFINE_INT_CURR("IRR ")		/* Iranian Rial  */
 DEFINE_INT_CURR("ISK ")		/* Iceland Krona  */
+DEFINE_INT_CURR("ITL ")		/* Italian Lira -> EUR  */
 DEFINE_INT_CURR("JMD ")		/* Jamaican Dollar  */
 DEFINE_INT_CURR("JOD ")		/* Jordanian Dinar  */
 DEFINE_INT_CURR("JPY ")		/* Japanese Yen  */
@@ -91,6 +101,7 @@
 DEFINE_INT_CURR("LRD ")		/* Liberian Dollar  */
 DEFINE_INT_CURR("LSL ")		/* Lesotho Maloti  */
 DEFINE_INT_CURR("LTL ")		/* Lithuanian Litas  */
+DEFINE_INT_CURR("LUF ")		/* Luxembourg Franc -> EUR  */
 DEFINE_INT_CURR("LVL ")		/* Latvia Lat  */
 DEFINE_INT_CURR("LYD ")		/* Libyan Arab Jamahiriya Dinar  */
 DEFINE_INT_CURR("MAD ")		/* Moroccan Dirham  */
@@ -111,6 +122,7 @@
 DEFINE_INT_CURR("NAD ")		/* Namibia Dollar  */
 DEFINE_INT_CURR("NGN ")		/* Nigeria Naira  */
 DEFINE_INT_CURR("NIO ")		/* Nicaragua Cordoba Oro  */
+DEFINE_INT_CURR("NLG ")		/* Netherlands Guilder -> EUR  */
 DEFINE_INT_CURR("NOK ")		/* Norwegian Krone  */
 DEFINE_INT_CURR("NPR ")		/* Nepalese Rupee  */
 DEFINE_INT_CURR("NZD ")		/* New Zealand Dollar  */
@@ -121,6 +133,7 @@
 DEFINE_INT_CURR("PHP ")		/* Philippines Peso  */
 DEFINE_INT_CURR("PKR ")		/* Pakistan Rupee  */
 DEFINE_INT_CURR("PLN ")		/* Polish Zloty  */
+DEFINE_INT_CURR("PTE ")		/* Portugese Escudo -> EUR  */
 DEFINE_INT_CURR("PYG ")		/* Paraguay Guarani  */
 DEFINE_INT_CURR("QAR ")		/* Qatar Rial  */
 DEFINE_INT_CURR("ROL ")		/* Romanian Leu  */
