--- glibc-2.3.2/elf/Makefile.rtld32-workaround	2003-07-07 17:14:51.000000000 +0200
+++ glibc-2.3.2/elf/Makefile	2003-08-12 10:42:39.000000000 +0200
@@ -286,6 +286,10 @@ $(objpfx)trusted-dirs.st: Makefile $(..)
 	$(move-if-change) ${@:st=T} ${@:st=h}
 	touch $@
 CPPFLAGS-dl-load.c = -I$(objpfx). -I$(csu-objpfx).
+# FIXME: workaround dl-load.c miscompilation with 3.3-hammer
+ifeq ($(shell uname -m),x86_64)
+CFLAGS-dl-load.c += -fno-strict-aliasing
+endif
 
 ifeq (yes,$(build-shared))
 $(inst_slibdir)/$(rtld-version-installed-name): $(objpfx)ld.so $(+force)
