--- quota-tools/quotagrpadmins.chmou	2002-07-22 20:03:35.000000000 +0200
+++ quota-tools/quotagrpadmins	2003-02-08 22:15:09.000000000 +0100
@@ -4,5 +4,5 @@
 #  Comments begin with hash in the beginning of the line
 
 # In this file you specify users responsible for space used by the group
-users: root
-mygroup: chief
+# users: root
+# mygroup: chief
--- quota-tools/quotatab.chmou	2001-11-09 21:03:35.000000000 +0100
+++ quota-tools/quotatab	2003-02-08 22:15:22.000000000 +0100
@@ -5,5 +5,5 @@
 #  Comments begin with hash in the beginning of the line
 
 # Example of description
-/dev/loop0: This is loopback device
-/dev/hda4: Your home directory
+#/dev/loop0: This is loopback device
+#/dev/hda4: Your home directory
