--- postfix-2.2.5/conf/master.cf.avx	2005-01-24 14:12:46.000000000 -0500
+++ postfix-2.2.5/conf/master.cf	2005-09-03 22:25:48.000000000 -0400
@@ -51,29 +51,6 @@
 # and other message envelope options.
 # ====================================================================
 #
-# maildrop. See the Postfix MAILDROP_README file for details.
-# Also specify in main.cf: maildrop_destination_recipient_limit=1
-#
-maildrop  unix  -       n       n       -       -       pipe
-  flags=DRhu user=vmail argv=/usr/local/bin/maildrop -d ${recipient}
-#
-# The Cyrus deliver program has changed incompatibly, multiple times.
-#
-old-cyrus unix  -       n       n       -       -       pipe
-  flags=R user=cyrus argv=/cyrus/bin/deliver -e -m ${extension} ${user}
-# Cyrus 2.1.5 (Amos Gouaux)
-# Also specify in main.cf: cyrus_destination_recipient_limit=1
-cyrus     unix  -       n       n       -       -       pipe
-  user=cyrus argv=/cyrus/bin/deliver -e -r ${sender} -m ${extension} ${user}
-#
-# See the Postfix UUCP_README file for configuration details.
-#
-uucp      unix  -       n       n       -       -       pipe
-  flags=Fqhu user=uucp argv=uux -r -n -z -a$sender - $nexthop!rmail ($recipient)
 #
 # Other external delivery methods.
 #
-ifmail    unix  -       n       n       -       -       pipe
-  flags=F user=ftn argv=/usr/lib/ifmail/ifmail -r $nexthop ($recipient)
-bsmtp     unix  -       n       n       -       -       pipe
-  flags=Fq. user=foo argv=/usr/local/sbin/bsmtp -f $sender $nexthop $recipient
--- postfix-2.1.5/conf/main.cf.avx	2004-06-06 15:53:58.000000000 -0600
+++ postfix-2.1.5/conf/main.cf	2004-10-15 12:45:15.000000000 -0600
@@ -38,7 +38,7 @@
 # daemon programs (i.e. programs listed in the master.cf file). This
 # directory must be owned by root.
 #
-daemon_directory = /usr/libexec/postfix
+daemon_directory = /usr/lib/postfix
 
 # QUEUE AND PROCESS OWNERSHIP
 #
@@ -103,6 +103,7 @@
 #
 # Note: you need to stop/start Postfix when this parameter changes.
 #
+inet_interfaces = localhost
 #inet_interfaces = all
 #inet_interfaces = $myhostname
 #inet_interfaces = $myhostname, localhost
@@ -150,7 +151,7 @@
 #
 # See also below, section "REJECTING MAIL FOR UNKNOWN LOCAL USERS".
 #
-#mydestination = $myhostname, localhost.$mydomain, localhost
+mydestination = $myhostname, localhost.$mydomain, localhost
 #mydestination = $myhostname, localhost.$mydomain, localhost, $mydomain
 #mydestination = $myhostname, localhost.$mydomain, localhost, $mydomain,
 #	mail.$mydomain, www.$mydomain, ftp.$mydomain
@@ -237,7 +238,7 @@
 # 
 #mynetworks_style = class
 #mynetworks_style = subnet
-#mynetworks_style = host
+mynetworks_style = host
 
 # Alternatively, you can specify the mynetworks list by hand, in
 # which case Postfix ignores the mynetworks_style setting.
@@ -372,7 +373,7 @@
 # "postfix reload" to eliminate the delay.
 #
 #alias_maps = dbm:/etc/aliases
-#alias_maps = hash:/etc/aliases
+alias_maps = hash:/etc/aliases
 #alias_maps = hash:/etc/aliases, nis:mail.aliases
 #alias_maps = netinfo:/aliases
 
@@ -412,7 +413,7 @@
 # system type.
 #
 #mail_spool_directory = /var/mail
-#mail_spool_directory = /var/spool/mail
+mail_spool_directory = /var/spool/mail
 
 # The mailbox_command parameter specifies the optional external
 # command to use instead of mailbox delivery. The command is run as
--- postfix-2.2.5/conf/postfix-files.avx	2005-03-14 08:13:16.000000000 -0500
+++ postfix-2.2.5/conf/postfix-files	2005-09-03 22:22:55.000000000 -0400
@@ -100,8 +100,6 @@
 $sendmail_path:f:root:-:755
 $newaliases_path:l:$sendmail_path
 $mailq_path:l:$sendmail_path
-$config_directory/LICENSE:f:root:-:644
-$config_directory/TLS_LICENSE:f:root:-:644
 $config_directory/access:f:root:-:644:p
 $config_directory/aliases:f:root:-:644:p
 $config_directory/canonical:f:root:-:644:p
@@ -111,6 +109,7 @@
 $config_directory/header_checks:f:root:-:644:p
 $config_directory/install.cf:f:root:-:644:o
 $config_directory/main.cf:f:root:-:644:p
+$config_directory/main.cf.dist:f:root:-:644
 $config_directory/main.cf.default:f:root:-:644
 $config_directory/makedefs.out:f:root:-:644
 $config_directory/master.cf:f:root:-:644:p
@@ -202,6 +201,7 @@
 $sample_directory/sample-pcre-access.cf:f:root:-:644:o:o
 $sample_directory/sample-pcre-body.cf:f:root:-:644:o:o
 $sample_directory/sample-pcre-header.cf:f:root:-:644:o:o
+$sample_directory/sample-pgsql.cf:f:root:-:644:o
 $sample_directory/sample-pgsql-aliases.cf:f:root:-:644:o
 $sample_directory/sample-qmqpd.cf:f:root:-:644:o
 $sample_directory/sample-rate.cf:f:root:-:644:o
@@ -250,7 +250,6 @@
 $readme_directory/PGSQL_README:f:root:-:644
 $readme_directory/QMQP_README:f:root:-:644
 $readme_directory/QSHAPE_README:f:root:-:644
-$readme_directory/RELEASE_NOTES:f:root:-:644
 $readme_directory/RESTRICTION_CLASS_README:f:root:-:644
 $readme_directory/SASL_README:f:root:-:644
 $readme_directory/SCHEDULER_README:f:root:-:644
--- postfix-2.1.5/src/util/sys_defs.h.avx	2004-10-15 12:50:03.000000000 -0600
+++ postfix-2.1.5/src/util/sys_defs.h	2004-10-15 12:53:41.000000000 -0600
@@ -540,7 +540,7 @@
 #define HAS_FSYNC
 #define HAS_DB
 #define DEF_DB_TYPE	"hash"
-#define ALIAS_DB_MAP	"hash:/etc/aliases"
+#define ALIAS_DB_MAP	"hash:/etc/postfix/aliases"
 #define HAS_NIS
 #define GETTIMEOFDAY(t)	gettimeofday(t,(struct timezone *) 0)
 #define ROOT_PATH	"/bin:/usr/bin:/sbin:/usr/sbin"
@@ -554,7 +554,7 @@
 #define NATIVE_MAILQ_PATH "/usr/bin/mailq"
 #define NATIVE_NEWALIAS_PATH "/usr/bin/newaliases"
 #define NATIVE_COMMAND_DIR "/usr/sbin"
-#define NATIVE_DAEMON_DIR "/usr/libexec/postfix"
+#define NATIVE_DAEMON_DIR "/usr/lib/postfix"
 #if __GLIBC__ >= 2 && __GLIBC_MINOR__ >= 1
 #define SOCKADDR_SIZE	socklen_t
 #define SOCKOPT_SIZE	socklen_t
@@ -572,7 +572,7 @@
 #define HAS_FSYNC
 #define HAS_DB
 #define DEF_DB_TYPE	"hash"
-#define ALIAS_DB_MAP	"hash:/etc/aliases"
+#define ALIAS_DB_MAP	"hash:/etc/postfix/aliases"
 #define HAS_NIS
 #define GETTIMEOFDAY(t)	gettimeofday(t,(struct timezone *) 0)
 #define ROOT_PATH	"/bin:/usr/bin:/sbin:/usr/sbin"
@@ -586,7 +586,7 @@
 #define NATIVE_MAILQ_PATH "/usr/bin/mailq"
 #define NATIVE_NEWALIAS_PATH "/usr/bin/newaliases"
 #define NATIVE_COMMAND_DIR "/usr/sbin"
-#define NATIVE_DAEMON_DIR "/usr/libexec/postfix"
+#define NATIVE_DAEMON_DIR "/usr/lib/postfix"
 #endif
 
  /*
--- postfix-2.1.5/src/global/mail_params.h.avx	2004-10-15 12:55:03.000000000 -0600
+++ postfix-2.1.5/src/global/mail_params.h	2004-10-15 12:55:27.000000000 -0600
@@ -373,11 +373,7 @@
   * Local delivery: alias databases.
   */
 #define VAR_ALIAS_MAPS		"alias_maps"
-#ifdef HAS_NIS
-#define DEF_ALIAS_MAPS		ALIAS_DB_MAP ", nis:mail.aliases"
-#else
 #define DEF_ALIAS_MAPS		ALIAS_DB_MAP
-#endif
 extern char *var_alias_maps;
 
  /*
