--- postfix-2.2.5/src/master/master.c.warnsetsid	2005-02-14 19:36:58.000000000 -0500
+++ postfix-2.2.5/src/master/master.c	2005-09-03 22:31:12.000000000 -0400
@@ -288,9 +288,10 @@
      * Run in a separate process group, so that "postfix stop" can terminate
      * all MTA processes cleanly. Give up if we can't separate from our
      * parent process. We're not supposed to blow away the parent.
+     * ANNVIX: to run master supervised, we change this from being fatal to being a warning
      */
-    if (debug_me == 0 && setsid() == -1)
-	msg_fatal("unable to set session and process group ID: %m");
+    if (debug_me == 0 && setsid() < 0)
+	msg_warn("unable to set session and process group ID: %m");
 
     /*
      * Make some room for plumbing with file descriptors. XXX This breaks
