--- perl-5.8.0-RC2/lib/ExtUtils/Liblist/Kid.pm.pix	2002-06-01 19:03:16.000000000 +0200
+++ perl-5.8.0-RC2/lib/ExtUtils/Liblist/Kid.pm	2002-06-23 21:39:37.000000000 +0200
@@ -46,6 +46,7 @@
     my(@searchpath); # from "-L/path" entries in $potential_libs
     my(@libpath) = split " ", $Config{'libpth'};
     my(@ldloadlibs, @bsloadlibs, @extralibs, @ld_run_path, %ld_run_path_seen);
+    $ld_run_path_seen{$_} = 1 foreach qw(/usr/X11R6/lib /usr/lib /lib);
     my(@libs, %libs_seen);
     my($fullname, $thislib, $thispth, @fullname);
     my($pwd) = cwd(); # from Cwd.pm
--- perl-5.8.0-RC2/lib/ExtUtils/MM_Unix.pm.pix	2002-06-16 16:27:03.000000000 +0200
+++ perl-5.8.0-RC2/lib/ExtUtils/MM_Unix.pm	2002-06-23 21:38:07.000000000 +0200
@@ -456,11 +456,18 @@
 };
     my($tmp);
     for $tmp (qw/
-	 EXTRALIBS LDLOADLIBS BSLOADLIBS LD_RUN_PATH
+	 EXTRALIBS LDLOADLIBS BSLOADLIBS
 	 /) {
 	next unless defined $self->{$tmp};
 	push @m, "$tmp = $self->{$tmp}\n";
     }
+    # don't set LD_RUN_PATH if empty
+    for $tmp (qw/
+	 LD_RUN_PATH
+	 /) {
+	next unless $self->{$tmp};
+	push @m, "$tmp = $self->{$tmp}\n";
+    }
     return join "", @m;
 }
 
