--- tar-1.13.22/src/tar.c--	Sat Sep  8 15:37:57 2001
+++ tar-1.13.22/src/tar.c	Sat Sep  8 15:51:21 2001
@@ -687,12 +687,18 @@
 	ignore_zeros_option = 1;
 	break;
 
-      case 'I':
-	USAGE_ERROR ((0, 0,
-		      _("Warning: the -I option is not supported;"
-			" perhaps you meant -j or -T?")));
-	break;
+      case 'y':
+	fprintf(stderr, "-y option is now deprecated for bzip2. Next time, please use -j instead.\n");
+	fprintf(stderr, "Compatibility will be removed soon.\n");
+	set_use_compress_program_option ("bzip2");
+	break;
 
+      case 'I':
+	fprintf(stderr, "-I option is now deprecated for bzip2. Next time, please use -j instead.\n");
+	fprintf(stderr, "Compatibility will be removed soon.\n");
+	set_use_compress_program_option ("bzip2");
+	break;
+	
       case 'j':
 	set_use_compress_program_option ("bzip2");
 	break;
@@ -869,12 +871,6 @@
 	    int e = errno;
 	    FATAL_ERROR ((0, e, "%s", quotearg_colon (optarg)));
 	  }
-	break;
-
-      case 'y':
-	USAGE_ERROR ((0, 0,
-		      _("Warning: the -y option is not supported;"
-			" perhaps you meant -j?")));
 	break;
 
       case 'z':
