--- bash-2.05/unwind_prot.c.orig	Thu Apr 26 13:17:27 2001
+++ bash-2.05/unwind_prot.c	Thu Apr 26 13:17:55 2001
@@ -312,7 +312,7 @@
 void
 unwind_protect_var (var, value, size)
      int *var;
-     char *value;
+     UWP value;
      int size;
 {
   SAVED_VAR *s = (SAVED_VAR *)xmalloc (sizeof (SAVED_VAR));
@@ -321,10 +321,10 @@
   if (size != sizeof (int))
     {
       s->desired_setting = (char *)xmalloc (size);
-      FASTCOPY (value, (char *)s->desired_setting, size);
+      FASTCOPY (value.s, (char *)s->desired_setting, size);
     }
   else
-    s->desired_setting = value;
+    s->desired_setting = value.i;
   s->size = size;
   add_unwind_protect ((Function *)restore_variable, (char *)s);
 }
