--- rpm-4.2/lib/fs.c.skip	2003-01-25 18:23:20.000000000 +0100
+++ rpm-4.2/lib/fs.c	2003-04-28 11:28:15.000000000 +0200
@@ -192,6 +192,7 @@
 	    mntdir = mounts[nextMount++].f_mntonname;
 #	endif
 
+	if (strncmp(mntdir, "/mnt/", 5)) {
 	if (stat(mntdir, &sb)) {
 	    rpmError(RPMERR_STAT, _("failed to stat %s: %s\n"), mntdir,
 			strerror(errno));
@@ -217,8 +218,11 @@
 		filesystems[numFilesystems].mntPoint);
 #endif
 	numFilesystems++;
+	} else {
+	    sb.st_dev = 0;
+	}
     }
-
+    
 #   if GETMNTENT_ONE || GETMNTENT_TWO
 	(void) fclose(mtab);
 #   elif HAVE_GETMNTINFO_R
--- rpm-4.2/lib/rpmts.c.skip	2003-02-02 16:52:34.000000000 +0100
+++ rpm-4.2/lib/rpmts.c	2003-04-28 11:23:55.000000000 +0200
@@ -936,7 +936,8 @@
     dsi = ts->dsi;
 
     if (dsi != NULL)
-    for (i = 0; (i < ts->filesystemCount) && dsi; i++, dsi++) {
+    for (i = 0; (i < ts->filesystemCount) && dsi; i++, dsi++)
+      if (strncmp(ts->filesystems[i], "/mnt/", 5)) {
 #if STATFS_IN_SYS_STATVFS
 	struct statvfs sfb;
 	memset(&sfb, 0, sizeof(sfb));
