diff -Naur php-5.2.3/scripts/Makefile.frag php-5.2.3.oden/scripts/Makefile.frag
--- php-5.2.3/scripts/Makefile.frag	2005-11-22 00:08:02.000000000 +0100
+++ php-5.2.3.oden/scripts/Makefile.frag	2007-06-20 13:15:17.000000000 +0200
@@ -4,7 +4,7 @@
 #
 
 phpincludedir = $(includedir)/php
-phpbuilddir = $(libdir)/build
+phpbuilddir = $(libdir)/php/build
 
 BUILD_FILES = \
 	scripts/phpize.m4 \
diff -Naur php-5.2.3/scripts/phpize.in php-5.2.3.oden/scripts/phpize.in
--- php-5.2.3/scripts/phpize.in	2005-12-12 12:51:45.000000000 +0100
+++ php-5.2.3.oden/scripts/phpize.in	2007-06-20 13:15:17.000000000 +0200
@@ -3,7 +3,7 @@
 # Variable declaration
 prefix='@prefix@'
 exec_prefix="`eval echo @exec_prefix@`"
-phpdir="`eval echo @libdir@`/build"
+phpdir="`eval echo @libdir@`/php/build"
 includedir="`eval echo @includedir@`/php"
 builddir="`pwd`"
 
