--- ext/xmlrpc/config.m4	2005-11-29 19:26:02.000000000 +0100
+++ ext/xmlrpc/config.m4.oden	2007-06-20 13:16:39.000000000 +0200
@@ -60,18 +60,6 @@
       AC_MSG_ERROR([XML-RPC support requires libexpat. Use --with-libexpat-dir=<DIR> (deprecated!)])
     fi
   fi
-
-  if test "$PHP_ICONV_DIR" != "no"; then
-    PHP_ICONV=$PHP_ICONV_DIR
-  fi
-  
-  if test -z "$PHP_ICONV" || test "$PHP_ICONV" = "no"; then
-    PHP_ICONV=yes
-  fi
-  
-  PHP_SETUP_ICONV(XMLRPC_SHARED_LIBADD, [], [
-    AC_MSG_ERROR([iconv not found, in order to build xmlrpc you need the iconv library])
-  ])
 fi
 
 if test "$PHP_XMLRPC" = "yes"; then
