--- lilo-orig/lilo.h	2004-02-03 20:16:51.000000000 +0100
+++ lilo-22.5.8/lilo.h	2004-02-03 20:34:34.000000000 +0100
@@ -120,7 +120,7 @@ source directory.
 #define MAJOR_EVMS	117 /* Enterprise Volume Management System */
 
 #define MAX_TOKEN	255 /* max device Token length */
-#define MAX_IMAGE_NAME	15 /* maximum name length (w/o terminating NUL) */
+#define MAX_IMAGE_NAME	31 /* maximum name length (w/o terminating NUL) */
 #define MAX_DESCR_SECTORS  12  /* upper limit on MAX_DESCR_SECS */
 
 #ifdef LCF_PASS160
