--- mod_auth_external-2.2.3/pwauth/pwauth.c.64bit-fixes	2003-04-18 08:03:38.000000000 -0400
+++ mod_auth_external-2.2.3/pwauth/pwauth.c	2003-04-18 08:09:50.000000000 -0400
@@ -127,7 +127,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <unistd.h>
-#include <strings.h>
+#include <string.h>
 #include <fcntl.h>
 #include <sys/types.h>
 #include <sys/stat.h>
@@ -563,7 +563,7 @@ char *hostaddr= getenv("IP");
  * we have no way to report failure counts.
  */
 
-check_fails()
+int check_fails()
 {
 int result= 1;
 #ifdef FAILLOG_JFH
@@ -586,7 +586,7 @@ int flfd
 /* LOG_FAILURE - Do whatever we need to do to log a failed login attempt.
  */
 
-log_failure()
+void log_failure()
 {
 #ifdef FAILLOG_JFH
 int flfd;
@@ -626,7 +626,7 @@ struct faillog flog;
  * sleep time, if other pwauth processes are in sleeps.
  */
 
-snooze(int seconds)
+void snooze(int seconds)
 {
 #ifdef SLEEP_LOCK
 int slfd;
@@ -652,13 +652,13 @@ int slfd;
 int server_uids[]= {SERVER_UIDS, 0};
 #endif
 
-main(int argc, char **argv)
+int main(int argc, char **argv)
 {
 #ifdef ENV_METHOD
 char *login, *passwd;
 #else
 char login[BFSZ+1], passwd[BFSZ+1];
-char *c, *strchr();
+char *c;
 #endif
 #ifdef SERVER_UIDS
 int uid;
