--- mod_auth_external-2.1.14/pwauth/config.h.orig	Wed Jan 30 12:48:20 2002
+++ mod_auth_external-2.1.14/pwauth/config.h	Wed Jan 30 12:47:33 2002
@@ -145,7 +145,7 @@
  * last).
  */
 
-#define SERVER_UIDS 48		/* user "apache" */
+/*#define SERVER_UIDS 48		/* user "apache" */
 
 
 /* If MIN_UNIX_UID is defined to an integer, logins with uid numbers less than
--- mod_auth_external-2.1.14/pwauth/pwauth.c.orig	Wed Jan 30 12:48:46 2002
+++ mod_auth_external-2.1.14/pwauth/pwauth.c	Wed Jan 30 12:47:38 2002
@@ -647,8 +647,10 @@
 #endif
 }
 
+#ifdef SERVER_UIDS
 /* Array of uid numbers that may run this program */
 int server_uids[]= {SERVER_UIDS, 0};
+#endif
 
 main(int argc, char **argv)
 {
@@ -669,6 +671,7 @@
 	rlim.rlim_cur = rlim.rlim_max = 0;
 	(void)setrlimit(RLIMIT_CORE, &rlim);
 
+#ifdef SERVER_UIDS
 	/* Check that we were invoked by one of the listed uids or by root */
 	uid= getuid();
 	for (i= 0; server_uids[i] != 0 && server_uids[i] != uid; i++)
@@ -676,6 +679,7 @@
 	if (uid != server_uids[i])
 	    exit(2);
 
+#endif
 	/* Get the arguments (login and password) */
 #ifdef ENV_METHOD
 	if ((login= getenv("USER")) == NULL ||
