--- ppp-2.4.1/pppd/plugins/pppoe/utils.c.varargs	2003-08-13 12:15:22.000000000 +0200
+++ ppp-2.4.1/pppd/plugins/pppoe/utils.c	2003-08-13 12:17:04.000000000 +0200
@@ -182,7 +182,7 @@ vpoe_slprintf (char *buf, int buflen, ch
       continue;
     case 'r':
       f = va_arg (args, char *);
-#ifndef __powerpc__
+#if !defined(__powerpc__) && !defined(__x86_64__)
       n = vpoe_slprintf (buf, buflen + 1, f, va_arg (args, va_list));
 #else
       /* On the powerpc, a va_list is an array of 1 structure */
