diff -urN vim63/runtime/tools/vimspell.sh vim63.new/runtime/tools/vimspell.sh
--- vim63/runtime/tools/vimspell.sh	1999-08-01 14:01:46.000000000 +0200
+++ vim63.new/runtime/tools/vimspell.sh	2005-01-18 16:20:40.774519152 +0100
@@ -13,9 +13,7 @@
 # March 1999
 
 INFILE=$1
-OUTFILE=/tmp/vimspell.$$
-# if you have "tempfile", use the following line
-#OUTFILE=`tempfile`
+OUTFILE=`mktemp -t vimspellXXXXXX` || exit 1
 
 #
 # local spellings
