--- udev-114/Makefile.libudevdir	2007-08-05 14:08:47.000000000 +0200
+++ udev-114/Makefile	2007-08-10 19:21:35.000000000 +0200
@@ -112,7 +112,7 @@ LD = $(CROSS_COMPILE)gcc
 AR = $(CROSS_COMPILE)ar
 RANLIB = $(CROSS_COMPILE)ranlib
 
-CFLAGS		+= -g -Wall -pipe -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64
+CFLAGS		+= -g -Wall -pipe -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -D_libudevdir=\"${libudevdir}\"
 WARNINGS	= -Wstrict-prototypes -Wsign-compare -Wshadow \
 		  -Wchar-subscripts -Wmissing-declarations -Wnested-externs \
 		  -Wpointer-arith -Wcast-align -Wsign-compare -Wmissing-prototypes
--- udev-114/udev_rules_parse.c.libudevdir	2007-08-05 14:08:47.000000000 +0200
+++ udev-114/udev_rules_parse.c	2007-08-10 19:21:35.000000000 +0200
@@ -447,7 +447,7 @@ static int add_to_rules(struct udev_rule
 
 				/* allow programs in /lib/udev called without the path */
 				if (strchr(file, '/') == NULL) {
-					strlcpy(file, "/lib/udev/", sizeof(file));
+					strlcpy(file, _libudevdir "/", sizeof(file));
 					strlcat(file, value, sizeof(file));
 					pos = strchr(file, ' ');
 					if (pos)
--- udev-114/udev_rules.c.libudevdir	2007-08-05 14:08:47.000000000 +0200
+++ udev-114/udev_rules.c	2007-08-10 19:27:09.000000000 +0200
@@ -172,7 +172,7 @@ static int run_program(const char *comma
 
 	/* allow programs in /lib/udev called without the path */
 	if (strchr(argv[0], '/') == NULL) {
-		strlcpy(program, "/lib/udev/", sizeof(program));
+		strlcpy(program, _libudevdir "/", sizeof(program));
 		strlcat(program, argv[0], sizeof(program));
 		argv[0] = program;
 	}
