--- Linux-PAM-0.99.6.0/modules/pam_console/50-default.perms.mdvperms	2006-06-30 04:56:16.000000000 -0400
+++ Linux-PAM-0.99.6.0/modules/pam_console/50-default.perms	2006-08-29 13:34:54.000000000 -0400
@@ -4,13 +4,16 @@
 <sound>=/dev/dsp* /dev/audio* /dev/midi* \
 	/dev/mixer* /dev/sequencer* \
 	/dev/sound/* /dev/beep \
+	/dev/admm* /dev/aload* /dev/amidi* \
+	/dev/dmfm* /dev/dmmidi* /dev/sndstat \
 	/dev/snd/* /dev/adsp*
 <cdrom>=/dev/cdrom* /dev/cdroms/* /dev/cdwriter* /mnt/cdrom*
 <pilot>=/dev/pilot
-<jaz>=/mnt/jaz*
+<jaz>=/mnt/jaz* /dev/jaz*
 <zip>=/mnt/pocketzip* /mnt/zip* /dev/zip*
 <ls120>=/dev/ls120 /mnt/ls120*
 <scanner>=/dev/scanner* /dev/usb/scanner*
+<lp>=/dev/parport*
 <rio500>=/dev/usb/rio500
 <camera>=/mnt/camera* /dev/usb/dc2xx* /dev/usb/mdc800*
 <memstick>=/mnt/memstick*
@@ -20,7 +23,7 @@
 <fb>=/dev/fb /dev/fb[0-9]* \
      /dev/fb/*
 <kbd>=/dev/kbd
-<joystick>=/dev/js[0-9]*
+<joystick>=/dev/js[0-9]* /dev/input/by-path/*-joystick
 <v4l>=/dev/video* /dev/radio* /dev/winradio* /dev/vtx* /dev/vbi* \
       /dev/video/*
 <gpm>=/dev/gpmctl
@@ -34,8 +37,8 @@
 
 # permission definitions
 <console>  0660 <floppy>     0660 root.floppy
-<console>  0600 <sound>      0600 root
-<console>  0600 <cdrom>      0660 root.disk
+<console>  0660 <sound>      0660 root.audio
+<console>  0600 <cdrom>      0660 root.cdrom
 <console>  0600 <pilot>      0660 root.uucp
 <console>  0600 <jaz>        0660 root.disk
 <console>  0600 <zip>        0660 root.disk
@@ -49,7 +52,7 @@
 <console>  0600 <fb>         0600 root
 <console>  0600 <kbd>        0600 root
 <console>  0600 <joystick>   0600 root
-<console>  0600 <v4l>        0600 root
+<console>  0660 <v4l>        0660 root.video
 <console>  0700 <gpm>	     0700 root
 <console>  0600 <mainboard>  0600 root
 <console>  0600 <rio500>     0600 root
@@ -57,7 +60,7 @@
 <console>  0600 <bluetooth>  0600 root
 <console>  0600 <raw1394>    0600 root
 <console>  0600 <irda>       0600 root
-<console>  0600 <dvb>        0600 root
+<console>  0660 <dvb>        0660 root.video
 
 <xconsole> 0600 /dev/console 0600 root.root
 <console>  0600 <dri>	     0600 root
