--- Linux-PAM-0.99.3.0/modules/pam_console/Makefile.am.pbuild-rh	2005-12-15 23:07:37.000000000 +0100
+++ Linux-PAM-0.99.3.0/modules/pam_console/Makefile.am	2006-01-28 00:26:20.000000000 +0100
@@ -57,6 +57,8 @@
 pam_console_la_CFLAGS = $(AM_CFLAGS)
 pam_console_apply_CFLAGS = $(AM_CFLAGS)
 
+configfile.tab.h: configfile.tab.c
+
 configfile.tab.c: configfile.y
 	$(YACC) $(BISON_OPTS) -o $@ -p _pc_yy $<
 	sh $(srcdir)/sed-static $@
