--- Linux-PAM-0.99.6.3/modules/Makefile.am.redhat-modules	2006-08-28 11:56:17.000000000 +0200
+++ Linux-PAM-0.99.6.3/modules/Makefile.am	2006-10-19 14:58:49.000000000 +0200
@@ -3,6 +3,8 @@
 #
 
 SUBDIRS = pam_access pam_cracklib pam_debug pam_deny pam_echo \
+	pam_chroot pam_console pam_postgresok pam_rps \
+	pam_stack pam_timestamp pam_tally2 \
 	pam_env pam_filter pam_ftp pam_group pam_issue pam_keyinit \
 	pam_lastlog pam_limits pam_listfile pam_localuser pam_mail \
 	pam_mkhomedir pam_motd pam_nologin pam_permit pam_rhosts pam_rootok \
--- Linux-PAM-0.99.6.3/libpam/libpam.map.redhat-modules	2006-06-14 17:28:44.000000000 +0200
+++ Linux-PAM-0.99.6.3/libpam/libpam.map	2006-10-19 14:57:31.000000000 +0200
@@ -45,3 +45,14 @@
     pam_modutil_read;
     pam_modutil_write;
 };
+
+LIBPAM_INTERNAL_DONTUSE_1.0 {
+  global:
+    _pam_dispatch;
+    _pam_drop_env;
+    _pam_free_handlers;
+    _pam_init_handlers;
+    _pam_make_env;
+    _pam_start_handlers;
+    _pam_strdup;
+};
--- Linux-PAM-0.99.6.3/configure.in.redhat-modules	2006-08-31 00:14:38.000000000 +0200
+++ Linux-PAM-0.99.6.3/configure.in	2006-10-19 14:58:41.000000000 +0200
@@ -472,6 +472,10 @@
 	libpam_misc/Makefile conf/Makefile conf/pam_conv1/Makefile \
 	po/Makefile.in \
 	modules/Makefile \
+	modules/pam_chroot/Makefile modules/pam_console/Makefile \
+	modules/pam_postgresok/Makefile \
+	modules/pam_rps/Makefile modules/pam_stack/Makefile \
+	modules/pam_timestamp/Makefile modules/pam_tally2/Makefile \
 	modules/pam_access/Makefile modules/pam_cracklib/Makefile \
         modules/pam_debug/Makefile modules/pam_deny/Makefile \
 	modules/pam_echo/Makefile modules/pam_env/Makefile \
