--- openldap-2.0.25/build/lib-shared.mk.liblber	2002-01-04 15:38:06.000000000 -0500
+++ openldap-2.0.25/build/lib-shared.mk	2002-07-26 06:04:46.000000000 -0400
@@ -31,7 +31,7 @@ MKDEPFLAG = -l
 DYN_EXT=@DYN_EXT@
 
 $(LIBRARY):  version.lo
-	$(LTLIBLINK) -o $@ $(OBJS) version.lo $(EXTRA_LIBS)
+	$(LTLIBLINK) -o $@ $(OBJS) version.lo $(EXTRA_LIBS) $(LTDEPS)
 	$(RM) ../$@
 	d=`$(PWD)`; d=`$(BASENAME) $$d`; cd ..; $(LN_S) $$d/$@ $@; \
 	t=`$(BASENAME) $@ .la`.a; $(RM) $$t; $(LN_S) $$d/.libs/$$t $$t
--- openldap-2.0.25/libraries/libldap/Makefile.in.liblber	2002-01-04 15:38:19.000000000 -0500
+++ openldap-2.0.25/libraries/libldap/Makefile.in	2002-07-26 06:04:46.000000000 -0400
@@ -9,6 +9,8 @@ XLIBRARY = ../libldap.a
 
 PROGRAMS = apitest ltest ttest
 
+LTDEPS = -L../liblber/.libs
+
 SRCS	= bind.c open.c result.c error.c compare.c search.c \
 	controls.c messages.c references.c extended.c cyrus.c \
 	modify.c add.c modrdn.c delete.c abandon.c cache.c \
--- openldap-2.0.25/libraries/libldap_r/Makefile.in.liblber	2002-01-04 15:38:23.000000000 -0500
+++ openldap-2.0.25/libraries/libldap_r/Makefile.in	2002-07-26 06:08:57.000000000 -0400
@@ -7,6 +7,8 @@
 LIBRARY = libldap_r.la
 XLIBRARY = ../libldap_r.a
 
+LTDEPS = -L../liblber/.libs
+
 PROGRAMS = apitest ltest ttest
 
 XXDIR = $(srcdir)/../libldap
