--- ./configure.color	2002-08-27 15:34:55.000000000 +0200
+++ ./configure	2002-08-27 15:34:58.000000000 +0200
@@ -83,6 +83,9 @@
       -confdir)
           setconfdir=true;
           ;;
+      -compatibility_mode_for_colored_groff )
+          setcompatibility_mode_for_colored_groff=true;
+          ;;
       +lang)
 	  setlang=true
 	  ;;
@@ -483,6 +486,12 @@
   FILTERS="troff nroff jnroff eqn neqn jneqn tbl col vgrind refer grap pic pager cmp cat awk"
 fi
 
+if [ "x$setcompatibility_mode_for_colored_groff" = "xtrue" ]; then
+	troff="$troff -c"
+	nroff="$nroff -c"
+	jnroff="$jnroff -c"
+fi
+
 if [ x$default = x ]; then
   echo ""
   echo "If you don't have one of the following programs, enter \`missing'."
