diff -ur man-1.5j/src/makewhatis.sh man-1.5h1-tv/src/makewhatis.sh
--- man-1.5j/src/makewhatis.sh.whatbz2	Tue Jun 29 04:20:59 1999
+++ man-1.5j/src/makewhatis.sh	Mon Jun 26 13:16:14 2000
@@ -3,6 +3,7 @@
 # Created: Sun Jun 14 10:49:37 1992
 # Revised: Sat Jan  8 14:12:37 1994 by faith@cs.unc.edu
 # Revised: Sat Mar 23 17:56:18 1996 by micheal@actrix.gen.nz
+# Linux-Mandrake adaptations by Chmouel Boudjnah <chmouel@mandrakesoft.com>
 # Copyright 1992, 1993, 1994 Rickard E. Faith (faith@cs.unc.edu)
 # May be freely distributed and modified as long as copyright is retained.
 #
@@ -37,6 +38,9 @@
 # and should be first.
 # It is a bug to add /var/cache/man to DEFCATPATH.
 DEFMANPATH=/usr/share/man:/usr/man
+# This files is specific to Linux-Mandrake.
+[ -f /etc/sysconfig/i18n ] && . /etc/sysconfig/i18n
+[ -d /usr/man/$LANG/ ] && DEFMANPATH="$DEFMANPATH:/usr/man/$LANG/"
 DEFCATPATH=/var/cache/man:/usr/man/preformat:/usr/share/man:/usr/man
 
 # In case /usr is read-only, make /usr/man/whatis (etc) a symlink to
@@ -66,6 +70,8 @@
 
 topath=manpath
 
+deffindarg="-size +0"
+
 defmanpath=$DEFMANPATH
 defcatpath=
 
@@ -152,10 +158,10 @@
 	    section=$i
 	    curdir=$mandir/${pages}$i
 	    export section verbose curdir
-	    find . -name '*' $findarg -print | $AWK '
+	    find . -name '*' $findarg $deffindarg -print | $AWK '
 
 	    function readline() {
-	      if (use_zcat) {
+	      if (use_zcat || use_bzcat) {
 		result = (pipe_cmd | getline);
 		if (result < 0) {
 		  print "Pipe error: " pipe_cmd " " ERRNO > "/dev/stderr";
@@ -170,7 +176,7 @@
 	    }
 	    
 	    function closeline() {
-	      if (use_zcat) {
+	      if (use_zcat || use_bzcat) {
 		return close(pipe_cmd);
 	      } else {
 		return close(filename);
@@ -185,6 +191,7 @@
 		print "adding " filename > "/dev/stderr"
 	      }
 	      
+ 	      use_bzcat = (filename ~ /\.bz2$/);
 	      use_zcat = match(filename,"\\.Z$") ||
 			 match(filename,"\\.z$") || match(filename,"\\.gz$");
 	      if(use_zcat) {
@@ -202,6 +209,8 @@
 	      sub(/\..*/, "", progname);
 	      if (use_zcat) {
 		pipe_cmd = "zcat \"" filename "\"";
+	      } else if (use_bzcat) {
+		pipe_cmd = "bzcat " filename;
 		# try to avoid suspicious stuff
 		if (filename ~ /[;&|`$(]/) {
 		  print "ignored strange file name " filename " in " curdir > "/dev/stderr";
