--- tzdata/tzdata2007k/backward.orig	2007-12-31 13:23:09.000000000 -0200
+++ tzdata/tzdata2007k/backward	2008-01-02 10:55:32.000000000 -0200
@@ -108,3 +108,18 @@ Link	Etc/UTC			UTC
 Link	Etc/UTC			Universal
 Link	Europe/Moscow		W-SU
 Link	Etc/UTC			Zulu
+
+# links requested by users (to access timezones by the name of main cities)
+# pablo@mandriva.com
+# Canada/Saskatchewan 
+Link	America/Regina	America/Saskatoon
+# Canada/Mountain (Alberta)
+Link	America/Edmonton	America/Calgary
+# Canada/Atlantic (New Brunswick)
+Link	America/Halifax		America/Fredericton
+# America/Eastern (Ontario)
+Link	America/Montreal	America/Ontario
+# PRC
+Link	Asia/Shanghai		Asia/Beijing
+# South Africa
+Link	Africa/Johannesburg	Africa/Pretoria
