--- rpm-4.4.1/lib/rpminstall.c.ordererase	2004-11-28 20:52:43.000000000 +0000
+++ rpm-4.4.1/lib/rpminstall.c	2005-04-27 15:56:11.000000000 +0100
@@ -835,14 +835,12 @@
 	ps = rpmpsFree(ps);
     }
 
-#ifdef	NOTYET
     if (!stopUninstall && !(ia->installInterfaceFlags & INSTALL_NOORDER)) {
 	if (rpmtsOrder(ts)) {
 	    numFailed += numPackages;
 	    stopUninstall = 1;
 	}
     }
-#endif
 
     if (!stopUninstall) {
 	(void) rpmtsSetFlags(ts, (rpmtsFlags(ts) | RPMTRANS_FLAG_REVERSE));
