--- syslinux-1.76/memdisk/init.S16.orig	2003-07-14 06:31:17.000000000 +0200
+++ syslinux-1.76/memdisk/init.S16	2003-07-14 06:30:43.000000000 +0200
@@ -40,7 +40,7 @@ boot_flag:	.word 0xAA55
 _start:
 		jmp	start
 
-# This is the setup header, and it must start at %cs:2 (old 0x9020:2)
+/*# This is the setup header, and it must start at %cs:2 (old 0x9020:2)*/
 	
 		.ascii	"HdrS"		# header signature
 		.word	0x0203		# header version number (>= 0x0105)
@@ -55,7 +55,7 @@ type_of_loader:	.byte	0		# = 0, old one 
 					# See Documentation/i386/boot.txt for
 					# assigned ids
 	
-# flags, unused bits must be zero (RFU) bit within loadflags
+/*# flags, unused bits must be zero (RFU) bit within loadflags*/
 loadflags:
 LOADED_HIGH	= 1			# If set, the kernel is loaded high
 CAN_USE_HEAP	= 0x80			# If set, the loader also has set
