diff -urN byacc-1.9-old/defs.h byacc-1.9/defs.h
--- byacc-1.9-old/defs.h	Sun Oct 11 21:27:49 1992
+++ byacc-1.9/defs.h	Sat Apr 14 08:57:18 2001
@@ -286,4 +286,4 @@
 extern char *calloc();
 extern char *malloc();
 extern char *realloc();
-extern char *strcpy();
+/*extern char *strcpy();*/
diff -urN byacc-1.9-old/main.c byacc-1.9/main.c
--- byacc-1.9-old/main.c	Sat Apr 14 07:48:25 2001
+++ byacc-1.9/main.c	Sat Apr 14 08:45:42 2001
@@ -1,4 +1,7 @@
 #include <signal.h>
+#include <unistd.h>
+#include <string.h>
+#include <stdlib.h>
 #include "defs.h"
 
 char dflag;
diff -urN byacc-1.9-old/output.c byacc-1.9/output.c
--- byacc-1.9-old/output.c	Mon Feb 22 01:32:46 1993
+++ byacc-1.9/output.c	Sat Apr 14 08:46:25 2001
@@ -1,3 +1,4 @@
+#include <string.h>
 #include "defs.h"
 
 static int nvectors;
diff -urN byacc-1.9-old/reader.c byacc-1.9/reader.c
--- byacc-1.9-old/reader.c	Mon Feb 22 01:32:46 1993
+++ byacc-1.9/reader.c	Sat Apr 14 08:47:02 2001
@@ -1,3 +1,4 @@
+#include <string.h>
 #include "defs.h"
 
 /*  The line size must be a positive integer.  One hundred was chosen	*/
diff -urN byacc-1.9-old/symtab.c byacc-1.9/symtab.c
--- byacc-1.9-old/symtab.c	Mon Feb  5 01:03:20 1990
+++ byacc-1.9/symtab.c	Sat Apr 14 08:47:32 2001
@@ -1,3 +1,4 @@
+#include <string.h>
 #include "defs.h"
 
 
