--- mailx-8.1.1/Makefile.makefile-create-dirs	2002-11-11 06:56:05.000000000 -0500
+++ mailx-8.1.1/Makefile	2002-11-11 07:00:51.000000000 -0500
@@ -4,6 +4,12 @@
 PROG=	mail
 CC=gcc
 
+bindir	= /bin
+etcdir	= /etc
+prefix	= /usr
+datadir	= $(prefix)/share
+mandir	= $(prefix)/share/man
+
 # use second line starting from hamm release
 #CPPFLAGS=-I/usr/include/bsd -D_BSD_SOURCE -DIOSAFE
 CPPFLAGS=-D_BSD_SOURCE 
@@ -39,8 +45,11 @@ default: all
 	rm -f $(PROG) *.o *~
  
  install:
-	install -c -m 2755 -s $(PROG) $(DESTDIR)/usr/bin/
-	install -c -m 644 $(MFILES) $(DESTDIR)/usr/man/man1/
-	cd misc && install -c -m 644 $(EFILES) $(DESTDIR)/etc/
-	mkdir -p $(DESTDIR)/usr/share/mailx/
-	cd misc && install -c -m 644 $(SFILES) $(DESTDIR)/usr/share/mailx/
+	mkdir -p $(DESTDIR)$(bindir)
+	install -c -m 2755 -s $(PROG) $(DESTDIR)$(bindir)/
+	mkdir -p $(DESTDIR)$(mandir)/man1
+	install -c -m 644 $(MFILES) $(DESTDIR)$(mandir)/man1/
+	mkdir -p $(DESTDIR)$(etcdir)
+	cd misc && install -c -m 644 $(EFILES) $(DESTDIR)$(etcdir)/
+	mkdir -p $(DESTDIR)$(datadir)/mailx/
+	cd misc && install -c -m 644 $(SFILES) $(DESTDIR)$(datadir)/mailx/
