--- rsbac-admin-v1.2.3/src/Makefile.am~	2004-04-29 10:30:11.000000000 +0200
+++ rsbac-admin-v1.2.3/src/Makefile.am	2004-07-19 17:38:48.605083779 +0200
@@ -137,7 +137,7 @@
 	ln -s -f $(RSBACLIB) librsbac.a
 
 $(RSBACSO) : $(ADD_OBJECTS) $(RSBACLIB)
-	gcc -shared -fPIC -o $(RSBACSO) $(ADD_OBJECTS)
+	gcc -shared -fPIC -Wl,-soname,librsbac.so.1 -o $(RSBACSO) $(ADD_OBJECTS)
 	ln -s -f $(RSBACSO) librsbac.so
 
 install-exec-local: $(RSBACLIB) $(RSBACSO)
