--- MCONFIG.avx	2005-03-01 12:13:21.000000000 -0700
+++ MCONFIG	2005-03-01 12:13:29.000000000 -0700
@@ -20,7 +20,7 @@
 
 # If HAVE_SELINUX is set to "yes", the login will make sure the user is
 # logged into an appropriate security context
-HAVE_SELINUX=yes
+HAVE_SELINUX=no
 
 # If HAVE_SHADOW is set to "yes", then login, chfn, chsh, newgrp, passwd,
 # and vipw will not be built or installed from the login-utils
