--- rpm-4.2/lib/rpmrc.c.amd64	2003-07-24 12:40:05.000000000 -0400
+++ rpm-4.2/lib/rpmrc.c	2003-07-24 12:40:05.000000000 -0400
@@ -1278,6 +1278,13 @@ static void defaultMachine(/*@out@*/ con
 	}
 #	endif
 
+#	if defined(__linux__) && defined(__amd64__)
+	{
+	  /* Defaults to amd64 if compiler supports __amd64__ */
+	  strcpy(un.machine, "amd64");
+	}
+#	endif
+
 #	if defined(__linux__) && defined(__powerpc__)
 	{
 	    unsigned pvr = 0;
--- rpm-4.2/build/parseSpec.c.amd64	2003-07-24 12:40:05.000000000 -0400
+++ rpm-4.2/build/parseSpec.c	2003-07-25 06:04:46.000000000 -0400
@@ -190,6 +190,16 @@ static int copyNextLine(Spec spec, OFI_t
 }
 /*@=boundswrite@*/
 
+static const char *getAlternateArch(const char *arch)
+{
+    const char *alternate_arch = NULL;
+    if (! strncmp("x86_64", arch, sizeof("x86_64")-1))
+      alternate_arch = "amd64";
+    else if (! strncmp("amd64", arch, sizeof("amd64")-1))
+      alternate_arch = "x86_64";
+    return alternate_arch;
+}
+
 /*@-boundswrite@*/
 int readLine(Spec spec, int strip)
 {
@@ -277,13 +287,15 @@ retry:
       match = 0;
 	} else if (! strncmp("%ifarch", s, sizeof("%ifarch")-1)) {
 	const char *arch = rpmExpand("%{_target_cpu}", NULL);
+	const char *alternate_arch = getAlternateArch(arch);
 	s += 7;
-	match = matchTok(arch, s);
+	match = matchTok(arch, s) || (alternate_arch && matchTok(alternate_arch, s));
 	arch = _free(arch);
     } else if (! strncmp("%ifnarch", s, sizeof("%ifnarch")-1)) {
 	const char *arch = rpmExpand("%{_target_cpu}", NULL);
+	const char *alternate_arch = getAlternateArch(arch);
 	s += 8;
-	match = !matchTok(arch, s);
+	match = !matchTok(arch, s) && (!alternate_arch || !matchTok(alternate_arch, s));
 	arch = _free(arch);
     } else if (! strncmp("%ifos", s, sizeof("%ifos")-1)) {
 	const char *os = rpmExpand("%{_target_os}", NULL);
--- rpm-4.2/rpmrc.in.amd64	2003-07-24 12:40:05.000000000 -0400
+++ rpm-4.2/rpmrc.in	2003-07-24 12:40:05.000000000 -0400
@@ -19,6 +19,7 @@ optflags: i686 -O2 -fomit-frame-pointer 
 optflags: athlon -O2 -fomit-frame-pointer -pipe -march=athlon %{debugcflags}
 optflags: ia64 -O2 -pipe %{debugcflags}
 optflags: x86_64 -O2 -pipe %{debugcflags}
+optflags: amd64 -O2 -pipe %{debugcflags}
   
   # XXX Please note that -mieee has been added in rpm-3.0.5.
 optflags: alpha -O2 -mieee -mcpu=ev5 -pipe %{debugcflags}
@@ -124,6 +125,7 @@ arch_canon:	ppc64:	ppc64	16
 arch_canon:	sh: sh		17
 arch_canon:	xtensa: xtensa	18
 
+arch_canon:	amd64:	amd64	19
 arch_canon:	x86_64:	x86_64	19
 
 #############################################################
@@ -202,7 +204,8 @@ buildarchtranslate: s390x: s390x
 
 buildarchtranslate: ia64: ia64
 
-buildarchtranslate: x86_64: x86_64
+buildarchtranslate: amd64: amd64
+buildarchtranslate: x86_64: amd64
 
 #############################################################
 # Architecture compatibility
@@ -270,6 +273,7 @@ arch_compat: s390x: s390 noarch
 
 arch_compat: ia64: i686 noarch
 
+arch_compat: amd64: x86_64
 arch_compat: x86_64: athlon noarch
 
 os_compat:   IRIX64: IRIX
@@ -356,6 +360,7 @@ buildarch_compat: s390x: noarch
 
 buildarch_compat: ia64: noarch
 
+buildarch_compat: amd64: noarch
 buildarch_compat: x86_64: noarch
 
 macrofiles:	@RPMCONFIGDIR@/macros:@RPMCONFIGDIR@/%{_target}/macros:@SYSCONFIGDIR@/macros.specspo:@SYSCONFIGDIR@/macros.prelink:@SYSCONFIGDIR@/macros.solve:@SYSCONFIGDIR@/macros.up2date:@SYSCONFIGDIR@/macros:@SYSCONFIGDIR@/%{_target}/macros:~/.rpmmacros
--- rpm-4.2/installplatform.amd64	2003-07-24 12:40:05.000000000 -0400
+++ rpm-4.2/installplatform	2003-07-24 12:40:05.000000000 -0400
@@ -35,7 +35,7 @@ case "$arch" in
   alpha*) SUBSTS='s_alpha_alpha_ s_alpha_alphaev5_ s_alpha_alphaev56_ s_alpha_alphapca56_ s_alpha_alphaev6_ s_alpha_alphaev67_' ;;
   sparc*) SUBSTS='s_sparc\(64\|v9\)_sparc_ s_sparc64_sparcv9_;s_sparc\([^v]\|$\)_sparcv9\1_ s_sparcv9_sparc64_;s_sparc\([^6]\|$\)_sparc64\1_' ;;
   powerpc*|ppc*) SUBSTS='s_ppc64_ppc_ s_ppc\([^6]\|$\)_ppc64\1_ s_${arch}_ppciseries_ s_${arch}_ppcpseries_' ;;
-  x86_64*) SUBSTS='s/x86_64/x86_64/' ;;
+  x86_64*|amd64*) SUBSTS="s/${arch}/x86_64/ s/${arch}/amd64/" ;;
   s390*) SUBSTS='s_s390x_s390_ s_s390\([^x]\|$\)_s390x\1_' ;;
   *) SUBSTS=y___ ;;
 esac
@@ -69,7 +69,7 @@ for SUBST in $SUBSTS ; do
     ppc-linux) MULTILIBNO=1 ;;
     ppc64-linux) LIB=lib64; MULTILIBNO=2 ;;
     i?86-linux|athlon-linux) MULTILIBNO=1 ;;
-    x86_64-linux) LIB=lib64; MULTILIBNO=2 ;;
+    x86_64-linux|amd64-linux) LIB=lib64; MULTILIBNO=2 ;;
   esac
 
   if [ -n "$MULTILIBNO" ]; then
