diff -Naur php-5.1.2/ext/snmp/config.m4 php-5.1.2.oden/ext/snmp/config.m4
--- php-5.1.2/ext/snmp/config.m4	2005-11-29 19:26:00.000000000 +0100
+++ php-5.1.2.oden/ext/snmp/config.m4	2006-01-15 10:18:20.000000000 +0100
@@ -46,12 +46,12 @@
         test -f $i/snmp/include/ucd-snmp/snmp.h && SNMP_INCDIR=$i/snmp/include/ucd-snmp
       done
       for i in /usr /usr/snmp /usr/local /usr/local/snmp; do
-        test -f $i/lib/libsnmp.a || test -f $i/lib/libsnmp.$SHLIB_SUFFIX_NAME && SNMP_LIBDIR=$i/lib
+        test -f $i/$PHP_LIBDIR/libsnmp.a || test -f $i/$PHP_LIBDIR/libsnmp.$SHLIB_SUFFIX_NAME && SNMP_LIBDIR=$i/$PHP_LIBDIR
       done
     else
       SNMP_INCDIR=$PHP_SNMP/include
       test -d $PHP_SNMP/include/ucd-snmp && SNMP_INCDIR=$PHP_SNMP/include/ucd-snmp
-      SNMP_LIBDIR=$PHP_SNMP/lib
+      SNMP_LIBDIR=$PHP_SNMP/$PHP_LIBDIR
     fi
 
     if test -z "$SNMP_INCDIR"; then
