
Save and restore umask across requests correctly.

--- php-4.3.10/sapi/apache2handler/sapi_apache2.c.orig	2004-12-06 11:55:16.000000000 -0700
+++ php-4.3.10/sapi/apache2handler/sapi_apache2.c	2004-12-16 20:59:11.000000000 -0700
@@ -399,6 +399,20 @@
 	return APR_SUCCESS;
 }
 
+static int saved_umask;
+
+static void php_save_umask(void)
+{
+	saved_umask = umask(0777);
+	umask(saved_umask);
+}
+
+static void php_restore_umask(void)
+{
+	umask(saved_umask);
+}
+
+
 static int php_apache_request_ctor(request_rec *r, php_struct *ctx TSRMLS_DC)
 {
 	char *content_type;
@@ -533,6 +547,8 @@
 	} else {
 		zend_file_handle zfd = {0};
 
+		php_save_umask();
+
 		zfd.type = ZEND_HANDLE_FILENAME;
 		zfd.filename = (char *) r->filename;
 		zfd.free_filename = 0;
@@ -543,6 +559,9 @@
 		} else {
 			zend_execute_scripts(ZEND_INCLUDE TSRMLS_CC, NULL, 1, &zfd);
 		}
+
+		php_restore_umask();
+
 #if MEMORY_LIMIT
 		{
 			char *mem_usage;
