diff -uNr tiff-v3.5.5/config.site tiff-v3.5.5.acme/config.site
--- tiff-v3.5.5/config.site	Sun Sep 26 19:03:46 1999
+++ tiff-v3.5.5.acme/config.site	Sun Dec 31 16:56:56 2000
@@ -41,8 +41,8 @@
 # Package controls.
 #
 #DSO="auto"			# auto|IRIX|IRIX52 enable DSO support for system
-#JPEG="no"			# yes|no configure JPEG support (see below)
-#ZIP="no"			# yes|no configure Deflate support (see below)
+JPEG="yes"			# yes|no configure JPEG support (see below)
+ZIP="yes"			# yes|no configure Deflate support (see below)
 #LIBGL="auto"			# yes|no|auto configure IRIS GL-based tools
 #LIBIMAGE="auto"		# yes|no|auto configure SGI RGB image tools
 #HTML="no"			# yes|no install HTML documentation
@@ -95,7 +95,7 @@
 # which you can find on the Internet at ftp.uu.net:/graphics/jpeg/.
 #
 #DIRS_LIBINC="$DIRS_LIBINC ../src/jpeg-6b" # directory for jpeg include files
-#DIR_JPEGLIB=../src/jpeg-6b		# directory for libjpeg
+DIR_JPEGLIB=/usr/lib		# directory for libjpeg
 
 #
 # Deflate-specific parameters; used when Deflate support is enabled (see above).
@@ -110,7 +110,7 @@
 #     Look for it also at ftp://ftp.uu.net/pub/archiving/zip/zlib.
 #
 #DIRS_LIBINC="$DIRS_LIBINC ../zlib"	# directory for zlib include files
-#DIR_GZLIB="../zlib"			# directory for libz
+DIR_GZLIB="/usr/lib"			# directory for libz
 
 #
 # Miscellaneous parameters.
diff -uNr tiff-v3.5.5/libtiff/Makefile.in tiff-v3.5.5.acme/libtiff/Makefile.in
--- tiff-v3.5.5/libtiff/Makefile.in	Sun Dec 31 16:59:06 2000
+++ tiff-v3.5.5.acme/libtiff/Makefile.in	Sun Dec 31 17:03:02 2000
@@ -152,7 +152,7 @@
 	${RANLIB} libtiff.a
 
 $(SHLIB): $(SHOBJS)
-	gcc -o $(SHLIB) -Wl,-soname,$(SONAME) -shared $(SHOBJS)
+	gcc -o $(SHLIB) -Wl,-soname,$(SONAME) -shared $(SHOBJS) -ljpeg -lz
 
 #
 # NB: The configure script verifies that the configured
