--- rpm-4.2/rpmdb/rpmdb.c.lock	2003-03-14 16:08:08.000000000 +0100
+++ rpm-4.2/rpmdb/rpmdb.c	2003-11-25 09:40:29.000000000 +0100
@@ -945,6 +945,14 @@
     }
 /*@=newreftrans@*/
 
+    /* unlock database */
+    {
+      struct flock l;
+      int ld = (db->db_perms >> 16) - 1;
+      memset(&l, 0, sizeof(l));
+      l.l_type = F_UNLCK;
+      fcntl(ld, F_SETLK, &l);
+    }
     /*@-refcounttrans@*/ db = _free(db); /*@=refcounttrans@*/
     /*@=usereleased@*/
 
@@ -1007,10 +1015,7 @@
     /*@=nullpass@*/
     if (!(db->db_home && db->db_home[0] != '%')) {
 	rpmError(RPMERR_DBOPEN, _("no dbpath has been set\n"));
-	db->db_root = _free(db->db_root);
-	db->db_home = _free(db->db_home);
-	db = _free(db);
-	/*@-globstate@*/ return NULL; /*@=globstate@*/
+	goto fail;
     }
     db->db_errpfx = rpmExpand( (epfx && *epfx ? epfx : _DB_ERRPFX), NULL);
     db->db_remove_env = 0;
@@ -1018,9 +1023,36 @@
     db->db_ndbi = dbiTagsMax;
     db->_dbi = xcalloc(db->db_ndbi, sizeof(*db->_dbi));
     db->nrefs = 0;
+    /* lock database using exclusive lock on write, and shared lock on read */
+    {
+      char *lockfilename = malloc(strlen(db->db_root) + strlen(db->db_home) + strlen("/RPMLOCK"));
+      int ld;
+      /* compute lockfilename and open it according to mode used */
+      sprintf(lockfilename, "%s%s/RPMLOCK", db->db_root, db->db_home);
+      ld = open(lockfilename, (mode == O_RDONLY ? O_RDONLY : O_WRONLY) | (getuid() ? 0 : O_CREAT), 0644);
+      if (ld >= 0) {
+	struct flock l;
+	do {
+	  memset(&l, 0, sizeof(l));
+	  l.l_type = mode == O_RDONLY ? F_RDLCK : F_WRLCK;
+	} while (fcntl(ld, F_SETLKW, &l) < 0);
+	/* HACK to keep ld without breaking compabilities (changing db size) */
+	db->db_perms |= (ld + 1) << 16;
+      } else{
+	rpmError( RPMERR_DBOPEN,
+		  _("cannot open lock file %s in %s mode\n"), lockfilename, (mode == O_RDONLY ? "shared" : "exclusive"));
+	goto fail;
+      }
+    }
     /*@-globstate@*/
     return rpmdbLink(db, "rpmdbCreate");
     /*@=globstate@*/
+
+ fail:
+    db->db_root = _free(db->db_root);
+    db->db_home = _free(db->db_home);
+    db = _free(db);
+    /*@-globstate@*/ return NULL; /*@=globstate@*/    
 }
 /*@=mods@*/
 
@@ -3549,6 +3581,10 @@
 	break;
     }
 
+    sprintf(filename, "%s/%s/RPMLOCK", prefix, dbpath);
+    (void)rpmCleanPath(filename);
+    xx = unlink(filename);
+
     sprintf(filename, "%s/%s", prefix, dbpath);
     (void)rpmCleanPath(filename);
     xx = rmdir(filename);
@@ -3656,6 +3692,11 @@
     case 0:
 	break;
     }
+
+    sprintf(ofilename, "%s/%s/RPMLOCK", prefix, olddbpath);
+    (void)rpmCleanPath(ofilename);
+    xx = unlink(ofilename);
+
     if (rc || _olddbapi == _newdbapi)
 	return rc;
 
