--- krb5-1.2.5/src/util/db2/configure.in.jj	1998-01-15 11:51:48.000000000 -0500
+++ krb5-1.2.5/src/util/db2/configure.in	2002-07-19 12:45:14.000000000 -0400
@@ -33,11 +33,11 @@ dnl checks for header files
 AC_CHECK_HEADERS(unistd.h)
 dnl checks for typedefs
 AC_TYPE_SIZE_T
+AC_HEADER_STDC
 
 dnl AC_COMPILE_TYPE(TYPE, DEFAULT)
 define(AC_COMPILE_TYPE,
-[AC_REQUIRE([AC_HEADER_STDC])dnl
-AC_MSG_CHECKING(for $1)
+[AC_MSG_CHECKING(for $1)
 AC_CACHE_VAL(ac_cv_type_$1,
 [AC_TRY_COMPILE([#include <sys/types.h>
 #if STDC_HEADERS
@@ -76,5 +76,14 @@ AC_SUBST(MEMMOVE_OBJ)
 AC_SUBST(MKSTEMP_OBJ)
 AC_SUBST(STRERROR_OBJ)
 AC_CHECK_FUNCS(memmove mkstemp strerror)
+
+AC_ARG_WITH([ccopts],
+[  --with-ccopts=CCOPTS    select compiler command line options],
+AC_MSG_RESULT(CCOPTS is $withval)
+CCOPTS=$withval
+CFLAGS="$CFLAGS $withval",
+CCOPTS=)dnl
+AC_SUBST(CCOPTS)
+
 dnl checks for system services
 AC_OUTPUT(Makefile obj/Makefile)
