The kdb stuff needs the db libraries linked in.

--- src/lib/kdb/Makefile.in	Mon Oct  4 22:28:37 1999
+++ src/lib/kdb/Makefile.in	Mon Oct  4 22:28:42 1999
@@ -16,7 +16,7 @@
 SHLIB_EXPDEPS = \
 	$(TOPLIBD)/libk5crypto$(SHLIBEXT) \
 	$(TOPLIBD)/libkrb5$(SHLIBEXT)
-SHLIB_EXPLIBS=-lkrb5 -lcom_err -lk5crypto $(LIBS)
+SHLIB_EXPLIBS=-lkrb5 -lcom_err -lk5crypto $(LIBS) $(DB_LIB)
 SHLIB_DIRS=-L$(TOPLIBD)
 SHLIB_RDIRS=$(KRB5_LIBDIR)
 
