--- mysql-4.0.13/acinclude.m4.quotes	2003-08-01 13:39:34.000000000 +0200
+++ mysql-4.0.13/acinclude.m4	2003-08-01 13:43:53.000000000 +0200
@@ -762,7 +762,7 @@ AC_MSG_CHECKING(for OpenSSL)
     case "$CLIENT_EXTRA_LDFLAGS $MYSQLD_EXTRA_LDFLAGS" in
 	*-all-static*) using_static="yes" ;;
     esac
-    if test $using_static = "yes"
+    if test "$using_static" = "yes"
     then
       echo "You can't use the --all-static link option when using openssl."
       exit 1
