--- mysql-4.0.11a-gamma/vio/Makefile.am.my_net	2003-03-02 08:19:06.000000000 +0100
+++ mysql-4.0.11a-gamma/vio/Makefile.am	2003-03-02 08:19:39.000000000 +0100
@@ -16,6 +16,7 @@
 
 INCLUDES =		-I$(srcdir)/../include -I../include $(openssl_includes)
 LDADD =			@CLIENT_EXTRA_LDFLAGS@ libvio.a $(openssl_libs) 
+LIBS =			@LIBS@ ../libmysql/.libs/libmysqlclient.so
 pkglib_LIBRARIES =	libvio.a
 noinst_PROGRAMS	=	test-ssl test-sslserver test-sslclient
 noinst_HEADERS =	
