diff -Naur php-4.3.0/ext/standard/credits.c php-4.3.0.oden/ext/standard/credits.c
--- php-4.3.0/ext/standard/credits.c	2002-12-27 00:39:51.000000000 +0100
+++ php-4.3.0.oden/ext/standard/credits.c	2003-01-14 08:29:46.000000000 +0100
@@ -108,6 +108,16 @@
 		php_info_print_table_end();
 	}
 
+	if (flag & PHP_CREDITS_RPM_PACKAGING) {
+		/* RPM Packaging Team */
+		php_info_print_table_start();
+		php_info_print_table_colspan_header(2, "Mandrake Linux PHP RPM Packaging Team");
+		php_info_print_table_header(2, "Name", "Organization");
+		CREDIT_LINE("Oden Eriksson", "Deserve-IT Networks");
+		CREDIT_LINE("Jean-Michel Dault", "MandrakeSoft");
+		php_info_print_table_end();
+	}
+
 	if (flag & PHP_CREDITS_FULLPAGE) {
 		PUTS("</center></body></html>\n");
 	}
diff -Naur php-4.3.0/ext/standard/credits.h php-4.3.0.oden/ext/standard/credits.h
--- php-4.3.0/ext/standard/credits.h	2002-02-28 09:26:44.000000000 +0100
+++ php-4.3.0.oden/ext/standard/credits.h	2003-01-14 07:12:22.000000000 +0100
@@ -33,6 +33,7 @@
 #define PHP_CREDITS_FULLPAGE		(1<<5)
 #define PHP_CREDITS_QA				(1<<6)
 #define PHP_CREDITS_WEB				(1<<7)
+#define PHP_CREDITS_RPM_PACKAGING		(1<<8)
 #define PHP_CREDITS_ALL				0xFFFFFFFF
 
 #endif /* HAVE_CREDITS_DEFS */
diff -Naur php-4.3.0/ext/standard/info.c php-4.3.0.oden/ext/standard/info.c
--- php-4.3.0/ext/standard/info.c	2002-11-14 22:17:00.000000000 +0100
+++ php-4.3.0.oden/ext/standard/info.c	2003-01-14 07:11:20.000000000 +0100
@@ -804,6 +804,7 @@
 	REGISTER_LONG_CONSTANT("CREDITS_DOCS",	PHP_CREDITS_DOCS, CONST_PERSISTENT|CONST_CS);
 	REGISTER_LONG_CONSTANT("CREDITS_FULLPAGE",	PHP_CREDITS_FULLPAGE, CONST_PERSISTENT|CONST_CS);
 	REGISTER_LONG_CONSTANT("CREDITS_QA",	PHP_CREDITS_QA, CONST_PERSISTENT|CONST_CS);
+	REGISTER_LONG_CONSTANT("CREDITS_RPM_PACKAGING",	PHP_CREDITS_RPM_PACKAGING, CONST_PERSISTENT|CONST_CS);
 	REGISTER_LONG_CONSTANT("CREDITS_ALL",	PHP_CREDITS_ALL, CONST_PERSISTENT|CONST_CS);
 }
 /* }}} */
diff -Naur php-4.3.0/ext/standard/info.h php-4.3.0.oden/ext/standard/info.h
--- php-4.3.0/ext/standard/info.h	2002-10-06 19:04:10.000000000 +0200
+++ php-4.3.0.oden/ext/standard/info.h	2003-01-14 07:12:40.000000000 +0100
@@ -46,6 +46,7 @@
 #define PHP_CREDITS_FULLPAGE		(1<<5)
 #define PHP_CREDITS_QA				(1<<6)
 #define PHP_CREDITS_WEB             (1<<7)
+#define PHP_CREDITS_RPM_PACKAGING               (1<<8)
 #define PHP_CREDITS_ALL				0xFFFFFFFF
 
 #endif /* HAVE_CREDITS_DEFS */
