--- ./src/ls.c.dumbterm	2002-10-08 17:00:53.000000000 +0200
+++ ./src/ls.c	2002-10-08 17:00:53.000000000 +0200
@@ -98,6 +98,7 @@
 
 #include "system.h"
 #include <fnmatch.h>
+#include <termcap.h>
 
 #include "acl.h"
 #include "argmatch.h"
@@ -1525,7 +1526,9 @@
 
 	    print_with_color = (i == color_always
 				|| (i == color_if_tty
-				    && isatty (STDOUT_FILENO)));
+				    && isatty (STDOUT_FILENO))
+				    && tgetent(NULL, getenv("TERM")) > 0
+				    && tgetstr("md", NULL) > 0);
 
 	    if (print_with_color)
 	      {
--- ./src/Makefile.in.dumbterm	2002-10-05 09:00:37.000000000 +0200
+++ ./src/Makefile.in	2002-10-08 17:02:43.000000000 +0200
@@ -189,10 +189,10 @@
 # replacement functions defined in libfetish.a.
 LDADD = ../lib/libfetish.a @LIBINTL@ ../lib/libfetish.a
 
-dir_LDADD = $(LDADD) @LIB_CLOCK_GETTIME@
-ls_LDADD = $(LDADD) @LIB_CLOCK_GETTIME@
+dir_LDADD = $(LDADD) @LIB_CLOCK_GETTIME@ -ltermcap
+ls_LDADD = $(LDADD) @LIB_CLOCK_GETTIME@ -ltermcap
 shred_LDADD = $(LDADD) @LIB_CLOCK_GETTIME@
-vdir_LDADD = $(LDADD) @LIB_CLOCK_GETTIME@
+vdir_LDADD = $(LDADD) @LIB_CLOCK_GETTIME@ -ltermcap
 
 sort_LDADD = $(LDADD) @POW_LIB@
 
