--- openldap-2.0.7/build/openldap.m4.old	Mon May  7 23:31:48 2001
+++ openldap-2.0.7/build/openldap.m4	Mon May  7 23:33:20 2001
@@ -252,7 +252,7 @@
 
 	AC_TRY_LINK([
 #ifdef HAVE_DB_185_H
-# include <db_185.h>
+# include <db3/db_185.h>
 #else
 # include <db.h>
 #endif
@@ -288,9 +288,9 @@
 AC_DEFUN([OL_BERKELEY_DB_LINK],
 [ol_cv_lib_db=no
 OL_BERKELEY_DB_TRY(ol_cv_db_none)
+OL_BERKELEY_DB_TRY(ol_cv_db_db3,[-ldb3]) 
 OL_BERKELEY_DB_TRY(ol_cv_db_db,[-ldb])
-OL_BERKELEY_DB_TRY(ol_cv_db_db3,[-ldb3])
 OL_BERKELEY_DB_TRY(ol_cv_db_db2,[-ldb2])
 OL_BERKELEY_DB_TRY(ol_cv_db_db1,[-ldb1])
 ])
 dnl
@@ -380,12 +380,12 @@
 dnl Find old Berkeley DB 1.85/1.86
 AC_DEFUN([OL_BERKELEY_COMPAT_DB],
 [ol_cv_berkeley_db=no
-AC_CHECK_HEADERS(db_185.h db.h)
+AC_CHECK_HEADERS(db3/db_185.h db.h)
 if test $ac_cv_header_db_185_h = yes -o $ac_cv_header_db_h = yes; then
 	AC_CACHE_CHECK([if Berkeley DB header compatibility], [ol_cv_header_db1],[
 		AC_EGREP_CPP(__db_version_1,[
 #if HAVE_DB_185_H
-#	include <db_185.h>
+#	include <db3/db_185.h>
 #else
 #	include <db.h>
 #endif


