--- vixie-cron-3.0.1/config.h.syslog	Mon Aug  7 22:25:09 2000
+++ vixie-cron-3.0.1/config.h	Mon Aug  7 22:25:15 2000
@@ -86,4 +86,4 @@
 			 * are both defined, then logging will go to both
 			 * places.
 			 */
-/*#define SYSLOG	 			/*-*/
+#define SYSLOG
--- vixie-cron-3.0.1/pathnames.h.syslog	Mon Aug  7 22:25:26 2000
+++ vixie-cron-3.0.1/pathnames.h	Mon Aug  7 22:25:50 2000
@@ -49,7 +49,7 @@
 			 */
 #define	ALLOW_FILE	"/etc/cron.allow"		/*-*/
 #define DENY_FILE	"/etc/cron.deny"		/*-*/
-#define LOG_FILE	"/var/log/cron"		/*-*/
+/* #define LOG_FILE	"/var/log/cron"		*/
 
 			/* where should the daemon stick its PID?
 			 */
