--- module-init-tools-3.0/depmod.c.failed-symb	2004-03-16 13:46:15.886366658 +0100
+++ module-init-tools-3.0/depmod.c	2004-03-16 13:48:10.977190877 +0100
@@ -105,6 +105,7 @@
 }
 
 static int print_unknown;
+static int unknown_symbol = 0;
 
 struct module *find_symbol(const char *name, const char *modname, int weak)
 {
@@ -119,9 +120,10 @@
 			return s->owner;
 	}
 
-	if (print_unknown && !weak)
+	if (print_unknown && !weak) {
 		warn("%s needs unknown symbol %s\n", modname, name);
-
+		unknown_symbol = 1;
+	}
 	return NULL;
 }
 
@@ -775,5 +777,5 @@
 	free(dirname);
 	free(version);
 	
-	return 0;
+	return unknown_symbol;
 }
