diff -up tt/Makefile.orig tt/Makefile
--- tt/Makefile.orig	2002-08-01 13:22:02.000000000 +0200
+++ tt/Makefile	2002-08-01 13:22:27.000000000 +0200
@@ -295,6 +295,11 @@ symlinks:
 	fi
 
 preconfig: include/linux/version.h symlinks
+	if [ -f .need_mrproper ]; then \
+		rm .need_mrproper; \
+		$(MAKE) mrproper;  \
+		$(MAKE) preconfig;  \
+	fi	
 
 oldconfig: preconfig
 	$(CONFIG_SHELL) scripts/Configure -d arch/$(ARCH)/config.in
