--- 6mdk/Makefile	2003-07-22 21:55:39.000000000 +0300
+++ tmb1/Makefile	2003-07-25 21:40:06.000000000 +0300
@@ -261,12 +261,15 @@ MRPROPER_FILES = \
 	$(TOPDIR)/include/linux/modversions.h \
 	3rdparty/Makefile.meta 3rdparty/Makefile 3rdparty/Makefile.drivers \
         3rdparty/Config.in \
-	kernel.spec
+	kernel.spec \
+	net/ipsec/alg/perlasm
+	
 
 # directories removed with 'make mrproper'
 MRPROPER_DIRS = \
 	include/config \
-	$(TOPDIR)/include/linux/modules
+	$(TOPDIR)/include/linux/modules \
+	net/ipsec/alg/lib*
 
 
 include arch/$(ARCH)/Makefile
