--- wget-1.8.2/src/url.c.fpons	2002-09-04 16:16:52.000000000 +0200
+++ wget-1.8.2/src/url.c	2002-09-04 16:32:14.000000000 +0200
@@ -499,14 +499,18 @@
 int
 url_skip_uname (const char *url)
 {
-  const char *p;
+  const char *p, *pp;
 
-  /* Look for '@' that comes before '/' or '?'. */
-  p = (const char *)strpbrk (url, "/?@");
-  if (!p || *p != '@')
-    return 0;
+  /* Look for last '@' that comes before '/' or '?'. */
+  pp = url;
+  while ((p = (const char *)strpbrk (pp, "/?@")) != NULL) {
+    if (*p != '@')
+      break;
+    /* Found '@' character so go on with possible next '@'. */
+    pp = p + 1;
+  }
 
-  return p - url + 1;
+  return pp != url ? pp - url: 0;
 }
 
 static int
