--- rpm-4.2/platform.in.configure-xpath	2003-07-08 04:49:53.000000000 -0400
+++ rpm-4.2/platform.in	2003-07-08 04:49:54.000000000 -0400
@@ -115,6 +115,8 @@
 @mandrake@  CXXFLAGS="${CXXFLAGS:-%optflags}" ; export CXXFLAGS ; \
 @mandrake@  FFLAGS="${FFLAGS:-%optflags}" ; export FFLAGS ; \
 @mandrake@  %{?__libtoolize:(cd $CONFIGURE_TOP; [ -f configure.in -o -f configure.ac ] && %{__libtoolize} --copy --force)} ; \
+@mandrake@  [ -f $CONFIGURE_TOP/configure.in -o -f $CONFIGURE_TOP/configure.ac ] && \
+@mandrake@        CONFIGURE_XPATH="--x-includes=%{_prefix}/X11R6/include --x-libraries=%{_prefix}/X11R6/%{_lib}" \
 @mandrake@  $CONFIGURE_TOP/configure --build=%{_target_platform} \\\
 @mandrake@ 	--prefix=%{_prefix} \\\
 @mandrake@	--exec-prefix=%{_exec_prefix} \\\
@@ -128,7 +130,7 @@
 @mandrake@	--localstatedir=%{_localstatedir} \\\
 @mandrake@	--sharedstatedir=%{_sharedstatedir} \\\
 @mandrake@	--mandir=%{_mandir} \\\
-@mandrake@	--infodir=%{_infodir}
+@mandrake@	--infodir=%{_infodir} $CONFIGURE_XPATH
 @mandrake@
 @mandrake@%makeinstall_std make DESTDIR=%{?buildroot:%{buildroot}} install
 @mandrake@
--- rpm-4.2/macros.in.configure-xpath	2003-07-08 04:49:53.000000000 -0400
+++ rpm-4.2/macros.in	2003-07-08 04:52:30.000000000 -0400
@@ -986,6 +986,8 @@ package or when debugging this package.\
   CXXFLAGS="${CXXFLAGS:-%optflags}" ; export CXXFLAGS ; \
   FFLAGS="${FFLAGS:-%optflags}" ; export FFLAGS ; \
   %{?__libtoolize:(cd $CONFIGURE_TOP; [ -f configure.in -o -f configure.ac ] && %{__libtoolize} --copy --force)} ; \
+  [ -f $CONFIGURE_TOP/configure.in -o -f $CONFIGURE_TOP/configure.ac ] && \
+        CONFIGURE_XPATH="--x-includes=%{_prefix}/X11R6/include --x-libraries=%{_prefix}/X11R6/%{_lib}" \
   $CONFIGURE_TOP/configure %{_target_platform} \\\
 	--program-prefix=%{?_program_prefix} \\\
  	--prefix=%{_prefix} \\\
@@ -1000,7 +1002,7 @@ package or when debugging this package.\
 	--localstatedir=%{_localstatedir} \\\
 	--sharedstatedir=%{_sharedstatedir} \\\
 	--mandir=%{_mandir} \\\
-	--infodir=%{_infodir}
+	--infodir=%{_infodir} $CONFIGURE_XPATH
 
 #------------------------------------------------------------------------------
 # The make install analogue of %configure:
