--- man-1.5i/src/apropos.sh.usr	Thu Jun  7 15:20:04 2001
+++ man-1.5i/src/apropos.sh	Thu Jun  7 15:20:04 2001
@@ -67,7 +67,7 @@
 found=0
 while [ $found = 0 -a -n "$1" ]
 do
-    for d in $manpath /usr/lib
+    for d in /var/cache/man /usr/lib
     do
         if [ -f $d/whatis ]
         then
@@ -101,7 +101,7 @@
     done
     nothing=
     found=0
-    for d in $manpath /usr/lib
+    for d in /var/cache/man /usr/lib
     do
         if [ -f $d/whatis ]
         then
--- man-1.5i/src/makewhatis.sh.usr	Thu Jun  7 15:20:04 2001
+++ man-1.5i/src/makewhatis.sh	Thu Jun  7 15:20:30 2001
@@ -81,7 +81,7 @@
 	continue;;
     -s) setsections=1
 	continue;;
-    -u) findarg="-ctime 0"
+    -u) findarg="-newer /var/cache/man/whatis"
 	update=1
 	continue;;
     -v) verbose=1
@@ -121,14 +121,7 @@
 # first truncate all the whatis files that will be created new,
 # then only update - we might visit the same directory twice
 if [ x$update = x ]; then
-   for pages in man cat
-   do
-      eval path="\$$pages"path
-      for mandir in $path
-      do
-	 cp /dev/null $mandir/whatis
-      done
-   done
+   cp /dev/null /var/cache/man/whatis
 fi
 
 for pages in man cat
@@ -140,11 +133,6 @@
      if [ x$verbose != x ]; then
 	echo "about to enter $mandir" > /dev/tty
      fi
-     if [ -s ${mandir}/whatis -a $pages = man -a x$update = x ]; then
-	if [ x$verbose != x ]; then
-	   echo skipping $mandir - we did it already > /dev/tty
-	fi
-     else      
        here=`pwd`
        cd $mandir
        for i in $sections
@@ -346,23 +334,14 @@
 
        cd $here
 
-       # kludge for Slackware's /usr/man/preformat
-       if [ $mandir = /usr/man/preformat ]
+       if [ -f /var/cache/man/whatis ]
        then
-	 mandir1=/usr/man
-       else
-	 mandir1=$mandir
+	 cat /var/cache/man/whatis >> $TMPFILE
        fi
+       sed '/^$/d' < $TMPFILE | sort | uniq > /var/cache/man/whatis
 
-       if [ -f ${mandir1}/whatis ]
-       then
-	 cat ${mandir1}/whatis >> $TMPFILE
-       fi
-       sed '/^$/d' < $TMPFILE | sort | uniq > ${mandir1}/whatis
-
-       chmod 644 ${mandir1}/whatis
+       chmod 644 /var/cache/man/whatis
        rm $TMPFILE
-     fi
    done
 done
 
