--- man-1.5h1/gencat/genlib.c.tv	Tue Jul 18 15:20:21 2000
+++ man-1.5h1/gencat/genlib.c	Tue Jul 18 15:20:14 2000
@@ -114,7 +114,7 @@
     error(NULL, "out of memory");
 }
 	    
-static char *getline(fd)
+static char *my_getline(fd)
 int fd;
 {
     static long	len = 0, curlen = BUFSIZ;
@@ -229,7 +229,7 @@
 	    ++cptr;
 	    switch (*cptr) {
 	      case '\0':
-		cptr = getline(fd);
+		cptr = my_getline(fd);
 		if (!cptr) error(NULL, "premature end of file");
 		msglen += strlen(cptr);
 		i = tptr - msg;
@@ -363,7 +363,7 @@
 
     hconst[0] = '\0';
     
-    while (cptr = getline(fd)) {
+    while (cptr = my_getline(fd)) {
 	if (*cptr == '$') {
 	    ++cptr;
 	    if (strncmp(cptr, "set", 3) == 0) {
--- man-1.5h1/man/Makefile.in.tv	Tue Jul 18 15:41:05 2000
+++ man-1.5h1/man/Makefile.in	Tue Jul 18 15:44:36 2000
@@ -34,6 +34,15 @@
 	for i in $(MAN8); \
 		do if test -f $$i.8; then $(INSTALL) $$i.8 $(PREFIX)$(mandir)/man8/$$i.@man8ext@; fi; done
 
+install-l10n: $(ALL)
+	mkdir -p $(PREFIX)$(mandir)/$(SLANG)/man1 $(PREFIX)$(mandir)/$(SLANG)/man5 $(PREFIX)$(mandir)/$(SLANG)/man8
+	for i in $(MAN1); \
+		do $(INSTALL) $$i.1 $(PREFIX)$(mandir)/$(SLANG)/man1/$$i.@man1ext@; done
+	for i in $(MAN5); \
+		do $(INSTALL) $$i.5 $(PREFIX)$(mandir)/$(SLANG)/man5/$$i.@man5ext@; done
+	for i in $(MAN8); \
+		do if test -f $$i.8; then $(INSTALL) $$i.8 $(PREFIX)$(mandir)/$(SLANG)/man8/$$i.@man8ext@; fi; done
+
 clean:
 	rm -f core *.in *.@man1ext@ *.@man5ext@ *~
 
@@ -49,7 +58,7 @@
 	@for i in @languages@; do if test -d $$i; then echo; \
 		echo "==== Installing the `cat $$i.txt` man pages. ===="; \
 		cd $$i; SLANG=/$$i; if test $$SLANG = /en; then SLANG= ; fi; \
-		export SLANG; make -f ../Makefile install; cd ..; \
+		export SLANG; make -f ../Makefile install-l10n; cd ..; \
 		else echo "==== No $$i man pages found. ===="; fi; done
 
 cleansubdirs:
