--- src/engine/Makefile.in.orig	2003-06-02 17:15:37.000000000 +0200
+++ src/engine/Makefile.in	2003-06-02 17:15:48.000000000 +0200
@@ -186,7 +186,7 @@
 DEFAULT_INCLUDES =  -I. -I$(srcdir) -I$(top_builddir)/autocfg
 CPPFLAGS = @CPPFLAGS@
 LDFLAGS = @LDFLAGS@
-LIBS = @LIBS@
+LIBS = @LIBS@ $(EXPATH) -lstdc++
 depcomp = $(SHELL) $(top_srcdir)/tools/depcomp
 @AMDEP_TRUE@DEP_FILES = $(DEPDIR)/arena.Plo $(DEPDIR)/base.Plo \
 @AMDEP_TRUE@	$(DEPDIR)/context.Plo $(DEPDIR)/datastr.Plo \
