--- lynx2-8-5/lynx.cfg.pix	2003-01-22 02:27:27.000000000 +0100
+++ lynx2-8-5/lynx.cfg	2003-01-22 02:28:02.000000000 +0100
@@ -1,6 +1,7 @@
 # lynx.cfg file.
-# The default placement for this file is /usr/local/lib/lynx.cfg (Unix)
-#                                     or Lynx_Dir:lynx.cfg (VMS)
+#
+# Please don't edit this file directly (it is updated with every Mandrake
+# Linux update, overwriting your changes). Instead, edit /etc/lynx-site.cfg.
 #
 # $Format: "#PRCS LYNX_VERSION \"$ProjectVersion$\""$
 #PRCS LYNX_VERSION "2.8.5dev.12"
@@ -92,7 +93,7 @@
 #
 # Normally we expect you will connect to a remote site, e.g., the Lynx starting
 # site:
-STARTFILE:http://lynx.browser.org/
+STARTFILE:http://www.mandrakelinux.com
 #
 # As an alternative, you may want to use a local URL.  A good choice for this is
 # the user's home directory:
@@ -115,9 +116,9 @@
 # http://www.trill-home.com/lynx/lynx_help/lynx_help_main.html
 #   This should be changed to the local path.
 #
-HELPFILE:http://www.trill-home.com/lynx/lynx_help/lynx_help_main.html
+HELPFILE:file://localhost/PATH_TO/lynx_help/lynx_help_main.html
 .ex
-#HELPFILE:file://localhost/PATH_TO/lynx_help/lynx_help_main.html
+#HELPFILE:http://www.trill-home.com/lynx/lynx_help/lynx_help_main.html
 
 
 .h2 DEFAULT_INDEX_FILE
@@ -494,7 +495,7 @@
 # Older browser versions will fail so keep NCR_IN_BOOKMARKS:FALSE if you
 # plan to use them.
 #
-#NCR_IN_BOOKMARKS:FALSE
+NCR_IN_BOOKMARKS:TRUE
 
 
 .h2 FORCE_8BIT_TOUPPER
@@ -2183,7 +2184,7 @@
 # viewer or don't want to disable the built-in default viewer
 # mappings for image types.
 #
-#XLOADIMAGE_COMMAND:xv %s
+XLOADIMAGE_COMMAND:xli %s
 
 # Unix:
 # =====
@@ -2250,11 +2251,11 @@
 #
 .ex 7
 #VIEWER:application/postscript:ghostview %s&:XWINDOWS
-#VIEWER:image/gif:xli %s&:XWINDOWS
-#VIEWER:image/x-xbm:xli %s&:XWINDOWS
-#VIEWER:image/png:xli %s&:XWINDOWS
-#VIEWER:image/tiff:xli %s&:XWINDOWS
-#VIEWER:image/jpeg:xli %s&:XWINDOWS
+VIEWER:image/gif:xli %s&:XWINDOWS
+VIEWER:image/x-xbm:xli %s&:XWINDOWS
+VIEWER:image/png:xli %s&:XWINDOWS
+VIEWER:image/tiff:xli %s&:XWINDOWS
+VIEWER:image/jpeg:xli %s&:XWINDOWS
 #VIEWER:video/mpeg:mpeg_play %s &:XWINDOWS
 
 
@@ -3270,3 +3271,6 @@
 #UUDECODE_PATH:
 #ZCAT_PATH:
 #ZIP_PATH:
+
+# included last to allow overriding configuration above
+INCLUDE:/etc/lynx-site.cfg
