--- nfs-utils-1.0.7/utils/svcgssd/svcgssd_proc.c.orig	2004-11-22 00:41:13.000000000 -0500
+++ nfs-utils-1.0.7/utils/svcgssd/svcgssd_proc.c	2005-05-19 08:32:56.000000000 -0400
@@ -243,7 +243,7 @@ handle_nullreq(FILE *f) {
 	/* XXX initialize to a random integer to reduce chances of unnecessary
 	 * invalidation of existing ctx's on restarting svcgssd. */
 	static u_int32_t	handle_seq = 0;
-	char			in_tok_buf[1023];
+	char			in_tok_buf[2047];
 	char			in_handle_buf[15];
 	char			out_handle_buf[15];
 	gss_buffer_desc		in_tok = {.value = in_tok_buf},
