--- xc/programs/Xserver/hw/xfree86/drivers/savage-1.1.26t/savage_accel.c	2002-10-07 15:17:32.000000000 -0400
+++ xc/programs/Xserver/hw/xfree86/drivers/savage/savage_accel.c	2002-11-08 13:03:32.000000000 -0500
@@ -417,7 +417,7 @@
 #if 1
     xaaptr->SetupForScreenToScreenCopy = SavageSetupForScreenToScreenCopy;
     xaaptr->SubsequentScreenToScreenCopy = SavageSubsequentScreenToScreenCopy;
-    xaaptr->ScreenToScreenCopyFlags = NO_TRANSPARENCY | ROP_NEEDS_SOURCE;
+    xaaptr->ScreenToScreenCopyFlags = NO_TRANSPARENCY | NO_PLANEMASK | ROP_NEEDS_SOURCE;
 #endif
 
 
@@ -438,6 +438,7 @@
     xaaptr->Mono8x8PatternFillFlags = 0
 	| HARDWARE_PATTERN_PROGRAMMED_BITS 
 	| HARDWARE_PATTERN_SCREEN_ORIGIN
+	| ROP_NEEDS_SOURCE
 	| BIT_ORDER_IN_BYTE_MSBFIRST
 	;
     if( psav->Chipset == S3_SAVAGE4 )
