--- XFree86-4.3/xc/programs/xkbcomp/compat/iso9995.fixxkb	2003-02-21 04:16:34.000000000 +0100
+++ XFree86-4.3/xc/programs/xkbcomp/compat/iso9995	2003-08-04 20:12:35.000000000 +0200
@@ -51,6 +51,12 @@
 	action= LockGroup(group=+1);
     };
 
+    interpret ISO_Group_Lock {
+	useModMapMods= level1;
+	virtualModifier= AltGr;
+	action= LockGroup(group=+2);
+    };
+					
     interpret ISO_Prev_Group {
 	useModMapMods= level1;
 	virtualModifier= AltGr;
--- XFree86-4.3/xc/programs/xkbcomp/symbols/pc/am.fixxkb	2002-11-22 05:03:28.000000000 +0100
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/pc/am	2003-08-04 20:12:35.000000000 +0200
@@ -64,3 +64,112 @@
     key <AB10> {	[ 0x100057c, 0x100054c	]       };
 
 };
+
+xkb_symbols "old" {
+    include "pc/am(basic)"
+	
+    key <BKSL> {	[ 0x1002026,  0x1000587 ]	};
+
+    key <AE01> {	[ 0x1000573,  0x1000543 ]	};
+    key <AE02> {	[ 0x100057b,  0x100054b ]	};
+    key <AE03> {	[ 0x100056a,  0x100053a ]	};
+    key <AE04> {	[ 0x1000571,  0x1000541 ]	};
+    key <AE05> {	[ 0x1000575,  0x1000545 ]	};
+    key <AE06> {	[ 0x100057c,  0x100054c ]	};
+    key <AE07> {	[ 0x1000580,  0x1000550 ]	};
+    key <AE08> {	[ 0x1000581,  0x1000551 ]	};
+    key <AE09> {	[ 0x1000567,  0x1000537 ]	};
+    key <AE10> {	[ 0x1000572,  0x1000542 ]	};
+    key <AE11> {	[ 0x1002013,  0x1002014 ]	};
+    key <AE12> {	[ parenleft,  parenright]	};
+	
+    key <AD01> {    [ 0x1000583,  0x1000553 ]	};
+    key <AD02> {    [ 0x1000562,  0x1000532 ]	};
+    key <AD03> {    [ 0x100057d,  0x100054d ]	};
+    key <AD04> {    [ 0x1000574,  0x1000544 ]	};
+    key <AD05> {    [ 0x1000578,  0x1000548 ]	};
+    key <AD06> {    [ 0x1000582,  0x1000552 ]	};
+    key <AD07> {    [ 0x100056f,  0x100053f ]	};
+    key <AD08> {    [ 0x1000568,  0x1000538 ]	};
+    key <AD09> {    [ 0x1000569,  0x1000539 ]	};
+    key <AD10> {    [ 0x100056e,  0x100053e ]	};
+    key <AD11> {    [ 0x1000585,  0x1000555 ]	};
+    key <AD12> {    [ 0x1000586,  0x1000556 ]	};
+
+    key <AC01> {	[ 0x100057e,  0x100054e ]	};
+    key <AC02> {	[ 0x1000563,  0x1000533 ]	};
+    key <AC03> {	[ 0x1000565,  0x1000535 ]	};
+    key <AC04> {	[ 0x1000561,  0x1000531 ]	};
+    key <AC05> {	[ 0x1000576,  0x1000546 ]	};
+    key <AC06> {	[ 0x100056b,  0x100053b ]	};
+    key <AC07> {	[ 0x100057f,  0x100054f ]	};
+    key <AC08> {	[ 0x1000570,  0x1000540 ]	};
+    key <AC09> {	[ 0x100057a,  0x100054a ]	};
+    key <AC10> {	[ 0x1000589,  0x100058a ]	};
+    key <AC11> {	[ 0x100055b,  0x10002bc ]	};
+
+    key <AB01> {	[ 0x1000564,  0x1000534 ]	};
+    key <AB02> {	[ 0x1000579,  0x1000549 ]	};
+    key <AB03> {	[ 0x1000566,  0x1000536 ]	};
+    key <AB04> {	[ 0x100056c,  0x100053c ]	};
+    key <AB05> {	[ 0x1000584,  0x1000554 ]	};
+    key <AB06> {	[ 0x100056d,  0x100053d ]	};
+    key <AB07> {	[ 0x1000577,  0x1000547 ]	};
+    key <AB08> {	[ comma,      guillemotleft ]	};
+    key <AB09> {	[ period,     guillemotright ]	};
+    key <AB10> {	[ 0x100055e,  0x1000530 ]	};
+    
+};
+
+xkb_symbols "phonetic" {
+    include "pc/am(old)"
+	
+    key <BKSL> {	[ 0x1000530,  backslash ]	};
+	
+    key <AE01> {	[ 0x1000567,  0x1000537 ]	};
+    key <AE02> {	[ 0x1000569,  0x1000539 ]	};
+    key <AE03> {	[ 0x1000583,  0x1000553 ]	};
+    key <AE05> {	[ 0x100057b,  0x100054b ]	};
+    key <AE06> {	[ parenright, parenleft ]	};
+    key <AE07> {	[ 0x1000587,  percent   ]	};
+    key <AE08> {	[ 0x100057c,  0x100054c ]	};
+    key <AE09> {	[ 0x1000579,  0x1000549 ]	};
+    key <AE10> {	[ 0x1000573,  0x1000543 ]	};
+    key <AE12> {	[ 0x100056a,  0x100053a ]	};
+
+    key <AD01> {    [ 0x1000584,  0x1000554 ]	};
+    key <AD02> {    [ 0x1000578,  0x1000548 ]	};
+    key <AD03> {    [ 0x1000565,  0x1000535 ]	};
+    key <AD04> {    [ 0x1000580,  0x1000550 ]	};
+    key <AD05> {    [ 0x100057f,  0x100054f ]	};
+    key <AD06> {    [ 0x1000568,  0x1000538 ]	};
+    key <AD07> {    [ 0x1000582,  0x1000552 ]	};
+    key <AD08> {    [ 0x100056b,  0x100053b ]	};
+    key <AD09> {    [ 0x1000585,  0x1000555 ]	};
+    key <AD10> {    [ 0x100057a,  0x100054a ]	};
+    key <AD11> {    [ 0x100056d,  0x100053d ]	};
+    key <AD12> {    [ 0x100056e,  0x100053e ]	};
+
+    key <AC01> {	[ 0x1000561,  0x1000531 ]	};
+    key <AC02> {	[ 0x100057d,  0x100054d ]	};
+    key <AC03> {	[ 0x1000564,  0x1000534 ]	};
+    key <AC04> {	[ 0x1000586,  0x1000556 ]	};
+    key <AC05> {	[ 0x1000563,  0x1000533 ]	};
+    key <AC06> {	[ 0x1000570,  0x1000540 ]	};
+    key <AC07> {	[ 0x1000575,  0x1000545 ]	};
+    key <AC08> {	[ 0x100056f,  0x100053f ]	};
+    key <AC09> {	[ 0x100056c,  0x100053c ]	};
+    key <AC10> {	[ 0x1000589,  0x1002026 ]	};
+
+    key <AB01> {	[ 0x1000566,  0x1000536 ]	};
+    key <AB02> {	[ 0x1000572,  0x1000542 ]	};
+    key <AB03> {	[ 0x1000581,  0x1000551 ]	};
+    key <AB04> {	[ 0x100057e,  0x100054e ]	};
+    key <AB05> {	[ 0x1000562,  0x1000532 ]	};
+    key <AB06> {	[ 0x1000576,  0x1000546 ]	};
+    key <AB07> {	[ 0x1000574,  0x1000544 ]	};
+    key <AB10> {	[ 0x1000577,  0x1000547 ]	};
+
+};
+
+
--- XFree86-4.3/xc/programs/xkbcomp/symbols/pc/be.fixxkb	2002-11-22 05:03:28.000000000 +0100
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/pc/be	2003-08-04 20:12:35.000000000 +0200
@@ -26,6 +26,7 @@
     key <AD01>	{ [         a,          A,           at,  Greek_OMEGA ]	};
     key <AD02>	{ [         z,          Z,      lstroke,      Lstroke ]	};
     key <AD03>	{ [         e,          E,     EuroSign,         cent ]	};
+    key <AD09>	{ [         o,          O,           oe,           OE ]	};
     key <AD11>	{ [dead_circumflex, dead_diaeresis,  bracketleft, dead_abovering ] };
     key <AD12>	{ [    dollar,   asterisk, bracketright,  dead_macron ]	};
 
@@ -35,6 +36,7 @@
     key <TLDE>	{ [twosuperior, threesuperior,  notsign,      notsign ]	};
 
     key <BKSL>	{ [        mu,   sterling,   dead_grave,   dead_breve ]	};
+    key <LSGT>	{ [      less,    greater,    backslash,    brokenbar ]	};
     key <AB01>	{ [         w,          W, guillemotleft,        less ]	};
     key <AB07>	{ [     comma,   question, dead_cedilla,    masculine ]	};
     key <AB08>	{ [ semicolon,     period, horizconnector,   multiply ]	};
--- XFree86-4.3/xc/programs/xkbcomp/symbols/pc/fr.fixxkb	2002-11-22 05:03:28.000000000 +0100
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/pc/fr	2003-08-04 20:12:35.000000000 +0200
@@ -26,6 +26,7 @@
     key <AD01>	{ [         a,          A,           ae,           AE ]	};
     key <AD02>	{ [         z,          Z, guillemotleft,        less ]	};
     key <AD03>	{ [         e,          E,     EuroSign,         cent ]	};
+    key <AD09>	{ [         o,          O,           oe,           OE ]	};
     key <AD11>	{ [dead_circumflex, dead_diaeresis, dead_diaeresis, dead_abovering ] };
     key <AD12>	{ [    dollar,   sterling,     currency,  dead_macron ]	};
 
--- XFree86-4.3/xc/programs/xkbcomp/symbols/pc/us.fixxkb	2003-02-15 23:21:24.000000000 +0100
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/pc/us	2003-08-04 20:12:35.000000000 +0200
@@ -13,7 +13,7 @@
     key <AE02> {	[	  2,	at		]	};
     key <AE03> {	[	  3,	numbersign	]	};
     key <AE04> {	[	  4,	dollar		]	};
-    key <AE05> {	[	  5,	percent		]	};
+    key <AE05> {	[	  5,	percent,     EuroSign ]	};
     key <AE06> {	[	  6,	asciicircum	]	};
     key <AE07> {	[	  7,	ampersand	]	};
     key <AE08> {	[	  8,	asterisk	]	};
--- XFree86-4.3/xc/programs/xkbcomp/symbols/pc/bs.fixxkb	2003-08-04 20:12:35.000000000 +0200
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/pc/bs	2003-08-04 20:12:35.000000000 +0200
@@ -0,0 +1,63 @@
+// Bosnian keyboards
+// Amila Akagic, <bono@linux.org.ba>
+// Zadnja promjena: 07.05.2002.
+// $XFree86: xc/programs/xkbcomp/symbols/bs,v 1.1 2002/12/13 04:18:03 dawes Exp $
+
+default partial alphanumeric_keys 
+xkb_symbols "basic" {
+
+    include "pc/latin(type3)"
+
+    // Alphanumeric section
+    name[Group1]= "Bosnian";
+    
+    key <TLDE> { [ dead_cedilla, dead_diaeresis	] };
+    key <AE01> { [        1,	exclam,	    asciitilde,    dead_tilde	] };
+    key <AE02> { [	  2,	quotedbl,   dead_caron,    caron	] };
+    key <AE03> { [	  3,    numbersign, asciicircum,   dead_circumflex ] };
+    key <AE04> { [	  4,	dollar,     dead_breve,    breve	] };
+    key <AE05> { [	  5,	percent,    degree,        dead_abovering ] };
+    key <AE06> { [	  6,	ampersand,  dead_ogonek,   ogonek	] };
+    key <AE07> { [	  7,	slash,      quoteleft,     dead_grave	] };
+    key <AE08> { [	  8,	parenleft,  dead_abovedot, abovedot	] };
+    key <AE09> { [	  9,	parenright, dead_acute,    quoteright	] };
+    key <AE10> { [	  0,	equal,      dead_doubleacute, doubleacute ] };
+    key <AE11> { [ apostrophe, 	question,   dead_diaeresis, diaeresis	] };
+    key <AE12> { [ plus,	asterisk,   dead_cedilla,  cedilla	] };
+
+    key <AD01> { [	  q,	Q,          backslash			] };
+    key <AD02> { [	  w,	W,          bar				] };
+    key <AD03> { [	  e,	E,          EuroSign			] };
+    key <AD05> { [	  t,	T		] };
+    key <AD06> { [	  z,	Z		] };
+    key <AD07> { [	  u,	U		] };
+    key <AD08> { [	  i,	I		] };
+    key <AD11> { [   scaron,	Scaron,     division	 		] };
+    key <AD12> { [  dstroke,    Dstroke,    multiply,	   dead_macron	] };
+    key <AC04> { [	  f,	F,          bracketleft			] };
+    key <AC05> { [	  g,	G,          bracketright		] };
+    key <AC08> { [	  k,	K,          lstroke			] };
+    key <AC09> { [	  l,	L,          Lstroke			] };
+    key <AC10> { [   ccaron,	Ccaron		] };
+    key <AC11> { [   cacute,	Cacute,     ssharp			] };
+
+    key <LSGT> { [      less,	greater,    bar	        		] };
+    key <AB01> { [	  y,	Y 		] };
+    key <AB04> { [	  v,	V,	    at				] };
+    key <AB05> { [	  b,	B,          braceleft			] };
+    key <AB06> { [	  n,	N,          braceright			] };
+    key <AB07> { [	  m,	M,          section			] };
+    key <AB08> { [     comma,	semicolon	] };
+    key <AB09> { [    period,	colon,      periodcentered		] };
+    key <AB10> { [     minus,	underscore	] };
+    key <BKSL> { [ zcaron,	Zcaron,     currency			] };
+    
+    key <RALT> { type[Group1]="TWO_LEVEL",
+                  [ ISO_Level3_Shift, Multi_key ]   };
+    modifier_map Mod5   { <RALT> };
+
+    // End alphanumeric section
+
+};
+
+
--- XFree86-4.3/xc/programs/xkbcomp/symbols/pc/ta.fixxkb	2003-08-04 20:12:35.000000000 +0200
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/pc/ta	2003-08-04 20:12:35.000000000 +0200
@@ -0,0 +1,224 @@
+default partial alphanumeric_keys 
+xkb_symbols "TSCII" {
+
+// Description	: A Tamil typewrite-style keymap 
+//		  loosely based on TamilNet'99 reommendations 
+// Encoding	: TSCII (http://www.tscii.org)
+// Author	: Thuraiappah Vaseeharan <vasee@ieee.org>
+// Last Modified: Sat Jan  5 17:11:26 CST 2002
+
+     	name[Group1]= "US/ASCII";
+     	name[Group2]= "TSCII";
+
+    key <AE01> { [ 0x10000b7, 0x10000a4 ] }; // aytham
+    key <AE02> { [ 0x1000082, 0x10000a5 ] }; // shri
+    key <AE03> { [ 0x1000083, 0x1000088 ] }; // ja
+    key <AE04> { [ 0x1000084, 0x1000089 ] }; // sha
+    key <AE05> { [ 0x1000085, 0x100008a ] }; // sa
+    key <AE06> { [ 0x1000086, 0x100008b ] }; // ha
+    key <AE07> { [ 0x1000087, 0x100008c ] }; // ksha
+	
+	
+    // Qrow
+    key <AD01> { [ 0x10000bb, 0x100009a ] }; // nja
+    key <AD02> { [ 0x10000c8, 0x10000da ] }; // Ra
+    key <AD03> { [ 0x10000bf, 0x10000d1 ] }; // NNa
+    key <AD04> { [ 0x10000ba, 0x10000cd ] }; // ca
+    key <AD05> { [ 0x10000c5, 0x10000d7 ] }; // va
+    key <AD06> { [ 0x10000c4, 0x10000d6 ] }; // la
+    key <AD07> { [ 0x10000c3, 0x10000d5 ] }; // ra
+    key <AD08> { [ 0x10000a8, 0x10000b3 ] }; // sangili, ai
+    key <AD09> { [ 0x10000ca, 0x10000cb ] }; // di, dI
+    key <AD10> { [ 0x10000a2, 0x10000a3 ] }; // visiri
+    key <AD11> { [ dead_acute, 0x10000a3 ] }; // Ukaaram
+    
+
+    // Arow
+    key <AC01> { [ 0x10000c2, 0x10000d4 ] }; // ya
+    key <AC02> { [ 0x10000c7, 0x10000d9 ] }; // La
+    key <AC03> { [ 0x10000c9, 0x10000db ] }; // na
+    key <AC04> { [ 0x10000b8, 0x10000cc ] }; // ka
+    key <AC05> { [ 0x10000c0, 0x10000d2 ] }; // pa
+    key <AC06> { [ dead_grave,0x10000a1 ] }; // pulli,aravu
+    key <AC07> { [ 0x10000be, 0x10000d0 ] }; // tha
+    key <AC08> { [ 0x10000c1, 0x10000d3 ] }; // ma
+    key <AC09> { [ 0x10000bc, 0x10000ce ] }; // da
+    key <AC10> { [ 0x10000c6, 0x10000d8 ] }; // zha
+    key <AC11> { [ 0x10000b9, 0x1000099 ] }; // nga
+    
+    // Zrow
+    key <AB01> { [ 0x10000bd, 0x10000cf ] }; // Na
+    key <AB02> { [ 0x10000b4, 0x10000b5 ] }; // o, O
+    key <AB03> { [ 0x10000af, 0x10000b0 ] }; // u, U
+    key <AB04> { [ 0x10000b1, 0x10000b2 ] }; // e, E
+    key <AB05> { [ 0x10000a6, 0x10000a7 ] }; // kombus
+    key <AB06> { [ 0x10000b6, 0x10000aa ] }; // au
+    key <AB07> { [ 0x10000ab, 0x10000ac ] };	// a, A
+    key <AB08> { [ 0x10000fe, 0x10000ae ] }; // i, I
+
+    key <RALT>	{ 	
+	symbols[Group1]= [	Mode_switch, Multi_key 	],
+	virtualMods= AltGr
+    };
+	
+    key <LALT>	{ 	
+	symbols[Group1]= [	Mode_switch, Multi_key 	],
+	virtualMods= AltGr
+    };
+
+    // begin modifier mappings
+    modifier_map Shift  { Shift_L };
+    modifier_map Lock   { Caps_Lock };
+    modifier_map Control{ Control_L };
+    modifier_map Mod3	{ Mode_switch };
+};
+
+partial alphanumeric_keys 
+xkb_symbols "TAB" {
+
+// Description: A keymap based on the TamilNet'99 typewriter keyboard 
+// Encoding: TAB (http://www.tamilnet99.org)
+// Author: Thuraiappah Vaseeharan <t_vasee@yahoo.com>
+// Date  : Sun Aug 12 02:23:00 CDT 2001
+
+  name[Group1] = "US/ASCII";
+  name[Group2]= "Tamil";
+
+	// numeral row
+	key <AE01> { [ 0x10000e7, 0x10000a7 ] } ;
+	key <AE02> { [ 0x10000fa, 0x10000a8 ] } ;
+	key <AE03> { [ 0x10000fb ] } ;
+	key <AE04> { [ 0x10000fc ] } ;
+	key <AE05> { [ 0x10000fd ] } ;
+	key <AE06> { [ 0x10000fe ] } ;
+	key <AE07> { [ 0x10000ff ] } ;
+	
+	
+    // q-row
+    key <AD01> { [ 0x10000eb, 0x10000b3 ] };
+    key <AD02> { [ 0x10000f8, 0x10000c1 ] };
+    key <AD03> { [ 0x10000ef, 0x10000b8 ] };
+    key <AD04> { [ 0x10000ea, 0x10000b2 ] };
+    key <AD05> { [ 0x10000f5, 0x10000be ] };
+    key <AD06> { [ 0x10000f4, 0x10000bd ] };
+    key <AD07> { [ 0x10000f3, 0x10000bc ] };
+    key <AD08> { [ 0x10000ac, 0x10000e4 ] };
+    key <AD09> { [ 0x10000ae, 0x10000af ] };
+    key <AD10> { [ 0x10000a4, 0x10000a6 ] };
+    key <AD11> { [ dead_circumflex, 0x10000a6 ] }; // Ukaaram
+    
+
+    // a-row
+    key <AC01> { [ 0x10000f2, 0x10000bb ] };
+    key <AC02> { [ 0x10000f7, 0x10000c0 ] };
+    key <AC03> { [ 0x10000f9, 0x10000c2 ] };
+    key <AC04> { [ 0x10000e8, 0x10000b0 ] };
+    key <AC05> { [ 0x10000f0, 0x10000b9 ] };
+    key <AC06> { [ 0x10000a2, 0x10000a3 ] };
+    key <AC07> { [ 0x10000ee, 0x10000b6 ] };
+    key <AC08> { [ 0x10000f1, 0x10000ba ] };
+    key <AC09> { [ 0x10000ec, 0x10000b4 ] };
+    key <AC10> { [ 0x10000f6, 0x10000bf ] };
+    key <AC11> { [ 0x10000e9, 0x10000b1 ] };
+    
+    // z-row
+    key <AB01> { [ 0x10000ed, 0x10000b5 ] };
+    key <AB02> { [ 0x10000e5, 0x10000e6 ] };
+    key <AB03> { [ 0x10000e0, 0x10000e1 ] };
+    key <AB04> { [ 0x10000e2, 0x10000e3 ] };
+    key <AB05> { [ 0x10000aa, 0x10000ab ] };
+    key <AB06> { [ 0x10000ac, 0x10000a3 ] };
+    key <AB07> { [ 0x10000dc, 0x10000dd ] };
+    key <AB08> { [ 0x10000de, 0x10000df ] };
+
+    key <RALT>	{ 	
+	symbols[Group1]= [	Mode_switch, Multi_key 	],
+	virtualMods= AltGr
+    };
+	
+    key <LALT>	{ 	
+	symbols[Group1]= [	Mode_switch, Multi_key 	],
+	virtualMods= AltGr
+    };
+ 
+    // begin modifier mappings
+    modifier_map Shift  { Shift_L };
+    modifier_map Lock   { Caps_Lock };
+    modifier_map Control{ Control_L };
+    modifier_map Mod3	{ Mode_switch };
+};
+
+partial alphanumeric_keys 
+xkb_symbols "UNI" {
+
+// Description: A keymap based on the TamilNet'99 typewriter keyboard 
+// Encoding: Unicode (http://www.unicode.org)
+// Author: Thuraiappah Vaseeharan <vasee@ieee.org>
+// Date  : Sat Feb  9 21:10:48 CST 2002
+// Mapping:
+  
+  name[Group1] = "US/ASCII";
+  name[Group2]= "Tamil";
+
+	// granthas
+	key <AE01> { [ 0x1000b83 ] } ;
+	key <AE03> { [ 0x1000b9c ] } ;
+	key <AE04> { [ 0x1000bb7 ] } ;
+	key <AE05> { [ 0x1000bb8 ] } ;
+	key <AE06> { [ 0x1000bb9 ] } ;
+	
+    // Qrow
+    key <AD01> { [ 0x1000b9e ] };
+    key <AD02> { [ 0x1000bb1 ] };
+    key <AD03> { [ 0x1000ba8 ] };
+    key <AD04> { [ 0x1000b9a ] };
+    key <AD05> { [ 0x1000bb5 ] };
+    key <AD06> { [ 0x1000bb2 ] };
+    key <AD07> { [ 0x1000bb0 ] };
+    key <AD08> { [ 0x1000bc8, 0x1000b90 ] };
+    key <AD09> { [ 0x1000bca, 0x1000bcb ] };
+    key <AD10> { [ 0x1000bbf, 0x1000bc0 ] };
+    key <AD11> { [ 0x1000bc1, 0x1000bc2 ] };
+    
+
+    // Arow
+    key <AC01> { [ 0x1000baf ] };
+    key <AC02> { [ 0x1000bb3 ] };
+    key <AC03> { [ 0x1000ba9 ] };
+    key <AC04> { [ 0x1000b95 ] };
+    key <AC05> { [ 0x1000baa ] };
+    key <AC06> { [ 0x1000bbe, 0x1000bb4 ] };
+    key <AC07> { [ 0x1000ba4 ] };
+    key <AC08> { [ 0x1000bae ] };
+    key <AC09> { [ 0x1000b9f ] };
+    key <AC10> { [ 0x1000bcd ] };
+    key <AC11> { [ 0x1000b99 ] };
+    
+    // Zrow
+    key <AB01> { [ 0x1000ba3 ]  };
+    key <AB02> { [ 0x1000b92, 0x1000b93 ]  };
+    key <AB03> { [ 0x1000b89, 0x1000b8a ]  };
+    key <AB04> { [ 0x1000b8e, 0x1000b8f ]  };
+    key <AB05> { [ 0x1000bc6, 0x1000bc7 ]  };
+    key <AB06> { [ 0x1000b94, 0x1000bcc ]  };
+    key <AB07> { [ 0x1000b85, 0x1000b86 ]  };
+    key <AB08> { [ 0x1000b87, 0x1000b88 ]  };
+
+    key <RALT>	{ 	
+	symbols[Group1]= [	Mode_switch, Multi_key 	],
+	virtualMods= AltGr
+    };
+	
+    key <LALT>	{ 	
+	symbols[Group1]= [	Mode_switch, Multi_key 	],
+	virtualMods= AltGr
+    };
+
+ 
+    // begin modifier mappings
+    modifier_map Shift  { Shift_L };
+    modifier_map Lock   { Caps_Lock };
+    modifier_map Control{ Control_L };
+    modifier_map Mod3	{ Mode_switch };
+
+};
--- XFree86-4.3/xc/programs/xkbcomp/symbols/pc/Imakefile.fixxkb	2003-01-26 03:38:14.000000000 +0100
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/pc/Imakefile	2003-08-04 20:29:16.000000000 +0200
@@ -5,6 +5,7 @@
      DATAFILES = al am ar \
 		be ben bg br by \
 		cz cz_qwerty \
+		bs ta uz \
 		de dev dk dvorak \
 		ee el en_US es \
 		fi fr fr-latin9 \
--- XFree86-4.3/xc/programs/xkbcomp/symbols/pc/ua.fixxkb	2003-08-04 20:26:53.000000000 +0200
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/pc/ua	2003-08-04 20:27:12.000000000 +0200
@@ -116,7 +116,7 @@
   key <LatE> {   [ Cyrillic_ie,       Cyrillic_IE         ]    };
   key <LatR> {   [ Cyrillic_er,       Cyrillic_ER         ]    };
   key <LatT> {   [ Cyrillic_te,       Cyrillic_TE         ]    };
-  key <LatY> {   [ Cyrillic_i,        Cyrillic_i          ]    };
+  key <LatY> {   [ Cyrillic_i,        Cyrillic_I          ]    };
   key <LatU> {   [ Cyrillic_u,        Cyrillic_U          ]    };
   key <LatI> {   [ Ukrainian_i,       Ukrainian_I         ]    };
   key <LatO> {   [ Cyrillic_o,        Cyrillic_O          ]    };
--- XFree86-4.3/xc/programs/xkbcomp/symbols/pc/ben.fixxkb	2003-08-04 20:27:47.000000000 +0200
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/pc/ben	2003-08-04 20:28:28.000000000 +0200
@@ -56,7 +56,7 @@
       key <AC11> { [      0x100099F, 0x10009A0 	]	};
       key <BKSL> { [      backslash, bar 	]	};
 
-      key <AB01> { [      z, Z  		]       };
+      key <AB01> { [      0x10009F1,  0x10009F0	]       };
       key <AB02> { [      0x1000982,  0x1000981 ]       };
       key <AB03> { [      0x10009AE,  0x10009A3 ]       };
       key <AB04> { [      0x10009A8,  0x10009A8 ]       };
--- XFree86-4.3/xc/programs/xkbcomp/symbols/pc/uz.fixxkb	2003-08-04 20:27:54.000000000 +0200
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/pc/uz	2003-08-04 20:32:58.000000000 +0200
@@ -0,0 +1,82 @@
+// Uzbek standard keyboard
+//
+// Mashrab Kuvatov <kmashrab@uni-bremen.de>
+//
+// History:
+//
+// Jun 2, 2003. Mashrab Kuvatov <kmashrab@uni-bremen.de>
+//
+// Created based on Russian standard keyboard. AEN <aen@logic.ru> 
+// The following symbols of Russian standard keyboard were replaced:
+// minus (key <AE11>) with Cyrillic_shortu (0x0100045e),
+// equal (key <AE12>) with Cyrillic_kadescender (0x0100049b)
+// Cyrillic_ha (key <AD11>) with Cyrillic_ghestoke (0x01000493)
+// Cyrillic_hardsign (key <AD11>) with Cyrillic_hadescender (0x01000493)
+// The following symbols of Russian standard keyboard were moved:
+// Cyrillic_ze       to key <AC02> (english 's'), 
+// Cyrillic_sha      to key <AD08> (english 'i'),
+// Cyrillic_ha       to key <AD09> (english 'o'),
+// Cyrillic_hardsign to key <AD10> (english 'p'). 
+
+
+partial default alphanumeric_keys 
+xkb_symbols "basic" {
+
+    name[Group1]= "Uzbek";
+
+    key	<AE01> {	[		1,	    exclam 	]	};
+    key	<AE02> {	[		2,        quotedbl	]	};
+    key	<AE03> {	[		3,      numbersign	]	};
+    key	<AE04> {	[		4,        asterisk	]	};
+    key	<AE05> {	[		5,	     colon	]	};
+    key	<AE06> {	[		6,	     comma	]	};
+    key	<AE07> {	[		7,	    period	]	};
+    key	<AE08> {	[		8,	 semicolon	]	};
+    key	<AE09> {	[		9,	 parenleft	]	};
+    key	<AE10> {	[		0,	parenright	]	};
+    key	<BKSL> {	[	backslash,             bar	]	};
+    key	<AB10> {	[	    slash,	  question	]	};
+    key <LSGT> {        [           slash,             bar      ]       };
+
+key.type[group1]="ALPHABETIC";
+
+    key	<AE11> {	[      0x0100045e,      0x0100040e      ]	};
+    key	<AE12> {	[      0x0100049b,      0x0100049a      ]	};
+    key	<TLDE> {	[     Cyrillic_io,     Cyrillic_IO	]	};
+    key	<AD01> {	[ Cyrillic_shorti, Cyrillic_SHORTI	]	};
+    key	<AD02> {	[    Cyrillic_tse,    Cyrillic_TSE	]	};
+    key	<AD03> {	[      Cyrillic_u,	Cyrillic_U	]	};
+    key	<AD04> {	[     Cyrillic_ka,     Cyrillic_KA	]	};
+    key	<AD05> {	[     Cyrillic_ie,     Cyrillic_IE	]	};
+    key	<AD06> {	[     Cyrillic_en,     Cyrillic_EN	]	};
+    key	<AD07> {	[    Cyrillic_ghe,    Cyrillic_GHE	]	};
+    key	<AD08> {	[    Cyrillic_sha,    Cyrillic_SHA	]	};
+    key	<AD09> {	[     Cyrillic_ha,     Cyrillic_HA	]	};
+    key	<AD10> {	[Cyrillic_hardsign,Cyrillic_HARDSIGN    ]	};
+    key	<AD11> {	[      0x01000493,      0x01000492      ]	};
+    key	<AD12> {	[      0x010004b3,      0x010004b2]	};
+    key	<AC01> {	[     Cyrillic_ef,     Cyrillic_EF	]	};
+    key	<AC02> {	[     Cyrillic_ze,     Cyrillic_ZE	]	};
+    key	<AC03> {	[     Cyrillic_ve,     Cyrillic_VE	]	};
+    key	<AC04> {	[      Cyrillic_a,	Cyrillic_A	]	};
+    key	<AC05> {	[     Cyrillic_pe,     Cyrillic_PE	]	};
+    key	<AC06> {	[     Cyrillic_er,     Cyrillic_ER	]	};
+    key	<AC07> {	[      Cyrillic_o,	Cyrillic_O	]	};
+    key	<AC08> {	[     Cyrillic_el,     Cyrillic_EL	]	};
+    key	<AC09> {	[     Cyrillic_de,     Cyrillic_DE	]	};
+    key	<AC10> {	[    Cyrillic_zhe,    Cyrillic_ZHE	]	};
+    key	<AC11> {	[      Cyrillic_e,	Cyrillic_E	]	};
+    key	<AB01> {	[     Cyrillic_ya,     Cyrillic_YA	]	};
+    key	<AB02> {	[    Cyrillic_che,    Cyrillic_CHE	]	};
+    key	<AB03> {	[     Cyrillic_es,     Cyrillic_ES	]	};
+    key	<AB04> {	[     Cyrillic_em,     Cyrillic_EM	]	};
+    key	<AB06> {	[     Cyrillic_te,     Cyrillic_TE	]	};
+    key	<AB05> {	[      Cyrillic_i,	Cyrillic_I	]	};
+    key	<AB07> {	[Cyrillic_softsign,Cyrillic_SOFTSIGN	]	};
+    key	<AB08> {	[     Cyrillic_be,     Cyrillic_BE	]	};
+    key	<AB09> {	[     Cyrillic_yu,     Cyrillic_YU	]	};
+
+key.type[group2]="TWO_LEVEL";
+
+    // End alphanumeric section
+};
--- XFree86-4.3/xc/programs/xkbcomp/symbols/pc/mng.fixxkb	2003-08-04 20:30:34.000000000 +0200
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/pc/mng	2003-08-04 20:34:35.000000000 +0200
@@ -0,0 +1,70 @@
+// Mongolian (cyrillic)
+
+partial default alphanumeric_keys 
+xkb_symbols "basic" {
+
+    name[Group1]= "Cyrillic";
+
+    key <AE01> { [ exclam,       1          	]	};
+    key <AE02> { [ minus,        2          	]	};
+    key <AE03> { [ quotedbl,     3          	]	};
+    key <AE04> { [ Cyrillic_ie,  4          	]	};
+    key <AE05> { [ colon,        5          	]	};
+    key <AE06> { [ period,       6          	]	};
+    key <AE07> { [ question,     7          	]	};
+    key <AE08> { [ comma,        8          	]	};
+    key <AE09> { [ parenleft,    9          	]	};
+    key <AE10> { [ parenright,   0          	]	};
+    
+key.type[group1]="ALPHABETIC";
+
+    key <AE11> { [ Cyrillic_ie,  Cyrillic_IE	]	};
+    key <AE12> { [ Cyrillic_shcha,Cyrillic_SHCHA]	};
+
+    key <TLDE> { [ 0x1002116,    percent		]	};
+
+	key <AD01> { [ Cyrillic_ef,  Cyrillic_EF	]	};
+    key <AD02> { [ Cyrillic_tse, Cyrillic_TSE	]	};
+    key <AD03> { [ Cyrillic_u,   Cyrillic_U    ]	};
+    key <AD04> { [ Cyrillic_zhe, Cyrillic_ZHE  ]	};
+    key <AD05> { [ Cyrillic_e,   Cyrillic_E    ]	};
+    key <AD06> { [ Cyrillic_en,  Cyrillic_EN   ]	};
+    key <AD07> { [ Cyrillic_ghe, Cyrillic_GHE  ]	};
+    key <AD08> { [ Cyrillic_sha, Cyrillic_SHA  ]	};
+    key <AD09> { [ 0x10004af,    0x10004ae     ]	};
+    key <AD10> { [ Cyrillic_ze,  Cyrillic_ZE   ]	};
+    key <AD11> { [ Cyrillic_ka,  Cyrillic_KA   ]	};
+    key <AD12> { [ Cyrillic_hardsign,Cyrillic_HARDSIGN	]	};
+
+    key <BKSL> { [ backslash,     slash	]	};
+
+    key <AC01> { [ Cyrillic_shorti,Cyrillic_SHORTI	]	};
+    key <AC02> { [ Cyrillic_yeru,Cyrillic_YERU	]	};
+    key <AC03> { [ Cyrillic_be,  Cyrillic_BE   ]	};
+    key <AC04> { [ 0x10004e9,    0x10004e8     ]	};
+    key <AC05> { [ Cyrillic_a,   Cyrillic_A    ]	};
+    key <AC06> { [ Cyrillic_ha,  Cyrillic_HA   ]	};
+    key <AC07> { [ Cyrillic_er,  Cyrillic_ER   ]	};
+    key <AC08> { [ Cyrillic_o,   Cyrillic_O    ]	};
+    key <AC09> { [ Cyrillic_el,  Cyrillic_EL   ]	};
+    key <AC10> { [ Cyrillic_de,  Cyrillic_DE   ]	};
+    key <AC11> { [ Cyrillic_pe,  Cyrillic_PE   ]	};
+
+    key <AB01> { [ Cyrillic_ya,  Cyrillic_YA   ]	};
+    key <AB02> { [ Cyrillic_che, Cyrillic_CHE  ]	};
+    key <AB03> { [ Cyrillic_io,  Cyrillic_IO   ]	};
+    key <AB04> { [ Cyrillic_es,  Cyrillic_ES   ]	};
+    key <AB05> { [ Cyrillic_em,  Cyrillic_EM   ]	};
+    key <AB06> { [ Cyrillic_i,   Cyrillic_I    ]	};
+    key <AB07> { [ Cyrillic_te,  Cyrillic_TE   ]	};
+    key <AB08> { [ Cyrillic_softsign,Cyrillic_SOFTSIGN	]	};
+    key <AB09> { [ Cyrillic_ve,  Cyrillic_VE   ]	};
+    key <AB10> { [ Cyrillic_yu,  Cyrillic_YU   ]	};
+	
+    key <LSGT> { [ less,         greater    	]	};
+
+key.type[group2]="TWO_LEVEL";
+	
+};
+
+
--- XFree86-4.3/xc/programs/xkbcomp/symbols/Imakefile.fixxkb	2003-02-15 23:22:44.000000000 +0100
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/Imakefile	2003-08-04 20:12:35.000000000 +0200
@@ -26,7 +26,8 @@
 		 nl no \
 		 ogham ori \
 		 pc104 pl pl2 pt \
-		 ralt ro ru ru_yawerty \
+ 		 ralt ro ro2 ru ru_yawerty \
+ 		 lao mng ta lt_a lt_p \
 		 sapmi se se_FI se_NO se_SE si sk sk_qwerty sr srvr_ctrl \
 		 syr syr_phonetic \
 		 tel th tj tml tr tr_f \
--- XFree86-4.3/xc/programs/xkbcomp/symbols/us.fixxkb	2002-11-20 05:49:03.000000000 +0100
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/us	2003-08-04 20:12:35.000000000 +0200
@@ -20,7 +20,7 @@
     key <AE02> {	[	  2,	at		]	};
     key <AE03> {	[	  3,	numbersign	]	};
     key <AE04> {	[	  4,	dollar		]	};
-    key <AE05> {	[	  5,	percent		]	};
+    key <AE05> {	[	  5,	percent		], [ EuroSign ]	};
     key <AE06> {	[	  6,	asciicircum	]	};
     key <AE07> {	[	  7,	ampersand	]	};
     key <AE08> {	[	  8,	asterisk	]	};
--- XFree86-4.3/xc/programs/xkbcomp/symbols/us_intl.fixxkb	2002-04-04 16:05:59.000000000 +0200
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/us_intl	2003-08-04 20:40:41.000000000 +0200
@@ -36,11 +36,10 @@
     // Alphanumeric section
     key <TLDE> {	[ dead_grave,	dead_tilde	],
 			[      grave,	asciitilde	]	};
+    key <AE05> {	[	  5,	percent		],
+ 			[ EuroSign					]	};
     key <AE06> {	[	  6,	dead_circumflex	],
 			[ asciicircum,	asciicircum	]	};
-    key <AC11> {	[ dead_acute,	dead_diaeresis	],
-			[ apostrophe,	quotedbl	]	};
-			
     key <AE09> {	[	  9,	parenleft	],
 			[      dead_breve,	dead_breve	]	};
     key <AE10> {	[	  0,	parenright	],
@@ -49,8 +48,12 @@
 			[      dead_macron,	dead_belowdot	]	};
     key <AE12> {	[     equal,	plus		],
 			[      dead_doubleacute,	dead_horn	]	};
+    key <AC02> {	[	  s,	S		],
+			[	  ssharp,	section		]	};
     key <AC10> {	[ semicolon,	colon		],
 			[      dead_ogonek,	dead_diaeresis	]	};
+    key <AC11> {	[ dead_acute,	dead_diaeresis	],
+			[ apostrophe,	quotedbl	]	};
     key <AB08> {	[     comma,	less		],
 			[      dead_cedilla,	dead_caron	]	};
     key <AB09> {	[    period,	greater		],
@@ -58,7 +61,7 @@
     key <AB10> {	[     slash,	question	],
 			[      dead_hook,	dead_hook	]	};
 
-
+    replace key <RWIN> { [ Multi_key ] };
     // End alphanumeric section
 };
 
--- XFree86-4.3/xc/programs/xkbcomp/symbols/lao.fixxkb	2003-08-04 20:12:35.000000000 +0200
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/lao	2003-08-04 20:12:35.000000000 +0200
@@ -0,0 +1,77 @@
+// $XConsortium: lo 2002/02/28 $
+//
+// $XFree86$
+//
+// Lao keymap, by Anousak Souphavanh <laonux@laonux.muanglao.com>
+//
+//
+//
+
+partial default alphanumeric_keys
+xkb_symbols "basic" {
+
+    // Describes the differences between a very simple US/ASCII
+    // keyboard and a Lao keyboard
+    // This keymap uses the Lao keyboard standard [kind of] widely used in Laos
+
+    name[Group2]= "Lao";
+
+    // Alphanumeric section
+    key <TLDE> {         [], [ 0x100200D        ]};
+    key <AE01> {         [], [ 0x1000ea2,  0x1000ed1  ]};
+    key <AE02> {         [], [ 0x1000e9f,       0x1000ed2   ]};
+    key <AE03> {         [], [ 0x1000ec2,       0x1000ed3   ]};
+    key <AE04> {         [], [ 0x1000e96,       0x1000ed4   ]};
+    key <AE05> {         [], [ 0x1000eb8,       0x1000ecc   ]};
+    key <AE06> {         [], [ 0x1000eb9,       0x1000ebc   ]};
+    key <AE07> {         [], [ 0x1000eb4,       0x1000ed5   ]};
+    key <AE08> {         [], [ 0x1000e95,       0x1000ed6   ]};
+    key <AE09> {         [], [ 0x1000e88,       0x1000ed7   ]};
+    key <AE10> {         [], [ 0x1000e82,       0x1000ed8   ]};
+    key <AE11> {         [], [ 0x1000e8a,       0x1000ed9   ]};
+    key <AE12> {         [], [ 0x1000ecd        ]};
+
+    key <AD01> {         [], [ 0x1000ebb        ]};
+    key <AD02> {         [], [ 0x1000ec4,       0x1000ed0   ]};
+    key <AD03> {         [], [ 0x1000eb3        ]};
+    key <AD04> {         [], [ 0x1000e9e,       underscore  ]};
+    key <AD05> {         [], [ 0x1000eb0,       plus  ]};
+    key <AD06> {         [], [ 0x1000eb4  ]};
+    key <AD07> {         [], [ 0x1000eb5  ]};
+    key <AD08> {         [], [0x1000ea3,  0x1000eae   ]};
+    key <AD09> {         [], [ 0x1000e99        ]};
+    key <AD10> {         [], [ 0x1000e8d,       0x1000eb9   ]};
+    key <AD11> {         [], [ 0x1000e9a,  minus            ]};
+    key <AD12> {         [], [ 0x1000ea5,       braceright  ]};
+
+    key <AC01> {         [], [ 0x1000eb1        ]};
+    key <AC02> {         [], [ 0x1000eab,       semicolon   ]};
+    key <AC03> {         [], [ 0x1000e81,       period      ]};
+    key <AC04> {         [], [ 0x1000e94,       comma ]};
+    key <AC05> {         [], [ 0x1000ec0,       colon ]};
+    key <AC06> {         [], [ 0x1000ec9,       0x1000eca   ]};
+    key <AC07> {         [], [ 0x1000ec8,       0x1000ecb   ]};
+    key <AC08> {         [], [ 0x1000eb2,       exclam      ]};
+    key <AC09> {         [], [ 0x1000eaa,       question    ]};
+    key <AC10> {         [], [ 0x1000ea7,       percent     ]};
+    key <AC11> {         [], [ 0x1000e87,       equal ]};
+
+    key <AB01> {         [], [ 0x1000e9c,       0x10020ad   ]};
+    key <AB02> {         [], [ 0x1000e9b,       parenleft   ]};
+    key <AB03> {         [], [ 0x1000ec1,       0x1000eaf   ]};
+    key <AB04> {         [], [ 0x1000ead        ]};
+    key <AB05> {         [], [ 0x1000eb6        ]};
+    key <AB06> {         [], [ 0x1000eb7        ]};
+    key <AB07> {         [], [ 0x1000e97,       0x1000ec6   ]};
+    key <AB08> {         [], [ 0x1000ea1, quoteleft   ]};
+    key <AB09> {         [], [ 0x1000ec3,       dollar      ]};
+    key <AB10> {         [], [ 0x1000e9d,       parenright  ]};
+
+    // Begin modifier mappings
+    modifier_map Shift                    { Shift_L };
+    modifier_map Lock                     { Caps_Lock, ISO_Lock };
+    modifier_map Control                  { Control_L };
+    modifier_map Mod3                     { Mode_switch };
+};
+
+
--- XFree86-4.3/xc/programs/xkbcomp/symbols/mng.fixxkb	2003-08-04 20:12:35.000000000 +0200
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/mng	2003-08-04 20:12:35.000000000 +0200
@@ -0,0 +1,75 @@
+// Mongolian (cyrillic)
+
+partial default alphanumeric_keys 
+xkb_symbols "basic" {
+
+    // Describes the differences between a very simple en_US
+    // keyboard and a very simple Mongolian cyrillic keyboard
+
+    name[Group1]= "US/ASCII";
+    name[Group2]= "Cyrillic";
+
+    key <TLDE> { [ ], [ 0x1002116,    percent		]	};
+
+    key <AE01> { [ ], [ exclam,       1          	]	};
+    key <AE02> { [ ], [ minus,        2          	]	};
+    key <AE03> { [ ], [ quotedbl,     3          	]	};
+    key <AE04> { [ ], [ Cyrillic_ie,  4          	]	};
+    key <AE05> { [ ], [ colon,        5          	]	};
+    key <AE06> { [ ], [ period,       6          	]	};
+    key <AE07> { [ ], [ question,     7          	]	};
+    key <AE08> { [ ], [ comma,        8          	]	};
+    key <AE09> { [ ], [ parenleft,    9          	]	};
+    key <AE10> { [ ], [ parenright,   0          	]	};
+    key <AE11> { [ ], [ Cyrillic_ie,  Cyrillic_IE	]	};
+    key <AE12> { [ ], [ Cyrillic_shcha,Cyrillic_SHCHA]	};
+    
+	key <AD01> { [ ], [ Cyrillic_ef,  Cyrillic_EF	]	};
+    key <AD02> { [ ], [ Cyrillic_tse, Cyrillic_TSE	]	};
+    key <AD03> { [ ], [ Cyrillic_u,   Cyrillic_U    ]	};
+    key <AD04> { [ ], [ Cyrillic_zhe, Cyrillic_ZHE  ]	};
+    key <AD05> { [ ], [ Cyrillic_e,   Cyrillic_E    ]	};
+    key <AD06> { [ ], [ Cyrillic_en,  Cyrillic_EN   ]	};
+    key <AD07> { [ ], [ Cyrillic_ghe, Cyrillic_GHE  ]	};
+    key <AD08> { [ ], [ Cyrillic_sha, Cyrillic_SHA  ]	};
+    key <AD09> { [ ], [ 0x10004af,    0x10004ae     ]	};
+    key <AD10> { [ ], [ Cyrillic_ze,  Cyrillic_ZE   ]	};
+    key <AD11> { [ ], [ Cyrillic_ka,  Cyrillic_KA   ]	};
+    key <AD12> { [ ], [ Cyrillic_hardsign,Cyrillic_HARDSIGN	]	};
+
+    key <BKSL> { [ ], [ backslash,     slash	]	};
+
+    key <AC01> { [ ], [ Cyrillic_shorti,Cyrillic_SHORTI	]	};
+    key <AC02> { [ ], [ Cyrillic_yeru,Cyrillic_YERU	]	};
+    key <AC03> { [ ], [ Cyrillic_be,  Cyrillic_BE   ]	};
+    key <AC04> { [ ], [ 0x10004e9,    0x10004e8     ]	};
+    key <AC05> { [ ], [ Cyrillic_a,   Cyrillic_A    ]	};
+    key <AC06> { [ ], [ Cyrillic_ha,  Cyrillic_HA   ]	};
+    key <AC07> { [ ], [ Cyrillic_er,  Cyrillic_ER   ]	};
+    key <AC08> { [ ], [ Cyrillic_o,   Cyrillic_O    ]	};
+    key <AC09> { [ ], [ Cyrillic_el,  Cyrillic_EL   ]	};
+    key <AC10> { [ ], [ Cyrillic_de,  Cyrillic_DE   ]	};
+    key <AC11> { [ ], [ Cyrillic_pe,  Cyrillic_PE   ]	};
+
+    key <AB01> { [ ], [ Cyrillic_ya,  Cyrillic_YA   ]	};
+    key <AB02> { [ ], [ Cyrillic_che, Cyrillic_CHE  ]	};
+    key <AB03> { [ ], [ Cyrillic_io,  Cyrillic_IO   ]	};
+    key <AB04> { [ ], [ Cyrillic_es,  Cyrillic_ES   ]	};
+    key <AB05> { [ ], [ Cyrillic_em,  Cyrillic_EM   ]	};
+    key <AB06> { [ ], [ Cyrillic_i,   Cyrillic_I    ]	};
+    key <AB07> { [ ], [ Cyrillic_te,  Cyrillic_TE   ]	};
+    key <AB08> { [ ], [ Cyrillic_softsign,Cyrillic_SOFTSIGN	]	};
+    key <AB09> { [ ], [ Cyrillic_ve,  Cyrillic_VE   ]	};
+    key <AB10> { [ ], [ Cyrillic_yu,  Cyrillic_YU   ]	};
+	
+    key <LSGT> { [ ], [ less,         greater    	]	};
+	
+    // Begin modifier mappings 
+
+    modifier_map Shift  { Shift_L };
+    modifier_map Lock   { Caps_Lock, ISO_Lock };
+    modifier_map Control{ Control_L };
+    modifier_map Mod3   { Mode_switch };
+};
+
+
--- XFree86-4.3/xc/programs/xkbcomp/symbols/ta.fixxkb	2003-08-04 20:12:35.000000000 +0200
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/ta	2003-08-04 20:12:35.000000000 +0200
@@ -0,0 +1,224 @@
+default partial alphanumeric_keys 
+xkb_symbols "TSCII" {
+
+// Description	: A Tamil typewrite-style keymap 
+//		  loosely based on TamilNet'99 reommendations 
+// Encoding	: TSCII (http://www.tscii.org)
+// Author	: Thuraiappah Vaseeharan <vasee@ieee.org>
+// Last Modified: Sat Jan  5 17:11:26 CST 2002
+
+     	name[Group1]= "US/ASCII";
+     	name[Group2]= "TSCII";
+
+    key <AE01> { [], [ 0x10000b7, 0x10000a4 ] }; // aytham
+    key <AE02> { [], [ 0x1000082, 0x10000a5 ] }; // shri
+    key <AE03> { [], [ 0x1000083, 0x1000088 ] }; // ja
+    key <AE04> { [], [ 0x1000084, 0x1000089 ] }; // sha
+    key <AE05> { [], [ 0x1000085, 0x100008a ] }; // sa
+    key <AE06> { [], [ 0x1000086, 0x100008b ] }; // ha
+    key <AE07> { [], [ 0x1000087, 0x100008c ] }; // ksha
+	
+	
+    // Qrow
+    key <AD01> { [], [ 0x10000bb, 0x100009a ] }; // nja
+    key <AD02> { [], [ 0x10000c8, 0x10000da ] }; // Ra
+    key <AD03> { [], [ 0x10000bf, 0x10000d1 ] }; // NNa
+    key <AD04> { [], [ 0x10000ba, 0x10000cd ] }; // ca
+    key <AD05> { [], [ 0x10000c5, 0x10000d7 ] }; // va
+    key <AD06> { [], [ 0x10000c4, 0x10000d6 ] }; // la
+    key <AD07> { [], [ 0x10000c3, 0x10000d5 ] }; // ra
+    key <AD08> { [], [ 0x10000a8, 0x10000b3 ] }; // sangili, ai
+    key <AD09> { [], [ 0x10000ca, 0x10000cb ] }; // di, dI
+    key <AD10> { [], [ 0x10000a2, 0x10000a3 ] }; // visiri
+    key <AD11> { [], [ dead_acute, 0x10000a3 ] }; // Ukaaram
+    
+
+    // Arow
+    key <AC01> { [],[ 0x10000c2, 0x10000d4 ] }; // ya
+    key <AC02> { [],[ 0x10000c7, 0x10000d9 ] }; // La
+    key <AC03> { [],[ 0x10000c9, 0x10000db ] }; // na
+    key <AC04> { [],[ 0x10000b8, 0x10000cc ] }; // ka
+    key <AC05> { [],[ 0x10000c0, 0x10000d2 ] }; // pa
+    key <AC06> { [],[ dead_grave,0x10000a1 ] }; // pulli,aravu
+    key <AC07> { [],[ 0x10000be, 0x10000d0 ] }; // tha
+    key <AC08> { [],[ 0x10000c1, 0x10000d3 ] }; // ma
+    key <AC09> { [],[ 0x10000bc, 0x10000ce ] }; // da
+    key <AC10> { [],[ 0x10000c6, 0x10000d8 ] }; // zha
+    key <AC11> { [],[ 0x10000b9, 0x1000099 ] }; // nga
+    
+    // Zrow
+    key <AB01> { [],[ 0x10000bd, 0x10000cf ] }; // Na
+    key <AB02> { [],[ 0x10000b4, 0x10000b5 ] }; // o, O
+    key <AB03> { [],[ 0x10000af, 0x10000b0 ] }; // u, U
+    key <AB04> { [],[ 0x10000b1, 0x10000b2 ] }; // e, E
+    key <AB05> { [],[ 0x10000a6, 0x10000a7 ] }; // kombus
+    key <AB06> { [],[ 0x10000b6, 0x10000aa ] }; // au
+    key <AB07> { [],[ 0x10000ab, 0x10000ac ] };	// a, A
+    key <AB08> { [],[ 0x10000fe, 0x10000ae ] }; // i, I
+
+    key <RALT>	{ 	
+	symbols[Group1]= [	Mode_switch, Multi_key 	],
+	virtualMods= AltGr
+    };
+	
+    key <LALT>	{ 	
+	symbols[Group1]= [	Mode_switch, Multi_key 	],
+	virtualMods= AltGr
+    };
+
+    // begin modifier mappings
+    modifier_map Shift  { Shift_L };
+    modifier_map Lock   { Caps_Lock };
+    modifier_map Control{ Control_L };
+    modifier_map Mod3	{ Mode_switch };
+};
+
+partial alphanumeric_keys 
+xkb_symbols "TAB" {
+
+// Description: A keymap based on the TamilNet'99 typewriter keyboard 
+// Encoding: TAB (http://www.tamilnet99.org)
+// Author: Thuraiappah Vaseeharan <t_vasee@yahoo.com>
+// Date  : Sun Aug 12 02:23:00 CDT 2001
+
+  name[Group1] = "US/ASCII";
+  name[Group2]= "Tamil";
+
+	// numeral row
+	key <AE01> { [], [ 0x10000e7, 0x10000a7 ] } ;
+	key <AE02> { [], [ 0x10000fa, 0x10000a8 ] } ;
+	key <AE03> { [], [ 0x10000fb ] } ;
+	key <AE04> { [], [ 0x10000fc ] } ;
+	key <AE05> { [], [ 0x10000fd ] } ;
+	key <AE06> { [], [ 0x10000fe ] } ;
+	key <AE07> { [], [ 0x10000ff ] } ;
+	
+	
+    // q-row
+    key <AD01> { [], [ 0x10000eb, 0x10000b3 ] };
+    key <AD02> { [], [ 0x10000f8, 0x10000c1 ] };
+    key <AD03> { [], [ 0x10000ef, 0x10000b8 ] };
+    key <AD04> { [], [ 0x10000ea, 0x10000b2 ] };
+    key <AD05> { [], [ 0x10000f5, 0x10000be ] };
+    key <AD06> { [], [ 0x10000f4, 0x10000bd ] };
+    key <AD07> { [], [ 0x10000f3, 0x10000bc ] };
+    key <AD08> { [], [ 0x10000ac, 0x10000e4 ] };
+    key <AD09> { [], [ 0x10000ae, 0x10000af ] };
+    key <AD10> { [], [ 0x10000a4, 0x10000a6 ] };
+    key <AD11> { [], [ dead_circumflex, 0x10000a6 ] }; // Ukaaram
+    
+
+    // a-row
+    key <AC01> { [], [ 0x10000f2, 0x10000bb ] };
+    key <AC02> { [], [ 0x10000f7, 0x10000c0 ] };
+    key <AC03> { [], [ 0x10000f9, 0x10000c2 ] };
+    key <AC04> { [], [ 0x10000e8, 0x10000b0 ] };
+    key <AC05> { [], [ 0x10000f0, 0x10000b9 ] };
+    key <AC06> { [], [ 0x10000a2, 0x10000a3 ] };
+    key <AC07> { [], [ 0x10000ee, 0x10000b6 ] };
+    key <AC08> { [], [ 0x10000f1, 0x10000ba ] };
+    key <AC09> { [], [ 0x10000ec, 0x10000b4 ] };
+    key <AC10> { [], [ 0x10000f6, 0x10000bf ] };
+    key <AC11> { [], [ 0x10000e9, 0x10000b1 ] };
+    
+    // z-row
+    key <AB01> { [], [ 0x10000ed, 0x10000b5 ] };
+    key <AB02> { [], [ 0x10000e5, 0x10000e6 ] };
+    key <AB03> { [], [ 0x10000e0, 0x10000e1 ] };
+    key <AB04> { [], [ 0x10000e2, 0x10000e3 ] };
+    key <AB05> { [], [ 0x10000aa, 0x10000ab ] };
+    key <AB06> { [], [ 0x10000ac, 0x10000a3 ] };
+    key <AB07> { [], [ 0x10000dc, 0x10000dd ] };
+    key <AB08> { [], [ 0x10000de, 0x10000df ] };
+
+    key <RALT>	{ 	
+	symbols[Group1]= [	Mode_switch, Multi_key 	],
+	virtualMods= AltGr
+    };
+	
+    key <LALT>	{ 	
+	symbols[Group1]= [	Mode_switch, Multi_key 	],
+	virtualMods= AltGr
+    };
+ 
+    // begin modifier mappings
+    modifier_map Shift  { Shift_L };
+    modifier_map Lock   { Caps_Lock };
+    modifier_map Control{ Control_L };
+    modifier_map Mod3	{ Mode_switch };
+};
+
+partial alphanumeric_keys 
+xkb_symbols "UNI" {
+
+// Description: A keymap based on the TamilNet'99 typewriter keyboard 
+// Encoding: Unicode (http://www.unicode.org)
+// Author: Thuraiappah Vaseeharan <vasee@ieee.org>
+// Date  : Sat Feb  9 21:10:48 CST 2002
+// Mapping:
+  
+  name[Group1] = "US/ASCII";
+  name[Group2]= "Tamil";
+
+	// granthas
+	key <AE01> { [], [ 0x1000b83 ] } ;
+	key <AE03> { [], [ 0x1000b9c ] } ;
+	key <AE04> { [], [ 0x1000bb7 ] } ;
+	key <AE05> { [], [ 0x1000bb8 ] } ;
+	key <AE06> { [], [ 0x1000bb9 ] } ;
+	
+    // Qrow
+    key <AD01> { [], [ 0x1000b9e ] };
+    key <AD02> { [], [ 0x1000bb1 ] };
+    key <AD03> { [], [ 0x1000ba8 ] };
+    key <AD04> { [], [ 0x1000b9a ] };
+    key <AD05> { [], [ 0x1000bb5 ] };
+    key <AD06> { [], [ 0x1000bb2 ] };
+    key <AD07> { [], [ 0x1000bb0 ] };
+    key <AD08> { [], [ 0x1000bc8, 0x1000b90 ] };
+    key <AD09> { [], [ 0x1000bca, 0x1000bcb ] };
+    key <AD10> { [], [ 0x1000bbf, 0x1000bc0 ] };
+    key <AD11> { [], [ 0x1000bc1, 0x1000bc2 ] };
+    
+
+    // Arow
+    key <AC01> { [],[ 0x1000baf ] };
+    key <AC02> { [],[ 0x1000bb3 ] };
+    key <AC03> { [],[ 0x1000ba9 ] };
+    key <AC04> { [],[ 0x1000b95 ] };
+    key <AC05> { [],[ 0x1000baa ] };
+    key <AC06> { [],[ 0x1000bbe, 0x1000bb4 ] };
+    key <AC07> { [],[ 0x1000ba4 ] };
+    key <AC08> { [],[ 0x1000bae ] };
+    key <AC09> { [],[ 0x1000b9f ] };
+    key <AC10> { [],[ 0x1000bcd ] };
+    key <AC11> { [],[ 0x1000b99 ] };
+    
+    // Zrow
+    key <AB01> { [],[ 0x1000ba3 ]  };
+    key <AB02> { [],[ 0x1000b92, 0x1000b93 ]  };
+    key <AB03> { [],[ 0x1000b89, 0x1000b8a ]  };
+    key <AB04> { [],[ 0x1000b8e, 0x1000b8f ]  };
+    key <AB05> { [],[ 0x1000bc6, 0x1000bc7 ]  };
+    key <AB06> { [],[ 0x1000b94, 0x1000bcc ]  };
+    key <AB07> { [],[ 0x1000b85, 0x1000b86 ]  };
+    key <AB08> { [],[ 0x1000b87, 0x1000b88 ]  };
+
+    key <RALT>	{ 	
+	symbols[Group1]= [	Mode_switch, Multi_key 	],
+	virtualMods= AltGr
+    };
+	
+    key <LALT>	{ 	
+	symbols[Group1]= [	Mode_switch, Multi_key 	],
+	virtualMods= AltGr
+    };
+
+ 
+    // begin modifier mappings
+    modifier_map Shift  { Shift_L };
+    modifier_map Lock   { Caps_Lock };
+    modifier_map Control{ Control_L };
+    modifier_map Mod3	{ Mode_switch };
+
+};
--- XFree86-4.3/xc/programs/xkbcomp/symbols/fr.fixxkb	2001-01-18 00:45:58.000000000 +0100
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/fr	2003-08-04 20:40:46.000000000 +0200
@@ -43,6 +43,8 @@
 			[   guillemotleft,            less	]	};
     key <AD03> {        [               e,               E      ],
                         [        EuroSign,            cent      ]	};
+    key <AD09> {        [               o,               O      ],
+                        [              oe,              OE      ]	};
     key <AD11> {	[ dead_circumflex,  dead_diaeresis	]	};
     key <AD12> {	[          dollar,        sterling	],
 			[        currency,     dead_macron	]	};
--- XFree86-4.3/xc/programs/xkbcomp/symbols/fr_CH.fixxkb	2001-01-18 00:45:58.000000000 +0100
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/fr_CH	2003-08-04 20:12:35.000000000 +0200
@@ -14,6 +14,10 @@
     include "de_CH(basic)"
     name[Group1]= "Swiss French";
 
+    override key <AD09> {
+	[ o,	O	],
+	[ oe,	OE	]
+    };
     override key <AD11> {
 	[ egrave,	udiaeresis	],
 	[ bracketleft			]
--- XFree86-4.3/xc/programs/xkbcomp/symbols/ca.fixxkb	2001-05-19 01:35:35.000000000 +0200
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/ca	2003-08-04 20:12:35.000000000 +0200
@@ -58,10 +58,8 @@
 		symbols[Group1]= [ equal, plus, notsign			],
 		symbols[Group2]= [ dead_cedilla, dead_ogonek		]
     };
-// once again, CAN/CSA-Z243-200.0 defines a key for which there is no
-// symbol.
-//    key <AD03> {	[	  e,	E		],
-//			[	 oe,	OE		]		};
+    key <AD03> {	[	  e,	E		],
+			[	 oe,	OE		]		};
     key <AD11> {
 		type[Group1]="THREE_LEVEL",
 		symbols[Group1]=[ dead_circumflex, dead_diaeresis, dead_grave ],
--- XFree86-4.3/xc/programs/xkbcomp/symbols/be.fixxkb	2001-01-18 00:45:57.000000000 +0100
+++ XFree86-4.3/xc/programs/xkbcomp/symbols/be	2003-08-04 20:12:35.000000000 +0200
@@ -36,6 +36,8 @@
     key <AD02> {	[               z,               Z 	]	};
     key <AD03> {        [               e,               E      ],
                         [        EuroSign,            cent      ]	};
+    key <AD09> {	[				o,				O	],
+					[              oe,             OE	]	};
     key <AD11> {	[ dead_circumflex,  dead_diaeresis 	],
 			[     bracketleft			]	};
     key <AD12> {	[          dollar,        asterisk	],
