--- contrib/redhat/sshd.pam.annvix	2004-08-19 10:22:53.000000000 -0600
+++ contrib/redhat/sshd.pam	2004-08-19 10:22:53.000000000 -0600
@@ -1,4 +1,5 @@
 #%PAM-1.0
+auth       required     pam_listfile.so item=user sense=deny file=/etc/ssh/denyusers.pam
 auth       required     pam_stack.so service=system-auth
 auth       required     pam_nologin.so
 account    required     pam_stack.so service=system-auth
--- ssh_config.annvix	2003-08-13 04:37:05.000000000 -0600
+++ ssh_config	2004-04-29 09:35:25.000000000 -0600
@@ -35,3 +35,8 @@
 #   Cipher 3des
 #   Ciphers aes128-cbc,3des-cbc,blowfish-cbc,cast128-cbc,arcfour,aes192-cbc,aes256-cbc
 #   EscapeChar ~
+
+Host *
+  ForwardX11 yes
+  Protocol 2,1
+  StrictHostKeyChecking ask
--- sshd_config.annvix	2004-05-23 18:36:24.000000000 -0600
+++ sshd_config	2004-09-02 11:25:59.000000000 -0600
@@ -11,15 +11,15 @@
 # default value.
 
 #Port 22
-#Protocol 2,1
+Protocol 2
 #ListenAddress 0.0.0.0
 #ListenAddress ::
 
 # HostKey for protocol version 1
 #HostKey /etc/ssh/ssh_host_key
 # HostKeys for protocol version 2
-#HostKey /etc/ssh/ssh_host_rsa_key
-#HostKey /etc/ssh/ssh_host_dsa_key
+HostKey /etc/ssh/ssh_host_rsa_key
+HostKey /etc/ssh/ssh_host_dsa_key
 
 # Lifetime and size of ephemeral version 1 server key
 #KeyRegenerationInterval 1h
@@ -33,9 +33,9 @@
 # Authentication:
 
 #LoginGraceTime 2m
-#PermitRootLogin yes
+PermitRootLogin no
 #StrictModes yes
-#MaxAuthTries 6
+MaxAuthTries 4
 
 #RSAAuthentication yes
 #PubkeyAuthentication yes
@@ -78,16 +78,16 @@
 # ChallengeResponseAuthentication=no
 #UsePAM no
 
-#AllowTcpForwarding yes
+AllowTcpForwarding no
 #GatewayPorts no
-#X11Forwarding no
+X11Forwarding yes
 #X11DisplayOffset 10
 #X11UseLocalhost yes
 #PrintMotd yes
 #PrintLastLog yes
 #TCPKeepAlive yes
 #UseLogin no
-#UsePrivilegeSeparation yes
+UsePrivilegeSeparation yes
 #PermitUserEnvironment no
 #Compression yes
 #ClientAliveInterval 0
