--- tetex-src-3.0/texk/xdvik/mime.c~	2005-02-07 13:28:27.359477658 +0100
+++ tetex-src-3.0/texk/xdvik/mime.c	2005-02-07 13:28:27.359477658 +0100
@@ -601,7 +601,7 @@
     /* See if filename extension is on the mime list: */
     extension = strrchr(filename, '.');
 
-    if (extension == NULL) {
+    if (extension == NULL || *++extension == '\0') {
 	TRACE_HTEX((stderr,
 		    "No extension, defaulting to noExtensionMimeType: %s\n",
 		    resource.no_mime_suffix));
