--- tetex-src-3.0/texk/tetex/epstopdf.epstopdf	2004-10-22 21:33:51.000000000 +0200
+++ tetex-src-3.0/texk/tetex/epstopdf	2005-02-07 13:19:43.180653605 +0100
@@ -57,8 +57,6 @@
 #      whitespace at the end of the eps file.
 #
 
-my $IsWin32 = ($^O =~ /MSWin32/i);
-
 ### program identification
 my $program = "epstopdf";
 my $filedate="2001/03/05";
@@ -70,17 +68,6 @@
 my $GS = "gs";
 $GS = "gswin32c" if $^O eq 'MSWin32';
 
-if ($IsWin32) {
-  $GS = `kpsecheck --ghostscript`;
-  $GS =~ m/^dll\s*:\s*(.+)/mio;
-  $GS = $1;
-  $GS =~ s/gsdll32.dll/gswin32c.exe/io;
-  if ($GS eq "") {
-    $GS = "gswin32c.exe";
-  }
-  $GS = "\"$GS\"" if ($GS =~ m/\s/);
-}
-
 ### options
 $::opt_help=0;
 $::opt_debug=0;
@@ -159,8 +146,12 @@
 }
 
 ### option compress
-my $GSOPTS = "";
-$GSOPTS = "-dUseFlateCompression=false " unless $::opt_compress;
+my $GSOPTS = "-dPDFSETTINGS=/prepress ";
+$GSOPTS .= "-dAutoFilterColorImages=false " unless !($::opt_compress);
+$GSOPTS .= "-sColorImageFilter=FlateEncode " unless !($::opt_compress);
+$GSOPTS .= "-dAutoFilterGrayImages=false " unless !($::opt_compress);
+$GSOPTS .= "-sGrayImageFilter=FlateEncode " unless !($::opt_compress);
+$GSOPTS .= "-dUseFlateCompression=false " unless $::opt_compress;
 
 ### option BoundingBox types
 my $BBName = "%%BoundingBox:";
