diff -uNp t1/arch/x86_64/Makefile.hb21.orig t1/arch/x86_64/Makefile
--- t1/arch/x86_64/Makefile.hb21.orig	2003-07-30 18:13:32.000000000 +0200
+++ t1/arch/x86_64/Makefile	2003-08-01 19:11:14.000000000 +0200
@@ -41,11 +41,11 @@ LINKFLAGS =-T $(TOPDIR)/arch/x86_64/vmli
 CFLAGS += -mno-red-zone
 CFLAGS += -mcmodel=kernel
 CFLAGS += -pipe
-CFLAGS += -fno-reorder-blocks
+# CFLAGS += -fno-reorder-blocks
 # needed for later gcc 3.1
 CFLAGS += -finline-limit=2000
 # needed for earlier gcc 3.1
-CFLAGS += -fno-strength-reduce
+# CFLAGS += -fno-strength-reduce
 CFLAGS += -Wno-sign-compare
 #CFLAGS += -g
 ifneq ($(CONFIG_X86_REMOTE_DEBUG),y)
