--- specs.orig	2005-07-21 18:08:17.000000000 -0400
+++ specs	2005-07-21 18:38:55.000000000 -0400
@@ -30,13 +30,13 @@
 cc1 -E %{traditional|ftraditional|traditional-cpp:-traditional-cpp}
 
 *cc1:
-%(cc1_cpu) %{profile:-p}
+%(cc1_ssp) %(cc1_cpu) %{profile:-p}
 
 *cc1_options:
 %{pg:%{fomit-frame-pointer:%e-pg and -fomit-frame-pointer are incompatible}} %1 %{!Q:-quiet} -dumpbase %B %{d*} %{m*} %{a*} %{c|S:%{o*:-auxbase-strip %*}%{!o*:-auxbase %b}}%{!c:%{!S:-auxbase %b}} %{g*} %{O*} %{W*&pedantic*} %{w} %{std*&ansi&trigraphs} %{v:-version} %{pg:-p} %{p} %{f*} %{undef} %{Qn:-fno-ident} %{--help:--help} %{--target-help:--target-help} %{!fsyntax-only:%{S:%W{o*}%{!o*:-o %b.s}}} %{fsyntax-only:-o %j} %{-param*}
 
 *cc1plus:
-
+%(cc1_ssp) 
 
 *link_gcc_c_sequence:
 %{static:--start-group} %G %L %{static:--end-group}%{!static:%G}
@@ -122,3 +122,6 @@
 *link_command:
 %{!fsyntax-only:%{!c:%{!M:%{!MM:%{!E:%{!S:    %(linker) %l %{pie:-pie} %X %{o*} %{A} %{d} %{e*} %{m} %{N} %{n} %{r}    %{s} %{t} %{u*} %{x} %{z} %{Z} %{!A:%{!nostdlib:%{!nostartfiles:%S}}}    %{static:} %{L*} %(link_libgcc) %o %{fprofile-arcs|fprofile-generate:-lgcov}    %{!nostdlib:%{!nodefaultlibs:%(link_gcc_c_sequence)}}    %{!A:%{!nostdlib:%{!nostartfiles:%E}}} %{T*} }}}}}}
 
+*cc1_ssp: 
+%{!fno-stack-protector*: -fstack-protector-all}
+
