--- postgresql-7.3.4/src/test/regress/resultmap.amd64-testsuite	2002-11-27 18:04:45.000000000 +0100
+++ postgresql-7.3.4/src/test/regress/resultmap	2003-08-31 18:45:35.000000000 +0200
@@ -39,6 +39,7 @@ geometry/powerpc.*-aix5=geometry-powerpc
 geometry/sparc-sun-solaris=geometry-solaris-precision
 geometry/sparc.*-linux-gnu=geometry-solaris-precision
 geometry/alpha.*-linux-gnu=geometry-solaris-precision
+geometry/x86_64.*-linux-gnu=geometry-alpha-precision
 geometry/.*-beos=geometry-intel-beos
 horology/.*-aix4=horology-solaris-1947
 horology/.*-aix5=horology-solaris-1947
