--- rpm-4.0.4/build/parseSpec.c.if	2002-01-20 22:03:52.000000000 +0100
+++ rpm-4.0.4/build/parseSpec.c	2002-06-29 12:24:36.000000000 +0200
@@ -258,7 +258,9 @@ retry:
     SKIPSPACE(s);
 
     match = -1;
-    if (! strncmp("%ifarch", s, sizeof("%ifarch")-1)) {
+    if (!spec->readStack->reading && ! strncmp("%if", s, sizeof("%if")-1)) {
+      match = 0;
+	} else if (! strncmp("%ifarch", s, sizeof("%ifarch")-1)) {
 	const char *arch = rpmExpand("%{_target_cpu}", NULL);
 	s += 7;
 	match = matchTok(arch, s);
