--- courier-imap-2.1.2/authlib/authdaemond.c.daemon	2004-04-23 08:31:21.000000000 -0600
+++ courier-imap-2.1.2/authlib/authdaemond.c	2004-04-23 08:41:29.000000000 -0600
@@ -640,7 +640,6 @@
 int start()
 {
 	int	s;
-	int	lockfd;
 	int	fd;
 	int	pipefd[2];
 	int	do_child;
@@ -648,9 +647,6 @@
 	for (fd=3; fd<256; fd++)
 		close(fd);
 
-	if ((lockfd=ll_daemon_start(AUTHDAEMONLOCK)) < 0)
-		return (1);
-
 	if (pipe(pipefd))
 	{
 		perror("pipe");
@@ -677,8 +673,6 @@
 	signal(SIGHUP, sighup);
 	signal(SIGTERM, sigterm);
 
-	ll_daemon_started(AUTHDAEMONPID, lockfd);
-
 	close(0);
 	if (open("/dev/null", O_RDWR) != 0)
 	{
@@ -750,23 +744,11 @@
 {
 	if (argc > 1)
 	{
-		if (strcmp(argv[1], "start") == 0)
-		{
-			exit(start());
-		}
-
-		if (strcmp(argv[1], "stop") == 0)
-		{
-			exit(ll_daemon_stop(AUTHDAEMONLOCK, AUTHDAEMONPID));
-		}
-
-		if (strcmp(argv[1], "restart") == 0)
-		{
-			exit(ll_daemon_restart(AUTHDAEMONLOCK, AUTHDAEMONPID));
-		}
+		fprintf(stderr, "Error: authdaemond no longer handles its own daemonizing.\n"
+			"Use new startup script.\n");
+		exit(1);
 	}
 
-	fprintf(stderr, "Usage: %s start|stop|restart\n", argv[0]);
-	exit (1);
+	start();
 	return (0);
 }
