diff -Naur sysvinit-2.85/src/hddown.c sysvinit-2.85.oden/src/hddown.c
--- sysvinit-2.85/src/hddown.c	2001-11-07 16:11:21.000000000 +0100
+++ sysvinit-2.85.oden/src/hddown.c	2004-01-25 22:23:20.000000000 +0100
@@ -15,7 +15,7 @@
 
 #include <sys/ioctl.h>
 
-#include <linux/hdreg.h>
+#include "/usr/include/linux/hdreg.h"
 
 #define MAX_DISKS	64
 #define PROC_IDE	"/proc/ide"
diff -Naur sysvinit-2.85/src/ifdown.c sysvinit-2.85.oden/src/ifdown.c
--- sysvinit-2.85/src/ifdown.c	1998-06-02 22:41:47.000000000 +0200
+++ sysvinit-2.85.oden/src/ifdown.c	2004-01-25 22:23:20.000000000 +0100
@@ -14,7 +14,7 @@
 #include <sys/ioctl.h>
 #include <sys/socket.h>
 #include <sys/time.h>
-#include <sys/errno.h>
+#include <errno.h>
 
 #include <net/if.h>
 #include <netinet/in.h>
diff -Naur sysvinit-2.85/src/last.c sysvinit-2.85.oden/src/last.c
--- sysvinit-2.85/src/last.c	2004-01-25 22:23:09.000000000 +0100
+++ sysvinit-2.85.oden/src/last.c	2004-01-25 22:23:20.000000000 +0100
@@ -19,7 +19,7 @@
 
 #include <sys/types.h>
 #include <sys/stat.h>
-#include <sys/fcntl.h>
+#include <fcntl.h>
 #include <time.h>
 #include <stdio.h>
 #include <ctype.h>
diff -Naur sysvinit-2.85/src/reboot.h sysvinit-2.85.oden/src/reboot.h
--- sysvinit-2.85/src/reboot.h	1997-09-24 10:55:52.000000000 +0200
+++ sysvinit-2.85.oden/src/reboot.h	2004-01-25 22:23:20.000000000 +0100
@@ -6,9 +6,7 @@
  *
  */
 
-#if defined(__GLIBC__)
-#  include <sys/reboot.h>
-#endif
+#include <sys/reboot.h>
 
 #define BMAGIC_HARD	0x89ABCDEF
 #define BMAGIC_SOFT	0
@@ -16,8 +14,4 @@
 #define BMAGIC_HALT	0xCDEF0123
 #define BMAGIC_POWEROFF	0x4321FEDC
 
-#if defined(__GLIBC__)
-  #define init_reboot(magic) reboot(magic)
-#else
-  #define init_reboot(magic) reboot(0xfee1dead, 672274793, magic)
-#endif
+#define init_reboot(magic) reboot(magic)
