--- chkfontpath-1.7/chkfontpath.c.old	Fri Oct 20 16:16:16 2000
+++ chkfontpath-1.7/chkfontpath.c	Fri Oct 20 16:16:29 2000
@@ -176,6 +176,7 @@
   FILE *f;
   char *fontsdir;
   char *fixDir = strdup(newDir);
+  char *fixDir2 = strdup(newDir);
 
   fixDirLen = strlen(fixDir);
   if (fixDir[fixDirLen - 1] == '/') {
@@ -274,7 +275,7 @@
     second++;
   }
   memmove(fpList + second + 1, fpList + second, sizeof(char *) * (fpCount - 1 - second));
-  fpList[second] = fixDir;
+  fpList[second] = fixDir2;
 }
 
 void removeDir(const char *delDir)
