--- tcllib-1.7/Makefile.in.no-apps	2004-12-03 14:21:34.436476024 +0100
+++ tcllib-1.7/Makefile.in	2004-12-03 14:17:26.758128856 +0100
@@ -53,19 +53,19 @@
 		-no-examples -no-html \
 		-pkg-path   $(DESTDIR)$(pkglibdir) \
 		-nroff-path $(DESTDIR)$(mandir)/mann \
-		-no-wait -no-gui
+		-no-wait -no-gui -no-apps
 
 install-libraries:
 	$(TCLSH_PROG) `$(CYGPATH) $(srcdir)/installer.tcl` \
 		-pkg-path   $(DESTDIR)$(pkglibdir) \
 		-no-examples -no-html -no-nroff \
-		-no-wait -no-gui
+		-no-wait -no-gui -no-apps
 
 install-doc:
 	$(TCLSH_PROG) `$(CYGPATH) $(srcdir)/installer.tcl` \
 		-nroff-path $(DESTDIR)$(mandir)/mann \
 		-no-examples -no-pkgs -no-html \
-		-no-wait -no-gui
+		-no-wait -no-gui -no-apps
 
 test:
 	$(TCLSH_PROG) `$(CYGPATH) $(srcdir)/sak.tcl` test
