--- elfutils-0.89/src/ldgeneric.c.fixlets	2003-08-18 01:49:51.000000000 +0200
+++ elfutils-0.89/src/ldgeneric.c	2003-11-05 14:55:37.000000000 +0100
@@ -1735,7 +1735,7 @@ extract_from_archive (struct usedfiles *
      the first definition.  */
   // XXX Is this a compatible behavior?
   bool any_used;
-  int round = 0;
+  int rounds = 0;
   do
     {
       any_used = false;
@@ -1793,7 +1793,7 @@ extract_from_archive (struct usedfiles *
 	    }
 	}
 
-      if (++round == 1)
+      if (++rounds == 1)
 	{
 	  /* This is an archive therefore it must have a number.  */
 	  assert (fileinfo->archive_seq != 0);
