--- mkinitrd-4.2.17/nash/nash.c.rtc	2007-08-24 11:46:32.000000000 +0200
+++ mkinitrd-4.2.17/nash/nash.c	2007-08-24 11:47:13.000000000 +0200
@@ -1992,6 +1992,7 @@ int mountdevCommand(char * cmd, char * e
     mount("/dev", "/dev", "tmpfs", MS_MGC_VAL, cmd?options:NULL);
     smartmknod("/dev/console", S_IFCHR | 0600, makedev(5, 1));
     smartmknod("/dev/null", S_IFCHR | 0666, makedev(1, 3));
+    smartmknod("/dev/rtc", S_IFCHR | 0664, makedev(10, 135));
     for (i=0;i<8;i++) {
 	snprintf(devName, 128, "/dev/tty%d", i);
 	smartmknod(devName, S_IFCHR | 0600, makedev(4, i));
--- mkinitrd-4.2.17/mkinitrd.rtc	2007-08-20 22:32:52.000000000 +0200
+++ mkinitrd-4.2.17/mkinitrd	2007-08-20 22:32:52.000000000 +0200
@@ -1094,6 +1094,7 @@ done
 # without devfs...
 mknod $MNTIMAGE/dev/console c 5 1
 mknod $MNTIMAGE/dev/null c 1 3
+mknod $MNTIMAGE/dev/rtc c 10 135
 for i in 1 2 3 4; do
     mknod $MNTIMAGE/dev/tty$i c 4 $i
 done
