--- nss_ldap-220/config.h.in.db4	2004-05-13 21:34:11.000000000 -0600
+++ nss_ldap-220/config.h.in	2004-06-30 10:09:22.000000000 -0600
@@ -150,6 +150,9 @@
 /* Define if you have the <db3/db_185.h> header file.  */
 #undef HAVE_DB3_DB_185_H
 
+/* Define if you have the <db4/db_185.h> header file.  */
+#undef HAVE_DB4_DB_185_H
+
 /* Define if you have the <db_185.h> header file.  */
 #undef HAVE_DB_185_H
 
--- nss_ldap-220/configure.in.db4	2004-06-30 10:09:40.000000000 -0600
+++ nss_ldap-220/configure.in	2004-06-30 10:10:03.000000000 -0600
@@ -133,6 +133,7 @@
 AC_CHECK_HEADERS(db1/db.h)
 AC_CHECK_HEADERS(db_185.h)
 AC_CHECK_HEADERS(db3/db_185.h)
+AC_CHECK_HEADERS(db4/db_185.h)
 AC_CHECK_HEADERS(alignof.h)
 AC_CHECK_HEADERS(rpc/rpcent.h)
 AC_CHECK_HEADERS(sys/byteorder.h)
--- nss_ldap-220/util.c.db4	2004-06-30 10:10:36.000000000 -0600
+++ nss_ldap-220/util.c	2004-06-30 10:12:13.000000000 -0600
@@ -84,7 +84,10 @@
  * right API!
  */
 #if defined(RFC2307BIS) || defined(AT_OC_MAP)
-#ifdef HAVE_DB3_DB_185_H
+#ifdef HAVE_DB4_DB_185_H
+#include <db4/db_185.h>
+#define DN2UID_CACHE
+#elif defined(HAVE_DB3_DB_185_H)
 #include <db3/db_185.h>
 #define DN2UID_CACHE
 #elif defined(HAVE_DB_185_H)
--- nss_ldap-220/ldap-nss.c.db4	2004-06-30 10:12:47.000000000 -0600
+++ nss_ldap-220/ldap-nss.c	2004-06-30 10:13:46.000000000 -0600
@@ -72,7 +72,9 @@
 #include <sasl.h>
 #endif
 #ifdef AT_OC_MAP
-#ifdef HAVE_DB3_DB_185_H
+#ifdef HAVE_DB4_DB_185_H
+#include <db4/db_185.h>
+#elif defined(HAVE_DB3_DB_185_H)
 #include <db3/db_185.h>
 #elif defined(HAVE_DB_185_H)
 #include <db_185.h>
