diff -u test/arch/i386/boot/bootsect.S.orig test/arch/i386/boot/bootsect.S
--- test/arch/i386/boot/bootsect.S.orig	2002-07-23 19:36:16.000000000 +0200
+++ test/arch/i386/boot/bootsect.S	2002-07-23 19:36:01.000000000 +0200
@@ -236,7 +236,7 @@
 rp_read:
 #ifdef __BIG_KERNEL__			# look in setup.S for bootsect_kludge
 	bootsect_kludge = 0x220		# 0x200 + 0x20 which is the size of the
-	lcall	bootsect_kludge		# bootsector + bootsect_kludge offset
+	lcall	*bootsect_kludge	# bootsector + bootsect_kludge offset
 #else
 	movw	%es, %ax
 	subw	$SYSSEG, %ax
