diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/drivers/hotplug/cpqphp_pci.c w1/drivers/hotplug/cpqphp_pci.c
--- w2/drivers/hotplug/cpqphp_pci.c	2003-08-20 19:13:30.000000000 +0200
+++ w1/drivers/hotplug/cpqphp_pci.c	2003-08-20 09:41:40.000000000 +0200
@@ -387,7 +387,8 @@ static int PCI_ScanBusNonBridge (u8 bus,
 
 static int PCI_ScanBusForNonBridge(struct controller *ctrl, u8 bus_num, u8 * dev_num)
 {
-	u8 tdevice;
+	u16 tdevice; /* u8 is not enough, we need to reach 0x100 to
+			exit loop */
 	u32 work;
 	u8 tbus;
 
