diff -uNp linux-2.4.21/kdb/modules/kdbm_task.c.orig linux-2.4.21/kdb/modules/kdbm_task.c
--- linux-2.4.21/kdb/modules/kdbm_task.c.orig	2003-08-08 00:47:03.000000000 +0200
+++ linux-2.4.21/kdb/modules/kdbm_task.c	2003-08-08 00:49:31.000000000 +0200
@@ -119,8 +119,8 @@ kdbm_task(int argc, const char **argv, c
 	kdb_printf(" lock_depth=%d\n", tp->lock_depth);
 
 	kdb_printf(
-	    "  fs=0x%p files=0x%p mm=0x%p nr_local_pages=%u\n",
-	    tp->fs, tp->files, tp->mm, tp->nr_local_pages);
+	    "  fs=0x%p files=0x%p mm=0x%p\n",
+	    tp->fs, tp->files, tp->mm);
 
 	kdb_printf(
 	    "  uid=%d euid=%d suid=%d fsuid=%d gid=%d egid=%d sgid=%d fsgid=%d\n",
