--- linux-2.4.22/arch/i386/kernel/dmi_scan.c.dm06-3.orig	2003-08-28 03:10:12.000000000 +0200
+++ linux-2.4.22/arch/i386/kernel/dmi_scan.c	2003-08-28 03:11:42.000000000 +0200
@@ -875,6 +875,12 @@ static __initdata struct dmi_blacklist d
 
 	/* Machines which have problems handling enabled local APICs */
 
+	{ local_apic_kills_bios, "HP ", {
+			MATCH(DMI_SYS_VENDOR, "Hewlett-Packard"),
+			MATCH(DMI_PRODUCT_NAME, "e-pc      "),
+			NO_MATCH, NO_MATCH
+			} },
+
 	{ local_apic_kills_bios, "Dell Inspiron", {
 			MATCH(DMI_SYS_VENDOR, "Dell Computer Corporation"),
 			MATCH(DMI_PRODUCT_NAME, "Inspiron"),
