diff -ru -x '*.[oas]' -x '*.[oa].flags' -x .depend -x .config -x modversions.h -x '*.ver' -x autoconf.h /usr/src/linux-2.4.20-2mdk/fs/fat/file.c linux-2.4.20-2mdk/fs/fat/file.c
--- /usr/src/linux-2.4.20-2mdk/fs/fat/file.c	2001-08-12 21:56:56.000000000 +0400
+++ linux-2.4.20-2mdk/fs/fat/file.c	2002-12-30 14:48:15.000000000 +0300
@@ -117,8 +117,17 @@
 	int cluster;
 
 	/* Why no return value?  Surely the disk could fail... */
+#if 0
+	/*
+	 * Why do we need this check? VFS layer should already have
+	 * checked user access before we ever get here; and we can 
+	 * get here from iput on read-only file system, consider
+	 * remounting file system read-only after changing some file.
+	 * That is exactly what happens with supermount and breaks it
+	 */
 	if (IS_RDONLY (inode))
 		return /* -EPERM */;
+#endif
 	if (IS_IMMUTABLE(inode))
 		return /* -EPERM */;
 	cluster = 1 << sbi->cluster_bits;
