--- mkinitrd.mkdevices	2005-02-27 23:01:34.000000000 -0500
+++ mkinitrd	2005-02-27 23:02:07.000000000 -0500
@@ -635,6 +635,10 @@
     done
 fi
 
+echo "echo Creating block devices" >> $RCFILE
+echo "mountdev size=1M,mode=0755" >> $RCFILE
+echo "mkdevices /dev" >> $RCFILE
+
 if [ -n "$loopDev" ]; then
     [ -d /initrd ] || mkdir /initrd
 
@@ -668,8 +672,8 @@
     echo "mount -o $rootopts --ro -t $rootfs $rootdev /sysroot" >> $RCFILE
 
     echo "pivot_root /sysroot /sysroot/initrd" >> $RCFILE
-    echo "echo Remounting devfs at correct place if necessary" >> $RCFILE
-    echo "handledevfs" >> $RCFILE
+#    echo "echo Remounting devfs at correct place if necessary" >> $RCFILE
+#    echo "handledevfs" >> $RCFILE
     echo "umount /initrd/proc" >> $RCFILE
 else
     echo "umount /proc" >> $RCFILE
