--- pciutils-2.1.99-test3/lib/configure.org	2004-03-15 10:15:21.000000000 -0700
+++ pciutils-2.1.99-test3/lib/configure	2004-03-15 10:15:21.000000000 -0700
@@ -47,7 +47,7 @@
 						echo >>$c '#define HAVE_PM_INTEL_CONF'
 						ok=1
 						;;
-				alpha|ia64|sparc|sparc64|ppc|ppc64)
+				alpha|ia64|sparc|sparc64|ppc|ppc64|x86_64|amd64)
 						echo >>$c '#define HAVE_64BIT_ADDRESS'
 						;;
 		esac
