--- xc/programs/Xserver/hw/xfree86/common/xf86PciInfo.h.2_rv280	Fri Mar  7 05:30:46 2003
+++ xc/programs/Xserver/hw/xfree86/common/xf86PciInfo.h	Fri Mar  7 05:30:45 2003
@@ -217,6 +217,14 @@
 #define PCI_CHIP_MACH64VT		0x5654
 #define PCI_CHIP_MACH64VU		0x5655
 #define PCI_CHIP_MACH64VV		0x5656
+#define PCI_CHIP_RV280_5960             0x5960
+#define PCI_CHIP_RV280_5961             0x5961
+#define PCI_CHIP_RV280_5962             0x5962
+#define PCI_CHIP_RV280_5963             0x5963
+#define PCI_CHIP_RV280_5968             0x5968
+#define PCI_CHIP_RV280_5969             0x5969
+#define PCI_CHIP_RV280_596A             0x596A
+#define PCI_CHIP_RV280_596B             0x596B
 
 /* Avance Logic */
 #define PCI_CHIP_ALG2064		0x2064
--- xc/programs/Xserver/hw/xfree86/drivers/ati/atichip.c.2_rv280	Mon Mar  3 23:13:18 2003
+++ xc/programs/Xserver/hw/xfree86/drivers/ati/atichip.c	Tue Mar  4 04:22:23 2003
@@ -675,6 +675,18 @@
         case NewChipID('L', 'g'):
             return ATI_CHIP_RADEONMOBILITY9;
 
+        case NewChipID('Y', '\''):
+        case NewChipID('Y', 'a'):
+        case NewChipID('I', 'b'):
+        case NewChipID('I', 'c'):
+            return ATI_CHIP_RV280;
+
+        case NewChipID('Y', 'h'):
+        case NewChipID('Y', 'i'):
+        case NewChipID('Y', 'j'):
+        case NewChipID('Y', 'k'):
+            return ATI_CHIP_RADEONMOBILITY9PLUS;
+
         case NewChipID('A', 'D'):
         case NewChipID('A', 'E'):
         case NewChipID('A', 'F'):
--- xc/programs/Xserver/hw/xfree86/drivers/ati/atichip.h.2_rv280	Mon Mar  3 23:13:27 2003
+++ xc/programs/Xserver/hw/xfree86/drivers/ati/atichip.h	Tue Mar  4 04:23:59 2003
@@ -96,6 +96,8 @@
     ATI_CHIP_RV200,             /* RV200 */
     ATI_CHIP_RV250,             /* RV250 */
     ATI_CHIP_RADEONMOBILITY9,   /* Radeon M9 */
+    ATI_CHIP_RV280,             /* RV250 */
+    ATI_CHIP_RADEONMOBILITY9PLUS,   /* Radeon M9+ */
     ATI_CHIP_R300,              /* R300 */
     ATI_CHIP_HDTV               /* HDTV */
 } ATIChipType;
--- xc/programs/Xserver/hw/xfree86/drivers/ati/atiprobe.c.2_rv280	Mon Mar  3 23:12:23 2003
+++ xc/programs/Xserver/hw/xfree86/drivers/ati/atiprobe.c	Tue Mar  4 04:24:47 2003
@@ -1738,6 +1738,8 @@
                     case ATI_CHIP_RV200:
                     case ATI_CHIP_RV250:
                     case ATI_CHIP_RADEONMOBILITY9:
+                    case ATI_CHIP_RV280:
+                    case ATI_CHIP_RADEONMOBILITY9PLUS:
                     case ATI_CHIP_R300:
                         DoRadeon = TRUE;
                         continue;
--- xc/programs/Xserver/hw/xfree86/drivers/ati/radeon.h.2_rv280	Tue Mar  4 04:08:22 2003
+++ xc/programs/Xserver/hw/xfree86/drivers/ati/radeon.h	Tue Mar  4 04:11:41 2003
@@ -257,6 +257,8 @@
     CHIP_FAMILY_R200,
     CHIP_FAMILY_RV250,
     CHIP_FAMILY_M9,
+    CHIP_FAMILY_RV280,
+    CHIP_FAMILY_M9PLUS,
     CHIP_FAMILY_R300
 } RADEONChipFamily;
 
--- xc/programs/Xserver/hw/xfree86/drivers/ati/radeon_driver.c.2_rv280	Mon Mar  3 23:11:01 2003
+++ xc/programs/Xserver/hw/xfree86/drivers/ati/radeon_driver.c	Tue Mar  4 04:15:35 2003
@@ -1422,6 +1422,20 @@
 	info->ChipFamily = CHIP_FAMILY_M9;
 	break;
 
+    case PCI_CHIP_RV280_5960:
+    case PCI_CHIP_RV280_5961:
+    case PCI_CHIP_RV280_5962:
+    case PCI_CHIP_RV280_5963:
+	info->ChipFamily = CHIP_FAMILY_RV280;
+	break;
+
+    case PCI_CHIP_RV280_5968:
+    case PCI_CHIP_RV280_5969:
+    case PCI_CHIP_RV280_596A:
+    case PCI_CHIP_RV280_596B:
+	info->ChipFamily = CHIP_FAMILY_M9PLUS;
+	break;
+
     case PCI_CHIP_R300_AD:
     case PCI_CHIP_R300_AE:
     case PCI_CHIP_R300_AF:
@@ -1641,6 +1655,14 @@
 	case PCI_CHIP_RV250_Le:
 	case PCI_CHIP_RV250_Lf:
 	case PCI_CHIP_RV250_Lg:
+	case PCI_CHIP_RV280_5960:
+	case PCI_CHIP_RV280_5961:
+	case PCI_CHIP_RV280_5962:
+	case PCI_CHIP_RV280_5963:
+	case PCI_CHIP_RV280_5968:
+	case PCI_CHIP_RV280_5969:
+	case PCI_CHIP_RV280_596A:
+	case PCI_CHIP_RV280_596B:
 	case PCI_CHIP_R300_AD:
 	case PCI_CHIP_R300_AE:
 	case PCI_CHIP_R300_AF:
--- xc/programs/Xserver/hw/xfree86/drivers/ati/radeon_probe.c.2_rv280	Mon Mar  3 23:11:18 2003
+++ xc/programs/Xserver/hw/xfree86/drivers/ati/radeon_probe.c	Tue Mar  4 02:56:59 2003
@@ -112,6 +112,14 @@
     { PCI_CHIP_RV250_Le, "ATI Radeon Mobility M9 Le (AGP)" },
     { PCI_CHIP_RV250_Lf, "ATI Radeon Mobility M9 Lf (AGP)" },
     { PCI_CHIP_RV250_Lg, "ATI Radeon Mobility M9 Lg (AGP)" },
+    { PCI_CHIP_RV280_5960, "ATI Radeon 9200 5960 (AGP)" },
+    { PCI_CHIP_RV280_5961, "ATI Radeon 9200 5961 (AGP)" },
+    { PCI_CHIP_RV280_5962, "ATI Radeon 9200 5962 (AGP)" },
+    { PCI_CHIP_RV280_5963, "ATI Radeon 9200 5963 (AGP)" },
+    { PCI_CHIP_RV280_5968, "ATI Radeon M9+ 5968 (AGP)" },
+    { PCI_CHIP_RV280_5969, "ATI Radeon M9+ 5969 (AGP)" },
+    { PCI_CHIP_RV280_596A, "ATI Radeon M9+ 596A (AGP)" },
+    { PCI_CHIP_RV280_596B, "ATI Radeon M9+ 596B (AGP)" },
     { PCI_CHIP_R300_AD, "ATI Radeon 9500 AD (AGP)" },
     { PCI_CHIP_R300_AE, "ATI Radeon 9500 AE (AGP)" },
     { PCI_CHIP_R300_AF, "ATI Radeon 9500 AF (AGP)" },
@@ -162,6 +170,14 @@
     { PCI_CHIP_RV250_Le, PCI_CHIP_RV250_Le, RES_SHARED_VGA },
     { PCI_CHIP_RV250_Lf, PCI_CHIP_RV250_Lf, RES_SHARED_VGA },
     { PCI_CHIP_RV250_Lg, PCI_CHIP_RV250_Lg, RES_SHARED_VGA },
+    { PCI_CHIP_RV280_5960, PCI_CHIP_RV280_5960, RES_SHARED_VGA },
+    { PCI_CHIP_RV280_5961, PCI_CHIP_RV280_5961, RES_SHARED_VGA },
+    { PCI_CHIP_RV280_5962, PCI_CHIP_RV280_5962, RES_SHARED_VGA },
+    { PCI_CHIP_RV280_5963, PCI_CHIP_RV280_5963, RES_SHARED_VGA },
+    { PCI_CHIP_RV280_5968, PCI_CHIP_RV280_5968, RES_SHARED_VGA },
+    { PCI_CHIP_RV280_5969, PCI_CHIP_RV280_5969, RES_SHARED_VGA },
+    { PCI_CHIP_RV280_596A, PCI_CHIP_RV280_596A, RES_SHARED_VGA },
+    { PCI_CHIP_RV280_596B, PCI_CHIP_RV280_596A, RES_SHARED_VGA },
     { PCI_CHIP_R300_AD, PCI_CHIP_R300_AD, RES_SHARED_VGA },
     { PCI_CHIP_R300_AE, PCI_CHIP_R300_AE, RES_SHARED_VGA },
     { PCI_CHIP_R300_AF, PCI_CHIP_R300_AF, RES_SHARED_VGA },
--- xc/programs/Xserver/hw/xfree86/drivers/ati/radeon_version.h.2_rv280	Fri Mar  7 06:56:11 2003
+++ xc/programs/Xserver/hw/xfree86/drivers/ati/radeon_version.h	Fri Mar  7 06:56:31 2003
@@ -39,6 +39,7 @@
 #define RADEON_DRIVER_NAME   "radeon"
 #define R200_DRIVER_NAME     "r200"
 #define RV250_DRIVER_NAME    "r200"
+#define RV280_DRIVER_NAME    "r200"
 
 #define RADEON_VERSION_MAJOR 4
 #define RADEON_VERSION_MINOR 0
