--- XFree86-4.3/xc/programs/xdm/socket.c.CAN-2004-0419	2004-07-19 13:22:02.000000000 -0600
+++ XFree86-4.3/xc/programs/xdm/socket.c	2004-07-19 13:22:51.000000000 -0600
@@ -66,6 +66,9 @@
     char *name = localHostname ();
     registerHostname (name, strlen (name));
 
+    if (request_port == 0)
+	return;
+		
     chooserFd = socket (AF_INET, SOCK_STREAM, 0);
     Debug ("Created chooser socket %d\n", chooserFd);
     if (chooserFd == -1)
