diff -urN openssh-3.1p1/entropy.c openssh-3.1p1-modified/entropy.c
--- openssh-3.1p1/entropy.c	Tue Jan 22 11:57:54 2002
+++ openssh-3.1p1-modified/entropy.c	Thu Mar  7 17:29:13 2002
@@ -136,7 +136,7 @@
 	 * OpenSSL version numbers: MNNFFPPS: major minor fix patch status
 	 * We match major, minor, fix and status (not patch)
 	 */
-	if ((SSLeay() ^ OPENSSL_VERSION_NUMBER) & ~0xff0L)
+	if ((SSLeay() >> 12) != (OPENSSL_VERSION_NUMBER >> 12))
 		fatal("OpenSSL version mismatch. Built against %lx, you "
 		    "have %lx", OPENSSL_VERSION_NUMBER, SSLeay());
 
