--- openssh-3.6.1p2/contrib/redhat/sshd.init.supervise	2003-12-01 23:13:57.000000000 -0700
+++ openssh-3.6.1p2/contrib/redhat/sshd.init	2003-12-01 23:15:26.000000000 -0700
@@ -31,14 +31,14 @@
 
 do_rsa1_keygen() {
 	if [ ! -s $RSA1_KEY ]; then
-		echo -n $"Generating SSH1 RSA host key: "
+		gprintf "Generating SSH1 RSA host key: "
 		if $KEYGEN -q -t rsa1 -f $RSA1_KEY -C '' -N '' >&/dev/null; then
 			chmod 600 $RSA1_KEY
 			chmod 644 $RSA1_KEY.pub
-			success $"RSA1 key generation"
+			success "RSA1 key generation"
 			echo
 		else
-			failure $"RSA1 key generation"
+			failure "RSA1 key generation"
 			echo
 			exit 1
 		fi
@@ -47,14 +47,14 @@
 
 do_rsa_keygen() {
 	if [ ! -s $RSA_KEY ]; then
-		echo -n $"Generating SSH2 RSA host key: "
+		gprintf "Generating SSH2 RSA host key: "
 		if $KEYGEN -q -t rsa -f $RSA_KEY -C '' -N '' >&/dev/null; then
 			chmod 600 $RSA_KEY
 			chmod 644 $RSA_KEY.pub
-			success $"RSA key generation"
+			success "RSA key generation"
 			echo
 		else
-			failure $"RSA key generation"
+			failure "RSA key generation"
 			echo
 			exit 1
 		fi
@@ -63,14 +63,14 @@
 
 do_dsa_keygen() {
 	if [ ! -s $DSA_KEY ]; then
-		echo -n $"Generating SSH2 DSA host key: "
+		gprintf "Generating SSH2 DSA host key: "
 		if $KEYGEN -q -t dsa -f $DSA_KEY -C '' -N '' >&/dev/null; then
 			chmod 600 $DSA_KEY
 			chmod 644 $DSA_KEY.pub
-			success $"DSA key generation"
+			success "DSA key generation"
 			echo
 		else
-			failure $"DSA key generation"
+			failure "DSA key generation"
 			echo
 			exit 1
 		fi
@@ -82,7 +82,7 @@
 	$SSHD -t
 	RETVAL=$?
 	if [ ! "$RETVAL" = 0 ]; then
-		failure $"Configuration file or keys are invalid"
+		failure "Configuration file or keys are invalid"
 		echo
 	fi
 }
@@ -94,7 +94,9 @@
 	do_rsa_keygen
 	do_dsa_keygen
 
-	echo -n $"Starting $prog:"
+	supervise_start sshd
+	[ "$?" = "1" ] && exit
+	gprintf "Starting %s:" "$prog"
 	initlog -c "$SSHD $OPTIONS" && success "startup" || failure "startup"
 	RETVAL=$?
 	[ "$RETVAL" = 0 ] && touch /var/lock/subsys/sshd
@@ -103,7 +105,9 @@
 
 stop()
 {
-	echo -n $"Stopping $prog:"
+	supervise_stop sshd
+	[ "$?" = "1" ] && exit
+	gprintf "Stopping %s:" "$prog"
 	if [ -r /var/run/sshd.pid ]; then
 	    kill -TERM `cat /var/run/sshd.pid`
 	    RETVAL=$?
@@ -118,7 +122,9 @@
 
 reload()
 {
-	echo -n $"Reloading $prog:"
+	supervise_hup sshd
+	[ "$?" = "1" ] && exit
+	gprintf "Reloading %s:" "$prog"
 	if [ -r /var/run/sshd.pid ]; then
 	    kill -HUP `cat /var/run/sshd.pid`
 	    RETVAL=$?
@@ -145,6 +151,8 @@
 		reload
 		;;
 	condrestart)
+		supervise_term sshd
+		[ "$?" = "1" ] && exit
 		if [ -f /var/lock/subsys/sshd ] ; then
 			do_restart_sanity_check
 			if [ "$RETVAL" = 0 ] ; then
@@ -156,11 +164,13 @@
 		fi
 		;;
 	status)
+		supervise_status sshd
+		[ "$?" = "1" ] && exit
 		status $SSHD
 		RETVAL=$?
 		;;
 	*)
-		echo $"Usage: $0 {start|stop|restart|reload|condrestart|status}"
+		gprintf "Usage: %s {start|stop|restart|reload|condrestart|status}\n" "$0"
 		RETVAL=1
 esac
 exit $RETVAL
