--- clamav-0.88.1/shared/output.c.stderr	2006-03-26 12:35:28.000000000 -0700
+++ clamav-0.88.1/shared/output.c	2006-04-11 10:20:56.000000000 -0600
@@ -123,7 +123,11 @@
 #endif
 	if(!logg_fs) {
 	    old_umask = umask(0037);
-	    if((logg_fs = fopen(logg_file, "a")) == NULL) {
+	    if (strcmp(logg_file,"/dev/stderr") == 0) {
+	        logg_fs = stderr;
+	        umask(old_umask);
+	    }
+	    else if((logg_fs = fopen(logg_file, "a")) == NULL) {
 		umask(old_umask);
 #ifdef CL_THREAD_SAFE
 		pthread_mutex_unlock(&logg_mutex);
