--- XFree86-4.3/xc/lib/X11/XlcDL.c._LP64-fix	2003-08-01 16:36:11.000000000 +0200
+++ XFree86-4.3/xc/lib/X11/XlcDL.c	2003-08-01 17:43:34.000000000 +0200
@@ -55,7 +55,7 @@ Sun Microsystems, Inc. or its licensors 
 #include "XlcPublic.h"
 #include "XlcPubI.h"
 
-#ifdef _LP64
+#if defined(_LP64) && !defined(__linux__)
 # if defined(__sparcv9)
 #  define	_MACH64_NAME		"sparcv9"
 # elif defined(__ia64__) 
@@ -248,7 +248,7 @@ const char *lc_dir;
     if (strstr (dl_name, "../"))
 	return NULL;
 
-#if defined (_LP64) && defined (_MACH64_NAME)
+#if defined (_LP64) && defined (_MACH64_NAME) && !defined(__linux__)
     len = (lc_dir ? strlen(lc_dir) : 0 ) +
 	(dl_name ? strlen(dl_name) : 0) + _MACH64_NAME_LEN + 10;
     path = Xmalloc(len + 1);
