diff -Naur syslinux-1.75.orig/bios.inc syslinux-1.75/bios.inc
--- syslinux-1.75.orig/bios.inc	2002-06-15 06:46:18.000000000 +0200
+++ syslinux-1.75/bios.inc	2002-07-17 16:19:51.000000000 +0200
@@ -35,4 +35,24 @@
                 absolute 0484h
 BIOS_vidrows    resb 1			; Number of screen rows
 
+		absolute 0500h		; all Mandrake VESA variable extensions.
+VESA_modeattr	resw 1
+VESA_winattrA	resb 1
+VESA_winattrB	resb 1
+VESA_wingran	resw 1
+VESA_winsz	resw 1
+VESA_winsegA	resw 1
+VESA_winsegB	resw 1
+VESA_farwinpos	resd 1
+VESA_byteline	resw 1
+ExtImageW	resw 1
+ExtImageH	resw 1
+ExtImageC	resw 1
+ExtImageCY	resw 1
+ExtImageX	resw 1
+ExtImageY	resw 1
+ExtImagePixel	resw 1
+ExtImagePal	resb 128 * 3
+		resb 128
+
 %endif ; _BIOS_INC
diff -Naur syslinux-1.75.orig/comboot.inc syslinux-1.75/comboot.inc
--- syslinux-1.75.orig/comboot.inc	2002-06-15 06:46:18.000000000 +0200
+++ syslinux-1.75/comboot.inc	2002-07-17 16:57:32.000000000 +0200
@@ -20,7 +20,7 @@
 ; Looks like a COMBOOT image but too large
 comboot_too_large:
 		mov si,err_comlarge
-		call cwritestr
+		call err_cwritestr
 cb_enter:	jmp enter_command
 
 ;
diff -Naur syslinux-1.75.orig/conio.inc syslinux-1.75/conio.inc
--- syslinux-1.75.orig/conio.inc	2002-06-15 06:46:18.000000000 +0200
+++ syslinux-1.75/conio.inc	2002-07-17 18:44:17.000000000 +0200
@@ -68,8 +68,6 @@
 		push cx
 		push edx
 		es lodsb
-                cmp al,1Ah                      ; DOS EOF?
-		je msg_done_pop
 		push si
 		mov cl,[UsingVGA]
 		inc cl				; 01h = text mode, 02h = graphics
@@ -91,6 +89,8 @@
 msg_putchar:                                    ; Normal character
                 cmp al,0Fh                      ; ^O = color code follows
                 je msg_ctrl_o
+		cmp al,0Eh			; ^N = extented code follows
+		je msg_ctrl_ext
                 cmp al,0Dh                      ; Ignore <CR>
                 je msg_ignore
                 cmp al,0Ah                      ; <LF> = newline
@@ -127,6 +127,9 @@
 msg_ctrl_o:                                     ; ^O = color code follows
                 mov word [NextCharJump],msg_setbg
                 ret
+msg_ctrl_ext:					; ^N = extended code follows
+		mov word [NextCharJump],msg_ext
+		ret
 msg_newline:                                    ; Newline char or end of line
 		mov si,crlf_msg
 		call write_serial_str_displaymask
@@ -140,7 +143,9 @@
                 ja msg_scroll
                 mov [CursorRow],al
                 jmp short msg_gotoxy
-msg_scroll:     xor cx,cx                       ; Upper left hand corner
+msg_scroll:	test byte [UsingVGA],10h
+		jnz msg_ignore			; If VESA enabled ignore output
+		xor cx,cx                       ; Upper left hand corner
                 mov dx,[ScreenSize]
                 mov [CursorRow],dh		; New cursor at the bottom
                 mov bh,[ScrollAttribute]
diff -Naur syslinux-1.75.orig/font.inc syslinux-1.75/font.inc
--- syslinux-1.75.orig/font.inc	2002-06-15 06:46:18.000000000 +0200
+++ syslinux-1.75/font.inc	2002-07-17 16:48:33.000000000 +0200
@@ -66,6 +66,8 @@
 		mov bh,[VGAFontSize]
 
 		xor bl,bl			; Needed by both INT 10h calls
+		test [UsingVGA], byte 10h	; Are we in VESA mode?
+		jnz .lf_ret			; If Yes, do nothing
 		cmp [UsingVGA], byte 1		; Are we in graphics mode?
 		jne .text
 
diff -Naur syslinux-1.75.orig/isolinux.asm syslinux-1.75/isolinux.asm
--- syslinux-1.75.orig/isolinux.asm	2002-06-15 06:46:18.000000000 +0200
+++ syslinux-1.75/isolinux.asm	2002-07-18 15:07:44.000000000 +0200
@@ -1105,7 +1105,7 @@
 		push di
                 call unmangle_name              ; Get human form
 		mov si,err_notfound		; Complain about missing kernel
-		call cwritestr
+		call err_cwritestr
 		pop si				; KernelCName
                 call cwritestr
                 mov si,crlf_msg
@@ -1343,7 +1343,7 @@
 		; If this returns, we have problems
 .bad_image:
 		mov si,err_disk_image
-		call cwritestr
+		call err_cwritestr
 		jmp enter_command
 
 ;
@@ -1432,7 +1432,7 @@
                 cli
 		lss sp,[cs:Stack]		; Reset the stack
                 sti
-                call cwritestr                  ; Expects SI -> error msg
+                call err_cwritestr              ; Expects SI -> error msg
 al_ok:          jmp enter_command               ; Return to command prompt
 ;
 ; End of abort_check
@@ -1673,6 +1673,8 @@
 		call write_serial	; write to serial port if needed
 		pushfd
 		pushad
+		test byte [UsingVGA],10h
+		jnz .ret		; really, do not write if VESA enabled
 		mov bh,[TextPage]
 		push ax
                 mov ah,03h              ; Read cursor position
@@ -1845,6 +1847,7 @@
 %include "loadhigh.inc"		; Load a file into high memory
 %include "font.inc"		; VGA font stuff
 %include "graphics.inc"		; VGA graphics
+%include "mdkvesa.inc"		; Mandrake VESA graphics extension
 
 ; -----------------------------------------------------------------------------
 ;  Begin data section
diff -Naur syslinux-1.75.orig/ldlinux.asm syslinux-1.75/ldlinux.asm
--- syslinux-1.75.orig/ldlinux.asm	2002-06-15 06:46:18.000000000 +0200
+++ syslinux-1.75/ldlinux.asm	2002-07-17 19:14:33.000000000 +0200
@@ -1259,7 +1259,7 @@
 		push di
                 call unmangle_name              ; Get human form
 		mov si,err_notfound		; Complain about missing kernel
-		call cwritestr
+		call err_cwritestr
 		pop si				; KernelCName
                 call cwritestr
                 mov si,crlf_msg
@@ -1388,7 +1388,7 @@
                 mov sp,StackBuf-2*3    		; Reset stack
                 mov ss,ax                       ; Just in case...
                 sti
-                call cwritestr                  ; Expects SI -> error msg
+                call err_cwritestr                  ; Expects SI -> error msg
 al_ok:          jmp enter_command               ; Return to command prompt
 ;
 ; End of abort_check
@@ -1480,6 +1480,8 @@
 ;
 writechr:
 		call write_serial	; write to serial port if needed
+		test byte [UsingVGA],10h
+		jnz dir_return		; really, do not write if VESA enabled
 		pushfd
 		pushad
 		mov ah,0Eh
@@ -1496,7 +1498,7 @@
 
 kaboom2:
 		mov si,err_bootfailed
-		call cwritestr
+		call err_cwritestr
 		call getchar
 		call vgaclearmode
 		int 19h			; And try once more to boot...
@@ -1625,6 +1627,7 @@
 %include "loadhigh.inc"		; Load a file into high memory
 %include "font.inc"		; VGA font stuff
 %include "graphics.inc"		; VGA graphics
+%include "mdkvesa.inc"		; Mandrake VESA graphics extension
 
 ; -----------------------------------------------------------------------------
 ;  Begin data section
diff -Naur syslinux-1.75.orig/loadhigh.inc syslinux-1.75/loadhigh.inc
--- syslinux-1.75.orig/loadhigh.inc	2002-06-15 06:46:18.000000000 +0200
+++ syslinux-1.75/loadhigh.inc	2002-07-17 19:45:51.000000000 +0200
@@ -46,6 +46,7 @@
 		push si
 		mov si,dot_msg
 		call cwritestr
+		call progress_bar
 		pop si
 		call abort_check
 
diff -Naur syslinux-1.75.orig/Makefile syslinux-1.75/Makefile
--- syslinux-1.75.orig/Makefile	2002-06-15 06:46:19.000000000 +0200
+++ syslinux-1.75/Makefile	2002-07-17 19:39:26.000000000 +0200
@@ -42,18 +42,18 @@
 	  pxelinux.asm mbr.asm isolinux.asm isolinux-debug.asm
 SOURCES = $(CSRC) $(NASMSRC) *.inc
 BTARGET = kwdhash.gen ldlinux.bss ldlinux.sys ldlinux.bin \
-	  pxelinux.0 mbr.bin isolinux.bin isolinux-debug.bin
-ITARGET = syslinux.com syslinux copybs.com gethostip
+	  pxelinux-graphic.0 mbr.bin isolinux-graphic.bin isolinux-debug-graphic.bin
+ITARGET = syslinux-graphic.com syslinux-graphic copybs.com gethostip
 DOCS    = COPYING NEWS README TODO *.doc sample
 OTHER   = Makefile bin2c.pl now.pl genhash.pl keywords findpatch.pl \
 	  keytab-lilo.pl version sys2ansi.pl ppmtolss16 lss16toppm memdisk
 OBSOLETE = pxelinux.bin
 
 # Things to install in /usr/bin
-INSTALL_BIN   =	syslinux gethostip ppmtolss16 lss16toppm
+INSTALL_BIN   =	syslinux-graphic gethostip ppmtolss16 lss16toppm
 # Things to install in /usr/lib/syslinux
-INSTALL_LIB   =	pxelinux.0 isolinux.bin isolinux-debug.bin \
-		syslinux.com copybs.com memdisk/memdisk
+INSTALL_LIB   =	pxelinux-graphic.0 isolinux-graphic.bin isolinux-debug-graphic.bin \
+		syslinux-graphic.com copybs.com memdisk/memdisk
 
 # The DATE is set on the make command line when building binaries for
 # official release.  Otherwise, substitute a hex string that is pretty much
@@ -92,19 +92,19 @@
 		-dHEXDATE="$(HEXDATE)" \
 		-l pxelinux.lst -o pxelinux.bin pxelinux.asm
 
-isolinux.bin: isolinux.asm kwdhash.gen
+isolinux-graphic.bin: isolinux.asm kwdhash.gen
 	$(NASM) -f bin -dVERSION="'$(VERSION)'" -dDATE_STR="'$(DATE)'" \
 		-dHEXDATE="$(HEXDATE)" \
-		-l isolinux.lst -o isolinux.bin isolinux.asm
+		-l isolinux.lst -o isolinux-graphic.bin isolinux.asm
 
-pxelinux.0: pxelinux.bin
-	cp pxelinux.bin pxelinux.0
+pxelinux-graphic.0: pxelinux.bin
+	cp pxelinux.bin pxelinux-graphic.0
 
 # Special verbose version of isolinux.bin
-isolinux-debug.bin: isolinux-debug.asm kwdhash.gen
+isolinux-debug-graphic.bin: isolinux-debug.asm kwdhash.gen
 	$(NASM) -f bin -dVERSION="'$(VERSION)'" -dDATE_STR="'$(DATE)'" \
 		-dHEXDATE="$(HEXDATE)" \
-		-l isolinux-debug.lst -o isolinux-debug.bin isolinux-debug.asm
+		-l isolinux-debug.lst -o isolinux-debug-graphic.bin isolinux-debug.asm
 
 ldlinux.bss: ldlinux.bin
 	dd if=ldlinux.bin of=ldlinux.bss bs=512 count=1
@@ -118,9 +118,9 @@
 mbr.bin: mbr.asm
 	$(NASM) -f bin -l mbr.lst -o mbr.bin mbr.asm
 
-syslinux.com: syslinux.asm ldlinux.bss ldlinux.sys patch.offset
+syslinux-graphic.com: syslinux.asm ldlinux.bss ldlinux.sys patch.offset
 	$(NASM) -f bin -DPATCH_OFFSET=`cat patch.offset` \
-		-l syslinux.lst -o syslinux.com syslinux.asm
+		-l syslinux.lst -o syslinux-graphic.com syslinux.asm
 
 copybs.com: copybs.asm
 	$(NASM) -f bin -l copybs.lst -o copybs.com copybs.asm
@@ -131,8 +131,8 @@
 ldlinux_bin.c: ldlinux.sys bin2c.pl
 	$(PERL) bin2c.pl ldlinux < ldlinux.sys > ldlinux_bin.c
 
-syslinux: syslinux.o bootsect_bin.o ldlinux_bin.o
-	$(CC) $(LDFLAGS) -o syslinux \
+syslinux-graphic: syslinux.o bootsect_bin.o ldlinux_bin.o
+	$(CC) $(LDFLAGS) -o syslinux-graphic \
 		syslinux.o bootsect_bin.o ldlinux_bin.o
 
 syslinux.o: syslinux.c patch.offset
diff -Naur syslinux-1.75.orig/mdkvesa.inc syslinux-1.75/mdkvesa.inc
--- syslinux-1.75.orig/mdkvesa.inc	1970-01-01 01:00:00.000000000 +0100
+++ syslinux-1.75/mdkvesa.inc	2002-07-17 19:05:31.000000000 +0200
@@ -0,0 +1,427 @@
+;; mdkvesa.inc
+;; -----------------------------------------------------------------------
+;;   
+;;   Copyright 2002 MandrakeSoft (Franois Pons <fpons@mandrakesoft.com>).
+;;   All Rights Reserved
+;;
+;;   This program is free software; you can redistribute it and/or modify
+;;   it under the terms of the GNU General Public License as published by
+;;   the Free Software Foundation, Inc., 53 Temple Place Ste 330,
+;;   Bostom MA 02111-1307, USA; either version 2 of the License, or
+;;   (at your option) any later version; incorporated herein by reference.
+;;
+;; -----------------------------------------------------------------------
+
+; ----------------------------------------------------------------------------
+;  VESA splash screen code
+; ----------------------------------------------------------------------------
+
+;
+; set_mode:
+;	Change graphic mode, support VESA.
+;
+; Input:
+;
+; BX	= video mode (BH not nul => VESA mode)
+;
+set_mode:	push es
+		push ds
+		push cs
+		pop ds
+		push cs
+		pop es
+		test bh,bh
+		jnz set_mode_vesa
+		mov ax,bx			; assume a standard BIOS mode (assume text)
+		int 10h
+		mov byte [UsingVGA],0
+		pop ds
+		pop es
+		jmp adjust_screen		; fall through adjust_screen
+set_mode_vesa:	mov cx,bx
+		mov di,VESA_modeattr
+		mov ax,4F01h			; get video mode information in 500h
+		int 10h
+		mov ax,[VESA_modeattr]
+		and ax,51h			; check for mode supported, tty output, graphics, bank switched supported
+		cmp ax,11h
+		jnz .ret			; hope it is enough
+		mov ax,4F02h			; initialize vesa mode, bx=mode
+		int 10h
+		sub ax,004Fh			; al=4fh if implemented, ah=0 if Ok
+		jnz .ret
+		mov byte [UsingVGA],13h		; 10h for VESA and 3h for serial output (incremented and tested with DisplayMask)
+.ret:		pop ds
+		pop es
+		ret
+
+msg_ext:
+		mov byte [ExtByte],al
+						; Fall through
+msg_ext_multiplex:
+		mov ah,80h			; Start at bit7
+		mov al,[ExtByte]
+msg_ext_mul_1:	test ah,al
+		jnz msg_ext_mul_2
+		shr ah,1
+		jnz msg_ext_mul_1
+		mov bx,msg_putchar
+msg_ext_ret:	mov [NextCharJump],bx
+		ret
+msg_ext_mul_2:	xor [ExtByte],ah		; Clear bit
+		mov bx,msg_ext_set_mode_lo	; set VESA mode
+		cmp ah,80h
+		jz msg_ext_ret
+		mov bx,msg_ext_gotoxy_lo
+		cmp ah,40h
+		jz msg_ext_ret
+		mov bx,msg_ext_progress_y_lo
+		cmp ah,20h
+		jz msg_ext_ret
+		mov bx,msg_ext_seek_y_lo
+		cmp ah,10h
+		jz msg_ext_ret
+		mov bx,msg_ext_ignore
+		mov byte [IgnoreCount],14	; used only by LILO graphic
+		cmp ah,08h
+		jz msg_ext_ret
+		mov bx,msg_ext_putimage
+		cmp ah,04h
+		jz msg_ext_ret
+		mov bx,msg_ext_putrect_w_lo
+		cmp ah,02h
+		jz msg_ext_ret
+		mov bx,msg_ext_putrect_raw
+		cmp ah,01h
+		jz msg_ext_ret
+		mov bx,msg_putchar
+		jmp short msg_ext_ret
+msg_ext_ignore:
+		dec byte [IgnoreCount]
+		jz msg_ext_multiplex
+		ret
+msg_ext_set_mode_lo:
+		mov [ModeLo],al
+		mov word [NextCharJump],msg_ext_set_mode_hi
+		ret
+msg_ext_set_mode_hi:
+		mov bh,al
+		mov bl,[ModeLo]
+		call set_mode
+msg_ext_follow_multiplex:
+		jmp msg_ext_multiplex
+msg_ext_gotoxy_lo:
+		mov [CursorRow],al
+		mov word [NextCharJump],msg_ext_gotoxy_hi
+		ret
+msg_ext_gotoxy_hi:
+		mov [CursorCol],al
+		call msg_gotoxy
+		jmp short msg_ext_follow_multiplex
+msg_ext_progress_y_lo:
+		mov [ProgressY],al
+		mov word [NextCharJump],msg_ext_progress_y_hi
+		ret
+msg_ext_progress_y_hi:
+		mov [ProgressY+1],al
+		mov word [NextCharJump],msg_ext_progress_x_lo
+		ret
+msg_ext_progress_x_lo:
+		mov [ProgressX],al
+		mov word [NextCharJump],msg_ext_progress_x_hi
+		ret
+msg_ext_progress_x_hi:
+		mov [ProgressX+1],al
+		mov word [NextCharJump],msg_ext_progress_yc
+		ret
+msg_ext_progress_yc:
+		mov [ProgressYC],al
+		mov word [NextCharJump],msg_ext_progress_xc
+		ret
+msg_ext_progress_xc:
+		mov [ProgressXC],al
+		mov word [NextCharJump],msg_ext_progress_pixel
+		ret
+msg_ext_progress_pixel:
+		mov [ProgressPixel],al
+		jmp short msg_ext_follow_multiplex
+msg_ext_seek_y_lo:
+		mov byte [ExtY],al
+		mov word [NextCharJump],msg_ext_seek_y_hi
+		ret
+msg_ext_seek_y_hi:
+		mov byte [ExtY+1],al
+		mov word [NextCharJump],msg_ext_seek_x_lo
+		ret
+msg_ext_seek_x_lo:
+		mov byte [ExtX],al
+		mov word [NextCharJump],msg_ext_seek_x_hi
+		ret
+msg_ext_seek_x_hi:
+		mov byte [ExtX+1],al
+msg_ext_follow_multiplex_2:
+		jmp short msg_ext_follow_multiplex
+msg_ext_putrect_w_lo:
+		mov [ExtPutRectW],al
+		mov word [NextCharJump],msg_ext_putrect_w_hi
+		ret
+msg_ext_putrect_w_hi:
+		mov [ExtPutRectW+1],al
+		mov word [NextCharJump],msg_ext_putrect_h_lo
+		ret
+msg_ext_putrect_h_lo:
+		mov [ExtPutRectH],al
+		mov word [NextCharJump],msg_ext_putrect_h_hi
+		ret
+msg_ext_putrect_h_hi:
+		mov [ExtPutRectH+1],al
+		mov word [NextCharJump],msg_ext_putrect_raw
+		ret
+msg_ext_putrect_raw:
+		pusha
+		push es
+		test byte [UsingVGA],10h
+		jz .ret
+		mov es,[VESA_winsegA]
+		mov bx,ax
+		mov cx,[ExtPutRectH]
+.loop_y:
+		mov si,[ExtPutRectW]
+		mov cx,[ExtX]
+.loop_x:
+		mov dx,[ExtY]
+		push cx
+		call putpixel
+		pop cx
+		inc cx
+		dec si
+		jnz .loop_x
+
+		inc word [ExtY]
+		dec word [ExtPutRectH]
+		jnz .loop_y
+
+		mov ax,1			; Restore default value
+		mov [ExtPutRectW],ax
+		mov [ExtPutRectH],ax
+		mov word [NextCharJump],msg_putchar
+.ret:
+		pop es
+		popa
+		ret
+
+putpixel:
+		mov ax,[VESA_byteline]
+		mul dx				; DX=rows, CX=cols, BL=color, only BX saved
+		add ax,cx
+		adc dx,byte 0
+		mov cx,[VESA_wingran]
+		shl cx,10
+		jz .64k_page
+		div cx
+		xchg ax,dx
+.64k_page:
+		cmp dx,[ExtPage]
+		jz .put_into_page
+		push bx
+		push ax
+		mov [ExtPage],dx
+		xor bx,bx
+		mov ax,4F05h
+		int 10h
+		pop ax
+		pop bx
+.put_into_page:
+		mov di,ax
+		mov ax,bx
+		stosb
+		ret
+
+msg_ext_putimage:
+		mov word [ExtImagePixel],0
+msg_ext_putimage_w_lo:
+		mov [ExtImageW],al
+		mov word [NextCharJump],msg_ext_putimage_w_hi
+		ret
+msg_ext_putimage_w_hi:
+		mov [ExtImageW+1],al
+		mov word [NextCharJump],msg_ext_putimage_h_lo
+		ret
+msg_ext_putimage_h_lo:
+		mov [ExtImageH],al
+		mov word [NextCharJump],msg_ext_putimage_h_hi
+		ret
+msg_ext_putimage_h_hi:
+		mov [ExtImageH+1],al
+		mov word [NextCharJump],msg_ext_putimage_c
+		ret
+msg_ext_putimage_c:
+		xor ah,ah
+		mov [ExtImageC],ax
+		mov word [NextCharJump],msg_ext_putimage_cy
+		ret
+msg_ext_putimage_cy:
+		xor ah,ah
+		mov [ExtImageCY],ax
+		mov word [NextCharJump],msg_ext_putimage_pal_r
+		ret
+msg_ext_putimage_pal_index:
+		mov bx,[ExtImagePixel]
+		mov cx,bx
+		shl bx,1
+		add bx,cx
+		shr al,2
+		ret
+msg_ext_putimage_pal_r:
+		call msg_ext_putimage_pal_index
+		mov [bx+ExtImagePal],al
+		mov word [NextCharJump],msg_ext_putimage_pal_g
+		ret
+msg_ext_putimage_pal_g:
+		call msg_ext_putimage_pal_index
+		mov [bx+1+ExtImagePal],al
+		mov word [NextCharJump],msg_ext_putimage_pal_b
+		ret
+msg_ext_putimage_pal_b:
+		call msg_ext_putimage_pal_index
+		mov [bx+2+ExtImagePal],al
+		mov bx,[ExtImagePixel]
+		inc bx
+		mov [ExtImagePixel],bx
+		cmp bx,[ExtImageC]
+		jb msg_ext_putimage_pal_loop
+		call msg_ext_putimage_set_colormap
+		xor ax,ax
+		mov [ExtImageX],ax
+		mov [ExtImageY],ax
+		or ax,[ExtImageCY]
+		jnz msg_ext_putimage_pal_cy
+msg_ext_putimage_cy_raw:
+		mov word [NextCharJump],msg_ext_putimage_raw
+		ret
+msg_ext_putimage_pal_loop:
+		mov word [NextCharJump],msg_ext_putimage_pal_r
+		ret
+msg_ext_putimage_pal_cy:
+		mov word [ExtImagePixel],0
+		mov word [NextCharJump],msg_ext_putimage_cy_loop
+		ret
+msg_ext_putimage_cy_loop:
+		mov bx,[ExtImagePixel]
+		inc bx
+		mov [ExtImagePixel],bx
+		cmp bx,[ExtImageCY]
+		jae msg_ext_putimage_cy_raw
+		ret
+progress_bar:	pushf
+		cs test byte [UsingVGA],10h
+		jz .ret
+		pushad
+		push es
+		mov es,[VESA_winsegA]
+		mov bl,[ProgressPixel]
+		mov cx,[ProgressX]
+		mov ah,[ProgressXC]
+.enlarge_x:	mov dx,[ProgressY]
+		mov al,[ProgressYC]
+.enlarge_y:	push ax
+		push cx
+		push dx
+		call putpixel
+		pop dx
+		pop cx
+		pop ax
+		inc dx
+		dec al
+		jnz .enlarge_y
+		inc cx
+		dec ah
+		jnz .enlarge_x
+		mov [ProgressX],cx
+		pop es
+		popad
+.ret
+		popf
+msg_ext_putimage_ret:
+		ret
+
+msg_ext_putimage_set_colormap:
+		test byte [UsingVGA],10h	; Silence if not VESA
+		jz msg_ext_putimage_ret
+		mov ax,1012h
+		mov bx,0040h
+		mov cx,[ExtImageC]
+		mov dx,ExtImagePal
+		push es
+		push cs
+		pop es
+		int 10h
+		pop es
+		ret
+msg_ext_putimage_raw:
+		test byte [UsingVGA],10h	; Silence if not VESA
+		jz msg_ext_putimage_ret
+		pusha
+		push es
+		mov es,[VESA_winsegA]
+		test al,80h
+		jnz msg_ext_putimage_raw_dup
+		mov [ExtImagePixel],al		; Save current pixel for looping
+		call msg_ext_putimage_raw_pixel
+		pop es
+		popa
+		ret
+msg_ext_putimage_raw_pixel:
+		mov cx,[ExtImageX]
+		mov dx,[ExtImageY]
+		add cx,[ExtX]
+		add dx,[ExtY]
+		mov bl,[ExtImagePixel]
+		add bl,40h
+		call putpixel
+		mov ax,[ExtImageX]
+		inc ax
+		cmp ax,[ExtImageW]
+		jb .cont
+		mov ax,[ExtImageY]
+		inc ax
+		mov [ExtImageY],ax
+		cmp ax,[ExtImageH]
+		jb .newline
+		mov word [NextCharJump],msg_putchar
+		ret
+.newline:
+		xor ax,ax
+.cont:
+		mov [ExtImageX],ax
+		ret
+msg_ext_putimage_raw_dup:
+		xor ah,ah
+		add al,82h
+.loop:
+		push ax
+		call msg_ext_putimage_raw_pixel
+		pop ax
+		cmp word [NextCharJump],msg_putchar
+		jnz .doloop
+		mov ax,0001h			; Avoid writing more than available image zone
+.doloop:
+		dec ax
+		jnz .loop
+		pop es
+		popa
+		ret
+
+ModeLo		db 0			; For setting VESA mode from message file
+ProgressX	dw 0			; Current Progress X position
+ProgressY	dw 0			; Current Progress Y position
+ProgressXC	db 2			; Current Progress X pixel Count
+ProgressYC	db 5			; Current Progress Y pixel Count
+ProgressPixel	db 0Fh			; Pixel color for progress bar
+IgnoreCount	db 0
+ExtByte		db 0			; Extended code
+ExtY		dw 0			; Extended Y
+ExtX		dw 0			; Extended X
+ExtPutRectW	dw 1			; Extended putrect Width
+ExtPutRectH	dw 1			; Extended putrect Height
+ExtPage		dw -1			; Extended page for high resolution
diff -Naur syslinux-1.75.orig/pxelinux.asm syslinux-1.75/pxelinux.asm
--- syslinux-1.75.orig/pxelinux.asm	2002-06-15 06:46:18.000000000 +0200
+++ syslinux-1.75/pxelinux.asm	2002-07-18 15:10:33.000000000 +0200
@@ -351,7 +351,7 @@
 		jnc have_pxenv
 
 no_pxe:		mov si,err_nopxe
-		call writestr
+		call err_cwritestr
 		jmp kaboom
 
 have_pxenv:
@@ -525,7 +525,7 @@
 
 .pxe_err:
 		mov si,err_pxefailed
-		call writestr
+		call err_cwritestr
 		call writehex4
 		mov al, ' '
 		call writechr
@@ -612,7 +612,7 @@
 		cmp word [pxe_udp_open_pkt.status], byte 0
 		je .success
 .failed:	mov si,err_udpinit
-		call writestr
+		call err_cwritestr
 		jmp kaboom
 .success:
 
@@ -997,7 +997,7 @@
 		push di
                 call unmangle_name              ; Get human form
 		mov si,err_notfound		; Complain about missing kernel
-		call cwritestr
+		call err_cwritestr
 		pop si				; KernelCName
                 call cwritestr
                 mov si,crlf_msg
@@ -1201,7 +1201,7 @@
                 cli
 		lss sp,[cs:Stack]		; Reset the stack
                 sti
-                call cwritestr                  ; Expects SI -> error msg
+                call err_cwritestr              ; Expects SI -> error msg
 al_ok:          jmp enter_command               ; Return to command prompt
 ;
 ; End of abort_check
@@ -1556,7 +1556,7 @@
 		call far [PXENVEntry]
 
 .no_tsize:	mov si,err_oldtftp
-		call writestr
+		call err_cwritestr
 		jmp kaboom
 
 .bailnow:	add sp,byte 8		; Immediate error - no retry
@@ -1637,6 +1637,8 @@
 		call write_serial	; write to serial port if needed
 		pushfd
 		pushad
+		test byte [UsingVGA],10h
+		jnz .ret		; really, do not write if VESA enabled
 		mov bh,[TextPage]
 		push ax
                 mov ah,03h              ; Read cursor position
@@ -2275,6 +2277,7 @@
 %include "loadhigh.inc"		; Load a file into high memory
 %include "font.inc"		; VGA font stuff
 %include "graphics.inc"		; VGA graphics
+%include "mdkvesa.inc"		; Mandrake VESA graphics extension
 
 ; -----------------------------------------------------------------------------
 ;  Begin data section
diff -Naur syslinux-1.75.orig/README.graphic syslinux-1.75/README.graphic
--- syslinux-1.75.orig/README.graphic	1970-01-01 01:00:00.000000000 +0100
+++ syslinux-1.75/README.graphic	2002-07-17 15:32:03.000000000 +0200
@@ -0,0 +1,233 @@
+Managing VESA graphic mode with syslinux.
+
+SYSLINUX-GRAPHIC is a modification of SYSLINUX for handling VESA graphic
+mode on boot.
+
+SYSLINUX is Copyright 1994-2002 H. Peter Anvin.
+
+******************************************************************************
+Support of graphic mode is made through escape code for changing video
+mode, inserting image and defining progression bar environment.
+
+Currently, changing text color is made with ^O escape char (0x0F) followed with
+hexadecimal background code and foreground code.
+
+graphic extension of SYSLINUX-GRAPHIC are available with ^N escape char (0x0E)
+followed by a byte containing a bit field for using specific extension.
+Just OR as many as you want of this value and include appropriate extension code:
+   0x80: VESA mode follow, first low byte followed by high byte. For
+         example 0x03 0x01 for 0x103 (VESA 800x600x256 colors).
+	 Other are 0x101 for 640x480, 0x105 for 1024x768 and 0x107 for
+	 1280x1024.
+         Only bankswitched 256 colors are supported for image.
+   0x40: gotoxy, first Row coordinate followed by Col coordinate.
+         ignored by lilo-graphic.
+   0x20: progress bar indication: lo byte of Y, hi byte of Y, lo byte of X,
+         hi byte of X, Height (one byte only), Width (one byte only) and
+         finally color of pixel.
+         ignored by lilo-graphic.
+   0x10: seek Y and X: lo byte of Y, hi byte of Y, lo byte of X, hi byte of X.
+   0x08: IGNORED to match lilo-graphic image format,
+         13 following bytes are not used.
+   0x04: putimage, builded by bmp2mdk (escape 0x0E and 0x04 are already in the
+         beginning of the file produced)
+   0x02: putrect: lo byte of Width, hi byte of Width, lo byte of Height, hi
+         byte of Height, color of pixel to draw.
+   0x01: putpixel: color of pixel.
+
+Image, rect and pixel are drawn from current Y and X, so seek to right position
+before. Image has 128 colors maximun reindexed from 0x40 to 0xBF and use a
+palette by default. Other colors are not modified.
+
+When VESA mode has been accepted, only graphic extension are drawn. If VESA
+mode has not been accepted on modified, only characters (and color extension,
+CR/LF) are printed. This means if VESA mode failed, it is still possible to
+print something on the screen, it can be done by concatening a graphic msg
+file with a text version and the right one will be printed.
+
+Since color palette is modifed whenever an image is drawn (unless there are no
+palette in it but the script at the end of this file only manage this such
+image), it is preferable to draw only one image for a screen.
+
+If SYSLINUX encounters an error when loading initrd or kernel, text mode 80x25
+is activated before printing the error message.
+
+If you have any questions/enhancements, please send mail
+to <fpons@mandrakesoft.com>.
+
+François Pons.
+
+******************************************************************************
+#!/usr/bin/perl
+
+#- Build a message file for syslinux-graphic from a BMP file.
+#- Copyright (C) 1999,200,2001,2002 MandrakeSoft (pixel@mandrakesoft.com,
+#- fpons@mandrakesoft.com).
+#-
+#- This program is free software; you can redistribute it and/or modify
+#- it under the terms of the GNU General Public License as published by
+#- the Free Software Foundation; either version 2, or (at your option)
+#- any later version.
+#-
+#- This program is distributed in the hope that it will be useful,
+#- but WITHOUT ANY WARRANTY; without even the implied warranty of
+#- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#- GNU General Public License for more details.
+#-
+#- You should have received a copy of the GNU General Public License
+#- along with this program; if not, write to the Free Software
+#- Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+
+#- usage:
+#-   bmp2mdk <picture.bmp >picture.msg
+#- NB: the input image (colormap indexed) should not have more than 128 colors
+#- and should not be compressed.
+
+#- additional modification, to include some escape sequence for the picture
+#- commented as follow:
+#-   mode:value                   - VESA mode to use, assume graphic if > 0x100,
+#-				    default is 0x101 (ie 640x480 in 256 colors).
+#-   timer:y,x,bg,fg              - timer position and colors (for LILO only).
+#-   entry:y,x,bg,fg,h_chr,w_chr  - entry position, colors and size in chars
+#-				    (for LILO only).
+#-   progress:y,x,h,w,c           - progress bar position, size (h and w in byte
+#-				    only) and color (for SYSLINUX only).
+#-   clear:h,w,c                  - clear before with color c, default is clear
+#-				    640x480 in color 0 (black).
+#- no options assume no addition initcode given.
+#- colors given in the range 0x40 and 0xBF are from picture palette used (shifted).
+
+use strict;
+
+my $initcode_str;
+my %initcode = ( #- this default are running quite nicely with first english LILO image.
+		mode        => 0x101,
+		timer_y     => 0,
+		timer_x     => 0,
+		timer_bg    => 0x00,
+		timer_fg    => 0x0f,
+		entry_y     => 0,
+		entry_x     => 0,
+		entry_bg    => 0x00,
+		entry_fg    => 0x0f,
+		entry_h_chr => 24,
+		entry_w_chr => 22,
+	        progress_y  => 0,
+                progress_x  => 0,
+	        progress_h  => 8,
+		progress_w  => 32,
+		progress_c  => 0x01,
+		clear_w     => 640,
+		clear_h     => 480,
+		clear_color => 0,
+		pos_y       => 0,
+		pos_x       => 0,
+	       );
+foreach (@ARGV) {
+    /^mode:(.*)/  and do { @initcode{qw(mode)} = map { eval "$_" } split /,/, $1; next };
+    /^timer:(.*)/ and do { @initcode{qw(timer_y timer_x timer_bg timer_fg)} = map { eval "$_" } split /,/, $1; next };
+    /^entry:(.*)/ and do { @initcode{qw(entry_y entry_x entry_bg entry_fg entry_h_chr entry_w_chr)} = map { eval "$_" } split /,/, $1; next };
+    /^progress:(.*)/ and do { @initcode{qw(progress_y progress_x progress_h progress_w progress_c)} = map { eval "$_" } split /,/, $1; next };
+    /^clear:(.*)/ and do { @initcode{qw(clear_h clear_w clear_color)} = map { eval "$_" } split /,/, $1; next };
+    /^pos:(.*)/   and do { @initcode{qw(pos_y pos_x)} = map { eval "$_" } split /,/, $1; next };
+    /^initcode$/  and next;
+    die "usage: bmp2mdk [options]
+where [options] are taken from the following list:
+   mode:value                   - VESA mode to use, assume graphic if > 0x100,
+                                  default is 0x101 (ie 640x480 in 256 colors).
+   timer:y,x,bg,fg              - timer position and colors (for LILO only).
+   entry:y,x,bg,fg,h_chr,w_chr  - entry position, colors and size in chars
+                                  (for LILO only).
+   progress:y,x,h,w,c           - progress bar position, size (h and w in byte
+                                  only) and color (for SYSLINUX only).
+   clear:h,w,c                  - clear before with color c, default is clear
+                                  640x480 in color 0 (black).
+   pos:y,x                      - set x,y for next pixmap, default is to (0,0).
+no options assume no addition initcode given.
+colors given in the range 0x40 and 0xBF are from picture palette used (shifted).
+input image should be a 128 colors uncompressed BMP file.
+
+unknown option [$_]\n";
+}
+if (@ARGV) {
+    my ($ext_code, $ext_data) = (0x00, '');
+
+    if ($initcode{mode}) {
+	$ext_code |= 0x80;
+	$ext_data .= pack "v", $initcode{mode};
+    }
+    if ($initcode{timer_y} || $initcode{timer_x} || $initcode{entry_y} || $initcode{entry_x}) {
+	$ext_code |= 0x08;
+	$ext_data .= pack "v2C2 v2C4", @initcode{qw(timer_y timer_x timer_bg timer_fg
+						    entry_y entry_x entry_bg entry_fg entry_h_chr entry_w_chr)};
+    }
+    if ($initcode{progress_y} || $initcode{progress_x}) {
+	$ext_code |= 0x20;
+	$ext_data .= pack "v2C3", @initcode{qw(progress_y progress_x progress_h progress_w progress_c)};
+    }
+    #- always set up clear.
+    $ext_code |= 0x02;
+    $ext_data .= pack "v2C", @initcode{qw(clear_w clear_h clear_color)};
+
+    print "\x0e";
+    print pack "C", $ext_code;
+    print $ext_data;
+
+    #- always set up pos too but in separate coded.
+    print "\x0e\x10";
+    print pack "v2", @initcode{qw(pos_y pos_x)};
+}
+
+undef $/;
+my $f = <STDIN>;
+
+my ($width, $height, $nbcolors) = unpack "x18 V V x20 V", $f;
+
+my $s = $nbcolors * 4;
+my ($palette, $image) = unpack "x54 a$s a*", $f;
+my @palette = unpack("C3x" x $nbcolors, $palette);
+
+for (my $i = 0; $i < 3 * $nbcolors; $i += 3) {
+    ($palette[$i], $palette[$i + 2]) = ($palette[$i + 2], $palette[$i]);
+}
+
+print "\x0e\x04";
+print pack "v v C C" . "C3" x $nbcolors, $width, $height, $nbcolors, 0, @palette;
+
+my $width_ = 4 * int (($width + 3) / 4);
+my ($oldv, $nb) = (-1024, 0);
+
+for (my $i = 0; $i < $height; $i++) {
+    for (my $j = 0; $j < $width; $j++) {
+	my $v = vec $image, ($height - $i - 1) * $width_+ $j, 8;
+	$v >= 128 and die "cannot use more than 128 colors in input file\n";
+
+	if ($v == $oldv) {
+	    $nb++;
+	} else {
+	    while ($nb > 129) {
+		print pack "C", 0xFF;
+		$nb -= 129;
+	    }
+	    if ($nb > 1) { #- necessary <= 129 here
+		print pack "C", (0x80 | ($nb - 2));
+	    } elsif ($nb == 1) {
+		print pack "C", $oldv;
+	    }
+	    print pack "C", $v;
+	    $oldv = $v;
+	    $nb = 0;
+	}
+    }
+}
+while ($nb > 129) {
+    print pack "C", 0xFF;
+    $nb -= 129;
+}
+if ($nb > 1) { #- necessary <= 129 here
+    print pack "C", (0x80 | ($nb - 2));
+} elsif ($nb == 1) {
+    print pack "C", $oldv;
+}
+
+1;
diff -Naur syslinux-1.75.orig/runkernel.inc syslinux-1.75/runkernel.inc
--- syslinux-1.75.orig/runkernel.inc	2002-06-15 06:46:18.000000000 +0200
+++ syslinux-1.75/runkernel.inc	2002-07-17 16:56:41.000000000 +0200
@@ -321,6 +321,7 @@
                 call cwritestr                  ; "Loading" message
                 mov si,dotdot_msg		; Print dots
                 call cwritestr
+		call progress_bar
 
                 mov eax,[HighMemSize]
 		sub eax,100000h			; Load address
@@ -344,6 +345,7 @@
 
                 mov si,dot_msg			; Progress report
                 call cwritestr
+		call progress_bar
                 call abort_check
 
 		pop ecx				; Number of bytes in the initial portion
@@ -360,6 +362,7 @@
 
                 mov si,dot_msg
                 call cwritestr
+		call progress_bar
 
 ;
 ; Now see if we have an initial RAMdisk; if so, do requisite computation
@@ -400,7 +403,7 @@
 
 initrd_notthere:
                 mov si,err_noinitrd
-                call cwritestr
+                call err_cwritestr
                 mov si,InitRDCName
                 call cwritestr
                 mov si,crlf_msg
@@ -551,6 +554,17 @@
                 int 16h
 %endif
 ;
+; Restore video mode to 80x25 if using VESA.
+;
+		test byte [UsingVGA],0x10	; VESA used? need to restore 80x25 for kernel boot
+		jz setup_real_mode_seg
+		pusha
+		mov ax,0003h
+		int 10h
+		popa
+
+setup_real_mode_seg:
+;
 ; Set up segment registers and the Linux real-mode stack
 ; Note: es == the real mode segment
 ;
@@ -605,6 +619,7 @@
                 call cwritestr
                 mov si,dotdot_msg		; Write dots
                 call cwritestr
+		call progress_bar
 		pop si
 
 		mov eax,[es:su_ramdisklen]
diff -Naur syslinux-1.75.orig/writestr.inc syslinux-1.75/writestr.inc
--- syslinux-1.75.orig/writestr.inc	2002-06-15 06:46:18.000000000 +0200
+++ syslinux-1.75/writestr.inc	2002-07-18 15:06:08.000000000 +0200
@@ -17,6 +17,19 @@
 ;; Code to write a simple string.
 ;;
 
+; err_cwritestr: write a null-terminated error string to the console, saving
+;                registers on entry, restore text mode if using VESA
+;
+err_cwritestr:	pushfd
+		pushad
+		test byte [UsingVGA],10h
+		jz cwritestr.top
+		mov bx,3
+		call set_mode
+		popad
+		pushad
+		jmp cwritestr.top
+
 ;
 ; crlf: Print a newline
 ;
@@ -32,6 +45,8 @@
 cwritestr:
 		pushfd
                 pushad
+		test byte [UsingVGA],10h
+		jnz .end
 .top:		lodsb
 		and al,al
                 jz .end
