--- syslinux-3.31/com32/lib/Makefile.syspng	2006-11-17 10:16:16.000000000 -0500
+++ syslinux-3.31/com32/lib/Makefile	2006-11-17 10:20:00.000000000 -0500
@@ -52,11 +52,7 @@
 	zlib/uncompr.o zlib/deflate.o zlib/trees.o zlib/zutil.o 	\
 	zlib/inflate.o zlib/infback.o zlib/inftrees.o zlib/inffast.o	\
 	\
-	libpng/png.o libpng/pngset.o libpng/pngget.o libpng/pngrutil.o  \
-	libpng/pngtrans.o libpng/pngwutil.o libpng/pngread.o		\
-	libpng/pngrio.o libpng/pngwio.o libpng/pngwrite.o		\
-	libpng/pngrtran.o libpng/pngwtran.o libpng/pngmem.o		\
-	libpng/pngerror.o libpng/pngpread.o				\
+	$(LIBDIR)/libpng.a						\
 	\
 	jpeg/tinyjpeg.o jpeg/jidctflt.o	jpeg/decode1.o jpeg/decode3.o   \
 	jpeg/rgb24.o jpeg/bgr24.o jpeg/yuv420p.o jpeg/grey.o		\
