--- at-3.1.8/at.1.in.shell	Wed Dec 13 11:50:23 2000
+++ at-3.1.8/at.1.in	Wed Dec 13 11:55:02 2000
@@ -1,4 +1,4 @@
-.Id $Id: at.1.in,v 1.8 1997/09/28 20:00:25 ig25 Exp $
+.Id $Id: at.1.in,v 1.8 1997/09/28 20:00:25 ig25 Exp 
 .TH AT 1 "Nov 1996" local "Linux Programmer's Manual"
 .SH NAME
 at, batch, atq, atrm \- queue, examine or delete jobs for later execution
@@ -39,7 +39,10 @@
 and
 .B batch
 read commands from standard input or a specified file which are to
-be executed at a later time, using
+be executed at a later time, using the shell set by the user's environment
+variable
+.BR SHELL,
+the user's login shell, or ultimately
 .BR /bin/sh .
 .TP 8
 .BR at
--- at-3.1.8/at.c.shell	Wed Dec 13 12:45:55 2000
+++ at-3.1.8/at.c	Wed Dec 13 12:46:25 2000
@@ -854,7 +854,7 @@
 	   It also alows a warning diagnostic to be printed.  Because of the
 	   possible variance, we always output the diagnostic. */
 
-	fprintf(stderr, "warning: commands will be executed using /bin/sh\n");
+	fprintf(stderr, "warning: commands will be executed using (in order) a) $SHELL b) login shell c) /bin/sh\n");
 
 	writefile(timer, queue);
 	break;
