--- at-3.1.8/Makefile.in.noroot	Thu Jan 17 12:30:30 2002
+++ at-3.1.8/Makefile.in	Thu Jan 17 12:33:21 2002
@@ -100,9 +100,9 @@
 	$(INSTALL) -s atrun $(IROOT)$(sbindir)
 	$(INSTALL) -m 644 at.1 $(IROOT)$(man1dir)/
 	cd $(IROOT)$(man1dir) && $(LN_S) -f at.1 atq.1 && $(LN_S) -f at.1 batch.1 && $(LN_S) -f at.1 atrm.1
-	$(INSTALL) -g root -o root -m 644 atd.8 $(IROOT)$(man8dir)/
-	$(INSTALL) -g root -o root -m 644 atrun.8 $(IROOT)$(man8dir)/
-	$(INSTALL) -g root -o root -m 644 at_allow.5 $(IROOT)$(man5dir)/
+	$(INSTALL) -m 644 atd.8 $(IROOT)$(man8dir)/
+	$(INSTALL) -m 644 atrun.8 $(IROOT)$(man8dir)/
+	$(INSTALL) -m 644 at_allow.5 $(IROOT)$(man5dir)/
 	cd $(IROOT)$(man5dir) && $(LN_S) -f at_allow.5 at_deny.5 
 
 dist: checkin $(DIST) $(LIST) Filelist.asc
