--- tcsh-6.12.00/tc.who.c.utmp	2002-07-01 23:12:04.000000000 +0200
+++ tcsh-6.12.00/tc.who.c	2002-08-20 18:55:09.000000000 +0200
@@ -163,12 +163,13 @@ void
 watch_login(force)
     int force;
 {
-    int     utmpfd, comp = -1, alldone;
+    int     comp = -1, alldone;
     int	    firsttime = stlast == 1;
 #ifdef BSDSIGS
     sigmask_t omask;
 #endif				/* BSDSIGS */
     struct utmp utmp;
+    struct utmp *uptr;
     struct who *wp, *wpnew;
     struct varent *v;
     Char  **vp = NULL;
@@ -261,18 +262,9 @@ watch_login(force)
 	return;
     }
     stlast = sta.st_mtime;
-    if ((utmpfd = open(_PATH_UTMP, O_RDONLY)) < 0) {
-	if (!force)
-	    xprintf(CGETS(26, 2,
-			  "%s cannot be opened.  Please \"unset watch\".\n"),
-		    _PATH_UTMP);
-# ifdef BSDSIGS
-	(void) sigsetmask(omask);
-# else
-	(void) sigrelse(SIGINT);
-# endif
-	return;
-    }
+
+    utmpname( _PATH_UTMP );
+    setutent();
 
     /*
      * xterm clears the entire utmp entry - mark everyone on the status list
@@ -287,7 +279,9 @@ watch_login(force)
      * Read in the utmp file, sort the entries, and update existing entries or
      * add new entries to the status list.
      */
-    while (read(utmpfd, (char *) &utmp, sizeof utmp) == sizeof utmp) {
+
+    while ((uptr = getutent())) {
+	memcpy(&utmp, uptr, sizeof(utmp));
 
 # ifdef DEAD_PROCESS
 #  ifndef IRIS4D
@@ -383,7 +377,7 @@ watch_login(force)
 	    wp->who_prev = wpnew;	/* linked in now */
 	}
     }
-    (void) close(utmpfd);
+    endutent();
 # if defined(UTHOST) && defined(_SEQUENT_)
     endutent();
 # endif
