--- man-1.5j/configure.utf8	Thu Nov 22 15:02:35 2001
+++ man-1.5j/configure	Thu Nov 22 15:03:15 2001
@@ -336,8 +336,8 @@
   Fgawk=/bin/gawk
   Fawk=/bin/gawk
   Fmawk=missing
-  troff=""
-  nroff=""
+  troff=/usr/bin/troff
+  nroff=/usr/bin/nroff
   jnroff=""
   eqn=""
   neqn=""
@@ -380,6 +380,7 @@
     done
   done
   Fgroff=/usr/bin/groff
+  Fnroff=/usr/bin/nroff
   Fgeqn=/usr/bin/geqn
   Fgtbl=/usr/bin/gtbl
   Fcol=/usr/bin/col
