--- ocaml-3.06/configure.lib64	2003-07-24 06:46:09.000000000 -0400
+++ ocaml-3.06/configure	2003-07-24 08:12:39.000000000 -0400
@@ -221,6 +221,34 @@ EOF
     esac;;
 esac
 
+# Determine library dir name
+
+case $host in
+*-*-linux*)
+  # Test if the compiler is 64bit
+  echo 'int i;' > conftest.c
+  ocaml_cv_cc_64bit_output=no
+  if $cc -c conftest.c; then
+    case `/usr/bin/file conftest.o` in
+    *"ELF 64"*)
+      ocaml_cv_cc_64bit_output=yes
+      ;;
+    esac
+  fi
+  rm -rf conftest*
+  ;;
+esac
+
+host_cpu=`uname -m`
+case $host_cpu:$ocaml_cv_cc_64bit_output in
+powerpc64:yes | s390x:yes | sparc64:yes | x86_64:yes)
+  lib="lib64"
+  ;;
+*:*)
+  lib="lib"
+  ;;
+esac
+
 # Configure the bytecode compiler
 
 bytecc="$cc"
@@ -1100,7 +1128,7 @@ done
 if test "$x11_include" = "not found"; then
   x11_try_lib_dir=''
 else
-  x11_try_lib_dir=`echo $x11_include | sed -e 's|include|lib|'`
+  x11_try_lib_dir=`echo $x11_include | sed -e "s|include|$lib|"`
 fi
 
 for dir in \
