--- dietlibc-0.22/include/assert.h.assert-fix	2002-12-20 13:06:51.000000000 -0500
+++ dietlibc-0.22/include/assert.h	2003-02-20 05:51:43.000000000 -0500
@@ -24,7 +24,7 @@ extern void __assert_fail (const char *_
 #else
 #ifdef expect
 # define assert(expr)							      \
-  ((void) (expect((int)expr,0) ? 0 :					      \
+  ((void) (expect((long)(expr),0) ? 0 :					      \
 	   (__assert_fail (#expr,				      \
 			   __FILE__, __LINE__, __ASSERT_FUNCTION), 0)))
 #else
