--- byacc-1.9/defs.h.includes	1992-10-12 02:27:49.000000000 +0100
+++ byacc-1.9/defs.h	2004-07-01 14:06:29.756267522 +0200
@@ -276,14 +276,9 @@ extern bucket *make_bucket();
 
 
 /* system variables */
-
-extern int errno;
+#include <errno.h>
 
 
 /* system functions */
-
-extern void free();
-extern char *calloc();
-extern char *malloc();
-extern char *realloc();
-extern char *strcpy();
+#include <stdlib.h>
+#include <string.h>
--- byacc-1.9/main.c.includes	2004-07-01 14:05:56.362536370 +0200
+++ byacc-1.9/main.c	2004-07-01 14:05:56.365536346 +0200
@@ -1,4 +1,7 @@
 #include <signal.h>
+#include <unistd.h>
+#include <string.h>
+#include <stdlib.h>
 #include "defs.h"
 
 char dflag;
--- byacc-1.9/output.c.includes	1993-02-22 07:32:46.000000000 +0100
+++ byacc-1.9/output.c	2004-07-01 14:05:56.366536338 +0200
@@ -1,3 +1,4 @@
+#include <string.h>
 #include "defs.h"
 
 static int nvectors;
--- byacc-1.9/reader.c.includes	1993-02-22 07:32:46.000000000 +0100
+++ byacc-1.9/reader.c	2004-07-01 14:05:56.367536330 +0200
@@ -1,3 +1,4 @@
+#include <string.h>
 #include "defs.h"
 
 /*  The line size must be a positive integer.  One hundred was chosen	*/
--- byacc-1.9/symtab.c.includes	1990-02-05 07:03:20.000000000 +0100
+++ byacc-1.9/symtab.c	2004-07-01 14:05:56.368536322 +0200
@@ -1,3 +1,4 @@
+#include <string.h>
 #include "defs.h"
 
 
