--- rpm-4.0.3/lib/tagName.c.pix	Fri Jun 29 00:12:06 2001
+++ rpm-4.0.3/lib/tagName.c	Sat Sep 15 22:24:45 2001
@@ -11,7 +11,7 @@
 {
     int i;
     static char nameBuf[128];	/* XXX yuk */
-    char *s;
+    char *s, *locale;
 
     switch (tag) {
     case RPMDBI_PACKAGES:
@@ -36,6 +36,11 @@
 	/*@notreached@*/ break;
     }
 
+    /* Turkic locales have tolower('I') != 'i' */
+    locale = setlocale(LC_ALL, NULL);
+    if (locale) locale = strdupa(locale);
+    setlocale(LC_ALL, "C");
+    
     strcpy(nameBuf, "(unknown)");
     for (i = 0; i < rpmTagTableSize; i++) {
 	if (tag != rpmTagTable[i].val)
@@ -47,12 +52,17 @@
 	    *s = xtolower(*s);
 	break;
     }
+    
+    /* restore back the locale setting */
+    setlocale(LC_ALL, locale);
+    
     return nameBuf;
 }
 
 int tagValue(const char * tagstr)
 {
     const struct headerTagTableEntry_s *t;
+    char *locale;
 
     if (!strcmp(tagstr, "Packages"))
 	return RPMDBI_PACKAGES;
@@ -65,9 +75,19 @@
     if (!strcmp(tagstr, "Available"))
 	return RPMDBI_AVAILABLE;
 
+    /* Turkic locales have tolower('I') != 'i' */
+    locale = setlocale(LC_ALL, NULL);
+    if (locale) locale = strdupa(locale);
+    setlocale(LC_ALL, "C");
+
     for (t = rpmTagTable; t->name != NULL; t++) {
-	if (!xstrcasecmp(t->name + 7, tagstr))
+	if (!xstrcasecmp(t->name + 7, tagstr)) {
+	    /* restore back the locale setting */
+	    setlocale(LC_ALL, locale);
 	    return t->val;
+	}
     }
+    /* restore back the locale setting */
+    setlocale(LC_ALL, locale);
     return -1;
 }
