--- rpm-4.2/scripts/find-lang.sh.find-lang_all-name	2002-06-28 16:48:28.000000000 +0200
+++ rpm-4.2/scripts/find-lang.sh	2003-04-28 11:56:03.000000000 +0200
@@ -12,6 +12,18 @@
 #work based on this file.
 
 #changes:
+# $Id: rpm-find-lang,v 1.3 2001/03/05 21:43:41 misiek Exp $
+# 2001-01-08 Micha? Kochanowicz <mkochano@pld.org.pl>
+#   * --all-name support for KDE.
+# 2000-11-28 Rafa? Cygnarowski <pascalek@pld.org.pl>
+#   * next simple rule for KDE
+# 2000-11-12 Rafa? Cygnarowski <pascalek@pld.org.pl>
+#   * simple rules for KDE help files
+# 2000-06-05 Micha? Kochanowicz <mkochano@pld.org.pl>
+#   * exact, not substring matching $NAME, i.e. find-lang top_dir NAME will
+#     no longer find /usr/share/locale/pl/LC_MESSAGES/<anything>NAME.mo.
+# 2000-04-17 Arkadiusz Mi?kiewicz <misiek@pld.org.pl>
+#   * exit 1 when no files found
 # 1999-10-19 Artur Frysiak <wiget@pld.org.pl>
 #   * added support for GNOME help files
 #   * start support for KDE help files
@@ -32,6 +44,7 @@
   --with-gnome		find GNOME help files
   --with-kde		find KDE help files (not implemented yet)
   --without-mo		not find locales files
+  --all-name		find all localized files
 EOF
 exit 1
 }
@@ -54,7 +67,8 @@
 KDE=#
 MO=
 MO_NAME=$NAME.lang
-
+ALL_NAME=#
+NO_ALL_NAME=
 while test $# -gt 0 ; do
     case "${1}" in
 	--with-gnome )
@@ -62,13 +76,18 @@
 		shift
 		;;
 	--with-kde )
-		KDE_HELP=
+		KDE=
 		shift
 		;;
 	--without-mo )
 		MO=#
 		shift
 		;;
+	--all-name )
+		ALL_NAME=
+		NO_ALL_NAME=#
+		shift
+		;;
 	* )
 		MO_NAME=${1}
 		shift
@@ -80,22 +99,36 @@
 1i\
 %defattr (644, root, root, 755)
 s:'"$TOP_DIR"'::
-'"$MO"'s:\(.*/share/locale/\)\([^/_]\+\)\(.*/'"$NAME"'\.mo$\):%lang(\2) \1\2\3:
-'"$GNOME"'s:\(.*/gnome/help/'"$NAME"'/\)\([^/_]\+\):%lang(\2) \1\2:
+'"$ALL_NAME$MO"'s:\(.*/share/locale/\)\([^/_]\+\)\(.*\.mo$\):%lang(\2) \1\2\3:
+'"$NO_ALL_NAME$MO"'s:\(.*/share/locale/\)\([^/]\+\)\(.*/'"$NAME"'\.mo$\):%lang(\2) \1\2\3:
 s:^\([^%].*\)::
 s:%lang(C) ::
-' > $MO_NAME
+s:^\$::' > $MO_NAME
 
 find $TOP_DIR -type d|sed '
 s:'"$TOP_DIR"'::
-'"$GNOME"'s:\(.*/gnome/help/'"$NAME"'$\):%dir \1:
-'"$GNOME"'s:\(.*/gnome/help/'"$NAME"'/\)\([^/_]\+\):%dir %lang(\2) \1\2:
+'"$NO_ALL_NAME$GNOME"'s:\(.*/gnome/help/'"$NAME"'$\):%dir \1:
+'"$NO_ALL_NAME$GNOME"'s:\(.*/gnome/help/'"$NAME"'/[a-zA-Z0-9\_\-]/.\+\)::
+'"$NO_ALL_NAME$GNOME"'s:\(.*/gnome/help/'"$NAME"'\/\)\([^/]\+\):%lang(\2) \1\2:
+'"$ALL_NAME$GNOME"'s:\(.*/gnome/help/[a-zA-Z0-9\_\-]\+$\):%dir \1:
+'"$ALL_NAME$GNOME"'s:\(.*/gnome/help/[a-zA-Z0-9\_\-]\+/[a-zA-Z0-9\_\-]/.\+\)::
+'"$ALL_NAME$GNOME"'s:\(.*/gnome/help/[a-zA-Z0-9\_\-]\+\/\)\([^/]\+\):%lang(\2) \1\2:
+s:%lang(.*) .*/gnome/help/[a-zA-Z0-9\_\-]\+/[a-zA-Z0-9\_\-]\+/.*::
+'#'s:\(.*/gnome/help/'"$NAME"'$\):%dir \1:
+'#'s:\(.*/gnome/help/'"$NAME"'/\)\([^/]\+\):%dir %lang(\2) \1\2:
 s:^\([^%].*\)::
-s:%lang(C) ::
-' >> $MO_NAME
+s:%lang(C) ::' >> $MO_NAME
+
+find $TOP_DIR -type d|sed '
+s:'"$TOP_DIR"'::
+'"$NO_ALL_NAME$KDE"'s:\(.*/doc/kde/HTML/\)\([^/_]\+\)\(.*/'"$NAME"'/\)::
+'"$NO_ALL_NAME$KDE"'s:\(.*/doc/kde/HTML/\)\([^/_]\+\)\(.*/'"$NAME"'\):%lang(\2) \1\2\3:
+'"$ALL_NAME$KDE"'s:\(.*/doc/kde/HTML/\)\([^/_]\+\)\(.*/[a-zA-Z0-9\_\-]\+/\)::
+'"$ALL_NAME$KDE"'s:\(.*/doc/kde/HTML/\)\([^/_]\+\)\(.*/[a-zA-Z0-9\_\-]\+$\):%lang(\2) \1\2\3:
+s:^\([^%].*\)::
+s:%lang(C) ::' >> $MO_NAME
 
-if ! grep -q / $MO_NAME; then
-	echo "No translations found for ${NAME} in ${TOP_DIR}"
-	exit 1
+if [ "$(cat $MO_NAME | egrep -v '(^%defattr|^$)' | wc -l)" -le 0  ]; then
+        echo 'Error: international files not found !' 1>&2
 fi
 exit 0
