--- rpm-4.2.3/build/parseSpec.c.amd64	2005-01-13 05:55:04.000000000 -0500
+++ rpm-4.2.3/build/parseSpec.c	2005-01-14 08:25:05.963038183 -0500
@@ -187,6 +187,16 @@ static int copyNextLine(Spec spec, OFI_t
 }
 /*@=boundswrite@*/
 
+static const char *getAlternateArch(const char *arch)
+{
+    const char *alternate_arch = NULL;
+    if (! strncmp("x86_64", arch, sizeof("x86_64")-1))
+      alternate_arch = "amd64";
+    else if (! strncmp("amd64", arch, sizeof("amd64")-1))
+      alternate_arch = "x86_64";
+    return alternate_arch;
+}
+
 /*@-boundswrite@*/
 int readLine(Spec spec, int strip)
 {
@@ -274,13 +284,15 @@ retry:
 	match = 0;
     } else if (! strncmp("%ifarch", s, sizeof("%ifarch")-1)) {
 	const char *arch = rpmExpand("%{_target_cpu}", NULL);
+	const char *alternate_arch = getAlternateArch(arch);
 	s += 7;
-	match = matchTok(arch, s);
+	match = matchTok(arch, s) || (alternate_arch && matchTok(alternate_arch, s));
 	arch = _free(arch);
     } else if (! strncmp("%ifnarch", s, sizeof("%ifnarch")-1)) {
 	const char *arch = rpmExpand("%{_target_cpu}", NULL);
+	const char *alternate_arch = getAlternateArch(arch);
 	s += 8;
-	match = !matchTok(arch, s);
+	match = !matchTok(arch, s) && (!alternate_arch || !matchTok(alternate_arch, s));
 	arch = _free(arch);
     } else if (! strncmp("%ifos", s, sizeof("%ifos")-1)) {
 	const char *os = rpmExpand("%{_target_os}", NULL);
--- rpm-4.2.3/installplatform.amd64	2005-01-14 08:25:05.882021840 -0500
+++ rpm-4.2.3/installplatform	2005-01-14 10:11:26.365102504 -0500
@@ -31,11 +31,11 @@ target="`$RPM --eval '%{_target}'|sed -e
 targetcpu="`$RPM --eval '%{_target_cpu}'`"
 
 case "$arch" in
-  i[3456]86|athlon) SUBSTS="s_${arch}_i386_ s_${arch}_i486_ s_${arch}_i586_ s_${arch}_k6_ s_${arch}_i686_ s_${arch}_athlon_" ;;
+  i[3456]86|athlon) SUBSTS="s/${arch}/i386/ s/${arch}/i486/ s/${arch}/i586/ s/${arch}/k6/ s/${arch}/i686/ s/${arch}/athlon/" ;;
   alpha*) SUBSTS='s_alpha_alpha_ s_alpha_alphaev5_ s_alpha_alphaev56_ s_alpha_alphapca56_ s_alpha_alphaev6_ s_alpha_alphaev67_' ;;
   sparc*) SUBSTS='s_sparc\(64\|v9\)_sparc_ s_sparc64_sparcv9_;s_sparc\([^v]\|$\)_sparcv9\1_ s_sparcv9_sparc64_;s_sparc\([^6]\|$\)_sparc64\1_' ;;
   powerpc*|ppc*) SUBSTS='s_ppc64_ppc_ s_ppc\([^6]\|$\)_ppc64\1_ s_${arch}_ppciseries_ s_${arch}_ppcpseries_' ;;
-  x86_64*) SUBSTS='s/x86_64/x86_64/' ;;
+  x86_64*|amd64*) SUBSTS="s/${arch}/x86_64/ s/${arch}/amd64/" ;;
   s390*) SUBSTS='s_s390x_s390_ s_s390\([^x]\|$\)_s390x\1_' ;;
   *) SUBSTS=y___ ;;
 esac
@@ -69,7 +69,7 @@ for SUBST in $SUBSTS ; do
     ppc-linux) MULTILIBNO=1 ;;
     ppc64-linux) LIB=lib64; MULTILIBNO=2 ;;
     i?86-linux|athlon-linux) MULTILIBNO=1 ;;
-    x86_64-linux) LIB=lib64; MULTILIBNO=2 ;;
+    x86_64-linux|amd64-linux|ia32e-linux) LIB=lib64; MULTILIBNO=2 ;;
   esac
 
   if [ -n "$MULTILIBNO" ]; then
