--- imap-2002d/src/c-client/mail.c.can-2005-2933	2005-10-24 23:39:48.000000000 -0400
+++ imap-2002d/src/c-client/mail.c	2005-10-24 23:43:49.000000000 -0400
@@ -637,8 +637,10 @@
       if (c == '=') {		/* parse switches which take arguments */
 	if (*t == '"') {	/* quoted string? */
 	  for (v = arg,i = 0,++t; (c = *t++) != '"';) {
+	    if (!c) return NIL; /* unterminated string */
 				/* quote next character */
 	    if (c == '\\') c = *t++;
+	    if (!c) return NIL; /* unterminated string */
 	    arg[i++] = c;
 	  }
 	  c = *t++;		/* remember delimiter for later */
