--- groff-1.19/aclocal.m4.mkstemp	2003-04-17 18:27:43.000000000 +0200
+++ groff-1.19/aclocal.m4	2004-03-05 16:24:58.000000000 +0100
@@ -623,9 +623,9 @@ AC_TRY_LINK([#include <stdlib.h>
 #include <unistd.h>
 int (*f) (char *);],
 [f = mkstemp;],
-AC_MSG_RESULT(yes);AC_DEFINE(HAVE_MKSTEMP, 1,
-			     [Define if you have mkstemp().]),
-AC_MSG_RESULT(no);_AC_LIBOBJ(mkstemp))
+[AC_MSG_RESULT(yes); AC_DEFINE(HAVE_MKSTEMP)], 1,
+			     [Define if you have mkstemp().],
+[AC_MSG_RESULT(no); _AC_LIBOBJ(mkstemp)])
 AC_LANG_POP(C++)])dnl
 dnl
 dnl
