--- dietlibc-0.30/Makefile.mdkconfig	2006-06-18 20:32:28.000000000 +0200
+++ dietlibc-0.30/Makefile	2007-07-17 14:18:13.000000000 +0200
@@ -1,13 +1,9 @@
 INSTALL=install
-prefix?=/opt/diet
+prefix?=/usr
 # Set the following to install to a different root
 #DESTDIR=/tmp/fefix
 # Use "make DEBUG=1" to compile a debug version.
 
-LIBDIR=${prefix}/lib
-BINDIR=${prefix}/bin
-MAN1DIR=${prefix}/man/man1
-
 MYARCH:=$(shell uname -m | sed -e 's/i[4-9]86/i386/' -e 's/armv[3-6]t\?e\?[lb]/arm/')
 
 # This extra-ugly cruft is here so make will not run uname and sed each
@@ -83,8 +79,13 @@ endif
 
 # ARCH=$(MYARCH)
 
+LIBDIR=${prefix}/lib
+BINDIR=${prefix}/bin
+MAN1DIR=${prefix}/share/man/man1
+INSTALLHOME=$(LIBDIR)/dietlibc
+
 OBJDIR=bin-$(ARCH)
-ILIBDIR=$(LIBDIR)-$(ARCH)
+ILIBDIR=$(INSTALLHOME)/lib-$(ARCH)
 
 HOME=$(shell pwd)
 
@@ -309,7 +310,7 @@ $(OBJDIR)/diet: $(OBJDIR)/start.o $(OBJD
 	$(CROSS)strip -R .comment -R .note $@
 
 $(OBJDIR)/diet-i: $(OBJDIR)/start.o $(OBJDIR)/dyn_start.o diet.c $(OBJDIR)/dietlibc.a $(OBJDIR)/dyn_stop.o
-	$(CROSS)$(CC) -isystem include $(CFLAGS) -nostdlib -o $@ $^ -DDIETHOME=\"$(prefix)\" -DVERSION=\"$(VERSION)\" -DINSTALLVERSION -lgcc
+	$(CROSS)$(CC) -isystem include $(CFLAGS) -nostdlib -o $@ $^ -DDIETHOME=\"$(INSTALLHOME)\" -DVERSION=\"$(VERSION)\" -DINSTALLVERSION -lgcc
 	$(CROSS)strip -R .comment -R .note $@
 
 $(PICODIR)/diet-dyn: $(PICODIR)/start.o $(PICODIR)/dyn_start.o diet.c
@@ -375,7 +376,7 @@ endif
 	-$(INSTALL) $(PICODIR)/dstart.o $(PICODIR)/dyn_so_start.o $(DESTDIR)$(ILIBDIR)
 	$(INSTALL) -m 644 diet.1 $(DESTDIR)$(MAN1DIR)/diet.1
 	if test -f $(PICODIR)/libc.so -a ! -f $(DESTDIR)/etc/diet.ld.conf; then echo "$(ILIBDIR)" > $(DESTDIR)/etc/diet.ld.conf; fi
-	for i in `find include -name \*.h`; do install -m 644 -D $$i $(DESTDIR)$(prefix)/$$i; done
+	for i in `find include -name \*.h`; do install -m 644 -D $$i $(DESTDIR)$(INSTALLHOME)/$$i; done
 
 uninstall:
 	for i in start.o libm.a libpthread.a librpc.a liblatin1.a libcompat.a libcrypt.a libc.a; do rm -f $(DESTDIR)$(ILIBDIR)/$$i; done
