diff -Naur mdadm-2.5/Makefile mdadm-2.5.oden/Makefile
--- mdadm-2.5/Makefile	2006-05-26 06:27:55.000000000 +0200
+++ mdadm-2.5.oden/Makefile	2006-05-28 21:41:26.000000000 +0200
@@ -32,7 +32,7 @@
 TCC = tcc
 UCLIBC_GCC = $(shell for nm in i386-uclibc-linux-gcc i386-uclibc-gcc; do which $$nm > /dev/null && { echo $$nm ; exit; } ; done; echo false No uclibc found )
 DIET_GCC = diet gcc
-LDLIBS=-lssl
+LDLIBS=
 
 KLIBC=/home/src/klibc/klibc-0.77
 
@@ -67,10 +67,10 @@
 
 OBJS =  mdadm.o config.o mdstat.o  ReadMe.o util.o Manage.o Assemble.o Build.o \
 	Create.o Detail.o Examine.o Grow.o Monitor.o dlink.o Kill.o Query.o \
-	mdopen.o super0.o super1.o bitmap.o restripe.o sysfs.o
+	mdopen.o super0.o super1.o bitmap.o restripe.o sysfs.o SHA1.o sha1.o
 SRCS =  mdadm.c config.c mdstat.c  ReadMe.c util.c Manage.c Assemble.c Build.c \
 	Create.c Detail.c Examine.c Grow.c Monitor.c dlink.c Kill.c Query.c \
-	mdopen.c super0.c super1.c bitmap.c restripe.c sysfs.c
+	mdopen.c super0.c super1.c bitmap.c restripe.c sysfs.c SHA1.c sha1.c
 
 ASSEMBLE_SRCS := mdassemble.c Assemble.c config.c dlink.c util.c super0.c super1.c
 ASSEMBLE_FLAGS:= -DMDASSEMBLE
diff -Naur mdadm-2.5/super0.c mdadm-2.5.oden/super0.c
--- mdadm-2.5/super0.c	2006-05-26 06:18:55.000000000 +0200
+++ mdadm-2.5.oden/super0.c	2006-05-28 21:41:41.000000000 +0200
@@ -28,11 +28,7 @@
  */
 
 #include "mdadm.h"
-#ifndef UCLIBC
-#include <openssl/sha.h> /* for SHA1 */
-#else
 extern unsigned char *SHA1(unsigned char *buf, int len, unsigned char *dest);
-#endif
 /*
  * All handling for the 0.90.0 version superblock is in
  * this file.
