--- mc-4.6.0/syntax/diff.syntax.diff_of_diff	2002-10-31 22:58:55.000000000 +0100
+++ mc-4.6.0/syntax/diff.syntax	2004-08-29 13:14:39.163240527 +0200
@@ -10,9 +10,9 @@
 context exclusive linestart @@*@@ \n brightcyan
 context exclusive linestart Index:\s \n black white
 context linestart diff \n white red
-context linestart --- \n brightmagenta
-context linestart \+\+\+ \n brightmagenta
-context linestart \*\*\* \n brightmagenta
+context linestart ---\s \n brightmagenta
+context linestart \+\+\+\s \n brightmagenta
+context linestart \*\*\*\s \n brightmagenta
 context linestart === \n brown
 context linestart \+ \n brightgreen
 context linestart > \n brightgreen
