--- crypt/crypt-entry.c	Thu Dec 10 12:49:04 1998
+++ crypt/crypt-entry.c	Sun Jun 11 03:14:57 2000
@@ -70,7 +70,7 @@
  */
 
 char *
-__crypt_r (key, salt, data)
+__des_crypt_r (key, salt, data)
      const char *key;
      const char *salt;
      struct crypt_data * __restrict data;
@@ -115,6 +115,7 @@
   _ufc_output_conversion_r (res[0], res[1], salt, data);
   return data->crypt_3_buf;
 }
+#if 0
 weak_alias (__crypt_r, crypt_r)
 
 char *
@@ -147,3 +147,4 @@
   return crypt (key, salt);
 }
 #endif
+#endif
--- crypt/Makefile	2001/07/06 04:54:45	1.9
+++ crypt/Makefile	2001/12/07 14:25:32
@@ -28,7 +28,8 @@
 extra-libs := libcrypt
 extra-libs-others := $(extra-libs)
 
-libcrypt-routines := crypt-entry md5-crypt md5 crypt crypt_util
+libcrypt-routines := crypt-entry md5-crypt md5 crypt crypt_util \
+		     crypt_blowfish crypt_gensalt wrapper
 
 tests = cert md5test md5c-test
 
--- crypt/ow-crypt.h	2001/12/07 15:50:08	1.1
+++ crypt/ow-crypt.h	2001/12/07 15:50:23
@@ -13,12 +13,6 @@
 #define __CONST
 #endif
 
-#ifndef __SKIP_GNU
-extern char *crypt(__CONST char *key, __CONST char *setting);
-extern char *crypt_r(__CONST char *key, __CONST char *setting, void *data);
-#endif
-
-#ifndef __SKIP_OW
 extern char *crypt_rn(__CONST char *key, __CONST char *setting,
 	void *data, int size);
 extern char *crypt_ra(__CONST char *key, __CONST char *setting,
@@ -29,6 +23,5 @@
 	__CONST char *input, int size, char *output, int output_size);
 extern char *crypt_gensalt_ra(__CONST char *prefix, unsigned long count,
 	__CONST char *input, int size);
-#endif
 
 #endif
--- crypt/crypt_blowfish.c	2001/12/07 16:52:58	1.1
+++ crypt/crypt_blowfish.c	2001/12/07 16:53:09
@@ -45,7 +45,7 @@
 #endif
 
 #ifdef __i386__
-#define BF_ASM				1
+#define BF_ASM				0
 #define BF_SCALE			1
 #elif defined(__alpha__)
 #define BF_ASM				0
--- crypt/wrapper.c	2001/12/09 19:36:57	1.1
+++ crypt/wrapper.c	2001/12/09 19:37:30
@@ -263,6 +263,7 @@
 weak_alias(__crypt_ra, crypt_ra)
 weak_alias(__crypt_r, crypt_r)
 weak_alias(__crypt, crypt)
+weak_alias(__crypt, fcrypt)
 weak_alias(__crypt_gensalt_rn, crypt_gensalt_rn)
 weak_alias(__crypt_gensalt_ra, crypt_gensalt_ra)
 weak_alias(__crypt_gensalt, crypt_gensalt)
