--- glibc-2.1.92/linuxthreads/man/pthread_create.man.chmou	Tue Jan  4 17:12:07 2000
+++ glibc-2.1.92/linuxthreads/man/pthread_create.man	Tue Oct 17 15:30:47 2000
@@ -31,10 +31,8 @@
 .SH ERRORS
 .TP
 !EAGAIN!
-not enough system resources to create a process for the new thread.
-.TP
-!EAGAIN!
-more than !PTHREAD_THREADS_MAX! threads are already active.
+not enough system resources to create a process for the new thread,
+or more than !PTHREAD_THREADS_MAX! threads are already active.
 
 .SH AUTHOR
 Xavier Leroy <Xavier.Leroy@inria.fr>
