--- glibc-2.3.1/inet/herrno.c.errno-compat	2003-02-09 12:23:12.000000000 -0500
+++ glibc-2.3.1/inet/herrno.c	2003-03-04 17:37:28.000000000 -0500
@@ -41,7 +41,7 @@ weak_alias (h_errno, _h_errno)
    visible at link time.  Programs must use the accessor functions.  */
 # if defined HAVE_ELF && defined SHARED && defined DO_VERSIONING
 #  include <shlib-compat.h>
-compat_symbol (libc, h_errno, h_errno, GLIBC_2_0);
-compat_symbol (libc, _h_errno, _h_errno, GLIBC_2_0);
+// compat_symbol (libc, h_errno, h_errno, GLIBC_2_0);
+// compat_symbol (libc, _h_errno, _h_errno, GLIBC_2_0);
 # endif
 #endif
--- glibc-2.3.1/resolv/res_libc.c.errno-compat	2003-02-09 12:23:13.000000000 -0500
+++ glibc-2.3.1/resolv/res_libc.c	2003-03-04 17:37:28.000000000 -0500
@@ -87,7 +87,7 @@ struct __res_state _res __attribute__((s
    visible at link time.  Programs must use the accessor functions.  */
 # if defined HAVE_ELF && defined SHARED && defined DO_VERSIONING
 #  include <shlib-compat.h>
-compat_symbol (libc, _res, _res, GLIBC_2_0);
+// compat_symbol (libc, _res, _res, GLIBC_2_0);
 # endif
 #endif
 
--- glibc-2.3.1/sysdeps/generic/errno.c.errno-compat	2002-11-17 14:55:51.000000000 -0500
+++ glibc-2.3.1/sysdeps/generic/errno.c	2003-03-04 17:37:28.000000000 -0500
@@ -36,7 +36,7 @@ strong_alias (errno, _errno)
    visible at link time.  Programs must use the accessor functions.  */
 # if defined HAVE_ELF && defined SHARED && defined DO_VERSIONING
 #  include <shlib-compat.h>
-compat_symbol (libc, errno, errno, GLIBC_2_0);
-compat_symbol (libc, _errno, _errno, GLIBC_2_0);
+// compat_symbol (libc, errno, errno, GLIBC_2_0);
+// compat_symbol (libc, _errno, _errno, GLIBC_2_0);
 # endif
 #endif
