--- quota-tools/warnquota.conf.chmou	2002-07-22 21:00:51.000000000 +0200
+++ quota-tools/warnquota.conf	2003-02-08 22:06:16.000000000 +0100
@@ -4,21 +4,21 @@
 # and even blank lines
 
 # values can be quoted:
-MAIL_CMD 	= "/usr/my/sendmail/instead/sendmail -t"
-FROM 		= "bas@localhost"
+MAIL_CMD 	= "/usr/sbin/sendmail -t"
+FROM 		= "root@localhost"
 # but they don't have to be:
-SUBJECT 	= Hey, user, clean up your account!
-CC_TO 		= "sysadm@localhost"
-SUPPORT 	= "support@myhost.com"
+SUBJECT 	= NOTE: You are exceeding your allocatted disk space limits
+CC_TO 		= "root@localhost"
+SUPPORT 	= "root@myhost.com"
 PHONE 		= "(123) 456-1111 or (222) 333-4444"
 # Text in the beginning of the mail (if not specified, default text is used)
 # This way text can be split to more lines
 # Line breaks are done by '|' character
-MESSAGE         = Hello, I've noticed you use too much space\
- on my disk|Delete your files on following filesystems:|
-# Text in the end of the mail (if not specified, default text using SUPPORT and PHONE
+MESSAGE         = Your disk usage has exceeded the agreeded limits\
+  on this server|Please delete any unnecessary files on following filesystems:|
+ Text in the end of the mail (if not specified, default text using SUPPORT and PHONE
 # is created)
-SIGNATURE	= See you!|			Your admin|
+SIGNATURE	= root@localhost
 # Following text is used for mails about group exceeding quotas
 # It should contain string %s exactly once - it will be substituted for a group name
 GROUP_MESSAGE	= Hello, a group '%s' you're member of use too much space.|\
