--- gcc-3.3/gcc/ada/7sosprim.adb.ada-64bit	2002-10-23 09:33:20.000000000 +0200
+++ gcc-3.3/gcc/ada/7sosprim.adb	2003-02-02 17:57:20.000000000 +0100
@@ -48,11 +48,11 @@ package body System.OS_Primitives is
    pragma Convention (C, struct_timezone);
    type struct_timezone_ptr is access all struct_timezone;
 
-   type time_t is new Integer;
+   type time_t is new Long_Integer;
 
    type struct_timeval is record
       tv_sec       : time_t;
-      tv_usec      : Integer;
+      tv_usec      : Long_Integer;
    end record;
    pragma Convention (C, struct_timeval);
 
