--- modules/experimental/util_ldap_cache.c	2004-02-11 13:07:46.000000000 -0500
+++ modules/experimental/util_ldap_cache.c.oden	2004-05-15 15:03:00.403776176 -0400
@@ -28,7 +28,7 @@
 #ifdef APU_HAS_LDAP
 
 #if APR_HAS_SHARED_MEMORY
-#define MODLDAP_SHMEM_CACHE "/tmp/mod_ldap_cache"
+#define MODLDAP_SHMEM_CACHE "/var/cache/httpd/mod_ldap_cache"
 #endif
 
 /* ------------------------------------------------------------------ */
