--- gawk-3.1.2/regex.h.regex	2003-02-25 10:58:52.000000000 +0100
+++ gawk-3.1.2/regex.h	2003-10-15 11:01:13.000000000 +0200
@@ -327,7 +327,7 @@
    private to the regex routines.  */
 
 #ifndef RE_TRANSLATE_TYPE
-# define RE_TRANSLATE_TYPE char *
+# define RE_TRANSLATE_TYPE unsigned char *
 #endif
 
 struct re_pattern_buffer
--- gawk-3.1.2/regexec.c.regex	2003-02-28 09:57:02.000000000 +0100
+++ gawk-3.1.2/regexec.c	2003-10-15 11:01:31.000000000 +0200
@@ -657,8 +657,8 @@
 	{
 	  if (BE (fast_translate, 1))
 	    {
-	      unsigned RE_TRANSLATE_TYPE t
-		= (unsigned RE_TRANSLATE_TYPE) preg->translate;
+	      RE_TRANSLATE_TYPE t
+		= (RE_TRANSLATE_TYPE) preg->translate;
 	      if (BE (range >= 0, 1))
 		{
 		  if (BE (t != NULL, 0))
