diff -Naur squid-2.6.STABLE16/src/cf.data.pre squid-2.6.STABLE16.oden/src/cf.data.pre
--- squid-2.6.STABLE16/src/cf.data.pre	2007-09-05 23:50:15.000000000 +0200
+++ squid-2.6.STABLE16.oden/src/cf.data.pre	2007-09-06 09:34:46.000000000 +0200
@@ -692,7 +692,7 @@
 #http_access allow our_networks
 
 # And finally deny all other access to this proxy
-http_access deny all
+http_access allow localhost
 NOCOMMENT_END
 DOC_END
 
@@ -2250,7 +2250,7 @@
 
 NAME: logfile_rotate
 TYPE: int
-DEFAULT: 10
+DEFAULT: 0
 LOC: Config.Log.rotateNumber
 DOC_START
 	Specifies the number of logfile rotations to make when you
@@ -3342,7 +3342,7 @@
 
 NAME: cache_mgr
 TYPE: string
-DEFAULT: webmaster
+DEFAULT: root
 LOC: Config.adminEmail
 DOC_START
 	Email-address of local cache manager who will receive
@@ -4405,12 +4405,22 @@
 	If you wish to create your own versions of the default
 	(English) error files, either to customize them to suit your
 	language or company copy the template English files to another
+	directory where the error files are read from.
+	/usr/lib/squid/errors contains sets of error files
+	in different languages. The default error directory
+	is /etc/squid/errors, which is a link to one of these
+	error sets.
+
+	If you wish to create your own versions of the error files,
+	either to customize them to suit your language or company,
+	copy the template English files to another
 	directory and point this tag at them.
 
 	The squid developers are interested in making squid available in
 	a wide variety of languages. If you are making translations for a
 	langauge that Squid does not currently provide please consider
 	contributing your translation back to the project.
+error_directory @DEFAULT_ERROR_DIR@
 DOC_END
 
 NAME: error_map
diff -Naur squid-2.6.STABLE16/src/defines.h squid-2.6.STABLE16.oden/src/defines.h
--- squid-2.6.STABLE16/src/defines.h	2007-02-03 23:58:20.000000000 +0100
+++ squid-2.6.STABLE16.oden/src/defines.h	2007-09-06 09:35:22.000000000 +0200
@@ -259,7 +259,7 @@
 
 /* were to look for errors if config path fails */
 #ifndef DEFAULT_SQUID_ERROR_DIR
-#define DEFAULT_SQUID_ERROR_DIR "/usr/local/squid/etc/errors"
+#define DEFAULT_SQUID_ERROR_DIR "/etc/squid/errors"
 #endif
 
 /* gb_type operations */
