diff -Naur Mesa-3.3/glx/configure.in Mesa-3.3.new/glx/configure.in
--- Mesa-3.3/glx/configure.in	Sun Aug 20 07:50:25 2000
+++ Mesa-3.3.new/glx/configure.in	Fri Aug 25 18:36:20 2000
@@ -231,7 +231,7 @@
 [  --with-moduledir=DIR    driver module (glx.so) in DIR [same as libdir] ],
 [moduledir=$withval],[moduledir="$libdir"])
 AC_SUBST(moduledir)
-glx_defines="$glx_defines -DGLX_SO_PATH='\"$moduledir/glx.so\"'"
+glx_defines="$glx_defines -DGLX_SO_PATH='\"$moduledir/glx-3.so\"'"
 
 dnl AC_ARG_WITH(x-srcs,
 dnl [  --with-x-srcs           specify the top level of the X build tree
diff -Naur Mesa-3.3/glx/libGL/direct_protocol.c Mesa-3.3.new/glx/libGL/direct_protocol.c
--- Mesa-3.3/glx/libGL/direct_protocol.c	Wed May 17 07:28:00 2000
+++ Mesa-3.3.new/glx/libGL/direct_protocol.c	Fri Aug 25 18:38:01 2000
@@ -241,7 +241,10 @@
 
  error_case:
    __glx_unload_direct();
-   fprintf(stderr, "failed to load glx.so\n");
+   /* Modification for Linux-Mandrake by François Pons <fpons@mandrakesoft.com>
+    * rename glx.so to glx-3.so to allow installation
+    * if using XFree 4.0 or above. */
+   fprintf(stderr, "failed to load glx-3.so\n");
    return 0;
 }
 
diff -Naur Mesa-3.3/glx/libGL/direct_protocol.h Mesa-3.3.new/glx/libGL/direct_protocol.h
--- Mesa-3.3/glx/libGL/direct_protocol.h	Wed May 17 07:28:00 2000
+++ Mesa-3.3.new/glx/libGL/direct_protocol.h	Fri Aug 25 18:41:21 2000
@@ -4,8 +4,11 @@
 #include "glxcommon.h"
 
 /* set a default path to look for glx.so if Makefile doesn't specify */
+/* Modification for Linux-Mandrake by François Pons <fpons@mandrakesoft.com>
+ * rename glx.so to glx-3.so to allow installation
+ * if using XFree86 4.0 or above. */
 #ifndef GLX_SO_PATH
-#define GLX_SO_PATH	"/usr/X11R6/lib/modules/glx.so"
+#define GLX_SO_PATH	"/usr/X11R6/lib/modules/glx-3.so"
 #endif
 
 struct _Client;
diff -Naur Mesa-3.3/glx/libGL/glx.c Mesa-3.3.new/glx/libGL/glx.c
--- Mesa-3.3/glx/libGL/glx.c	Thu Jun 29 18:20:25 2000
+++ Mesa-3.3.new/glx/libGL/glx.c	Fri Aug 25 18:41:13 2000
@@ -868,7 +868,11 @@
 		  __glx_direct_count - 1);
 
 	  if (--__glx_direct_count == 0) {
-	     fprintf(stderr, "Unloading glx.so\n");
+	     /* Modification for Linux-Mandrake by François Pons
+	      * <fpons@mandrakesoft.com>
+	      * rename glx.so to glx-3.so to allow installation
+	      * if using XFree86 4.0 or above. */
+	     fprintf(stderr, "Unloading glx-3.so\n");
 	     __glx_unload_direct();
 	  }
 
diff -Naur Mesa-3.3/glx/servGL/Makefile.in Mesa-3.3.new/glx/servGL/Makefile.in
--- Mesa-3.3/glx/servGL/Makefile.in	Tue Aug 22 02:04:12 2000
+++ Mesa-3.3.new/glx/servGL/Makefile.in	Fri Aug 25 18:41:01 2000
@@ -94,8 +94,12 @@
 # In this case, the targets are passed on to the subdirectories
 # -------------------------------------------------------------
 
+# Modification for Linux-Mandrake by François Pons <fpons@mandrakesoft.com>
+# rename glx.so to glx-3.so to allow installation
+# if using XFree86 4.0 or above.
+
 ifeq "${USE_DYNAMIC_MODULE}" "yes"
- GLXMODULES = glx.so
+ GLXMODULES = glx-3.so
  all : subdirs ${GLXMODULES}
 else
  all : subdirs
@@ -110,7 +114,7 @@
 	for i in ${SUBDIRS} ; do \
 		( cd $$i && ${MAKE} USE_HW_MGA=yes USE_HW_TNT=yes USE_HW_MACH64=yes USE_HW_S3VIRGE=yes USE_HW_S3SAVAGE=yes USE_HW_SIS6326=yes real-clean ) || exit 1 ; \
 	done
-	${RM} glx.so* ${GLXMODULES}
+	${RM} glx-3.so* ${GLXMODULES}
 	${RM} -f .deps
 
 depend :
@@ -153,7 +157,7 @@
 # Solaris ld has issues linking this with the options it ends up with when gcc
 # is passed "-shared", so we'll kludge around it for now :-(
 
-glx.so : ${SO_DEPENDS}
+glx-3.so : ${SO_DEPENDS}
 	${RM} $@~
 	if [ `uname` = "SunOS" ]; then \
 		${CC} -o $@~ -Wl,-G,-dy,-ztextwarn ${SO_OBJS} ${MATHLIBS}; \
