--- src/support/nfs/exports.c.org	2003-08-06 01:19:42.000000000 -0400
+++ src/support/nfs/exports.c	2004-06-14 06:48:29.000000000 -0400
@@ -442,7 +442,7 @@ bad_option:
 	ep->e_nsqgids = nsqgids;
 
 out:
-	if (warn && !had_sync_opt)
+	if (warn && !had_sync_opt && !(ep->e_flags & NFSEXP_READONLY))
 		xlog(L_WARNING, "%s [%d]: No 'sync' or 'async' option specified for export \"%s:%s\".\n"
 				"  Assuming default behaviour ('sync').\n"
 		     		"  NOTE: this default has changed from previous versions\n",
