diff -Nurp src/3rdparty/iteraid.old/iteraid.c src/3rdparty/iteraid/iteraid.c
--- src/3rdparty/iteraid.old/iteraid.c	2004-06-01 11:22:13.000000000 +0300
+++ src/3rdparty/iteraid/iteraid.c	2004-06-01 12:15:30.000000000 +0300
@@ -252,9 +252,9 @@ MODULE_DESCRIPTION ("ITE IT8212 RAID Con
 #include <asm/irq.h>
 #include <asm/uaccess.h>
 
-#include "sd.h"
-#include "scsi.h"
-#include "hosts.h"
+#include "../../drivers/scsi/sd.h"
+#include "../../drivers/scsi/scsi.h"
+#include "../../drivers/scsi/hosts.h"
 
 #include "iteraid.h"
 
@@ -5833,4 +5833,4 @@ itedev_close(struct inode *inode,  struc
  * each time a new driver is written.
  */
 static Scsi_Host_Template driver_template = ITERAID;
-#include "scsi_module.c"
+#include "../../drivers/scsi/scsi_module.c"
