--- courier-imap-4.1.1/imap/imapd.dist.in.avx	2006-12-30 23:39:48.000000000 -0700
+++ courier-imap-4.1.1/imap/imapd.dist.in	2006-12-30 23:53:10.000000000 -0700
@@ -8,34 +8,32 @@
 #  Copyright 1998 - 2006 Double Precision, Inc.  See COPYING for
 #  distribution information.
 #
-#  This configuration file sets various options for the Courier-IMAP server
-#  when used with the couriertcpd server.
-#  A lot of the stuff here is documented in the manual page for couriertcpd.
+# imapd configuration file for courier-imap on Annvix
+#
+# This configuration file is used instead of what you might be used to
+# seeing from a "stock" courier-imap because Annvix does not use
+# couriertcpd but uses tcpsvd instead so a lot of the stuff that is
+# normally here is not required
+#
+# Many options can be set in /var/service/courier-imapd/env using the
+# environment files MAX_CONN, MAX_MEM, IP, and PORT:
+#
+#   MAX_MEM	 - the maximum memory to allow courier-imap to use
+#   MAX_CONN	 - the maximum number of concurrent connections
+#   MAX_PER_HOST - the maximum number of conncurrent connections per host
+#		   (the old MAXPERIP setting)
+#   IP		 - the IP address to bind to (default: 0 (all))
+#		   (the old ADDRESS setting)
+#   PORT	 - the TCP port to listen to (default: 143)
+#
 #
 #  NOTE - do not use \ to split long variable contents on multiple lines.
 #  This will break the default imapd.rc script, which parses this file.
 #
+# use ./env/IP
 ##NAME: ADDRESS:0
-#
-#  Address to listen on, can be set to a single IP address.
-#
-# ADDRESS=127.0.0.1
-
-ADDRESS=0
-
+# use ./env/PORT
 ##NAME: PORT:1
-#
-#  Port numbers that connections are accepted on.  The default is 143,
-#  the standard IMAP port.
-#
-#  Multiple port numbers can be separated by commas.  When multiple port
-#  numbers are used it is possible to select a specific IP address for a
-#  given port as "ip.port".  For example, "127.0.0.1.900,192.68.0.1.900"
-#  accepts connections on port 900 on IP addresses 127.0.0.1 and 192.68.0.1
-#  The previous ADDRESS setting is a default for ports that do not have
-#  a specified IP address.
-
-PORT=143
 
 ##NAME: AUTHSERVICE:0
 #
@@ -46,50 +44,15 @@
 #  AUTHSERVICE143=imap
 #  AUTHSERVICE993=imaps
 
+# use ./env/MAX_CONN
 ##NAME: MAXDAEMONS:0
-#
-#  Maximum number of IMAP servers started
-#
-
-MAXDAEMONS=40
-
+# use ./env/MAX_PER_HOST
 ##NAME: MAXPERIP:0
-#
-#  Maximum number of connections to accept from the same IP address
-
-MAXPERIP=4
-
+# these are unused
 ##NAME: PIDFILE:0
-#
-#  File where couriertcpd will save its process ID
-#
-
-PIDFILE=@piddir@/imapd.pid
-
 ##NAME: TCPDOPTS:0
-#
-# Miscellaneous couriertcpd options that shouldn't be changed.
-#
-
-TCPDOPTS="-nodnslookup -noidentlookup"
-
 ##NAME: LOGGEROPTS:0
-#
-# courierlogger(1) options.                                        
-#
-
-LOGGEROPTS="-name=imapd"
-
 ##NAME: DEFDOMAIN:0
-#
-# Optional default domain. If the username does not contain the         
-# first character of DEFDOMAIN, then it is appended to the username.
-# If DEFDOMAIN and DOMAINSEP are both set, then DEFDOMAIN is appended
-# only if the username does not contain any character from DOMAINSEP.
-# You can set different default domains based on the the interface IP
-# address using the -access and -accesslocal options of couriertcpd(1).
-
-#DEFDOMAIN="@example.com"
 
 ##NAME: IMAP_CAPABILITY:1
 #
@@ -384,23 +347,8 @@
 #
 # OUTBOX_MULTIPLE_SEND=1
 
+# this has absolutely no impact; use srv to control whether or not to start
 ##NAME: IMAPDSTART:0
-#
-# IMAPDSTART is not used directly.  Rather, this is a convenient flag to
-# be read by your system startup script in /etc/rc.d, like this:
-#
-#  . @sysconfdir@/imapd
-#
-#  case x$IMAPDSTART in
-#  x[yY]*)
-#        @libexecdir@/imapd.rc start
-#        ;;
-#  esac
-#
-# The default setting is going to be NO, so you'll have to manually flip
-# it to yes.
-
-IMAPDSTART=NO
 
 ##NAME: MAILDIRPATH:0
 #
--- courier-imap-4.1.1/imap/imapd-ssl.dist.in.avx	2006-12-30 23:53:40.000000000 -0700
+++ courier-imap-4.1.1/imap/imapd-ssl.dist.in	2006-12-30 23:56:11.000000000 -0700
@@ -8,71 +8,38 @@
 #  Copyright 2000 - 2004 Double Precision, Inc.  See COPYING for
 #  distribution information.
 #
+#
+# imapds configuration file for courier-imap on Annvix
+#
+# This configuration file is used instead of what you might be used to
+# seeing from a "stock" courier-imap because Annvix does not use
+# couriertcpd but uses tcpsvd instead so a lot of the stuff that is
+# normally here is not required
+#
+# Many options can be set in /var/service/courier-imapds/env using the
+# environment files MAX_CONN, MAX_MEM, IP, and PORT:
+#
+#   MAX_MEM	 - the maximum memory to allow courier-imap to use
+#   MAX_CONN	 - the maximum number of concurrent connections
+#   MAX_PER_HOST - the maximum number of conncurrent connections per host
+#		   (the old MAXPERIP setting)
+#   IP		 - the IP address to bind to (default: 0 (all))
+#		   (the old ADDRESS setting)
+#   PORT	 - the TCP port to listen to (default: 143)
+#
 #  This configuration file sets various options for the Courier-IMAP server
 #  when used to handle SSL IMAP connections.
 #
-#  SSL and non-SSL connections are handled by a dedicated instance of the
-#  couriertcpd daemon.  If you are accepting both SSL and non-SSL IMAP
-#  connections, you will start two instances of couriertcpd, one on the
-#  IMAP port 143, and another one on the IMAP-SSL port 993.
-#
 #  Download OpenSSL from http://www.openssl.org/
 #
+# use ./env/PORT
 ##NAME: SSLPORT:1
-#
-#  Options in the imapd-ssl configuration file AUGMENT the options in the
-#  imapd configuration file.  First the imapd configuration file is read,
-#  then the imapd-ssl configuration file, so we do not have to redefine
-#  anything.
-#
-#  However, some things do have to be redefined.  The port number is
-#  specified by SSLPORT, instead of PORT.  The default port is port 993.
-#
-#  Multiple port numbers can be separated by commas.  When multiple port
-#  numbers are used it is possibly to select a specific IP address for a
-#  given port as "ip.port".  For example, "127.0.0.1.900,192.68.0.1.900"
-#  accepts connections on port 900 on IP addresses 127.0.0.1 and 192.68.0.1
-#  The SSLADDRESS setting is a default for ports that do not have
-#  a specified IP address.
-
-SSLPORT=993
-
+# use ./env/IP
 ##NAME: SSLADDRESS:0
-#
-#  Address to listen on, can be set to a single IP address.
-#
-# SSLADDRESS=127.0.0.1
-
-SSLADDRESS=0
-
+# unused
 ##NAME: SSLPIDFILE:0
-#
-# That's the SSL IMAP port we'll listen on.
-# Feel free to redefine MAXDAEMONS, TCPDOPTS, and MAXPERIP.
-
-SSLPIDFILE=@piddir@/imapd-ssl.pid
-
 ##NAME: SSLLOGGEROPTS:0
-#
-# courierlogger(1) options.                                        
-#
-
-SSLLOGGEROPTS="-name=imapd-ssl"
-
 ##NAME: IMAPDSSLSTART:0
-#
-# Different pid files, so that both instances of couriertcpd can coexist
-# happily.
-#
-# You can also redefine IMAP_CAPABILITY, although I can't
-# think of why you'd want to do that.
-#
-#
-# Ok, the following settings are new to imapd-ssl:
-#
-#  Whether or not to start IMAP over SSL on simap port:
-
-IMAPDSSLSTART=NO
 
 ##NAME: IMAPDSTARTTLS:0
 #
--- courier-imap-4.1.1/imap/pop3d.dist.in.avx	2006-12-30 23:56:36.000000000 -0700
+++ courier-imap-4.1.1/imap/pop3d.dist.in	2006-12-30 23:58:28.000000000 -0700
@@ -10,23 +10,31 @@
 #
 #  Courier POP3 daemon configuration
 #
-##NAME: PIDFILE:0
-#
-
-PIDFILE=@piddir@/pop3d.pid
-
-##NAME: MAXDAEMONS:0
 #
-#  Maximum number of POP3 servers started
+# pop3d configuration file for courier-imap on Annvix
 #
+# This configuration file is used instead of what you might be used to
+# seeing from a "stock" courier-imap because Annvix does not use
+# couriertcpd but uses tcpsvd instead so a lot of the stuff that is
+# normally here is not required
+#
+# Many options can be set in /var/service/courier-pop3d/env using the
+# environment files MAX_CONN, MAX_MEM, IP, and PORT:
+#
+#   MAX_MEM	 - the maximum memory to allow courier-imap to use
+#   MAX_CONN	 - the maximum number of concurrent connections
+#   MAX_PER_HOST - the maximum number of conncurrent connections per host
+#		   (the old MAXPERIP setting)
+#   IP		 - the IP address to bind to (default: 0 (all))
+#		   (the old ADDRESS setting)
+#   PORT	 - the TCP port to listen to (default: 143)
 
-MAXDAEMONS=40
-
+# unused
+##NAME: PIDFILE:0
+# use ./env/MAX_CONN
+##NAME: MAXDAEMONS:0
+# use ./env/MAX_PER_HOST
 ##NAME: MAXPERIP:4
-#
-#  Maximum number of connections to accept from the same IP address
-
-MAXPERIP=4
 
 ##NAME: POP3AUTH:1
 #
@@ -76,68 +84,15 @@
 
 # PROXY_HOSTNAME=
 
+# use ./env/PORT
 ##NAME: PORT:1
-#
-# Port to listen on for connections.  The default is port 110.
-#
-#  Multiple port numbers can be separated by commas.  When multiple port
-#  numbers are used it is possibly to select a specific IP address for a
-#  given port as "ip.port".  For example, "127.0.0.1.900,192.68.0.1.900"
-#  accepts connections on port 900 on IP addresses 127.0.0.1 and 192.68.0.1
-#  The ADDRESS setting is a default for ports that do not have a specified
-#  IP address.
-
-PORT=110
-
+# use ./env/IP
 ##NAME: ADDRESS:0
-#
-# IP address to listen on.  0 means all IP addresses.
-
-ADDRESS=0
-
+# unused
 ##NAME: TCPDOPTS:0
-#
-# Other couriertcpd(1) options.  The following defaults should be fine.
-#
-
-TCPDOPTS="-nodnslookup -noidentlookup"
-
 ##NAME: LOGGEROPTS:0
-#
-# courierlogger(1) options.
-#
-
-LOGGEROPTS="-name=pop3d"
-
 ##NAME: DEFDOMAIN:0
-#
-# Optional default domain. If the username does not contain the
-# first character of DEFDOMAIN, then it is appended to the username.
-# If DEFDOMAIN and DOMAINSEP are both set, then DEFDOMAIN is appended
-# only if the username does not contain any character from DOMAINSEP.
-# You can set different default domains based on the the interface IP
-# address using the -access and -accesslocal options of couriertcpd(1).
-
-#DEFDOMAIN="@example.com"
-
 ##NAME: POP3DSTART:0
-#
-# POP3DSTART is not referenced anywhere in the standard Courier programs
-# or scripts.  Rather, this is a convenient flag to be read by your system
-# startup script in /etc/rc.d, like this:
-#
-#  . @sysconfdir@/pop3d
-#  case x$POP3DSTART in
-#  x[yY]*)
-#        @libexecdir@/pop3d.rc start
-#        ;;
-#  esac
-#
-# The default setting is going to be NO, until Courier is shipped by default
-# with enough platforms so that people get annoyed with having to flip it to
-# YES every time.
-
-POP3DSTART=NO
 
 ##NAME: MAILDIRPATH:0
 #
--- courier-imap-4.1.1/imap/pop3d-ssl.dist.in.avx	2006-12-30 23:59:04.000000000 -0700
+++ courier-imap-4.1.1/imap/pop3d-ssl.dist.in	2006-12-31 00:00:18.000000000 -0700
@@ -8,62 +8,38 @@
 #  Copyright 2000-2004 Double Precision, Inc.  See COPYING for
 #  distribution information.
 #
+# pop3ds configuration file for courier-imap on Annvix
+#
+# This configuration file is used instead of what you might be used to
+# seeing from a "stock" courier-imap because Annvix does not use
+# couriertcpd but uses tcpsvd instead so a lot of the stuff that is
+# normally here is not required
+#
+# Many options can be set in /var/service/courier-pop3ds/env using the
+# environment files MAX_CONN, MAX_MEM, IP, and PORT:
+#
+#   MAX_MEM	 - the maximum memory to allow courier-imap to use
+#   MAX_CONN	 - the maximum number of concurrent connections
+#   MAX_PER_HOST - the maximum number of conncurrent connections per host
+#		   (the old MAXPERIP setting)
+#   IP		 - the IP address to bind to (default: 0 (all))
+#		   (the old ADDRESS setting)
+#   PORT	 - the TCP port to listen to (default: 143)
+#
+#
 #  This configuration file sets various options for the Courier-IMAP server
 #  when used to handle SSL POP3 connections.
 #
-#  SSL and non-SSL connections are handled by a dedicated instance of the
-#  couriertcpd daemon.  If you are accepting both SSL and non-SSL POP3
-#  connections, you will start two instances of couriertcpd, one on the
-#  POP3 port 110, and another one on the POP3-SSL port 995.
-#
 #  Download OpenSSL from http://www.openssl.org/
 #
+# use ./env/PORT
 ##NAME: SSLPORT:0
-#
-#  Options in the pop3d-ssl configuration file AUGMENT the options in the
-#  pop3d configuration file.  First the pop3d configuration file is read,
-#  then the pop3d-ssl configuration file, so we do not have to redefine
-#  anything.
-#
-#  However, some things do have to be redefined.  The port number is
-#  specified by SSLPORT, instead of PORT.  The default port is port 995.
-#
-#  Multiple port numbers can be separated by commas.  When multiple port
-#  numbers are used it is possibly to select a specific IP address for a
-#  given port as "ip.port".  For example, "127.0.0.1.900,192.68.0.1.900"
-#  accepts connections on port 900 on IP addresses 127.0.0.1 and 192.68.0.1
-#  The SSLADDRESS setting is a default for ports that do not have
-#  a specified IP address.
-
-SSLPORT=995
-
+# use ./env/IP
 ##NAME: SSLADDRESS:0
-#
-#  Address to listen on, can be set to a single IP address.
-#
-# SSLADDRESS=127.0.0.1
-
-SSLADDRESS=0
-
+# unused
 ##NAME: SSLPIDFILE:0
-#
-#
-#
-
-SSLPIDFILE=@piddir@/pop3d-ssl.pid
-
 ##NAME: SSLLOGGEROPTS:0
-#
-# courierlogger(1) options.                                              
-#
-
-SSLLOGGEROPTS="-name=pop3d-ssl"
-
 ##NAME: POP3DSSLSTART:0
-#
-#  Whether or not to start POP3 over SSL on spop3 port:
-
-POP3DSSLSTART=NO
 
 ##NAME: POP3_STARTTLS:0
 #
