--- rpm-4.2/rpmio/Makefile.am.mad-relink	2003-07-08 13:03:14.000000000 -0400
+++ rpm-4.2/rpmio/Makefile.am	2003-07-08 13:03:14.000000000 -0400
@@ -22,7 +22,7 @@ noinst_HEADERS = rpmio_internal.h
 
 BEECRYPTLOBJS = $(shell cat $(top_builddir)/beecrypt/listobjs)
 
-LDFLAGS = -L$(DESTDIR)$(usrlibdir)
+LDFLAGS = -L$(RPM_BUILD_ROOT)$(usrlibdir) -L$(DESTDIR)$(usrlibdir)
 
 usrlibdir = $(libdir)@MARK64@
 usrlib_LTLIBRARIES = librpmio.la
--- rpm-4.2/rpmio/Makefile.in.mad-relink	2003-07-08 13:03:14.000000000 -0400
+++ rpm-4.2/rpmio/Makefile.in	2003-07-08 13:03:14.000000000 -0400
@@ -227,7 +227,7 @@ noinst_HEADERS = rpmio_internal.h
 
 BEECRYPTLOBJS = $(shell cat $(top_builddir)/beecrypt/listobjs)
 
-LDFLAGS = -L$(DESTDIR)$(usrlibdir)
+LDFLAGS = -L$(RPM_BUILD_ROOT)$(usrlibdir) -L$(DESTDIR)$(usrlibdir)
 
 usrlibdir = $(libdir)@MARK64@
 usrlib_LTLIBRARIES = librpmio.la
--- rpm-4.2/rpmdb/Makefile.am.mad-relink	2003-07-08 13:03:14.000000000 -0400
+++ rpm-4.2/rpmdb/Makefile.am	2003-07-08 13:03:14.000000000 -0400
@@ -38,7 +38,7 @@ LIBS =
 
 DB3LOBJS = $(shell cat $(top_builddir)/$(WITH_DB_SUBDIR)/db3lobjs)
 
-LDFLAGS = -L$(DESTDIR)$(usrlibdir)
+LDFLAGS = -L$(RPM_BUILD_ROOT)$(usrlibdir) -L$(DESTDIR)$(usrlibdir)
 
 usrlibdir = $(libdir)@MARK64@
 usrlib_LTLIBRARIES = librpmdb.la
--- rpm-4.2/rpmdb/Makefile.in.mad-relink	2003-07-08 13:03:14.000000000 -0400
+++ rpm-4.2/rpmdb/Makefile.in	2003-07-08 13:03:14.000000000 -0400
@@ -244,7 +244,7 @@ LIBS = 
 
 DB3LOBJS = $(shell cat $(top_builddir)/$(WITH_DB_SUBDIR)/db3lobjs)
 
-LDFLAGS = -L$(DESTDIR)$(usrlibdir)
+LDFLAGS = -L$(RPM_BUILD_ROOT)$(usrlibdir) -L$(DESTDIR)$(usrlibdir)
 
 usrlibdir = $(libdir)@MARK64@
 usrlib_LTLIBRARIES = librpmdb.la
--- rpm-4.2/lib/Makefile.am.mad-relink	2003-07-08 13:03:14.000000000 -0400
+++ rpm-4.2/lib/Makefile.am	2003-07-08 13:03:14.000000000 -0400
@@ -24,7 +24,7 @@ noinst_HEADERS = \
 mylibs = librpm.la
 LIBS =
 
-LDFLAGS = -L$(DESTDIR)$(usrlibdir)
+LDFLAGS = -L$(RPM_BUILD_ROOT)$(usrlibdir) -L$(DESTDIR)$(usrlibdir)
 
 usrlibdir = $(libdir)@MARK64@
 usrlib_LTLIBRARIES = librpm.la
--- rpm-4.2/lib/Makefile.in.mad-relink	2003-07-08 13:03:14.000000000 -0400
+++ rpm-4.2/lib/Makefile.in	2003-07-08 13:03:14.000000000 -0400
@@ -230,7 +230,7 @@ noinst_HEADERS = \
 mylibs = librpm.la
 LIBS = 
 
-LDFLAGS = -L$(DESTDIR)$(usrlibdir)
+LDFLAGS = -L$(RPM_BUILD_ROOT)$(usrlibdir) -L$(DESTDIR)$(usrlibdir)
 
 usrlibdir = $(libdir)@MARK64@
 usrlib_LTLIBRARIES = librpm.la
--- rpm-4.2/build/Makefile.am.mad-relink	2003-07-08 13:03:14.000000000 -0400
+++ rpm-4.2/build/Makefile.am	2003-07-08 13:03:14.000000000 -0400
@@ -19,7 +19,7 @@ pkgincdir = $(pkgincludedir)
 pkginc_HEADERS = rpmbuild.h rpmfc.h rpmfile.h rpmspec.h
 noinst_HEADERS = buildio.h
 
-LDFLAGS = -L$(DESTDIR)$(usrlibdir)
+LDFLAGS = -L$(RPM_BUILD_ROOT)$(usrlibdir) -L$(DESTDIR)$(usrlibdir)
 
 usrlibdir = $(libdir)@MARK64@
 usrlib_LTLIBRARIES = librpmbuild.la
--- rpm-4.2/build/Makefile.in.mad-relink	2003-07-08 13:03:14.000000000 -0400
+++ rpm-4.2/build/Makefile.in	2003-07-08 13:03:14.000000000 -0400
@@ -223,7 +223,7 @@ pkgincdir = $(pkgincludedir)
 pkginc_HEADERS = rpmbuild.h rpmfc.h rpmfile.h rpmspec.h
 noinst_HEADERS = buildio.h
 
-LDFLAGS = -L$(DESTDIR)$(usrlibdir)
+LDFLAGS = -L$(RPM_BUILD_ROOT)$(usrlibdir) -L$(DESTDIR)$(usrlibdir)
 
 usrlibdir = $(libdir)@MARK64@
 usrlib_LTLIBRARIES = librpmbuild.la
