--- rpm-4.2/lib/rpmal.c.rpmal-fix-crash	2003-03-03 19:17:30.000000000 +0100
+++ rpm-4.2/lib/rpmal.c	2003-06-04 20:24:19.000000000 +0200
@@ -752,7 +752,7 @@ rpmalAllSatisfiesDepend(const rpmal al, 
 
     if (al->list != NULL)	/* XXX always true */
     for (ret = NULL, found = 0;
-	 match <= ai->index + ai->size && indexcmp(match, needle) == 0;
+	 match < ai->index + ai->size && indexcmp(match, needle) == 0;
 	 match++)
     {
 	alp = al->list + alKey2Num(al, match->pkgKey);
