--- libXt-1.0.2/include/X11/VendorP.h.orig	2006-08-28 18:39:47.000000000 -0300
+++ libXt-1.0.2/include/X11/VendorP.h	2006-08-28 18:47:03.000000000 -0300
@@ -68,6 +68,10 @@
 #ifndef  _XtVendorPrivate_h
 #define _XtVendorPrivate_h
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
 #include <X11/Vendor.h>
 
 /* New fields for the VendorShell widget class record */
@@ -100,4 +104,8 @@
 	VendorShellPart	vendor;
 } VendorShellRec, *VendorShellWidget;
 
+#ifdef __cplusplus
+}
+#endif
+
 #endif  /* _XtVendorPrivate_h */
