--- php-4.3.0/pear/System.php	2002-12-21 21:43:19.000000000 -0400
+++ php-4.3.0.mdk/pear/System.php	2003-01-04 09:46:04.000000000 -0400
@@ -233,7 +233,7 @@
                     $dir = dirname($dir);
                 }
                 while ($newdir = array_shift($dirstack)) {
-                    if (!call_user_func('mkdir', $newdir, $mode)) {
+                    if (!@call_user_func('mkdir', $newdir, $mode)) {
                         $ret = false;
                     }
                 }
