--- ./lib/psm.c.syslog	2005-01-04 17:35:05.000000000 +0100
+++ ./lib/psm.c	2005-04-25 17:20:05.827325744 +0200
@@ -2104,6 +2104,16 @@
 	    }
 	}
 	(void) rpmswExit(rpmtsOp(ts, RPMTS_OP_DBADD), 0);
+#if HAVE_SYSLOG_H
+        {
+	  char *s;
+	  
+	  s = headerSprintf(fi->h, "%{NAME}-%{VERSION}-%{RELEASE}",
+			    rpmTagTable, rpmHeaderFormats, NULL);
+	  syslog(LOG_NOTICE, "[RPM] %s installed\n", s);
+	  s = _free(s);
+	 }
+#endif
 	break;
     case PSM_RPMDB_REMOVE:
 	if (rpmtsFlags(ts) & RPMTRANS_FLAG_TEST)	break;
@@ -2138,6 +2148,16 @@
 	}
 
 	(void) rpmswExit(rpmtsOp(ts, RPMTS_OP_DBREMOVE), 0);
+#if HAVE_SYSLOG_H
+        {
+	  char *s;
+	  
+	  s = headerSprintf(fi->h, "%{NAME}-%{VERSION}-%{RELEASE}",
+			    rpmTagTable, rpmHeaderFormats, NULL);
+	  syslog(LOG_NOTICE, "[RPM] %s removed\n", s);
+	  s = _free(s);
+	}
+#endif
 	break;
 
     default:
--- ./system.h.syslog	2005-02-05 22:13:36.000000000 +0100
+++ ./system.h	2005-04-25 17:20:05.888315329 +0200
@@ -9,6 +9,10 @@
 #include "config.h"
 #endif
 
+#if HAVE_SYSLOG_H
+#include <syslog.h>
+#endif
+
 #include <sys/types.h>
 
 #if defined(__LCLINT__)
--- ./configure.ac.syslog	2005-04-25 17:46:11.021257674 +0200
+++ ./configure.ac	2005-04-25 17:47:06.284808990 +0200
@@ -372,6 +372,12 @@
 fi
 
 dnl
+dnl Checking syslog
+dnl
+
+AC_CHECK_HEADERS(syslog.h)
+
+dnl
 dnl Check for features
 dnl
 
