--- rpm-4.2.2/scripts/brp-redhat.in.avx	2004-08-04 12:57:59.000000000 -0600
+++ rpm-4.2.2/scripts/brp-redhat.in	2004-08-04 12:58:42.000000000 -0600
@@ -1,17 +1,17 @@
 #!/bin/sh
 
-# These are the build root policies that Mandrake Linux invokes at the end
+# These are the build root policies that Annvix invokes at the end
 # of the %install scriptlet.
 
 if [ -x /usr/share/spec-helper/spec-helper ];then
     /usr/share/spec-helper/spec-helper
 else
-    # Compress man pages (Mandrake Linux uses GNU bzip2)
+    # Compress man pages (Annvix uses GNU bzip2)
     @RPMCONFIGDIR@/brp-compress
 
-    # Strip ELF binaries (Mandrake Linux uses GNU binutils).
+    # Strip ELF binaries (Annvix uses GNU binutils).
     @RPMCONFIGDIR@/brp-strip
 
-    # Strip even more sections (Mandrake Linux uses GNU binutils).
+    # Strip even more sections (Annvix uses GNU binutils).
     @RPMCONFIGDIR@/brp-strip-comment-note
 fi
--- rpm-4.2.3/platform.in.annvix	2005-07-26 02:17:07.000000000 -0400
+++ rpm-4.2.3/platform.in	2005-07-26 02:20:30.000000000 -0400
@@ -103,10 +103,10 @@
 @mandrake@%_localstatedir	%{_var}/lib
 @mandrake@%_sysconfdir		/etc
 @mandrake@
-@mandrake@# XXX A copy of brp-mandrake, please. I'd suggest break down by
+@mandrake@# XXX A copy of brp-annvix, please. I'd suggest break down by
 @mandrake@#     functionality rather than vendor <shrug>.
 @mandrake@%__os_install_post    \
-@mandrake@    @RPMCONFIGDIR@/brp-mandrake \
+@mandrake@    @RPMCONFIGDIR@/brp-annvix \
 @mandrake@%{nil}
 @mandrake@
 @mandrake@%configure2_5x \
@@ -217,7 +217,7 @@
 @mandrake@%{nil}
 @mandrake@
 @mandrake@%_add_user_helper /usr/share/rpm-helper/add-user
-@mandrake@%_pre_useradd() %_add_user_helper %{name} $1 %{1} %{2} %{3} \
+@mandrake@%_pre_useradd() %_add_user_helper %{name} $1 %{1} %{2} %{3} %{4} \
 @mandrake@%{nil}
 @mandrake@
 @mandrake@%_del_user_helper /usr/share/rpm-helper/del-user
@@ -225,7 +225,7 @@
 @mandrake@%{nil}
 @mandrake@
 @mandrake@%_add_group_helper /usr/share/rpm-helper/add-group
-@mandrake@%_pre_groupadd() %_add_group_helper %{name} $1 %{1} %{?2:%2} \
+@mandrake@%_pre_groupadd() %_add_group_helper %{name} $1 %{1} %{2} %{?3:%3} \
 @mandrake@%{nil}
 @mandrake@
 @mandrake@%_del_group_helper /usr/share/rpm-helper/del-group
@@ -239,6 +239,31 @@
 @mandrake@%_del_shell_helper /usr/share/rpm-helper/del-shell
 @mandrake@%_preun_shelldel() %_del_shell_helper %{name} $1 \
 @mandrake@%{nil}
-@mandrake@
-@mandrake@%mkrel(c:) %{-c:0.%{-c*}.}%{!?_with_unstable:%(perl -e '$_="%{1}";m/(.*)(\\d+)$/;$rel=$2-1;re;print "$1$rel";').%{?subrel:%subrel}%{!?subrel:1}.%{?distversion:%distversion}%{?!distversion:%(echo $[%{mdkversion}/10])}}%{?_with_unstable:%{1}}%{?distsuffix:%distsuffix}%{?!distsuffix:mdk}
 
+@mandrake@#---------------------------------------------------------------------
+@mandrake@#	Annvix vendor specific macros
+@mandrake@# 	Vincent Danen <vdanen@annvix.org>
+@mandrake@#
+@mandrake@#
+@mandrake@%_srvdir		%{_var}/service
+@mandrake@%_srvlogdir		%{_var}/log/supervise
+@mandrake@%_vendor		Annvix
+@mandrake@%_real_vendor		annvix
+@mandrake@%_target_platform	%{_target_cpu}-%{_real_vendor}-%{_target_os}%{?_gnu}
+@mandrake@%_buildroot		/override
+@mandrake@%_ext			avx
+@mandrake@
+#@mandrake@# Default to building with ssp
+#@mandrake@%build_ssp %([ -n "$STACK_PROTECTOR" ] && echo 1 || echo 0)
+#@mandrake@
+@mandrake@%_add_srv_helper /usr/share/rpm-helper/add-srv
+@mandrake@%_post_srv() %_add_srv_helper %{name} $1 %{1} \
+@mandrake@%{nil}
+@mandrake@
+@mandrake@%_del_srv_helper /usr/share/rpm-helper/del-srv
+@mandrake@%_preun_srv() %_del_srv_helper %{name} $1 %{1} \
+@mandrake@%{nil}
+@mandrake@
+@mandrake@%_mkafterboot	if [[ -f %{_datadir}/afterboot/mkafterboot ]]; then %{_datadir}/afterboot/mkafterboot; fi \
+@mandrake@%{nil}
+@mandrake@
--- rpm-4.2.3/rpmrc.in.avx	2005-07-26 02:25:06.000000000 -0400
+++ rpm-4.2.3/rpmrc.in	2005-07-26 02:26:26.000000000 -0400
@@ -14,7 +14,7 @@
 optflags: i386 -O2 -fomit-frame-pointer -pipe -march=i386 %{debugcflags}
 optflags: i486 -O2 -fomit-frame-pointer -pipe -march=i486 %{debugcflags}
 optflags: k6   -O2 -fomit-frame-pointer -pipe -march=k6 %{debugcflags}
-optflags: i586 -O2 -fomit-frame-pointer -pipe -march=i586 -mcpu=pentiumpro %{debugcflags}
+optflags: i586 -O2 -fomit-frame-pointer -pipe -march=i586 -mtune=pentiumpro %{debugcflags}
 optflags: i686 -O2 -fomit-frame-pointer -pipe -march=i686 %{debugcflags}
 optflags: athlon -O2 -fomit-frame-pointer -pipe -march=athlon %{debugcflags}
 optflags: ia64 -O2 -pipe %{debugcflags}
