diff -Naur php-4.3.10/acinclude.m4 php-4.3.10.oden/acinclude.m4
--- php-4.3.10/acinclude.m4	2004-12-11 12:17:21.000000000 +0100
+++ php-4.3.10.oden/acinclude.m4	2004-12-16 10:13:03.295235088 +0100
@@ -180,7 +180,7 @@
   unset ac_new_flags
   for i in [$]$1; do
     case [$]i in
-    -L/usr/lib|-L/usr/lib/[)] ;;
+    -L/usr/lib|-L/usr/lib/|-L/usr/lib64|-L/usr/lib64/) ;;
     *[)] ac_new_flags="[$]ac_new_flags [$]i" ;;
     esac
   done
@@ -836,7 +836,7 @@
 dnl add a library to linkpath/runpath
 dnl
 AC_DEFUN([PHP_ADD_LIBPATH],[
-  if test "$1" != "/usr/lib"; then
+  if test "$1" != "/usr/lib" && test "$1" != "/usr/lib64"; then
     PHP_EXPAND_PATH($1, ai_p)
     ifelse([$2],,[
       _PHP_ADD_LIBPATH_GLOBAL([$ai_p])
@@ -1539,6 +1539,40 @@
   	fi
 ])
 
+dnl PHP_CHECK_LIBNAME
+dnl
+dnl Determine libdir name.
+dnl
+AC_DEFUN([PHP_CHECK_LIBNAME], [
+  AC_REQUIRE([AC_CANONICAL_HOST])dnl
+  AC_REQUIRE([AC_CANONICAL_BUILD])dnl
+  AC_MSG_CHECKING([libdir name])
+  case $host in
+  *-*-linux*)
+ # Test if the compiler is 64bit
+ echo 'int i;' > conftest.$ac_ext
+ php_cv_cc_64bit_output=no
+ if AC_TRY_EVAL(ac_compile); then
+   case `/usr/bin/file conftest.$ac_objext` in
+   *"ELF 64"*)
+     php_cv_cc_64bit_output=yes
+     ;;
+   esac
+ fi
+ rm -rf conftest*
+ ;;
+  esac
+  
+  case $host_cpu:$php_cv_cc_64bit_output in
+  powerpc64:yes | s390x:yes | sparc64:yes | x86_64:yes)
+ PHP_LIBNAME="lib64"
+ ;;
+  *:*)
+ PHP_LIBNAME="lib"
+ ;;
+  esac
+  AC_MSG_RESULT([$PHP_LIBNAME])
+])
 
 dnl
 dnl PHP_CHECK_LIBRARY(library, function [, action-found [, action-not-found [, extra-libs]]])
@@ -1626,8 +1660,8 @@
       if test -r $i/include/openssl/evp.h; then
         OPENSSL_INCDIR=$i/include
       fi
-      if test -r $i/lib/libssl.a -o -r $i/lib/libssl.$SHLIB_SUFFIX_NAME; then
-        OPENSSL_LIBDIR=$i/lib
+      if test -r $i/$PHP_LIBNAME/libssl.a -o -r $i/$PHP_LIBNAME/libssl.$SHLIB_SUFFIX_NAME; then
+        OPENSSL_LIBDIR=$i/$PHP_LIBNAME
       fi
       test -n "$OPENSSL_INCDIR" && test -n "$OPENSSL_LIBDIR" && break
     done
@@ -1748,8 +1782,8 @@
       AC_MSG_ERROR([Please specify the install prefix of iconv with --with-iconv=<DIR>])
     fi
   
-    if test -f $ICONV_DIR/lib/lib$iconv_lib_name.a ||
-       test -f $ICONV_DIR/lib/lib$iconv_lib_name.$SHLIB_SUFFIX_NAME
+    if test -f $ICONV_DIR/$PHP_LIBNAME/lib$iconv_lib_name.a ||
+       test -f $ICONV_DIR/$PHP_LIBNAME/lib$iconv_lib_name.$SHLIB_SUFFIX_NAME
     then
       PHP_CHECK_LIBRARY($iconv_lib_name, libiconv, [
         found_iconv=yes
@@ -1759,10 +1793,10 @@
           found_iconv=yes
           PHP_DEFINE(HAVE_ICONV,1,[ext/iconv])
         ], [], [
-          -L$ICONV_DIR/lib
+          -L$ICONV_DIR/$PHP_LIBNAME
         ])
       ], [
-        -L$ICONV_DIR/lib
+        -L$ICONV_DIR/$PHP_LIBNAME
       ])
     fi
   fi
diff -Naur php-4.3.10/aclocal.m4 php-4.3.10.oden/aclocal.m4
--- php-4.3.10/aclocal.m4	2004-12-14 18:55:17.000000000 +0100
+++ php-4.3.10.oden/aclocal.m4	2004-12-16 10:13:03.297234784 +0100
@@ -180,7 +180,7 @@
   unset ac_new_flags
   for i in [$]$1; do
     case [$]i in
-    -L/usr/lib|-L/usr/lib/[)] ;;
+    -L/usr/lib|-L/usr/lib/|-L/usr/lib64|-L/usr/lib64/) ;;
     *[)] ac_new_flags="[$]ac_new_flags [$]i" ;;
     esac
   done
@@ -836,7 +836,7 @@
 dnl add a library to linkpath/runpath
 dnl
 AC_DEFUN([PHP_ADD_LIBPATH],[
-  if test "$1" != "/usr/lib"; then
+  if test "$1" != "/usr/lib" && test "$1" != "/usr/lib64"; then
     PHP_EXPAND_PATH($1, ai_p)
     ifelse([$2],,[
       _PHP_ADD_LIBPATH_GLOBAL([$ai_p])
@@ -1539,6 +1539,41 @@
   	fi
 ])
 
+dnl
+dnl PHP_CHECK_LIBNAME
+dnl
+dnl Determine libdir name.
+dnl
+AC_DEFUN([PHP_CHECK_LIBNAME], [
+  AC_REQUIRE([AC_CANONICAL_HOST])dnl
+  AC_REQUIRE([AC_CANONICAL_BUILD])dnl
+  AC_MSG_CHECKING([libdir name])
+  case $host in
+  *-*-linux*)
+ # Test if the compiler is 64bit
+ echo 'int i;' > conftest.$ac_ext
+ php_cv_cc_64bit_output=no
+ if AC_TRY_EVAL(ac_compile); then
+   case `/usr/bin/file conftest.$ac_objext` in
+   *"ELF 64"*)
+     php_cv_cc_64bit_output=yes
+     ;;
+   esac
+ fi
+ rm -rf conftest*
+ ;;
+  esac
+  
+  case $host_cpu:$php_cv_cc_64bit_output in
+  powerpc64:yes | s390x:yes | sparc64:yes | x86_64:yes)
+ PHP_LIBNAME="lib64"
+ ;;
+  *:*)
+ PHP_LIBNAME="lib"
+ ;;
+  esac
+  AC_MSG_RESULT([$PHP_LIBNAME])
+])
 
 dnl
 dnl PHP_CHECK_LIBRARY(library, function [, action-found [, action-not-found [, extra-libs]]])
@@ -1626,8 +1661,8 @@
       if test -r $i/include/openssl/evp.h; then
         OPENSSL_INCDIR=$i/include
       fi
-      if test -r $i/lib/libssl.a -o -r $i/lib/libssl.$SHLIB_SUFFIX_NAME; then
-        OPENSSL_LIBDIR=$i/lib
+      if test -r $i/$PHP_LIBNAME/libssl.a -o -r $i/$PHP_LIBNAME/libssl.$SHLIB_SUFFIX_NAME; then
+        OPENSSL_LIBDIR=$i/$PHP_LIBNAME
       fi
       test -n "$OPENSSL_INCDIR" && test -n "$OPENSSL_LIBDIR" && break
     done
@@ -1748,8 +1783,8 @@
       AC_MSG_ERROR([Please specify the install prefix of iconv with --with-iconv=<DIR>])
     fi
   
-    if test -f $ICONV_DIR/lib/lib$iconv_lib_name.a ||
-       test -f $ICONV_DIR/lib/lib$iconv_lib_name.$SHLIB_SUFFIX_NAME
+    if test -f $ICONV_DIR/$PHP_LIBNAME/lib$iconv_lib_name.a ||
+       test -f $ICONV_DIR/$PHP_LIBNAME/lib$iconv_lib_name.$SHLIB_SUFFIX_NAME
     then
       PHP_CHECK_LIBRARY($iconv_lib_name, libiconv, [
         found_iconv=yes
@@ -1759,10 +1794,10 @@
           found_iconv=yes
           PHP_DEFINE(HAVE_ICONV,1,[ext/iconv])
         ], [], [
-          -L$ICONV_DIR/lib
+          -L$ICONV_DIR/$PHP_LIBNAME
         ])
       ], [
-        -L$ICONV_DIR/lib
+        -L$ICONV_DIR/$PHP_LIBNAME
       ])
     fi
   fi
diff -Naur php-4.3.10/configure.in php-4.3.10.oden/configure.in
--- php-4.3.10/configure.in	2004-12-14 17:07:49.000000000 +0100
+++ php-4.3.10.oden/configure.in	2004-12-16 10:13:03.297234784 +0100
@@ -128,6 +128,10 @@
 PHP_RUNPATH_SWITCH
 
 PHP_PROG_RE2C
+
+dnl check for libdir name
+PHP_CHECK_LIBNAME
+
 AC_PROG_RANLIB
 AC_PROG_LN_S
 AC_PROG_AWK
diff -Naur php-4.3.10/ext/bz2/config.m4 php-4.3.10.oden/ext/bz2/config.m4
--- php-4.3.10/ext/bz2/config.m4	2004-01-21 06:04:50.000000000 +0100
+++ php-4.3.10.oden/ext/bz2/config.m4	2004-12-16 10:41:29.092914424 +0100
@@ -27,12 +27,12 @@
   PHP_CHECK_LIBRARY(bz2, BZ2_bzerror, 
   [
     PHP_ADD_INCLUDE($BZIP_DIR/include)
-    PHP_ADD_LIBRARY_WITH_PATH(bz2, $BZIP_DIR/lib, BZ2_SHARED_LIBADD)
+    PHP_ADD_LIBRARY_WITH_PATH(bz2, $BZIP_DIR/$PHP_LIBNAME, BZ2_SHARED_LIBADD)
     AC_DEFINE(HAVE_BZ2,1,[ ])
   ], [
     AC_MSG_ERROR(bz2 module requires libbz2 >= 1.0.0)
   ], [
-    -L$BZIP_DIR/lib
+    -L$BZIP_DIR/$PHP_LIBNAME
   ])
 
   PHP_NEW_EXTENSION(bz2, bz2.c, $ext_shared)
diff -Naur php-4.3.10/ext/cpdf/config.m4 php-4.3.10.oden/ext/cpdf/config.m4
--- php-4.3.10/ext/cpdf/config.m4	2003-10-01 04:53:53.000000000 +0200
+++ php-4.3.10.oden/ext/cpdf/config.m4	2004-12-16 10:41:29.093914272 +0100
@@ -4,7 +4,7 @@
   AC_ARG_WITH(jpeg-dir,
   [  --with-jpeg-dir[=DIR]     CPDF: Set the path to libjpeg install prefix.],[
     for i in $withval /usr/local /usr; do
-      if test -f "$i/lib/libjpeg.$SHLIB_SUFFIX_NAME" -o -f "$i/lib/libjpeg.a"; then
+      if test -f "$i/$PHP_LIBNAME/libjpeg.$SHLIB_SUFFIX_NAME" -o -f "$i/$PHP_LIBNAME/libjpeg.a"; then
         CPDF_JPEG_DIR=$i
         break;
       fi
@@ -14,11 +14,11 @@
       AC_MSG_ERROR([libjpeg.(a|so) not found.])
     fi
     PHP_CHECK_LIBRARY(jpeg, jpeg_read_header, [
-      PHP_ADD_LIBRARY_WITH_PATH(jpeg, $CPDF_JPEG_DIR/lib, CPDF_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(jpeg, $CPDF_JPEG_DIR/$PHP_LIBNAME, CPDF_SHARED_LIBADD)
     ] ,[
       AC_MSG_ERROR([CPDF: Problem with libjpeg.(a|so). Please check config.log for more information.])
     ], [
-      -L$CPDF_JPEG_DIR/lib
+      -L$CPDF_JPEG_DIR/$PHP_LIBNAME
     ])
   ],)
 ])
@@ -27,7 +27,7 @@
   AC_ARG_WITH(tiff-dir,
   [  --with-tiff-dir[=DIR]     CPDF: Set the path to libtiff install prefix.],[
     for i in $withval /usr/local /usr; do
-      if test -f "$i/lib/libtiff.$SHLIB_SUFFIX_NAME" -o -f "$i/lib/libtiff.a"; then
+      if test -f "$i/$PHP_LIBNAME/libtiff.$SHLIB_SUFFIX_NAME" -o -f "$i/$PHP_LIBNAME/libtiff.a"; then
         CPDF_TIFF_DIR=$i
         break;
       fi
@@ -37,11 +37,11 @@
       AC_MSG_ERROR([libtiff.(a|so) not found.])
     fi
     PHP_CHECK_LIBRARY(tiff, TIFFOpen, [
-      PHP_ADD_LIBRARY_WITH_PATH(tiff, $CPDF_TIFF_DIR/lib, CPDF_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(tiff, $CPDF_TIFF_DIR/$PHP_LIBNAME, CPDF_SHARED_LIBADD)
     ] ,[
       AC_MSG_ERROR([CPDF: Problem with libtiff.(a|so). Please check config.log for more information.])
     ], [
-      -L$CPDF_TIFF_DIR/lib
+      -L$CPDF_TIFF_DIR/$PHP_LIBNAME
     ])
   ],)
 ])
@@ -66,12 +66,12 @@
 
         PHP_CHECK_LIBRARY(cpdf, cpdf_open, [
           PHP_ADD_INCLUDE($CPDFLIB_INCLUDE)
-          PHP_ADD_LIBRARY_WITH_PATH(cpdf, $i/lib, CPDF_SHARED_LIBADD)
+          PHP_ADD_LIBRARY_WITH_PATH(cpdf, $i/$PHP_LIBNAME, CPDF_SHARED_LIBADD)
           AC_DEFINE(HAVE_CPDFLIB,1,[Whether you have cpdflib])
         ], [
           AC_MSG_ERROR([Cpdflib module requires cpdflib >= 2.])
         ], [
-          -L$i/lib $CPDF_SHARED_LIBADD
+          -L$i/$PHP_LIBNAME $CPDF_SHARED_LIBADD
         ])
         break
       fi
diff -Naur php-4.3.10/ext/crack/config.m4 php-4.3.10.oden/ext/crack/config.m4
--- php-4.3.10/ext/crack/config.m4	2003-10-01 04:53:54.000000000 +0200
+++ php-4.3.10.oden/ext/crack/config.m4	2004-12-16 10:41:29.101913056 +0100
@@ -7,7 +7,7 @@
 
 if test "$PHP_CRACK" != "no"; then
 
-  for i in $PHP_CRACK/lib $PHP_CRACK/cracklib /usr/local/lib /usr/lib; do
+  for i in $PHP_CRACK/$PHP_LIBNAME $PHP_CRACK/cracklib /usr/local/$PHP_LIBNAME /usr/$PHP_LIBNAME; do
     test -f $i/libcrack.$SHLIB_SUFFIX_NAME -o -f $i/libcrack.a && CRACK_LIBDIR=$i && break
   done
 
diff -Naur php-4.3.10/ext/curl/config.m4 php-4.3.10.oden/ext/curl/config.m4
--- php-4.3.10/ext/curl/config.m4	2003-03-23 19:16:12.000000000 +0100
+++ php-4.3.10.oden/ext/curl/config.m4	2004-12-16 10:41:29.093914272 +0100
@@ -51,7 +51,7 @@
 
   PHP_ADD_INCLUDE($CURL_DIR/include)
   PHP_EVAL_LIBLINE($CURL_LIBS, CURL_SHARED_LIBADD)
-  PHP_ADD_LIBRARY_WITH_PATH(curl, $CURL_DIR/lib, CURL_SHARED_LIBADD)
+  PHP_ADD_LIBRARY_WITH_PATH(curl, $CURL_DIR/$PHP_LIBNAME, CURL_SHARED_LIBADD)
 
   PHP_CHECK_LIBRARY(curl,curl_easy_perform, 
   [ 
@@ -59,14 +59,14 @@
   ],[
     AC_MSG_ERROR(There is something wrong. Please check config.log for more information.)
   ],[
-    $CURL_LIBS -L$CURL_DIR/lib
+    $CURL_LIBS -L$CURL_DIR/$PHP_LIBNAME
   ])
 
   PHP_CHECK_LIBRARY(curl,curl_version_info,
   [
     AC_DEFINE(HAVE_CURL_VERSION_INFO,1,[ ])
   ],[],[
-    $CURL_LIBS -L$CURL_DIR/lib
+    $CURL_LIBS -L$CURL_DIR/$PHP_LIBNAME
   ])
 
 dnl  if test "$PHP_CURLWRAPPERS" != "no" ; then
diff -Naur php-4.3.10/ext/cyrus/config.m4 php-4.3.10.oden/ext/cyrus/config.m4
--- php-4.3.10/ext/cyrus/config.m4	2003-10-03 07:25:32.000000000 +0200
+++ php-4.3.10.oden/ext/cyrus/config.m4	2004-12-16 10:41:29.101913056 +0100
@@ -16,7 +16,7 @@
   for i in $PHP_CYRUS /usr/local /usr; do
     if test -r $i/include/cyrus/imclient.h && test "$found_cyrus" = "no"; then
       PHP_ADD_INCLUDE($i/include)
-      PHP_ADD_LIBRARY_WITH_PATH(cyrus, $i/lib, CYRUS_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(cyrus, $i/$PHP_LIBNAME, CYRUS_SHARED_LIBADD)
       found_cyrus=yes
       CYRUS_INCPATH=$i/include
       break 2
@@ -50,10 +50,10 @@
     fi
 
     if test "$found_sasl" = "yes"; then
-      if test -f $i/lib/libsasl2.a || test -f $i/lib/libsasl2.$SHLIB_SUFFIX_NAME; then
-        PHP_ADD_LIBRARY_WITH_PATH(sasl2, $i/lib, CYRUS_SHARED_LIBADD)
+      if test -f $i/$PHP_LIBNAME/libsasl2.a || test -f $i/$PHP_LIBNAME/libsasl2.$SHLIB_SUFFIX_NAME; then
+        PHP_ADD_LIBRARY_WITH_PATH(sasl2, $i/$PHP_LIBNAME, CYRUS_SHARED_LIBADD)
       else
-        PHP_ADD_LIBRARY_WITH_PATH(sasl, $i/lib, CYRUS_SHARED_LIBADD)
+        PHP_ADD_LIBRARY_WITH_PATH(sasl, $i/$PHP_LIBNAME, CYRUS_SHARED_LIBADD)
       fi
       break 2
     fi
@@ -70,8 +70,8 @@
   if test "$PHP_OPENSSL" = "no"; then
     for i in $PHP_CYRUS /usr/local /usr; do
       if test -r $i/include/openssl/ssl.h; then
-        PHP_ADD_LIBRARY_WITH_PATH(ssl, $i/lib, CYRUS_SHARED_LIBADD)
-        PHP_ADD_LIBRARY_WITH_PATH(crypto, $i/lib, CYRUS_SHARED_LIBADD)
+        PHP_ADD_LIBRARY_WITH_PATH(ssl, $i/$PHP_LIBNAME, CYRUS_SHARED_LIBADD)
+        PHP_ADD_LIBRARY_WITH_PATH(crypto, $i/$PHP_LIBNAME, CYRUS_SHARED_LIBADD)
         break 2
       fi
     done
diff -Naur php-4.3.10/ext/dba/config.m4 php-4.3.10.oden/ext/dba/config.m4
--- php-4.3.10/ext/dba/config.m4	2004-09-25 17:32:20.000000000 +0200
+++ php-4.3.10.oden/ext/dba/config.m4	2004-12-16 10:41:29.092914424 +0100
@@ -21,7 +21,7 @@
 dnl Assign INCLUDE/LFLAGS from PREFIX
 AC_DEFUN(PHP_DBA_STD_ASSIGN,[
   if test -n "$THIS_PREFIX" && test "$THIS_PREFIX" != "/usr"; then
-    THIS_LFLAGS=$THIS_PREFIX/lib
+    THIS_LFLAGS=$THIS_PREFIX/$PHP_LIBNAME
   fi
 ])
 
@@ -86,7 +86,7 @@
         AC_DEFINE_UNQUOTED(GDBM_INCLUDE_FILE, "$THIS_INCLUDE", [ ])
         AC_DEFINE(DBA_GDBM, 1, [ ]) 
         THIS_LIBS=gdbm
-      ], [], [-L$THIS_PREFIX/lib])
+      ], [], [-L$THIS_PREFIX/$PHP_LIBNAME])
     fi
     
     PHP_DBA_STD_ASSIGN
@@ -118,7 +118,7 @@
           AC_DEFINE_UNQUOTED(NDBM_INCLUDE_FILE, "$THIS_INCLUDE", [ ])
           AC_DEFINE(DBA_NDBM, 1, [ ]) 
           THIS_LIBS=$LIB
-        ], [], [-L$THIS_PREFIX/lib])
+        ], [], [-L$THIS_PREFIX/$PHP_LIBNAME])
         if test -n "$THIS_LIBS"; then
           break
         fi
@@ -136,8 +136,8 @@
 dnl parameters(version, library list, function)
 AC_DEFUN(PHP_DBA_DB_CHECK,[
   for LIB in $2; do
-    if test -f $THIS_PREFIX/lib/lib$LIB.a -o -f $THIS_PREFIX/lib/lib$LIB.$SHLIB_SUFFIX_NAME; then
-      PHP_TEMP_LDFLAGS(-L$THIS_PREFIX/lib, -l$LIB,[
+    if test -f $THIS_PREFIX/$PHP_LIBNAME/lib$LIB.a -o -f $THIS_PREFIX/$PHP_LIBNAME/lib$LIB.$SHLIB_SUFFIX_NAME; then
+      PHP_TEMP_LDFLAGS(-L$THIS_PREFIX/$PHP_LIBNAME, -l$LIB,[
         AC_TRY_LINK([
 #include "$THIS_INCLUDE"
         ],[
@@ -329,7 +329,7 @@
           fi
           AC_DEFINE(DBA_DBM, 1, [ ]) 
           THIS_LIBS=$LIB
-        ], [], [-L$THIS_PREFIX/lib])
+        ], [], [-L$THIS_PREFIX/$PHP_LIBNAME])
         if test -n "$THIS_LIBS"; then
           break
         fi
@@ -371,7 +371,7 @@
           AC_DEFINE_UNQUOTED(CDB_INCLUDE_FILE, "$THIS_INCLUDE", [ ])
           AC_DEFINE(DBA_CDB, 1, [ ]) 
           THIS_LIBS=$LIB
-        ], [], [-L$THIS_PREFIX/lib])
+        ], [], [-L$THIS_PREFIX/$PHP_LIBNAME])
         if test -n "$THIS_LIBS"; then
           break
         fi
diff -Naur php-4.3.10/ext/domxml/config.m4 php-4.3.10.oden/ext/domxml/config.m4
--- php-4.3.10/ext/domxml/config.m4	2002-10-30 18:42:36.000000000 +0100
+++ php-4.3.10.oden/ext/domxml/config.m4	2004-12-16 10:41:29.104912600 +0100
@@ -50,7 +50,7 @@
 
   PHP_DOM_CHECK_VERSION
 
-  if test -f $DOMXML_DIR/lib/libxml2.a -o -f $DOMXML_DIR/lib/libxml2.$SHLIB_SUFFIX_NAME ; then
+  if test -f $DOMXML_DIR/$PHP_LIBNAME/libxml2.a -o -f $DOMXML_DIR/$PHP_LIBNAME/libxml2.$SHLIB_SUFFIX_NAME ; then
     DOM_LIBNAME=xml2
   else
     DOM_LIBNAME=xml
@@ -62,7 +62,7 @@
     DOM_LIBS=`$XML2_CONFIG --libs`
     PHP_EVAL_LIBLINE($DOM_LIBS, DOMXML_SHARED_LIBADD)
   else 
-    PHP_ADD_LIBRARY_WITH_PATH($DOM_LIBNAME, $DOMXML_DIR/lib, DOMXML_SHARED_LIBADD)
+    PHP_ADD_LIBRARY_WITH_PATH($DOM_LIBNAME, $DOMXML_DIR/$PHP_LIBNAME, DOMXML_SHARED_LIBADD)
   fi
 
   PHP_ADD_INCLUDE($DOMXML_DIR/include$DOMXML_DIR_ADD)
@@ -70,7 +70,7 @@
   if test "$PHP_ZLIB_DIR" = "no"; then
     AC_MSG_ERROR(DOMXML requires ZLIB. Use --with-zlib-dir=<DIR>)
   else
-    PHP_ADD_LIBRARY_WITH_PATH(z, $PHP_ZLIB_DIR/lib, DOMXML_SHARED_LIBADD)
+    PHP_ADD_LIBRARY_WITH_PATH(z, $PHP_ZLIB_DIR/$PHP_LIBNAME, DOMXML_SHARED_LIBADD)
   fi
   
   AC_DEFINE(HAVE_DOMXML,1,[ ])
@@ -137,7 +137,7 @@
 
   PHP_DOM_XSLT_CHECK_VERSION
 
-  PHP_ADD_LIBRARY_WITH_PATH(xslt, $DOMXSLT_DIR/lib, DOMXML_SHARED_LIBADD)
+  PHP_ADD_LIBRARY_WITH_PATH(xslt, $DOMXSLT_DIR/$PHP_LIBNAME, DOMXML_SHARED_LIBADD)
 
   PHP_ADD_INCLUDE($DOMXSLT_DIR/include)
 
@@ -145,17 +145,17 @@
     AC_MSG_ERROR(DOMXSLT requires DOMXML. Use --with-dom=<DIR>)
   fi
   
-  if test -f $DOMXML_DIR/lib/libxml2.a -o -f $DOMXML_DIR/lib/libxml2.$SHLIB_SUFFIX_NAME ; then
+  if test -f $DOMXML_DIR/$PHP_LIBNAME/libxml2.a -o -f $DOMXML_DIR/$PHP_LIBNAME/libxml2.$SHLIB_SUFFIX_NAME ; then
     DOM_LIBNAME=xml2
   else
     DOM_LIBNAME=xml
   fi
 
-  PHP_ADD_LIBRARY_WITH_PATH($DOM_LIBNAME, $DOMXML_DIR/lib, DOMXML_SHARED_LIBADD)
+  PHP_ADD_LIBRARY_WITH_PATH($DOM_LIBNAME, $DOMXML_DIR/$PHP_LIBNAME, DOMXML_SHARED_LIBADD)
   PHP_ADD_INCLUDE($DOMXML_DIR/include$DOMXML_DIR_ADD)
 
-  if test -f $DOMXML_DIR/lib/libxsltbreakpoint.a -o -f $DOMXML_DIR/lib/libxsltbreakpoint.$SHLIB_SUFFIX_NAME ; then
-    PHP_ADD_LIBRARY_WITH_PATH(xsltbreakpoint, $DOMXML_DIR/lib, DOMXML_SHARED_LIBADD)
+  if test -f $DOMXML_DIR/$PHP_LIBNAME/libxsltbreakpoint.a -o -f $DOMXML_DIR/$PHP_LIBNAME/libxsltbreakpoint.$SHLIB_SUFFIX_NAME ; then
+    PHP_ADD_LIBRARY_WITH_PATH(xsltbreakpoint, $DOMXML_DIR/$PHP_LIBNAME, DOMXML_SHARED_LIBADD)
   fi
 
   AC_DEFINE(HAVE_DOMXSLT,1,[ ])
@@ -187,7 +187,7 @@
 
   PHP_DOM_EXSLT_CHECK_VERSION
 
-  PHP_ADD_LIBRARY_WITH_PATH(exslt, $DOMEXSLT_DIR/lib, DOMXML_SHARED_LIBADD)
+  PHP_ADD_LIBRARY_WITH_PATH(exslt, $DOMEXSLT_DIR/$PHP_LIBNAME, DOMXML_SHARED_LIBADD)
 
   PHP_ADD_INCLUDE($DOMEXSLT_DIR/include)
 
diff -Naur php-4.3.10/ext/fbsql/config.m4 php-4.3.10.oden/ext/fbsql/config.m4
--- php-4.3.10/ext/fbsql/config.m4	2003-10-01 04:53:55.000000000 +0200
+++ php-4.3.10.oden/ext/fbsql/config.m4	2004-12-16 10:41:29.102912904 +0100
@@ -13,7 +13,7 @@
   FBSQL_INSTALLATION_DIR=""
   if test "$PHP_FBSQL" = "yes"; then
 
-    for i in /Local/Library /usr/local /usr /opt /Library /usr/lib; do
+    for i in /Local/Library /usr/local /usr /opt /Library /usr/$PHP_LIBNAME; do
       if test -f $i/FrontBase/include/FBCAccess/FBCAccess.h; then
         FBSQL_INSTALLATION_DIR=$i/FrontBase
         break
@@ -33,11 +33,11 @@
     fi
   fi  
 
-  if test ! -r "$FBSQL_INSTALLATION_DIR/lib/libFBCAccess.a"; then
-     AC_MSG_ERROR(Could not find $FBSQL_INSTALLATION_DIR/lib/libFBCAccess.a)
+  if test ! -r "$FBSQL_INSTALLATION_DIR/$PHP_LIBNAME/libFBCAccess.a"; then
+     AC_MSG_ERROR(Could not find $FBSQL_INSTALLATION_DIR/$PHP_LIBNAME/libFBCAccess.a)
   fi
 
-  PHP_ADD_LIBRARY_WITH_PATH(FBCAccess, $FBSQL_INSTALLATION_DIR/lib, FBSQL_SHARED_LIBADD)
+  PHP_ADD_LIBRARY_WITH_PATH(FBCAccess, $FBSQL_INSTALLATION_DIR/$PHP_LIBNAME, FBSQL_SHARED_LIBADD)
   PHP_ADD_INCLUDE($FBSQL_INSTALLATION_DIR/include)
   PHP_SUBST(FBSQL_SHARED_LIBADD)
 fi
diff -Naur php-4.3.10/ext/gd/config.m4 php-4.3.10.oden/ext/gd/config.m4
--- php-4.3.10/ext/gd/config.m4	2004-07-27 14:44:44.000000000 +0200
+++ php-4.3.10.oden/ext/gd/config.m4	2004-12-16 10:41:29.092914424 +0100
@@ -51,7 +51,7 @@
   if test "$PHP_JPEG_DIR" != "no"; then
 
     for i in $PHP_JPEG_DIR /usr/local /usr; do
-      test -f $i/lib/libjpeg.$SHLIB_SUFFIX_NAME -o -f $i/lib/libjpeg.a && GD_JPEG_DIR=$i && break
+      test -f $i/$PHP_LIBNAME/libjpeg.$SHLIB_SUFFIX_NAME -o -f $i/$PHP_LIBNAME/libjpeg.a && GD_JPEG_DIR=$i && break
     done
 
     if test -z "$GD_JPEG_DIR"; then
@@ -61,11 +61,11 @@
     PHP_CHECK_LIBRARY(jpeg,jpeg_read_header,
     [
       PHP_ADD_INCLUDE($GD_JPEG_DIR/include)
-      PHP_ADD_LIBRARY_WITH_PATH(jpeg, $GD_JPEG_DIR/lib, GD_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(jpeg, $GD_JPEG_DIR/$PHP_LIBNAME, GD_SHARED_LIBADD)
     ],[
       AC_MSG_ERROR([Problem with libjpeg.(a|so). Please check config.log for more information.]) 
     ],[
-      -L$GD_JPEG_DIR/lib
+      -L$GD_JPEG_DIR/$PHP_LIBNAME
     ])
   else 
     AC_MSG_RESULT([If configure fails try --with-jpeg-dir=<DIR>])
@@ -76,7 +76,7 @@
   if test "$PHP_PNG_DIR" != "no"; then
 
     for i in $PHP_PNG_DIR /usr/local /usr; do
-      test -f $i/lib/libpng.$SHLIB_SUFFIX_NAME -o -f $i/lib/libpng.a && GD_PNG_DIR=$i && break
+      test -f $i/$PHP_LIBNAME/libpng.$SHLIB_SUFFIX_NAME -o -f $i/$PHP_LIBNAME/libpng.a && GD_PNG_DIR=$i && break
     done
 
     if test -z "$GD_PNG_DIR"; then
@@ -94,12 +94,12 @@
     PHP_CHECK_LIBRARY(png,png_write_image,
     [
       PHP_ADD_INCLUDE($GD_PNG_DIR/include)
-      PHP_ADD_LIBRARY_WITH_PATH(z, $PHP_ZLIB_DIR/lib, GD_SHARED_LIBADD)
-      PHP_ADD_LIBRARY_WITH_PATH(png, $GD_PNG_DIR/lib, GD_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(z, $PHP_ZLIB_DIR/$PHP_LIBNAME, GD_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(png, $GD_PNG_DIR/$PHP_LIBNAME, GD_SHARED_LIBADD)
     ],[
       AC_MSG_ERROR([Problem with libpng.(a|so) or libz.(a|so). Please check config.log for more information.]) 
     ],[
-      -L$PHP_ZLIB_DIR/lib -lz -L$GD_PNG_DIR/lib
+      -L$PHP_ZLIB_DIR/$PHP_LIBNAME -lz -L$GD_PNG_DIR/$PHP_LIBNAME
     ])
 
   else 
@@ -111,7 +111,7 @@
   if test "$PHP_XPM_DIR" != "no"; then
 
     for i in $PHP_XPM_DIR /usr/local /usr/X11R6 /usr; do
-      test -f $i/lib/libXpm.$SHLIB_SUFFIX_NAME -o -f $i/lib/libXpm.a && GD_XPM_DIR=$i && break
+      test -f $i/$PHP_LIBNAME/libXpm.$SHLIB_SUFFIX_NAME -o -f $i/$PHP_LIBNAME/libXpm.a && GD_XPM_DIR=$i && break
     done
 
     if test -z "$GD_XPM_DIR"; then
@@ -129,12 +129,12 @@
     PHP_CHECK_LIBRARY(Xpm,XpmFreeXpmImage, 
     [
       PHP_ADD_INCLUDE($GD_XPM_INC)
-      PHP_ADD_LIBRARY_WITH_PATH(Xpm, $GD_XPM_DIR/lib, GD_SHARED_LIBADD)
-      PHP_ADD_LIBRARY_WITH_PATH(X11, $GD_XPM_DIR/lib, GD_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(Xpm, $GD_XPM_DIR/$PHP_LIBNAME, GD_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(X11, $GD_XPM_DIR/$PHP_LIBNAME, GD_SHARED_LIBADD)
     ],[
       AC_MSG_ERROR([Problem with libXpm.(a|so) or libX11.(a|so). Please check config.log for more information.]) 
     ],[
-      -L$GD_XPM_DIR/lib -lX11
+      -L$GD_XPM_DIR/$PHP_LIBNAME -lX11
     ])
   else 
     AC_MSG_RESULT(If configure fails try --with-xpm-dir=<DIR>)
@@ -163,7 +163,7 @@
       fi
       if test -n "$TTF_DIR" ; then
         AC_DEFINE(HAVE_LIBTTF,1,[ ])
-        PHP_ADD_LIBRARY_WITH_PATH(ttf, $TTF_DIR/lib, GD_SHARED_LIBADD)
+        PHP_ADD_LIBRARY_WITH_PATH(ttf, $TTF_DIR/$PHP_LIBNAME, GD_SHARED_LIBADD)
       fi
       if test -z "$TTF_INC_DIR"; then
         TTF_INC_DIR=$TTF_DIR/include
@@ -187,7 +187,7 @@
     done
     
     if test -n "$FREETYPE2_DIR" ; then
-      PHP_ADD_LIBRARY_WITH_PATH(freetype, $FREETYPE2_DIR/lib, GD_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(freetype, $FREETYPE2_DIR/$PHP_LIBNAME, GD_SHARED_LIBADD)
       PHP_ADD_INCLUDE($FREETYPE2_DIR/include)
       PHP_ADD_INCLUDE($FREETYPE2_INC_DIR)
       AC_DEFINE(USE_GD_IMGSTRTTF, 1, [ ])
@@ -215,11 +215,11 @@
     [
       AC_DEFINE(HAVE_LIBT1,1,[ ])
       PHP_ADD_INCLUDE($GD_T1_DIR/include)
-      PHP_ADD_LIBRARY_WITH_PATH(t1, $GD_T1_DIR/lib, GD_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(t1, $GD_T1_DIR/$PHP_LIBNAME, GD_SHARED_LIBADD)
     ],[
       AC_MSG_ERROR([Problem with libt1.(a|so). Please check config.log for more information.]) 
     ],[
-      -L$GD_T1_DIR/lib
+      -L$GD_T1_DIR/$PHP_LIBNAME
     ])
   fi
 ])
@@ -362,7 +362,7 @@
   done
 
 dnl Library path
-  for i in lib/gd1.3 lib/gd lib gd1.3 gd ""; do
+  for i in lib/gd1.3 lib/gd $PHP_LIBNAME gd1.3 gd ""; do
     test -f "$PHP_GD/$i/libgd.$SHLIB_SUFFIX_NAME" -o -f "$PHP_GD/$i/libgd.a" && GD_LIB="$PHP_GD/$i"
   done
 
diff -Naur php-4.3.10/ext/gettext/config.m4 php-4.3.10.oden/ext/gettext/config.m4
--- php-4.3.10/ext/gettext/config.m4	2003-10-01 04:53:58.000000000 +0200
+++ php-4.3.10.oden/ext/gettext/config.m4	2004-12-16 10:41:29.104912600 +0100
@@ -14,7 +14,7 @@
     AC_MSG_ERROR(Cannot locate header file libintl.h)
   fi
 
-  GETTEXT_LIBDIR=$GETTEXT_DIR/lib
+  GETTEXT_LIBDIR=$GETTEXT_DIR/$PHP_LIBNAME
   GETTEXT_INCDIR=$GETTEXT_DIR/include
   
   O_LDFLAGS=$LDFLAGS
diff -Naur php-4.3.10/ext/gmp/config.m4 php-4.3.10.oden/ext/gmp/config.m4
--- php-4.3.10/ext/gmp/config.m4	2003-11-19 05:44:24.000000000 +0100
+++ php-4.3.10.oden/ext/gmp/config.m4	2004-12-16 10:41:29.092914424 +0100
@@ -21,13 +21,13 @@
     [],[
       AC_MSG_ERROR([GNU MP Library version 4.1.2 or greater required.])
     ],[
-      -L$GMP_DIR/lib
+      -L$GMP_DIR/$PHP_LIBNAME
     ])
   ],[
-    -L$GMP_DIR/lib
+    -L$GMP_DIR/$PHP_LIBNAME
   ])
 
-  PHP_ADD_LIBRARY_WITH_PATH(gmp, $GMP_DIR/lib, GMP_SHARED_LIBADD)
+  PHP_ADD_LIBRARY_WITH_PATH(gmp, $GMP_DIR/$PHP_LIBNAME, GMP_SHARED_LIBADD)
   PHP_ADD_INCLUDE($GMP_DIR/include)
 
   PHP_NEW_EXTENSION(gmp, gmp.c, $ext_shared)
diff -Naur php-4.3.10/ext/hwapi/config.m4 php-4.3.10.oden/ext/hwapi/config.m4
--- php-4.3.10/ext/hwapi/config.m4	2002-09-19 22:14:33.000000000 +0200
+++ php-4.3.10.oden/ext/hwapi/config.m4	2004-12-16 10:41:29.102912904 +0100
@@ -23,7 +23,7 @@
 
   PHP_ADD_INCLUDE($HWAPI_DIR/include/hwapi)
   PHP_SUBST(HWAPI_SHARED_LIBADD)
-  PHP_ADD_LIBRARY_WITH_PATH(hwapi, $HWAPI_DIR/lib, HWAPI_SHARED_LIBADD)
+  PHP_ADD_LIBRARY_WITH_PATH(hwapi, $HWAPI_DIR/$PHP_LIBNAME, HWAPI_SHARED_LIBADD)
   AC_DEFINE(HAVE_HWAPI,1,[ ])
 
   PHP_REQUIRE_CXX
diff -Naur php-4.3.10/ext/iconv/config.m4 php-4.3.10.oden/ext/iconv/config.m4
--- php-4.3.10/ext/iconv/config.m4	2004-12-06 21:37:56.000000000 +0100
+++ php-4.3.10.oden/ext/iconv/config.m4	2004-12-16 10:41:29.102912904 +0100
@@ -24,7 +24,7 @@
     fi
 
     CFLAGS="-I$PHP_ICONV_PREFIX/include $CFLAGS"
-    LDFLAGS="-L$PHP_ICONV_PREFIX/lib $LDFLAGS"
+    LDFLAGS="-L$PHP_ICONV_PREFIX/$PHP_LIBNAME $LDFLAGS"
 
     if test -r $PHP_ICONV_PREFIX/include/giconv.h; then
       PHP_ICONV_H_PATH="$PHP_ICONV_PREFIX/include/giconv.h"
diff -Naur php-4.3.10/ext/imap/config.m4 php-4.3.10.oden/ext/imap/config.m4
--- php-4.3.10/ext/imap/config.m4	2004-01-17 01:01:06.000000000 +0100
+++ php-4.3.10.oden/ext/imap/config.m4	2004-12-16 10:41:29.093914272 +0100
@@ -71,7 +71,7 @@
     fi
 
     for i in $SEARCH_PATHS; do
-      if test -f $i/lib/libkrb5.a || test -f $i/lib/libkrb5.$SHLIB_SUFFIX_NAME; then
+      if test -f $i/$PHP_LIBNAME/libkrb5.a || test -f $i/$PHP_LIBNAME/libkrb5.$SHLIB_SUFFIX_NAME; then
         PHP_KERBEROS_DIR=$i
         break
       fi
@@ -84,7 +84,7 @@
       ])
     fi
     AC_DEFINE(HAVE_IMAP_KRB,1,[ ])
-    PHP_ADD_LIBPATH($PHP_KERBEROS_DIR/lib, IMAP_SHARED_LIBADD)
+    PHP_ADD_LIBPATH($PHP_KERBEROS_DIR/$PHP_LIBNAME, IMAP_SHARED_LIBADD)
     PHP_ADD_LIBRARY(gssapi_krb5, 1, IMAP_SHARED_LIBADD)
     PHP_ADD_LIBRARY(krb5, 1, IMAP_SHARED_LIBADD)
     PHP_ADD_LIBRARY(k5crypto, 1, IMAP_SHARED_LIBADD)
@@ -116,11 +116,11 @@
   AC_MSG_CHECKING([whether SSL libraries are needed for c-client])
 
   if test "$PHP_IMAP_SSL" != "no"; then
-    AC_MSG_RESULT([$PHP_IMAP_SSL/lib])
+    AC_MSG_RESULT([$PHP_IMAP_SSL/$PHP_LIBNAME])
     AC_DEFINE(HAVE_IMAP_SSL,1,[ ])
     PHP_ADD_LIBRARY_DEFER(ssl,    1, IMAP_SHARED_LIBADD)
     PHP_ADD_LIBRARY_DEFER(crypto, 1, IMAP_SHARED_LIBADD)
-    PHP_ADD_LIBPATH($PHP_IMAP_SSL/lib, IMAP_SHARED_LIBADD)
+    PHP_ADD_LIBPATH($PHP_IMAP_SSL/$PHP_LIBNAME, IMAP_SHARED_LIBADD)
   else
     AC_MSG_RESULT(no)
   fi
@@ -178,13 +178,13 @@
 
     if test -r "$IMAP_DIR/c-client/c-client.a"; then
       ln -s "$IMAP_DIR/c-client/c-client.a" "$IMAP_DIR/c-client/libc-client.a" >/dev/null 2>&1
-    elif test -r "$IMAP_DIR/lib/c-client.a"; then
-      ln -s "$IMAP_DIR/lib/c-client.a" "$IMAP_DIR/lib/libc-client.a" >/dev/null 2>&1
+    elif test -r "$IMAP_DIR/$PHP_LIBNAME/c-client.a"; then
+      ln -s "$IMAP_DIR/$PHP_LIBNAME/c-client.a" "$IMAP_DIR/$PHP_LIBNAME/libc-client.a" >/dev/null 2>&1
     fi
 
     for lib in c-client4 c-client imap; do
       IMAP_LIB=$lib
-      IMAP_LIB_CHK(lib)
+      IMAP_LIB_CHK($PHP_LIBNAME)
       IMAP_LIB_CHK(c-client)
     done
 
diff -Naur php-4.3.10/ext/ingres_ii/config.m4 php-4.3.10.oden/ext/ingres_ii/config.m4
--- php-4.3.10/ext/ingres_ii/config.m4	2003-01-23 06:33:17.000000000 +0100
+++ php-4.3.10.oden/ext/ingres_ii/config.m4	2004-12-16 10:41:29.103912752 +0100
@@ -23,10 +23,10 @@
     AC_MSG_ERROR(Cannot find iiapi.h under $II_DIR/files)
   fi
 
-  if test -r $II_DIR/lib/libiiapi.a; then
-    II_LIB_DIR=$II_DIR/lib
+  if test -r $II_DIR/$PHP_LIBNAME/libiiapi.a; then
+    II_LIB_DIR=$II_DIR/$PHP_LIBNAME
   else
-    AC_MSG_ERROR(Cannot find libiiapi.a under $II_DIR/lib)
+    AC_MSG_ERROR(Cannot find libiiapi.a under $II_DIR/$PHP_LIBNAME)
   fi
 
   PHP_ADD_LIBRARY_WITH_PATH(iiapi, $II_LIB_DIR, II_SHARED_LIBADD)
diff -Naur php-4.3.10/ext/ircg/config.m4 php-4.3.10.oden/ext/ircg/config.m4
--- php-4.3.10/ext/ircg/config.m4	2003-01-21 06:09:33.000000000 +0100
+++ php-4.3.10.oden/ext/ircg/config.m4	2004-12-16 10:41:29.099913360 +0100
@@ -26,7 +26,7 @@
   
   PHP_EVAL_LIBLINE(`$IRCG_CONFIG --ldflags`)
   PHP_EVAL_INCLINE(`$IRCG_CONFIG --cppflags`)
-  PHP_ADD_LIBRARY_WITH_PATH(ircg, $PHP_IRCG/lib)
+  PHP_ADD_LIBRARY_WITH_PATH(ircg, $PHP_IRCG/$PHP_LIBNAME)
   PHP_ADD_INCLUDE($PHP_IRCG/include)
   if test "$PHP_SAPI" = "thttpd"; then
     AC_DEFINE(IRCG_WITH_THTTPD, 1, [Whether thttpd is available])
diff -Naur php-4.3.10/ext/java/config.m4 php-4.3.10.oden/ext/java/config.m4
--- php-4.3.10/ext/java/config.m4	2003-04-04 19:43:13.000000000 +0200
+++ php-4.3.10.oden/ext/java/config.m4	2004-12-16 10:41:29.099913360 +0100
@@ -57,17 +57,17 @@
 
 AC_DEFUN(JAVA_CHECK_LIB, [
  AC_MSG_CHECKING([Checking for libjava])
- if test -d $PHP_JAVA/lib/kaffe; then
-   PHP_ADD_LIBPATH($PHP_JAVA/lib)
+ if test -d $PHP_JAVA/$PHP_LIBNAME/kaffe; then
+   PHP_ADD_LIBPATH($PHP_JAVA/$PHP_LIBNAME)
    JAVA_CFLAGS=-DKAFFE
    JAVA_INCLUDE=-I$PHP_JAVA/include/kaffe
    JAVA_CLASSPATH=$PHP_JAVA/share/kaffe/Klasses.jar
    JAVA_LIB=kaffevm
-   JAVA_LIBPATH=$PHP_JAVA/lib/kaffe
+   JAVA_LIBPATH=$PHP_JAVA/$PHP_LIBNAME/kaffe
    java_libext=kaffevm
 
-   test -f $PHP_JAVA/lib/$JAVA_LIB       && JAVA_LIBPATH=$PHP_JAVA/lib
-   test -f $PHP_JAVA/lib/kaffe/$JAVA_LIB && JAVA_LIBPATH=$PHP_JAVA/lib/kaffe
+   test -f $PHP_JAVA/$PHP_LIBNAME/$JAVA_LIB       && JAVA_LIBPATH=$PHP_JAVA/$PHP_LIBNAME
+   test -f $PHP_JAVA/$PHP_LIBNAME/kaffe/$JAVA_LIB && JAVA_LIBPATH=$PHP_JAVA/$PHP_LIBNAME/kaffe
 
    dnl
    dnl accomodate old versions of kaffe which don't support jar
@@ -75,15 +75,15 @@
    if kaffe -version 2>&1 | grep 1.0b > /dev/null; then
      JAVA_JAR='zip -q0'
    fi
- elif test -f $PHP_JAVA/lib/$java_libext; then
+ elif test -f $PHP_JAVA/$PHP_LIBNAME/$java_libext; then
    JAVA_LIB=java
-   JAVA_LIBPATH=$PHP_JAVA/lib
+   JAVA_LIBPATH=$PHP_JAVA/$PHP_LIBNAME
    JAVA_INCLUDE=-I$PHP_JAVA/include
 
-   test -f $PHP_JAVA/lib/classes.zip && JAVA_CFLAGS=-DJNI_11
-   test -f $PHP_JAVA/lib/jvm.jar     && JAVA_CFLAGS=-DJNI_12
-   test -f $PHP_JAVA/lib/classes.zip && JAVA_CLASSPATH=$PHP_JAVA/lib/classes.zip
-   test -f $PHP_JAVA/lib/jvm.jar     && JAVA_CLASSPATH=$PHP_JAVA/lib/jvm.jar
+   test -f $PHP_JAVA/$PHP_LIBNAME/classes.zip && JAVA_CFLAGS=-DJNI_11
+   test -f $PHP_JAVA/$PHP_LIBNAME/jvm.jar     && JAVA_CFLAGS=-DJNI_12
+   test -f $PHP_JAVA/$PHP_LIBNAME/classes.zip && JAVA_CLASSPATH=$PHP_JAVA/$PHP_LIBNAME/classes.zip
+   test -f $PHP_JAVA/$PHP_LIBNAME/jvm.jar     && JAVA_CLASSPATH=$PHP_JAVA/$PHP_LIBNAME/jvm.jar
 
    for i in $PHP_JAVA/include/*; do
      test -f $i/jni_md.h && JAVA_INCLUDE="$JAVA_INCLUDE $i"
diff -Naur php-4.3.10/ext/ldap/config.m4 php-4.3.10.oden/ext/ldap/config.m4
--- php-4.3.10/ext/ldap/config.m4	2003-07-01 02:02:29.000000000 +0200
+++ php-4.3.10.oden/ext/ldap/config.m4	2004-12-16 10:41:29.100913208 +0100
@@ -6,15 +6,15 @@
   if test -f $1/include/ldap.h; then
     LDAP_DIR=$1
     LDAP_INCDIR=$1/include
-    LDAP_LIBDIR=$1/lib
+    LDAP_LIBDIR=$1/$PHP_LIBNAME
   elif test -f $1/include/umich-ldap/ldap.h; then
     LDAP_DIR=$1
     LDAP_INCDIR=$1/include/umich-ldap
-    LDAP_LIBDIR=$1/lib
+    LDAP_LIBDIR=$1/$PHP_LIBNAME
   elif test -f $1/ldap/public/ldap.h; then
     LDAP_DIR=$1
     LDAP_INCDIR=$1/ldap/public
-    LDAP_LIBDIR=$1/lib
+    LDAP_LIBDIR=$1/$PHP_LIBNAME
   fi
 ])
 
diff -Naur php-4.3.10/ext/mcal/config.m4 php-4.3.10.oden/ext/mcal/config.m4
--- php-4.3.10/ext/mcal/config.m4	2002-03-27 11:26:16.000000000 +0100
+++ php-4.3.10.oden/ext/mcal/config.m4	2004-12-16 10:41:29.100913208 +0100
@@ -19,7 +19,7 @@
     fi
   done
 
-  for i in $MCAL_DIR $MCAL_DIR/mcal $MCAL_DIR/mcal/lib $MCAL_DIR/lib/mcal $MCAL_DIR/lib; do
+  for i in $MCAL_DIR $MCAL_DIR/mcal $MCAL_DIR/mcal/$PHP_LIBNAME $MCAL_DIR/$PHP_LIBNAME/mcal $MCAL_DIR/$PHP_LIBNAME; do
     if test -r "$i/libmcal.a"; then
       MCAL_LIBRARY=$i
     fi
diff -Naur php-4.3.10/ext/mcrypt/config.m4 php-4.3.10.oden/ext/mcrypt/config.m4
--- php-4.3.10/ext/mcrypt/config.m4	2003-10-01 04:54:00.000000000 +0200
+++ php-4.3.10.oden/ext/mcrypt/config.m4	2004-12-16 10:41:29.104912600 +0100
@@ -23,7 +23,7 @@
     [
       AC_DEFINE(HAVE_MCRYPT_GENERIC_DEINIT,1,[ ])
     ],[],[
-      -L$MCRYPT_DIR/lib
+      -L$MCRYPT_DIR/$PHP_LIBNAME
     ])
 
   ],[
@@ -37,7 +37,7 @@
       [
         AC_DEFINE(HAVE_MCRYPT_GENERIC_DEINIT,1,[ ])
       ],[],[
-        -L$MCRYPT_DIR/lib
+        -L$MCRYPT_DIR/$PHP_LIBNAME
       ])
     ],[
       PHP_CHECK_LIBRARY(mcrypt, init_mcrypt, 
@@ -46,14 +46,14 @@
       ],[
         AC_MSG_ERROR([Sorry, I was not able to diagnose which libmcrypt version you have installed.])
       ],[
-        -L$MCRYPT_DIR/lib
+        -L$MCRYPT_DIR/$PHP_LIBNAME
       ])
     ],[])
   ],[
-    -L$MCRYPT_DIR/lib -lltdl
+    -L$MCRYPT_DIR/$PHP_LIBNAME -lltdl
   ])
 
-  PHP_ADD_LIBRARY_WITH_PATH(mcrypt, $MCRYPT_DIR/lib, MCRYPT_SHARED_LIBADD)
+  PHP_ADD_LIBRARY_WITH_PATH(mcrypt, $MCRYPT_DIR/$PHP_LIBNAME, MCRYPT_SHARED_LIBADD)
   PHP_ADD_INCLUDE($MCRYPT_DIR/include)
 
   PHP_NEW_EXTENSION(mcrypt, mcrypt.c, $ext_shared)
diff -Naur php-4.3.10/ext/mcve/config.m4 php-4.3.10.oden/ext/mcve/config.m4
--- php-4.3.10/ext/mcve/config.m4	2004-12-08 07:48:11.000000000 +0100
+++ php-4.3.10.oden/ext/mcve/config.m4	2004-12-16 10:41:29.100913208 +0100
@@ -68,7 +68,7 @@
   CPPFLAGS=$saved_CPPFLAGS
 
   PHP_ADD_INCLUDE($MCVE_DIR/include)
-  PHP_ADD_LIBRARY_WITH_PATH(mcve, $MCVE_DIR/lib, MCVE_SHARED_LIBADD)
+  PHP_ADD_LIBRARY_WITH_PATH(mcve, $MCVE_DIR/$PHP_LIBNAME, MCVE_SHARED_LIBADD)
 
   dnl Build test. Check for libmcve or libmonetra
   dnl libmonetra_compat.h handle's MCVE_ -> M_ conversion
diff -Naur php-4.3.10/ext/mhash/config.m4 php-4.3.10.oden/ext/mhash/config.m4
--- php-4.3.10/ext/mhash/config.m4	2003-10-01 04:54:01.000000000 +0200
+++ php-4.3.10.oden/ext/mhash/config.m4	2004-12-16 10:41:29.102912904 +0100
@@ -15,7 +15,7 @@
   fi
 
   PHP_ADD_INCLUDE($MHASH_DIR/include)
-  PHP_ADD_LIBRARY_WITH_PATH(mhash, $MHASH_DIR/lib, MHASH_SHARED_LIBADD)
+  PHP_ADD_LIBRARY_WITH_PATH(mhash, $MHASH_DIR/$PHP_LIBNAME, MHASH_SHARED_LIBADD)
 
   PHP_NEW_EXTENSION(mhash, mhash.c, $ext_shared)
   PHP_SUBST(MHASH_SHARED_LIBADD)
diff -Naur php-4.3.10/ext/ming/config.m4 php-4.3.10.oden/ext/ming/config.m4
--- php-4.3.10/ext/ming/config.m4	2003-07-11 09:48:33.000000000 +0200
+++ php-4.3.10.oden/ext/ming/config.m4	2004-12-16 10:41:29.100913208 +0100
@@ -9,7 +9,7 @@
   AC_CHECK_LIB(m, sin)
 
   for i in $PHP_MING /usr/local /usr; do
-    if test -f $i/lib/libming.$SHLIB_SUFFIX_NAME -o -f $i/lib/libming.a; then
+    if test -f $i/$PHP_LIBNAME/libming.$SHLIB_SUFFIX_NAME -o -f $i/$PHP_LIBNAME/libming.a; then
       MING_DIR=$i
     fi
   done
@@ -33,11 +33,11 @@
   ],[
     AC_MSG_ERROR([Ming library 0.2a or greater required.])
   ],[
-    -L$MING_DIR/lib
+    -L$MING_DIR/$PHP_LIBNAME
   ])
   
   PHP_ADD_INCLUDE($MING_INC_DIR)
-  PHP_ADD_LIBRARY_WITH_PATH(ming, $MING_DIR/lib, MING_SHARED_LIBADD)
+  PHP_ADD_LIBRARY_WITH_PATH(ming, $MING_DIR/$PHP_LIBNAME, MING_SHARED_LIBADD)
 
   AC_MSG_CHECKING([for destroySWFBlock])
   AC_TRY_RUN([
diff -Naur php-4.3.10/ext/msession/config.m4 php-4.3.10.oden/ext/msession/config.m4
--- php-4.3.10/ext/msession/config.m4	2002-03-12 17:25:11.000000000 +0100
+++ php-4.3.10.oden/ext/msession/config.m4	2004-12-16 10:41:29.106912296 +0100
@@ -6,12 +6,12 @@
 [  --with-msession[=DIR]   Include msession support])
  
 if test "$PHP_MSESSION" != "no"; then
-	if test -r $PHP_MSESSION/lib/libphoenix.a; then
+	if test -r $PHP_MSESSION/$PHP_LIBNAME/libphoenix.a; then
 		PHOENIX_DIR=$PHP_MSESSION
 	else
 		AC_MSG_CHECKING(for msession in default path)
 		for i in /usr/local/phoenix /usr/local /usr; do
-			if test -r $i/lib/libphoenix.a; then
+			if test -r $i/$PHP_LIBNAME/libphoenix.a; then
 				PHOENIX_DIR=$i
 				AC_MSG_RESULT(found in $i)
 			fi
@@ -25,7 +25,7 @@
 
 
   	AC_DEFINE(HAVE_MSESSION, 1, [ ])
-  	PHP_ADD_LIBRARY_WITH_PATH(phoenix, "$PHOENIX_DIR/lib", PHOENIX_LIB)
+  	PHP_ADD_LIBRARY_WITH_PATH(phoenix, "$PHOENIX_DIR/$PHP_LIBNAME", PHOENIX_LIB)
 	PHOENIX_INCLUDE="-I$PHOENIX_DIR"
 	PHP_SUBST(PHOENIX_LIB)
 
diff -Naur php-4.3.10/ext/msql/config.m4 php-4.3.10.oden/ext/msql/config.m4
--- php-4.3.10/ext/msql/config.m4	2002-03-12 17:25:28.000000000 +0100
+++ php-4.3.10.oden/ext/msql/config.m4	2004-12-16 10:41:29.100913208 +0100
@@ -31,7 +31,7 @@
       MSQL_LIBDIR=/usr/local/Hughes/lib
     else
       MSQL_INCDIR=$PHP_MSQL/include
-      MSQL_LIBDIR=$PHP_MSQL/lib
+      MSQL_LIBDIR=$PHP_MSQL/$PHP_LIBNAME
     fi
     PHP_ADD_LIBRARY_WITH_PATH(msql, $MSQL_LIBDIR, MSQL_SHARED_LIBADD)
     PHP_ADD_INCLUDE($MSQL_INCDIR)
diff -Naur php-4.3.10/ext/mssql/config.m4 php-4.3.10.oden/ext/mssql/config.m4
--- php-4.3.10/ext/mssql/config.m4	2004-12-08 19:59:11.000000000 +0100
+++ php-4.3.10.oden/ext/mssql/config.m4	2004-12-16 10:41:29.102912904 +0100
@@ -32,12 +32,12 @@
     fi
   fi  
 
-  if test ! -r "$FREETDS_INSTALLATION_DIR/lib/libtds.a" && test ! -r "$FREETDS_INSTALLATION_DIR/lib/libtds.so"; then
-     AC_MSG_ERROR(Could not find $FREETDS_INSTALLATION_DIR/lib/libtds.[a|so])
+  if test ! -r "$FREETDS_INSTALLATION_DIR/$PHP_LIBNAME/libtds.a" && test ! -r "$FREETDS_INSTALLATION_DIR/$PHP_LIBNAME/libtds.so"; then
+     AC_MSG_ERROR(Could not find $FREETDS_INSTALLATION_DIR/$PHP_LIBNAME/libtds.[a|so])
   fi
 
   MSSQL_INCDIR=$FREETDS_INSTALLATION_DIR/include
-  MSSQL_LIBDIR=$FREETDS_INSTALLATION_DIR/lib
+  MSSQL_LIBDIR=$FREETDS_INSTALLATION_DIR/$PHP_LIBNAME
 
   PHP_ADD_INCLUDE($MSSQL_INCDIR)
   PHP_ADD_LIBRARY_WITH_PATH(sybdb, $MSSQL_LIBDIR, MSSQL_SHARED_LIBADD)
diff -Naur php-4.3.10/ext/mysql/config.m4 php-4.3.10.oden/ext/mysql/config.m4
--- php-4.3.10/ext/mysql/config.m4	2003-12-31 12:28:38.000000000 +0100
+++ php-4.3.10.oden/ext/mysql/config.m4	2004-12-16 10:41:29.102912904 +0100
@@ -114,7 +114,7 @@
 
   MYSQL_MODULE_TYPE=external
 
-  for i in lib lib/mysql; do
+  for i in $PHP_LIBNAME $PHP_LIBNAME/mysql; do
     MYSQL_LIB_CHK($i)
   done
 
@@ -129,9 +129,9 @@
       PHP_CHECK_LIBRARY(mysqlclient, mysql_error, [], [
         AC_MSG_ERROR([mysql configure failed. Please check config.log for more information.])
       ], [
-        -L$PHP_ZLIB_DIR/lib -L$MYSQL_LIB_DIR 
+        -L$PHP_ZLIB_DIR/$PHP_LIBNAME -L$MYSQL_LIB_DIR 
       ])  
-      MYSQL_LIBS="-L$PHP_ZLIB_DIR/lib -lz"
+      MYSQL_LIBS="-L$PHP_ZLIB_DIR/$PHP_LIBNAME -lz"
     else
       PHP_ADD_LIBRARY(z,, MYSQL_SHARED_LIBADD)
       PHP_CHECK_LIBRARY(mysqlclient, mysql_errno, [], [
diff -Naur php-4.3.10/ext/ncurses/config.m4 php-4.3.10.oden/ext/ncurses/config.m4
--- php-4.3.10/ext/ncurses/config.m4	2003-12-09 21:33:22.000000000 +0100
+++ php-4.3.10.oden/ext/ncurses/config.m4	2004-12-16 10:41:29.104912600 +0100
@@ -40,18 +40,18 @@
 
    PHP_CHECK_LIBRARY($LIBNAME, $LIBSYMBOL, [
      AC_DEFINE(HAVE_NCURSESLIB,1,[ ])
-     PHP_ADD_LIBRARY_WITH_PATH($LIBNAME, $NCURSES_DIR/lib, NCURSES_SHARED_LIBADD)
+     PHP_ADD_LIBRARY_WITH_PATH($LIBNAME, $NCURSES_DIR/$PHP_LIBNAME, NCURSES_SHARED_LIBADD)
 
      PHP_CHECK_LIBRARY(panel, new_panel, [
        AC_DEFINE(HAVE_NCURSES_PANEL,1,[ ])
-       PHP_ADD_LIBRARY_WITH_PATH(panel, $NCURSES_DIR/lib, NCURSES_SHARED_LIBADD)
+       PHP_ADD_LIBRARY_WITH_PATH(panel, $NCURSES_DIR/$PHP_LIBNAME, NCURSES_SHARED_LIBADD)
      ], [], [ 
-       -L$NCURSES_DIR/lib -l$LIBNAME -lm
+       -L$NCURSES_DIR/$PHP_LIBNAME -l$LIBNAME -lm
      ])
    ], [
      AC_MSG_ERROR(Wrong ncurses lib version or lib not found)
    ], [
-     -L$NCURSES_DIR/lib -lm
+     -L$NCURSES_DIR/$PHP_LIBNAME -lm
    ])
  
    AC_CHECK_LIB($LIBNAME, color_set,   [AC_DEFINE(HAVE_NCURSES_COLOR_SET,  1, [ ])])
diff -Naur php-4.3.10/ext/oci8/config.m4 php-4.3.10.oden/ext/oci8/config.m4
--- php-4.3.10/ext/oci8/config.m4	2004-02-03 15:22:34.000000000 +0100
+++ php-4.3.10.oden/ext/oci8/config.m4	2004-12-16 10:41:29.100913208 +0100
@@ -23,16 +23,16 @@
   if test -s "$OCI8_DIR/orainst/unix.rgs"; then
     OCI8_VERSION=`grep '"ocommon"' $OCI8_DIR/orainst/unix.rgs | sed 's/[ ][ ]*/:/g' | cut -d: -f 6 | cut -c 2-4`
     test -z "$OCI8_VERSION" && OCI8_VERSION=7.3
-  elif test -f $OCI8_DIR/lib/libclntsh.$SHLIB_SUFFIX_NAME.10.1; then
+  elif test -f $OCI8_DIR/$PHP_LIBNAME/libclntsh.$SHLIB_SUFFIX_NAME.10.1; then
     OCI8_VERSION=10.1    
-  elif test -f $OCI8_DIR/lib/libclntsh.$SHLIB_SUFFIX_NAME.9.0; then
+  elif test -f $OCI8_DIR/$PHP_LIBNAME/libclntsh.$SHLIB_SUFFIX_NAME.9.0; then
     OCI8_VERSION=9.0
-  elif test -f $OCI8_DIR/lib/libclntsh.$SHLIB_SUFFIX_NAME.8.0; then
+  elif test -f $OCI8_DIR/$PHP_LIBNAME/libclntsh.$SHLIB_SUFFIX_NAME.8.0; then
     OCI8_VERSION=8.1
-  elif test -f $OCI8_DIR/lib/libclntsh.$SHLIB_SUFFIX_NAME.1.0; then
+  elif test -f $OCI8_DIR/$PHP_LIBNAME/libclntsh.$SHLIB_SUFFIX_NAME.1.0; then
     OCI8_VERSION=8.0
-  elif test -f $OCI8_DIR/lib/libclntsh.a; then 
-    if test -f $OCI8_DIR/lib/libcore4.a; then 
+  elif test -f $OCI8_DIR/$PHP_LIBNAME/libclntsh.a; then 
+    if test -f $OCI8_DIR/$PHP_LIBNAME/libcore4.a; then 
       OCI8_VERSION=8.0
     else
       OCI8_VERSION=8.1
@@ -72,10 +72,10 @@
     OCI8_INCLUDES="$OCI8_INCLUDES -I$OCI8_DIR/plsql/public"
   fi
 
-  if test -f "$OCI8_DIR/lib/sysliblist"; then
-    PHP_EVAL_LIBLINE(`cat $OCI8_DIR/lib/sysliblist`, OCI8_SYSLIB)
-  elif test -f "$OCI8_DIR/rdbms/lib/sysliblist"; then
-    PHP_EVAL_LIBLINE(`cat $OCI8_DIR/rdbms/lib/sysliblist`, OCI8_SYSLIB)
+  if test -f "$OCI8_DIR/$PHP_LIBNAME/sysliblist"; then
+    PHP_EVAL_LIBLINE(`cat $OCI8_DIR/$PHP_LIBNAME/sysliblist`, OCI8_SYSLIB)
+  elif test -f "$OCI8_DIR/rdbms/$PHP_LIBNAME/sysliblist"; then
+    PHP_EVAL_LIBLINE(`cat $OCI8_DIR/rdbms/$PHP_LIBNAME/sysliblist`, OCI8_SYSLIB)
   fi
 
   AC_OCI8_VERSION($OCI8_DIR)
@@ -84,12 +84,12 @@
       PHP_ADD_LIBRARY_WITH_PATH(nlsrtl3, "", OCI8_SHARED_LIBADD)
       PHP_ADD_LIBRARY_WITH_PATH(core4, "", OCI8_SHARED_LIBADD)
       PHP_ADD_LIBRARY_WITH_PATH(psa, "", OCI8_SHARED_LIBADD)
-      PHP_ADD_LIBRARY_WITH_PATH(clntsh, $OCI8_DIR/lib, OCI8_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(clntsh, $OCI8_DIR/$PHP_LIBNAME, OCI8_SHARED_LIBADD)
       ;;
 
     8.1)
       PHP_ADD_LIBRARY(clntsh, 1, OCI8_SHARED_LIBADD)
-      PHP_ADD_LIBPATH($OCI8_DIR/lib, OCI8_SHARED_LIBADD)
+      PHP_ADD_LIBPATH($OCI8_DIR/$PHP_LIBNAME, OCI8_SHARED_LIBADD)
 
       dnl 
       dnl OCI_ATTR_STATEMENT is not available in all 8.1.x versions
@@ -99,7 +99,7 @@
 
     9.0)
       PHP_ADD_LIBRARY(clntsh, 1, OCI8_SHARED_LIBADD)
-      PHP_ADD_LIBPATH($OCI8_DIR/lib, OCI8_SHARED_LIBADD)
+      PHP_ADD_LIBPATH($OCI8_DIR/$PHP_LIBNAME, OCI8_SHARED_LIBADD)
       AC_DEFINE(HAVE_OCI8_ATTR_STATEMENT,1,[ ])
 
       dnl These functions are only available in version >= 9.2
@@ -110,15 +110,15 @@
           AC_DEFINE(HAVE_OCI_9_2,1,[ ])
           OCI8_VERSION=9.2
         ], [], [
-          -L$OCI8_DIR/lib $OCI8_SHARED_LIBADD
+          -L$OCI8_DIR/$PHP_LIBNAME $OCI8_SHARED_LIBADD
         ])
       ], [], [
-        -L$OCI8_DIR/lib $OCI8_SHARED_LIBADD
+        -L$OCI8_DIR/$PHP_LIBNAME $OCI8_SHARED_LIBADD
       ])
       ;;
     10.1)
       PHP_ADD_LIBRARY(clntsh, 1, OCI8_SHARED_LIBADD)
-      PHP_ADD_LIBPATH($OCI8_DIR/lib, OCI8_SHARED_LIBADD)
+      PHP_ADD_LIBPATH($OCI8_DIR/$PHP_LIBNAME, OCI8_SHARED_LIBADD)
       AC_DEFINE(HAVE_OCI8_ATTR_STATEMENT,1,[ ])
       AC_DEFINE(HAVE_OCI_9_2,1,[ ])
       ;;
@@ -139,10 +139,10 @@
       PHP_ADD_LIBRARY(ocijdbc8, 1, OCI8_SHARED_LIBADD)
       AC_DEFINE(HAVE_OCI8_TEMP_LOB,1,[ ])
     ], [], [
-      -L$OCI8_DIR/lib $OCI8_SHARED_LIBADD
+      -L$OCI8_DIR/$PHP_LIBNAME $OCI8_SHARED_LIBADD
     ])
   ], [
-    -L$OCI8_DIR/lib $OCI8_SHARED_LIBADD
+    -L$OCI8_DIR/$PHP_LIBNAME $OCI8_SHARED_LIBADD
   ])
 
   PHP_NEW_EXTENSION(oci8, oci8.c, $ext_shared)
diff -Naur php-4.3.10/ext/odbc/config.m4 php-4.3.10.oden/ext/odbc/config.m4
--- php-4.3.10/ext/odbc/config.m4	2003-11-25 09:08:30.000000000 +0100
+++ php-4.3.10.oden/ext/odbc/config.m4	2004-12-16 10:41:29.101913056 +0100
@@ -101,8 +101,8 @@
   fi
   if test "$withval" != "no"; then
     PHP_ADD_INCLUDE($withval/incl)
-    PHP_ADD_LIBPATH($withval/lib)
-    ODBC_OBJS="$withval/lib/odbclib.a"
+    PHP_ADD_LIBPATH($withval/$PHP_LIBNAME)
+    ODBC_OBJS="$withval/$PHP_LIBNAME/odbclib.a"
     ODBC_LIB="$abs_builddir/ext/odbc/libodbc_adabas.a"
     $srcdir/build/shtool mkdir -f -p ext/odbc
     rm -f "$ODBC_LIB"
@@ -133,7 +133,7 @@
   fi
   if test "$withval" != "no"; then
     PHP_ADD_INCLUDE($withval/incl)
-    PHP_ADD_LIBPATH($withval/lib)
+    PHP_ADD_LIBPATH($withval/$PHP_LIBNAME)
     PHP_ADD_LIBRARY(sqlod)
     ODBC_TYPE=sapdb
     AC_DEFINE(HAVE_SAPDB,1,[ ])
@@ -159,7 +159,7 @@
   fi
   if test "$withval" != "no"; then
     ODBC_INCDIR=$withval/include
-    ODBC_LIBDIR=$withval/lib
+    ODBC_LIBDIR=$withval/$PHP_LIBNAME
     ODBC_INCLUDE=-I$ODBC_INCDIR
     ODBC_TYPE=solid
     if test -f $ODBC_LIBDIR/soc*35.a; then
@@ -192,7 +192,7 @@
       ODBC_LIBDIR=/home/db2inst1/sqllib/lib
     else
       ODBC_INCDIR=$withval/include
-      ODBC_LIBDIR=$withval/lib
+      ODBC_LIBDIR=$withval/$PHP_LIBNAME
     fi
     ODBC_INCLUDE=-I$ODBC_INCDIR
     ODBC_LFLAGS=-L$ODBC_LIBDIR
@@ -302,7 +302,7 @@
         ODBC_LIBDIR=/usr/local/birdstep/lib
     else
         ODBC_INCDIR=$withval/include
-        ODBC_LIBDIR=$withval/lib
+        ODBC_LIBDIR=$withval/$PHP_LIBNAME
     fi
     ODBC_INCLUDE=-I$ODBC_INCDIR
     ODBC_TYPE=birdstep
@@ -340,7 +340,7 @@
   fi
   if test "$withval" != "no"; then
     ODBC_INCDIR=$withval/include
-    ODBC_LIBDIR=$withval/lib
+    ODBC_LIBDIR=$withval/$PHP_LIBNAME
     ODBC_LFLAGS=-L$ODBC_LIBDIR
     ODBC_INCLUDE=-I$ODBC_INCDIR
     ODBC_LIBS=$CUSTOM_ODBC_LIBS
@@ -366,11 +366,11 @@
     withval=/usr/local
   fi
   if test "$withval" != "no"; then
-    PHP_ADD_LIBRARY_WITH_PATH(iodbc, $withval/lib)
+    PHP_ADD_LIBRARY_WITH_PATH(iodbc, $withval/$PHP_LIBNAME)
     PHP_ADD_INCLUDE($withval/include, 1)
     ODBC_TYPE=iodbc
     ODBC_INCLUDE=-I$withval/include
-    ODBC_LFLAGS=-L$withval/lib
+    ODBC_LFLAGS=-L$withval/$PHP_LIBNAME
     ODBC_LIBS=-liodbc
     AC_DEFINE(HAVE_IODBC,1,[ ])
     AC_DEFINE(HAVE_ODBC2,1,[ ])
@@ -396,7 +396,7 @@
   fi
   if test "$withval" != "no"; then
     ODBC_INCDIR=$withval/include
-    ODBC_LIBDIR=$withval/lib
+    ODBC_LIBDIR=$withval/$PHP_LIBNAME
     ODBC_LFLAGS=-L$ODBC_LIBDIR
     ODBC_INCLUDE=-I$ODBC_INCDIR
     ODBC_LIBS=-lesoobclient
@@ -423,7 +423,7 @@
   fi
   if test "$withval" != "no"; then
     ODBC_INCDIR=$withval/include
-    ODBC_LIBDIR=$withval/lib
+    ODBC_LIBDIR=$withval/$PHP_LIBNAME
     ODBC_LFLAGS=-L$ODBC_LIBDIR
     ODBC_INCLUDE=-I$ODBC_INCDIR
     ODBC_LIBS=-lodbc
@@ -472,7 +472,7 @@
 
   if test "$withval" != "no"; then
     ODBC_INCDIR=$withval/include
-    ODBC_LIBDIR=$withval/lib
+    ODBC_LIBDIR=$withval/$PHP_LIBNAME
     ODBC_INCLUDE=-I$ODBC_INCDIR
     ODBC_LFLAGS=-L$ODBC_LIBDIR
     ODBC_INCLUDE=-I$ODBC_INCDIR
diff -Naur php-4.3.10/ext/pcre/config.m4 php-4.3.10.oden/ext/pcre/config.m4
--- php-4.3.10/ext/pcre/config.m4	2003-12-16 23:14:55.000000000 +0100
+++ php-4.3.10.oden/ext/pcre/config.m4	2004-12-16 10:41:29.101913056 +0100
@@ -25,7 +25,7 @@
       AC_MSG_ERROR([Could not find pcre.h in $PHP_PCRE_REGEX])
     fi
 
-    for j in $PHP_PCRE_REGEX $PHP_PCRE_REGEX/lib; do
+    for j in $PHP_PCRE_REGEX $PHP_PCRE_REGEX/$PHP_LIBNAME; do
       test -f $j/libpcre.a -o -f $j/libpcre.$SHLIB_SUFFIX_NAME && PCRE_LIBDIR=$j
     done
     
diff -Naur php-4.3.10/ext/pdf/config.m4 php-4.3.10.oden/ext/pdf/config.m4
--- php-4.3.10/ext/pdf/config.m4	2004-11-29 10:13:39.000000000 +0100
+++ php-4.3.10.oden/ext/pdf/config.m4	2004-12-16 10:41:29.093914272 +0100
@@ -40,11 +40,11 @@
   if test "$PHP_JPEG_DIR" != "no"; then
     PHP_CHECK_LIBRARY(jpeg,jpeg_read_header, 
     [
-      PHP_ADD_LIBRARY_WITH_PATH(jpeg, $PHP_JPEG_DIR/lib, PDF_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(jpeg, $PHP_JPEG_DIR/$PHP_LIBNAME, PDF_SHARED_LIBADD)
     ],[
       AC_MSG_ERROR([libjpeg not found!])
     ],[
-      -L$PHP_JPEG_DIR/lib
+      -L$PHP_JPEG_DIR/$PHP_LIBNAME
     ])
   else
     AC_MSG_WARN([If configure fails, try --with-jpeg-dir=<DIR>])
@@ -54,11 +54,11 @@
   if test "$PHP_PNG_DIR" != "no"; then
     PHP_CHECK_LIBRARY(png,png_create_info_struct, 
     [
-      PHP_ADD_LIBRARY_WITH_PATH(png, $PHP_PNG_DIR/lib, PDF_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(png, $PHP_PNG_DIR/$PHP_LIBNAME, PDF_SHARED_LIBADD)
     ],[
       AC_MSG_ERROR([libpng not found!])
     ],[
-      -L$PHP_PNG_DIR/lib
+      -L$PHP_PNG_DIR/$PHP_LIBNAME
     ])
   else
     AC_MSG_WARN([If configure fails, try --with-png-dir=<DIR>])
@@ -68,11 +68,11 @@
   if test "$PHP_TIFF_DIR" != "no"; then
     PHP_CHECK_LIBRARY(tiff,TIFFOpen, 
     [
-      PHP_ADD_LIBRARY_WITH_PATH(tiff, $PHP_TIFF_DIR/lib, PDF_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(tiff, $PHP_TIFF_DIR/$PHP_LIBNAME, PDF_SHARED_LIBADD)
     ],[
       AC_MSG_ERROR([libtiff not found!])
     ],[
-      -L$PHP_TIFF_DIR/lib
+      -L$PHP_TIFF_DIR/$PHP_LIBNAME
     ])
   else
     AC_MSG_WARN([If configure fails, try --with-tiff-dir=<DIR>])
@@ -84,7 +84,7 @@
     AC_MSG_RESULT([no. If configure fails, try --with-zlib-dir=<DIR>])
   else           
     AC_MSG_RESULT([$PHP_ZLIB_DIR])
-    PHP_ADD_LIBRARY_WITH_PATH(z, $PHP_ZLIB_DIR/lib, PDF_SHARED_LIBADD)
+    PHP_ADD_LIBRARY_WITH_PATH(z, $PHP_ZLIB_DIR/$PHP_LIBNAME, PDF_SHARED_LIBADD)
   fi
 
   dnl #
@@ -118,7 +118,7 @@
         PHP_CHECK_LIBRARY(pdf, PDF_show_boxed, 
         [
           AC_DEFINE(HAVE_PDFLIB,1,[ ]) 
-          PHP_ADD_LIBRARY_WITH_PATH(pdf, $PHP_PDFLIB/lib, PDF_SHARED_LIBADD)
+          PHP_ADD_LIBRARY_WITH_PATH(pdf, $PHP_PDFLIB/$PHP_LIBNAME, PDF_SHARED_LIBADD)
           PHP_ADD_INCLUDE($PHP_PDFLIB/include)
         ],[
           AC_MSG_ERROR([
@@ -127,7 +127,7 @@
 See config.log for more information.
 ])
         ],[
-          -L$PHP_PDFLIB/lib $PDF_SHARED_LIBADD
+          -L$PHP_PDFLIB/$PHP_LIBNAME $PDF_SHARED_LIBADD
         ])
       else
         AC_MSG_ERROR([pdflib.h not found! Check the path passed to --with-pdflib=<PATH>. PATH should be the install prefix directory.])
diff -Naur php-4.3.10/ext/pfpro/config.m4 php-4.3.10.oden/ext/pfpro/config.m4
--- php-4.3.10/ext/pfpro/config.m4	2003-10-01 04:54:03.000000000 +0200
+++ php-4.3.10.oden/ext/pfpro/config.m4	2004-12-16 10:41:29.103912752 +0100
@@ -14,16 +14,16 @@
       PFPRO_INC_DIR=$i
     elif test -r $i/include/$PFPRO_HDR; then
       PFPRO_INC_DIR=$i/include
-    elif test -r $i/lib/$PFPRO_HDR; then
-      PFPRO_INC_DIR=$i/lib
+    elif test -r $i/$PHP_LIBNAME/$PFPRO_HDR; then
+      PFPRO_INC_DIR=$i/$PHP_LIBNAME
     elif test -r $i/bin/$PFPRO_HDR; then
       PFPRO_INC_DIR=$i/bin
     fi
 
     if test -r $i/$PFPRO_LIB; then
       PFPRO_LIB_DIR=$i
-    elif test -r $i/lib/$PFPRO_LIB; then
-      PFPRO_LIB_DIR=$i/lib
+    elif test -r $i/$PHP_LIBNAME/$PFPRO_LIB; then
+      PFPRO_LIB_DIR=$i/$PHP_LIBNAME
     fi
 
 	test -n "$PFPRO_INC_DIR" && test -n "$PFPRO_LIB_DIR" && break
diff -Naur php-4.3.10/ext/pgsql/config.m4 php-4.3.10.oden/ext/pgsql/config.m4
--- php-4.3.10/ext/pgsql/config.m4	2002-06-26 15:07:40.000000000 +0200
+++ php-4.3.10.oden/ext/pgsql/config.m4	2004-12-16 10:41:29.103912752 +0100
@@ -29,7 +29,7 @@
       fi
     done
 
-    for j in lib lib/pgsql lib/postgres lib/postgresql ""; do
+    for j in $PHP_LIBNAME lib/pgsql lib/postgres lib/postgresql ""; do
       if test -f "$i/$j/libpq.so" || test -f "$i/$j/libpq.a"; then 
         PGSQL_LIBDIR=$i/$j
       fi
diff -Naur php-4.3.10/ext/pspell/config.m4 php-4.3.10.oden/ext/pspell/config.m4
--- php-4.3.10/ext/pspell/config.m4	2003-08-09 20:43:58.000000000 +0200
+++ php-4.3.10.oden/ext/pspell/config.m4	2004-12-16 10:41:29.105912448 +0100
@@ -27,7 +27,7 @@
 		AC_MSG_ERROR(Cannot find pspell)
 	fi
 
-	PSPELL_LIBDIR=$PSPELL_DIR/lib
+	PSPELL_LIBDIR=$PSPELL_DIR/$PHP_LIBNAME
 
 	AC_DEFINE(HAVE_PSPELL,1,[ ])
 	PHP_SUBST(PSPELL_SHARED_LIBADD)
diff -Naur php-4.3.10/ext/qtdom/config.m4 php-4.3.10.oden/ext/qtdom/config.m4
--- php-4.3.10/ext/qtdom/config.m4	2002-11-18 06:06:37.000000000 +0100
+++ php-4.3.10.oden/ext/qtdom/config.m4	2004-12-16 10:41:29.103912752 +0100
@@ -9,16 +9,16 @@
   case $PHP_QTDOM in
     yes)
       if test -f $QTDIR/include/qdom.h; then
-        QTDOM_LIBDIR=$QTDIR/lib
+        QTDOM_LIBDIR=$QTDIR/$PHP_LIBNAME
         QTDOM_INCDIR=$QTDIR/include
-      elif test -f /usr/lib/qt2/include/qdom.h; then
-        QTDOM_LIBDIR=/usr/lib
-        QTDOM_INCDIR=/usr/lib/qt2/include
+      elif test -f /usr/$PHP_LIBNAME/qt2/include/qdom.h; then
+        QTDOM_LIBDIR=/usr/$PHP_LIBNAME
+        QTDOM_INCDIR=/usr/$PHP_LIBNAME/qt2/include
       fi
       ;;
     *)
       if test -f $PHP_QTDOM/include/qdom.h; then
-        QTDOM_LIBDIR=$PHP_QTDOM/lib
+        QTDOM_LIBDIR=$PHP_QTDOM/$PHP_LIBNAME
         QTDOM_INCDIR=$PHP_QTDOM/include
       fi
       ;;
diff -Naur php-4.3.10/ext/readline/config.m4 php-4.3.10.oden/ext/readline/config.m4
--- php-4.3.10/ext/readline/config.m4	2003-10-01 04:54:04.000000000 +0200
+++ php-4.3.10.oden/ext/readline/config.m4	2004-12-16 10:41:29.106912296 +0100
@@ -31,20 +31,20 @@
 
   PHP_CHECK_LIBRARY(readline, readline,
   [
-    PHP_ADD_LIBRARY_WITH_PATH(readline, $READLINE_DIR/lib, READLINE_SHARED_LIBADD)
+    PHP_ADD_LIBRARY_WITH_PATH(readline, $READLINE_DIR/$PHP_LIBNAME, READLINE_SHARED_LIBADD)
   ], [
     AC_MSG_ERROR(readline library not found)
   ], [
-    -L$READLINE_DIR/lib 
+    -L$READLINE_DIR/$PHP_LIBNAME
   ])
 
   PHP_CHECK_LIBRARY(history, add_history,
   [
-    PHP_ADD_LIBRARY_WITH_PATH(history, $READLINE_DIR/lib, READLINE_SHARED_LIBADD)
+    PHP_ADD_LIBRARY_WITH_PATH(history, $READLINE_DIR/$PHP_LIBNAME, READLINE_SHARED_LIBADD)
   ], [
     AC_MSG_ERROR(history library required by readline not found)
   ], [
-    -L$READLINE_DIR/lib 
+    -L$READLINE_DIR/$PHP_LIBNAME
   ])
 
   PHP_NEW_EXTENSION(readline, readline.c, $ext_shared, cli)
@@ -75,11 +75,11 @@
 
   PHP_CHECK_LIBRARY(edit, readline,
   [
-    PHP_ADD_LIBRARY_WITH_PATH(edit, $LIBEDIT_DIR/lib, READLINE_SHARED_LIBADD)  
+    PHP_ADD_LIBRARY_WITH_PATH(edit, $LIBEDIT_DIR/$PHP_LIBNAME, READLINE_SHARED_LIBADD)  
   ], [
     AC_MSG_ERROR(edit library required by readline not found)
   ], [
-    -L$READLINE_DIR/lib 
+    -L$READLINE_DIR/$PHP_LIBNAME
   ])
 
   PHP_NEW_EXTENSION(readline, readline.c, $ext_shared, cli)
diff -Naur php-4.3.10/ext/session/config.m4 php-4.3.10.oden/ext/session/config.m4
--- php-4.3.10/ext/session/config.m4	2003-10-01 04:54:06.000000000 +0200
+++ php-4.3.10.oden/ext/session/config.m4	2004-12-16 10:41:29.103912752 +0100
@@ -25,7 +25,7 @@
     AC_MSG_ERROR(cannot find mm library)
   fi
   
-  PHP_ADD_LIBRARY_WITH_PATH(mm, $MM_DIR/lib, SESSION_SHARED_LIBADD)
+  PHP_ADD_LIBRARY_WITH_PATH(mm, $MM_DIR/$PHP_LIBNAME, SESSION_SHARED_LIBADD)
   PHP_ADD_INCLUDE($MM_DIR/include)
   AC_DEFINE(HAVE_LIBMM, 1, [Whether you have libmm])
 fi
diff -Naur php-4.3.10/ext/snmp/config.m4 php-4.3.10.oden/ext/snmp/config.m4
--- php-4.3.10/ext/snmp/config.m4	2003-09-23 10:16:47.000000000 +0200
+++ php-4.3.10.oden/ext/snmp/config.m4	2004-12-16 10:41:29.101913056 +0100
@@ -46,12 +46,12 @@
         test -f $i/snmp/include/ucd-snmp/snmp.h && SNMP_INCDIR=$i/snmp/include/ucd-snmp
       done
       for i in /usr /usr/snmp /usr/local /usr/local/snmp; do
-        test -f $i/lib/libsnmp.a -o -f $i/lib/libsnmp.$SHLIB_SUFFIX_NAME && SNMP_LIBDIR=$i/lib
+        test -f $i/$PHP_LIBNAME/libsnmp.a -o -f $i/$PHP_LIBNAME/libsnmp.$SHLIB_SUFFIX_NAME && SNMP_LIBDIR=$i/$PHP_LIBNAME
       done
     else
       SNMP_INCDIR=$PHP_SNMP/include
       test -d $PHP_SNMP/include/ucd-snmp && SNMP_INCDIR=$PHP_SNMP/include/ucd-snmp
-      SNMP_LIBDIR=$PHP_SNMP/lib
+      SNMP_LIBDIR=$PHP_SNMP/$PHP_LIBNAME
     fi
 
     if test -z "$SNMP_INCDIR"; then
diff -Naur php-4.3.10/ext/swf/config.m4 php-4.3.10.oden/ext/swf/config.m4
--- php-4.3.10/ext/swf/config.m4	2002-03-12 17:35:59.000000000 +0100
+++ php-4.3.10.oden/ext/swf/config.m4	2004-12-16 10:41:29.093914272 +0100
@@ -6,12 +6,12 @@
 [  --with-swf[=DIR]        Include swf support])
 
 if test "$PHP_SWF" != "no"; then
-  if test -r $PHP_SWF/lib/libswf.a; then
+  if test -r $PHP_SWF/$PHP_LIBNAME/libswf.a; then
     SWF_DIR=$PHP_SWF
   else
     AC_MSG_CHECKING(for libswf in default path)
     for i in /usr/local /usr; do
-      if test -r $i/lib/libswf.a; then
+      if test -r $i/$PHP_LIBNAME/libswf.a; then
         SWF_DIR=$i
         AC_MSG_RESULT(found in $i)
       fi
@@ -28,7 +28,7 @@
   PHP_ADD_INCLUDE($SWF_DIR/include)
 
   PHP_SUBST(SWF_SHARED_LIBADD)
-  PHP_ADD_LIBRARY_WITH_PATH(swf, $SWF_DIR/lib, SWF_SHARED_LIBADD)
+  PHP_ADD_LIBRARY_WITH_PATH(swf, $SWF_DIR/$PHP_LIBNAME, SWF_SHARED_LIBADD)
   AC_DEFINE(HAVE_SWF,1,[ ])
 
   PHP_NEW_EXTENSION(swf, swf.c, $ext_shared)
diff -Naur php-4.3.10/ext/xml/config.m4 php-4.3.10.oden/ext/xml/config.m4
--- php-4.3.10/ext/xml/config.m4	2003-10-03 07:25:42.000000000 +0200
+++ php-4.3.10.oden/ext/xml/config.m4	2004-12-16 10:41:29.093914272 +0100
@@ -31,7 +31,7 @@
     PHP_NEW_EXTENSION(xml, xml.c, $ext_shared)
 
     for i in $PHP_XML $PHP_EXPAT_DIR; do
-      if test -f $i/lib/libexpat.a -o -f $i/lib/libexpat.$SHLIB_SUFFIX_NAME ; then
+      if test -f $i/$PHP_LIBNAME/libexpat.a -o -f $i/$PHP_LIBNAME/libexpat.$SHLIB_SUFFIX_NAME ; then
         EXPAT_DIR=$i
       fi
     done
@@ -41,7 +41,7 @@
     fi
 
     PHP_ADD_INCLUDE($EXPAT_DIR/include)
-    PHP_ADD_LIBRARY_WITH_PATH(expat, $EXPAT_DIR/lib, XML_SHARED_LIBADD)
+    PHP_ADD_LIBRARY_WITH_PATH(expat, $EXPAT_DIR/$PHP_LIBNAME, XML_SHARED_LIBADD)
     PHP_SUBST(XML_SHARED_LIBADD)
   fi
 fi
diff -Naur php-4.3.10/ext/xmlrpc/config.m4 php-4.3.10.oden/ext/xmlrpc/config.m4
--- php-4.3.10/ext/xmlrpc/config.m4	2003-10-03 07:25:43.000000000 +0200
+++ php-4.3.10.oden/ext/xmlrpc/config.m4	2004-12-16 10:41:29.105912448 +0100
@@ -23,9 +23,9 @@
 
   testval=no
   for i in $PHP_EXPAT_DIR $XMLRPC_DIR /usr/local /usr; do
-    if test -f $i/lib/libexpat.a -o -f $i/lib/libexpat.$SHLIB_SUFFIX_NAME; then
+    if test -f $i/$PHP_LIBNAME/libexpat.a -o -f $i/$PHP_LIBNAME/libexpat.$SHLIB_SUFFIX_NAME; then
       AC_DEFINE(HAVE_LIBEXPAT2,1,[ ])
-      PHP_ADD_LIBRARY_WITH_PATH(expat, $i/lib, XMLRPC_SHARED_LIBADD)
+      PHP_ADD_LIBRARY_WITH_PATH(expat, $i/$PHP_LIBNAME, XMLRPC_SHARED_LIBADD)
       PHP_ADD_INCLUDE($i/include)
       testval=yes
       break
@@ -88,5 +88,5 @@
   fi
 
   PHP_ADD_INCLUDE($XMLRPC_DIR)
-  PHP_ADD_LIBRARY_WITH_PATH(xmlrpc, $XMLRPC_DIR/lib, XMLRPC_SHARED_LIBADD)
+  PHP_ADD_LIBRARY_WITH_PATH(xmlrpc, $XMLRPC_DIR/$PHP_LIBNAME, XMLRPC_SHARED_LIBADD)
 fi
diff -Naur php-4.3.10/ext/xslt/config.m4 php-4.3.10.oden/ext/xslt/config.m4
--- php-4.3.10/ext/xslt/config.m4	2004-08-18 07:14:16.000000000 +0200
+++ php-4.3.10.oden/ext/xslt/config.m4	2004-12-16 10:41:29.101913056 +0100
@@ -98,10 +98,10 @@
 
     found_expat=no
     for i in $PHP_EXPAT_DIR $XSLT_DIR /usr/local /usr; do
-      if test -f $i/lib/libexpat.a -o -f $i/lib/libexpat.$SHLIB_SUFFIX_NAME; then
+      if test -f $i/$PHP_LIBNAME/libexpat.a -o -f $i/$PHP_LIBNAME/libexpat.$SHLIB_SUFFIX_NAME; then
         AC_DEFINE(HAVE_LIBEXPAT2, 1, [ ])
         PHP_ADD_INCLUDE($i/include)
-        PHP_ADD_LIBRARY_WITH_PATH(expat, $i/lib, XSLT_SHARED_LIBADD)
+        PHP_ADD_LIBRARY_WITH_PATH(expat, $i/$PHP_LIBNAME, XSLT_SHARED_LIBADD)
         found_expat=yes
         break
       fi
@@ -125,7 +125,7 @@
      
     if test "$PHP_SABLOT_JS" != "no"; then
       for i in $PHP_SABLOT_JS /usr/local /usr; do
-        if test -f $i/lib/libjs.a -o -f $i/lib/libjs.$SHLIB_SUFFIX_NAME; then
+        if test -f $i/$PHP_LIBNAME/libjs.a -o -f $i/$PHP_LIBNAME/libjs.$SHLIB_SUFFIX_NAME; then
           PHP_SABLOT_JS_DIR=$i
           break
         fi
@@ -133,12 +133,12 @@
 
       PHP_CHECK_LIBRARY(js, JS_GetRuntime,
       [
-        PHP_ADD_LIBRARY_WITH_PATH(js, $PHP_SABLOT_JS_DIR/lib, XSLT_SHARED_LIBADD)
-        PHP_SABLOT_JS_LIBS="-L$PHP_SABLOT_JS_DIR/lib -ljs"
+        PHP_ADD_LIBRARY_WITH_PATH(js, $PHP_SABLOT_JS_DIR/$PHP_LIBNAME, XSLT_SHARED_LIBADD)
+        PHP_SABLOT_JS_LIBS="-L$PHP_SABLOT_JS_DIR/$PHP_LIBNAME -ljs"
       ], [
         AC_MSG_ERROR([libjs not found. Please check config.log for more information.])
       ], [
-        -L$PHP_SABLOT_JS_DIR/lib
+        -L$PHP_SABLOT_JS_DIR/$PHP_LIBNAME
       ])
     fi
 
@@ -146,7 +146,7 @@
     [
       AC_DEFINE(HAVE_SABLOT_SET_ENCODING, 1, [ ])
     ], [], [
-      -L$XSLT_DIR/lib $PHP_SABLOT_JS_LIBS
+      -L$XSLT_DIR/$PHP_LIBNAME $PHP_SABLOT_JS_LIBS
     ])
 
     dnl SablotSetOptions implemented in Sablotron CVS > 2002/10/31
@@ -154,14 +154,14 @@
     [
       AC_DEFINE(HAVE_SABLOT_GET_OPTIONS, 1, [Whether Sablotron supports SablotGetOptions])
     ], [], [
-      -L$XSLT_DIR/lib $PHP_SABLOT_JS_LIBS
+      -L$XSLT_DIR/$PHP_LIBNAME $PHP_SABLOT_JS_LIBS
     ])
 
     AC_DEFINE(HAVE_SABLOT_BACKEND, 1, [ ])
   fi
 
   PHP_ADD_INCLUDE($XSLT_DIR/include)
-  PHP_ADD_LIBRARY_WITH_PATH($XSLT_LIBNAME, $XSLT_DIR/lib, XSLT_SHARED_LIBADD)
+  PHP_ADD_LIBRARY_WITH_PATH($XSLT_LIBNAME, $XSLT_DIR/$PHP_LIBNAME, XSLT_SHARED_LIBADD)
 
   AC_DEFINE(HAVE_XSLT, 1, [ ])
 fi
diff -Naur php-4.3.10/ext/zip/config.m4 php-4.3.10.oden/ext/zip/config.m4
--- php-4.3.10/ext/zip/config.m4	2002-12-09 19:28:03.000000000 +0100
+++ php-4.3.10.oden/ext/zip/config.m4	2004-12-16 10:41:29.093914272 +0100
@@ -20,7 +20,7 @@
     AC_MSG_ERROR(Cannot find libzzip)
   fi
 
-  ZZIPLIB_LIBDIR=$ZZIPLIB_DIR/lib
+  ZZIPLIB_LIBDIR=$ZZIPLIB_DIR/$PHP_LIBNAME
 
   PHP_CHECK_LIBRARY(zzip, zzip_open,
   [
diff -Naur php-4.3.10/ext/zlib/config0.m4 php-4.3.10.oden/ext/zlib/config0.m4
--- php-4.3.10/ext/zlib/config0.m4	2003-07-02 01:54:44.000000000 +0200
+++ php-4.3.10.oden/ext/zlib/config0.m4	2004-12-16 10:41:29.101913056 +0100
@@ -41,10 +41,10 @@
   ],[
     AC_MSG_ERROR(ZLIB extension requires zlib >= 1.0.9)
   ],[
-    -L$ZLIB_DIR/lib
+    -L$ZLIB_DIR/$PHP_LIBNAME
   ])
 
-  PHP_ADD_LIBPATH($ZLIB_DIR/lib, ZLIB_SHARED_LIBADD)
+  PHP_ADD_LIBPATH($ZLIB_DIR/$PHP_LIBNAME, ZLIB_SHARED_LIBADD)
 
   PHP_ZLIB_DIR=$ZLIB_DIR
   PHP_ADD_LIBRARY(z,, ZLIB_SHARED_LIBADD)
diff -Naur php-4.3.10/scripts/Makefile.frag php-4.3.10.oden/scripts/Makefile.frag
--- php-4.3.10/scripts/Makefile.frag	2003-12-01 17:12:42.000000000 +0100
+++ php-4.3.10.oden/scripts/Makefile.frag	2004-12-16 10:13:03.302234024 +0100
@@ -4,7 +4,7 @@
 #
 
 phpincludedir = $(includedir)/php
-phpbuilddir = $(prefix)/lib/php/build
+phpbuilddir = $(libdir)/php/build
 
 BUILD_FILES = \
 	scripts/phpize.m4 \
diff -Naur php-4.3.10/scripts/phpize.in php-4.3.10.oden/scripts/phpize.in
--- php-4.3.10/scripts/phpize.in	2004-06-19 22:37:55.000000000 +0200
+++ php-4.3.10.oden/scripts/phpize.in	2004-12-16 10:13:03.302234024 +0100
@@ -1,7 +1,7 @@
 #! /bin/sh
 
 prefix='@prefix@'
-phpdir="$prefix/lib/php/build"
+phpdir="@libdir@/php/build"
 includedir="$prefix/include/php"
 builddir="`pwd`"
 
diff -Naur php-4.3.10/scripts/phpize.m4 php-4.3.10.oden/scripts/phpize.m4
--- php-4.3.10/scripts/phpize.m4	2003-09-03 10:07:51.000000000 +0200
+++ php-4.3.10.oden/scripts/phpize.m4	2004-12-16 10:13:03.302234024 +0100
@@ -47,6 +47,7 @@
 AC_PROG_CC_C_O
 
 PHP_RUNPATH_SWITCH
+PHP_CHECK_LIBNAME
 PHP_SHLIB_SUFFIX_NAME
 PHP_WITH_PHP_CONFIG
 
