diff -upN cpio-2.5.orig/main.c cpio-2.5/main.c
--- cpio-2.5.orig/main.c	Thu Aug  8 12:24:05 2002
+++ cpio-2.5/main.c	Thu Aug  8 12:25:45 2002
@@ -21,6 +21,7 @@
 
 #include <stdio.h>
 #include <getopt.h>
+#include <locale.h>
 #include <sys/types.h>
 #include <sys/stat.h>
 #include "filetypes.h"
@@ -511,6 +512,8 @@ main (argc, argv)
      int argc;
      char *argv[];
 {
+  setlocale (LC_ALL, "");
+
   program_name = argv[0];
   umask (0);
 
