--- linux-atm-2.4.1/src/lib/sapequal.c.orig	2001-09-03 20:41:05.000000000 +0200
+++ linux-atm-2.4.1/src/lib/sapequal.c	2004-06-08 05:16:59.376468991 +0200
@@ -65,6 +65,7 @@ static int match_blli(const struct atm_b
 	    CHECK(l2.itu.window,a.l2.itu.window > b.l2.itu.window);
 	    break;
 	default:
+	    break;
     }
     switch (a.l3_proto) {
 	case ATM_L3_X25:
@@ -83,6 +84,7 @@ static int match_blli(const struct atm_b
 	    if (a.l3.user != b.l3.user) return 0;
 	    break;
 	default:
+	    break;
     }
     return 1;
 }
--- linux-atm-2.4.1/src/sigd/proto.c.orig	2001-09-03 20:41:06.000000000 +0200
+++ linux-atm-2.4.1/src/sigd/proto.c	2004-06-08 05:17:49.020500967 +0200
@@ -259,6 +259,7 @@ void send_release(SOCKET *sock,unsigned 
 		break;
 	    }
 	default:
+	    break;
     }
     va_end(ap);
     if ((size = q_close(&dsc)) >= 0) to_signaling(sock->sig,q_buffer,size);
@@ -288,6 +289,7 @@ void send_release_complete(SIG_ENTITY *s
 		}
 		break;
 	    default:
+		break;
 	}
 	va_end(ap);
     }
--- linux-atm-2.4.1/src/maint/enitune.c.orig	2004-06-08 05:18:30.325527510 +0200
+++ linux-atm-2.4.1/src/maint/enitune.c	2004-06-08 05:18:43.578536015 +0200
@@ -44,6 +44,7 @@ int main(int argc,char **argv)
 		if (*end || mult.rx <= 100) usage(name);
 		break;
 	    default:
+		break;
 	}
     if (argc != optind+1) usage(name);
     sioc.number = strtol(argv[optind],&end,0);
