--- tiff-v3.6.1/libtiff/Makefile.in.codecs	2004-05-27 11:49:53.207449056 +0200
+++ tiff-v3.6.1/libtiff/Makefile.in	2004-05-27 11:49:53.218447384 +0200
@@ -160,7 +160,7 @@
 	${RANLIB} libtiff.a
 
 $(SHLIB): $(SHOBJS)
-	gcc -o $(SHLIB) -Wl,-soname,$(SONAME) -shared $(SHOBJS)
+	gcc -o $(SHLIB) -Wl,-soname,$(SONAME) -shared $(SHOBJS) -ljpeg -lz
 
 #
 # NB: The configure script verifies that the configured
--- tiff-v3.6.1/config.site.codecs	2002-10-12 14:52:46.000000000 +0200
+++ tiff-v3.6.1/config.site	2004-05-27 11:51:39.443298752 +0200
@@ -41,8 +41,8 @@
 # Package controls.
 #
 #DSO="auto"			# auto|IRIX|IRIX52 enable DSO support for system
-#JPEG="yes"			# yes|no configure JPEG support (see below)
-#ZIP="yes"			# yes|no configure Deflate support (see below)
+JPEG="yes"			# yes|no configure JPEG support (see below)
+ZIP="yes"			# yes|no configure Deflate support (see below)
 #LIBGL="auto"			# yes|no|auto configure IRIS GL-based tools
 #LIBIMAGE="auto"		# yes|no|auto configure SGI RGB image tools
 #HTML="yes"			# yes|no install HTML documentation
@@ -97,7 +97,7 @@
 # which you can find on the Internet at ftp.uu.net:/graphics/jpeg/.
 #
 #DIRS_LIBINC="$DIRS_LIBINC ../src/jpeg-6b" # directory for jpeg include files
-#DIR_JPEGLIB=../src/jpeg-6b		# directory for libjpeg
+DIR_JPEGLIB=/usr/lib		# directory for libjpeg
 
 # The following will enable OJPEG support.  Note that a patch libjpeg is
 # required before this can be enabled.  It is also only supported if JPEG is
@@ -118,7 +118,7 @@
 #     Look for it also at ftp://ftp.uu.net/pub/archiving/zip/zlib.
 #
 #DIRS_LIBINC="$DIRS_LIBINC ../zlib"	# directory for zlib include files
-#DIR_GZLIB="../zlib"			# directory for libz
+DIR_GZLIB="/usr/lib"			# directory for libz
 
 #
 # Miscellaneous parameters.
