diff -Naur net-snmp-5.2.2/perl/agent/agent.xs net-snmp-5.2.2.oden/perl/agent/agent.xs
--- net-snmp-5.2.2/perl/agent/agent.xs	2005-05-06 01:13:49.000000000 +0200
+++ net-snmp-5.2.2.oden/perl/agent/agent.xs	2005-12-20 22:16:54.000000000 +0100
@@ -394,19 +394,19 @@
         PUSHMARK(sp);
         rarg = newSViv(0);
         arg = newSVrv(rarg, "NetSNMP::agent::netsnmp_mib_handler");
-        sv_setiv(arg, (int) handler);
+        sv_setiv(arg, (IV) handler);
         XPUSHs(sv_2mortal(rarg));
         rarg = newSViv(0);
         arg = newSVrv(rarg, "NetSNMP::agent::reginfo");
-        sv_setiv(arg, (int) reginfo);
+        sv_setiv(arg, (IV) reginfo);
         XPUSHs(sv_2mortal(rarg));
         rarg = newSViv(0);
         arg = newSVrv(rarg, "NetSNMP::agent::netsnmp_agent_request_info");
-        sv_setiv(arg, (int) reqinfo);
+        sv_setiv(arg, (IV) reqinfo);
         XPUSHs(sv_2mortal(rarg));
         rarg = newSViv(0);
         arg = newSVrv(rarg, "NetSNMP::agent::netsnmp_request_infoPtr");
-        sv_setiv(arg, (int) requests);
+        sv_setiv(arg, (IV) requests);
         XPUSHs(sv_2mortal(rarg));
         PUTBACK;
         if (SvTYPE(cb) == SVt_PVCV) {
@@ -567,7 +567,7 @@
 
         rarg = newSViv((int) 0);
         arg = newSVrv(rarg, "netsnmp_oidPtr");
-        sv_setiv(arg, (int) o);
+        sv_setiv(arg, (IV) o);
 
         XPUSHs(rarg);
 
@@ -606,7 +606,7 @@
 
         rarg = newSViv((int) 0);
         arg = newSVrv(rarg, "netsnmp_oidPtr");
-        sv_setiv(arg, (int) o);
+        sv_setiv(arg, (IV) o);
 
         XPUSHs(rarg);
 
@@ -996,7 +996,7 @@
                 request = request->next;
                 rarg = newSViv(0);
                 arg = newSVrv(rarg, "NetSNMP::agent::netsnmp_request_infoPtr");
-                sv_setiv(arg, (int) request);
+                sv_setiv(arg, (IV) request);
                 ST(0) = rarg;
             } else {
                 ST(0) = &sv_undef;
diff -Naur net-snmp-5.2.2/perl/OID/OID.xs net-snmp-5.2.2.oden/perl/OID/OID.xs
--- net-snmp-5.2.2/perl/OID/OID.xs	2005-09-09 12:09:25.000000000 +0200
+++ net-snmp-5.2.2.oden/perl/OID/OID.xs	2005-12-20 22:13:52.000000000 +0100
@@ -178,16 +178,19 @@
 nso_newptr(initstring)
     char *initstring
     CODE:
+        size_t len;
         if (get_tree_head() == NULL)
             init_mib();
         RETVAL = SNMP_MALLOC_TYPEDEF(netsnmp_oid);
         RETVAL->name = RETVAL->namebuf;
         RETVAL->len = sizeof(RETVAL->namebuf)/sizeof(RETVAL->namebuf[0]);
-        if (!snmp_parse_oid(initstring, (oid *) RETVAL->name, &RETVAL->len)) {
+        if (!snmp_parse_oid(initstring, (oid *) RETVAL->name, &len)) {
             snmp_log(LOG_ERR, "Can't parse: %s\n", initstring);
             RETVAL->len = 0;
             RETVAL = NULL;
         }
+        else
+            RETVAL->len = len;
     OUTPUT:
         RETVAL
 
diff -Naur net-snmp-5.2.2/perl/SNMP/SNMP.xs net-snmp-5.2.2.oden/perl/SNMP/SNMP.xs
--- net-snmp-5.2.2/perl/SNMP/SNMP.xs	2005-10-05 14:26:19.000000000 +0200
+++ net-snmp-5.2.2.oden/perl/SNMP/SNMP.xs	2005-12-20 22:14:28.000000000 +0100
@@ -117,15 +117,15 @@
                               netsnmp_variable_list*, struct tree *,
                              int, int));
 static int __sprint_num_objid _((char *, oid *, int));
-static int __scan_num_objid _((char *, oid *, int *));
+static int __scan_num_objid _((char *, oid *, size_t *));
 static int __get_type_str _((int, char *));
 static int __get_label_iid _((char *, char **, char **, int));
 static int __oid_cmp _((oid *, int, oid *, int));
 static int __tp_sprint_num_objid _((char*,SnmpMibNode *));
 static SnmpMibNode * __get_next_mib_node _((SnmpMibNode *));
 static struct tree * __oid2tp _((oid*, int, struct tree *, int*));
-static struct tree * __tag2oid _((char *, char *, oid  *, int  *, int *, int));
-static int __concat_oid_str _((oid *, int *, char *));
+static struct tree * __tag2oid _((char *, char *, oid  *, size_t  *, int *, int));
+static int __concat_oid_str _((oid *, size_t *, char *));
 static int __add_var_val_str _((netsnmp_pdu *, oid *, int, char *,
                                  int, int));
 static int __send_sync_pdu _((netsnmp_session *, netsnmp_pdu *,
@@ -565,7 +565,7 @@
 __scan_num_objid (buf, objid, len)
 char *buf;
 oid *objid;
-int *len;
+size_t *len;
 {
    char *cp;
    *len = 0;
@@ -856,14 +856,14 @@
 char * tag;
 char * iid;
 oid  * oid_arr;
-int  * oid_arr_len;
+size_t  * oid_arr_len;
 int  * type;
 int    best_guess;
 {
    struct tree *tp = NULL;
    struct tree *rtp = NULL;
    oid newname[MAX_OID_LEN], *op;
-   int newname_len = 0;
+   size_t newname_len = 0;
    const char *cp = NULL;
    char *module = NULL;
 
@@ -1004,7 +1004,7 @@
 static int
 __concat_oid_str(doid_arr, doid_arr_len, soid_str)
 oid *doid_arr;
-int *doid_arr_len;
+size_t *doid_arr_len;
 char * soid_str;
 {
    char soid_buf[STR_BUF_SIZE];
@@ -2992,7 +2992,7 @@
            netsnmp_pdu *pdu, *response;
            struct tree *tp;
 	   oid *oid_arr;
-	   int oid_arr_len = MAX_OID_LEN;
+	   size_t oid_arr_len = MAX_OID_LEN;
            char *tag_pv;
            snmp_xs_cb_data *xs_cb_data;
            SV **sess_ptr_sv;
@@ -3188,7 +3188,7 @@
            struct tree *tp;
            int len;
 	   oid *oid_arr;
-	   int oid_arr_len = MAX_OID_LEN;
+	   size_t oid_arr_len = MAX_OID_LEN;
            SV *tmp_sv;
            int type;
 	   char tmp_type_str[MAX_TYPE_NAME_LEN];
@@ -3403,7 +3403,7 @@
            struct tree *tp;
            int len;
 	   oid *oid_arr;
-	   int oid_arr_len = MAX_OID_LEN;
+	   size_t oid_arr_len = MAX_OID_LEN;
            SV *tmp_sv;
            int type;
 	   char tmp_type_str[MAX_TYPE_NAME_LEN];
@@ -3644,7 +3644,7 @@
            struct tree *tp;
            int len;
 	   oid *oid_arr;
-	   int oid_arr_len = MAX_OID_LEN;
+	   size_t oid_arr_len = MAX_OID_LEN;
            SV *tmp_sv;
            int type;
 	   char tmp_type_str[MAX_TYPE_NAME_LEN];
@@ -3868,7 +3868,7 @@
            netsnmp_session *ss;
            netsnmp_pdu *pdu = NULL;
 	   oid oid_arr[MAX_OID_LEN];
-	   int oid_arr_len;
+	   size_t oid_arr_len;
            SV **sess_ptr_sv;
            SV **err_str_svp;
            SV **err_num_svp;
@@ -4086,7 +4086,7 @@
 	      /* Sent okay...  Return the request ID in 'pdu' as an SvIV. */
 	      DBPRT(1,(DBOUT "Okay, request id is %d\n", (int)pdu));
 /*	      XSRETURN_IV((int)pdu); */
-	      XPUSHs(sv_2mortal(newSViv((int)pdu)));
+	      XPUSHs(sv_2mortal(newSViv((IV)pdu)));
 	      XSRETURN(1);
 	   }
 
@@ -4179,7 +4179,7 @@
            netsnmp_pdu *pdu = NULL;
            struct tree *tp;
 	   oid *oid_arr;
-	   int oid_arr_len = MAX_OID_LEN;
+	   size_t oid_arr_len = MAX_OID_LEN;
            SV **sess_ptr_sv;
            SV **err_str_svp;
            SV **err_num_svp;
@@ -4318,7 +4318,7 @@
            netsnmp_pdu *pdu = NULL;
            struct tree *tp;
 	   oid *oid_arr;
-	   int oid_arr_len = MAX_OID_LEN;
+	   size_t oid_arr_len = MAX_OID_LEN;
            SV **sess_ptr_sv;
            SV **err_str_svp;
            SV **err_num_svp;
@@ -4455,7 +4455,7 @@
            netsnmp_pdu *response;
            struct tree *tp;
 	   oid *oid_arr;
-	   int oid_arr_len = MAX_OID_LEN;
+	   size_t oid_arr_len = MAX_OID_LEN;
            snmp_xs_cb_data *xs_cb_data;
            SV **sess_ptr_sv;
            SV **err_str_svp;
@@ -4686,7 +4686,7 @@
            char str_buf[STR_BUF_SIZE];
            char str_buf_temp[STR_BUF_SIZE];
            oid oid_arr[MAX_OID_LEN];
-           int oid_arr_len = MAX_OID_LEN;
+           size_t oid_arr_len = MAX_OID_LEN;
            char * label;
            char * iid;
            int status = FAILURE;
diff -Naur net-snmp-5.2.2/snmplib/scapi.c net-snmp-5.2.2.oden/snmplib/scapi.c
--- net-snmp-5.2.2/snmplib/scapi.c	2005-05-04 12:16:11.000000000 +0200
+++ net-snmp-5.2.2.oden/snmplib/scapi.c	2005-12-20 22:17:19.000000000 +0100
@@ -271,8 +271,10 @@
     int             properlength;
 
     u_char          buf[SNMP_MAXBUF_SMALL];
-#if  defined(USE_OPENSSL) || defined(USE_PKCS)
-    size_t             buf_len = sizeof(buf);
+#ifdef USE_OPENSSL
+    unsigned int    buf_len = sizeof(buf);
+#elif defined(USE_PKCS)
+    size_t          buf_len = sizeof(buf);
 #endif
 
     DEBUGTRACE;
