--- net-snmp-5.3/include/net-snmp/types.h.old	2005-11-07 04:34:48.000000000 +0100
+++ net-snmp-5.3/include/net-snmp/types.h	2006-02-02 08:30:56.000000000 +0100
@@ -144,7 +144,7 @@
  */
 
     typedef struct netsnmp_index_s {
-       int          len;
+       size_t      len;
        oid         *oids;
     } netsnmp_index;
 
--- net-snmp-5.3/agent/agent_registry.c.old	2006-02-02 08:20:02.000000000 +0100
+++ net-snmp-5.3/agent/agent_registry.c	2006-02-02 08:29:56.000000000 +0100
@@ -1451,7 +1451,7 @@
     lookup_cache *lookup_cache = NULL;
     netsnmp_subtree *myptr = NULL, *previous = NULL;
     int cmp = 1;
-    int ll_off = 0;
+    size_t ll_off = 0;
 
     if (subtree) {
         myptr = subtree;
