Index: squid/configure.in
diff -c squid/configure.in:1.251.2.95 squid/configure.in:1.251.2.96
*** squid/configure.in:1.251.2.95	Sat Sep  3 02:20:40 2005
--- squid/configure.in	Sat Sep 10 19:35:13 2005
***************
*** 1337,1342 ****
--- 1337,1343 ----
  	sys/mount.h \
  	sys/msg.h \
  	sys/param.h \
+ 	sys/prctl.h \
  	sys/resource.h \
  	sys/select.h\
  	sys/socket.h \
***************
*** 1814,1819 ****
--- 1815,1821 ----
  	mktime \
  	mstats \
  	poll \
+ 	prctl \
  	pthread_attr_setscope \
  	pthread_setschedparam \
  	pthread_attr_setschedparam \
Index: squid/src/tools.c
diff -c squid/src/tools.c:1.213.2.18 squid/src/tools.c:1.213.2.19
*** squid/src/tools.c:1.213.2.18	Sat Sep  3 03:14:43 2005
--- squid/src/tools.c	Sat Sep 10 19:35:14 2005
***************
*** 35,40 ****
--- 35,44 ----
  
  #include "squid.h"
  
+ #if HAVE_SYS_PRCTL_H
+ #include <sys/prctl.h>
+ #endif
+ 
  #define DEAD_MSG "\
  The Squid Cache (version %s) died.\n\
  \n\
***************
*** 557,562 ****
--- 561,571 ----
      if (setuid(Config2.effectiveUserID) < 0)
  	debug(50, 0) ("ALERT: setuid: %s\n", xstrerror());
  #endif
+ #if HAVE_PRCTL && defined(PR_SET_DUMPABLE)
+     /* Set Linux DUMPABLE flag */
+     if (Config.coredump_dir && prctl(PR_SET_DUMPABLE, 1) != 0)
+ 	debug(50, 0) ("ALERT: prctl: %s\n", xstrerror());
+ #endif
  }
  
  /* Enter a privilegied section */
***************
*** 569,574 ****
--- 578,588 ----
  #else
      setuid(0);
  #endif
+ #if HAVE_PRCTL && defined(PR_SET_DUMPABLE)
+     /* Set Linux DUMPABLE flag */
+     if (Config.coredump_dir && prctl(PR_SET_DUMPABLE, 1) != 0)
+ 	debug(50, 0) ("ALERT: prctl: %s\n", xstrerror());
+ #endif
  }
  
  /* Give up the posibility to gain privilegies.
***************
*** 589,594 ****
--- 603,613 ----
      if (setuid(uid) < 0)
  	debug(50, 1) ("no_suid: setuid: %s\n", xstrerror());
  #endif
+ #if HAVE_PRCTL && defined(PR_SET_DUMPABLE)
+     /* Set Linux DUMPABLE flag */
+     if (Config.coredump_dir && prctl(PR_SET_DUMPABLE, 1) != 0)
+ 	debug(50, 0) ("ALERT: prctl: %s\n", xstrerror());
+ #endif
  }
  
  void
Index: squid/configure
diff -c squid/configure:1.248.2.97 squid/configure:1.248.2.98
*** squid/configure:1.248.2.97	Sat Sep  3 18:12:57 2005
--- squid/configure	Sat Sep 10 19:40:01 2005
***************
*** 4097,4102 ****
--- 4097,4103 ----
  	sys/mount.h \
  	sys/msg.h \
  	sys/param.h \
+ 	sys/prctl.h \
  	sys/resource.h \
  	sys/select.h\
  	sys/socket.h \
***************
*** 7735,7740 ****
--- 7736,7742 ----
  	mktime \
  	mstats \
  	poll \
+ 	prctl \
  	pthread_attr_setscope \
  	pthread_setschedparam \
  	pthread_attr_setschedparam \
Index: squid/include/autoconf.h.in
diff -c squid/include/autoconf.h.in:1.109.2.12 squid/include/autoconf.h.in:1.109.2.13
*** squid/include/autoconf.h.in:1.109.2.12	Fri Apr 22 19:32:27 2005
--- squid/include/autoconf.h.in	Sat Sep 10 19:40:02 2005
***************
*** 506,511 ****
--- 506,514 ----
  /* Define if you have the poll function.  */
  #undef HAVE_POLL
  
+ /* Define if you have the prctl function.  */
+ #undef HAVE_PRCTL
+ 
  /* Define if you have the pthread_attr_setschedparam function.  */
  #undef HAVE_PTHREAD_ATTR_SETSCHEDPARAM
  
***************
*** 794,799 ****
--- 797,805 ----
  /* Define if you have the <sys/param.h> header file.  */
  #undef HAVE_SYS_PARAM_H
  
+ /* Define if you have the <sys/prctl.h> header file.  */
+ #undef HAVE_SYS_PRCTL_H
+ 
  /* Define if you have the <sys/resource.h> header file.  */
  #undef HAVE_SYS_RESOURCE_H
  
