diff -Naur mysql-4.1.7-orig/scripts/mysql_config.sh mysql-4.1.7/scripts/mysql_config.sh
--- mysql-4.1.7-orig/scripts/mysql_config.sh	2004-10-23 03:28:44.000000000 -0400
+++ mysql-4.1.7/scripts/mysql_config.sh	2004-10-27 21:15:48.762963874 -0400
@@ -75,7 +75,6 @@
 execdir='@libexecdir@'
 bindir='@bindir@'
 pkglibdir='@pkglibdir@'
-fix_path pkglibdir lib/mysql lib
 pkgincludedir='@pkgincludedir@'
 fix_path pkgincludedir include/mysql include
 version='@VERSION@'
