--- fileutils-4.1.1/src/touch.c.FBopts	Sat Oct 20 21:10:57 2001
+++ fileutils-4.1.1/src/touch.c	Mon Nov  5 17:36:36 2001
@@ -89,6 +89,9 @@
 /* New time to use when setting time. */
 static time_t newtime;
 
+/* Time modifier, allowing to "make file x 10 seconds older than file y" */
+static int time_modifier = 0;
+
 /* File to use for -r. */
 static char *ref_file;
 
@@ -108,6 +111,8 @@
   {"no-create", no_argument, 0, 'c'},
   {"date", required_argument, 0, 'd'},
   {"file", required_argument, 0, 'r'}, /* FIXME: phase out --file */
+  {"forward", required_argument, 0, 'F'},
+  {"backward", required_argument, 0, 'B'},
   {"reference", required_argument, 0, 'r'},
   {GETOPT_HELP_OPTION_DECL},
   {GETOPT_VERSION_OPTION_DECL},
@@ -176,9 +181,19 @@
 
   if (amtime_now)
     {
-      /* Pass NULL to utime so it will not fail if we just have
-	 write access to the file, but don't own it.  */
-      status = utime (file, NULL);
+      if (time_modifier == 0)
+	{
+          /* Pass NULL to utime so it will not fail if we just have
+	  write access to the file, but don't own it.  */
+          status = utime (file, NULL);
+	}
+      else
+	{
+	  struct utimbuf t;
+	  t.actime = time(NULL) + time_modifier;
+	  t.modtime = time(NULL) + time_modifier;
+	  status = utime (file, &t);
+	}
     }
   else
     {
@@ -190,17 +205,17 @@
 
       if (use_ref)
 	{
-	  utb.actime = ref_stats.st_atime;
-	  utb.modtime = ref_stats.st_mtime;
+	  utb.actime = ref_stats.st_atime + time_modifier;
+	  utb.modtime = ref_stats.st_mtime + time_modifier;
 	}
       else
 	utb.actime = utb.modtime = newtime;
 
       if (!(change_times & CH_ATIME))
-	utb.actime = sbuf.st_atime;
+	utb.actime = sbuf.st_atime + time_modifier;
 
       if (!(change_times & CH_MTIME))
-	utb.modtime = sbuf.st_mtime;
+	utb.modtime = sbuf.st_mtime + time_modifier;
 
       status = utime (file, &utb);
     }
@@ -232,8 +247,10 @@
 Update the access and modification times of each FILE to the current time.\n\
 \n\
   -a                     change only the access time\n\
+  -B SEC, --backward=SEC date back SEC seconds\n\
   -c, --no-create        do not create any files\n\
   -d, --date=STRING      parse STRING and use it instead of current time\n\
+  -F SEC, --forward=SEC  date forward SEC seconds\n\
   -f                     (ignored)\n\
   -m                     change only the modification time\n\
   -r, --reference=FILE   use this file's times instead of current time\n\
@@ -268,13 +285,21 @@
   change_times = no_create = use_ref = posix_date = flexible_date = 0;
   newtime = (time_t) -1;
 
-  while ((c = getopt_long (argc, argv, "acd:fmr:t:", longopts, NULL)) != -1)
+  while ((c = getopt_long (argc, argv, "B:F:acd:fmr:t:", longopts, NULL)) != -1)
     {
       switch (c)
 	{
 	case 0:
 	  break;
 
+	case 'B':
+	  time_modifier = -atoi(optarg);
+	  break;
+
+	case 'F':
+	  time_modifier = atoi(optarg);
+	  break;
+
 	case 'a':
 	  change_times |= CH_ATIME;
 	  break;
--- fileutils-4.1.1/doc/coreutils.texi.FBopts	Mon Nov  5 17:36:36 2001
+++ fileutils-4.1.1/doc/coreutils.texi	Mon Nov  5 17:36:45 2001
@@ -7216,6 +7216,14 @@
 @opindex use @r{time, changing}
 Change the access time only.
 
+@item -B
+@itemx --backward=seconds
+@opindex -B
+@opindex --backward
+Modify the time by going back @var{seconds} seconds.
+For example, @code{touch -r foo -B 5 bar} will make the file bar 5 seconds
+older than file foo.
+
 @item -c
 @itemx --no-create
 @opindex -c
@@ -7230,6 +7238,14 @@
 Use @var{time} instead of the current time.  It can contain month names,
 time zones, @samp{am} and @samp{pm}, etc.  @xref{Date input formats}.
 
+@item -F
+@itemx --forward=seconds
+@opindex -F
+@opindex --forward
+Modify the time by going forward @var{seconds} seconds.
+For example, @code{touch -r foo -F 5 bar} will make the file bar 5 seconds
+newer than file foo.
+
 @item -f
 @opindex -f
 @cindex BSD @code{touch} compatibility
--- fileutils-4.1.1/doc/coreutils.info.FBopts	Mon Nov  5 17:36:36 2001
+++ fileutils-4.1.1/doc/coreutils.info	Mon Nov  5 17:36:50 2001
@@ -5694,6 +5694,12 @@
 `--time=use'
      Change the access time only.
 
+`-B'
+`--backward=seconds'
+     Modify the time by going back `seconds' seconds.
+     For example, `touch -r foo -B 5 bar' will make the file bar 5 seconds
+     older than file foo.
+
 `-c'
 `--no-create'
      Do not create files that do not exist.
@@ -5703,6 +5709,12 @@
      Use TIME instead of the current time.  It can contain month names,
      time zones, `am' and `pm', etc.  *Note Date input formats::.
 
+`-F'
+'--forward=seconds'
+     Modify the time by going forward `seconds' seconds.
+     For example, `touch -r foo -F 5 bar' will make the file bar 5 seconds
+     newer than file foo.
+
 `-f'
      Ignored; for compatibility with BSD versions of `touch'.
 
--- fileutils-4.1.1/man/touch.1.FBopts	Sun Oct 21 10:03:14 2001
+++ fileutils-4.1.1/man/touch.1	Mon Nov  5 17:36:50 2001
@@ -16,12 +16,22 @@
 \fB\-a\fR
 change only the access time
 .TP
+\fB\-B\fR, \fB\-\-backward\fR=\fISECONDS\fR
+Modify the time by going back SECONDS seconds.
+For example, \fBtouch -r foo -B 5 bar\fR will make the file bar 5 seconds
+older than file foo.
+.TP
 \fB\-c\fR, \fB\-\-no\-create\fR
 do not create any files
 .TP
 \fB\-d\fR, \fB\-\-date\fR=\fISTRING\fR
 parse STRING and use it instead of current time
 .TP
+\fB\-F\fR, \fB\-\-forward\fR=\fISECONDS\fR
+Modify the time by going forward SECONDS seconds.
+For example, \fBtouch -r foo -F 5 bar\fR will make the file bar 5 seconds
+newer than file foo.
+.TP
 \fB\-f\fR
 (ignored)
 .TP
