--- cups/auth.c.orig	2003-12-15 22:16:52.000000000 +0100
+++ cups/auth.c	2003-12-15 22:17:19.000000000 +0100
@@ -108,8 +108,10 @@
     if ((password = cupsGetPassword(prompt)) == NULL)
       return (-1);
 
+    /*
     if (!password[0])
       return (-1);
+    */
 
     snprintf(http->userpass, sizeof(http->userpass), "%s:%s", cupsUser(),
              password);
--- scheduler/auth.c.orig	2003-12-15 22:16:59.000000000 +0100
+++ scheduler/auth.c	2003-12-15 22:17:35.000000000 +0100
@@ -991,8 +991,10 @@
     * Not doing local certificate-based authentication; check the password...
     */
 
+    /*
     if (!con->password[0])
       return (HTTP_UNAUTHORIZED);
+    */
 
    /*
     * See what kind of authentication we are doing...
@@ -1089,9 +1091,9 @@
 
 	  if (spw == NULL && strcmp(pw->pw_passwd, "x") == 0)
 	  {					/* Don't allow blank passwords! */
-	    LogMessage(L_WARN, "IsAuthorized: Username \"%s\" has no shadow password; access denied.",
+	    /* LogMessage(L_WARN, "IsAuthorized: Username \"%s\" has no shadow password; access denied.",
         	       con->username);
-	    return (HTTP_UNAUTHORIZED);	/* No such user or bad shadow file */
+	       return (HTTP_UNAUTHORIZED);*/	/* No such user or bad shadow file */
 	  }
 
 #    ifdef DEBUG
@@ -1106,9 +1108,9 @@
 	  if (pw->pw_passwd[0] == '\0')
 #  endif /* HAVE_SHADOW_H */
 	  {					/* Don't allow blank passwords! */
-	    LogMessage(L_WARN, "IsAuthorized: Username \"%s\" has no password; access denied.",
+	    /* LogMessage(L_WARN, "IsAuthorized: Username \"%s\" has no password; access denied.",
         	       con->username);
-	    return (HTTP_UNAUTHORIZED);
+	       return (HTTP_UNAUTHORIZED); */
 	  }
 
 	 /*
