--- openldap-2.3.4/contrib/slapd-modules/smbk5pwd/Makefile.only-smb	2004-04-02 13:06:38.000000000 +0200
+++ openldap-2.3.4/contrib/slapd-modules/smbk5pwd/Makefile	2005-08-05 16:25:45.000000000 +0200
@@ -14,7 +14,7 @@
 CC=gcc
 
 # Omit DO_KRB5 or DO_SAMBA if you don't want to support it.
-DEFS=-DDO_KRB5 -DDO_SAMBA
+DEFS=-DDO_SAMBA
 
 HEIMDAL_INC=-I/usr/heimdal/include
 SSL_INC=
@@ -23,8 +23,8 @@
 
 HEIMDAL_LIB=-L/usr/heimdal/lib -lkrb5 -lkadm5srv
 SSL_LIB=-lcrypto
-LDAP_LIB=-lldap_r -llber
-LIBS=$(LDAP_LIB) $(HEIMDAL_LIB) $(SSL_LIB)
+LDAP_LIB=-L../../../libraries/libldap_r/.libs -L../../../libraries/liblber/.libs -lldap_r -llber
+LIBS=$(LDAP_LIB) $(SSL_LIB)
 
 all:	smbk5pwd.la
 
