--- ./src/ls.c.dumbterm	2004-01-12 22:43:13.000000000 +0100
+++ ./src/ls.c	2004-01-12 22:43:13.000000000 +0100
@@ -90,6 +90,7 @@
 
 #include "system.h"
 #include <fnmatch.h>
+#include <termcap.h>
 
 #include "acl.h"
 #include "argmatch.h"
@@ -1593,7 +1594,9 @@
 
 	    print_with_color = (i == color_always
 				|| (i == color_if_tty
-				    && isatty (STDOUT_FILENO)));
+				    && isatty (STDOUT_FILENO))
+				    && tgetent(NULL, getenv("TERM")) > 0
+				    && tgetstr("md", NULL) > 0);
 
 	    if (print_with_color)
 	      {
--- ./src/Makefile.am.tv	2004-06-23 18:10:19.713468443 +0200
+++ ./src/Makefile.am	2004-06-23 18:10:37.195040604 +0200
@@ -32,10 +32,10 @@
 # replacement functions defined in libfetish.a.
 LDADD = ../lib/libfetish.a $(LIBINTL) ../lib/libfetish.a
 
-dir_LDADD = $(LDADD) $(LIB_CLOCK_GETTIME)
-ls_LDADD = $(LDADD) $(LIB_CLOCK_GETTIME)
+dir_LDADD = $(LDADD) $(LIB_CLOCK_GETTIME) -ltermcap
+ls_LDADD = $(LDADD) $(LIB_CLOCK_GETTIME) -ltermcap
 shred_LDADD = $(LDADD) $(LIB_CLOCK_GETTIME)
-vdir_LDADD = $(LDADD) $(LIB_CLOCK_GETTIME)
+vdir_LDADD = $(LDADD) $(LIB_CLOCK_GETTIME) -ltermcap
 
 ## If necessary, add -lm to resolve use of pow in lib/strtod.c.
 sort_LDADD = $(LDADD) $(POW_LIB)
