--- xterm-172/input.c.chmou	2002-12-27 22:05:22.000000000 +0100
+++ xterm-172/input.c	2003-02-08 07:24:01.000000000 +0100
@@ -721,6 +721,16 @@
 		eightbit = False;
 		unparseputc(ESC, pty);	/* escape */
 	    }
+
+	    /* Make alt like meta don't know if the semantic is right
+	       but For Mdk we need that*/
+	    if (screen->meta_sends_esc 
+		&& ((event->state & term->misc.alt_left) != 0 
+		    || (event->state & term->misc.alt_right)) != 0) {
+	      TRACE(("...input-char is modified by ESC\n"));
+	      eightbit = False;
+	      unparseputc(ESC, pty);	/* escape */
+	    }
 #endif
 	    if (eightbit && screen->input_eight_bits) {
 		if (CharOf(*string) < 128) {
--- xterm-172/charproc.c.chmou	2002-12-27 22:05:21.000000000 +0100
+++ xterm-172/charproc.c	2003-02-08 07:21:11.000000000 +0100
@@ -405,7 +405,7 @@
     Bres(XtNjumpScroll, XtCJumpScroll, screen.jumpscroll, TRUE),
     Bres(XtNloginShell, XtCLoginShell, misc.login_shell, FALSE),
     Bres(XtNmarginBell, XtCMarginBell, screen.marginbell, FALSE),
-    Bres(XtNmetaSendsEscape, XtCMetaSendsEscape, screen.meta_sends_esc, FALSE),
+    Bres(XtNmetaSendsEscape, XtCMetaSendsEscape, screen.meta_sends_esc, TRUE),
     Bres(XtNmultiScroll, XtCMultiScroll, screen.multiscroll, FALSE),
     Bres(XtNoldXtermFKeys, XtCOldXtermFKeys, screen.old_fkeys, FALSE),
     Bres(XtNpopOnBell, XtCPopOnBell, screen.poponbell, FALSE),
