--- e2fsprogs-1.38/po/Makefile.in.in.pottcdate	2005-06-29 19:42:20.000000000 +0200
+++ e2fsprogs-1.38/po/Makefile.in.in	2005-11-10 17:35:39.000000000 +0100
@@ -117,10 +117,12 @@
 	      rm -f $(DOMAIN).1po $(DOMAIN).2po $(DOMAIN).po; \
 	    else \
 	      rm -f $(DOMAIN).1po $(DOMAIN).2po $(srcdir)/$(DOMAIN).pot && \
-	      mv $(DOMAIN).po $(srcdir)/$(DOMAIN).pot; \
+	      sed -f remove-potcdate.sed < $(DOMAIN).po > $(srcdir)/$(DOMAIN).pot && \
+	      rm -f $(DOMAIN).po; \
 	    fi; \
 	  else \
-	    mv $(DOMAIN).po $(srcdir)/$(DOMAIN).pot; \
+	    sed -f remove-potcdate.sed < $(DOMAIN).po > $(srcdir)/$(DOMAIN).pot && \
+	    rm -f $(DOMAIN).po; \
 	  fi; \
 	}
 
