--- openldap-2.1.25/include/ldap_defaults.h.data-locations	2004-01-25 22:24:35.000000000 +0200
+++ openldap-2.1.25/include/ldap_defaults.h	2004-01-26 20:06:01.000000000 +0200
@@ -72,7 +72,7 @@
  */
 	/* location of the default slapd config file */
 #define SLAPD_DEFAULT_CONFIGFILE	LDAP_SYSCONFDIR LDAP_DIRSEP "slapd.conf"
-#define SLAPD_DEFAULT_DB_DIR		LDAP_DATADIR LDAP_DIRSEP "openldap-data"
+#define SLAPD_DEFAULT_DB_DIR		"/var/lib/ldap"
 #define SLAPD_DEFAULT_DB_MODE		0600
 #define SLAPD_DEFAULT_UCDATA		LDAP_DATADIR LDAP_DIRSEP "ucdata"
 	/* default max deref depth for aliases */
--- openldap-2.1.25/servers/slurpd/slurp.h.data-locations	2003-09-14 09:21:56.000000000 +0200
+++ openldap-2.1.25/servers/slurpd/slurp.h	2004-01-26 18:42:19.000000000 +0200
@@ -53,7 +53,7 @@
 #define SERVICE_NAME	OPENLDAP_PACKAGE "-slurpd"
 
 /* Default directory for slurpd's private copy of replication logs */
-#define	DEFAULT_SLURPD_REPLICA_DIR	LDAP_RUNDIR LDAP_DIRSEP "openldap-slurp"
+#define	DEFAULT_SLURPD_REPLICA_DIR	 "/var/lib/ldap"
 
 /* Default name for slurpd's private copy of the replication log */
 #define	DEFAULT_SLURPD_REPLOGFILE	"slurpd.replog"
