--- compress_files.avx	2007-12-10 07:44:15.371310750 +0000
+++ compress_files	2007-12-10 07:44:15.371310750 +0000
@@ -112,10 +112,10 @@
     system "ln", "-sf",$sodest,$sofile;
 }
 
-my @files = split(/\n/, `find usr/info usr/share/info usr/man usr/share/man usr/X11*/man usr/lib/perl5/man -type f ! -name "*.gz" -a ! -name "*.bz2" -a ! -name "*.lzma"  ! -name 'dir' ! -name 'whatis' 2>/dev/null || true`);
+my @files = split(/\n/, `find usr/info usr/share/info usr/local/share/info usr/man usr/share/man usr/local/share/man usr/X11*/man usr/lib/perl5/man -type f ! -name "*.gz" -a ! -name "*.bz2" -a ! -name "*.lzma"  ! -name 'dir' ! -name 'whatis' 2>/dev/null || true`);
 
 if ($ext ne '.gz') {
-    my @gz_files = split(/\n/, `find usr/info usr/share/info usr/man usr/share/man usr/X11*/man usr/lib/perl5/man -type f -name "*.gz" 2>/dev/null || true`);
+    my @gz_files = split(/\n/, `find usr/info usr/share/info usr/local/share/info usr/man usr/share/man usr/local/share/man usr/X11*/man usr/lib/perl5/man -type f -name "*.gz" 2>/dev/null || true`);
     if (@gz_files) {
         xargs(\@gz_files, "gzip", "-d");
         die "Something wrong with the decompression of the gzip man/info file"
@@ -124,7 +124,7 @@
     }
 }
 if ($ext ne '.bz2') {
-    my @bz_files = split(/\n/, `find usr/info usr/share/info usr/man usr/share/man usr/X11*/man usr/lib/perl5/man -type f -name "*.bz2" 2>/dev/null || true`);
+    my @bz_files = split(/\n/, `find usr/info usr/share/info usr/local/share/info usr/man usr/share/man usr/local/share/man usr/X11*/man usr/lib/perl5/man -type f -name "*.bz2" 2>/dev/null || true`);
     if (@bz_files) {
         xargs(\@bz_files, "bzip2", "-d");
         die "Something wrong with the decompression of the bzip2 man/info file"
@@ -133,7 +133,7 @@
     }
 }
 if ($ext ne '.lzma') {
-    my @lzma_files = split(/\n/, `find usr/info usr/share/info usr/man usr/share/man usr/X11*/man usr/lib/perl5/man -type f -name "*.lzma" 2>/dev/null || true`);
+    my @lzma_files = split(/\n/, `find usr/info usr/share/info usr/local/share/info usr/man usr/share/man usr/local/share/man usr/X11*/man usr/lib/perl5/man -type f -name "*.lzma" 2>/dev/null || true`);
     if (@lzma_files) { 
         xargs(\@lzma_files, "lzmash", "-d");
         die "Something wrong with the decompression of the lzma man/info file"
