--- rpm-4.2/lib/rpmvercmp.c.rpmvercmp	2003-01-10 20:49:29.000000000 +0100
+++ rpm-4.2/lib/rpmvercmp.c	2003-04-28 11:45:39.000000000 +0200
@@ -45,7 +45,7 @@
 	/* grab first completely alpha or completely numeric segment */
 	/* leave one and two pointing to the start of the alpha or numeric */
 	/* segment and walk str1 and str2 to end of segment */
-	if (xisdigit(*str1)) {
+	if (xisdigit(*str1) || xisdigit(*str2)) {
 	    while (*str1 && xisdigit(*str1)) str1++;
 	    while (*str2 && xisdigit(*str2)) str2++;
 	    isnum = 1;
@@ -64,12 +64,14 @@
 	*str2 = '\0';
 /*@=boundswrite@*/
 
+#if 0
 	/* take care of the case where the two version segments are */
 	/* different types: one numeric, the other alpha (i.e. empty) */
 	if (one == str1) return -1;	/* arbitrary */
 	/* XXX See patch #60884 (and details) from bugzilla #50977. */
 	if (two == str2) return (isnum ? 1 : -1);
-
+#endif
+	
 	if (isnum) {
 	    /* this used to be done by converting the digit segments */
 	    /* to ints using atoi() - it's changed because long  */
