--- Linux-PAM-0.99.3.0/modules/pam_limits/limits.conf.enable_rt	2005-08-16 16:02:28.000000000 +0200
+++ Linux-PAM-0.99.3.0/modules/pam_limits/limits.conf	2006-01-28 14:51:28.000000000 +0100
@@ -47,4 +47,9 @@
 #ftp             hard    nproc           0
 #@student        -       maxlogins       4
 
+*               -       rtprio           0
+*               -       nice             0
+@audio          -       rtprio           50
+@audio          -       nice             -10
+
 # End of file
