--- exim-4.33/src/verify.c.can-2004-0400	2004-05-05 04:08:36.000000000 -0600
+++ exim-4.33/src/verify.c	2004-05-08 07:50:54.000000000 -0600
@@ -1264,19 +1264,13 @@
 
     if (recipient == NULL && Ustrcmp(errmess, "empty address") != 0)
       {
-      uschar hname[64];
-      uschar *t = h->text;
-      uschar *tt = hname;
       uschar *verb = US"is";
+      uschar *t = ss;
       int len;
 
-      while (*t != ':') *tt++ = *t++;
-      *tt = 0;
-
       /* Arrange not to include any white space at the end in the
       error message. */
 
-      t = ss;
       while (t > s && isspace(t[-1])) t--;
 
       /* Add the address which failed to the error message, since in a
