--- gcc-4.0.3/libjava/configure.libtool1.4-lib64	2005-11-04 11:50:21.000000000 +0100
+++ gcc-4.0.3/libjava/configure	2005-11-04 11:59:16.000000000 +0100
@@ -4827,6 +4827,7 @@ x86_64-*linux*|ppc*-*linux*|powerpc*-*li
   (exit $ac_status); }; then
     case "`/usr/bin/file conftest.o`" in
     *32-bit*)
+      lt_cv_cc_64bit_output=no
       case $host in
         x86_64-*linux*)
           LD="${LD-ld} -m elf_i386"
@@ -4843,6 +4844,7 @@ x86_64-*linux*|ppc*-*linux*|powerpc*-*li
       esac
       ;;
     *64-bit*)
+      lt_cv_cc_64bit_output=yes
       case $host in
         x86_64-*linux*)
           LD="${LD-ld} -m elf_x86_64"
--- gcc-4.0.3/ltconfig.libtool1.4-lib64	2005-11-03 15:52:21.000000000 +0100
+++ gcc-4.0.3/ltconfig	2005-11-04 11:55:46.000000000 +0100
@@ -1265,6 +1265,13 @@ linux-gnu*)
   # before this can be enabled.
   hardcode_into_libs=yes
 
+  case $host_cpu:$lt_cv_cc_64bit_output in
+  powerpc64:yes | s390x:yes | sparc64:yes | x86_64:yes)
+    sys_lib_dlsearch_path_spec="/lib64 /usr/lib64 /usr/X11R6/lib64"
+    sys_lib_search_path_spec="/lib64 /usr/lib64 /usr/local/lib64 /usr/X11R6/lib64"
+    ;;
+  esac
+
   # We used to test for /lib/ld.so.1 and disable shared libraries on
   # powerpc, because MkLinux only supported shared libraries with the
   # GNU dynamic linker.  Since this was broken with cross compilers,
--- gcc-4.0.3/libtool.m4.libtool1.4-lib64	2005-11-03 15:52:21.000000000 +0100
+++ gcc-4.0.3/libtool.m4	2005-11-04 11:58:33.000000000 +0100
@@ -195,6 +195,7 @@ x86_64-*linux*|ppc*-*linux*|powerpc*-*li
   if AC_TRY_EVAL(ac_compile); then
     case "`/usr/bin/file conftest.o`" in
     *32-bit*)
+      lt_cv_cc_64bit_output=no
       case $host in
         x86_64-*linux*)
           LD="${LD-ld} -m elf_i386"
@@ -211,6 +212,7 @@ x86_64-*linux*|ppc*-*linux*|powerpc*-*li
       esac
       ;;
     *64-bit*)
+      lt_cv_cc_64bit_output=yes
       case $host in
         x86_64-*linux*)
           LD="${LD-ld} -m elf_x86_64"
--- gcc-4.0.3/ltmain.sh.libtool1.4-lib64	2005-11-04 08:26:19.000000000 +0100
+++ gcc-4.0.3/ltmain.sh	2005-11-04 12:26:44.000000000 +0100
@@ -2048,6 +2048,13 @@ EOF
 
 	  if test "$linkmode,$pass" = "prog,link" ||
 	   { test $linkmode = lib && test $hardcode_into_libs = yes; }; then
+	    # canonicalize libdir path
+	    case "$libdir" in
+	    [\\/]*) c_libdir=`cd "$libdir" && pwd`;;
+	    esac
+	    if test -z "$c_libdir"; then
+		c_libdir="$libdir"
+	    fi
 	    # Hardcode the library path.
 	    # Skip directories that are in the system default run-time
 	    # search path.
@@ -2061,10 +2068,10 @@ EOF
 	      ;;
 	    esac
 	    case " $sys_lib_dlsearch_path " in
-	    *" $libdir "*) ;;
+	    *" $c_libdir "*) ;;
 	    *)
 	      case "$finalize_rpath " in
-	      *" $libdir "*) ;;
+	      *" $c_libdir "*) ;;
 	      *) finalize_rpath="$finalize_rpath $libdir"
 	      esac
 	      ;;
@@ -3661,9 +3668,16 @@ EOF
       if test -n "$rpath$xrpath"; then
 	# If the user specified any rpath flags, then add them.
 	for libdir in $rpath $xrpath; do
+	  # canonicalize libdir path
+	  case "$libdir" in
+	  [\\/]*) c_libdir=`cd "$libdir" && pwd`;;
+	  esac
+	  if test -z "$c_libdir"; then
+	    c_libdir="$libdir"
+	  fi
 	  # This is the magic to use -rpath.
-	  case "$finalize_rpath " in
-	  *" $libdir "*) ;;
+	  case " $finalize_rpath " in
+	  *" $c_libdir "*|"  ");;
 	  *) finalize_rpath="$finalize_rpath $libdir" ;;
 	  esac
 	done
