--- linux-2.4.20/arch/x86_64/config.in.mb02-2.orig	2003-04-14 23:17:03.000000000 -0400
+++ linux-2.4.20/arch/x86_64/config.in	2003-04-14 23:21:42.000000000 -0400
@@ -254,3 +254,6 @@ fi
 endmenu
 
 source lib/Config.in
+
+source 3rdparty/Config.in
+
--- linux-2.4.20/arch/i386/config.in.mb02-2.orig	2003-04-14 22:56:14.000000000 -0400
+++ linux-2.4.20/arch/i386/config.in	2003-04-14 23:21:14.000000000 -0400
@@ -507,3 +507,5 @@ fi
 endmenu
 
 source lib/Config.in
+
+source 3rdparty/Config.in
diff -uNp t1/Makefile.mb02-3.orig t1/Makefile
--- t1/Makefile.mb02-3.orig	2003-08-01 18:40:50.000000000 +0200
+++ t1/Makefile	2003-08-01 19:12:32.000000000 +0200
@@ -130,7 +130,7 @@ CORE_FILES	=kernel/kernel.o mm/mm.o fs/f
 NETWORKS	=net/network.o
 
 LIBS		=$(TOPDIR)/lib/lib.a
-SUBDIRS		=kernel drivers mm fs net ipc lib crypto
+SUBDIRS		=kernel drivers mm fs net ipc lib crypto 3rdparty
 
 ifeq ($(CONFIG_KDB),y)
 CORE_FILES	+= kdb/kdb.o
@@ -206,6 +206,9 @@ DRIVERS-$(CONFIG_CRYPTO) += crypto/crypt
 DRIVERS-$(CONFIG_SENSORS) += drivers/sensors/sensor.o
 DRIVERS := $(DRIVERS-y)
 
+ifeq (3rdparty/Makefile.drivers ,$(wildcard 3rdparty/Makefile.drivers))
+include 3rdparty/Makefile.drivers
+endif
 
 # files removed with 'make clean'
 CLEAN_FILES = \
@@ -258,6 +261,8 @@ MRPROPER_FILES = \
 	kdb/gen-kdb_cmds.c \
 	.hdepend scripts/mkdep scripts/split-include scripts/docproc \
 	$(TOPDIR)/include/linux/modversions.h \
+	3rdparty/Makefile.meta 3rdparty/Makefile 3rdparty/Makefile.drivers \
+        3rdparty/Config.in \
 	kernel.spec
 
 # directories removed with 'make mrproper'
@@ -337,7 +342,11 @@ symlinks:
 		mkdir include/linux/modules; \
 	fi
 
-preconfig: include/linux/version.h symlinks
+# Create Makefile for all the 3rdparty stuff
+3rdparty/Makefile.drivers:
+	(cd 3rdparty ; perl ./mkbuild.pl)
+
+preconfig: include/linux/version.h symlinks 3rdparty/Makefile.drivers
 	if [ -f .need_mrproper ]; then \
 		rm .need_mrproper; \
 		$(MAKE) mrproper;  \
