--- rpm-4.2.2/lib/rpmrc.c.ppc64	2004-11-23 16:42:54.000000000 +0000
+++ rpm-4.2.2/lib/rpmrc.c	2004-11-24 17:50:50.691415615 +0000
@@ -1279,6 +1279,10 @@ static void defaultMachine(/*@out@*/ con
 #	endif
 
 #	if defined(__linux__) && defined(__powerpc__)
+#	if defined(__powerpc64__)
+	/* TODO: power4, power5? */
+	strcpy(un.machine, "ppc64");
+#	else
 	{
 	    unsigned pvr = 0;
 	    __sighandler_t oldh = signal(SIGILL, mfspr_ill);
@@ -1301,6 +1305,7 @@ static void defaultMachine(/*@out@*/ con
 	    }
 	}
 #	endif
+#	endif
 
 	/* the uname() result goes through the arch_canon table */
 	canon = lookupInCanonTable(un.machine,
