--- autofs-4.0.0/modules/Makefile.fpic	Fri Oct  5 15:47:06 2001
+++ autofs-4.0.0/modules/Makefile	Fri Oct  5 15:53:17 2001
@@ -38,7 +41,7 @@
   MODS += lookup_ldap.so
 endif
 
-CFLAGS += -I../include -fpic -DAUTOFS_LIB_DIR=\"$(autofslibdir)\" -DPATH_AUTOMOUNT=\"$(sbindir)/automount\"
+CFLAGS += -I../include -fPIC -DAUTOFS_LIB_DIR=\"$(autofslibdir)\" -DPATH_AUTOMOUNT=\"$(sbindir)/automount\"
 
 all: $(MODS)
 
