Index: server/Makefile.in
===================================================================
RCS file: /store/cvs/root/httpd-2.0/server/Makefile.in,v
retrieving revision 1.85
retrieving revision 1.86
diff -u -r1.85 -r1.86
--- server/Makefile.in	17 Feb 2003 01:00:50 -0000	1.85
+++ server/Makefile.in	6 Apr 2003 19:59:02 -0000	1.86
@@ -70,6 +70,9 @@
 # wtf does this have to be explicit????
 exports.lo: exports.c
 
+# Needed to allow exports.c to be generated in a parallel build successfully
+.NOTPARALLEL: $(top_builddir)/server/exports.c
+
 # Rule to make def file for OS/2 core dll
 ApacheCoreOS2.def: exports.c export_vars.h $(top_srcdir)/os/$(OS_DIR)/core_header.def
 	cat $(top_srcdir)/os/$(OS_DIR)/core_header.def > $@
