Newer autoheader complains about missing defaults.
--- shadow-20000902/acconfig.h	Fri Jan 25 16:51:15 2002
+++ shadow-20000902/acconfig.h	Fri Jan 25 16:51:10 2002
@@ -128,6 +128,9 @@
 /* Define if you want gdbm for TCFS. */
 #undef TCFS_GDBM_SUPPORT
 
+/* Define to support USG (Unix Systems Group?) behavior. */
+#undef USG
+
 /* Define to support Pluggable Authentication Modules.  */
 #undef USE_PAM
 
