diff -urNp patch-2.5.4.orig/Makefile.in patch-2.5.4/Makefile.in
--- patch-2.5.4.orig/Makefile.in	Mon Aug 30 08:37:54 1999
+++ patch-2.5.4/Makefile.in	Tue Nov 20 10:53:51 2001
@@ -97,11 +97,11 @@ patch$(EXEEXT): $(OBJS)
 	$(CC) -o $@ $(CFLAGS) $(LDFLAGS) $(OBJS) $(LIBS)
 
 install:: all installdirs
-	$(INSTALL_PROGRAM) patch$(EXEEXT) $(bindir)/$(patch_name)$(EXEEXT)
-	-$(INSTALL_DATA) $(srcdir)/patch.man $(man1dir)/$(patch_name)$(man1ext)
+	$(INSTALL_PROGRAM) patch$(EXEEXT) $(DESTDIR)/$(bindir)/$(patch_name)$(EXEEXT)
+	-$(INSTALL_DATA) $(srcdir)/patch.man $(DESTDIR)/$(man1dir)/$(patch_name)$(man1ext)
 
 installdirs::
-	$(SHELL) $(srcdir)/mkinstalldirs $(bindir) $(man1dir)
+	$(SHELL) $(srcdir)/mkinstalldirs $(DESTDIR)/$(bindir) $(DESTDIR)/$(man1dir)
 
 install-strip::
 	$(MAKE) INSTALL_PROGRAM='$(INSTALL_PROGRAM) -s' install
