--- libpng-1.2.4/pngerror.c.org	2004-04-19 13:32:31.000000000 -0600
+++ libpng-1.2.4/pngerror.c	2004-04-19 13:32:31.000000000 -0600
@@ -135,10 +135,12 @@
       buffer[iout] = 0;
    else
    {
+      int len = strnlen(error_message, 64);
+      
       buffer[iout++] = ':';
       buffer[iout++] = ' ';
-      png_memcpy(buffer+iout, error_message, 64);
-      buffer[iout+63] = 0;
+      png_memcpy(buffer+iout, error_message, len);
+      buffer[iout+len-1] = 0;
    }
 }
 
