--- ./lib/signature.c.region_trailer	2005-02-02 19:26:14.000000000 +0100
+++ ./lib/signature.c	2005-05-08 20:17:24.993026392 +0200
@@ -262,7 +262,7 @@ rpmRC rpmReadSignature(FD_t fd, Header *
 
 	xx = headerVerifyInfo(1, dl, info, &entry->info, 1);
 	if (xx != -1 ||
-	    !(entry->info.tag == RPMTAG_HEADERSIGNATURES
+	    !((entry->info.tag == RPMTAG_HEADERSIGNATURES || entry->info.tag == RPMTAG_HEADERIMAGE)
 	   && entry->info.type == RPM_BIN_TYPE
 	   && entry->info.count == REGION_TAG_COUNT))
 	{
