--- tix8.2.0/library/Balloon.tcl.perf	Tue Nov  7 22:31:12 2000
+++ tix8.2.0/library/Balloon.tcl	Fri Jul  6 11:50:45 2001
@@ -118,21 +118,23 @@
 # "RAW" event bindings:
 #----------------------------------------------------------------------
 
-bind all <B1-Motion> 	    "+tixBalloon_XXMotion %X %Y 1"
-bind all <B2-Motion> 	    "+tixBalloon_XXMotion %X %Y 2"
-bind all <B3-Motion> 	    "+tixBalloon_XXMotion %X %Y 3"
-bind all <B4-Motion> 	    "+tixBalloon_XXMotion %X %Y 4"
-bind all <B5-Motion> 	    "+tixBalloon_XXMotion %X %Y 5"
-bind all <Any-Motion> 	    "+tixBalloon_XXMotion %X %Y 0"
-bind all <Leave>      	    "+tixBalloon_XXMotion %X %Y %b"
-bind all <Button>      	    "+tixBalloon_XXButton   %X %Y %b"
-bind all <ButtonRelease>    "+tixBalloon_XXButtonUp %X %Y %b"
+#bind all <B1-Motion> 	    "+tixBalloon_XXMotion %X %Y 1"
+#bind all <B2-Motion> 	    "+tixBalloon_XXMotion %X %Y 2"
+#bind all <B3-Motion> 	    "+tixBalloon_XXMotion %X %Y 3"
+#bind all <B4-Motion> 	    "+tixBalloon_XXMotion %X %Y 4"
+#bind all <B5-Motion> 	    "+tixBalloon_XXMotion %X %Y 5"
+#bind all <Any-Motion> 	    "+tixBalloon_XXMotion %X %Y 0"
+bind all <Enter>	"+tixBalloon_XXMotion %W %X %Y 0"
+bind all <Leave>	"+tixBalloon_XXMotion %W %X %Y %b"
+#bind all <Button>      	    "+tixBalloon_XXButton   %X %Y %b"
+#bind all <ButtonRelease>    "+tixBalloon_XXButtonUp %X %Y %b"
 
-proc tixBalloon_XXMotion {rootX rootY b} {
+proc tixBalloon_XXMotion {cw rootX rootY b} {
     global tixBalloon
 
+#   set cw [winfo containing $rootX $rootY]
     foreach w $tixBalloon(bals) {
-	tixBalloon:XXMotion $w $rootX $rootY $b
+	tixBalloon:XXMotion $w $cw $rootX $rootY $b
     }
 }
 
@@ -184,7 +186,7 @@
     return 1
 }
 
-proc tixBalloon:XXMotion {w rootX rootY b} {
+proc tixBalloon:XXMotion {w cw rootX rootY b} {
     upvar #0 $w data
 
     if {[info exists $data(-state)]} {
@@ -353,7 +355,7 @@
 
     set data(isActive) 1
 
-    after 200 tixBalloon:Verify $w
+#    after 200 tixBalloon:Verify $w
 }
 
 
