--- screen-3.9.13/configure.in.no-libelf	2003-01-20 15:16:24.000000000 +0100
+++ screen-3.9.13/configure.in	2003-01-20 15:19:23.000000000 +0100
@@ -196,7 +196,7 @@ AC_EGREP_CPP(yes,
 ], LIBS="$LIBS -lsocket -linet";seqptx=1)
 
 oldlibs="$LIBS"
-LIBS="$LIBS -lelf"
+LIBS="$LIBS"
 AC_CHECKING(SVR4)
 AC_TRY_LINK([#include <utmpx.h>
 ],,
--- screen-3.9.13/configure.no-libelf	2002-08-12 16:34:23.000000000 +0200
+++ screen-3.9.13/configure	2003-01-20 15:19:50.000000000 +0100
@@ -3083,7 +3083,7 @@ rm -f conftest*
 
 
 oldlibs="$LIBS"
-LIBS="$LIBS -lelf"
+LIBS="$LIBS"
 { echo "$as_me:$LINENO: checking SVR4..." >&5
 echo "$as_me: checking SVR4..." >&6;}
 echo "$as_me:$LINENO: checking for ANSI C header files" >&5
