--- xorg-x11-6.8.2/xc/programs/Xserver/hw/xnest/Screen.c.xnest-stacking	2004-08-12 00:40:14.000000000 +0200
+++ xorg-x11-6.8.2/xc/programs/Xserver/hw/xnest/Screen.c	2005-05-25 20:31:49.757395371 +0200
@@ -249,6 +249,16 @@
     xnestHeight = gattributes.height;
   }
 
+  /* myNum */
+  /* id */
+  miScreenInit(pScreen, NULL, xnestWidth, xnestHeight, 1, 1, xnestWidth,
+	       rootDepth,
+	       numDepths, depths,
+	       defaultVisual, /* root visual */
+	       numVisuals, visuals);
+
+/*  miInitializeBackingStore(pScreen); */
+
   pScreen->defColormap = (Colormap) FakeClientID(0);
   pScreen->minInstalledCmaps = MINCMAPS;
   pScreen->maxInstalledCmaps = MAXCMAPS;
@@ -332,15 +342,6 @@
   pScreen->WakeupHandler = (ScreenWakeupHandlerProcPtr)NoopDDA;
   pScreen->blockData = NULL;
   pScreen->wakeupData = NULL;
-  /* myNum */
-  /* id */
-  miScreenInit(pScreen, NULL, xnestWidth, xnestHeight, 1, 1, xnestWidth,
-	       rootDepth,
-	       numDepths, depths,
-	       defaultVisual, /* root visual */
-	       numVisuals, visuals);
-
-/*  miInitializeBackingStore(pScreen); */
 
   miPointerInitialize (pScreen, &xnestPointerSpriteFuncs, 
 		       &xnestPointerCursorFuncs, True);
