Assume the file size can be a long long.

--- krb5-1.2.2/src/appl/gssftp/ftpd/ftpcmd.y	Wed Jun 27 14:25:16 2001
+++ krb5-1.2.2/src/appl/gssftp/ftpd/ftpcmd.y	Wed Jun 27 14:25:18 2001
@@ -1506,12 +1506,12 @@
 		    (stbuf.st_mode&S_IFMT) != S_IFREG)
 			reply(550, "%s: not a plain file.", filename);
 		else
-			reply(213, "%lu", stbuf.st_size);
+			reply(213, "%llu", stbuf.st_size);
 		break;}
 	case TYPE_A: {
 		FILE *fin;
 		register int c;
-		register long count;
+		register long long count;
 		struct stat stbuf;
 		fin = fopen(filename, "r");
 		if (fin == NULL) {
@@ -1533,7 +1533,7 @@
 		}
 		(void) fclose(fin);
 
-		reply(213, "%ld", count);
+		reply(213, "%lld", count);
 		break;}
 	default:
 		reply(504, "SIZE not implemented for Type %c.", "?AEIL"[type]);
