--- glibc-2.2.5/sysdeps/unix/sysv/linux/powerpc/Makefile.ppc	Tue Apr 30 07:42:01 2002
+++ glibc-2.2.5/sysdeps/unix/sysv/linux/powerpc/Makefile	Tue Apr 30 07:42:37 2002
@@ -6,3 +6,9 @@
 ifeq ($(subdir),resource)
 sysdep_routines += oldgetrlimit64
 endif
+
+ifeq ($(subdir),elf)
+sysdep-others += lddlibc4
+install-bin += lddlibc4
+endif
+
--- glibc-2.2.5/shlib-versions.ppc	Tue Apr 30 07:46:42 2002
+++ glibc-2.2.5/shlib-versions	Tue Apr 30 09:01:35 2002
@@ -73,6 +73,8 @@
 sparc64-.*-linux.*	ld=ld-linux.so.2	GLIBC_2.2
 sparc.*-.*-linux.*	ld=ld-linux.so.2
 alpha.*-.*-linux.*	ld=ld-linux.so.2
+ppc.*-.*-linux.*	ld=ld-linux.so.2
+powerpc.*-.*-linux.*	ld=ld-linux.so.2
 arm.*-.*-linux.*	ld=ld-linux.so.2
 sh.*-.*-linux.*		ld=ld-linux.so.2	GLIBC_2.2
 ia64-.*-linux.*		ld=ld-linux-ia64.so.2	GLIBC_2.2
