--- tmdns-0.1/configure.in.libresolv	2003-08-19 16:05:45.000000000 +0200
+++ tmdns-0.1/configure.in	2003-08-19 16:36:25.000000000 +0200
@@ -117,7 +117,8 @@ dnl AC_PATH_PROG(DVIPS, dvips)
 dnl AC_PATH_PROG(LYNX, lynx)
 
 dnl Checks for libraries.
-AC_CHECK_LIB(resolv, dn_expand)
+AC_CHECK_LIB(resolv, dn_expand, [],
+AC_CHECK_LIB(resolv, __dn_expand))
 AC_CHECK_LIB(socket, socket)
 
 AC_HEADER_SYS_WAIT
