--- net/ipsvd-0.12.1/src/Makefile.avx	2006-02-04 12:16:47.000000000 -0700
+++ net/ipsvd-0.12.1/src/Makefile	2006-12-30 12:30:41.000000000 -0700
@@ -15,10 +15,10 @@
 
 sslsvd: load sslsvd.o ssl_io.o sslerror_str.o ipsvd_log.o ipsvd_fmt.o \
 	  ipsvd_check.o ipsvd_hostname.o ipsvd_phcc.o ipsvd_scan.o uidgid.o \
-	  cdb.a dns.a unix.a byte.a time.a socket.lib matrixssl.a
+	  cdb.a dns.a unix.a byte.a time.a socket.lib /usr/lib/dietlibc/lib-{ARCH}/libmatrixssl.a
 	./load sslsvd ssl_io.o sslerror_str.o ipsvd_log.o ipsvd_fmt.o \
 	  ipsvd_check.o ipsvd_hostname.o ipsvd_phcc.o ipsvd_scan.o uidgid.o \
-	  cdb.a dns.a unix.a byte.a time.a `cat socket.lib` matrixssl.a
+	  cdb.a dns.a unix.a byte.a time.a `cat socket.lib` /usr/lib/dietlibc/lib-{ARCH}/libmatrixssl.a
 
 udpsvd: load udpsvd.o ipsvd_log.o ipsvd_fmt.o ipsvd_check.o ipsvd_hostname.o \
 	  ipsvd_scan.o uidgid.o cdb.a dns.a unix.a byte.a time.a socket.lib
@@ -32,21 +32,9 @@
 	  time.a `cat socket.lib`
 
 sslio: load sslio.o ssl_io.o uidgid.o sslerror_str.o unix.a byte.a time.a \
-	  matrixssl.a
+	  /usr/lib/dietlibc/lib-{ARCH}/libmatrixssl.a
 	./load sslio ssl_io.o uidgid.o sslerror_str.o unix.a byte.a time.a \
-	  matrixssl.a
-
-matrixssl.a: makelib matrixssl
-	install -m0644 Makefile.matrixssl matrixssl/src/Makefile.matrixssl
-	(cd matrixssl/src && $(MAKE) -fMakefile.matrixssl matrixssl.a)
-	install -m0644 matrixssl/src/matrixssl.a matrixssl.a
-
-matrixssl: matrixssl.tar.gz compile load
-	gzip -dc matrixssl.tar.gz |tar xf -
-	sed -e 's/#define USE_MULTITHREADING/#undef USE_MULTITHREADING/' \
-	  <matrixssl/src/matrixConfig.h >matrixConfig.h && \
-	    mv -f matrixConfig.h matrixssl/src/matrixConfig.h
-	touch matrixssl
+	  /usr/lib/dietlibc/lib-{ARCH}/libmatrixssl.a
 
 check-tcpsvd: load check-tcpsvd.o unix.a byte.a socket.lib
 	./load check-tcpsvd unix.a byte.a `cat socket.lib`
@@ -60,11 +48,11 @@
 tcpsvd.o: compile sysdeps tcpsvd.c ipsvd_log.h
 	./compile tcpsvd.c
 
-sslsvd.o: compile sysdeps tcpsvd.c ssl_io.c ipsvd_log.h matrixssl
-	./compile -I./matrixssl sslsvd.c
+sslsvd.o: compile sysdeps tcpsvd.c ssl_io.c ipsvd_log.h
+	./compile sslsvd.c
 
-ssl_io.o: compile sysdeps ssl_io.c matrixssl
-	./compile -I./matrixssl ssl_io.c
+ssl_io.o: compile sysdeps ssl_io.c
+	./compile ssl_io.c
 
 udpsvd.o: compile sysdeps udpsvd.c ipsvd_log.h
 	./compile udpsvd.c
@@ -72,8 +60,8 @@
 ipsvd-cdb.o: compile sysdeps ipsvd-cdb.c
 	./compile ipsvd-cdb.c
 
-sslio.o: compile sysdeps sslio.c matrixssl
-	./compile -I./matrixssl sslio.c
+sslio.o: compile sysdeps sslio.c
+	./compile sslio.c
 
 check-tcpsvd.o: compile check-tcpsvd.c
 	./compile check-tcpsvd.c
@@ -105,8 +93,8 @@
 uidgid.o: compile uidgid.c uidgid.h
 	./compile uidgid.c
 
-sslerror_str.o: compile sslerror_str.c sslerror_str.h matrixssl
-	./compile -I./matrixssl sslerror_str.c
+sslerror_str.o: compile sslerror_str.c sslerror_str.h
+	./compile sslerror_str.c
 
 socket.lib: compile load trysocketlib.c
 	./compile trysocketlib.c
@@ -118,7 +106,7 @@
 
 command.ssl: command.sslcheck
 command.sslcheck:
-	( (test ! -r matrixssl.tar.gz && echo nossl) || \
+	( (test ! -r /usr/lib/dietlibc/lib-{ARCH}/libmatrixssl.a && echo nossl) || \
 	  echo sslio sslsvd) >command.sslcheck
 	mv -f command.sslcheck command.ssl
 
