--- pcre-4.3/Makefile.in.linkpath	2003-05-22 21:25:53.000000000 +0800
+++ pcre-4.3/Makefile.in	2003-09-12 20:03:25.000000000 +0800
@@ -117,9 +117,9 @@
 		$(LINKLIB) -rpath $(LIBDIR) -version-info \
 		'$(PCRELIBVERSION)' -o libpcre.la $(LOBJ)
 
-libpcreposix.la: pcreposix.@OBJEXT@
+libpcreposix.la: pcreposix.@OBJEXT@ libpcre.la
 		-rm -f libpcreposix.la
-		$(LINKLIB) -rpath $(LIBDIR) -L. -lpcre -version-info \
+		$(LINKLIB) -rpath $(LIBDIR) libpcre.la -version-info \
 		'$(PCREPOSIXLIBVERSION)' -o libpcreposix.la pcreposix.lo
 
 pcre.@OBJEXT@:  $(top_srcdir)/chartables.c $(top_srcdir)/pcre.c \
