--- tk8.3.3/unix/configure.makecfg	Thu Apr  5 22:03:05 2001
+++ tk8.3.3/unix/configure	Mon Jul  9 21:21:41 2001
@@ -1768,11 +1768,11 @@
 	    CFLAGS_OPTIMIZE="${CFLAGS_OPTIMIZE} -D__NO_STRING_INLINES -D__NO_MATH_INLINES"
 
 	    if test "$have_dl" = yes; then
-		SHLIB_LD="${CC} -shared"
+		SHLIB_LD="${CC} -shared -Wl,-soname,libtk.so.0"
 		DL_OBJS="tclLoadDl.o"
 		DL_LIBS="-ldl"
 		LDFLAGS="-rdynamic"
-		LD_SEARCH_FLAGS='-Wl,-rpath,${LIB_RUNTIME_DIR}'
+		LD_SEARCH_FLAGS=''
 	    else
 		ac_safe=`echo "dld.h" | sed 'y%./+-%__p_%'`
 echo $ac_n "checking for dld.h""... $ac_c" 1>&6
--- tk8.3.3/unix/tkConfig.sh.in.makecfg	Thu Apr  5 22:03:05 2001
+++ tk8.3.3/unix/tkConfig.sh.in	Tue Jul 10 00:57:43 2001
@@ -29,7 +29,7 @@
 TK_DBGX=@TK_DBGX@
 
 # The name of the Tk library (may be either a .a file or a shared library):
-TK_LIB_FILE='@TK_LIB_FILE@'
+TK_LIB_FILE='libtk${TCL_DBGX}.so'
 
 # Additional libraries to use when linking Tk.
 TK_LIBS='@XLIBSW@ @DL_LIBS@ @LIBS@ @MATH_LIBS@'
@@ -49,15 +49,11 @@
 TK_XLIBSW='@XLIBSW@'
 
 # -l flag to pass to the linker to pick up the Tcl library
-TK_LIB_FLAG='@TK_LIB_FLAG@'
-
-# String to pass to linker to pick up the Tk library from its
-# build directory.
-TK_BUILD_LIB_SPEC='@TK_BUILD_LIB_SPEC@'
+TK_LIB_FLAG='-ltk${TK_DBGX}'
 
 # String to pass to linker to pick up the Tk library from its
 # installed directory.
-TK_LIB_SPEC='@TK_LIB_SPEC@'
+TK_LIB_SPEC='-L/usr/lib -ltk${TK_DBGX}'
 
 # Location of the top-level source directory from which Tk was built.
 # This is the directory that contains a README file as well as
@@ -67,27 +63,31 @@
 # compiled.
 TK_SRC_DIR='@TK_SRC_DIR@'
 
+# String to pass to linker to pick up the Tk library from its
+# build directory.
+TK_BUILD_LIB_SPEC='-L@TK_SRC_DIR@/unix -ltk${TK_DBGX}'
+
 # Needed if you want to make a 'fat' shared library library
 # containing tk objects or link a different wish.
 TK_CC_SEARCH_FLAGS='@TK_CC_SEARCH_FLAGS@'
 TK_LD_SEARCH_FLAGS='@TK_LD_SEARCH_FLAGS@'
 
 # The name of the Tk stub library (.a):
-TK_STUB_LIB_FILE='@TK_STUB_LIB_FILE@'
+TK_STUB_LIB_FILE='libtkstub.a'
 
 # -l flag to pass to the linker to pick up the Tk stub library
-TK_STUB_LIB_FLAG='@TK_STUB_LIB_FLAG@'
+TK_STUB_LIB_FLAG='-ltkstub'
 
 # String to pass to linker to pick up the Tk stub library from its
 # build directory.
-TK_BUILD_STUB_LIB_SPEC='@TK_BUILD_STUB_LIB_SPEC@'
+TK_BUILD_STUB_LIB_SPEC='-L@TK_SRC_DIR@/unix -ltkstub'
 
 # String to pass to linker to pick up the Tk stub library from its
 # installed directory.
-TK_STUB_LIB_SPEC='@TK_STUB_LIB_SPEC@'
+TK_STUB_LIB_SPEC='-L/usr/lib -ltkstub'
 
 # Path to the Tk stub library in the build directory.
 TK_BUILD_STUB_LIB_PATH='@TK_BUILD_STUB_LIB_PATH@'
 
 # Path to the Tk stub library in the install directory.
-TK_STUB_LIB_PATH='@TK_STUB_LIB_PATH@'
+TK_STUB_LIB_PATH='/usr/lib/libtkstub.a'
--- tk8.3.3/unix/Makefile.in.makecfg	Thu Apr  5 22:03:05 2001
+++ tk8.3.3/unix/Makefile.in	Wed Aug  8 15:58:04 2001
@@ -89,18 +89,18 @@
 TK_DBGX			= @TK_DBGX@
 
 # warning flags
-CFLAGS_WARNING		= @CFLAGS_WARNING@
+CFLAGS_WARNING		= -Wall
 
 # The default switches for optimization or debugging
 CFLAGS_DEBUG		= @CFLAGS_DEBUG@
-CFLAGS_OPTIMIZE		= @CFLAGS_OPTIMIZE@
+CFLAGS_OPTIMIZE		= -O2
 
 # To change the compiler switches, for example to change from optimization to
 # debugging symbols, change the following line:
 #CFLAGS			= $(CFLAGS_DEBUG)
 #CFLAGS			= $(CFLAGS_OPTIMIZE)
 #CFLAGS			= $(CFLAGS_DEBUG) $(CFLAGS_OPTIMIZE)
-CFLAGS			= @CFLAGS@
+CFLAGS			= -O2
 
 # A "-I" switch that can be used when compiling to make all of the
 # X11 include files accessible (the configure script will try to
@@ -173,19 +173,19 @@
 SHLIB_CFLAGS		= @SHLIB_CFLAGS@
 
 # To enable support for stubs in Tcl.
-STUB_LIB_FILE		= @STUB_LIB_FILE@
+STUB_LIB_FILE		= libtkstub.a
 
-TK_STUB_LIB_FILE	= @TK_STUB_LIB_FILE@
-#TK_STUB_LIB_FILE	= libtkstub.a
+#TK_STUB_LIB_FILE	= @TK_STUB_LIB_FILE@
+TK_STUB_LIB_FILE	= libtkstub.a
 
-TK_STUB_LIB_FLAG	= @TK_STUB_LIB_FLAG@
-#TK_STUB_LIB_FLAG	= -ltkstub
+#TK_STUB_LIB_FLAG	= @TK_STUB_LIB_FLAG@
+TK_STUB_LIB_FLAG	= -ltkstub
 
-TK_LIB_FILE		= @TK_LIB_FILE@
-#TK_LIB_FILE		= libtk.a
+#TK_LIB_FILE		= @TK_LIB_FILE@
+TK_LIB_FILE		= libtk.so.0
 
-TK_LIB_FLAG		= @TK_LIB_FLAG@
-#TK_LIB_FLAG		= -ltk
+#TK_LIB_FLAG		= @TK_LIB_FLAG@
+TK_LIB_FLAG		= -ltk
 
 #TCL_LIB_SPEC		= @TCL_BUILD_LIB_SPEC@
 TCL_LIB_SPEC		= @TCL_LIB_SPEC@
@@ -350,7 +350,7 @@
 
 all: binaries libraries doc
 
-binaries: ${TK_LIB_FILE} ${STUB_LIB_FILE} wish
+binaries: ${TK_LIB_FILE} libtk.so ${STUB_LIB_FILE} wish
 
 libraries:
 
@@ -364,9 +364,12 @@
 # a shared library or non-shared library for Tk.
 ${TK_LIB_FILE}: ${OBJS}
 	rm -f ${TK_LIB_FILE}
-	@MAKE_LIB@
+	${SHLIB_LD} -o ${TK_LIB_FILE} ${OBJS} $(TK_LD_SEARCH_FLAGS) -L$(TCL_BIN_DIR) $(TCL_STUB_LIB_FLAG) ${SHLIB_LD_LIBS} -ltcl
 	$(RANLIB) ${TK_LIB_FILE}
 
+libtk.so: ${TK_LIB_FILE}
+	ln -sf $< $@
+
 ${STUB_LIB_FILE}: ${STUB_LIB_OBJS}
 	rm -f ${STUB_LIB_FILE}
 	@MAKE_STUB_LIB@
@@ -385,7 +388,7 @@
 
 wish: $(WISH_OBJS) $(TK_LIB_FILE) $(TK_STUB_LIB_FILE)
 	$(CC) @LDFLAGS@ $(WISH_OBJS) \
-		@TK_BUILD_LIB_SPEC@ \
+		-L$(UNIX_DIR) -ltk \
 		$(WISH_LIBS) $(TK_CC_SEARCH_FLAGS) -o wish
 
 # This rule is executed if the user tried to run tktest without first
@@ -396,12 +399,12 @@
 
 tktest: ${TCL_BIN_DIR}/tcltest $(TKTEST_OBJS) $(TK_LIB_FILE)
 	${CC} @LDFLAGS@ $(TKTEST_OBJS) \
-		@TK_BUILD_LIB_SPEC@ \
+		-L$(UNIX_DIR) -ltk \
 		$(WISH_LIBS) $(TK_CC_SEARCH_FLAGS) -o tktest
 
 xttest: test.o tkTest.o tkSquare.o $(TK_LIB_FILE)
 	${CC} @LDFLAGS@ test.o tkTest.o tkSquare.o \
-		@TK_BUILD_LIB_SPEC@ \
+		-L$(UNIX_DIR) -ltk \
 		$(WISH_LIBS) $(TK_LD_SEARCH_FLAGS) -lXt -o xttest
 
 # Note, in the target below TCL_LIBRARY needs to be set or else
@@ -456,10 +459,11 @@
 	    fi
 	@echo "Installing $(TK_LIB_FILE) to $(LIB_INSTALL_DIR)/"
 	@$(INSTALL_DATA) $(TK_LIB_FILE) $(LIB_INSTALL_DIR)/$(TK_LIB_FILE)
+	@cp -d libtk.so $(LIB_INSTALL_DIR)/libtk.so
 	@(cd $(LIB_INSTALL_DIR); $(RANLIB) $(TK_LIB_FILE))
 	@chmod 555 $(LIB_INSTALL_DIR)/$(TK_LIB_FILE)
-	@echo "Installing wish as $(BIN_INSTALL_DIR)/wish$(VERSION)"
-	@$(INSTALL_PROGRAM) wish $(BIN_INSTALL_DIR)/wish$(VERSION)
+	@echo "Installing wish as $(BIN_INSTALL_DIR)/wish"
+	@$(INSTALL_PROGRAM) wish $(BIN_INSTALL_DIR)/wish
 	@echo "Installing tkConfig.sh to $(LIB_INSTALL_DIR)/"
 	@$(INSTALL_DATA) tkConfig.sh $(LIB_INSTALL_DIR)/tkConfig.sh
 	@if test "$(TK_BUILD_EXP_FILE)" != ""; then \
