--- pam_ldap-164/pam_ldap.c.duplicate_definition	2003-08-05 15:27:36.000000000 +0200
+++ pam_ldap-164/pam_ldap.c	2003-08-05 15:28:35.000000000 +0200
@@ -234,7 +234,7 @@
 static int _open_session (pam_ldap_session_t *);
 static int _connect_anonymously (pam_ldap_session_t *);
 #if defined(LDAP_API_FEATURE_X_OPENLDAP) && (LDAP_API_VERSION > 2000)
-static int _rebind_proc (LDAP * ld, LDAP_CONST char *url, int request, ber_int_t msgid);
+/*static int _rebind_proc (LDAP * ld, LDAP_CONST char *url, int request, ber_int_t msgid);*/
 #else
 #if LDAP_SET_REBIND_PROC_ARGS == 3
 static int _rebind_proc (LDAP *, char **, char **, int *, int, void *);
