--- glibc-2.5.orig/nscd/nscd.h
+++ glibc-2.5/nscd/nscd.h
@@ -89,9 +89,9 @@ struct database_dyn
 
 
 /* Paths of the file for the persistent storage.  */
-#define _PATH_NSCD_PASSWD_DB	"/var/db/nscd/passwd"
-#define _PATH_NSCD_GROUP_DB	"/var/db/nscd/group"
-#define _PATH_NSCD_HOSTS_DB	"/var/db/nscd/hosts"
+#define _PATH_NSCD_PASSWD_DB	"/var/lib/nscd/passwd"
+#define _PATH_NSCD_GROUP_DB	"/var/lib/nscd/group"
+#define _PATH_NSCD_HOSTS_DB	"/var/lib/nscd/hosts"
 
 /* Path used when not using persistent storage.  */
 #define _PATH_NSCD_XYZ_DB_TMP	"/var/run/nscd/dbXXXXXX"
