--- coreutils-6.9/src/dircolors.hin.colors_mdkconf	2007-03-18 22:36:43.000000000 +0100
+++ coreutils-6.9/src/dircolors.hin	2007-09-01 18:56:07.000000000 +0200
@@ -9,6 +9,10 @@
 # The keywords COLOR, OPTIONS, and EIGHTBIT (honored by the
 # slackware version of dircolors) are recognized but ignored.
 
+# This file goes in the /etc directory, and must be world readable.
+# You can copy this file to .dir_colors in your $HOME directory to override
+# the system defaults.
+
 # Below, there should be one TERM entry for each termtype that is colorizable
 TERM Eterm
 TERM ansi
@@ -67,7 +71,8 @@ SOCK 01;35	# socket
 DOOR 01;35	# door
 BLK 40;33;01	# block device driver
 CHR 40;33;01	# character device driver
-ORPHAN 40;31;01 # symlink to nonexistent file, or non-stat'able file
+ORPHAN 01;05;37;41 # symlink to nonexistent file, or non-stat'able file
+MISSING 01;05;37;41 # ... and the files they point to
 SETUID 37;41	# file that is setuid (u+s)
 SETGID 30;43	# file that is setgid (g+s)
 STICKY_OTHER_WRITABLE 30;42 # dir that is sticky and other-writable (+t,o+w)
@@ -82,15 +87,15 @@ EXEC 01;32
 # (and any comments you want to add after a '#')
 
 # If you use DOS-style suffixes, you may want to uncomment the following:
-#.cmd 01;32 # executables (bright green)
-#.exe 01;32
-#.com 01;32
-#.btm 01;32
-#.bat 01;32
+.cmd 01;32 # executables (bright green)
+.exe 01;32
+.com 01;32
+.btm 01;32
+.bat 01;32
 # Or if you want to colorize scripts even if they do not have the
 # executable bit actually set.
-#.sh  01;32
-#.csh 01;32
+.sh  01;32
+.csh 01;32
 
  # archives or compressed (bright red)
 .tar 01;31
@@ -106,8 +111,11 @@ EXEC 01;32
 .bz  01;31
 .tbz2 01;31
 .tz  01;31
+.lzma 01;31
+.tlz 01;31
 .deb 01;31
 .rpm 01;31
+.cpio 01;31
 .jar 01;31
 .rar 01;31
 .ace 01;31
@@ -115,10 +123,13 @@ EXEC 01;32
 .cpio 01;31
 .7z  01;31
 .rz  01;31
+.iso 01;31
 
-# image formats
+# image and video formats
 .jpg 01;35
 .jpeg 01;35
+.JPG 01;35
+.JPEG 01;35
 .gif 01;35
 .bmp 01;35
 .pbm 01;35
@@ -156,6 +167,14 @@ EXEC 01;32
 .xcf 01;35
 .xwd 01;35
 .yuv 01;35
+.divx 01;35
+.xvid 01;35
+.asf 01;35
+.wmv 01;35
+.mp4 01;35
+.3gp 01;35
+.flv 01;35
+.ico 01;35
 
 # audio formats
 .aac 00;36
@@ -169,3 +188,73 @@ EXEC 01;32
 .ogg 00;36
 .ra 00;36
 .wav 00;36
+.mp2 00;36
+.mod 00;36
+.xm 00;36
+.s3m 00;36
+.it 00;36
+.wma 00;36
+
+# document and various text(ish) formats
+#.txt 01;37;40
+#.lyx 01;37;40
+#.doc 01;37;40
+#.wri 01;37;40
+#.odt 01;37;40
+#.wpd 01;37;40
+#.abw 01;37;40
+#.rtf 01;37;40
+#.sam 01;37;40
+#.pdb 01;37;40
+#.sxw 01;37;40
+#.kwd 01;37;40
+#.sgml 01;37;40
+#.html 01;37;40
+#.htm 01;37;40
+#.wml 01;37;40
+#.xml 01;37;40
+#.pdf 01;37;40
+#.ps 01;37;40
+#.man 01;37;40
+#.info 01;37;40
+#.tex 01;37;40
+#.conf 01;37;40
+#.cnf 01;37;40
+#.cfg 01;37;40
+#.config 01;37;40
+#.ini 01;37;40
+#.spec 01;37;40
+#.log 01;37;40
+#.c 01;37;40
+#.cpp 01;37;40
+#.cxx 01;37;40
+#.h 01;37;40
+#.hpp 01;37;40
+#.java 01;37;40
+#.pl 01;37;40
+#.py 01;37;40
+#.patch 01;37;40
+#*rc 01;37;40
+#*README 01;37;40
+#*COPYING 01;37;40
+#*ChangeLog 01;37;40
+#*NEWS 01;37;40
+#*AUTHORS 01;37;40
+#*ABOUT-NLS 01;37;40
+#*THANKS 01;37;40
+#*TODO 01;37;40
+#*FAQ 01;37;40
+#*INSTALL 01;37;40
+#*CHANGES 01;37;40
+
+# backup formats
+*~ 47;30
+.bak 47;30
+.swp 47;30
+.bck 47;30
+.bk 47;30
+.old 47;30
+.tmp 47;30
+.save 47;30
+.rpmsave 47;30
+.rpmnew 47;30
