--- coreutils-5.97/src/cp.c.override	2005-09-16 09:50:33.000000000 +0200
+++ coreutils-5.97/src/cp.c	2007-01-31 21:42:55.000000000 +0100
@@ -848,6 +848,7 @@ main (int argc, char **argv)
 	  break;
 
 	case 'f':
+	  x.interactive = I_ALWAYS_YES;
 	  x.unlink_dest_after_failed_open = true;
 	  break;
 
--- coreutils-5.97/tests/mv/i-2.override	2001-09-15 13:18:40.000000000 +0200
+++ coreutils-5.97/tests/mv/i-2	2007-01-31 22:41:15.000000000 +0100
@@ -42,11 +42,12 @@ mv -fi c d < y >/dev/null 2>&1 || fail=1
 # Before 4.0s, this would not prompt.
 cp -if e f < y > out 2>&1 || fail=1
 
+#(peroyvind) this test is bogus, we don't want to be prompted when forcing
 # Make sure out contains the prompt.
-case "`cat out`" in
-  "cp: overwrite \`f', overriding mode 0000?"*) ;;
-  *) fail=1 ;;
-esac
+#case "`cat out`" in
+#  "cp: overwrite \`f', overriding mode 0000?"*) ;;
+#  *) fail=1 ;;
+#esac
 
 test -f e || fail=1
 test -f f || fail=1
