--- gzip-1.2.4a/gzip.c.can-2005-1228	2005-05-18 14:39:55.734861260 -0600
+++ gzip-1.2.4a/gzip.c	2005-05-18 14:40:28.984881434 -0600
@@ -1269,6 +1269,8 @@
 			error("corrupted input -- file name too large");
 		    }
 		}
+		char *base2 = basename (base);   /* strip any paths that may exists */
+		strcpy(base, base2);              /* in the output name */
                 /* If necessary, adapt the name to local OS conventions: */
                 if (!list) {
                    MAKE_LEGAL_NAME(base);
