diff -Naur nfs-utils-1.0.7/utils/gssd/Makefile nfs-utils-1.0.7.oden/utils/gssd/Makefile
--- nfs-utils-1.0.7/utils/gssd/Makefile	2004-10-19 02:23:08.000000000 +0200
+++ nfs-utils-1.0.7.oden/utils/gssd/Makefile	2005-01-09 12:39:37.536662377 +0100
@@ -7,7 +7,7 @@
 OBJS	= gssd.o gssd_main_loop.o gssd_proc.o err_util.o gss_util.o \
 	  gss_oids.o context.o context_heimdal.o krb5_util.o
 LIBDEPS	= $(TOP)support/lib/librpc.a $(TOP)support/lib/libgssapi.a
-LIBS	= -Wl,-rpath=$(KRBDIR)/lib -lrpc -lgssapi -ldl $(KRBLIB)
+LIBS	= -lrpc -lgssapi -ldl $(KRBLIB)
 MAN8	= gssd
 
 include $(TOP)rules.mk
diff -Naur nfs-utils-1.0.7/utils/svcgssd/Makefile nfs-utils-1.0.7.oden/utils/svcgssd/Makefile
--- nfs-utils-1.0.7/utils/svcgssd/Makefile	2004-11-22 06:41:13.000000000 +0100
+++ nfs-utils-1.0.7.oden/utils/svcgssd/Makefile	2005-01-09 12:40:58.133490943 +0100
@@ -7,7 +7,7 @@
 OBJS	= svcgssd.o svcgssd_main_loop.o svcgssd_proc.o err_util.o gss_util.o \
 	  gss_oids.o context.o context_heimdal.o cacheio.o svcgssd_mech2file.o
 LIBDEPS	= $(TOP)support/lib/librpc.a $(TOP)support/lib/libgssapi.a
-LIBS	= -Wl,-rpath=$(KRBDIR)/lib -lrpc -lgssapi -ldl $(KRBLIB) -lnfsidmap
+LIBS	= -lrpc -lgssapi -ldl $(KRBLIB) -lnfsidmap
 MAN8	= svcgssd
 
 predep ::
