--- sysvinit-2.86/src/init.c.varargs	2005-08-23 08:09:20.000000000 -0400
+++ sysvinit-2.86/src/init.c	2005-08-23 11:15:30.000000000 -0400
@@ -1884,7 +1884,7 @@ void re_exec(void)
 	 *	The existing init process execs a new init binary.
 	 */
 	env = init_buildenv(0);
-	execl(myname, myname, "--init", NULL, env);
+	execle(myname, myname, "--init", NULL, env);
 
 	/*
 	 *	We shouldn't be here, something failed. 
