--- config.c	2005-11-12 20:07:56.000000000 +0100
+++ config.c.oden	2005-12-29 17:55:13.000000000 +0100
@@ -1011,12 +1011,7 @@
 		rc = glob(argv[argNum], GLOB_NOCHECK, globerr,
 			  &globResult);
 		if (rc == GLOB_ABORTED) {
-		    if (newlog->flags & LOG_FLAG_MISSINGOK)
 			continue;
-
-		    message(MESS_ERROR, "%s:%d glob failed for %s\n",
-			    configFile, lineNum, argv[argNum]);
-		    return 1;
 		}
 
 		newlog->files =
