--- ./build/pack.c.buildsubdir	2005-08-25 17:17:30.000000000 +0200
+++ ./build/pack.c	2005-08-25 17:17:39.000000000 +0200
@@ -138,7 +138,8 @@
     FILE * f;
     FD_t fd;
 
-    fn = rpmGetPath("%{_builddir}/%{?_buildsubdir:%{_buildsubdir}/}", file, NULL);
+    /* XXX use rpmGenPath(rootdir, "%{_buildir}/%{_buildsubdir}/", file) */
+    fn = rpmGetPath("%{_builddir}/", spec->buildSubdir, "/", file, NULL);
 
     fd = Fopen(fn, "r.ufdio");
     if (fn != buf) fn = _free(fn);
