--- ipsvd-0.10.1/src/Makefile.oden	2005-01-16 11:34:35.000000000 +0100
+++ ipsvd-0.10.1/src/Makefile	2005-01-18 12:03:20.365791969 +0100
@@ -24,24 +24,9 @@
 	./load ipsvd-cdb ipsvd_fmt.o ipsvd_check.o cdb.a dns.a unix.a byte.a \
 	  time.a `cat socket.lib`
 
-sslio: load sslio.o uidgid.o sslerror_str.o unix.a byte.a time.a matrixssl.a
-	./load sslio uidgid.o sslerror_str.o unix.a byte.a time.a matrixssl.a
-
-matrixssl.a: makelib matrixssl
-	install -m0644 Makefile.matrixssl matrixssl/src/Makefile.matrixssl
-	( cd matrixssl/src && $(MAKE) -fMakefile.matrixssl matrixssl.a )
-	install -m0644 matrixssl/src/matrixssl.a matrixssl.a
-
-matrixSsl.h: matrixssl
-	rm -f matrixSsl.h
-	ln -s matrixssl/matrixSsl.h matrixSsl.h
-
-matrixssl: matrixssl.tar.gz compile load
-	gzip -dc matrixssl.tar.gz |tar xf -
-	sed -e 's/#define USE_MULTITHREADING/#undef USE_MULTITHREADING/' \
-	  <matrixssl/src/matrixConfig.h >matrixConfig.h && \
-	    mv -f matrixConfig.h matrixssl/src/matrixConfig.h
-	touch matrixssl
+sslio: load sslio.o uidgid.o sslerror_str.o unix.a byte.a time.a /usr/lib/dietlibc/lib/libmatrixssl.a
+	cp /usr/lib/dietlibc/include/matrixSsl.h .
+	./load sslio uidgid.o sslerror_str.o unix.a byte.a time.a /usr/lib/dietlibc/lib/libmatrixssl.a
 
 check-tcpsvd: load check-tcpsvd.o unix.a byte.a socket.lib
 	./load check-tcpsvd unix.a byte.a `cat socket.lib`
@@ -61,7 +46,7 @@
 ipsvd-cdb.o: compile sysdeps ipsvd-cdb.c
 	./compile ipsvd-cdb.c
 
-sslio.o: compile sysdeps matrixSsl.h sslio.c
+sslio.o: compile sysdeps sslio.c
 	./compile sslio.c
 
 check-tcpsvd.o: compile check-tcpsvd.c
@@ -94,7 +79,7 @@
 uidgid.o: compile uidgid.c uidgid.h
 	./compile uidgid.c
 
-sslerror_str.o: compile matrixSsl.h sslerror_str.c sslerror_str.h 
+sslerror_str.o: compile sslerror_str.c sslerror_str.h 
 	./compile sslerror_str.c
 
 socket.lib: compile load trysocketlib.c
@@ -108,7 +93,6 @@
 
 command.linux: sysdeps
 	( ( test "`cut -c1-5 <systype`" != 'linux' && echo nolinux ) || \
-	  ( test ! -r matrixssl.tar.gz && echo nomatrixssl ) || \
 	  echo sslio ) >command.linux
 
 nolinux:
