Index: lesspipe.sh.in
==================================================================
--- lesspipe-1.42/lesspipe.sh.in
+++ lesspipe-1.42/lesspipe.sh.in	2004-05-06 11:10:03.968283846 +0200
@@ -396,7 +396,13 @@
 #ifdef html2text
   elif [[ "$1" = *HTML* ]]; then
     echo "==> append $sep to filename to view the HTML source"
-    html2text -style pretty "$2"
+    if [ -x /usr/bin/lynx ]; then
+        /usr/bin/lynx -nobrowse -dump "$2"
+    elif [ -x /usr/bin/html2text ]; then
+        /usr/bin/html2text -style pretty "$2"
+    else
+        cat "$2"
+    fi
 #elif lynx
   elif [[ "$1" = *HTML* ]]; then
     echo "==> append $sep to filename to view the HTML source"
