--- vim60af/src/Makefile.manpath	Sun May 13 21:05:49 2001
+++ vim60af/src/Makefile	Mon May 14 15:49:51 2001
@@ -785,9 +785,9 @@
 # Uncomment the next line to install Vim in "/usr/bin"
 #BINDIR   = /usr/bin
 # Uncomment the next line to install Vim manuals in "/usr/share/man/man1"
-#MANDIR   = /usr/share/man
+MANDIR   = ${prefix}/share/man
 # Uncomment the next line to install Vim help files in "/usr/share/vim"
-#DATADIR  = /usr/share
+DATADIR  = ${prefix}/share
 
 ### DESTDIR	root of the installation tree.  This is prepended to the other
 #		directories.  This directory must exist.
@@ -1666,40 +1666,40 @@
 installvimdiff: $(DEST_BIN)/$(VIMDIFFTARGET) $(DEST_BIN)/$(GVIMDIFFTARGET)
 
 $(DEST_BIN)/$(EXTARGET):
-	cd $(DEST_BIN); ln -s $(VIMTARGET) $(EXTARGET)
+	cd $(DEST_BIN); ln -sf $(VIMTARGET) $(EXTARGET)
 
 $(DEST_BIN)/$(VIEWTARGET):
-	cd $(DEST_BIN); ln -s $(VIMTARGET) $(VIEWTARGET)
+	cd $(DEST_BIN); ln -sf $(VIMTARGET) $(VIEWTARGET)
 
 $(DEST_BIN)/$(GVIMTARGET):
-	cd $(DEST_BIN); ln -s $(VIMTARGET) $(GVIMTARGET)
+	cd $(DEST_BIN); ln -sf $(VIMTARGET) $(GVIMTARGET)
 
 $(DEST_BIN)/$(GVIEWTARGET):
-	cd $(DEST_BIN); ln -s $(VIMTARGET) $(GVIEWTARGET)
+	cd $(DEST_BIN); ln -sf $(VIMTARGET) $(GVIEWTARGET)
 
 $(DEST_BIN)/$(RVIMTARGET):
-	cd $(DEST_BIN); ln -s $(VIMTARGET) $(RVIMTARGET)
+	cd $(DEST_BIN); ln -sf $(VIMTARGET) $(RVIMTARGET)
 
 $(DEST_BIN)/$(RVIEWTARGET):
-	cd $(DEST_BIN); ln -s $(VIMTARGET) $(RVIEWTARGET)
+	cd $(DEST_BIN); ln -sf $(VIMTARGET) $(RVIEWTARGET)
 
 $(DEST_BIN)/$(RGVIMTARGET):
-	cd $(DEST_BIN); ln -s $(VIMTARGET) $(RGVIMTARGET)
+	cd $(DEST_BIN); ln -sf $(VIMTARGET) $(RGVIMTARGET)
 
 $(DEST_BIN)/$(RGVIEWTARGET):
-	cd $(DEST_BIN); ln -s $(VIMTARGET) $(RGVIEWTARGET)
+	cd $(DEST_BIN); ln -sf $(VIMTARGET) $(RGVIEWTARGET)
 
 $(DEST_BIN)/$(VIMDIFFTARGET):
-	cd $(DEST_BIN); ln -s $(VIMTARGET) $(VIMDIFFTARGET)
+	cd $(DEST_BIN); ln -sf $(VIMTARGET) $(VIMDIFFTARGET)
 
 $(DEST_BIN)/$(GVIMDIFFTARGET):
-	cd $(DEST_BIN); ln -s $(VIMTARGET) $(GVIMDIFFTARGET)
+	cd $(DEST_BIN); ln -sf $(VIMTARGET) $(GVIMDIFFTARGET)
 
 $(DEST_BIN)/$(EVIMTARGET):
-	cd $(DEST_BIN); ln -s $(VIMTARGET) $(EVIMTARGET)
+	cd $(DEST_BIN); ln -sf $(VIMTARGET) $(EVIMTARGET)
 
 $(DEST_BIN)/$(EVIEWTARGET):
-	cd $(DEST_BIN); ln -s $(VIMTARGET) $(EVIEWTARGET)
+	cd $(DEST_BIN); ln -sf $(VIMTARGET) $(EVIEWTARGET)
 
 # create links for the manual pages with various names to vim.	This is only
 # done when the links (or manpages with the same name) don't exist yet.
