--- clamav-0.90.2/shared/output.c.stderr	2007-03-01 17:43:18.000000000 -0700
+++ clamav-0.90.2/shared/output.c	2007-05-07 09:46:04.133279250 -0600
@@ -150,7 +150,11 @@
     if(logg_file) {
 	if(!logg_fd) {
 	    old_umask = umask(0037);
-	    if((logg_fd = fopen(logg_file, "at")) == NULL) {
+	    if (strcmp(logg_file,"/dev/stderr") == 0) {
+	        logg_fd = stderr;
+	        umask(old_umask);
+            }
+	    else if((logg_fd = fopen(logg_file, "at")) == NULL) {
 		umask(old_umask);
 #ifdef CL_THREAD_SAFE
 		pthread_mutex_unlock(&logg_mutex);
