--- ./mount/mount.c.encrypted	2002-11-06 09:02:48.000000000 +0100
+++ ./mount/mount.c	2002-11-06 09:03:11.000000000 +0100
@@ -125,10 +125,11 @@
 #define MS_USER		0x20000000
 #define MS_OWNER	0x10000000
 #define MS_NETDEV	0x00020000
+#define MS_ENCRYPT	0x00030000
 #define MS_LOOP		0x00010000
 
 /* Options that we keep the mount system call from seeing.  */
-#define MS_NOSYS	(MS_NOAUTO|MS_USERS|MS_USER|MS_NETDEV|MS_LOOP)
+#define MS_NOSYS	(MS_NOAUTO|MS_USERS|MS_USER|MS_NETDEV|MS_ENCRYPT|MS_LOOP)
 
 /* Options that we keep from appearing in the options field in the mtab.  */
 #define MS_NOMTAB	(MS_REMOUNT|MS_NOAUTO|MS_USERS|MS_USER)
@@ -162,6 +163,7 @@
   { "nouser",	0, 1, MS_USER	},	/* Forbid ordinary user to mount */
   { "owner",	0, 0, MS_OWNER  },	/* Let the owner of the device mount */
   { "noowner",	0, 1, MS_OWNER  },	/* Device owner has no special privs */
+  { "encrypted", 0, 0, MS_ENCRYPT },	/* Device encrypted */
   { "_netdev",	0, 0, MS_NETDEV },	/* Device accessible only via network */
   /* add new options here */
 #ifdef MS_NOSUB
