--- util-linux-2.12/login-utils/chfn.c.pamsec	2004-04-22 08:42:00.000000000 -0600
+++ util-linux-2.12/login-utils/chfn.c	2004-04-22 08:43:00.000000000 -0600
@@ -157,9 +157,9 @@
 	    exit(1);
 	}
         retcode = pam_acct_mgmt(pamh, 0);
-        if (retcode == PAM_NEW_AUTHTOK_REQD) {
+        if (retcode == PAM_NEW_AUTHTOK_REQD)
 	    retcode = pam_chauthtok(pamh, PAM_CHANGE_EXPIRED_AUTHTOK);
-        } else if (retcode) {
+        if (retcode) {
 	    puts(_("Password error."));
 	    exit(1);
 	}
--- util-linux-2.12/login-utils/chsh.c.pamsec	2004-04-22 08:43:23.000000000 -0600
+++ util-linux-2.12/login-utils/chsh.c	2004-04-22 08:43:52.000000000 -0600
@@ -148,9 +148,9 @@
 	    exit(1);
 	}
         retcode = pam_acct_mgmt(pamh, 0);
-        if (retcode == PAM_NEW_AUTHTOK_REQD) {
+        if (retcode == PAM_NEW_AUTHTOK_REQD)
 	    retcode = pam_chauthtok(pamh, PAM_CHANGE_EXPIRED_AUTHTOK);
-        } else if (retcode) {
+        if (retcode) {
 	    puts(_("Password error."));
 	    exit(1);
 	}
