diff -ruN util-linux-2.11t/mount.orig/swapon.c util-linux-2.11t/mount/swapon.c
--- util-linux-2.12/mount/swapon.c.swapon-encrypted	2004-01-30 12:11:43.482690622 +0100
+++ util-linux-2.12/mount/swapon.c	2004-01-30 12:14:37.391171752 +0100
@@ -593,6 +593,7 @@
 		if (streq (program_name, "swapon")) {
 		while ((fstab = getmntent(fp)) != NULL) {
 		        int i;
+			int encrypted = 0;
 			char *special = fstab->mnt_fsname;
 
 			if (streq(fstab->mnt_type, MNTTYPE_SWAP)) {
@@ -607,7 +608,12 @@
 						loop = opt + 5;
 					if (strncmp(opt, "encryption=", 11) == 0)
 						encryption = opt + 11;
+					if (strncmp(opt, "encrypted", 9) == 0)
+						encrypted=1;
 				}
+				if (encrypted) 
+					continue; 
+
 				if (loop && encryption) {
 					if(!is_in_proc_swaps(loop) && (!ifexists || !access(special, R_OK))) {
 						if (!prepare_encrypted_swap(special, loop, encryption)) {
