--- chkauth~	2004-03-09 20:06:14.000000000 -0700
+++ chkauth	2004-03-09 20:06:14.000000000 -0700
@@ -85,7 +85,7 @@
  
 auth       required       pam_nologin.so
 auth       required       pam_env.so
-auth       sufficient     pam_pwdb.so likeauth nullok
+auth       sufficient     pam_unix.so likeauth nullok
 auth       sufficient     pam_ldap.so use_first_pass
 auth       required       pam_deny.so
  
