--- gcc-3.3/gcc/ada/Makefile.in.ada-makefile	2003-02-02 23:57:52.000000000 +0100
+++ gcc-3.3/gcc/ada/Makefile.in	2003-02-03 09:12:09.000000000 +0100
@@ -143,7 +143,7 @@ MISCLIB =
 objext = .o
 exeext =
 arext  = .a
-soext  = .so
+soext  = .so.1
 shext  =
 
 HOST_CC=$(CC)
@@ -658,7 +658,7 @@ ifeq ($(strip $(filter-out %86 linux%,$(
   system.ads<5lsystem.ads
 
   TOOLS_TARGET_PAIRS = mlib-tgt.adb<5lml-tgt.adb
-  SYMLIB = -laddr2line -lbfd $(INTLLIBS)
+  SYMLIB =  $(INTLLIBS)
   THREADSLIB = -lpthread
   GNATLIB_SHARED = gnatlib-shared-dual
   GMEM_LIB = gmemlib
@@ -1822,7 +1822,7 @@ gnatlib: ../stamp-gnatlib1 ../stamp-gnat
 	$(MAKE) -C rts CC="../../xgcc -B../../" \
 	        INCLUDES="$(INCLUDES_FOR_SUBDIR) -I./../.." \
                 CFLAGS="$(GNATLIBCFLAGS_FOR_C)" \
-		srcdir=$(fsrcdir) \
+		srcdir=$(fsrcdir) VPATH=$(fsrcdir) \
 	        -f ../Makefile $(LIBGNAT_OBJS) prefix.o
 	$(MAKE) -C rts CC="../../xgcc -B../../" \
 	        ADA_INCLUDES="$(ADA_INCLUDES_FOR_SUBDIR)" \
