--- xc/programs/xkbcomp/symbols/inet.orig	Thu Jun  6 10:34:43 2002
+++ xc/programs/xkbcomp/symbols/inet	Thu Jun  6 10:39:55 2002
@@ -14,6 +14,114 @@
     key <I20>	{	[ XF86AudioMute		]	};
 };
 
+xkb_symbols "hp_multimedia" {
+
+    // Describes the extra keys on an HP "Internet" keyboard.
+
+    name[Group1]= "HPMultimedia";
+    key <I1E>  {       [ XF86Mail              ]       };
+    key <I12>  {       [ XF86Search            ]       };
+    key <I32>   {      [ XF86WWW               ]       };
+    key <I17>  {       [ Help                  ]       };
+    key <I25>   {      [ XF86ScreenSaver       ]       };
+    key <I23>   {      [ XF86Launch2           ]       };
+    key <I22>  {       [ XF86Launch3           ]       };
+    key <I24>  {       [ XF86Launch4           ]       };
+    key <I21>  {       [ XF86Launch5           ]       };
+    key <I26>  {       [ XF86Launch6           ]       };
+    key <I20>  {       [ XF86AudioMute         ]       };
+    key <I2E>  {       [ XF86AudioLowerVolume  ]       };
+    key <I30>  {       [ XF86AudioRaiseVolume  ]       };
+};
+
+partial alphanumeric_keys 
+xkb_symbols "hp_multimedia2" {
+
+    // Describes the extra keys on an HP "Internet" keyboard.
+
+    name[Group1]= "HPMultimedia2";
+    key <I2D>  {       [ XF86Mail              ]       };
+    key <I13>   {      [ XF86WWW               ]       };
+    key <I26>  {       [ XF86HomePage          ]       };
+    key <I18>   {      [ XF86Launch1           ]       };
+    key <I15>   {      [ XF86Launch2           ]       };
+    key <I31>  {       [ XF86Launch3           ]       };
+};
+
+partial alphanumeric_keys
+xkb_symbols "xe3gc" {
+
+    // Describes the OneTouch buttons on HP Omnibook XE3 GC models.
+    // See http://sourceforge.net/projects/omke for details on enabling
+    // these keys
+
+    name[Group1]= "XE3GC";
+    key <I73>  {       [ XF86WWW               ]       };
+    key <I74>  {       [ XF86Mail              ]       };
+    key <I72>  {       [ XF86Launch1           ]       };
+    key <I71>  {       [ Help                  ]       };
+    key <I10>  {       [ XF86AudioPrev         ]       };
+    key <I22>  {       [ XF86AudioPlay, XF86AudioPause ]       };
+    key <I24>  {       [ XF86AudioStop, XF86Eject      ]       };
+    key <I19>  {       [ XF86AudioNext         ]       };
+};
+
+partial alphanumeric_keys
+xkb_symbols "xe3gf" {
+
+    // Describes the OneTouch buttons on HP Omnibook XE3 GF models.
+    // See http://sourceforge.net/projects/omke for details on enabling
+    // these keys
+
+    name[Group1]= "XE3GF";
+    key <I32>  {       [ XF86WWW               ]       };
+    key <I6C>  {       [ XF86Mail              ]       };
+    key <I74>  {       [ XF86Launch1           ]       };
+    key <I73>  {       [ Help                  ]       };
+    key <I10>  {       [ XF86AudioPrev         ]       };
+    key <I22>  {       [ XF86AudioPlay, XF86AudioPause ]       };
+    key <I24>  {       [ XF86AudioStop, XF86Eject      ]       };
+    key <I19>  {       [ XF86AudioNext         ]       };
+    key <I2E>  {       [ XF86AudioLowerVolume  ]       };
+    key <I30>  {       [ XF86AudioRaiseVolume  ]       };
+    key <I20>  {       [ XF86AudioMute         ]       };
+};
+
+partial alphanumeric_keys
+xkb_symbols "xt1000" {
+
+    // Describes the OneTouch buttons on HP Omnibook XT1000 models.
+    // See http://sourceforge.net/projects/omke for details on enabling
+    // these keys
+
+    name[Group1]= "XT1000";
+    key <I6C>  {       [ XF86Mail              ]       };
+    key <I74>  {       [ XF86Launch1           ]       };
+    key <I32>  {       [ XF86WWW               ]       };
+    key <I73>  {       [ XF86Launch2           ]       };
+    key <I72>  {       [ Help                  ]       };
+    key <I71>  {       [ XF86Launch3           ]       };
+    key <I10>  {       [ XF86AudioPrev         ]       };
+    key <I19>  {       [ XF86AudioNext         ]       };
+    key <I22>  {       [ XF86AudioPlay, XF86AudioPause ]       };
+    key <I24>  {       [ XF86AudioStop, XF86Eject      ]       };
+    key <I2E>  {       [ XF86AudioLowerVolume  ]       };
+    key <I30>  {       [ XF86AudioRaiseVolume  ]       };
+    key <I20>  {       [ XF86AudioMute         ]       };
+};
+
+partial alphanumeric_keys
+xkb_symbols "omnibook500fa" {
+
+    // Describes the OneTouch buttons on HP Omnibook 500 FA models.
+    // See http://sourceforge.net/projects/omke for details on enabling
+    // these keys
+
+    name[Group1]= "OmniBook500FA";
+    key <I74>  {       [ XF86Launch1           ]       };
+    key <I73>  {       [ Help                  ]       };
+};
+
 partial alphanumeric_keys 
 xkb_symbols "compaq" {
 
