--- xorg-x11-6.8.2/xc/programs/Xserver/hw/xfree86/drivers/nv/nv_driver.c.nv-ids	2004-12-16 02:42:35.000000000 +0100
+++ xorg-x11-6.8.2/xc/programs/Xserver/hw/xfree86/drivers/nv/nv_driver.c	2005-03-08 11:15:50.504668670 +0100
@@ -216,14 +216,16 @@
   { 0x10DE0045, "GeForce 6800 GT" },
   { 0x10DE0049, "0x0049" },
   { 0x10DE004E, "Quadro FX 4000" },
-  { 0x10DE004D, "Quadro FX 4400" },
+
   { 0x10DE00C0, "0x00C0" },
-  { 0x10DE00C1, "0x00C1" },
+  { 0x10DE00C1, "GeForce 6800" },
   { 0x10DE00C2, "GeForce 6800 LE" },
-  { 0x10DE00C8, "0x00C8" },
-  { 0x10DE00C9, "0x00C9" },
-  { 0x10DE00CC, "0x00CC" },
-  { 0x10DE00CE, "0x00CE" },
+  { 0x10DE00C8, "GeForce Go 6800" },
+  { 0x10DE00C9, "GeForce Go 6800 Ultra" },
+  { 0x10DE00CC, "Quadro FX Go1400" },
+  { 0x10DE00CD, "0x00CD" },
+  { 0x10DE00CE, "Quadro FX 1400" },
+
   { 0x10DE0140, "GeForce 6600 GT" },
   { 0x10DE0141, "GeForce 6600" },
   { 0x10DE0142, "0x0142" },
@@ -239,12 +241,32 @@
   { 0x10DE014D, "0x014D" },
   { 0x10DE014E, "Quadro FX 540" },
   { 0x10DE014F, "GeForce 6200" },
+
   { 0x10DE0160, "0x0160" },
-  { 0x10DE0166, "0x0166" },
+  { 0x10DE0161, "GeForce 6200 TurboCache(TM)" },
+  { 0x10DE0162, "0x0162" },
+  { 0x10DE0163, "0x0163" },
+  { 0x10DE0164, "GeForce Go 6200" },
+  { 0x10DE0165, "0x0163" },
+  { 0x10DE0166, "GeForce Go 6250" },
+  { 0x10DE0167, "GeForce Go 6200" },
+  { 0x10DE0168, "GeForce Go 6250" },
+  { 0x10DE0169, "0x0169" },
+  { 0x10DE016B, "0x016B" },
+  { 0x10DE016C, "0x016C" },
+  { 0x10DE016D, "0x016D" },
+  { 0x10DE016E, "0x016E" },
+
   { 0x10DE0210, "0x0210" },
-  { 0x10DE0211, "0x0211" },
-  { 0x10DE021D, "0x021D" },
-  { 0x10DE021E, "0x021E" },
+  { 0x10DE0211, "GeForce 6800" },
+  { 0x10DE0212, "GeForce 6800 LE" },
+  { 0x10DE0215, "GeForce 6800 GT" },
+
+  { 0x10DE0220, "0x0220" },
+  { 0x10DE0221, "0x0221" },
+  { 0x10DE0222, "0x0222" },
+  { 0x10DE0228, "0x0228" },
+
   {-1, NULL}
 };
 
@@ -640,6 +662,8 @@
                case 0x01D0:
                case 0x0090:
                case 0x0210:
+               case 0x0220:
+               case 0x0230:
                    NVChipsets[numUsed].token = pciid;
                    NVChipsets[numUsed].name = "Unknown NVIDIA chip";
                    NVPciChipsets[numUsed].numChipset = pciid;
@@ -1328,6 +1352,8 @@
     case 0x01D0:
     case 0x0090:
     case 0x0210:
+    case 0x0220:
+    case 0x0230:
          pNv->Architecture =  NV_ARCH_40;
          break;
     default:
--- xorg-x11-6.8.2/xc/programs/Xserver/hw/xfree86/drivers/nv/nv_hw.c.nv-ids	2004-12-16 02:42:35.000000000 +0100
+++ xorg-x11-6.8.2/xc/programs/Xserver/hw/xfree86/drivers/nv/nv_hw.c	2005-03-08 11:16:13.319120380 +0100
@@ -1173,6 +1173,7 @@
 
               switch(pNv->Chipset & 0xfff0) {
               case 0x0040:
+              case 0x0210:
                  pNv->PGRAPH[0x09b8/4] = 0x0078e366;
                  pNv->PGRAPH[0x09bc/4] = 0x0000014c;
                  pNv->PFB[0x033C/4] &= 0xffff7fff;
@@ -1182,6 +1183,7 @@
                  pNv->PGRAPH[0x082C/4] = 0x00000108;
                  break;
               case 0x0160:
+              case 0x01D0:
                  pNv->PMC[0x1700/4] = pNv->PFB[0x020C/4];
                  pNv->PMC[0x1704/4] = 0;
                  pNv->PMC[0x1708/4] = 0;
@@ -1194,6 +1196,15 @@
                  pNv->PGRAPH[0x0828/4] = 0x0072cb77;
                  pNv->PGRAPH[0x082C/4] = 0x00000108;
                  break;
+              case 0x0220:
+              case 0x0230:
+                 pNv->PGRAPH[0x0860/4] = 0;
+                 pNv->PGRAPH[0x0864/4] = 0;
+                 pNv->PRAMDAC[0x0608/4] |= 0x00100000;
+                 break;
+              case 0x0090:
+                 pNv->PRAMDAC[0x0608/4] |= 0x00100000;
+                 break;
               default:
                  break;
               };
