--- xorg-x11-6.8.2/xc/nls/Compose/pt_BR.UTF-8.compose-pt_br-utf8	2005-02-01 04:17:00.000000000 +0100
+++ xorg-x11-6.8.2/xc/nls/Compose/pt_BR.UTF-8	2005-03-22 22:53:41.060459115 +0100
@@ -30,6 +30,8 @@
 <dead_caron> <dead_caron>		: "ˇ"	caron
 <dead_cedilla> <space>			: "¸"	cedilla
 <dead_cedilla> <dead_cedilla>		: "¸"	cedilla
+<dead_cedilla> <c> : "ç" ccedilla
+<dead_cedilla> <C> : "Ç" Ccedilla
 <dead_ogonek> <dead_space>		: "˛"	ogonek
 <dead_ogonek> <dead_ogonek>		: "˛"	ogonek
 <dead_iota> <dead_space>		: "ͺ"	U37a
