--- LVM2.2.02.27/tools/lvmcmdline.c	2007-06-28 10:33:44.000000000 -0700
+++ LVM2.2.02.27/tools/lvmcmdline.c.new	2007-08-21 19:53:49.000000000 -0700
@@ -1127,7 +1127,7 @@ static void _exec_lvm1_command(char **ar
 {
 	char path[PATH_MAX];
 
-	if (dm_snprintf(path, sizeof(path), "%s.lvm1", argv[0]) < 0) {
+	if (dm_snprintf(path, sizeof(path), "lvm1-%s", argv[0]) < 0) {
 		log_error("Failed to create LVM1 tool pathname");
 		return;
 	}
@@ -1149,7 +1149,9 @@ int lvm2_main(int argc, char **argv, uns
 	while (*base == '/')
 		base++;
 	if (strcmp(base, "lvm") && strcmp(base, "lvm.static") &&
-	    strcmp(base, "initrd-lvm"))
+ 	    strcmp(base, "lvm2") && strcmp(base, "lvm2.static") &&
+ 	    strcmp(base, "lvm-static") && strcmp(base, "lvm2-static") &&
+ 	    strcmp(base, "initrd-lvm") && strcmp(base, "initrd-lvm2"))
 		alias = 1;
 
 	if (is_static && strcmp(base, "lvm.static") && 
