--- mc-4.6.0/acinclude.m4.jj	2003-01-30 10:48:32.000000000 -0500
+++ mc-4.6.0/acinclude.m4	2003-02-20 06:43:22.000000000 -0500
@@ -704,11 +704,11 @@ AC_DEFUN([MC_WITH_SLANG], [
     with_screen=slang
 
     dnl Unless external S-Lang was requested, reject S-Lang with UTF-8 hacks
-    m4_if([$1], strict, ,
-	  [AC_CHECK_LIB([slang], [SLsmg_write_nwchars],
-			[AC_MSG_WARN([Rejecting S-Lang with UTF-8 support, \
-it doesn't work well])
-			with_screen=mcslang])])
+dnl    m4_if([$1], strict, ,
+dnl	  [AC_CHECK_LIB([slang], [SLsmg_write_nwchars],
+dnl			[AC_MSG_WARN([Rejecting S-Lang with UTF-8 support, \
+dnl it doesn't work well])
+dnl			with_screen=mcslang])])
 
     if test x$with_screen = xslang; then
 	AC_CHECK_LIB([slang], [SLang_init_tty], [MCLIBS="$MCLIBS -lslang"],
--- mc-4.6.0/aclocal.m4.jj	2003-02-05 13:08:57.000000000 -0500
+++ mc-4.6.0/aclocal.m4	2003-02-20 06:44:28.000000000 -0500
@@ -717,11 +717,11 @@ AC_DEFUN([MC_WITH_SLANG], [
     with_screen=slang
 
     dnl Unless external S-Lang was requested, reject S-Lang with UTF-8 hacks
-    m4_if([$1], strict, ,
-	  [AC_CHECK_LIB([slang], [SLsmg_write_nwchars],
-			[AC_MSG_WARN([Rejecting S-Lang with UTF-8 support, \
-it doesn't work well])
-			with_screen=mcslang])])
+dnl    m4_if([$1], strict, ,
+dnl	  [AC_CHECK_LIB([slang], [SLsmg_write_nwchars],
+dnl			[AC_MSG_WARN([Rejecting S-Lang with UTF-8 support, \
+dnl it doesn't work well])
+dnl			with_screen=mcslang])])
 
     if test x$with_screen = xslang; then
 	AC_CHECK_LIB([slang], [SLang_init_tty], [MCLIBS="$MCLIBS -lslang"],
--- mc-4.6.0/configure.jj	2003-02-20 06:35:34.000000000 -0500
+++ mc-4.6.0/configure	2003-02-20 06:44:48.000000000 -0500
@@ -16483,7 +16483,6 @@ xslang)
     with_screen=slang
 
 
-
     if test x$with_screen = xslang; then
 	echo "$as_me:$LINENO: checking for SLang_init_tty in -lslang" >&5
 echo $ECHO_N "checking for SLang_init_tty in -lslang... $ECHO_C" >&6
@@ -17364,68 +17363,6 @@ x)
 
     with_screen=slang
 
-        echo "$as_me:$LINENO: checking for SLsmg_write_nwchars in -lslang" >&5
-echo $ECHO_N "checking for SLsmg_write_nwchars in -lslang... $ECHO_C" >&6
-if test "${ac_cv_lib_slang_SLsmg_write_nwchars+set}" = set; then
-  echo $ECHO_N "(cached) $ECHO_C" >&6
-else
-  ac_check_lib_save_LIBS=$LIBS
-LIBS="-lslang  $LIBS"
-cat >conftest.$ac_ext <<_ACEOF
-#line $LINENO "configure"
-/* confdefs.h.  */
-_ACEOF
-cat confdefs.h >>conftest.$ac_ext
-cat >>conftest.$ac_ext <<_ACEOF
-/* end confdefs.h.  */
-
-/* Override any gcc2 internal prototype to avoid an error.  */
-#ifdef __cplusplus
-extern "C"
-#endif
-/* We use char because int might match the return type of a gcc2
-   builtin and then its argument prototype would still apply.  */
-char SLsmg_write_nwchars ();
-int
-main ()
-{
-SLsmg_write_nwchars ();
-  ;
-  return 0;
-}
-_ACEOF
-rm -f conftest.$ac_objext conftest$ac_exeext
-if { (eval echo "$as_me:$LINENO: \"$ac_link\"") >&5
-  (eval $ac_link) 2>&5
-  ac_status=$?
-  echo "$as_me:$LINENO: \$? = $ac_status" >&5
-  (exit $ac_status); } &&
-         { ac_try='test -s conftest$ac_exeext'
-  { (eval echo "$as_me:$LINENO: \"$ac_try\"") >&5
-  (eval $ac_try) 2>&5
-  ac_status=$?
-  echo "$as_me:$LINENO: \$? = $ac_status" >&5
-  (exit $ac_status); }; }; then
-  ac_cv_lib_slang_SLsmg_write_nwchars=yes
-else
-  echo "$as_me: failed program was:" >&5
-sed 's/^/| /' conftest.$ac_ext >&5
-
-ac_cv_lib_slang_SLsmg_write_nwchars=no
-fi
-rm -f conftest.$ac_objext conftest$ac_exeext conftest.$ac_ext
-LIBS=$ac_check_lib_save_LIBS
-fi
-echo "$as_me:$LINENO: result: $ac_cv_lib_slang_SLsmg_write_nwchars" >&5
-echo "${ECHO_T}$ac_cv_lib_slang_SLsmg_write_nwchars" >&6
-if test $ac_cv_lib_slang_SLsmg_write_nwchars = yes; then
-  { echo "$as_me:$LINENO: WARNING: Rejecting S-Lang with UTF-8 support, \
-it doesn't work well" >&5
-echo "$as_me: WARNING: Rejecting S-Lang with UTF-8 support, \
-it doesn't work well" >&2;}
-			with_screen=mcslang
-fi
-
 
     if test x$with_screen = xslang; then
 	echo "$as_me:$LINENO: checking for SLang_init_tty in -lslang" >&5
