diff -uNp w1/scripts/tkgen.c.orig w1/scripts/tkgen.c
--- w1/scripts/tkgen.c.orig	2003-08-14 20:09:36.000000000 +0200
+++ w1/scripts/tkgen.c	2003-08-14 20:15:38.000000000 +0200
@@ -115,6 +115,8 @@
 #include "tkparse.h"
 
 
+#define MAX_MENUS	150
+
 /*
  * Total number of menus.
  */
@@ -123,8 +125,8 @@ static int tot_menu_num = 0;
 /*
  * Pointers to mainmenu_option and endmenu of each menu.
  */
-struct kconfig * menu_first [100];
-struct kconfig * menu_last  [100];
+struct kconfig * menu_first [MAX_MENUS];
+struct kconfig * menu_last  [MAX_MENUS];
 
 /*
  * Generate portion of wish script for the beginning of a submenu.
@@ -1147,7 +1149,7 @@ void dump_tk_script( struct kconfig * sc
 	case token_mainmenu_option:
 	    if ( ++menu_depth >= 64 )
 		{ fprintf( stderr, "menus too deep\n" ); exit( 1 ); }
-	    if ( ++tot_menu_num >= 100 )
+	    if ( ++tot_menu_num >= MAX_MENUS )
 		{ fprintf( stderr, "too many menus\n" ); exit( 1 ); }
 	    menu_num   [menu_depth]   = tot_menu_num;
 	    menu_first [tot_menu_num] = cfg;
