--- mod_auth_radius.c	2002-08-05 18:51:24.000000000 +0200
+++ mod_auth_radius.c.oden	2003-07-10 21:43:34.000000000 +0200
@@ -1897,9 +1897,7 @@
                                 char *ipnum_str_p = inet_ntoa (*client_addr);
                                 char *ipnum_str   = apr_pstrdup (r->pool, ipnum_str_p);
                                 apr_snprintf (errormsg, MAX_STRING_LEN,
-                                             "user %s is not allowed to connect from this address: %s see AuthRadiusNetMaskFile directive pache 
-configfile",
-                                             user, ipnum_str);
+                                             "user %s is not allowed to connect from this address: %s see AuthRadiusNetMaskFile directive", user, ipnum_str);
                                 return (HTTP_UNAUTHORIZED);
                         } else {        /* no tags requred */
                                 return (OK);
