--- ppp-2.4.1/pppd/demand.c~	2004-06-27 23:06:09.000000000 -0600
+++ ppp-2.4.1/pppd/demand.c	2004-06-27 23:06:09.000000000 -0600
@@ -34,7 +34,7 @@
 #include <sys/socket.h>
 #ifdef PPP_FILTER
 #include <net/if.h>
-#include <net/bpf.h>
+#include <pcap-bpf.h>
 #include <pcap.h>
 #endif
 
--- ppp-2.4.1/pppd/sys-linux.c~	2004-06-27 23:07:04.000000000 -0600
+++ ppp-2.4.1/pppd/sys-linux.c	2004-06-27 23:07:04.000000000 -0600
@@ -85,7 +85,7 @@
 #endif /* IPX_CHANGE */
 
 #ifdef PPP_FILTER
-#include <net/bpf.h>
+#include <pcap-bpf.h>
 #include <linux/filter.h>
 #endif /* PPP_FILTER */
 
