--- ash-0.3.8/Makefile.dietlibc	2004-04-15 09:35:03.682776386 -0300
+++ ash-0.3.8/Makefile	2004-04-15 09:39:55.726000364 -0300
@@ -1,5 +1,5 @@
-CC	= gcc
-LD	= gcc
+CC	= diet gcc
+LD	= diet gcc
 PROG	= sh
 SHSRCS	= alias.c cd.c error.c eval.c exec.c expand.c bltin/echo.c \
 	  histedit.c input.c jobs.c mail.c main.c memalloc.c miscbltin.c \
@@ -14,6 +14,7 @@
 LFLAGS	= -8	# 8-bit lex scanner for arithmetic
 YFLAGS	= -d
 CFLAGS	= -Wall -DBSD=1 -DSMALL -DGLOB_BROKEN \
+          -D__DIETLIBC__=1 \
           -DIFS_BROKEN -DFNMATCH_BROKEN \
           -D__COPYRIGHT\(x\)= -D__RCSID\(x\)= -D_DIAGASSERT\(x\)= -I. \
 	  -Os -fstrict-aliasing -fomit-frame-pointer -DSHELL -DHAVE_STDLIB_H
