--- ifplugd-0.15/src/ifplugd.c.force-up-on-error	2003-08-20 12:17:09.000000000 +0200
+++ ifplugd-0.15/src/ifplugd.c	2003-08-20 12:30:31.000000000 +0200
@@ -339,7 +339,8 @@
     int send_retval = 1;
     int paused = 0;
     static char log_ident[256];
-
+    int ret = 0;
+    
     snprintf(log_ident, sizeof(log_ident), "ifplugd(%s)", interface);
 
     daemon_log_ident = log_ident;
@@ -367,11 +368,13 @@
             break;
     }
 
-    if ((fd = open_iface(interface)) < 0)
+    if ((fd = open_iface(interface)) < 0) {
+        ret = -1;
         goto finish;
+    }
 
     if ((status = detect_beat(fd, interface)) == IFSTATUS_ERR)
-        goto finish;
+        goto cleanup;
     
     daemon_log(LOG_INFO, "ifplugd "VERSION" successfully initialized, link beat %sdetected.", status == IFSTATUS_UP ? "" : "not ");
     beep(status == IFSTATUS_UP ? 0 : 1);
@@ -448,8 +451,10 @@
 
 
         if (!paused) {
-            if ((r = detect_beat(fd, interface)) == IFSTATUS_ERR)
+            if ((r = detect_beat(fd, interface)) == IFSTATUS_ERR) {
+				status = r;
                 break;
+			}
 
             if (status != r) {
                 status = r;
@@ -476,7 +481,7 @@
         if (t && t < time(NULL)) {
             t = 0;
 
-            if (action(status) < 0)
+            if ((ret = action(status)) < 0)
                 goto finish;
         }
     }
@@ -485,8 +490,14 @@
     if (!no_shutdown_script && status == IFSTATUS_UP) {
         setenv(IFPLUGD_ENV_PREVIOUS, strstatus(status), 1);
         setenv(IFPLUGD_ENV_CURRENT, strstatus(-1), 1);
-        action(IFSTATUS_DOWN);
+        ret = action(IFSTATUS_DOWN);
         beep(1);
+    } else {
+		/* Mandrake Linux specific: if not supported, put the device up */
+        if (status == IFSTATUS_ERR) {
+			setenv("MII_NOT_SUPPORTED", "yes", 1);
+			ret = action(IFSTATUS_UP);
+		}
     }
     
 finish:
@@ -495,7 +506,7 @@
         close(fd);
 
     if (send_retval && daemonize && wait_on_fork)
-        daemon_retval_send(1);
+        daemon_retval_send(ret == 0 ? 1 : 4);
 
     daemon_pid_file_remove();
     daemon_signal_done();
