--- tiff-v3.5.7/libtiff/Makefile.in.deps	2003-10-09 17:58:34.000000000 +0200
+++ tiff-v3.5.7/libtiff/Makefile.in	2003-10-09 18:00:56.000000000 +0200
@@ -155,7 +155,7 @@ libtiff.a: ${OBJS}
 	${RANLIB} libtiff.a
 
 $(SHLIB): $(SHOBJS)
-	gcc -o $(SHLIB) -Wl,-soname,$(SONAME) -shared $(SHOBJS) -ljpeg -lz
+	gcc -o $(SHLIB) -Wl,-soname,$(SONAME) -shared $(SHOBJS) -ljpeg -lz -lm
 
 #
 # NB: The configure script verifies that the configured
