--- ./mount/mount.c.autodav	2003-12-11 11:33:32.000000000 +0100
+++ ./mount/mount.c	2003-12-11 11:34:08.000000000 +0100
@@ -829,6 +829,7 @@
   /*
    * Call mount.TYPE for types that require a separate mount program.
    * For the moment these types are ncpfs and smbfs. Maybe also vxfs.
+   * For the moment these types are ncpfs and smbfs and davfs. Maybe also vxfs.
    * All such special things must occur isolated in the types string.
    */
   if (check_special_mountprog (spec, node, types, flags, extra_opts, &status))
@@ -1172,7 +1173,12 @@
   }
 
   if (types == NULL && !mounttype) {
-      if (strchr (spec, ':') != NULL) {
+      if((!strncmp(spec, "http://", 7))||(!strncmp(spec, "https://", 8))) {
+	types = "davfs";
+	if (verbose)
+	  printf(_("mount: no type was given - "
+		   "I'll assume dav because of the http[s]:// prefix\n"));
+      } else if (strchr (spec, ':') != NULL) {
 	types = "nfs";
 	if (verbose)
 	  printf(_("mount: no type was given - "
