--- pppd/plugins/pppoe/pppoe.c~	2003-06-19 23:43:39.000000000 +0200
+++ pppd/plugins/pppoe/pppoe.c	2003-06-19 23:44:09.000000000 +0200
@@ -197,7 +197,7 @@
 
     if (pppoe_srv_name !=NULL) {
 	if (strlen (pppoe_srv_name) > 255) {
-	    poe_error (ses," Service name too long
+	    poe_error (ses," Service name too long \
 	                (maximum allowed 256 chars)");
 	    poe_die(-1);
 	}
--- pppd/plugins/pppoe/pppoed.c~	2003-06-19 23:43:39.000000000 +0200
+++ pppd/plugins/pppoe/pppoed.c	2003-06-19 23:44:58.000000000 +0200
@@ -168,7 +168,7 @@
 	case 'A':			/* AC override */
 	    poe_info (sess,"AC name override to %s", optarg);
 	    if (strlen (optarg) > 255) {
-		poe_error (sess," AC name too long 
+		poe_error (sess," AC name too long \
 			  (maximum allowed 256 chars)");
 		poe_die(-1);
 	    }
@@ -186,7 +186,7 @@
 	case 'E':			/* AC service name override */
 	    poe_info (sess,"AC service name override to %s", optarg);
 	    if (strlen (optarg) > 255) {
-		poe_error (sess," Service name too long 
+		poe_error (sess," Service name too long \
 	          (maximum allowed 256 chars)");
 		poe_die(-1);
 	    }
