--- SSLeay.pm	2006/01/27 11:36:03	1.1
+++ SSLeay.pm	2006/01/27 11:38:17
@@ -1603,14 +1603,14 @@
 
 sub ssl_read_all {
     my ($ssl,$how_much) = @_;
-    $how_much = 2000000000 unless $how_much;
+    $how_much = undef unless $how_much;
     my ($got, $errs);
     my $reply = '';
 
-    while ($how_much > 0) {
-	$got = Net::SSLeay::read($ssl,$how_much);
+    while (! defined $how_much || $how_much > 0) {
+	$got = Net::SSLeay::read($ssl,$how_much > 0 ? $how_much : 512);
 	last if $errs = print_errs('SSL_read');
-	$how_much -= blength($got);
+	defined $how_much and $how_much -= blength($got);
 	debug_read(\$reply, \$got) if $trace>1;
 	last if $got eq '';  # EOF
 	$reply .= $got;
