diff -uNp t1/kernel/printk.c.hb03-1.orig t1/kernel/printk.c
--- t1/kernel/printk.c.hb03-1.orig	2003-10-07 19:22:38.000000000 +0200
+++ t1/kernel/printk.c	2003-10-07 19:24:39.000000000 +0200
@@ -31,7 +31,7 @@
 #include <asm/uaccess.h>
 
 #if !defined(CONFIG_LOG_BUF_SHIFT) || (CONFIG_LOG_BUF_SHIFT == 0)
-#if defined(CONFIG_MULTIQUAD) || defined(CONFIG_IA64)
+#if defined(CONFIG_MULTIQUAD) || defined(CONFIG_IA64) || defined(CONFIG_X86_64)
 #define LOG_BUF_LEN	(65536)
 #elif defined(CONFIG_ARCH_S390)
 #define LOG_BUF_LEN	(131072)
