diff -urN openswan-2.3.1/programs/ikeping/ikeping.c openswan-2.3.1g-m/programs/ikeping/ikeping.c
--- openswan-2.3.1/programs/ikeping/ikeping.c        2004-10-18 05:28:59.000000000 +0530
+++ openswan-2.3.1-m/programs/ikeping/ikeping.c      2005-03-27 18:22:32.326750488 +0530
@@ -32,6 +32,7 @@
 #include <poll.h>

 #include <openswan.h>
+#include "pfkeyv2.h"

 #include "constants.h"
 #include "packet.h"
@@ -203,7 +204,7 @@
	rport = 500;
	xchg  = 0;
	sendlen=sizeof(sender);
-	n = recvfrom(s, &ih, sizeof(ih), 0, (struct sockaddr *)&sender, &sendlen);
+	n = recvfrom(s, &ih, sizeof(ih), 0, (struct sockaddr *)&sender, (socklen_t*)&sendlen);

	addrtot(&sender, 0, buf, sizeof(buf));
	switch(afamily) {
diff -urN openswan-2.3.1/programs/pluto/connections.c openswan-2.3.1-m/programs/pluto/connections.c
--- openswan-2.3.1/programs/pluto/connections.c      2005-03-21 09:19:42.000000000 +0530
+++ openswan-2.3.1-m/programs/pluto/connections.c    2005-03-27 18:22:50.305017376 +0530
@@ -30,6 +30,7 @@
 
 #include <openswan.h>
 #include <openswan/ipsec_policy.h>
+#include "pfkeyv2.h"
 #include "kameipsec.h"
 
 #include "constants.h"
diff -urN openswan-2.3.1/programs/pluto/ipsec_doi.c openswan-2.3.1-m/programs/pluto/ipsec_doi.c
--- openswan-2.3.1/programs/pluto/ipsec_doi.c        2005-03-20 07:57:50.000000000 +0530
+++ openswan-2.3.1-m/programs/pluto/ipsec_doi.c      2005-03-27 18:19:43.768375256 +0530
@@ -31,6 +31,7 @@
 
 #include <openswan.h>
 #include <openswan/ipsec_policy.h>
+#include "pfkeyv2.h"
 
 #include "constants.h"
 #include "defs.h"
diff -urN openswan-2.3.1/programs/pluto/log.c openswan-2.3.1-m/programs/pluto/log.c
--- openswan-2.3.1/programs/pluto/log.c      2004-12-29 11:38:20.000000000 +0530
+++ openswan-2.3.1-m/programs/pluto/log.c    2005-03-27 18:19:13.939909872 +0530
@@ -30,6 +30,7 @@
 #include <sys/types.h>
 
 #include <openswan.h>
+#include "pfkeyv2.h"
 
 #include "constants.h"
 #include "oswlog.h"
diff -urN openswan-2.3.1/programs/pluto/rcv_whack.c openswan-2.3.1-m/programs/pluto/rcv_whack.c
--- openswan-2.3.1/programs/pluto/rcv_whack.c        2005-03-21 09:21:59.000000000 +0530
+++ openswan-2.3.1-m/programs/pluto/rcv_whack.c      2005-03-27 18:20:12.287039760 +0530
@@ -31,6 +31,7 @@
 #include <sys/queue.h>
 
 #include <openswan.h>
+#include "pfkeyv2.h"
 
 #include "constants.h"
 #include "defs.h"
diff -urN openswan-2.3.1/programs/pluto/spdb.c openswan-2.3.1-m/programs/pluto/spdb.c
--- openswan-2.3.1/programs/pluto/spdb.c     2004-11-30 07:58:28.000000000 +0530
+++ openswan-2.3.1-m/programs/pluto/spdb.c   2005-03-27 18:20:37.837155552 +0530
@@ -24,6 +24,7 @@
 
 #include <openswan.h>
 #include <openswan/ipsec_policy.h>
+#include "pfkeyv2.h"
 
 #include "constants.h"
 #include "oswlog.h"
diff -urN openswan-2.3.1/programs/pluto/spdb_struct.c openswan-2.3.1-m/programs/pluto/spdb_struct.c
--- openswan-2.3.1/programs/pluto/spdb_struct.c      2005-02-15 07:24:20.000000000 +0530
+++ openswan-2.3.1-m/programs/pluto/spdb_struct.c    2005-03-27 18:20:59.474866120 +0530
@@ -24,6 +24,7 @@
 
 #include <openswan.h>
 #include <openswan/ipsec_policy.h>
+#include "pfkeyv2.h"
 
 #include "constants.h"
 #include "oswlog.h"
diff -urN openswan-2.3.1/programs/pluto/ikev1_quick_orig.c openswan-2.3.1/programs/pluto/ikev1_quick.c
--- openswan-2.3.1/programs/pluto/ikev1_quick_orig.c    2005-08-18 16:22:38.000000000 +0200
+++ openswan-2.3.1/programs/pluto/ikev1_quick.c 2005-08-18 16:22:51.000000000 +0200
@@ -33,6 +33,7 @@

 #include <openswan.h>
 #include <openswan/ipsec_policy.h>
+#include "pfkeyv2.h"

 #include "constants.h"
 #include "defs.h"
