--- authlib/onfigure.nores_query	2004-03-09 14:25:33.000000000 -0700
+++ authlib/configure	2004-03-09 14:29:50.000000000 -0700
@@ -5122,9 +5122,9 @@
   echo "$as_me: failed program was:" >&5
 sed 's/^/| /' conftest.$ac_ext >&5
 
-{ { echo "$as_me:$LINENO: error: Cannot find function res_query" >&5
-echo "$as_me: error: Cannot find function res_query" >&2;}
-   { (exit 1); exit 1; }; }
+#{ { echo "$as_me:$LINENO: error: Cannot find function res_query" >&5
+#echo "$as_me: error: Cannot find function res_query" >&2;}
+#   { (exit 1); exit 1; }; }
 fi
 rm -f conftest.$ac_objext conftest$ac_exeext conftest.$ac_ext
 fi
