--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/input/acecad/acecad.c~	2003-02-23 10:22:04.000000000 +0100
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/input/acecad/acecad.c	2003-02-23 10:22:04.000000000 +0100
@@ -33,7 +33,6 @@
  *	Standard Headers
  ****************************************************************************/
 
-#ifdef LINUX_INPUT
 #include <asm/types.h>
 #include <linux/input.h>
 #ifdef BUS_PCI
@@ -42,7 +41,6 @@
 #ifdef BUS_ISA
 #undef BUS_ISA
 #endif
-#endif
 
 #include <misc.h>
 #include <xf86.h>
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/input/confdrv.sh.debug	1999-06-12 17:37:09.000000000 +0200
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/input/confdrv.sh	2003-02-23 20:52:30.000000000 +0100
@@ -23,9 +23,13 @@
 Args="$BUILTIN `echo $* | tr '[a-z]' '[A-Z]'`"
 set - $Args
 while [ $# -gt 1 ]; do
+ if [ $1 != "ACECAD" ]; then
   echo "#undef $1" >> $DRVCONF
   echo "	$1," >> $DRVCONF
   shift
+ else
+  shift
+ fi
 done
 echo "#undef $1" >> $DRVCONF
 echo "	$1;" >> $DRVCONF
@@ -35,7 +39,9 @@
 {
 EOF
 for i in $Args; do
-  echo "	&$i," >> $DRVCONF
+  if [ $i != "ACECAD" ]; then
+    echo "	&$i," >> $DRVCONF
+  fi
 done
 cat >> $DRVCONF <<EOF
 };
