--- libpng-1.2.5/scripts/makefile.linux.mdkconf	2002-10-03 13:32:34.000000000 +0200
+++ libpng-1.2.5/scripts/makefile.linux	2003-08-27 10:32:17.000000000 +0200
@@ -15,12 +15,6 @@ CC=gcc
 # Prefix must be a full pathname.
 prefix=/usr/local
 
-# Where the zlib library and include files are located.
-#ZLIBLIB=/usr/local/lib
-#ZLIBINC=/usr/local/include
-ZLIBLIB=../zlib
-ZLIBINC=../zlib
-
 ALIGN=
 # for i386:
 #ALIGN=-malign-loops=2 -malign-functions=2
@@ -30,12 +24,11 @@ WARNMORE=-Wwrite-strings -Wpointer-arith
 	-Wstrict-prototypes -Wmissing-prototypes #-Wconversion
 
 # for pgcc version 2.95.1, -O3 is buggy; don't use it.
+CFLAGS=$(RPM_OPT_FLAGS) -Wall
 
-CFLAGS=-I$(ZLIBINC) -Wall -O3 -funroll-loops \
-	$(ALIGN) # $(WARNMORE) -g -DPNG_DEBUG=5
-
-LDFLAGS=-L. -Wl,-rpath,. -L$(ZLIBLIB) -Wl,-rpath,$(ZLIBLIB) -lpng12 -lz -lm
-LDFLAGS_A=-L$(ZLIBLIB) -Wl,-rpath,$(ZLIBLIB) libpng.a -lz -lm
+LIBPNG_LDFLAGS=-lz -lm
+LDFLAGS=-L. -Wl,-rpath,. -lpng12
+LDFLAGS_A=libpng.a $(LIBPNG_LDFLAGS)
 
 RANLIB=ranlib
 #RANLIB=echo
@@ -84,9 +77,9 @@ libpng-config:
 	( cat scripts/libpng-config-head.in; \
 	echo prefix=\"$(prefix)\"; \
 	echo I_opts=\"-I$(INCPATH)/$(LIBNAME)\"; \
-	echo L_opts=\"-L$(LIBPATH)\"; \
-	echo R_opts=\"-Wl,-rpath,$(LIBPATH)\"; \
-	echo libs=\"-lpng12 -lz -lm\"; \
+	echo L_opts=\"\"; \
+	echo R_opts=\"\"; \
+	echo libs=\"-lpng12\"; \
 	cat scripts/libpng-config-body.in ) > libpng-config
 	chmod +x libpng-config
 
@@ -99,11 +92,13 @@ $(LIBNAME).so.$(PNGMAJ): $(LIBNAME).so.$
 $(LIBNAME).so.$(PNGVER): $(OBJSDLL)
 	$(CC) -shared -Wl,-soname,$(LIBNAME).so.$(PNGMAJ) \
 	-o $(LIBNAME).so.$(PNGVER) \
+	$(LIBPNG_LDFLAGS) \
 	$(OBJSDLL)
 
 libpng.so.3.$(PNGMIN): $(OBJSDLL)
 	$(CC) -shared -Wl,-soname,libpng.so.3 \
 	-o libpng.so.3.$(PNGMIN) \
+	$(LIBPNG_LDFLAGS) \
 	$(OBJSDLL)
 
 pngtest: pngtest.o $(LIBNAME).so
@@ -186,9 +181,8 @@ install: install-static install-shared i
 # move the library to its final location.
 
 test-installed:
-	$(CC) -I$(ZLIBINC) \
+	$(CC) \
 	   `$(BINPATH)/libpng12-config --cflags` pngtest.c \
-	   -L$(ZLIBLIB) -Wl,-rpath,$(ZLIBLIB) \
 	   -o pngtesti `$(BINPATH)/libpng12-config --ldflags`
 	./pngtesti pngtest.png
 
