--- sysconfig.txt.opensls	2004-03-05 13:01:32.000000000 -0700
+++ sysconfig.txt	2004-03-05 13:01:33.000000000 -0700
@@ -554,7 +554,7 @@
       /sbin/pump
       /sbin/dhcpxd
   if you have a preferred client define the DHCP_CLIENT variable. 
-  When you use dhclient (which it is the default on a Mandrake-Linux
+  When you use dhclient (which is the default on an OpenSLS
   system) you can specify a timeout of waiting the answer from server,
   the option is called DHCP_TIMEOUT the default is 60 sec.
 
--- sysctl.conf.opensls	2004-05-11 09:15:15.000000000 -0600
+++ sysctl.conf	2004-05-11 09:18:19.000000000 -0600
@@ -1,4 +1,4 @@
-# Kernel sysctl configuration file for Mandrakelinux
+# Kernel sysctl configuration file for OpenSLS
 #
 # For binary values, 0 is disabled, 1 is enabled.  See sysctl(8) and
 # sysctl.conf(5) for more details.
@@ -21,8 +21,3 @@
 
 # on Andrey Borzenkov suggestion
 dev.cdrom.autoclose=0
-# removed to fix some digital extraction problems 
-# dev.cdrom.check_media=1
-
-# to be able to eject via the device eject button (magicdev)
-dev.cdrom.lock=0
--- inittab.opensls	2004-05-11 09:15:15.000000000 -0600
+++ inittab	2004-05-11 09:19:22.000000000 -0600
@@ -6,7 +6,7 @@
 #               Modified for RHS Linux by Marc Ewing and Donnie Barnes
 #
 
-# Default runlevel. The runlevels used by Mandrakelinux are:
+# Default runlevel. The runlevels used by OpenSLS are:
 #   0 - halt (Do NOT set initdefault to this)
 #   1 - Single user mode
 #   2 - Multiuser, without NFS (The same as 3, if you do not have networking)
--- rc.d/rc.local.opensls	2004-05-11 09:15:17.000000000 -0600
+++ rc.d/rc.local	2004-05-11 09:22:28.000000000 -0600
@@ -15,13 +15,13 @@
     fi
 fi
 
-SYSTEM=${SYSTEM=Mandrakelinux}
+SYSTEM=${SYSTEM=OpenSLS}
 
 # Source functions
 . /etc/init.d/functions
 
-if [ "$REWRITEISSUE" = "rewrite" -a -f /etc/mandrake-release ]; then
-    R=$(cat /etc/mandrake-release)
+if [ "$REWRITEISSUE" = "rewrite" -a -f /etc/opensls-release ]; then
+    R=$(cat /etc/opensls-release)
 
     arch=$(uname -m)
     a="a"
--- rc.d/rc.sysinit.opensls	2004-05-11 09:15:17.000000000 -0600
+++ rc.d/rc.sysinit	2004-05-11 09:23:51.000000000 -0600
@@ -98,8 +98,8 @@
 # Print a banner. ;)
 # C-like escape sequences don't work as 2nd and up parameters of gprintf,
 # so real escap chars were written
-PRODUCT=`sed "s/.*release \([0-9.]*\).*/\1/g" /etc/mandrake-release 2> /dev/null`
-SYSTEM=${SYSTEM=Mandrakelinux}
+PRODUCT=`sed "s/.*release \([0-9.]*\).*/\1/g" /etc/opensls-release 2> /dev/null`
+SYSTEM=${SYSTEM=OpenSLS}
 if [ "$BOOTUP" != "serial" ]; then
  gprintf "\t\t\tWelcome to %s" "`echo -en '\\033[1;36m'`$SYSTEM`echo -en '\\033[0;39m'` $PRODUCT"
 else
@@ -1118,13 +1118,13 @@
 fi
 
 if [ -f /var/lock/TMP_1ST ];then
- if [ -f /etc/init.d/mandrake_firstime ];then
-	/bin/sh /etc/init.d/mandrake_firstime
+ if [ -f /etc/init.d/opensls_firstime ];then
+	/bin/sh /etc/init.d/opensls_firstime
  fi
 fi
 
-if [ -f /etc/init.d/mandrake_everytime ]; then 
-	/bin/sh /etc/init.d/mandrake_everytime
+if [ -f /etc/init.d/opensls_everytime ]; then 
+	/bin/sh /etc/init.d/opensls_everytime
 fi
 
 # (pixel) a kind of profile for XF86Config
--- mandrake/mandrake_firstime.opensls	2004-05-11 09:15:15.000000000 -0600
+++ mandrake/mandrake_firstime	2004-05-11 09:24:57.000000000 -0600
@@ -1,5 +1,5 @@
 #!/bin/sh
-# Description: Launch Mandrake Configuration for the first time.
+# Description: Launch OpenSLS Configuration for the first time.
 # (c) MandrakeSoft, Chmouel Boudjnah <chmouel@mandrakesoft.com>
 # 	$Id: mandrake_firstime,v 1.18 2003/09/04 14:10:01 tvignaud Exp $	
 
@@ -11,16 +11,6 @@
 # If fail erase the temporary file
 trap 'rm -f /var/lock/TMP_1ST' 1 2 3 15
 
-#Rebuild Mozilla registry
-# be sure that mozilla registry is ok
-for MOZ_REBUILD_DATABASE in /usr/lib*/mozilla-*/mozilla-rebuild-databases.pl; do
-[ -x $MOZ_REBUILD_DATABASE ] &&
-{
- action "Building Mozilla registry" $MOZ_REBUILD_DATABASE 2>> $FLO
-}
-done
-
-
 [ -s $FLO ] || rm -f $FLO
 
 # Finish
--- mandrake/mandrake_everytime.opensls	2004-05-11 09:15:15.000000000 -0600
+++ mandrake/mandrake_everytime	2004-05-11 09:26:01.000000000 -0600
@@ -1,5 +1,5 @@
 #!/bin/bash
-# Description: Launch diverses Mandrake Scripts on boot.
+# Description: Launch diverses OpenSLS Scripts on boot.
 # (c) MandrakeSoft, Chmouel Boudjnah <chmouel@mandrakesoft.com>
 # 	$Id: mandrake_everytime,v 1.19 2003/07/11 21:02:33 flepied Exp $	
 
@@ -7,15 +7,6 @@
 
 [ -e /etc/sysconfig/system ] && . /etc/sysconfig/system
 
-# Disabling supermount if not here
-if [[ $(uname -m) != sparc*  ]];then
-    if grep -q '^\/.*supermount.*' /etc/fstab && \
-	! grep -q '[[:space:]]supermount$' /proc/filesystems  && \
-	! modprobe -n supermount >/dev/null 2>/dev/null && [ -x /usr/bin/perl ];then
-	action "Disabling Supermount" /usr/sbin/supermount -i disable
-    fi
-fi
-
 # setup eth* (without configuring it) interfaces before the
 # hotplug/pcmcia started so we are sure that what we have configured
 # in /etc/modules.conf get the device we have affected.
