--- dcron/Makefile.avx	2004-09-21 13:51:29.000000000 -0600
+++ dcron/Makefile	2004-09-21 13:52:06.000000000 -0600
@@ -15,8 +15,8 @@
 BINDIR=/usr/bin
 SBINDIR=/usr/bin
 MANDIR=/usr/man
-CRONTAB_OWNER=root.cron
-CROND_OWNER=root.wheel
+CRONTAB_OWNER=root:cron
+CROND_OWNER=root:wheel
 CRONTAB_PERM=4750
 CROND_PERM=0700
 
--- dcron/defs.h.avx	2004-09-21 13:52:11.000000000 -0600
+++ dcron/defs.h	2004-09-21 13:53:54.000000000 -0600
@@ -26,7 +26,7 @@
 #define arysize(ary)	(sizeof(ary)/sizeof((ary)[0]))
 
 #ifndef CRONTABS
-#define CRONTABS	"/var/spool/cron/crontabs"
+#define CRONTABS	"/var/spool/dcron/crontabs"
 #endif
 #ifndef TMPDIR
 #define TMPDIR		"/tmp"
@@ -50,7 +50,7 @@
 #define PATH_VI		"/usr/bin/vi"	/* location of vi	*/
 #endif
 #ifndef PIDFILE
-#define PIDFILE		"/var/run/crond.pid"
+#define PIDFILE		"/var/run/dcron/dcrond.pid"
 #endif
 
 
