diff -ruN util-linux-2.11t/mount.orig/swapon.c util-linux-2.11t/mount/swapon.c
--- util-linux-2.11t/mount.orig/swapon.c	2002-07-29 15:43:49.000000000 +0200
+++ util-linux-2.11t/mount/swapon.c	2002-07-29 15:46:33.000000000 +0200
@@ -342,6 +342,7 @@
 		if (streq (program_name, "swapon")) {
 		while ((fstab = getmntent(fp)) != NULL) {
 		        int i;
+			int encrypted = 0;
 			char *special = fstab->mnt_fsname;
 
 			if (streq(fstab->mnt_type, MNTTYPE_SWAP) &&
@@ -350,10 +351,13 @@
 				char *opt, *opts = strdup(fstab->mnt_opts);
 	   
 				for (opt = strtok(opts, ","); opt != NULL;
-				     opt = strtok(NULL, ","))
-					if (strncmp(opt, "pri=", 4) == 0)
-						priority = atoi(opt+4);
-
+				     opt = strtok (NULL, ",")) {
+				  if (strncmp(opt, "pri=", 4) == 0)
+				    priority = atoi(opt+4);
+				  if (strncmp(opt, "encrypted", 9) == 0)
+				    encrypted=1;
+				}
+				if (encrypted) continue; 
 				for (i = 0; i < numSwaps; i++)
 				     if (!strcmp(swapList[i], fstab->mnt_fsname)) break;
           
