--- XFree86-4.2.99.5/xc/programs/xkbcomp/rules/xfree86.old_kbd	2003-01-26 03:38:14.000000000 +0100
+++ XFree86-4.2.99.5/xc/programs/xkbcomp/rules/xfree86	2003-01-29 19:03:56.000000000 +0100
@@ -9,7 +9,15 @@
 //
 
 // Layouts that still need be composed by old rules
-! $oldlayouts = az bs ca ca_enhanced de_CH fr_CH hu hu_qwerty lt_std mn vn
+// all latin layouts with a third and fourth state accessibles trough AltGr
+// are put back to old models, as the new model gives too much problems
+// currently
+! $oldlayouts = al az be br bs ca ca_enhanced cz cz_qwerty czsk\
+	de de(nodeadkeys) de_CH dk dvorak dvorak(no) dvorak(se) en_US ee es\
+	fi fr fr_CH gb hr hu hu_qwerty hu_US ie is iso9995-3 it\
+	la lt lt_a lt_p lt_std lv mt mt_us nl no pl pl2 pt ro ro2\
+	sapmi se se_FI se_NO se_SE si sk sk_qwerty tr tr_f\
+	us_intl vn yu		jp tscii mn
 
 // If you want non-latin layouts implicitly include the en_US layout
 // uncomment lines below
