--- bin/dig/nslookup.c	2003-07-25 06:36:44.000000000 +0200
+++ bin/dig/nslookup.c.oden	2003-09-20 15:19:55.000000000 +0200
@@ -62,7 +62,7 @@
 extern char *progname;
 
 static isc_boolean_t short_form = ISC_TRUE,
-	tcpmode = ISC_FALSE, deprecation_msg = ISC_TRUE,
+	tcpmode = ISC_FALSE, deprecation_msg = ISC_FALSE;
 	identify = ISC_FALSE, stats = ISC_TRUE,
 	comments = ISC_TRUE, section_question = ISC_TRUE,
 	section_answer = ISC_TRUE, section_authority = ISC_TRUE,
