--- perl-5.8.1-RC3/lib/Pod/Perldoc/ToMan.pm.pix	2003-07-27 23:12:13.000000000 +0200
+++ perl-5.8.1-RC3/lib/Pod/Perldoc/ToMan.pm	2003-07-31 13:39:39.000000000 +0200
@@ -52,7 +52,7 @@
       ($self->{'__bindir'}  || die "no bindir set?!"  ),
       ($self->{'__pod2man'} || die "no pod2man set?!" ),
     )
-    . " $switches --lax $file | $render -man"
+    . " $switches --lax $file | $render -c -man"
   ;               # no temp file, just a pipe!
 
   # I hear persistent reports that adding a -c switch to $render
@@ -60,6 +60,8 @@
   # don't have a -c switch, so that adding it here would presumably
   # be a Bad Thing   -- sburke@cpan.org
 
+  # [Pixel] our man do have the -c switch, using it!!
+
   $command .= " | col -x" if $^O =~ /hpux/;
   
   defined(&Pod::Perldoc::DEBUG)
