--- linux-2.4.20/net/ipv4/netfilter/ip_fw_compat.c.za01-5.orig	2003-04-16 04:10:57.000000000 +0200
+++ linux-2.4.20/net/ipv4/netfilter/ip_fw_compat.c	2003-04-16 04:12:05.000000000 +0200
@@ -304,3 +304,5 @@ static void __exit fini(void)
 
 module_init(init);
 module_exit(fini);
+
+MODULE_LICENSE("GPL");
--- linux-2.4.20/net/ipv6/netfilter/ip6t_multiport.c.za01-5.orig	2001-10-31 00:08:12.000000000 +0100
+++ linux-2.4.20/net/ipv6/netfilter/ip6t_multiport.c	2003-04-16 04:12:05.000000000 +0200
@@ -99,3 +99,5 @@ static void __exit fini(void)
 
 module_init(init);
 module_exit(fini);
+
+MODULE_LICENSE("GPL");
