diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff c2/drivers/usb/serial/safe_serial.c x2/drivers/usb/serial/safe_serial.c
--- c2/drivers/usb/serial/safe_serial.c	2002-11-18 17:20:54.000000000 +0100
+++ x2/drivers/usb/serial/safe_serial.c	2002-11-18 21:21:09.000000000 +0100
@@ -416,11 +416,9 @@ static int safe_startup (struct usb_seri
 }
 
 struct usb_serial_device_type safe_device = {
+	owner:			THIS_MODULE,
 	name:			"Safe",
 	id_table:		id_table,
-	needs_interrupt_in:	DONT_CARE,
-	needs_bulk_in:		DONT_CARE,
-	needs_bulk_out:		DONT_CARE,
 	num_interrupt_in:	NUM_DONT_CARE,
 	num_bulk_in:		NUM_DONT_CARE,
 	num_bulk_out:		NUM_DONT_CARE,
