Index: util-linux-2.12q/mount/swapon.c
===================================================================
--- util-linux-2.12q.orig/mount/swapon.c	2005-04-29 14:09:23.000000000 +0200
+++ util-linux-2.12q/mount/swapon.c	2005-04-29 14:13:01.000000000 +0200
@@ -334,7 +334,8 @@ swapon_all(void) {
 		fclose(swaps);
 	}
 
-	while ((fstab = getmntent(fp)) != NULL) {
+	if (streq (progname, "swapon")) {
+	    while ((fstab = getmntent(fp)) != NULL) {
 		const char *orig_special = fstab->mnt_fsname;
 		const char *special;
 		int skip = 0;
@@ -368,6 +369,11 @@ swapon_all(void) {
 			         (!ifexists || !access(fstab->mnt_fsname, R_OK)))
 				status |= do_swapon(special, pri);
 		}
+	    }
+	} else { /* swapoff -- we don't need to parse fstab */
+		       int i;
+		       for(i=0; i<numSwaps; i++)
+			      swapoff(swapList[i]);
 	}
 	fclose(fp);
 
