--- ppp-2.4.1/pppdump/Makefile.linux	Mon Jul 26 13:09:29 1999
+++ ppp-2.4.1/pppdump/Makefile.geoff	Sun Apr  8 08:29:24 2001
@@ -1,4 +1,4 @@
-CFLAGS= -O -I../include/net
+CFLAGS=  -I../include/net $(RPM_OPT_FLAGS)
 OBJS = pppdump.o bsd-comp.o deflate.o zlib.o
 
 INSTALL= install
@@ -14,4 +14,4 @@
 install:
 	mkdir -p $(BINDIR) $(MANDIR)/man8
 	$(INSTALL) -s -c pppdump $(BINDIR)
-	$(INSTALL) -c -m 444 pppdump.8 $(MANDIR)/man8
+	$(INSTALL) -c pppdump.8 $(MANDIR)/man8
