--- symlinks-1.2/Makefile.orig	Tue Nov 30 21:26:13 1999
+++ symlinks-1.2/Makefile	Tue Nov 30 21:33:10 1999
@@ -6,8 +6,8 @@
 	gcc -Wall -Wstrict-prototypes -O3 -fomit-frame-pointer -fno-exceptions -fno-rtti -pipe -s -mpentium -mcpu=pentium -march=pentium -ffast-math -fexpensive-optimizations -N -s -o symlinks symlinks.c
  
 install: all symlinks.8
-	install -m 755 -o root -g root symlinks /usr/local/bin
-	install -m 644 -o root -g root symlinks.8 /usr/local/man/man8
+	install -m 755   symlinks /usr/local/bin
+	install -m 644   symlinks.8 /usr/local/man/man8
 
 clean:
 	rm -f symlinks *.o core
