--- xpm-3.4k/lib/CrBufFrI.c.xfree43merge	1998-03-19 20:50:59.000000000 +0100
+++ xpm-3.4k/lib/CrBufFrI.c	2003-11-13 17:07:39.000000000 +0100
@@ -31,6 +31,7 @@
 *                                                                             *
 *  Developed by Arnaud Le Hors                                                *
 \*****************************************************************************/
+/* $XFree86: xc/extras/Xpm/lib/CrBufFrI.c,v 1.2 2001/10/28 03:32:09 tsi Exp $ */
 
 #include "XpmI.h"
 
@@ -269,7 +270,7 @@ WriteColors(dataptr, data_size, used_siz
 	s += cpp;
 
 	for (key = 1; key <= NKEYS; key++, defaults++) {
-	    if (s2 = *defaults) {
+	    if ((s2 = *defaults)) {
 #ifndef VOID_SPRINTF
 		s +=
 #endif
--- xpm-3.4k/lib/CrDatFrI.c.xfree43merge	1998-03-19 20:50:59.000000000 +0100
+++ xpm-3.4k/lib/CrDatFrI.c	2003-11-13 17:07:39.000000000 +0100
@@ -31,6 +31,7 @@
 *                                                                             *
 *  Developed by Arnaud Le Hors                                                *
 \*****************************************************************************/
+/* $XFree86: xc/extras/Xpm/lib/CrDatFrI.c,v 1.2 2001/10/28 03:32:09 tsi Exp $ */
 
 #include "XpmI.h"
 
@@ -229,7 +230,7 @@ CreateColors(dataptr, data_size, colors,
 	s = buf + cpp;
 
 	for (key = 1; key <= NKEYS; key++, defaults++) {
-	    if (s2 = *defaults) {
+	    if ((s2 = *defaults)) {
 #ifndef VOID_SPRINTF
 		s +=
 #endif
--- xpm-3.4k/lib/RdFToI.c.xfree43merge	1998-03-19 20:51:00.000000000 +0100
+++ xpm-3.4k/lib/RdFToI.c	2003-11-13 17:07:39.000000000 +0100
@@ -31,6 +31,7 @@
 *                                                                             *
 *  Developed by Arnaud Le Hors                                                *
 \*****************************************************************************/
+/* $XFree86: xc/extras/Xpm/lib/RdFToI.c,v 1.2 2001/10/28 03:32:09 tsi Exp $ */
 
 #include "XpmI.h"
 #include <sys/stat.h>
@@ -128,8 +129,9 @@ OpenReadFile(filename, mdata)
     xpmData *mdata;
 {
 #ifndef NO_ZPIPE
-    char *compressfile, buf[BUFSIZ];
+    char buf[BUFSIZ];
 # ifdef STAT_ZFILE
+    char *compressfile;
     struct stat status;
 # endif
 #endif
--- xpm-3.4k/lib/WrFFrI.c.xfree43merge	1998-03-19 20:51:00.000000000 +0100
+++ xpm-3.4k/lib/WrFFrI.c	2003-11-13 17:07:39.000000000 +0100
@@ -31,6 +31,7 @@
 *                                                                             *
 *  Developed by Arnaud Le Hors                                                *
 \*****************************************************************************/
+/* $XFree86: xc/extras/Xpm/lib/WrFFrI.c,v 1.2 2001/10/28 03:32:09 tsi Exp $ */
 
 /*
  * The code related to AMIGA has been added by
@@ -119,23 +120,23 @@ XpmWriteFileFromXpmImage(filename, image
 	    name++;
 #endif
 	/* let's try to make a valid C syntax name */
-	if (dot = index(name, '.')) {
+	if (index(name, '.')) {
 	    strcpy(new_name, name);
 	    /* change '.' to '_' */
 	    name = s = new_name;
-	    while (dot = index(s, '.')) {
+	    while ((dot = index(s, '.'))) {
 		*dot = '_';
 		s = dot;
 	    }
 	}
-	if (dot = index(name, '-')) {
+	if (index(name, '-')) {
 	    if (name != new_name) {
 		strcpy(new_name, name);
 		name = new_name;
 	    }
 	    /* change '-' to '_' */
 	    s = name;
-	    while (dot = index(s, '-')) {
+	    while ((dot = index(s, '-'))) {
 		*dot = '_';
 		s = dot;
 	    }
@@ -226,7 +227,7 @@ WriteColors(file, colors, ncolors)
 	fprintf(file, "\"%s", *defaults++);
 
 	for (key = 1; key <= NKEYS; key++, defaults++) {
-	    if (s = *defaults)
+	    if ((s = *defaults))
 		fprintf(file, "\t%s %s", xpmColorKeys[key - 1], s);
 	}
 	fprintf(file, "\",\n");
--- xpm-3.4k/lib/Xpm-def.cpp.xfree43merge	1998-03-19 20:51:00.000000000 +0100
+++ xpm-3.4k/lib/Xpm-def.cpp	2003-11-13 17:07:39.000000000 +0100
@@ -35,3 +35,5 @@ EXPORTS
  XpmWriteFileFromImage
  XpmWriteFileFromPixmap
  XpmWriteFileFromXpmImage
+
+/* $XFree86: xc/lib/Xpm/Xpm-def.cpp,v 1.1 2000/08/09 23:40:13 dawes Exp $ */
--- xpm-3.4k/lib/create.c.xfree43merge	1998-03-19 20:51:00.000000000 +0100
+++ xpm-3.4k/lib/create.c	2003-11-13 17:07:39.000000000 +0100
@@ -32,6 +32,7 @@
 *                                                                             *
 *  Developed by Arnaud Le Hors                                                *
 \*****************************************************************************/
+/* $XFree86: xc/extras/Xpm/lib/create.c,v 1.3 2002/01/07 19:40:49 dawes Exp $ */
 
 /*
  * The code related to FOR_MSW has been added by
@@ -212,8 +213,7 @@ typedef struct {
 }      CloseColor;
 
 static int
-closeness_cmp(a, b)
-    Const void *a, *b;
+closeness_cmp(Const void *a, Const void *b)
 {
     CloseColor *x = (CloseColor *) a, *y = (CloseColor *) b;
 
@@ -517,7 +517,7 @@ CreateColors(display, attributes, colors
     /* variables stored in the XpmAttributes structure */
     Visual *visual;
     Colormap colormap;
-    XpmColorSymbol *colorsymbols;
+    XpmColorSymbol *colorsymbols = NULL;
     unsigned int numsymbols;
     XpmAllocColorFunc allocColor;
     void *closure;
@@ -525,7 +525,7 @@ CreateColors(display, attributes, colors
     char *colorname;
     unsigned int color, key;
     Bool pixel_defined;
-    XpmColorSymbol *symbol;
+    XpmColorSymbol *symbol = NULL;
     char **defaults;
     int ErrorStatus = XpmSuccess;
     char *s;
@@ -1755,7 +1755,6 @@ PutPixel1(ximage, x, y, pixel)
     register char *src;
     register char *dst;
     register int i;
-    register char *data;
     Pixel px;
     int nbytes;
 
@@ -1788,7 +1787,6 @@ PutPixel(ximage, x, y, pixel)
     register char *src;
     register char *dst;
     register int i;
-    register char *data;
     Pixel px;
     int nbytes, ibpp;
 
@@ -2186,7 +2184,7 @@ xpmParseDataAndCreate(display, data, ima
     /*
      * parse extensions
      */
-    if (info && (info->valuemask & XpmReturnExtensions))
+    if (info && (info->valuemask & XpmReturnExtensions)) {
 	if (extensions) {
 	    ErrorStatus = xpmParseExtensions(data, &info->extensions,
 					     &info->nextensions);
@@ -2196,7 +2194,7 @@ xpmParseDataAndCreate(display, data, ima
 	    info->extensions = NULL;
 	    info->nextensions = 0;
 	}
-
+    }
     /*
      * store found informations in the XpmImage structure
      */
--- xpm-3.4k/lib/data.c.xfree43merge	1998-03-19 20:51:00.000000000 +0100
+++ xpm-3.4k/lib/data.c	2003-11-13 17:07:39.000000000 +0100
@@ -31,14 +31,16 @@
 *                                                                             *
 *  Developed by Arnaud Le Hors                                                *
 \*****************************************************************************/
+/* $XFree86: xc/extras/Xpm/lib/data.c,v 1.4 2002/01/07 19:40:49 dawes Exp $ */
 
 #ifndef CXPMPROG
+#if 0
 /* Official version number */
 static char *RCS_Version = "$XpmVersion: 3.4k $";
 
 /* Internal version number */
-static char *RCS_Id = "$Id: xpm.shar,v 3.71 1998/03/19 19:47:14 lehors Exp $";
-
+static char *RCS_Id = "Id: xpm.shar,v 3.71 1998/03/19 19:47:14 lehors Exp $";
+#endif
 #include "XpmI.h"
 #endif
 #include <ctype.h>
@@ -49,8 +51,7 @@ static char *RCS_Id = "$Id: xpm.shar,v 3
 #endif
 
 static int
-ParseComment(data)
-    xpmData *data;
+ParseComment(xpmData *data)
 {
     if (data->type == XPMBUFFER) {
 	register char c;
@@ -386,14 +387,10 @@ xpmGetCmt(data, cmt)
 
 xpmDataType xpmDataTypes[] =
 {
-    "", "!", "\n", '\0', '\n', "", "", "", "",	/* Natural type */
-    "C", "/*", "*/", '"', '"', ",\n", "static char *", "[] = {\n", "};\n",
-    "Lisp", ";", "\n", '"', '"', "\n", "(setq ", " '(\n", "))\n",
-#ifdef VMS
-    NULL
-#else
-    NULL, NULL, NULL, 0, 0, NULL, NULL, NULL, NULL
-#endif
+    {"", "!", "\n", '\0', '\n', "", "", "", ""},	/* Natural type */
+    {"C", "/*", "*/", '"', '"', ",\n", "static char *", "[] = {\n", "};\n"},
+    {"Lisp", ";", "\n", '"', '"', "\n", "(setq ", " '(\n", "))\n"},
+    {NULL, NULL, NULL, 0, 0, NULL, NULL, NULL, NULL}
 };
 
 /*
--- xpm-3.4k/lib/parse.c.xfree43merge	1998-03-19 20:51:00.000000000 +0100
+++ xpm-3.4k/lib/parse.c	2003-11-13 17:07:39.000000000 +0100
@@ -22,6 +22,7 @@
  * used in advertising or otherwise to promote the sale, use or other dealings
  * in this Software without prior written authorization from GROUPE BULL.
  */
+/* $XFree86: xc/extras/Xpm/lib/parse.c,v 1.3 2001/10/28 03:32:10 tsi Exp $ */
 
 /*****************************************************************************\
 * parse.c:                                                                    *
@@ -32,6 +33,7 @@
 *                                                                             *
 *  Developed by Arnaud Le Hors                                                *
 \*****************************************************************************/
+/* $XFree86: xc/extras/Xpm/lib/parse.c,v 1.3 2001/10/28 03:32:10 tsi Exp $ */
 
 /*
  * The code related to FOR_MSW has been added by
@@ -40,6 +42,7 @@
 
 #include "XpmI.h"
 #include <ctype.h>
+#include <string.h>
 
 LFUNC(ParsePixels, int, (xpmData *data, unsigned int width,
 			 unsigned int height, unsigned int ncolors,
@@ -172,7 +175,7 @@ xpmParseColors(data, ncolors, cpp, color
     XpmColor **colorTablePtr;
     xpmHashTable *hashtable;
 {
-    unsigned int key, l, a, b;
+    unsigned int key = 0, l, a, b;
     unsigned int curkey;		/* current color key */
     unsigned int lastwaskey;		/* key read */
     char buf[BUFSIZ];
@@ -222,7 +225,7 @@ xpmParseColors(data, ncolors, cpp, color
 	    curkey = 0;
 	    lastwaskey = 0;
 	    *curbuf = '\0';		/* init curbuf */
-	    while (l = xpmNextWord(data, buf, BUFSIZ)) {
+	    while ((l = xpmNextWord(data, buf, BUFSIZ))) {
 		if (!lastwaskey) {
 		    for (key = 0, sptr = xpmColorKeys; key < NKEYS; key++,
 			 sptr++)
@@ -302,7 +305,7 @@ xpmParseColors(data, ncolors, cpp, color
 	     */
 	    xpmNextString(data);	/* get to the next string */
 	    *curbuf = '\0';		/* init curbuf */
-	    while (l = xpmNextWord(data, buf, BUFSIZ)) {
+	    while ((l = xpmNextWord(data, buf, BUFSIZ))) {
 		if (*curbuf != '\0')
 		    strcat(curbuf, " ");/* append space */
 		buf[l] = '\0';
@@ -693,7 +696,7 @@ xpmParseData(data, image, info)
     /*
      * parse extensions
      */
-    if (info && (info->valuemask & XpmReturnExtensions))
+    if (info && (info->valuemask & XpmReturnExtensions)) {
 	if (extensions) {
 	    ErrorStatus = xpmParseExtensions(data, &info->extensions,
 					     &info->nextensions);
@@ -703,6 +706,7 @@ xpmParseData(data, image, info)
 	    info->extensions = NULL;
 	    info->nextensions = 0;
 	}
+    }
 
     /*
      * store found informations in the XpmImage structure
--- xpm-3.4k/lib/scan.c.xfree43merge	1998-03-19 20:51:00.000000000 +0100
+++ xpm-3.4k/lib/scan.c	2003-11-13 17:07:39.000000000 +0100
@@ -31,6 +31,7 @@
 *                                                                             *
 *  Developed by Arnaud Le Hors                                                *
 \*****************************************************************************/
+/* $XFree86: xc/extras/Xpm/lib/scan.c,v 1.3 2002/01/07 19:40:49 dawes Exp $ */
 
 /*
  * The code related to FOR_MSW has been added by
@@ -73,6 +74,9 @@ LFUNC(storePixel, int, (Pixel pixel, Pix
 LFUNC(storeMaskPixel, int, (Pixel pixel, PixelsMap *pmap,
 			    unsigned int *index_return));
 
+typedef int (*storeFuncPtr)(Pixel pixel, PixelsMap *pmap,
+			    unsigned int *index_return);
+
 #ifndef FOR_MSW
 # ifndef AMIGA
 LFUNC(GetImagePixels, int, (XImage *image, unsigned int width,
@@ -89,16 +93,16 @@ LFUNC(GetImagePixels8, int, (XImage *ima
 
 LFUNC(GetImagePixels1, int, (XImage *image, unsigned int width,
 			     unsigned int height, PixelsMap *pmap,
-			     int (*storeFunc) ()));
+			     storeFuncPtr storeFunc));
 # else /* AMIGA */
 LFUNC(AGetImagePixels, int, (XImage *image, unsigned int width,
 			     unsigned int height, PixelsMap *pmap,
-			     int (*storeFunc) ()));
+			     storeFuncPtr storeFunc));
 # endif/* AMIGA */
 #else  /* ndef FOR_MSW */
 LFUNC(MSWGetImagePixels, int, (Display *d, XImage *image, unsigned int width,
 			       unsigned int height, PixelsMap *pmap,
-			       int (*storeFunc) ()));
+			       storeFuncPtr storeFunc));
 #endif
 LFUNC(ScanTransparentColor, int, (XpmColor *color, unsigned int cpp,
 				  XpmAttributes *attributes));
@@ -191,7 +195,7 @@ XpmCreateXpmImageFromImage(display, imag
     /* variables to return */
     PixelsMap pmap;
     XpmColor *colorTable = NULL;
-    int ErrorStatus;
+    int ErrorStatus = 0;
 
     /* calculation variables */
     unsigned int width = 0;
@@ -385,7 +389,7 @@ ScanTransparentColor(color, cpp, attribu
 		((XpmColor **) attributes->colorTable)[attributes->mask_pixel];
 /* end 3.2 bc */
 	for (key = 1; key <= NKEYS; key++) {
-	    if (s = mask_defaults[key]) {
+	    if ((s = mask_defaults[key])) {
 		defaults[key] = (char *) xpmstrdup(s);
 		if (!defaults[key])
 		    return (XpmNoMemory);
@@ -423,10 +427,10 @@ ScanOtherColors(display, colors, ncolors
     XpmColor *color;
     XColor *xcolors = NULL, *xcolor;
     char *colorname, *s;
-    XpmColor *colorTable, **oldColorTable = NULL;
+    XpmColor *colorTable = NULL, **oldColorTable = NULL;
     unsigned int ancolors = 0;
-    Pixel *apixels;
-    unsigned int mask_pixel;
+    Pixel *apixels = NULL;
+    unsigned int mask_pixel = 0;
     Bool found;
 
     /* retrieve information from the XpmAttributes */
@@ -521,7 +525,7 @@ ScanOtherColors(display, colors, ncolors
 
 		found = True;
 		for (key = 1; key <= NKEYS; key++) {
-		    if (s = adefaults[key])
+		    if ((s = adefaults[key]))
 			defaults[key] = (char *) xpmstrdup(s);
 		}
 	    }
@@ -845,7 +849,7 @@ GetImagePixels1(image, width, height, pm
     unsigned int width;
     unsigned int height;
     PixelsMap *pmap;
-    int (*storeFunc) ();
+    storeFuncPtr storeFunc;
 {
     unsigned int *iptr;
     int x, y;
--- xpm-3.4k/lib/xpm.h.xfree43merge	1998-03-19 20:51:00.000000000 +0100
+++ xpm-3.4k/lib/xpm.h	2003-11-13 17:07:39.000000000 +0100
@@ -22,6 +22,7 @@
  * used in advertising or otherwise to promote the sale, use or other dealings
  * in this Software without prior written authorization from GROUPE BULL.
  */
+/* $XFree86: xc/extras/Xpm/lib/xpm.h,v 1.2 2001/08/22 23:36:44 dawes Exp $ */
 
 /*****************************************************************************\
 * xpm.h:                                                                      *
@@ -392,7 +393,7 @@ extern "C" {
     FUNC(XpmReadFileToData, int, (char *filename, char ***data_return));
     FUNC(XpmWriteFileFromData, int, (char *filename, char **data));
 
-    FUNC(XpmAttributesSize, int, ());
+    FUNC(XpmAttributesSize, int, (void));
     FUNC(XpmFreeAttributes, void, (XpmAttributes *attributes));
     FUNC(XpmFreeExtensions, void, (XpmExtension *extensions,
 				   int nextensions));
@@ -400,7 +401,7 @@ extern "C" {
     FUNC(XpmFreeXpmImage, void, (XpmImage *image));
     FUNC(XpmFreeXpmInfo, void, (XpmInfo *info));
     FUNC(XpmGetErrorString, char *, (int errcode));
-    FUNC(XpmLibraryVersion, int, ());
+    FUNC(XpmLibraryVersion, int, (void));
 
     /* XpmImage functions */
     FUNC(XpmReadFileToXpmImage, int, (char *filename,
