# Allow tmpfile(3) to use $TMPDIR.

--- glibc-2.5.orig/libio/oldtmpfile.c
+++ glibc-2.5/libio/oldtmpfile.c
@@ -36,7 +36,7 @@ __old_tmpfile (void)
   int fd;
   FILE *f;
 
-  if (__path_search (buf, FILENAME_MAX, NULL, "tmpf", 0))
+  if (__path_search (buf, FILENAME_MAX, NULL, "tmpf", 1))
     return NULL;
   fd = __gen_tempname (buf, __GT_FILE);
   if (fd < 0)
--- glibc-2.5.orig/stdio-common/tmpfile.c
+++ glibc-2.5/stdio-common/tmpfile.c
@@ -43,7 +43,7 @@ tmpfile (void)
   int fd;
   FILE *f;
 
-  if (__path_search (buf, FILENAME_MAX, NULL, "tmpf", 0))
+  if (__path_search (buf, FILENAME_MAX, NULL, "tmpf", 1))
     return NULL;
   fd = __gen_tempname (buf, GEN_THIS);
   if (fd < 0)
