# Export __libc_enable_secure.

--- glibc-2.5.orig/elf/Versions
+++ glibc-2.5/elf/Versions
@@ -32,6 +32,7 @@ ld {
     calloc; free; malloc; realloc; __libc_memalign;
 
     _r_debug;
+    __libc_enable_secure;
   }
   GLIBC_2.1 {
     # functions used in other libraries
@@ -49,7 +50,6 @@ ld {
   }
   GLIBC_PRIVATE {
     # Those are in the dynamic linker, but used by libc.so.
-    __libc_enable_secure;
     _dl_argv;
     _dl_out_of_memory;
     _dl_starting_up;
