--- findutils-4.1.7/xargs/xargs.c.usage	2002-07-03 17:25:26.000000000 +0100
+++ findutils-4.1.7/xargs/xargs.c	2002-07-03 17:25:42.000000000 +0100
@@ -922,6 +922,6 @@
        [--max-args=max-args] [--no-run-if-empty] [--version] [--help]\n\
        [command [initial-arguments]]\n"),
 	   program_name);
-  fputs (_("\nReport bugs to <bug-findutils@gnu.org>."), stream);
+  fputs (_("\nReport bugs to <bug-findutils@gnu.org>.\n"), stream);
   exit (status);
 }
