--- mkbootdisk-1.5.1/mkbootdisk-redhat	2004-05-13 20:47:28.944464878 +0200
+++ mkbootdisk-1.5.1/mkbootdisk	2004-05-13 20:52:15.770560851 +0200
@@ -12,6 +12,7 @@
 unset mountopts
 unset isoimage
 unset realdev
+unset appendargs
 size=1440
 failed=0
 
@@ -24,7 +25,8 @@
 usage () {
     cat >&2 <<EOF
 usage: `basename $0` [--version] [--noprompt] [--device <devicefile>] 
-       [--verbose -v] [--iso] [--kernelargs <args>] [--size <size>]  <kernel>
+       [--verbose -v] [--iso] [--kernelargs <args>] [--size <size>]
+       [--appendargs <args>] <kernel>
        (ex: `basename $0` --device /dev/fd1 2.0.31)
 EOF
     exit $1
@@ -36,6 +38,10 @@
 	    shift
 	    device=$1
 	    ;;
+	--appendargs)
+	    shift
+	    appendargs=$1
+	    ;;
 	--kernelargs)
 	    shift
 	    kernelargs=$1
@@ -88,7 +94,8 @@
 }
 
 
-rootdev=`awk '$1 ~ /^[^#]/ && $2 ~ /^\/$/ { print $1 ; exit }' /etc/fstab`
+# yves from chmou fix loopback devices
+rootdev=`awk '$1 ~ /^[^#]/ && $2 == "/" { print $4 ~ /loop/ ? "/dev/loop7" : $1 }' /etc/fstab`
 
 if [ $(echo $rootdev | cut -c1-6) = "LABEL=" ]; then
     rootlabel=$(echo $rootdev | cut -c7-)
@@ -222,7 +229,7 @@
 timeout 100
 label linux
 	kernel vmlinuz
-	append $INITRDARG $kernelargs root=$rootdev
+	append $INITRDARG root=$rootdev $appendargs
 EOF
 
 [ $? = 0 ] || failed=1
@@ -231,7 +238,7 @@
 
 cat >> $bootmsg <<EOF
 
-Press <return> (or wait 10 seconds) to boot your Red Hat Linux system from
+Press <return> (or wait 10 seconds) to boot your Mandrakelinux system from
 $rootdev. You may override the default linux kernel parameters by typing
 "linux <params>", followed by <return> if you like.
 
@@ -243,8 +250,8 @@
 
 if [ -n "$isoimage" ]; then
     mkisofs \
-	-A "Red Hat Linux Boot CD" \
-	-V "Red Hat Linux Boot CD" \
+	-A "Mandrakelinux Boot CD" \
+	-V "Mandrakelinux Boot CD" \
 	-J -R -T -quiet \
 	-o $device \
 	-b isolinux/isolinux.bin -c isolinux/boot.cat -no-emul-boot \
