diff -uNp t1/arch/i386/kernel/pci-irq.c.ba11-1.orig t1/arch/i386/kernel/pci-irq.c
--- t1/arch/i386/kernel/pci-irq.c.ba11-1.orig	2003-10-03 13:27:41.000000000 +0200
+++ t1/arch/i386/kernel/pci-irq.c	2003-10-07 14:43:29.000000000 +0200
@@ -611,7 +611,8 @@ static __init int via_router_probe(struc
 		case PCI_DEVICE_ID_VIA_82C596:
 		case PCI_DEVICE_ID_VIA_82C686:
 		case PCI_DEVICE_ID_VIA_8231:
-		/* FIXME: add new ones for 8233/5 */
+		case PCI_DEVICE_ID_VIA_8233A:
+		case PCI_DEVICE_ID_VIA_8235:
 			r->name = "VIA";
 			r->get = pirq_via_get;
 			r->set = pirq_via_set;
