diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/drivers/video/riva/fbdev.c w1/drivers/video/riva/fbdev.c
--- w2/drivers/video/riva/fbdev.c	2003-06-13 16:51:37.000000000 +0200
+++ w1/drivers/video/riva/fbdev.c	2003-08-18 17:45:13.000000000 +0200
@@ -2221,7 +2221,6 @@ err_out_free_base1:
 	release_mem_region(rinfo->fb_base_phys, rinfo->base1_region_size);
 err_out_free_base0:
 	release_mem_region(rinfo->ctrl_base_phys, rinfo->base0_region_size);
-err_out_kfree:
 	kfree(rinfo);
 err_out:
 	return -ENODEV;
