--- linux/include/asm-i386/pgalloc.h.chmou	Mon Aug 20 19:50:45 2001
+++ linux/include/asm-i386/pgalloc.h	Mon Aug 20 19:52:32 2001
@@ -102,11 +102,16 @@
 
 static inline pte_t *pte_alloc_one(struct mm_struct *mm, unsigned long address)
 {
-	pte_t *pte;
-
-	pte = (pte_t *) __get_free_page(GFP_KERNEL);
-	if (pte)
-		clear_page(pte);
+	pte_t *pte=NULL;
+	int count = 0;
+   
+   	do {
+		pte = (pte_t *) __get_free_page(GFP_KERNEL);
+		if (pte)
+			clear_page(pte);
+		else
+			schedule_timeout(HZ);
+	} while ((pte==NULL)&&(count++<10));
 	return pte;
 }
 
