--- linux/drivers/net/ppp_generic.c.mppe	2002-05-21 13:25:54.000000000 -0400
+++ linux/drivers/net/ppp_generic.c	2002-05-21 13:29:52.000000000 -0400
@@ -64,6 +64,26 @@
 #define MIN_FRAG_SIZE	64
 
 /*
+* brute force -- redefine PPP_HDRLEN to allow a little
+* extra room for MPPE headers
+*/
+#undef PPP_HDRLEN
+#define PPP_HDRLEN 8
+/* 
+* End
+*/
+
+/*
+* increase CCP_MAX_OPTION_LENGTH to allow room for 2 128-bit
+* MPPE keys
+*/
+#undef CCP_MAX_OPTION_LENGTH
+#define CCP_MAX_OPTION_LENGTH 64
+/*
+* End
+*/
+
+/*
  * An instance of /dev/ppp can be associated with either a ppp
  * interface unit or a ppp channel.  In both cases, file->private_data
  * points to one of these.
