diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff --exclude='defconfig*' --exclude='log.*' --exclude='log2?' w2/drivers/block/cciss.c w1/drivers/block/cciss.c
--- w2/drivers/block/cciss.c	2003-08-20 19:13:51.000000000 +0200
+++ w1/drivers/block/cciss.c	2003-08-20 09:31:12.000000000 +0200
@@ -107,7 +107,7 @@ static struct board_type products[] = {
 #define NR_CMDS		 128 /* #commands that can be outstanding */
 #define MAX_CTLR 8
 
-#define CCISS_DMA_MASK 0xFFFFFFFFFFFFFFFF /* 64 bit DMA */
+#define CCISS_DMA_MASK 0xFFFFFFFFFFFFFFFFULL /* 64 bit DMA */
 
 static ctlr_info_t *hba[MAX_CTLR];
 
@@ -762,9 +762,6 @@ static int cciss_ioctl(struct inode *ino
 				 	!= XFER_NONE)) {	
 			return -EINVAL;
 		} 
-		/* Check kmalloc limits */
-		if (iocommand.buf_size > 128000)
-			return -EINVAL;
 		if (iocommand.buf_size > 0) {
 			buff =  kmalloc(iocommand.buf_size, GFP_KERNEL);
 			if (buff == NULL) 
