--- q2/arch/i386/kernel/dmi_scan.c	2003-11-28 01:35:52.000000000 +0100
+++ q1/arch/i386/kernel/dmi_scan.c	2003-11-28 01:31:33.000000000 +0100
@@ -902,6 +902,12 @@ static __initdata struct dmi_blacklist d
 			NO_MATCH, NO_MATCH
 			} },
 
+	{ local_apic_kills_bios, "Toshiba Sattelite 2435", {
+			MATCH(DMI_BOARD_VENDOR, "TOSHIBA"),
+			MATCH(DMI_PRODUCT_NAME, "BTS88"),
+			NO_MATCH, NO_MATCH
+			} },
+
 	{ init_ints_after_s1, "Toshiba Satellite 4030cdt", { /* Reinitialization of 8259 is needed after S1 resume */
 			MATCH(DMI_PRODUCT_NAME, "S4030CDT/4.3"),
 			NO_MATCH, NO_MATCH, NO_MATCH
