diff -uNp q1/drivers/usb/storage/unusual_devs.h.du21.orig q1/drivers/usb/storage/unusual_devs.h
--- q1/drivers/usb/storage/unusual_devs.h.du21.orig	2004-02-14 16:40:43.000000000 +0100
+++ q1/drivers/usb/storage/unusual_devs.h	2004-02-14 16:52:07.000000000 +0100
@@ -80,7 +80,7 @@ UNUSUAL_DEV(  0x03f0, 0x0307, 0x0001, 0x
 UNUSUAL_DEV(  0x0411, 0x001c, 0x0113, 0x0113,
 		"Buffalo",
 		"DUB-P40G HDD",
-		US_SC_DEVICE, US_PR_DEVICE, NULL,
+		US_SC_SCSI, US_PR_DEVICE, NULL,
 		US_FL_FIX_INQUIRY ),
 
 #ifdef CONFIG_USB_STORAGE_DPCM
@@ -90,6 +90,19 @@ UNUSUAL_DEV(  0x0436, 0x0005, 0x0100, 0x
  		US_SC_SCSI, US_PR_DPCM_USB, NULL, 0 ),
 #endif
 
+/* Made with the help of Edd Dumbill <edd@usefulinc.com> */
+UNUSUAL_DEV(  0x0451, 0x5409, 0x0001, 0x0001,
+		"Frontier Labs",
+		"Nex II Digital",
+		US_SC_SCSI, US_PR_BULK, NULL, 0),
+
+/* Reported by Thomas Rabe <raven@drehmoment.org> */
+UNUSUAL_DEV(  0x0461, 0x0822, 0x0000, 0x9999,
+		"Vivitar",
+		"Vivicam 3610",
+		US_SC_SCSI, US_PR_BULK, NULL,
+		US_FL_FIX_INQUIRY | US_FL_MODE_XLATE),
+
 /* Patch submitted by Philipp Friedrich <philipp@void.at> */
 UNUSUAL_DEV(  0x0482, 0x0100, 0x0100, 0x0100,
 		"Kyocera",
@@ -168,6 +181,21 @@ UNUSUAL_DEV(  0x04ce, 0x0002, 0x0260, 0x
 		US_SC_SCSI, US_PR_BULK, NULL,
 		US_FL_SL_IDE_BUG ),
 
+/* Reported by Leif Sawyer <leif@gci.net> */
+UNUSUAL_DEV(  0x04ce, 0x0002, 0x0240, 0x0240,
+		"H45 ScanLogic",
+		"SL11R-IDE 9951SQFP-1.2 K004",
+		US_SC_SCSI, US_PR_BULK, NULL,
+		US_FL_FIX_INQUIRY | US_FL_SL_IDE_BUG ),
+
+/* Reported by Rene Engelhard <mail@rene-engelhard.de> and
+    Dylan Egan <crack_me@bigpond.com.au> */
+UNUSUAL_DEV(  0x04ce, 0x0002, 0x0260, 0x0260,
+		"ScanLogic",
+		"SL11R-IDE unknown HW rev",
+		US_SC_SCSI, US_PR_BULK, NULL,
+		US_FL_SL_IDE_BUG ),
+
 /* Most of the following entries were developed with the help of
  * Shuttle/SCM directly.
  */
@@ -260,6 +288,13 @@ UNUSUAL_DEV(  0x054c, 0x0010, 0x0106, 0x
 		US_SC_SCSI, US_PR_DEVICE, NULL,
 		US_FL_SINGLE_LUN | US_FL_MODE_XLATE ),
 
+/* This entry is needed because the device reports Sub=ff */
+UNUSUAL_DEV(  0x054c, 0x0010, 0x0106, 0x0432, 
+		"Sony",
+		"DSC-F707/U10/U20", 
+		US_SC_SCSI, US_PR_CB, NULL,
+		US_FL_SINGLE_LUN | US_FL_MODE_XLATE ),
+
 /* Reported by wim@geeks.nl */
 UNUSUAL_DEV(  0x054c, 0x0025, 0x0100, 0x0100, 
 		"Sony",
@@ -289,11 +324,17 @@ UNUSUAL_DEV(  0x054c, 0x002e, 0x0106, 0x
 		US_FL_SINGLE_LUN | US_FL_MODE_XLATE),
 
 UNUSUAL_DEV(  0x054c, 0x0032, 0x0000, 0x9999,
-		"Sony",
+                "Sony",
 		"Memorystick MSC-U01N",
-		US_SC_DEVICE, US_PR_DEVICE, NULL,
+		US_SC_UFI, US_PR_DEVICE, NULL,
 		US_FL_SINGLE_LUN ),
 		
+UNUSUAL_DEV(  0x054c, 0x0058, 0x0000, 0x9999,
+                "Sony",
+		"PEG-N760C Mass Storage",
+		US_SC_8070, US_PR_CBI, NULL,
+		US_FL_FIX_INQUIRY ),
+		
 UNUSUAL_DEV(  0x054c, 0x0069, 0x0000, 0x9999,
 		"Sony",
 		"Memorystick MSC-U03",
@@ -374,6 +415,24 @@ UNUSUAL_DEV(  0x05ab, 0x0581, 0x0100, 0x
                 US_SC_SCSI, US_PR_USS725, uss725_init, 0 ),
 #endif
 
+#ifdef CONFIG_USB_STORAGE_USS725
+UNUSUAL_DEV(  0x05ab, 0x0200, 0x0100, 0x0110,
+                "In-System",
+                "USS-725 USB/IDE Bridge (ATA/ATAPI)",
+                US_SC_SCSI, US_PR_USS725, uss725_init, 0 ),
+
+/* Reported by <dwiebold@cactus.org> for original Pockey */
+UNUSUAL_DEV(  0x05ab, 0x0202, 0x0100, 0x0155,
+                "In-System",
+                "USS-725 USB/IDE Bridge (ATA/ATAPI)",
+                US_SC_SCSI, US_PR_USS725, uss725_init, 0 ),
+
+UNUSUAL_DEV(  0x05ab, 0x0581, 0x0100, 0x0110,
+                "In-System",
+                "USS-725 USB/IDE Bridge (ATA/ATAPI)",
+                US_SC_SCSI, US_PR_USS725, uss725_init, 0 ),
+#endif
+
 #ifdef CONFIG_USB_STORAGE_JUMPSHOT
 UNUSUAL_DEV(  0x05dc, 0x0001, 0x0000, 0x0001,
 		"Lexar",
@@ -474,6 +533,12 @@ UNUSUAL_DEV( 0x0686, 0x400f, 0x0001, 0x0
              US_SC_SCSI, US_PR_DEVICE, NULL,
              0 ),
 
+/* Submitted by Benny Sjostrand <benny@hostmobility.com> */
+UNUSUAL_DEV( 0x0686, 0x4011, 0x0001, 0x0001,
+		"Minolta",
+		"Dimage F300",
+		US_SC_SCSI, US_PR_BULK, NULL, 0 ),
+
 UNUSUAL_DEV(  0x0693, 0x0002, 0x0100, 0x0100, 
 		"Hagiwara",
 		"FlashGate SmartMedia",
@@ -497,10 +562,10 @@ UNUSUAL_DEV(  0x0781, 0x0002, 0x0009, 0x
 		US_FL_IGNORE_SER ),
 
 UNUSUAL_DEV(  0x0781, 0x0100, 0x0100, 0x0100,
-		"Sandisk",
-		"ImageMate SDDR-12",
-		US_SC_SCSI, US_PR_CB, NULL,
-		US_FL_SINGLE_LUN ),
+                "Sandisk",
+                "ImageMate SDDR-12",
+                US_SC_SCSI, US_PR_CB, NULL,
+                US_FL_SINGLE_LUN ),
 
 #ifdef CONFIG_USB_STORAGE_SDDR09
 UNUSUAL_DEV(  0x0781, 0x0200, 0x0000, 0x9999, 
@@ -510,6 +575,12 @@ UNUSUAL_DEV(  0x0781, 0x0200, 0x0000, 0x
 		US_FL_SINGLE_LUN ),
 #endif
 
+UNUSUAL_DEV(  0x0784, 0x1688, 0x0000, 0x9999,
+		"Vivitar",
+		"Vivicam 36xx",
+		US_SC_SCSI, US_PR_BULK, NULL,
+		US_FL_FIX_INQUIRY | US_FL_MODE_XLATE),
+
 #ifdef CONFIG_USB_STORAGE_FREECOM
 UNUSUAL_DEV(  0x07ab, 0xfc01, 0x0000, 0x9999,
 		"Freecom",
@@ -611,6 +682,16 @@ UNUSUAL_DEV(  0x07c4, 0xa109, 0x0000, 0x
 		US_FL_SINGLE_LUN ),
 #endif
 
+/* Datafab KECF-USB Ver A /Jenoptik Jenreader
+ * Note: there seem to be two versions of the KECF-USB device.
+ * Submitted by Chris Clayton (chris@theclaytons.freeserve.co.uk)
+ */
+UNUSUAL_DEV(  0x07c4, 0xb000, 0x0000, 0xffff,
+		"Datafab",
+		"KECF-USB Ver A",
+		US_SC_SCSI, US_PR_BULK, NULL,
+		US_FL_FIX_INQUIRY ),
+
 /* Datafab KECF-USB / Sagatek DCS-CF / Simpletech Flashlink UCF-100
  * Only revision 1.13 tested (same for all of the above devices,
  * based on the Datafab DF-UG-07 chip).  Needed for US_FL_FIX_INQUIRY.
@@ -643,6 +724,13 @@ UNUSUAL_DEV( 0x0839, 0x000a, 0x0001, 0x0
 		US_SC_DEVICE, US_PR_DEVICE, NULL,
 		US_FL_FIX_INQUIRY),
 
+/* Info by Vincent Meyer <meyerv@winflex.com>*/
+UNUSUAL_DEV( 0x084d, 0x0011, 0x0000, 0x0000,
+		"Argus",
+		"DC-3500",
+		US_SC_DEVICE, US_PR_DEVICE, NULL,
+		US_FL_FIX_INQUIRY),
+
 /* Aiptek PocketCAM 3Mega
  * Nicolas DUPEUX <nicolas@dupeux.net> 
  */
@@ -794,3 +882,31 @@ UNUSUAL_DEV(  0x0482, 0x0105, 0x0100, 0x
 		"Finecam L3",
 		US_SC_SCSI, US_PR_BULK, NULL,
 		US_FL_FIX_INQUIRY),
+
+/*
+ * Panasonic/OEMs compact USB CDROMs status
+ * KXL-840(CD-ROM11): usb_stor_Bulk_max_lun() is danger, need US_FL_SINGLE_LUN
+ * KXL-RW11(CDRRW02): usb_stor_Bulk_max_lun() is danger, need US_FL_SINGLE_LUN
+ * KXL-RW20(CDRRW03): original IClass is 0xFF, use US_PR_CB and need init reset
+ * KXL-RW21(CDRRW06): original IClass is 0xFF, use US_PR_CB and need init reset
+ * KXL-RW31(CDRRW05): work fine with current code
+ * KXL-RW32(CDRRW09): work fine with current code
+ *      Checked: Sun Feb 9 JST 2003 Go Taniguchi <go@turbolinux.co.jp>
+ */
+UNUSUAL_DEV( 0x04da, 0x0d01, 0x0000, 0xffff,
+		"MATSHITA",
+		"CD-ROM11",
+		US_SC_8020, US_PR_BULK, NULL, US_FL_SINGLE_LUN),
+UNUSUAL_DEV( 0x04da, 0x0d02, 0x0000, 0xffff,
+		"MATSHITA",
+		"CDRRW02",
+		US_SC_8020, US_PR_BULK, NULL, US_FL_SINGLE_LUN),
+UNUSUAL_DEV( 0x04da, 0x0d03, 0x0000, 0xffff,
+		"MATSHITA",
+		"CDRRW03",
+		US_SC_8020, US_PR_CB, NULL, US_FL_INIT_RESET),
+UNUSUAL_DEV( 0x04da, 0x0d06, 0x0000, 0xffff,
+		"MATSHITA",
+		"CDRRW06",
+		US_SC_8020, US_PR_CB, NULL, US_FL_INIT_RESET),
+
