--- linux-2.4.24/kernel/ksyms.c.fn01.orig	2004-01-19 20:07:12.466895592 +0100
+++ linux-2.4.24/kernel/ksyms.c	2004-01-19 20:08:10.349096160 +0100
@@ -193,7 +193,9 @@ EXPORT_SYMBOL(d_alloc);
 EXPORT_SYMBOL(d_lookup);
 EXPORT_SYMBOL(__d_path);
 EXPORT_SYMBOL(mark_buffer_dirty);
-EXPORT_SYMBOL(set_buffer_async_io); /* for reiserfs_writepage */
+EXPORT_SYMBOL(end_buffer_io_sync);
+EXPORT_SYMBOL(set_buffer_async_io);
+EXPORT_SYMBOL(__mark_dirty);
 EXPORT_SYMBOL(end_buffer_io_async);
 EXPORT_SYMBOL(__mark_buffer_dirty);
 EXPORT_SYMBOL(__mark_inode_dirty);
