--- /usr/src/linux-2.4.21-0.pre3.1mdk/drivers/scsi/ppa.c	2002-11-29 02:53:14.000000000 +0300
+++ linux-2.4.21-0.pre3.1mdk/drivers/scsi/ppa.c	2003-01-21 22:15:44.000000000 +0300
@@ -23,6 +23,8 @@
 int ppa_release(struct Scsi_Host *);
 static void ppa_reset_pulse(unsigned int base);
 
+static int no_cable_warning;
+
 typedef struct {
     struct pardevice *dev;	/* Parport device entry         */
     int base;			/* Actual port address          */
@@ -217,12 +219,14 @@
     }
     if (nhosts == 0) {
 	if (try_again == 1) {
-	    printk("WARNING - no ppa compatible devices found.\n");
-	    printk("  As of 31/Aug/1998 Iomega started shipping parallel\n");
-	    printk("  port ZIP drives with a different interface which is\n");
-	    printk("  supported by the imm (ZIP Plus) driver. If the\n");
-	    printk("  cable is marked with \"AutoDetect\", this is what has\n");
-	    printk("  happened.\n");
+	    if (!no_cable_warning) {
+		printk("WARNING - no ppa compatible devices found.\n");
+		printk("  As of 31/Aug/1998 Iomega started shipping parallel\n");
+		printk("  port ZIP drives with a different interface which is\n");
+		printk("  supported by the imm (ZIP Plus) driver. If the\n");
+		printk("  cable is marked with \"AutoDetect\", this is what has\n");
+		printk("  happened.\n");
+	    }
 	    spin_lock_irq(&io_request_lock);
 	    return 0;
 	}
@@ -1146,3 +1150,17 @@
     return 1;
 }
 MODULE_LICENSE("GPL");
+MODULE_PARM(no_cable_warning, "i");
+MODULE_PARM_DESC(no_cable_warning, "Supress annoying cable warning");
+
+#ifndef MODULE
+static int __init ppa_setup(char *str)
+{
+	if (!strcmp(str, "no_cable_warning"))
+		no_cable_warning = 1;
+
+	return 1;
+}
+
+__setup("ppa=", ppa_setup);
+#endif
