--- mod_auth_remote.c	2003-01-21 21:20:37.000000000 +0100
+++ mod_auth_remote.c.oden	2003-01-21 21:23:22.000000000 +0100
@@ -175,8 +175,16 @@
    return  HTTP_UNAUTHORIZED;
 }
 
+static int auth_remote_init_handler(apr_pool_t *p, apr_pool_t *plog,
+                            	apr_pool_t *ptemp, server_rec *s)
+{
+   ap_add_version_component(p, "mod_auth_remote/1.0");
+   return OK;
+}
+
 static void register_hooks(apr_pool_t *p)
 {
+  ap_hook_post_config(auth_remote_init_handler, NULL, NULL, APR_HOOK_MIDDLE);
   ap_hook_check_user_id(authenticate_basic_user, NULL, NULL, APR_HOOK_MIDDLE);
   /**
          check_user_access will return OK without performing __any__ checks.
