--- devfsd/modules.devfs.orig	Sun Mar 17 21:38:23 2002
+++ devfsd/modules.devfs	Sun Mar 17 21:41:39 2002
@@ -49,8 +52,9 @@
 alias     /dev/sd*		/dev/sd
 
 # SCSI CD-ROMs
 probeall  /dev/sr		scsi_hostadapter sr_mod
-alias     /dev/sr*		/dev/sr
+probeall  /dev/scd		scsi_hostadapter sr_mod
+alias     /dev/scd*		/dev/scd
-alias     /dev/scsi/*/cd	sr_mod
+alias     /dev/scsi/*/cd	/dev/scd
 
 # SCSI tapes
--- devfsd0/compat_name.c	2002-08-17 18:46:21.000000000 +0200
+++ devfsd/compat_name.c	2002-08-17 19:07:26.000000000 +0200
@@ -201,7 +196,7 @@
 	}
 	else if (strcmp (devname + namelen - 2, "cd") == 0)
 	{
-	    sprintf (buffer, "sr%u", minor);
+	    sprintf (buffer, "scd%u", minor);
 	    compat_name = buffer;
 	}
 	else if (strcmp (devname + namelen - 4, "disc") == 0)
diff -ru devfsd/devfsd.c devfsd/devfsd.c
--- devfsd/devfsd.c	2004-11-15 14:17:00.000000000 +0100
+++ devfsd/devfs.dc	2004-11-15 13:33:12.499665608 +0100
@@ -1573,6 +1573,18 @@
 	    if (trace_level > 1)
 		fprintf (stderr, "made symlink: \"%s\" for dev: %u,%u\n",
 			compat_name, info->major, info->minor);
+	if (strncmp(compat_name, "scd", 3) == 0) {
+	  char *str = strdup(compat_name);
+	  char *new_dest = str+1;
+	  str[1] = 's';
+	  str[2] = 'r';
+	  
+	  if (mksymlink (dest_name, new_dest) == 0)
+	    if (trace_level > 1)
+	      fprintf (stderr, "made symlink: \"%s\" for dev: %u,%u\n",
+		       new_dest, info->major, info->minor);
+	  free(str);
+	}
 	break;
       case AC_RMOLDCOMPAT:
       case AC_RMNEWCOMPAT:
