--- ./compat_name.c.usb_serial	2002-08-17 19:09:26.000000000 +0200
+++ ./compat_name.c	2002-08-17 19:09:26.000000000 +0200
@@ -131,6 +131,7 @@
     {"cua/F",      "cuf%s"},        /*  Computone serial driver callout      */
     {"tts/C",      "ttyC%s"},       /*  Cyclades serial driver               */
     {"cua/C",      "cub%s"},        /*  Cyclades serial driver callout       */
+    {"usb/tts/",   "ttyUSB%s"},     /*  USB serial driver                    */
     {"tts/",       "ttyS%s"},       /*  Generic serial: must be after others */
     {"cua/",       "cua%s"},        /*  Generic serial: must be after others */
     {"input/js",   "js%s"},         /*  Joystick driver                      */
