--- cvs-1.11.1/src/logmsg.c.fred	Tue Apr 24 20:14:53 2001
+++ cvs-1.11.1/src/logmsg.c	Tue Jul 10 08:41:03 2001
@@ -247,7 +247,7 @@
     }
 
     (void) fprintf (fp,
-  "%s----------------------------------------------------------------------\n",
+  "\n%s----------------------------------------------------------------------\n",
 		    CVSEDITPREFIX);
     (void) fprintf (fp,
   "%sEnter Log.  Lines beginning with `%.*s' are removed automatically\n%s\n",
