--- ./rpmqv.c.shortcircuit	2005-05-12 14:01:07.634264808 +0200
+++ ./rpmqv.c	2005-05-12 14:21:41.191735568 +0200
@@ -688,16 +688,18 @@
 	case 'b':
 	    ba->buildAmount |= RPMBUILD_PACKAGEBINARY;
 	    ba->buildAmount |= RPMBUILD_CLEAN;
+	    if (ba->shortCircuit)
+		break;
 	    /*@fallthrough@*/
 	case 'i':
 	    ba->buildAmount |= RPMBUILD_INSTALL;
 	    ba->buildAmount |= RPMBUILD_CHECK;
-	    if ((ba->buildChar == 'i') && ba->shortCircuit)
+	    if (/* (ba->buildChar == 'i') && */ ba->shortCircuit)
 		/*@innerbreak@*/ break;
 	    /*@fallthrough@*/
 	case 'c':
 	    ba->buildAmount |= RPMBUILD_BUILD;
-	    if ((ba->buildChar == 'c') && ba->shortCircuit)
+	    if ( /* (ba->buildChar == 'c') && */ ba->shortCircuit)
 		/*@innerbreak@*/ break;
 	    /*@fallthrough@*/
 	case 'p':
