--- ncurses-5.2/include/termcap.h.in.64bit	Fri Nov 17 12:40:25 2000
+++ ncurses-5.2/include/termcap.h.in	Fri Nov 17 12:41:23 2000
@@ -45,6 +45,7 @@
 #endif /* __cplusplus */
 
 #include <sys/types.h>
+@OSPEED_INCLUDES@
 
 #undef  NCURSES_CONST 
 #define NCURSES_CONST @NCURSES_CONST@ 
@@ -52,7 +53,7 @@
 extern char PC;
 extern char *UP;
 extern char *BC;
-extern short ospeed; 
+extern @OSPEED_TYPE@ ospeed; 
 
 #if !defined(_NCU_TERM_H)
 extern char *tgetstr(NCURSES_CONST char *, char **);
--- ncurses-5.2/ncurses/tinfo/lib_baudrate.c.64bit	Fri Nov 17 12:41:42 2000
+++ ncurses-5.2/ncurses/tinfo/lib_baudrate.c	Fri Nov 17 12:43:45 2000
@@ -51,7 +51,7 @@
  */
 
 struct speed {
-    int s;	/* value for 'ospeed' is an index */
+    speed_t s;	/* value for 'ospeed' is an index */
     int sp;	/* the actual speed */
 };
 
@@ -128,7 +128,7 @@
 int
 _nc_ospeed(int BaudRate)
 {
-    int result = 1;
+    speed_t result = 1;
     unsigned i;
 
     if (BaudRate >= 0) {
--- ncurses-5.2/ncurses/tinfo/lib_tputs.c.64bit	Fri Nov 17 12:43:57 2000
+++ ncurses-5.2/ncurses/tinfo/lib_tputs.c	Fri Nov 17 12:44:22 2000
@@ -48,7 +48,7 @@
 MODULE_ID("$Id: lib_tputs.c,v 1.51 2000/10/08 00:22:24 tom Exp $")
 
 char PC = 0;			/* used by termcap library */
-short ospeed = 0;		/* used by termcap library */
+speed_t ospeed = 0;		/* used by termcap library */
 
 int _nc_nulls_sent = 0;		/* used by 'tack' program */
 
--- ncurses-5.2/ncurses/tty/lib_mvcur.c.64bit	Fri Nov 17 12:45:04 2000
+++ ncurses-5.2/ncurses/tty/lib_mvcur.c	Fri Nov 17 12:45:24 2000
@@ -924,7 +924,7 @@
 }
 
 char PC = 0;			/* used by termcap library */
-short ospeed = 0;		/* used by termcap library */
+speed_t ospeed = 0;		/* used by termcap library */
 int _nc_nulls_sent = 0;		/* used by 'tack' program */
 
 int
--- ncurses-5.2/ncurses/llib-lncurses.64bit	Fri Nov 17 12:45:36 2000
+++ ncurses-5.2/ncurses/llib-lncurses	Fri Nov 17 12:46:19 2000
@@ -2400,7 +2400,7 @@
 #include <termcap.h>
 
 struct speed {
-    int s;	
+    speed_t s;	
     int sp;	
 };
 
@@ -2716,7 +2716,7 @@
 #undef PC
 char	PC;
 #undef ospeed
-short	ospeed;
+speed_t	ospeed;
 #undef _nc_nulls_sent
 int	_nc_nulls_sent;
 
--- ncurses-5.2/progs/tset.c.64bit	Fri Nov 17 12:46:37 2000
+++ ncurses-5.2/progs/tset.c	Fri Nov 17 12:46:54 2000
@@ -234,7 +234,7 @@
     const char *porttype;	/* Port type, or "" for any. */
     const char *type;		/* Terminal type to select. */
     int conditional;		/* Baud rate conditionals bitmask. */
-    int speed;			/* Baud rate to compare against. */
+    speed_t speed;		/* Baud rate to compare against. */
 } MAP;
 
 static MAP *cur, *maplist;
--- ncurses-5.2/aclocal.m4.64bit	Fri Nov 17 12:47:32 2000
+++ ncurses-5.2/aclocal.m4	Fri Nov 17 12:50:25 2000
@@ -2123,6 +2123,30 @@
 fi
 ])dnl
 dnl ---------------------------------------------------------------------------
+dnl Check for datatype 'speed_t', which is normally declared via either
+dnl sys/types.h or termios.h
+AC_DEFUN([CF_SPEED_TYPE],
+[
+AC_MSG_CHECKING(for speed_t)
+OSPEED_INCLUDES=
+AC_TRY_COMPILE([#include <sys/types.h>],
+	[speed_t some_variable = 0],
+	[OSPEED_TYPE=speed_t],
+	[OSPEED_TYPE=unsigned])
+AC_TRY_COMPILE([#include <termios.h>],
+	[speed_t some_variable = 0],
+	[OSPEED_TYPE=speed_t
+	 OSPEED_INCLUDES="#include <termios.h>"],[])
+AC_SUBST(OSPEED_TYPE)
+AC_SUBST(OSPEED_INCLUDES)
+if test "$OSPEED_TYPE" = "unsigned" ; then
+	AC_MSG_RESULT(no)
+	AC_DEFINE(speed_t,unsigned)
+else
+	AC_MSG_RESULT(yes)
+fi
+])dnl
+dnl ---------------------------------------------------------------------------
 dnl For each parameter, test if the source-directory exists, and if it contains
 dnl a 'modules' file.  If so, add to the list $cf_cv_src_modules which we'll
 dnl use in CF_LIB_RULES.
--- ncurses-5.2/configure.in.64bit	Fri Nov 17 12:50:31 2000
+++ ncurses-5.2/configure.in	Fri Nov 17 12:50:53 2000
@@ -858,6 +858,7 @@
 ###	Checks for external-data
 CF_ERRNO
 CF_LINK_DATAONLY
+CF_SPEED_TYPE
 
 ###	Checks for library functions.
 AC_CHECK_FUNCS( \
--- ncurses-5.2/configure.64bit	Fri Nov 17 12:51:02 2000
+++ ncurses-5.2/configure	Fri Nov 17 12:55:13 2000
@@ -1646,7 +1646,7 @@
 # sed with no file args requires a program.
 test "$program_transform_name" = "" && program_transform_name="s,x,x,"
 
-for ac_prog in mawk gawk nawk awk
+for ac_prog in gawk mawk nawk awk
 do
 # Extract the first word of "$ac_prog", so it can be a program name with args.
 set dummy $ac_prog; ac_word=$2
@@ -5404,6 +5404,55 @@
 #define BROKEN_LINKER 1
 EOF
 
+echo $ac_n "checking for speed_t""... $ac_c" 1>&6
+echo "configure:5339: checking for speed_t" >&5
+OSPEED_INCLUDES=
+cat > conftest.$ac_ext <<EOF
+#line 5342 "configure"
+#include "confdefs.h"
+#include <sys/types.h>
+int main() {
+speed_t some_variable = 0
+; return 0; }
+EOF
+if { (eval echo configure:5349: \"$ac_compile\") 1>&5; (eval $ac_compile) 2>&5; }; then
+  rm -rf conftest*
+  OSPEED_TYPE=speed_t
+else
+  echo "configure: failed program was:" >&5
+  cat conftest.$ac_ext >&5
+  rm -rf conftest*
+  OSPEED_TYPE=unsigned
+fi
+rm -f conftest*
+cat > conftest.$ac_ext <<EOF
+#line 5360 "configure"
+#include "confdefs.h"
+#include <termios.h>
+int main() {
+speed_t some_variable = 0
+; return 0; }
+EOF
+if { (eval echo configure:5367: \"$ac_compile\") 1>&5; (eval $ac_compile) 2>&5; }; then
+  rm -rf conftest*
+  OSPEED_TYPE=speed_t
+        OSPEED_INCLUDES="#include <termios.h>"
+else
+  echo "configure: failed program was:" >&5
+  cat conftest.$ac_ext >&5
+fi
+rm -f conftest*
+
+if test "$OSPEED_TYPE" = "unsigned" ; then
+	echo "$ac_t""no" 1>&6
+	cat >> confdefs.h <<\EOF
+#define speed_t unsigned
+EOF
+
+else
+	echo "$ac_t""yes" 1>&6
+fi
+
 
 
 ###	Checks for library functions.
@@ -7474,6 +7523,8 @@
 s%@cf_cv_1UL@%$cf_cv_1UL%g
 s%@cf_cv_shift_limit@%$cf_cv_shift_limit%g
 s%@cf_cv_widec_shift@%$cf_cv_widec_shift%g
+s%@OSPEED_TYPE@%$OSPEED_TYPE%g
+s%@OSPEED_INCLUDES@%$OSPEED_INCLUDES%g
 s%@CXXCPP@%$CXXCPP%g
 s%@CXXLDFLAGS@%$CXXLDFLAGS%g
 s%@CXXLIBS@%$CXXLIBS%g
