--- fixpamd.avx	2006-07-08 14:14:52.000000000 -0600
+++ fixpamd	2006-07-08 14:15:12.000000000 -0600
@@ -1,7 +1,7 @@
 #!/bin/sh
 
 if test -z "$DONT_FIX_PAMD_CONFIGS"; then
-    configs=`ls $RPM_BUILD_ROOT/etc/pam.d/* 2> /dev/null`
+    configs=`find $RPM_BUILD_ROOT/etc/pam.d/ -type f 2>/dev/null`
     if [ -n "$configs" ]; then
         echo -n "Fixing pam.d config files..."
         perl -pi -e "s,/(lib|lib64)/security/,," $configs
