--- file-4.02/src/Makefile.am.deps	2003-03-24 15:51:58.000000000 +0100
+++ file-4.02/src/Makefile.am	2003-04-03 13:58:16.000000000 +0200
@@ -12,4 +12,5 @@ libmagic_la_SOURCES = magic.c apprentice
 libmagic_la_LDFLAGS = -version-info 1:0:0
 
 file_SOURCES = file.c
-file_LDFLAGS = -lmagic
+file_LDFLAGS = libmagic.la
+file_DEPENDENCIES = libmagic.la
--- file-4.02/src/Makefile.in.deps	2003-03-28 22:12:42.000000000 +0100
+++ file-4.02/src/Makefile.in	2003-04-03 13:58:31.000000000 +0200
@@ -137,7 +137,7 @@ libmagic_la_SOURCES = magic.c apprentice
 libmagic_la_LDFLAGS = -version-info 1:0:0
 
 file_SOURCES = file.c
-file_LDFLAGS = -lmagic
+file_LDFLAGS = libmagic.la
 subdir = src
 mkinstalldirs = $(SHELL) $(top_srcdir)/mkinstalldirs
 CONFIG_HEADER = $(top_builddir)/config.h
@@ -155,7 +155,7 @@ PROGRAMS = $(bin_PROGRAMS)
 am_file_OBJECTS = file.$(OBJEXT)
 file_OBJECTS = $(am_file_OBJECTS)
 file_LDADD = $(LDADD)
-file_DEPENDENCIES =
+file_DEPENDENCIES = libmagic.la
 
 DEFAULT_INCLUDES =  -I. -I$(srcdir) -I$(top_builddir)
 depcomp = $(SHELL) $(top_srcdir)/depcomp
