--- dietlibc-0.22/libcruft/dnscruft2.c.resolv	2002-11-17 20:06:22.000000000 -0500
+++ dietlibc-0.22/libcruft/dnscruft2.c	2003-02-21 03:49:38.000000000 -0500
@@ -43,16 +43,14 @@ int __dns_gethostbyx_r(const char* name,
   int size;
 
   if (lookfor==1) {
-    result->h_aliases=(char**)(buf+8*4);
     result->h_addrtype=AF_INET;
     result->h_length=4;
-    result->h_addr_list=(char**)buf;
   } else {
-    result->h_aliases=(char**)(buf+8*16);
     result->h_addrtype=AF_INET6;
     result->h_length=16;
-    result->h_addr_list=(char**)buf;
   }
+  result->h_aliases=(char**)(buf+8*sizeof(char*));
+  result->h_addr_list=(char**)buf;
   result->h_aliases[0]=0;
 
   cur=buf+16*sizeof(char*);
@@ -113,7 +111,7 @@ invalidpacket:
 	  else
 	    result->h_aliases[names-1]=cur;
 	  result->h_aliases[names]=0;
-	  ++names;
+	  if (names<8) ++names;
 /*		cur+=slen+1; */
 	  cur+=(slen|3)+1;
 	  result->h_addr_list[ips++] = cur;
--- dietlibc-0.22/test/gethostbyaddr.c.resolv	2003-02-21 03:13:12.000000000 -0500
+++ dietlibc-0.22/test/gethostbyaddr.c	2003-02-21 03:49:45.000000000 -0500
@@ -13,6 +13,8 @@ int main() {
     int i;
     for (i=0; foo->h_addr_list[i]; ++i)
       printf("%s -> %s\n",foo->h_name,inet_ntoa(*(struct in_addr*)foo->h_addr_list[i]));
+    for (i=0; foo->h_aliases[i]; ++i)
+      printf("  also known as %s\n",foo->h_aliases[i]);
   }
   return 0;
 }
--- dietlibc-0.22/test/gethostbyname.c.resolv	2002-09-18 12:44:28.000000000 -0400
+++ dietlibc-0.22/test/gethostbyname.c	2003-02-21 03:49:48.000000000 -0500
@@ -6,8 +6,13 @@ int main() {
   struct in_addr bar;
   struct hostent *foo;
 //  foo=gethostbyname("zeit.fu-berlin.de");
-  foo=gethostbyname("ftp.fu-berlin.de");
-  if (foo)
+  foo=gethostbyname("zeit.fu-berlin.de");
+  if (foo) {
+    int i;
     printf("%s -> %s\n",foo->h_name,inet_ntoa(*(struct in_addr*)foo->h_addr));
+    for (i=0; foo->h_aliases[i]; ++i) {
+      printf("  also known as %s\n",foo->h_aliases[i]);
+    }
+  }
 /*  printf("%g %g\n",1e-10,1e10); */
 }
