--- util-linux-2.11z/MCONFIG.rhconfig	2002-11-25 05:29:48.000000000 -0500
+++ util-linux-2.11z/MCONFIG	2003-04-14 01:24:37.000000000 -0400
@@ -16,17 +16,17 @@ ARCH=$(shell echo $(CPU) | sed 's/i.86/i
 # If HAVE_PAM is set to "yes", then login, chfn, chsh, and newgrp
 # will use PAM for authentication. Additionally, passwd will not be
 # installed as it is not PAM aware.
-HAVE_PAM=no
+HAVE_PAM=yes
 
 # If HAVE_SHADOW is set to "yes", then login, chfn, chsh, newgrp, passwd,
 # and vipw will not be built or installed from the login-utils
 # subdirectory.  
-HAVE_SHADOW=yes
+HAVE_SHADOW=no
 
 # If HAVE_PASSWD is set to "yes", then passwd will not be built or
 # installed from the login-utils subdirectory (but login, chfn, chsh,
 # newgrp, and vipw *will* be installed).
-HAVE_PASSWD=no
+HAVE_PASSWD=yes
 
 # If you use chfn and chsh from this package, REQUIRE_PASSWORD will require
 # non-root users to enter the account password before updating /etc/passwd.
@@ -79,7 +79,7 @@ HAVE_KILL=no
 # to the current user, allowing her to make a screendump and do other
 # nifty things on the console, but also allowing him to keep an open
 # file descriptor after logging out to trick the next user.
-ALLOW_VCS_USE=yes
+ALLOW_VCS_USE=no
 
 # If HAVE_RESET is set to "yes", then reset won't be installed.  The version
 # of reset that comes with the ncurses package is less aggressive.
@@ -102,6 +102,9 @@ SILENT_PG=no
 # DISABLE_NLS to "yes".
 DISABLE_NLS=no
 
+# If ADD_RAW is set to "yes", then raw gets built
+ADD_RAW=yes
+
 # Different optimizations for different cpus.
 # gcc 3.0 likes options -mcpu=i486 instead of -m486
 ifeq "$(ARCH)" "intel"
@@ -133,9 +136,9 @@ ERR_O=$(LIB)/err.o
 
 # Set HAVE_SLANG to yes if you have slang (and prefer to use that for cfdisk)
 # (If neither HAVE_NCURSES nor HAVE_SLANG is defined, cfdisk is not made.)
-# HAVE_SLANG=yes
+HAVE_SLANG=yes
 # There is a subdirectory /usr/include/slang containing slcurses.h
-# SLANGFLAGS=-I/usr/include/slang
+SLANGFLAGS=-I/usr/include/slang
 # No such subdirectory - slcurses.h lives in /usr/include
 #   (no extra definition required).
 LIBSLANG=-lslang
@@ -198,7 +201,7 @@ INSTALLDIR=	$(INSTALL) -d -m $(DIRMODE)
 INSTALLBIN=	$(INSTALL) -m $(BINMODE)
 INSTALLMAN=	$(INSTALL) -m $(MANMODE)
 INSTALLDAT=	$(INSTALL) -m $(DATMODE)
-INSTALLSUID=    $(INSTALL) -m $(SUIDMODE) -o root
+INSTALLSUID=    $(INSTALL) -m $(SUIDMODE)
 
 ifeq "$(DISABLE_NLS)" "yes"
 NLSFLAGS = -DDISABLE_NLS
