--- ./src/Makefile.am.nolibrt	2002-09-22 00:40:35.000000000 +0200
+++ ./src/Makefile.am	2002-10-08 17:03:54.000000000 +0200
@@ -35,10 +35,10 @@
 # replacement functions defined in libfetish.a.
 LDADD = ../lib/libfetish.a @LIBINTL@ ../lib/libfetish.a
 
-dir_LDADD = $(LDADD) @LIB_CLOCK_GETTIME@
-ls_LDADD = $(LDADD) @LIB_CLOCK_GETTIME@
+dir_LDADD = $(LDADD)
+ls_LDADD = $(LDADD)
 shred_LDADD = $(LDADD) @LIB_CLOCK_GETTIME@
-vdir_LDADD = $(LDADD) @LIB_CLOCK_GETTIME@
+vdir_LDADD = $(LDADD)
 
 ## If necessary, add -lm to resolve use of pow in lib/strtod.c.
 sort_LDADD = $(LDADD) @POW_LIB@
--- ./src/Makefile.in.nolibrt	2002-10-08 17:03:54.000000000 +0200
+++ ./src/Makefile.in	2002-10-08 17:06:04.000000000 +0200
@@ -189,10 +189,10 @@
 # replacement functions defined in libfetish.a.
 LDADD = ../lib/libfetish.a @LIBINTL@ ../lib/libfetish.a
 
-dir_LDADD = $(LDADD) @LIB_CLOCK_GETTIME@ -ltermcap
-ls_LDADD = $(LDADD) @LIB_CLOCK_GETTIME@ -ltermcap
-shred_LDADD = $(LDADD) @LIB_CLOCK_GETTIME@
-vdir_LDADD = $(LDADD) @LIB_CLOCK_GETTIME@ -ltermcap
+dir_LDADD = $(LDADD) -ltermcap
+ls_LDADD = $(LDADD) -ltermcap
+shred_LDADD = $(LDADD)
+vdir_LDADD = $(LDADD) -ltermcap
 
 sort_LDADD = $(LDADD) @POW_LIB@
 
--- ./src/ls.c.nolibrt	2002-10-08 17:03:54.000000000 +0200
+++ ./src/ls.c	2002-10-08 17:03:55.000000000 +0200
@@ -2742,7 +2742,7 @@
 static void
 get_current_time (void)
 {
-#if HAVE_CLOCK_GETTIME && defined CLOCK_REALTIME
+#if 0 && HAVE_CLOCK_GETTIME && defined CLOCK_REALTIME
   {
     struct timespec timespec;
     if (clock_gettime (CLOCK_REALTIME, &timespec) == 0)
--- ./src/shred.c.nolibrt	2002-10-08 17:03:54.000000000 +0200
+++ ./src/shred.c	2002-10-08 17:03:55.000000000 +0200
@@ -626,7 +626,7 @@
     hrtime_t t = gethrtime ();
     ISAAC_SEED (s, t);
 #else
-# if HAVE_CLOCK_GETTIME && defined(CLOCK_REALTIME) /* POSIX ns-resolution */
+# if 0 && HAVE_CLOCK_GETTIME && defined(CLOCK_REALTIME) /* POSIX ns-resolution */
     struct timespec t;
     clock_gettime (CLOCK_REALTIME, &t);
 # else
