--- texinfo-4.11/util/defs.in.orig	2007-11-11 16:33:54.000000000 +0100
+++ texinfo-4.11/util/defs.in	2007-11-11 16:52:26.000000000 +0100
@@ -198,8 +198,7 @@
     set x --debug --verbose ${1+"$@"}
     shift
   fi
-  which texi2dvi
-  texi2dvi ${1+"$@"} 2>stderr >stdout || exitcode=$?
+  $TEXI2DVI ${1+"$@"} 2>stderr >stdout || exitcode=$?
   cat stderr
   cat stdout
   test $exitcode = $expected_exitcode || exit 1
