--- autofs-5.0.2/daemon/automount.c.avx	2007-11-10 14:16:42.825498000 -0700
+++ autofs-5.0.2/daemon/automount.c	2007-11-10 14:17:37.016884750 -0700
@@ -789,7 +789,7 @@
 {
 	FILE *pidfp;
 	char buf[MAX_ERR_BUF];
-	unsigned to_stderr = 0;
+	unsigned to_stderr = 1;
 	pid_t pid;
 
 	/* Don't BUSY any directories unnecessarily */
@@ -1370,7 +1370,7 @@
 		"	-v --verbose	be verbose\n"
 		"	-d --debug	log debuging info\n"
 		"	-D --define	define global macro variable\n"
-		/*"	-f --foreground do not fork into background\n" */
+		"	-f --foreground do not fork into background\n"
 		"	-r --random-multimount-selection\n"
 		"			use ramdom replicated server selection\n"
 		"	-O --global-options\n"
