--- audit-1.6.1/system-config-audit/src/server.c~	2007-09-24 18:41:56.000000000 -0300
+++ audit-1.6.1/system-config-audit/src/server.c	2007-09-24 18:59:14.000000000 -0300
@@ -200,7 +200,7 @@
       err = errno;
       goto err_fd;
     }
-  if (st.st_size > (off_t)SIZE_MAX)
+  if (st.st_size > SIZE_MAX)
     {
             printf("size: %lu, SIZE_MAX: %lu\n", st.st_size, SIZE_MAX);
       err = EFBIG;
