--- itcl3.2.1/itcl/Makefile.in.destdir	2001-05-29 17:51:16.000000000 +0200
+++ itcl3.2.1/itcl/Makefile.in	2004-12-03 12:49:05.738006408 +0100
@@ -269,16 +269,16 @@
 #========================================================================
 
 install-libraries: libraries
-	$(mkinstalldirs) $(includedir)
+	$(mkinstalldirs) $(DESTDIR)$(includedir)
 	@echo "Installing header files in $(includedir)"
 	@for i in $(GENERIC_HDRS) ; do \
 	    echo "Installing $$i" ; \
-	    $(INSTALL_DATA) $$i $(includedir) ; \
+	    $(INSTALL_DATA) $$i $(DESTDIR)$(includedir) ; \
 	done;
 	@echo "Installing library files in $(ITCL_LIBRARY)"
 	@for i in $(srcdir)/library/*.tcl ; do \
 	    echo "Installing $$i" ; \
-	    $(INSTALL_DATA) $$i $(ITCL_LIBRARY) ; \
+	    $(INSTALL_DATA) $$i $(DESTDIR)$(ITCL_LIBRARY) ; \
 	done;
 
 #========================================================================
@@ -287,13 +287,13 @@
 #========================================================================
 
 install-doc: doc
-	$(mkinstalldirs) $(mandir)/mann
+	$(mkinstalldirs) $(DESTDIR)$(mandir)/mann
 	@echo "Installing man pages in $(mandir)"
 	@for i in $(srcdir)/doc/*.n; \
 	    do \
 	    echo "Installing $$i"; \
 	    rm -f $(mandir)/mann/`basename $$i`; \
-	    $(INSTALL_DATA) $$i $(mandir)/mann ; \
+	    $(INSTALL_DATA) $$i $(DESTDIR)$(mandir)/mann ; \
 	    done
 
 test: $(TCLSH_PROG)
--- itcl3.2.1/itk/Makefile.in.destdir	2001-05-29 17:51:16.000000000 +0200
+++ itcl3.2.1/itk/Makefile.in	2004-12-03 12:47:16.744575928 +0100
@@ -264,16 +264,16 @@
 #========================================================================
 
 install-libraries: libraries
-	$(mkinstalldirs) $(includedir)
+	$(mkinstalldirs) $(DESTDIR)$(includedir)
 	@echo "Installing header files in $(includedir)"
 	@for i in $(GENERIC_HDRS) ; do \
 	    echo "Installing $$i" ; \
-	    $(INSTALL_DATA) $$i $(includedir) ; \
+	    $(INSTALL_DATA) $$i $(DESTDIR)$(includedir) ; \
 	done;
 	@echo "Installing library files in $(ITK_LIBRARY)"
 	@for i in $(srcdir)/library/*.* $(srcdir)/library/tclIndex ; do \
 	    echo "Installing $$i" ; \
-	    $(INSTALL_DATA) $$i $(ITK_LIBRARY) ; \
+	    $(INSTALL_DATA) $$i $(DESTDIR)$(ITK_LIBRARY) ; \
 	done;
 
 #========================================================================
@@ -282,13 +282,13 @@
 #========================================================================
 
 install-doc: doc
-	$(mkinstalldirs) $(mandir)/mann
+	$(mkinstalldirs) $(DESTDIR)$(mandir)/mann
 	@echo "Installing man pages in $(mandir)"
 	@for i in $(srcdir)/doc/*.n; \
 	    do \
 	    echo "Installing $$i info $(mandir)/mann"; \
 	    rm -f $(mandir)/mann/`basename $$i`; \
-	    $(INSTALL_DATA) $$i $(mandir)/mann ; \
+	    $(INSTALL_DATA) $$i $(DESTDIR)$(mandir)/mann ; \
 	    done
 
 test: $(TCLSH_PROG)
