--- net-tools-1.60/lib/x25_sr.c.x25	2000-05-20 15:38:10.000000000 +0200
+++ net-tools-1.60/lib/x25_sr.c	2004-03-05 10:32:32.000000000 +0100
@@ -77,7 +77,7 @@ static int X25_setroute(int action, int 
   rt.sigdigits=sigdigits;
 
   /* x25_route_struct.address isn't type struct sockaddr_x25, Why? */
-  memcpy(&rt.address, &sx25.sx25_addr, sizeof(x25_address));
+  memcpy(&rt.address, &sx25.sx25_addr, sizeof(struct x25_address));
 
   while (*args) {
 	if (!strcmp(*args,"device") || !strcmp(*args,"dev")) {
