--- XFree86-4.0.2/xc/config/cf/xfree86.cf--	Tue Jan  9 11:16:29 2001
+++ XFree86-4.0.2/xc/config/cf/xfree86.cf	Tue Jan  9 11:16:59 2001
@@ -521,7 +521,7 @@
 /* Pure PCI drivers should go first */
 # ifndef XF86CardDrivers
 #  define XF86CardDrivers	mga glint s3virge sis savage\
-				trident chips fbdev \
+				trident chips fbdev ati \
 				DevelDrivers vga \
 				XF86OSCardDrivers XF86ExtraCardDrivers
 # endif
