--- XFree86-4.3/xc/config/cf/X11.tmpl.branch-4.3	2003-02-25 16:57:52.000000000 -0500
+++ XFree86-4.3/xc/config/cf/X11.tmpl	2003-07-08 00:10:39.000000000 -0400
@@ -2181,14 +2181,14 @@
 #define GLwUseXmStubs		NO	/* create stub (weak) Motif symbols */
 #endif
 #else
-#undef  SharedLibGlw
-#define SharedLibGlw		NO
-#undef  NormalLibGlw
-#define NormalLibGlw		NO
-#undef  DebugLibGlw
-#define DebugLibGlw		NO
-#undef  ProfileLibGlw
-#define ProfileLibGlw		NO
+#undef  SharedLibGLw
+#define SharedLibGLw		NO
+#undef  NormalLibGLw
+#define NormalLibGLw		NO
+#undef  DebugLibGLw
+#define DebugLibGLw		NO
+#undef  ProfileLibGLw
+#define ProfileLibGLw		NO
 #endif
 
 #ifndef SharedLibXext
--- XFree86-4.3/xc/lib/GL/mesa/src/drv/i830/Imakefile.branch-4.3	2003-07-08 00:08:05.000000000 -0400
+++ XFree86-4.3/xc/lib/GL/mesa/src/drv/i830/Imakefile	2003-07-08 00:10:39.000000000 -0400
@@ -87,3 +87,15 @@
 InstallDriverSDKNonExecFile(i830_tris.h,$(DRIVERSDKDIR)/lib/GL/mesa/src/drv/i830)
 InstallDriverSDKNonExecFile(i830_vb.h,$(DRIVERSDKDIR)/lib/GL/mesa/src/drv/i830)
 
+InstallDriverSDKNonExecFile(Imakefile,$(DRIVERSDKDIR)/lib/GL/mesa/src/drv/i830)
+InstallDriverSDKNonExecFile(i830_3d_reg.h,$(DRIVERSDKDIR)/lib/GL/mesa/src/drv/i830)
+InstallDriverSDKNonExecFile(i830_context.h,$(DRIVERSDKDIR)/lib/GL/mesa/src/drv/i830)
+InstallDriverSDKNonExecFile(i830_debug.h,$(DRIVERSDKDIR)/lib/GL/mesa/src/drv/i830)
+InstallDriverSDKNonExecFile(i830_ioctl.h,$(DRIVERSDKDIR)/lib/GL/mesa/src/drv/i830)
+InstallDriverSDKNonExecFile(i830_screen.h,$(DRIVERSDKDIR)/lib/GL/mesa/src/drv/i830)
+InstallDriverSDKNonExecFile(i830_span.h,$(DRIVERSDKDIR)/lib/GL/mesa/src/drv/i830)
+InstallDriverSDKNonExecFile(i830_state.h,$(DRIVERSDKDIR)/lib/GL/mesa/src/drv/i830)
+InstallDriverSDKNonExecFile(i830_tex.h,$(DRIVERSDKDIR)/lib/GL/mesa/src/drv/i830)
+InstallDriverSDKNonExecFile(i830_tris.h,$(DRIVERSDKDIR)/lib/GL/mesa/src/drv/i830)
+InstallDriverSDKNonExecFile(i830_vb.h,$(DRIVERSDKDIR)/lib/GL/mesa/src/drv/i830)
+
--- XFree86-4.3/xc/lib/GL/mesa/src/drv/r200/r200_swtcl.c.branch-4.3	2003-07-08 00:08:15.000000000 -0400
+++ XFree86-4.3/xc/lib/GL/mesa/src/drv/r200/r200_swtcl.c	2003-07-08 00:10:39.000000000 -0400
@@ -1006,6 +1006,14 @@
 }
 
 
+void r200FlushVertices( GLcontext *ctx, GLuint flags )
+{
+   _tnl_flush_vertices( ctx, flags );
+
+   if (flags & FLUSH_STORED_VERTICES) 
+      R200_FIREVERTICES( R200_CONTEXT( ctx ) );
+}
+
 /**********************************************************************/
 /*           Transition to/from hardware rasterization.               */
 /**********************************************************************/
--- XFree86-4.3/xc/lib/GL/mesa/src/drv/radeon/radeon_context.c.branch-4.3	2003-02-08 16:26:45.000000000 -0500
+++ XFree86-4.3/xc/lib/GL/mesa/src/drv/radeon/radeon_context.c	2003-07-08 00:10:39.000000000 -0400
@@ -382,6 +382,7 @@
     */
    _tnl_destroy_pipeline( ctx );
    _tnl_install_pipeline( ctx, radeon_pipeline );
+   ctx->Driver.FlushVertices = radeonFlushVertices;
 
    /* Try and keep materials and vertices separate:
     */
--- XFree86-4.3/xc/lib/GL/mesa/src/drv/radeon/radeon_swtcl.c.branch-4.3	2003-02-15 17:18:48.000000000 -0500
+++ XFree86-4.3/xc/lib/GL/mesa/src/drv/radeon/radeon_swtcl.c	2003-07-08 00:10:39.000000000 -0400
@@ -45,6 +45,7 @@
 #include "math/m_translate.h"
 #include "tnl/tnl.h"
 #include "tnl/t_context.h"
+#include "tnl/t_imm_exec.h"
 #include "tnl/t_pipeline.h"
 
 #include "radeon_context.h"
@@ -1134,6 +1135,14 @@
 }
 
 
+void radeonFlushVertices( GLcontext *ctx, GLuint flags )
+{
+   _tnl_flush_vertices( ctx, flags );
+
+   if (flags & FLUSH_STORED_VERTICES) 
+      RADEON_FIREVERTICES( RADEON_CONTEXT( ctx ) );
+}
+
 /**********************************************************************/
 /*                            Initialization.                         */
 /**********************************************************************/
--- XFree86-4.3/xc/lib/GL/mesa/src/drv/radeon/radeon_swtcl.h.branch-4.3	2002-10-30 07:51:57.000000000 -0500
+++ XFree86-4.3/xc/lib/GL/mesa/src/drv/radeon/radeon_swtcl.h	2003-07-08 00:10:39.000000000 -0400
@@ -43,6 +43,7 @@
 extern void radeonInitSwtcl( GLcontext *ctx );
 extern void radeonDestroySwtcl( GLcontext *ctx );
 
+extern void radeonFlushVertices( GLcontext *ctx, GLuint flags );
 extern void radeonChooseRenderState( GLcontext *ctx );
 extern void radeonChooseVertexState( GLcontext *ctx );
 
--- XFree86-4.3/xc/lib/GL/mesa/src/drv/radeon/radeon_vtxfmt.c.branch-4.3	2002-12-16 11:18:59.000000000 -0500
+++ XFree86-4.3/xc/lib/GL/mesa/src/drv/radeon/radeon_vtxfmt.c	2003-07-08 00:10:39.000000000 -0400
@@ -38,6 +38,7 @@
 #include "radeon_ioctl.h"
 #include "radeon_tex.h"
 #include "radeon_tcl.h"
+#include "radeon_swtcl.h"
 #include "radeon_vtxfmt.h"
 
 #include "api_noop.h"
@@ -59,7 +60,7 @@
 
 struct radeon_vb vb;
 
-static void radeonFlushVertices( GLcontext *, GLuint );
+static void radeonVtxfmtFlushVertices( GLcontext *, GLuint );
 
 static void count_func( const char *name,  struct dynfn *l )
 {
@@ -336,12 +337,13 @@
       fprintf(stderr, "%s from %s\n", __FUNCTION__, caller);
 
    if (ctx->Driver.NeedFlush) 
-      radeonFlushVertices( ctx, ctx->Driver.NeedFlush );
+      radeonVtxfmtFlushVertices( ctx, ctx->Driver.NeedFlush );
 
    if (ctx->NewState)
       _mesa_update_state( ctx ); /* clear state so fell_back sticks */
 
    _tnl_wakeup_exec( ctx );
+   ctx->Driver.FlushVertices = radeonFlushVertices;
 
    assert( rmesa->dma.flush == 0 );
    rmesa->vb.fell_back = GL_TRUE;
@@ -382,6 +384,7 @@
    prim = rmesa->vb.prim[0];
    ctx->Driver.CurrentExecPrimitive = GL_POLYGON+1;
    _tnl_wakeup_exec( ctx );
+   ctx->Driver.FlushVertices = radeonFlushVertices;
 
    assert(rmesa->dma.flush == 0);
    rmesa->vb.fell_back = GL_TRUE;
@@ -731,7 +734,7 @@
 	    fprintf(stderr, "reinstall (new install)\n");
 
 	 _mesa_install_exec_vtxfmt( ctx, &rmesa->vb.vtxfmt );
-	 ctx->Driver.FlushVertices = radeonFlushVertices;
+	 ctx->Driver.FlushVertices = radeonVtxfmtFlushVertices;
 	 ctx->Driver.NewList = radeonNewList;
 	 rmesa->vb.installed = GL_TRUE;
 	 vb.context = ctx;
@@ -747,6 +750,7 @@
 	 if (rmesa->dma.flush)
 	    rmesa->dma.flush( rmesa );
 	 _tnl_wakeup_exec( ctx );
+	 ctx->Driver.FlushVertices = radeonFlushVertices;
 	 rmesa->vb.installed = GL_FALSE;
 	 vb.context = 0;
       }
@@ -905,7 +909,7 @@
    return GL_TRUE;
 }
 
-static void radeonFlushVertices( GLcontext *ctx, GLuint flags )
+static void radeonVtxfmtFlushVertices( GLcontext *ctx, GLuint flags )
 {
    radeonContextPtr rmesa = RADEON_CONTEXT( ctx );
 
--- XFree86-4.3/xc/lib/GLU/libnurbs/internals/Imakefile.branch-4.3	2002-12-13 23:41:12.000000000 -0500
+++ XFree86-4.3/xc/lib/GLU/libnurbs/internals/Imakefile	2003-07-08 00:10:40.000000000 -0400
@@ -149,6 +149,8 @@
 # if OSMajorVersion <= 5
 OSDEFINES = -DNEEDCEILF
 # endif
+#elif defined(LinuxArchitecture) && (LinuxCLibMajorVersion < 6)
+OSDEFINES = -DNEEDCEILF
 #else
 OSDEFINES = -D_EXTENSIONS_
 #endif
--- XFree86-4.3/xc/lib/X11/XlcDL.c.branch-4.3	2002-11-25 09:04:53.000000000 -0500
+++ XFree86-4.3/xc/lib/X11/XlcDL.c	2003-07-08 00:10:40.000000000 -0400
@@ -406,7 +406,7 @@
 
     if (lc_name == NULL) return (XLCd)NULL;
 
-    if (_XlcLocaleDirName(lc_dir, (char *)lc_name) == (char*)NULL)
+    if (_XlcLocaleDirName(lc_dir, BUFSIZE, (char *)lc_name) == (char*)NULL)
 	return (XLCd)NULL;
 
     resolve_object(lc_dir, lc_name);
@@ -452,7 +452,7 @@
 
   lc_name = lcd->core->name;
 
-  if (_XlcLocaleDirName(lc_dir, lc_name) == NULL) return (XIM)0;
+  if (_XlcLocaleDirName(lc_dir, BUFSIZE, lc_name) == NULL) return (XIM)0;
 
   count = lc_count;
   for (; count-- > 0; objects_list++) {
@@ -498,7 +498,7 @@
 
   lc_name = lcd->core->name;
 
-  if (_XlcLocaleDirName(lc_dir, lc_name) == NULL) return False;
+  if (_XlcLocaleDirName(lc_dir, BUFSIZE, lc_name) == NULL) return False;
 
   count = lc_count;
   for (; count-- > 0; objects_list++) {
@@ -543,7 +543,7 @@
 #endif
 
   lc_name = lcd->core->name;
-  if (_XlcLocaleDirName(lc_dir, lc_name) == NULL) return False;
+  if (_XlcLocaleDirName(lc_dir, BUFSIZE, lc_name) == NULL) return False;
 
   count = lc_count;
   for (; count-- > 0; objects_list++) {
@@ -610,7 +610,7 @@
 
   lc_name = lcd->core->name;
 
-  if (_XlcLocaleDirName(lc_dir, lc_name) == NULL) return (XOM)0;
+  if (_XlcLocaleDirName(lc_dir, BUFSIZE, lc_name) == NULL) return (XOM)0;
 
   count = lc_count;
   for (; count-- > 0; objects_list++) {
--- XFree86-4.3/xc/lib/X11/XlcPubI.h.branch-4.3	2001-11-15 19:52:27.000000000 -0500
+++ XFree86-4.3/xc/lib/X11/XlcPubI.h	2003-07-08 00:10:40.000000000 -0400
@@ -217,6 +217,7 @@
 extern char *_XlcLocaleDirName(
 #if NeedFunctionPrototypes
      char*             /* dir_name */,
+     size_t,	       /* dir_len */
      char*             /* lc_name */
 #endif
 );
--- XFree86-4.3/xc/lib/X11/lcFile.c.branch-4.3	2002-11-25 09:04:53.000000000 -0500
+++ XFree86-4.3/xc/lib/X11/lcFile.c	2003-07-08 00:10:40.000000000 -0400
@@ -429,8 +429,9 @@
 }
 
 char *
-_XlcLocaleDirName(dir_name, lc_name)
+_XlcLocaleDirName(dir_name, dir_len, lc_name)
      char *dir_name;
+     size_t dir_len;
      char *lc_name;
 {
     char dir[PATH_MAX], buf[PATH_MAX], *name = NULL;
@@ -486,9 +487,16 @@
  	target_dir = args[0];
  	target_name = lc_name;
     }
-    strcpy(dir_name, target_dir);
-    strcat(dir_name, "/");
-    strcat(dir_name, target_name);
+    /* snprintf(dir_name, dir_len, "%s/%", target_dir, target_name); */
+    strncpy(dir_name, target_dir, dir_len - 1);
+    if (strlen(target_dir) >= dir_len - 1) {
+	dir_name[dir_len - 1] = '\0';
+    } else  {
+	strcat(dir_name, "/");
+	strncat(dir_name, target_name, dir_len - strlen(dir_name) - 1);
+	if (strlen(target_name) >= dir_len - strlen(dir_name) - 1) 
+	    dir_name[dir_len - 1] = '\0';
+    }
     if (target_name != lc_name)
  	Xfree(target_name);
     return dir_name;
--- XFree86-4.3/xc/lib/font/FreeType/fttools.c.branch-4.3	2003-02-25 16:36:54.000000000 -0500
+++ XFree86-4.3/xc/lib/font/FreeType/fttools.c	2003-07-08 00:10:40.000000000 -0400
@@ -51,7 +51,9 @@
 {
     if(rc == 0x40)
         return AllocError;
-    else return BadFontFormat;
+    /* Anything else stops the font matching mechanism */
+    else return BadFontName;
+
 }
 
 /* Convert slen bytes from UCS-2 to ISO 8859-1.  Byte specifies the
@@ -132,7 +134,7 @@
         len = name.string_len;
         if(len > name_len)
             len = name_len;
-        memcpy(name_return, name.string, name_len);
+        memcpy(name_return, name.string, len);
         return len;
     }
 
--- XFree86-4.3/xc/lib/font/fontfile/fontdir.c.branch-4.3	2002-05-31 14:45:50.000000000 -0400
+++ XFree86-4.3/xc/lib/font/fontfile/fontdir.c	2003-07-08 00:10:40.000000000 -0400
@@ -728,10 +728,13 @@
 		    if (!(existing->u.scalable.fileName = FontFileSaveString (fileName)))
 			return FALSE;
 		}
-		FontFileCompleteXLFD(&vals, &vals);
-		FontFileAddScaledInstance (existing, &vals, NullFont,
-					   bitmap->name.name);
-		return TRUE;
+                if(bitmap)
+                {
+                    FontFileCompleteXLFD(&vals, &vals);
+                    FontFileAddScaledInstance (existing, &vals, NullFont,
+                                               bitmap->name.name);
+                    return TRUE;
+                }
 	    }
 	}
 	if (!(entry.u.scalable.fileName = FontFileSaveString (fileName)))
@@ -795,9 +798,12 @@
 	}
 	if (vals.values_supplied & SIZE_SPECIFY_MASK)
 	{
-	    FontFileCompleteXLFD(&vals, &vals);
-	    FontFileAddScaledInstance (scalable, &vals, NullFont,
-				       bitmap->name.name);
+            if(bitmap)
+            {
+                FontFileCompleteXLFD(&vals, &vals);
+                FontFileAddScaledInstance (scalable, &vals, NullFont,
+                                           bitmap->name.name);
+            }
 	}
       }
     }
--- XFree86-4.3/xc/programs/Xserver/Xext/Imakefile.branch-4.3	2002-03-06 16:12:32.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/Xext/Imakefile	2003-07-08 00:10:40.000000000 -0400
@@ -170,3 +170,4 @@
 
 InstallDriverSDKNonExecFile(dgaproc.h,$(DRIVERSDKINCLUDEDIR))
 InstallDriverSDKNonExecFile(xvdix.h,$(DRIVERSDKINCLUDEDIR))
+InstallDriverSDKNonExecFile(xvmcext.h,$(DRIVERSDKINCLUDEDIR))
--- XFree86-4.3/xc/programs/Xserver/hw/darwin/quartz/XDarwin.pbproj/project.pbxproj.branch-4.3	2003-01-28 20:11:06.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/darwin/quartz/XDarwin.pbproj/project.pbxproj	2003-07-08 00:10:40.000000000 -0400
@@ -371,7 +371,7 @@
 	<key>CFBundleExecutable</key>
 	<string>XDarwin</string>
 	<key>CFBundleGetInfoString</key>
-	<string>XDarwin 1.2b4, Â©2001-2003 XFree86 Project, Inc.</string>
+	<string>XDarwin 1.2.0, Â©2001-2003 XFree86 Project, Inc.</string>
 	<key>CFBundleIconFile</key>
 	<string>XDarwin.icns</string>
 	<key>CFBundleIdentifier</key>
@@ -383,7 +383,7 @@
 	<key>CFBundlePackageType</key>
 	<string>APPL</string>
 	<key>CFBundleShortVersionString</key>
-	<string>XDarwin 1.2b4</string>
+	<string>XDarwin 1.2.0</string>
 	<key>CFBundleSignature</key>
 	<string>????</string>
 	<key>CFBundleVersion</key>
--- XFree86-4.3/xc/programs/Xserver/hw/darwin/Imakefile.branch-4.3	2002-10-11 20:32:43.000000000 -0400
+++ XFree86-4.3/xc/programs/Xserver/hw/darwin/Imakefile	2003-07-08 00:10:40.000000000 -0400
@@ -19,7 +19,8 @@
 
 INCLUDES = -I. -I$(SERVERSRC)/mi -I$(SERVERSRC)/fb -I$(EXTINCSRC) \
            -I$(SERVERSRC)/render -I$(SERVERSRC)/include -I$(XINCLUDESRC) \
-           -I$(SERVERSRC)/os -I$(INCLUDESRC) -I$(FONTINCSRC)
+           -I$(SERVERSRC)/os -I$(INCLUDESRC) -I$(FONTINCSRC) \
+           -I$(SERVERSRC)/hw/xfree86
 
 OSNAME = OSName
 OSVENDOR = OSVendor
@@ -33,6 +34,10 @@
 BUILDERMSG = -DBUILDERSTRING='$(BUILDERSTRING)'
 #endif
 
+#if OSMajorVersion >= 6
+CLUTDEF = -DUSE_NEW_CLUT
+#endif
+
 #if DarwinQuartzSupport
 # if HasNSCarbonWindow
 SUBDIRS = bundle quartz utils
@@ -47,7 +52,8 @@
 DEFINES = $(QUARTZDEF)
 EXTRAMANDEFS = $(QUARTZDEF) -D__logdir__=$(LOGDIRECTORY)
 
-SpecialCObjectRule(darwin,$(ICONFIGFILES),$(OSNAMEDEF) $(BUILDERMSG) $(CUSTOMVERDEF))
+SpecialCObjectRule(darwin,$(ICONFIGFILES),$(OSNAMEDEF) $(BUILDERMSG) \
+                   $(CUSTOMVERDEF) $(CLUTDEF))
 
 NormalLibraryObjectRule()
 NormalLibraryTarget(darwin,$(OBJS))
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/common/Imakefile.branch-4.3	2003-02-17 12:06:41.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/common/Imakefile	2003-07-08 00:10:41.000000000 -0400
@@ -288,3 +288,5 @@
 InstallDriverSDKNonExecFile(xf86str.h,$(DRIVERSDKINCLUDEDIR))
 InstallDriverSDKNonExecFile(xf86xv.h,$(DRIVERSDKINCLUDEDIR))
 InstallDriverSDKNonExecFile(xf86xvmc.h,$(DRIVERSDKINCLUDEDIR))
+InstallDriverSDKNonExecFile(xf86Bus.h,$(DRIVERSDKINCLUDEDIR))
+InstallDriverSDKNonExecFile(xf86pciBus.h,$(DRIVERSDKINCLUDEDIR))
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/common/xf86Mode.c.branch-4.3	2003-01-22 16:44:09.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/common/xf86Mode.c	2003-07-08 00:10:41.000000000 -0400
@@ -1639,7 +1639,7 @@
 	/*
 	 * Don't let non-user defined modes increase the virtual size
 	 */
-	if (!(p->type & M_T_USERDEF)) {
+	if (!(p->type & M_T_USERDEF) && (numModes > 0)) {
 	    if (p->HDisplay > virtX) {
 		p->status = MODE_VIRTUAL_X;
 		goto lookupNext;
@@ -1732,16 +1732,17 @@
     scrp->virtualX = virtX;
     scrp->virtualY = virtY;
     scrp->displayWidth = linePitch;
+
+    if (numModes <= 0)
+	return 0;
     
     /* Make the mode list into a circular list by joining up the ends */
-    if (numModes > 0) {
-	p = scrp->modes;
-	while (p->next != NULL)
-	    p = p->next;
-	/* p is now the last mode on the list */
-	p->next = scrp->modes;
-	scrp->modes->prev = p;
-    }
+    p = scrp->modes;
+    while (p->next != NULL)
+	p = p->next;
+    /* p is now the last mode on the list */
+    p->next = scrp->modes;
+    scrp->modes->prev = p;
 
     if (minHeight > 0 && virtY < minHeight) {
 	xf86DrvMsg(scrp->scrnIndex, X_ERROR,
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/common/xf86pciBus.c.branch-4.3	2003-02-18 10:42:11.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/common/xf86pciBus.c	2003-07-08 00:10:41.000000000 -0400
@@ -741,7 +741,8 @@
 savePciBusState(BusAccPtr ptr)
 {
     ptr->busdep.pci.save.control =
-	pciReadWord(ptr->busdep.pci.acc, PCI_PCI_BRIDGE_CONTROL_REG);
+	pciReadWord(ptr->busdep.pci.acc, PCI_PCI_BRIDGE_CONTROL_REG) &
+	~PCI_PCI_BRIDGE_SECONDARY_RESET;
     /* Allow master aborts to complete normally on non-root buses */
     if (ptr->busdep.pci.save.control & PCI_PCI_BRIDGE_MASTER_ABORT_EN)
 	pciWriteWord(ptr->busdep.pci.acc, PCI_PCI_BRIDGE_CONTROL_REG,
@@ -2164,8 +2165,7 @@
 		}
 		*pnPciBus = PciBus = xnfcalloc(1, sizeof(PciBusRec));
 		pnPciBus = &PciBus->next;
-		PciBus->primary = -1;
-		PciBus->secondary = i;
+		PciBus->primary = PciBus->secondary = i;
 		PciBus->subclass = PCI_SUBCLASS_BRIDGE_HOST;
 		PciBus->brcontrol = PCI_PCI_BRIDGE_VGA_EN;
 		PciBus->preferred_io =
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/ati/Imakefile.branch-4.3	2003-02-17 12:06:41.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/ati/Imakefile	2003-07-08 00:10:42.000000000 -0400
@@ -207,6 +207,7 @@
 InstallObjectModule(r128,$(MODULEDIR),drivers)
 InstallObjectModule(radeon,$(MODULEDIR),drivers)
 
+#if !defined(XF86DriverSDK)
 #if 0
 InstallModuleManPage(ati)
 #endif
@@ -214,6 +215,7 @@
 InstallModuleManPage(r128)
 
 InstallModuleManPage(radeon)
+#endif
 
 DependTarget()
 
@@ -332,6 +334,7 @@
 InstallDriverSDKNonExecFile(radeon_version.h,$(DRIVERSDKDIR)/drivers/ati)
 InstallDriverSDKNonExecFile(radeon_video.c,$(DRIVERSDKDIR)/drivers/ati)
 InstallDriverSDKNonExecFile(radeon_common.h,$(DRIVERSDKDIR)/drivers/ati)
+InstallDriverSDKNonExecFile(radeon_accelfuncs.c,$(DRIVERSDKDIR)/drivers/ati)
 
 InstallDriverSDKObjectModule(ati,$(DRIVERSDKMODULEDIR),drivers)
 InstallDriverSDKObjectModule(atimisc,$(DRIVERSDKMODULEDIR),drivers)
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/ati/radeon_cursor.c.branch-4.3	2003-02-24 15:34:55.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/ati/radeon_cursor.c	2003-07-08 00:10:42.000000000 -0400
@@ -370,7 +370,7 @@
     }
 
     if (info->IsSecondary || info->Clone) {
-	save2 = INREG(RADEON_CRTC_GEN_CNTL) & ~(CARD32) (3 << 20);
+	save2 = INREG(RADEON_CRTC2_GEN_CNTL) & ~(CARD32) (3 << 20);
 	save2 |= (CARD32) (2 << 20);
 	OUTREG(RADEON_CRTC2_GEN_CNTL, save2 & (CARD32)~RADEON_CRTC2_CUR_EN);
     }
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/cyrix/Imakefile.branch-4.3	2003-02-17 12:06:42.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/cyrix/Imakefile	2003-07-08 00:10:42.000000000 -0400
@@ -43,6 +43,7 @@
 InstallDriverSDKNonExecFile(cyrix_bank.c,$(DRIVERSDKDIR)/drivers/cyrix)
 InstallDriverSDKNonExecFile(cyrix_driver.c,$(DRIVERSDKDIR)/drivers/cyrix)
 InstallDriverSDKNonExecFile(cyrix_helper.c,$(DRIVERSDKDIR)/drivers/cyrix)
+InstallDriverSDKNonExecFile(cyrix_shadow.c,$(DRIVERSDKDIR)/drivers/cyrix)
 
 InstallDriverSDKObjectModule(cyrix,$(DRIVERSDKMODULEDIR),drivers)
 
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/glint/pm3_video.c.branch-4.3	2002-05-22 04:12:02.000000000 -0400
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/glint/pm3_video.c	2003-07-08 00:10:43.000000000 -0400
@@ -746,6 +746,10 @@
 	drw_w = (dstBox->x2 - dstBox->x1);
     }
 
+    /* Avoid divide by zero in compute_scale_factor. */
+    if (drw_w < 8)
+	return;
+
     /* Let's adjust the width of source and dest to be compliant with 
      * the Permedia3 overlay unit requirement, and compute the X deltas. */
     newx1 = src_w; newx2 = drw_w;
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/i740/Imakefile.branch-4.3	2003-02-17 12:06:42.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/i740/Imakefile	2003-07-08 00:10:43.000000000 -0400
@@ -47,6 +47,10 @@
 InstallDriverSDKNonExecFile(i740_io.c,$(DRIVERSDKDIR)/drivers/i740)
 InstallDriverSDKNonExecFile(i740_macros.h,$(DRIVERSDKDIR)/drivers/i740)
 InstallDriverSDKNonExecFile(i740_reg.h,$(DRIVERSDKDIR)/drivers/i740)
+InstallDriverSDKNonExecFile(i740_dga.h,$(DRIVERSDKDIR)/drivers/i740)
+InstallDriverSDKNonExecFile(i740_dga.c,$(DRIVERSDKDIR)/drivers/i740)
+InstallDriverSDKNonExecFile(i740_i2c.c,$(DRIVERSDKDIR)/drivers/i740)
+InstallDriverSDKNonExecFile(i740_video.c,$(DRIVERSDKDIR)/drivers/i740)
 
 InstallDriverSDKObjectModule(i740,$(DRIVERSDKMODULEDIR),drivers)
 
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/i810/i810_dri.c.branch-4.3	2002-12-09 20:27:04.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/i810/i810_dri.c	2003-07-08 00:10:43.000000000 -0400
@@ -490,7 +490,11 @@
       DRICloseScreen(pScreen);
       return FALSE;
    } else {
-      back_size = i810_pitches[pitch_idx] * (pScrn->virtualY + 4);
+      /* for tiled memory to work, the buffer needs to have the
+       * number of lines as a multiple of 16 (the tile size),
+       *  - airlied */
+      int lines = (pScrn->virtualY + 15) / 16 * 16;
+      back_size = i810_pitches[pitch_idx] * lines;
       back_size = ((back_size + 4096 - 1) / 4096) * 4096;
    }
 
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/mga/Imakefile.branch-4.3	2003-02-17 12:06:42.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/mga/Imakefile	2003-07-08 00:10:43.000000000 -0400
@@ -68,7 +68,7 @@
 OBJS = $(MGAOBJS) $(MGAHALOBJS)
 
 #if defined(XF86DriverSDK)
-INCLUDES = -I. -I../../include
+INCLUDES = -I. -I../../include $(MGAHALINCLUDES)
 #else
 INCLUDES = -I. $(MGAHALINCLUDES) -I$(XF86COMSRC) -I$(XF86OSSRC) \
            -I$(SERVERSRC)/fb \
@@ -133,6 +133,20 @@
 InstallDriverSDKNonExecFile(mga_video.c,$(DRIVERSDKDIR)/drivers/mga)
 InstallDriverSDKNonExecFile(mga_halmod.c,$(DRIVERSDKDIR)/drivers/mga)
 InstallDriverSDKNonExecFile(mga_common.h,$(DRIVERSDKDIR)/drivers/mga)
+InstallDriverSDKNonExecFile(mga_dripriv.h,$(DRIVERSDKDIR)/drivers/mga)
+InstallDriverSDKNonExecFile(mga_dri.h,$(DRIVERSDKDIR)/drivers/mga)
+InstallDriverSDKNonExecFile(mga_merge.h,$(DRIVERSDKDIR)/drivers/mga)
+InstallDriverSDKNonExecFile(mga_merge.c,$(DRIVERSDKDIR)/drivers/mga)
+InstallDriverSDKNonExecFile(mga_g450pll.c,$(DRIVERSDKDIR)/drivers/mga)
+InstallDriverSDKNonExecFile(mga_dh.c,$(DRIVERSDKDIR)/drivers/mga)
+InstallDriverSDKNonExecFile(mga_esc.c,$(DRIVERSDKDIR)/drivers/mga)
+InstallDriverSDKNonExecFile(mga_dri.c,$(DRIVERSDKDIR)/drivers/mga)
+InstallDriverSDKNonExecFile(mgareg_flags.h,$(DRIVERSDKDIR)/drivers/mga)
+InstallDriverSDKNonExecFile(mga_sarea.h,$(DRIVERSDKDIR)/drivers/mga)
+InstallDriverSDKNonExecFile(mga_ucode.h,$(DRIVERSDKDIR)/drivers/mga)
+InstallDriverSDKNonExecFile(client.h,$(DRIVERSDKDIR)/drivers/mga)
+InstallDriverSDKNonExecFile(clientlx.c,$(DRIVERSDKDIR)/drivers/mga)
+InstallDriverSDKNonExecFile(HALlib/binding.h,$(DRIVERSDKDIR)/drivers/mga/HALlib)
 
 InstallDriverSDKObjectModule(mga,$(DRIVERSDKMODULEDIR),drivers)
 #if BuildMatroxHal
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/neomagic/Imakefile.branch-4.3	2003-02-17 12:06:43.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/neomagic/Imakefile	2003-07-08 00:10:43.000000000 -0400
@@ -56,5 +56,7 @@
 InstallDriverSDKNonExecFile(neo_shadow.c,$(DRIVERSDKDIR)/drivers/neomagic)
 InstallDriverSDKNonExecFile(neo_macros.h,$(DRIVERSDKDIR)/drivers/neomagic)
 InstallDriverSDKNonExecFile(neo_reg.h,$(DRIVERSDKDIR)/drivers/neomagic)
+InstallDriverSDKNonExecFile(neo_video.c,$(DRIVERSDKDIR)/drivers/neomagic)
+InstallDriverSDKNonExecFile(neo_video.h,$(DRIVERSDKDIR)/drivers/neomagic)
 
 InstallDriverSDKObjectModule(neomagic,$(DRIVERSDKMODULEDIR),drivers)
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/newport/Imakefile.branch-4.3	2002-12-09 23:03:00.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/newport/Imakefile	2003-07-08 00:10:43.000000000 -0400
@@ -30,9 +30,9 @@
 ObjectModuleTarget(newport,$(OBJS))
 
 InstallObjectModule(newport,$(MODULEDIR),drivers)
-InstallNamedNonExec($(XF86CONFIG),$(XF86CONFIG),$(LIBDIR))
 
 #if !defined(XF86DriverSDK)
+InstallNamedNonExec($(XF86CONFIG),$(XF86CONFIG),$(LIBDIR))
 InstallModuleManPage(newport)
 #endif
 
@@ -43,6 +43,7 @@
 InstallDriverSDKNonExecFile(newport_regs.c,$(DRIVERSDKDIR)/drivers/newport)
 InstallDriverSDKNonExecFile(newport_cmap.c,$(DRIVERSDKDIR)/drivers/newport)
 InstallDriverSDKNonExecFile(newport_shadow.c,$(DRIVERSDKDIR)/drivers/newport)
+InstallDriverSDKNonExecFile(newport_cursor.c,$(DRIVERSDKDIR)/drivers/newport)
 InstallDriverSDKNonExecFile(newport_regs.h,$(DRIVERSDKDIR)/drivers/newport)
 InstallDriverSDKNonExecFile(newport.h,$(DRIVERSDKDIR)/drivers/newport)
 
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/nsc/Imakefile.branch-4.3	2003-02-19 05:10:49.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/nsc/Imakefile	2003-07-08 00:10:43.000000000 -0400
@@ -61,7 +61,7 @@
 		nsc_gx2_shadow.o $(I86OBJ) $(STBOBJS) $(DURANGOOBJS)
 
 #if defined(XF86DriverSDK)
-INCLUDES = -I. -I../../include
+INCLUDES = -I. -I../../include $(EXTINCLUDES)
 #else
 INCLUDES = -I. -I$(XF86COMSRC) -I$(XF86OSSRC) -I$(SERVERSRC)/Xext \
 		-I$(SERVERSRC)/mfb -I$(SERVERSRC)/mi \
@@ -138,6 +138,7 @@
 InstallDriverSDKNonExecFile(gfx/gfx_rndr.c,$(DRIVERSDKDIR)/drivers/nsc/gfx)
 InstallDriverSDKNonExecFile(gfx/gfx_rtns.h,$(DRIVERSDKDIR)/drivers/nsc/gfx)
 InstallDriverSDKNonExecFile(gfx/gfx_tv.c,$(DRIVERSDKDIR)/drivers/nsc/gfx)
+InstallDriverSDKNonExecFile(gfx/gfx_tv.h,$(DRIVERSDKDIR)/drivers/nsc/gfx)
 InstallDriverSDKNonExecFile(gfx/gfx_type.h,$(DRIVERSDKDIR)/drivers/nsc/gfx)
 InstallDriverSDKNonExecFile(gfx/gfx_vga.c,$(DRIVERSDKDIR)/drivers/nsc/gfx)
 InstallDriverSDKNonExecFile(gfx/gfx_vid.c,$(DRIVERSDKDIR)/drivers/nsc/gfx)
@@ -155,11 +156,15 @@
 InstallDriverSDKNonExecFile(gfx/tv_1200.c,$(DRIVERSDKDIR)/drivers/nsc/gfx)
 InstallDriverSDKNonExecFile(gfx/tv_fs450.c,$(DRIVERSDKDIR)/drivers/nsc/gfx)
 InstallDriverSDKNonExecFile(gfx/tv_fs450.h,$(DRIVERSDKDIR)/drivers/nsc/gfx)
+InstallDriverSDKNonExecFile(gfx/tv_fs451.c,$(DRIVERSDKDIR)/drivers/nsc/gfx)
+InstallDriverSDKNonExecFile(gfx/tv_geode.c,$(DRIVERSDKDIR)/drivers/nsc/gfx)
 InstallDriverSDKNonExecFile(gfx/vga_gu1.c,$(DRIVERSDKDIR)/drivers/nsc/gfx)
 InstallDriverSDKNonExecFile(gfx/vid_1200.c,$(DRIVERSDKDIR)/drivers/nsc/gfx)
+InstallDriverSDKNonExecFile(gfx/vid_1400.c,$(DRIVERSDKDIR)/drivers/nsc/gfx)
 InstallDriverSDKNonExecFile(gfx/vid_5530.c,$(DRIVERSDKDIR)/drivers/nsc/gfx)
 InstallDriverSDKNonExecFile(gfx/vid_rdcl.c,$(DRIVERSDKDIR)/drivers/nsc/gfx)
 InstallDriverSDKNonExecFile(gfx/vip_1200.c,$(DRIVERSDKDIR)/drivers/nsc/gfx)
+InstallDriverSDKNonExecFile(gfx/vip_1400.c,$(DRIVERSDKDIR)/drivers/nsc/gfx)
 InstallDriverSDKNonExecFile(panel/92xx.h,$(DRIVERSDKDIR)/drivers/nsc/panel)
 InstallDriverSDKNonExecFile(panel/cen9211.c,$(DRIVERSDKDIR)/drivers/nsc/panel)
 InstallDriverSDKNonExecFile(panel/cen9211.h,$(DRIVERSDKDIR)/drivers/nsc/panel)
@@ -167,6 +172,8 @@
 InstallDriverSDKNonExecFile(panel/dora9211.h,$(DRIVERSDKDIR)/drivers/nsc/panel)
 InstallDriverSDKNonExecFile(panel/drac9210.c,$(DRIVERSDKDIR)/drivers/nsc/panel)
 InstallDriverSDKNonExecFile(panel/drac9210.h,$(DRIVERSDKDIR)/drivers/nsc/panel)
+InstallDriverSDKNonExecFile(panel/gx2_9211.c,$(DRIVERSDKDIR)/drivers/nsc/panel)
+InstallDriverSDKNonExecFile(panel/gx2_9211.h,$(DRIVERSDKDIR)/drivers/nsc/panel)
 InstallDriverSDKNonExecFile(panel/panel.c,$(DRIVERSDKDIR)/drivers/nsc/panel)
 InstallDriverSDKNonExecFile(panel/panel.h,$(DRIVERSDKDIR)/drivers/nsc/panel)
 InstallDriverSDKNonExecFile(panel/platform.c,$(DRIVERSDKDIR)/drivers/nsc/panel)
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/nv/nv_setup.c.branch-4.3	2003-02-10 18:42:51.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/nv/nv_setup.c	2003-07-08 00:10:44.000000000 -0400
@@ -416,6 +416,9 @@
         break;
     }
 
+    if(pNv->riva.Architecture == 3)
+       pNv->riva.PCRTC0 = pNv->riva.PGRAPH;
+
     if(pNv->SecondCRTC) {
        pNv->riva.PCIO = pNv->riva.PCIO0 + 0x2000;
        pNv->riva.PCRTC = pNv->riva.PCRTC0 + 0x800;
@@ -484,7 +487,6 @@
                                      frameBase+0x00C00000, 0x00008000);
             
     NVCommonSetup(pScrn);
-    pNv->riva.PCRTC = pNv->riva.PCRTC0 = pNv->riva.PGRAPH;
 }
 
 void
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/s3/Imakefile.branch-4.3	2003-02-17 12:06:43.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/s3/Imakefile	2003-07-08 00:10:44.000000000 -0400
@@ -46,6 +46,7 @@
 SpecialCObjectRule(s3_video,$(_NOOP_), -DS3_NEWMMIO)
            
 InstallDriverSDKNonExecFile(Imakefile,$(DRIVERSDKDIR)/drivers/s3)  
+InstallDriverSDKNonExecFile(newmmio.h,$(DRIVERSDKDIR)/drivers/s3)
 InstallDriverSDKNonExecFile(s3.h,$(DRIVERSDKDIR)/drivers/s3)
 InstallDriverSDKNonExecFile(s3_reg.h,$(DRIVERSDKDIR)/drivers/s3)
 InstallDriverSDKNonExecFile(s3_driver.c,$(DRIVERSDKDIR)/drivers/s3)
@@ -57,4 +58,3 @@
 InstallDriverSDKNonExecFile(s3_Trio64DAC.c,$(DRIVERSDKDIR)/drivers/s3)
 InstallDriverSDKNonExecFile(s3_IBMRGB.c,$(DRIVERSDKDIR)/drivers/s3)
 InstallDriverSDKNonExecFile(s3_Ti.c,$(DRIVERSDKDIR)/drivers/s3)
-
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/s3virge/Imakefile.branch-4.3	2003-02-17 12:06:43.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/s3virge/Imakefile	2003-07-08 00:10:44.000000000 -0400
@@ -63,5 +63,6 @@
 InstallDriverSDKNonExecFile(s3v_macros.h,$(DRIVERSDKDIR)/drivers/s3virge)
 InstallDriverSDKNonExecFile(s3v_rop.h,$(DRIVERSDKDIR)/drivers/s3virge)
 InstallDriverSDKNonExecFile(s3v_shadow.c,$(DRIVERSDKDIR)/drivers/s3virge)
+InstallDriverSDKNonExecFile(s3v_xv.c,$(DRIVERSDKDIR)/drivers/s3virge)
 
 InstallDriverSDKObjectModule(s3virge,$(DRIVERSDKMODULEDIR),drivers)
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/s3virge/s3v_driver.c.branch-4.3	2003-02-03 21:20:50.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/s3virge/s3v_driver.c	2003-07-08 00:10:44.000000000 -0400
@@ -1781,8 +1781,8 @@
       VGAOUT8(vgaCRIndex, 0x93);
       save->CR93 = VGAIN8(vgaCRReg);
    }
-   if (ps3v->Chipset == S3_ViRGE_DXGX || S3_ViRGE_GX2_SERIES(ps3v->Chipset) || 
-       S3_ViRGE_MX_SERIES(ps3v->Chipset)) {
+   if (ps3v->Chipset == S3_ViRGE_DXGX || S3_ViRGE_GX2_SERIES(ps3v->Chipset) ||
+       S3_ViRGE_MX_SERIES(ps3v->Chipset) || S3_TRIO_3D_SERIES(ps3v->Chipset)) {
       VGAOUT8(vgaCRIndex, 0x90);
       save->CR90 = VGAIN8(vgaCRReg);
       VGAOUT8(vgaCRIndex, 0x91);
@@ -2109,7 +2109,7 @@
       VGAOUT8(vgaCRReg, restore->CR93);
    }
    if (ps3v->Chipset == S3_ViRGE_DXGX || S3_ViRGE_GX2_SERIES(ps3v->Chipset) ||
-       S3_ViRGE_MX_SERIES(ps3v->Chipset)) {
+       S3_ViRGE_MX_SERIES(ps3v->Chipset) || S3_TRIO_3D_SERIES(ps3v->Chipset)) {
       VGAOUT8(vgaCRIndex, 0x90);
       VGAOUT8(vgaCRReg, restore->CR90);
       VGAOUT8(vgaCRIndex, 0x91);
@@ -2268,7 +2268,6 @@
    else
      VGAOUT8(vgaCRReg, cr3a);
 
-
    if (xf86GetVerbosity() > 1) {
       xf86DrvMsgVerb(pScrn->scrnIndex, X_INFO, VERBLEV, 
          "ViRGE driver: done restoring mode, dumping CR registers:\n");
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/savage/savage_driver.c.branch-4.3	2003-06-17 01:24:50.000000000 -0400
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/savage/savage_driver.c	2003-07-08 00:10:44.000000000 -0400
@@ -2103,9 +2103,19 @@
     VGAOUT8(vgaCRIndex, 0x3a);
     VGAOUT8(vgaCRReg, cr3a);
 
-    if( Entering )
+    if( Entering ) {
+    	/* We reinit the engine here just as in the UseBIOS case
+	 * as otherwise we lose performance because the engine
+	 * isn't setup properly (Alan Hourihane - alanh@fairlite.demon.co.uk).
+	 */
+	SavageInitialize2DEngine(pScrn);
 	SavageSetGBD(pScrn);
 
+	VGAOUT16(vgaCRIndex, 0x0140);
+
+	SavageSetGBD(pScrn);
+    }
+
     vgaHWProtect(pScrn, FALSE);
 
     return;
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/siliconmotion/smi_driver.c.branch-4.3	2003-02-05 12:45:29.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/siliconmotion/smi_driver.c	2003-07-08 00:10:45.000000000 -0400
@@ -1410,7 +1410,7 @@
 	ENTER_PROC("SMI_Save");
 
 	/* Save the standard VGA registers */
-	vgaHWSave(pScrn, vgaSavePtr, VGA_SR_MODE);
+	vgaHWSave(pScrn, vgaSavePtr, VGA_SR_ALL);
 	save->smiDACMask = VGAIN8(pSmi, VGA_DAC_MASK);
 	VGAOUT8(pSmi, VGA_DAC_READ_ADDR, 0);
 	for (i = 0; i < 256; i++)
@@ -1601,7 +1601,7 @@
 		VGAOUT8_INDEX(pSmi, VGA_SEQ_INDEX, VGA_SEQ_DATA, 0xA0, restore->SRA0);
 
 		/* Restore the standard VGA registers */
-		vgaHWRestore(pScrn, vgaSavePtr, VGA_SR_MODE);
+		vgaHWRestore(pScrn, vgaSavePtr, VGA_SR_ALL);
 		if (restore->smiDACMask)
 		{
 			VGAOUT8(pSmi, VGA_DAC_MASK, restore->smiDACMask);
@@ -1672,7 +1672,7 @@
 		} 
 
 		if (restore->modeInit)
-		    vgaHWRestore(pScrn, vgaSavePtr, VGA_SR_MODE);
+		    vgaHWRestore(pScrn, vgaSavePtr, VGA_SR_ALL);
 
 		if (!SMI_LYNXM_SERIES(pSmi->Chipset))
 		{
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/siliconmotion/smi_video.c.branch-4.3	2003-01-11 22:55:49.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/siliconmotion/smi_video.c	2003-07-08 00:10:45.000000000 -0400
@@ -277,7 +277,7 @@
     {XvSettable | XvGettable,        0,             1, XV_INTERLACED_NAME},
 };
 
-static XF86AttributeRec SMI_VideoAttributes[N_ATTRS] = {
+static XF86AttributeRec SMI_VideoAttributes[2] = {
     {XvSettable | XvGettable,        0,           255, XV_BRIGHTNESS_NAME},
     {XvSettable | XvGettable, 0x000000,      0xFFFFFF, XV_COLORKEY_NAME},
 };
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/sis/Imakefile.branch-4.3	2003-02-17 12:06:44.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/sis/Imakefile	2003-07-08 00:10:45.000000000 -0400
@@ -73,6 +73,7 @@
 InstallDriverSDKNonExecFile(sis300_accel.h,$(DRIVERSDKDIR)/drivers/sis)
 InstallDriverSDKNonExecFile(sis310_accel.c,$(DRIVERSDKDIR)/drivers/sis)
 InstallDriverSDKNonExecFile(sis310_accel.h,$(DRIVERSDKDIR)/drivers/sis)
+InstallDriverSDKNonExecFile(sis_accel.h,$(DRIVERSDKDIR)/drivers/sis)
 InstallDriverSDKNonExecFile(sis_accel.c,$(DRIVERSDKDIR)/drivers/sis)
 InstallDriverSDKNonExecFile(init.c,$(DRIVERSDKDIR)/drivers/sis)
 InstallDriverSDKNonExecFile(init.h,$(DRIVERSDKDIR)/drivers/sis)
@@ -103,5 +104,8 @@
 InstallDriverSDKNonExecFile(sis_vga.c,$(DRIVERSDKDIR)/drivers/sis)
 InstallDriverSDKNonExecFile(sis_video.c,$(DRIVERSDKDIR)/drivers/sis)
 InstallDriverSDKNonExecFile(sis6326_video.c,$(DRIVERSDKDIR)/drivers/sis)
+InstallDriverSDKNonExecFile(sis_accel.c,$(DRIVERSDKDIR)/drivers/sis)
+InstallDriverSDKNonExecFile(vgatypes.h,$(DRIVERSDKDIR)/drivers/sis)
+InstallDriverSDKNonExecFile(vstruct.h,$(DRIVERSDKDIR)/drivers/sis)
 
 InstallDriverSDKObjectModule(sis,$(DRIVERSDKMODULEDIR),drivers)
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/trident/Imakefile.branch-4.3	2003-02-17 12:06:44.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/trident/Imakefile	2003-07-08 00:10:45.000000000 -0400
@@ -62,5 +62,6 @@
 InstallDriverSDKNonExecFile(tridentramdac.c,$(DRIVERSDKDIR)/drivers/trident)
 InstallDriverSDKNonExecFile(trident_video.c,$(DRIVERSDKDIR)/drivers/trident)
 InstallDriverSDKNonExecFile(tvga_dac.c,$(DRIVERSDKDIR)/drivers/trident)
+InstallDriverSDKNonExecFile(xp_accel.c,$(DRIVERSDKDIR)/drivers/trident)
 
 InstallDriverSDKObjectModule(trident,$(DRIVERSDKMODULEDIR),drivers)
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/vesa/Imakefile.branch-4.3	2003-02-17 12:06:44.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/vesa/Imakefile	2003-07-08 00:10:45.000000000 -0400
@@ -40,6 +40,7 @@
 
 InstallDriverSDKNonExecFile(Imakefile,$(DRIVERSDKDIR)/drivers/vesa)
 InstallDriverSDKNonExecFile(vesa.c,$(DRIVERSDKDIR)/drivers/vesa)
+InstallDriverSDKNonExecFile(vesa.h,$(DRIVERSDKDIR)/drivers/vesa)
 
 InstallDriverSDKObjectModule(vesa,$(DRIVERSDKMODULEDIR),drivers)
 
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/vmware/Imakefile.branch-4.3	2003-02-17 12:06:45.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/drivers/vmware/Imakefile	2003-07-08 00:10:45.000000000 -0400
@@ -43,5 +43,20 @@
 DependTarget()
 
 InstallDriverSDKNonExecFile(Imakefile,$(DRIVERSDKDIR)/drivers/vmware)
+InstallDriverSDKNonExecFile(vmware.c,$(DRIVERSDKDIR)/drivers/vmware)
+InstallDriverSDKNonExecFile(vmwarecurs.c,$(DRIVERSDKDIR)/drivers/vmware)
+InstallDriverSDKNonExecFile(vmwarexaa.c,$(DRIVERSDKDIR)/drivers/vmware)
+InstallDriverSDKNonExecFile(guest_os.h,$(DRIVERSDKDIR)/drivers/vmware)
+InstallDriverSDKNonExecFile(includeCheck.h,$(DRIVERSDKDIR)/drivers/vmware)
+InstallDriverSDKNonExecFile(svga_limits.h,$(DRIVERSDKDIR)/drivers/vmware)
+InstallDriverSDKNonExecFile(svga_reg.h,$(DRIVERSDKDIR)/drivers/vmware)
+InstallDriverSDKNonExecFile(svga_struct.h,$(DRIVERSDKDIR)/drivers/vmware)
+InstallDriverSDKNonExecFile(vm_basic_types.h,$(DRIVERSDKDIR)/drivers/vmware)
+InstallDriverSDKNonExecFile(vm_device_version.h,$(DRIVERSDKDIR)/drivers/vmware)
+InstallDriverSDKNonExecFile(vmware.h,$(DRIVERSDKDIR)/drivers/vmware)
+InstallDriverSDKNonExecFile(bits2pixels.c,$(DRIVERSDKDIR)/drivers/vmware)
+InstallDriverSDKNonExecFile(bits2pixels.h,$(DRIVERSDKDIR)/drivers/vmware)
+InstallDriverSDKNonExecFile(offscreen_manager.c,$(DRIVERSDKDIR)/drivers/vmware)
+InstallDriverSDKNonExecFile(offscreen_manager.h,$(DRIVERSDKDIR)/drivers/vmware)
 
 InstallDriverSDKObjectModule(vmware,$(DRIVERSDKMODULEDIR),drivers)
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/etc/Xinstall.sh.branch-4.3	2003-02-23 23:24:17.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/etc/Xinstall.sh	2003-07-08 00:10:45.000000000 -0400
@@ -561,8 +561,11 @@
 				2)
 					DistName="Linux-ix86-glibc22"
 					;;
+				3)
+					DistName="Linux-ix86-glibc23"
+					;;
 				*)
-					Message="No dist available for glibc 2.$OsLibcMinor.  Try Linux-ix86-glibc22"
+					Message="No dist available for glibc 2.$OsLibcMinor.  Try Linux-ix86-glibc23"
 					;;
 				esac
 				;;
@@ -596,8 +599,11 @@
 			6.1)
 				DistName="Linux-alpha-glibc21"
 				;;
+			6.2)
+				DistName="Linux-alpha-glibc22"
+				;;
 			6.*)
-				Message="No Linux/alpha binaries for glibc 2.$OsLibcMinor.  Try Linux-alpha-glibc21"
+				Message="No Linux/alpha binaries for glibc 2.$OsLibcMinor.  Try Linux-alpha-glibc22"
 				;;
 			*)
 				Message="No Linux/alpha binaries for this libc version"
@@ -632,7 +638,7 @@
 					DistName="NetBSD-1.4.x"
 					;;
 				*)
-					DistName="NetBSD-1.5"
+					DistName="NetBSD-1.6"
 					;;
 				esac
 				;;
@@ -650,8 +656,8 @@
 		case "$OsArch" in
 		i386)
 			case "$OsVersion" in
-			3.0*)	# Check this
-				DistName="OpenBSD-3.0"
+			3.2*)	# Check this
+				DistName="OpenBSD-3.2"
 				;;
 			*)
 				Message="No OpenBSD/i386 binaries available for this version"
@@ -947,7 +953,7 @@
             SERVDIST="Xxserv.tgz"
         fi
 	;;
-FreeBSD|NetBSD|OpenBSD)
+FreeBSD|OpenBSD)
 	VARDIST="Xvar.tgz"
 	XKBDBDIR="$VARDIR/db/xkb"
 	;;
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/int10/generic.c.branch-4.3	2002-04-04 09:05:51.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/int10/generic.c	2003-07-08 00:10:45.000000000 -0400
@@ -108,7 +108,7 @@
     MapVRam(pInt);
 #ifdef _PC
     if (!sysMem)
-	sysMem = xf86MapVidMem(screen, VIDMEM_FRAMEBUFFER, V_BIOS,
+	sysMem = xf86MapVidMem(screen, VIDMEM_MMIO, V_BIOS,
 			       BIOS_SIZE + SYS_BIOS - V_BIOS);
     INTPriv(pInt)->sysMem = sysMem;
 
@@ -446,12 +446,14 @@
 #define OFF(addr) ((addr) & 0xffff)
 #if defined _PC
 # define HIGH_OFFSET (INTPriv(pInt)->highMemory)
+# define HIGH_BASE   V_BIOS
 #else
 # define HIGH_OFFSET SYS_BIOS
+# define HIGH_BASE   SYS_BIOS
 #endif
 # define SYS(addr) ((addr) >= HIGH_OFFSET)
 #define V_ADDR(addr) \
-	  (SYS(addr) ? ((char*)INTPriv(pInt)->sysMem) + (addr - HIGH_OFFSET) \
+	  (SYS(addr) ? ((char*)INTPriv(pInt)->sysMem) + (addr - HIGH_BASE) \
 	   : (((char*)(INTPriv(pInt)->base) + addr)))
 #define VRAM_ADDR(addr) (addr - V_RAM)
 #define VRAM_BASE (INTPriv(pInt)->vRam)
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/int10/xf86int10.c.branch-4.3	2002-11-25 09:05:01.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/int10/xf86int10.c	2003-07-08 00:10:45.000000000 -0400
@@ -664,7 +664,7 @@
 #endif
 	return 1;
     case 0xb108:
-	if ((tag = findPci(pInt, X86_EBX))) {
+	if ((tag = findPci(pInt, X86_EBX)) != PCI_NOT_FOUND) {
 	    X86_CL = pciReadByte(tag, X86_EDI);
 	    X86_EAX = X86_AL | (SUCCESSFUL << 8);
 	    X86_EFLAGS &= ~((unsigned long)0x01); /* clear carry flag */
@@ -677,7 +677,7 @@
 #endif
 	return 1;
     case 0xb109:
-	if ((tag = findPci(pInt, X86_EBX))) {
+	if ((tag = findPci(pInt, X86_EBX)) != PCI_NOT_FOUND) {
 	    X86_CX = pciReadWord(tag, X86_EDI);
 	    X86_EAX = X86_AL | (SUCCESSFUL << 8);
 	    X86_EFLAGS &= ~((unsigned long)0x01); /* clear carry flag */
@@ -690,7 +690,7 @@
 #endif
 	return 1;
     case 0xb10a:
-	if ((tag = findPci(pInt, X86_EBX))) {
+	if ((tag = findPci(pInt, X86_EBX)) != PCI_NOT_FOUND) {
 	    X86_ECX = pciReadLong(tag, X86_EDI);
 	    X86_EAX = X86_AL | (SUCCESSFUL << 8);
 	    X86_EFLAGS &= ~((unsigned long)0x01); /* clear carry flag */
@@ -703,7 +703,7 @@
 #endif
 	return 1;
     case 0xb10b:
-	if ((tag = findPci(pInt, X86_EBX))) {
+	if ((tag = findPci(pInt, X86_EBX)) != PCI_NOT_FOUND) {
 	    pciWriteByte(tag, X86_EDI, X86_CL);
 	    X86_EAX = X86_AL | (SUCCESSFUL << 8);
 	    X86_EFLAGS &= ~((unsigned long)0x01); /* clear carry flag */
@@ -716,7 +716,7 @@
 #endif
 	return 1;
     case 0xb10c:
-	if ((tag = findPci(pInt, X86_EBX))) {
+	if ((tag = findPci(pInt, X86_EBX)) != PCI_NOT_FOUND) {
 	    pciWriteWord(tag, X86_EDI, X86_CX);
 	    X86_EAX = X86_AL | (SUCCESSFUL << 8);
 	    X86_EFLAGS &= ~((unsigned long)0x01); /* clear carry flag */
@@ -729,7 +729,7 @@
 #endif
 	return 1;
     case 0xb10d:
-	if ((tag = findPci(pInt, X86_EBX))) {
+	if ((tag = findPci(pInt, X86_EBX)) != PCI_NOT_FOUND) {
 	    pciWriteLong(tag, X86_EDI, X86_ECX);
 	    X86_EAX = X86_AL | (SUCCESSFUL << 8);
 	    X86_EFLAGS &= ~((unsigned long)0x01); /* clear carry flag */
@@ -759,7 +759,7 @@
     int func = bx & 0x7;
     if (xf86IsPciDevPresent(bus, dev, func))
 	return pciTag(bus, dev, func);
-    return 0;
+    return PCI_NOT_FOUND;
 }
 
 static CARD32
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/loader/Imakefile.branch-4.3	2003-02-26 18:32:12.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/loader/Imakefile	2003-07-08 00:10:45.000000000 -0400
@@ -30,10 +30,6 @@
     SHM_DEFINES = -DHAS_SHM
 #endif
 
-#if HasGlibc21Sigsetjmp
-  SETJMPDEFINES = -DHAS_GLIBC_SIGSETJMP=1
-#endif
-
 #if defined (x86_64Architecture)
 ARCHDEFINES = -DDoMMAPedMerge -DMmapPageAlign
 #endif
@@ -77,7 +73,7 @@
 NormalLibraryTarget(xloader,$(XOBJS) )
 
 SpecialCObjectRule(loadmod,NullParameter,$(MODULEDEFINES) $(EXT_DEFINES))
-SpecialCObjectRule(xf86sym,NullParameter,$(EXT_DEFINES) $(SETJMPDEFINES))
+SpecialCObjectRule(xf86sym,NullParameter,$(EXT_DEFINES))
 SpecialCObjectRule(dixsym,NullParameter,$(EXT_DEFINES))
 
 #ifdef SparcArchitecture
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/loader/xf86sym.c.branch-4.3	2003-02-26 15:08:02.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/loader/xf86sym.c	2003-07-08 00:10:45.000000000 -0400
@@ -63,8 +63,9 @@
 #include "compiler.h"
 
 #ifndef HAS_GLIBC_SIGSETJMP
-#if defined(setjmp) && \
-    defined(__GLIBC__) && __GLIBC__ == 2 && __GLIBC_MINOR__ < 2
+#if defined(setjmp) && defined(__GNU_LIBRARY__) && \
+    (!defined(__GLIBC__) || (__GLIBC__ < 2) || \
+     ((__GLIBC__ == 2) && (__GLIBC_MINOR__ < 3)))
 #define HAS_GLIBC_SIGSETJMP 1
 #endif
 #endif
@@ -897,9 +898,15 @@
    SYMFUNC(xf86shmctl)
 #ifdef HAS_GLIBC_SIGSETJMP
    SYMFUNC(xf86setjmp)
+   SYMFUNC(xf86setjmp0)
+#if defined(__GLIBC__) && (__GLIBC__ >= 2)
    SYMFUNCALIAS("xf86setjmp1",__sigsetjmp)
 #else
+   SYMFUNC(xf86setjmp1)		/* For libc5 */
+#endif
+#else
    SYMFUNCALIAS("xf86setjmp",setjmp)
+   SYMFUNCALIAS("xf86setjmp0",setjmp)
    SYMFUNC(xf86setjmp1)
 #endif
    SYMFUNCALIAS("xf86longjmp",longjmp)
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/os-support/bsd/alpha_video.c.branch-4.3	2002-10-29 18:19:13.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/os-support/bsd/alpha_video.c	2003-07-08 00:10:45.000000000 -0400
@@ -33,9 +33,13 @@
 #include <sys/param.h>
 #ifndef __NetBSD__
 #  include <sys/sysctl.h>
+#  ifdef __FreeBSD__
+#      include <machine/sysarch.h>
+#   endif
 # else
 #  include <machine/sysarch.h>
 #endif
+
 #include "xf86Axp.h"
 
 #include "xf86_OSlib.h"
@@ -51,6 +55,8 @@
 #define MAP_FAILED ((caddr_t)-1)
 #endif
 
+axpDevice bsdGetAXP(void);
+
 #ifndef __NetBSD__
 extern unsigned long dense_base(void);
 
@@ -248,7 +254,6 @@
 #ifndef HAS_APERTURE_DRV
            xf86Msg(X_WARNING, "checkDevMem: failed to open/mmap %s (%s)\n",
                    DEV_MEM, strerror(errno));
-           xf86ErrorF("\tlinear framebuffer access unavailable\n");
 #else
 #ifndef __OpenBSD__
            xf86Msg(X_WARNING, "checkDevMem: failed to open %s and %s\n"
@@ -258,12 +263,11 @@
                    "\t(%s)\n%s", DEV_APERTURE, DEV_MEM, strerror(errno),
                    SYSCTL_MSG);
 #endif /* __OpenBSD__ */
-	   
+#endif
            xf86ErrorF("\tlinear framebuffer access unavailable\n");
 	}
 	useDevMem = FALSE;
 	return;
-#endif
 }
 
 void
@@ -486,6 +490,9 @@
 sethae(u_int64_t hae)
 {
 #ifdef __FreeBSD__
+#ifndef ALPHA_SETHAE
+#define ALPHA_SETHAE 0
+#endif
 	struct parms p;
 	p.hae = hae;
 	return (sysarch(ALPHA_SETHAE, (char *)&p));
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/os-support/bus/Pci.c.branch-4.3	2003-01-23 11:22:13.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/os-support/bus/Pci.c	2003-07-08 00:10:45.000000000 -0400
@@ -573,7 +573,7 @@
 pciMfDev(int busnum, int devnum)
 {
     PCITAG tag0, tag1;
-    unsigned long id0, id1;
+    unsigned long id0, id1, val;
 
     /* Detect a multi-function device that complies to the PCI 2.0 spec */
 
@@ -582,7 +582,8 @@
     if (id0 == 0xffffffff)
 	return FALSE;
 
-    if (pciReadLong(tag0, PCI_HEADER_MISC) & PCI_HEADER_MULTIFUNCTION)
+    val = pciReadLong(tag0, PCI_HEADER_MISC) & 0x00ff0000;
+    if ((val != 0x00ff0000) && (val & PCI_HEADER_MULTIFUNCTION))
 	return TRUE;
 
     /*
@@ -969,6 +970,16 @@
 	for (i = 0; i < 17; i++)  /* PCI hdr plus 1st dev spec dword */
 	    devp->cfgspc.dwords[i] = pciReadLong(tag, i * sizeof(CARD32));
 
+#ifdef ARCH_PCI_HOST_BRIDGE
+	if ((devp->pci_base_class == PCI_CLASS_BRIDGE) &&
+	    (devp->pci_sub_class == PCI_SUBCLASS_BRIDGE_HOST))
+	    ARCH_PCI_HOST_BRIDGE(devp);
+#endif
+
+	/* Some broken devices don't implement this field... */
+	if (devp->pci_header_type == 0xff)
+	    devp->pci_header_type = 0;
+
 	switch (devp->pci_header_type & 0x7f) {
 	case 0:
 	    /* Get base address sizes for type 0 headers */
@@ -1025,9 +1036,6 @@
 		break;
 	    pciBusInfo[devp->busnum]->bridge = devp;
 	    pciBusInfo[devp->busnum]->primary_bus = devp->busnum;
-#ifdef ARCH_PCI_HOST_BRIDGE
-	    ARCH_PCI_HOST_BRIDGE(devp);
-#endif
 	    break;
 
 	case 1:
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/os-support/bus/Pci.h.branch-4.3	2002-12-23 10:37:26.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/os-support/bus/Pci.h	2003-07-08 00:10:45.000000000 -0400
@@ -97,8 +97,6 @@
 # define MAX_PCI_BUSES   256	/* Max number of PCI buses           */
 #endif
 
-#define PCI_NOT_FOUND   0xffffffff
-
 #define DEVID(vendor, device) \
     ((CARD32)((PCI_CHIP_##device << 16) | PCI_VENDOR_##vendor))
 
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/os-support/bus/xf86Pci.h.branch-4.3	2003-02-18 10:42:12.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/os-support/bus/xf86Pci.h	2003-07-08 00:10:45.000000000 -0400
@@ -81,6 +81,8 @@
 #include "Xfuncproto.h"
 #include "misc.h"
 
+#define PCI_NOT_FOUND	0xFFFFFFFFU
+
 /*
  * PCI cfg space definitions (e.g. stuff right out of the PCI spec)
  */
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/os-support/shared/libc_wrapper.c.branch-4.3	2003-02-22 01:00:39.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/os-support/shared/libc_wrapper.c	2003-07-08 00:10:45.000000000 -0400
@@ -165,8 +165,9 @@
 #endif
 #include <setjmp.h>
 
-#if defined(setjmp) && \
-    defined(__GLIBC__) && __GLIBC__ == 2 && __GLIBC_MINOR__ < 2
+#if defined(setjmp) && defined(__GNU_LIBRARY__) && \
+    (!defined(__GLIBC__) || (__GLIBC__ < 2) || \
+     ((__GLIBC__ == 2) && (__GLIBC_MINOR__ < 3)))
 #define HAS_GLIBC_SIGSETJMP 1
 #endif
 
@@ -1961,23 +1962,48 @@
 }
 
 #ifdef HAS_GLIBC_SIGSETJMP
+
 int
 xf86setjmp(xf86jmp_buf env)
 {
+#if defined(__GLIBC__) && (__GLIBC__ >= 2)
+    return __sigsetjmp(env, xf86setjmp1_arg2());
+#else
+    return xf86setjmp1(env, xf86setjmp1_arg2());
+#endif
+}
+
+int
+xf86setjmp0(xf86jmp_buf env)
+{
     FatalError("setjmp: type 0 called instead of type %d\n", xf86getjmptype());
 }
-#else
+
+#if !defined(__GLIBC__) || (__GLIBC__ < 2)	/* libc5 */
+
 int
 xf86setjmp1(xf86jmp_buf env, int arg2)
 {
-    FatalError("setjmp: type 1 called instead of type %d\n", xf86getjmptype());
+    __sigjmp_save((void *)env, arg2);
+    return __setjmp((void *)env);
 }
+
 #endif
 
+#else	/* HAS_GLIBC_SIGSETJMP */
+
+int
+xf86setjmp1(xf86jmp_buf env, int arg2)
+{
+    FatalError("setjmp: type 1 called instead of type %d\n", xf86getjmptype());
+}
+
+#endif  /* HAS_GLIBC_SIGSETJMP */
+
 int
 xf86setjmp1_arg2()
 {
-    return 0;
+    return 1;
 }
 
 int
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/os-support/Imakefile.branch-4.3	2003-02-17 12:06:45.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/os-support/Imakefile	2003-07-08 00:10:45.000000000 -0400
@@ -119,3 +119,5 @@
 InstallDriverSDKNonExecFile(xf86_ansic.h,$(DRIVERSDKINCLUDEDIR))
 InstallDriverSDKNonExecFile(xf86_libc.h,$(DRIVERSDKINCLUDEDIR))
 InstallDriverSDKNonExecFile(xf86drm.h,$(DRIVERSDKINCLUDEDIR))
+InstallDriverSDKNonExecFile(xf86drmCompat.h,$(DRIVERSDKINCLUDEDIR))
+InstallDriverSDKNonExecFile(assyntax.h,$(DRIVERSDKINCLUDEDIR))
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/os-support/xf86_ansic.h.branch-4.3	2003-02-22 01:00:39.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/os-support/xf86_ansic.h	2003-07-08 00:10:45.000000000 -0400
@@ -307,13 +307,14 @@
 extern int xf86shmctl(int id, int xf86cmd, pointer buf);
 
 extern int xf86setjmp(xf86jmp_buf env);
+extern int xf86setjmp0(xf86jmp_buf env);
 extern int xf86setjmp1(xf86jmp_buf env, int);
 extern int xf86setjmp1_arg2(void);
 extern int xf86setjmperror(xf86jmp_buf env);
 extern int xf86getjmptype(void);
 extern void xf86longjmp(xf86jmp_buf env, int val);
 #define xf86setjmp_macro(env) \
-	(xf86getjmptype() == 0 ? xf86setjmp((env)) : \
+	(xf86getjmptype() == 0 ? xf86setjmp0((env)) : \
 	(xf86getjmptype() == 1 ? xf86setjmp1((env), xf86setjmp1_arg2()) : \
 		xf86setjmperror((env))))
 
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/shadowfb/shadow.c.branch-4.3	2003-02-21 10:06:19.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/shadowfb/shadow.c	2003-07-08 00:10:45.000000000 -0400
@@ -461,18 +461,28 @@
     ShadowScreenPtr pPriv = GET_SCREEN_PRIVATE(pScreen);
     PictureScreenPtr ps = GetPictureScreen(pScreen);
     BoxRec box;
+    BoxPtr extents;
     Bool boxNotEmpty = FALSE;
 
-    box.x1 = pDst->pDrawable->x + xDst;
-    box.y1 = pDst->pDrawable->y + yDst;
-    box.x2 = box.x1 + width;
-    box.y2 = box.y1 + height;
-
     if (pPriv->vtSema
-	&& pDst->pDrawable->type == DRAWABLE_WINDOW && BOX_NOT_EMPTY(box)) {
-        if (pPriv->preRefresh)
-            (*pPriv->preRefresh)(pPriv->pScrn, 1, &box);
-        boxNotEmpty = TRUE;
+	&& pDst->pDrawable->type == DRAWABLE_WINDOW) {
+
+	box.x1 = pDst->pDrawable->x + xDst;
+	box.y1 = pDst->pDrawable->y + yDst;
+	box.x2 = box.x1 + width;
+	box.y2 = box.y1 + height;
+
+	extents = &pDst->pCompositeClip->extents;
+	if(box.x1 < extents->x1) box.x1 = extents->x1;
+	if(box.x2 > extents->x2) box.x2 = extents->x2;
+	if(box.y1 < extents->y1) box.y1 = extents->y1;
+	if(box.y2 > extents->y2) box.y2 = extents->y2;
+	
+	if (BOX_NOT_EMPTY(box)) {
+	    if (pPriv->preRefresh)
+		(*pPriv->preRefresh)(pPriv->pScrn, 1, &box);
+	    boxNotEmpty = TRUE;
+	}
     }
     
     ps->Composite = pPriv->Composite;
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/vbe/Imakefile.branch-4.3	2003-02-17 12:06:45.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/vbe/Imakefile	2003-07-08 00:10:45.000000000 -0400
@@ -33,4 +33,5 @@
 InstallDriverSDKLibraryModule(vbe,$(DRIVERSDKMODULEDIR),.)
 
 InstallDriverSDKNonExecFile(vbe.h,$(DRIVERSDKINCLUDEDIR))
+InstallDriverSDKNonExecFile(vbeModes.h,$(DRIVERSDKINCLUDEDIR))
 
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/xf86cfg/Imakefile.branch-4.3	2003-02-26 15:08:03.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/xf86cfg/Imakefile	2003-07-08 00:10:45.000000000 -0400
@@ -96,15 +96,11 @@
 CURSESDEFINES  = -DHAS_NCURSES
 #endif
 
-#if HasGlibc21Sigsetjmp
-SETJMPDEFINES = -DHAS_GLIBC_SIGSETJMP=1
-#endif
-
 XF86CONFIGFILE = XConfigFile
 XF86CONFIGDIR = XConfigDir
 
 DEFINES        = -DXF86CONFIG=\"$(XF86CONFIGFILE)\" $(MODULEDEFINES) \
-		 $(CURSESDEFINES) $(SETJMPDEFINES) \
+		 $(CURSESDEFINES) \
 		 -DPROJECT_ROOT=\"$(PROJECTROOT)\" \
 		 -DXF86CONFIGDIR=\"$(XF86CONFIGDIR)\" XFree86ConsoleDefines
 
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/xf86cfg/loadmod.c.branch-4.3	2003-02-26 15:08:03.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/xf86cfg/loadmod.c	2003-07-08 00:15:13.000000000 -0400
@@ -33,8 +33,9 @@
 #include <setjmp.h>
 
 #ifndef HAS_GLIBC_SIGSETJMP
-#if defined(setjmp) && \
-    defined(__GLIBC__) && __GLIBC__ == 2 && __GLIBC_MINOR__ < 2
+#if defined(setjmp) && defined(__GNU_LIBRARY__) && \
+    (!defined(__GLIBC__) || (__GLIBC__ < 2) || \
+     ((__GLIBC__ == 2) && (__GLIBC_MINOR__ < 3)))
 #define HAS_GLIBC_SIGSETJMP 1
 #endif
 #endif
@@ -275,9 +276,15 @@
    SYMFUNC(xf86shmctl)
 #ifdef HAS_GLIBC_SIGSETJMP
    SYMFUNC(xf86setjmp)
+   SYMFUNC(xf86setjmp0)
+#if defined(__GLIBC__) && (__GLIBC__ >= 2)
    SYMFUNCALIAS("xf86setjmp1",__sigsetjmp)
 #else
+   SYMFUNC(xf86setjmp1)
+#endif
+#else
    SYMFUNCALIAS("xf86setjmp",setjmp)
+   SYMFUNCALIAS("xf86setjmp0",setjmp)
    SYMFUNC(xf86setjmp1)
 #endif
    SYMFUNCALIAS("xf86longjmp",longjmp)
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/CHANGELOG.branch-4.3	2003-02-26 23:56:44.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/CHANGELOG	2003-07-08 00:10:40.000000000 -0400
@@ -1,3 +1,66 @@
+XFree86 4.3.0.2 (xx June 2003)
+ 996. Pull twm fixes (signal handler, empty windows menu) from -current.
+ 995. Fix for a crash if a scalable font has a bitmap entry in 
+      fonts.dir (#5687, Bugzilla #332, Juliusz Chroboczek).
+ 994. Fix for xfs crashes in Freetype backend (Bugzilla #242, 
+      Juliusz Chroboczek). 
+ 993. Fix repeated image problem when using a vesa video mode before starting
+      the Xserver on the S3 Trio3D cards (Alan Hourihane).
+ 992. Fix XDMCP queries in xdm on systems using getifaddrs().
+      (Bugzilla #277, Joel Ray Holveck).
+ 991. Fixed Imakefiles so the driver SDK builds again (Sven Luther).
+ 990. i810/815 depth buffer needs to be a multiple of the tile size.  This
+      fixes 3D corruption near the bottom of the screen at 800x600@16bpp
+      (Bugzilla #283, Dave Airlie).
+ 989. Typo fix in radeon driver cursor code, fixes the primary head switching
+      off when the cursor moves to the secondary head on dualhead cards
+      (Michel Dänzer)
+ 988. Fix a problem with savage driver when UseBIOS is off, such that the
+      performance drops dramatically. Now the performance matches or
+      exceeds that of the option UseBIOS (Alan Hourihane).
+
+XFree86 4.3.0.1 (9 May 2003)
+ 987. Fix a FreeBSD/alpha build problem (#5679, Fred Clift).
+ 986. Fix SiliconMotion driver for mode switching and SEGV problem when
+      initializing Xv functionality (Bugzilla #50, Alan Hourihane).
+ 985. Fix a SEGV that can happen with Riva128 cards (Mark Vojkovich).
+ 984. Prevent a SIGFPE with the glint/pm3 driver when attempting to display
+      an XVideo image less than 8 pixels wide (Måns Rullgård).
+ 983. Fix double free bug when a Mac-specific font fails to load with
+      Freetype (Torrey T. Lyons).
+ 982. Check for NULL tObj in the i830 3D driver's TexEnv function.  This
+      fixes a FlightGear crash (Keith Whitwell).
+ 981. Fix lockup on server reset in radeon driver
+      (Michel Dänzer, Keith Whitwell).
+ 980. Set Mesa hooks to flush vertices on state changes in Radeon 3D drivers
+      (Keith Whitwell).
+ 979. Fix to prevent PCI and CardBus resets when switching out of the server's
+      virtual console (Marc La France).
+ 978. Fix infinite loop that occurs on systems whose PCI configuration space
+      doesn't advertise a host bridge (Marc La France).
+ 977. Workaround for broken devices that don't implement the header type field
+      in their PCI configuration space (Marc La France).
+ 976. Fix programming error in ix86 motherboard chipset determination
+      (Marc La France).
+ 975. Fix bug in mode validation that occurs when the XF86Config doesn't
+      specify any mode or virtual resolution information (Marc La France).
+ 974. Another int10 fix.  This time for adapters found at PCI:0:0:0.  This
+      fix is particularly important for ZX1-based systems (Marc La France).
+ 973. int10 fix for all ix86 non-Linux systems (Marc La France).
+ 972. Fix compatibility problem between modules generated without the recent
+      setjmp/longjmp work and a server generated using glibc <= 2.2.*
+      (Marc La France).
+ 971. setjmp/longjmp related fixes for Linux/libc5 systems (Marc La France).
+ 970. Make setjmp/longjmp emulation save/restore blocked signal masks on all
+      libc5 & glibc systems (Marc La France).
+ 969. Fix setjmp/longjmp emulation for glibc 2.2.[01] systems and remove
+      HasGlibc21Sigsetjmp override (Marc La France).
+ 968. setjmp/longjmp-related compilation fixes for libc5 systems
+      (Marc La France).
+ 967. Fix possible overflow in _XlcLocaleDirName. (Matthieu Herrb).
+ 966. Fix XDarwin build broken by #960 (Shantonu Sen).
+ 965. Fix StaticColor colormap on Darwin/x86 6.x (Rob Braun).
+
 XFree86 4.3.0 (27 February 2003)
  964. Add an imake option to allow the glibc21-style setjmp() behaviour
       to be forced when auto-detecting it fails (this is needed for RH 7.0).
--- XFree86-4.3/xc/programs/Xserver/hw/xfree86/xf86Date.h.branch-4.3	2003-02-26 23:56:45.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/hw/xfree86/xf86Date.h	2003-07-08 00:14:32.000000000 -0400
@@ -2,6 +2,6 @@
 
 #ifndef XF86_DATE
 
-#define XF86_DATE	"27 February 2003"
+#define XF86_DATE	"9 May 2003"
 
 #endif
--- XFree86-4.3/xc/programs/Xserver/mi/Imakefile.branch-4.3	2002-05-22 17:38:31.000000000 -0400
+++ XFree86-4.3/xc/programs/Xserver/mi/Imakefile	2003-07-08 00:10:47.000000000 -0400
@@ -85,3 +85,4 @@
 InstallDriverSDKNonExecFile(mipointer.h,$(DRIVERSDKINCLUDEDIR))
 InstallDriverSDKNonExecFile(mipointrst.h,$(DRIVERSDKINCLUDEDIR))
 InstallDriverSDKNonExecFile(mizerarc.h,$(DRIVERSDKINCLUDEDIR))
+InstallDriverSDKNonExecFile(micoord.h,$(DRIVERSDKINCLUDEDIR))
--- XFree86-4.3/xc/programs/Xserver/os/access.c.branch-4.3	2002-07-07 16:11:52.000000000 -0400
+++ XFree86-4.3/xc/programs/Xserver/os/access.c	2003-07-08 00:10:47.000000000 -0400
@@ -730,6 +730,7 @@
 	if (ifr->ifa_addr.sa_family == AF_DECnet) 
 	    continue;
 #endif /* DNETCONN */
+	len = sizeof(*(ifr->ifa_addr));
 	family = ConvertAddr(ifr->ifa_addr, &len, (pointer *)&addr);
 	if (family == -1 || family == FamilyLocal) 
 	    continue;
--- XFree86-4.3/xc/programs/Xserver/render/Imakefile.branch-4.3	2002-11-22 21:38:15.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/render/Imakefile	2003-07-08 00:10:47.000000000 -0400
@@ -36,3 +36,9 @@
 NormalLintTarget($(SRCS))
 
 DependTarget()
+
+InstallDriverSDKNonExecFile(glyphstr.h,$(DRIVERSDKINCLUDEDIR))
+InstallDriverSDKNonExecFile(mipict.h,$(DRIVERSDKINCLUDEDIR))
+InstallDriverSDKNonExecFile(picture.h,$(DRIVERSDKINCLUDEDIR))
+InstallDriverSDKNonExecFile(picturestr.h,$(DRIVERSDKINCLUDEDIR))
+
--- XFree86-4.3/xc/programs/Xserver/xkb/xkbInit.c.branch-4.3	2003-02-09 01:29:20.000000000 -0500
+++ XFree86-4.3/xc/programs/Xserver/xkb/xkbInit.c	2003-07-08 00:10:47.000000000 -0400
@@ -713,7 +713,7 @@
     if (names->types) names->types = _XkbDupString(names->types);
     if (names->compat) names->compat = _XkbDupString(names->compat);
     if (names->geometry) names->geometry = _XkbDupString(names->geometry);
-    if (names->symbols) names->geometry = _XkbDupString(names->symbols);
+    if (names->symbols) names->symbols = _XkbDupString(names->symbols);
 
     if (defs.model && defs.layout && rules) {
 	XkbComponentNamesRec	rNames;
--- XFree86-4.3/xc/programs/twm/events.c.branch-4.3	2001-12-14 15:01:06.000000000 -0500
+++ XFree86-4.3/xc/programs/twm/events.c	2003-07-08 00:10:47.000000000 -0400
@@ -350,7 +350,7 @@
 void
 HandleEvents()
 {
-    while (!TimeToYield)
+    while (TRUE)
     {
 	if (enter_flag && !QLength(dpy)) {
 	    if (enter_win && enter_win != raise_win) {
@@ -369,12 +369,6 @@
 	else
 	    XtDispatchEvent (&Event);
     }
-    if (dpy)
-    {
-	Reborder (CurrentTime);
-	XCloseDisplay(dpy);
-    }
-    exit(0);
 }
 
 
--- XFree86-4.3/xc/programs/twm/list.c.branch-4.3	2002-09-24 17:00:27.000000000 -0400
+++ XFree86-4.3/xc/programs/twm/list.c	2003-07-08 00:10:47.000000000 -0400
@@ -110,7 +110,7 @@
 	twmrc_error_prefix();
 	fprintf (stderr, "unable to allocate %ld bytes for name_list\n",
 		 (unsigned long)sizeof(name_list));
-	Done(0);
+	Done(NULL, NULL);
     }
 
     nptr->next = *list_head;
--- XFree86-4.3/xc/programs/twm/menus.c.branch-4.3	2002-10-19 16:04:20.000000000 -0400
+++ XFree86-4.3/xc/programs/twm/menus.c	2003-07-08 00:10:47.000000000 -0400
@@ -1007,7 +1007,7 @@
     int x, y;
     Bool center;
 {
-    int WindowNameOffset, WindowNameCount;
+    int WindowNameCount;
     TwmWindow **WindowNames;
     TwmWindow *tmp_win2,*tmp_win3;
     int i;
@@ -1033,37 +1033,38 @@
 	menu->mapped = NEVER_MAPPED;
   	AddToMenu(menu, "TWM Windows", NULLSTR, NULL, F_TITLE,NULLSTR,NULLSTR);
   
-        WindowNameOffset=(char *)Scr->TwmRoot.next->name -
-                               (char *)Scr->TwmRoot.next;
         for(tmp_win = Scr->TwmRoot.next , WindowNameCount=0;
             tmp_win != NULL;
             tmp_win = tmp_win->next)
           WindowNameCount++;
-        WindowNames =
-          (TwmWindow **)malloc(sizeof(TwmWindow *)*WindowNameCount);
-        WindowNames[0] = Scr->TwmRoot.next;
-        for(tmp_win = Scr->TwmRoot.next->next , WindowNameCount=1;
-            tmp_win != NULL;
-            tmp_win = tmp_win->next,WindowNameCount++)
+        if (WindowNameCount != 0)
         {
-            tmp_win2 = tmp_win;
-            for (i=0;i<WindowNameCount;i++)
+            WindowNames =
+              (TwmWindow **)malloc(sizeof(TwmWindow *)*WindowNameCount);
+            WindowNames[0] = Scr->TwmRoot.next;
+            for(tmp_win = Scr->TwmRoot.next->next , WindowNameCount=1;
+                tmp_win != NULL;
+                tmp_win = tmp_win->next,WindowNameCount++)
             {
-                if ((*compar)(tmp_win2->name,WindowNames[i]->name) < 0)
+                tmp_win2 = tmp_win;
+                for (i=0;i<WindowNameCount;i++)
                 {
-                    tmp_win3 = tmp_win2;
-                    tmp_win2 = WindowNames[i];
-                    WindowNames[i] = tmp_win3;
+                    if ((*compar)(tmp_win2->name,WindowNames[i]->name) < 0)
+                    {
+                        tmp_win3 = tmp_win2;
+                        tmp_win2 = WindowNames[i];
+                        WindowNames[i] = tmp_win3;
+                    }
                 }
+                WindowNames[WindowNameCount] = tmp_win2;
             }
-            WindowNames[WindowNameCount] = tmp_win2;
-        }
-        for (i=0; i<WindowNameCount; i++)
-        {
-            AddToMenu(menu, WindowNames[i]->name, (char *)WindowNames[i],
-                      NULL, F_POPUP,NULL,NULL);
+            for (i=0; i<WindowNameCount; i++)
+            {
+                AddToMenu(menu, WindowNames[i]->name, (char *)WindowNames[i],
+                          NULL, F_POPUP,NULL,NULL);
+            }
+            free(WindowNames);
         }
-        free(WindowNames);
 
 	MakeMenu(menu);
     }
@@ -1236,17 +1237,22 @@
      Window w;
      TwmWindow *tmp_win;
 {
-  int lastx, lasty, width, height, bw2;
-  int namelen;
+  int lastx, lasty, bw2;
   XEvent event;
+#if 0
+  int namelen;
+  int width, height;
 
   namelen = strlen (tmp_win->name);
+#endif
   bw2 = tmp_win->frame_bw * 2;
   AddingW = tmp_win->attr.width + bw2;
   AddingH = tmp_win->attr.height + tmp_win->title_height + bw2;
+#if 0
   width = (SIZE_HINDENT + MyFont_TextWidth (&Scr->SizeFont,
 					     tmp_win->name, namelen));
   height = Scr->SizeFont.height + SIZE_VINDENT * 2;
+#endif
   XGetGeometry(dpy, w, &JunkRoot, &origDragX, &origDragY,
 	       (unsigned int *)&DragWidth, (unsigned int *)&DragHeight, 
 	       &JunkBW, &JunkDepth);
@@ -1255,7 +1261,7 @@
   XQueryPointer (dpy, Scr->Root, &JunkRoot, 
 		 &JunkChild, &JunkX, &JunkY,
 		 &AddingX, &AddingY, &JunkMask);
-/*****
+#if 0
   Scr->SizeStringOffset = width +
     MyFont_TextWidth(&Scr->SizeFont, ": ", 2);
   XResizeWindow (dpy, Scr->SizeWindow, Scr->SizeStringOffset +
@@ -1263,16 +1269,16 @@
   MyFont_DrawImageString (dpy, Scr->SizeWindow, &Scr->SizeFont, Scr->NormalGC,
 		    width, SIZE_VINDENT + Scr->SizeFont.ascent,
 		    ": ", 2);
-*****/
+#endif
   lastx = -10000;
   lasty = -10000;
-/*****
+#if 0
   MoveOutline(Scr->Root,
 	      origDragX - JunkBW, origDragY - JunkBW,
 	      DragWidth * JunkBW, DragHeight * JunkBW,
 	      tmp_win->frame_bw,
 	      tmp_win->title_height);
-*****/
+#endif
   MenuStartResize(tmp_win, origDragX, origDragY, DragWidth, DragHeight);
   while (TRUE)
     {
@@ -2303,7 +2309,7 @@
 	break;
 
     case F_QUIT:
-	Done(0);
+	Done(NULL, NULL);
 	break;
 
     case F_PRIORITY:
--- XFree86-4.3/xc/programs/twm/session.c.branch-4.3	2001-12-14 15:01:10.000000000 -0500
+++ XFree86-4.3/xc/programs/twm/session.c	2003-07-08 00:10:48.000000000 -0400
@@ -977,7 +977,7 @@
 {
     SmcCloseConnection (smcConn, 0, NULL);
     XtRemoveInput (iceInputId);
-    Done(0);
+    Done(NULL, NULL);
 }
 
 
--- XFree86-4.3/xc/programs/twm/twm.c.branch-4.3	2001-12-14 15:01:10.000000000 -0500
+++ XFree86-4.3/xc/programs/twm/twm.c	2003-07-08 00:10:48.000000000 -0400
@@ -85,6 +85,7 @@
 #include <X11/Xlocale.h>
 
 XtAppContext appContext;	/* Xt application context */
+XtSignalId si;
 
 Display *dpy = NULL;		/* which display are we talking to */
 Window ResizeWindow;		/* the window we are resizing */
@@ -99,11 +100,11 @@
 ScreenInfo *Scr = NULL;		/* the cur and prev screens */
 int PreviousScreen;		/* last screen that we were on */
 int FirstScreen;		/* TRUE ==> first screen of display */
-volatile Bool TimeToYield = FALSE;	/* TRUE ==> exit requested */
 Bool PrintErrorMessages = False;	/* controls error messages */
 static int RedirectError;	/* TRUE ==> another window manager running */
 static int TwmErrorHandler ( Display *dpy, XErrorEvent *event );	/* for settting RedirectError */
 static int CatchRedirectError ( Display *dpy, XErrorEvent *event );	/* for everything else */
+static SIGNAL_T sigHandler(int);
 char Info[INFO_LINES][INFO_SIZE];		/* info strings to print */
 int InfoLines;
 char *InitFile = NULL;
@@ -235,8 +236,9 @@
     }
 
 #define newhandler(sig) \
-    if (signal (sig, SIG_IGN) != SIG_IGN) (void) signal (sig, Done)
+    if (signal (sig, SIG_IGN) != SIG_IGN) (void) signal (sig, sigHandler)
 
+    
     newhandler (SIGINT);
     newhandler (SIGHUP);
     newhandler (SIGQUIT);
@@ -269,6 +271,8 @@
     XtToolkitInitialize ();
     appContext = XtCreateApplicationContext ();
 
+    si = XtAppAddSignal(appContext, Done, NULL);
+    
     if (!(dpy = XtOpenDisplay (appContext, display_name, "twm", "twm",
 	NULL, 0, &zero, NULL))) {
 	fprintf (stderr, "%s:  unable to open display \"%s\"\n",
@@ -825,26 +829,6 @@
 }
 
 
-/***********************************************************************
- *
- *  Procedure:
- *	Done - cleanup and exit twm
- *
- *  Returned Value:
- *	none
- *
- *  Inputs:
- *	none
- *
- *  Outputs:
- *	none
- *
- *  Special Considerations:
- *	none
- *
- ***********************************************************************
- */
-
 void 
 Reborder (time)
 Time time;
@@ -872,13 +856,42 @@
     SetFocus ((TwmWindow*)NULL, time);
 }
 
-SIGNAL_T 
-Done(int sig)
+static SIGNAL_T 
+sigHandler(int sig)
 {
-    TimeToYield = True;
+    XtNoticeSignal(si);
     SIGNAL_RETURN;
 }
 
+/***********************************************************************
+ *
+ *  Procedure:
+ *	Done - cleanup and exit twm
+ *
+ *  Returned Value:
+ *	none
+ *
+ *  Inputs:
+ *	none
+ *
+ *  Outputs:
+ *	none
+ *
+ *  Special Considerations:
+ *	none
+ *
+ ***********************************************************************
+ */
+void
+Done(XtPointer client_data, XtSignalId *si)
+{
+    if (dpy) 
+    {
+	Reborder(CurrentTime);
+	XCloseDisplay(dpy);
+    }
+    exit(0);
+}
 
 /*
  * Error Handlers.  If a client dies, we'll get a BadWindow error (except for
--- XFree86-4.3/xc/programs/twm/twm.h.branch-4.3	2001-12-14 15:01:10.000000000 -0500
+++ XFree86-4.3/xc/programs/twm/twm.h	2003-07-08 00:10:48.000000000 -0400
@@ -352,7 +352,7 @@
 extern void CreateFonts ( void );
 extern void RestoreWithdrawnLocation ( TwmWindow *tmp );
 extern void Reborder( Time time);
-extern SIGNAL_T Done( int sig );
+extern void Done( XtPointer, XtSignalId * );
 extern void ComputeCommonTitleOffsets ( void );
 extern void ComputeTitleLocation ( TwmWindow *tmp );
 extern void ComputeWindowTitleOffsets ( TwmWindow *tmp_win, int width, Bool squeeze );
@@ -399,7 +399,6 @@
 extern Pixmap CreateMenuIcon ( int height, unsigned int *widthp, unsigned int *heightp );
 
 extern Bool ErrorOccurred;
-extern volatile Bool TimeToYield;
 extern XErrorEvent LastErrorEvent;
 
 #define ResetError() (ErrorOccurred = False)
