--- DB_File-1.805/Makefile.PL.fpons	2002-10-16 15:29:05.000000000 +0200
+++ DB_File-1.805/Makefile.PL	2002-10-16 15:29:20.000000000 +0200
@@ -51,7 +51,7 @@
 	'depend'	=> { 'Makefile'          => 'config.in',
                              'version$(OBJ_EXT)' => 'version.c'},
 	'clean'		=> { FILES => 'constants.h constants.xs' },
-	'macro'		=> { INSTALLDIRS => 'perl', my_files => "@files" },
+	#'macro'		=> { INSTALLDIRS => 'perl', my_files => "@files" },
         'dist'          => { COMPRESS => 'gzip', SUFFIX => 'gz',
 			     DIST_DEFAULT => 'MyDoubleCheck tardist'},    
 	);
