--- util-linux-2.12q/misc-utils/script.c.varargs	2004-03-26 12:07:16.000000000 -0500
+++ util-linux-2.12q/misc-utils/script.c	2005-08-23 08:45:18.000000000 -0400
@@ -325,9 +325,9 @@ doshell() {
 		shname = shell;
 
 	if (cflg)
-		execl(shell, shname, "-c", cflg, 0);
+		execl(shell, shname, "-c", cflg, NULL);
 	else
-		execl(shell, shname, "-i", 0);
+		execl(shell, shname, "-i", NULL);
 
 	perror(shell);
 	fail();
